
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Genotype PLINK File Quality Control &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Principal Component Analysis" href="PCA.html" />
    <link rel="prev" title="Genotype VCF File Quality Control" href="VCF_QC.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../trans_analysis/rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../trans_analysis/polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../trans_analysis/MRAID_QTL.html">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/gregor.html">
   Enrichment analysis: TORUS and GREGOR
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../pecotmr_analysis/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/ptwas_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/data_preprocessing/genotype/GWAS_QC.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype/GWAS_QC.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/genotype/GWAS_QC.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods">
   Methods
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#default-parameters-qc">
   Default Parameters: QC
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#perform-qc-on-both-rare-and-common-variants">
     Perform QC on both rare and common variants
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sample-match-with-genotype-1-minute">
     Sample match with genotype (&lt; 1 minute)
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#kinship-qc-2-minutes">
     Kinship QC (&lt; 2 minutes)
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#prepare-unrelated-individuals-data-for-pca-2-minutes">
     Prepare unrelated individuals data for PCA (&lt; 2 minutes)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#estimate-kinship-in-the-sample">
   Estimate kinship in the sample
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#genotype-and-sample-qc">
   Genotype and sample QC
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#extract-genotype-based-on-overlap-with-phenotype">
   Extract genotype based on overlap with phenotype
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="genotype-plink-file-quality-control">
<h1>Genotype PLINK File Quality Control<a class="headerlink" href="#genotype-plink-file-quality-control" title="Permalink to this headline">¶</a></h1>
<p>This workflow implements some preliminary data QC steps for PLINK input files. VCF format of inputs will be converted to PLINK before performing QC.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This notebook includes workflow for</p>
<ul class="simple">
<li><p>Compute kinship matrix in sample and estimate related individuals</p></li>
<li><p>Genotype and sample QC: by MAF, missing data and HWE</p></li>
<li><p>LD pruning for follow up PCA analysis on genotype, as needed</p></li>
</ul>
<p>A potential limitation is that the workflow requires all samples and chromosomes to be merged as one single file, in order to perform both sample and variant level QC. However, in our experience using this pipeline with 200K exomes with 15 million variants, this pipeline works on the single merged PLINK file.</p>
</div>
<div class="section" id="methods">
<h2>Methods<a class="headerlink" href="#methods" title="Permalink to this headline">¶</a></h2>
<p>Depending on the context of your problem, the workflow can be executed in two ways:</p>
<ol class="simple">
<li><p>Run <code class="docutils literal notranslate"><span class="pre">qc</span></code> command to perform genotype data QC and LD pruning to generate a subset of variants in preparation for analysis such as PCA.</p></li>
<li><p>Run <code class="docutils literal notranslate"><span class="pre">king</span></code> first on either the original or a subset of common variants to identify unrelated individuals. The <code class="docutils literal notranslate"><span class="pre">king</span></code> pipeline will split samples to related and unrelated individuals. Then you perform <code class="docutils literal notranslate"><span class="pre">qc</span></code> on these individuals only and finally extract the same set of QC-ed variants for related individuals.</p></li>
</ol>
</div>
<div class="section" id="default-parameters-qc">
<h2>Default Parameters: QC<a class="headerlink" href="#default-parameters-qc" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Kinship coefficient for related individuals: 0.0625</p></li>
<li><p>MAF and MAC default: 0</p>
<ul>
<li><p>Above default includes both common and are variant</p></li>
<li><p>Recommand MAF for PCA: 0.01, <a class="reference external" href="https://bmcgenomdata.biomedcentral.com/articles/10.1186/s12863-020-0833-x">we should stick to common variants</a></p></li>
<li><p>Recommand MAC for single variant analysis: 5.</p></li>
</ul>
</li>
<li><p>Variant level missingness threshold: 0.1</p></li>
<li><p>Sample level missingness threshold: 0.1</p></li>
<li><p>LD pruning via PLINK for PCA analysis:</p>
<ul>
<li><p>window 50</p></li>
<li><p>shift 10</p></li>
<li><p>r2 0.1</p></li>
</ul>
</li>
<li><p>HWE default: 1E-15 which is very lenient</p></li>
</ul>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<p>The whole genome PLINK bim/bed/fam bundle. For input in VCF format and/or per-chromosome VCF or PLINK format, please use <code class="docutils literal notranslate"><span class="pre">vcf_to_plink</span></code> and <code class="docutils literal notranslate"><span class="pre">merge_plink</span></code> in <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/genotype_formatting.html">genotype formatting</a> pipeline to convert them to PLINK file bundle.</p>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>Minimal working example data-set as well as the singularity container <code class="docutils literal notranslate"><span class="pre">bioinfo.sif</span></code> can be downloaded from <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn36416559/files/">Synapse</a>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">chr1_chr6</span></code> data-set was merged from <code class="docutils literal notranslate"><span class="pre">chr1</span></code> and <code class="docutils literal notranslate"><span class="pre">chr6</span></code> data, using <code class="docutils literal notranslate"><span class="pre">merge_plink</span></code> command from <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/genotype_formatting.html">genotype formatting</a> pipeline.</p>
<div class="section" id="perform-qc-on-both-rare-and-common-variants">
<h3>Perform QC on both rare and common variants<a class="headerlink" href="#perform-qc-on-both-rare-and-common-variants" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">Genotype</span> \
   <span class="o">--</span><span class="n">genoFile</span> <span class="n">Genotype</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">bcftools_qc</span><span class="o">.</span><span class="n">bed</span> \
   <span class="o">--</span><span class="n">geno</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> \
   <span class="o">--</span><span class="n">mind</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> \
   <span class="o">--</span><span class="n">hwe</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">1e-08</span>   \
   <span class="o">--</span><span class="n">mac</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> \
   <span class="o">--</span><span class="n">container</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> \
   <span class="o">-</span><span class="n">J</span> <span class="mi">1</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">150</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="sample-match-with-genotype-1-minute">
<h3>Sample match with genotype (&lt; 1 minute)<a class="headerlink" href="#sample-match-with-genotype-1-minute" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">genotype_phenotype_sample_overlap</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">sample_meta</span> \
        <span class="o">--</span><span class="n">genoFile</span> <span class="nb">input</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">fam</span>  \
        <span class="o">--</span><span class="n">phenoFile</span> <span class="nb">input</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">csv</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> \
        <span class="o">--</span><span class="kp">mem</span> <span class="mi">5</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">genotype_phenotype_sample_overlap</span><span class="p">:</span> <span class="n">This</span> <span class="n">workflow</span> <span class="n">extracts</span> <span class="n">overlapping</span> <span class="n">samples</span> <span class="k">for</span> <span class="n">genotype</span> <span class="n">data</span> <span class="k">with</span> <span class="n">phenotype</span> <span class="n">data</span><span class="p">,</span> <span class="ow">and</span> <span class="n">output</span> <span class="n">the</span> <span class="n">filtered</span> <span class="n">sample</span> <span class="n">genotype</span> <span class="nb">list</span> <span class="k">as</span> <span class="n">well</span> <span class="k">as</span> <span class="n">sample</span> <span class="n">phenotype</span> <span class="nb">list</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">genotype_phenotype_sample_overlap</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">genotype_phenotype_sample_overlap</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">sample_meta</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sample_overlap</span><span class="o">.</span><span class="n">txt</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">sample_meta</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sample_genotypes</span><span class="o">.</span><span class="n">txt</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">genotype_phenotype_sample_overlap</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w71b4e35979654867</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">1</span> <span class="n">completed</span> <span class="n">step</span><span class="o">.</span>
</pre></div>
</div>
</div>
<div class="section" id="kinship-qc-2-minutes">
<h3>Kinship QC (&lt; 2 minutes)<a class="headerlink" href="#kinship-qc-2-minutes" title="Permalink to this headline">¶</a></h3>
<p>To accuratly estimate the PCs for the genotype. We split participants based on their kinship coefficients, estimated by KING.</p>
<ol class="simple">
<li><p>Variant level and sample level QC on unrelated individuals using missingness &gt; 10%, and LD-prunning in preparation for PCA analysis.</p></li>
<li><p>There is no related samples in these ROSMAP samples, so there is an additional step to only keep those samples in <code class="docutils literal notranslate"><span class="pre">rosmap_pheno.sample_genotypes.txt</span></code> to do PCA.</p></li>
</ol>
<p><strong>Be aware:</strong></p>
<p><strong>If the message from <code class="docutils literal notranslate"><span class="pre">king_2</span></code> shown as <code class="docutils literal notranslate"><span class="pre">No</span> <span class="pre">related</span> <span class="pre">individuals</span> <span class="pre">detected</span> <span class="pre">from</span> <span class="pre">*.kin0</span></code>, this means no related individuals detected for the samples in <code class="docutils literal notranslate"><span class="pre">--keep_samples</span></code>. In this case, there will be no output for unrelated individuals from this step.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">king</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">kinship</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="nb">input</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">pQTL</span> \
    <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">sample_meta</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sample_genotypes</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">maximize</span><span class="o">-</span><span class="n">unrelated</span> \
    <span class="o">--</span><span class="kp">mem</span> <span class="mi">40</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">king_1</span><span class="p">:</span> <span class="n">Inference</span> <span class="n">of</span> <span class="n">relationships</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">sample</span> <span class="n">to</span> <span class="n">identify</span> <span class="n">closely</span> <span class="n">related</span> <span class="n">individuals</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_1</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_1</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">kin0</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">king_2</span><span class="p">:</span> <span class="n">Select</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">unrelated</span> <span class="n">individual</span> <span class="k">with</span> <span class="n">an</span> <span class="n">attempt</span> <span class="n">to</span> <span class="n">maximize</span> <span class="n">the</span> <span class="n">unrelated</span> <span class="n">individuals</span> <span class="n">selected</span> <span class="kn">from</span> <span class="nn">the</span> <span class="n">data</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_2</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_2</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">related_id</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">king_3</span><span class="p">:</span> <span class="n">Split</span> <span class="n">genotype</span> <span class="n">data</span> <span class="n">into</span> <span class="n">related</span> <span class="ow">and</span> <span class="n">unrelated</span> <span class="n">samples</span><span class="p">,</span> <span class="k">if</span> <span class="n">related</span> <span class="n">individuals</span> <span class="n">are</span> <span class="n">detected</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_3</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_3</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">king</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w7fad1d8b027ec781</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">3</span> <span class="n">completed</span> <span class="n">steps</span><span class="o">.</span>
</pre></div>
</div>
</div>
<div class="section" id="prepare-unrelated-individuals-data-for-pca-2-minutes">
<h3>Prepare unrelated individuals data for PCA (&lt; 2 minutes)<a class="headerlink" href="#prepare-unrelated-individuals-data-for-pca-2-minutes" title="Permalink to this headline">¶</a></h3>
<p>Here we write data to <code class="docutils literal notranslate"><span class="pre">cache</span></code> folder instead of <code class="docutils literal notranslate"><span class="pre">output</span></code> because this genotype data can be removed later after PCA. Also filter out minor allel accout &lt; 5.</p>
<p><strong>If your data has <code class="docutils literal notranslate"><span class="pre">*.unrelated.bed</span></code> generated, that means there are related individuals in your data. In cases, we will use output from the KING step for unrelated individuals.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">cache</span> \
   <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> \
   <span class="o">--</span><span class="n">mac</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">5</span> \
   <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> \
   <span class="o">--</span><span class="kp">mem</span> <span class="mi">16</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span><span class="p">:</span> <span class="n">Filter</span> <span class="n">SNPs</span> <span class="ow">and</span> <span class="n">select</span> <span class="n">individuals</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">LD</span> <span class="n">pruning</span><span class="p">:</span> <span class="n">LD</span> <span class="n">prunning</span> <span class="ow">and</span> <span class="n">remove</span> <span class="n">related</span> <span class="n">individuals</span> <span class="p">(</span><span class="n">both</span> <span class="n">ind</span> <span class="n">of</span> <span class="n">a</span> <span class="n">pair</span><span class="p">)</span> <span class="n">Plink2</span> <span class="n">has</span> <span class="n">multi</span><span class="o">-</span><span class="n">threaded</span> <span class="n">calculation</span> <span class="k">for</span> <span class="n">LD</span> <span class="n">prunning</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">bed</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">qc</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w3a34828bd2888342</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">2</span> <span class="n">completed</span> <span class="n">steps</span><span class="o">.</span>
</pre></div>
</div>
<p><strong>In other cases eg ROSMAP proteomics data, message <code class="docutils literal notranslate"><span class="pre">No</span> <span class="pre">related</span> <span class="pre">individuals</span> <span class="pre">detected</span> <span class="pre">from</span> <span class="pre">*.kin0</span></code> occured, there is no separate genotype data generated for unrelated individuals. In this case, we need to work from the original genotype data and must use <code class="docutils literal notranslate"><span class="pre">--keep-samples</span></code> to run <code class="docutils literal notranslate"><span class="pre">qc</span></code> to extract samples for PCA.</strong> For example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">cache</span> \
   <span class="o">--</span><span class="n">genoFile</span> <span class="nb">input</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">bed</span> \
   <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">sample_meta</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sample_genotypes</span><span class="o">.</span><span class="n">txt</span> \
   <span class="o">--</span><span class="kp">name</span> <span class="n">pQTL</span> \
   <span class="o">--</span><span class="n">mac</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">5</span> \
   <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> \
   <span class="o">--</span><span class="kp">mem</span> <span class="mi">40</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span><span class="p">:</span> <span class="n">Filter</span> <span class="n">SNPs</span> <span class="ow">and</span> <span class="n">select</span> <span class="n">individuals</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">LD</span> <span class="n">pruning</span><span class="p">:</span> <span class="n">LD</span> <span class="n">prunning</span> <span class="ow">and</span> <span class="n">remove</span> <span class="n">related</span> <span class="n">individuals</span> <span class="p">(</span><span class="n">both</span> <span class="n">ind</span> <span class="n">of</span> <span class="n">a</span> <span class="n">pair</span><span class="p">)</span> <span class="n">Plink2</span> <span class="n">has</span> <span class="n">multi</span><span class="o">-</span><span class="n">threaded</span> <span class="n">calculation</span> <span class="k">for</span> <span class="n">LD</span> <span class="n">prunning</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">bed</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">alexmccreight</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">qc</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w3d8b01776519226f</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">2</span> <span class="n">completed</span> <span class="n">steps</span><span class="o">.</span>
</pre></div>
</div>
<p>[FIXME:] Extract previously selected variants from related individuals in preparation for PCA, only applying missingness filter at sample level,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">genotype</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">chr1_chr6</span><span class="o">.</span><span class="mf">20220110.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">variants</span> <span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">chr1_chr6</span><span class="o">.</span><span class="mf">20220110.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">for_pca</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span> \
    <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">geno</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">mind</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">for_pca</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run GWAS_QC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  king
  qc_no_prune
  qc
  genotype_phenotype_sample_overlap

Global Workflow Options:
  --cwd output (as path)
                        the output directory for generated files
  --name &#39;&#39;
                        A string to identify your analysis run
  --genoFile  paths

                        PLINK binary files
  --remove-samples . (as path)
                        The path to the file that contains the list of samples
                        to remove (format FID, IID)
  --keep-samples . (as path)
                        The path to the file that contains the list of samples
                        to keep (format FID, IID)
  --keep-variants . (as path)
                        The path to the file that contains the list of variants
                        to keep
  --exclude-variants . (as path)
                        The path to the file that contains the list of variants
                        to exclude
  --kinship 0.0625 (as float)
                        Kinship coefficient threshold for related individuals
                        (e.g first degree above 0.25, second degree above 0.125,
                        third degree above 0.0625)
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 20 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option
  --entrypoint (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + container.split(&#39;/&#39;)[-1][:-4]) if container.endswith(&#39;.sif&#39;) else &quot;&quot;


Sections
  king_1:               Inference of relationships in the sample to identify
                        closely related individuals
    Workflow Options:
      --kin-maf 0.01 (as float)
                        PLINK binary file
  king_2:               Select a list of unrelated individual with an attempt to
                        maximize the unrelated individuals selected from the
                        data
    Workflow Options:
      --[no-]maximize-unrelated (default to False)
                        If set to true, the unrelated individuals in a family
                        will be kept without being reported. Otherwise (use
                        `--no-maximize-unrelated`) the entire family will be
                        removed Note that attempting to maximize unrelated
                        individuals is computationally intensive on large data.
  king_3:               Split genotype data into related and unrelated samples,
                        if related individuals are detected
  qc_no_prune, qc_1:    Filter SNPs and select individuals
    Workflow Options:
      --maf-filter 0.0 (as float)
                        minimum MAF filter to use. 0 means do not apply this
                        filter.
      --maf-max-filter 0.0 (as float)
                        maximum MAF filter to use. 0 means do not apply this
                        filter.
      --mac-filter 0.0 (as float)
                        minimum MAC filter to use. 0 means do not apply this
                        filter.
      --mac-max-filter 0.0 (as float)
                        maximum MAC filter to use. 0 means do not apply this
                        filter.
      --geno-filter 0.1 (as float)
                        Maximum missingess per-variant
      --mind-filter 0.1 (as float)
                        Maximum missingness per-sample
      --hwe-filter 1e-15 (as float)
                        HWE filter -- a very lenient one
      --other-args  (as list)
                        Other PLINK arguments e.g snps_only, write-samples, etc
      --[no-]meta-only (default to False)
                        Only output SNP and sample list, rather than the PLINK
                        binary format of subset data
      --[no-]rm-dups (default to False)
                        Remove duplicate variants
  qc_2:                 LD prunning and remove related individuals (both ind of
                        a pair) Plink2 has multi-threaded calculation for LD
                        prunning
    Workflow Options:
      --window 50 (as int)
                        Window size
      --shift 10 (as int)
                        Shift window every 10 snps
      --r2 0.1 (as float)
  genotype_phenotype_sample_overlap: This workflow extracts overlapping samples
                        for genotype data with phenotype data, and output the
                        filtered sample genotype list as well as sample
                        phenotype list
    Workflow Options:
      --phenoFile VAL (as path, required)
                        A phenotype file, can be bed.gz or tsv
      --sample-participant-lookup . (as path)
                        If this file is provided, a genotype/phenotype sample
                        name match will be performed It must contain two column
                        names: genotype_id, sample_id
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># the output directory for generated files</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># A string to identify your analysis run</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># PLINK binary files</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># The path to the file that contains the list of samples to remove (format FID, IID)</span>
<span class="kn">parameter:</span> <span class="n">remove_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># The path to the file that contains the list of samples to keep (format FID, IID)</span>
<span class="kn">parameter:</span> <span class="n">keep_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># The path to the file that contains the list of variants to keep</span>
<span class="kn">parameter:</span> <span class="n">keep_variants</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># The path to the file that contains the list of variants to exclude</span>
<span class="kn">parameter:</span> <span class="n">exclude_variants</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># Kinship coefficient threshold for related individuals</span>
<span class="c1"># (e.g first degree above 0.25, second degree above 0.125, third degree above 0.0625)</span>
<span class="kn">parameter:</span> <span class="n">kinship</span> <span class="o">=</span> <span class="mf">0.0625</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># use this function to edit memory string for PLINK input</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="estimate-kinship-in-the-sample">
<h2>Estimate kinship in the sample<a class="headerlink" href="#estimate-kinship-in-the-sample" title="Permalink to this headline">¶</a></h2>
<p>The output is a list of related individuals, as well as the kinship matrix</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Inference of relationships in the sample to identify closely related individuals</span>
<span class="p">[</span><span class="n">king_1</span><span class="p">]</span>
<span class="c1"># PLINK binary file</span>
<span class="kn">parameter:</span> <span class="n">kin_maf</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="kn">input:</span> <span class="n">genoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}{</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="o">+</span><span class="kp">name</span><span class="p">)</span> <span class="k">if</span> <span class="kp">name</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.kin0&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">king</span><span class="o">-</span><span class="n">table</span> \
      <span class="o">--</span><span class="n">king</span><span class="o">-</span><span class="n">table</span><span class="o">-</span><span class="nb">filter</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--keep </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--remove </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">remove_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">remove_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">af</span> <span class="err">$</span><span class="p">{</span><span class="n">kin_maf</span><span class="p">}</span> \
      <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">af</span> <span class="err">$</span><span class="p">{</span><span class="mi">1</span><span class="o">-</span><span class="n">kin_maf</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span> 
    
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">i</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span>
    <span class="n">output_size</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">output_rows</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">output_column</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span><span class="p">)</span>
    <span class="n">output_preview</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
    
    <span class="n">printf</span> <span class="s2">&quot;output_info: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_size: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_rows: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_column: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_preview:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> \
        <span class="s2">&quot;$i&quot;</span> <span class="s2">&quot;$output_size&quot;</span> <span class="s2">&quot;$output_rows&quot;</span> <span class="s2">&quot;$output_column&quot;</span> <span class="s2">&quot;$output_preview&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select a list of unrelated individual with an attempt to maximize the unrelated individuals selected from the data </span>
<span class="p">[</span><span class="n">king_2</span><span class="p">]</span>
<span class="c1"># If set to true, the unrelated individuals in a family will be kept without being reported. </span>
<span class="c1"># Otherwise (use `--no-maximize-unrelated`) the entire family will be removed</span>
<span class="c1"># Note that attempting to maximize unrelated individuals is computationally intensive on large data.</span>
<span class="kn">parameter:</span> <span class="n">maximize_unrelated</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">related_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">_input</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">)]</span>
<span class="n">done_if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">related_id</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;No related individuals detected from </span><span class="si">{</span><span class="n">_input</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.related_id&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span>  <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">igraph</span><span class="p">)</span>
    <span class="c1"># Remove related individuals while keeping maximum number of individuals</span>
    <span class="c1"># this function is simplified from: </span>
    <span class="c1"># https://rdrr.io/cran/plinkQC/src/R/utils.R</span>
    <span class="c1">#&#39; @param relatedness [data.frame] containing pair-wise relatedness estimates</span>
    <span class="c1">#&#39; (in column [relatednessRelatedness]) for individual 1 (in column</span>
    <span class="c1">#&#39; [relatednessIID1] and individual 2 (in column [relatednessIID1]). Columns</span>
    <span class="c1">#&#39; relatednessIID1, relatednessIID2 and relatednessRelatedness have to present,</span>
    <span class="c1">#&#39; while additional columns such as family IDs can be present. Default column</span>
    <span class="c1">#&#39; names correspond to column names in output of plink --genome</span>
    <span class="c1">#&#39; (\url{https://www.cog-genomics.org/plink/1.9/ibd}). All original</span>
    <span class="c1">#&#39; columns for pair-wise highIBDTh fails will be returned in fail_IBD.</span>
    <span class="c1">#&#39; @param relatednessTh [double] Threshold for filtering related individuals.</span>
    <span class="c1">#&#39; Individuals, whose pair-wise relatedness estimates are greater than this</span>
    <span class="c1">#&#39; threshold are considered related.</span>
    <span class="n">relatednessFilter</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> 
                                  <span class="n">relatednessTh</span><span class="p">,</span>
                                  <span class="n">relatednessIID1</span><span class="o">=</span><span class="s2">&quot;IID1&quot;</span><span class="p">,</span> 
                                  <span class="n">relatednessIID2</span><span class="o">=</span><span class="s2">&quot;IID2&quot;</span><span class="p">,</span>
                                  <span class="n">relatednessRelatedness</span><span class="o">=</span><span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># format data</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessIID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessIID1</span><span class="p">,</span> <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessIID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessIID1</span><span class="p">,</span> <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessRelatedness</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessRelatedness</span><span class="p">,</span>
                       <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>

        <span class="n">iid1_index</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessIID1</span><span class="p">)</span>
        <span class="n">iid2_index</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessIID2</span><span class="p">)</span>

        <span class="n">relatedness</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">relatedness</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">])</span>
        <span class="n">relatedness</span><span class="p">[,</span><span class="n">iid2_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">relatedness</span><span class="p">[,</span><span class="n">iid2_index</span><span class="p">])</span>

        <span class="n">relatedness_names</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">iid1_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;IID1&quot;</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">iid2_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;IID2&quot;</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessRelatedness</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;M&quot;</span>

        <span class="c1"># Remove symmetric IID rows</span>
        <span class="n">relatedness_original</span> <span class="o">&lt;-</span> <span class="n">relatedness</span>
        <span class="n">relatedness</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">select_</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="o">~</span><span class="n">IID1</span><span class="p">,</span> <span class="o">~</span><span class="n">IID2</span><span class="p">,</span> <span class="o">~</span><span class="n">M</span><span class="p">)</span>

        <span class="n">sortedIDs</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">apply</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">c</span><span class="p">(</span><span class="n">sort</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">2</span><span class="p">])))</span>
            <span class="p">})),</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
        <span class="n">keepIndex</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">sortedIDs</span><span class="p">))</span>

        <span class="n">relatedness_original</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">keepIndex</span><span class="p">,]</span>
        <span class="n">relatedness</span> <span class="o">&lt;-</span> <span class="n">relatedness</span><span class="p">[</span><span class="n">keepIndex</span><span class="p">,]</span>

        <span class="c1"># individuals with at least one pair-wise comparison &gt; relatednessTh</span>
        <span class="c1"># return NULL to failIDs if no one fails the relatedness check</span>
        <span class="n">highRelated</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">filter_</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="o">~</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">highRelated</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatednessFails</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span> <span class="n">failIDs</span><span class="o">=</span><span class="n">NULL</span><span class="p">))</span>
        <span class="p">}</span>

        <span class="c1"># all samples with related individuals</span>
        <span class="n">allRelated</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">highRelated</span><span class="err">$</span><span class="n">IID1</span><span class="p">,</span> <span class="n">highRelated</span><span class="err">$</span><span class="n">IID2</span><span class="p">)</span>
        <span class="n">uniqueIIDs</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">allRelated</span><span class="p">)</span>

        <span class="c1"># Further selection of samples with relatives in cohort</span>
        <span class="n">multipleRelative</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">allRelated</span><span class="p">[</span><span class="n">duplicated</span><span class="p">(</span><span class="n">allRelated</span><span class="p">)])</span>
        <span class="n">singleRelative</span> <span class="o">&lt;-</span> <span class="n">uniqueIIDs</span><span class="p">[</span><span class="err">!</span><span class="n">uniqueIIDs</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span><span class="p">]</span>

        <span class="n">highRelatedMultiple</span> <span class="o">&lt;-</span> <span class="n">highRelated</span><span class="p">[</span><span class="n">highRelated</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span> <span class="o">|</span>
                                            <span class="n">highRelated</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span><span class="p">,]</span>
        <span class="n">highRelatedSingle</span> <span class="o">&lt;-</span> <span class="n">highRelated</span><span class="p">[</span><span class="n">highRelated</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">singleRelative</span> <span class="o">&amp;</span>
                                           <span class="n">highRelated</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">singleRelative</span><span class="p">,]</span>

        <span class="c1"># Only one related samples per individual</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">singleRelative</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
          <span class="c1"># randomly choose one to exclude</span>
          <span class="n">failIDs_single</span> <span class="o">&lt;-</span> <span class="n">highRelatedSingle</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span>
            
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="n">failIDs_single</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span>
  
        <span class="c1"># An individual has multiple relatives</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">relatedPerID</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">highRelatedMultiple</span><span class="p">[</span><span class="n">rowSums</span><span class="p">(</span>
                    <span class="n">cbind</span><span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">x</span><span class="p">,</span>
                          <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">x</span><span class="p">))</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
                <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">tmp</span><span class="p">))</span>
                <span class="k">return</span><span class="p">(</span><span class="n">rel</span><span class="p">)</span>
            <span class="p">})</span>
            <span class="n">names</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">multipleRelative</span>

            <span class="n">keepIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">pairwise</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">combn</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
                <span class="n">index</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span> <span class="o">&amp;</span>
                              <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">2</span><span class="p">])</span> <span class="o">|</span>
                    <span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">2</span><span class="p">]</span> <span class="o">&amp;</span>
                         <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">IID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">combination</span> <span class="o">&lt;-</span> <span class="n">highRelatedMultiple</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
                <span class="n">combination_graph</span> <span class="o">&lt;-</span> <span class="n">igraph</span><span class="p">::</span><span class="n">graph_from_data_frame</span><span class="p">(</span><span class="n">combination</span><span class="p">,</span>
                                                                   <span class="n">directed</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
                <span class="n">all_iv_set</span> <span class="o">&lt;-</span> <span class="n">igraph</span><span class="p">::</span><span class="n">ivs</span><span class="p">(</span><span class="n">combination_graph</span><span class="p">)</span>
                <span class="n">length_iv_set</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="n">all_iv_set</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">length</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>

                <span class="k">if</span> <span class="p">(</span><span class="nb">all</span><span class="p">(</span><span class="n">length_iv_set</span> <span class="o">==</span> <span class="mi">1</span><span class="p">))</span> <span class="p">{</span>
                    <span class="c1"># check how often they occurr elsewhere</span>
                    <span class="n">occurrence</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="nb">id</span><span class="p">)</span> <span class="p">{</span>
                        <span class="nb">sum</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">idlist</span><span class="p">)</span> <span class="nb">id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">idlist</span><span class="p">))</span>
                    <span class="p">})</span>
                    <span class="c1"># if occurrence the same everywhere, pick the first, else keep</span>
                    <span class="c1"># the one with minimum occurrence elsewhere</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">occurrence</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">sort</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                        <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">occurrence</span><span class="p">)[</span><span class="n">which</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">occurrence</span><span class="p">)]</span>
                    <span class="p">}</span>
                <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">all_iv_set</span><span class="p">[</span><span class="n">which</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">length_iv_set</span><span class="p">)]</span>
                <span class="p">}</span>
                <span class="k">return</span><span class="p">(</span><span class="n">nonRelated</span><span class="p">)</span>
            <span class="p">})</span>
            <span class="n">keepIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">keepIDs_multiple</span><span class="p">))</span>
            <span class="n">failIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">[</span><span class="err">!</span><span class="n">multipleRelative</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span>
                                                       <span class="n">keepIDs_multiple</span><span class="p">])</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
            <span class="n">failIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span>
        <span class="n">allFailIIDs</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">failIDs_single</span><span class="p">,</span> <span class="n">failIDs_multiple</span><span class="p">)</span>
        <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">allFailIIDs</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="nb">id</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">fail_inorder</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="err">$</span><span class="n">IID1</span> <span class="o">==</span> <span class="nb">id</span> <span class="o">&amp;</span>
                <span class="n">relatedness_original</span><span class="err">$</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span>
            <span class="n">fail_inreverse</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="err">$</span><span class="n">IID2</span> <span class="o">==</span> <span class="nb">id</span> <span class="o">&amp;</span>
                <span class="n">relatedness_original</span><span class="err">$</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span>
            <span class="k">if</span> <span class="p">(</span><span class="nb">any</span><span class="p">(</span><span class="n">fail_inreverse</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">inreverse</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">fail_inreverse</span><span class="p">,</span> <span class="p">]</span>
                <span class="n">id1</span> <span class="o">&lt;-</span> <span class="n">iid1_index</span>
                <span class="n">id2</span> <span class="o">&lt;-</span> <span class="n">iid2_index</span>
                <span class="n">inreverse</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="n">id1</span><span class="p">,</span> <span class="n">id2</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="n">inreverse</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="n">id2</span><span class="p">,</span> <span class="n">id1</span><span class="p">)]</span>
                <span class="n">names</span><span class="p">(</span><span class="n">inreverse</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">inreverse</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
            <span class="p">}</span>
            <span class="n">inorder</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">fail_inorder</span><span class="p">,</span> <span class="p">]</span>
            <span class="n">names</span><span class="p">(</span><span class="n">inorder</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="k">return</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">inorder</span><span class="p">,</span> <span class="n">inreverse</span><span class="p">))</span>
        <span class="p">})</span>
        <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">relatednessFails</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
            <span class="n">failIDs</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
            <span class="n">names</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="n">rownames</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="mi">1</span><span class="p">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span>
            <span class="n">uniqueFails</span> <span class="o">&lt;-</span> <span class="n">relatednessFails</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">]),]</span>
            <span class="n">failIDs</span> <span class="o">&lt;-</span> <span class="n">uniqueFails</span><span class="p">[,</span><span class="n">iid1_index</span><span class="p">]</span>
        <span class="p">}</span>
        <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatednessFails</span><span class="o">=</span><span class="n">relatednessFails</span><span class="p">,</span> <span class="n">failIDs</span><span class="o">=</span><span class="n">failIDs</span><span class="p">))</span>
    <span class="p">}</span>
    
  
    <span class="c1"># main code</span>
    <span class="n">kin0</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">stringsAsFactor</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">kin0</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID1&quot;</span><span class="p">,</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span><span class="s2">&quot;FID2&quot;</span><span class="p">,</span><span class="s2">&quot;ID2&quot;</span><span class="p">,</span><span class="s2">&quot;NSNP&quot;</span><span class="p">,</span><span class="s2">&quot;HETHET&quot;</span><span class="p">,</span><span class="s2">&quot;IBS0&quot;</span><span class="p">,</span><span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">maximize_unrelated</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span> <span class="p">{</span>
        <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">relatednessFilter</span><span class="p">(</span><span class="n">kin0</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">},</span> <span class="s2">&quot;ID1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID2&quot;</span><span class="p">,</span> <span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">failIDs</span>
        <span class="n">tmp1</span> <span class="o">&lt;-</span> <span class="n">kin0</span><span class="p">[,</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">tmp2</span> <span class="o">&lt;-</span> <span class="n">kin0</span><span class="p">[,</span><span class="mi">3</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">tmp1</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">tmp2</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">)</span>
        <span class="c1"># Get the family ID for these rels so there are two columns FID and IID in the output</span>
        <span class="n">lookup</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">distinct</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">tmp1</span><span class="p">,</span><span class="n">tmp2</span><span class="p">))</span>
        <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">lookup</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">lookup</span><span class="p">[,</span><span class="mi">2</span><span class="p">]</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">rel</span><span class="p">),]</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">kin0</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">KINSHIP</span> <span class="o">&gt;=</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">})</span>
        <span class="n">dat</span> <span class="o">=</span> <span class="n">rbind</span><span class="p">(</span><span class="n">rel</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID1&quot;</span><span class="p">,</span><span class="s2">&quot;ID1&quot;</span><span class="p">)],</span><span class="n">setNames</span><span class="p">(</span><span class="n">rel</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID2&quot;</span><span class="p">,</span><span class="s2">&quot;ID2&quot;</span><span class="p">)],</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID1&quot;</span><span class="p">,</span><span class="s2">&quot;ID1&quot;</span><span class="p">)))</span>
        <span class="n">dat</span> <span class="o">=</span> <span class="n">dat</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">dat</span><span class="p">),]</span> <span class="c1">## This is to remove duplicated FID and IID caused by one sample being related to multiple samples</span>
       <span class="p">}</span>    

    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;There are&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="p">(</span><span class="n">dat</span><span class="p">),</span><span class="s2">&quot;related individuals using a kinship threshold of $</span><span class="si">{kinship}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">i</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span>
    <span class="n">output_size</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">output_rows</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">output_column</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span><span class="p">)</span>
    <span class="n">output_preview</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span>
    
    <span class="n">printf</span> <span class="s2">&quot;output_info: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_size: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_rows: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_column: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_preview:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> \
        <span class="s2">&quot;$i&quot;</span> <span class="s2">&quot;$output_size&quot;</span> <span class="s2">&quot;$output_rows&quot;</span> <span class="s2">&quot;$output_column&quot;</span> <span class="s2">&quot;$output_preview&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Split genotype data into related and unrelated samples, if related individuals are detected</span>
<span class="p">[</span><span class="n">king_3</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">genoFile</span>
<span class="kn">output:</span> <span class="n">unrelated_bed</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.unrelated.bed&#39;</span><span class="p">,</span>
        <span class="n">related_bed</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.related.bed&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">remove</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--keep </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span> <span class="o">--</span><span class="n">new</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="nb">max</span><span class="o">-</span><span class="n">allele</span><span class="o">-</span><span class="nb">len</span> <span class="mi">1000</span> <span class="o">--</span><span class="nb">set</span><span class="o">-</span><span class="nb">all</span><span class="o">-</span><span class="n">var</span><span class="o">-</span><span class="n">ids</span> <span class="nb">chr</span><span class="o">@</span><span class="p">:</span><span class="c1">#_\$r_\$a </span>

    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">keep</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> \
      <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span> <span class="o">--</span><span class="n">new</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="nb">max</span><span class="o">-</span><span class="n">allele</span><span class="o">-</span><span class="nb">len</span> <span class="mi">1000</span> <span class="o">--</span><span class="nb">set</span><span class="o">-</span><span class="nb">all</span><span class="o">-</span><span class="n">var</span><span class="o">-</span><span class="n">ids</span> <span class="nb">chr</span><span class="o">@</span><span class="p">:</span><span class="c1">#_\$r_\$a </span>
        
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">};</span> <span class="n">do</span>     
        <span class="n">output_size</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
        <span class="n">printf</span> <span class="s2">&quot;output_info: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_size: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="s2">&quot;$i&quot;</span> <span class="s2">&quot;$output_size&quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">stdout</span>
    <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="genotype-and-sample-qc">
<h2>Genotype and sample QC<a class="headerlink" href="#genotype-and-sample-qc" title="Permalink to this headline">¶</a></h2>
<p>QC the genetic data based on MAF, sample and variant missigness and Hardy-Weinberg Equilibrium (HWE).</p>
<p>In this step you may also provide a list of samples to keep, for example in the case when you would like to subset a sample based on their ancestries to perform independent analyses on each of these groups.</p>
<p>The default parameters are set to reflect some suggestions in Table 1 of <a class="reference external" href="https://dx.doi.org/10.1002%2Fmpr.1608">this paper</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter SNPs and select individuals </span>
<span class="p">[</span><span class="n">qc_no_prune</span><span class="p">,</span> <span class="n">qc_1</span> <span class="p">(</span><span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span><span class="p">)]</span>
<span class="c1"># minimum MAF filter to use. 0 means do not apply this filter.</span>
<span class="kn">parameter:</span> <span class="n">maf_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># maximum MAF filter to use. 0 means do not apply this filter.</span>
<span class="kn">parameter:</span> <span class="n">maf_max_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># minimum MAC filter to use. 0 means do not apply this filter.</span>
<span class="kn">parameter:</span> <span class="n">mac_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># maximum MAC filter to use. 0 means do not apply this filter.</span>
<span class="kn">parameter:</span> <span class="n">mac_max_filter</span> <span class="o">=</span> <span class="mf">0.0</span> 
<span class="c1"># Maximum missingess per-variant</span>
<span class="kn">parameter:</span> <span class="n">geno_filter</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="c1"># Maximum missingness per-sample</span>
<span class="kn">parameter:</span> <span class="n">mind_filter</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="c1"># HWE filter -- a very lenient one</span>
<span class="kn">parameter:</span> <span class="n">hwe_filter</span> <span class="o">=</span> <span class="mf">1e-15</span>
<span class="c1"># Other PLINK arguments e.g snps_only, write-samples, etc</span>
<span class="kn">parameter:</span> <span class="n">other_args</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Only output SNP and sample list, rather than the PLINK binary format of subset data</span>
<span class="kn">parameter:</span> <span class="n">meta_only</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Remove duplicate variants</span>
<span class="kn">parameter:</span> <span class="n">rm_dups</span> <span class="o">=</span> <span class="kc">False</span>

<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">keep_samples</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Cannot find ``</span><span class="si">{</span><span class="n">keep_samples</span><span class="si">}</span><span class="s1">``&#39;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">keep_variants</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Cannot find ``</span><span class="si">{</span><span class="n">keep_variants</span><span class="si">}</span><span class="s1">``&#39;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">remove_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">remove_samples</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Cannot find ``</span><span class="si">{</span><span class="n">remove_samples</span><span class="si">}</span><span class="s1">``&#39;</span><span class="p">)</span>

<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}{</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="o">+</span><span class="kp">name</span><span class="p">)</span> <span class="k">if</span> <span class="kp">name</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.plink_qc</span><span class="si">{</span><span class="s2">&quot;.extracted&quot;</span> <span class="k">if</span> <span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}{</span><span class="s2">&quot;.bed&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">meta_only</span> <span class="k">else</span> <span class="s2">&quot;.snplist&quot;</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--max-maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_max_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_max_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--mac </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">mac_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">mac_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--max-mac </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">mac_max_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">mac_max_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--geno </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">geno_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">geno_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--hwe </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hwe_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">hwe_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--mind </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">mind_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">mind_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--keep </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--remove </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">remove_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">remove_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--exclude </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">exclude_variants</span><span class="p">)</span> <span class="k">if</span> <span class="n">exclude_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--extract </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_variants</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--make-bed&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">meta_only</span> <span class="k">else</span> <span class="s2">&quot;--write-snplist --write-samples&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">rm_dups</span> <span class="k">else</span> <span class="s2">&quot;--rm-dup force-first &#39;list&#39;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">([</span><span class="s2">&quot;--</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">other_args</span><span class="p">])</span> <span class="k">if</span> <span class="n">other_args</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span> <span class="o">--</span><span class="n">new</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="nb">max</span><span class="o">-</span><span class="n">allele</span><span class="o">-</span><span class="nb">len</span> <span class="mi">1000</span> <span class="o">--</span><span class="nb">set</span><span class="o">-</span><span class="nb">all</span><span class="o">-</span><span class="n">var</span><span class="o">-</span><span class="n">ids</span> <span class="nb">chr</span><span class="o">@</span><span class="p">:</span><span class="c1">#_\$r_\$a </span>
        
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">i</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span>
    <span class="n">output_size</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">printf</span> <span class="s2">&quot;output_info: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_size: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="s2">&quot;$i&quot;</span> <span class="s2">&quot;$output_size&quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">stdout</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># LD prunning and remove related individuals (both ind of a pair)</span>
<span class="c1"># Plink2 has multi-threaded calculation for LD prunning</span>
<span class="p">[</span><span class="n">qc_2</span> <span class="p">(</span><span class="n">LD</span> <span class="n">pruning</span><span class="p">)]</span>
<span class="c1"># Window size</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">50</span>
<span class="c1"># Shift window every 10 snps</span>
<span class="kn">parameter:</span> <span class="n">shift</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">r2</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="n">stop_if</span><span class="p">(</span><span class="n">r2</span><span class="o">==</span><span class="mi">0</span><span class="p">)</span>
<span class="kn">output:</span> <span class="n">bed</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.prune.bed&#39;</span><span class="p">,</span> <span class="n">prune</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.prune.in&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">plink2</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">indep</span><span class="o">-</span><span class="n">pairwise</span> <span class="err">$</span><span class="p">{</span><span class="n">window</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">shift</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">r2</span><span class="p">}</span>  \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s2">&quot;prune&quot;</span><span class="p">]:</span><span class="n">nn</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span>
   
    <span class="n">plink2</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">extract</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s1">&#39;prune&#39;</span><span class="p">]}</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s1">&#39;bed&#39;</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span>
    
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">i</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{_output[0]}</span><span class="s2">&quot;</span>
    <span class="n">output_size</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">printf</span> <span class="s2">&quot;output_info: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_size: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="s2">&quot;$i&quot;</span> <span class="s2">&quot;$output_size&quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">stdout</span>
    <span class="n">i</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{_output[1]}</span><span class="s2">&quot;</span>
    <span class="n">output_size</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">output_rows</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">output_column</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span><span class="p">)</span>
    <span class="n">output_preview</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">)</span> 
    <span class="n">printf</span> <span class="s2">&quot;output_info: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_size: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_rows: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_column: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">output_preview:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> \
        <span class="s2">&quot;$i&quot;</span> <span class="s2">&quot;$output_size&quot;</span> <span class="s2">&quot;$output_rows&quot;</span> <span class="s2">&quot;$output_column&quot;</span> <span class="s2">&quot;$output_preview&quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span><span class="o">.</span><span class="n">stdout</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="extract-genotype-based-on-overlap-with-phenotype">
<h2>Extract genotype based on overlap with phenotype<a class="headerlink" href="#extract-genotype-based-on-overlap-with-phenotype" title="Permalink to this headline">¶</a></h2>
<p>This is an auxiliary step to match genotype and phenotype based on the data and look-up table. The look up table should contain two columns: <code class="docutils literal notranslate"><span class="pre">sample_id</span></code>, <code class="docutils literal notranslate"><span class="pre">genotype_id</span></code>. If the look up table is not provided or look-up table file not found, then we will assume the names have already been matched.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># This workflow extracts overlapping samples for genotype data with phenotype data, and output the filtered sample genotype list as well as sample phenotype list</span>
<span class="p">[</span><span class="n">genotype_phenotype_sample_overlap</span><span class="p">]</span>
<span class="c1"># A genotype fam file</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># A phenotype file, can be bed.gz or tsv</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># If this file is provided, a genotype/phenotype sample name match will be performed</span>
<span class="c1"># It must contain two column names: genotype_id, sample_id</span>
<span class="kn">parameter:</span> <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">phenoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">path</span><span class="p">(</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.sample_overlap.txt&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">path</span><span class="p">(</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.sample_genotypes.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="c1"># Load required libraries</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>

    <span class="c1"># Read data files; let read_delim auto-determine the delimiter</span>
    <span class="n">genoFam</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">phenoFile</span> <span class="o">&lt;-</span> <span class="n">read_delim</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">col_names</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">sample_participant_lookup</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span> <span class="p">{</span>
        <span class="n">sample_lookup</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">sample_participant_lookup</span><span class="p">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">sample_lookup</span> <span class="o">&lt;-</span> <span class="n">cbind</span><span class="p">(</span><span class="n">genoFam</span><span class="p">[,</span><span class="mi">2</span><span class="p">],</span> <span class="n">genoFam</span><span class="p">[,</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">sample_lookup</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;genotype_id&quot;</span><span class="p">,</span> <span class="s2">&quot;sample_id&quot;</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">sample_lookup</span> <span class="o">&lt;-</span> <span class="n">sample_lookup</span> <span class="o">%&gt;%</span>
    <span class="nb">filter</span><span class="p">(</span>
        <span class="n">genotype_id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">genoFam</span><span class="err">$</span><span class="n">V2</span><span class="p">,</span>
        <span class="n">sample_id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">colnames</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span>
    <span class="p">)</span>
    
    <span class="n">genoFam</span> <span class="o">%&gt;%</span>
    <span class="nb">filter</span><span class="p">(</span>
        <span class="n">V2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">sample_lookup</span><span class="err">$</span><span class="n">genotype_id</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="n">select</span><span class="p">(</span><span class="n">V1</span><span class="p">,</span> <span class="n">V2</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="n">fwrite</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">sample_lookup</span> <span class="o">%&gt;%</span>
    <span class="n">fwrite</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/data_preprocessing/genotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="VCF_QC.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Genotype VCF File Quality Control</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="PCA.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Principal Component Analysis</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>