
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Genotype PLINK File Quality Control &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/genotype/GWAS_QC';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Principal Component Analysis" href="PCA.html" />
    <link rel="prev" title="Genotype VCF File Quality Control" href="VCF_QC.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../genotype_preprocessing.html">Genotype data preprocessing</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/data_preprocessing/genotype/GWAS_QC.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype/GWAS_QC.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/data_preprocessing/genotype/GWAS_QC.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Genotype PLINK File Quality Control</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods">Methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-qc">Default Parameters: QC</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-perform-qc-on-both-rare-and-common-variants">iii. Perform QC on both rare and common variants</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#v-sample-match-with-genotype">v. Sample match with genotype</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-kinship-qc">vi. Kinship QC</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vii-prepare-unrelated-individuals-data-for-pca">vii. Prepare unrelated individuals data for PCA</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#estimate-kinship-in-the-sample">Estimate kinship in the sample</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-and-sample-qc">Genotype and sample QC</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-genotype-based-on-overlap-with-phenotype">Extract genotype based on overlap with phenotype</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="genotype-plink-file-quality-control">
<h1>Genotype PLINK File Quality Control<a class="headerlink" href="#genotype-plink-file-quality-control" title="Link to this heading">#</a></h1>
<p>This workflow implements some preliminary data QC steps for PLINK input files. It supports both PLINK1 binary format (BED/BIM/FAM) and PLINK2 format (PGEN/PVAR/PSAM). VCF format of inputs will be converted to PLINK before performing QC.</p>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>This notebook includes workflow for</p>
<ul class="simple">
<li><p>Compute kinship matrix in sample and estimate related individuals</p></li>
<li><p>Genotype and sample QC: by MAF, missing data and HWE</p></li>
<li><p>LD pruning for follow up PCA analysis on genotype, as needed</p></li>
</ul>
<p>A potential limitation is that the workflow requires all samples and chromosomes to be merged as one single file, in order to perform both sample and variant level QC. However, in our experience using this pipeline with 200K exomes with 15 million variants, this pipeline works on the single merged PLINK file.</p>
</section>
<section id="methods">
<h2>Methods<a class="headerlink" href="#methods" title="Link to this heading">#</a></h2>
<p>Depending on the context of your problem, the workflow can be executed in two ways:</p>
<ol class="arabic simple">
<li><p>Run <code class="docutils literal notranslate"><span class="pre">qc</span></code> command to perform genotype data QC and LD pruning to generate a subset of variants in preparation for analysis such as PCA.</p></li>
<li><p>Run <code class="docutils literal notranslate"><span class="pre">king</span></code> first on either the original or a subset of common variants to identify unrelated individuals. The <code class="docutils literal notranslate"><span class="pre">king</span></code> pipeline will split samples to related and unrelated individuals. Then you perform <code class="docutils literal notranslate"><span class="pre">qc</span></code> on these individuals only and finally extract the same set of QC-ed variants for related individuals.</p></li>
</ol>
</section>
<section id="default-parameters-qc">
<h2>Default Parameters: QC<a class="headerlink" href="#default-parameters-qc" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Kinship coefficient for related individuals: 0.0625</p></li>
<li><p>MAF and MAC default: 0</p>
<ul>
<li><p>Above default includes both common and are variant</p></li>
<li><p>Recommand MAF for PCA: 0.01, <a class="reference external" href="https://bmcgenomdata.biomedcentral.com/articles/10.1186/s12863-020-0833-x">we should stick to common variants</a></p></li>
<li><p>Recommand MAC for single variant analysis: 5.</p></li>
</ul>
</li>
<li><p>Variant level missingness threshold: 0.1</p></li>
<li><p>Sample level missingness threshold: 0.1</p></li>
<li><p>LD pruning via PLINK for PCA analysis:</p>
<ul>
<li><p>window 50</p></li>
<li><p>shift 10</p></li>
<li><p>r2 0.1</p></li>
</ul>
</li>
<li><p>HWE default: 1E-15 which is very lenient</p></li>
</ul>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<p>The genotype files in either PLINK1 binary format (BED/BIM/FAM) or PLINK2 format (PGEN/PVAR/PSAM). For input in VCF format and/or per-chromosome VCF or PLINK format, please use <code class="docutils literal notranslate"><span class="pre">vcf_to_plink</span></code> and <code class="docutils literal notranslate"><span class="pre">merge_plink</span></code> in <a class="reference external" href="https://statfungen.github.io/xqtl-protocol/code/data_preprocessing/genotype/genotype_formatting.html">genotype formatting</a> pipeline to convert them to PLINK file bundle.</p>
</section>
<section id="minimal-working-example">
<h2>Minimal Working Example<a class="headerlink" href="#minimal-working-example" title="Link to this heading">#</a></h2>
<p>Minimal working example data-set as well as the singularity container <code class="docutils literal notranslate"><span class="pre">bioinfo.sif</span></code> can be downloaded from <a class="reference external" href="https://www.synapse.org/#!Synapse:syn36416559/files/">Synapse</a>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">chr1_chr6</span></code> data-set was merged from <code class="docutils literal notranslate"><span class="pre">chr1</span></code> and <code class="docutils literal notranslate"><span class="pre">chr6</span></code> data, using <code class="docutils literal notranslate"><span class="pre">merge_plink</span></code> command from <a class="reference external" href="https://statfungen.github.io/xqtl-protocol/code/data_preprocessing/genotype/genotype_formatting.html">genotype formatting</a> pipeline.</p>
<section id="iii-perform-qc-on-both-rare-and-common-variants">
<h3>iii. Perform QC on both rare and common variants<a class="headerlink" href="#iii-perform-qc-on-both-rare-and-common-variants" title="Link to this heading">#</a></h3>
<p>Timing: &lt;1 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc_no_prune \
   --cwd output/plink \
   --genoFile output/plink/wgs.merged.bed \
   --geno-filter 0.1 \
   --mind-filter 0.1 \
   --hwe-filter 1e-08 \
   --mac-filter 0 
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">qc_no_prune</span><span class="p">:</span> <span class="n">Filter</span> <span class="n">SNPs</span> <span class="ow">and</span> <span class="n">select</span> <span class="n">individuals</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">qc_no_prune</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">qc_no_prune</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">plink</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">qc_no_prune</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w6697f77cea0f6dc2</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">1</span> <span class="n">completed</span> <span class="n">step</span><span class="o">.</span>
</pre></div>
</div>
</section>
<section id="v-sample-match-with-genotype">
<h3>v. Sample match with genotype<a class="headerlink" href="#v-sample-match-with-genotype" title="Link to this heading">#</a></h3>
<p>Timing: &lt;1 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb genotype_phenotype_sample_overlap \
        --cwd output/genotype/ \
        --genoFile output/plink/wgs.merged.plink_qc.fam  \
        --phenoFile output/rnaseq/bulk_rnaseq_tmp_matrix.low_expression_filtered.outlier_removed.tmm.expression.bed.bed.gz
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">genotype_phenotype_sample_overlap</span><span class="p">:</span> <span class="n">This</span> <span class="n">workflow</span> <span class="n">extracts</span> <span class="n">overlapping</span> <span class="n">samples</span> <span class="k">for</span> <span class="n">genotype</span> <span class="n">data</span> <span class="k">with</span> <span class="n">phenotype</span> <span class="n">data</span><span class="p">,</span> <span class="ow">and</span> <span class="n">output</span> <span class="n">the</span> <span class="n">filtered</span> <span class="n">sample</span> <span class="n">genotype</span> <span class="nb">list</span> <span class="k">as</span> <span class="n">well</span> <span class="k">as</span> <span class="n">sample</span> <span class="n">phenotype</span> <span class="nb">list</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">genotype_phenotype_sample_overlap</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">genotype_phenotype_sample_overlap</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">bulk_rnaseq_tmp_matrix</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">sample_overlap</span><span class="o">.</span><span class="n">txt</span> <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">bulk_rnaseq_tmp_matrix</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">sample_genotypes</span><span class="o">.</span><span class="n">txt</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">genotype_phenotype_sample_overlap</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w3667bce205c8538b</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">1</span> <span class="n">completed</span> <span class="n">step</span><span class="o">.</span>
</pre></div>
</div>
</section>
<section id="vi-kinship-qc">
<h3>vi. Kinship QC<a class="headerlink" href="#vi-kinship-qc" title="Link to this heading">#</a></h3>
<p>To accuratly estimate the PCs for the genotype. We split participants based on their kinship coefficients, estimated by KING.</p>
<ol class="arabic simple">
<li><p>Variant level and sample level QC on unrelated individuals using missingness &gt; 10%, and LD-prunning in preparation for PCA analysis.</p></li>
<li><p>There is no related samples in these ROSMAP samples, so there is an additional step to only keep those samples in <code class="docutils literal notranslate"><span class="pre">rosmap_pheno.sample_genotypes.txt</span></code> to do PCA.</p></li>
</ol>
<p><strong>Be aware:</strong></p>
<p><strong>If the message from <code class="docutils literal notranslate"><span class="pre">king_2</span></code> shown as <code class="docutils literal notranslate"><span class="pre">No</span> <span class="pre">related</span> <span class="pre">individuals</span> <span class="pre">detected</span> <span class="pre">from</span> <span class="pre">*.kin0</span></code>, this means no related individuals detected for the samples in <code class="docutils literal notranslate"><span class="pre">--keep_samples</span></code>. In this case, there will be no output for related individuals from this step.</strong></p>
<p>Timing: &lt;2 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb king \
    --cwd output/genotype/kinship \
    --genoFile output/plink/wgs.merged.plink_qc.bed \
    --name wgs.merged.king \
    --keep-samples output/genotype/bulk_rnaseq_tmp_matrix.low_expression_filtered.outlier_removed.tmm.expression.bed.bed.sample_genotypes.txt
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">king_1</span><span class="p">:</span> <span class="n">Inference</span> <span class="n">of</span> <span class="n">relationships</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">sample</span> <span class="n">to</span> <span class="n">identify</span> <span class="n">closely</span> <span class="n">related</span> <span class="n">individuals</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_1</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_1</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">king</span><span class="o">.</span><span class="n">kin0</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">king_2</span><span class="p">:</span> <span class="n">Select</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">unrelated</span> <span class="n">individual</span> <span class="k">with</span> <span class="n">an</span> <span class="n">attempt</span> <span class="n">to</span> <span class="n">maximize</span> <span class="n">the</span> <span class="n">unrelated</span> <span class="n">individuals</span> <span class="n">selected</span> <span class="kn">from</span><span class="w"> </span><span class="nn">the</span> <span class="n">data</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">No</span> <span class="n">related</span> <span class="n">individuals</span> <span class="n">detected</span> <span class="kn">from</span><span class="w"> </span><span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">king</span><span class="o">.</span><span class="n">kin0</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_2</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">king</span><span class="o">.</span><span class="n">related_id</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">king_3</span><span class="p">:</span> <span class="n">Split</span> <span class="n">genotype</span> <span class="n">data</span> <span class="n">into</span> <span class="n">related</span> <span class="ow">and</span> <span class="n">unrelated</span> <span class="n">samples</span><span class="p">,</span> <span class="k">if</span> <span class="n">related</span> <span class="n">individuals</span> <span class="n">are</span> <span class="n">detected</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_3</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">king_3</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">king</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">kinship</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">king</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">king</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w414e098f8f3bd993</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">2</span> <span class="n">completed</span> <span class="n">steps</span> <span class="ow">and</span> <span class="mi">1</span> <span class="n">ignored</span> <span class="n">step</span><span class="o">.</span>
</pre></div>
</div>
</section>
<section id="vii-prepare-unrelated-individuals-data-for-pca">
<h3>vii. Prepare unrelated individuals data for PCA<a class="headerlink" href="#vii-prepare-unrelated-individuals-data-for-pca" title="Link to this heading">#</a></h3>
<p>Here we write data to <code class="docutils literal notranslate"><span class="pre">cache</span></code> folder instead of <code class="docutils literal notranslate"><span class="pre">output</span></code> because this genotype data can be removed later after PCA. Also filter out minor allel accout &lt; 5.</p>
<p><strong>If your data has <code class="docutils literal notranslate"><span class="pre">*.unrelated.bed</span></code> generated, that means there are related individuals in your data. In cases, we will use output from the KING step for unrelated individuals.</strong></p>
<p>Timing: &lt;1 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc \
   --cwd output/genotype/ \
   --genoFile output/genotype/kinship/wgs.merged.plink_qc.wgs.merged.king.unrelated.bed \
   --mac-filter 5 
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span><span class="p">:</span> <span class="n">Filter</span> <span class="n">SNPs</span> <span class="ow">and</span> <span class="n">select</span> <span class="n">individuals</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">LD</span> <span class="n">pruning</span><span class="p">:</span> <span class="n">LD</span> <span class="n">prunning</span> <span class="ow">and</span> <span class="n">remove</span> <span class="n">related</span> <span class="n">individuals</span> <span class="p">(</span><span class="n">both</span> <span class="n">ind</span> <span class="n">of</span> <span class="n">a</span> <span class="n">pair</span><span class="p">)</span> <span class="n">Plink2</span> <span class="n">has</span> <span class="n">multi</span><span class="o">-</span><span class="n">threaded</span> <span class="n">calculation</span> <span class="k">for</span> <span class="n">LD</span> <span class="n">prunning</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">bed</span> <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">qc</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w5583b53e391fc494</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">2</span> <span class="n">completed</span> <span class="n">steps</span><span class="o">.</span>
</pre></div>
</div>
<p><strong>In other cases eg ROSMAP proteomics data, message <code class="docutils literal notranslate"><span class="pre">No</span> <span class="pre">related</span> <span class="pre">individuals</span> <span class="pre">detected</span> <span class="pre">from</span> <span class="pre">*.kin0</span></code> occured, there is no separate genotype data generated for unrelated individuals. In this case, we need to work from the original genotype data and must use <code class="docutils literal notranslate"><span class="pre">--keep-samples</span></code> to run <code class="docutils literal notranslate"><span class="pre">qc</span></code> to extract samples for PCA.</strong> For example:</p>
<p>Timing: &lt;1 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc \
   --cwd output/genotype/ \
   --genoFile output/plink/wgs.merged.plink_qc.bed \
   --mac-filter 5 
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span><span class="p">:</span> <span class="n">Filter</span> <span class="n">SNPs</span> <span class="ow">and</span> <span class="n">select</span> <span class="n">individuals</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">basic</span> <span class="n">QC</span> <span class="n">filters</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">LD</span> <span class="n">pruning</span><span class="p">:</span> <span class="n">LD</span> <span class="n">prunning</span> <span class="ow">and</span> <span class="n">remove</span> <span class="n">related</span> <span class="n">individuals</span> <span class="p">(</span><span class="n">both</span> <span class="n">ind</span> <span class="n">of</span> <span class="n">a</span> <span class="n">pair</span><span class="p">)</span> <span class="n">Plink2</span> <span class="n">has</span> <span class="n">multi</span><span class="o">-</span><span class="n">threaded</span> <span class="n">calculation</span> <span class="k">for</span> <span class="n">LD</span> <span class="n">prunning</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">LD</span> <span class="n">pruning</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">bed</span> <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">qc</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w5583b53e391fc494</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">2</span> <span class="n">completed</span> <span class="n">steps</span><span class="o">.</span>
</pre></div>
</div>
<p>[FIXME:] Extract previously selected variants from related individuals in preparation for PCA, only applying missingness filter at sample level,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc_no_prune \
   --cwd output/plink \
   --genoFile output/plink/wgs.merged.bed \
   --geno-filter 0 \
   --mind-filter 0.1 \
   --maf-filter 0 \
   --keep-variants wgs.merged.plink_qc.plink_qc.prune.in \
   --name for_pca
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run GWAS_QC.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run GWAS_QC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  king
  qc_no_prune
  qc
  genotype_phenotype_sample_overlap

Global Workflow Options:
  --cwd output (as path)
                        the output directory for generated files
  --name &#39;&#39;
                        A string to identify your analysis run
  --genoFile  paths

                        PLINK binary files
  --remove-samples . (as path)
                        The path to the file that contains the list of samples
                        to remove (format FID, IID)
  --keep-samples . (as path)
                        The path to the file that contains the list of samples
                        to keep (format FID, IID)
  --keep-variants . (as path)
                        The path to the file that contains the list of variants
                        to keep
  --exclude-variants . (as path)
                        The path to the file that contains the list of variants
                        to exclude
  --kinship 0.0625 (as float)
                        Kinship coefficient threshold for related individuals
                        (e.g first degree above 0.25, second degree above 0.125,
                        third degree above 0.0625)
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 20 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option
  --entrypoint  (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;


Sections
  king_1:               Inference of relationships in the sample to identify
                        closely related individuals
    Workflow Options:
      --kin-maf 0.01 (as float)
                        PLINK binary file
  king_2:               Select a list of unrelated individual with an attempt to
                        maximize the unrelated individuals selected from the
                        data
  king_3:               Split genotype data into related and unrelated samples,
                        if related individuals are detected
  qc_no_prune, qc_1:    Filter SNPs and select individuals
    Workflow Options:
      --maf-filter 0.0 (as float)
                        minimum MAF filter to use. 0 means do not apply this
                        filter.
      --maf-max-filter 0.0 (as float)
                        maximum MAF filter to use. 0 means do not apply this
                        filter.
      --mac-filter 0.0 (as float)
                        minimum MAC filter to use. 0 means do not apply this
                        filter.
      --mac-max-filter 0.0 (as float)
                        maximum MAC filter to use. 0 means do not apply this
                        filter.
      --geno-filter 0.1 (as float)
                        Maximum missingess per-variant
      --mind-filter 0.1 (as float)
                        Maximum missingness per-sample
      --hwe-filter 1e-15 (as float)
                        HWE filter -- a very lenient one
      --other-args  (as list)
                        Other PLINK arguments e.g snps_only, write-samples, etc
      --[no-]meta-only (default to False)
                        Only output SNP and sample list, rather than the PLINK
                        binary format of subset data
      --[no-]rm-dups (default to False)
                        Remove duplicate variants
  qc_2:                 LD prunning and remove related individuals (both ind of
                        a pair) Plink2 has multi-threaded calculation for LD
                        prunning
    Workflow Options:
      --window 50 (as int)
                        Window size
      --shift 10 (as int)
                        Shift window every 10 snps
      --r2 0.1 (as float)
  genotype_phenotype_sample_overlap: This workflow extracts overlapping samples
                        for genotype data with phenotype data, and output the
                        filtered sample genotype list as well as sample
                        phenotype list
    Workflow Options:
      --phenoFile VAL (as path, required)
                        A phenotype file, can be bed.gz or tsv
      --sample-participant-lookup . (as path)
                        If this file is provided, a genotype/phenotype sample
                        name match will be performed It must contain two column
                        names: genotype_id, sample_id
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# the output directory for generated files
parameter: cwd = path(&quot;output&quot;)
# A string to identify your analysis run
parameter: name = &quot;&quot;
# PLINK binary files (either BED/BIM/FAM or PGEN/PVAR/PSAM format)
parameter: genoFile = paths
# The path to the file that contains the list of samples to remove (format FID, IID)
parameter: remove_samples = path(&#39;.&#39;)
# The path to the file that contains the list of samples to keep (format FID, IID)
parameter: keep_samples = path(&#39;.&#39;)
# The path to the file that contains the list of variants to keep
parameter: keep_variants = path(&#39;.&#39;)
# The path to the file that contains the list of variants to exclude
parameter: exclude_variants = path(&#39;.&#39;)
# Kinship coefficient threshold for related individuals
# (e.g first degree above 0.25, second degree above 0.125, third degree above 0.0625)
parameter: kinship = 0.0625
# For cluster jobs, number commands to run per job
parameter: job_size = 1
# Wall clock time expected
parameter: walltime = &quot;5h&quot;
# Memory expected
parameter: mem = &quot;16G&quot;
# Number of threads
parameter: numThreads = 20
# Software container option
parameter: container = &quot;&quot;
parameter: entrypoint= &quot;&quot;
# use this function to edit memory string for PLINK input
from sos.utils import expand_size
cwd = path(f&quot;{cwd:a}&quot;)

# Determine if the file is in PLINK1 (BED/BIM/FAM) or PLINK2 (PGEN/PVAR/PSAM) format
def determine_plink_format(file_path):
    &quot;&quot;&quot;
    Determine the PLINK file format based on file extensions and companion files.
    
    Args:
        file_path (str or Path): Path to the input file
    
    Returns:
        str: &#39;plink1&#39; or &#39;plink2&#39;
    &quot;&quot;&quot;
    # Convert to string if it&#39;s a Path object
    file_path = str(file_path)
    
    # Check direct file extensions
    if file_path.endswith(&#39;.bed&#39;):
        return &#39;plink1&#39;
    elif file_path.endswith(&#39;.pgen&#39;):
        return &#39;plink2&#39;
    
    # If the file doesn&#39;t have a standard extension, try to infer format
    try:
        # Remove the file extension if present
        base_path = file_path.rsplit(&#39;.&#39;, 1)[0] if &#39;.&#39; in file_path else file_path
        
        # Check for PLINK1 companion files
        plink1_companion_files = [
            f&quot;{base_path}.bim&quot;,
            f&quot;{base_path}.fam&quot;
        ]
        
        # Check for PLINK2 companion files
        plink2_companion_files = [
            f&quot;{base_path}.pvar&quot;,
            f&quot;{base_path}.psam&quot;
        ]
        
        # Check PLINK1 format
        if all(os.path.exists(f) for f in plink1_companion_files):
            return &#39;plink1&#39;
        
        # Check PLINK2 format
        if all(os.path.exists(f) for f in plink2_companion_files):
            return &#39;plink2&#39;
    
    except Exception as e:
        print(f&quot;Error determining PLINK format: {e}&quot;)
    
    # Default to PLINK1 if can&#39;t determine
    return &#39;plink1&#39;


# Get the appropriate PLINK command based on the input file format
def get_plink_command_prefix(file_path):
    format_type = determine_plink_format(file_path)
    if format_type == &#39;plink1&#39;:
        return &quot;--bfile&quot;
    else:  # plink2
        return &quot;--pfile&quot;
        
# Generate the appropriate file extension based on the requested format
def get_output_extension(output_format, is_prune=False):
    if output_format == &#39;plink1&#39;:
        return &#39;.bed&#39; if not is_prune else &#39;.prune.bed&#39;
    else:  # plink2
        return &#39;.pgen&#39; if not is_prune else &#39;.prune.pgen&#39;
        
# Choose the make-bed or make-pgen command based on desired output format
def get_make_command(output_format):
    if output_format == &#39;plink1&#39;:
        return &#39;--make-bed&#39;
    else:  # plink2
        return &#39;--make-pgen&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="estimate-kinship-in-the-sample">
<h2>Estimate kinship in the sample<a class="headerlink" href="#estimate-kinship-in-the-sample" title="Link to this heading">#</a></h2>
<p>The output is a list of related individuals, as well as the kinship matrix</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Inference of relationships in the sample to identify closely related individuals
[king_1]
# PLINK binary file
parameter: kin_maf = 0.01
input: genoFile
output: f&#39;{cwd}/{_input:bn}{(&quot;.&quot;+name) if name else &quot;&quot;}.kin0&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;

plink_command = get_plink_command_prefix(genoFile)

bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, entrypoint=entrypoint
    plink2 \
      ${plink_command} ${_input:n} \
      --make-king-table \
      --king-table-filter ${kinship} \
      ${(&#39;--keep %s&#39; % keep_samples) if keep_samples.is_file() else &quot;&quot;} \
      ${(&#39;--remove %s&#39; % remove_samples) if remove_samples.is_file() else &quot;&quot;} \
      --min-af ${kin_maf} \
      --max-af ${1-kin_maf} \
      --out ${_output:n} \
      --threads ${numThreads} \
      --memory ${int(expand_size(mem) * 0.9)/1e6} 
    
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint=entrypoint
    i=&quot;${_output}&quot;
    output_size=$(ls -lh $i | cut -f 5 -d &#39; &#39;)
    output_rows=$(cat $i | wc -l | cut -f 1 -d &#39; &#39;)
    output_column=$(cat $i | head -1 | wc -w)
    output_preview=$(cat $i | grep -v &quot;##&quot; | head | cut -f 1,2,3,4,5,6)
    
    printf &quot;output_info: %s\noutput_size: %s\noutput_rows: %s\noutput_column: %s\noutput_preview:\n%s\n&quot; \
        &quot;$i&quot; &quot;$output_size&quot; &quot;$output_rows&quot; &quot;$output_column&quot; &quot;$output_preview&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Select a list of unrelated individual with an attempt to maximize the unrelated individuals selected from the data 
[king_2: shared = &quot;related_id&quot; ]
related_id = [x.strip() for x in open(_input).readlines() if not x.startswith(&quot;#&quot;)]
output: f&#39;{_input:n}.related_id&#39;
with open(_output, &#39;a&#39;):
    pass
done_if(len(related_id) == 0, msg = f&quot;No related individuals detected from {_input}.&quot;)
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R:  container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, entrypoint=entrypoint

    # By Rui Dong and Derek Lamb, Columbia Neurology
    # This function enhances the standard plinkQC::relatednessFilter approach through:
    #
    # 1. Graph-based preprocessing:
    #    - For large relatedness networks, breaks down dense clusters of related individuals
    #    - Removes highly connected nodes (individuals with many relatives) first
    #    - Prevents memory issues when processing large components (&gt;20 individuals)
    #    - Reduces the input size for the more intensive plinkQC filtering
    #
    # 2. Additional checks:
    #    - Verifies that all related pairs are properly handled
    #    - Runs up to 20 iterations to ensure complete filtering
    #    - Catches edge cases that might be missed in a single pass
    #
    # This implementation is particularly helpful for datasets with large family 
    # clusters or complex relatedness structures where the standard approach
    # might struggle.    

    filter_relatedness &lt;- function(
      relatedness,                                  # Path to the relatedness file (required)
      relatednessTh = 0.0625,                       # Threshold for relatedness
      analysis_type = &quot;maximize_unrelated&quot;,         # Type of analysis: &quot;maximize_unrelated&quot; or &quot;maximize_cases&quot;
      output_prefix = &quot;output_prefix&quot;,              # Base name for output file
      otherCriterion = NULL,                        # Optional additional criterion for filtering
      otherCriterionTh = NULL,                      # Threshold for the additional criterion
      otherCriterionThDirection = &quot;ge&quot;,             # Direction of the threshold
      relatednessIID1 = &quot;IID1&quot;,                     # Column name for first individual ID
      relatednessIID2 = &quot;IID2&quot;,                     # Column name for second individual ID
      relatednessFID1 = NULL,                       # Column name for first family ID
      relatednessFID2 = NULL,                       # Column name for second family ID
      relatednessRelatedness = &quot;PI_HAT&quot;,            # Column name for relatedness value
      pheno_file = NULL,                            # File for phenotype information
      pheno_col = &quot;pheno&quot;,                          # Column name for phenotype
      otherCriterionIID = &quot;IID&quot;,                    # Column name for additional criterion ID
      otherCriterionMeasure = NULL,                 # Column name for additional criterion measure
      verbose = FALSE                               # Whether to print verbose output
    ) {
      # Load required packages
      suppressMessages({
        library(tidyverse)
        library(igraph)
        library(plinkQC)
        library(data.table)
      })
      
      # Validate inputs
      if (is.null(relatedness)) {
        stop(&quot;Relatedness file path is required&quot;)
      }
      
      if (analysis_type == &quot;maximize_cases&quot; &amp;&amp; is.null(pheno_file)) {
        stop(&quot;Must provide phenotype file when analysis_type is &#39;maximize_cases&#39;&quot;)
      }
      
      # Read the relatedness file
      full_kin0 &lt;- as.data.frame(relatedness)
      
      # Initial pruning of highly related individuals
      working_graph &lt;- full_kin0 |&gt;
        filter(!!sym(relatednessRelatedness) &gt;= relatednessTh) |&gt;
        select(!!sym(relatednessIID1), !!sym(relatednessIID2)) |&gt;
        graph_from_data_frame(directed = FALSE)
      
      working_comp &lt;- components(working_graph)
      
      # Set parameters
      graph_size_th = 20
      reduce_fraction = 0.05
      high_related_indiv &lt;- c()
      
      # Reduce components to max 20
      while (max(working_comp$csize &gt; graph_size_th)) {
        
        if (verbose) {
          print(paste0(&quot;The largest component graph has &quot;, max(working_comp$csize), 
                       &quot; individuals. Removing the most connected &quot;, 
                       scales::percent(reduce_fraction), &quot; of them.&quot;))
        }
        
        # Prepare for loop
        large_comp_ids &lt;- which(working_comp$csize &gt; graph_size_th)
        nodes_to_remove &lt;- c()
        
        for (comp_id in large_comp_ids) {
          # Get nodes/degrees for component
          comp_nodes &lt;- V(working_graph)[working_comp$membership == comp_id]
          comp_degrees &lt;- degree(working_graph, v = comp_nodes)
          
          # Remove highly related individuals
          num_to_remove &lt;- ceiling(length(comp_nodes) * reduce_fraction)
          high_degree_nodes &lt;- names(sort(comp_degrees, decreasing = TRUE))[1:num_to_remove]
          
          # Append to removal list
          nodes_to_remove &lt;- c(nodes_to_remove, high_degree_nodes)
        }
        
        # Update for next loop
        high_related_indiv &lt;- c(high_related_indiv, nodes_to_remove)
        working_graph &lt;- delete_vertices(working_graph, nodes_to_remove)
        working_comp &lt;- components(working_graph)
      }
      
      kin0 &lt;- full_kin0 |&gt;
        filter(!(!!sym(relatednessIID1) %in% high_related_indiv) &amp; 
               !(!!sym(relatednessIID2) %in% high_related_indiv))
      
      # Process based on analysis type
      if (analysis_type == &quot;maximize_unrelated&quot;) {
        # Run the plinkQC relatedness Filter
        rel &lt;- plinkQC::relatednessFilter(
          relatedness = kin0,
          otherCriterion = otherCriterion,
          relatednessTh = relatednessTh,
          relatednessIID1 = relatednessIID1,
          relatednessIID2 = relatednessIID2,
          otherCriterionTh = otherCriterionTh,
          otherCriterionThDirection = otherCriterionThDirection,
          relatednessFID1 = relatednessFID1,
          relatednessFID2 = relatednessFID2,
          relatednessRelatedness = relatednessRelatedness,
          otherCriterionIID = otherCriterionIID,
          otherCriterionMeasure = otherCriterionMeasure,
          verbose = verbose
        )$failIDs
        
        all_exclude &lt;- rel$IID
        
      } else if (analysis_type == &quot;maximize_cases&quot;) {
        # Load phenotype information
        related_individuals &lt;- unique(c(kin0 |&gt; pull(relatednessIID1), kin0 |&gt; pull(relatednessIID2)))
        
        df_pheno &lt;- fread(pheno_file) |&gt;
          drop_na(!!sym(pheno_col)) |&gt;
          filter(IID %in% related_individuals)
        
        # Match kinship and phenotype information
        related_pheno_individuals &lt;- df_pheno |&gt; 
          pull(IID)
        
        related_cases &lt;- df_pheno |&gt; 
          filter(!!sym(pheno_col) == 1) |&gt; 
          pull(IID)
        
        related_controls &lt;- df_pheno |&gt; 
          filter(!!sym(pheno_col) == 0) |&gt; 
          pull(IID)
        
        kin0 &lt;- kin0 |&gt;
          filter(!!sym(relatednessIID1) %in% related_pheno_individuals &amp; 
                 !!sym(relatednessIID2) %in% related_pheno_individuals)
        
        # Optimize cases
        df_case_kin &lt;- kin0 |&gt;
          filter(!!sym(relatednessIID1) %in% related_cases &amp; 
                 !!sym(relatednessIID2) %in% related_cases)
        
        rel_cases &lt;- plinkQC::relatednessFilter(
          relatedness = df_case_kin,
          otherCriterion = otherCriterion,
          relatednessTh = relatednessTh,
          relatednessIID1 = relatednessIID1,
          relatednessIID2 = relatednessIID2,
          otherCriterionTh = otherCriterionTh,
          otherCriterionThDirection = otherCriterionThDirection,
          relatednessFID1 = relatednessFID1,
          relatednessFID2 = relatednessFID2,
          relatednessRelatedness = relatednessRelatedness,
          otherCriterionIID = otherCriterionIID,
          otherCriterionMeasure = otherCriterionMeasure,
          verbose = verbose
        )$failIDs
        
        # Remove controls related to cases
        related_cases_keep &lt;- setdiff(related_cases, rel_cases$IID)
        
        related_controls_exclude &lt;- c()
        
        for (i in 1:nrow(kin0)) {
          iid1 &lt;- kin0 |&gt; pull(!!sym(relatednessIID1)) |&gt; nth(i) 
          iid2 &lt;- kin0 |&gt; pull(!!sym(relatednessIID2)) |&gt; nth(i)
          
          if (iid1 %in% related_cases_keep &amp; iid2 %in% related_controls) {
            related_controls_exclude &lt;- c(related_controls_exclude, iid2)
            
          } else if(iid2 %in% related_cases_keep &amp; iid1 %in% related_controls) {
            related_controls_exclude &lt;- c(related_controls_exclude, iid1)
            
          }
        }
        
        # Subset controls to only unrelated controls
        related_controls_keep &lt;- setdiff(related_controls, related_controls_exclude)
        
        df_control_kin &lt;- kin0 |&gt;
          filter(!!sym(relatednessIID1) %in% related_controls_keep &amp; 
                 !!sym(relatednessIID2) %in% related_controls_keep)
        
        rel_controls &lt;- plinkQC::relatednessFilter(
          relatedness = df_control_kin,
          otherCriterion = otherCriterion,
          relatednessTh = relatednessTh,
          relatednessIID1 = relatednessIID1,
          relatednessIID2 = relatednessIID2,
          otherCriterionTh = otherCriterionTh,
          otherCriterionThDirection = otherCriterionThDirection,
          relatednessFID1 = relatednessFID1,
          relatednessFID2 = relatednessFID2,
          relatednessRelatedness = relatednessRelatedness,
          otherCriterionIID = otherCriterionIID,
          otherCriterionMeasure = otherCriterionMeasure,
          verbose = verbose
        )$failIDs
        
        # Handle leftover individuals
        all_exclude &lt;- c(rel_cases$IID, related_controls_exclude, rel_controls$IID)
        
      } else {
        # Directly filter on kinship threshold
        rel &lt;- kin0 %&gt;% dplyr::filter(!!sym(relatednessRelatedness) &gt;= relatednessTh)
        
        all_exclude &lt;- unique(c(rel |&gt; pull(relatednessIID1), rel |&gt; pull(relatednessIID2)))
      }
      
      # Check for leftovers
      df_related &lt;- kin0 |&gt;
        filter(!(!!sym(relatednessIID1) %in% all_exclude) &amp; 
               !(!!sym(relatednessIID2) %in% all_exclude)) |&gt;
        filter(!!sym(relatednessRelatedness) &gt; relatednessTh)
      print(&quot;df_related&quot;)
      print(df_related)
      if (nrow(df_related) &gt; 0) {
        if (verbose) {
          print(paste0(&quot;Warning: After first plinkQC pass, there are still &quot;, 
                      nrow(df_related), &quot; related subjects in dataset.&quot;))
        }
        
        # Iterate until all subjects are removed
        iter &lt;- 0
        while (nrow(df_related) &gt; 0 &amp; iter &lt; 20) {
          # Rerun plinkQC
          additional_exclude &lt;- plinkQC::relatednessFilter(
            relatedness = df_related,
            otherCriterion = otherCriterion,
            relatednessTh = relatednessTh,
            relatednessIID1 = relatednessIID1,
            relatednessIID2 = relatednessIID2,
            otherCriterionTh = otherCriterionTh,
            otherCriterionThDirection = otherCriterionThDirection,
            relatednessFID1 = relatednessFID1,
            relatednessFID2 = relatednessFID2,
            relatednessRelatedness = relatednessRelatedness,
            otherCriterionIID = otherCriterionIID,
            otherCriterionMeasure = otherCriterionMeasure,
            verbose = verbose
          )$failIDs
          
          # Add to exclude list
          all_exclude &lt;- c(all_exclude, additional_exclude$IID)
          
          # Update df_related and iter
          df_related &lt;- kin0 |&gt;
            filter(!(!!sym(relatednessIID1) %in% all_exclude) &amp; 
                   !(!!sym(relatednessIID2) %in% all_exclude)) |&gt;
            filter(!!sym(relatednessRelatedness) &gt; relatednessTh)
          iter &lt;- iter + 1
        }
        
        # Print status
        if (verbose) {
          if (nrow(df_related) == 0) {
            print(paste0(&quot;All related subjects successfully removed after &quot;, iter, &quot; iterations.&quot;))
          } else if (nrow(df_related) &gt; 0) {
            stop(paste0(&quot;Error: After 20 plinkQC passes, there are still &quot;, 
                       nrow(df_related), &quot; subjects in dataset.&quot;))
          }
        }
      }
      
      # Add in high-related individuals
      all_exclude &lt;- c(all_exclude, high_related_indiv)
      
      # List of all excluded subjects
      dat &lt;- data.frame(IID = all_exclude, FID = as.character(all_exclude)) # Because IID and FID are the same
      
      # Output the related ids
      output_related &lt;- paste0(output_prefix, &quot;_kinship_cutoff_&quot;, relatednessTh, &quot;.related_id&quot;)
      write.table(dat, output_related, quote = FALSE, row.names = FALSE, col.names = FALSE)
      
      if (verbose) {
        cat(&quot;\n[INFO] There are&quot;, nrow(dat), &quot;excluded individuals identified using a kinship threshold of&quot;, relatednessTh, &quot;\n&quot;)
        cat(&quot;[OUTPUT] Excluded individual IDs have been written to the file:\n&quot;)
        cat(&quot;      &quot;, normalizePath(output_related), &quot;\n&quot;)
      }
      
      # Return the excluded IDs
      return(dat)
    }
    
  
    # main code
    suppressMessages({
      library(tidyverse)
      library(data.table)
    })
    
    # Read input and set column names to match your original code
    kin0 &lt;- read.table(${_input:r}, header=FALSE, stringsAsFactors=FALSE)
    colnames(kin0) &lt;- c(&quot;FID1&quot;,&quot;ID1&quot;,&quot;FID2&quot;,&quot;ID2&quot;,&quot;NSNP&quot;,&quot;HETHET&quot;,&quot;IBS0&quot;,&quot;KINSHIP&quot;)
    
    # Always use the filter_relatedness function with maximize_unrelated
    result &lt;- filter_relatedness(
      relatedness = kin0,  
      relatednessTh = ${kinship},
      analysis_type = &quot;maximize_unrelated&quot;,
      relatednessIID1 = &quot;ID1&quot;,
      relatednessIID2 = &quot;ID2&quot;,
      relatednessFID1 = &quot;FID1&quot;, 
      relatednessFID2 = &quot;FID2&quot;,
      relatednessRelatedness = &quot;KINSHIP&quot;,
      verbose = FALSE
    )
    
    # Format output to match your original code
    tmp1 &lt;- kin0[,1:2]
    tmp2 &lt;- kin0[,3:4]
    colnames(tmp1) = colnames(tmp2) = c(&quot;FID&quot;, &quot;ID&quot;)
    lookup &lt;- dplyr::distinct(rbind(tmp1,tmp2))
    dat &lt;- lookup[which(lookup[,2] %in% result$IID),]
    # Write output in the exact format of the original script
    cat(&quot;There are&quot;, nrow(dat), &quot;related individuals using a kinship threshold of ${kinship}\n&quot;)
    write.table(dat, ${_output:r}, quote=FALSE, row.names=FALSE, col.names=FALSE)
    
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint=entrypoint
    i=&quot;${_output}&quot;
    output_size=$(ls -lh $i | cut -f 5 -d &#39; &#39;)
    output_rows=$(cat $i | wc -l | cut -f 1 -d &#39; &#39;)
    output_column=$(cat $i | head -1 | wc -w)
    output_preview=$(cat $i | grep -v &quot;##&quot; | head | cut -f 1,2,3,4,5,6)
    
    printf &quot;output_info: %s\noutput_size: %s\noutput_rows: %s\noutput_column: %s\noutput_preview:\n%s\n&quot; \
        &quot;$i&quot; &quot;$output_size&quot; &quot;$output_rows&quot; &quot;$output_column&quot; &quot;$output_preview&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Split genotype data into related and unrelated samples, if related individuals are detected
[king_3]
depends: sos_variable(&quot;related_id&quot;)
input: output_from(2), genoFile
plink_command = get_plink_command_prefix(_input[1])
output_format = determine_plink_format(_input[1])
make_command = get_make_command(output_format)
output_ext = get_output_extension(output_format)
output: unrelated_bed = f&#39;{cwd}/{_input[0]:bn}.unrelated{output_ext}&#39;,
        related_bed = f&#39;{cwd}/{_input[0]:bn}.related{output_ext}&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    plink2 \
      ${plink_command} ${_input[1]:n} \
      --remove ${_input[0]} \
      ${(&#39;--keep %s&#39; % keep_samples) if keep_samples.is_file() else &quot;&quot;} \
      ${make_command} \
      --out ${_output[0]:n} \
      --threads ${numThreads} \
      --memory ${int(expand_size(mem) * 0.9)/1e6} --new-id-max-allele-len 1000 --set-all-var-ids chr@:#_\$r_\$a 

    if [ ${len(related_id)} -ne 0 ] ; then
    plink2 \
      ${plink_command} ${_input[1]:n} \
      --keep ${_input[0]} \
      ${make_command} \
      --out ${_output[1]:n} \
      --threads ${numThreads} \
      --memory ${int(expand_size(mem) * 0.9)/1e6} --new-id-max-allele-len 1000 --set-all-var-ids chr@:#_\$r_\$a 
    else
       touch ${_output[1]}
    fi
</pre></div>
</div>
</div>
</div>
</section>
<section id="genotype-and-sample-qc">
<h2>Genotype and sample QC<a class="headerlink" href="#genotype-and-sample-qc" title="Link to this heading">#</a></h2>
<p>QC the genetic data based on MAF, sample and variant missigness and Hardy-Weinberg Equilibrium (HWE).</p>
<p>In this step you may also provide a list of samples to keep, for example in the case when you would like to subset a sample based on their ancestries to perform independent analyses on each of these groups.</p>
<p>The default parameters are set to reflect some suggestions in Table 1 of <a class="reference external" href="https://dx.doi.org/10.1002%2Fmpr.1608">this paper</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Filter SNPs and select individuals 
[qc_no_prune, qc_1 (basic QC filters)]
# minimum MAF filter to use. 0 means do not apply this filter.
parameter: maf_filter = 0.0
# maximum MAF filter to use. 0 means do not apply this filter.
parameter: maf_max_filter = 0.0
# minimum MAC filter to use. 0 means do not apply this filter.
parameter: mac_filter = 0.0
# maximum MAC filter to use. 0 means do not apply this filter.
parameter: mac_max_filter = 0.0 
# Maximum missingess per-variant
parameter: geno_filter = 0.1
# Maximum missingness per-sample
parameter: mind_filter = 0.1
# HWE filter -- a very lenient one
parameter: hwe_filter = 1e-15
# Other PLINK arguments e.g snps_only, write-samples, etc
parameter: other_args = []
# Only output SNP and sample list, rather than the PLINK binary format of subset data
parameter: meta_only = False
# Remove duplicate variants
parameter: rm_dups = False
# Add option to process dosage
parameter: treat_dosage_missing = False

fail_if(not (keep_samples.is_file() or keep_samples == path(&#39;.&#39;)), msg = f&#39;Cannot find ``{keep_samples}``&#39;)
fail_if(not (keep_variants.is_file() or keep_variants == path(&#39;.&#39;)), msg = f&#39;Cannot find ``{keep_variants}``&#39;)
fail_if(not (remove_samples.is_file() or remove_samples == path(&#39;.&#39;)), msg = f&#39;Cannot find ``{remove_samples}``&#39;)

input: genoFile, group_by=1
plink_command = get_plink_command_prefix(_input)
output_format = determine_plink_format(_input)
make_command = get_make_command(output_format) if not meta_only else &quot;--write-snplist --write-samples&quot;
output_ext = get_output_extension(output_format) if not meta_only else &quot;.snplist&quot;
output: f&#39;{cwd}/{_input:bn}{(&quot;.&quot; + name) if name else &quot;&quot;}.plink_qc{&quot;.extracted&quot; if keep_variants.is_file() else &quot;&quot;}{output_ext}&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    plink2 \
      ${plink_command} ${_input:n} \
      --allow-extra-chr \
      ${(&#39;--maf %s&#39; % maf_filter) if maf_filter &gt; 0 else &#39;&#39;} \
      ${(&#39;--max-maf %s&#39; % maf_max_filter) if maf_max_filter &gt; 0 else &#39;&#39;} \
      ${(&#39;--mac %s&#39; % mac_filter) if mac_filter &gt; 0 else &#39;&#39;} \
      ${(&#39;--max-mac %s&#39; % mac_max_filter) if mac_max_filter &gt; 0 else &#39;&#39;} \
      ${(&#39;--geno %s&#39; % geno_filter) if geno_filter &gt; 0 else &#39;&#39;} ${(&#39;dosage&#39; if treat_dosage_missing else &#39;&#39;)} \
      ${(&#39;--hwe %s&#39; % hwe_filter) if hwe_filter &gt; 0 else &#39;&#39;} \
      ${(&#39;--mind %s&#39; % mind_filter) if mind_filter &gt; 0 else &#39;&#39;} ${(&#39;dosage&#39; if treat_dosage_missing else &#39;&#39;)} \
      ${(&#39;--keep %s&#39; % keep_samples) if keep_samples.is_file() else &quot;&quot;} \
      ${(&#39;--remove %s&#39; % remove_samples) if remove_samples.is_file() else &quot;&quot;} \
      ${(&#39;--exclude %s&#39; % exclude_variants) if exclude_variants.is_file() else &quot;&quot;} \
      ${(&#39;--extract %s&#39; % keep_variants) if keep_variants.is_file() else &quot;&quot;} \
      ${make_command} \
      ${(&quot;&quot;) if not rm_dups else &quot;--rm-dup force-first &#39;list&#39;&quot;} \
      ${paths([&quot;--%s&quot; % x for x in other_args]) if other_args else &quot;&quot;} \
      --out ${_output:n} \
      --threads ${numThreads} \
      --memory ${int(expand_size(mem) * 0.9)/1e6} --new-id-max-allele-len 1000 --set-all-var-ids chr@:#_\$r_\$a 
        
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint=entrypoint
    i=&quot;${_output}&quot;
    output_size=$(ls -lh $i | cut -f 5 -d &#39; &#39;)
    printf &quot;output_info: %s\noutput_size: %s\n&quot; &quot;$i&quot; &quot;$output_size&quot; &gt;&gt; ${_output:n}.stdout
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># LD prunning and remove related individuals (both ind of a pair)
# Plink2 has multi-threaded calculation for LD prunning
[qc_2 (LD pruning)]
# Window size
parameter: window = 50
# Shift window every 10 snps
parameter: shift = 10
parameter: r2 = 0.1
stop_if(r2==0)
plink_command = get_plink_command_prefix(_input)
output_format = determine_plink_format(_input)
make_command = get_make_command(output_format)
output_ext = get_output_extension(output_format, is_prune=True)
output: bed=f&#39;{cwd}/{_input:bn}{output_ext}&#39;, prune=f&#39;{cwd}/{_input:bn}.prune.in&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    plink2 \
    ${plink_command} ${_input:n} \
    --allow-extra-chr \
    --rm-dup force-first \
    --indep-pairwise ${window} ${shift} ${r2}  \
    --out ${_output[&quot;prune&quot;]:nn} \
    --threads ${numThreads} \
    --memory ${int(expand_size(mem) * 0.9)/1e6}
   
    plink2 \
    ${plink_command} ${_input:n} \
    --allow-extra-chr \
    --extract ${_output[&#39;prune&#39;]} \
    ${make_command} \
    --out ${_output[&#39;bed&#39;]:n} \
    --threads ${numThreads} \
    --memory ${int(expand_size(mem) * 0.9)/1e6}
    
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    i=&quot;${_output[0]}&quot;
    output_size=$(ls -lh $i | cut -f 5 -d &#39; &#39;)
    printf &quot;output_info: %s\noutput_size: %s\n&quot; &quot;$i&quot; &quot;$output_size&quot; &gt;&gt; ${_output[0]:n}.stdout
    i=&quot;${_output[1]}&quot;
    output_size=$(ls -lh $i | cut -f 5 -d &#39; &#39;)
    output_rows=$(zcat $i | wc -l | cut -f 1 -d &#39; &#39;)
    output_column=$(zcat $i | head -1 | wc -w)
    output_preview=$(cat $i | grep -v &quot;##&quot; | head | cut -f 1,2,3,4,5,6) 
    printf &quot;output_info: %s\noutput_size: %s\noutput_rows: %s\noutput_column: %s\noutput_preview:\n%s\n&quot; \
        &quot;$i&quot; &quot;$output_size&quot; &quot;$output_rows&quot; &quot;$output_column&quot; &quot;$output_preview&quot; &gt;&gt; ${_output[1]}.stdout
</pre></div>
</div>
</div>
</div>
</section>
<section id="extract-genotype-based-on-overlap-with-phenotype">
<h2>Extract genotype based on overlap with phenotype<a class="headerlink" href="#extract-genotype-based-on-overlap-with-phenotype" title="Link to this heading">#</a></h2>
<p>This is an auxiliary step to match genotype and phenotype based on the data and look-up table. The look up table should contain two columns: <code class="docutils literal notranslate"><span class="pre">sample_id</span></code>, <code class="docutils literal notranslate"><span class="pre">genotype_id</span></code>. If the look up table is not provided or look-up table file not found, then we will assume the names have already been matched.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># This workflow extracts overlapping samples for genotype data with phenotype data, and output the filtered sample genotype list as well as sample phenotype list
[genotype_phenotype_sample_overlap]
# A genotype fam file
parameter: genoFile = path
# A phenotype file, can be bed.gz or tsv
parameter: phenoFile = path
# If this file is provided, a genotype/phenotype sample name match will be performed
# It must contain two column names: genotype_id, sample_id
parameter: sample_participant_lookup = path(&quot;.&quot;)
depends: executable(&#39;tabix&#39;), executable(&#39;bgzip&#39;)
input: genoFile, phenoFile
output: f&#39;{cwd:a}/{path(_input[1]):bn}.sample_overlap.txt&#39;, f&#39;{cwd:a}/{path(_input[1]):bn}.sample_genotypes.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    # Load required libraries
    library(dplyr)
    library(readr)
    library(data.table)

    # Read data files; let read_delim auto-determine the delimiter
    input_file &lt;- &quot;${_input[0]}&quot;
    
    # Determine if the input is a FAM file (PLINK1) or a PSAM file (PLINK2)
    if (grepl(&quot;\\.psam$&quot;, input_file)) {
        # For PSAM file (PLINK2 format)
        genoFam &lt;- fread(${_input[0]:ar}, header=TRUE)
        # Standardize column names (PSAM typically uses &#39;#IID&#39; or &#39;#FID IID&#39;)
        if (&quot;#FID&quot; %in% colnames(genoFam) &amp;&amp; &quot;IID&quot; %in% colnames(genoFam)) {
            genoFam &lt;- genoFam %&gt;% select(`#FID`, IID)
            colnames(genoFam) &lt;- c(&quot;V1&quot;, &quot;V2&quot;)  # Make column names match PLINK1 format for consistency
        } else if (&quot;#IID&quot; %in% colnames(genoFam)) {
            genoFam &lt;- genoFam %&gt;% mutate(V1 = `#IID`) %&gt;% select(V1, `#IID`)
            colnames(genoFam) &lt;- c(&quot;V1&quot;, &quot;V2&quot;)  # Make column names match PLINK1 format for consistency
        } else {
            # Try to handle other possible PSAM formats
            if (ncol(genoFam) &gt;= 2) {
                genoFam &lt;- genoFam[, 1:2]
                colnames(genoFam) &lt;- c(&quot;V1&quot;, &quot;V2&quot;)  # Make column names match PLINK1 format for consistency
            } else {
                stop(&quot;Cannot determine the structure of the PSAM file. Expected columns not found.&quot;)
            }
        }
    } else {
        # For FAM file (PLINK1 format) - assume standard format with no header
        genoFam &lt;- fread(${_input[0]:ar}, header=FALSE)
    }
    
    phenoFile &lt;- read_delim(${_input[1]:ar}, col_names=TRUE)
    if (${&quot;TRUE&quot; if sample_participant_lookup.is_file() else &quot;FALSE&quot;}) {
        sample_lookup &lt;- fread(${sample_participant_lookup:ar}, header=TRUE)
        colnames(sample_lookup) &lt;- c(&quot;sample_id&quot;, &quot;genotype_id&quot;) # FIXME: This is for the old lookup table with columns c(&quot;sample_id&quot;, &quot;participant_id&quot;)
        # rename phenotype file according to lookup file
        colnames(phenoFile)[-c(1:4)] &lt;- phenoFile %&gt;%
          colnames() %&gt;%
          .[-c(1:4)] %&gt;%
          match(sample_lookup$sample_id) %&gt;%
          sample_lookup$genotype_id[.]
      output_file &lt;- paste0(${phenoFile:nnr}, &#39;.rename_sample.bed&#39;)
      # rename phenotype file
      phenoFile$start &lt;- as.integer(phenoFile$start) 
      phenoFile$end &lt;- as.integer(phenoFile$end) 
      fwrite(phenoFile, output_file, sep = &#39;\t&#39;)
      # Compress the file using bgzip
      system(paste(&quot;bgzip -c&quot;, output_file, &quot;&gt;&quot;, paste0(output_file, &quot;.gz&quot;)))
      # Create the index file using tabix
      system(paste(&quot;tabix -p bed&quot;, paste0(output_file, &quot;.gz&quot;)))
      system(paste(&quot;rm&quot;, output_file))
    } else {
        sample_lookup &lt;- cbind(genoFam[,2], genoFam[,2])
        colnames(sample_lookup) &lt;- c(&quot;genotype_id&quot;, &quot;sample_id&quot;)
    }
    sample_lookup &lt;- sample_lookup %&gt;%
    filter(
        genotype_id %in% genoFam$V2,
        sample_id %in% colnames(phenoFile)
    )
    
    genoFam %&gt;%
    filter(
        V2 %in% sample_lookup$genotype_id,
    ) %&gt;%
    select(V1, V2) %&gt;%
    fwrite(${_output[1]:r}, col.names=FALSE, sep=&quot;\t&quot;)

    sample_lookup %&gt;%
    fwrite(${_output[0]:r}, sep=&quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing/genotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="VCF_QC.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Genotype VCF File Quality Control</p>
      </div>
    </a>
    <a class="right-next"
       href="PCA.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Principal Component Analysis</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods">Methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-qc">Default Parameters: QC</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-perform-qc-on-both-rare-and-common-variants">iii. Perform QC on both rare and common variants</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#v-sample-match-with-genotype">v. Sample match with genotype</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-kinship-qc">vi. Kinship QC</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vii-prepare-unrelated-individuals-data-for-pca">vii. Prepare unrelated individuals data for PCA</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#estimate-kinship-in-the-sample">Estimate kinship in the sample</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-and-sample-qc">Genotype and sample QC</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-genotype-based-on-overlap-with-phenotype">Extract genotype based on overlap with phenotype</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>