
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>Principal component analysis &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Genomic relationship matrices" href="GRM.html" />
    <link rel="prev" title="Genotype PLINK file quality control" href="GWAS_QC.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   Methylation calling and QC
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   APA calling
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Quantifying methylation data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantifying methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Quantifying alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/data_preprocessing/genotype/PCA.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype/PCA.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/genotype/PCA.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#general-workflow">
   General workflow
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#method">
   Method
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-1-estimate-kinship-in-the-sample">
     Step 1: Estimate kinship in the sample
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-2-and-3-sample-selection-and-qc-the-genotype-data-for-pca">
     Step 2 and 3: Sample selection and QC the genotype data for PCA
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-4-pca-analysis-for-unrelated-samples">
     Step 4: PCA analysis for unrelated samples
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-5-projection-of-related-individuals">
     Step 5: Projection of related individuals
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-5-2-finalize-genotype-qc-by-pca-for-homogenous-population">
     Step 5-2: Finalize genotype QC by PCA for homogenous population
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-6-split-data-by-population">
     Step 6: Split data by population
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-7-for-each-population-do-step-2">
     Step 7: For each population do step 2
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-8-for-each-population-do-step-3-if-applicable">
     Step 8: For each population do step 3 if applicable
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-9-for-each-population-do-step-4">
     Step 9: For each population do step 4
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-10-for-each-population-do-step-5-if-applicable">
     Step 10: For each population do step 5 if applicable
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#final-step-for-each-population-do-their-own-qc-to-finalize">
     Final step: For each population do their own QC to finalize
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pca-analysis">
   PCA analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#project-related-individuals-back">
   Project related individuals back
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plot-pca-results">
   Plot PCA results
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#detect-outliers">
   Detect outliers
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#add-plot-and-outlier-detection-to-pca-steps">
   Add plot and outlier detection to PCA steps
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="principal-component-analysis">
<h1>Principal component analysis<a class="headerlink" href="#principal-component-analysis" title="Permalink to this headline">¶</a></h1>
<p>The intention of this notebook is to perform the PCA analysis on genotype data and generate plots.</p>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Steps to generate a PCA include</p>
<ul class="simple">
<li><p>removing related individuals</p></li>
<li><p>pruning variants in linkage disequilibrium (LD)</p></li>
<li><p>perform PCA analysis on genotype of unrelated individuals</p></li>
<li><p>excluding outlier samples in the PCA space for individuals of homogeneous self-reported ancestry. These outliers may suggest poor genotyping quality or distant relatedness.</p></li>
</ul>
<p>Pitfalls</p>
<ol class="simple">
<li><p>Some of the PCs may capture LD structure rather than population structure (decrease in power to detect associations in these regions of high LD)</p></li>
<li><p>When projecting a new study dataset to the PCA space computed from a reference dataset: projected PCs are shrunk toward 0 in the new dataset</p></li>
<li><p>PC scores may capture outliers that are due to family structure, population structure or other reasons; it might be beneficial to detect and remove these individuals to maximize the population structure captured by PCA (in the case of removing a few outliers) or to restrict analyses to genetically homogeneous samples</p></li>
</ol>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Genotype in PLINK format: common variants, LD pruned.</p></li>
<li><p>Phenotype files in text format and must contain a column ‘IID’ for individual ID. Optionally it can contain a column named ‘FID’ for family ID. <strong>In this case samples will be matched by both FID and IID</strong>. If you do not have a phenotype file you can input a PLINK fam file. The phenotype file can additionally contain population information and possibly disease or other labelling information, to label the PCA plots.</p></li>
</ol>
<p>The inputs should be splitted into sets of related and unrelated individuals. Additionally you may want to prepare data per population. See “How to run this workflow” section for more details.</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>PCA models (inside RDS file)</p></li>
<li><p>PCA scores (inside RDS file)</p></li>
<li><p>Mahalanobis distances and outliers to remove</p></li>
<li><p>Various plots</p></li>
</ol>
</section>
<section id="general-workflow">
<h2>General workflow<a class="headerlink" href="#general-workflow" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Estimate relatedness of the individuals in the sample by PLINK 2 that implements the KING algorithm</p></li>
<li><p>Select specific SNPs and samples using PLINK and remove related individuals</p></li>
<li><p>SNPs thining by doing LD-pruning</p></li>
</ol>
<p>The above steps are implemented in <code class="docutils literal notranslate"><span class="pre">GWAS_QC.ipynb</span></code> workflow.</p>
<ol class="simple">
<li><p>Run PCA using only unrelated individuals for all populations, and examine the resulting plot</p></li>
<li><p>Project back related individuals, and generate a list of suggested samples to remove based on Mahalanobis distance test statistic per population. Default criteria is 0.997 percentile (two-sided) but we recommend checking the output plot before and after removal and rethink about it.</p></li>
</ol>
<p>The analysis above can be performed with reference data eg 1000 Genomes integrated, to help diagnose population substructure in data.</p>
<p>If you have subpopulations in the data, then additional steps should be applied for:</p>
<ol class="simple">
<li><p>Split data into different populations, each population data should have both related vs unrelated individual data-sets</p></li>
<li><p>For each population, perform QC</p></li>
<li><p>For each population, re-calculate per population PC’s for unrelated individuals</p></li>
<li><p>For each population, project related samples back to the PC space</p></li>
<li><p>Remove outliers based on list previously generated</p></li>
</ol>
</section>
<section id="method">
<h2>Method<a class="headerlink" href="#method" title="Permalink to this headline">¶</a></h2>
<p>Here is a quick recap of PCA analysis for those not immediately familiar with the method. PCA is a mathematical method to reduce dimensionality of the data while retaining most of the variation in the dataset.
This is accomplished by identifying directions or Principal Components (PC’s) that account for the maximum variation in the data.</p>
<p>One common approach to PCA is based on the singular-value decomposition of the the data matrix <span class="math notranslate nohighlight">\(X\)</span> (in our case the genotype matrix),</p>
<div class="math notranslate nohighlight">
\[X = U D V^T,\]</div>
<p>where <span class="math notranslate nohighlight">\(U\)</span> are the left eigenvectors, <span class="math notranslate nohighlight">\(D\)</span> is the diagonal matrix of singular values, and <span class="math notranslate nohighlight">\(V\)</span> are the right eigenvectors (also called loadings).</p>
<p>PCA can also be done using the eigen-decomposition of <span class="math notranslate nohighlight">\(X X^T\)</span>:</p>
<div class="math notranslate nohighlight">
\[X X^T = U S U^T,\]</div>
<p>where <span class="math notranslate nohighlight">\(S=D^2\)</span> is the diagonal matrix of eigenvalues.
<span class="math notranslate nohighlight">\(X\)</span> is usually centred (mean-subtracted) or standardised (mean subtracted, then divided by standard deviation) before PCA.</p>
<p>For PCA of SNP genotypes (at least in diploid organisms), the common standardisation is</p>
<div class="math notranslate nohighlight">
\[X_{ij}^{\prime} = \frac{X_{ij} - 2p_j}{\sqrt{2 p_j (1 - p_j)}},\]</div>
<p>where <span class="math notranslate nohighlight">\(X_{ij}\)</span> is the genotype (minor allele dosage <span class="math notranslate nohighlight">\(\{0, 1, 2\}\)</span>) for the <span class="math notranslate nohighlight">\(i\)</span>th individual and the <span class="math notranslate nohighlight">\(j\)</span>th SNP, and <span class="math notranslate nohighlight">\(p_j\)</span> is the minor allele frequency (MAF) for the <span class="math notranslate nohighlight">\(j\)</span>th SNP. In addition, the eigenvalues are scaled by the number of SNPs <span class="math notranslate nohighlight">\(m\)</span> (equivalent to performing the eigen-decomposition of <span class="math notranslate nohighlight">\(XX^T/m\)</span>).</p>
</section>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>The data can be found on <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1_dmlPf_7-4xpBKqBmhqRhuHnOUKTAVJi">Google Drive</a>.</p>
<p><strong>Note: parameter set for the MWE are meant to let the MWE work to show the workflow procedures. They may be unrealistic and should not be used in practice. The pipeline has reasonable default values for what we suggest to use in practice for most of the parameters.</strong></p>
<section id="step-1-estimate-kinship-in-the-sample">
<h3>Step 1: Estimate kinship in the sample<a class="headerlink" href="#step-1-estimate-kinship-in-the-sample" title="Permalink to this headline">¶</a></h3>
<p>Aim: identify and output closely related individuals prior to PCA analysis.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">king</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">rename_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">kinship</span> <span class="mf">0.13</span> \
    <span class="o">--</span><span class="n">name</span> <span class="mi">20220110</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p><strong><code class="docutils literal notranslate"><span class="pre">--kinship</span> <span class="pre">0.13</span></code> is meant for use with this MWE. The default (without using <code class="docutils literal notranslate"><span class="pre">--kinship</span></code> parameter) is 0.0625 for 3rd degree of relatedness.</strong></p>
</section>
<section id="step-2-and-3-sample-selection-and-qc-the-genotype-data-for-pca">
<h3>Step 2 and 3: Sample selection and QC the genotype data for PCA<a class="headerlink" href="#step-2-and-3-sample-selection-and-qc-the-genotype-data-for-pca" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>QC based on MAF, sample and variant missigness and Hardy-Weinberg Equilibrium. You can provide a list of samples to keep, or to remove. For example:</p>
<ul class="simple">
<li><p>Only extract data for one population</p></li>
<li><p>Only extract data for related individuals</p></li>
<li><p>Only extract data for unrelated individuals
In current context we would like to extract data for unrelated individuals and proceed with the rest of the QC steps.</p></li>
</ul>
</li>
<li><p>LD pruning. Prune SNPs in linkage dissequilibrium to make sure the PCA actually captures population structure and not LD structure (which could reduce the power of detecting genetic associations in these LD-regions).</p></li>
</ol>
<p><code class="docutils literal notranslate"><span class="pre">king</span></code> will split data into unrelated and related individuals. We will QC on unrelated individuals but <strong>extract the same variants from related individuals</strong>. We therefore only run <code class="docutils literal notranslate"><span class="pre">qc</span></code> with a list of variants extracted, and other filtering parameters set to 0.</p>
<p>Step 2: Variant level and sample level QC on unrelated individuals, in preparation for PCA analysis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.01</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">for_pca</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Step 3: Extract previously selected variants from related individuals in preparation for PCA, only applying missingness filter at sample level,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">variants</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">for_pca</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span> \
    <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">geno</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">mind</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> <span class="o">--</span><span class="n">hwe</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">for_pca</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</section>
<section id="step-4-pca-analysis-for-unrelated-samples">
<h3>Step 4: PCA analysis for unrelated samples<a class="headerlink" href="#step-4-pca-analysis-for-unrelated-samples" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">flashpca</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span> \
        <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">for_pca</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">bed</span> \
        <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">txt</span> \
        <span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
        <span class="o">--</span><span class="n">pop</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
        <span class="o">--</span><span class="n">maha</span><span class="o">-</span><span class="n">k</span> <span class="mi">2</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">flashpcaR</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>For starters we set <code class="docutils literal notranslate"><span class="pre">maha_k</span></code> abritarily to 2 for outlier calculations. You should look at the scree plot and determine how many PCs explain the most variation, then based on the scree plot generated to determine how many PCs you would like to consider for outlier checks.</p>
<p><strong>If you would like to skip outlier check and labelling, you can use <code class="docutils literal notranslate"><span class="pre">--prob</span> <span class="pre">1</span></code> to effectively not calling any samples an outlier</strong></p>
<p>If your <code class="docutils literal notranslate"><span class="pre">phenoFile</span></code> has a categorical variable with the ethnicity or population you can use <code class="docutils literal notranslate"><span class="pre">--label-col</span></code> parameter to define it, so that the plot generated has different colors depending on the population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">~/</span><span class="n">tmp</span><span class="o">/</span><span class="mi">25</span><span class="o">-</span><span class="n">Jan</span><span class="o">-</span><span class="mi">2022</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">scree</span><span class="o">.</span><span class="n">png</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /home/gw/tmp/25-Jan-2022/output/pca/MWE_pheno.pca.scree.png (56.4 KiB):</div></div><img alt="../../../_images/PCA_8_1.png" src="../../../_images/PCA_8_1.png" />
</div>
</div>
</section>
<section id="step-5-projection-of-related-individuals">
<h3>Step 5: Projection of related individuals<a class="headerlink" href="#step-5-projection-of-related-individuals" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">project_samples</span> \
  <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span> \
  <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">for_pca</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">bed</span> \
  <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">pca</span><span class="o">-</span><span class="n">model</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">rds</span> \
  <span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
  <span class="o">--</span><span class="n">pop</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
  <span class="o">--</span><span class="n">maha</span><span class="o">-</span><span class="n">k</span> <span class="mi">2</span> \
  <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">flashpcaR</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Same as above, if you have an ethnicity variable you can specify it by using <code class="docutils literal notranslate"><span class="pre">label_col</span></code> and here you would also need to add <code class="docutils literal notranslate"><span class="pre">pop_col</span></code> that points to the same variable. This is necessary, otherwise the detect_outliers part of the workflow will error.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">~/</span><span class="n">tmp</span><span class="o">/</span><span class="mi">25</span><span class="o">-</span><span class="n">Jan</span><span class="o">-</span><span class="mi">2022</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">pc</span><span class="o">.</span><span class="n">png</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /home/gw/tmp/25-Jan-2022/output/pca/MWE_pheno.pca.projected.pc.png (570.8 KiB):</div></div><img alt="../../../_images/PCA_10_1.png" src="../../../_images/PCA_10_1.png" />
</div>
</div>
</section>
<section id="step-5-2-finalize-genotype-qc-by-pca-for-homogenous-population">
<h3>Step 5-2: Finalize genotype QC by PCA for homogenous population<a class="headerlink" href="#step-5-2-finalize-genotype-qc-by-pca-for-homogenous-population" title="Permalink to this headline">¶</a></h3>
<p>Now pretend you analyze a homogenous population and you only did PCA to remove PC outliers. You have decided the outliers to remove and you would like to finalize the genotype QC.</p>
<p>First, remove these outliers and perform genotype data QC, similar to our initial QC but without LD pruning:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">remove</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">outliers</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">no_outlier</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Also remove outliers and keep the same variant as unrelated individuals, in related individuals:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">remove</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">outliers</span> \
    <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">variants</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">no_outlier</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">bim</span> \
    <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">geno</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">mind</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> <span class="o">--</span><span class="n">hwe</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">no_outlier</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Finally, merge back related and unrelated individuals as finalized genotype in PLINK format,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">genotype_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">merge_plink</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">no_outlier</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">bed</span> \
               <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">no_outlier</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">genotype_final</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">chr22_20220110_qced</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>You can stop here if you analyze a homogenous population.</p>
</section>
<section id="step-6-split-data-by-population">
<h3>Step 6: Split data by population<a class="headerlink" href="#step-6-split-data-by-population" title="Permalink to this headline">¶</a></h3>
<p>If your data contains multiple populations, then for each population, you can still use the <code class="docutils literal notranslate"><span class="pre">GWAS_QC.ipynb</span></code> and <code class="docutils literal notranslate"><span class="pre">PCA.ipynb</span></code> workflows, except that at this step you have to split data by population using manually created files to include or to remove samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">pheno</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s2">&quot;data/MWE_pheno.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
<span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">){</span>
 <span class="n">race</span> <span class="o">=</span> <span class="n">subset</span><span class="p">(</span><span class="n">pheno</span><span class="p">,</span> <span class="n">RACE</span> <span class="o">==</span> <span class="n">i</span><span class="p">)</span>
 <span class="n">race_id</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">race</span><span class="p">[,</span><span class="mi">2</span><span class="p">],</span><span class="n">race</span><span class="p">[,</span> <span class="mi">2</span><span class="p">])</span>
 <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">race_id</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;output/ID.&quot;</span><span class="p">,</span> <span class="s2">&quot;race&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">),</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>Below we perform analysis on each population. Here we only show it for Populations 1 and 3</p>
</section>
<section id="step-7-for-each-population-do-step-2">
<h3>Step 7: For each population do step 2<a class="headerlink" href="#step-7-for-each-population-do-step-2" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">race1</span> <span class="n">race3</span><span class="p">;</span> <span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
        <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> \
        <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">ID</span><span class="o">.</span><span class="err">$</span><span class="n">i</span> \
        <span class="o">--</span><span class="n">name</span> <span class="n">for_pca_</span><span class="err">$</span><span class="n">i</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
<span class="n">done</span>
</pre></div>
</div>
</section>
<section id="step-8-for-each-population-do-step-3-if-applicable">
<h3>Step 8: For each population do step 3 if applicable<a class="headerlink" href="#step-8-for-each-population-do-step-3-if-applicable" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">race1</span> <span class="n">race3</span><span class="p">;</span> <span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
        <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span> \
        <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">variants</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">for_pca_</span><span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span> \
        <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">ID</span><span class="o">.</span><span class="err">$</span><span class="n">i</span> \
        <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">geno</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">mind</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> <span class="o">--</span><span class="n">hwe</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> \
        <span class="o">--</span><span class="n">name</span> <span class="n">for_pca_</span><span class="err">$</span><span class="n">i</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> <span class="o">-</span><span class="n">s</span> <span class="n">force</span>
<span class="n">done</span>
</pre></div>
</div>
<p>Here in fact you will run into an error message for population 1, which you can find inside <code class="docutils literal notranslate"><span class="pre">ename_chr22.20220110.related.for_pca_race1.filtered.extracted.stderr</span></code> that says: <code class="docutils literal notranslate"><span class="pre">Error:</span> <span class="pre">No</span> <span class="pre">people</span> <span class="pre">remaining</span> <span class="pre">after</span> <span class="pre">--keep.</span></code>. That means there is no samples from population 1 among the related individuals. You can safely ignore this, and only handle related individuals for population 3 in upcoming analysis.</p>
</section>
<section id="step-9-for-each-population-do-step-4">
<h3>Step 9: For each population do step 4<a class="headerlink" href="#step-9-for-each-population-do-step-4" title="Permalink to this headline">¶</a></h3>
<p>Here I have to set the partial PCA to only computer 5 PCs because sample size is too small to work for the default 20 PCs.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">race1</span> <span class="n">race3</span><span class="p">;</span> <span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">flashpca</span> \
        <span class="o">--</span><span class="n">name</span> <span class="err">$</span><span class="n">i</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span> \
        <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">for_pca_</span><span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">bed</span> \
        <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">txt</span> \
        <span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
        <span class="o">--</span><span class="n">pop</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
        <span class="o">--</span><span class="n">maha</span><span class="o">-</span><span class="n">k</span> <span class="mi">2</span> \
        <span class="o">--</span><span class="n">k</span> <span class="mi">5</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">flashpcaR</span><span class="o">.</span><span class="n">sif</span>
<span class="n">done</span>
</pre></div>
</div>
</section>
<section id="step-10-for-each-population-do-step-5-if-applicable">
<h3>Step 10: For each population do step 5 if applicable<a class="headerlink" href="#step-10-for-each-population-do-step-5-if-applicable" title="Permalink to this headline">¶</a></h3>
<p>Here we only do it for population 3 (there was no related individuals in population 1):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">race3</span><span class="p">;</span> <span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">project_samples</span> \
      <span class="o">--</span><span class="n">name</span> <span class="err">$</span><span class="n">i</span> \
      <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span> \
      <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">for_pca_</span><span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">bed</span> \
      <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">txt</span> \
      <span class="o">--</span><span class="n">pca</span><span class="o">-</span><span class="n">model</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">rds</span> \
      <span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
      <span class="o">--</span><span class="n">pop</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
      <span class="o">--</span><span class="n">maha</span><span class="o">-</span><span class="n">k</span> <span class="mi">2</span> \
      <span class="o">--</span><span class="n">k</span> <span class="mi">5</span> \
      <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">flashpcaR</span><span class="o">.</span><span class="n">sif</span>
<span class="n">done</span>
</pre></div>
</div>
<p>To visualize the results at this point,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">~/</span><span class="n">tmp</span><span class="o">/</span><span class="mi">19</span><span class="o">-</span><span class="n">Jan</span><span class="o">-</span><span class="mi">2022</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">race3</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">pc</span><span class="o">.</span><span class="n">png</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /home/gw/tmp/19-Jan-2022/output/pca/MWE_pheno.race3.pca.projected.pc.png (124.0 KiB):</div></div><img alt="../../../_images/PCA_20_1.png" src="../../../_images/PCA_20_1.png" />
</div>
</div>
</section>
<section id="final-step-for-each-population-do-their-own-qc-to-finalize">
<h3>Final step: For each population do their own QC to finalize<a class="headerlink" href="#final-step-for-each-population-do-their-own-qc-to-finalize" title="Permalink to this headline">¶</a></h3>
<p>Same as to Step 5-2, taking into consideration of outliers detected. Please read the <code class="docutils literal notranslate"><span class="pre">GWAS_QC.ipynb</span></code> document to see available QC options and recommendations.</p>
</section>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run PCA.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  pca_plink
  flashpca
  project_samples
  plot_pca
  detect_outliers

Global Workflow Options:
  --cwd output (as path)
                        the output directory for generated files
  --name &#39;&#39;
                        A string to identify your analysis run
  --pop-col &#39;&#39;
                        Name of the population column in the phenoFile
  --pops  (as list)
  --label-col &#39;&#39;
                        Name of the color label column in the phenoFile; can be
                        the same as population column
  --k 20 (as int)
                        Number of Principal Components to output (flashpca
                        partial PCA method).
  --maha-k 5 (as int)
                        Number of Principal Components based on which outliers
                        should be evaluated. Default is 5 but this should be
                        based on examine the scree plot
  --[no-]homogeneous (default to False)
                        Homogeneity of populations. Set to --homogeneous when
                        true and --no-homogeneous when false
  --container &#39;&#39;
                        Software container option
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 10 (as int)
                        Number of threads

Sections
  pca_plink:
    Workflow Options:
      --genoFile VAL (as path, required)
                        Plink binary file
  flashpca_1:           Run PCA analysis using flashpca
    Workflow Options:
      --genoFile VAL (as path, required)
                        Plink binary file
      --phenoFile  path(f&#39;{genoFile}&#39;.replace(&quot;.bed&quot;,&quot;.fam&quot;))

                        The phenotypic file
      --min-pop-size 2 (as int)
                        minimum population size to consider in the analysis
      --stand binom2
                        How to standardize X before PCA
  project_samples_1:    Project back to PCA model additional samples
    Workflow Options:
      --genoFile VAL (as path, required)
                        Plink binary file
      --phenoFile  path(f&#39;{genoFile}&#39;.replace(&quot;.bed&quot;,&quot;.fam&quot;))

                        The phenotypic file
      --pca-model  f&#39;{cwd}/{phenoFile:bn}{(&quot;.&quot;+name) if name else &quot;&quot;}.{(suffix+&quot;.&quot;) if suffix != &quot;&quot; else &quot;&quot;}pca.rds&#39;

  plot_pca:             Plot PCA results. Can be used independently as
                        &quot;plot_pca&quot; or combined with other workflow as eg
                        &quot;flashpca+plot_pca&quot;
    Workflow Options:
      --outlier-file . (as path)
      --plot-data VAL (as path, required)
      --min-axis &#39;&#39;
      --max-axis &#39;&#39;
  detect_outliers:      Calculate Mahalanobis distance per population and report
                        outliers
    Workflow Options:
      --prob 0.997 (as float)
                        Set the probability to remove outliers eg 0.95 or 0.997
      --pval 0.05 (as float)
                        Mahalanobis distance p-value cutoff
      --[no-]robust (default to True)
                        Robust Mahalanobis to outliers
      --pca-result VAL (as path, required)
  flashpca_2, project_samples_2:
    Workflow Options:
      --prob 0.997 (as float)
                        Set the probability to remove outliers eg 0.95 or 0.997
      --[no-]robust (default to True)
                        Robust Mahalanobis to outliers
  flashpca_3, project_samples_3:
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># the output directory for generated files</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># A string to identify your analysis run</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># Name of the population column in the phenoFile</span>
<span class="kn">parameter:</span> <span class="n">pop_col</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># Name of the populations (from the population column) you would like to plot and show on the PCA plot</span>
<span class="kn">parameter:</span> <span class="n">pops</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Name of the color label column in the phenoFile; can be the same as population column. Can also be a separate column eg a &quot;super population&quot; column as a way to enable you to combine selected populations based on another column.</span>
<span class="kn">parameter:</span> <span class="n">label_col</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># Number of Principal Components to output (flashpca partial PCA method).</span>
<span class="kn">parameter:</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># Number of Principal Components based on which outliers should be evaluated. Default is 5 but this should be based on examine the scree plot</span>
<span class="kn">parameter:</span> <span class="n">maha_k</span> <span class="o">=</span> <span class="mi">5</span>
<span class="c1"># Homogeneity of populations. Set to --homogeneous when true and --no-homogeneous when false</span>
<span class="kn">parameter:</span> <span class="n">homogeneous</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">suffix</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pops</span><span class="p">)</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">pop_col</span><span class="p">:</span>
    <span class="n">homogeneous</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># PCA command with PLINK, as a sanity check</span>
<span class="p">[</span><span class="n">pca_plink</span><span class="p">]</span>
<span class="c1"># PLINK binary file</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">genoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.pca.eigenvec&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">plink</span> <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> <span class="o">--</span><span class="n">pca</span> <span class="err">$</span><span class="p">{</span><span class="n">k</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="pca-analysis">
<h2>PCA analysis<a class="headerlink" href="#pca-analysis" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run PCA analysis using flashpca </span>
<span class="p">[</span><span class="n">flashpca_1</span><span class="p">]</span>
<span class="c1"># Plink binary file</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># The phenotypic file</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span><span class="s2">&quot;.fam&quot;</span><span class="p">))</span>
<span class="c1"># minimum population size to consider in the analysis</span>
<span class="kn">parameter:</span> <span class="n">min_pop_size</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># How to standardize X before PCA</span>
<span class="kn">parameter:</span> <span class="n">stand</span> <span class="o">=</span> <span class="s2">&quot;binom2&quot;</span>
<span class="c1">## Input genoFile here is for unrelated samples</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">phenoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">phenoFile</span><span class="si">:</span><span class="s1">bn</span><span class="si">}{</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="o">+</span><span class="kp">name</span><span class="p">)</span> <span class="k">if</span> <span class="kp">name</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="p">(</span><span class="n">suffix</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">suffix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">pca.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="c1"># Load required libraries</span>
    <span class="n">library</span><span class="p">(</span><span class="n">flashpcaR</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">pops</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">(</span><span class="n">pops</span><span class="p">):</span><span class="n">r</span><span class="p">,})</span>
    <span class="n">f</span> <span class="o">&lt;-</span> <span class="n">flashpca</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nr</span><span class="p">},</span> <span class="n">ndim</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">k</span><span class="p">},</span> <span class="n">stand</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{stand}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">do_loadings</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">check_geno</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">loadings</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input[0]:n}</span><span class="s1">.bim&#39;</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span><span class="n">F</span><span class="p">)[,</span><span class="mi">2</span><span class="p">]</span>
    <span class="c1"># Use the projection file to generate pca plot</span>
    <span class="n">pca</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">)</span>
    <span class="n">pca</span> <span class="o">&lt;-</span> <span class="n">tibble</span><span class="p">::</span><span class="n">rownames_to_column</span><span class="p">(</span><span class="n">pca</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">pca</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;PC&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">k</span><span class="p">}))</span>
  
    <span class="c1"># Read fam file with phenotypes</span>
    <span class="k">if</span><span class="p">(</span><span class="n">stringr</span><span class="p">::</span><span class="n">str_detect</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span><span class="s2">&quot;.fam$&quot;</span><span class="p">)){</span>
        <span class="n">pheno</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">F</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span><span class="n">F</span><span class="p">)</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;IID&quot;</span><span class="p">,</span> <span class="s2">&quot;MID&quot;</span><span class="p">,</span> <span class="s2">&quot;PID&quot;</span><span class="p">,</span> <span class="s2">&quot;SEX&quot;</span><span class="p">,</span> <span class="s2">&quot;STATUS&quot;</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">pheno</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span><span class="n">F</span><span class="p">)</span>
        <span class="k">if</span><span class="p">(</span><span class="s2">&quot;IID&quot;</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span> <span class="o">==</span> <span class="n">FALSE</span><span class="p">)</span> <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;No IID column in the phenoFile. Please rename the header of the phenoFile&quot;</span><span class="p">)</span>
        <span class="k">if</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span> <span class="o">==</span> <span class="n">FALSE</span><span class="p">)</span> <span class="n">pheno</span><span class="err">$</span><span class="n">FID</span> <span class="o">=</span> <span class="n">pheno</span><span class="err">$</span><span class="n">IID</span>
    <span class="p">}</span>
    <span class="c1"># Make the unique ID by merge FID and IID</span>
    <span class="n">pheno</span><span class="err">$</span><span class="n">ID</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="n">pheno</span><span class="err">$</span><span class="n">FID</span><span class="p">,</span><span class="n">pheno</span><span class="err">$</span><span class="n">IID</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">)</span>
    <span class="c1">#check duplicated ID</span>
    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">pheno</span><span class="err">$</span><span class="n">ID</span><span class="p">))</span><span class="o">!=</span><span class="n">length</span><span class="p">(</span><span class="n">pheno</span><span class="err">$</span><span class="n">ID</span><span class="p">))</span> <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;There are duplicated names in IID column of phenoFile&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">pops</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span> <span class="n">pheno</span> <span class="o">&lt;-</span> <span class="n">pheno</span> <span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">pop_col</span> <span class="k">if</span> <span class="n">pop_col</span> <span class="k">else</span>  <span class="s2">&quot;pop&quot;</span><span class="p">}</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pops</span> <span class="o">|</span> <span class="err">$</span><span class="p">{</span><span class="n">label_col</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span>  <span class="s2">&quot;pop&quot;</span><span class="p">}</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pops</span><span class="p">)</span>
    <span class="n">pca</span> <span class="o">&lt;-</span><span class="n">merge</span><span class="p">(</span><span class="n">pheno</span><span class="p">,</span> <span class="n">pca</span><span class="p">,</span><span class="n">by</span> <span class="o">=</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span> <span class="nb">all</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span> 
    <span class="c1">#</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">pop_col</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span> <span class="p">{</span>
        <span class="c1"># remove populations have less than ${min_pop_size} samples</span>
        <span class="n">pop</span><span class="o">&lt;-</span><span class="n">names</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">pca</span><span class="err">$$</span><span class="p">{</span><span class="n">pop_col</span> <span class="k">if</span> <span class="n">pop_col</span> <span class="k">else</span> <span class="s2">&quot;pop&quot;</span><span class="p">}))</span>
        <span class="n">pop_filter</span><span class="o">&lt;-</span><span class="n">pop</span><span class="p">[</span><span class="n">table</span><span class="p">(</span><span class="n">pca</span><span class="err">$</span><span class="n">pop</span><span class="p">)</span><span class="o">&lt;</span><span class="err">$</span><span class="p">{</span><span class="n">min_pop_size</span><span class="p">}]</span> <span class="c1"># pop to be removed</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">pop_filter</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">warning</span><span class="p">(</span><span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="n">pop_filter</span><span class="p">){</span><span class="n">cat</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;;&#39;</span><span class="p">)},</span><span class="s1">&#39;these &#39;</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">pop_filter</span><span class="p">),</span><span class="s2">&quot; population will be removed due to having less than $</span><span class="si">{min_pop_size}</span><span class="s2"> samples in data.&quot;</span><span class="p">)</span>
            <span class="c1"># remove</span>
            <span class="n">pca</span><span class="o">&lt;-</span><span class="n">pca</span><span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;!</span><span class="si">{</span><span class="n">pop_col</span><span class="si">}</span><span class="s1">%in%pop_filter&#39;</span> <span class="k">if</span> <span class="n">pop_col</span> <span class="k">else</span> <span class="n">pop_col</span><span class="p">})</span>
        <span class="p">}</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
      <span class="n">pca</span><span class="err">$</span><span class="n">pop</span> <span class="o">&lt;-</span> <span class="mi">1</span> 
    <span class="p">}</span>

    <span class="c1"># Write the PC scores to a file</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">pca</span><span class="p">,</span><span class="s2">&quot;$</span><span class="si">{_output:n}</span><span class="s2">.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pca_model</span> <span class="o">=</span> <span class="n">f</span><span class="p">,</span> <span class="n">pc_scores</span> <span class="o">=</span> <span class="n">pca</span><span class="p">,</span> <span class="n">meta</span> <span class="o">=</span> <span class="s2">&quot;${_input[1]:bn} $</span><span class="si">{suffix}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># compute centroids before projecting back the samples</span>
    <span class="c1"># (calculate mean/median/cov per pop)</span>
    <span class="k">if</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;FALSE&quot;</span> <span class="k">if</span> <span class="n">homogeneous</span> <span class="k">else</span> <span class="s2">&quot;TRUE&quot;</span><span class="p">}){</span>
        <span class="n">pop_group</span> <span class="o">&lt;-</span> <span class="n">split</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span><span class="p">[</span> <span class="p">,</span><span class="n">c</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;PC&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">maha_k</span><span class="p">}))],</span> <span class="nb">list</span><span class="p">(</span><span class="n">Group</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span><span class="err">$$</span><span class="p">{</span><span class="n">pop_col</span> <span class="k">if</span> <span class="n">pop_col</span> <span class="k">else</span> <span class="s2">&quot;pop&quot;</span><span class="p">}))</span>
        <span class="n">dat</span><span class="err">$</span><span class="n">pc_cov</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">pop_group</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">cov</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="n">dat</span><span class="err">$</span><span class="n">pc_mean</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">pop_group</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">sapply</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">mean</span><span class="p">))</span>
        <span class="n">dat</span><span class="err">$</span><span class="n">pc_median</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">pop_group</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">sapply</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">median</span><span class="p">))</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">dat</span><span class="err">$</span><span class="n">pc_cov</span> <span class="o">&lt;-</span> <span class="n">cov</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">[,</span><span class="mi">1</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">maha_k</span><span class="p">}])</span>
        <span class="n">dat</span><span class="err">$</span><span class="n">pc_mean</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">[,</span><span class="mi">1</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">maha_k</span><span class="p">}],</span> <span class="mi">2</span><span class="p">,</span> <span class="n">mean</span><span class="p">)</span>
        <span class="n">dat</span><span class="err">$</span><span class="n">pc_median</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">[,</span><span class="mi">1</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">maha_k</span><span class="p">}],</span> <span class="mi">2</span><span class="p">,</span> <span class="n">median</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="c1"># save results</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="project-related-individuals-back">
<h2>Project related individuals back<a class="headerlink" href="#project-related-individuals-back" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Project back to PCA model additional samples</span>
<span class="p">[</span><span class="n">project_samples_1</span><span class="p">]</span>
<span class="c1"># Plink binary file</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># The phenotypic file</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span><span class="s2">&quot;.fam&quot;</span><span class="p">))</span>
<span class="kn">parameter:</span> <span class="n">pca_model</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">phenoFile</span><span class="si">:</span><span class="s1">bn</span><span class="si">}{</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="o">+</span><span class="kp">name</span><span class="p">)</span> <span class="k">if</span> <span class="kp">name</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="p">(</span><span class="n">suffix</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">suffix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">pca.rds&#39;</span>
<span class="c1">## Input genoFile here is for related samples</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">phenoFile</span><span class="p">,</span> <span class="n">pca_model</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">phenoFile</span><span class="si">:</span><span class="s1">bn</span><span class="si">}{</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="o">+</span><span class="kp">name</span><span class="p">)</span> <span class="k">if</span> <span class="kp">name</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.pca.projected.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="c1"># Load required libraries</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">flashpcaR</span><span class="p">)</span>
    <span class="c1"># Read the PLINK binary files</span>
    <span class="n">frel</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nr</span><span class="p">}</span>
    <span class="c1"># Read loadings, center and scale from previous PCA</span>
    <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">f</span> <span class="o">&lt;-</span> <span class="n">dat</span><span class="err">$</span><span class="n">pca_model</span>                           
    <span class="c1"># Read the bim file to obtain reference alleles</span>
    <span class="n">bim</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input[0]:n}</span><span class="s1">.bim&#39;</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span><span class="n">F</span><span class="p">)</span>
    <span class="n">ref</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">bim</span><span class="p">[,</span><span class="mi">5</span><span class="p">])</span>
    <span class="n">names</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">bim</span><span class="p">[,</span><span class="mi">2</span><span class="p">]</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">loadings</span><span class="p">)</span> <span class="o">!=</span> <span class="n">length</span><span class="p">(</span><span class="n">ref</span><span class="p">))</span> <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;Input number of variants should be the same as used in the previous PCA model&quot;</span><span class="p">)</span>
    <span class="n">overlapped_variants</span> <span class="o">&lt;-</span> <span class="n">match</span><span class="p">(</span><span class="n">names</span><span class="p">(</span><span class="n">ref</span><span class="p">),</span> <span class="n">rownames</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">loadings</span><span class="p">))</span>
    <span class="c1"># Project related samples</span>
    <span class="n">fpro</span> <span class="o">&lt;-</span> <span class="n">project</span><span class="p">(</span><span class="n">frel</span><span class="p">,</span> <span class="n">loadings</span><span class="o">=</span><span class="n">f</span><span class="err">$</span><span class="n">loadings</span><span class="p">[</span><span class="n">overlapped_variants</span><span class="p">,],</span> <span class="n">orig_mean</span><span class="o">=</span><span class="n">f</span><span class="err">$</span><span class="n">center</span><span class="p">[</span><span class="n">overlapped_variants</span><span class="p">],</span> <span class="n">orig_sd</span><span class="o">=</span><span class="n">f</span><span class="err">$</span><span class="n">scale</span><span class="p">[</span><span class="n">overlapped_variants</span><span class="p">],</span> <span class="n">ref_allele</span><span class="o">=</span><span class="n">ref</span><span class="p">)</span>
    <span class="n">pca</span> <span class="o">&lt;-</span> <span class="n">fpro</span><span class="err">$</span><span class="n">projection</span>
    <span class="n">k</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">k</span><span class="p">}</span>
    <span class="n">pca</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">fpro</span><span class="err">$</span><span class="n">projection</span><span class="p">)</span>
    <span class="n">pca</span> <span class="o">&lt;-</span> <span class="n">tibble</span><span class="p">::</span><span class="n">rownames_to_column</span><span class="p">(</span><span class="n">pca</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">pca</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;PC&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="n">k</span><span class="p">))</span>
    
    <span class="c1"># Read fam file with phenotypes</span>
    <span class="k">if</span><span class="p">(</span><span class="n">stringr</span><span class="p">::</span><span class="n">str_detect</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span><span class="s2">&quot;.fam$&quot;</span><span class="p">)){</span>
        <span class="n">pheno</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">F</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span><span class="n">F</span><span class="p">)</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;IID&quot;</span><span class="p">,</span> <span class="s2">&quot;MID&quot;</span><span class="p">,</span> <span class="s2">&quot;PID&quot;</span><span class="p">,</span> <span class="s2">&quot;SEX&quot;</span><span class="p">,</span> <span class="s2">&quot;STATUS&quot;</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">pheno</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span><span class="n">F</span><span class="p">)</span>
        <span class="k">if</span><span class="p">(</span><span class="s2">&quot;IID&quot;</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span> <span class="o">==</span> <span class="n">FALSE</span><span class="p">)</span> <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;No IID column in the phenoFile. Please rename the header of the phenoFile&quot;</span><span class="p">)</span>
        <span class="k">if</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span> <span class="o">==</span> <span class="n">FALSE</span><span class="p">)</span> <span class="n">pheno</span><span class="err">$</span><span class="n">FID</span> <span class="o">=</span> <span class="n">pheno</span><span class="err">$</span><span class="n">IID</span>
    <span class="p">}</span>
    <span class="c1"># Make the unique ID by merge FID and IID</span>
    <span class="n">pheno</span><span class="err">$</span><span class="n">ID</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="n">pheno</span><span class="err">$</span><span class="n">FID</span><span class="p">,</span><span class="n">pheno</span><span class="err">$</span><span class="n">IID</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">)</span>
    <span class="c1">#check duplicated ID</span>
    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">pheno</span><span class="err">$</span><span class="n">ID</span><span class="p">))</span><span class="o">!=</span><span class="n">length</span><span class="p">(</span><span class="n">pheno</span><span class="err">$</span><span class="n">ID</span><span class="p">))</span> <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;There are duplicated names in IID column of phenoFile&quot;</span><span class="p">)</span>

    <span class="n">pca</span> <span class="o">&lt;-</span><span class="n">merge</span><span class="p">(</span><span class="n">pheno</span><span class="p">,</span> <span class="n">pca</span><span class="p">,</span><span class="n">by</span> <span class="o">=</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span> <span class="nb">all</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span> <span class="o">=</span> <span class="n">bind_rows</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span><span class="p">,</span> <span class="n">pca</span><span class="p">)</span>
    
    <span class="c1"># Write the PC scores to a file</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span><span class="p">,</span><span class="s2">&quot;$</span><span class="si">{_output:n}</span><span class="s2">.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="c1"># save results</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="plot-pca-results">
<h2>Plot PCA results<a class="headerlink" href="#plot-pca-results" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot PCA results. </span>
<span class="c1"># Can be used independently as &quot;plot_pca&quot; or combined with other workflow as eg &quot;flashpca+plot_pca&quot;</span>
<span class="p">[</span><span class="n">plot_pca</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">outlier_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">plot_data</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">min_axis</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">max_axis</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">input:</span> <span class="n">plot_data</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.pc.png&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.scree.png&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.scree.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">outlier_file</span><span class="si">:</span><span class="s2">ad</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">outlier_file</span><span class="si">:</span><span class="s2">ad</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>   
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">matrixStats</span><span class="p">)</span>
    <span class="n">pops</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">(</span><span class="n">pops</span><span class="p">):</span><span class="n">r</span><span class="p">,})</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">f</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">pca_model</span>
    <span class="n">pca_final</span><span class="o">&lt;-</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span>
    
    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">pops</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">)</span> <span class="n">pca_final</span> <span class="o">&lt;-</span> <span class="n">pca_final</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">pop_col</span> <span class="k">if</span> <span class="n">pop_col</span> <span class="k">else</span> <span class="s2">&quot;pop&quot;</span><span class="p">}</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pops</span> <span class="o">|</span> <span class="err">$</span><span class="p">{</span><span class="n">label_col</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span>  <span class="s2">&quot;pop&quot;</span><span class="p">}</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pops</span><span class="p">)</span> 
    <span class="n">pca_final</span> <span class="o">&lt;-</span> <span class="n">pca_final</span> <span class="o">%&gt;%</span> <span class="n">mutate</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">label_col</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span> <span class="s2">&quot;pop&quot;</span><span class="p">}</span><span class="o">=</span><span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">label_col</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span>  <span class="s2">&quot;pop&quot;</span><span class="p">}))</span>
    <span class="n">k</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">k</span><span class="p">}</span>
    <span class="c1"># manually set colors for PCA plotting, to avoid similar colors in one plot:</span>
    <span class="c1"># generated by https://mokole.com/palette.html</span>
    <span class="nb">set</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">999</span><span class="p">)</span>
    <span class="n">colors_40</span> <span class="o">=</span> <span class="n">sample</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;#a9a9a9&quot;</span><span class="p">,</span> <span class="s2">&quot;#2f4f4f&quot;</span><span class="p">,</span> <span class="s2">&quot;#556b2f&quot;</span><span class="p">,</span> <span class="s2">&quot;#a0522d&quot;</span><span class="p">,</span> <span class="s2">&quot;#7f0000&quot;</span><span class="p">,</span> <span class="s2">&quot;#006400&quot;</span><span class="p">,</span> <span class="s2">&quot;#808000&quot;</span><span class="p">,</span> <span class="s2">&quot;#483d8b&quot;</span><span class="p">,</span> <span class="s2">&quot;#3cb371&quot;</span><span class="p">,</span> <span class="s2">&quot;#bdb76b&quot;</span><span class="p">,</span> <span class="s2">&quot;#4682b4&quot;</span><span class="p">,</span> <span class="s2">&quot;#9acd32&quot;</span><span class="p">,</span> 
                          <span class="s2">&quot;#20b2aa&quot;</span><span class="p">,</span> <span class="s2">&quot;#00008b&quot;</span><span class="p">,</span> <span class="s2">&quot;#32cd32&quot;</span><span class="p">,</span> <span class="s2">&quot;#daa520&quot;</span><span class="p">,</span> <span class="s2">&quot;#7f007f&quot;</span><span class="p">,</span> <span class="s2">&quot;#b03060&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff0000&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff8c00&quot;</span><span class="p">,</span> <span class="s2">&quot;#ffff00&quot;</span><span class="p">,</span> <span class="s2">&quot;#0000cd&quot;</span><span class="p">,</span> <span class="s2">&quot;#00ff00&quot;</span><span class="p">,</span> <span class="s2">&quot;#9400d3&quot;</span><span class="p">,</span> 
                          <span class="s2">&quot;#00fa9a&quot;</span><span class="p">,</span> <span class="s2">&quot;#00ffff&quot;</span><span class="p">,</span> <span class="s2">&quot;#00bfff&quot;</span><span class="p">,</span> <span class="s2">&quot;#f4a460&quot;</span><span class="p">,</span> <span class="s2">&quot;#f08080&quot;</span><span class="p">,</span> <span class="s2">&quot;#adff2f&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff6347&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff00ff&quot;</span><span class="p">,</span> <span class="s2">&quot;#1e90ff&quot;</span><span class="p">,</span> <span class="s2">&quot;#dda0dd&quot;</span><span class="p">,</span> <span class="s2">&quot;#7b68ee&quot;</span><span class="p">,</span> <span class="s2">&quot;#afeeee&quot;</span><span class="p">,</span> 
                          <span class="s2">&quot;#ee82ee&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff69b4&quot;</span><span class="p">,</span> <span class="s2">&quot;#ffe4c4&quot;</span><span class="p">,</span> <span class="s2">&quot;#ffc0cb&quot;</span><span class="p">))</span>
    <span class="n">colors_20</span> <span class="o">=</span> <span class="n">sample</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;#2f4f4f&quot;</span><span class="p">,</span> <span class="s2">&quot;#2e8b57&quot;</span><span class="p">,</span> <span class="s2">&quot;#8b0000&quot;</span><span class="p">,</span> <span class="s2">&quot;#808000&quot;</span><span class="p">,</span> <span class="s2">&quot;#00008b&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff0000&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff8c00&quot;</span><span class="p">,</span> <span class="s2">&quot;#00ff00&quot;</span><span class="p">,</span> <span class="s2">&quot;#4169e1&quot;</span><span class="p">,</span> <span class="s2">&quot;#00ffff&quot;</span><span class="p">,</span> <span class="s2">&quot;#00bfff&quot;</span><span class="p">,</span> <span class="s2">&quot;#0000ff&quot;</span><span class="p">,</span> 
                          <span class="s2">&quot;#da70d6&quot;</span><span class="p">,</span> <span class="s2">&quot;#d8bfd8&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff00ff&quot;</span><span class="p">,</span> <span class="s2">&quot;#eee8aa&quot;</span><span class="p">,</span> <span class="s2">&quot;#ffff54&quot;</span><span class="p">,</span> <span class="s2">&quot;#ff1493&quot;</span><span class="p">,</span> <span class="s2">&quot;#ffa07a&quot;</span><span class="p">,</span> <span class="s2">&quot;#98fb98&quot;</span><span class="p">))</span>

    <span class="c1"># assign colors to each ethnicity:</span>
    <span class="n">num_col</span><span class="o">=</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">pca_final</span><span class="err">$$</span><span class="p">{</span><span class="n">label_col</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span> <span class="s2">&quot;pop&quot;</span><span class="p">}))</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">num_col</span> <span class="o">&lt;=</span> <span class="mi">20</span><span class="p">)</span> <span class="p">{</span>
       <span class="n">color_list</span> <span class="o">&lt;-</span> <span class="n">colors_20</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="n">num_col</span><span class="p">]</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
       <span class="n">color_list</span> <span class="o">&lt;-</span> <span class="n">colors_40</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="n">num_col</span><span class="p">]</span>
    <span class="p">}</span>

    <span class="c1">###</span>
    <span class="c1"># Make the plots</span>
    <span class="c1">###</span>
    <span class="c1"># Get the min and max values for x and y-axes</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">min_axis</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">max_axis</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span> <span class="p">{</span>
        <span class="n">min_axis</span> <span class="o">&lt;-</span> <span class="nb">round</span><span class="p">(</span><span class="n">colMins</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">[</span><span class="n">sapply</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">,</span> <span class="ow">is</span><span class="o">.</span><span class="n">numeric</span><span class="p">)])),</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">max_axis</span> <span class="o">&lt;-</span> <span class="nb">round</span><span class="p">(</span><span class="n">colMaxs</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">[</span><span class="n">sapply</span><span class="p">(</span><span class="n">f</span><span class="err">$</span><span class="n">projection</span><span class="p">,</span> <span class="ow">is</span><span class="o">.</span><span class="n">numeric</span><span class="p">)])),</span><span class="mi">1</span><span class="p">)</span>  
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">min_axis</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">double</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">min_axis</span><span class="p">})</span>
        <span class="n">max_axis</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">double</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">max_axis</span><span class="p">})</span>
    <span class="p">}</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">outlier_file</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span> <span class="p">{</span>
        <span class="n">outliers</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">outlier_file</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;IID&quot;</span><span class="p">),</span><span class="n">stringsAsFactors</span> <span class="o">=</span><span class="n">F</span><span class="p">)</span>
        <span class="n">plot_pcs</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">pca_final</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">ggplot</span><span class="p">(</span><span class="n">pca_final</span><span class="p">,</span> <span class="n">aes_string</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">))</span> <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;aes(color=</span><span class="si">{</span><span class="n">label_col</span><span class="si">}</span><span class="s1">)&#39;</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">})</span> <span class="o">+</span> 
              <span class="c1"># add circles for these ouliters:</span>
              <span class="n">geom_point</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="nb">filter</span><span class="p">(</span><span class="n">pca_final</span><span class="p">,</span> <span class="n">IID</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">outliers</span><span class="err">$</span><span class="n">IID</span><span class="p">,</span> <span class="n">FID</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">outliers</span><span class="err">$</span><span class="n">FID</span><span class="p">),</span> <span class="n">shape</span> <span class="o">=</span> <span class="mi">21</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">stroke</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">)</span> <span class="o">+</span>
              <span class="c1"># add outliers dots:</span>
              <span class="n">geom_point</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="nb">filter</span><span class="p">(</span><span class="n">pca_final</span><span class="p">,</span> <span class="n">IID</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">outliers</span><span class="err">$</span><span class="n">IID</span><span class="p">,</span> <span class="n">FID</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">outliers</span><span class="err">$</span><span class="n">FID</span><span class="p">),</span> <span class="n">shape</span> <span class="o">=</span> <span class="mi">16</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;,aes(color=</span><span class="si">{</span><span class="n">label_col</span><span class="si">}</span><span class="s1">)&#39;</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> <span class="p">)</span> <span class="o">+</span> 
              <span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">)</span> <span class="o">+</span>
              <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">limits</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="n">min_axis</span><span class="p">,</span> <span class="n">max_axis</span><span class="p">))</span> <span class="o">+</span>
              <span class="n">scale_x_continuous</span><span class="p">(</span><span class="n">limits</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="n">min_axis</span><span class="p">,</span> <span class="n">max_axis</span><span class="p">))</span> <span class="o">+</span>
              <span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">color_list</span><span class="p">)</span> <span class="o">+</span>
              <span class="n">theme_classic</span><span class="p">()</span>
        <span class="p">}}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">plot_pcs</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">pca_final</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="p">{</span>
          <span class="n">ggplot</span><span class="p">(</span><span class="n">pca_final</span><span class="p">,</span> <span class="n">aes_string</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">))</span> <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;aes(color=</span><span class="si">{</span><span class="n">label_col</span><span class="si">}</span><span class="s1">)&#39;</span> <span class="k">if</span> <span class="n">label_col</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">})</span> <span class="o">+</span> 
              <span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">)</span> <span class="o">+</span>
              <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">limits</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="n">min_axis</span><span class="p">,</span> <span class="n">max_axis</span><span class="p">))</span> <span class="o">+</span>
              <span class="n">scale_x_continuous</span><span class="p">(</span><span class="n">limits</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="n">min_axis</span><span class="p">,</span> <span class="n">max_axis</span><span class="p">))</span> <span class="o">+</span>
              <span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">color_list</span><span class="p">)</span> <span class="o">+</span>
              <span class="n">theme_classic</span><span class="p">()</span>
        <span class="p">}}</span>
    <span class="n">unit</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">n_col</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
    <span class="n">n_row</span> <span class="o">=</span> <span class="n">ceiling</span><span class="p">(</span><span class="n">k</span> <span class="o">/</span> <span class="n">n_col</span><span class="p">)</span>
    <span class="n">plots</span> <span class="o">=</span> <span class="n">lapply</span><span class="p">(</span><span class="mi">1</span><span class="p">:(</span><span class="n">k</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="n">plot_pcs</span><span class="p">(</span><span class="n">pca_final</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;PC&quot;</span><span class="p">,</span><span class="n">i</span><span class="p">),</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;PC&quot;</span><span class="p">,</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">dat</span><span class="err">$</span><span class="n">meta</span><span class="p">))</span>
    <span class="n">png</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output[0]}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="n">unit</span> <span class="o">*</span> <span class="n">n_col</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="n">unit</span> <span class="o">*</span> <span class="n">n_row</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">::</span><span class="n">grid</span><span class="o">.</span><span class="n">arrange</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="n">plots</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">ncol</span> <span class="o">=</span> <span class="n">n_col</span><span class="p">,</span> <span class="n">nrow</span> <span class="o">=</span> <span class="n">n_row</span><span class="p">)))</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    <span class="c1"># Create scree plot</span>
    <span class="n">PVE</span> <span class="o">&lt;-</span> <span class="n">f</span><span class="err">$</span><span class="n">values</span>
    <span class="n">PVE</span> <span class="o">&lt;-</span> <span class="nb">round</span><span class="p">(</span><span class="n">PVE</span><span class="o">/</span><span class="nb">sum</span><span class="p">(</span><span class="n">PVE</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">PVEplot</span> <span class="o">&lt;-</span> <span class="n">qplot</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">PVE</span><span class="p">)),</span> <span class="n">PVE</span><span class="p">)</span> <span class="o">+</span> <span class="n">geom_line</span><span class="p">()</span> <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;Principal Component&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s2">&quot;PVE&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">ggtitle</span><span class="p">(</span><span class="s2">&quot;Scree Plot&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span><span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">limits</span><span class="o">=</span><span class="n">factor</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">PVE</span><span class="p">)))</span>
    <span class="n">PVE_cum</span> <span class="o">&lt;-</span> <span class="n">cumsum</span><span class="p">(</span><span class="n">PVE</span><span class="p">)</span><span class="o">/</span><span class="nb">sum</span><span class="p">(</span><span class="n">PVE</span><span class="p">)</span>
    <span class="n">cumPVEplot</span> <span class="o">&lt;-</span> <span class="n">qplot</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">PVE</span><span class="p">)),</span> <span class="n">cumsum</span><span class="p">(</span><span class="n">PVE</span><span class="p">))</span> <span class="o">+</span> <span class="n">geom_line</span><span class="p">()</span> <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;Principal Component&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s2">&quot;PVE&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">ggtitle</span><span class="p">(</span><span class="s2">&quot;Cumulative PVE Plot&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">limits</span><span class="o">=</span><span class="n">factor</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">PVE</span><span class="p">)))</span>
    <span class="n">png</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output[1]}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">grid</span><span class="o">.</span><span class="n">arrange</span><span class="p">(</span><span class="n">PVEplot</span><span class="p">,</span> <span class="n">cumPVEplot</span><span class="p">,</span> <span class="n">nrow</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    <span class="c1">## Textual Output</span>
    
    <span class="n">PVE_output</span> <span class="o">=</span> <span class="n">tibble</span><span class="p">(</span><span class="n">PCs</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">PVE</span><span class="p">),</span> <span class="n">PVE</span> <span class="o">=</span> <span class="n">PVE</span><span class="p">,</span> <span class="n">PVE_cum</span> <span class="o">=</span> <span class="n">PVE_cum</span> <span class="p">)</span>
    <span class="n">PVE_output</span><span class="o">%&gt;%</span><span class="n">readr</span><span class="p">::</span><span class="n">write_delim</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="detect-outliers">
<h2>Detect outliers<a class="headerlink" href="#detect-outliers" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate Mahalanobis distance per population and report outliers</span>
<span class="p">[</span><span class="n">detect_outliers</span><span class="p">]</span>
<span class="c1"># Set the probability to remove outliers eg 0.95 or 0.997</span>
<span class="kn">parameter:</span> <span class="n">prob</span> <span class="o">=</span> <span class="mf">0.997</span>
<span class="c1"># Mahalanobis distance p-value cutoff</span>
<span class="kn">parameter:</span> <span class="n">pval</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="c1"># Robust Mahalanobis to outliers</span>
<span class="kn">parameter:</span> <span class="n">robust</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">parameter:</span> <span class="n">pca_result</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">pca_result</span>
<span class="kn">output:</span> <span class="n">distance</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis.rds&#39;</span><span class="p">,</span>
        <span class="n">identified_outliers</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.outliers&#39;</span><span class="p">,</span>
        <span class="n">analysis_summary</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.analysis_summary.md&#39;</span><span class="p">,</span>
        <span class="n">qqplot_mahalanobis</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis_qq.png&#39;</span><span class="p">,</span>
        <span class="n">hist_mahalanobis</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis_hist.png&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span>
    <span class="n">echo</span> <span class="s1">&#39;&#39;&#39;---</span>
<span class="s1">    theme: base-theme</span>
<span class="s1">    style: |</span>
<span class="s1">      img {</span>
<span class="s1">        height: 80%;</span>
<span class="s1">        display: block;</span>
<span class="s1">        margin-left: auto;</span>
<span class="s1">        margin-right: auto;</span>
<span class="s1">      }</span>
<span class="s1">    ---    </span>
<span class="s1">    &#39;&#39;&#39;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]}</span>
    
<span class="kn">R:</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="c1"># Load required libraries</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span>
  
    <span class="c1"># invert a known covariance matrix but allow them to be numerically singular matrix (still assuming full rank)</span>
    <span class="n">robust_inv</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1">#tryCatch(solve(s), error=function(cond) solve(Matrix::nearPD(s)$mat))</span>
        <span class="n">tryCatch</span><span class="p">(</span><span class="n">solve</span><span class="p">(</span><span class="n">s</span><span class="p">),</span> <span class="n">error</span><span class="o">=</span><span class="n">function</span><span class="p">(</span><span class="n">cond</span><span class="p">)</span> <span class="n">MASS</span><span class="p">::</span><span class="n">ginv</span><span class="p">(</span><span class="n">s</span><span class="p">))</span>
    <span class="p">}</span>

    <span class="c1"># Calculate mahalanobis distance  </span>
    <span class="n">calc_mahalanobis_dist</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">s</span><span class="p">,</span> <span class="kp">name</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">prob</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">prob</span><span class="p">})</span> <span class="p">{</span>
        <span class="n">pc</span> <span class="o">&lt;-</span> <span class="n">x</span> <span class="o">%&gt;%</span>
          <span class="n">select</span><span class="p">(</span><span class="s2">&quot;IID&quot;</span><span class="p">,</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;${pop_col if pop_col else &quot;</span><span class="n">pop</span><span class="s2">&quot;}&quot;</span><span class="p">,</span> <span class="n">starts_with</span><span class="p">(</span><span class="s2">&quot;PC&quot;</span><span class="p">))</span>
        <span class="n">mu_pc</span> <span class="o">&lt;-</span> <span class="n">pc</span><span class="p">[,</span><span class="mi">4</span><span class="p">:(</span><span class="mi">4</span> <span class="o">+</span> <span class="n">length</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="n">pc</span><span class="err">$</span><span class="n">mahal</span> <span class="o">=</span> <span class="n">mahalanobis</span><span class="p">(</span><span class="n">mu_pc</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">robust_inv</span><span class="p">(</span><span class="n">s</span><span class="p">),</span> <span class="n">inverted</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">pc</span><span class="err">$</span><span class="n">p</span> <span class="o">&lt;-</span> <span class="n">pchisq</span><span class="p">(</span><span class="n">pc</span><span class="err">$</span><span class="n">mahal</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">nrow</span><span class="p">(</span><span class="n">s</span><span class="p">),</span> <span class="n">lower</span><span class="o">.</span><span class="n">tail</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
        <span class="n">manh_dis_sq_cutoff</span> <span class="o">=</span> <span class="n">quantile</span><span class="p">(</span><span class="n">pc</span><span class="err">$</span><span class="n">mahal</span><span class="p">,</span> <span class="n">probs</span><span class="o">=</span><span class="n">prob</span><span class="p">)</span>
        <span class="c1"># Obtain outliers</span>
        <span class="n">outliers</span> <span class="o">=</span> <span class="n">pc</span><span class="p">[(</span><span class="n">pc</span><span class="err">$</span><span class="n">mahal</span> <span class="o">&gt;</span> <span class="n">manh_dis_sq_cutoff</span> <span class="o">&amp;</span> <span class="n">pc</span><span class="err">$</span><span class="n">p</span> <span class="o">&lt;</span> <span class="err">$</span><span class="p">{</span><span class="n">pval</span><span class="p">}),]</span>
        <span class="n">d_summary</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">capture</span><span class="o">.</span><span class="n">output</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">pc</span><span class="err">$</span><span class="n">mahal</span><span class="p">)),</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="s2">&quot;result summary</span><span class="se">\n</span><span class="s2">## Mahalanobis distance summary:</span><span class="se">\n</span><span class="s2">```</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">d_summary</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">```</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> 
            <span class="n">paste</span><span class="p">(</span><span class="s2">&quot;The cut-off for outlier removal is set to:&quot;</span><span class="p">,</span> <span class="n">manh_dis_sq_cutoff</span><span class="p">,</span> <span class="s2">&quot;and the number of individuals to remove is:&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="p">(</span><span class="n">outliers</span><span class="p">),</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">),</span>
            <span class="n">paste</span><span class="p">(</span><span class="s2">&quot;The new sample size after outlier removal is:&quot;</span><span class="p">,</span><span class="n">nrow</span><span class="p">(</span><span class="n">pc</span><span class="p">)</span> <span class="o">-</span> <span class="n">nrow</span><span class="p">(</span><span class="n">outliers</span><span class="p">),</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">))</span>
        <span class="c1">#</span>
        <span class="n">outliers</span> <span class="o">&lt;-</span> <span class="n">outliers</span> <span class="o">%&gt;%</span>
        <span class="n">select</span><span class="p">(</span><span class="n">FID</span><span class="p">,</span><span class="n">IID</span><span class="p">)</span>
      <span class="nb">list</span><span class="p">(</span><span class="n">pc</span><span class="o">=</span><span class="n">pc</span><span class="p">,</span> <span class="n">manh_dis_sq_cutoff</span><span class="o">=</span><span class="n">manh_dis_sq_cutoff</span><span class="p">,</span> <span class="n">msg</span><span class="o">=</span><span class="n">msg</span><span class="p">,</span> <span class="n">outliers</span><span class="o">=</span><span class="n">outliers</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">list</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_mean</span><span class="p">))</span> <span class="p">{</span>
      <span class="n">pops</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_mean</span><span class="p">)</span>
      <span class="n">pop_group</span> <span class="o">=</span> <span class="n">split</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span><span class="p">,</span> <span class="n">f</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span><span class="err">$$</span><span class="p">{</span><span class="n">pop_col</span> <span class="k">if</span> <span class="n">pop_col</span> <span class="k">else</span> <span class="s2">&quot;pop&quot;</span><span class="p">})</span>
      <span class="n">res</span> <span class="o">=</span> <span class="n">lapply</span><span class="p">(</span><span class="n">pops</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="n">calc_mahalanobis_dist</span><span class="p">(</span><span class="n">pop_group</span><span class="p">[[</span><span class="n">p</span><span class="p">]],</span> <span class="n">dat</span><span class="err">$$</span><span class="p">{</span><span class="s2">&quot;pc_mean&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">robust</span> <span class="k">else</span> <span class="s2">&quot;pc_median&quot;</span><span class="p">}[[</span><span class="n">p</span><span class="p">]],</span> <span class="n">dat</span><span class="err">$</span><span class="n">pc_cov</span><span class="p">[[</span><span class="n">p</span><span class="p">]],</span> <span class="kp">name</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">meta</span><span class="p">,</span> <span class="n">p</span><span class="p">)))</span>
      <span class="n">names</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">=</span> <span class="n">pops</span>
      <span class="n">res</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
          <span class="n">msg</span> <span class="o">=</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">paste</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="n">lapply</span><span class="p">(</span><span class="n">pops</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="n">res</span><span class="p">[[</span><span class="n">p</span><span class="p">]]</span><span class="err">$</span><span class="n">msg</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)),</span>
          <span class="n">manh_dis_sq_cutoff</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">pops</span><span class="p">,</span> <span class="n">sapply</span><span class="p">(</span><span class="n">pops</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="n">res</span><span class="p">[[</span><span class="n">p</span><span class="p">]]</span><span class="err">$</span><span class="n">manh_dis_sq_cutoff</span><span class="p">)),</span>
          <span class="n">outliers</span> <span class="o">=</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="n">lapply</span><span class="p">(</span><span class="n">pops</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="n">res</span><span class="p">[[</span><span class="n">p</span><span class="p">]]</span><span class="err">$</span><span class="n">outliers</span><span class="p">))),</span>
          <span class="n">pc</span> <span class="o">=</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="n">lapply</span><span class="p">(</span><span class="n">pops</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="n">res</span><span class="p">[[</span><span class="n">p</span><span class="p">]]</span><span class="err">$</span><span class="n">pc</span><span class="p">)))</span>
          <span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
      <span class="n">res</span> <span class="o">=</span> <span class="n">calc_mahalanobis_dist</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">pc_scores</span><span class="p">,</span> <span class="n">dat</span><span class="err">$$</span><span class="p">{</span><span class="s2">&quot;pc_mean&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">robust</span> <span class="k">else</span> <span class="s2">&quot;pc_median&quot;</span><span class="p">},</span> <span class="n">dat</span><span class="err">$</span><span class="n">pc_cov</span><span class="p">,</span> <span class="kp">name</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">meta</span><span class="p">)</span>
    <span class="p">}</span>
      
    <span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="err">$</span><span class="n">msg</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>   
    <span class="c1"># Plot mahalanobis</span>
    <span class="n">k</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">k</span><span class="p">}</span>
    <span class="n">png</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output[3]}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">qqplot</span><span class="p">(</span><span class="n">qchisq</span><span class="p">(</span><span class="n">ppoints</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="n">df</span><span class="o">=</span><span class="n">k</span><span class="p">),</span> <span class="n">res</span><span class="err">$</span><span class="n">pc</span><span class="err">$</span><span class="n">mahal</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="n">expression</span><span class="p">(</span><span class="s2">&quot;Mahalanobis&quot;</span> <span class="o">*</span> <span class="o">~</span><span class="n">D</span><span class="o">^</span><span class="mi">2</span> <span class="o">*</span> <span class="s2">&quot; vs. quantiles of&quot;</span> <span class="o">*</span> <span class="o">~</span> <span class="n">chi</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">^</span><span class="mi">2</span><span class="p">),</span> <span class="n">xlab</span> <span class="o">=</span> <span class="n">expression</span><span class="p">(</span><span class="n">chi</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">^</span><span class="mi">2</span> <span class="o">*</span> <span class="s2">&quot;, probability points = 100&quot;</span><span class="p">),</span> <span class="n">ylab</span> <span class="o">=</span> <span class="n">expression</span><span class="p">(</span><span class="n">D</span><span class="o">^</span><span class="mi">2</span><span class="p">),</span> <span class="n">pch</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
    <span class="n">abline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span> 
    <span class="n">png</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output[4]}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">res</span><span class="err">$</span><span class="n">pc</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mahal</span><span class="p">))</span> <span class="o">+</span> <span class="n">geom_histogram</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">y</span> <span class="o">=</span> <span class="o">..</span><span class="n">count</span><span class="o">..</span><span class="p">),</span> <span class="n">binwidth</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="s2">&quot;#1F3552&quot;</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s2">&quot;#4271AE&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">scale_x_continuous</span><span class="p">(</span><span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;Mahalanobis distance&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
  
    <span class="c1"># Save results and outliers</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">res</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">res</span><span class="err">$</span><span class="n">outliers</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="add-plot-and-outlier-detection-to-pca-steps">
<h2>Add plot and outlier detection to PCA steps<a class="headerlink" href="#add-plot-and-outlier-detection-to-pca-steps" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">flashpca_2</span><span class="p">,</span> <span class="n">project_samples_2</span><span class="p">]</span>
<span class="c1"># Set the probability to remove outliers eg 0.95 or 0.997</span>
<span class="kn">parameter:</span> <span class="n">prob</span> <span class="o">=</span> <span class="mf">0.997</span>
<span class="c1"># Robust Mahalanobis to outliers</span>
<span class="kn">parameter:</span> <span class="n">robust</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">output:</span> <span class="n">distance</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis.rds&#39;</span><span class="p">,</span>
        <span class="n">identified_outliers</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.outliers&#39;</span><span class="p">,</span>
        <span class="n">analysis_summary</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.analysis_summary.md&#39;</span><span class="p">,</span>
        <span class="n">qqplot_mahalanobis</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis_qq.png&#39;</span><span class="p">,</span>
        <span class="n">hist_mahalanobis</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis_hist.png&#39;</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s2">&quot;detect_outliers&quot;</span><span class="p">,</span> <span class="n">pca_result</span><span class="o">=</span><span class="n">_input</span><span class="p">,</span> <span class="n">prob</span><span class="o">=</span><span class="n">prob</span><span class="p">,</span> <span class="n">robust</span><span class="o">=</span><span class="n">robust</span><span class="p">)</span>

<span class="p">[</span><span class="n">flashpca_3</span><span class="p">,</span> <span class="n">project_samples_3</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">output_from</span><span class="p">(</span><span class="mi">2</span><span class="p">)[</span><span class="s1">&#39;identified_outliers&#39;</span><span class="p">]</span>
<span class="n">outliers</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">.pc.png&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">.scree.png&quot;</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s2">&quot;plot_pca&quot;</span><span class="p">,</span> <span class="n">plot_data</span> <span class="o">=</span> <span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">outlier_file</span> <span class="o">=</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">outliers</span><span class="p">)</span> <span class="k">else</span> <span class="n">path</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/data_preprocessing/genotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="GWAS_QC.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Genotype PLINK file quality control</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="GRM.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genomic relationship matrices</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>