
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Genotype VCF File Quality Control &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/genotype/VCF_QC';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Genotype PLINK File Quality Control" href="GWAS_QC.html" />
    <link rel="prev" title="Genotype data preprocessing" href="../genotype_preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../genotype_preprocessing.html">Genotype data preprocessing</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL Modifier Score</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_modifier_score/ems_training.html">EMS Training</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/data_preprocessing/genotype/VCF_QC.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype/VCF_QC.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/data_preprocessing/genotype/VCF_QC.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Genotype VCF File Quality Control</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-vcf-qc-filters">Default Parameters: VCF QC Filters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-note-on-ts-tv-summary-from-vcf-genotype-data">A note on TS/TV summary from VCF genotype data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-resource-usage">Default Parameters: Resource Usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-rename-chromosomes">(Optional) Rename Chromosomes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-dbsnp-annotation">(Optional) dbSNP Annotation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-quality-control">i. Quality Control</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#global-parameters">Global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotate-known-and-novel-variants">Annotate known and novel variants</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-qc">Genotype QC</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="genotype-vcf-file-quality-control">
<h1>Genotype VCF File Quality Control<a class="headerlink" href="#genotype-vcf-file-quality-control" title="Link to this heading">#</a></h1>
<p>This implements some recommendations from UK Biobank on <a class="reference external" href="https://www.medrxiv.org/content/10.1101/2020.11.02.20222232v1.full-text">sequence data quality control</a>.</p>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>A major challenge in biomedical research is the quality control (QC) of sequencing data. False positive variant calls can hinder the ability to detect disease associated variants or introduce spurious associations, therefore the need for a rigorous QC. Our pipeline focuses on QC after the variant calling stage and requires project Variant Calling Format (pVCF) as input files. We have defined default thresholds for genotype and variant-level hard filtering based on recommendations from the UK Biobank team and a thorough review of the literature [<a class="reference external" href="https://doi.org/10.1186/1471-2105-15-125">cf. Carson et al. BMC Bioinformatics (2014)</a>,<a class="reference external" href="https://doi.org/10.1038/nature19057">cf. Lek et al. Nature (2016)</a>,<a class="reference external" href="https://doi.org/10.1038/s41588-021-00885-0">cf. Szustakowski et al. Nature Genetics (2021)</a>]. Bcftools is used in our QC steps. We first handle multi-allelic sites by splitting them into bi-allelic records. We include an optional workflow to keep only bi-allelic sites in the data. Variants are then annotated based on dbSNP data. Genotypes are kept if they have a Genotype Depth (DP) &gt;= 10 and a Genotype Quality (GQ) &gt;= 20. Variants are included if at least one sample has an allelic balance (AB) &gt;= 0.15 for Single Nucleotide Variants (SNVs) and AB&gt;=0.2 for indels, variant missigness is below 20% and Hardy-Weinberg Equilibrium p-value is &gt; 1e-08. Allele balance is calculated for heterozygotes as the number of bases supporting the least-represented allele over the total number of base observations. Output summary statistics, such as transistion/transversion ratios (TS/TV ratio) are calculated to determine the effectiveness of QC.</p>
</section>
<section id="default-parameters-vcf-qc-filters">
<h2>Default Parameters: VCF QC Filters<a class="headerlink" href="#default-parameters-vcf-qc-filters" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Genotype depth filters: For WES data, UK Biobank recommends <strong>SNPs DP&gt;10 and Indels DP&gt;10 for indels.</strong> However we think for WGS we can be less stringent, or simply rely on GQ. Users can set it to 1 eg, <code class="docutils literal notranslate"><span class="pre">--DP</span> <span class="pre">1</span> <span class="pre">--DP-indel</span> <span class="pre">1</span> </code></p></li>
<li><p>Genotype quality GQ&gt;20.</p></li>
<li><p>At least one sample per site passed the allele balance threshold &gt;= 0.15 for SNPs and &gt;=0.20 for indels (heterozygous variants).</p>
<ul class="simple">
<li><p>Allele balance is calculated for heterozygotes as the number of bases supporting the least-represented allele over the total number of base observations.</p></li>
</ul>
</li>
<li><p>This module also allows for filtering by HWE and missingness although by default we don’t filter on them.</p></li>
</ol>
<p>Filtering are done with <code class="docutils literal notranslate"><span class="pre">bcftools</span></code>. Here is a <a class="reference external" href="https://gist.github.com/elowy01/93922762e131d7abd3c7e8e166a74a0b">useful cheatsheet from github user &#64;elowy01</a>.</p>
<section id="a-note-on-ts-tv-summary-from-vcf-genotype-data">
<h3>A note on TS/TV summary from VCF genotype data<a class="headerlink" href="#a-note-on-ts-tv-summary-from-vcf-genotype-data" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">stats</span></code> command provides useful summary statistics including TS/TV ratio, which is routinely used as a quality measure of variant calls. With dbSNP based annotation of novel and known variants, <code class="docutils literal notranslate"><span class="pre">bcftools</span></code> can compute TS/TV for novel and known variants at variant level, and at sample level. It should be noted that variant level TS/TV does not take sample genotype into consideration – it simply counts the TS and TV event for observed SNPs in the data. Other tools, such as <code class="docutils literal notranslate"><span class="pre">snpsift</span></code>, implements variant level TS/TV by counting TS and TV events in sample genotypes and compute the ratio after summing up TS and TV across all samples. See <a class="reference external" href="https://github.com/samtools/bcftools/issues/1526">here</a> some discussions on this issue. We provide these TS/TV calculations before and after QC but users should be aware of the difference when interpreting the results.</p>
</section>
</section>
<section id="default-parameters-resource-usage">
<h2>Default Parameters: Resource Usage<a class="headerlink" href="#default-parameters-resource-usage" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Memory:   Usually a whole genome VCF.gz file has the size of 200+GB, after testing, a minimum of 60GB of mem is requried.</p></li>
<li><p>Walltimes: For every hour qc_1 or qc_2 can process ~14G of data. The default is set to be 24h, corresponding to ~300GB of input. Please set the –walltime parameter according to the size of your input files.</p></li>
</ul>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>The target <code class="docutils literal notranslate"><span class="pre">vcf.gz</span></code> file</p>
<ul class="simple">
<li><p>If its chromosome name does not have the <code class="docutils literal notranslate"><span class="pre">chr</span></code> prefix and you need it to match with reference <code class="docutils literal notranslate"><span class="pre">fasta</span></code> file, please run <code class="docutils literal notranslate"><span class="pre">rename_chrs</span></code> workflow to add <code class="docutils literal notranslate"><span class="pre">chr</span></code>.</p></li>
<li><p>The vcf.gz file needs to be compressed by bgzip, instead of simple gzip</p></li>
<li><p>It should have a index file accompanying it. The index file can be generated by tabix</p></li>
<li><p>It must be a valid vcf.gz file that can pass bcftools sanity check: i.e. all tags are defined properly</p></li>
<li><p>It must contains following fields:</p>
<ol class="arabic simple">
<li><p>##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description=”Approximate read depth (reads with MQ=255 or with bad mates are filtered)”&gt;</p></li>
<li><p>##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description=”Genotype Quality”&gt;</p></li>
<li><p>##FORMAT=&lt;ID=AB,Number=1,Type=Float,Description=”Allele balance for each het genotype”&gt;</p></li>
</ol>
</li>
</ul>
</li>
<li><p>dbSNP database in <code class="docutils literal notranslate"><span class="pre">VCF</span></code> format</p></li>
<li><p>A reference sequence <code class="docutils literal notranslate"><span class="pre">fasta</span></code> file</p></li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>QC-ed genotype data in VCF format. You can use <code class="docutils literal notranslate"><span class="pre">vcf_to_plink</span></code> implemented in <code class="docutils literal notranslate"><span class="pre">genotype_formatting.ipynb</span></code> to further convert it to PLINK format.</p></li>
<li><p>A set of sumstats to help evaluate quality of genotype before and after QC</p>
<ul class="simple">
<li><p>Particularly useful is the TS/TV ratio</p></li>
</ul>
</li>
</ol>
</section>
<section id="minimal-working-example">
<h2>Minimal Working Example<a class="headerlink" href="#minimal-working-example" title="Link to this heading">#</a></h2>
<p>The MWE is generated via</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bcftools</span> <span class="n">query</span> <span class="o">-</span><span class="n">l</span> <span class="n">get</span><span class="o">-</span><span class="n">dosage</span><span class="o">.</span><span class="n">ALL</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">40</span> <span class="o">&gt;</span> <span class="n">MWE_sample_list</span>
<span class="n">bcftools</span> <span class="n">view</span> <span class="o">-</span><span class="n">S</span> <span class="n">MWE_sample_list</span>  <span class="n">get</span><span class="o">-</span><span class="n">dosage</span><span class="o">.</span><span class="n">ALL</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">&gt;</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">&amp;</span>
<span class="n">bgzip</span> <span class="o">-</span><span class="n">c</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">&gt;</span>  <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
<span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">vcf</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
<span class="n">bcftools</span> <span class="n">view</span> <span class="o">--</span><span class="n">regions</span> <span class="n">chr1</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">&gt;</span> <span class="n">chr1_sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">&amp;</span>
<span class="n">cat</span> <span class="n">chr1_sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">20000</span> <span class="o">&gt;</span> <span class="n">MWE_genotype</span><span class="o">.</span><span class="n">vcf</span>
</pre></div>
</div>
<p>and was stored here: <a class="reference external" href="https://drive.google.com/file/d/1sxxPdPIyKma0mAl8TKwhgyRHlOh0Oyrc/view?usp=sharing">https://drive.google.com/file/d/1sxxPdPIyKma0mAl8TKwhgyRHlOh0Oyrc/view?usp=sharing</a></p>
<p>FIXME: point this to the synapse folder.</p>
<section id="optional-rename-chromosomes">
<h3>(Optional) Rename Chromosomes<a class="headerlink" href="#optional-rename-chromosomes" title="Link to this heading">#</a></h3>
<p>Timing: 40 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run VCF_QC.ipynb rename_chrs \
    --genoFile /restricted/projectnb/xqtl/xqtl_protocol/reference_data/00-All.vcf.gz \
    --cwd /restricted/projectnb/xqtl/xqtl_protocol/reference_data
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INFO: Running [32mrename_chrs[0m: 
INFO: t0bb13a5cc2cae479 [32msubmitted[0m to neurology with job id Your job 8313589 (&quot;job_t0bb13a5cc2cae479&quot;) has been submitted
INFO: Waiting for the completion of [32m1[0m task.
INFO: [32mrename_chrs[0m output:   [32m/restricted/projectnb/xqtl/xqtl_protocol/reference_data/00-All.add_chr.vcf.gz[0m
INFO: Workflow rename_chrs (ID=w4f0deb3b5b4ae8ce) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</section>
<section id="optional-dbsnp-annotation">
<h3>(Optional) dbSNP Annotation<a class="headerlink" href="#optional-dbsnp-annotation" title="Link to this heading">#</a></h3>
<p>Timing: 4 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run VCF_QC.ipynb dbsnp_annotate \
    --genoFile /restricted/projectnb/xqtl/xqtl_protocol/reference_data/00-All.add_chr.vcf.gz \
    --cwd /restricted/projectnb/xqtl/xqtl_protocol/reference_data
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INFO: Running [32mdbsnp_annotate[0m: 
INFO: t6a0e22783b2666a3 [32msubmitted[0m to neurology with job id Your job 8314421 (&quot;job_t6a0e22783b2666a3&quot;) has been submitted
INFO: Waiting for the completion of [32m1[0m task.
INFO: [32mdbsnp_annotate[0m output:   [32m/restricted/projectnb/xqtl/xqtl_protocol/reference_data/00-All.add_chr.variants.gz[0m
INFO: Workflow dbsnp_annotate (ID=wf0ebf0af2d3d81ad) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</section>
<section id="i-quality-control">
<h3>i. Quality Control<a class="headerlink" href="#i-quality-control" title="Link to this heading">#</a></h3>
<p>Timing: X min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run VCF_QC.ipynb qc    \
    --genoFile data/MWE/MWE_genotype.vcf     \
    --dbsnp-variants data/reference_data/00-All.add_chr.variants.gz  \
    --reference-genome data/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta   \
    --cwd MWE/output/genotype_1
</pre></div>
</div>
</div>
</div>
<p>To run in parallel for all genotype data listed in <code class="docutils literal notranslate"><span class="pre">mwe_genotype_list</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run VCF_QC.ipynb qc    \
    --genoFile data/mwe/mwe_genotype_list    \
    --dbsnp-variants data/reference_data/00-All.add_chr.variants.gz  \
    --reference-genome data/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta   \
    --cwd MWE/output/genotype_4
</pre></div>
</div>
</div>
</div>
<p>Producing the following results:</p>
<ul class="simple">
<li><p>Total TS/TV for 19639 known variants before QC: 2.599</p></li>
<li><p>Total TS/TV for 19573 known variants after QC: 2.600</p></li>
<li><p>There is no novel variants included in the MWE.</p></li>
</ul>
<p>The Total TS/TV is extracted from the last step of QC. For known variant before QC:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>grep Ts/Tv MWE_genotype.leftnorm.known_variant.snipsift_tstv | rev | cut -d&#39;,&#39; -f1 | rev
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2.599
</pre></div>
</div>
</div>
</div>
<p>For known variant after QC:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>grep Ts/Tv MWE_genotype.leftnorm.filtered.*_variant.snipsift_tstv | rev | cut -d&#39;,&#39; -f1 | rev
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2.600
</pre></div>
</div>
</div>
</div>
<p>For novel variant before/after QC, TS/TV is not avaible since no novel_variants presented in the MWE</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>grep Ts/Tv MWE_genotype.leftnorm.novel_variant.snipsift_tstv | rev | cut -d&#39;,&#39; -f1 | rev
grep Ts/Tv MWE_genotype.leftnorm.filtered.novel_variant.snipsift_tstv | rev | cut -d&#39;,&#39; -f1 | rev
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run VCF_QC.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run VCF_QC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  rename_chrs
  dbsnp_annotate
  qc

Global Workflow Options:
  --genoFile  paths

                        input can either be 1 vcf genoFile, or a list of vcf
                        genoFile.
  --remove-samples . (as path)
                        The path to the file that contains the list of samples
                        to remove (format FID, IID)
  --keep-samples . (as path)
                        The path to the file that contains the list of samples
                        to keep (format FID, IID)
  --cwd output (as path)
                        Workdir
  --numThreads 1 (as int)
                        Number of threads
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 24h
                        Walltime
  --mem 60G
                        Usually a whole genome VCF.gz file has the size of
                        200+GB, after testing, a minimum of 60GB of mem is
                        requried.
  --container &#39;&#39;
                        Software container option
  --entrypoint (&#39;micromamba run -n&#39; + &#39; &#39; + container.split(&#39;/&#39;)[-1][:-4] + &quot; --no-capture-output&quot;) if container.endswith(&#39;.sif&#39;) else &quot;&quot;

  --[no-]add-chr (default to False)
                        use this function to edit memory string for PLINK input

Sections
  rename_chrs:
  dbsnp_annotate:
  qc_1:                 Handel multi-allelic sites, left normalization of indels
                        and add variant ID
    Workflow Options:
      --dbsnp-variants VAL (as path, required)
                        Path to dbSNP variants generated previously
      --reference-genome VAL (as path, required)
                        Path to fasta file for HG reference genome, eg
                        GRCh38_full_analysis_set_plus_decoy_hla.fa
      --[no-]bi-allelic (default to False)
      --[no-]snp-only (default to False)
  qc_2:                 genotype QC
    Workflow Options:
      --geno-filter 0.2 (as float)
                        Maximum missingess per-variant, default to 0.2
      --DP-snp 10 (as int)
                        Sample level QC - read depth (DP) to filter out SNPs
                        below this value Default to 10, with WES data in mind
                        But for WGS, setting it to 2 may be fine considering the
                        WGS may have low DP but the GQ filter should be good
                        enough
      --GQ 20 (as int)
                        Sample level QC - genotype quality (GQ) of specific
                        sample. This measure tells you how confident we are that
                        the genotype we assigned to a particular sample is
                        correct
      --DP-indel 10 (as int)
                        Sample level QC - read depth (DP) to filter out indels
                        below this value
      --AB-snp 0.15 (as float)
                        Allele balance for snps
      --AB-indel 0.2 (as float)
                        Allele balance for indels
      --hwe-filter 0.0 (as float)
                        HWE filter, default to 0.0 which means no HWE filter is
                        applied
  qc_3:
</pre></div>
</div>
</div>
</div>
</section>
<section id="global-parameters">
<h2>Global parameters<a class="headerlink" href="#global-parameters" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# input can either be 1 vcf genoFile, or a list of vcf genoFile.
parameter: genoFile = paths
# The path to the file that contains the list of samples to remove (format FID, IID)
parameter: remove_samples = path(&#39;.&#39;)
# The path to the file that contains the list of samples to keep (format FID, IID)
parameter: keep_samples = path(&#39;.&#39;)
# Workdir
parameter: cwd = path(&quot;output&quot;)
# Number of threads
parameter: numThreads = 1
# For cluster jobs, number commands to run per job
parameter: job_size = 1
# Walltime 
parameter: walltime = &#39;24h&#39;
# Usually a whole genome VCF.gz file has the size of 200+GB, after testing, a minimum of 60GB of mem is requried.
parameter: mem = &#39;60G&#39;
# Software container option
parameter: container = &quot;&quot;
import re
parameter: entrypoint=&quot;&quot;
#whether things like DP in the vcf header should be ignored when filtering.
parameter: skip_vcf_header_filtering = False
# use this function to edit memory string for PLINK input
from sos.utils import expand_size
cwd = path(f&quot;{cwd:a}&quot;)

import os
def get_genotype_file(geno_file_paths):
    #
    def valid_geno_file(x):
        suffixes = path(x).suffixes
        if suffixes[-1] == &#39;.bed&#39;:
            return True
        elif suffixes[-1] == &#39;.vcf&#39;:
            return True
        elif len(suffixes)&gt;1 and &#39;&#39;.join(suffixes[-2:]) == &quot;.vcf.gz&quot; :
            return True
        return False
    #
    def complete_geno_path(x, geno_file):
        if not valid_geno_file(x):
            raise ValueError(f&quot;Genotype file {x} should be VCF (end with .vcf.gz) or PLINK bed file (end with .bed)&quot;)
        if not os.path.isfile(x):
            # relative path
            if not os.path.isfile(f&#39;{geno_file:ad}/&#39; + x):
                raise ValueError(f&quot;Cannot find genotype file {x}&quot;)
            else:
                x = f&#39;{geno_file:ad}/&#39; + x
        return x
    # 
    def format_chrom(chrom):
        if chrom.startswith(&#39;chr&#39;):
            chrom = chrom[3:]
        return chrom
    # Inputs are either VCF or bed, or a vector of them 
    if len(geno_file_paths) &gt; 1:
        if all([valid_geno_file(x) for x in geno_file_paths]):
            return paths(geno_file_paths)
        else: 
            raise ValueError(f&quot;Invalid input {geno_file_paths}&quot;)
    # Input is one genotype file or text list of genotype files
    geno_file = geno_file_paths[0]
    if valid_geno_file(geno_file):
        return paths(geno_file)
    else: 
        units = [x.strip().split() for x in open(geno_file).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
        if all([len(x) == 1 for x in units]):
            return paths([complete_geno_path(x[0], geno_file) for x in units])
        elif all([len(x) == 2 for x in units]):
            genos = dict([(format_chrom(x[0]), path(complete_geno_path(x[1], geno_file))) for x in units])
        else:
            raise ValueError(f&quot;{geno_file} should contain one column of file names, or two columns of chrom number and corresponding file name&quot;)
        return genos
    
genoFile = get_genotype_file(genoFile)
</pre></div>
</div>
</div>
</div>
</section>
<section id="annotate-known-and-novel-variants">
<h2>Annotate known and novel variants<a class="headerlink" href="#annotate-known-and-novel-variants" title="Link to this heading">#</a></h2>
<p>You can download the known variant reference from <a class="reference external" href="https://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-All.vcf.gz">this link</a>.</p>
<p>For a detailed explanation of the procedure and its rationale, please refer to <a class="reference external" href="https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionVI/lessons/03_annotation-snpeff.html">this post</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rename_chrs: provides = &#39;{genoFile:nn}.add_chr.vcf.gz&#39;]
# This file can be downloaded from https://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-All.vcf.gz.
input: genoFile
output: f&#39;{_input:nn}.add_chr.vcf.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container = container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:nn}.stderr&#39;, stdout = f&#39;{_output:nn}.stdout&#39;, entrypoint=entrypoint
    for i in {1..22} X Y MT; do echo &quot;$i chr$i&quot;; done &gt; ${_output:nn}.chr_name_conv.txt
    bcftools annotate --rename-chrs ${_output:nn}.chr_name_conv.txt ${_input} -Oz -o ${_output}
    tabix -p vcf ${_output}
    rm -f ${_output:nn}.chr_name_conv.txt
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[dbsnp_annotate]
input: genoFile
output: f&quot;{cwd}/{_input:bnn}.variants.gz&quot;
task: trunk_workers = 1, trunk_size=job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container = container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    # Extract specific fields from the VCF file using bcftools
    bcftools query -f&#39;%CHROM\t%POS\t%ID\t%REF\t%ALT\n&#39; ${_input} | \
    awk &#39;BEGIN { 
            OFS=&quot;\t&quot; 
         } 
         {
            # Calculate end position based on the length of REF or ALT
            if (length($4) &gt; length($5)) {
                end_pos = $2 + (length($4) - 1)
            } else {
                end_pos = $2 + (length($5) - 1)
            }
            print $1, $2, end_pos, $3
         }&#39; | \
    # Compress the output using bgzip
    bgzip -c &gt; ${_output}
</pre></div>
</div>
</div>
</div>
</section>
<section id="genotype-qc">
<h2>Genotype QC<a class="headerlink" href="#genotype-qc" title="Link to this heading">#</a></h2>
<p>This step handles multi-allelic sites and annotate variants to known and novel. We add an RS ID to variants in dbSNP. Variants without rsID are considered novel variants. For every hour it can produce ~14Gb of data, please set the –walltime parameter according to the size of your input files.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Handel multi-allelic sites, left normalization of indels and add variant ID
[qc_1 (variant preprocessing)]
# Path to dbSNP variants generated previously
parameter: dbsnp_variants = path
# Path to fasta file for HG reference genome, eg GRCh38_full_analysis_set_plus_decoy_hla.fa
parameter: reference_genome = path
parameter: bi_allelic = False
parameter: snp_only = False
input: genoFile, group_by = 1
output: f&#39;{cwd}/{_input:bnn}.{&quot;leftnorm&quot; if not bi_allelic else &quot;biallelic&quot;}{&quot;.snp&quot; if snp_only else &quot;&quot;}.vcf.gz&#39;
task: trunk_workers = 1, trunk_size=job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container = container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    
    if [ &quot;${skip_vcf_header_filtering}&quot; = &quot;True&quot; ]; then
        printf &quot;not filtering for vcf header&quot; &gt;&gt; ${_output:n}.stdout
        # split multiallelic sites into biallelic records 
        ${&#39;bcftools norm -m-any&#39; if not bi_allelic else &#39;bcftools view -m2 -M2&#39;} ${&#39;-v snps&#39; if snp_only else &quot;&quot;} ${_input} |\
        # Fix incorrect or missing REF alleles and warn about them
        bcftools norm -d exact -N --check-ref ws -f ${reference_genome} --threads ${numThreads} |\
        # Remove ID and replace with CHROM:POS:REF:ALT format
        bcftools annotate -x ID -I +&#39;%CHROM:%POS:%REF:%ALT&#39; | \
        # Annotate with dbSNP rsID
        bcftools annotate -a ${dbsnp_variants}  -h &lt;(echo &#39;##INFO=&lt;ID=RSID,Number=1,Type=String,Description=&quot;dbSNP rsID&quot;&gt;&#39;) \
            -c CHROM,FROM,TO,INFO/RSID -Oz --threads ${numThreads} -o ${_output}

    else
        printf &quot;filtering for vcf header&quot; &gt;&gt; ${_output:n}.stdout
        # split multiallelic sites into biallelic records 
        ${&#39;bcftools norm -m-any&#39; if not bi_allelic else &#39;bcftools view -m2 -M2&#39;} ${&#39;-v snps&#39; if snp_only else &quot;&quot;} ${_input} |\
        # Fix incorrect or missing REF alleles and warn about them
        bcftools norm -d exact -N --check-ref ws -f ${reference_genome} --threads ${numThreads} |\
        # Fill missing tags
        bcftools +fill-tags -- -t all,F_MISSING,&#39;VD=sum(FMT/DP)&#39; | \
        # Remove ID and replace with CHROM:POS:REF:ALT format
        bcftools annotate -x ID -I +&#39;%CHROM:%POS:%REF:%ALT&#39; | \
        # Annotate with dbSNP rsID
        bcftools annotate -a ${dbsnp_variants}  -h &lt;(echo &#39;##INFO=&lt;ID=RSID,Number=1,Type=String,Description=&quot;dbSNP rsID&quot;&gt;&#39;) \
            -c CHROM,FROM,TO,INFO/RSID -Oz --threads ${numThreads} -o ${_output}

    fi
    
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint=entrypoint
    for i in ${_output} ; do 
        # Capture file metadata
        output_info=&quot;$i&quot;
        output_size=$(ls -lh &quot;$i&quot; | awk &#39;{print $5}&#39;)
        output_rows=$(zcat &quot;$i&quot; | wc -l)
        output_column=$(zcat &quot;$i&quot; | grep -v &quot;##&quot; | head -1 | wc -w)
        output_header_row=$(zcat &quot;$i&quot; | grep &quot;##&quot; | wc -l)
        output_preview=$(zcat &quot;$i&quot; | grep -v &quot;##&quot; | head | cut -f 1-11)

        # Write captured information to the stdout file
        printf &quot;output_info: %s\noutput_size: %s\noutput_rows: %d\noutput_column: %d\noutput_header_row: %d\noutput_preview:\n%s\n&quot; \
            &quot;$output_info&quot; &quot;$output_size&quot; &quot;$output_rows&quot; &quot;$output_column&quot; &quot;$output_header_row&quot; &quot;$output_preview&quot; &gt;&gt; ${_output:n}.stdout
    done
</pre></div>
</div>
</div>
</div>
<p>This step filter variants based on FILTER PASS, DP and QC, fraction of missing genotypes (all samples), and on HWE, for snps and indels. It will also remove monomorphic sites – using <code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">view</span> <span class="pre">-c1</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># genotype QC
[qc_2 (variant level QC)]
# Maximum missingess per-variant, default to 0.2
parameter: geno_filter = 0.2
# Sample level QC - read depth (DP) to filter out SNPs below this value
# Default to 10, with WES data in mind 
# But for WGS, setting it to 2 may be fine considering the WGS may have low DP but the GQ filter should be good enough
parameter: DP_snp = 10
# Sample level QC - genotype quality (GQ) of specific sample. This measure tells you how confident we are that the genotype we assigned to a particular sample is correct
parameter: GQ = 20
# Sample level QC - read depth (DP) to filter out indels below this value
parameter: DP_indel = 10
# Allele balance for snps
parameter: AB_snp = 0.15
# Allele balance for indels
parameter: AB_indel = 0.2
# HWE filter, default to 0.0 which means no HWE filter is applied
parameter: hwe_filter = 0.0
output: f&quot;{_input:nn}.bcftools_qc.vcf.gz&quot;
task: trunk_workers = 1, trunk_size=job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container = container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint

    if [ &quot;${skip_vcf_header_filtering}&quot; = &quot;True&quot; ]; then
        printf &quot;not filtering for vcf header&quot; &gt;&gt; ${_output:n}.stdout

        bcftools view -c1 ${_input} | \
        bcftools view -f PASS | \

        # Filter based on missingness and Hardy-Weinberg equilibrium 
        bcftools filter -i &#39;F_MISSING&lt;${geno_filter}&#39; \
         -Oz --threads ${numThreads} -o ${_output}
    else

        printf &quot;filtering for vcf header&quot; &gt;&gt; ${_output:n}.stdout
        # Initial filtering based on depth and genotype quality for SNPs and INDELs. Will set to missing genotypes that do not meet both conditions (DP&gt;=DP_spns &amp; GQ&gt;=GQ), similarly for indels
        bcftools filter -S . -i \
        &#39;(TYPE=&quot;SNP&quot; &amp; (FMT/DP)&gt;=${DP_snp} &amp; (FMT/GQ)&gt;=${GQ}) | 
         (TYPE=&quot;INDEL&quot; &amp; (FMT/DP)&gt;=${DP_indel} &amp; (FMT/GQ)&gt;=${GQ})&#39; ${_input} | \

        # Further filtering to retain only variants that are PASS and have at least one non-reference allele 
        bcftools view -c1 | \
        bcftools view -f PASS | \

        # Filter based on genotype (hom/het) and allelic balance for SNPs and INDELs
        bcftools filter -i \
        &#39;GT=&quot;hom&quot; | 
         TYPE=&quot;snp&quot; &amp; GT=&quot;het&quot; &amp; (FORMAT/AD[*:1])/(FORMAT/AD[*:0] + FORMAT/AD[*:1]) &gt;= ${AB_snp} | 
         TYPE=&quot;indel&quot; &amp; GT=&quot;het&quot; &amp; (FORMAT/AD[*:1])/(FORMAT/AD[*:0] + FORMAT/AD[*:1]) &gt;= ${AB_indel}&#39; | \

        # Filter based on missingness and Hardy-Weinberg equilibrium 
        bcftools filter -i &#39;F_MISSING&lt;${geno_filter} &amp; HWE&gt;${hwe_filter}&#39; \
        -Oz --threads ${numThreads} -o ${_output}
    fi

bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint=entrypoint
    for i in ${_output} ; do 
        # Capture file metadata
        output_info=&quot;$i&quot;
        output_size=$(ls -lh &quot;$i&quot; | awk &#39;{print $5}&#39;)
        output_rows=$(zcat &quot;$i&quot; | wc -l)
        output_column=$(zcat &quot;$i&quot; | grep -v &quot;##&quot; | head -1 | wc -w)
        output_header_row=$(zcat &quot;$i&quot; | grep &quot;##&quot; | wc -l)
        output_preview=$(zcat &quot;$i&quot; | grep -v &quot;##&quot; | head | cut -f 1-11)

        # Write captured information to the stdout file
        printf &quot;output_info: %s\noutput_size: %s\noutput_rows: %d\noutput_column: %d\noutput_header_row: %d\noutput_preview:\n%s\n&quot; \
            &quot;$output_info&quot; &quot;$output_size&quot; &quot;$output_rows&quot; &quot;$output_column&quot; &quot;$output_header_row&quot; &quot;$output_preview&quot; &gt;&gt; ${_output:n}.stdout
    done
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[qc_3 (genotype data summary statistics)]
input: output_from(&#39;qc_1&#39;), output_from(&#39;qc_2&#39;), group_by = 1
output: f&quot;{cwd}/{_input:bnn}.novel_variant_sumstats&quot;, 
        f&quot;{cwd}/{_input:bnn}.known_variant_sumstats&quot;, 
        f&quot;{cwd}/{_input:bnn}.novel_variant.snipsift_tstv&quot;,
        f&quot;{cwd}/{_input:bnn}.known_variant.snipsift_tstv&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: container = container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint

    # Compute statistics for novel variants (RSID is missing)
    bcftools stats -i &#39;RSID=&quot;.&quot;&#39; -v ${_input} &gt; ${_output[0]}

    # Compute statistics for known variants (RSID is present)
    bcftools stats -i &#39;RSID!=&quot;.&quot;&#39; -v ${_input} &gt; ${_output[1]}

    # Compute TS/TV for novel variants
    bcftools filter -i &#39;RSID=&quot;.&quot;&#39; ${_input} | \
        SnpSift tstv - &gt; ${_output[2]}

    # Compute TS/TV for known variants
    bcftools filter -i &#39;RSID!=&quot;.&quot;&#39; ${_input} | \
        SnpSift tstv - &gt; ${_output[3]}

bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    for i in ${_output} ; do 
        # Capture file metadata
        output_info=&quot;$i&quot;
        output_size=$(ls -lh &quot;$i&quot; | awk &#39;{print $5}&#39;)
        output_rows=$(zcat &quot;$i&quot; | wc -l)
        output_column=$(zcat &quot;$i&quot; | grep -v &quot;##&quot; | head -1 | wc -w)
        output_header_row=$(zcat &quot;$i&quot; | grep &quot;##&quot; | wc -l)
        output_preview=$(zcat &quot;$i&quot; | grep -v &quot;##&quot; | head | cut -f 1-11)

        # Write captured information to the stdout file
        printf &quot;output_info: %s\noutput_size: %s\noutput_rows: %d\noutput_column: %d\noutput_header_row: %d\noutput_preview:\n%s\n&quot; \
            &quot;$output_info&quot; &quot;$output_size&quot; &quot;$output_rows&quot; &quot;$output_column&quot; &quot;$output_header_row&quot; &quot;$output_preview&quot; &gt;&gt; ${_output[0]:n}.stdout
    done
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing/genotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../genotype_preprocessing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Genotype data preprocessing</p>
      </div>
    </a>
    <a class="right-next"
       href="GWAS_QC.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genotype PLINK File Quality Control</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-vcf-qc-filters">Default Parameters: VCF QC Filters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-note-on-ts-tv-summary-from-vcf-genotype-data">A note on TS/TV summary from VCF genotype data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-resource-usage">Default Parameters: Resource Usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-rename-chromosomes">(Optional) Rename Chromosomes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-dbsnp-annotation">(Optional) dbSNP Annotation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-quality-control">i. Quality Control</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#global-parameters">Global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotate-known-and-novel-variants">Annotate known and novel variants</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-qc">Genotype QC</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>