
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>Quantifying methylation data &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../bulk_expression.html">
   Methylation calling and QC
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../apa.html">
   APA calling
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/molecular_phenotypes/calling/methylation_calling.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/molecular_phenotypes/calling/methylation_calling.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/molecular_phenotypes/calling/methylation_calling.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods-overview">
   Methods overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup-and-global-parameters">
   Setup and global parameters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-1">
   Step 1:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-2">
   Step 2
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="quantifying-methylation-data">
<h1>Quantifying methylation data<a class="headerlink" href="#quantifying-methylation-data" title="Permalink to this headline">¶</a></h1>
<section id="methods-overview">
<h2>Methods overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">¶</a></h2>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">phenoFile</span></code>: The input of this module is a folder containing all the IDAT file, 1 pair for each sample. Directly under the folder, there should be one companion csv file that documenting all the meta-information of the bisulfite sequencing. Please specify the <code class="docutils literal notranslate"><span class="pre">phenoFile</span></code> as the path to this companion SampleSheet csv file</p></li>
<li><p>Optional: <code class="docutils literal notranslate"><span class="pre">cross_reactive</span></code>: A list of cpg probe that are reported to <a class="reference external" href="https://academic.oup.com/nargab/article/2/4/lqaa105/6040968">map to multiple regions in the genome.</a></p></li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>A pair bed.gz file for both beta and m value.</p>
</section>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/methylation_calling.ipynb methylation <span class="se">\</span>
    --phenoFile data/MWE/MWE_Sample_sheet.csv <span class="se">\</span>
    --container containers/methylation.sif 
</pre></div>
</div>
</div>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run methylation_calling.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run methylation_calling.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  methylation

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files.
  --phenoFile VAL (as path, required)
                        The companion sample sheet csv file as outlined in the
                        input section.
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option

Sections
  methylation_1:
    Workflow Options:
      --samples-pval-tre 0.05 (as float)
      --probe-det-pval-tre 0.01 (as float)
      --bisulfite-conversion-tre 85 (as int)
      --cross-reactive /opt/cross_reactive_probe_Hop2020.txt (as path)
                        Use the default list in our docker, if want to skip
                        methylation, specify it as &quot;.&quot;
      --hg 38 (as int)
  methylation_2:
</pre></div>
</div>
</div>
</div>
</section>
<section id="setup-and-global-parameters">
<h2>Setup and global parameters<a class="headerlink" href="#setup-and-global-parameters" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>global<span class="o">]</span>
<span class="c1"># The output directory for generated files.</span>
parameter: <span class="nv">cwd</span> <span class="o">=</span> path<span class="o">(</span><span class="s2">&quot;output&quot;</span><span class="o">)</span>
<span class="c1"># The companion sample sheet csv file as outlined in the input section.</span>
parameter: <span class="nv">phenoFile</span> <span class="o">=</span> path
<span class="c1"># For cluster jobs, number commands to run per job</span>
parameter: <span class="nv">job_size</span> <span class="o">=</span> <span class="m">1</span>
<span class="c1"># Wall clock time expected</span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
parameter: <span class="nv">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>

<span class="c1"># Number of threads</span>
parameter: <span class="nv">numThreads</span> <span class="o">=</span> <span class="m">8</span>
<span class="c1"># Software container option</span>
parameter: <span class="nv">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
from sos.utils import expand_size
<span class="nv">cwd</span> <span class="o">=</span> path<span class="o">(</span>f<span class="s1">&#39;{cwd:a}&#39;</span><span class="o">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-1">
<h2>Step 1:<a class="headerlink" href="#step-1" title="Permalink to this headline">¶</a></h2>
<p>The first step of methylation processing is to acquired the beta and M value after preliminary QC and filtering of the IDAT data.</p>
<p>By default, for epic data ,the data will be annotated based on hg38 using <a class="reference external" href="https://github.com/achilleasNP/IlluminaHumanMethylationEPICanno.ilm10b5.hg38">this annotation</a>, alternatively user can set the <code class="docutils literal notranslate"><span class="pre">--hg</span></code> parameter back to 19 to use the <a class="reference external" href="https://bioconductor.org/packages/release/data/annotation/html/IlluminaHumanMethylationEPICanno.ilm10b4.hg19.html">hg19 annotation</a>.</p>
<p>For 450K data however, only hg19 annotation is availble. And that is what we would use.</p>
<ol class="simple">
<li><p>Based on 1 input csv file, all the IDAT file in the folder and sub-folder will be loaded</p></li>
<li><p>The methylation data samples will first be filtered based on <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4527772/">bisulphite conversation rate</a>. This operation is done using the <a class="reference external" href="http://www.bioconductor.org/packages/release/bioc/vignettes/wateRmelon/inst/doc/wateRmelon.html#introduction">bscon function from watermelon package</a></p></li>
<li><p>samples will then be filtered based on a <a class="reference external" href="https://www.rdocumentation.org/packages/minfi/versions/1.18.4/topics/detectionP">detection pvalue</a>, which indicates the quality of the signal at each genomics position</p></li>
<li><p><a class="reference external" href="https://rdrr.io/bioc/minfi/man/preprocessQuantile.html">Stratified Quantile Normalization</a> will then be applied.</p></li>
<li><p>features will be filtered if they are on sex chr, known to be <a class="reference external" href="https://academic.oup.com/nargab/article/2/4/lqaa105/6040968">cross-reactive,maping to multiple regions in the genome.</a>, overlapping with snps, or having too low a detection P. The list of cross-reactive probe can be found as <code class="docutils literal notranslate"><span class="pre">/opt/cross_reactive_probe_Hop2020.txt</span></code> in our docker and <a class="reference external" href="https://raw.githubusercontent.com/hsun3163/xqtl-pipeline/main/data/cross_reactive_probe_Hop2020.txt">here</a>.</p></li>
<li><p>Beta and M value will for all the probes/samples will then each be saved to a indexed bed.gz file.</p></li>
</ol>
<p><a class="reference external" href="https://github.com/cumc/xqtl-pipeline/issues/312">As documented here</a> when the batch of IDAT data are different, there will be a problem reading the IDAT file without specifing the force = TRUE option in the <code class="docutils literal notranslate"><span class="pre">read.metharray.exp(targets</span> <span class="pre">=</span> <span class="pre">targets,force</span> <span class="pre">=</span> <span class="pre">TRUE)</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>methylation_1<span class="o">]</span>
<span class="c1"># treshold to filter out samples based on detection P value</span>
parameter: <span class="nv">samples_pval_tre</span> <span class="o">=</span> <span class="m">0</span>.05
<span class="c1"># treshold to filter out probes based on detection P value</span>
parameter: <span class="nv">probe_det_pval_tre</span> <span class="o">=</span> <span class="m">0</span>.01
ersion_rate.
parameter: <span class="nv">bisulfite_conversion_tre</span> <span class="o">=</span> <span class="m">85</span> <span class="c1"># 85 in# treshold to remove samples base on bisulfite_convdicates a conversion rate of 85% </span>
<span class="c1">## Use the default list in our docker, if want to skip methylation, specify it as &quot;.&quot;</span>
parameter: <span class="nv">cross_reactive_list</span> <span class="o">=</span> path<span class="o">(</span><span class="s2">&quot;/opt/cross_reactive_probe_Hop2020.txt&quot;</span><span class="o">)</span>
parameter: <span class="nv">hg</span> <span class="o">=</span> <span class="m">38</span> <span class="c1">#hg = 38 or 19 for epic data, by default 38. Noted for 450K data only hg19 is availble</span>
input: phenoFile
output: f<span class="s1">&#39;{cwd}/{_input:bn}.minfi.rds&#39;</span>,f<span class="s1">&#39;{cwd}/{_input:bn}.methyl.beta.bed&#39;</span>,f<span class="s1">&#39;{cwd}/{_input:bn}.methyl.M.bed&#39;</span>,f<span class="s1">&#39;{cwd}/{_input:bn}.methyl.M.region_list&#39;</span>  
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span> <span class="o">=</span> job_size, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads
R: <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>, <span class="nv">container</span><span class="o">=</span>container
    <span class="c1">## load libraries</span>
    library<span class="o">(</span>dplyr<span class="o">)</span>
    library<span class="o">(</span>readr<span class="o">)</span>
    library<span class="o">(</span>tibble<span class="o">)</span>
    library<span class="o">(</span>minfi<span class="o">)</span>
    sessionInfo<span class="o">()</span>
    <span class="nv">cross_reactive</span> <span class="o">=</span> readr::read_delim<span class="o">(</span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">cross_reactive_list</span><span class="si">}</span><span class="s2">&quot;</span>,<span class="s2">&quot;\t&quot;</span><span class="o">)</span><span class="nv">$probe</span>
    
    <span class="c1">## Define functions</span>
    bscon_minfi &lt;- <span class="k">function</span><span class="o">(</span>RGsetEx<span class="o">){</span>
        
        <span class="c1">#getting the values from the green channel</span>
        <span class="o">(</span>csp.green &lt;- <span class="k">function</span> <span class="o">(</span>RGsetEx, <span class="nv">controls</span> <span class="o">=</span> c<span class="o">(</span><span class="s2">&quot;BISULFITE CONVERSION I&quot;</span>, <span class="s2">&quot;BISULFITE CONVERSION II&quot;</span><span class="o">))</span>
        <span class="o">{</span>
          minfi:::.isRGOrStop<span class="o">(</span>RGsetEx<span class="o">)</span>
          r &lt;- getRed<span class="o">(</span>RGsetEx<span class="o">)</span>
          g &lt;- getGreen<span class="o">(</span>RGsetEx<span class="o">)</span>
          sapply <span class="o">(</span>controls, <span class="k">function</span><span class="o">(</span> controlType <span class="o">)</span> <span class="o">{</span>
        
           ctrlAddress &lt;- try <span class="o">(</span>getControlAddress<span class="o">(</span>RGsetEx, <span class="nv">controlType</span> <span class="o">=</span> controlType<span class="o">)</span>, <span class="nv">silent</span> <span class="o">=</span> T<span class="o">)</span>
           <span class="k">if</span> <span class="o">(</span>!inherits <span class="o">(</span>ctrlAddress, <span class="s1">&#39;try-error&#39;</span><span class="o">)){</span>ctrlAddress &lt;- getControlAddress<span class="o">(</span>RGsetEx, <span class="nv">controlType</span> <span class="o">=</span> controlType<span class="o">)}</span>
           <span class="k">else</span>
             stop <span class="o">(</span><span class="s2">&quot;450k QC data could not be found&quot;</span><span class="o">)</span>
        
        
           g<span class="o">[</span>ctrlAddress, <span class="o">]</span>
          <span class="o">})})</span>
        
        green &lt;- csp.green<span class="o">(</span>RGsetEx<span class="o">)</span>
        
        <span class="c1">#Getting values from the red channel</span>
        <span class="o">(</span>csp.red &lt;- <span class="k">function</span> <span class="o">(</span>RGsetEx, <span class="nv">controls</span> <span class="o">=</span> c<span class="o">(</span><span class="s2">&quot;BISULFITE CONVERSION I&quot;</span>, <span class="s2">&quot;BISULFITE CONVERSION II&quot;</span><span class="o">))</span>
        <span class="o">{</span>
          minfi:::.isRGOrStop<span class="o">(</span>RGsetEx<span class="o">)</span>
          r &lt;- getRed<span class="o">(</span>RGsetEx<span class="o">)</span>
          g &lt;- getGreen<span class="o">(</span>RGsetEx<span class="o">)</span>
          sapply <span class="o">(</span>controls, <span class="k">function</span><span class="o">(</span> controlType <span class="o">)</span> <span class="o">{</span>
        
            ctrlAddress &lt;- getControlAddress<span class="o">(</span>RGsetEx, <span class="nv">controlType</span> <span class="o">=</span> controlType<span class="o">)</span>
        
            r<span class="o">[</span>ctrlAddress, <span class="o">]</span>
          <span class="o">})})</span>
        
        red &lt;- csp.red <span class="o">(</span>RGsetEx<span class="o">)</span>
        
        <span class="c1">#selecting only the Bisulfite conversion I values from both green and red</span>
        bsI.green &lt;- green$<span class="sb">`</span>BISULFITE CONVERSION I<span class="sb">`</span>
        bsI.red &lt;- red$<span class="sb">`</span>BISULFITE CONVERSION I<span class="sb">`</span>
        <span class="c1">#selecting only the Bisulfite conversion II values from both green and red</span>
        bsII.green &lt;- green$<span class="sb">`</span>BISULFITE CONVERSION II<span class="sb">`</span>
        bsII.red &lt;- red$<span class="sb">`</span>BISULFITE CONVERSION II<span class="sb">`</span>
        
        <span class="c1"># calculate BS conv type I betas as an example of using an index vector</span>
        <span class="k">if</span><span class="o">(</span>nrow<span class="o">(</span>bsI.green<span class="o">)</span> &gt; <span class="m">11</span><span class="o">){</span> <span class="c1"># 450K</span>
          BSI.betas &lt;- rbind<span class="o">(</span>bsI.green<span class="o">[</span><span class="m">1</span>:3,<span class="o">]</span>, bsI.red<span class="o">[</span><span class="m">7</span>:9,<span class="o">])</span>/<span class="o">((</span>rbind<span class="o">(</span>bsI.green<span class="o">[</span><span class="m">1</span>:3,<span class="o">]</span>, bsI.red<span class="o">[</span><span class="m">7</span>:9,<span class="o">]))</span> + rbind<span class="o">(</span>bsI.green<span class="o">[</span><span class="m">4</span>:6,<span class="o">]</span>, bsI.red<span class="o">[</span><span class="m">10</span>:12,<span class="o">]))</span>
        <span class="o">}</span> <span class="k">else</span> <span class="o">{</span> <span class="c1"># EPIC</span>
          BSI.betas &lt;- rbind<span class="o">(</span>bsI.green<span class="o">[</span><span class="m">1</span>:2,<span class="o">]</span>, bsI.red<span class="o">[</span><span class="m">6</span>:7,<span class="o">])</span>/<span class="o">((</span>rbind<span class="o">(</span>bsI.green<span class="o">[</span><span class="m">1</span>:2,<span class="o">]</span>, bsI.red<span class="o">[</span><span class="m">6</span>:7,<span class="o">]))</span> + rbind<span class="o">(</span>bsI.green<span class="o">[</span><span class="m">3</span>:4,<span class="o">]</span>, bsI.red<span class="o">[</span> <span class="m">8</span>:9 ,<span class="o">]))</span>
        <span class="o">}</span>
        
        <span class="c1">#calculation of BS con in Type II data</span>
        BSII.betas &lt;- bsII.red/<span class="o">(</span>bsII.red + bsII.green<span class="o">)</span>
        
        apply<span class="o">(</span>rbind<span class="o">(</span>BSI.betas, BSII.betas<span class="o">)</span>, <span class="m">2</span>, median<span class="o">)</span>*100 <span class="c1">## this is the value you are interested in</span>
        <span class="o">}</span>

    <span class="c1">## 1. read idat files</span>
    targets &lt;- read.metharray.sheet<span class="o">(</span><span class="si">${</span><span class="nv">_input</span><span class="p">:</span><span class="nv">adr</span><span class="si">}</span><span class="o">)</span>
    colnames<span class="o">(</span>targets<span class="o">)[</span><span class="m">1</span><span class="o">]</span> <span class="o">=</span> <span class="s2">&quot;Sample_Name&quot;</span>
    <span class="c1">### Only read samples with data</span>
    <span class="nv">Missing_sample</span> <span class="o">=</span> targets%&gt;%filter<span class="o">(</span>!stringr::str_detect<span class="o">(</span>targets<span class="nv">$Basename</span> ,<span class="s2">&quot;/&quot;</span><span class="o">)</span> <span class="o">)</span>%&gt;%pull<span class="o">(</span>Sample_Name<span class="o">)</span> 
    <span class="nv">targets</span> <span class="o">=</span> targets%&gt;%filter<span class="o">(</span>stringr::str_detect<span class="o">(</span>targets<span class="nv">$Basename</span> ,<span class="s2">&quot;/&quot;</span><span class="o">)</span> <span class="o">)</span> 
    message<span class="o">(</span>paste0<span class="o">(</span><span class="s2">&quot;Following samples: &quot;</span>,paste0<span class="o">(</span>Missing_sample,collapse <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">)</span>, <span class="s2">&quot; don&#39;t have IDAT data&quot;</span> <span class="o">))</span>
    
    rgSet &lt;- read.metharray.exp<span class="o">(</span><span class="nv">targets</span> <span class="o">=</span> targets<span class="o">)</span>
    <span class="k">if</span><span class="o">(</span><span class="si">${</span><span class="nv">hg</span><span class="si">}</span> <span class="o">==</span> <span class="m">38</span> <span class="o">&amp;&amp;</span> rgSet@annotation<span class="o">[</span><span class="s2">&quot;array&quot;</span><span class="o">]</span> <span class="o">==</span> <span class="s1">&#39;IlluminaHumanMethylationEPIC&#39;</span> <span class="o">){</span>rgSet@annotation<span class="o">[</span><span class="s1">&#39;annotation&#39;</span><span class="o">]</span> <span class="o">=</span> <span class="s2">&quot;ilm10b5.hg38&quot;</span><span class="o">)}</span>
    message<span class="o">(</span><span class="s2">&quot;rgset created&quot;</span><span class="o">)</span>
    
    <span class="c1">###### Quality Control and Normalization ###############</span>
    <span class="c1">## 2. bisulphite conversation rate filtering</span>
    <span class="nv">rgSet_bcr</span> <span class="o">=</span> bscon_minfi<span class="o">(</span>rgSet<span class="o">)</span>
    <span class="nv">rgSet</span> <span class="o">=</span> rgSet<span class="o">[</span>,names<span class="o">(</span>which<span class="o">(</span>rgSet_bcr &gt; <span class="si">${</span><span class="nv">bisulfite_conversion_tre</span><span class="si">}</span> <span class="o">))]</span>
    
    <span class="c1">## 3. QC based on p-value, remove samples with average p value less than 0.05</span>
    
    detP &lt;- detectionP<span class="o">(</span>rgSet<span class="o">)</span>
    keep &lt;- colMeans<span class="o">(</span>detP<span class="o">)</span> &lt; <span class="si">${</span><span class="nv">samples_pval_tre</span><span class="si">}</span>
    rgSet &lt;- rgSet<span class="o">[</span>,keep<span class="o">]</span>
    targets &lt;- targets<span class="o">[</span>keep,<span class="o">]</span>
    message<span class="o">(</span><span class="s2">&quot;samples with avg det p-val &lt; </span><span class="si">${</span><span class="nv">samples_pval_tre</span><span class="si">}</span><span class="s2"> removed&quot;</span><span class="o">)</span>

    <span class="c1">## 4. Normalize the data - Quantile</span>
    mSetSq &lt;- preprocessQuantile<span class="o">(</span>rgSet<span class="o">)</span>
    message<span class="o">(</span><span class="s2">&quot;data quantile-normalized&quot;</span><span class="o">)</span>
    <span class="c1">## 5. Remove XY chr - probes</span>
    
    mSetSq &lt;- mSetSq<span class="o">[</span>!as.vector<span class="o">(</span>mSetSq@rowRanges@seqnames<span class="o">)</span>%in% c<span class="o">(</span><span class="s2">&quot;chrX&quot;</span>,<span class="s2">&quot;chrY&quot;</span><span class="o">)</span>,<span class="o">]</span>
    message<span class="o">(</span><span class="s2">&quot;sex probes removed&quot;</span><span class="o">)</span>
    
    <span class="c1">## 6. Remove cross-reactive probes</span>
    no_cross_reactive &lt;- !<span class="o">(</span>featureNames<span class="o">(</span>mSetSq<span class="o">)</span> %in% cross_reactive<span class="o">)</span>
    mSetSq &lt;- mSetSq<span class="o">[</span>no_cross_reactive, <span class="o">]</span>
    message<span class="o">(</span><span class="s2">&quot;cross-reactive probes removed&quot;</span><span class="o">)</span>
    
    <span class="c1">## 7. Drop probes that are also SNPs</span>
    
    mSetSq &lt;- dropLociWithSnps<span class="o">(</span>mSetSq<span class="o">)</span>
    message<span class="o">(</span><span class="s2">&quot;probes overlapping with snps removed&quot;</span><span class="o">)</span>
    
    <span class="c1">## 8. Remove probes with &lt; ${probe_det_pval_tre} detectin p-values</span>
    detP &lt;- detP<span class="o">[</span>match<span class="o">(</span>featureNames<span class="o">(</span>mSetSq<span class="o">)</span>,rownames<span class="o">(</span>detP<span class="o">))</span>,<span class="o">]</span>
    keep &lt;- rowSums<span class="o">(</span>detP &lt; <span class="si">${</span><span class="nv">probe_det_pval_tre</span><span class="si">}</span><span class="o">)</span> <span class="o">==</span> ncol<span class="o">(</span>mSetSq<span class="o">)</span>
    mSetSq &lt;- mSetSq<span class="o">[</span>keep,<span class="o">]</span>
    
    <span class="c1">## 9. get Beta and Mvalues</span>
    mSetSqbval &lt;- getBeta<span class="o">(</span>mSetSq<span class="o">)</span>%&gt;%as_tibble<span class="o">(</span><span class="nv">rownames</span> <span class="o">=</span> <span class="s2">&quot;ID&quot;</span><span class="o">)</span>
    mSetSqMval &lt;- getM<span class="o">(</span>mSetSq<span class="o">)</span>%&gt;%as_tibble<span class="o">(</span><span class="nv">rownames</span> <span class="o">=</span> <span class="s2">&quot;ID&quot;</span><span class="o">)</span>
    <span class="nv">cpg_regions</span> <span class="o">=</span> mSetSq@rowRanges%&gt;%as.data.frame<span class="o">()</span>%&gt;%as_tibble<span class="o">(</span><span class="nv">rownames</span> <span class="o">=</span> <span class="s2">&quot;ID&quot;</span><span class="o">)</span>%&gt;%select<span class="o">(</span><span class="s2">&quot;#chr&quot;</span> <span class="o">=</span> seqnames, start, end, ID, -width,-strand<span class="o">)</span>
    <span class="nv">mSetSqbval</span> <span class="o">=</span> cpg_regions%&gt;%right_join<span class="o">(</span>mSetSqbval, <span class="nv">by</span> <span class="o">=</span> <span class="s2">&quot;ID&quot;</span><span class="o">)</span>%&gt;%mutate<span class="o">(</span><span class="nv">end</span> <span class="o">=</span> start + <span class="m">1</span><span class="o">)</span>
    <span class="nv">mSetSqMval</span> <span class="o">=</span> cpg_regions%&gt;%right_join<span class="o">(</span>mSetSqMval, <span class="nv">by</span> <span class="o">=</span> <span class="s2">&quot;ID&quot;</span><span class="o">)</span>%&gt;%mutate<span class="o">(</span><span class="nv">end</span> <span class="o">=</span> start + <span class="m">1</span><span class="o">)</span>
    message<span class="o">(</span><span class="s2">&quot;Beta-values and M-values obtained&quot;</span><span class="o">)</span>
    <span class="c1">## 10. output data</span>
    <span class="nv">mSetSqMval</span> <span class="o">=</span> mSetSqMval%&gt;%rename_at<span class="o">(</span>vars<span class="o">(</span>rgSet@colData%&gt;%rownames<span class="o">())</span>, <span class="k">function</span><span class="o">(</span>x<span class="o">)</span> rgSet@colData<span class="o">[</span>x,<span class="o">]</span>%&gt;%as_tibble%&gt;%pull<span class="o">(</span>Sample_Name<span class="o">)</span> <span class="o">)</span>
    <span class="nv">mSetSqbval</span> <span class="o">=</span> mSetSqbval%&gt;%rename_at<span class="o">(</span>vars<span class="o">(</span>rgSet@colData%&gt;%rownames<span class="o">())</span>, <span class="k">function</span><span class="o">(</span>x<span class="o">)</span> rgSet@colData<span class="o">[</span>x,<span class="o">]</span>%&gt;%as_tibble%&gt;%pull<span class="o">(</span>Sample_Name<span class="o">)</span> <span class="o">)</span>
    mSetSqMval<span class="o">[</span>,1:4<span class="o">]</span>%&gt;%write_delim<span class="o">(</span><span class="si">${</span><span class="nv">_output</span><span class="p">[3]:</span><span class="nv">r</span><span class="si">}</span>,<span class="s2">&quot;\t&quot;</span><span class="o">)</span>
    mSetSqbval%&gt;%readr::write_delim<span class="o">(</span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">_output</span><span class="p">[1]</span><span class="si">}</span><span class="s2">&quot;</span>,<span class="s2">&quot;\t&quot;</span><span class="o">)</span>
    mSetSqMval%&gt;%readr::write_delim<span class="o">(</span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">_output</span><span class="p">[2]</span><span class="si">}</span><span class="s2">&quot;</span>,<span class="s2">&quot;\t&quot;</span><span class="o">)</span>
    <span class="nv">output</span> <span class="o">=</span> list<span class="o">(</span><span class="s2">&quot;rgSet&quot;</span> <span class="o">=</span> rgSet,mSetSq <span class="o">=</span> <span class="s2">&quot;mSetSq&quot;</span>, <span class="nv">mSetSqbval</span> <span class="o">=</span> <span class="s2">&quot;mSetSqbval&quot;</span>, <span class="nv">mSetSqMval</span> <span class="o">=</span> <span class="s2">&quot;mSetSqMval&quot;</span><span class="o">)</span>
    output%&gt;%saveRDS<span class="o">(</span><span class="si">${</span><span class="nv">_output</span><span class="p">[0]:</span><span class="nv">r</span><span class="si">}</span><span class="o">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-2">
<h2>Step 2<a class="headerlink" href="#step-2" title="Permalink to this headline">¶</a></h2>
<p>After the beta and m value are generated. The output will be saved as a bed file</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>methylation_2<span class="o">]</span>
output: f<span class="s1">&#39;{_input[0]:nn}.methyl.beta.bed.gz&#39;</span>,f<span class="s1">&#39;{_input[0]:nn}.methyl.M.bed.gz&#39;</span> 
bash: <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>, <span class="nv">container</span><span class="o">=</span>container
    bgzip -f <span class="si">${</span><span class="nv">_input</span><span class="p">[1]</span><span class="si">}</span> 
    tabix <span class="si">${</span><span class="nv">_output</span><span class="p">[0]</span><span class="si">}</span>
    bgzip -f <span class="si">${</span><span class="nv">_input</span><span class="p">[2]</span><span class="si">}</span> 
    tabix <span class="si">${</span><span class="nv">_output</span><span class="p">[1]</span><span class="si">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">#[combat_batch]</span>
parameter: <span class="nv">batchFile</span> <span class="o">=</span> path<span class="o">(</span><span class="s2">&quot;.&quot;</span><span class="o">)</span>
input: output_from<span class="o">(</span><span class="s2">&quot;methylation&quot;</span><span class="o">)</span>
output:f<span class="s1">&#39;{_input}.MDS.pdf&#39;</span>,f<span class="s1">&#39;{_input}.beta.batch_corrected.txt&#39;</span>
R: <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>, <span class="nv">container</span><span class="o">=</span>container
    library<span class="o">(</span>ggplot2<span class="o">)</span>
    library<span class="o">(</span>metaMA<span class="o">)</span>
    library<span class="o">(</span>sva<span class="o">)</span>
    library<span class="o">(</span>lme4<span class="o">)</span>
    <span class="c1">## Define functions</span>
    posibatches &lt;- <span class="k">function</span><span class="o">(</span>dat, Sentrix, <span class="nv">batch</span><span class="o">=</span>TRUE, par.prior<span class="o">=</span>TRUE, prior.plots<span class="o">=</span>FALSE, mean.only.posi<span class="o">=</span>FALSE, mean.only.batch<span class="o">=</span>FALSE<span class="o">)</span> <span class="o">{</span>
      require<span class="o">(</span>sva<span class="o">)</span>
      require<span class="o">(</span>lme4<span class="o">)</span>
      <span class="c1">#get the position and batch information----------------------------------------</span>
      <span class="c1">### Extraction of the position numbers and chip numbers--------------------</span>
      <span class="c1">#if (is.null(SentrixVector)){</span>
      <span class="k">if</span> <span class="o">(</span>is.null<span class="o">(</span>Sentrix<span class="o">)){</span>
        stop<span class="o">(</span><span class="s1">&#39;Sentrix informations must be provided.&#39;</span><span class="o">)</span>
      <span class="o">}</span>
      <span class="c1">#chips&lt;-as.numeric(factor(substr(Sentrix$sampleNames, 1, 10))) </span>
      <span class="c1">#positions&lt;-as.numeric(factor(substr(Sentrix$sampleNames, 12, 17))) </span>
      
      chips &lt;- as.numeric<span class="o">(</span>factor<span class="o">(</span>sapply<span class="o">(</span>strsplit<span class="o">(</span>as.character<span class="o">(</span>Sentrix<span class="nv">$sampleNames</span><span class="o">)</span>, <span class="s2">&quot;_&quot;</span><span class="o">)</span>, <span class="k">function</span><span class="o">(</span>x<span class="o">)</span> x<span class="o">[</span><span class="m">1</span><span class="o">])))</span>
      positions &lt;- as.numeric<span class="o">(</span>factor<span class="o">(</span>sapply<span class="o">(</span>strsplit<span class="o">(</span>as.character<span class="o">(</span>Sentrix<span class="nv">$sampleNames</span><span class="o">)</span>, <span class="s2">&quot;_&quot;</span><span class="o">)</span>, <span class="k">function</span><span class="o">(</span>x<span class="o">)</span> x<span class="o">[</span><span class="m">2</span><span class="o">])))</span>
      
      <span class="k">if</span> <span class="o">(</span>length<span class="o">(</span>positions<span class="o">)</span>!<span class="o">=</span>length<span class="o">(</span>chips<span class="o">)){</span>
        stop<span class="o">(</span><span class="s1">&#39;positions and chips must have the same length&#39;</span><span class="o">)</span>
      <span class="o">}</span>
      <span class="k">if</span> <span class="o">(</span>sum<span class="o">(</span>positions&gt;12<span class="o">)</span>&gt;0<span class="o">){</span>
        stop<span class="o">(</span><span class="s1">&#39;Position number cannot be greater than 12&#39;</span><span class="o">)</span>
      <span class="o">}</span>
      <span class="k">if</span> <span class="o">(</span>sum<span class="o">(</span>is.na<span class="o">(</span>chips<span class="o">))</span>&gt;1 <span class="o">||</span> sum<span class="o">(</span>is.na<span class="o">(</span>positions<span class="o">))</span>&gt;1<span class="o">){</span>
        stop<span class="o">(</span><span class="s1">&#39;One or more position or chip numbers missing&#39;</span><span class="o">)</span>
      <span class="o">}</span>
      
      <span class="k">if</span> <span class="o">(</span>sum<span class="o">(</span>is.na<span class="o">(</span>Sentrix<span class="nv">$batches</span><span class="o">))==</span>length<span class="o">(</span>chips<span class="o">)){</span>
        batches&lt;-chips
      <span class="o">}</span> <span class="k">else</span> <span class="o">{</span>
        batches&lt;-as.factor<span class="o">(</span>Sentrix<span class="nv">$batches</span><span class="o">)</span>
      <span class="o">}</span>
      
      <span class="k">if</span><span class="o">(</span><span class="nv">batch</span><span class="o">==</span>TRUE<span class="o">){</span>
        
        <span class="c1">################################################################################</span>
        <span class="c1">################################################################################</span>
        <span class="nv">pct_threshold</span> <span class="o">=</span> .8 <span class="c1"># Amount of variability desired to be explained by the principal components.  Set to match the results in book chapter and SAS code.  User can adjust this to a higher (&gt;= 0.8) number but &lt; 1.0</span>
        dataRowN &lt;- nrow<span class="o">(</span>dat<span class="o">)</span>
        dataColN &lt;- ncol<span class="o">(</span>dat<span class="o">)</span>
        
        <span class="c1">########## Center the data (center rows) ##########</span>
        datCentered &lt;- matrix<span class="o">(</span><span class="nv">data</span> <span class="o">=</span> <span class="m">0</span>, <span class="nv">nrow</span> <span class="o">=</span> dataRowN, <span class="nv">ncol</span> <span class="o">=</span> dataColN<span class="o">)</span>
        <span class="nv">datCentered_transposed</span> <span class="o">=</span> apply<span class="o">(</span>dat, <span class="m">1</span>, scale, <span class="nv">center</span> <span class="o">=</span> TRUE, <span class="nv">scale</span> <span class="o">=</span> FALSE<span class="o">)</span>
        <span class="nv">datCentered</span> <span class="o">=</span> t<span class="o">(</span>datCentered_transposed<span class="o">)</span>
        
        exp_design&lt;-data.frame<span class="o">(</span>cbind<span class="o">(</span>positions,batches<span class="o">))</span>
        expDesignRowN &lt;- nrow<span class="o">(</span>exp_design<span class="o">)</span>
        expDesignColN &lt;- ncol<span class="o">(</span>exp_design<span class="o">)</span>
        myColNames &lt;- names<span class="o">(</span>exp_design<span class="o">)</span>
        
        
        <span class="c1">########## Compute correlation matrix ##########</span>
        
        theDataCor &lt;- cor<span class="o">(</span>datCentered<span class="o">)</span>
        
        <span class="c1">########## Obtain eigenvalues ##########</span>
        
        eigenData &lt;- eigen<span class="o">(</span>theDataCor<span class="o">)</span>
        <span class="nv">eigenValues</span> <span class="o">=</span> eigenData<span class="nv">$values</span>
        ev_n &lt;- length<span class="o">(</span>eigenValues<span class="o">)</span>
        <span class="nv">eigenVectorsMatrix</span> <span class="o">=</span> eigenData<span class="nv">$vectors</span>
        <span class="nv">eigenValuesSum</span> <span class="o">=</span> sum<span class="o">(</span>eigenValues<span class="o">)</span>
        <span class="nv">percents_PCs</span> <span class="o">=</span> eigenValues /eigenValuesSum
        
        <span class="c1">########## Merge experimental file and eigenvectors for n components ##########</span>
        
        <span class="nv">my_counter_2</span> <span class="o">=</span> <span class="m">0</span>
        <span class="nv">my_sum_2</span> <span class="o">=</span> <span class="m">1</span>
        <span class="k">for</span> <span class="o">(</span>i <span class="k">in</span> ev_n:1<span class="o">){</span>
          <span class="nv">my_sum_2</span>  <span class="o">=</span> my_sum_2 - percents_PCs<span class="o">[</span>i<span class="o">]</span>
          <span class="k">if</span> <span class="o">((</span>my_sum_2<span class="o">)</span> &lt;<span class="o">=</span> pct_threshold <span class="o">){</span>
            <span class="nv">my_counter_2</span> <span class="o">=</span> my_counter_2 + <span class="m">1</span>
          <span class="o">}</span>
          
        <span class="o">}</span>
        <span class="k">if</span> <span class="o">(</span>my_counter_2 &lt; <span class="m">3</span><span class="o">){</span>
          <span class="nv">pc_n</span>  <span class="o">=</span> <span class="m">3</span>
          
        <span class="o">}</span><span class="k">else</span> <span class="o">{</span>
          <span class="nv">pc_n</span> <span class="o">=</span> my_counter_2
        <span class="o">}</span>
        
        <span class="c1"># pc_n is the number of principal components to model</span>
        
        pc_data_matrix &lt;- matrix<span class="o">(</span><span class="nv">data</span> <span class="o">=</span> <span class="m">0</span>, <span class="nv">nrow</span> <span class="o">=</span> <span class="o">(</span>expDesignRowN*pc_n<span class="o">)</span>, <span class="nv">ncol</span> <span class="o">=</span> <span class="m">1</span><span class="o">)</span>
        <span class="nv">mycounter</span> <span class="o">=</span> <span class="m">0</span>
        <span class="k">for</span> <span class="o">(</span>i <span class="k">in</span> <span class="m">1</span>:pc_n<span class="o">){</span>
          <span class="k">for</span> <span class="o">(</span>j <span class="k">in</span> <span class="m">1</span>:expDesignRowN<span class="o">){</span>
            mycounter &lt;- mycounter + <span class="m">1</span>
            pc_data_matrix<span class="o">[</span>mycounter,1<span class="o">]</span> <span class="o">=</span> eigenVectorsMatrix<span class="o">[</span>j,i<span class="o">]</span>
            
          <span class="o">}</span>
        <span class="o">}</span>
        
        AAA &lt;- exp_design<span class="o">[</span>rep<span class="o">(</span><span class="m">1</span>:expDesignRowN,pc_n<span class="o">)</span>,<span class="o">]</span>
        
        Data &lt;- cbind<span class="o">(</span>AAA,pc_data_matrix<span class="o">)</span>
        
        <span class="c1">####### Edit these variables according to your factors #######</span>
        
        variables &lt;- c<span class="o">(</span>colnames<span class="o">(</span>exp_design<span class="o">))</span>
        <span class="k">for</span> <span class="o">(</span>i <span class="k">in</span> <span class="m">1</span>:length<span class="o">(</span>variables<span class="o">))</span> <span class="o">{</span>
          Data<span class="nv">$variables</span><span class="o">[</span>i<span class="o">]</span> &lt;- as.factor<span class="o">(</span>Data<span class="nv">$variables</span><span class="o">[</span>i<span class="o">])</span>
        <span class="o">}</span>
        
        <span class="c1">########## Mixed linear model ##########</span>
        
        p &lt;- options<span class="o">(</span><span class="nv">warn</span> <span class="o">=</span> <span class="o">(</span>-1<span class="o">))</span>
        <span class="c1">#effects_n = expDesignColN + choose(expDesignColN, 2) + 1</span>
        <span class="nv">effects_n</span> <span class="o">=</span> expDesignColN  + <span class="m">1</span>
        randomEffectsMatrix &lt;- matrix<span class="o">(</span><span class="nv">data</span> <span class="o">=</span> <span class="m">0</span>, <span class="nv">nrow</span> <span class="o">=</span> pc_n, <span class="nv">ncol</span> <span class="o">=</span> effects_n<span class="o">)</span>
        
        model.func &lt;- c<span class="o">()</span>
        index &lt;- <span class="m">1</span>
        <span class="k">for</span> <span class="o">(</span>i <span class="k">in</span> <span class="m">1</span>:length<span class="o">(</span>variables<span class="o">))</span> <span class="o">{</span>
          <span class="nv">mod</span> <span class="o">=</span> paste<span class="o">(</span><span class="s2">&quot;(1|&quot;</span>, variables<span class="o">[</span>i<span class="o">]</span>, <span class="s2">&quot;)&quot;</span>, <span class="nv">sep</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">)</span>
          model.func<span class="o">[</span>index<span class="o">]</span> <span class="o">=</span> mod
          <span class="nv">index</span> <span class="o">=</span> index + <span class="m">1</span>
        <span class="o">}</span>
        
        <span class="k">function</span>.mods &lt;- paste<span class="o">(</span>model.func, <span class="nv">collapse</span> <span class="o">=</span> <span class="s2">&quot; + &quot;</span><span class="o">)</span>
        
        <span class="k">for</span> <span class="o">(</span>i <span class="k">in</span> <span class="m">1</span>:pc_n<span class="o">)</span> <span class="o">{</span>
          <span class="nv">y</span> <span class="o">=</span> <span class="o">(((</span>i - <span class="m">1</span><span class="o">)</span> * expDesignRowN<span class="o">)</span> + <span class="m">1</span><span class="o">)</span>
          funct &lt;- paste<span class="o">(</span><span class="s2">&quot;pc_data_matrix&quot;</span>, <span class="k">function</span>.mods, <span class="nv">sep</span> <span class="o">=</span> <span class="s2">&quot; ~ &quot;</span><span class="o">)</span>
          Rm1ML &lt;- lmer<span class="o">(</span>funct, Data<span class="o">[</span>y:<span class="o">(((</span>i - <span class="m">1</span><span class="o">)</span> * expDesignRowN<span class="o">)</span> +
                                         expDesignRowN<span class="o">)</span>, <span class="o">]</span>, <span class="nv">REML</span> <span class="o">=</span> TRUE, <span class="nv">control</span><span class="o">=</span>lmerControl<span class="o">(</span>check.nobs.vs.nlev <span class="o">=</span> <span class="s2">&quot;ignore&quot;</span>,check.nobs.vs.rankZ <span class="o">=</span> <span class="s2">&quot;ignore&quot;</span>,check.nobs.vs.nRE<span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="o">)</span>,verbose <span class="o">=</span> FALSE,
                        na.action <span class="o">=</span> na.omit<span class="o">)</span>
          randomEffects &lt;- Rm1ML
          randomEffectsMatrix<span class="o">[</span>i, <span class="o">]</span> &lt;- c<span class="o">(</span>unlist<span class="o">(</span>VarCorr<span class="o">(</span>Rm1ML<span class="o">))</span>,
                                        <span class="nv">resid</span> <span class="o">=</span> sigma<span class="o">(</span>Rm1ML<span class="o">)</span>^2<span class="o">)</span>
        <span class="o">}</span>
        effectsNames &lt;- c<span class="o">(</span>names<span class="o">(</span>getME<span class="o">(</span>Rm1ML, <span class="s2">&quot;cnms&quot;</span><span class="o">))</span>, <span class="s2">&quot;resid&quot;</span><span class="o">)</span>
        <span class="c1">########## Standardize Variance ##########</span>
        
        randomEffectsMatrixStdze &lt;- matrix<span class="o">(</span><span class="nv">data</span> <span class="o">=</span> <span class="m">0</span>, <span class="nv">nrow</span> <span class="o">=</span> pc_n, <span class="nv">ncol</span> <span class="o">=</span> effects_n<span class="o">)</span>
        <span class="k">for</span> <span class="o">(</span>i <span class="k">in</span> <span class="m">1</span>:pc_n<span class="o">){</span>
          <span class="nv">mySum</span> <span class="o">=</span> sum<span class="o">(</span>randomEffectsMatrix<span class="o">[</span>i,<span class="o">])</span>
          <span class="k">for</span> <span class="o">(</span>j <span class="k">in</span> <span class="m">1</span>:effects_n<span class="o">){</span>
            randomEffectsMatrixStdze<span class="o">[</span>i,j<span class="o">]</span> <span class="o">=</span> randomEffectsMatrix<span class="o">[</span>i,j<span class="o">]</span>/mySum
          <span class="o">}</span>
        <span class="o">}</span>
        
        <span class="c1">########## Compute Weighted Proportions ##########</span>
        
        randomEffectsMatrixWtProp &lt;- matrix<span class="o">(</span><span class="nv">data</span> <span class="o">=</span> <span class="m">0</span>, <span class="nv">nrow</span> <span class="o">=</span> pc_n, <span class="nv">ncol</span> <span class="o">=</span> effects_n<span class="o">)</span>
        <span class="k">for</span> <span class="o">(</span>i <span class="k">in</span> <span class="m">1</span>:pc_n<span class="o">){</span>
          <span class="nv">weight</span> <span class="o">=</span> eigenValues<span class="o">[</span>i<span class="o">]</span>/eigenValuesSum
          <span class="k">for</span> <span class="o">(</span>j <span class="k">in</span> <span class="m">1</span>:effects_n<span class="o">){</span>
            randomEffectsMatrixWtProp<span class="o">[</span>i,j<span class="o">]</span> <span class="o">=</span> randomEffectsMatrixStdze<span class="o">[</span>i,j<span class="o">]</span>*weight
          <span class="o">}</span>
        <span class="o">}</span>
        <span class="c1">######### Compute Weighted Ave Proportions ##########</span>
        
        randomEffectsSums &lt;- matrix<span class="o">(</span><span class="nv">data</span> <span class="o">=</span> <span class="m">0</span>, <span class="nv">nrow</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">ncol</span> <span class="o">=</span> effects_n<span class="o">)</span>
        randomEffectsSums &lt;-colSums<span class="o">(</span>randomEffectsMatrixWtProp<span class="o">)</span>
        <span class="nv">totalSum</span> <span class="o">=</span> sum<span class="o">(</span>randomEffectsSums<span class="o">)</span>
        randomEffectsMatrixWtAveProp &lt;- matrix<span class="o">(</span><span class="nv">data</span> <span class="o">=</span> <span class="m">0</span>, <span class="nv">nrow</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">ncol</span> <span class="o">=</span> effects_n<span class="o">)</span>
        
        <span class="k">for</span> <span class="o">(</span>j <span class="k">in</span> <span class="m">1</span>:effects_n<span class="o">){</span>
          randomEffectsMatrixWtAveProp<span class="o">[</span>j<span class="o">]</span> <span class="o">=</span> randomEffectsSums<span class="o">[</span>j<span class="o">]</span>/totalSum
          
        <span class="o">}</span>
        
        <span class="k">if</span><span class="o">(</span>randomEffectsMatrixWtAveProp<span class="o">[</span>,1<span class="o">]</span>&lt;randomEffectsMatrixWtAveProp<span class="o">[</span>,2<span class="o">]){</span>
          afterbatchExp&lt;-ComBat<span class="o">(</span><span class="nv">dat</span> <span class="o">=</span> dat, <span class="nv">batch</span> <span class="o">=</span> batches, par.prior <span class="o">=</span> par.prior, prior.plots <span class="o">=</span> prior.plots, mean.only <span class="o">=</span> mean.only.batch<span class="o">)</span>
          afterposiExp&lt;-ComBat<span class="o">(</span><span class="nv">dat</span> <span class="o">=</span> afterbatchExp, <span class="nv">batch</span> <span class="o">=</span> positions, par.prior <span class="o">=</span> par.prior, prior.plots <span class="o">=</span> prior.plots, mean.only <span class="o">=</span> mean.only.posi<span class="o">)</span>
          dat&lt;-afterposiExp
        <span class="o">}</span> <span class="k">else</span><span class="o">{</span>
          afterposiExp&lt;-ComBat<span class="o">(</span><span class="nv">dat</span> <span class="o">=</span> dat, <span class="nv">batch</span> <span class="o">=</span> positions, par.prior <span class="o">=</span> par.prior, prior.plots <span class="o">=</span> prior.plots, mean.only <span class="o">=</span> mean.only.posi<span class="o">)</span>
          afterbatchExp&lt;-ComBat<span class="o">(</span><span class="nv">dat</span> <span class="o">=</span> afterposiExp, <span class="nv">batch</span> <span class="o">=</span> batches, par.prior <span class="o">=</span> par.prior, prior.plots <span class="o">=</span> prior.plots, mean.only <span class="o">=</span> mean.only.batch<span class="o">)</span>
          dat&lt;-afterbatchExp
        <span class="o">}</span>
      <span class="o">}</span><span class="k">else</span><span class="o">{</span>
        afterposiExp&lt;-ComBat<span class="o">(</span><span class="nv">dat</span> <span class="o">=</span> dat, <span class="nv">batch</span> <span class="o">=</span> positions, par.prior <span class="o">=</span> par.prior, prior.plots <span class="o">=</span> prior.plots, mean.only <span class="o">=</span> mean.only.posi<span class="o">)</span>
        dat&lt;-afterposiExp
      <span class="o">}</span>
      <span class="k">return</span><span class="o">(</span>dat<span class="o">)</span>
    <span class="o">}</span>
    <span class="c1">## Combat batch effect</span>
    <span class="nv">mSetSq</span> <span class="o">=</span> readRDS<span class="o">(</span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">_input</span><span class="p">[0]</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">)</span><span class="nv">$mSetSq</span>
    <span class="nv">Sentrix</span> <span class="o">=</span> read.csv<span class="o">(</span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">batchFile</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">)</span>
    results &lt;- posibatches<span class="o">(</span>mSetSq, Sentrix, <span class="nv">batch</span><span class="o">=</span>TRUE, par.prior<span class="o">=</span>TRUE, prior.plots<span class="o">=</span>FALSE, mean.only.posi<span class="o">=</span>FALSE, mean.only.batch<span class="o">=</span>FALSE<span class="o">)</span>
    results%&gt;%readr::write_delim<span class="o">(</span><span class="si">${</span><span class="nv">_output</span><span class="p">[1]:</span><span class="nv">r</span><span class="si">}</span><span class="o">)</span>
    <span class="c1">## Plot MDS</span>
    data4mds &lt;- <span class="k">function</span><span class="o">(</span>data, topvals<span class="o">){</span>
      b &lt;- data
      o &lt;- order<span class="o">(</span>rowVars<span class="o">(</span>b<span class="o">)</span>, <span class="nv">decreasing</span> <span class="o">=</span> TRUE<span class="o">)[</span>seq_len<span class="o">(</span>topvals<span class="o">)]</span>
      d &lt;- dist<span class="o">(</span>t<span class="o">(</span>b<span class="o">[</span>o, <span class="o">]))</span>
      fit &lt;- cmdscale<span class="o">(</span>d<span class="o">)</span>
      <span class="k">return</span><span class="o">(</span>fit<span class="o">)</span>
    <span class="o">}</span>
    
    myfit &lt;- as.data.frame<span class="o">(</span>data4mds<span class="o">(</span>mSetSq, <span class="m">1000</span><span class="o">))</span>
    colnames<span class="o">(</span>myfit<span class="o">)</span> &lt;- c<span class="o">(</span><span class="s2">&quot;fit1&quot;</span>, <span class="s2">&quot;fit2&quot;</span><span class="o">)</span>

    <span class="nv">MDSplot1</span> <span class="o">=</span> ggplot<span class="o">(</span>myfit, aes<span class="o">(</span><span class="nv">x</span> <span class="o">=</span> fit1, <span class="nv">y</span> <span class="o">=</span> fit2<span class="o">))</span> + geom_point<span class="o">(</span><span class="nv">size</span> <span class="o">=</span> <span class="m">1</span>.5<span class="o">)</span> + 
      labs<span class="o">(</span><span class="nv">title</span> <span class="o">=</span> <span class="s2">&quot;MDS plot&quot;</span>,  <span class="nv">x</span> <span class="o">=</span> <span class="s2">&quot;Coordinate 1&quot;</span>,  <span class="nv">y</span> <span class="o">=</span> <span class="s2">&quot;Coordinate 2&quot;</span><span class="o">)</span> + 
      theme<span class="o">(</span>axis.title.x <span class="o">=</span> element_text<span class="o">(</span><span class="nv">face</span> <span class="o">=</span> <span class="s2">&quot;bold&quot;</span>, <span class="nv">size</span> <span class="o">=</span> <span class="m">26</span><span class="o">)</span>, axis.text.x <span class="o">=</span> element_text<span class="o">(</span><span class="nv">face</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span>, <span class="nv">size</span> <span class="o">=</span> <span class="m">20</span><span class="o">))</span> + theme<span class="o">(</span>axis.text.y <span class="o">=</span> element_text<span class="o">(</span><span class="nv">face</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span>, <span class="nv">size</span> <span class="o">=</span> <span class="m">20</span><span class="o">)</span>, axis.title.y <span class="o">=</span> element_text<span class="o">(</span><span class="nv">face</span> <span class="o">=</span> <span class="s2">&quot;bold&quot;</span>, <span class="nv">size</span> <span class="o">=</span> <span class="m">26</span><span class="o">))</span> +  
      theme<span class="o">(</span>panel.background <span class="o">=</span> element_rect<span class="o">(</span><span class="nv">fill</span> <span class="o">=</span> <span class="s2">&quot;white&quot;</span>, <span class="nv">colour</span> <span class="o">=</span> <span class="s2">&quot;white&quot;</span><span class="o">))</span> + 
      theme<span class="o">(</span>panel.border <span class="o">=</span> element_blank<span class="o">()</span>, panel.grid.major <span class="o">=</span> element_blank<span class="o">()</span>, panel.grid.minor <span class="o">=</span> element_blank<span class="o">()</span>, axis.line <span class="o">=</span> element_line<span class="o">(</span><span class="nv">size</span> <span class="o">=</span> <span class="m">0</span>.5, <span class="nv">linetype</span> <span class="o">=</span> <span class="s2">&quot;solid&quot;</span>, <span class="nv">colour</span> <span class="o">=</span> <span class="s2">&quot;black&quot;</span><span class="o">))</span> +  
      theme<span class="o">(</span>plot.title <span class="o">=</span> element_text<span class="o">(</span><span class="nv">face</span> <span class="o">=</span> <span class="s2">&quot;bold&quot;</span>, <span class="nv">size</span> <span class="o">=</span> <span class="m">30</span>, <span class="nv">hjust</span> <span class="o">=</span> <span class="m">0</span>.5<span class="o">))</span> + 
      theme<span class="o">(</span>legend.position<span class="o">=</span><span class="s2">&quot;bottom&quot;</span>, legend.text<span class="o">=</span>element_text<span class="o">(</span><span class="nv">size</span><span class="o">=</span><span class="m">16</span><span class="o">))</span>
      MDSplot1%&gt;%ggsave<span class="o">(</span><span class="si">${</span><span class="nv">_output</span><span class="p">[0]:</span><span class="nv">r</span><span class="si">}</span><span class="o">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "bash"
        },
        kernelOptions: {
            kernelName: "bash",
            path: "./code/molecular_phenotypes/calling"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'bash'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>