
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Quantifying expression from RNA-seq data &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Sample level RNA-seq quality control" href="../QC/bulk_expression_QC.html" />
    <link rel="prev" title="RNA-seq expression" href="../bulk_expression.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../bulk_expression.html">
   RNA-seq expression
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     Data overview
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/molecular_phenotypes/calling/RNA_calling.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/molecular_phenotypes/calling/RNA_calling.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/molecular_phenotypes/calling/RNA_calling.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods-overview">
   Methods overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gene-level-quantifications">
     Gene-level quantifications
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#exon-level-quantifications">
     Exon-level quantifications
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#paired-end">
     Paired end:
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#single-end">
     Single end:
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#strand-option">
   <code class="docutils literal notranslate">
    <span class="pre">
     strand
    </span>
   </code>
   option
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#read-length-option">
   <code class="docutils literal notranslate">
    <span class="pre">
     read_length
    </span>
   </code>
   option
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup-and-global-parameters">
   Setup and global parameters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-0-qc-before-alignment">
   Step 0: QC before alignment
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-inputs">
     Step Inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-outputs">
     Step Outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-1-remove-adaptor-through-fastp">
   Step 1: Remove adaptor through
   <code class="docutils literal notranslate">
    <span class="pre">
     fastp
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Step Inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Step Outputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-options">
     Step options
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-1-alternative-remove-adaptor-through-trimmomatic">
   Step 1 Alternative: Remove adaptor through
   <code class="docutils literal notranslate">
    <span class="pre">
     Trimmomatic
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Step Inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id4">
     Step Outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-2-alignment-through-star">
   Step 2: Alignment through
   <code class="docutils literal notranslate">
    <span class="pre">
     STAR
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id5">
     Step Inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id6">
     Step Outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-3-mark-duplicates-reads-qc-through-picard">
   Step 3: Mark duplicates reads &amp; QC through
   <code class="docutils literal notranslate">
    <span class="pre">
     Picard
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id7">
     Step Inputs:
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id8">
     Step Outputs:
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-4-post-aligment-qc-through-rna-seqc">
   Step 4: Post aligment QC through
   <code class="docutils literal notranslate">
    <span class="pre">
     RNA-SeQC
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id9">
     Step Inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id10">
     Step Outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-6-quantify-expression-through-rsem">
   Step 6: Quantify expression through
   <code class="docutils literal notranslate">
    <span class="pre">
     RSEM
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-input">
     Step Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id11">
     Step Outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-7-summarize-with-multiqc">
   Step 7: summarize with MultiQC
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="quantifying-expression-from-rna-seq-data">
<h1>Quantifying expression from RNA-seq data<a class="headerlink" href="#quantifying-expression-from-rna-seq-data" title="Permalink to this headline">¶</a></h1>
<p>This pipeline aims to call expression levels from RNA-seq data starting from <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> sequences. It implements the GTEx pipeline for GTEx / TOPMed project. Please refer to <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md">this page</a> for detail.</p>
<p><strong>Various reference data needs to be prepared before using this workflow</strong>. <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/reference_data.html">Here we provide a module</a> to download and prepare the reference data.</p>
<div class="section" id="methods-overview">
<h2>Methods overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">../../</span><span class="n">images</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">jpeg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ../../images/rna_quantification.jpeg (495.9 KiB):</div></div><img alt="../../../_images/RNA_calling_2_1.png" src="../../../_images/RNA_calling_2_1.png" />
</div>
</div>
<p>The procedure is decribed also in <a class="reference external" href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">GTEx portal</a>. Here is a recap of some important details.</p>
<div class="section" id="gene-level-quantifications">
<h3>Gene-level quantifications<a class="headerlink" href="#gene-level-quantifications" title="Permalink to this headline">¶</a></h3>
<p>read counts and TPM values were produced with RNA-SeQC v2.4.2 (DeLuca et al., Bioinformatics, 2012 ), using the following read-level filters:</p>
<ul class="simple">
<li><p>Reads were uniquely mapped (corresponding to a mapping quality of 255 for START BAMs).</p></li>
<li><p>Reads were aligned in proper pairs.</p></li>
<li><p>The read alignment distance was &lt;=6 (i.e., alignments must not contain more than six non-reference bases).</p></li>
<li><p>Reads were fully contained within exon boundaries. Reads overlapping introns were not counted.</p></li>
</ul>
<p>These filters were applied using the “-strictMode” flag in RNA-SeQC. The TPM values have not been normalized or corrected for any covariates.</p>
</div>
<div class="section" id="exon-level-quantifications">
<h3>Exon-level quantifications<a class="headerlink" href="#exon-level-quantifications" title="Permalink to this headline">¶</a></h3>
<p>for exon-level read counts, if a read overlapped multiple exons, then a fractional value equal to the portion of the read contained within that exon was allotted. Transcript-level quantifications were calculated using RSEM v1.3.0.</p>
</div>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<div class="section" id="paired-end">
<h3>Paired end:<a class="headerlink" href="#paired-end" title="Permalink to this headline">¶</a></h3>
<p>A meta-data file, tab delimited with header, containing 3 required columns: ID, fq1, fq2 and 2 optional columns: strand and read_length:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ID</span> <span class="n">fq1</span> <span class="n">fq2</span> <span class="n">strand</span> <span class="n">read_length</span>
<span class="n">sample_1</span> <span class="n">samp1_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">samp1_r2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rf</span> <span class="mi">100</span>
<span class="n">sample_2</span> <span class="n">samp2_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">samp2_r2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">fr</span> <span class="mi">150</span>
<span class="n">sample_3</span> <span class="n">samp3_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">samp3_r2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">strand_missing</span> <span class="mi">75</span>
</pre></div>
</div>
</div>
<div class="section" id="single-end">
<h3>Single end:<a class="headerlink" href="#single-end" title="Permalink to this headline">¶</a></h3>
<p>A similar meta-data file, tab delimited with header, containing 2 required columns: ID, fq1 and 2 optional columns: strand and read_length:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ID</span> <span class="n">fq1</span> <span class="n">fq2</span> <span class="n">strand</span> <span class="n">read_length</span>
<span class="n">sample_1</span> <span class="n">samp1_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rf</span> <span class="mi">100</span>
<span class="n">sample_2</span> <span class="n">samp2_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">fr</span> <span class="mi">150</span>
<span class="n">sample_3</span> <span class="n">samp3_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">strand_missing</span> <span class="mi">75</span>
</pre></div>
</div>
<p>The interface for single end and paired end data are the same. Only difference lied in whether there is a fq2 column in the meta data.</p>
<p>All the fastq files should be available under specified folder (default assumes the same folder as where the meta-data file is).</p>
</div>
</div>
<div class="section" id="strand-option">
<h2><code class="docutils literal notranslate"><span class="pre">strand</span></code> option<a class="headerlink" href="#strand-option" title="Permalink to this headline">¶</a></h2>
<p>Some steps requires you specify strand option via <code class="docutils literal notranslate"><span class="pre">--strand</span></code> and should be one of<code class="docutils literal notranslate"><span class="pre">rf</span></code>, <code class="docutils literal notranslate"><span class="pre">fr</span></code>,<code class="docutils literal notranslate"><span class="pre">unstranded</span></code> , or <code class="docutils literal notranslate"><span class="pre">strand_missing</span></code>. Please refer to <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-022-04572-7">Signal et al (2022)</a> for an explanation what these library types mean. <strong>If you are not sure about the strandedness of your data, our strand detection steps will be automatically ran to detect the strand of input fastq file levaging the <a class="reference external" href="https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf">gene count table output from STAR</a> . The definition of strandness are following <a class="reference external" href="https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/">this page</a>. The percentage cutoff are based on <a class="reference external" href="https://github.com/signalbash/how_are_we_stranded_here">how are we stranded here</a>.</strong></p>
</div>
<div class="section" id="read-length-option">
<h2><code class="docutils literal notranslate"><span class="pre">read_length</span></code> option<a class="headerlink" href="#read-length-option" title="Permalink to this headline">¶</a></h2>
<p>The the read length for each sample should be specified in the read_length column in the input meta-data file (see input section). If you don’t know the read length for a sample, enter 0 and the pipeline will use 100 as the default value. You can change this default value by using the <code class="docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> parameter. If none of the samples have read length information, please left out the read_length column and the pipeline will use the default one for each sample.</p>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>TPM count for gene level RNA expression via <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code>, and isoform level RNA expression via <code class="docutils literal notranslate"><span class="pre">RSEM</span></code>.</p>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>A toy <code class="docutils literal notranslate"><span class="pre">fastq</span></code> data can be found on <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn36417253">synapse</a>.</p>
<p>To generate <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> report,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fastqc</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> 
</pre></div>
</div>
</div>
</div>
<p>To cut adaptor, after cutting the adaptor, a new sample list will be generated with the trimmed fastq file. Please noted that there is no adaptors to be trimmed in the test data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fastp_trim_adaptor</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span> 
</pre></div>
</div>
</div>
</div>
<p>To align the reads with STAR and generate the bam_list recipe for downstream molecular phenotype count matrixes. The <code class="docutils literal notranslate"><span class="pre">-J</span> <span class="pre">4</span> <span class="pre">-c</span> <span class="pre">csg.yml</span> <span class="pre">-q</span> <span class="pre">csg</span></code> part is crucial for it ask for the required memory to conduct the STAR alignment. The STAR command can only be ran with more than 40G of memory or with the counter-part of <code class="docutils literal notranslate"><span class="pre">-J</span> <span class="pre">4</span> <span class="pre">-c</span> <span class="pre">csg.yml</span> <span class="pre">-q</span> <span class="pre">csg</span></code></p>
<p><em><strong>Warning</strong></em>: It is of crucial importance that, the STAR Alignment shall be ran with the gtf file that were generated <code class="docutils literal notranslate"><span class="pre">before</span> <span class="pre">collapsing</span> <span class="pre">to</span> <span class="pre">gene</span></code>, i.e. the one that are used to generated RSEM index. Other wise there will be an <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/issues/390">error while running RSEM later on</a></p>
<p><em><strong>Warning</strong></em>: It is also crucial that, the <code class="docutils literal notranslate"><span class="pre">STAR_align</span></code>,<code class="docutils literal notranslate"><span class="pre">picard_qc</span></code>,<code class="docutils literal notranslate"><span class="pre">strand_detect</span></code> are not to be ran saperately, but instead they need to be run together by running <code class="docutils literal notranslate"><span class="pre">STAR_output</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">STAR_output</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">4</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">uncompressed</span> <span class="o">&amp;</span> 
</pre></div>
</div>
</div>
</div>
<p>To call gene-level RNA expression using <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code>.</p>
<p><em><strong>Warning</strong></em>: Unlike STAR_output and rsem, rnaseqc required the gtf <code class="docutils literal notranslate"><span class="pre">after</span> <span class="pre">collapsing</span> <span class="pre">to</span> <span class="pre">gene</span> </code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rnaseqc_call</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">collapse_only</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> \
    <span class="o">--</span><span class="n">bam_list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">finalizing</span><span class="o">/</span><span class="n">output_test</span><span class="o">/</span><span class="n">test_data_bam_list</span> 
</pre></div>
</div>
</div>
</div>
<p>To call transcript level RNA expression using <code class="docutils literal notranslate"><span class="pre">RSEM</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rsem_call</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">data</span><span class="o">/</span><span class="n">sample_fastq</span><span class="o">.</span><span class="n">list</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">data</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">RSEM</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">RSEM_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> \
    <span class="o">--</span><span class="n">bam_list</span> <span class="n">output3</span><span class="o">/</span><span class="n">sample_fastq_bam_list</span> <span class="o">-</span><span class="n">J</span> <span class="mi">3</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rsem_call</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">collapse_only</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> \
    <span class="o">--</span><span class="n">bam_list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">finalizing</span><span class="o">/</span><span class="n">output_test</span><span class="o">/</span><span class="n">test_data_bam_list</span> \
    <span class="o">--</span><span class="n">RSEM</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">RSEM_Index</span><span class="o">/</span> 
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run RNA_calling.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  fastqc
  fastp_trim_adaptor
  trimmomatic
  STAR_align
  rnaseqc_call
  rsem_call
  picard_qc

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files.
  --samples VAL (as path, required)
                        Sample meta data list
  --data-dir  path(f&quot;{samples:d}&quot;)

                        Raw data directory, default to the same directory as
                        sample list
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --java-mem 6G
                        Memory for Java virtual mechine (`picard`)
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option
  --gtf VAL (as path, required)
                        Reference gene model

Sections
  fastqc:
  fastp_trim_adaptor:
    Workflow Options:
      --window-size 4 (as int)
                        sliding window setting
      --required-quality 20 (as int)
      --leading 20 (as int)
                        the mean quality requirement option for cut_front
      --trailing 20 (as int)
                        the mean quality requirement option for cut_tail
      --min-len 15 (as int)
                        reads shorter than length_required will be discarded
  trimmomatic:
    Workflow Options:
      --trimmomatic-jar &#39;/opt/Trimmomatic-0.39/trimmomatic-0.39.jar&#39;
                        Path to the software. Default set to using our
                        rna_quantification.sif image
      --fasta-with-adapters-etc VAL (as str, required)
                        Illumina clip setting Path to the reference adaptors
      --seed-mismatches 2 (as int)
      --palindrome-clip-threshold 30 (as int)
      --simple-clip-threshold 10 (as int)
      --window-size 4 (as int)
                        sliding window setting
      --required-quality 20 (as int)
      --leading 3 (as int)
                        Other settings
      --trailing 3 (as int)
      --min-len 50 (as int)
  STAR_align:
    Workflow Options:
      --STAR-index VAL (as path, required)
                        STAR indexing file
      --outFilterMultimapNmax 20 (as int)
                        Alignment parameter
      --alignSJoverhangMin 8 (as int)
      --alignSJDBoverhangMin 1 (as int)
      --outFilterMismatchNmax 999 (as int)
      --outFilterMismatchNoverLmax 0.1 (as float)
      --alignIntronMin 20 (as int)
      --alignIntronMax 1000000 (as int)
      --alignMatesGapMax 1000000 (as int)
      --outFilterType BySJout
      --outFilterScoreMinOverLread 0.33 (as float)
      --outFilterMatchNminOverLread 0.33 (as float)
      --limitSjdbInsertNsj 1200000 (as int)
      --outSAMstrandField intronMotif
      --outFilterIntronMotifs None
      --alignSoftClipAtReferenceEnds Yes
      --quantMode TranscriptomeSAM GeneCounts (as list)
      --outSAMattrRGline ID:rg1 SM:sm1 (as list)
      --outSAMattributes NH HI AS nM NM ch (as list)
      --chimSegmentMin 15 (as int)
      --chimJunctionOverhangMin 15 (as int)
      --chimOutType Junctions WithinBAM SoftClip (as list)
      --chimMainSegmentMultNmax 1 (as int)
  rnaseqc_call_1, rsem_call_1, picard_qc:
    Workflow Options:
      --ref-flat . (as path)
                        Path to flat reference file, for computing QC metric
      --picard-jar &#39;/opt/picard-tools/picard.jar&#39;
                        Path to the software. Default set to using our
                        rna_quantification.sif image
      --reference-fasta VAL (as path, required)
                        The fasta reference file used to generate star index
      --optical-distance 100 (as int)
                        For the patterned flowcell models (HiSeq X), change to
                        2500
  rnaseqc_call_2:
    Workflow Options:
      --strand VAL (as str, required)
                        Options are fr or rf or unstranded
  rnaseqc_call_3:
  rsem_call_2:
    Workflow Options:
      --RSEM-index VAL (as path, required)
      --max-frag-len 1000 (as int)
      --strand VAL (as str, required)
                        Options are fr or rf or unstranded
  rsem_call_3:
  rsem_call_4, rnaseqc_call_4:
  rsem_call_5, rnaseqc_call_5:
    Workflow Options:
      --ref-flat . (as path)
                        Path to flat reference file, for computing QC metric
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="setup-and-global-parameters">
<h2>Setup and global parameters<a class="headerlink" href="#setup-and-global-parameters" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># The output directory for generated files.</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># Sample meta data list</span>
<span class="kn">parameter:</span> <span class="n">samples</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Raw data directory, default to the same directory as sample list</span>
<span class="kn">parameter:</span> <span class="n">data_dir</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Memory for Java virtual mechine (`picard`)</span>
<span class="kn">parameter:</span> <span class="n">java_mem</span> <span class="o">=</span> <span class="s2">&quot;6G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span><span class="p">{(</span><span class="s1">&#39;micromamba run -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">])</span> <span class="k">if</span> <span class="n">container</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.sif&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;&#39;</span><span class="p">}</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1">## Whether the fasta/fastq file is compressed or not.</span>
<span class="kn">parameter:</span> <span class="n">uncompressed</span> <span class="o">=</span> <span class="kc">False</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1">## FIX: The way to get sample needs to be revamped to 1. Accomodate rf/fr as column 2. accomodate single end read (Only 2 fq/samples)</span>
<span class="n">sample_inv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1">## Extract strand information if user have specified the strand</span>
<span class="n">strand_inv</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">if</span> <span class="s2">&quot;strand&quot;</span> <span class="ow">in</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">strand_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">strand</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">sample_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">stop_if</span><span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;fr&quot;</span><span class="p">,</span> <span class="s2">&quot;rf&quot;</span><span class="p">,</span> <span class="s2">&quot;unstranded&quot;</span><span class="p">,</span><span class="s2">&quot;strand_missing&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">strand_inv</span> <span class="p">]),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;strand columns should only include ``fr``, ``rf``, ``strand_missing`` or ``unstranded``&quot;</span><span class="p">)</span>
            
<span class="c1">## Extract read_length if user have specified read_length</span>
<span class="n">read_length</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_inv</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="k">if</span> <span class="s2">&quot;read_length&quot;</span> <span class="ow">in</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">read_length</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">read_length</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">sample_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;read_length&quot;</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    
    
<span class="c1">## Extract sample_id</span>
<span class="n">sample_inv_list</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">sample_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sample_inv_list</span><span class="p">]</span>


    
<span class="c1">## Get the file name for single/paired end data</span>
<span class="n">file_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sample_inv_list</span><span class="p">]</span>
<span class="n">file_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">file_inv</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>

<span class="n">fastq</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">file_inv</span><span class="p">]</span>


<span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">fastq</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">y</span><span class="si">}</span><span class="s2"> does not exist&quot;</span><span class="p">)</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fastq</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">fastq</span><span class="p">)):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Duplicated files are found (but should not be allowed) in fastq file list&quot;</span><span class="p">)</span>

<span class="c1"># Is the RNA-seq data pair-end</span>
<span class="n">is_paired_end</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fastq</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_id</span><span class="p">)</span> <span class="k">else</span> <span class="mi">1</span> 
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="kp">env</span>
<span class="kp">env</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Input samples are </span><span class="si">{</span><span class="s2">&quot;paired-end&quot;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="s2">&quot;single-end&quot;</span><span class="si">}</span><span class="s1"> sequences.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="step-0-qc-before-alignment">
<h2>Step 0: QC before alignment<a class="headerlink" href="#step-0-qc-before-alignment" title="Permalink to this headline">¶</a></h2>
<p>This step utilize <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> and will generate two QC report in <code class="docutils literal notranslate"><span class="pre">html</span></code> format. It should be noted that the <a class="reference external" href="https://www.biostars.org/p/190584/">paired_end reads will also be qc separately</a></p>
<div class="section" id="step-inputs">
<h3>Step Inputs<a class="headerlink" href="#step-inputs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastq1</span></code> and <code class="docutils literal notranslate"><span class="pre">fastq2</span></code>: paths to original <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file.</p></li>
</ul>
</div>
<div class="section" id="step-outputs">
<h3>Step Outputs<a class="headerlink" href="#step-outputs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Two <code class="docutils literal notranslate"><span class="pre">html</span></code> file for QC report</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">fastqc</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">fastq</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span>  <span class="mi">1</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">_fastqc.html&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">_fastqc.zip&#39;</span> 
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">fastqc</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-1-remove-adaptor-through-fastp">
<h2>Step 1: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">fastp</span></code><a class="headerlink" href="#step-1-remove-adaptor-through-fastp" title="Permalink to this headline">¶</a></h2>
<p>Documentation: <a class="reference external" href="https://github.com/OpenGene/fastp">fastp</a></p>
<p>We use <code class="docutils literal notranslate"><span class="pre">fastp</span></code> in place of the <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> for fastp’s ability to detect adapter from the reads. It was a c++ command line tool published in <a class="reference external" href="https://academic.oup.com/bioinformatics/article/34/17/i884/5093234">Sept 2018</a>. It  will use the following algorithm to detect the adaptors:</p>
<blockquote>
<div><p>The adapter-sequence detection algorithm is based on two assumptions: the first is that only one adapter exists in the data; the second is that adapter sequences exist only in the read tails. These two assumptions are valid for major next-generation sequencers like Illumina HiSeq series, NextSeq series and NovaSeq series. We compute the k-mer (k = 10) of first N reads (N = 1 M). From this k-mer, the sequences with high occurrence frequencies (&gt;0.0001) are considered as adapter seeds. Low-complexity sequences are removed because they are usually caused by sequencing artifacts. The adapter seeds are sorted by its occurrence frequencies. A tree-based algorithm is applied to extend the adapter seeds to find the real complete adapter</p>
</div></blockquote>
<p>It was demostrated that fastp can remove all the adaptor automatically and completely faster than Trimmomatic and cutadapt</p>
<div class="section" id="id1">
<h3>Step Inputs<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastq</span></code>: 1 set of fq.gz file for each sample. (2 files if paired end, 1 file if single end )</p></li>
</ul>
</div>
<div class="section" id="id2">
<h3>Step Outputs<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>1 set of  <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file for alignment. (2 files if paired end, 1 file if single end )</p></li>
<li><p>The unpaired reads (i.e.where a read survived, but the partner read did not.) will be discarded by default as those were not used in the following steps. This feature can be added were it was needed.</p></li>
<li><p>1 set of html documenting the quality of input reads(1 html file and 1 json file)</p></li>
</ul>
</div>
<div class="section" id="step-options">
<h3>Step options<a class="headerlink" href="#step-options" title="Permalink to this headline">¶</a></h3>
<p>A few options were selected to be customizable so that fastp step can have the same level of flexiblity as the Trimmomatic step had. They are:</p>
<ul class="simple">
<li><p>min_len: length_required, reads shorter than this will be discarded, default is 15. (int [=15])</p></li>
<li><p>window_size: cut_window_size, the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])</p></li>
<li><p>leading/trailing : cut_front_mean_quality/cut_tail_mean_quality, the mean quality requirement option for cut_front/cut_tail, which move a sliding window from front/tail, drop the bases in the window if its mean quality &lt; threshold. <strong>Notice the choice of quality score in <code class="docutils literal notranslate"><span class="pre">fastp</span></code> (N=20) is a lot higher than that of <code class="docutils literal notranslate"><span class="pre">trimmomatic</span></code> (N=3)</strong>. Default <code class="docutils literal notranslate"><span class="pre">fastp</span></code> setting is in line with that of <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/guide.html"><code class="docutils literal notranslate"><span class="pre">cutadapt</span></code></a>.</p></li>
</ul>
<p>The default value are set to be the same as the default value of the <code class="docutils literal notranslate"><span class="pre">fastp</span></code> software.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">fastp_trim_adaptor_1</span><span class="p">]</span>
<span class="c1"># sliding window setting</span>
<span class="kn">parameter:</span> <span class="n">window_size</span> <span class="o">=</span> <span class="mi">4</span>
<span class="kn">parameter:</span> <span class="n">required_quality</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># the mean quality requirement option for cut_front</span>
<span class="kn">parameter:</span> <span class="n">leading</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># the mean quality requirement option for cut_tail</span>
<span class="kn">parameter:</span> <span class="n">trailing</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># reads shorter than length_required will be discarded</span>
<span class="kn">parameter:</span> <span class="n">min_len</span> <span class="o">=</span> <span class="mi">15</span>
<span class="c1"># Path to the reference adaptors</span>
<span class="kn">parameter:</span> <span class="n">fasta_with_adapters_etc</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
<span class="n">warn_if</span><span class="p">(</span><span class="n">fasta_with_adapters_etc</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span><span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Use input fasta and adaptor detection of paired-end read was disabled&quot;</span> <span class="p">)</span>

<span class="kn">input:</span> <span class="n">fastq</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">is_paired_end</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;sample_id&quot;</span>
<span class="kn">output:</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.trimmed.fq.gz&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">]</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">bash:</span>  <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span><span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
        <span class="n">fastp</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1"> -I </span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="n">_input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1"> -O </span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="n">_output</span> <span class="p">}</span> \
            <span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;--adapter_fasta </span><span class="si">{</span><span class="n">fasta_with_adapters_etc</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">fasta_with_adapters_etc</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;--detect_adapter_for_pe&quot;</span><span class="p">}</span>  <span class="o">-</span><span class="n">V</span> <span class="o">-</span><span class="n">h</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">html</span> <span class="o">-</span><span class="n">j</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">json</span> <span class="o">-</span><span class="n">w</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
            <span class="o">--</span><span class="n">length_required</span> <span class="err">$</span><span class="p">{</span><span class="n">min_len</span><span class="p">}</span>  <span class="o">-</span><span class="n">W</span> <span class="err">$</span><span class="p">{</span><span class="n">window_size</span><span class="p">}</span> <span class="o">-</span><span class="n">M</span> <span class="err">$</span><span class="p">{</span><span class="n">required_quality</span><span class="p">}</span> <span class="o">-</span><span class="mi">5</span> <span class="o">-</span><span class="mi">3</span> <span class="o">--</span><span class="n">cut_front_mean_quality</span> <span class="err">$</span><span class="p">{</span><span class="n">leading</span><span class="p">}</span> <span class="o">--</span><span class="n">cut_tail_mean_quality</span> <span class="err">$</span><span class="p">{</span><span class="n">leading</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">fastp_trim_adaptor_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.trimmed.txt&#39;</span>
<span class="n">sample_inv_tmp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="k">if</span> <span class="n">is_paired_end</span><span class="p">:</span>
    <span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">fq1</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">r</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">][::</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">fq2</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">r</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">][</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]</span>
<span class="kn">else:</span>
    <span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">fq1</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">r</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">]</span>
<span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">_output</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-1-alternative-remove-adaptor-through-trimmomatic">
<h2>Step 1 Alternative: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code><a class="headerlink" href="#step-1-alternative-remove-adaptor-through-trimmomatic" title="Permalink to this headline">¶</a></h2>
<p>Documentation: <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a></p>
<p>We have replaced this with <code class="docutils literal notranslate"><span class="pre">fastp</span></code>, see above, which performs better than <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> in terms of removing adaptors that <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> cannot detect. <code class="docutils literal notranslate"><span class="pre">fastp</span></code> can also automatically guess the adapter sequences from data and by default no adapter sequence is required for input to <code class="docutils literal notranslate"><span class="pre">fastp</span></code>.</p>
<div class="section" id="id3">
<h3>Step Inputs<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">software_dir</span></code>: directory for the software</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fasta_with_adapters_etc</span></code>: <strong>filename</strong> for the adapter reference file. According to <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> documention,</p></li>
</ul>
<blockquote>
<div><p>As a rule of thumb newer libraries will use <code class="docutils literal notranslate"><span class="pre">TruSeq3</span></code>, but this really depends on your service provider. If you use FASTQC, the “Overrepresented Sequences” report can help indicate which adapter file is best suited for your data. “Illumina Single End” or “Illumina Paired End” sequences indicate single-end or paired-end <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq2-SE.fa</span></code> and <code class="docutils literal notranslate"><span class="pre">TruSeq2-PE.fa</span></code> respectively. “TruSeq Universal Adapter” or “TruSeq Adapter, Index …” indicates <code class="docutils literal notranslate"><span class="pre">TruSeq-3</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq3-SE.fa</span></code> or <code class="docutils literal notranslate"><span class="pre">TruSeq3-PE.fa</span></code>, for single-end and paired-end data respectively. Adapter sequences for <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> multiplexed libraries, indicated by “Illumina Multiplexing
…”, and the various RNA library preparations are not currently included.</p>
</div></blockquote>
<p>We have <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> workflow previously defined and executed. Users should decide what fasta adapter reference to use based on <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> results (or their own knowledge).</p>
</div>
<div class="section" id="id4">
<h3>Step Outputs<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Two paired <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file for alignment</p></li>
<li><p>Two unpaired <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code></p></li>
</ul>
<blockquote>
<div><p>For single-ended data, one input and one output file are specified, plus the processing steps. For paired-end data, two input files are specified, and 4 output files, 2 for the ‘paired’ output where both reads survived the processing, and 2 for corresponding ‘unpaired’ output where a read survived, but the partner read did not.</p>
</div></blockquote>
<p><strong>You need to figure out from fastqc results what adapter reference sequence to use.</strong>, eg <code class="docutils literal notranslate"><span class="pre">--fasta_with_adapters_etc</span> <span class="pre">TruSeq3-PE.fa</span></code>. These files can be downloaded from <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> github repo.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">trimmomatic_trim_adaptor</span><span class="p">]</span>
<span class="c1"># Path to the software. Default set to using our rna_quantification.sif image</span>
<span class="kn">parameter:</span> <span class="n">trimmomatic_jar</span> <span class="o">=</span> <span class="s2">&quot;/opt/Trimmomatic-0.39/trimmomatic-0.39.jar&quot;</span>
<span class="c1"># Illumina clip setting</span>
<span class="c1"># Path to the reference adaptors</span>
<span class="kn">parameter:</span> <span class="n">fasta_with_adapters_etc</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">seed_mismatches</span> <span class="o">=</span> <span class="mi">2</span>
<span class="kn">parameter:</span> <span class="n">palindrome_clip_threshold</span> <span class="o">=</span> <span class="mi">30</span>
<span class="kn">parameter:</span> <span class="n">simple_clip_threshold</span> <span class="o">=</span> <span class="mi">10</span>
<span class="c1"># sliding window setting</span>
<span class="kn">parameter:</span> <span class="n">window_size</span> <span class="o">=</span> <span class="mi">4</span>
<span class="kn">parameter:</span> <span class="n">required_quality</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># Other settings</span>
<span class="kn">parameter:</span> <span class="n">leading</span> <span class="o">=</span> <span class="mi">3</span>
<span class="kn">parameter:</span> <span class="n">trailing</span> <span class="o">=</span> <span class="mi">3</span>
<span class="kn">parameter:</span> <span class="n">min_len</span> <span class="o">=</span> <span class="mi">50</span>
<span class="kn">input:</span> <span class="n">fastq</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">is_paired_end</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;sample_id&quot;</span>
<span class="kn">output:</span> <span class="p">([</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">_paired_</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.gz&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">_unpaired_</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.gz&#39;</span><span class="p">,</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">_paired_</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.gz&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">_unpaired_</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.gz&#39;</span> <span class="p">]</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">_trimmed_</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.gz&#39;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">jar</span> <span class="o">-</span><span class="n">Xmx</span><span class="err">$</span><span class="p">{</span><span class="n">java_mem</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">trimmomatic_jar</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;PE&quot;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="s2">&quot;SE&quot;</span><span class="p">}</span>  <span class="o">-</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">}</span> \
        <span class="n">ILLUMINACLIP</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">fasta_with_adapters_etc</span><span class="p">}:</span><span class="err">$</span><span class="p">{</span><span class="n">seed_mismatches</span><span class="p">}:</span><span class="err">$</span><span class="p">{</span><span class="n">palindrome_clip_threshold</span><span class="p">}:</span><span class="err">$</span><span class="p">{</span><span class="n">simple_clip_threshold</span><span class="p">}</span> \
        <span class="n">LEADING</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">leading</span><span class="p">}</span> <span class="n">TRAILING</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">trailing</span><span class="p">}</span> <span class="n">SLIDINGWINDOW</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">window_size</span><span class="p">}:</span><span class="err">$</span><span class="p">{</span><span class="n">required_quality</span><span class="p">}</span> <span class="n">MINLEN</span><span class="p">:</span><span class="err">$</span><span class="p">{</span><span class="n">min_len</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-2-alignment-through-star">
<h2>Step 2: Alignment through <code class="docutils literal notranslate"><span class="pre">STAR</span></code><a class="headerlink" href="#step-2-alignment-through-star" title="Permalink to this headline">¶</a></h2>
<p>Documentation : <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> and <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/rnaseq/src/run_STAR.py">Script in docker</a></p>
<p>This step is the main step for <code class="docutils literal notranslate"><span class="pre">STAR</span></code> alignment.</p>
<p>Originally, the STAR alignment is implemented using the run_STAR.py command from gtex. However, in order to <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/issues/318">accomodate different read length among samples</a>, We reimplement what was included in the wrapper, including the re-alignment of STAR unsorted bam using samtools to avoid high mem consumption.</p>
<div class="section" id="id5">
<h3>Step Inputs<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>paths to trimmed <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file from Step 1 as documented in the new sample list.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">STAR_index</span></code>: directory for the STAR aligment index</p></li>
</ul>
</div>
<div class="section" id="id6">
<h3>Step Outputs<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>bam file output <code class="docutils literal notranslate"><span class="pre">${cwd}/{sample_id}.Aligned.sortedByCoord.bam</span></code>, will be used in step 3 and 4</p></li>
<li><p>bam file output <code class="docutils literal notranslate"><span class="pre">${cwd}/{sample_id}.Aligned.toTranscriptome.bam</span></code>, will be used in step 5</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">STAR_align_1</span><span class="p">]</span>
<span class="c1"># Reference gene model</span>
<span class="kn">parameter:</span> <span class="n">gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># STAR indexing file</span>
<span class="kn">parameter:</span> <span class="n">STAR_index</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">outFilterMultimapNmax</span> <span class="o">=</span> <span class="mi">20</span> 
<span class="kn">parameter:</span> <span class="n">alignSJoverhangMin</span> <span class="o">=</span> <span class="mi">8</span> 
<span class="kn">parameter:</span> <span class="n">alignSJDBoverhangMin</span> <span class="o">=</span> <span class="mi">1</span> 
<span class="kn">parameter:</span> <span class="n">outFilterMismatchNmax</span> <span class="o">=</span> <span class="mi">999</span> 
<span class="kn">parameter:</span> <span class="n">outFilterMismatchNoverLmax</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">alignIntronMin</span> <span class="o">=</span> <span class="mi">20</span> 
<span class="kn">parameter:</span> <span class="n">alignIntronMax</span> <span class="o">=</span> <span class="mi">1000000</span> 
<span class="kn">parameter:</span> <span class="n">alignMatesGapMax</span> <span class="o">=</span> <span class="mi">1000000</span> 
<span class="kn">parameter:</span> <span class="n">outFilterType</span> <span class="o">=</span>  <span class="s2">&quot;BySJout&quot;</span> 
<span class="kn">parameter:</span> <span class="n">outFilterScoreMinOverLread</span> <span class="o">=</span> <span class="mf">0.33</span> 
<span class="kn">parameter:</span> <span class="n">outFilterMatchNminOverLread</span> <span class="o">=</span> <span class="mf">0.33</span> 
<span class="kn">parameter:</span> <span class="n">limitSjdbInsertNsj</span> <span class="o">=</span> <span class="mi">1200000</span> 
<span class="kn">parameter:</span> <span class="n">outSAMstrandField</span> <span class="o">=</span> <span class="s2">&quot;intronMotif&quot;</span> 
<span class="kn">parameter:</span> <span class="n">outFilterIntronMotifs</span> <span class="o">=</span> <span class="s2">&quot;None&quot;</span> 
<span class="kn">parameter:</span> <span class="n">alignSoftClipAtReferenceEnds</span> <span class="o">=</span> <span class="s2">&quot;Yes&quot;</span> 
<span class="kn">parameter:</span> <span class="n">quantMode</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;TranscriptomeSAM&quot;</span><span class="p">,</span> <span class="s2">&quot;GeneCounts&quot;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">outSAMattrRGline</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ID:rg1&quot;</span><span class="p">,</span> <span class="s2">&quot;SM:sm1&quot;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">outSAMattributes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;NH&quot;</span><span class="p">,</span> <span class="s2">&quot;HI&quot;</span><span class="p">,</span> <span class="s2">&quot;AS&quot;</span><span class="p">,</span> <span class="s2">&quot;nM&quot;</span><span class="p">,</span> <span class="s2">&quot;NM&quot;</span><span class="p">,</span> <span class="s2">&quot;ch&quot;</span><span class="p">]</span> 
<span class="kn">parameter:</span> <span class="n">chimSegmentMin</span> <span class="o">=</span> <span class="mi">15</span> 
<span class="kn">parameter:</span> <span class="n">chimJunctionOverhangMin</span> <span class="o">=</span> <span class="mi">15</span> 
<span class="kn">parameter:</span> <span class="n">chimOutType</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Junctions&quot;</span><span class="p">,</span> <span class="s2">&quot;WithinBAM&quot;</span><span class="p">,</span> <span class="s2">&quot;SoftClip&quot;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">chimMainSegmentMultNmax</span> <span class="o">=</span> <span class="mi">1</span> 
<span class="kn">parameter:</span> <span class="n">sjdbOverhang</span> <span class="o">=</span> <span class="mi">100</span>
<span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="kp">mem</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">))</span> <span class="o">&lt;</span>  <span class="mi">40</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Insufficent memory for STAR, changing to 40G&quot;</span><span class="p">)</span>
    <span class="n">star_mem</span> <span class="o">=</span> <span class="s1">&#39;40G&#39;</span>
<span class="kn">else:</span>
    <span class="n">star_mem</span> <span class="o">=</span> <span class="kp">mem</span>
<span class="c1"># This option is commented out because it will force the downstream analysis to use 40G, which significantlly slow down the process.</span>
<span class="kn">input:</span> <span class="n">fastq</span><span class="p">,</span><span class="n">group_by</span> <span class="o">=</span> <span class="n">is_paired_end</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span><span class="n">group_with</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span><span class="s2">&quot;read_length&quot;</span><span class="p">}</span>
<span class="kn">output:</span> <span class="n">cord_bam</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.Aligned.sortedByCoord.out.bam&#39;</span><span class="p">,</span>
        <span class="n">trans_bam</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.Aligned.toTranscriptome.out.bam&#39;</span>
<span class="k">if</span> <span class="n">_read_length</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using specified --sjdbOverhang as read length&quot;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using read length specified in the sample list&quot;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="n">star_mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_sample_id</span><span class="p">}</span><span class="o">.*.</span><span class="n">out</span><span class="o">.*.</span><span class="n">gz</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_sample_id</span><span class="p">}</span><span class="o">.</span><span class="n">_STARpass1</span>
    <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
    <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">star_start</span><span class="o">.</span><span class="n">timestamp</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">alignReads</span> \
        <span class="o">--</span><span class="n">runThreadN</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">genomeDir</span>  <span class="err">$</span><span class="p">{</span><span class="n">STAR_index</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">readFilesIn</span>  <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">readFilesCommand</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;cat&quot;</span> <span class="k">if</span> <span class="n">uncompressed</span> <span class="k">else</span> <span class="s2">&quot;zcat&quot;</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span> \
        <span class="o">--</span><span class="n">outSAMstrandField</span> <span class="err">$</span><span class="p">{</span><span class="n">outSAMstrandField</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">twopassMode</span> <span class="n">Basic</span> \
        <span class="o">--</span><span class="n">outFilterMultimapNmax</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMultimapNmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignSJoverhangMin</span> <span class="err">$</span><span class="p">{</span><span class="n">alignSJoverhangMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignSJDBoverhangMin</span> <span class="err">$</span><span class="p">{</span><span class="n">alignSJDBoverhangMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterMismatchNmax</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMismatchNmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterMismatchNoverLmax</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMismatchNoverLmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignIntronMin</span> <span class="err">$</span><span class="p">{</span><span class="n">alignIntronMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignIntronMax</span> <span class="err">$</span><span class="p">{</span><span class="n">alignIntronMax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignMatesGapMax</span> <span class="err">$</span><span class="p">{</span><span class="n">alignMatesGapMax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterType</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterType</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterScoreMinOverLread</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterScoreMinOverLread</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterMatchNminOverLread</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMatchNminOverLread</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">limitSjdbInsertNsj</span> <span class="err">$</span><span class="p">{</span><span class="n">limitSjdbInsertNsj</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterIntronMotifs</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterIntronMotifs</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignSoftClipAtReferenceEnds</span> <span class="err">$</span><span class="p">{</span><span class="n">alignSoftClipAtReferenceEnds</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">quantMode</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">quantMode</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">Unsorted</span> \
        <span class="o">--</span><span class="n">outSAMunmapped</span> <span class="n">Within</span> \
        <span class="o">--</span><span class="n">genomeLoad</span> <span class="n">NoSharedMemory</span> \
        <span class="o">--</span><span class="n">chimSegmentMin</span> <span class="err">$</span><span class="p">{</span><span class="n">chimSegmentMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">chimJunctionOverhangMin</span> <span class="err">$</span><span class="p">{</span><span class="n">chimJunctionOverhangMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">chimOutType</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chimOutType</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">chimMainSegmentMultNmax</span> <span class="err">$</span><span class="p">{</span><span class="n">chimMainSegmentMultNmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">chimOutJunctionFormat</span> <span class="mi">0</span> \
        <span class="o">--</span><span class="n">outSAMattributes</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outSAMattributes</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">outSAMattrRGline</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outSAMattrRGline</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">sjdbOverhang</span> <span class="err">$</span><span class="p">{</span><span class="n">sjdbOverhang</span> <span class="k">if</span> <span class="n">_read_length</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">_read_length</span>  <span class="p">}</span> \
        <span class="o">--</span><span class="n">sjdbGTFfile</span> <span class="err">$</span><span class="p">{</span><span class="n">gtf</span><span class="p">}</span> \

    <span class="n">rm</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">_STARgenome</span>
    <span class="err">$</span><span class="p">{</span> <span class="sa">f</span><span class="s1">&#39;rm -rf </span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nnnn</span><span class="si">}</span><span class="s1">._STARtmp&#39;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span>
    <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">sort_start</span><span class="o">.</span><span class="n">timestamp</span>
    <span class="n">samtools</span> <span class="n">sort</span> <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">Aligned</span><span class="o">.</span><span class="n">sortedByCoord</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">Aligned</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span> <span class="c1"># According to GTEX, this can help reducing the amount of memory consumption.</span>
    <span class="n">rm</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">Aligned</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span> 
    <span class="n">samtools</span> <span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">STAR_sQTL_align_1</span><span class="p">]</span>
<span class="c1"># Reference gene model</span>
<span class="kn">parameter:</span> <span class="n">gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># STAR indexing file</span>
<span class="kn">parameter:</span> <span class="n">STAR_index</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">outFilterMultimapNmax</span> <span class="o">=</span> <span class="mi">20</span> 
<span class="kn">parameter:</span> <span class="n">alignSJoverhangMin</span> <span class="o">=</span> <span class="mi">8</span> 
<span class="kn">parameter:</span> <span class="n">alignSJDBoverhangMin</span> <span class="o">=</span> <span class="mi">1</span> 
<span class="kn">parameter:</span> <span class="n">outFilterMismatchNmax</span> <span class="o">=</span> <span class="mi">999</span> 
<span class="kn">parameter:</span> <span class="n">outFilterMismatchNoverLmax</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">alignIntronMin</span> <span class="o">=</span> <span class="mi">20</span> 
<span class="kn">parameter:</span> <span class="n">alignIntronMax</span> <span class="o">=</span> <span class="mi">1000000</span> 
<span class="kn">parameter:</span> <span class="n">alignMatesGapMax</span> <span class="o">=</span> <span class="mi">1000000</span> 
<span class="kn">parameter:</span> <span class="n">outFilterType</span> <span class="o">=</span>  <span class="s2">&quot;BySJout&quot;</span> 
<span class="kn">parameter:</span> <span class="n">outFilterScoreMinOverLread</span> <span class="o">=</span> <span class="mf">0.33</span> 
<span class="kn">parameter:</span> <span class="n">outFilterMatchNminOverLread</span> <span class="o">=</span> <span class="mf">0.33</span> 
<span class="kn">parameter:</span> <span class="n">limitSjdbInsertNsj</span> <span class="o">=</span> <span class="mi">1200000</span> 
<span class="kn">parameter:</span> <span class="n">outSAMstrandField</span> <span class="o">=</span> <span class="s2">&quot;intronMotif&quot;</span> 
<span class="kn">parameter:</span> <span class="n">outFilterIntronMotifs</span> <span class="o">=</span> <span class="s2">&quot;None&quot;</span> 
<span class="kn">parameter:</span> <span class="n">alignSoftClipAtReferenceEnds</span> <span class="o">=</span> <span class="s2">&quot;Yes&quot;</span> 
<span class="kn">parameter:</span> <span class="n">quantMode</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;TranscriptomeSAM&quot;</span><span class="p">,</span> <span class="s2">&quot;GeneCounts&quot;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">outSAMattrRGline</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ID:rg1&quot;</span><span class="p">,</span> <span class="s2">&quot;SM:sm1&quot;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">outSAMattributes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;NH&quot;</span><span class="p">,</span> <span class="s2">&quot;HI&quot;</span><span class="p">,</span> <span class="s2">&quot;AS&quot;</span><span class="p">,</span> <span class="s2">&quot;nM&quot;</span><span class="p">,</span> <span class="s2">&quot;NM&quot;</span><span class="p">,</span> <span class="s2">&quot;ch&quot;</span><span class="p">]</span> 
<span class="kn">parameter:</span> <span class="n">chimSegmentMin</span> <span class="o">=</span> <span class="mi">15</span> 
<span class="kn">parameter:</span> <span class="n">chimJunctionOverhangMin</span> <span class="o">=</span> <span class="mi">15</span> 
<span class="kn">parameter:</span> <span class="n">chimOutType</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Junctions&quot;</span><span class="p">,</span> <span class="s2">&quot;WithinBAM&quot;</span><span class="p">,</span> <span class="s2">&quot;SoftClip&quot;</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">chimMainSegmentMultNmax</span> <span class="o">=</span> <span class="mi">1</span> 
<span class="kn">parameter:</span> <span class="n">sjdbOverhang</span> <span class="o">=</span> <span class="mi">100</span>
<span class="kn">parameter:</span> <span class="n">wasp_VCF</span> <span class="o">=</span> <span class="n">path</span>
<span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="kp">mem</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">))</span> <span class="o">&lt;</span>  <span class="mi">40</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Insufficent memory for STAR, changing to 40G&quot;</span><span class="p">)</span>
    <span class="n">star_mem</span> <span class="o">=</span> <span class="s1">&#39;40G&#39;</span>
<span class="kn">else:</span>
    <span class="n">star_mem</span> <span class="o">=</span> <span class="kp">mem</span>
<span class="c1"># This option is commented out because it will force the downstream analysis to use 40G, which significantlly slow down the process.</span>
<span class="kn">input:</span> <span class="n">fastq</span><span class="p">,</span><span class="n">group_by</span> <span class="o">=</span> <span class="n">is_paired_end</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span><span class="n">group_with</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span><span class="s2">&quot;read_length&quot;</span><span class="p">}</span>
<span class="kn">output:</span> <span class="n">cord_bam</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.Aligned.sortedByCoord.out.bam&#39;</span><span class="p">,</span>
        <span class="n">trans_bam</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.Aligned.toTranscriptome.out.bam&#39;</span>
<span class="k">if</span> <span class="n">_read_length</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using specified --sjdbOverhang as read length&quot;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using read length specified in the sample list&quot;</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="n">star_mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_sample_id</span><span class="p">}</span><span class="o">.*.</span><span class="n">out</span><span class="o">.*.</span><span class="n">gz</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_sample_id</span><span class="p">}</span><span class="o">.</span><span class="n">_STARpass1</span>
    <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
    <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">star_start</span><span class="o">.</span><span class="n">timestamp</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">alignReads</span> \
        <span class="o">--</span><span class="n">runThreadN</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">genomeDir</span>  <span class="err">$</span><span class="p">{</span><span class="n">STAR_index</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">readFilesIn</span>  <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">readFilesCommand</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;cat&quot;</span> <span class="k">if</span> <span class="n">uncompressed</span> <span class="k">else</span> <span class="s2">&quot;zcat&quot;</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span> \
        <span class="o">--</span><span class="n">outSAMstrandField</span> <span class="err">$</span><span class="p">{</span><span class="n">outSAMstrandField</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">twopassMode</span> <span class="n">Basic</span> \
        <span class="o">--</span><span class="n">outFilterMultimapNmax</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMultimapNmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignSJoverhangMin</span> <span class="err">$</span><span class="p">{</span><span class="n">alignSJoverhangMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignSJDBoverhangMin</span> <span class="err">$</span><span class="p">{</span><span class="n">alignSJDBoverhangMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterMismatchNmax</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMismatchNmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterMismatchNoverLmax</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMismatchNoverLmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignIntronMin</span> <span class="err">$</span><span class="p">{</span><span class="n">alignIntronMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignIntronMax</span> <span class="err">$</span><span class="p">{</span><span class="n">alignIntronMax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignMatesGapMax</span> <span class="err">$</span><span class="p">{</span><span class="n">alignMatesGapMax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterType</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterType</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterScoreMinOverLread</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterScoreMinOverLread</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterMatchNminOverLread</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterMatchNminOverLread</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">limitSjdbInsertNsj</span> <span class="err">$</span><span class="p">{</span><span class="n">limitSjdbInsertNsj</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">outFilterIntronMotifs</span> <span class="err">$</span><span class="p">{</span><span class="n">outFilterIntronMotifs</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">alignSoftClipAtReferenceEnds</span> <span class="err">$</span><span class="p">{</span><span class="n">alignSoftClipAtReferenceEnds</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">quantMode</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">quantMode</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">Unsorted</span> \
        <span class="o">--</span><span class="n">outSAMunmapped</span> <span class="n">Within</span> \
        <span class="o">--</span><span class="n">genomeLoad</span> <span class="n">NoSharedMemory</span> \
        <span class="o">--</span><span class="n">chimSegmentMin</span> <span class="err">$</span><span class="p">{</span><span class="n">chimSegmentMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">chimJunctionOverhangMin</span> <span class="err">$</span><span class="p">{</span><span class="n">chimJunctionOverhangMin</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">chimOutType</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chimOutType</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">chimMainSegmentMultNmax</span> <span class="err">$</span><span class="p">{</span><span class="n">chimMainSegmentMultNmax</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">chimOutJunctionFormat</span> <span class="mi">0</span> \
        <span class="o">--</span><span class="n">outSAMattributes</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outSAMattributes</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">outSAMattrRGline</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outSAMattrRGline</span><span class="p">)}</span> \
        <span class="o">--</span><span class="n">sjdbOverhang</span> <span class="err">$</span><span class="p">{</span><span class="n">sjdbOverhang</span> <span class="k">if</span> <span class="n">_read_length</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">_read_length</span>  <span class="p">}</span> \
        <span class="o">--</span><span class="n">sjdbGTFfile</span> <span class="err">$</span><span class="p">{</span><span class="n">gtf</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">varVCFfile</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">zcat</span> <span class="err">$</span><span class="p">{</span><span class="n">wasp_VCF</span><span class="p">})</span> \
        <span class="o">--</span><span class="n">waspOutputMode</span> <span class="n">SAMtag</span> \

    <span class="n">rm</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">_STARgenome</span>
    <span class="err">$</span><span class="p">{</span> <span class="sa">f</span><span class="s1">&#39;rm -rf </span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nnnn</span><span class="si">}</span><span class="s1">._STARtmp&#39;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span>
    <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">sort_start</span><span class="o">.</span><span class="n">timestamp</span>
    <span class="n">samtools</span> <span class="n">sort</span> <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">Aligned</span><span class="o">.</span><span class="n">sortedByCoord</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">Aligned</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span> <span class="c1"># According to GTEX, this can help reducing the amount of memory consumption.</span>
    <span class="n">rm</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">}</span><span class="o">.</span><span class="n">Aligned</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span> 
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">h</span> <span class="o">-</span><span class="n">q</span> <span class="mi">255</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;vW:i:[2-7]&quot;</span> <span class="o">|</span> <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">b</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">bam</span>
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">h</span> <span class="o">-</span><span class="n">q</span> <span class="mi">255</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;vW:i:[2-7]&quot;</span> <span class="o">|</span> <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">b</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">bam</span>

    <span class="n">samtools</span> <span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">strand_detected_1</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s2">&quot;step_strand_detected&quot;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;STAR_align_1&quot;</span><span class="p">)[</span><span class="s2">&quot;trans_bam&quot;</span><span class="p">],</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;sample_id&quot;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">ReadsPerGene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.ReadsPerGene.out.tab&#39;</span> <span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span> <span class="p">)</span>
<span class="n">ReadsPerGene_list</span> <span class="o">=</span> <span class="n">ReadsPerGene</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">3</span><span class="p">::,]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">numeric_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">strand_percentage</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">/</span><span class="n">ReadsPerGene_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ReadsPerGene_list</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;for sample </span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.9</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Counts for the 1st read strand aligned with RNA is </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">, &gt; 90% of aligned count&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data is likely FR/fr-secondstrand&#39;</span><span class="p">)</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s2">&quot;fr&quot;</span>
<span class="k">elif</span>  <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.9</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Counts for the 2nd read strand aligned with RNA is </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1">, &gt; 90% of aligned count&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data is likely RF/fr-firststrand&#39;</span><span class="p">)</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s2">&quot;rf&quot;</span>
<span class="k">elif</span> <span class="nb">max</span><span class="p">(</span> <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>  <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mf">0.6</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Both </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1"> and  </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1"> are under 60% of reads explained by one direction&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data is likely unstranded&#39;</span><span class="p">)</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s2">&quot;unstranded&quot;</span>
<span class="kn">else:</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s1">&#39;strand_missing&#39;</span> 
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data does not fall into a likely stranded (max percent explained </span><span class="si">{</span><span class="nb">max</span><span class="p">(</span> <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>  <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">}</span><span class="s1"> &gt; 0.9) or unstranded layout (max percent explained </span><span class="si">{</span><span class="nb">max</span><span class="p">(</span> <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>  <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">}</span><span class="s1"> &lt; 0.6), please check your data and manually specified the ``--strand`` option as ``fr``, ``rf`` or ``unstranded``&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">strand_detected_2</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">parameter:</span> <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">strand_inv</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">strand_inv</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">strand</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">strand</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;strand_missing&quot;</span><span class="p">:</span>
                <span class="n">strand</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">step_strand_detected</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using strand specified in the input samples list </span><span class="si">{</span><span class="n">strand</span><span class="si">}</span><span class="s1">, replacing strand_missing with detected strand&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">warn_if</span><span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span> <span class="ow">is</span> <span class="n">step_strand_detected</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">step_strand_detected</span><span class="p">),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;strands detected are different among samples, please check your protocol, we will use the detected strand for each samples&quot;</span><span class="p">)</span>    
        <span class="n">strand</span> <span class="o">=</span> <span class="n">step_strand_detected</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using detected strand for each samples </span><span class="si">{</span><span class="n">strand</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">stop_if</span><span class="p">(</span><span class="n">strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;fr&quot;</span><span class="p">,</span> <span class="s2">&quot;rf&quot;</span><span class="p">,</span> <span class="s2">&quot;unstranded&quot;</span><span class="p">],</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;``--strand`` option should be ``fr``, ``rf`` or ``unstranded``&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using ``--strand`` overwrite option for all the samples </span><span class="si">{</span><span class="n">strand</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span> 
    <span class="n">strand</span> <span class="o">=</span> <span class="p">[</span><span class="n">strand</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">step_strand_detected</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-3-mark-duplicates-reads-qc-through-picard">
<h2>Step 3: Mark duplicates reads &amp; QC through <code class="docutils literal notranslate"><span class="pre">Picard</span></code><a class="headerlink" href="#step-3-mark-duplicates-reads-qc-through-picard" title="Permalink to this headline">¶</a></h2>
<p>This step is the first QC step after <code class="docutils literal notranslate"><span class="pre">STAR</span></code> alignment. This step will performed QC to collect multipe metrics regarding the RNASeq using Picard. Then it will also generate a new <code class="docutils literal notranslate"><span class="pre">.bam</span></code> file with duplication marked with the hexadecimal value of <code class="docutils literal notranslate"><span class="pre">0x0400</span></code>, which corresponds to a decimal value of 1024</p>
<div class="section" id="id7">
<h3>Step Inputs:<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">STAR_bam</span></code>: path to the output in Step 2.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">reference_fasta</span></code>:  The fasta reference file used to generate star index, it is critical that the this fasta is <em>exactly</em> the same as those that generate the STAR Index, else this error will occurs: <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/issues/357">https://github.com/cumc/xqtl-pipeline/issues/357</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">RefFlat</span> <span class="pre">file</span></code></p></li>
</ul>
<p>This file is needed for picard CollectRnaSeqMetrics module, which in turn</p>
<blockquote>
<div><p>produces metrics describing the distribution of the bases within the transcripts. It calculates the total numbers and the fractions of nucleotides within specific genomic regions including untranslated regions (UTRs), introns, intergenic sequences (between discrete genes), and peptide-coding sequences (exons). This tool also determines the numbers of bases that pass quality filters that are specific to Illumina data (PF_BASES).</p>
</div></blockquote>
<p>The refFlat file can be generated by the reference_data module, RefFlat_generation step.</p>
</div>
<div class="section" id="id8">
<h3>Step Outputs:<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>A collection of metrics file for each of the samples</p></li>
<li><p>A new <code class="docutils literal notranslate"><span class="pre">.bam</span></code> file with duplication marked with the hexadecimal value of <code class="docutils literal notranslate"><span class="pre">0x0400</span></code>, which corresponds to a decimal value of 1024</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">picard_qc</span><span class="p">]</span>
<span class="c1"># Reference gene model</span>
<span class="kn">parameter:</span> <span class="n">gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s1">&#39;strand&#39;</span><span class="p">)</span>
<span class="c1"># Path to flat reference file, for computing QC metric</span>
<span class="kn">parameter:</span> <span class="n">ref_flat</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Path to the software. Default set to using our rna_quantification.sif image</span>
<span class="kn">parameter:</span> <span class="n">picard_jar</span> <span class="o">=</span> <span class="s2">&quot;/opt/picard-tools/picard.jar&quot;</span>
<span class="c1"># The fasta reference file used to generate star index</span>
<span class="kn">parameter:</span> <span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># For the patterned flowcell models (HiSeq X), change to 2500</span>
<span class="kn">parameter:</span> <span class="n">optical_distance</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">picard_strand_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;rf&quot;</span><span class="p">:</span><span class="s2">&quot;SECOND_READ_TRANSCRIPTION_STRAND&quot;</span><span class="p">,</span><span class="s2">&quot;fr&quot;</span><span class="p">:</span> <span class="s2">&quot;FIRST_READ_TRANSCRIPTION_STRAND&quot;</span><span class="p">,</span><span class="s2">&quot;unstranded&quot;</span><span class="p">:</span><span class="s2">&quot;NONE&quot;</span> <span class="p">}</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;STAR_align_1&quot;</span><span class="p">),</span><span class="n">group_by</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span><span class="s2">&quot;strand&quot;</span><span class="p">}</span>
<span class="kn">output:</span> <span class="n">picard_metrics</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.alignment_summary_metrics&#39;</span><span class="p">,</span>
        <span class="n">picard_rna_metrics</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rna_metrics&#39;</span><span class="p">,</span>
        <span class="n">md_bam</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.Aligned.sortedByCoord.out.md.bam&#39;</span><span class="p">,</span>
        <span class="n">md_metrics</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.Aligned.sortedByCoord.md.metrics&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
        <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
        <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">CollectMultipleMetrics_start</span><span class="o">.</span><span class="n">timestamp</span>
        <span class="n">java</span> <span class="o">-</span><span class="n">jar</span> <span class="o">-</span><span class="n">Xmx</span><span class="err">$</span><span class="p">{</span><span class="n">java_mem</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_jar</span><span class="p">}</span> <span class="n">CollectMultipleMetrics</span> \
            <span class="o">-</span><span class="n">REFERENCE_SEQUENCE</span> <span class="err">$</span><span class="p">{</span><span class="n">reference_fasta</span><span class="p">}</span> \
            <span class="o">-</span><span class="n">PROGRAM</span> <span class="n">CollectAlignmentSummaryMetrics</span> \
            <span class="o">-</span><span class="n">PROGRAM</span> <span class="n">CollectInsertSizeMetrics</span> \
            <span class="o">-</span><span class="n">PROGRAM</span> <span class="n">QualityScoreDistribution</span> \
            <span class="o">-</span><span class="n">PROGRAM</span> <span class="n">MeanQualityByCycle</span> \
            <span class="o">-</span><span class="n">PROGRAM</span> <span class="n">CollectBaseDistributionByCycle</span> \
            <span class="o">-</span><span class="n">PROGRAM</span> <span class="n">CollectGcBiasMetrics</span> \
            <span class="o">-</span><span class="n">VALIDATION_STRINGENCY</span> <span class="n">STRICT</span> \
            <span class="o">-</span><span class="n">INPUT</span>  <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]}</span> \
            <span class="o">-</span><span class="n">OUTPUT</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span>

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
        <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
        <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">MarkDuplicates_start</span><span class="o">.</span><span class="n">timestamp</span>
        <span class="n">java</span> <span class="o">-</span><span class="n">Xmx</span><span class="err">$</span><span class="p">{</span><span class="n">java_mem</span><span class="p">}</span> <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_jar</span><span class="p">}</span> <span class="n">MarkDuplicates</span> \
            <span class="o">-</span><span class="n">I</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]}</span>  \
            <span class="o">-</span><span class="n">O</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]}</span> \
            <span class="o">-</span><span class="n">PROGRAM_RECORD_ID</span> <span class="n">null</span> \
            <span class="o">-</span><span class="n">M</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span> \
            <span class="o">-</span><span class="n">TMP_DIR</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span>\
            <span class="o">-</span><span class="n">MAX_RECORDS_IN_RAM</span> <span class="mi">500000</span> <span class="o">-</span><span class="n">SORTING_COLLECTION_SIZE_RATIO</span> <span class="mf">0.25</span> \
            <span class="o">-</span><span class="n">ASSUME_SORT_ORDER</span> <span class="n">coordinate</span> \
            <span class="o">-</span><span class="n">TAGGING_POLICY</span> <span class="n">DontTag</span> \
            <span class="o">-</span><span class="n">OPTICAL_DUPLICATE_PIXEL_DISTANCE</span> <span class="err">$</span><span class="p">{</span><span class="n">optical_distance</span><span class="p">}</span> \
            <span class="o">-</span><span class="n">CREATE_INDEX</span> <span class="n">true</span> \
            <span class="o">-</span><span class="n">CREATE_MD5_FILE</span> <span class="n">true</span> \
            <span class="o">-</span><span class="n">VALIDATION_STRINGENCY</span> <span class="n">STRICT</span> \
            <span class="o">-</span><span class="n">REMOVE_SEQUENCING_DUPLICATES</span> <span class="n">false</span> \
            <span class="o">-</span><span class="n">REMOVE_DUPLICATES</span> <span class="n">false</span> 

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.rna_metrics.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.rna_metrics.stderr&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
        <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
        <span class="c1"># Get only line with rRNA and transcript_id</span>
        <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">gtf</span><span class="p">}</span><span class="o">|</span> <span class="n">grep</span> <span class="n">rRNA</span> <span class="o">|</span> <span class="n">grep</span> <span class="n">transcript_id</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">gtf</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span><span class="o">.</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]:</span><span class="n">bnnnn</span><span class="p">}</span>  <span class="c1"># To Avoid data racing problem</span>
        <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">H</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]}</span>  <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]}</span><span class="o">.</span><span class="n">RI</span>  

<span class="kn">python:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.rna_metrics.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.rna_metrics.stderr&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
        <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
        <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">end</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">tag</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">annotation_gtf</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{gtf}</span><span class="s2">.tmp.${_input[&quot;</span><span class="n">cord_bam</span><span class="s2">&quot;]:bnnnn}&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">annotation_gtf</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gtf</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gtf</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;#&#39;</span> <span class="ow">or</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;transcript&quot;</span><span class="p">:</span> <span class="k">continue</span> <span class="c1"># skip header</span>
                <span class="n">chrom</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">start</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">end</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
                <span class="n">strand</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
                <span class="n">attributes</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                    <span class="n">kv</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">kv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">!=</span><span class="s1">&#39;tag&#39;</span><span class="p">:</span>
                        <span class="n">attributes</span><span class="p">[</span><span class="n">kv</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">kv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">attributes</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s1">&#39;tags&#39;</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">kv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">tag</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">attributes</span><span class="p">)</span>
        <span class="n">transcript_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;transcript_id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tag</span><span class="p">]</span>
        <span class="n">RI</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;chr&#39;</span><span class="p">:</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">:</span><span class="n">start</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">:</span><span class="n">end</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">:</span><span class="n">strand</span><span class="p">,</span><span class="s1">&#39;transcript_id&#39;</span> <span class="p">:</span> <span class="n">transcript_id</span> <span class="p">})</span>
        <span class="n">RI</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input[&quot;cord_bam&quot;]}</span><span class="s2">.RI&quot;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">)</span>

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.rna_metrics.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.rna_metrics.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
        <span class="nb">set</span> <span class="o">-</span><span class="n">e</span>
        <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">CollectRnaSeqMetrics_start</span><span class="o">.</span><span class="n">timestamp</span>
        <span class="n">java</span> <span class="o">-</span><span class="n">jar</span> <span class="o">-</span><span class="n">Xmx</span><span class="err">$</span><span class="p">{</span><span class="n">java_mem</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_jar</span><span class="p">}</span> <span class="n">CollectRnaSeqMetrics</span> \
            <span class="o">-</span><span class="n">REF_FLAT</span> <span class="err">$</span><span class="p">{</span><span class="n">ref_flat</span><span class="p">}</span> \
            <span class="o">-</span><span class="n">RIBOSOMAL_INTERVALS</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]}</span><span class="o">.</span><span class="n">RI</span> \
            <span class="o">-</span><span class="n">STRAND_SPECIFICITY</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_strand_dict</span><span class="p">[</span><span class="n">_strand</span><span class="p">]}</span> \
            <span class="o">-</span><span class="n">CHART_OUTPUT</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">rna_metrics</span><span class="o">.</span><span class="n">pdf</span> \
            <span class="o">-</span><span class="n">VALIDATION_STRINGENCY</span> <span class="n">STRICT</span> \
            <span class="o">-</span><span class="n">INPUT</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]}</span>  \
            <span class="o">-</span><span class="n">OUTPUT</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">rna_metrics</span>
        <span class="n">rm</span> <span class="err">$</span><span class="p">{</span><span class="n">gtf</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span><span class="o">.</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]:</span><span class="n">bnnnn</span><span class="p">}</span>

<span class="n">_input</span><span class="p">[</span><span class="s2">&quot;cord_bam&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">zap</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">STAR_output</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;picard_qc&quot;</span><span class="p">),</span>  <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">_bam_list&#39;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">coord_bam_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">b</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">4</span><span class="p">]]</span>
<span class="n">SJ_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">bnnnnn</span><span class="si">}</span><span class="s1">.SJ.out.tab&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">4</span><span class="p">]]</span>
<span class="n">Trans_bam_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">bnnnn</span><span class="si">}</span><span class="s1">.toTranscriptome.out.bam&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">4</span><span class="p">]]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">SJ_list</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;sample_id&quot;</span> <span class="p">:</span> <span class="n">sample_id</span><span class="p">,</span><span class="s2">&quot;strand&quot;</span> <span class="p">:</span> <span class="n">strand</span> <span class="p">,</span> <span class="s2">&quot;coord_bam_list&quot;</span> <span class="p">:</span> <span class="n">coord_bam_list</span><span class="p">,</span> <span class="s2">&quot;SJ_list&quot;</span> <span class="p">:</span><span class="n">SJ_list</span><span class="p">,</span><span class="s2">&quot;trans_bam_list&quot;</span> <span class="p">:</span> <span class="n">Trans_bam_list</span>  <span class="p">})</span>
<span class="n">out</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">_output</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-4-post-aligment-qc-through-rna-seqc">
<h2>Step 4: Post aligment QC through <code class="docutils literal notranslate"><span class="pre">RNA-SeQC</span></code><a class="headerlink" href="#step-4-post-aligment-qc-through-rna-seqc" title="Permalink to this headline">¶</a></h2>
<p>Documentation : <a class="reference external" href="https://github.com/getzlab/rnaseqc">RNA-SeQC</a> and <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/rnaseq/src/run_rnaseqc.py">Script in docker</a></p>
<p>This step is second QC step after <code class="docutils literal notranslate"><span class="pre">STAR</span></code> alignment. It will perform RNA-seq quantification as well.</p>
<div class="section" id="id9">
<h3>Step Inputs<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">bam_list</span></code>: path to the output of STAR_output, which outlined the strands and bam file used for each samples.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code>: reference genome <code class="docutils literal notranslate"><span class="pre">.gtf</span></code> file, this gtf file need to have the same chr name format as the index used to generate the bam file and must be on collapsed gene gtf</p></li>
</ul>
</div>
<div class="section" id="id10">
<h3>Step Outputs<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<p>The following output files are generated in the output directory you provide:</p>
<ul class="simple">
<li><p>{sample}.metrics.tsv : A tab-delimited list of (Statistic, Value) pairs of all statistics and metrics recorded.</p></li>
<li><p>{sample}.exon_reads.gct : A tab-delimited GCT file with (Exon ID, Gene Name, coverage) tuples for all exons which had at least part of one read mapped.</p></li>
<li><p>{sample}.gene_reads.gct : A tab-delimited GCT file with (Gene ID, Gene Name, coverage) tuples for all genes which had at least one read map to at least one of its exons</p></li>
<li><p>{sample}.gene_tpm.gct : A tab-delimited GCT file with (Gene ID, Gene Name, TPM) tuples for all genes reported in the gene_reads.gct file. Note: this file is renamed to .gene_rpkm.gct if the –rpkm flag is present.</p></li>
<li><p>{sample}.fragmentSizes.txt : A list of fragment sizes recorded, if a BED file was provided</p></li>
<li><p>{sample}.coverage.tsv : A tab-delimited list of (Gene ID, Transcript ID, Mean Coverage, Coverage Std, Coverage CV) tuples for all transcripts encountered in the GTF.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">rnaseqc_call_1</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">parameter:</span> <span class="n">bam_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Reference gene model</span>
<span class="kn">parameter:</span> <span class="n">gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">sample_inv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">bam_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">detection_threshold</span> <span class="o">=</span> <span class="mi">5</span>
<span class="kn">parameter:</span> <span class="n">mapping_quality</span> <span class="o">=</span> <span class="mi">255</span>
<span class="c1">## Extract strand information if user have specified the strand</span>
<span class="n">strand_list</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">strand</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">stop_if</span><span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;fr&quot;</span><span class="p">,</span> <span class="s2">&quot;rf&quot;</span><span class="p">,</span> <span class="s2">&quot;unstranded&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">strand_list</span> <span class="p">]),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;strand columns should only include ``fr``, ``rf`` or ``unstranded``, please check the bam_list&quot;</span><span class="p">)</span>     
<span class="c1">## Extract sample_id</span>
<span class="n">sample_id</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    
<span class="c1">## Get the file name for cood_bam data</span>
<span class="n">coord_bam_list_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">coord_bam_list</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">coord_bam_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">coord_bam_list_inv</span> <span class="p">]</span>
<span class="kn">input:</span>  <span class="n">coord_bam_list</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span><span class="s2">&quot;strand_list&quot;</span><span class="p">}</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rnaseqc.gene_tpm.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rnaseqc.gene_reads.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rnaseqc.exon_reads.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rnaseqc.metrics.tsv&#39;</span>

<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">cd</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">rnaseqc</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">gtf</span><span class="p">:</span><span class="n">a</span><span class="p">}</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">a</span><span class="p">}</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span>  \
        <span class="err">$</span><span class="p">{(</span><span class="s2">&quot;--stranded &quot;</span> <span class="o">+</span> <span class="n">_strand_list</span><span class="p">)</span> <span class="k">if</span> <span class="n">_strand_list</span> <span class="o">!=</span> <span class="s2">&quot;unstranded&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
        <span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;--detection-threshold </span><span class="si">{</span><span class="n">detection_threshold</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;--mapping-quality </span><span class="si">{</span><span class="n">mapping_quality</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span> <span class="s1">&#39;&#39;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="s1">&#39;-u&#39;</span><span class="p">}</span>

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">b</span><span class="p">}</span><span class="o">.</span><span class="n">gene_tpm</span><span class="o">.</span><span class="n">gct</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">b</span><span class="p">}</span><span class="o">.</span><span class="n">gene_reads</span><span class="o">.</span><span class="n">gct</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">b</span><span class="p">}</span><span class="o">.</span><span class="n">exon_reads</span><span class="o">.</span><span class="n">gct</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">b</span><span class="p">}</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">tsv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]}</span>
    <span class="n">gzip</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>The RNASEQC results were merged in the following step,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">rnaseqc_call_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rnaseqc.gene_tpm.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rnaseqc.gene_readsCount.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rnaseqc.exon_readsCount.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rnaseqc.metrics.tsv&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">python:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">os</span>
    <span class="k">def</span> <span class="nf">make_gct</span><span class="p">(</span><span class="n">gct_path</span><span class="p">):</span>
        <span class="c1"># sample_name</span>
        <span class="n">sample_name</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">gct_path</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">4</span><span class="p">])</span>
        <span class="c1"># read_input</span>
        <span class="n">pre_gct</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gct_path</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                              <span class="n">skiprows</span><span class="o">=</span> <span class="mi">2</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="s2">&quot;Name&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;Description&quot;</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">pre_gct</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;gene_ID&quot;</span>
        <span class="n">pre_gct</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_name</span><span class="p">]</span>
        <span class="k">return</span><span class="p">(</span><span class="n">pre_gct</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">merge_gct</span><span class="p">(</span><span class="n">gct_path_list</span><span class="p">):</span>
        <span class="n">gct</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">gct_path</span> <span class="ow">in</span> <span class="n">gct_path_list</span><span class="p">:</span>
            <span class="c1">#check duplicated indels and remove them.</span>
            <span class="n">gct_col</span> <span class="o">=</span> <span class="n">make_gct</span><span class="p">(</span><span class="n">gct_path</span><span class="p">)</span>
            <span class="n">gct</span> <span class="o">=</span> <span class="n">gct</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">gct_col</span><span class="p">,</span><span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">left_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span><span class="n">how</span> <span class="o">=</span> <span class="s2">&quot;outer&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">gct</span>

    <span class="n">input_list</span> <span class="o">=</span> <span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}]</span>
    <span class="n">tpm_list</span> <span class="o">=</span> <span class="n">input_list</span><span class="p">[</span><span class="mi">0</span><span class="p">::</span><span class="mi">4</span><span class="p">]</span>
    <span class="n">gc_list</span> <span class="o">=</span> <span class="n">input_list</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">4</span><span class="p">]</span>
    <span class="n">ec_list</span> <span class="o">=</span> <span class="n">input_list</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">4</span><span class="p">]</span>
    <span class="n">gct_path_list_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">tpm_list</span><span class="p">,</span><span class="n">gc_list</span><span class="p">,</span><span class="n">ec_list</span><span class="p">]</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">,}][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">output_path</span><span class="p">)):</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">merge_gct</span><span class="p">(</span><span class="n">gct_path_list_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">output</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">metrics_list</span> <span class="o">=</span> <span class="n">input_list</span><span class="p">[</span><span class="mi">3</span><span class="p">::</span><span class="mi">4</span><span class="p">]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{cwd}</span><span class="s2">/${samples:bn}.rnaseqc.metrics_output_list&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">metrics_list</span><span class="p">))</span>

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">aggregate_rnaseqc_metrics</span><span class="o">.</span><span class="n">py</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span><span class="n">_output_list</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">nn</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-6-quantify-expression-through-rsem">
<h2>Step 6: Quantify expression through <code class="docutils literal notranslate"><span class="pre">RSEM</span></code><a class="headerlink" href="#step-6-quantify-expression-through-rsem" title="Permalink to this headline">¶</a></h2>
<p>Documentation : <a class="reference external" href="https://deweylab.github.io/RSEM/rsem-calculate-expression.html">RSEM</a> and <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/rnaseq/src/run_RSEM.py">Script in docker</a></p>
<p>This step generate the expression matrix from STAR output. Estimate gene and isoform expression from RNA-Seq data are generated.</p>
<div class="section" id="step-input">
<h3>Step Input<a class="headerlink" href="#step-input" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">bam_list</span></code>: path to the output of STAR_output, which outlined the strands and bam file used for each samples.</p></li>
<li><p>transcript-level BAM file: path to the output of Step 3.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">RSEM_index</span></code>: path to RSEM index</p></li>
</ul>
</div>
<div class="section" id="id11">
<h3>Step Outputs<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<p>Please see the output section of <a class="reference external" href="https://deweylab.github.io/RSEM/rsem-calculate-expression.html">https://deweylab.github.io/RSEM/rsem-calculate-expression.html</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">rsem_call_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">RSEM_index</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">max_frag_len</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">estimate_rspd</span> <span class="o">=</span> <span class="kc">True</span>

<span class="kn">parameter:</span> <span class="n">bam_list</span> <span class="o">=</span> <span class="n">path</span>

<span class="n">sample_inv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">bam_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1">## Extract strand information if user have specified the strand</span>
<span class="n">strand_list</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">strand</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">stop_if</span><span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;fr&quot;</span><span class="p">,</span> <span class="s2">&quot;rf&quot;</span><span class="p">,</span> <span class="s2">&quot;unstranded&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">strand_list</span> <span class="p">]),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;strand columns should only include ``fr``, ``rf`` or ``unstranded``, please check the bam_list&quot;</span><span class="p">)</span>     
<span class="c1">## Extract sample_id</span>
<span class="n">sample_id</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    
<span class="c1">## Get the file name for trans_bam_list data</span>
<span class="n">trans_bam_list_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">trans_bam_list</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">trans_bam_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">trans_bam_list_inv</span> <span class="p">]</span>
<span class="kn">input:</span> <span class="n">trans_bam_list</span> <span class="p">,</span>  <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span><span class="s2">&quot;strand_list&quot;</span><span class="p">}</span> 
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rsem.isoforms.results&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rsem.genes.results&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rsem.stat/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.rsem.cnt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">run_RSEM</span><span class="o">.</span><span class="n">py</span> <span class="err">$</span><span class="p">{</span><span class="n">RSEM_index</span><span class="p">:</span><span class="n">a</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">a</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_sample_id</span><span class="p">}</span> \
        <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">max_frag_len</span> <span class="err">$</span><span class="p">{</span><span class="n">max_frag_len</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">estimate_rspd</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;true&#39;</span> <span class="k">if</span> <span class="n">estimate_rspd</span> <span class="k">else</span> <span class="s1">&#39;false&#39;</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">paired_end</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;true&quot;</span> <span class="k">if</span> <span class="n">is_paired_end</span> <span class="k">else</span> <span class="s2">&quot;false&quot;</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">is_stranded</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;true&quot;</span> <span class="k">if</span> <span class="n">_strand_list</span> <span class="o">!=</span> <span class="s2">&quot;unstranded&quot;</span> <span class="k">else</span> <span class="s2">&quot;false&quot;</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>The RSEM results were merged in the following steps, seven files (four for each columns in the isoform output and 3 for each of the genes output) will be generated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">rsem_call_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem_transcripts_expected_count.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem_transcripts_tpm.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem_transcripts_fpkm.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem_transcripts_isopct.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem_genes_expected_count.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem_genes_tpm.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem_genes_fpkm.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.rsem.aggregated_quality.metrics.tsv&#39;</span>

<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span>
<span class="kn">python:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">input_list</span> <span class="o">=</span> <span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{cwd}</span><span class="s1">/${samples:bn}.rsem.isoforms_output_list&#39;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">input_list</span><span class="p">[</span><span class="mi">0</span><span class="p">::</span><span class="mi">3</span><span class="p">]))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{cwd}</span><span class="s1">/${samples:bn}.rsem.genes_output_list&#39;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">input_list</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">3</span><span class="p">]))</span>

<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
         <span class="n">aggregate_rsem_results</span><span class="o">.</span><span class="n">py</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samples</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">rsem</span><span class="o">.</span><span class="n">isoforms_output_list</span> <span class="p">{</span><span class="n">expected_count</span><span class="p">,</span><span class="n">TPM</span><span class="p">,</span><span class="n">FPKM</span><span class="p">,</span><span class="n">IsoPct</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnn</span><span class="p">}</span>
         <span class="n">aggregate_rsem_results</span><span class="o">.</span><span class="n">py</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samples</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">.</span><span class="n">rsem</span><span class="o">.</span><span class="n">genes_output_list</span> <span class="p">{</span><span class="n">expected_count</span><span class="p">,</span><span class="n">TPM</span><span class="p">,</span><span class="n">FPKM</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">nnn</span><span class="p">}</span> 

<span class="kn">R:</span>  <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
     <span class="n">readRSEM</span><span class="o">.</span><span class="n">cnt</span> <span class="o">&lt;-</span> <span class="n">function</span> <span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>
            <span class="c1"># RSEM .cnt files gives statistics about the (transcriptome) alignment passed to RSEM:</span>
            <span class="c1"># Row 1: N0 (# unalignable reads);</span>
            <span class="c1">#        N1 (# alignable reads);</span>
            <span class="c1">#        N2 (# filtered reads due to too many alignments);</span>
            <span class="c1">#        N_tot (N0+N1+N2)</span>
            <span class="c1"># Row 2: nUnique (# reads aligned uniquely to a gene);</span>
            <span class="c1">#        nMulti (# reads aligned to multiple genes);</span>
            <span class="c1">#        nUncertain (# reads aligned to multiple locations in the given reference sequences, which include isoform-level multi-mapping reads)</span>
            <span class="c1"># Row 3: nHits (# total alignments);</span>
            <span class="c1">#        read_type (0: single-end read, no quality; 1: single-end read, with quality score; 2: paired-end read, no quality score; 3: paired-end read, with quality score)</span>
            <span class="c1"># Source: https://groups.google.com/forum/#!topic/rsem-users/usmPKgsC5LU</span>
            <span class="c1"># Note: N1 = nUnique + nMulti</span>

            <span class="n">stopifnot</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
            <span class="n">isDir</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="err">$</span><span class="n">isdir</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">isDir</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;find&quot;</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;-name </span><span class="se">\&quot;</span><span class="s2">*.rsem.cnt</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">intern</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
                <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
                <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;_rsem.cnt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
            <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">source</span>
                <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.rsem.cnt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">.</span><span class="n">char</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
                <span class="n">metrics</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span><span class="o">=</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                <span class="n">File</span><span class="o">=</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                <span class="n">TotalReads</span><span class="o">=</span><span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
                <span class="n">AlignedReads</span><span class="o">=</span><span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
                <span class="n">UniquelyAlignedReads</span><span class="o">=</span><span class="n">m</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
                <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
            <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">metrics</span><span class="err">$</span><span class="n">Sample</span>

            <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="n">sourceRSEM</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,})</span>
        <span class="n">sourceRSEM</span> <span class="o">=</span> <span class="n">sourceRSEM</span><span class="p">[</span><span class="n">seq</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="n">length</span><span class="p">(</span><span class="n">sourceRSEM</span><span class="p">),</span><span class="mi">3</span><span class="p">)]</span>
        <span class="n">metrics</span><span class="o">.</span><span class="n">RSEM</span> <span class="o">=</span> <span class="n">readRSEM</span><span class="o">.</span><span class="n">cnt</span><span class="p">(</span><span class="n">sourceRSEM</span><span class="p">)</span>
        <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">RSEM</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{_output[-1]}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-7-summarize-with-multiqc">
<h2>Step 7: summarize with MultiQC<a class="headerlink" href="#step-7-summarize-with-multiqc" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://multiqc.info/docs/#using-multiqc">MultiQC</a></p>
<blockquote>
<div><p>MultiQC is a reporting tool that parses summary statistics from results and log files generated by other bioinformatics tools. MultiQC doesn’t run other tools for you - it’s designed to be placed at the end of analysis pipelines or to be run manually when you’ve finished running your tools. When you launch MultiQC, it recursively searches through any provided file paths and finds files that it recognises. It parses relevant information from these and generates a single stand-alone HTML report file. It also saves a directory of data files with all parsed data for further downstream use.</p>
</div></blockquote>
<p>MultiQC will automatically generate QC report for anything embedded within the given directory. Therefore providing the directory containing all the output will surfice.</p>
<p>The output of MultiQC is a multi-module report each corresponding to the quality report of each step of analysis previously performed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">rsem_call_3</span><span class="p">,</span><span class="n">rnaseqc_call_3</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.multiqc_report.html&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span>
<span class="kn">report:</span> <span class="n">output</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.multiqc_config.yml&quot;</span>
  <span class="n">extra_fn_clean_exts</span><span class="p">:</span>
      <span class="o">-</span> <span class="s1">&#39;_rsem&#39;</span>
  <span class="n">fn_ignore_dirs</span><span class="p">:</span>
      <span class="o">-</span> <span class="s1">&#39;*_STARpass1&#39;</span>
<span class="kn">bash:</span>  <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span><span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">multiqc</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">d</span><span class="p">}</span> <span class="o">-</span><span class="n">v</span> <span class="o">-</span><span class="n">n</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">b</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">d</span><span class="p">}</span> <span class="o">-</span><span class="n">c</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">multiqc_config</span><span class="o">.</span><span class="n">yml</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">rsem_call_4</span><span class="p">,</span> <span class="n">rnaseqc_call_4</span><span class="p">]</span>
<span class="c1"># Path to flat reference file, for computing QC metric</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">b</span><span class="si">}</span><span class="s1">.picard.aggregated_quality.metrics.tsv&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="c1">## Define Function</span>
    <span class="n">readPicard</span><span class="o">.</span><span class="n">alignment_summary_metrics</span> <span class="o">&lt;-</span> <span class="n">function</span> <span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
      <span class="n">stopifnot</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source</span><span class="p">))</span>
      <span class="n">isDir</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="err">$</span><span class="n">isdir</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">isDir</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;find -L&quot;</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;-name </span><span class="se">\&quot;</span><span class="s2">*.alignment_summary_metrics</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">intern</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.alignment_summary_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">source</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.alignment_summary_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>
    
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">.</span><span class="n">char</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">is_paired_end</span><span class="p">}</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span><span class="o">=</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">File</span><span class="o">=</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">PF_READS</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="err">$</span><span class="n">PF_READS</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]),</span>
                                   <span class="n">PF_READS_ALIGNED</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="err">$</span><span class="n">PF_READS_ALIGNED</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]),</span>
                                   <span class="n">PCT_PF_READS_ALIGNED</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="err">$</span><span class="n">PF_READS_ALIGNED</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">])</span><span class="o">/</span><span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="err">$</span><span class="n">PF_READS</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]),</span>
                                   <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
      <span class="p">}</span>
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
      <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">metrics</span><span class="err">$</span><span class="n">Sample</span>
    
      <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">readPicard</span><span class="o">.</span><span class="n">rna_metrics</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>

      <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
      <span class="n">stopifnot</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source</span><span class="p">))</span>
      <span class="n">isDir</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="err">$</span><span class="n">isdir</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">isDir</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;find -L&quot;</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;-name </span><span class="se">\&quot;</span><span class="s2">*.rna_metrics</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">intern</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.rna_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">source</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.rna_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>

      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">.</span><span class="n">char</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span><span class="o">=</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">File</span><span class="o">=</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">PCT_RIBOSOMAL_BASES</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_RIBOSOMAL_BASES</span><span class="p">,</span>
                                   <span class="n">PCT_CODING_BASES</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_CODING_BASES</span><span class="p">,</span>
                                   <span class="n">PCT_UTR_BASES</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_UTR_BASES</span><span class="p">,</span>
                                   <span class="n">PCT_INTRONIC_BASES</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_INTRONIC_BASES</span><span class="p">,</span>
                                   <span class="n">PCT_INTERGENIC_BASES</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_INTERGENIC_BASES</span><span class="p">,</span>
                                   <span class="n">PCT_MRNA_BASES</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_MRNA_BASES</span><span class="p">,</span>
                                   <span class="n">PCT_USABLE_BASES</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_USABLE_BASES</span><span class="p">,</span>
                                   <span class="n">MEDIAN_CV_COVERAGE</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MEDIAN_CV_COVERAGE</span><span class="p">,</span>
                                   <span class="n">MEDIAN_5PRIME_BIAS</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MEDIAN_5PRIME_BIAS</span><span class="p">,</span>
                                   <span class="n">MEDIAN_3PRIME_BIAS</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MEDIAN_3PRIME_BIAS</span><span class="p">,</span>
                                   <span class="n">MEDIAN_5PRIME_TO_3PRIME_BIAS</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MEDIAN_5PRIME_TO_3PRIME_BIAS</span><span class="p">,</span>
                                   <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
      <span class="p">}</span>
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
      <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">metrics</span><span class="err">$</span><span class="n">Sample</span>
    
      <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
    <span class="p">}</span>


    <span class="n">readPicard</span><span class="o">.</span><span class="n">duplicate_metrics</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>

      <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
      <span class="n">stopifnot</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source</span><span class="p">))</span>
      <span class="n">isDir</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="err">$</span><span class="n">isdir</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">isDir</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;find -L&quot;</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;-name </span><span class="se">\&quot;</span><span class="s2">*.Aligned.sortedByCoord.md.metrics</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">intern</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.Aligned.sortedByCoord.md.metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">source</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.Aligned.sortedByCoord.md.metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>

      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">.</span><span class="n">char</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span><span class="o">=</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">File</span><span class="o">=</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">PERCENT_DUPLICATION</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PERCENT_DUPLICATION</span><span class="p">,</span>
                                   <span class="n">ESTIMATED_LIBRARY_SIZE</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">ESTIMATED_LIBRARY_SIZE</span><span class="p">,</span>
                                   <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
      <span class="p">}</span>
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
      <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">metrics</span><span class="err">$</span><span class="n">Sample</span>
    
      <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">readPicard</span><span class="o">.</span><span class="n">wgs_metrics</span> <span class="o">&lt;-</span> <span class="n">function</span> <span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>

      <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
      <span class="n">stopifnot</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source</span><span class="p">))</span>
      <span class="n">isDir</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="err">$</span><span class="n">isdir</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">isDir</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;find -L&quot;</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;-name </span><span class="se">\&quot;</span><span class="s2">*.wgs_metrics</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">intern</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.wgs_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">source</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.wgs_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>

      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">.</span><span class="n">char</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span><span class="o">=</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">File</span><span class="o">=</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">MEDIAN_COVERAGE</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MEDIAN_COVERAGE</span><span class="p">,</span>
                                   <span class="n">MAD_COVERAGE</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MAD_COVERAGE</span><span class="p">,</span>
                                   <span class="n">PCT_EXC_DUPE</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_EXC_DUPE</span><span class="p">,</span>
                                   <span class="n">PCT_EXC_TOTAL</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">PCT_EXC_TOTAL</span><span class="p">,</span>
                                   <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
      <span class="p">}</span>
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
      <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">metrics</span><span class="err">$</span><span class="n">Sample</span>

      <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
    <span class="p">}</span>


    <span class="n">readPicard</span><span class="o">.</span><span class="n">insert_size_metrics</span> <span class="o">&lt;-</span> <span class="n">function</span> <span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>

      <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
      <span class="n">stopifnot</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source</span><span class="p">))</span>
      <span class="n">isDir</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="err">$</span><span class="n">isdir</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">isDir</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;find -L&quot;</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;-name </span><span class="se">\&quot;</span><span class="s2">*.insert_size_metrics</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">intern</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.insert_size_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">source</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.insert_size_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>

      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">.</span><span class="n">char</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span><span class="o">=</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">File</span><span class="o">=</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">MEDIAN_INSERT_SIZE</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MEDIAN_INSERT_SIZE</span><span class="p">,</span>
                                   <span class="n">MODE_INSERT_SIZE</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MODE_INSERT_SIZE</span><span class="p">,</span>
                                   <span class="n">MEDIAN_ABSOLUTE_DEVIATION</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">MEDIAN_ABSOLUTE_DEVIATION</span><span class="p">,</span>
                                   <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
      <span class="p">}</span>
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
      <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">metrics</span><span class="err">$</span><span class="n">Sample</span>
    
      <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">readPicard</span><span class="o">.</span><span class="n">gc_bias</span><span class="o">.</span><span class="n">summary_metrics</span> <span class="o">&lt;-</span> <span class="n">function</span> <span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>

      <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
      <span class="n">stopifnot</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source</span><span class="p">))</span>
      <span class="n">isDir</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="err">$</span><span class="n">isdir</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">isDir</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;find -L&quot;</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;-name </span><span class="se">\&quot;</span><span class="s2">*.gc_bias.summary_metrics</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">intern</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.gc_bias.summary_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">files</span> <span class="o">&lt;-</span> <span class="n">source</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.gc_bias.summary_metrics&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">files</span><span class="p">),</span> <span class="n">fixed</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>

      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">.</span><span class="n">char</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span><span class="o">=</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">File</span><span class="o">=</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                   <span class="n">AT_DROPOUT</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">AT_DROPOUT</span><span class="p">,</span>
                                   <span class="n">GC_DROPOUT</span><span class="o">=</span><span class="n">m</span><span class="err">$</span><span class="n">GC_DROPOUT</span><span class="p">,</span>
                                   <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
      <span class="p">}</span>
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
      <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">metrics</span><span class="err">$</span><span class="n">Sample</span>

      <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
    <span class="p">}</span>




    <span class="n">readPicard</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">source</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">metrics</span><span class="o">.</span><span class="n">aln</span> <span class="o">&lt;-</span> <span class="n">readPicard</span><span class="o">.</span><span class="n">alignment_summary_metrics</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
      <span class="n">metrics</span><span class="o">.</span><span class="n">rna</span> <span class="o">&lt;-</span> <span class="n">readPicard</span><span class="o">.</span><span class="n">rna_metrics</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
      <span class="n">metrics</span><span class="o">.</span><span class="n">dup</span> <span class="o">&lt;-</span> <span class="n">readPicard</span><span class="o">.</span><span class="n">duplicate_metrics</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

      <span class="n">stopifnot</span><span class="p">(</span><span class="nb">all</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">aln</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">rna</span><span class="p">))</span> <span class="o">&amp;</span>
                <span class="nb">all</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">rna</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">dup</span><span class="p">))</span> <span class="o">&amp;</span>
                <span class="nb">all</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">dup</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">aln</span><span class="p">)))</span>

      <span class="n">metrics</span><span class="o">.</span><span class="n">aln</span><span class="err">$</span><span class="n">File</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
      <span class="n">metrics</span><span class="o">.</span><span class="n">rna</span><span class="err">$</span><span class="n">File</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
      <span class="n">metrics</span><span class="o">.</span><span class="n">dup</span><span class="err">$</span><span class="n">File</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
      <span class="n">metrics</span><span class="o">.</span><span class="n">rna</span><span class="err">$</span><span class="n">Sample</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
      <span class="n">metrics</span><span class="o">.</span><span class="n">dup</span><span class="err">$</span><span class="n">Sample</span> <span class="o">&lt;-</span> <span class="n">NULL</span>

      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">cbind</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">aln</span><span class="p">,</span> <span class="n">metrics</span><span class="o">.</span><span class="n">rna</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">aln</span><span class="p">),</span> <span class="p">])</span>
      <span class="n">metrics</span> <span class="o">&lt;-</span> <span class="n">cbind</span><span class="p">(</span><span class="n">metrics</span><span class="p">,</span> <span class="n">metrics</span><span class="o">.</span><span class="n">dup</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">aln</span><span class="p">),</span> <span class="p">])</span>

      <span class="k">return</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1">## Execution  </span>
    <span class="n">sourcePicard</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span><span class="n">dr</span><span class="p">}</span>

    <span class="n">Picard_qualityMetrics</span> <span class="o">&lt;-</span> <span class="n">readPicard</span><span class="p">(</span><span class="n">sourcePicard</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">Picard_qualityMetrics</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/molecular_phenotypes/calling"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../bulk_expression.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">RNA-seq expression</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../QC/bulk_expression_QC.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Sample level RNA-seq quality control</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>