
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Genotype data preprocessing &#8212; ADSP-FG Project</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Genotype VCF file quality control" href="genotype/VCF_QC.html" />
    <link rel="prev" title="Normalization and phenotype table generation for splicingQTL analysis" href="../molecular_phenotypes/QC/splicing_normalization.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   ADSP-FG xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq xQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="current reference internal" href="#">
   Genotype data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/summary_stats_standardizer.html">
   Overview of design
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/data_preprocessing/genotype_preprocessing.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype_preprocessing.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/genotype_preprocessing.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#analysis-steps">
     Analysis steps
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input-data-requirement">
     Input data requirement
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#qc-for-vcf-files">
   QC for VCF files
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#optional-merged-output-of-vcf-qc">
   Optional: Merged output of VCF QC
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-overlap-samples-for-genotypes-and-phenotype-data">
   Get overlap samples for genotypes and phenotype data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#genotype-qc-for-pca">
   Genotype QC for PCA
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#principle-component-analysis">
   Principle component analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#finalized-genotype-qc">
   Finalized genotype QC
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#for-common-and-infrequent-variants">
     For common and infrequent variants
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#optional-remove-outlier">
     Optional: Remove outlier
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#optional-extract-common-variants">
     Optional: Extract common variants
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#genotype-reformatting">
   Genotype Reformatting
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#partition-by-chromosome">
     Partition by Chromosome
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#optional-convert-to-vcf">
     Optional: convert to VCF
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="genotype-data-preprocessing">
<h1>Genotype data preprocessing<a class="headerlink" href="#genotype-data-preprocessing" title="Permalink to this headline">¶</a></h1>
<p>This section documents output from the genotype section (colored in light yellow) of xqtl discovery pipeline MWE and explained the purpose for each of the command. The file used in this page can be found within the protocol_data folder of the synapse repo as described on the landing page.</p>
<p><strong>For mutli-theme study using same set of genotype (i.e. different tissues for the same participant) , the VCF QC and final Genotype QC steps are ran once. The related/unrelated genotype QC and PCA were ran once per tissue, as only the participant also have molecular phenotype data will be used for PCA</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview ../images/eqtl_command.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ../images/eqtl_command.png (254.4 KiB):</div></div><img alt="../../_images/genotype_preprocessing_1_1.png" src="../../_images/genotype_preprocessing_1_1.png" />
</div>
</div>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<div class="section" id="analysis-steps">
<h3>Analysis steps<a class="headerlink" href="#analysis-steps" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>Genotype data quality control (QC). See here for the <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/GWAS_QC.html">QC default settings</a>.</p></li>
<li><p>Principle component analysis (PCA) based QC, and PC computation for each sub-population available in the genotype data.</p></li>
<li><p>Genomic relationship matrix (GRM) computation (optional, when in Step 1 genotype QC detects related individuals or the input is known to have related individuals).</p></li>
</ol>
</div>
<div class="section" id="input-data-requirement">
<h3>Input data requirement<a class="headerlink" href="#input-data-requirement" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>Genotype data. See here for <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/genotype_formatting.html">format details</a>.</p></li>
<li><p>[Optional] a sample information file to specific population information, if external data such as HapMap or 1000 Genomes are to be integrated to the PCA analysis to visualize and assess population structure in the genotype data. See here for <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/genotype_formatting.html">format details</a>.</p></li>
</ol>
</div>
</div>
<div class="section" id="qc-for-vcf-files">
<h2>QC for VCF files<a class="headerlink" href="#qc-for-vcf-files" title="Permalink to this headline">¶</a></h2>
<p>This step extract the rsID for the known varriants, so that we can distinguish them from the novel variants we had in our study. The procedure/rationale is <a class="reference external" href="https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionVI/lessons/03_annotation-snpeff.html">explained in this post</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/VCF_QC.ipynb dbsnp_annotate <span class="se">\</span>
    --genoFile reference_data/00-All.vcf.gz <span class="se">\</span>
    --cwd output/reference_data <span class="se">\</span>
    --container containers/bioinfo.sif <span class="se">\</span>
    -J <span class="m">50</span> -c csg.yml -q csg --add_chr
</pre></div>
</div>
</div>
</div>
<p>Perform QC on VCF files. The QC-ed data will also be exported to PLINK format for next steps analysis. The QC procedure includes:</p>
<ol>
<li><p>Handling the formatting of multi-allelic sites,</p></li>
<li><p>Genotype and variant level filtering based on genotype calling qualities.</p>
<p>2.1. Genotype depth filters: For WES data, UK Biobank recommends <strong>SNPs DP&gt;10 and Indels DP&gt;10 for indels.</strong> However we think for WGS we can be less stringent, or simply rely on GQ. Users can set it to 1 eg, <code class="docutils literal notranslate"><span class="pre">--DP</span> <span class="pre">1</span> <span class="pre">--DP-indel</span> <span class="pre">1</span> </code></p>
<p>2.2. Genotype quality GQ&gt;20.</p>
<p>2.3. At least one sample per site passed the allele balance threshold &gt;= 0.15 for SNPs and &gt;=0.20 for indels (heterozygous variants). Allele balance is calculated for heterozygotes as the number of bases supporting the least-represented allele over the total number of base observations.</p>
</li>
<li><p>Known/novel variants annotation</p></li>
<li><p>Summary statistics before and after QC, in particular the ts/tv ratio, to assess the effectiveness of QC.</p></li>
</ol>
<p>3 and 4 above are for explorative analysis on the overall quality assessment of genotype data in the VCF files. We annotate known and novel variants because ts/tv are expected to be different between known and novel variants, and is important QC metric to assess the effectiveness of our QC.
After the procedure was done, the output will be converted to a plink bed/bim/fam binary format for downstream analysis.</p>
<p>It will take ~25G of memory and ~5min to complete the following step</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/VCF_QC.ipynb qc <span class="se">\</span>
            --genoFile ROSMAP_data/Genotype/DEJ_11898_B01_GRM_WGS_2017-05-15_21.recalibrated_variants.xqtl_protocol_data.add_chr.vcf.gz <span class="se">\</span>
                       ROSMAP_data/Genotype//DEJ_11898_B01_GRM_WGS_2017-05-15_22.recalibrated_variants.xqtl_protocol_data.add_chr.vcf.gz <span class="se">\</span>
            --dbsnp-variants reference_data/00-All.add_chr.variants.gz <span class="se">\</span>
            --reference-genome reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta <span class="se">\</span>
            --cwd output/genotype --container containers/bioinfo.sif 
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="optional-merged-output-of-vcf-qc">
<h2>Optional: Merged output of VCF QC<a class="headerlink" href="#optional-merged-output-of-vcf-qc" title="Permalink to this headline">¶</a></h2>
<p>To accomodate the genotype file with extra large size,we allow our user to perform QC of vcf file on a per chromomosome basis. However, to accurately perform filtering base on minor allele count, HWE deviation, and missingness, the input to GWAS QC need to be the whole genome genotype file. Therefore we provide this utlity for the user to merge their output from vcf QC</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/genotype_formatting.ipynb merge_plink <span class="se">\</span>
            --genoFile <span class="sb">`</span>ls output/genotype/DEJ_11898_B01_GRM_WGS_2017-05-15_21.recalibrated_variants*.leftnorm.filtered.bed<span class="sb">`</span> <span class="se">\</span>
            --cwd output/genotype --container containers/bioinfo.sif --name xqtl_protocol_data  
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="get-overlap-samples-for-genotypes-and-phenotype-data">
<h2>Get overlap samples for genotypes and phenotype data<a class="headerlink" href="#get-overlap-samples-for-genotypes-and-phenotype-data" title="Permalink to this headline">¶</a></h2>
<p>In most of the time, the number of samples in genotype and phenotype data will not be an exact match. Therefore, we provides a utility to produce a list of overlaps samples between genotypes and each phenotypes. Following command retrieve the overlapped sample between Genotypes and methylation data</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/sample_matcher.ipynb filtered_sample_list <span class="se">\</span>
        --cwd output/  <span class="se">\</span>
        --phenoFile output/methylation/xqtl_protocol_data_arrayMethylation_covariates.sesame.M.bed.gz  <span class="se">\</span>
        --genoFile output/genotype/xqtl_protocol_data.fam  <span class="se">\</span>
        --sample-participant-lookup reference_data/sample_participant_lookup.rnaseq <span class="se">\</span>
        --container containers/bioinfo.sif <span class="se">\</span>
        --translated_phenoFile
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="genotype-qc-for-pca">
<h2>Genotype QC for PCA<a class="headerlink" href="#genotype-qc-for-pca" title="Permalink to this headline">¶</a></h2>
<p>To accuratly estimate the PCs for the genotype. We split participants based on their kinship coefficients, estimated by king. We have implemented a graph-based method to avoid removing unrelated participants that are in a family. However, this option is disable by default due to the long running time it required. User can enable this option by changing the <code class="docutils literal notranslate"><span class="pre">--no-maximize_unrelated</span></code> to <code class="docutils literal notranslate"><span class="pre">--maximize_unrelated</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb king <span class="se">\</span>
   --cwd output/pca_methylation <span class="se">\</span>
   --genoFile output/genotype/xqtl_protocol_data.bed <span class="se">\</span>
   --keep-samples output/sample_participant_lookup.filtered_geno.txt <span class="se">\</span>
   --container containers/bioinfo.sif <span class="se">\</span>
   --walltime 48h   --no-maximize_unrelated

sos run pipeline/GWAS_QC.ipynb king <span class="se">\</span>
   --cwd output/pca <span class="se">\</span>
   --genoFile output/genotype/xqtl_protocol_data.bed <span class="se">\</span>
   --container containers/bioinfo.sif <span class="se">\</span>
   --walltime 48h   --no-maximize_unrelated
</pre></div>
</div>
</div>
</div>
<p>Variant level and sample level QC on unrelated individuals using missingness &gt; 10% ,MAF&gt;1% , and LD-prunning in preparation for PCA analysis:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc <span class="se">\</span>
   --cwd output/genotype <span class="se">\</span>
   --genoFile output/pca/xqtl_protocol_data.unrelated.bed <span class="se">\</span>
   --maf-filter <span class="m">0</span>.01 <span class="se">\</span>
   --container containers/bioinfo.sif 
</pre></div>
</div>
</div>
</div>
<p>Extract previously selected variants from related individuals in preparation for PCA, only applying missingness filter at sample level,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc_no_prune <span class="se">\</span>
   --cwd output/genotype <span class="se">\</span>
   --genoFile output/pca/xqtl_protocol_data.related.bed <span class="se">\</span>
   --maf-filter <span class="m">0</span> <span class="se">\</span>
   --geno-filter <span class="m">0</span> <span class="se">\</span>
   --mind-filter <span class="m">0</span>.1 <span class="se">\</span>
   --hwe-filter <span class="m">0</span> <span class="se">\</span>
   --keep-variants output/genotype/xqtl_protocol_data.unrelated.filtered.prune.in  <span class="se">\</span>
   --container containers/bioinfo.sif  <span class="se">\</span>
   --mem 40G
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="principle-component-analysis">
<h2>Principle component analysis<a class="headerlink" href="#principle-component-analysis" title="Permalink to this headline">¶</a></h2>
<p>Analysis below uses data from previous genotype QC.</p>
<p>First, perform PCA on unrelated individuals after QC in previous step,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span> sos run pipeline/PCA.ipynb flashpca <span class="se">\</span>
        --cwd output/pca <span class="se">\</span>
        --genoFile output/genotype/xqtl_protocol_data.unrelated.filtered.prune.bed  <span class="se">\</span>
        --container containers/flashpcaR.sif 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.pc.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.pc.png (611.8 KiB):</div></div><img alt="../../_images/genotype_preprocessing_22_1.png" src="../../_images/genotype_preprocessing_22_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.scree.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.scree.png (56.6 KiB):</div></div><img alt="../../_images/genotype_preprocessing_23_1.png" src="../../_images/genotype_preprocessing_23_1.png" />
</div>
</div>
<p>Then project back related individuals to the space spanned by PC from unrelated individuals, and detect PC outliers.</p>
<p><strong>Notice here that you need to decide how many PCs to use to compute outliers (the <code class="docutils literal notranslate"><span class="pre">--maha-k</span></code> parameter), based on the PC scree plot. You should include a decent number of PCs that capture enough variation in genotype data. Please manually examine the PC plot before running this command.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/PCA.ipynb project_samples <span class="se">\</span>
        --cwd output/pca <span class="se">\</span>
        --genoFile output/genotype/xqtl_protocol_data.related.filtered.extracted.bed  <span class="se">\</span>
        --pca-model  output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.rds <span class="se">\</span>
        --container containers/flashpcaR.sif --maha-k <span class="sb">`</span>awk <span class="s1">&#39;$3 &lt; 0.7&#39;</span> output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.scree.txt  <span class="p">|</span> tail -1 <span class="p">|</span> cut -f <span class="m">1</span>  <span class="sb">`</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.pc.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.pc.png (615.6 KiB):</div></div><img alt="../../_images/genotype_preprocessing_26_1.png" src="../../_images/genotype_preprocessing_26_1.png" />
</div>
</div>
<p>Two outliers was removed, as shown below</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.outliers
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0	SM-CJGN1
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="finalized-genotype-qc">
<h2>Finalized genotype QC<a class="headerlink" href="#finalized-genotype-qc" title="Permalink to this headline">¶</a></h2>
<div class="section" id="for-common-and-infrequent-variants">
<h3>For common and infrequent variants<a class="headerlink" href="#for-common-and-infrequent-variants" title="Permalink to this headline">¶</a></h3>
<p>In our demo, we only keep variants having sample minor allele count MAC &gt;= 5, for the sake of convenience focusing our single variant analysis. It should be noted that mac and maf don’t overwrite each other within our module. It is up to plink to decide how to handle the secnario of both <code class="docutils literal notranslate"><span class="pre">--maf-filter</span></code> and <code class="docutils literal notranslate"><span class="pre">--mac-filter</span></code> &gt; 0. Therefore, the user should not specify both <code class="docutils literal notranslate"><span class="pre">--maf-filter</span></code> and <code class="docutils literal notranslate"><span class="pre">--mac-filter</span></code> without good reason.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc_no_prune <span class="se">\</span>
   --cwd output/genotype <span class="se">\</span>
   --genoFile output/genotype/xqtl_protocol_data.bed <span class="se">\</span>
   --geno-filter <span class="m">0</span>.1 <span class="se">\</span>
   --mind-filter <span class="m">0</span>.1 <span class="se">\</span>
   --hwe-filter 1e-06   <span class="se">\</span>
   --mac-filter <span class="m">5</span> <span class="se">\</span>
   --container containers/bioinfo.sif  <span class="se">\</span>
   --mem 40G
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="optional-remove-outlier">
<h3>Optional: Remove outlier<a class="headerlink" href="#optional-remove-outlier" title="Permalink to this headline">¶</a></h3>
<p>PCA analysis (see covariate formatting page) can possibly provide us a list of outlier samples to be removed from the genotype data. We can optionally remove them and perform genotype data QC.</p>
<p>Removal of samples is a complicated decision that require human descretion. Therefore outlier removal is not performed automatically. The <code class="docutils literal notranslate"><span class="pre">--remove-samples</span></code> parameter take a two columns table that illustrated by  <code class="docutils literal notranslate"><span class="pre">output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.outliers</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc_no_prune <span class="se">\</span>
    --cwd output/genotype <span class="se">\</span>
    --genoFile output/genotype/xqtl_protocol_data.bed <span class="se">\</span>
    --geno-filter <span class="m">0</span>.1 <span class="se">\</span>
    --mind-filter <span class="m">0</span>.1 <span class="se">\</span>
    --hwe-filter 1e-06   <span class="se">\</span>
    --mac-filter <span class="m">5</span> <span class="se">\</span>
    --container containers/bioinfo.sif  <span class="se">\</span>
    --mem 40G
    --remove-samples output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.outliers
</pre></div>
</div>
</div>
</div>
<p>To fully illustrate how PCA and PCA based QC is done on sample with <strong>multiple populations (a more complicated but potentially very useful setting)</strong>, please try out the example documented in the <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/PCA.html">PCA analysis module</a> with a different example data-set.</p>
</div>
<div class="section" id="optional-extract-common-variants">
<h3>Optional: Extract common variants<a class="headerlink" href="#optional-extract-common-variants" title="Permalink to this headline">¶</a></h3>
<p>In case you would like to focus on common variants only, the genotype data can also filtered by the criteria of missingness &gt; 10% and MAF&gt;5% as shown below to retain the common variants, User can detemined the level of filtering by changing the generated command manually.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc_no_prune <span class="se">\</span>
   --cwd output/genotype <span class="se">\</span>
   --genoFile output/genotype/xqtl_protocol_data.bed <span class="se">\</span>
   --geno-filter <span class="m">0</span>.1 <span class="se">\</span>
   --mind-filter <span class="m">0</span>.1 <span class="se">\</span>
   --hwe-filter 1e-06   <span class="se">\</span>
   --maf-filter <span class="m">0</span>.05 <span class="se">\</span>
   --container containers/bioinfo.sif  <span class="se">\</span>
   --mem 40G
</pre></div>
</div>
</div>
</div>
<p>If outliers are to be removed, following command could be used:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/GWAS_QC.ipynb qc_no_prune <span class="se">\</span>
   --cwd output/genotype <span class="se">\</span>
   --genoFile output/genotype/xqtl_protocol_data.bed <span class="se">\</span>
   --geno-filter <span class="m">0</span>.1 <span class="se">\</span>
   --mind-filter <span class="m">0</span>.1 <span class="se">\</span>
   --hwe-filter 1e-06   <span class="se">\</span>
   --maf-filter <span class="m">0</span>.05 <span class="se">\</span>
   --container containers/bioinfo.sif  <span class="se">\</span>
   --mem 40G 
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="genotype-reformatting">
<h2>Genotype Reformatting<a class="headerlink" href="#genotype-reformatting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="partition-by-chromosome">
<h3>Partition by Chromosome<a class="headerlink" href="#partition-by-chromosome" title="Permalink to this headline">¶</a></h3>
<p>The association analysis of TensorQTL required the chromosome be seperated by chromosome to allowed for parallel analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/genotype_formatting.ipynb plink_by_chrom <span class="se">\</span>
    --genoFile output/genotype/xqtl_protocol_data.filtered.bed <span class="se">\</span>
    --cwd output/data_preprocessing <span class="se">\</span>
    --chrom  <span class="sb">`</span>cut -f <span class="m">1</span>  output/genotype/xqtl_protocol_data.filtered.bim  <span class="p">|</span> uniq <span class="p">|</span> sed <span class="s2">&quot;s/chr//g&quot;</span><span class="sb">`</span> <span class="se">\</span>
    --container containers/bioinfo.sif 
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="optional-convert-to-vcf">
<h3>Optional: convert to VCF<a class="headerlink" href="#optional-convert-to-vcf" title="Permalink to this headline">¶</a></h3>
<p>If association analysis by APEX is needed, following command convert the per-chromosome plink binaray generated above into indexed vcf.gz.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/genotype_formatting.ipynb plink_to_vcf <span class="se">\</span>
    --genoFile plink_files_list.txt  <span class="se">\</span>
    --cwd output/data_preprocessing <span class="se">\</span>
    --container containers/bioinfo.sif 
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="toctree-wrapper compound">
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "bash"
        },
        kernelOptions: {
            kernelName: "bash",
            path: "./code/data_preprocessing"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'bash'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../molecular_phenotypes/QC/splicing_normalization.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Normalization and phenotype table generation for splicingQTL analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="genotype/VCF_QC.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genotype VCF file quality control</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021+, ADSP-FG xQTL methods and data integration working group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>