
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Reference data standardization &#8212; ADSP-FG Project</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="RNA-seq expression" href="../molecular_phenotypes/bulk_expression.html" />
    <link rel="prev" title="Bulk RNA-seq xQTL analysis" href="../commands_generator/eQTL_analysis_commands.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   ADSP-FG xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq xQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/summary_stats_standardizer.html">
   Overview of design
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/data_preprocessing/reference_data.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/reference_data.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/reference_data.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reference-files-for-rna-seq-expression-quantification">
     Reference files for RNA-seq expression quantification
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reference-files-for-methylation">
     Reference files for Methylation
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reference-files-for-alternative-splicing">
     Reference files for alternative splicing
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#details">
   Details
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#convert-transcript-feature-file-gff3-to-gtf">
     Convert transcript feature file gff3 to gtf
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#collapse-transcript-features-into-genes">
     Collapse transcript features into genes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generate-star-index-based-on-gtf-and-reference-fasta">
     Generate STAR index based on gtf and reference fasta
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generate-rsem-index-based-on-gtf-and-reference-fasta">
     Generate RSEM index based on gtf and reference fasta
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#list-of-commands-to-prepare-reference-data">
   List of commands to prepare reference data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-download">
   Data download
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#gff3-to-gtf-formatting">
   GFF3 to GTF formatting
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#hg-reference-file-preprocessing">
   HG reference file preprocessing
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#transcript-and-gene-model-reference-processing">
   Transcript and gene model reference processing
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#generating-index-file-for-star">
   Generating index file for
   <code class="docutils literal notranslate">
    <span class="pre">
     STAR
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-inputs">
     Step Inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-output">
     Step Output
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#generating-index-file-for-rsem">
   Generating index file for
   <code class="docutils literal notranslate">
    <span class="pre">
     RSEM
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Step Inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-outputs">
     Step Outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#generation-of-refflat-file">
   Generation of RefFlat file
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#generation-of-suppa-annotation-for-psichomics">
   Generation of SUPPA annotation for psichomics.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#generate-psichomics-hg38-splicing-annotation">
   Generate psichomics Hg38 splicing annotation
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="reference-data-standardization">
<h1>Reference data standardization<a class="headerlink" href="#reference-data-standardization" title="Permalink to this headline">¶</a></h1>
<p>This module provides reference data download, indexing and preprocessing (if necessary), in preparation for use throughout the pipeline.</p>
<p>We have included the PDF document compiled by Data Standardization Working Group in the <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn36416587">on Synapse</a> as well as on <a class="reference external" href="https://www.niagads.org/adsp/content/adspgcadgenomeresources-v2pdf">ADSP Dashboard</a>. It contains the reference data to use for the project.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This module is based on the <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md">TOPMed workflow from Broad</a>. The reference data after we process it (details see Methods section and the rest of the analysis) can be found <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn36416587">in this folder on Synapse</a>.</p>
<div class="section" id="reference-files-for-rna-seq-expression-quantification">
<h3>Reference files for RNA-seq expression quantification<a class="headerlink" href="#reference-files-for-rna-seq-expression-quantification" title="Permalink to this headline">¶</a></h3>
<p><strong>We have decided to use these preprocessed reference files for RNA-seq expression and splicing quantification.</strong> Specifically:</p>
<ol class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.{dict,fasta,fasta.fai}</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf</span></code> for stranded protocol, and <code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.103.chr.reformatted.gene.ERCC.gtf</span></code> for unstranded protocol.</p></li>
<li><p>Everything under <code class="docutils literal notranslate"><span class="pre">STAR_Index</span></code> folder</p></li>
<li><p>Everything under <code class="docutils literal notranslate"><span class="pre">RSEM_Index</span></code> folder</p></li>
<li><p>Optionally, for quality control, <code class="docutils literal notranslate"><span class="pre">gtf_ref.flat</span></code></p></li>
</ol>
</div>
<div class="section" id="reference-files-for-methylation">
<h3>Reference files for Methylation<a class="headerlink" href="#reference-files-for-methylation" title="Permalink to this headline">¶</a></h3>
<p>FIXME &#64;hao</p>
</div>
<div class="section" id="reference-files-for-alternative-splicing">
<h3>Reference files for alternative splicing<a class="headerlink" href="#reference-files-for-alternative-splicing" title="Permalink to this headline">¶</a></h3>
<p>FIXME &#64;xuanhe</p>
</div>
</div>
<div class="section" id="details">
<h2>Details<a class="headerlink" href="#details" title="Permalink to this headline">¶</a></h2>
<p>Workflows implemented include:</p>
<div class="section" id="convert-transcript-feature-file-gff3-to-gtf">
<h3>Convert transcript feature file gff3 to gtf<a class="headerlink" href="#convert-transcript-feature-file-gff3-to-gtf" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Input: an uncompressed gff3 file.(i.e. can be view via cat)</p></li>
<li><p>Output: a gtf file.</p></li>
</ul>
</div>
<div class="section" id="collapse-transcript-features-into-genes">
<h3>Collapse transcript features into genes<a class="headerlink" href="#collapse-transcript-features-into-genes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Input: a gtf file.</p></li>
<li><p>Output: a gtf file with collapesed gene model.</p></li>
</ul>
</div>
<div class="section" id="generate-star-index-based-on-gtf-and-reference-fasta">
<h3>Generate STAR index based on gtf and reference fasta<a class="headerlink" href="#generate-star-index-based-on-gtf-and-reference-fasta" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Input: a gtf file and an acompanying fasta file.</p></li>
<li><p>Output: A folder of STAR index.</p></li>
</ul>
</div>
<div class="section" id="generate-rsem-index-based-on-gtf-and-reference-fasta">
<h3>Generate RSEM index based on gtf and reference fasta<a class="headerlink" href="#generate-rsem-index-based-on-gtf-and-reference-fasta" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Input: a gtf file and an acompanying fasta file.</p></li>
<li><p>Output: A folder of RSEM index.</p></li>
</ul>
</div>
</div>
<div class="section" id="list-of-commands-to-prepare-reference-data">
<h2>List of commands to prepare reference data<a class="headerlink" href="#list-of-commands-to-prepare-reference-data" title="Permalink to this headline">¶</a></h2>
<p>To download reference data,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_hg_reference</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_gene_annotation</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_ercc_reference</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_dbsnp</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: an hour, depending on the network.</p></li>
</ul>
<p>To format reference data,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">hg_reference</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">ercc</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">ERCC92</span><span class="o">.</span><span class="n">fa</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">fa</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">hg_gtf</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">stranded</span>
</pre></div>
</div>
</div>
</div>
<p>To format gene feature data:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">gene_annotation</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">ercc</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">ERCC92</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">stranded</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: 10min</p></li>
<li><p>Memory: 16GB</p></li>
</ul>
<p><strong>Notice that for un-stranded RNA-seq protocol please use switch <code class="docutils literal notranslate"><span class="pre">--no-stranded</span></code> to the command above instead of <code class="docutils literal notranslate"><span class="pre">--stranded</span></code>. More details can be found later in the document.</strong></p>
<p>Generating STAR index without the GTF annotation file allow customize read lenght lateron in STAR alignment.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">STAR_index</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="kp">mem</span> <span class="mi">40</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: 30 min</p></li>
<li><p>Memory: 40 GB</p></li>
</ul>
<p>To generate RSEM index with the gtf file <strong>prior</strong> to the gene collapsing step ( <strong>without</strong> the gene tag in its file name.)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">RSEM_index</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>  
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: 1 min</p></li>
</ul>
<p>To generate RefFlat annotation for Picard QC,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">RefFlat_generation</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<p>To generate the SUPPA annotation for psichomics to detect RNA alternative splicing events,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">SUPPA_annotation</span> \
    <span class="o">--</span><span class="n">hg_gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">psichomics</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>Approximate time: ? min</p></li>
</ul>
</div>
<div class="section" id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run reference_data.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  download_hg_reference
  download_gene_annotation
  download_ercc_reference
  download_dbsnp
  gff3_to_gtf
  hg_reference
  hg_gtf
  ercc_gtf
  gene_annotation
  STAR_index
  RSEM_index
  RefFlat_generation
  SUPPA_annotation
  psi_hg38_annotation_modification

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files.
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option

Sections
  download_hg_reference:
  download_gene_annotation:
  download_ercc_reference:
  download_dbsnp:
  gff3_to_gtf:
    Workflow Options:
      --gff3-file VAL (as path, required)
  hg_reference_1:
    Workflow Options:
      --hg-reference VAL (as path, required)
                        Path to HG reference file
  hg_reference_2:
    Workflow Options:
      --ercc-reference VAL (as path, required)
  hg_reference_3:
  hg_gtf_1:
    Workflow Options:
      --hg-reference VAL (as path, required)
      --hg-gtf VAL (as path, required)
  hg_gtf_2:
    Workflow Options:
      --[no-]stranded (required)
  ercc_gtf:
    Workflow Options:
      --ercc-gtf VAL (as path, required)
  gene_annotation:
  STAR_index:
    Workflow Options:
      --hg-reference VAL (as path, required)
  RSEM_index:
    Workflow Options:
      --hg-gtf VAL (as path, required)
      --hg-reference VAL (as path, required)
  RefFlat_generation:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  SUPPA_annotation_1:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  SUPPA_annotation_2:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  psi_hg38_annotation_modification_1:
    Workflow Options:
      --hg-gtf VAL (as path, required)
      --hgrc-db VAL (as path, required)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># The output directory for generated files.</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="data-download">
<h2>Data download<a class="headerlink" href="#data-download" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_hg_reference</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/GRCh38_full_analysis_set_plus_decoy_hla.fa&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="mi">1000</span><span class="n">genomes</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">ftp</span><span class="o">/</span><span class="n">technical</span><span class="o">/</span><span class="n">reference</span><span class="o">/</span><span class="n">GRCh38_reference_genome</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_gene_annotation</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/Homo_sapiens.GRCh38.103.chr.gtf&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">decompress</span><span class="o">=</span><span class="kc">True</span>
    <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ensembl</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">pub</span><span class="o">/</span><span class="n">release</span><span class="o">-</span><span class="mi">103</span><span class="o">/</span><span class="n">gtf</span><span class="o">/</span><span class="n">homo_sapiens</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_ercc_reference</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/ERCC92.gtf&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/ERCC92.fa&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">decompress</span><span class="o">=</span><span class="kc">True</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">tools</span><span class="o">.</span><span class="n">thermofisher</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">content</span><span class="o">/</span><span class="n">sfs</span><span class="o">/</span><span class="n">manuals</span><span class="o">/</span><span class="n">ERCC92</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">download_dbsnp</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/00-All.vcf.gz&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/00-All.vcf.gz.tbi&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">snp</span><span class="o">/</span><span class="n">organisms</span><span class="o">/</span><span class="n">human_9606</span><span class="o">/</span><span class="n">VCF</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">snp</span><span class="o">/</span><span class="n">organisms</span><span class="o">/</span><span class="n">human_9606</span><span class="o">/</span><span class="n">VCF</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">tbi</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="gff3-to-gtf-formatting">
<h2>GFF3 to GTF formatting<a class="headerlink" href="#gff3-to-gtf-formatting" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gff3_to_gtf</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">gff3_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">gff3_file</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.gtf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="n">gffread</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">-</span><span class="n">T</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="hg-reference-file-preprocessing">
<h2>HG reference file preprocessing<a class="headerlink" href="#hg-reference-file-preprocessing" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Remove the HLA/ALT/Decoy record from the fasta – because none of the downstreams RNA-seq calling pipeline component can handle them properly.</p></li>
<li><p>Adding in ERCC information to the fasta file – even if ERCC is not included in the RNA-seq library it does not harm to add them.</p></li>
<li><p>Generating index for the fasta file</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hg_reference_1</span> <span class="p">(</span><span class="n">HLA</span> <span class="n">ALT</span> <span class="n">Decoy</span> <span class="n">removal</span><span class="p">)]</span>
<span class="c1"># Path to HG reference file</span>
<span class="kn">parameter:</span> <span class="n">hg_reference</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">hg_reference</span>
<span class="kn">output:</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.noALT_noHLA_noDecoy.fasta&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta</span><span class="p">:</span>
        <span class="n">contigs</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
        <span class="n">contigs</span> <span class="o">=</span> <span class="n">contigs</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
        <span class="n">contig_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">contigs</span><span class="p">]</span>

        <span class="c1"># exclude ALT, HLA and decoy contigs</span>
        <span class="n">filtered_fasta</span> <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">c</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">contig_ids</span><span class="p">,</span> <span class="n">contigs</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="o">-</span><span class="mi">4</span><span class="p">:]</span><span class="o">==</span><span class="s1">&#39;_alt&#39;</span> <span class="ow">or</span> <span class="n">i</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;HLA&#39;</span> <span class="ow">or</span> <span class="n">i</span><span class="p">[</span><span class="o">-</span><span class="mi">6</span><span class="p">:]</span><span class="o">==</span><span class="s1">&#39;_decoy&#39;</span><span class="p">)])</span>
    
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta</span><span class="p">:</span>
        <span class="n">fasta</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">filtered_fasta</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hg_reference_2</span> <span class="p">(</span><span class="n">merge</span> <span class="k">with</span> <span class="n">ERCC</span> <span class="n">reference</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">ercc_reference</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">_ERCC.fasta&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">sed</span> <span class="s1">&#39;s/ERCC-/ERCC_/g&#39;</span> <span class="err">$</span><span class="p">{</span><span class="n">ercc_reference</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">ercc_reference</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">patched</span><span class="o">.</span><span class="n">fa</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">ercc_reference</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">patched</span><span class="o">.</span><span class="n">fa</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hg_reference_3</span> <span class="p">(</span><span class="n">index</span> <span class="n">the</span> <span class="n">fasta</span> <span class="n">file</span><span class="p">)]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.dict&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">samtools</span> <span class="n">faidx</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">jar</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">picard</span><span class="o">-</span><span class="n">tools</span><span class="o">/</span><span class="n">picard</span><span class="o">.</span><span class="n">jar</span> \
        <span class="n">CreateSequenceDictionary</span> \
        <span class="n">R</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> \
        <span class="n">O</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="transcript-and-gene-model-reference-processing">
<h2>Transcript and gene model reference processing<a class="headerlink" href="#transcript-and-gene-model-reference-processing" title="Permalink to this headline">¶</a></h2>
<p>This step modify the <code class="docutils literal notranslate"><span class="pre">gtf</span></code> file for following reasons:</p>
<ol class="simple">
<li><p>RSEM require GTF input to have the same chromosome name format (with <code class="docutils literal notranslate"><span class="pre">chr</span></code> prefix) as the fasta file. <strong>Although for STAR, this problem can be solved by the <code class="docutils literal notranslate"><span class="pre">--sjdbGTFchrPrefix</span> <span class="pre">&quot;chr&quot;</span></code> option, we have to add <code class="docutils literal notranslate"><span class="pre">chr</span></code> to it for use with RSEM anyways.</strong> That is why we would like to do it here.</p></li>
<li><p>Gene model collapsing script <code class="docutils literal notranslate"><span class="pre">collapse_annotation.py</span></code> from GTEx require the gtf have <code class="docutils literal notranslate"><span class="pre">transcript_type</span></code> instead <code class="docutils literal notranslate"><span class="pre">transcript_biotype</span></code> in its annotation. We rename it here, although <strong>this problem can also be solved by modifying the collapse_annotation.py while building the docker, since we are already customizing the reference file for 1 above, we add this in as well, as another customization made.</strong></p></li>
<li><p>Adding in ERCC information to the <code class="docutils literal notranslate"><span class="pre">gtf</span></code> reference.</p></li>
</ol>
<p>We may reimplement 1 and 2 with the alternative approaches discussed above if the problem with RSEM is solved, or when RSEM is no longer needed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hg_gtf_1</span> <span class="p">(</span><span class="n">add</span> <span class="nb">chr</span> <span class="n">prefix</span> <span class="n">to</span> <span class="n">gtf</span> <span class="n">file</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">hg_reference</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">hg_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">hg_reference</span><span class="p">,</span> <span class="n">hg_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.reformatted.gtf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;stringr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">options</span><span class="p">(</span><span class="n">scipen</span> <span class="o">=</span> <span class="mi">999</span><span class="p">)</span>
    <span class="n">con</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input[0]}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
    <span class="n">fasta</span> <span class="o">&lt;-</span> <span class="n">readLines</span><span class="p">(</span><span class="n">con</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">close</span><span class="p">(</span><span class="n">con</span><span class="p">)</span>
    <span class="n">gtf</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input[1]}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>  <span class="n">col_names</span>  <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">col_types</span><span class="o">=</span><span class="s2">&quot;ccccccccc&quot;</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="err">!</span><span class="n">str_detect</span><span class="p">(</span><span class="n">fasta</span><span class="p">,</span><span class="s2">&quot;&gt;chr&quot;</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">gtf_mod</span> <span class="o">=</span> <span class="n">gtf</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">X1</span> <span class="o">=</span> <span class="n">str_remove_all</span><span class="p">(</span><span class="n">X1</span><span class="p">,</span><span class="s2">&quot;chr&quot;</span><span class="p">))</span>
    <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="nb">any</span><span class="p">(</span><span class="n">str_detect</span><span class="p">(</span><span class="n">gtf</span><span class="err">$</span><span class="n">X1</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;chr&quot;</span><span class="p">)))</span> <span class="p">{</span>
        <span class="n">gtf_mod</span> <span class="o">=</span> <span class="n">gtf</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">X1</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="n">X1</span><span class="p">))</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">(</span><span class="n">gtf_mod</span> <span class="o">=</span> <span class="n">gtf</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">any</span><span class="p">(</span><span class="n">str_detect</span><span class="p">(</span><span class="n">gtf_mod</span><span class="err">$</span><span class="n">X9</span><span class="p">,</span> <span class="s2">&quot;transcript_biotype&quot;</span><span class="p">)))</span> <span class="p">{</span>
      <span class="n">gtf_mod</span> <span class="o">=</span> <span class="n">gtf_mod</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">X9</span> <span class="o">=</span> <span class="n">str_replace_all</span><span class="p">(</span><span class="n">X9</span><span class="p">,</span><span class="s2">&quot;transcript_biotype&quot;</span><span class="p">,</span><span class="s2">&quot;transcript_type&quot;</span><span class="p">))</span>
    <span class="p">}</span>
    <span class="n">gtf_mod</span><span class="o">%&gt;%</span><span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">quote</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span><span class="n">col</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span><span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Text below is taken from <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/tree/master/gene_model">https://github.com/broadinstitute/gtex-pipeline/tree/master/gene_model</a></strong></p>
<p>Gene-level expression and eQTLs from the GTEx project are calculated based on a collapsed gene model (i.e., combining all isoforms of a gene into a single transcript), according to the following rules:</p>
<ol class="simple">
<li><p>Transcripts annotated as “retained_intron” or “read_through” are excluded. Additionally, transcripts that overlap with annotated read-through transcripts may be blacklisted (blacklists for GENCODE v19, 24 &amp; 25 are provided in this repository; no transcripts were blacklisted for v26).</p></li>
<li><p>The union of all exon intervals of each gene is calculated.</p></li>
<li><p>Overlapping intervals between genes are excluded from all genes.</p></li>
</ol>
<p>The purpose of step 3 is primarily to exclude overlapping regions from genes annotated on both strands, which can’t be unambiguously quantified from unstranded RNA-seq (GTEx samples were sequenced using an unstranded protocol). For stranded protocols, this step can be skipped by adding the <code class="docutils literal notranslate"><span class="pre">--collapse_only</span></code> flag.</p>
<p>Further documentation is available on the <a class="reference external" href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">GTEx Portal</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">hg_gtf_2</span> <span class="p">(</span><span class="n">collapsed</span> <span class="n">gene</span> <span class="n">model</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">stranded</span> <span class="o">=</span> <span class="nb">bool</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}{</span><span class="s2">&quot;.collapse_only&quot;</span> <span class="k">if</span> <span class="n">stranded</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.gene.gtf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">collapse_annotation</span><span class="o">.</span><span class="n">py</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;--collapse_only&quot;</span> <span class="k">if</span> <span class="n">stranded</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ercc_gtf</span> <span class="p">(</span><span class="n">Preprocess</span> <span class="n">ERCC</span> <span class="n">gtf</span> <span class="n">file</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">ercc_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">ercc_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.genes.patched.gtf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">exon_gtf</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_output}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gene_gtf</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">exon_gtf</span><span class="p">:</span>
            <span class="n">f</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>  <span class="c1"># required for RNA-SeQC/GATK (no &#39;-&#39; in contig name)</span>
        
            <span class="n">attr</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">attr</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;;&#39;</span><span class="p">:</span>
                <span class="n">attr</span> <span class="o">=</span> <span class="n">attr</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">attr</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">attr</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;; &#39;</span><span class="p">)])</span>
            <span class="c1"># add gene_name, gene_type</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ercc_control&#39;</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;KNOWN&#39;</span>
            <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;level&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;type&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;status&#39;</span><span class="p">]:</span>
                <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;transcript_&#39;</span><span class="o">+</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;gene_&#39;</span><span class="o">+</span><span class="n">k</span><span class="p">]</span>
        
            <span class="n">attr_str</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript_id&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_type&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_status&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span>
                <span class="s1">&#39;transcript_type&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript_status&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript_name&#39;</span><span class="p">]:</span>
                <span class="n">attr_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0:s}</span><span class="s1"> &quot;</span><span class="si">{1:s}</span><span class="s1">&quot;;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">attr</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span>
            <span class="n">attr_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0:s}</span><span class="s1"> </span><span class="si">{1:d}</span><span class="s1">;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;level&#39;</span><span class="p">,</span> <span class="n">attr</span><span class="p">[</span><span class="s1">&#39;level&#39;</span><span class="p">]))</span>
            <span class="n">f</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">attr_str</span><span class="p">)</span>
        
            <span class="c1"># write gene, transcript, exon</span>
            <span class="n">gene_gtf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">f</span><span class="p">[</span><span class="mi">3</span><span class="p">:])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">gene_gtf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;transcript&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">f</span><span class="p">[</span><span class="mi">3</span><span class="p">:])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">f</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">attr_str</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">gene_gtf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="s1">&#39;exon&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">f</span><span class="p">[</span><span class="mi">3</span><span class="p">:])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene_annotation</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;hg_gtf_1&quot;</span><span class="p">),</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;hg_gtf_2&quot;</span><span class="p">),</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;ercc_gtf&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.ERCC.gtf&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.ERCC.gtf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="generating-index-file-for-star">
<h2>Generating index file for <code class="docutils literal notranslate"><span class="pre">STAR</span></code><a class="headerlink" href="#generating-index-file-for-star" title="Permalink to this headline">¶</a></h2>
<p>This step generate the index file for STAR alignment. This file just need to generate once and can be re-used.</p>
<p><strong>At least 40GB of memory is needed</strong>.</p>
<div class="section" id="step-inputs">
<h3>Step Inputs<a class="headerlink" href="#step-inputs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code> and <code class="docutils literal notranslate"><span class="pre">fasta</span></code>: path to reference sequence. Both of them needs to be unzipped. <code class="docutils literal notranslate"><span class="pre">gtf</span></code> should be the one prior to collapse by gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sjdbOverhang</span></code>: specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads. We use 100 here as recommended by the TOPMed pipeline. See here for <a class="reference external" href="https://groups.google.com/g/rna-star/c/h9oh10UlvhI/m/BfSPGivUHmsJ">some additional discussions</a>.</p></li>
</ul>
</div>
<div class="section" id="step-output">
<h3>Step Output<a class="headerlink" href="#step-output" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Indexing file stored in <code class="docutils literal notranslate"><span class="pre">{cwd}/STAR_index</span></code>, which will be used by <code class="docutils literal notranslate"><span class="pre">STAR</span></code></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">STAR_index</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">hg_reference</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads.</span>
<span class="c1"># Default choice follows from TOPMed pipeline recommendation.</span>
<span class="k">if</span> <span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">expand_size</span><span class="p">(</span><span class="s1">&#39;40G&#39;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Insufficent memory for STAR, changing to 40G&quot;</span><span class="p">)</span>
    <span class="n">star_mem</span> <span class="o">=</span> <span class="s1">&#39;40G&#39;</span>
<span class="kn">else:</span>
    <span class="n">star_mem</span> <span class="o">=</span> <span class="kp">mem</span>
<span class="kn">input:</span> <span class="n">hg_reference</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/chrName.txt&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/SAindex&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/SA&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/genomeParameters.txt&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/chrStart.txt&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/chrLength.txt&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/Genome&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/STAR_Index/chrNameLength.txt&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bd</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> \
         <span class="o">--</span><span class="n">genomeDir</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span> \
         <span class="o">--</span><span class="n">genomeFastaFiles</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> \
         <span class="o">--</span><span class="n">runThreadN</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="generating-index-file-for-rsem">
<h2>Generating index file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code><a class="headerlink" href="#generating-index-file-for-rsem" title="Permalink to this headline">¶</a></h2>
<p>This step generate the indexing file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code>. This file just need to generate once.</p>
<div class="section" id="id1">
<h3>Step Inputs<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code> and <code class="docutils literal notranslate"><span class="pre">fasta</span></code>: path to reference sequence. <code class="docutils literal notranslate"><span class="pre">gtf</span></code> should be the one prior to collapse by gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sjdbOverhang</span></code>: specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads.</p></li>
</ul>
</div>
<div class="section" id="step-outputs">
<h3>Step Outputs<a class="headerlink" href="#step-outputs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Indexing file stored in <code class="docutils literal notranslate"><span class="pre">RSEM_index_dir</span></code>, which will be used by <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">RSEM_index</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">hg_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">hg_reference</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">hg_reference</span><span class="p">,</span> <span class="n">hg_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.n2g.idx.fa&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.grp&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.idx.fa&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.ti&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.chrlist&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.seq&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.transcripts.fa&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bd</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">rsem</span><span class="o">-</span><span class="n">prepare</span><span class="o">-</span><span class="n">reference</span> \
            <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> \
            <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
            <span class="o">--</span><span class="n">gtf</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> \
            <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="generation-of-refflat-file">
<h2>Generation of RefFlat file<a class="headerlink" href="#generation-of-refflat-file" title="Permalink to this headline">¶</a></h2>
<p>This file is needed for picard CollectRnaSeqMetrics module, which in turn</p>
<blockquote>
<div><p>produces metrics describing the distribution of the bases within the transcripts. It calculates the total numbers and the fractions of nucleotides within specific genomic regions including untranslated regions (UTRs), introns, intergenic sequences (between discrete genes), and peptide-coding sequences (exons). This tool also determines the numbers of bases that pass quality filters that are specific to Illumina data (PF_BASES).</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">RefFlat_generation</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">hg_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">hg_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.ref.flat&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bd</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">gtfToGenePred</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span> <span class="o">-</span><span class="n">genePredExt</span> <span class="o">-</span><span class="n">geneNameAsName2</span>
    <span class="n">awk</span> <span class="o">-</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">-</span><span class="n">v</span> <span class="n">OFS</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="s1">&#39;{$1=$12 OFS $1;}7&#39;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="o">-</span><span class="mi">11</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
    <span class="n">rm</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span><span class="o">.</span><span class="n">tmp</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="generation-of-suppa-annotation-for-psichomics">
<h2>Generation of SUPPA annotation for psichomics.<a class="headerlink" href="#generation-of-suppa-annotation-for-psichomics" title="Permalink to this headline">¶</a></h2>
<p>The generation of custom alternative splicing annotation is based on <a class="reference external" href="https://rpubs.com/nuno-agostinho/preparing-AS-annotation">this tutorial</a>. The way to generate the local alternative splicing suppa output is documented on <a class="reference external" href="https://github.com/comprna/SUPPA">SUPPA github page</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">SUPPA_annotation</span> \
    <span class="o">--</span><span class="n">hg_gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">psochimics</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">SUPPA_annotation_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">hg_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">hg_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/hg38.</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">_SE_strict.ioe&#39;</span> <span class="c1"># The stderr file must not shared the same start with the output file</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">python</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">SUPPA</span><span class="o">/</span><span class="n">suppa</span><span class="o">.</span><span class="n">py</span> <span class="n">generateEvents</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="n">hg38</span><span class="o">.</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span> <span class="o">-</span><span class="n">f</span> <span class="n">ioe</span> <span class="o">-</span><span class="n">e</span> <span class="n">SE</span> <span class="n">SS</span> <span class="n">MX</span> <span class="n">RI</span> <span class="n">FL</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">SUPPA_annotation_2</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">hg_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">hg_gtf</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.SUPPA_annotation.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;psichomics&quot;</span><span class="p">)</span>
    <span class="n">suppa</span> <span class="o">&lt;-</span> <span class="n">parseSuppaAnnotation</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input:d}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">)</span> 
    <span class="n">annot</span> <span class="o">&lt;-</span> <span class="n">prepareAnnotationFromEvents</span><span class="p">(</span><span class="n">suppa</span><span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">annot</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="generate-psichomics-hg38-splicing-annotation">
<h2>Generate psichomics Hg38 splicing annotation<a class="headerlink" href="#generate-psichomics-hg38-splicing-annotation" title="Permalink to this headline">¶</a></h2>
<p>Since the original annotation provided by psichomics package is using gene symbols, we modified it to use Ensembl IDs. The modified annotation will be used for <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/molecular_phenotypes/calling/splicing_calling.html">detecting RNA alternative splicing using psichomics</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">psi_hg38_annotation</span> \
    <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformated</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">hgrc</span><span class="o">-</span><span class="n">db</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">hgnc_database</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">psichomics</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">psi_hg38_annotation</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">hg_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># FIXME: Please document what this file is and where do we get it @xuanhe.</span>
<span class="kn">parameter:</span> <span class="n">hgrc_db</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">hg_gtf</span><span class="p">,</span> <span class="n">hgrc_db</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/psichomics_hg38_annotation.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;psichomics&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tidyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;data.table&quot;</span><span class="p">)</span>
  
    <span class="c1"># load psicomics default annotation, option of hg38 from listSplicingAnnotations()</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span> <span class="n">loadAnnotation</span><span class="p">(</span><span class="s2">&quot;AH63657&quot;</span><span class="p">)</span>

    
    <span class="c1"># reduce the demension of annotation file</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span> 
      <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
                       <span class="n">tidyr</span><span class="p">::</span><span class="n">unnest</span><span class="p">(</span><span class="n">cols</span> <span class="o">=</span> <span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">))</span>
  
    <span class="c1"># Create empty colomns for each event for easier mapping</span>
    <span class="n">annotation</span><span class="p">[[</span><span class="s2">&quot;Tandem UTR&quot;</span><span class="p">]][[</span><span class="s2">&quot;SUPPA.Event.ID&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">NA</span>
    <span class="n">annotation</span><span class="p">[[</span><span class="s2">&quot;Tandem UTR&quot;</span><span class="p">]][[</span><span class="s2">&quot;VAST-TOOLS.Event.ID&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">NA</span>
    <span class="n">annotation</span><span class="p">[[</span><span class="s2">&quot;Alternative first exon&quot;</span><span class="p">]][[</span><span class="s2">&quot;VAST-TOOLS.Event.ID&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">NA</span>
    <span class="n">annotation</span><span class="p">[[</span><span class="s2">&quot;Alternative last exon&quot;</span><span class="p">]][[</span><span class="s2">&quot;VAST-TOOLS.Event.ID&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">NA</span>
    <span class="n">annotation</span><span class="p">[[</span><span class="s2">&quot;Mutually exclusive exon&quot;</span><span class="p">]][[</span><span class="s2">&quot;VAST-TOOLS.Event.ID&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">NA</span>
  
    <span class="c1"># extract Ensembl ID substring from original SUPPA.ID and VASTTOOL.ID</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span> 
      <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="n">ENSG</span><span class="o">.</span><span class="n">SUPPA</span> <span class="o">=</span> <span class="n">substr</span><span class="p">(</span><span class="err">`</span><span class="n">SUPPA</span><span class="o">.</span><span class="n">Event</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="n">ENSG</span><span class="o">.</span><span class="n">VAST</span> <span class="o">=</span> <span class="n">substr</span><span class="p">(</span><span class="err">`</span><span class="n">VAST</span><span class="o">-</span><span class="n">TOOLS</span><span class="o">.</span><span class="n">Event</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">)))</span>
  
    <span class="c1"># Load gtf file</span>
    <span class="n">gtf_sample</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input[0]}</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
  
    <span class="c1"># from the gtf file, seperate gene names and corresponding Ensembl ID</span>
    <span class="n">gtf_sample</span> <span class="o">&lt;-</span> <span class="n">separate</span><span class="p">(</span><span class="n">gtf_sample</span><span class="p">,</span> <span class="n">V9</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;;&quot;</span><span class="p">,</span><span class="n">into</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;transcript_id&quot;</span><span class="p">,</span> <span class="s2">&quot;exon_number&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">))</span>
    <span class="n">gtf_sample</span> <span class="o">&lt;-</span> <span class="n">separate</span><span class="p">(</span><span class="n">gtf_sample</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">,</span><span class="n">into</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id_val&quot;</span><span class="p">))</span>
    <span class="n">gtf_sample</span> <span class="o">&lt;-</span> <span class="n">separate</span><span class="p">(</span><span class="n">gtf_sample</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;e &quot;</span><span class="p">,</span><span class="n">into</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;gene_name&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name_val&quot;</span><span class="p">))</span>
  
    <span class="n">gtf_name_id_match</span> <span class="o">&lt;-</span> <span class="n">gtf_sample</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;gene_id_val&quot;</span><span class="p">,</span><span class="s2">&quot;gene_name_val&quot;</span><span class="p">)]</span>
    <span class="n">gtf_name_id_match</span> <span class="o">&lt;-</span> <span class="n">gtf_name_id_match</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">gtf_name_id_match</span><span class="p">),</span> <span class="p">]</span>
  
    <span class="c1"># For any matched approved id in the psi hg38 annotation and gtf file, record the corresponding Ensembl ID</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span>
      <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">GTF</span><span class="err">`</span> <span class="o">=</span> <span class="n">gtf_name_id_match</span><span class="err">$</span><span class="n">gene_id_val</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">,</span> <span class="n">gtf_name_id_match</span><span class="err">$</span><span class="n">gene_name_val</span><span class="p">)]))</span>
  
    <span class="c1"># load hgnc database</span>
    <span class="n">hgnc_db</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_input[1]}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
  
    <span class="c1"># Combine the `Ensembl.ID.supplied.by.Ensembl.` and `Ensembl.gene.ID` column, if there are any conflict use the former</span>
    <span class="c1"># For conflict ones (15 total) both the former and latter records are poiting to the same gene name in Ensembl website so the order should not matter</span>
    <span class="n">hgnc_db</span> <span class="o">&lt;-</span> <span class="n">hgnc_db</span> <span class="o">%&gt;%</span>
    <span class="n">mutate</span><span class="p">(</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="err">`</span><span class="n">Ensembl</span> <span class="n">ID</span><span class="p">(</span><span class="n">supplied</span> <span class="n">by</span> <span class="n">Ensembl</span><span class="p">)</span><span class="err">`</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="err">`</span><span class="n">Ensembl</span> <span class="n">gene</span> <span class="n">ID</span><span class="err">`</span><span class="p">,</span> <span class="err">`</span><span class="n">Ensembl</span> <span class="n">ID</span><span class="p">(</span><span class="n">supplied</span> <span class="n">by</span> <span class="n">Ensembl</span><span class="p">)</span><span class="err">`</span><span class="p">))</span>
  
    <span class="c1"># Create a one to one reference list for approved names, previous names and aliases</span>
    <span class="c1"># There is no duplicate symbol and Ensembl id info for approved symbol so no need for chromosome verification</span>
    <span class="n">hgnc_name_id_match</span> <span class="o">&lt;-</span> <span class="n">hgnc_db</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;Approved symbol&quot;</span><span class="p">,</span><span class="s2">&quot;ENSG.ID&quot;</span><span class="p">)]</span>
    <span class="n">hgnc_name_prev_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_db</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;Previous symbols&quot;</span><span class="p">,</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span><span class="s2">&quot;ENSG.ID&quot;</span><span class="p">)]</span>
    <span class="n">hgnc_name_alias_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_db</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;Alias symbols&quot;</span><span class="p">,</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span><span class="s2">&quot;ENSG.ID&quot;</span><span class="p">)]</span>
  
    <span class="c1"># Remove NAs</span>
    <span class="n">hgnc_name_prev_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_name_prev_check</span><span class="p">[</span><span class="n">hgnc_name_prev_check</span><span class="err">$</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">,]</span>
    <span class="n">hgnc_name_alias_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_name_alias_check</span><span class="p">[</span><span class="n">hgnc_name_alias_check</span><span class="err">$</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">,]</span>

    <span class="n">hgnc_name_prev_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_name_prev_check</span><span class="p">[</span><span class="n">hgnc_name_prev_check</span><span class="err">$</span><span class="s2">&quot;Previous symbols&quot;</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">,]</span> 
    <span class="n">hgnc_name_alias_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_name_alias_check</span><span class="p">[</span><span class="n">hgnc_name_alias_check</span><span class="err">$</span><span class="s2">&quot;Alias symbols&quot;</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">,]</span>

    <span class="c1"># Seperate symbol column values from list of sybols to individual rows with one each</span>
    <span class="n">hgnc_name_prev_check</span> <span class="o">&lt;-</span> <span class="n">separate_rows</span><span class="p">(</span><span class="n">hgnc_name_prev_check</span><span class="p">,</span> <span class="s2">&quot;Previous symbols&quot;</span><span class="p">,</span> <span class="n">convert</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="n">hgnc_name_alias_check</span> <span class="o">&lt;-</span> <span class="n">separate_rows</span><span class="p">(</span><span class="n">hgnc_name_alias_check</span><span class="p">,</span> <span class="s2">&quot;Alias symbols&quot;</span><span class="p">,</span> <span class="n">convert</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
  
    <span class="c1"># Convert chomosome info in hgnc database to number for matching with other database</span>
    <span class="n">hgnc_name_prev_check</span> <span class="o">&lt;-</span> <span class="n">separate</span><span class="p">(</span><span class="n">hgnc_name_prev_check</span><span class="p">,</span> <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="n">into</span> <span class="o">=</span> <span class="s2">&quot;Chrp&quot;</span><span class="p">,</span> <span class="n">remove</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="n">hgnc_name_prev_check</span> <span class="o">&lt;-</span> <span class="n">separate</span><span class="p">(</span><span class="n">hgnc_name_prev_check</span><span class="p">,</span> <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;q&#39;</span><span class="p">,</span> <span class="n">into</span> <span class="o">=</span> <span class="s2">&quot;Chrq&quot;</span><span class="p">,</span> <span class="n">remove</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="n">hgnc_name_prev_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_name_prev_check</span><span class="o">%&gt;%</span>
                                <span class="n">mutate</span><span class="p">(</span><span class="n">Chr</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="n">nchar</span><span class="p">(</span><span class="n">hgnc_name_prev_check</span><span class="err">$</span><span class="n">Chrp</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">Chrp</span><span class="p">,</span> <span class="n">Chrq</span><span class="p">))</span>
    
    <span class="n">hgnc_name_alias_check</span> <span class="o">&lt;-</span> <span class="n">separate</span><span class="p">(</span><span class="n">hgnc_name_alias_check</span><span class="p">,</span> <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="n">into</span> <span class="o">=</span> <span class="s2">&quot;Chrp&quot;</span><span class="p">,</span> <span class="n">remove</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="n">hgnc_name_alias_check</span> <span class="o">&lt;-</span> <span class="n">separate</span><span class="p">(</span><span class="n">hgnc_name_alias_check</span><span class="p">,</span> <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;q&#39;</span><span class="p">,</span> <span class="n">into</span> <span class="o">=</span> <span class="s2">&quot;Chrq&quot;</span><span class="p">,</span> <span class="n">remove</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="n">hgnc_name_alias_check</span> <span class="o">&lt;-</span> <span class="n">hgnc_name_alias_check</span><span class="o">%&gt;%</span>
                                <span class="n">mutate</span><span class="p">(</span><span class="n">Chr</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="n">nchar</span><span class="p">(</span><span class="n">hgnc_name_alias_check</span><span class="err">$</span><span class="n">Chrp</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">Chrp</span><span class="p">,</span> <span class="n">Chrq</span><span class="p">))</span>
  
    <span class="c1"># For any matched approved id in the psi hg38 annotation and hgnc database, record the corresponding Ensembl ID</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span>
      <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span> <span class="o">=</span> <span class="n">hgnc_name_id_match</span><span class="err">$`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">,</span> <span class="n">hgnc_name_id_match</span><span class="err">$</span><span class="s2">&quot;Approved symbol&quot;</span><span class="p">)]))</span>
  
    <span class="c1"># Drop hypothetical genes</span>
    <span class="n">annotation</span><span class="o">&lt;-</span>
      <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
            <span class="n">subset</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span> <span class="o">!=</span> <span class="s1">&#39;Hypothetical&#39;</span><span class="p">))</span>
  
    <span class="c1"># IN remaining NAs, for any matched alias/previous names and chromosome in the psi hg38 annotation and hgnc database, record the corresponding Ensembl ID</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span>
      <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
                      <span class="n">mutate</span><span class="p">(</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span><span class="p">)</span> <span class="o">|</span> <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
                                                <span class="n">hgnc_name_alias_check</span><span class="err">$`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">,</span> <span class="n">hgnc_name_alias_check</span><span class="err">$</span><span class="s2">&quot;Alias symbols&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">match</span><span class="p">(</span><span class="err">`</span><span class="n">Chromosome</span><span class="err">`</span><span class="p">,</span> <span class="n">hgnc_name_alias_check</span><span class="err">$</span><span class="n">Chr</span><span class="p">)],</span>
                                                <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span><span class="p">))</span><span class="o">%&gt;%</span>
                      <span class="n">mutate</span><span class="p">(</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span><span class="p">)</span> <span class="o">|</span> <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
                                                <span class="n">hgnc_name_prev_check</span><span class="err">$`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">,</span> <span class="n">hgnc_name_prev_check</span><span class="err">$</span><span class="s2">&quot;Previous symbols&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">match</span><span class="p">(</span><span class="err">`</span><span class="n">Chromosome</span><span class="err">`</span><span class="p">,</span> <span class="n">hgnc_name_prev_check</span><span class="err">$</span><span class="n">Chr</span><span class="p">)],</span>
                                                <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span><span class="p">)))</span>
  
    <span class="c1"># Build the final Ensembl id column base on the gtf file first, then for remaining NAs check the HGNC database record, SUPPA and VASTTOOL record,</span>
    <span class="c1"># Drop special cases that Ensembl ID is not recorded in VASTTOOLs and SUPPA</span>
    <span class="c1"># finnally in the remaining NA Ensmbl ID if the gene name in original annotation is NCBI IDs just use it</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span>
      <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span> <span class="o">=</span> <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">GTF</span><span class="err">`</span><span class="p">)</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">),</span>
                                               <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">HGNC</span><span class="err">`</span><span class="p">,</span>
                                                <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">))</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">),</span>
                                               <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">VAST</span><span class="err">`</span><span class="p">,</span>
                                                <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">))</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">),</span>
                                               <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">SUPPA</span><span class="err">`</span><span class="p">,</span>
                                                <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">))</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="n">substr</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;ENSG&#39;</span><span class="p">,</span>
                                                 <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">,</span>
                                                 <span class="n">NA</span><span class="p">))</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span> <span class="o">=</span> <span class="n">ifelse</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">substr</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;LOC&#39;</span><span class="p">,</span>
                                               <span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">,</span>
                                                <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">)))</span>

    <span class="c1"># Use the Ensembl IDs to replace gene names, drop remaining NAs</span>
    <span class="n">annotation</span> <span class="o">&lt;-</span>
    <span class="nb">map</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span>
                       <span class="n">mutate</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span> <span class="o">=</span> <span class="err">`</span><span class="n">ENSG</span><span class="o">.</span><span class="n">ID</span><span class="err">`</span><span class="p">)</span><span class="o">%&gt;%</span>
            <span class="n">drop_na</span><span class="p">(</span><span class="err">`</span><span class="n">Gene</span><span class="err">`</span><span class="p">)</span>
          <span class="p">)</span>

    <span class="c1"># save modified annotation</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/data_preprocessing"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../commands_generator/eQTL_analysis_commands.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Bulk RNA-seq xQTL analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../molecular_phenotypes/bulk_expression.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">RNA-seq expression</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021+, ADSP-FG xQTL methods and data integration working group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>