
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>RNA-seq calling and QC &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Univariate xQTL Discovery" href="eQTL_analysis_commands.html" />
    <link rel="prev" title="Illustration of xQTL protocol" href="../xqtl_protocol_demo.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pecotmr_integration/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/fine_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/commands_generator/bulk_expression_commands.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/commands_generator/bulk_expression_commands.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/commands_generator/bulk_expression_commands.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#download-process-and-index-reference-data-first-time-use">
   Download, Process, and index reference data (first time use)
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#download-reference-and-annotation-data-first-time-use">
     Download reference and annotation data (first time use)
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reformat-reference-and-annotation-data-first-time-use">
     Reformat reference and annotation data (first time use)
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generation-of-index-first-time-only">
     Generation of Index (First time only)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#perform-data-quality-summary-via-fastqc">
   Perform data quality summary via
   <code class="docutils literal notranslate">
    <span class="pre">
     fastqc
    </span>
   </code>
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multi-sample-rna-seq-calling-qc-and-normalizations-automatically-for-every-studies-using-the-same-set-of-ref-index-gtf">
   Multi-sample RNA-seq Calling, QC, and normalizations (Automatically for every studies using the same set of ref/index/gtf)
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#call-gene-level-rna-expression-via-rnaseqc">
     Call gene-level RNA expression via
     <code class="docutils literal notranslate">
      <span class="pre">
       rnaseqc
      </span>
     </code>
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#call-transcript-level-rna-expression-via-rsem">
     Call transcript level RNA expression via
     <code class="docutils literal notranslate">
      <span class="pre">
       RSEM
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multi-sample-rna-seq-qc-for-every-set-of-samples">
   Multi-sample RNA-seq QC (For every set of samples)
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multi-sample-read-count-normalization-for-every-set-of-samples">
   Multi-sample read count normalization (For every set of samples)
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="rna-seq-calling-and-qc">
<h1>RNA-seq calling and QC<a class="headerlink" href="#rna-seq-calling-and-qc" title="Permalink to this headline">¶</a></h1>
<p>This document shows the use of various modules to prepare reference data, perform RNA-seq calling, expression level quantification and quality control. In particular,</p>
<ol class="simple">
<li><p><a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/reference_data.html"><code class="docutils literal notranslate"><span class="pre">reference_data.ipynb</span></code></a></p></li>
<li><p><a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/molecular_phenotypes/calling/RNA_calling.html"><code class="docutils literal notranslate"><span class="pre">RNA_calling.ipynb</span></code></a></p></li>
<li><p><a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/molecular_phenotypes/QC/readCount_QC.html"><code class="docutils literal notranslate"><span class="pre">readCount_QC.ipynb</span></code></a></p></li>
</ol>
<p>A minimal working example is available on <a class="reference external" href="https://drive.google.com/drive/u/0/folders/11kQv7PXozsKkgeqADH-28bC_kZ-w_oHo">Google Drive</a>.</p>
<p>This is a master control notebook for the molecular_phenotype_calling section of our pipeline. When following example commands was excuted, it will checked if the reference data was presented and generate the reference data if not. After that, molecular phenotype calling will be done based on the recipe that were fed into it. For each rows in the recipe, It will output a bed.gz file that are ready for downstream analysis.</p>
<p>By default, the reference data is assumed to be stored in the reference_data folder under the cwd. However, user can specify whether the</p>
<ul>
<li><p>Input:</p>
<p>A recipe file, with columns: <code class="docutils literal notranslate"><span class="pre">&quot;Theme&quot;,</span> <span class="pre">&quot;Sample_fastq_list&quot;,</span> <span class="pre">&quot;Data_dir&quot;</span></code></p>
</li>
<li><p>Output:</p>
<p>A bed.gz file for each row in the Input</p>
</li>
<li><p>parameters:</p>
<ul class="simple">
<li><p>flag <code class="docutils literal notranslate"><span class="pre">--stranded</span></code> or <code class="docutils literal notranslate"><span class="pre">--no-stranded</span></code> : whether the rna seq data intended to be called is stranded or not.</p></li>
<li><p>flag <code class="docutils literal notranslate"><span class="pre">--pair-ended</span></code> or <code class="docutils literal notranslate"><span class="pre">--no-pair-ended</span></code> : whether the rna seq data intended to be called is paired-ended or not.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--exe_dir</span></code> : diretory to the workflow notebook</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--sample-participant-lookup</span></code> : a sample participant lookup table with the first column being sample ID and second column being the cooresponding sample ids in the genotype file</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">bulk_expression_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run bulk_expression_commands.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  ref_data
  gene_annotation
  indexing
  fastqc
  rnaseqc
  RSEM
  QC
  normalization

Global Workflow Options:
  --recipe VAL (as path, required)
                        The aforementioned input recipe
  --cwd . (as path)
                        Overall wd, the file structure of analysis is
                        wd/[steps]/[sub_dir for each steps]
  --refd  path(f&#39;{cwd}/reference_data/&#39;)

                        dir to store the reference data
  --exe-dir /home/gw/GIT/xqtl-pipeline/pipeline (as path)
                        Diretory to the excutable
  --container-rnaquant &#39;/mnt/mfs/statgen/containers/rna_quantification.sif&#39;
  --yml csg.yml (as path)
  --jobs 50 (as int)
  --queue &#39;csg,q&#39;

Sections
  ref_data_1:
  ref_data_2:
  gene_annotation:
    Workflow Options:
      --[no-]stranded (required)
  indexing:
  fastqc:
  rnaseqc:
    Workflow Options:
      --fasta-with-adapters-etc &#39;TruSeq3-PE.fa&#39;
  RSEM:
    Workflow Options:
      --fasta-with-adapters-etc &#39;TruSeq3-PE.fa&#39;
      --[no-]is-paired-end (required)
  QC:
    Workflow Options:
      --low-expr-TPM 0.1 (as float)
      --low-expr-TPM-percent 0.2 (as float)
      --RLEFilterPercent 0.05 (as float)
      --DSFilterPercent 0.05 (as float)
      --topk-genes 100 (as int)
      --cluster-percent 0.6 (as float)
      --pvalues-cut 0.05 (as float)
      --cluster-level 5 (as int)
  normalization:
    Workflow Options:
      --sample-participant-lookup VAL (as path, required)
      --tpm-threshold 0.1 (as float)
      --count-threshold 6 (as int)
      --sample-frac-threshold 0.2 (as float)
      --normalization-method tmm
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1">## The aforementioned input recipe</span>
<span class="kn">parameter:</span> <span class="n">recipe</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1">## Overall wd, the file structure of analysis is wd/[steps]/[sub_dir for each steps]</span>
<span class="kn">parameter:</span>  <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1">## dir to store the reference data</span>
<span class="kn">parameter:</span> <span class="n">refd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/reference_data/&#39;</span><span class="p">)</span>
<span class="c1">## Diretory to the excutable</span>
<span class="kn">parameter:</span> <span class="n">exe_dir</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;~/GIT/xqtl-pipeline/pipeline/&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">container_rnaquant</span> <span class="o">=</span> <span class="s1">&#39;/mnt/mfs/statgen/containers/rna_quantification.sif&#39;</span>
<span class="c1">## Cloud computing options:</span>
<span class="kn">parameter:</span> <span class="n">yml</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./csg.yml&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">jobs</span> <span class="o">=</span> <span class="mi">50</span> <span class="c1"># Number of jobs that are submitted to the cluster</span>
<span class="kn">parameter:</span> <span class="kp">queue</span> <span class="o">=</span> <span class="s2">&quot;csg&quot;</span> <span class="c1"># The queue that jobs are submitted to</span>
<span class="n">submission</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;-J </span><span class="si">{</span><span class="n">jobs</span><span class="si">}</span><span class="s1"> -c </span><span class="si">{</span><span class="n">yml</span><span class="si">}</span><span class="s1"> -q </span><span class="si">{</span><span class="kp">queue</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">parameter:</span> <span class="n">yml</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;csg.yml&quot;</span><span class="p">)</span> 
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">input_inv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">recipe</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;records&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">recipe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span> <span class="p">{</span><span class="s2">&quot;Theme&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;AC&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC&quot;</span><span class="p">,</span><span class="s2">&quot;PCC&quot;</span><span class="p">]</span> <span class="p">,</span>
                <span class="s2">&quot;Sample_fastq_list&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;AC_sample_fastq.list&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;DLPFC_sample_fastq.list&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;PCC_sample_fastq&quot;</span><span class="p">],</span>
                <span class="s2">&quot;Data_dir&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/&quot;</span><span class="p">]})</span>
<span class="n">recipe</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/mwe_recipe.txt&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">recipe</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Theme</th>
      <th>Sample_fastq_list</th>
      <th>Data_dir</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AC</td>
      <td>AC_sample_fastq.list</td>
      <td>/mnt/mfs/statgen/xqtl_workflow_testing/rna_qua...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>DLPFC</td>
      <td>DLPFC_sample_fastq.list</td>
      <td>/mnt/mfs/statgen/xqtl_workflow_testing/rna_qua...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>PCC</td>
      <td>PCC_sample_fastq</td>
      <td>/mnt/mfs/statgen/xqtl_workflow_testing/rna_qua...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">dryrun</span> <span class="o">./</span><span class="n">bulk_expression_master</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fastqc</span> \
    <span class="o">--</span><span class="n">recipe</span> <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/mwe_recipe.txt&quot;</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span>
    

<span class="n">sos</span> <span class="n">dryrun</span> <span class="o">./</span><span class="n">bulk_expression_master</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">normalization</span> \
    <span class="o">--</span><span class="n">recipe</span> <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/mwe_recipe.txt&quot;</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span> \
    <span class="o">--</span><span class="n">sample</span><span class="o">-</span><span class="n">participant</span><span class="o">-</span><span class="n">lookup</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">fasta</span><span class="o">-</span><span class="k">with</span><span class="o">-</span><span class="n">adapters</span><span class="o">-</span><span class="n">etc</span>  <span class="s2">&quot;TruSeq3-PE.fa&quot;</span> \
    <span class="o">--</span><span class="n">stranded</span> \
    <span class="o">--</span><span class="ow">is</span><span class="o">-</span><span class="n">paired</span><span class="o">-</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">example_script</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">phenotype_master_control</span>
<span class="n">cat</span> <span class="n">example_script</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">HINT</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/pipeline/reference_data.ipynb download_hg_reference --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data
sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/pipeline/reference_data.ipynb download_gene_annotation --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data
sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/pipeline/reference_data.ipynb download_ercc_reference --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/reference_data.ipynb hg_reference \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/GRCh38_full_analysis_set_plus_decoy_hla /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92 /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data \
    --ercc-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.fa \
    --hg-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/reference_data.ipynb gene_annotation \
        --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data \
        --ercc-gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.gtf \
        --hg-gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.gtf \
        --hg-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/GRCh38_full_analysis_set_plus_decoy_hla /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92 /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.noALT_noHLA_noDecoy_ERCC.fasta \
        --container /mnt/mfs/statgen/containers/rna_quantification.sif \
        --stranded  -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/reference_data.ipynb STAR_index RSEM_index \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data \
    --hg-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/GRCh38_full_analysis_set_plus_decoy_hla /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92 /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.noALT_noHLA_noDecoy_ERCC.fasta \
    --hg-gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run  /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  AC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run  /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  DLPFC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run  /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  PCC_sample_fastq \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  AC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --RSEM-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/RSEM_Index \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \ 
    --mem 40G \
    --is_paired_end -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  DLPFC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --RSEM-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/RSEM_Index \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \ 
    --mem 40G \
    --is_paired_end -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  PCC_sample_fastq \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --RSEM-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/RSEM_Index \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \ 
    --mem 40G \
    --is_paired_end -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_QC.ipynb qc \
    --cwd . \
    --tpm-gct AC.rnaseqc.gene_tpm.gct.gz \
    --counts-gct  AC.rnaseqc.gene_readsCount.gct.gz \
    --low_expr_TPM 0.1  \
    --low_expr_TPM_percent 0.2 \
    --RLEFilterPercent 0.05 \
    --DSFilterPercent 0.05 \
    --topk_genes 100 \
    --cluster_percent 0.6 \
    --pvalues_cut 0.05 \
    --cluster_level 5 \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_QC.ipynb qc \
    --cwd . \
    --tpm-gct DLPFC.rnaseqc.gene_tpm.gct.gz \
    --counts-gct  DLPFC.rnaseqc.gene_readsCount.gct.gz \
    --low_expr_TPM 0.1  \
    --low_expr_TPM_percent 0.2 \
    --RLEFilterPercent 0.05 \
    --DSFilterPercent 0.05 \
    --topk_genes 100 \
    --cluster_percent 0.6 \
    --pvalues_cut 0.05 \
    --cluster_level 5 \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_QC.ipynb qc \
    --cwd . \
    --tpm-gct PCC.rnaseqc.gene_tpm.gct.gz \
    --counts-gct  PCC.rnaseqc.gene_readsCount.gct.gz \
    --low_expr_TPM 0.1  \
    --low_expr_TPM_percent 0.2 \
    --RLEFilterPercent 0.05 \
    --DSFilterPercent 0.05 \
    --topk_genes 100 \
    --cluster_percent 0.6 \
    --pvalues_cut 0.05 \
    --cluster_level 5 \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_normalization.ipynb normalize \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control \
    --tpm-gct AC.rnaseqc.low_expression_filtered.outlier_removed.tpm.gct.gz \
    --counts-gct AC.rnaseqc.low_expression_filtered.outlier_removed.geneCount.gct.gz \
    --annotation-gtf  /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf  \
    --sample-participant-lookup /mnt/mfs/statgen/snuc_pseudo_bulk/data/reference_data/sampleSheetAfterQC.txt \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif  \
    --tpm_threshold 0.1 \
    --count_threshold 6  \
    --sample_frac_threshold 0.2 \
    --normalization_method tmm  -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_normalization.ipynb normalize \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control \
    --tpm-gct DLPFC.rnaseqc.low_expression_filtered.outlier_removed.tpm.gct.gz \
    --counts-gct DLPFC.rnaseqc.low_expression_filtered.outlier_removed.geneCount.gct.gz \
    --annotation-gtf  /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf  \
    --sample-participant-lookup /mnt/mfs/statgen/snuc_pseudo_bulk/data/reference_data/sampleSheetAfterQC.txt \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif  \
    --tpm_threshold 0.1 \
    --count_threshold 6  \
    --sample_frac_threshold 0.2 \
    --normalization_method tmm  -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_normalization.ipynb normalize \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control \
    --tpm-gct PCC.rnaseqc.low_expression_filtered.outlier_removed.tpm.gct.gz \
    --counts-gct PCC.rnaseqc.low_expression_filtered.outlier_removed.geneCount.gct.gz \
    --annotation-gtf  /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf  \
    --sample-participant-lookup /mnt/mfs/statgen/snuc_pseudo_bulk/data/reference_data/sampleSheetAfterQC.txt \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif  \
    --tpm_threshold 0.1 \
    --count_threshold 6  \
    --sample_frac_threshold 0.2 \
    --normalization_method tmm  -J 50 -c csg.yml -q csg,q
</pre></div>
</div>
</div>
</div>
<div class="section" id="download-process-and-index-reference-data-first-time-use">
<h2>Download, Process, and index reference data (first time use)<a class="headerlink" href="#download-process-and-index-reference-data-first-time-use" title="Permalink to this headline">¶</a></h2>
<div class="section" id="download-reference-and-annotation-data-first-time-use">
<h3>Download reference and annotation data (first time use)<a class="headerlink" href="#download-reference-and-annotation-data-first-time-use" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ref_data_1</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/GRCh38_full_analysis_set_plus_decoy_hla.fa&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/Homo_sapiens.GRCh38.103.chr.gtf&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/ERCC92.gtf&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/ERCC92.fa&quot;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_hg_reference</span> <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span>
        <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_gene_annotation</span> <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span>
        <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">download_ercc_reference</span> <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="reformat-reference-and-annotation-data-first-time-use">
<h3>Reformat reference and annotation data (first time use)<a class="headerlink" href="#reformat-reference-and-annotation-data-first-time-use" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ref_data_2</span><span class="p">]</span>
<span class="kn">output:</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.noALT_noHLA_noDecoy_ERCC.fasta&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.noALT_noHLA_noDecoy_ERCC.dict&#39;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">hg_reference</span> \
            <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span> \
            <span class="o">--</span><span class="n">ercc</span><span class="o">-</span><span class="n">reference</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">3</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">**</span><span class="n">Notice</span><span class="p">:</span> <span class="n">you</span> <span class="n">need</span> <span class="n">to</span> <span class="n">know</span> <span class="k">if</span> <span class="n">your</span> <span class="n">RNA</span><span class="o">-</span><span class="n">seq</span> <span class="n">data</span> <span class="ow">is</span> <span class="n">stranded</span> <span class="ow">or</span> <span class="n">unstranded</span><span class="o">.</span> <span class="n">If</span> <span class="n">it</span> <span class="ow">is</span> <span class="n">stranded</span><span class="p">,</span> <span class="n">you</span> <span class="n">should</span> <span class="n">add</span> <span class="err">`</span><span class="o">--</span><span class="n">stranded</span><span class="err">`</span> <span class="n">to</span> <span class="n">the</span> <span class="n">command</span> <span class="n">below</span><span class="o">.**</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gene_annotation</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">stranded</span> <span class="o">=</span> <span class="nb">bool</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;ref_data_1&quot;</span><span class="p">),</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;ref_data_2&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.reformatted.ERCC.gtf&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.reformatted</span><span class="si">{</span><span class="s2">&quot;.collapse_only&quot;</span> <span class="k">if</span> <span class="n">stranded</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.gene.ERCC.gtf&#39;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">gene_annotation</span> \
            <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span> \
            <span class="o">--</span><span class="n">ercc</span><span class="o">-</span><span class="n">gtf</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span> \
            <span class="err">$</span><span class="p">[</span><span class="s2">&quot;--stranded&quot;</span> <span class="k">if</span> <span class="n">stranded</span> <span class="k">else</span> <span class="s2">&quot;--no-stranded&quot;</span><span class="p">]</span>  <span class="err">$</span><span class="p">[</span><span class="n">submission</span> <span class="k">if</span> <span class="n">yml</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="generation-of-index-first-time-only">
<h3>Generation of Index (First time only)<a class="headerlink" href="#generation-of-index-first-time-only" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">indexing</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;gene_annotation&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;ref_data_2&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.n2g.idx.fa&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.grp&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.idx.fa&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.ti&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.chrlist&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.seq&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/RSEM_Index/rsem_reference.transcripts.fa&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/chrName.txt&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/Log.out&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/transcriptInfo.tab&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/exonGeTrInfo.tab&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/SAindex&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/SA&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/genomeParameters.txt&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/chrStart.txt&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/sjdbList.out.tab&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/exonInfo.tab&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/sjdbList.fromGTF.out.tab&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/chrLength.txt&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/sjdbInfo.txt&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/Genome&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/chrNameLength.txt&quot;</span><span class="p">,</span> 
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">refd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/STAR_Index/geneInfo.tab&quot;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">reference_data</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">STAR_index</span> <span class="n">RSEM_index</span> \
            <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">refd</span><span class="p">:</span><span class="n">a</span><span class="p">]</span> \
            <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">reference</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">hg</span><span class="o">-</span><span class="n">gtf</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span> \
            <span class="o">--</span><span class="kp">mem</span> <span class="mi">40</span><span class="n">G</span> <span class="err">$</span><span class="p">[</span><span class="n">submission</span> <span class="k">if</span> <span class="n">yml</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="perform-data-quality-summary-via-fastqc">
<h2>Perform data quality summary via <code class="docutils literal notranslate"><span class="pre">fastqc</span></code><a class="headerlink" href="#perform-data-quality-summary-via-fastqc" title="Permalink to this headline">¶</a></h2>
<p><strong>You need to figure out from fastqc results what adapter reference sequence to use for <code class="docutils literal notranslate"><span class="pre">--fasta_with_adapters_etc</span></code>. See section Step 1 below for details</strong>. We chose <code class="docutils literal notranslate"><span class="pre">--fasta_with_adapters_etc</span> <span class="pre">TruSeq3-PE.fa</span></code> for our toy example.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fasta_with_adapters_etc</span></code>: <strong>filename</strong> for the adapter reference file. According to <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> documention,</p></li>
</ul>
<blockquote>
<div><p>As a rule of thumb newer libraries will use <code class="docutils literal notranslate"><span class="pre">TruSeq3</span></code>, but this really depends on your service provider. If you use FASTQC, the “Overrepresented Sequences” report can help indicate which adapter file is best suited for your data. “Illumina Single End” or “Illumina Paired End” sequences indicate single-end or paired-end <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq2-SE.fa</span></code> and <code class="docutils literal notranslate"><span class="pre">TruSeq2-PE.fa</span></code> respectively. “TruSeq Universal Adapter” or “TruSeq Adapter, Index …” indicates <code class="docutils literal notranslate"><span class="pre">TruSeq-3</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq3-SE.fa</span></code> or <code class="docutils literal notranslate"><span class="pre">TruSeq3-PE.fa</span></code>, for single-end and paired-end data respectively. Adapter sequences for <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> multiplexed libraries, indicated by “Illumina Multiplexing
…”, and the various RNA library preparations are not currently included.</p>
</div></blockquote>
<p>We have <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> workflow defined as followed. Users should decide what fasta adapter reference to use based on <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> results (or their own knowledge).
It is assumed that all studies that were ran in one run are using the same adapter file</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">fastqc</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;input_inv&quot;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fastqc</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">cwd</span><span class="p">:</span><span class="n">a</span><span class="p">]</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">fastqc</span><span class="o">/</span> \
        <span class="o">--</span><span class="n">samples</span> <span class="err">$</span><span class="p">[</span><span class="n">_input_inv</span><span class="p">[</span><span class="s1">&#39;Sample_fastq_list&#39;</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">[</span><span class="n">_input_inv</span><span class="p">[</span><span class="s1">&#39;Data_dir&#39;</span><span class="p">]]</span>   \
        <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span> <span class="err">$</span><span class="p">[</span><span class="n">submission</span> <span class="k">if</span> <span class="n">yml</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="multi-sample-rna-seq-calling-qc-and-normalizations-automatically-for-every-studies-using-the-same-set-of-ref-index-gtf">
<h2>Multi-sample RNA-seq Calling, QC, and normalizations (Automatically for every studies using the same set of ref/index/gtf)<a class="headerlink" href="#multi-sample-rna-seq-calling-qc-and-normalizations-automatically-for-every-studies-using-the-same-set-of-ref-index-gtf" title="Permalink to this headline">¶</a></h2>
<div class="section" id="call-gene-level-rna-expression-via-rnaseqc">
<h3>Call gene-level RNA expression via <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code><a class="headerlink" href="#call-gene-level-rna-expression-via-rnaseqc" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">rnaseqc</span><span class="p">]</span>
<span class="kn">parameter:</span><span class="n">fasta_with_adapters_etc</span> <span class="o">=</span> <span class="s2">&quot;TruSeq3-PE.fa&quot;</span>
<span class="kn">input:</span>  <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;gene_annotation&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;indexing&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span><span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;input_inv&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rnaseqc.gene_tpm.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rnaseqc.gene_readsCount.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rnaseqc.exon_readsCount.gct.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rnaseqc.metrics.tsv&#39;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="n">sos</span> <span class="n">run</span>  <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rnaseqc_call</span> \
            <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span><span class="o">/</span> \
            <span class="o">--</span><span class="n">samples</span>  <span class="err">$</span><span class="p">[</span><span class="n">_input_inv</span><span class="p">[</span><span class="s1">&#39;Sample_fastq_list&#39;</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">[</span><span class="n">_input_inv</span><span class="p">[</span><span class="s1">&#39;Data_dir&#39;</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">fasta_with_adapters_etc</span> <span class="err">$</span><span class="p">[</span><span class="n">fasta_with_adapters_etc</span><span class="p">]</span>  \
            <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span> \
            <span class="o">--</span><span class="n">gtf</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span> \
            <span class="o">--</span><span class="kp">mem</span> <span class="mi">40</span><span class="n">G</span> <span class="err">$</span><span class="p">[</span><span class="n">submission</span> <span class="k">if</span> <span class="n">yml</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="call-transcript-level-rna-expression-via-rsem">
<h3>Call transcript level RNA expression via <code class="docutils literal notranslate"><span class="pre">RSEM</span></code><a class="headerlink" href="#call-transcript-level-rna-expression-via-rsem" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">RSEM</span><span class="p">]</span>
<span class="kn">parameter:</span><span class="n">fasta_with_adapters_etc</span> <span class="o">=</span> <span class="s2">&quot;TruSeq3-PE.fa&quot;</span>
<span class="kn">parameter:</span> <span class="n">paired_end</span> <span class="o">=</span> <span class="nb">bool</span>
<span class="kn">input:</span>  <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;indexing&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;indexing&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span><span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;input_inv&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rsem_transcripts_expected_count.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rsem_transcripts_tpm.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rsem_transcripts_fpkm.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rsem_transcripts_isopct.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rsem_genes_expected_count.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rsem_genes_tpm.txt.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input_inv</span><span class="p">[</span><span class="s2">&quot;Theme&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rsem_genes_fpkm.txt.gz&#39;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rnaseqc_call</span> \
            <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span><span class="o">/</span> \
            <span class="o">--</span><span class="n">samples</span>  <span class="err">$</span><span class="p">[</span><span class="n">_input_inv</span><span class="p">[</span><span class="s1">&#39;Sample_fastq_list&#39;</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">[</span><span class="n">_input_inv</span><span class="p">[</span><span class="s1">&#39;Data_dir&#39;</span><span class="p">]]</span> \
            <span class="o">--</span><span class="n">fasta_with_adapters_etc</span> <span class="err">$</span><span class="p">[</span><span class="n">fasta_with_adapters_etc</span><span class="p">]</span>  \
            <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span> \
            <span class="o">--</span><span class="n">RSEM</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span> \
            <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span> \ 
            <span class="o">--</span><span class="kp">mem</span> <span class="mi">40</span><span class="n">G</span> \
            <span class="err">$</span><span class="p">[</span><span class="s2">&quot;--paired_end&quot;</span> <span class="k">if</span> <span class="n">paired_end</span> <span class="k">else</span> <span class="s2">&quot; &quot;</span><span class="p">]</span> <span class="err">$</span><span class="p">[</span><span class="n">submission</span> <span class="k">if</span> <span class="n">yml</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="multi-sample-rna-seq-qc-for-every-set-of-samples">
<h2>Multi-sample RNA-seq QC (For every set of samples)<a class="headerlink" href="#multi-sample-rna-seq-qc-for-every-set-of-samples" title="Permalink to this headline">¶</a></h2>
<p>We need to use a different MWE data-set that contains multiple samples – here is the <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1Rv2bWHBbX_tastTh49ToYVDMV6rFP5Wk">Google Drive link</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">QC</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">low_expr_TPM</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">low_expr_TPM_percent</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="kn">parameter:</span> <span class="n">RLEFilterPercent</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="kn">parameter:</span> <span class="n">DSFilterPercent</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="kn">parameter:</span> <span class="n">topk_genes</span> <span class="o">=</span> <span class="mi">100</span>
<span class="kn">parameter:</span> <span class="n">cluster_percent</span> <span class="o">=</span> <span class="mf">0.6</span>
<span class="kn">parameter:</span> <span class="n">pvalues_cut</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="kn">parameter:</span> <span class="n">cluster_level</span> <span class="o">=</span> <span class="mi">5</span>
<span class="kn">depends:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;RSEM&quot;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;rnaseqc&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.low_expression_filtered.outlier_removed.tpm.gct.gz&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.low_expression_filtered.outlier_removed.geneCount.gct.gz&#39;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">bulk_expression_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">tpm</span><span class="o">-</span><span class="n">gct</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">counts</span><span class="o">-</span><span class="n">gct</span>  <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">low_expr_TPM</span> <span class="err">$</span><span class="p">[</span><span class="n">low_expr_TPM</span><span class="p">]</span>  \
        <span class="o">--</span><span class="n">low_expr_TPM_percent</span> <span class="err">$</span><span class="p">[</span><span class="n">low_expr_TPM_percent</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">RLEFilterPercent</span> <span class="err">$</span><span class="p">[</span><span class="n">RLEFilterPercent</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">DSFilterPercent</span> <span class="err">$</span><span class="p">[</span><span class="n">DSFilterPercent</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">topk_genes</span> <span class="err">$</span><span class="p">[</span><span class="n">topk_genes</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">cluster_percent</span> <span class="err">$</span><span class="p">[</span><span class="n">cluster_percent</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">pvalues_cut</span> <span class="err">$</span><span class="p">[</span><span class="n">pvalues_cut</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">cluster_level</span> <span class="err">$</span><span class="p">[</span><span class="n">cluster_level</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span> <span class="err">$</span><span class="p">[</span><span class="n">submission</span> <span class="k">if</span> <span class="n">yml</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="multi-sample-read-count-normalization-for-every-set-of-samples">
<h2>Multi-sample read count normalization (For every set of samples)<a class="headerlink" href="#multi-sample-read-count-normalization-for-every-set-of-samples" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">normalization</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">tpm_threshold</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">count_threshold</span> <span class="o">=</span> <span class="mi">6</span>
<span class="kn">parameter:</span> <span class="n">sample_frac_threshold</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="kn">parameter:</span> <span class="n">normalization_method</span> <span class="o">=</span> <span class="s1">&#39;tmm&#39;</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;QC&quot;</span><span class="p">),</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;gene_annotation&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">normalization_method</span><span class="si">}</span><span class="s1">.expression.bed.gz&#39;</span>
<span class="kn">report:</span>  <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">bulk_expression_normalization</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">normalize</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">d</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">tpm</span><span class="o">-</span><span class="n">gct</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">counts</span><span class="o">-</span><span class="n">gct</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">annotation</span><span class="o">-</span><span class="n">gtf</span>  <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>  \
        <span class="o">--</span><span class="n">sample</span><span class="o">-</span><span class="n">participant</span><span class="o">-</span><span class="n">lookup</span> <span class="err">$</span><span class="p">[</span><span class="n">sample_participant_lookup</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">container</span> <span class="err">$</span><span class="p">[</span><span class="n">container_rnaquant</span><span class="p">]</span>  \
        <span class="o">--</span><span class="n">tpm_threshold</span> <span class="err">$</span><span class="p">[</span><span class="n">tpm_threshold</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">count_threshold</span> <span class="err">$</span><span class="p">[</span><span class="n">count_threshold</span><span class="p">]</span>  \
        <span class="o">--</span><span class="n">sample_frac_threshold</span> <span class="err">$</span><span class="p">[</span><span class="n">sample_frac_threshold</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">normalization_method</span> <span class="err">$</span><span class="p">[</span><span class="n">normalization_method</span><span class="p">]</span>  <span class="err">$</span><span class="p">[</span><span class="n">submission</span> <span class="k">if</span> <span class="n">yml</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/commands_generator"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../xqtl_protocol_demo.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Illustration of xQTL protocol</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="eQTL_analysis_commands.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Univariate xQTL Discovery</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>