{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5b07a337-e451-4ea3-8798-4264dc21b6de",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(readr)\n",
    "library(dplyr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8d2f5081-8218-4259-948b-d498abf96be7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "Loading required package: Rsamtools\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘VariantAnnotation’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    tabulate\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"VariantAnnotation\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f6c9af1b-24a5-40bb-92bb-0978e200aa3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `#chrom` = \u001b[32mcol_double()\u001b[39m,\n",
      "  pos = \u001b[32mcol_double()\u001b[39m,\n",
      "  ref = \u001b[31mcol_character()\u001b[39m,\n",
      "  alt = \u001b[31mcol_character()\u001b[39m,\n",
      "  gene = \u001b[31mcol_character()\u001b[39m,\n",
      "  beta = \u001b[32mcol_double()\u001b[39m,\n",
      "  se = \u001b[32mcol_double()\u001b[39m,\n",
      "  pval = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "a = read_delim(\"/mnt/mfs/statgen//neuro-apex/demo_with_mwe/cis/MWE.21.cis_long_table.txt.gz\",\"\\t\") -> a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "393df42f-d95a-4ab2-99cc-ebbd9f78fc48",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = a%>%dplyr::filter(alt != \"*\", ref != \"*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2f97e6f7-315e-421c-8a4c-76e20237bee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "       create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, \n",
    "        se = NULL, pval = NULL, n = NULL, ncase = NULL, name = NULL,cs = NULL, pip = NULL) \n",
    "    {\n",
    "        stopifnot(length(chrom) == length(pos))\n",
    "        if (is.null(snp)) {\n",
    "            snp <- paste0(chrom, \":\", pos)\n",
    "        }\n",
    "        snp <- paste0(chrom, \":\", pos)\n",
    "        nsnp <- length(chrom)\n",
    "        gen <- list()\n",
    "        ## Setupt data content for each sample column\n",
    "        if (!is.null(ea_af)) \n",
    "            gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
    "        if (!is.null(effect)) \n",
    "            gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
    "        if (!is.null(se)) \n",
    "            gen[[\"SE\"]] <- matrix(se, nsnp)\n",
    "        if (!is.null(pval)) \n",
    "            gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
    "        if (!is.null(n)) \n",
    "            gen[[\"SS\"]] <- matrix(n, nsnp)\n",
    "        if (!is.null(ncase)) \n",
    "            gen[[\"NC\"]] <- matrix(ncase, nsnp)\n",
    "        gen <- S4Vectors::SimpleList(gen)\n",
    "        \n",
    "      ## Setup snps info for the fix columns\n",
    "        gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
    "            end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
    "         coldata <- S4Vectors::DataFrame(Studies = name, row.names = name)\n",
    "    ## Setup header informations\n",
    "        hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
    "            row.names = \"fileformat\")), sample = name)\n",
    "        VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "            \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
    "            \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
    "            \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", \n",
    "            \"Alternate allele frequency in the association study\", \n",
    "            \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), \n",
    "            row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))\n",
    "      ## Save only the meta information in the sample columns \n",
    "        VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
    "            rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
    "      ## Save VCF \n",
    "        vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
    "            exptData = list(header = hdr), geno = gen)\n",
    "        VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
    "        VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
    "      ## Add fixed values\n",
    "        VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\"\n",
    "      ## Add Info values\n",
    "        if (!is.null(cs)) {\n",
    "        VariantAnnotation::info(vcf)$CS <- cs\n",
    "            }\n",
    "        if (!is.null(pip)) {\n",
    "        VariantAnnotation::info(vcf)$PIP <- pip\n",
    "            }\n",
    "      ## Add header for info\n",
    "        VariantAnnotation::info(header(vcf)) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "            \"A\"), Type = c(\"Float\", \"Float\"), Description = c(\"The credible set this snv located in\", \n",
    "            \"The posterior inclusion probability for this snv in the credible set\"), \n",
    "            row.names = c(\"CS\",\"PIP\"))\n",
    "        #VariantAnnotation::info(hdr) <- subset(VariantAnnotation::info(hdr), \n",
    "         #   rownames(VariantAnnotation::info(hdr)) %in% names(info(vcf)))\n",
    "           return(sort(vcf))\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e3dee685-17ca-4039-8868-c5b7dc457300",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in info((b)): could not find function \"info\"\n",
     "output_type": "error",
     "traceback": [
      "Error in info((b)): could not find function \"info\"\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "names(info((b)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "c2fef1d2-d19d-4c4a-9247-d9582ddcb300",
   "metadata": {},
   "outputs": [],
   "source": [
    " hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
    "            row.names = \"fileformat\")), sample = \"TTT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "66a2ba7d-4969-4b2b-9f0e-dfd59bc8f708",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 2 rows and 3 columns\n",
       "         Number        Type\n",
       "    <character> <character>\n",
       "CS            A       Float\n",
       "PIP           A       Float\n",
       "                                                             Description\n",
       "                                                             <character>\n",
       "CS                                  The credible set this snv located in\n",
       "PIP The posterior inclusion probability for this snv in the credible set"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "info(hdr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "8cde30ff-5785-472e-a3f7-b4f066a01299",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: CollapsedVCF \n",
       "dim: 102064 1 \n",
       "rowRanges(vcf):\n",
       "  GRanges with 4 metadata columns: REF, ALT, QUAL, FILTER\n",
       "info(vcf):\n",
       "  DataFrame with 2 columns: CS, PIP\n",
       "info(header(vcf)):\n",
       "       Number Type  Description                                                \n",
       "   CS  A      Float The credible set this snv located in                       \n",
       "   PIP A      Float The posterior inclusion probability for this snv in the ...\n",
       "geno(vcf):\n",
       "  List of length 3: ES, SE, LP\n",
       "geno(header(vcf)):\n",
       "      Number Type  Description                                            \n",
       "   ES A      Float Effect size estimate relative to the alternative allele\n",
       "   SE A      Float Standard error of effect size estimate                 \n",
       "   LP A      Float -log10 p-value for effect estimate                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    " VariantAnnotation::info(header(b)) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "            \"A\"), Type = c(\"Float\", \"Float\"), Description = c(\"The credible set this snv located in\", \n",
    "            \"The posterior inclusion probability for this snv in the credible set\"), \n",
    "            row.names = c(\"CS\",\"PIP\"))\n",
    "sort(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "00d44fd9-8d9d-4cf0-a46a-c99c3c5939c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: CollapsedVCF \n",
       "dim: 102064 1 \n",
       "rowRanges(vcf):\n",
       "  GRanges with 4 metadata columns: REF, ALT, QUAL, FILTER\n",
       "info(vcf):\n",
       "  DataFrame with 2 columns: CS, PIP\n",
       "info(header(vcf)):\n",
       "       Number Type  Description                                                \n",
       "   CS  A      Float The credible set this snv located in                       \n",
       "   PIP A      Float The posterior inclusion probability for this snv in the ...\n",
       "geno(vcf):\n",
       "  List of length 3: ES, SE, LP\n",
       "geno(header(vcf)):\n",
       "      Number Type  Description                                            \n",
       "   ES A      Float Effect size estimate relative to the alternative allele\n",
       "   SE A      Float Standard error of effect size estimate                 \n",
       "   LP A      Float -log10 p-value for effect estimate                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "28512e98-2b96-4f20-af0a-f14452c703a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in matrix(-log10(pval), nsnp):\n",
      "“NaNs produced”\n",
      "Warning message:\n",
      "“info fields with no header: CS”\n",
      "Warning message:\n",
      "“info fields with no header: CS,PIP”\n"
     ]
    }
   ],
   "source": [
    "d = create_vcf(\n",
    "   chrom = a$`#chrom`,\n",
    "    pos = a$pos,\n",
    "    ea = a$ref,\n",
    "    nea = a$alt,\n",
    "  effect = a$beta,\n",
    "  se = a$se,\n",
    "  pval = a$pval,\n",
    "    name = \"TTT\", cs = a$pval , pip =  a$pval )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6239c1cb-1697-4fac-bc0e-81d64babfbb0",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'd' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'd' not found\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "43926020-9586-4306-b66b-c88b129f6c4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'DFrame'"
      ],
      "text/latex": [
       "'DFrame'"
      ],
      "text/markdown": [
       "'DFrame'"
      ],
      "text/plain": [
       "[1] \"DFrame\"\n",
       "attr(,\"package\")\n",
       "[1] \"S4Vectors\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class(info(header(b)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74ce5641-2705-4a04-aafd-6bb175cf563a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "afb799fb-af44-4640-9806-85d36705ed6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“info fields with no header: TEST”\n"
     ]
    }
   ],
   "source": [
    "VariantAnnotation::info(d)$TEST = a$pval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0a0e8b7b-b814-49fd-8632-947087b30091",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'd' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'd' not found\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4863bddc-0c08-4199-870b-cd19f49cc63c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: CollapsedVCF \n",
       "dim: 102064 1 \n",
       "rowRanges(vcf):\n",
       "  GRanges with 4 metadata columns: REF, ALT, QUAL, FILTER\n",
       "info(vcf):\n",
       "  DataFrame with 1 column: TEST\n",
       "  Fields with no header: TEST \n",
       "geno(vcf):\n",
       "  List of length 3: ES, SE, LP\n",
       "geno(header(vcf)):\n",
       "      Number Type  Description                                            \n",
       "   ES A      Float Effect size estimate relative to the alternative allele\n",
       "   SE A      Float Standard error of effect size estimate                 \n",
       "   LP A      Float -log10 p-value for effect estimate                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "b\n",
    "VariantAnnotation::writeVcf(d,\"~/test.vcf\",index = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7d91144b-4c0b-451b-a23a-857b1b9328df",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = create_vcf(\n",
    "   chrom = a$`#chrom`,\n",
    "    pos = a$pos,\n",
    "    ea = a$ref,\n",
    "    nea = a$alt,\n",
    "  effect = cbind(a$se,a$beta+1),\n",
    "    pval = cbind(a$se,a$beta+1),\n",
    "    name = c(\"AC\",\"DLPFC_PCC\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c9d90fe6-8688-48af-a1fc-482e3c238e5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: CollapsedVCF \n",
       "dim: 102064 2 \n",
       "rowRanges(vcf):\n",
       "  GRanges with 4 metadata columns: REF, ALT, QUAL, FILTER\n",
       "info(vcf):\n",
       "  DataFrame with 0 columns: \n",
       "geno(vcf):\n",
       "  List of length 2: ES, LP\n",
       "geno(header(vcf)):\n",
       "      Number Type  Description                                            \n",
       "   ES A      Float Effect size estimate relative to the alternative allele\n",
       "   LP A      Float -log10 p-value for effect estimate                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0a43bd6a-9c23-44fa-80f7-a54f913d9c7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>102064</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 102064\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 102064\n",
       "2. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 102064      2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "fe7dbe7a-1be7-4360-8996-a3f745184145",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, \n",
       "<span style=white-space:pre-wrap>    se = NULL, pval = NULL, n = NULL, ncase = NULL, name = NULL) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    stopifnot(length(chrom) == length(pos))</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(snp)) {</span>\n",
       "<span style=white-space:pre-wrap>        snp &lt;- paste0(chrom, \":\", pos)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    nsnp &lt;- length(chrom)</span>\n",
       "<span style=white-space:pre-wrap>    gen &lt;- list()</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(ea_af)) </span>\n",
       "<span style=white-space:pre-wrap>        gen[[\"AF\"]] &lt;- matrix(ea_af, nsnp)</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(effect)) </span>\n",
       "<span style=white-space:pre-wrap>        gen[[\"ES\"]] &lt;- matrix(effect, nsnp)</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(se)) </span>\n",
       "<span style=white-space:pre-wrap>        gen[[\"SE\"]] &lt;- matrix(se, nsnp)</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(pval)) </span>\n",
       "<span style=white-space:pre-wrap>        gen[[\"LP\"]] &lt;- matrix(-log10(pval), nsnp)</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(n)) </span>\n",
       "<span style=white-space:pre-wrap>        gen[[\"SS\"]] &lt;- matrix(n, nsnp)</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(ncase)) </span>\n",
       "<span style=white-space:pre-wrap>        gen[[\"NC\"]] &lt;- matrix(ncase, nsnp)</span>\n",
       "<span style=white-space:pre-wrap>    gen &lt;- S4Vectors::SimpleList(gen)</span>\n",
       "<span style=white-space:pre-wrap>    gr &lt;- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, </span>\n",
       "<span style=white-space:pre-wrap>        end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))</span>\n",
       "<span style=white-space:pre-wrap>    coldata &lt;- S4Vectors::DataFrame(Samples = length(name), row.names = name)</span>\n",
       "<span style=white-space:pre-wrap>    hdr &lt;- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", </span>\n",
       "<span style=white-space:pre-wrap>        row.names = \"fileformat\")), sample = name)</span>\n",
       "<span style=white-space:pre-wrap>    VariantAnnotation::geno(hdr) &lt;- S4Vectors::DataFrame(Number = c(\"A\", </span>\n",
       "<span style=white-space:pre-wrap>        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", </span>\n",
       "<span style=white-space:pre-wrap>        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", </span>\n",
       "<span style=white-space:pre-wrap>        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", </span>\n",
       "<span style=white-space:pre-wrap>        \"Alternate allele frequency in the association study\", </span>\n",
       "<span style=white-space:pre-wrap>        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), </span>\n",
       "<span style=white-space:pre-wrap>        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))</span>\n",
       "<span style=white-space:pre-wrap>    VariantAnnotation::geno(hdr) &lt;- subset(VariantAnnotation::geno(hdr), </span>\n",
       "<span style=white-space:pre-wrap>        rownames(VariantAnnotation::geno(hdr)) %in% names(gen))</span>\n",
       "<span style=white-space:pre-wrap>    vcf &lt;- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, </span>\n",
       "<span style=white-space:pre-wrap>        exptData = list(header = hdr), geno = gen)</span>\n",
       "<span style=white-space:pre-wrap>    VariantAnnotation::alt(vcf) &lt;- Biostrings::DNAStringSetList(as.list(ea))</span>\n",
       "<span style=white-space:pre-wrap>    VariantAnnotation::ref(vcf) &lt;- Biostrings::DNAStringSet(nea)</span>\n",
       "<span style=white-space:pre-wrap>    VariantAnnotation::fixed(vcf)$FILTER &lt;- \"PASS\"</span>\n",
       "<span style=white-space:pre-wrap>    return(sort(vcf))</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (chrom, pos, nea, ea, snp = NULL, ea\\_af = NULL, effect = NULL, \n",
       "    se = NULL, pval = NULL, n = NULL, ncase = NULL, name = NULL) \n",
       "\\{\n",
       "    stopifnot(length(chrom) == length(pos))\n",
       "    if (is.null(snp)) \\{\n",
       "        snp <- paste0(chrom, \":\", pos)\n",
       "    \\}\n",
       "    nsnp <- length(chrom)\n",
       "    gen <- list()\n",
       "    if (!is.null(ea\\_af)) \n",
       "        gen{[}{[}\"AF\"{]}{]} <- matrix(ea\\_af, nsnp)\n",
       "    if (!is.null(effect)) \n",
       "        gen{[}{[}\"ES\"{]}{]} <- matrix(effect, nsnp)\n",
       "    if (!is.null(se)) \n",
       "        gen{[}{[}\"SE\"{]}{]} <- matrix(se, nsnp)\n",
       "    if (!is.null(pval)) \n",
       "        gen{[}{[}\"LP\"{]}{]} <- matrix(-log10(pval), nsnp)\n",
       "    if (!is.null(n)) \n",
       "        gen{[}{[}\"SS\"{]}{]} <- matrix(n, nsnp)\n",
       "    if (!is.null(ncase)) \n",
       "        gen{[}{[}\"NC\"{]}{]} <- matrix(ncase, nsnp)\n",
       "    gen <- S4Vectors::SimpleList(gen)\n",
       "    gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
       "        end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
       "    coldata <- S4Vectors::DataFrame(Samples = length(name), row.names = name)\n",
       "    hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
       "        row.names = \"fileformat\")), sample = name)\n",
       "    VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
       "        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
       "        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
       "        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", \n",
       "        \"Alternate allele frequency in the association study\", \n",
       "        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), \n",
       "        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))\n",
       "    VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
       "        rownames(VariantAnnotation::geno(hdr)) \\%in\\% names(gen))\n",
       "    vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
       "        exptData = list(header = hdr), geno = gen)\n",
       "    VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
       "    VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
       "    VariantAnnotation::fixed(vcf)\\$FILTER <- \"PASS\"\n",
       "    return(sort(vcf))\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, \n",
       "    se = NULL, pval = NULL, n = NULL, ncase = NULL, name = NULL) \n",
       "{\n",
       "    stopifnot(length(chrom) == length(pos))\n",
       "    if (is.null(snp)) {\n",
       "        snp <- paste0(chrom, \":\", pos)\n",
       "    }\n",
       "    nsnp <- length(chrom)\n",
       "    gen <- list()\n",
       "    if (!is.null(ea_af)) \n",
       "        gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
       "    if (!is.null(effect)) \n",
       "        gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
       "    if (!is.null(se)) \n",
       "        gen[[\"SE\"]] <- matrix(se, nsnp)\n",
       "    if (!is.null(pval)) \n",
       "        gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
       "    if (!is.null(n)) \n",
       "        gen[[\"SS\"]] <- matrix(n, nsnp)\n",
       "    if (!is.null(ncase)) \n",
       "        gen[[\"NC\"]] <- matrix(ncase, nsnp)\n",
       "    gen <- S4Vectors::SimpleList(gen)\n",
       "    gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
       "        end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
       "    coldata <- S4Vectors::DataFrame(Samples = length(name), row.names = name)\n",
       "    hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
       "        row.names = \"fileformat\")), sample = name)\n",
       "    VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
       "        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
       "        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
       "        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", \n",
       "        \"Alternate allele frequency in the association study\", \n",
       "        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), \n",
       "        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))\n",
       "    VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
       "        rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
       "    vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
       "        exptData = list(header = hdr), geno = gen)\n",
       "    VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
       "    VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
       "    VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\"\n",
       "    return(sort(vcf))\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, \n",
       "    se = NULL, pval = NULL, n = NULL, ncase = NULL, name = NULL) \n",
       "{\n",
       "    stopifnot(length(chrom) == length(pos))\n",
       "    if (is.null(snp)) {\n",
       "        snp <- paste0(chrom, \":\", pos)\n",
       "    }\n",
       "    nsnp <- length(chrom)\n",
       "    gen <- list()\n",
       "    if (!is.null(ea_af)) \n",
       "        gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
       "    if (!is.null(effect)) \n",
       "        gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
       "    if (!is.null(se)) \n",
       "        gen[[\"SE\"]] <- matrix(se, nsnp)\n",
       "    if (!is.null(pval)) \n",
       "        gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
       "    if (!is.null(n)) \n",
       "        gen[[\"SS\"]] <- matrix(n, nsnp)\n",
       "    if (!is.null(ncase)) \n",
       "        gen[[\"NC\"]] <- matrix(ncase, nsnp)\n",
       "    gen <- S4Vectors::SimpleList(gen)\n",
       "    gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
       "        end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
       "    coldata <- S4Vectors::DataFrame(Samples = length(name), row.names = name)\n",
       "    hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
       "        row.names = \"fileformat\")), sample = name)\n",
       "    VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
       "        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
       "        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
       "        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", \n",
       "        \"Alternate allele frequency in the association study\", \n",
       "        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), \n",
       "        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))\n",
       "    VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
       "        rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
       "    vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
       "        exptData = list(header = hdr), geno = gen)\n",
       "    VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
       "    VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
       "    VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\"\n",
       "    return(sort(vcf))\n",
       "}\n",
       "<bytecode: 0x558c86492638>\n",
       "<environment: namespace:gwasvcf>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "create_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6f206cdb-108d-4e48-b3bd-5b6a0c0e8aed",
   "metadata": {},
   "outputs": [],
   "source": [
    "effect = cbind(a$se,a$beta+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a27d7169-9f68-4923-8247-742a71cb2860",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 102064 x 2 of type dbl</caption>\n",
       "<tbody>\n",
       "\t<tr><td>0.02678740</td><td>1.0045513</td></tr>\n",
       "\t<tr><td>0.00863702</td><td>1.0191654</td></tr>\n",
       "\t<tr><td>0.07035390</td><td>1.0317659</td></tr>\n",
       "\t<tr><td>0.07023000</td><td>1.0879099</td></tr>\n",
       "\t<tr><td>0.07042450</td><td>0.9787639</td></tr>\n",
       "\t<tr><td>0.00561773</td><td>1.0015735</td></tr>\n",
       "\t<tr><td>0.00830861</td><td>0.9954964</td></tr>\n",
       "\t<tr><td>0.00400209</td><td>0.9949001</td></tr>\n",
       "\t<tr><td>0.00441183</td><td>0.9912124</td></tr>\n",
       "\t<tr><td>0.02500750</td><td>0.9926339</td></tr>\n",
       "\t<tr><td>0.01431390</td><td>1.0034420</td></tr>\n",
       "\t<tr><td>0.00975617</td><td>0.9795864</td></tr>\n",
       "\t<tr><td>0.00701652</td><td>0.9822078</td></tr>\n",
       "\t<tr><td>0.02507760</td><td>1.0006945</td></tr>\n",
       "\t<tr><td>0.07019540</td><td>1.2147080</td></tr>\n",
       "\t<tr><td>0.07034460</td><td>1.0543885</td></tr>\n",
       "\t<tr><td>0.00813211</td><td>0.9819778</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.07031120</td><td>0.9203712</td></tr>\n",
       "\t<tr><td>0.03518680</td><td>0.9399555</td></tr>\n",
       "\t<tr><td>0.00801391</td><td>0.9824509</td></tr>\n",
       "\t<tr><td>0.07043840</td><td>1.0530584</td></tr>\n",
       "\t<tr><td>0.01358680</td><td>0.9845577</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.04995950</td><td>1.0683947</td></tr>\n",
       "\t<tr><td>0.01358680</td><td>0.9845577</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.01358680</td><td>0.9845577</td></tr>\n",
       "\t<tr><td>...</td><td>...</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.05903760</td><td>1.0671182</td></tr>\n",
       "\t<tr><td>0.05887050</td><td>1.0197502</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.05910890</td><td>1.0514288</td></tr>\n",
       "\t<tr><td>0.00768803</td><td>0.9907117</td></tr>\n",
       "\t<tr><td>0.04161290</td><td>1.0078385</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.01105300</td><td>1.0137439</td></tr>\n",
       "\t<tr><td>0.05899770</td><td>1.0202583</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.00768803</td><td>0.9907117</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.05894790</td><td>1.0847073</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>0.00547492</td><td>0.9967546</td></tr>\n",
       "\t<tr><td>0.05888330</td><td>0.9520156</td></tr>\n",
       "\t<tr><td>0.05892490</td><td>1.0935775</td></tr>\n",
       "\t<tr><td>0.02945750</td><td>1.0222687</td></tr>\n",
       "\t<tr><td>0.00325643</td><td>1.0053196</td></tr>\n",
       "\t<tr><td>0.03401580</td><td>1.0113699</td></tr>\n",
       "\t<tr><td>0.05894710</td><td>1.0143673</td></tr>\n",
       "\t<tr><td>0.00768803</td><td>0.9907117</td></tr>\n",
       "\t<tr><td>0.05897960</td><td>1.0538694</td></tr>\n",
       "\t<tr><td>0.06068900</td><td>1.0365998</td></tr>\n",
       "\t<tr><td>0.05909110</td><td>1.0068038</td></tr>\n",
       "\t<tr><td>0.05908000</td><td>0.9643547</td></tr>\n",
       "\t<tr><td>0.00000000</td><td>1.0000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 102064 x 2 of type dbl\n",
       "\\begin{tabular}{ll}\n",
       "\t 0.02678740 & 1.0045513\\\\\n",
       "\t 0.00863702 & 1.0191654\\\\\n",
       "\t 0.07035390 & 1.0317659\\\\\n",
       "\t 0.07023000 & 1.0879099\\\\\n",
       "\t 0.07042450 & 0.9787639\\\\\n",
       "\t 0.00561773 & 1.0015735\\\\\n",
       "\t 0.00830861 & 0.9954964\\\\\n",
       "\t 0.00400209 & 0.9949001\\\\\n",
       "\t 0.00441183 & 0.9912124\\\\\n",
       "\t 0.02500750 & 0.9926339\\\\\n",
       "\t 0.01431390 & 1.0034420\\\\\n",
       "\t 0.00975617 & 0.9795864\\\\\n",
       "\t 0.00701652 & 0.9822078\\\\\n",
       "\t 0.02507760 & 1.0006945\\\\\n",
       "\t 0.07019540 & 1.2147080\\\\\n",
       "\t 0.07034460 & 1.0543885\\\\\n",
       "\t 0.00813211 & 0.9819778\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.07031120 & 0.9203712\\\\\n",
       "\t 0.03518680 & 0.9399555\\\\\n",
       "\t 0.00801391 & 0.9824509\\\\\n",
       "\t 0.07043840 & 1.0530584\\\\\n",
       "\t 0.01358680 & 0.9845577\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.04995950 & 1.0683947\\\\\n",
       "\t 0.01358680 & 0.9845577\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.01358680 & 0.9845577\\\\\n",
       "\t ... & ...\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.05903760 & 1.0671182\\\\\n",
       "\t 0.05887050 & 1.0197502\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.05910890 & 1.0514288\\\\\n",
       "\t 0.00768803 & 0.9907117\\\\\n",
       "\t 0.04161290 & 1.0078385\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.01105300 & 1.0137439\\\\\n",
       "\t 0.05899770 & 1.0202583\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.00768803 & 0.9907117\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.05894790 & 1.0847073\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\t 0.00547492 & 0.9967546\\\\\n",
       "\t 0.05888330 & 0.9520156\\\\\n",
       "\t 0.05892490 & 1.0935775\\\\\n",
       "\t 0.02945750 & 1.0222687\\\\\n",
       "\t 0.00325643 & 1.0053196\\\\\n",
       "\t 0.03401580 & 1.0113699\\\\\n",
       "\t 0.05894710 & 1.0143673\\\\\n",
       "\t 0.00768803 & 0.9907117\\\\\n",
       "\t 0.05897960 & 1.0538694\\\\\n",
       "\t 0.06068900 & 1.0365998\\\\\n",
       "\t 0.05909110 & 1.0068038\\\\\n",
       "\t 0.05908000 & 0.9643547\\\\\n",
       "\t 0.00000000 & 1.0000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 102064 x 2 of type dbl\n",
       "\n",
       "| 0.02678740 | 1.0045513 |\n",
       "| 0.00863702 | 1.0191654 |\n",
       "| 0.07035390 | 1.0317659 |\n",
       "| 0.07023000 | 1.0879099 |\n",
       "| 0.07042450 | 0.9787639 |\n",
       "| 0.00561773 | 1.0015735 |\n",
       "| 0.00830861 | 0.9954964 |\n",
       "| 0.00400209 | 0.9949001 |\n",
       "| 0.00441183 | 0.9912124 |\n",
       "| 0.02500750 | 0.9926339 |\n",
       "| 0.01431390 | 1.0034420 |\n",
       "| 0.00975617 | 0.9795864 |\n",
       "| 0.00701652 | 0.9822078 |\n",
       "| 0.02507760 | 1.0006945 |\n",
       "| 0.07019540 | 1.2147080 |\n",
       "| 0.07034460 | 1.0543885 |\n",
       "| 0.00813211 | 0.9819778 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.07031120 | 0.9203712 |\n",
       "| 0.03518680 | 0.9399555 |\n",
       "| 0.00801391 | 0.9824509 |\n",
       "| 0.07043840 | 1.0530584 |\n",
       "| 0.01358680 | 0.9845577 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.04995950 | 1.0683947 |\n",
       "| 0.01358680 | 0.9845577 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.01358680 | 0.9845577 |\n",
       "| ... | ... |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.05903760 | 1.0671182 |\n",
       "| 0.05887050 | 1.0197502 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.05910890 | 1.0514288 |\n",
       "| 0.00768803 | 0.9907117 |\n",
       "| 0.04161290 | 1.0078385 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.01105300 | 1.0137439 |\n",
       "| 0.05899770 | 1.0202583 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.00768803 | 0.9907117 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.05894790 | 1.0847073 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "| 0.00547492 | 0.9967546 |\n",
       "| 0.05888330 | 0.9520156 |\n",
       "| 0.05892490 | 1.0935775 |\n",
       "| 0.02945750 | 1.0222687 |\n",
       "| 0.00325643 | 1.0053196 |\n",
       "| 0.03401580 | 1.0113699 |\n",
       "| 0.05894710 | 1.0143673 |\n",
       "| 0.00768803 | 0.9907117 |\n",
       "| 0.05897960 | 1.0538694 |\n",
       "| 0.06068900 | 1.0365998 |\n",
       "| 0.05909110 | 1.0068038 |\n",
       "| 0.05908000 | 0.9643547 |\n",
       "| 0.00000000 | 1.0000000 |\n",
       "\n"
      ],
      "text/plain": [
       "      [,1]       [,2]     \n",
       " [1,] 0.02678740 1.0045513\n",
       " [2,] 0.00863702 1.0191654\n",
       " [3,] 0.07035390 1.0317659\n",
       " [4,] 0.07023000 1.0879099\n",
       " [5,] 0.07042450 0.9787639\n",
       " [6,] 0.00561773 1.0015735\n",
       " [7,] 0.00830861 0.9954964\n",
       " [8,] 0.00400209 0.9949001\n",
       " [9,] 0.00441183 0.9912124\n",
       "[10,] 0.02500750 0.9926339\n",
       "[11,] 0.01431390 1.0034420\n",
       "[12,] 0.00975617 0.9795864\n",
       "[13,] 0.00701652 0.9822078\n",
       "[14,] 0.02507760 1.0006945\n",
       "[15,] 0.07019540 1.2147080\n",
       "[16,] 0.07034460 1.0543885\n",
       "[17,] 0.00813211 0.9819778\n",
       "[18,] 0.00000000 1.0000000\n",
       "[19,] 0.07031120 0.9203712\n",
       "[20,] 0.03518680 0.9399555\n",
       "[21,] 0.00801391 0.9824509\n",
       "[22,] 0.07043840 1.0530584\n",
       "[23,] 0.01358680 0.9845577\n",
       "[24,] 0.00000000 1.0000000\n",
       "[25,] 0.00000000 1.0000000\n",
       "[26,] 0.00000000 1.0000000\n",
       "[27,] 0.04995950 1.0683947\n",
       "[28,] 0.01358680 0.9845577\n",
       "[29,] 0.00000000 1.0000000\n",
       "[30,] 0.01358680 0.9845577\n",
       "[31,] ...        ...      \n",
       "[32,] 0.00000000 1.0000000\n",
       "[33,] 0.05903760 1.0671182\n",
       "[34,] 0.05887050 1.0197502\n",
       "[35,] 0.00000000 1.0000000\n",
       "[36,] 0.05910890 1.0514288\n",
       "[37,] 0.00768803 0.9907117\n",
       "[38,] 0.04161290 1.0078385\n",
       "[39,] 0.00000000 1.0000000\n",
       "[40,] 0.01105300 1.0137439\n",
       "[41,] 0.05899770 1.0202583\n",
       "[42,] 0.00000000 1.0000000\n",
       "[43,] 0.00768803 0.9907117\n",
       "[44,] 0.00000000 1.0000000\n",
       "[45,] 0.00000000 1.0000000\n",
       "[46,] 0.05894790 1.0847073\n",
       "[47,] 0.00000000 1.0000000\n",
       "[48,] 0.00000000 1.0000000\n",
       "[49,] 0.00547492 0.9967546\n",
       "[50,] 0.05888330 0.9520156\n",
       "[51,] 0.05892490 1.0935775\n",
       "[52,] 0.02945750 1.0222687\n",
       "[53,] 0.00325643 1.0053196\n",
       "[54,] 0.03401580 1.0113699\n",
       "[55,] 0.05894710 1.0143673\n",
       "[56,] 0.00768803 0.9907117\n",
       "[57,] 0.05897960 1.0538694\n",
       "[58,] 0.06068900 1.0365998\n",
       "[59,] 0.05909110 1.0068038\n",
       "[60,] 0.05908000 0.9643547\n",
       "[61,] 0.00000000 1.0000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "matrix(effect,nrow(effect))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "385d402c-b94c-4448-8622-8c2b1d8ac8de",
   "metadata": {},
   "outputs": [],
   "source": [
    "rowRanges(k) = rowRanges(k)[,1:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "9ca530b4-9800-4187-babe-cb1686425b68",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in .::S4Vectors: unused argument (SimpleList)\n",
     "output_type": "error",
     "traceback": [
      "Error in .::S4Vectors: unused argument (SimpleList)\nTraceback:\n",
      "1. merge_vcf(b, d)",
      "2. lapply(names(VariantAnnotation::geno(ab)), function(x) rbind(VariantAnnotation::geno(ab)[x], \n .     VariantAnnotation::geno(b)[x])) %>% S4Vectors::SimpleList"
     ]
    }
   ],
   "source": [
    "merge_vcf(b,d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "347cb558-5baf-46c3-91b1-c70f52dca7c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede85c02-3a16-40bb-8954-07b40a6024f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cbind(b,d)-> k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "d3686048-6b20-4b74-97ed-0d125bf66565",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in all_dims[, 1L]: incorrect number of dimensions\n",
     "output_type": "error",
     "traceback": [
      "Error in all_dims[, 1L]: incorrect number of dimensions\nTraceback:\n",
      "1. VariantAnnotation::`geno<-`(`*tmp*`, value = `*vtmp*`)   # at line 13-14 of file <text>",
      "2. VariantAnnotation::`geno<-`(`*tmp*`, value = `*vtmp*`)",
      "3. `assays<-`(`*tmp*`, withDimnames = FALSE, value = value)",
      "4. `assays<-`(`*tmp*`, withDimnames = FALSE, value = value)",
      "5. Assays(value, as.null.if.zero.assay = TRUE)",
      "6. validObject(assays)",
      "7. anyStrings(validityMethod(as(object, superClass)))",
      "8. isTRUE(x)",
      "9. validityMethod(as(object, superClass))",
      "10. method(object)"
     ]
    }
   ],
   "source": [
    "a <- VariantAnnotation::expand(a)\n",
    "    b <- VariantAnnotation::expand(b)\n",
    "    o <- dplyr::tibble(from = which(names(a) %in% names(b)), \n",
    "        to = match(names(a)[.data$from], names(b)))\n",
    "    a <- a[o[[\"from\"]], ]\n",
    "    b <- b[o[[\"to\"]], ]\n",
    "    allele_match <- VariantAnnotation::ref(a) == VariantAnnotation::ref(b) & \n",
    "        VariantAnnotation::alt(a) == VariantAnnotation::alt(b)\n",
    "    switch <- VariantAnnotation::ref(a) == VariantAnnotation::alt(b) & \n",
    "        VariantAnnotation::ref(b) == VariantAnnotation::alt(a)\n",
    "    if (any(switch)) {\n",
    "        for (i in 1:ncol(VariantAnnotation::geno(b)[[\"ES\"]])) {\n",
    "            VariantAnnotation::geno(b)[[\"ES\"]][, i][switch] <- lapply(VariantAnnotation::geno(b)[[\"ES\"]][, \n",
    "                i][switch], function(x) x * -1)\n",
    "        }\n",
    "    }\n",
    "    a <- a[allele_match | switch, ]\n",
    "    b <- b[allele_match | switch, ]\n",
    "    ab <- a\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "e4debb11-657e-451b-b2d1-f824b25d8b87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "78900"
      ],
      "text/latex": [
       "78900"
      ],
      "text/markdown": [
       "78900"
      ],
      "text/plain": [
       "[1] 78900"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sum(switch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "23c97716-c26a-492b-8fe4-2dd0fc4698f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "    temp <- lapply(names(VariantAnnotation::geno(ab)), function(x) rbind(VariantAnnotation::geno(ab)[[x]], \n",
    "        VariantAnnotation::geno(b)[x])) %>% S4Vectors::SimpleList()\n",
    "    names(temp) <- names(VariantAnnotation::geno(ab))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "db310ec3-3ae3-4237-b4db-4d17758ba832",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp <- lapply(names(VariantAnnotation::geno(ab)), function(x) rbind(VariantAnnotation::geno(ab)[[x]], \n",
    "        VariantAnnotation::geno(b)[[x]]))%>%SimpleList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "163ae623-ae2c-4e60-a9b9-bbd85fa3ebd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "List of length 3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "fd92a074-9da0-40ba-b50c-5280db8857c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (a, b) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    a &lt;- VariantAnnotation::expand(a)</span>\n",
       "<span style=white-space:pre-wrap>    b &lt;- VariantAnnotation::expand(b)</span>\n",
       "<span style=white-space:pre-wrap>    o &lt;- dplyr::tibble(from = which(names(a) %in% names(b)), </span>\n",
       "<span style=white-space:pre-wrap>        to = match(names(a)[.data$from], names(b)))</span>\n",
       "<span style=white-space:pre-wrap>    a &lt;- a[o[[\"from\"]], ]</span>\n",
       "<span style=white-space:pre-wrap>    b &lt;- b[o[[\"to\"]], ]</span>\n",
       "<span style=white-space:pre-wrap>    allele_match &lt;- VariantAnnotation::ref(a) == VariantAnnotation::ref(b) &amp; </span>\n",
       "<span style=white-space:pre-wrap>        VariantAnnotation::alt(a) == VariantAnnotation::alt(b)</span>\n",
       "<span style=white-space:pre-wrap>    switch &lt;- VariantAnnotation::ref(a) == VariantAnnotation::alt(b) &amp; </span>\n",
       "<span style=white-space:pre-wrap>        VariantAnnotation::ref(b) == VariantAnnotation::alt(a)</span>\n",
       "<span style=white-space:pre-wrap>    if (any(switch)) {</span>\n",
       "<span style=white-space:pre-wrap>        for (i in 1:ncol(VariantAnnotation::geno(b)[[\"ES\"]])) {</span>\n",
       "<span style=white-space:pre-wrap>            VariantAnnotation::geno(b)[[\"ES\"]][, i][switch] &lt;- lapply(VariantAnnotation::geno(b)[[\"ES\"]][, </span>\n",
       "<span style=white-space:pre-wrap>                i][switch], function(x) x * -1)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    a &lt;- a[allele_match | switch, ]</span>\n",
       "<span style=white-space:pre-wrap>    b &lt;- b[allele_match | switch, ]</span>\n",
       "<span style=white-space:pre-wrap>    ab &lt;- a</span>\n",
       "<span style=white-space:pre-wrap>    temp &lt;- lapply(names(VariantAnnotation::geno(ab)), function(x) rbind(VariantAnnotation::geno(ab)[x], </span>\n",
       "<span style=white-space:pre-wrap>        VariantAnnotation::geno(b)[x])) %&gt;% S4Vectors::SimpleList</span>\n",
       "<span style=white-space:pre-wrap>    names(temp) &lt;- names(VariantAnnotation::geno(ab))</span>\n",
       "<span style=white-space:pre-wrap>    VariantAnnotation::geno(ab) &lt;- temp</span>\n",
       "<span style=white-space:pre-wrap>    h &lt;- VariantAnnotation::header(a)</span>\n",
       "<span style=white-space:pre-wrap>    VCFHeader(reference = VariantAnnotation::reference(h), samples = c(VariantAnnotation::samples(h), </span>\n",
       "<span style=white-space:pre-wrap>        VariantAnnotation::samples(VariantAnnotation::header(b))), </span>\n",
       "<span style=white-space:pre-wrap>        meta = VariantAnnotation::meta(h))</span>\n",
       "<span style=white-space:pre-wrap>    VariantAnnotation::samples(h) &lt;- c(VariantAnnotation::samples(h), </span>\n",
       "<span style=white-space:pre-wrap>        VariantAnnotation::samples(VariantAnnotation::header(b)))</span>\n",
       "<span style=white-space:pre-wrap>    return(S4Vectors::SimpleList())</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (a, b) \n",
       "\\{\n",
       "    a <- VariantAnnotation::expand(a)\n",
       "    b <- VariantAnnotation::expand(b)\n",
       "    o <- dplyr::tibble(from = which(names(a) \\%in\\% names(b)), \n",
       "        to = match(names(a){[}.data\\$from{]}, names(b)))\n",
       "    a <- a{[}o{[}{[}\"from\"{]}{]}, {]}\n",
       "    b <- b{[}o{[}{[}\"to\"{]}{]}, {]}\n",
       "    allele\\_match <- VariantAnnotation::ref(a) == VariantAnnotation::ref(b) \\& \n",
       "        VariantAnnotation::alt(a) == VariantAnnotation::alt(b)\n",
       "    switch <- VariantAnnotation::ref(a) == VariantAnnotation::alt(b) \\& \n",
       "        VariantAnnotation::ref(b) == VariantAnnotation::alt(a)\n",
       "    if (any(switch)) \\{\n",
       "        for (i in 1:ncol(VariantAnnotation::geno(b){[}{[}\"ES\"{]}{]})) \\{\n",
       "            VariantAnnotation::geno(b){[}{[}\"ES\"{]}{]}{[}, i{]}{[}switch{]} <- lapply(VariantAnnotation::geno(b){[}{[}\"ES\"{]}{]}{[}, \n",
       "                i{]}{[}switch{]}, function(x) x * -1)\n",
       "        \\}\n",
       "    \\}\n",
       "    a <- a{[}allele\\_match \\textbar{} switch, {]}\n",
       "    b <- b{[}allele\\_match \\textbar{} switch, {]}\n",
       "    ab <- a\n",
       "    temp <- lapply(names(VariantAnnotation::geno(ab)), function(x) rbind(VariantAnnotation::geno(ab){[}x{]}, \n",
       "        VariantAnnotation::geno(b){[}x{]})) \\%>\\% S4Vectors::SimpleList\n",
       "    names(temp) <- names(VariantAnnotation::geno(ab))\n",
       "    VariantAnnotation::geno(ab) <- temp\n",
       "    h <- VariantAnnotation::header(a)\n",
       "    VCFHeader(reference = VariantAnnotation::reference(h), samples = c(VariantAnnotation::samples(h), \n",
       "        VariantAnnotation::samples(VariantAnnotation::header(b))), \n",
       "        meta = VariantAnnotation::meta(h))\n",
       "    VariantAnnotation::samples(h) <- c(VariantAnnotation::samples(h), \n",
       "        VariantAnnotation::samples(VariantAnnotation::header(b)))\n",
       "    return(S4Vectors::SimpleList())\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (a, b) \n",
       "{\n",
       "    a <- VariantAnnotation::expand(a)\n",
       "    b <- VariantAnnotation::expand(b)\n",
       "    o <- dplyr::tibble(from = which(names(a) %in% names(b)), \n",
       "        to = match(names(a)[.data$from], names(b)))\n",
       "    a <- a[o[[\"from\"]], ]\n",
       "    b <- b[o[[\"to\"]], ]\n",
       "    allele_match <- VariantAnnotation::ref(a) == VariantAnnotation::ref(b) & \n",
       "        VariantAnnotation::alt(a) == VariantAnnotation::alt(b)\n",
       "    switch <- VariantAnnotation::ref(a) == VariantAnnotation::alt(b) & \n",
       "        VariantAnnotation::ref(b) == VariantAnnotation::alt(a)\n",
       "    if (any(switch)) {\n",
       "        for (i in 1:ncol(VariantAnnotation::geno(b)[[\"ES\"]])) {\n",
       "            VariantAnnotation::geno(b)[[\"ES\"]][, i][switch] <- lapply(VariantAnnotation::geno(b)[[\"ES\"]][, \n",
       "                i][switch], function(x) x * -1)\n",
       "        }\n",
       "    }\n",
       "    a <- a[allele_match | switch, ]\n",
       "    b <- b[allele_match | switch, ]\n",
       "    ab <- a\n",
       "    temp <- lapply(names(VariantAnnotation::geno(ab)), function(x) rbind(VariantAnnotation::geno(ab)[x], \n",
       "        VariantAnnotation::geno(b)[x])) %>% S4Vectors::SimpleList\n",
       "    names(temp) <- names(VariantAnnotation::geno(ab))\n",
       "    VariantAnnotation::geno(ab) <- temp\n",
       "    h <- VariantAnnotation::header(a)\n",
       "    VCFHeader(reference = VariantAnnotation::reference(h), samples = c(VariantAnnotation::samples(h), \n",
       "        VariantAnnotation::samples(VariantAnnotation::header(b))), \n",
       "        meta = VariantAnnotation::meta(h))\n",
       "    VariantAnnotation::samples(h) <- c(VariantAnnotation::samples(h), \n",
       "        VariantAnnotation::samples(VariantAnnotation::header(b)))\n",
       "    return(S4Vectors::SimpleList())\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (a, b) \n",
       "{\n",
       "    a <- VariantAnnotation::expand(a)\n",
       "    b <- VariantAnnotation::expand(b)\n",
       "    o <- dplyr::tibble(from = which(names(a) %in% names(b)), \n",
       "        to = match(names(a)[.data$from], names(b)))\n",
       "    a <- a[o[[\"from\"]], ]\n",
       "    b <- b[o[[\"to\"]], ]\n",
       "    allele_match <- VariantAnnotation::ref(a) == VariantAnnotation::ref(b) & \n",
       "        VariantAnnotation::alt(a) == VariantAnnotation::alt(b)\n",
       "    switch <- VariantAnnotation::ref(a) == VariantAnnotation::alt(b) & \n",
       "        VariantAnnotation::ref(b) == VariantAnnotation::alt(a)\n",
       "    if (any(switch)) {\n",
       "        for (i in 1:ncol(VariantAnnotation::geno(b)[[\"ES\"]])) {\n",
       "            VariantAnnotation::geno(b)[[\"ES\"]][, i][switch] <- lapply(VariantAnnotation::geno(b)[[\"ES\"]][, \n",
       "                i][switch], function(x) x * -1)\n",
       "        }\n",
       "    }\n",
       "    a <- a[allele_match | switch, ]\n",
       "    b <- b[allele_match | switch, ]\n",
       "    ab <- a\n",
       "    temp <- lapply(names(VariantAnnotation::geno(ab)), function(x) rbind(VariantAnnotation::geno(ab)[x], \n",
       "        VariantAnnotation::geno(b)[x])) %>% S4Vectors::SimpleList\n",
       "    names(temp) <- names(VariantAnnotation::geno(ab))\n",
       "    VariantAnnotation::geno(ab) <- temp\n",
       "    h <- VariantAnnotation::header(a)\n",
       "    VCFHeader(reference = VariantAnnotation::reference(h), samples = c(VariantAnnotation::samples(h), \n",
       "        VariantAnnotation::samples(VariantAnnotation::header(b))), \n",
       "        meta = VariantAnnotation::meta(h))\n",
       "    VariantAnnotation::samples(h) <- c(VariantAnnotation::samples(h), \n",
       "        VariantAnnotation::samples(VariantAnnotation::header(b)))\n",
       "    return(S4Vectors::SimpleList())\n",
       "}\n",
       "<bytecode: 0x55a09e013a88>\n",
       "<environment: namespace:gwasvcf>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "518686c9-5389-46f5-8160-f01618e5cd3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_vcf = function (a, b) \n",
    "{\n",
    "    a <- VariantAnnotation::expand(a)\n",
    "    b <- VariantAnnotation::expand(b)\n",
    "    o <- dplyr::tibble(from = which(names(a) %in% names(b)), \n",
    "        to = match(names(a)[.data$from], names(b)))\n",
    "    a <- a[o[[\"from\"]], ]\n",
    "    b <- b[o[[\"to\"]], ]\n",
    "    allele_match <- VariantAnnotation::ref(a) == VariantAnnotation::ref(b) & \n",
    "        VariantAnnotation::alt(a) == VariantAnnotation::alt(b)\n",
    "    switch <- VariantAnnotation::ref(a) == VariantAnnotation::alt(b) & \n",
    "        VariantAnnotation::ref(b) == VariantAnnotation::alt(a)\n",
    "    if (any(switch)) {\n",
    "        for (i in 1:ncol(VariantAnnotation::geno(b)[[\"ES\"]])) {\n",
    "            VariantAnnotation::geno(b)[[\"ES\"]][, i][switch] <- lapply(VariantAnnotation::geno(b)[[\"ES\"]][, \n",
    "                i][switch], function(x) x * -1)\n",
    "        }\n",
    "    }\n",
    "    a <- a[allele_match | switch, ]\n",
    "    b <- b[allele_match | switch, ]\n",
    "    ab <- a\n",
    "    temp <- lapply(names(VariantAnnotation::geno(ab)), function(x) cbind(VariantAnnotation::geno(ab)[[x]], \n",
    "        VariantAnnotation::geno(b)[[x]])) %>% S4Vectors::SimpleList()\n",
    "    names(temp) <- names(VariantAnnotation::geno(ab))\n",
    "    VariantAnnotation::geno(ab) <- temp\n",
    "    h <- VariantAnnotation::header(a)\n",
    "    VCFHeader(reference = VariantAnnotation::reference(h), samples = c(VariantAnnotation::samples(h), \n",
    "        VariantAnnotation::samples(VariantAnnotation::header(b))), \n",
    "        meta = VariantAnnotation::meta(h))\n",
    "    VariantAnnotation::samples(h) <- c(VariantAnnotation::samples(h), \n",
    "        VariantAnnotation::samples(VariantAnnotation::header(b)))\n",
    "    return(S4Vectors::SimpleList())\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "6303aafd-0bc1-4e4a-8677-56edb6221eb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "List of length 3\n",
       "names(3): ES SE LP"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "    VariantAnnotation::geno(ab) <- temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "db9ed618-0315-43bc-92de-65564a7bf110",
   "metadata": {},
   "outputs": [],
   "source": [
    "tet = temp\n",
    "tet$ES = temp$ES[[1]]\n",
    "tet$SE = temp$SE[[1]]\n",
    "tet$LP = temp$LP[[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "8e0949c8-f9a1-47a3-ac3f-498864388006",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 78900 x 1 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>TTT</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>21:30873191</th><td> 0.004551320</td></tr>\n",
       "\t<tr><th scope=row>21:30873237</th><td> 0.019165400</td></tr>\n",
       "\t<tr><th scope=row>21:30873308</th><td> 0.031765900</td></tr>\n",
       "\t<tr><th scope=row>21:30873319</th><td> 0.087909900</td></tr>\n",
       "\t<tr><th scope=row>21:30873432</th><td>-0.021236100</td></tr>\n",
       "\t<tr><th scope=row>21:30873441</th><td> 0.001573480</td></tr>\n",
       "\t<tr><th scope=row>21:30873445</th><td>-0.004503620</td></tr>\n",
       "\t<tr><th scope=row>21:30873446</th><td>-0.005099940</td></tr>\n",
       "\t<tr><th scope=row>21:30873447</th><td>-0.008787590</td></tr>\n",
       "\t<tr><th scope=row>21:30873449</th><td>-0.007366060</td></tr>\n",
       "\t<tr><th scope=row>21:30873450</th><td> 0.003442040</td></tr>\n",
       "\t<tr><th scope=row>21:30873452</th><td>-0.020413600</td></tr>\n",
       "\t<tr><th scope=row>21:30873453</th><td>-0.017792200</td></tr>\n",
       "\t<tr><th scope=row>21:30873467</th><td> 0.000694478</td></tr>\n",
       "\t<tr><th scope=row>21:30873514</th><td> 0.214708000</td></tr>\n",
       "\t<tr><th scope=row>21:30873598</th><td> 0.054388500</td></tr>\n",
       "\t<tr><th scope=row>21:30873668</th><td>-0.018022200</td></tr>\n",
       "\t<tr><th scope=row>21:30873820</th><td>-0.079628800</td></tr>\n",
       "\t<tr><th scope=row>21:30873834</th><td>-0.060044500</td></tr>\n",
       "\t<tr><th scope=row>21:30873877</th><td>-0.017549100</td></tr>\n",
       "\t<tr><th scope=row>21:30873901</th><td> 0.053058400</td></tr>\n",
       "\t<tr><th scope=row>21:30873969</th><td>-0.015442300</td></tr>\n",
       "\t<tr><th scope=row>21:30874189</th><td> 0.068394700</td></tr>\n",
       "\t<tr><th scope=row>21:30874520</th><td>-0.015442300</td></tr>\n",
       "\t<tr><th scope=row>21:30874599</th><td>-0.015442300</td></tr>\n",
       "\t<tr><th scope=row>21:30874625</th><td> 0.015948000</td></tr>\n",
       "\t<tr><th scope=row>21:30874736</th><td>-0.013622100</td></tr>\n",
       "\t<tr><th scope=row>21:30874842</th><td> 0.039402700</td></tr>\n",
       "\t<tr><th scope=row>21:30874864</th><td>-0.077661000</td></tr>\n",
       "\t<tr><th scope=row>21:30874896</th><td> 0.006205330</td></tr>\n",
       "\t<tr><th scope=row>...</th><td>...</td></tr>\n",
       "\t<tr><th scope=row>21:45325701</th><td>-0.076992300</td></tr>\n",
       "\t<tr><th scope=row>21:45325708</th><td> 0.019555700</td></tr>\n",
       "\t<tr><th scope=row>21:45325792</th><td> 0.113967000</td></tr>\n",
       "\t<tr><th scope=row>21:45325803</th><td> 0.063246100</td></tr>\n",
       "\t<tr><th scope=row>21:45325808</th><td> 0.011982100</td></tr>\n",
       "\t<tr><th scope=row>21:45325862</th><td> 0.019102800</td></tr>\n",
       "\t<tr><th scope=row>21:45325916</th><td>-0.000255327</td></tr>\n",
       "\t<tr><th scope=row>21:45326033</th><td> 0.037275200</td></tr>\n",
       "\t<tr><th scope=row>21:45326122</th><td>-0.060114200</td></tr>\n",
       "\t<tr><th scope=row>21:45326139</th><td> 0.067118200</td></tr>\n",
       "\t<tr><th scope=row>21:45326143</th><td> 0.019750200</td></tr>\n",
       "\t<tr><th scope=row>21:45326214</th><td> 0.051428800</td></tr>\n",
       "\t<tr><th scope=row>21:45326398</th><td>-0.009288300</td></tr>\n",
       "\t<tr><th scope=row>21:45326458</th><td> 0.007838520</td></tr>\n",
       "\t<tr><th scope=row>21:45326609</th><td> 0.013743900</td></tr>\n",
       "\t<tr><th scope=row>21:45326657</th><td> 0.020258300</td></tr>\n",
       "\t<tr><th scope=row>21:45326665</th><td>-0.009288300</td></tr>\n",
       "\t<tr><th scope=row>21:45326804</th><td> 0.084707300</td></tr>\n",
       "\t<tr><th scope=row>21:45326910</th><td>-0.003245450</td></tr>\n",
       "\t<tr><th scope=row>21:45326914</th><td>-0.047984400</td></tr>\n",
       "\t<tr><th scope=row>21:45326915</th><td> 0.093577500</td></tr>\n",
       "\t<tr><th scope=row>21:45326931</th><td> 0.022268700</td></tr>\n",
       "\t<tr><th scope=row>21:45326939</th><td> 0.005319590</td></tr>\n",
       "\t<tr><th scope=row>21:45326975</th><td> 0.011369900</td></tr>\n",
       "\t<tr><th scope=row>21:45326976</th><td> 0.014367300</td></tr>\n",
       "\t<tr><th scope=row>21:45327035</th><td>-0.009288300</td></tr>\n",
       "\t<tr><th scope=row>21:45327054</th><td> 0.053869400</td></tr>\n",
       "\t<tr><th scope=row>21:45327109</th><td> 0.036599800</td></tr>\n",
       "\t<tr><th scope=row>21:45327110</th><td> 0.006803810</td></tr>\n",
       "\t<tr><th scope=row>21:45327186</th><td>-0.035645300</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 78900 x 1 of type dbl\n",
       "\\begin{tabular}{r|l}\n",
       "  & TTT\\\\\n",
       "\\hline\n",
       "\t21:30873191 &  0.004551320\\\\\n",
       "\t21:30873237 &  0.019165400\\\\\n",
       "\t21:30873308 &  0.031765900\\\\\n",
       "\t21:30873319 &  0.087909900\\\\\n",
       "\t21:30873432 & -0.021236100\\\\\n",
       "\t21:30873441 &  0.001573480\\\\\n",
       "\t21:30873445 & -0.004503620\\\\\n",
       "\t21:30873446 & -0.005099940\\\\\n",
       "\t21:30873447 & -0.008787590\\\\\n",
       "\t21:30873449 & -0.007366060\\\\\n",
       "\t21:30873450 &  0.003442040\\\\\n",
       "\t21:30873452 & -0.020413600\\\\\n",
       "\t21:30873453 & -0.017792200\\\\\n",
       "\t21:30873467 &  0.000694478\\\\\n",
       "\t21:30873514 &  0.214708000\\\\\n",
       "\t21:30873598 &  0.054388500\\\\\n",
       "\t21:30873668 & -0.018022200\\\\\n",
       "\t21:30873820 & -0.079628800\\\\\n",
       "\t21:30873834 & -0.060044500\\\\\n",
       "\t21:30873877 & -0.017549100\\\\\n",
       "\t21:30873901 &  0.053058400\\\\\n",
       "\t21:30873969 & -0.015442300\\\\\n",
       "\t21:30874189 &  0.068394700\\\\\n",
       "\t21:30874520 & -0.015442300\\\\\n",
       "\t21:30874599 & -0.015442300\\\\\n",
       "\t21:30874625 &  0.015948000\\\\\n",
       "\t21:30874736 & -0.013622100\\\\\n",
       "\t21:30874842 &  0.039402700\\\\\n",
       "\t21:30874864 & -0.077661000\\\\\n",
       "\t21:30874896 &  0.006205330\\\\\n",
       "\t... & ...\\\\\n",
       "\t21:45325701 & -0.076992300\\\\\n",
       "\t21:45325708 &  0.019555700\\\\\n",
       "\t21:45325792 &  0.113967000\\\\\n",
       "\t21:45325803 &  0.063246100\\\\\n",
       "\t21:45325808 &  0.011982100\\\\\n",
       "\t21:45325862 &  0.019102800\\\\\n",
       "\t21:45325916 & -0.000255327\\\\\n",
       "\t21:45326033 &  0.037275200\\\\\n",
       "\t21:45326122 & -0.060114200\\\\\n",
       "\t21:45326139 &  0.067118200\\\\\n",
       "\t21:45326143 &  0.019750200\\\\\n",
       "\t21:45326214 &  0.051428800\\\\\n",
       "\t21:45326398 & -0.009288300\\\\\n",
       "\t21:45326458 &  0.007838520\\\\\n",
       "\t21:45326609 &  0.013743900\\\\\n",
       "\t21:45326657 &  0.020258300\\\\\n",
       "\t21:45326665 & -0.009288300\\\\\n",
       "\t21:45326804 &  0.084707300\\\\\n",
       "\t21:45326910 & -0.003245450\\\\\n",
       "\t21:45326914 & -0.047984400\\\\\n",
       "\t21:45326915 &  0.093577500\\\\\n",
       "\t21:45326931 &  0.022268700\\\\\n",
       "\t21:45326939 &  0.005319590\\\\\n",
       "\t21:45326975 &  0.011369900\\\\\n",
       "\t21:45326976 &  0.014367300\\\\\n",
       "\t21:45327035 & -0.009288300\\\\\n",
       "\t21:45327054 &  0.053869400\\\\\n",
       "\t21:45327109 &  0.036599800\\\\\n",
       "\t21:45327110 &  0.006803810\\\\\n",
       "\t21:45327186 & -0.035645300\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 78900 x 1 of type dbl\n",
       "\n",
       "| <!--/--> | TTT |\n",
       "|---|---|\n",
       "| 21:30873191 |  0.004551320 |\n",
       "| 21:30873237 |  0.019165400 |\n",
       "| 21:30873308 |  0.031765900 |\n",
       "| 21:30873319 |  0.087909900 |\n",
       "| 21:30873432 | -0.021236100 |\n",
       "| 21:30873441 |  0.001573480 |\n",
       "| 21:30873445 | -0.004503620 |\n",
       "| 21:30873446 | -0.005099940 |\n",
       "| 21:30873447 | -0.008787590 |\n",
       "| 21:30873449 | -0.007366060 |\n",
       "| 21:30873450 |  0.003442040 |\n",
       "| 21:30873452 | -0.020413600 |\n",
       "| 21:30873453 | -0.017792200 |\n",
       "| 21:30873467 |  0.000694478 |\n",
       "| 21:30873514 |  0.214708000 |\n",
       "| 21:30873598 |  0.054388500 |\n",
       "| 21:30873668 | -0.018022200 |\n",
       "| 21:30873820 | -0.079628800 |\n",
       "| 21:30873834 | -0.060044500 |\n",
       "| 21:30873877 | -0.017549100 |\n",
       "| 21:30873901 |  0.053058400 |\n",
       "| 21:30873969 | -0.015442300 |\n",
       "| 21:30874189 |  0.068394700 |\n",
       "| 21:30874520 | -0.015442300 |\n",
       "| 21:30874599 | -0.015442300 |\n",
       "| 21:30874625 |  0.015948000 |\n",
       "| 21:30874736 | -0.013622100 |\n",
       "| 21:30874842 |  0.039402700 |\n",
       "| 21:30874864 | -0.077661000 |\n",
       "| 21:30874896 |  0.006205330 |\n",
       "| ... | ... |\n",
       "| 21:45325701 | -0.076992300 |\n",
       "| 21:45325708 |  0.019555700 |\n",
       "| 21:45325792 |  0.113967000 |\n",
       "| 21:45325803 |  0.063246100 |\n",
       "| 21:45325808 |  0.011982100 |\n",
       "| 21:45325862 |  0.019102800 |\n",
       "| 21:45325916 | -0.000255327 |\n",
       "| 21:45326033 |  0.037275200 |\n",
       "| 21:45326122 | -0.060114200 |\n",
       "| 21:45326139 |  0.067118200 |\n",
       "| 21:45326143 |  0.019750200 |\n",
       "| 21:45326214 |  0.051428800 |\n",
       "| 21:45326398 | -0.009288300 |\n",
       "| 21:45326458 |  0.007838520 |\n",
       "| 21:45326609 |  0.013743900 |\n",
       "| 21:45326657 |  0.020258300 |\n",
       "| 21:45326665 | -0.009288300 |\n",
       "| 21:45326804 |  0.084707300 |\n",
       "| 21:45326910 | -0.003245450 |\n",
       "| 21:45326914 | -0.047984400 |\n",
       "| 21:45326915 |  0.093577500 |\n",
       "| 21:45326931 |  0.022268700 |\n",
       "| 21:45326939 |  0.005319590 |\n",
       "| 21:45326975 |  0.011369900 |\n",
       "| 21:45326976 |  0.014367300 |\n",
       "| 21:45327035 | -0.009288300 |\n",
       "| 21:45327054 |  0.053869400 |\n",
       "| 21:45327109 |  0.036599800 |\n",
       "| 21:45327110 |  0.006803810 |\n",
       "| 21:45327186 | -0.035645300 |\n",
       "\n"
      ],
      "text/plain": [
       "            TTT         \n",
       "21:30873191  0.004551320\n",
       "21:30873237  0.019165400\n",
       "21:30873308  0.031765900\n",
       "21:30873319  0.087909900\n",
       "21:30873432 -0.021236100\n",
       "21:30873441  0.001573480\n",
       "21:30873445 -0.004503620\n",
       "21:30873446 -0.005099940\n",
       "21:30873447 -0.008787590\n",
       "21:30873449 -0.007366060\n",
       "21:30873450  0.003442040\n",
       "21:30873452 -0.020413600\n",
       "21:30873453 -0.017792200\n",
       "21:30873467  0.000694478\n",
       "21:30873514  0.214708000\n",
       "21:30873598  0.054388500\n",
       "21:30873668 -0.018022200\n",
       "21:30873820 -0.079628800\n",
       "21:30873834 -0.060044500\n",
       "21:30873877 -0.017549100\n",
       "21:30873901  0.053058400\n",
       "21:30873969 -0.015442300\n",
       "21:30874189  0.068394700\n",
       "21:30874520 -0.015442300\n",
       "21:30874599 -0.015442300\n",
       "21:30874625  0.015948000\n",
       "21:30874736 -0.013622100\n",
       "21:30874842  0.039402700\n",
       "21:30874864 -0.077661000\n",
       "21:30874896  0.006205330\n",
       "...         ...         \n",
       "21:45325701 -0.076992300\n",
       "21:45325708  0.019555700\n",
       "21:45325792  0.113967000\n",
       "21:45325803  0.063246100\n",
       "21:45325808  0.011982100\n",
       "21:45325862  0.019102800\n",
       "21:45325916 -0.000255327\n",
       "21:45326033  0.037275200\n",
       "21:45326122 -0.060114200\n",
       "21:45326139  0.067118200\n",
       "21:45326143  0.019750200\n",
       "21:45326214  0.051428800\n",
       "21:45326398 -0.009288300\n",
       "21:45326458  0.007838520\n",
       "21:45326609  0.013743900\n",
       "21:45326657  0.020258300\n",
       "21:45326665 -0.009288300\n",
       "21:45326804  0.084707300\n",
       "21:45326910 -0.003245450\n",
       "21:45326914 -0.047984400\n",
       "21:45326915  0.093577500\n",
       "21:45326931  0.022268700\n",
       "21:45326939  0.005319590\n",
       "21:45326975  0.011369900\n",
       "21:45326976  0.014367300\n",
       "21:45327035 -0.009288300\n",
       "21:45327054  0.053869400\n",
       "21:45327109  0.036599800\n",
       "21:45327110  0.006803810\n",
       "21:45327186 -0.035645300"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "e596da8e-a7a0-4a08-bec9-f1033b170006",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: CollapsedVCF \n",
       "dim: 102064 1 \n",
       "rowRanges(vcf):\n",
       "  GRanges with 4 metadata columns: REF, ALT, QUAL, FILTER\n",
       "info(vcf):\n",
       "  DataFrame with 0 columns: \n",
       "geno(vcf):\n",
       "  List of length 3: ES, SE, LP\n",
       "geno(header(vcf)):\n",
       "      Number Type  Description                                            \n",
       "   ES A      Float Effect size estimate relative to the alternative allele\n",
       "   SE A      Float Standard error of effect size estimate                 \n",
       "   LP A      Float -log10 p-value for effect estimate                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d776db-8d61-441d-8ef1-b6e530080c0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0238e106-80bd-4180-95f6-ed398fa4e296",
   "metadata": {},
   "outputs": [],
   "source": [
    "c = vcf_to_tibble(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "1c76ef69-d67a-450d-a0fa-e1d78e1813c5",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:2:0: unexpected end of input\n1: d = \n   ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:2:0: unexpected end of input\n1: d = \n   ^\nTraceback:\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "a407b73a-2294-456b-9341-3b5901c1dad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "   chrom = a$`#chrom`\n",
    "    pos = a$pos\n",
    "    ea = a$ref\n",
    "    nea = a$alt\n",
    "  effect = cbind(a$se,a$beta+1)\n",
    "  se =  cbind(a$se,a$beta+1)\n",
    "    pval = cbind(a$se,a$beta+1)\n",
    "    name = c(\"sss\",\"qqq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "878cb6c4-438c-427b-8673-76eb59588ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ea_af = NULL\n",
    "n = NULL\n",
    "ncase = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "754394d7-243b-4960-962d-c14321fc2072",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, \n",
    "    se = NULL, pval = NULL, n = NULL, ncase = NULL, name = NULL) \n",
    "{\n",
    "    stopifnot(length(chrom) == length(pos))\n",
    "    if (is.null(snp)) {\n",
    "        snp <- paste0(chrom, \":\", pos)\n",
    "    }\n",
    "    snp <- paste0(chrom, \":\", pos)\n",
    "    nsnp <- length(chrom)\n",
    "    gen <- list()\n",
    "    if (!is.null(ea_af)) \n",
    "        gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
    "    if (!is.null(effect)) \n",
    "        gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
    "    if (!is.null(se)) \n",
    "        gen[[\"SE\"]] <- matrix(se, nsnp)\n",
    "    if (!is.null(pval)) \n",
    "        gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
    "    if (!is.null(n)) \n",
    "        gen[[\"SS\"]] <- matrix(n, nsnp)\n",
    "    if (!is.null(ncase)) \n",
    "        gen[[\"NC\"]] <- matrix(ncase, nsnp)\n",
    "    gen <- S4Vectors::SimpleList(gen)\n",
    "    gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
    "        end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
    "     coldata <- S4Vectors::DataFrame(Studies = name, row.names = name)\n",
    "    hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
    "        row.names = \"fileformat\")), sample = name)\n",
    "    VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
    "        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
    "        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", \n",
    "        \"Alternate allele frequency in the association study\", \n",
    "        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), \n",
    "        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))\n",
    "    VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
    "        rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
    "    vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
    "        exptData = list(header = hdr), geno = gen)\n",
    "    VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
    "    VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
    "    VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\"\n",
    "        return(sort(vcf))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "954684f8-9492-4c01-a534-30c3d1d3e11a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'sss'</li><li>'qqq'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'sss'\n",
       "\\item 'qqq'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'sss'\n",
       "2. 'qqq'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"sss\" \"qqq\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "l$Studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a7d5d3fa-73aa-452a-a94c-5d14b7f05c7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 102064 rows and 0 columns"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "info(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73fa9a77-9319-4bf5-90ec-e962c4097348",
   "metadata": {},
   "outputs": [],
   "source": [
    "    coldata <- S4Vectors::DataFrame(Samples = length(name), row.names = name)\n",
    "    hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
    "        row.names = \"fileformat\")), sample = name)\n",
    "    VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
    "        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
    "        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", \n",
    "        \"Alternate allele frequency in the association study\", \n",
    "        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), \n",
    "        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))\n",
    "    VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
    "        rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
    "    vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
    "        exptData = list(header = hdr), geno = gen)\n",
    "    VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
    "    VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
    "    VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "39afa4c8-6f39-4aaa-81f1-2b437d447b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    snp <- paste0(chrom, \":\", pos)\n",
    "    nsnp <- length(chrom)\n",
    "    gen <- list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "33af3ae5-7177-4ca7-8432-3073fefc22cc",
   "metadata": {},
   "outputs": [],
   "source": [
    " if (!is.null(ea_af)) \n",
    "        gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
    "    if (!is.null(effect)) \n",
    "        gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
    "    if (!is.null(se)) \n",
    "        gen[[\"SE\"]] <- matrix(se, nsnp)\n",
    "    if (!is.null(pval)) \n",
    "        gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
    "    if (!is.null(n)) \n",
    "        gen[[\"SS\"]] <- matrix(n, nsnp)\n",
    "    if (!is.null(ncase)) \n",
    "        gen[[\"NC\"]] <- matrix(ncase, nsnp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "11a3085f-f86e-4f76-bffe-2acee0a8ac58",
   "metadata": {},
   "outputs": [],
   "source": [
    "    gen <- S4Vectors::SimpleList(gen)\n",
    "    gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
    "        end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
    "    coldata <- S4Vectors::DataFrame(Samples = length(name), row.names = name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2a43d752-fce7-43a6-b7a5-ff2779f65e35",
   "metadata": {},
   "outputs": [],
   "source": [
    "hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
    "        row.names = \"fileformat\")), sample = name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ed9014fa-30dd-40db-8af7-89e11d83cc8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
    "        rownames(VariantAnnotation::geno(hdr)) %in% names(gen))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "211cc502-9383-470b-a600-36e4c32884c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
    "        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
    "        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\", \n",
    "        \"Alternate allele frequency in the association study\", \n",
    "        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"), \n",
    "        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0978f8d7-d4b4-4aad-957e-17c0ceb268d8",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in .Call2(\"new_XStringSet_from_CHARACTER\", class(x0), elementType(x0), : key 42 (char '*') not in lookup table\n",
     "output_type": "error",
     "traceback": [
      "Error in .Call2(\"new_XStringSet_from_CHARACTER\", class(x0), elementType(x0), : key 42 (char '*') not in lookup table\nTraceback:\n",
      "1. Biostrings::DNAStringSetList(as.list(ea))",
      "2. XStringSetList(\"DNA\", ..., use.names = use.names)",
      "3. .new_XStringSetList_from_list(seqtype, x)",
      "4. XStringSet(seqtype, unlisted_y)",
      "5. XStringSet(seqtype, unlisted_y)",
      "6. .charToXStringSet(seqtype, x, start, end, width, use.names)",
      "7. make_XStringSet_from_strings(x0, strings, start(solved_SEW), \n .     width(solved_SEW))",
      "8. make_XStringSet_from_strings(x0, strings, start(solved_SEW), \n .     width(solved_SEW))",
      "9. .Call2(\"new_XStringSet_from_CHARACTER\", class(x0), elementType(x0), \n .     strings, start, width, lkup, PACKAGE = \"Biostrings\")"
     ]
    }
   ],
   "source": [
    "    VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
    "    VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ec776869-db9e-4a61-9bb9-782963442ba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
    "        exptData = list(header = hdr), geno = gen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7604b3f3-c974-4a87-8f34-bef7c2b90b6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 23250 x 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>#chrom</th><th scope=col>pos</th><th scope=col>ref</th><th scope=col>alt</th><th scope=col>gene</th><th scope=col>beta</th><th scope=col>se</th><th scope=col>pval</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>21</td><td>30873708</td><td>A    </td><td>G  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30874181</td><td>T    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30874182</td><td>C    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30874183</td><td>C    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30874585</td><td>G    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30874921</td><td>C    </td><td>T  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30875013</td><td>C    </td><td>CT </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30875332</td><td>T    </td><td>C  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30875460</td><td>T    </td><td>G  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30875630</td><td>A    </td><td>G  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30876630</td><td>C    </td><td>T  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30876874</td><td>C    </td><td>T  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30877210</td><td>T    </td><td>C  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30877267</td><td>A    </td><td>G  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30877651</td><td>T    </td><td>TTA</td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30877931</td><td>A    </td><td>G  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30878324</td><td>T    </td><td>C  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30878658</td><td>G    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30878965</td><td>T    </td><td>TA </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30879287</td><td>C    </td><td>T  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30879543</td><td>C    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30879890</td><td>CTTTA</td><td>C  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30879943</td><td>A    </td><td>T  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30880138</td><td>G    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30880180</td><td>G    </td><td>C  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30880327</td><td>G    </td><td>A  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30880400</td><td>C    </td><td>T  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30880918</td><td>C    </td><td>T  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30881081</td><td>A    </td><td>G  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>30881086</td><td>A    </td><td>G  </td><td>ENSG00000142149</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td></tr>\n",
       "\t<tr><td>21</td><td>45322820</td><td>G     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45322850</td><td>C     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45323085</td><td>T     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45323179</td><td>G     </td><td>A </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45323358</td><td>G     </td><td>C </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45323513</td><td>A     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45323572</td><td>G     </td><td>A </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45324060</td><td>C     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45324274</td><td>AG    </td><td>A </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45324338</td><td>G     </td><td>A </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45324502</td><td>C     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45324687</td><td>C     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45325032</td><td>A     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45325035</td><td>AATATT</td><td>AT</td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45325123</td><td>T     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45325232</td><td>G     </td><td>A </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45325593</td><td>T     </td><td>C </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45325841</td><td>TA    </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45325987</td><td>G     </td><td>A </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326086</td><td>C     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326104</td><td>C     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326128</td><td>G     </td><td>C </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326146</td><td>C     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326542</td><td>C     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326662</td><td>C     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326711</td><td>C     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326731</td><td>C     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326837</td><td>G     </td><td>A </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45326895</td><td>C     </td><td>T </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "\t<tr><td>21</td><td>45327278</td><td>A     </td><td>G </td><td>ENSG00000141959</td><td>0</td><td>0</td><td>-1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 23250 x 8\n",
       "\\begin{tabular}{llllllll}\n",
       " \\#chrom & pos & ref & alt & gene & beta & se & pval\\\\\n",
       " <dbl> & <dbl> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 21 & 30873708 & A     & G   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30874181 & T     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30874182 & C     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30874183 & C     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30874585 & G     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30874921 & C     & T   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30875013 & C     & CT  & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30875332 & T     & C   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30875460 & T     & G   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30875630 & A     & G   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30876630 & C     & T   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30876874 & C     & T   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30877210 & T     & C   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30877267 & A     & G   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30877651 & T     & TTA & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30877931 & A     & G   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30878324 & T     & C   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30878658 & G     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30878965 & T     & TA  & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30879287 & C     & T   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30879543 & C     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30879890 & CTTTA & C   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30879943 & A     & T   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30880138 & G     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30880180 & G     & C   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30880327 & G     & A   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30880400 & C     & T   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30880918 & C     & T   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30881081 & A     & G   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t 21 & 30881086 & A     & G   & ENSG00000142149 & 0 & 0 & -1\\\\\n",
       "\t ... & ... & ... & ... & ... & ... & ... & ...\\\\\n",
       "\t 21 & 45322820 & G      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45322850 & C      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45323085 & T      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45323179 & G      & A  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45323358 & G      & C  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45323513 & A      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45323572 & G      & A  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45324060 & C      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45324274 & AG     & A  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45324338 & G      & A  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45324502 & C      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45324687 & C      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45325032 & A      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45325035 & AATATT & AT & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45325123 & T      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45325232 & G      & A  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45325593 & T      & C  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45325841 & TA     & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45325987 & G      & A  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326086 & C      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326104 & C      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326128 & G      & C  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326146 & C      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326542 & C      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326662 & C      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326711 & C      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326731 & C      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326837 & G      & A  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45326895 & C      & T  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\t 21 & 45327278 & A      & G  & ENSG00000141959 & 0 & 0 & -1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 23250 x 8\n",
       "\n",
       "| #chrom &lt;dbl&gt; | pos &lt;dbl&gt; | ref &lt;chr&gt; | alt &lt;chr&gt; | gene &lt;chr&gt; | beta &lt;dbl&gt; | se &lt;dbl&gt; | pval &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 21 | 30873708 | A     | G   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30874181 | T     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30874182 | C     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30874183 | C     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30874585 | G     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30874921 | C     | T   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30875013 | C     | CT  | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30875332 | T     | C   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30875460 | T     | G   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30875630 | A     | G   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30876630 | C     | T   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30876874 | C     | T   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30877210 | T     | C   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30877267 | A     | G   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30877651 | T     | TTA | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30877931 | A     | G   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30878324 | T     | C   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30878658 | G     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30878965 | T     | TA  | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30879287 | C     | T   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30879543 | C     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30879890 | CTTTA | C   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30879943 | A     | T   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30880138 | G     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30880180 | G     | C   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30880327 | G     | A   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30880400 | C     | T   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30880918 | C     | T   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30881081 | A     | G   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| 21 | 30881086 | A     | G   | ENSG00000142149 | 0 | 0 | -1 |\n",
       "| ... | ... | ... | ... | ... | ... | ... | ... |\n",
       "| 21 | 45322820 | G      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45322850 | C      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45323085 | T      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45323179 | G      | A  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45323358 | G      | C  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45323513 | A      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45323572 | G      | A  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45324060 | C      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45324274 | AG     | A  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45324338 | G      | A  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45324502 | C      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45324687 | C      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45325032 | A      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45325035 | AATATT | AT | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45325123 | T      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45325232 | G      | A  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45325593 | T      | C  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45325841 | TA     | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45325987 | G      | A  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326086 | C      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326104 | C      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326128 | G      | C  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326146 | C      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326542 | C      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326662 | C      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326711 | C      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326731 | C      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326837 | G      | A  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45326895 | C      | T  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "| 21 | 45327278 | A      | G  | ENSG00000141959 | 0 | 0 | -1 |\n",
       "\n"
      ],
      "text/plain": [
       "      #chrom pos      ref    alt gene            beta se  pval\n",
       "1     21     30873708 A      G   ENSG00000142149 0    0   -1  \n",
       "2     21     30874181 T      A   ENSG00000142149 0    0   -1  \n",
       "3     21     30874182 C      A   ENSG00000142149 0    0   -1  \n",
       "4     21     30874183 C      A   ENSG00000142149 0    0   -1  \n",
       "5     21     30874585 G      A   ENSG00000142149 0    0   -1  \n",
       "6     21     30874921 C      T   ENSG00000142149 0    0   -1  \n",
       "7     21     30875013 C      CT  ENSG00000142149 0    0   -1  \n",
       "8     21     30875332 T      C   ENSG00000142149 0    0   -1  \n",
       "9     21     30875460 T      G   ENSG00000142149 0    0   -1  \n",
       "10    21     30875630 A      G   ENSG00000142149 0    0   -1  \n",
       "11    21     30876630 C      T   ENSG00000142149 0    0   -1  \n",
       "12    21     30876874 C      T   ENSG00000142149 0    0   -1  \n",
       "13    21     30877210 T      C   ENSG00000142149 0    0   -1  \n",
       "14    21     30877267 A      G   ENSG00000142149 0    0   -1  \n",
       "15    21     30877651 T      TTA ENSG00000142149 0    0   -1  \n",
       "16    21     30877931 A      G   ENSG00000142149 0    0   -1  \n",
       "17    21     30878324 T      C   ENSG00000142149 0    0   -1  \n",
       "18    21     30878658 G      A   ENSG00000142149 0    0   -1  \n",
       "19    21     30878965 T      TA  ENSG00000142149 0    0   -1  \n",
       "20    21     30879287 C      T   ENSG00000142149 0    0   -1  \n",
       "21    21     30879543 C      A   ENSG00000142149 0    0   -1  \n",
       "22    21     30879890 CTTTA  C   ENSG00000142149 0    0   -1  \n",
       "23    21     30879943 A      T   ENSG00000142149 0    0   -1  \n",
       "24    21     30880138 G      A   ENSG00000142149 0    0   -1  \n",
       "25    21     30880180 G      C   ENSG00000142149 0    0   -1  \n",
       "26    21     30880327 G      A   ENSG00000142149 0    0   -1  \n",
       "27    21     30880400 C      T   ENSG00000142149 0    0   -1  \n",
       "28    21     30880918 C      T   ENSG00000142149 0    0   -1  \n",
       "29    21     30881081 A      G   ENSG00000142149 0    0   -1  \n",
       "30    21     30881086 A      G   ENSG00000142149 0    0   -1  \n",
       "...   ...    ...      ...    ... ...             ...  ... ... \n",
       "23221 21     45322820 G      T   ENSG00000141959 0    0   -1  \n",
       "23222 21     45322850 C      G   ENSG00000141959 0    0   -1  \n",
       "23223 21     45323085 T      G   ENSG00000141959 0    0   -1  \n",
       "23224 21     45323179 G      A   ENSG00000141959 0    0   -1  \n",
       "23225 21     45323358 G      C   ENSG00000141959 0    0   -1  \n",
       "23226 21     45323513 A      G   ENSG00000141959 0    0   -1  \n",
       "23227 21     45323572 G      A   ENSG00000141959 0    0   -1  \n",
       "23228 21     45324060 C      T   ENSG00000141959 0    0   -1  \n",
       "23229 21     45324274 AG     A   ENSG00000141959 0    0   -1  \n",
       "23230 21     45324338 G      A   ENSG00000141959 0    0   -1  \n",
       "23231 21     45324502 C      T   ENSG00000141959 0    0   -1  \n",
       "23232 21     45324687 C      G   ENSG00000141959 0    0   -1  \n",
       "23233 21     45325032 A      G   ENSG00000141959 0    0   -1  \n",
       "23234 21     45325035 AATATT AT  ENSG00000141959 0    0   -1  \n",
       "23235 21     45325123 T      G   ENSG00000141959 0    0   -1  \n",
       "23236 21     45325232 G      A   ENSG00000141959 0    0   -1  \n",
       "23237 21     45325593 T      C   ENSG00000141959 0    0   -1  \n",
       "23238 21     45325841 TA     T   ENSG00000141959 0    0   -1  \n",
       "23239 21     45325987 G      A   ENSG00000141959 0    0   -1  \n",
       "23240 21     45326086 C      G   ENSG00000141959 0    0   -1  \n",
       "23241 21     45326104 C      G   ENSG00000141959 0    0   -1  \n",
       "23242 21     45326128 G      C   ENSG00000141959 0    0   -1  \n",
       "23243 21     45326146 C      T   ENSG00000141959 0    0   -1  \n",
       "23244 21     45326542 C      G   ENSG00000141959 0    0   -1  \n",
       "23245 21     45326662 C      T   ENSG00000141959 0    0   -1  \n",
       "23246 21     45326711 C      T   ENSG00000141959 0    0   -1  \n",
       "23247 21     45326731 C      T   ENSG00000141959 0    0   -1  \n",
       "23248 21     45326837 G      A   ENSG00000141959 0    0   -1  \n",
       "23249 21     45326895 C      T   ENSG00000141959 0    0   -1  \n",
       "23250 21     45327278 A      G   ENSG00000141959 0    0   -1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a[which(is.na(gen[[\"LP\"]])),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6b8becc-2703-4fe3-9e92-4b2f2390d2a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
