{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b7594e23-1f73-47e1-ac13-8e2fd51af73f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "130a614b-3eba-495e-ba64-c3c825076ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "Recipe_temp = pd.DataFrame( {\"Theme\" : [\"AC\",\"DLPFC\",\"PCC\"] ,\n",
    "                \"genotype_list\" : [\"/home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_list\",\n",
    "                   \"/home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_list\",\n",
    "                   \"/home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_list\"],\n",
    "                \"molecular_pheno\" : [\"/home/hs3163/Project/Rosmap/data/gene_exp/AC/geneTpmResidualsAgeGenderAdj_rename.txt\",\n",
    "                     \"/home/hs3163/Project/Rosmap/data/gene_exp/DLPFC/geneTpmResidualsAgeGenderAdj_rename.txt\",\n",
    "                     \"/home/hs3163/Project/Rosmap/data/gene_exp/PCC/geneTpmResidualsAgeGenderAdj_rename.txt\"],\n",
    "                \"region_list\" : [\"~/GIT/ADSPFG-xQTL/MWE/mwe_region\",\n",
    "                \"~/GIT/ADSPFG-xQTL/MWE/mwe_region\",\n",
    "                \"~/GIT/ADSPFG-xQTL/MWE/mwe_region\"] ,\n",
    "                \"covariate_file\" : [\"~/GIT/ADSPFG-xQTL/MWE/mwe_cov\",\"~/GIT/ADSPFG-xQTL/MWE/mwe_cov\",\"~/GIT/ADSPFG-xQTL/MWE/mwe_cov\"],\n",
    "                \"factor_analysis_opt\" : [\"APEX\",\"APEX\",\"APEX\"],\n",
    "                \"LD_options\": [\"In-sample\",\"In-sample\",\"In-sample\"],\n",
    "                \"QTL_tool_option\" : [\"APEX\",\"TensorQTL\",\"APEX\"],\n",
    "                \"QTL_analysis_option\" : [\"cis\",\"cis\",\"cis\"],\n",
    "                \"cis_windows\" : [500000,500000,5000000],\n",
    "                \"Metal\" : [\"T\",\"T\",\"F\"]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0c9981ea-2961-4e18-a0eb-9ad2204461d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Theme</th>\n",
       "      <th>genotype_list</th>\n",
       "      <th>molecular_pheno</th>\n",
       "      <th>region_list</th>\n",
       "      <th>covariate_file</th>\n",
       "      <th>factor_analysis_opt</th>\n",
       "      <th>LD_options</th>\n",
       "      <th>QTL_tool_option</th>\n",
       "      <th>QTL_analysis_option</th>\n",
       "      <th>cis_windows</th>\n",
       "      <th>Metal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC</td>\n",
       "      <td>/home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_...</td>\n",
       "      <td>/home/hs3163/Project/Rosmap/data/gene_exp/AC/g...</td>\n",
       "      <td>~/GIT/ADSPFG-xQTL/MWE/mwe_region</td>\n",
       "      <td>~/GIT/ADSPFG-xQTL/MWE/mwe_cov</td>\n",
       "      <td>APEX</td>\n",
       "      <td>In-sample</td>\n",
       "      <td>APEX</td>\n",
       "      <td>cis</td>\n",
       "      <td>500000</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DLPFC</td>\n",
       "      <td>/home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_...</td>\n",
       "      <td>/home/hs3163/Project/Rosmap/data/gene_exp/DLPF...</td>\n",
       "      <td>~/GIT/ADSPFG-xQTL/MWE/mwe_region</td>\n",
       "      <td>~/GIT/ADSPFG-xQTL/MWE/mwe_cov</td>\n",
       "      <td>APEX</td>\n",
       "      <td>In-sample</td>\n",
       "      <td>TensorQTL</td>\n",
       "      <td>cis</td>\n",
       "      <td>500000</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PCC</td>\n",
       "      <td>/home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_...</td>\n",
       "      <td>/home/hs3163/Project/Rosmap/data/gene_exp/PCC/...</td>\n",
       "      <td>~/GIT/ADSPFG-xQTL/MWE/mwe_region</td>\n",
       "      <td>~/GIT/ADSPFG-xQTL/MWE/mwe_cov</td>\n",
       "      <td>APEX</td>\n",
       "      <td>In-sample</td>\n",
       "      <td>APEX</td>\n",
       "      <td>cis</td>\n",
       "      <td>5000000</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Theme                                      genotype_list  \\\n",
       "0     AC  /home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_...   \n",
       "1  DLPFC  /home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_...   \n",
       "2    PCC  /home/hs3163/GIT/ADSPFG-xQTL/MWE/mwe_genotype_...   \n",
       "\n",
       "                                     molecular_pheno  \\\n",
       "0  /home/hs3163/Project/Rosmap/data/gene_exp/AC/g...   \n",
       "1  /home/hs3163/Project/Rosmap/data/gene_exp/DLPF...   \n",
       "2  /home/hs3163/Project/Rosmap/data/gene_exp/PCC/...   \n",
       "\n",
       "                        region_list                 covariate_file  \\\n",
       "0  ~/GIT/ADSPFG-xQTL/MWE/mwe_region  ~/GIT/ADSPFG-xQTL/MWE/mwe_cov   \n",
       "1  ~/GIT/ADSPFG-xQTL/MWE/mwe_region  ~/GIT/ADSPFG-xQTL/MWE/mwe_cov   \n",
       "2  ~/GIT/ADSPFG-xQTL/MWE/mwe_region  ~/GIT/ADSPFG-xQTL/MWE/mwe_cov   \n",
       "\n",
       "  factor_analysis_opt LD_options QTL_tool_option QTL_analysis_option  \\\n",
       "0                APEX  In-sample            APEX                 cis   \n",
       "1                APEX  In-sample       TensorQTL                 cis   \n",
       "2                APEX  In-sample            APEX                 cis   \n",
       "\n",
       "   cis_windows Metal  \n",
       "0       500000     T  \n",
       "1       500000     T  \n",
       "2      5000000     F  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Recipe_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b393f2af-18be-494f-880e-b47cbcb8c648",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"dplyr\")\n",
    "library(\"readr\")\n",
    "library(\"purrr\")\n",
    "## Define a wrapper, modified from the gwasvcf packages, to create the vcf of needed.\n",
    "\n",
    "    create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL,\n",
    "    se = NULL, pval = NULL, n = NULL, ncase = NULL, name = NULL)\n",
    "{\n",
    "    stopifnot(length(chrom) == length(pos))\n",
    "    if (is.null(snp)) {\n",
    "        snp <- paste0(chrom, \":\", pos)\n",
    "    }\n",
    "    snp <- paste0(chrom, \":\", pos)\n",
    "    nsnp <- length(chrom)\n",
    "    gen <- list()\n",
    "    ## Setupt data content for each sample column\n",
    "    if (!is.null(ea_af))\n",
    "        gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
    "    if (!is.null(effect))\n",
    "        gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
    "    if (!is.null(se))\n",
    "        gen[[\"SE\"]] <- matrix(se, nsnp)\n",
    "    if (!is.null(pval))\n",
    "        gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
    "    if (!is.null(n))\n",
    "        gen[[\"SS\"]] <- matrix(n, nsnp)\n",
    "    if (!is.null(ncase))\n",
    "        gen[[\"NC\"]] <- matrix(ncase, nsnp)\n",
    "    gen <- S4Vectors::SimpleList(gen)\n",
    "\n",
    "  ## Setup snps info for the fix columns\n",
    "    gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos,\n",
    "        end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
    "  ## Setup meta informations\n",
    "     coldata <- S4Vectors::DataFrame(Studies = name, row.names = name)\n",
    "    hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\",\n",
    "        row.names = \"fileformat\")), sample = name)\n",
    "    VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\",\n",
    "        \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\",\n",
    "        \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\",\n",
    "        \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\",\n",
    "        \"Alternate allele frequency in the association study\",\n",
    "        \"Sample size used to estimate genetic effect\", \"Number of cases used to estimate genetic effect\"),\n",
    "        row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"SS\", \"NC\"))\n",
    "  ## Save only the meta information in the sample columns\n",
    "    VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr),\n",
    "        rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
    "  ## Save VCF values\n",
    "    vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata,\n",
    "        exptData = list(header = hdr), geno = gen)\n",
    "    VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
    "    VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
    "  ## Write fixed values\n",
    "    VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\"\n",
    "        return(sort(vcf))\n",
    "    }\n",
    "\n",
    "    input = readRDS('/mnt/mfs/statgen/xqtl_workflow_testing/testing_6/Meta_analysis/sumstat/mash_rds/ENSG00000141959.posterior.rds')\n",
    "    input_effect = input$PosteriorMean\n",
    "    if(is.null(input$PosteriorSD)){\n",
    "       input$PosteriorSD = matrix(1,nrow = nrow(input_effect),ncol = ncol(input_effect) )\n",
    "      }\n",
    "    input_se = input$PosteriorSD\n",
    "    df = tibble(snps = input$snps)\n",
    "    df = df%>%mutate(    chr = map_dbl(snps,~read.table(text = .x,sep = \":\",as.is = T)$V1),\n",
    "                 pos_alt_ref = map_chr(snps,~read.table(text = .x,sep = \":\",as.is = TRUE)$V2),\n",
    "                 pos = map_dbl(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE)$V1),\n",
    "                 alt = map_chr(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE, colClass = \"character\")$V2),\n",
    "                 ref = map_chr(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE, colClass = \"character\")$V3))\n",
    "\n",
    "\n",
    "    vcf = create_vcf(\n",
    "       chrom = df$chr,\n",
    "        pos = df$pos,\n",
    "        ea = df$alt,\n",
    "        nea = df$ref,\n",
    "        effect = input_effect ,\n",
    "        se =  input_se,\n",
    "        name = colnames(input_effect))\n",
    "      \n",
    "  VariantAnnotation::writeVcf(vcf,'/mnt/mfs/statgen/xqtl_workflow_testing/testing_6/Meta_analysis/mash_vcf/mash_vcf/ENSG00000141959.posterior.vcf',index = TRUE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
