
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>FunGen-xQTL Computational Protocol &#8212; FunGen-xQTL Consortium</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Illustration of xQTL protocol" href="code/xqtl_protocol_demo.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1 current">
  <a class="reference internal" href="#">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="code/xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="code/commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="code/commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="code/data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="code/molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="code/data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/data_preprocessing/phenotype_preprocessing.html">
   Phenotype preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/association_scan/qtl_association_testing.html">
   TensorQTL association analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="code/association_scan/qtl_association_postprocessing.html">
   QTL association results postprocessing and standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/fine_mapping/univariate_fine_mapping.html">
   Fine-mapping xQTL using SuSiE
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with individual level data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="code/fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/multivariate/MASH/mashr.html">
     Data overview
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/integrative_analysis/COLOC/fastenloc_dap.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="code/integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="code/integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="code/integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="code/misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="code/misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/README.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2FREADME.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/README.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-of-the-protocol">
   Overview of the protocol
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#standardized-reference-data">
     Standardized reference data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#software-environment">
     Software environment
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#pipeline-execution">
     Pipeline execution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#source-code">
     Source code
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#how-to-use-the-resource">
   How to use the resource
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#organization-of-the-resource">
     Organization of the resource
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#computing-environment-setup">
     Computing environment setup
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#see-also">
     See Also
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#our-team">
   Our team
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#developers">
     Developers
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#leadership">
     Leadership
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="fungen-xqtl-computational-protocol">
<h1>FunGen-xQTL Computational Protocol<a class="headerlink" href="#fungen-xqtl-computational-protocol" title="Permalink to this headline">¶</a></h1>
<p>Developed for reproducible &amp; reusable molecular QTL analyses for the NIH/NIA Alzheimer’s Disease Sequencing Project (ADSP) Functional Genomics xQTL (FunGen-xQTL) Project.</p>
<p><img alt="QTL Diagram" src="_images/xQTL_Protocol_June_2023.png" /></p>
<div class="section" id="overview-of-the-protocol">
<h2>Overview of the protocol<a class="headerlink" href="#overview-of-the-protocol" title="Permalink to this headline">¶</a></h2>
<div class="section" id="standardized-reference-data">
<h3>Standardized reference data<a class="headerlink" href="#standardized-reference-data" title="Permalink to this headline">¶</a></h3>
<p>Reference data are standardized and curated by the ADSP FGC Standardization Workgroup in coordination with <a class="reference external" href="https://www.nia.nih.gov/research/ad-genetics">NIAGCADS</a>. Please find reference data specifications on <a class="reference external" href="https://www.niagads.org/adsp/content/adspgcadgenomeresources-v2pdf">ADSP Dashboard</a>.</p>
</div>
<div class="section" id="software-environment">
<h3>Software environment<a class="headerlink" href="#software-environment" title="Permalink to this headline">¶</a></h3>
<p>We have prepared containerized software environment through both Docker and Singularity virtualization systems to facilicate software environment setup and to aid in software reproducibility. For those not familiar with this concept please <a class="reference external" href="https://en.wikipedia.org/wiki/OS-level_virtualization">check out this wiki page of virtualization</a> and <a class="reference external" href="https://www.docker.com/resources/what-container/">an explanation on Docker website</a>.</p>
</div>
<div class="section" id="pipeline-execution">
<h3>Pipeline execution<a class="headerlink" href="#pipeline-execution" title="Permalink to this headline">¶</a></h3>
<p>Pipelines in this repository are written in the <a class="reference external" href="https://vatlab.github.io/sos-docs/">Script of Scripts (SoS) workflow language</a>. Like most other workflow languages, SoS workflows can <strong>distribute and execute computing jobs directly in High Performance Computing cluster</strong>. It can also use <strong>containers (Docker or Singularity)</strong> to help with setting up computational environment and improve reproducibility. Unlike most other workflow languages, SoS workflows are created using SoS Notebooks (based on Ipython Notebook and developed in <a class="reference external" href="https://jupyter.org/">Jupyter</a>) which allow for both <strong>scientific narrative and pipeline scripts in the same document</strong>. Unlike typical Jupyter Notebooks intended for interactive data analysis, SoS workflows written in Jupyter Notebooks can be executed directly as command line scripts either on a local computer or in a HPC environment.</p>
<p>We provide this <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/blob/main/code/misc/Job_Example.ipynb">toy example for running SoS pipeline on a typical HPC cluster environment</a>. First time users are encouraged to try it out in order to help setting up the computational environment necessary to run the analysis in this protocol.</p>
</div>
<div class="section" id="source-code">
<h3>Source code<a class="headerlink" href="#source-code" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Source code of pipelines and containers implemented in this repository are available at <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/tree/main/code">https://github.com/cumc/xqtl-pipeline/tree/main/code</a>.</p></li>
<li><p>Container specifications for required software environments are available at <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/tree/main/container">https://github.com/cumc/xqtl-pipeline/tree/main/container</a>.</p></li>
</ul>
</div>
</div>
<div class="section" id="how-to-use-the-resource">
<h2>How to use the resource<a class="headerlink" href="#how-to-use-the-resource" title="Permalink to this headline">¶</a></h2>
<div class="section" id="organization-of-the-resource">
<h3>Organization of the resource<a class="headerlink" href="#organization-of-the-resource" title="Permalink to this headline">¶</a></h3>
<p>The website <a class="reference external" href="https://cumc.github.io/xqtl-pipeline">https://cumc.github.io/xqtl-pipeline</a> is generated from files under the <code class="docutils literal notranslate"><span class="pre">code</span></code> folder of the source code repository. The <code class="docutils literal notranslate"><span class="pre">pipeline</span></code> folder contains symbolic links automatically generated for pipeline files under <code class="docutils literal notranslate"><span class="pre">code.</span></code> The logic of the entire xQTL analysis workflow is roughly reflected on the <strong>left sidebar</strong>:</p>
<ul class="simple">
<li><p>The <strong>GETTING STARTED</strong>  section serves as the main landing page or index of the xQTL protocol, guiding users through the various pipelines implemented in this repository. It’s structured to mirror the logic of the xQTL analysis we’ve crafted. Because this page provides pointers to other sections, users can primarily focus here without having to sift through the rest of the pages pages in this repository.</p></li>
<li><p>The <strong>COMMAND GENERATOR</strong> section is designed as a one-stop hub for “push button” commands, enabling users to generate the full QTL analysis pipeline workflow scripts from a straightforward configuration file. Notebooks within this section are intended to be <strong>executed as command line software</strong> for data analysis command generation. Users can then execute these generated commands directly to conduct all preset analyses. Alternatively, they can tweak the commands to cater to particular analysis requirements. The configuration file serves a dual purpose: streamlining control and maintaining a record of executed workflows.</p></li>
<li><p>Other sections in bold fonts provide an array of available analyses, presented roughly from upstream to downstream processes. Most of these sections feature <em><strong>mini-protocols</strong></em>, represented as clickable, non-bold text under each analysis category, leading to specific notebooks. These notebooks detail the commands necessary for the analyses defined in the respective mini-protocols. Predominantly tutorial-based, they are designed to be <strong>executed interactively in Jupyter or via the command terminal</strong>, allowing users to navigate through the SoS pipelines step by step. A few of these sections serve as actual <em><strong>pipeline modules</strong></em> which we’ll discuss next (see below).</p></li>
<li><p><em>Mini-protocols</em>, as mentioned earlier, can be expanded by clicking the downward arrows, revealing the SoS implementations of <em><strong>pipeline modules</strong></em>. These represent the crux of the pipeline implementations and are intended to be <strong>executed as command line software</strong>. They’re also <strong>self-contained</strong>, allowing for reusability beyond the specific context of xQTL data analysis.</p></li>
</ul>
</div>
<div class="section" id="computing-environment-setup">
<h3>Computing environment setup<a class="headerlink" href="#computing-environment-setup" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>In order to run the xQTL protocol on your computer (or a High Performance Computing cluster), please install Script of Scripts (<strong>version &gt;= 0.24.1</strong>) <a class="reference external" href="https://wanggroup.org/orientation/jupyter-setup#setup-the-script-of-scripts-computing-environment">(see here for a tutorial to set it up with <code class="docutils literal notranslate"><span class="pre">micromamba</span></code>)</a>.</p>
<ul>
<li><p>For Linux and Mac desktop users you can either install the container <a class="reference external" href="https://docs.sylabs.io/guides/3.2/user-guide/installation.html"><code class="docutils literal notranslate"><span class="pre">Singularity</span></code></a> or <a class="reference external" href="https://www.docker.com/"><code class="docutils literal notranslate"><span class="pre">Docker</span></code></a>. In the xQTL project we primarily use <code class="docutils literal notranslate"><span class="pre">Singularity</span></code>.</p></li>
<li><p>For Linux-based HPC users, your system may already have <code class="docutils literal notranslate"><span class="pre">Singularity</span></code> installed. If not please communicate with the IT support for the HPC. Typically Docker is not allowed on HPC.</p></li>
<li><p>For Windows users, you will need to install <a class="reference external" href="https://learn.microsoft.com/en-us/windows/wsl/install">WSL</a> (we have tested it on WSL2 and not on WSL1) and then install <code class="docutils literal notranslate"><span class="pre">Singularity</span></code> within WSL as instructed in this <a class="reference external" href="https://www.blopig.com/blog/2021/09/using-singularity-on-windows-with-wsl2/">post</a>.</p></li>
</ul>
</li>
<li><p>We have provided example data-sets and <code class="docutils literal notranslate"><span class="pre">Singularity</span></code> container images in <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn36416559/files/">this Synapse folder</a>. For guidance on downloading the data programmatically, refer to <a class="reference external" href="https://help.synapse.org/docs/Upload-and-Download-Data-in-Bulk.2003796248.html">this documentation</a>. If you need to set up a Synapse client, consult <a class="reference external" href="https://help.synapse.org/docs/Installing-Synapse-API-Clients.1985249668.html">this guide</a>.</p>
<ul>
<li><p>Within the <code class="docutils literal notranslate"><span class="pre">test_data</span></code> folder, datasets prefixed with <strong>MWE</strong> (Minimal Working Example) are provided. These are used for unit testing each module, ensuring the integrity of the code.</p></li>
</ul>
</li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">protocol_data</span></code> folder houses a comprehensive set of data, illustrating the full extent of our protocol. This is showcased in <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/xqtl_protocol_demo.html">this notebook</a>, with the <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/blob/main/code/xqtl_protocol_demo.ipynb">source code</a> available for reference.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">container/singularity</span></code> folder contains the released Singularity images for the software environment. For Docker users (e.g., on Linux or Mac Desktops), downloading this folder is <strong>not</strong> necessary.</p></li>
<li><p>Please clone this repository <a class="reference external" href="https://github.com/cumc/xqtl-pipeline">https://github.com/cumc/xqtl-pipeline</a> onto your computer. This is the source code for this resource. All pipelines are symbolic links in the <code class="docutils literal notranslate"><span class="pre">pipeline</span></code> folder. Users are encouraged to execute from the root of the repository folders by typing</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/&lt;</span><span class="n">pipeline_file</span><span class="o">&gt;.</span><span class="n">ipynb</span>
</pre></div>
</div>
<p>that is, executing the symbolic links directly to perform the analysis.</p>
</div>
<div class="section" id="see-also">
<h3>See Also<a class="headerlink" href="#see-also" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Analysis from FunGen-xQTL consortium using this protocol can be found at <a class="reference external" href="https://github.com/cumc/fungen-xqtl-analysis">https://github.com/cumc/fungen-xqtl-analysis</a></p></li>
</ul>
</div>
</div>
<div class="section" id="our-team">
<h2>Our team<a class="headerlink" href="#our-team" title="Permalink to this headline">¶</a></h2>
<p>This repository is developed by the Analysis Working Group of the NIA FunGen-xQTL consortium.</p>
<div class="section" id="developers">
<h3>Developers<a class="headerlink" href="#developers" title="Permalink to this headline">¶</a></h3>
<p><strong>Lead developers</strong></p>
<ul class="simple">
<li><p>Hao Sun, Department of Neurology, Columbia University</p></li>
<li><p>Gao Wang, Department of Neurology, Columbia University</p></li>
</ul>
<p><strong>Main contributors</strong> (largely based on GitHub Pull Requests)</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Affiliation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Xuanhe Chen</p></td>
<td><p>Department of Biostatistics, Columbia University</p></td>
</tr>
<tr class="row-odd"><td><p>Wenhao Gou</p></td>
<td><p>Department of Biostatistics, Columbia University</p></td>
</tr>
<tr class="row-even"><td><p>Liucheng Shi</p></td>
<td><p>Department of Biostatistics, Columbia University</p></td>
</tr>
<tr class="row-odd"><td><p>Haochen Sun</p></td>
<td><p>Department of Biostatistics, Columbia University</p></td>
</tr>
<tr class="row-even"><td><p>Zining Qi</p></td>
<td><p>Department of Biostatistics, Columbia University</p></td>
</tr>
<tr class="row-odd"><td><p>Ru Feng</p></td>
<td><p>Department of Neurology, Columbia University</p></td>
</tr>
<tr class="row-even"><td><p>Alexandre Pelletier</p></td>
<td><p>Department of Medicine, Boston University</p></td>
</tr>
<tr class="row-odd"><td><p>Travyse Edwards</p></td>
<td><p>Mount Sinai &amp; University of Pennsylvania</p></td>
</tr>
<tr class="row-even"><td><p>Daniel Nachun</p></td>
<td><p>Department of Pathology, Stanford University</p></td>
</tr>
<tr class="row-odd"><td><p>Jiacheng Li</p></td>
<td><p>Department of Neurology, Columbia University</p></td>
</tr>
<tr class="row-even"><td><p>Mintao Lin</p></td>
<td><p>Department of Medicine, Boston University</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="leadership">
<h3>Leadership<a class="headerlink" href="#leadership" title="Permalink to this headline">¶</a></h3>
<p><strong>FunGen-AD</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Affiliation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Philip De Jager</p></td>
<td><p>Department of Neurology, Columbia University</p></td>
</tr>
<tr class="row-odd"><td><p>Carlos Crunchaga</p></td>
<td><p>Department of Psychiatry, Neurology and Genetics, Washington University in St. Louis</p></td>
</tr>
</tbody>
</table>
<p><strong>FunGen-xQTL Analysis Working Group</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Affiliation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Gao Wang</p></td>
<td><p>Department of Neurology, Columbia University</p></td>
</tr>
<tr class="row-odd"><td><p>Xiaoling Zhang</p></td>
<td><p>Departments of Medicine and Biostatistics, Boston University</p></td>
</tr>
<tr class="row-even"><td><p>Edoardo Marcora</p></td>
<td><p>Departments of Neuroscience, Genetics and Genomic Sciences, Icahn School of Medicine at Mount Sinai</p></td>
</tr>
<tr class="row-odd"><td><p>Fanny Leung</p></td>
<td><p>Department of the Pathology and Laboratory Medicine, University of Pennsylvania</p></td>
</tr>
<tr class="row-even"><td><p>Julia TCW</p></td>
<td><p>Department of Medicine, Boston University</p></td>
</tr>
<tr class="row-odd"><td><p>Kushal K. Dey</p></td>
<td><p>Memorial Sloan Kettering</p></td>
</tr>
<tr class="row-even"><td><p>Alan Renton</p></td>
<td><p>Departments of Neuroscience, Genetics and Genomic Sciences, Icahn School of Medicine at Mount Sinai</p></td>
</tr>
<tr class="row-odd"><td><p>Stephen Montgomery</p></td>
<td><p>Department of Pathology, Stanford University</p></td>
</tr>
<tr class="row-even"><td><p>Xiaoquan Wen</p></td>
<td><p>Department of Biostatistics, University of Michigan</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='right-next' id="next-link" href="code/xqtl_protocol_demo.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Illustration of xQTL protocol</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>