
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Meta-analysis with METAL &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Multivariate association with MASH" href="../MASH_pipeline.html" />
    <link rel="prev" title="Meta-analysis with METAL" href="../METAL_pipeline.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/qtl_association_testing.html">
   QTL association analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with individual level data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../MASH/mashr.html">
     Data overview
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/multivariate/METAL/METAL.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate/METAL/METAL.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/multivariate/METAL/METAL.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-1-script-generation">
   Step.1 Script generation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-2-metal-analysis">
   Step 2. Metal analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#step-3-reformat-the-output-so-that-they-are-in-the-same-formmat-as-the-non-metaled-one">
   Step 3. Reformat the output so that they are in the same formmat as the non metaled one
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="meta-analysis-with-metal">
<h1>Meta-analysis with METAL<a class="headerlink" href="#meta-analysis-with-metal" title="Permalink to this headline">¶</a></h1>
<p>METAT is a command line tool, takes in a script that documenting the input sumstat files, the field of each of those sumstat input file, and then the end of scipt.</p>
<p>METAL analysis is enssentially a weighted sum of Z score, therefore the input of snps from each chrm can be processed seperately. To make sure the same region was analysis, the input of this workflow shall be a list of path to the sumstat, join by ” ” to be analyzed at once.</p>
<p>Also, The scheme of weighting at the moment is set to be standard error, changing it to sample size would require the upstream input have a columns named sample size, which they dont have it yet.</p>
<p>At the moment, the AVERAGEFREQ, MINMAXFREQ, and GENOMICCONTROL are by default set to off, turning on each of the option require additional input from upstream file.</p>
<p>The output file name of the seconde step will by default contains a 1 linking the desinateted prefix and surfix, which lead to the f’{wd}/{name}1.METAL.txt’ file name</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sumstat_list_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">variant_id</span> <span class="o">=</span> <span class="s1">&#39;variant_id&#39;</span>
<span class="kn">parameter:</span> <span class="n">Sample_size</span> <span class="o">=</span> <span class="s1">&#39;N&#39;</span>
<span class="kn">parameter:</span> <span class="n">ALLELE</span> <span class="o">=</span> <span class="s1">&#39;alt ref&#39;</span>
<span class="kn">parameter:</span> <span class="n">FREQ</span>   <span class="o">=</span> <span class="s1">&#39;EFFECT_ALLELE_FREQ&#39;</span>
<span class="kn">parameter:</span> <span class="n">Beta</span> <span class="o">=</span> <span class="s1">&#39;beta&#39;</span>
<span class="kn">parameter:</span> <span class="n">SE</span> <span class="o">=</span> <span class="s1">&#39;se&#39;</span>
<span class="kn">parameter:</span> <span class="n">PVAL</span>   <span class="o">=</span> <span class="s1">&#39;pval&#39;</span>
<span class="kn">parameter:</span> <span class="n">wd</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">SCHEME</span> <span class="o">=</span> <span class="s1">&#39;STDERR&#39;</span>
<span class="kn">parameter:</span> <span class="n">AVERAGEFREQ</span> <span class="o">=</span> <span class="s2">&quot;OFF&quot;</span>
<span class="kn">parameter:</span> <span class="n">MINMAXFREQ</span>  <span class="o">=</span> <span class="s2">&quot;OFF&quot;</span>
<span class="kn">parameter:</span> <span class="n">GENOMICCONTROL</span>  <span class="o">=</span> <span class="s2">&quot;OFF&quot;</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">sumstat_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sumstat_list_path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">sumstat_inv</span> <span class="o">=</span> <span class="n">sumstat_list</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sumstat_list</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">sumstat_list</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="step-1-script-generation">
<h2>Step.1 Script generation<a class="headerlink" href="#step-1-script-generation" title="Permalink to this headline">¶</a></h2>
<p>The METAL script contains three sections, the first is the analysis parameters, the second is the inputs files, and the third is the ending of script.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">METAL_1</span><span class="p">,</span><span class="n">generate_METAL_script</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;sumstat_inv&quot;</span>
<span class="kn">output:</span> <span class="n">METAL_script</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_sumstat_inv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.METAL_script.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;2h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;55G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
        <span class="n">echo</span> <span class="s1">&#39;</span>
        <span class="c1"># Meta-analysis weighted by:</span>
        <span class="n">SCHEME</span>   <span class="err">$</span><span class="p">[</span><span class="n">SCHEME</span><span class="p">]</span>
        
        <span class="c1"># Whether do genomics control</span>
        <span class="n">GENOMICCONTROL</span> <span class="err">$</span><span class="p">[</span><span class="n">GENOMICCONTROL</span><span class="p">]</span>
        
        <span class="c1"># Whether do AVERAGEFREQ or MINMAXFREQ control</span>
        <span class="n">AVERAGEFREQ</span> <span class="err">$</span><span class="p">[</span><span class="n">AVERAGEFREQ</span><span class="p">]</span>
        <span class="n">MINMAXFREQ</span> <span class="err">$</span><span class="p">[</span><span class="n">MINMAXFREQ</span><span class="p">]</span> 
        <span class="n">OUTFILE</span> <span class="err">$</span><span class="p">[</span><span class="n">wd</span><span class="p">:</span><span class="n">a</span><span class="p">]</span><span class="o">/</span><span class="err">$</span><span class="p">[</span><span class="kp">name</span><span class="p">]</span><span class="o">.</span><span class="err">$</span><span class="p">[</span><span class="n">_sumstat_inv</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span> <span class="o">.</span><span class="n">METAL</span><span class="o">.</span><span class="n">txt</span><span class="s1">&#39; &gt; $[_output]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">_sumstat_inv</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">_sumstat_inv</span><span class="p">)])]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;     </span>
        <span class="n">MARKER</span>   <span class="err">$</span><span class="p">[</span><span class="n">variant_id</span><span class="p">]</span>
        <span class="n">WEIGHT</span>   <span class="err">$</span><span class="p">[</span><span class="n">Sample_size</span><span class="p">]</span>
        <span class="n">ALLELE</span>   <span class="err">$</span><span class="p">[</span><span class="n">ALLELE</span><span class="p">]</span>
        <span class="c1">#FREQ     $[FREQ]</span>
        <span class="n">EFFECT</span>   <span class="err">$</span><span class="p">[</span><span class="n">Beta</span><span class="p">]</span>
        <span class="n">STDERR</span>   <span class="err">$</span><span class="p">[</span><span class="n">SE</span><span class="p">]</span>
        <span class="n">PVAL</span>     <span class="err">$</span><span class="p">[</span><span class="n">PVAL</span><span class="p">]</span>
        <span class="n">PROCESS</span>  <span class="err">$</span><span class="n">i</span> <span class="s2">&quot; &gt;&gt; $[_output];</span>
        <span class="n">done</span> 
        
        <span class="n">echo</span> <span class="s1">&#39;</span>
        <span class="n">ANALYZE</span><span class="s1">&#39; &gt;&gt; $[_output]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="step-2-metal-analysis">
<h2>Step 2. Metal analysis<a class="headerlink" href="#step-2-metal-analysis" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">METAL_2</span><span class="p">,</span><span class="n">Excute_METAL_script</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;sumstat_inv&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_sumstat_inv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.1.METAL.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;55G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
       <span class="n">metal</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="step-3-reformat-the-output-so-that-they-are-in-the-same-formmat-as-the-non-metaled-one">
<h2>Step 3. Reformat the output so that they are in the same formmat as the non metaled one<a class="headerlink" href="#step-3-reformat-the-output-so-that-they-are-in-the-same-formmat-as-the-non-metaled-one" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">METAL_3</span><span class="p">,</span><span class="n">Output_reformatting</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;sumstat_inv&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_sumstat_inv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.METAL.vcf.bgz&#39;</span><span class="p">,</span>
        <span class="n">sumstat</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_sumstat_inv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.METAL.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;2h&#39;</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;55G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span> 
        <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">],</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;variant_id&quot;</span><span class="p">,</span><span class="s2">&quot;alt&quot;</span><span class="p">,</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span><span class="s2">&quot;se&quot;</span><span class="p">,</span><span class="s2">&quot;pval&quot;</span><span class="p">,</span><span class="s2">&quot;Direction&quot;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span> 
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="c1">## Define a wrapper, modified from the gwasvcf packages, to create the vcf of needed.</span>
        <span class="n">create_vcf</span> <span class="o">=</span> <span class="n">function</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">nea</span><span class="p">,</span> <span class="n">ea</span><span class="p">,</span> <span class="n">snp</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">ea_af</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">effect</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> 
        <span class="n">se</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">pval</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">n</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">ncase</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="kp">name</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">)</span> 
    <span class="p">{</span>
        <span class="n">stopifnot</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="o">==</span> <span class="n">length</span><span class="p">(</span><span class="n">pos</span><span class="p">))</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">snp</span><span class="p">))</span> <span class="p">{</span>
            <span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="p">)</span>
        <span class="n">nsnp</span> <span class="o">&lt;-</span> <span class="n">length</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
        <span class="n">gen</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
        <span class="c1">## Setupt data content for each sample column</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">ea_af</span><span class="p">))</span> 
            <span class="n">gen</span><span class="p">[[</span><span class="s2">&quot;AF&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">ea_af</span><span class="p">,</span> <span class="n">nsnp</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">effect</span><span class="p">))</span> 
            <span class="n">gen</span><span class="p">[[</span><span class="s2">&quot;ES&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">effect</span><span class="p">,</span> <span class="n">nsnp</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">se</span><span class="p">))</span> 
            <span class="n">gen</span><span class="p">[[</span><span class="s2">&quot;SE&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">se</span><span class="p">,</span> <span class="n">nsnp</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">pval</span><span class="p">))</span> 
            <span class="n">gen</span><span class="p">[[</span><span class="s2">&quot;LP&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="o">-</span><span class="n">log10</span><span class="p">(</span><span class="n">pval</span><span class="p">),</span> <span class="n">nsnp</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">n</span><span class="p">))</span> 
            <span class="n">gen</span><span class="p">[[</span><span class="s2">&quot;SS&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">nsnp</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">ncase</span><span class="p">))</span> 
            <span class="n">gen</span><span class="p">[[</span><span class="s2">&quot;NC&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">ncase</span><span class="p">,</span> <span class="n">nsnp</span><span class="p">)</span>
        <span class="n">gen</span> <span class="o">&lt;-</span> <span class="n">S4Vectors</span><span class="p">::</span><span class="n">SimpleList</span><span class="p">(</span><span class="n">gen</span><span class="p">)</span>
        
      <span class="c1">## Setup snps info for the fix columns</span>
        <span class="n">gr</span> <span class="o">&lt;-</span> <span class="n">GenomicRanges</span><span class="p">::</span><span class="n">GRanges</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">IRanges</span><span class="p">::</span><span class="n">IRanges</span><span class="p">(</span><span class="n">start</span> <span class="o">=</span> <span class="n">pos</span><span class="p">,</span> 
            <span class="n">end</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">pmax</span><span class="p">(</span><span class="n">nchar</span><span class="p">(</span><span class="n">nea</span><span class="p">),</span> <span class="n">nchar</span><span class="p">(</span><span class="n">ea</span><span class="p">))</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="n">snp</span><span class="p">))</span>
      <span class="c1">## Setup meta informations</span>
         <span class="n">coldata</span> <span class="o">&lt;-</span> <span class="n">S4Vectors</span><span class="p">::</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">Studies</span> <span class="o">=</span> <span class="kp">name</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="kp">name</span><span class="p">)</span>
        <span class="n">hdr</span> <span class="o">&lt;-</span> <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">VCFHeader</span><span class="p">(</span><span class="n">header</span> <span class="o">=</span> <span class="n">IRanges</span><span class="p">::</span><span class="n">DataFrameList</span><span class="p">(</span><span class="n">fileformat</span> <span class="o">=</span> <span class="n">S4Vectors</span><span class="p">::</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">Value</span> <span class="o">=</span> <span class="s2">&quot;VCFv4.2&quot;</span><span class="p">,</span> 
            <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="s2">&quot;fileformat&quot;</span><span class="p">)),</span> <span class="n">sample</span> <span class="o">=</span> <span class="kp">name</span><span class="p">)</span>
        <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">hdr</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">S4Vectors</span><span class="p">::</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">Number</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> 
            <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">),</span> <span class="n">Type</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;Float&quot;</span><span class="p">,</span> <span class="s2">&quot;Float&quot;</span><span class="p">,</span> 
            <span class="s2">&quot;Float&quot;</span><span class="p">,</span> <span class="s2">&quot;Float&quot;</span><span class="p">,</span> <span class="s2">&quot;Float&quot;</span><span class="p">,</span> <span class="s2">&quot;Float&quot;</span><span class="p">),</span> <span class="n">Description</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;Effect size estimate relative to the alternative allele&quot;</span><span class="p">,</span> 
            <span class="s2">&quot;Standard error of effect size estimate&quot;</span><span class="p">,</span> <span class="s2">&quot;-log10 p-value for effect estimate&quot;</span><span class="p">,</span> 
            <span class="s2">&quot;Alternate allele frequency in the association study&quot;</span><span class="p">,</span> 
            <span class="s2">&quot;Sample size used to estimate genetic effect&quot;</span><span class="p">,</span> <span class="s2">&quot;Number of cases used to estimate genetic effect&quot;</span><span class="p">),</span> 
            <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;ES&quot;</span><span class="p">,</span> <span class="s2">&quot;SE&quot;</span><span class="p">,</span> <span class="s2">&quot;LP&quot;</span><span class="p">,</span> <span class="s2">&quot;AF&quot;</span><span class="p">,</span> <span class="s2">&quot;SS&quot;</span><span class="p">,</span> <span class="s2">&quot;NC&quot;</span><span class="p">))</span>
      <span class="c1">## Save only the meta information in the sample columns </span>
        <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">hdr</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">subset</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">hdr</span><span class="p">),</span> 
            <span class="n">rownames</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">hdr</span><span class="p">))</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">gen</span><span class="p">))</span>
      <span class="c1">## Save VCF values</span>
        <span class="n">vcf</span> <span class="o">&lt;-</span> <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">VCF</span><span class="p">(</span><span class="n">rowRanges</span> <span class="o">=</span> <span class="n">gr</span><span class="p">,</span> <span class="n">colData</span> <span class="o">=</span> <span class="n">coldata</span><span class="p">,</span> 
            <span class="n">exptData</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">header</span> <span class="o">=</span> <span class="n">hdr</span><span class="p">),</span> <span class="n">geno</span> <span class="o">=</span> <span class="n">gen</span><span class="p">)</span>
        <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">alt</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">Biostrings</span><span class="p">::</span><span class="n">DNAStringSetList</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">list</span><span class="p">(</span><span class="n">ea</span><span class="p">))</span>
        <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">ref</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">Biostrings</span><span class="p">::</span><span class="n">DNAStringSet</span><span class="p">(</span><span class="n">nea</span><span class="p">)</span>
      <span class="c1">## Write fixed values</span>
        <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">fixed</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span><span class="err">$</span><span class="n">FILTER</span> <span class="o">&lt;-</span> <span class="s2">&quot;PASS&quot;</span>
            <span class="k">return</span><span class="p">(</span><span class="n">sort</span><span class="p">(</span><span class="n">vcf</span><span class="p">))</span>
        <span class="p">}</span>
  
        <span class="nb">input</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">],</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">input_effect</span> <span class="o">=</span> <span class="nb">input</span><span class="err">$</span><span class="n">beta</span>
        <span class="n">input_se</span> <span class="o">=</span> <span class="nb">input</span><span class="err">$</span><span class="n">se</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">tibble</span><span class="p">(</span><span class="n">snps</span> <span class="o">=</span> <span class="nb">input</span><span class="err">$</span><span class="n">variant_id</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span>    <span class="nb">chr</span> <span class="o">=</span> <span class="n">map_dbl</span><span class="p">(</span><span class="n">snps</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span><span class="err">$</span><span class="n">V1</span><span class="p">),</span>
                     <span class="n">pos_alt_ref</span> <span class="o">=</span> <span class="n">map_chr</span><span class="p">(</span><span class="n">snps</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span><span class="err">$</span><span class="n">V2</span><span class="p">),</span>
                     <span class="n">pos</span> <span class="o">=</span> <span class="n">map_dbl</span><span class="p">(</span><span class="n">pos_alt_ref</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span><span class="err">$</span><span class="n">V1</span><span class="p">),</span>
                     <span class="n">alt</span> <span class="o">=</span> <span class="n">map_chr</span><span class="p">(</span><span class="n">pos_alt_ref</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">colClass</span> <span class="o">=</span> <span class="s2">&quot;character&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">V3</span><span class="p">),</span>
                     <span class="n">ref</span> <span class="o">=</span> <span class="n">map_chr</span><span class="p">(</span><span class="n">pos_alt_ref</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">colClass</span> <span class="o">=</span> <span class="s2">&quot;character&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">V2</span><span class="p">))</span>
        <span class="n">vcf</span> <span class="o">=</span> <span class="n">create_vcf</span><span class="p">(</span>
           <span class="n">chrom</span> <span class="o">=</span> <span class="n">df</span><span class="err">$</span><span class="nb">chr</span><span class="p">,</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">df</span><span class="err">$</span><span class="n">pos</span><span class="p">,</span>
            <span class="n">ea</span> <span class="o">=</span> <span class="n">df</span><span class="err">$</span><span class="n">alt</span><span class="p">,</span>
            <span class="n">nea</span> <span class="o">=</span> <span class="n">df</span><span class="err">$</span><span class="n">ref</span><span class="p">,</span>
            <span class="n">effect</span> <span class="o">=</span> <span class="n">input_effect</span> <span class="p">,</span>
            <span class="n">se</span> <span class="o">=</span>  <span class="n">input_se</span><span class="p">,</span>
            <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;$[name]&quot;</span><span class="p">)</span>
      <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">writeVcf</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nr</span><span class="p">],</span><span class="n">index</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">METAL_4</span><span class="p">,</span><span class="n">recipe</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;METAL_3&quot;</span><span class="p">)[</span><span class="s2">&quot;sumstat&quot;</span><span class="p">],</span><span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.METAL_list.txt&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span> 
        <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;#chr&quot;</span> <span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">sumstat_list_path</span><span class="p">:</span><span class="n">r</span><span class="p">],</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="s2">&quot;#chr&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="p">,</span>  <span class="s2">&quot;$[name]&quot;</span>  <span class="p">:</span> <span class="p">[</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,]]</span> <span class="p">})</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output]&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/multivariate/METAL"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../METAL_pipeline.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Meta-analysis with METAL</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../MASH_pipeline.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Multivariate association with MASH</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>