
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Generation of Topologically Associated Domains and their Boundaries &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/reference_data/generalized_TADB';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Independent list of variants using LD clumping" href="ld_prune_reference.html" />
    <link rel="prev" title="Reference Data Standardization" href="reference_data_preparation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="reference_data.html">Reference Data</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL Modifier Score</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_modifier_score/ems_training.html">xQTL Model Training Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../xqtl_modifier_score/ems_prediction.html">EMS Prediction for xQTL Modifier Scores</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/reference_data/generalized_TADB.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/reference_data/generalized_TADB.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/reference_data/generalized_TADB.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Generation of Topologically Associated Domains and their Boundaries</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-manage-tad-redundancy">i. Manage TAD Redundancy</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-tads-with-neighboring-tabs">Merge TADs With Neighboring TABs</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#preliminary-merges">Preliminary Merges</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#determining-overlap">Determining Overlap</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#overlap-check">Overlap Check</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#recursively-merge-tads">Recursively Merge TADs</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-generate-tadb-enhanced-cis-windows-and-extended-tadbs">ii. Generate TADB-enhanced cis windows and extended TADBs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-find-boundaries-and-build-tadb">How to find boundaries and build TADB?</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tadb-coding">TADB - coding</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#find-left-boundaries">Find left boundaries</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#find-right-boundaries">Find right boundaries</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#map-each-tad-to-the-closest-left-right-boundary">Map each TAD to the closest left_right boundary</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#length-distribution-of-generalized-tads">Length distribution of generalized TADs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#length-distribution-of-general-tadbs">Length distribution of general TADBs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compare-tadb-and-cis-windows">Compare TADB and cis windows</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#construct-tadb-enhanced-cis-window">Construct TADB-enhanced cis window</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extended-tadb">Extended TADB</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="generation-of-topologically-associated-domains-and-their-boundaries">
<h1>Generation of Topologically Associated Domains and their Boundaries<a class="headerlink" href="#generation-of-topologically-associated-domains-and-their-boundaries" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>During fine mapping we often use cis window (up &amp; downstream 1M base of te start of gene) as the fine mapping region. However, this method lacks justification in biological level, so we will use topologically associated domains (TAD) to create a list of fine mapping regions.</p>
<p>We use generalized TAD region lists created from the merging of TAD data from hippocampus and cortex brain tissues obtained from [<a class="reference external" href="https://doi.org/10.1016%2Fj.ajhg.2021.01.001">cf. McArthur et al (2021)</a> and [<a class="reference external" href="https://doi.org/10.1016%2Fj.celrep.2016.10.061">cf. Schmitt et al (2017)</a>. Variants within boundaries may also have an effect on gene expresion, so we explore how to extend TAD regions to the adjacent boundaries (TADB), in an attempt to not leave out any possible causal variants.</p>
<p>However, we are trying to be conservative, so we do not want to lose information gained from the use of 1Mb cis windows. Therefore, we want to extend those regions. Here we extend both TADB and TADB-enhanced cis windows since each has a different goal.</p>
<p>To build a list of extended TADB, we start with generalized TADB. We find all genes within each TADB and get their start and end positions extended by 1Mb cis windows. We then take the outmost boundary of all the !M cis windows and this TADB as the boundary of the extended TADB. We end with 1,381 TADBs which may then be used as functional units of epigenetic analysis.</p>
<p>To build a list of TADB-enhanced cis windows, we start with the cis window of each gene. We take the outermost boundary of the generalized TADB the gene is in and the 1Mb cis window of the gene as the boundary of the TADB-enhanced cis window. If one gene is in two generalized TADBs, we take the outermost of both the TADB and cis window. We end with a cis window for each gene.</p>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Cortex_DLPFC_Schmitt2016-raw_TADs.txt</span></code> and <code class="docutils literal notranslate"><span class="pre">Hippocampus_Schmitt2016-raw_TADs.txt</span></code> for initial TAD data. Obtained from [<a class="reference external" href="https://doi.org/10.1016%2Fj.ajhg.2021.01.001">cf. McArthur et al (2021)</a> including hg38 TAD regions in cortex and hippocampal tissues collected by [<a class="reference external" href="https://doi.org/10.1016%2Fj.celrep.2016.10.061">cf. Schmitt et al (2017)</a>. These are used to define the genotype cis region for fine mapping with different phenotypes. These TAD regions may be downloaded <a class="reference external" href="http://3dgenome.fsm.northwestern.edu/publications.html">here</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf</span></code>，downloaded from <a class="reference external" href="https://www.synapse.org/#!Synapse:syn36419586">https://www.synapse.org/#!Synapse:syn36419586</a>.</p></li>
</ul>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Generalized TAD: <code class="docutils literal notranslate"><span class="pre">generalized_TAD.tsv</span></code></p></li>
<li><p>Generalized TADB : <code class="docutils literal notranslate"><span class="pre">generalized_TADB.tsv</span></code></p></li>
<li><p>TADB-enhanced cis window: (also in fungen-xqtl github) <code class="docutils literal notranslate"><span class="pre">TADB_enhanced_cis.bed</span></code></p></li>
<li><p>Extended TADB: (also in fungen-xqtl github)
<code class="docutils literal notranslate"><span class="pre">extended_TADB.bed</span></code></p></li>
</ul>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<section id="i-manage-tad-redundancy">
<h3>i. Manage TAD Redundancy<a class="headerlink" href="#i-manage-tad-redundancy" title="Link to this heading">#</a></h3>
<p>Timing: ~20 min</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># merge the TADs to one file</span>
<span class="n">cat</span> <span class="o">../../</span><span class="n">reference_data</span><span class="o">/</span><span class="n">TAD</span><span class="o">/</span><span class="n">Hippocampus_Schmitt2016</span><span class="o">-</span><span class="n">raw_TADs</span><span class="o">.</span><span class="n">txt</span> <span class="o">../../</span><span class="n">reference_data</span><span class="o">/</span><span class="n">TAD</span><span class="o">/</span><span class="n">Cortex_DLPFC_Schmitt2016</span><span class="o">-</span><span class="n">raw_TADs</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="o">../../</span><span class="n">reference_data</span><span class="o">/</span><span class="n">TAD</span><span class="o">/</span><span class="n">brain_TADs</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">find_TAD_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">inputDF</span><span class="p">){</span>
<span class="w">    </span><span class="n">rowChr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="p">[</span><span class="s">&#39;chr&#39;</span><span class="p">]</span>
<span class="w">    </span><span class="n">rowStart</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;start&#39;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowEnd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;end&#39;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowTADIndex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="p">[</span><span class="s">&#39;TAD_index&#39;</span><span class="p">]</span>

<span class="w">    </span><span class="n">TADsubset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">inputDF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">rowChr</span><span class="p">)</span><span class="w"> </span>
<span class="w">    </span><span class="n">TADsubset</span><span class="o">$</span><span class="n">start</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">TADsubset</span><span class="o">$</span><span class="n">start</span><span class="p">)</span>
<span class="w">    </span><span class="n">TADsubset</span><span class="o">$</span><span class="n">end</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">TADsubset</span><span class="o">$</span><span class="n">end</span><span class="p">)</span>

<span class="w">    </span><span class="n">priorTADsubset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">TADsubset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">filter</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">rowStart</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">rowStart</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">&amp;</span>
<span class="w">              </span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowStart</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowEnd</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">arrange</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
<span class="w">    </span><span class="n">nextTADsubset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">TADsubset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">filter</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">rowEnd</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">rowEnd</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">&amp;</span>
<span class="w">              </span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowStart</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowEnd</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">arrange</span><span class="p">(</span><span class="o">-</span><span class="n">end</span><span class="p">)</span>
<span class="w">    </span><span class="n">completeOverlapSubset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">TADsubset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">filter</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">rowStart</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">rowEnd</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">&amp;</span>
<span class="w">              </span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowStart</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowEnd</span><span class="p">))</span>
<span class="w">    </span>
<span class="w">    </span><span class="n">priorOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="w">    </span><span class="n">prior_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowTADIndex</span>
<span class="w">    </span><span class="n">nextOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="w">    </span><span class="n">next_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowTADIndex</span>
<span class="w">    </span><span class="n">completeOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">FALSE</span>

<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">priorTADsubset</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">priorOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">priorTADsubset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">            </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">                </span><span class="n">inner_TAD_Length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">rowStart</span><span class="p">,</span>
<span class="w">                </span><span class="n">outer_TAD_Length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">,</span>
<span class="w">            </span><span class="n">overlap</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">(</span><span class="n">inner_TAD_Length</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">outer_TAD_Length</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">            </span><span class="nf">arrange</span><span class="p">(</span><span class="o">-</span><span class="n">overlap</span><span class="p">)</span>
<span class="w">        </span><span class="n">prior_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">priorOverlap</span><span class="o">$</span><span class="n">TAD_index</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">        </span><span class="n">priorOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">priorOverlap</span><span class="o">$</span><span class="n">overlap</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">        </span><span class="kr">if</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">prior_TAD_index</span><span class="p">))</span><span class="w"> </span><span class="nf">stop</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;The following TAD has an issue:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">rowTADIndex</span><span class="p">))</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">nextTADsubset</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">nextOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">nextTADsubset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">            </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">                </span><span class="n">inner_TAD_Length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowEnd</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">,</span>
<span class="w">                </span><span class="n">outer_TAD_Length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">,</span>
<span class="w">            </span><span class="n">overlap</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">(</span><span class="n">inner_TAD_Length</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">outer_TAD_Length</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">            </span><span class="nf">arrange</span><span class="p">(</span><span class="o">-</span><span class="n">overlap</span><span class="p">)</span>
<span class="w">        </span><span class="n">next_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">nextOverlap</span><span class="o">$</span><span class="n">TAD_index</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">        </span><span class="n">nextOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">nextOverlap</span><span class="o">$</span><span class="n">overlap</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">        </span><span class="kr">if</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">next_TAD_index</span><span class="p">))</span><span class="w"> </span><span class="nf">stop</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;The following TAD has an issue:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">rowTADIndex</span><span class="p">))</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">completeOverlapSubset</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">completeOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">TRUE</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">prior</span><span class="o">=</span><span class="nf">as.character</span><span class="p">(</span><span class="n">priorOverlap</span><span class="p">),</span><span class="w"> </span><span class="n">subsequent</span><span class="o">=</span><span class="nf">as.character</span><span class="p">(</span><span class="n">nextOverlap</span><span class="p">),</span><span class="w"> </span><span class="n">com</span><span class="o">=</span><span class="nf">as.character</span><span class="p">(</span><span class="n">completeOverlap</span><span class="p">),</span><span class="w"> </span><span class="n">prior_tad</span><span class="o">=</span><span class="nf">as.character</span><span class="p">(</span><span class="n">prior_TAD_index</span><span class="p">),</span><span class="w"> </span><span class="n">next_tad</span><span class="o">=</span><span class="nf">as.character</span><span class="p">(</span><span class="n">next_TAD_index</span><span class="p">)))</span>
<span class="p">}</span>

<span class="n">merge_TADs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">inputDF</span><span class="p">,</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">80</span><span class="p">){</span>
<span class="w">    </span><span class="n">rowChr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="p">[</span><span class="s">&quot;chr&quot;</span><span class="p">]</span>
<span class="w">    </span><span class="n">rowStart</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&quot;start&quot;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowEnd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&quot;end&quot;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowTADIndex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;TAD_index&#39;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowPriorOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&quot;prior_overlap&quot;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowPriorTADIndex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&quot;prior_TAD_index&quot;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowNextOverlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&quot;next_overlap&quot;</span><span class="p">])</span>
<span class="w">    </span><span class="n">rowNextTADIndex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&quot;next_TAD_index&quot;</span><span class="p">])</span>
<span class="w">    </span>
<span class="w">    </span><span class="n">newStart</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowStart</span>
<span class="w">    </span><span class="n">newEnd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowEnd</span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">rowNextOverlap</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="n">rowNextTADIndex</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowTADIndex</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">newEnd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">inputDF</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">            </span><span class="nf">filter</span><span class="p">(</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">rowNextTADIndex</span><span class="p">)</span>
<span class="w">        </span><span class="kr">if</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">newEnd</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">end</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">length</span><span class="p">()))</span><span class="w"> </span><span class="nf">print</span><span class="p">(</span><span class="n">newEnd</span><span class="o">$</span><span class="n">end</span><span class="p">[</span><span class="m">1</span><span class="p">])</span>
<span class="w">        </span><span class="n">newEnd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">newEnd</span><span class="o">$</span><span class="n">end</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">rowPriorOverlap</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">rowPriorTADIndex</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">rowTADIndex</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">newStart</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">inputDF</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">            </span><span class="nf">filter</span><span class="p">(</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">rowPriorTADIndex</span><span class="p">)</span>
<span class="w">        </span><span class="n">newStart</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">newStart</span><span class="o">$</span><span class="n">start</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">newstart</span><span class="o">=</span><span class="n">newStart</span><span class="p">,</span><span class="w"> </span><span class="n">newend</span><span class="o">=</span><span class="n">newEnd</span><span class="p">))</span>
<span class="p">}</span>

<span class="n">recursive_merge</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">tadDF</span><span class="p">){</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;TAD_&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">tadDF</span><span class="p">)))</span>

<span class="w">    </span><span class="n">overlap_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">apply</span><span class="p">(</span><span class="n">tadDF</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">find_TAD_overlap</span><span class="p">,</span><span class="w"> </span><span class="n">tadDF</span><span class="p">)</span><span class="w"> </span><span class="c1"># First Call</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior&#39;</span><span class="p">))</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">prior_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior_tad&#39;</span><span class="p">))</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;subsequent&#39;</span><span class="p">))</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">next_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;next_tad&#39;</span><span class="p">))</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.logical</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;com&#39;</span><span class="p">))</span>

<span class="w">    </span><span class="n">merge_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">apply</span><span class="p">(</span><span class="n">tadDF</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">merge_TADs</span><span class="p">,</span><span class="w"> </span><span class="n">tadDF</span><span class="p">,</span><span class="w"> </span><span class="m">80</span><span class="p">)</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">end</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">merge_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;newend&#39;</span><span class="p">))</span>
<span class="w">    </span><span class="n">tadDF</span><span class="o">$</span><span class="n">start</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">merge_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;newstart&#39;</span><span class="p">))</span>
<span class="w">    </span><span class="n">candidateDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tadDF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">distinct</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="n">.keep_all</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">tadDF</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">candidateDF</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="kr">return</span><span class="p">(</span><span class="n">candidateDF</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;TAD_&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">candidateDF</span><span class="p">)))</span>
<span class="w">        </span><span class="n">overlap_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">apply</span><span class="p">(</span><span class="n">candidateDF</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">find_TAD_overlap</span><span class="p">,</span><span class="w"> </span><span class="n">candidateDF</span><span class="p">)</span><span class="w"> </span><span class="c1"># First Call</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">prior_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior_tad&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;subsequent&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">next_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;next_tad&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.logical</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;com&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">candidateDF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>

<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;TAD_&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">candidateDF</span><span class="p">)))</span>
<span class="w">        </span><span class="n">overlap_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">apply</span><span class="p">(</span><span class="n">candidateDF</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">find_TAD_overlap</span><span class="p">,</span><span class="w"> </span><span class="n">candidateDF</span><span class="p">)</span><span class="w"> </span><span class="c1"># Second Call</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">prior_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior_tad&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;subsequent&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">next_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;next_tad&#39;</span><span class="p">))</span>
<span class="w">        </span><span class="n">candidateDF</span><span class="o">$</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.logical</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;com&#39;</span><span class="p">))</span>

<span class="w">        </span><span class="kr">return</span><span class="p">(</span><span class="nf">recursive_merge</span><span class="p">(</span><span class="n">candidateDF</span><span class="p">))</span>
<span class="w">    </span><span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<section id="merge-tads-with-neighboring-tabs">
<h4>Merge TADs With Neighboring TABs<a class="headerlink" href="#merge-tads-with-neighboring-tabs" title="Link to this heading">#</a></h4>
<p>Includes a step to manually correct the start and end position of each chromosome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">general_tad_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;../../reference_data/TAD/brain_TADs.txt&quot;</span>
<span class="n">general_TAD_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="n">general_tad_path</span><span class="p">,</span><span class="w"> </span><span class="n">col_names</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;chr&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;start&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;end&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">show_col_types</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<span class="n">general_TAD_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">general_TAD_DF</span><span class="p">[</span><span class="nf">with</span><span class="p">(</span><span class="n">general_TAD_DF</span><span class="p">,</span><span class="w"> </span><span class="nf">order</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">end</span><span class="p">)),]</span>
<span class="n">general_TAD_DF</span><span class="o">$</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;TAD_&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="nf">nrow</span><span class="p">(</span><span class="n">general_TAD_DF</span><span class="p">)))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="preliminary-merges">
<h4>Preliminary Merges<a class="headerlink" href="#preliminary-merges" title="Link to this heading">#</a></h4>
<p>Remove duplicates (same chr, start, and end).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">general_TAD_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">general_TAD_DF</span><span class="p">[</span><span class="nf">with</span><span class="p">(</span><span class="n">general_TAD_DF</span><span class="p">,</span><span class="w"> </span><span class="nf">order</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">end</span><span class="p">)),]</span>

<span class="nf">head</span><span class="p">(</span><span class="n">general_TAD_DF</span><span class="p">)</span>

<span class="c1"># Initial Number of TADs</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Initial number of TADs before removing redundancy:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">general_TAD_DF</span><span class="p">))</span>

<span class="c1"># Remove duplicate TAD sites</span>
<span class="n">general_TAD_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">general_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">distinct</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="n">.keep_all</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">arrange</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="p">)</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs left after removing duplicate TADs:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">general_TAD_DF</span><span class="p">))</span>

<span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">general_TAD_DF</span>
<span class="n">reduced_TAD_DF</span><span class="o">$</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;TAD_&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="p">)))</span>

<span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 6 × 4</caption>
<thead>
	<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TAD_index</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1</td><td> 760000</td><td>6160000</td><td>TAD_1</td></tr>
	<tr><td>chr1</td><td> 800000</td><td>6120000</td><td>TAD_2</td></tr>
	<tr><td>chr1</td><td>6480000</td><td>7640000</td><td>TAD_3</td></tr>
	<tr><td>chr1</td><td>6480000</td><td>7600000</td><td>TAD_4</td></tr>
	<tr><td>chr1</td><td>7920000</td><td>9480000</td><td>TAD_5</td></tr>
	<tr><td>chr1</td><td>7960000</td><td>9520000</td><td>TAD_6</td></tr>
</tbody>
</table>
</div><div class="output text_html">'Initial number of TADs before removing redundancy: 2973'</div><div class="output text_html">'Number of TADs left after removing duplicate TADs: 2652'</div><div class="output text_html">2652</div><div class="output text_html"><table class="dataframe">
<caption>A tibble: 6 × 4</caption>
<thead>
	<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TAD_index</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1</td><td> 760000</td><td>6160000</td><td>TAD_1</td></tr>
	<tr><td>chr1</td><td> 800000</td><td>6120000</td><td>TAD_2</td></tr>
	<tr><td>chr1</td><td>6480000</td><td>7600000</td><td>TAD_3</td></tr>
	<tr><td>chr1</td><td>6480000</td><td>7640000</td><td>TAD_4</td></tr>
	<tr><td>chr1</td><td>7920000</td><td>9480000</td><td>TAD_5</td></tr>
	<tr><td>chr1</td><td>7960000</td><td>9520000</td><td>TAD_6</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
<section id="determining-overlap">
<h4>Determining Overlap<a class="headerlink" href="#determining-overlap" title="Link to this heading">#</a></h4>
<p>Determining the overlap between TADs is important for the later merging steps. In this step, we find:</p>
<ul class="simple">
<li><p>Overlap between a TAD and the preceding TAD</p></li>
<li><p>Overlap between a TAD and the following TAD</p></li>
<li><p>TADs that are entirely contained within another TAD</p></li>
</ul>
<section id="overlap-check">
<h5>Overlap Check<a class="headerlink" href="#overlap-check" title="Link to this heading">#</a></h5>
<p>There are two values for the following overlap checks. The first value is the overlap with the preceding TAD, and the second value is the overlap of the following TAD.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute overlap</span>
<span class="n">overlap_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">apply</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">find_TAD_overlap</span><span class="p">,</span><span class="w"> </span><span class="n">reduced_TAD_DF</span><span class="p">)</span>
<span class="n">reduced_TAD_DF</span><span class="o">$</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior&#39;</span><span class="p">))</span>
<span class="n">reduced_TAD_DF</span><span class="o">$</span><span class="n">prior_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior_tad&#39;</span><span class="p">))</span>
<span class="n">reduced_TAD_DF</span><span class="o">$</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;subsequent&#39;</span><span class="p">))</span>
<span class="n">reduced_TAD_DF</span><span class="o">$</span><span class="n">next_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;next_tad&#39;</span><span class="p">))</span>
<span class="n">reduced_TAD_DF</span><span class="o">$</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.logical</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;com&#39;</span><span class="p">))</span>

<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with 100% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 80% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">80</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">80</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 50% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">50</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">50</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 25% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 10% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">10</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">10</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'Number of TADs with 100% overlap: 1030'</div><div class="output text_html">'Number of TADs with greater than 80% overlap: 1386 1379'</div><div class="output text_html">'Number of TADs with greater than 50% overlap: 1531 1546'</div><div class="output text_html">'Number of TADs with greater than 25% overlap: 1630 1634'</div><div class="output text_html">'Number of TADs with greater than 10% overlap: 1682 1686'</div></div>
</div>
</section>
</section>
<section id="recursively-merge-tads">
<h4>Recursively Merge TADs<a class="headerlink" href="#recursively-merge-tads" title="Link to this heading">#</a></h4>
<section id="overview">
<h5>Overview<a class="headerlink" href="#overview" title="Link to this heading">#</a></h5>
<ol class="arabic simple">
<li><p>Extend the given TAD to the overlapped TAD position</p>
<ul class="simple">
<li><p>If a given TAD overlaps the prededing TAD, extend the start position</p></li>
<li><p>If a given TAD overlaps the subsequent TAD, extend the end position</p></li>
<li><p>If a given TAD overlaps both ends, extend both</p></li>
</ul>
</li>
<li><p>Remove TADs that are completely overlapped</p></li>
<li><p>If the TAD is reduced, compute the overlap for the reduced dataframe and rerun the merging function until the dataframe dimensions stop changing</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">recursive_merge</span><span class="p">(</span><span class="n">reduced_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs in the finalized TAD dataset:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="p">))</span>
<span class="nf">print</span><span class="p">(</span><span class="s">&quot;Preview of the finalized TAD dataset:&quot;</span><span class="p">)</span>
<span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">select</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;end&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'Number of TADs in the finalized TAD dataset: 1401'</div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Preview of the finalized TAD dataset:&quot;
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 6 × 3</caption>
<thead>
	<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1</td><td>  760000</td><td> 6160000</td></tr>
	<tr><td>chr1</td><td> 6480000</td><td> 7640000</td></tr>
	<tr><td>chr1</td><td> 7920000</td><td> 9520000</td></tr>
	<tr><td>chr1</td><td> 9880000</td><td>10480000</td></tr>
	<tr><td>chr1</td><td>10520000</td><td>12040000</td></tr>
	<tr><td>chr1</td><td>12040000</td><td>12840000</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">overlap_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">apply</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">find_TAD_overlap</span><span class="p">,</span><span class="w"> </span><span class="n">final_brain_TAD_DF</span><span class="p">)</span>
<span class="n">final_brain_TAD_DF</span><span class="o">$</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior&#39;</span><span class="p">))</span>
<span class="n">final_brain_TAD_DF</span><span class="o">$</span><span class="n">prior_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;prior_tad&#39;</span><span class="p">))</span>
<span class="n">final_brain_TAD_DF</span><span class="o">$</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;subsequent&#39;</span><span class="p">))</span>
<span class="n">final_brain_TAD_DF</span><span class="o">$</span><span class="n">next_TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;next_tad&#39;</span><span class="p">))</span>
<span class="n">final_brain_TAD_DF</span><span class="o">$</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.logical</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">overlap_results</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;[[&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;com&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with 100% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 80% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">80</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">80</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 50% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">50</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">50</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 25% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">)))</span>
<span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Number of TADs with greater than 10% overlap:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">prior_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">10</span><span class="p">)),</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">next_overlap</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">10</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'Number of TADs with 100% overlap: 0'</div><div class="output text_html">'Number of TADs with greater than 80% overlap: 0 0'</div><div class="output text_html">'Number of TADs with greater than 50% overlap: 14 19'</div><div class="output text_html">'Number of TADs with greater than 25% overlap: 26 31'</div><div class="output text_html">'Number of TADs with greater than 10% overlap: 39 39'</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">formatted_final_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;chr&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">22</span><span class="p">))</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">complete_overlap</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">select</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;end&quot;</span><span class="p">))</span><span class="w"> </span>
<span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">TAD_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;TAD&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="nf">nrow</span><span class="p">(</span><span class="n">formatted_final_DF</span><span class="p">)))</span>
<span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">end</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">end</span><span class="p">)</span>
<span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">start</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">start</span><span class="p">)</span>

<span class="n">formatted_final_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">formatted_final_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">    </span><span class="n">TAD_Length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span>

<span class="nf">quantile</span><span class="p">(</span><span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">TAD_Length</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.19</span><span class="p">,</span><span class="m">0.25</span><span class="p">,</span><span class="m">0.5</span><span class="p">,</span><span class="m">0.75</span><span class="p">,</span><span class="m">0.95</span><span class="p">,</span><span class="m">0.99</span><span class="p">,</span><span class="m">0.995</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.dl-inline {width: auto; margin:0; padding: 0}
.dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}
.dl-inline>dt::after {content: ":\0020"; padding-right: .5ex}
.dl-inline>dt:not(:first-of-type) {padding-left: .5ex}
</style><dl class=dl-inline><dt>19%</dt><dd>917200</dd><dt>25%</dt><dd>1040000</dd><dt>50%</dt><dd>1680000</dd><dt>75%</dt><dd>2520000</dd><dt>95%</dt><dd>3880000</dd><dt>99%</dt><dd>5200000</dd><dt>99.5%</dt><dd>5551800.00000001</dd><dt>100%</dt><dd>6080000</dd></dl>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">formatted_final_DF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">final_brain_TAD_DF</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">select</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;end&quot;</span><span class="p">))</span>
<span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">start</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">format</span><span class="p">(</span><span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">scientific</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">end</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">format</span><span class="p">(</span><span class="n">formatted_final_DF</span><span class="o">$</span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="n">scientific</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">write.table</span><span class="p">(</span>
<span class="w">    </span><span class="n">formatted_final_DF</span><span class="p">,</span>
<span class="w">    </span><span class="s">&quot;../../reference_data/TAD/generalized_TAD.tsv&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">append</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span>
<span class="w">    </span><span class="n">row.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span>
<span class="w">    </span><span class="n">col.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span>
<span class="w">    </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
</section>
<section id="ii-generate-tadb-enhanced-cis-windows-and-extended-tadbs">
<h2>ii. Generate TADB-enhanced cis windows and extended TADBs<a class="headerlink" href="#ii-generate-tadb-enhanced-cis-windows-and-extended-tadbs" title="Link to this heading">#</a></h2>
<p>Timing: ~20 min</p>
<section id="how-to-find-boundaries-and-build-tadb">
<h3>How to find boundaries and build TADB?<a class="headerlink" href="#how-to-find-boundaries-and-build-tadb" title="Link to this heading">#</a></h3>
<p>After we had the generalized TAD, it’s not the simple case that all intervals between TADs are the boundaries.
The ideal (and most of) generalized TADs have a structure like this:</p>
<p><img alt="" src="../../_images/TADB_idealcase.png" /></p>
<p>The black line is the genome. Red and blue shaded triangles are TAD results from hippo/cortex, respectively.
In the later code we will denote them as <strong>specific TAD</strong> because they are specific to one tissue.</p>
<p>The two specific TAD lists (blue and red) don’t have overlaps internally. But if we pool them together, red and blue overlap. If their overlap exceeds <strong>80%</strong>, they will be merged to a larger TAD, called <strong>Generalized TAD</strong> (Green triangle). That’s how we get the generalized TAD list.</p>
<p>However, it’s not necessary that specific TAD line up in this way, so there are more cases. For example, if two specific TAD <strong>do not overlap enough</strong>, or one specific TAD does not even overlap with any other TADs, then they will both one generalized TAD <strong>solely</strong>. The diagram is shown as below.</p>
<p><img alt="" src="../../_images/sole_case.png" /></p>
<p>We can see from the ideal case, at the boundary of generalized TAD, there are regions not covered by all two specific TADs, which means not all information belongs to a TAD, so it can belong to TAB (boundary regions).
That’s how we define a <strong>Generalized TAB</strong>, not only the intervals, but also include the ambigous regions. They are shown as the <strong>yellow</strong> triangles in the diagram.</p>
<p><img alt="" src="../../_images/TADB_diagram.png" /></p>
<p>Then to get TADBs, we will include the generalized TAD and adjacent TABs. In the diagram for generalized TAD2, the TADB will be (generalized) TAB1 + TAD2 + TAB2.</p>
<p>Solely formed generalized TAD have no ambiguous regions, so their boundaries will also be boundaries of generalized TAB.
Here we give a example.</p>
<p>We search for the last TADs right boundary and extend there, then search for the next TADs left boundary and extend there.
Note: if two generalized TAD overlap, we will serach for the boundary of non-overlapping one.</p>
<p>It’s not hard to find that under this definition, TAD2 and TAD3 share the same TADB.</p>
<p><img alt="" src="../../_images/TADB_solely.png" /></p>
<p>So now our strategy is to work out a list of left boundaries and right boundaries. For a generalized TAD formed by two or more specific TADs, the left boundary will be the start of the second specific TAD forming this generalized TAD.
For a generalized TAD formed solely by one, then the left boundary is just the start of this generlized TAD.</p>
<p>It’s similar for the right boundaries, the only difference we use the last but one specific TADs end as the right boundary. Here are the diagrams.</p>
<p>Left boundary will be labled as red vertical lines, right will be labeld as blue verticle lines. The definition also fits for more complicated cases.</p>
<p><img alt="" src="../../_images/left_right.png" /></p>
<p>After having these boundaries our strategy to get TADB will be to search from start of generalized TAD, extend left side to last right boundary (including the start of chromosome, 0). Search from end of generalized TAD, extend right side to the next left boundary (including the end of each chromosomes).</p>
</section>
<section id="tadb-coding">
<h3>TADB - coding<a class="headerlink" href="#tadb-coding" title="Link to this heading">#</a></h3>
<p>Total number of generalized TADs: 1401.</p>
<p>Then we checked how many specific TAD forms each generalized TAD. That’s because as shown above, those generalized TAD formed by one vs more specific TADs, will have different “boundaries”.</p>
<p>Here we begin with a gtf to get the start and end of a gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">import</span><span class="w"> </span><span class="n">pandas</span><span class="w"> </span><span class="n">as</span><span class="w"> </span><span class="n">pd</span>
<span class="n">import</span><span class="w"> </span><span class="n">numpy</span><span class="w"> </span><span class="n">as</span><span class="w"> </span><span class="n">np</span>
<span class="n">from</span><span class="w"> </span><span class="n">collections</span><span class="w"> </span><span class="n">import</span><span class="w"> </span><span class="n">defaultdict</span>
<span class="n">import</span><span class="w"> </span><span class="n">subprocess</span>
<span class="n">import</span><span class="w"> </span><span class="n">gzip</span>

<span class="n">def</span><span class="w"> </span><span class="nf">gtf_to_start_end_bed</span><span class="p">(</span><span class="n">annotation_gtf</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="o">=</span><span class="s">&#39;gene&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">exclude_chrs</span><span class="o">=</span><span class="p">[],</span><span class="w"> </span><span class="n">phenotype_id</span><span class="o">=</span><span class="s">&#39;gene_id&#39;</span><span class="p">)</span><span class="o">:</span>
<span class="w">    </span><span class="s">&quot;&quot;&quot;Parse genes and starts and ends from GTF and return DataFrame for BED output&quot;&quot;&quot;</span>
<span class="w">    </span><span class="n">chrom</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>
<span class="w">    </span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>
<span class="w">    </span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>
<span class="w">    </span><span class="n">gene_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>
<span class="w">    </span><span class="n">gene_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[]</span>

<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="nf">annotation_gtf.endswith</span><span class="p">(</span><span class="s">&#39;.gz&#39;</span><span class="p">)</span><span class="o">:</span>
<span class="w">        </span><span class="n">opener</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">gzip.open</span><span class="p">(</span><span class="n">annotation_gtf</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;rt&#39;</span><span class="p">)</span>
<span class="w">    </span><span class="kr">else</span><span class="o">:</span>
<span class="w">        </span><span class="n">opener</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">open</span><span class="p">(</span><span class="n">annotation_gtf</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;r&#39;</span><span class="p">)</span>

<span class="w">    </span><span class="n">with</span><span class="w"> </span><span class="n">opener</span><span class="w"> </span><span class="n">as</span><span class="w"> </span><span class="n">gtf</span><span class="o">:</span>
<span class="w">        </span><span class="kr">for</span><span class="w"> </span><span class="n">row</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="n">gtf</span><span class="o">:</span>
<span class="w">            </span><span class="n">row</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">row.strip</span><span class="p">()</span><span class="nf">.split</span><span class="p">(</span><span class="s">&#39;\t&#39;</span><span class="p">)</span>
<span class="w">            </span><span class="kr">if</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="m">0</span><span class="p">][</span><span class="m">0</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&#39;#&#39;</span><span class="w"> </span><span class="n">or</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">feature</span><span class="o">:</span><span class="w"> </span><span class="n">continue</span><span class="w"> </span><span class="c1"># skip header</span>
<span class="w">            </span><span class="nf">chrom.append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="m">0</span><span class="p">])</span>

<span class="w">            </span><span class="c1"># extract start and end</span>
<span class="w">            </span><span class="kr">if</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="m">6</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&#39;+&#39;</span><span class="o">:</span>
<span class="w">                </span><span class="nf">start.append</span><span class="p">(</span><span class="nf">np.int64</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="m">3</span><span class="p">])</span><span class="w"> </span><span class="m">-1</span><span class="p">)</span>
<span class="w">                </span><span class="nf">end.append</span><span class="p">(</span><span class="nf">np.int64</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="m">4</span><span class="p">]))</span>
<span class="w">            </span><span class="n">elif</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="m">6</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&#39;-&#39;</span><span class="o">:</span>
<span class="w">                </span><span class="nf">start.append</span><span class="p">(</span><span class="nf">np.int64</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="m">3</span><span class="p">]))</span><span class="w">  </span><span class="c1"># last base of gene</span>
<span class="w">                </span><span class="nf">end.append</span><span class="p">(</span><span class="nf">np.int64</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="m">4</span><span class="p">]))</span>
<span class="w">            </span><span class="kr">else</span><span class="o">:</span>
<span class="w">                </span><span class="n">raise</span><span class="w"> </span><span class="nf">ValueError</span><span class="p">(</span><span class="s">&#39;Strand not specified.&#39;</span><span class="p">)</span>

<span class="w">            </span><span class="n">attributes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">defaultdict</span><span class="p">()</span>
<span class="w">            </span><span class="kr">for</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="n">row</span><span class="p">[</span><span class="m">8</span><span class="p">]</span><span class="nf">.replace</span><span class="p">(</span><span class="s">&#39;&quot;&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="p">)</span><span class="nf">.split</span><span class="p">(</span><span class="s">&#39;;&#39;</span><span class="p">)[</span><span class="o">:</span><span class="m">-1</span><span class="p">]</span><span class="o">:</span>
<span class="w">                </span><span class="n">kv</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">a.strip</span><span class="p">()</span><span class="nf">.split</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">)</span>
<span class="w">                </span><span class="kr">if</span><span class="w"> </span><span class="n">kv</span><span class="p">[</span><span class="m">0</span><span class="p">]</span><span class="o">!=</span><span class="s">&#39;tag&#39;</span><span class="o">:</span>
<span class="w">                    </span><span class="n">attributes</span><span class="p">[</span><span class="n">kv</span><span class="p">[</span><span class="m">0</span><span class="p">]]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">kv</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">                </span><span class="kr">else</span><span class="o">:</span>
<span class="w">                    </span><span class="nf">attributes.setdefault</span><span class="p">(</span><span class="s">&#39;tags&#39;</span><span class="p">,</span><span class="w"> </span><span class="p">[])</span><span class="nf">.append</span><span class="p">(</span><span class="n">kv</span><span class="p">[</span><span class="m">1</span><span class="p">])</span>

<span class="w">            </span><span class="nf">gene_id.append</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;gene_id&#39;</span><span class="p">])</span>
<span class="w">            </span><span class="nf">gene_name.append</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;gene_name&#39;</span><span class="p">])</span>

<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="n">phenotype_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&#39;gene_id&#39;</span><span class="o">:</span>
<span class="w">        </span><span class="n">bed_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pd.DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;chr&#39;</span><span class="o">:</span><span class="n">chrom</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;start&#39;</span><span class="o">:</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;end&#39;</span><span class="o">:</span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;gene_id&#39;</span><span class="o">:</span><span class="n">gene_id</span><span class="p">},</span><span class="w"> </span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;chr&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;start&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;end&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;gene_id&#39;</span><span class="p">],</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">gene_id</span><span class="p">)</span>
<span class="w">    </span><span class="n">elif</span><span class="w"> </span><span class="n">phenotype_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&#39;gene_name&#39;</span><span class="o">:</span>
<span class="w">        </span><span class="n">bed_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pd.DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;chr&#39;</span><span class="o">:</span><span class="n">chrom</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;start&#39;</span><span class="o">:</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;end&#39;</span><span class="o">:</span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;gene_id&#39;</span><span class="o">:</span><span class="n">gene_name</span><span class="p">},</span><span class="w"> </span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;chr&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;start&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;end&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;gene_id&#39;</span><span class="p">],</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">gene_name</span><span class="p">)</span>
<span class="w">    </span><span class="c1"># drop rows corresponding to excluded chromosomes</span>
<span class="w">    </span><span class="n">mask</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">np.ones</span><span class="p">(</span><span class="nf">len</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span><span class="w"> </span><span class="n">dtype</span><span class="o">=</span><span class="n">bool</span><span class="p">)</span>
<span class="w">    </span><span class="kr">for</span><span class="w"> </span><span class="n">k</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="n">exclude_chrs</span><span class="o">:</span>
<span class="w">        </span><span class="n">mask</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mask</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="p">(</span><span class="n">bed_df</span><span class="p">[</span><span class="s">&#39;chr&#39;</span><span class="p">]</span><span class="o">!=</span><span class="n">k</span><span class="p">)</span>
<span class="w">    </span><span class="n">bed_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bed_df</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>

<span class="w">    </span><span class="c1"># sort by start position</span>
<span class="w">    </span><span class="n">bed_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">bed_df.groupby</span><span class="p">(</span><span class="s">&#39;chr&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">sort</span><span class="o">=</span><span class="n">False</span><span class="p">,</span><span class="w"> </span><span class="n">group_keys</span><span class="o">=</span><span class="n">False</span><span class="p">)</span><span class="nf">.apply</span><span class="p">(</span><span class="n">lambda</span><span class="w"> </span><span class="n">x</span><span class="o">:</span><span class="w"> </span><span class="nf">x.sort_values</span><span class="p">(</span><span class="s">&#39;start&#39;</span><span class="p">))</span>

<span class="w">    </span><span class="kr">return</span><span class="w"> </span><span class="n">bed_df</span>
<span class="c1">#../../reference_data/TAD/generalized_TAD.ts</span>
<span class="c1"># save the bed files recording start and end of a chromosome</span>
<span class="n">bed_template_df_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">gtf_to_start_end_bed</span><span class="p">(</span><span class="s">&quot;../../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="o">=</span><span class="s">&#39;gene&#39;</span><span class="p">,</span><span class="n">phenotype_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;gene_id&quot;</span><span class="w"> </span><span class="p">)</span>
<span class="n">bed_template_df_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">gtf_to_start_end_bed</span><span class="p">(</span><span class="s">&quot;../../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="o">=</span><span class="s">&#39;gene&#39;</span><span class="p">,</span><span class="n">phenotype_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;gene_name&quot;</span><span class="p">)</span>
<span class="n">bed_template_df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">bed_template_df_id.merge</span><span class="p">(</span><span class="n">bed_template_df_name</span><span class="p">,</span><span class="w"> </span><span class="n">on</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="s">&quot;end&quot;</span><span class="p">])</span>
<span class="n">bed_template_df.columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="s">&quot;#chr&quot;</span><span class="p">,</span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="s">&quot;end&quot;</span><span class="p">,</span><span class="s">&quot;gene_id&quot;</span><span class="p">,</span><span class="s">&quot;gene_name&quot;</span><span class="p">]</span>
<span class="nf">bed_template_df.to_csv</span><span class="p">(</span><span class="s">&quot;../../reference_data/TAD/gene_start_end.tsv&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># generalized TAD, already merged if have &gt; 80% overlap</span>
<span class="nf">library</span><span class="p">(</span><span class="s">&quot;tidyverse&quot;</span><span class="p">)</span>
<span class="n">general</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;../../reference_data/TAD/generalized_TAD.tsv&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">col_names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;end&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">row_index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">row_number</span><span class="p">())</span>

<span class="c1"># Specific TADs list -- belong to one of the tissue we interested in</span>
<span class="n">specific</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;../../reference_data/TAD/brain_TADs.txt&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">col_names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;start&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;end&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">row_index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">row_number</span><span class="p">())</span>

<span class="c1"># for each generalized TAD, how many specific TADs they are formed from?</span>

<span class="n">check_TAD</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">start_pos</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="p">){</span>
<span class="w">    </span><span class="n">within_TAD</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">chrm</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">start_pos</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end_pos</span><span class="p">)</span>
<span class="w">    </span><span class="n">within_TAD_num</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">within_TAD</span><span class="p">)</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">within_TAD_num</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">general_number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">general</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">within_num</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmap</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">chrm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span>
<span class="w">            </span><span class="n">start_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="p">),</span><span class="w"> </span><span class="n">check_TAD</span><span class="p">))</span>

<span class="n">general_number</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">group_by</span><span class="p">(</span><span class="n">within_num</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold">Rows: </span><span class=" -Color -Color-Blue">1401</span> <span class=" -Color -Color-Bold">Columns: </span><span class=" -Color -Color-Blue">3</span>
<span class=" -Color -Color-Cyan">──</span> <span class=" -Color -Color-Bold">Column specification</span> <span class=" -Color -Color-Cyan">────────────────────────────────────────────────────────</span>
<span class=" -Color -Color-Bold">Delimiter:</span> &quot;\t&quot;
<span class=" -Color -Color-Red">chr</span> (1): chr
<span class=" -Color -Color-Green">dbl</span> (2): start, end

<span class=" -Color -Color-Cyan">ℹ</span> Use `spec()` to retrieve the full column specification for this data.
<span class=" -Color -Color-Cyan">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.
<span class=" -Color -Color-Bold">Rows: </span><span class=" -Color -Color-Blue">2973</span> <span class=" -Color -Color-Bold">Columns: </span><span class=" -Color -Color-Blue">3</span>
<span class=" -Color -Color-Cyan">──</span> <span class=" -Color -Color-Bold">Column specification</span> <span class=" -Color -Color-Cyan">────────────────────────────────────────────────────────</span>
<span class=" -Color -Color-Bold">Delimiter:</span> &quot;\t&quot;
<span class=" -Color -Color-Red">chr</span> (1): chr
<span class=" -Color -Color-Green">dbl</span> (2): start, end

<span class=" -Color -Color-Cyan">ℹ</span> Use `spec()` to retrieve the full column specification for this data.
<span class=" -Color -Color-Cyan">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A grouped_df: 4 × 2</caption>
<thead>
	<tr><th scope=col>within_num</th><th scope=col>n</th></tr>
	<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
	<tr><td>2</td><td>1163</td></tr>
	<tr><td>3</td><td> 188</td></tr>
	<tr><td>4</td><td>  11</td></tr>
	<tr><td>1</td><td>  39</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Most of the generalized TADs are include 2 specific TADs.</p>
<p>There are cases that one generalized TAD is forming from 4, but even for these cases, our strategy still works.</p>
<section id="find-left-boundaries">
<h4>Find left boundaries<a class="headerlink" href="#find-left-boundaries" title="Link to this heading">#</a></h4>
<p>Left boundaries should include the <strong>end</strong> of the chromosome – because left boundaries of TADs will be the right boundary of former TADB. For the last one TADB, is should have the end of chromosome as its right boundary.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># a list showing the end position of each chromosome, including chrX</span>

<span class="n">chromosomes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;chr&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">))</span>
<span class="n">chromosomes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span><span class="s">&quot;chrX&quot;</span><span class="p">)</span>
<span class="n">bplengths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">248956422</span><span class="p">,</span><span class="w"> </span><span class="m">242193529</span><span class="p">,</span><span class="w"> </span><span class="m">198295559</span><span class="p">,</span><span class="w"> </span><span class="m">190214555</span><span class="p">,</span><span class="w"> </span><span class="m">181538259</span><span class="p">,</span><span class="w"> </span><span class="m">170805979</span><span class="p">,</span><span class="w"> </span><span class="m">159345973</span><span class="p">,</span><span class="w"> </span><span class="m">145138636</span><span class="p">,</span>
<span class="w">                 </span><span class="m">138394717</span><span class="p">,</span><span class="w"> </span><span class="m">133797422</span><span class="p">,</span><span class="w"> </span><span class="m">135086622</span><span class="p">,</span><span class="w"> </span>
<span class="w">               </span><span class="m">133275309</span><span class="p">,</span><span class="w"> </span><span class="m">114364328</span><span class="p">,</span><span class="w"> </span><span class="m">107043718</span><span class="p">,</span><span class="w"> </span><span class="m">101991189</span><span class="p">,</span><span class="w"> </span><span class="m">90338345</span><span class="p">,</span><span class="w"> </span><span class="m">83257441</span><span class="p">,</span><span class="w"> </span>
<span class="w">               </span><span class="m">80373285</span><span class="p">,</span><span class="w"> </span><span class="m">58617616</span><span class="p">,</span><span class="w"> </span><span class="m">64444167</span><span class="p">,</span><span class="w"> </span><span class="m">46709983</span><span class="p">,</span><span class="w"> </span><span class="m">50818468</span><span class="p">,</span><span class="w"> </span><span class="m">156040895</span><span class="p">)</span>

<span class="n">chrDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chromosomes</span><span class="p">,</span><span class="w"> </span><span class="n">left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bplengths</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">find_left_boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">start_pos</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="p">){</span>
<span class="w">    </span><span class="n">within_TAD</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">chrm</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">start_pos</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end_pos</span><span class="p">)</span>
<span class="w">    </span><span class="n">within_TAD_num</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">within_TAD</span><span class="p">)</span>
<span class="w">    </span><span class="kr">if</span><span class="p">(</span><span class="n">within_TAD_num</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span>
<span class="w">        </span><span class="c1"># if formed only by one, then the first start of specific TAD</span>
<span class="w">        </span><span class="n">left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">within_TAD</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">()</span>
<span class="w">    </span><span class="p">}</span><span class="kr">else</span><span class="p">{</span>
<span class="w">        </span><span class="c1"># if formed by more than 2, then the second start of specific TAD</span>
<span class="w">        </span><span class="n">TAD_start_list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">within_TAD</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">sort</span><span class="p">()</span>
<span class="w">        </span><span class="n">left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TAD_start_list</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">()</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">left</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># find all left_boundaries</span>

<span class="n">left_table</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">general</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmap</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">chrm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span>
<span class="w">            </span><span class="n">start_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="p">),</span><span class="w"> </span><span class="n">find_left_boundary</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">            </span><span class="nf">select</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">left</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">left</span><span class="p">))</span>

<span class="c1"># combine left boundaries with end of chromosome to form the final left boundary list</span>

<span class="n">left_table_final</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rbind</span><span class="p">(</span><span class="n">left_table</span><span class="p">,</span><span class="w"> </span><span class="n">chrDF</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="find-right-boundaries">
<h4>Find right boundaries<a class="headerlink" href="#find-right-boundaries" title="Link to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">find_right_boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">start_pos</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="p">){</span>
<span class="w">    </span><span class="n">within_TAD</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">chrm</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">start_pos</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end_pos</span><span class="p">)</span>
<span class="w">    </span><span class="n">within_TAD_num</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">within_TAD</span><span class="p">)</span>
<span class="w">    </span><span class="kr">if</span><span class="p">(</span><span class="n">within_TAD_num</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span>
<span class="w">        </span><span class="n">right</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">within_TAD</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">end</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">()</span>
<span class="w">    </span><span class="p">}</span><span class="kr">else</span><span class="p">{</span>
<span class="w">        </span><span class="n">TAD_end_list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">within_TAD</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">end</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">sort</span><span class="p">(</span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">T</span><span class="p">)</span>
<span class="w">        </span><span class="n">right</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TAD_end_list</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">()</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">right_table</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">general</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">right</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmap</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">chrm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span>
<span class="w">            </span><span class="n">start_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="p">),</span><span class="w"> </span><span class="n">find_right_boundary</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">            </span><span class="nf">select</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">right</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">right</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">right</span><span class="p">))</span>

<span class="c1"># Similarly, we should include start of the chromosome to be the first right boundaries</span>

<span class="n">chr_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">tibble</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chromosomes</span><span class="p">,</span><span class="w"> </span>
<span class="w">                  </span><span class="n">right</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span>

<span class="n">right_table_final</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rbind</span><span class="p">(</span><span class="n">right_table</span><span class="p">,</span><span class="w"> </span><span class="n">chr_start</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="map-each-tad-to-the-closest-left-right-boundary">
<h4>Map each TAD to the closest left_right boundary<a class="headerlink" href="#map-each-tad-to-the-closest-left-right-boundary" title="Link to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># In the list of left boundary, find the closet next one </span>

<span class="n">find_next_boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="p">){</span>
<span class="w">    </span><span class="n">left_boundaries</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">left_table_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">left</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">end_pos</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">pull</span><span class="p">(</span><span class="n">left</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">sort</span><span class="p">()</span>
<span class="w">    </span><span class="n">first_left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">left_boundaries</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">first_left</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># In the list of right boundary, find the closet previous one </span>
<span class="n">find_previous_boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">start_pos</span><span class="p">){</span>
<span class="w">    </span><span class="n">right_boundaries</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">right_table_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">right</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">start_pos</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">pull</span><span class="p">(</span><span class="n">right</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">sort</span><span class="p">(</span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">T</span><span class="p">)</span>
<span class="w">    </span><span class="n">first_right</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">right_boundaries</span><span class="p">[</span><span class="m">1</span><span class="p">]</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">first_right</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">TADB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">general</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">previous_bound</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmap</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">),</span><span class="w"> </span><span class="n">find_previous_boundary</span><span class="p">),</span>
<span class="w">    </span><span class="n">next_bound</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmap</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="p">),</span><span class="w"> </span><span class="n">find_next_boundary</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">end</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">rename</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">previous_bound</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">rename</span><span class="p">(</span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">next_bound</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">start</span><span class="p">),</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">end</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">select</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="p">)</span>

<span class="nf">nrow</span><span class="p">(</span><span class="n">TADB</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">1401</div></div>
</div>
<p>After extending to TADB, the number of regions did not change, which should be that case.
However, after merging, some of the TADBs will be totally covered by other TADBS. An example is shown here.</p>
<p>Genralized TAD0 and 4 not drawn. But for generalized TAD1, TADB fromed by it will be totally a subset of TADB 2, which is the extension of generalized TAD2.
(Blue: right boundary, red: left_boundary)</p>
<p><img alt="" src="../../_images/overlap.png" /></p>
<p>Here we will use two steps to exclude those TADBs that are fully covered by other single TADBs (including the case that two TADBs are identical)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">TADB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">distinct</span><span class="p">()</span>

<span class="c1"># if we don&#39;t use dintinct above, then the identical TADBS will all be removed</span>
<span class="c1"># that&#39;s not we want, we still want to leave one copy there</span>

<span class="n">if_fully_cover</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">start_pos</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="p">){</span>
<span class="w">    </span><span class="n">potential</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">chrm</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">start_pos</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">end_pos</span><span class="p">)</span>
<span class="w">    </span><span class="kr">if</span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">potential</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="m">1</span><span class="p">){</span>
<span class="w">        </span><span class="kr">return</span><span class="p">(</span><span class="m">0</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span><span class="kr">else</span><span class="p">{</span>
<span class="w">        </span><span class="kr">return</span><span class="p">(</span><span class="m">1</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span>
<span class="p">}</span>

<span class="n">TADB_final</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">cover_status</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmap</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">chrm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="p">),</span><span class="w"> </span><span class="n">if_fully_cover</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">filter</span><span class="p">(</span><span class="n">cover_status</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;TADB&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">row_number</span><span class="p">()))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="w"> </span><span class="n">cover_status</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">rename</span><span class="p">(</span><span class="n">`#chr`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chr</span><span class="p">)</span>

<span class="nf">nrow</span><span class="p">(</span><span class="n">TADB_final</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">1381</div></div>
</div>
<p>After removing subsets and duplicates, there will be 1381 TADBs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># save the TADB final result</span>
<span class="nf">write_tsv</span><span class="p">(</span><span class="n">TADB_final</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;../../reference_data/TAD/generalized_TADB.tsv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="length-distribution-of-generalized-tads">
<h3>Length distribution of generalized TADs<a class="headerlink" href="#length-distribution-of-generalized-tads" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## Length distribution of generalized TADs</span>

<span class="n">general</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">length</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2000000</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">)</span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">))</span><span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../../_images/96a56b9865abe6c3cba8ff63844e897055a335bfac49b04d67febfdda7117d38.png"><img alt="../../_images/96a56b9865abe6c3cba8ff63844e897055a335bfac49b04d67febfdda7117d38.png" src="../../_images/96a56b9865abe6c3cba8ff63844e897055a335bfac49b04d67febfdda7117d38.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
<p>The average length is a little bit shorter than 2M.</p>
</section>
<section id="length-distribution-of-general-tadbs">
<h3>Length distribution of general TADBs<a class="headerlink" href="#length-distribution-of-general-tadbs" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## Length distribution of general TADBs</span>

<span class="n">TADB_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">length</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2000000</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">)</span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">))</span><span class="w">  </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../../_images/8fe69faf164190edfc28587a97d14e6584e7c0c346f670440ff004692190bcf4.png"><img alt="../../_images/8fe69faf164190edfc28587a97d14e6584e7c0c346f670440ff004692190bcf4.png" src="../../_images/8fe69faf164190edfc28587a97d14e6584e7c0c346f670440ff004692190bcf4.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
<p>The average length is a little bit longer than 2M. Yet, the extreme case number increases.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># average length of generalized TAD</span>

<span class="n">general</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">length</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mean</span><span class="p">()</span>

<span class="c1"># average length of TADB</span>

<span class="n">TADB_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">length</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mean</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">1920021.81441827</div><div class="output text_html">2512992.33671253</div></div>
</div>
<p>TADB, as expected, is longer than TAD because it includes boundaries.</p>
</section>
<section id="compare-tadb-and-cis-windows">
<h3>Compare TADB and cis windows<a class="headerlink" href="#compare-tadb-and-cis-windows" title="Link to this heading">#</a></h3>
<p>Each TADB covers many genes. For each gene we will extend a 1M window for finemapping. We now checking if the TADBs can cover all 1M windows for genes inside of them.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">all_gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;../../reference_data/TAD/gene_start_end.tsv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>New names:
<span class=" -Color -Color-Cyan">•</span> `` -&gt; `...1`
<span class=" -Color -Color-Bold">Rows: </span><span class=" -Color -Color-Blue">60672</span> <span class=" -Color -Color-Bold">Columns: </span><span class=" -Color -Color-Blue">6</span>
<span class=" -Color -Color-Cyan">──</span> <span class=" -Color -Color-Bold">Column specification</span> <span class=" -Color -Color-Cyan">────────────────────────────────────────────────────────</span>
<span class=" -Color -Color-Bold">Delimiter:</span> &quot;\t&quot;
<span class=" -Color -Color-Red">chr</span> (3): #chr, gene_id, gene_name
<span class=" -Color -Color-Green">dbl</span> (3): ...1, start, end

<span class=" -Color -Color-Cyan">ℹ</span> Use `spec()` to retrieve the full column specification for this data.
<span class=" -Color -Color-Cyan">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">all_gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">all_gene</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">`...1`</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">gene_name</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">rename</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">`#chr`</span><span class="p">,</span><span class="w"> </span><span class="n">gene_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">start</span><span class="p">,</span>
<span class="w">                                                              </span><span class="n">gene_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="p">)</span><span class="w">  </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">distinct</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">nrow</span><span class="p">(</span><span class="n">all_gene</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">60668</div></div>
</div>
<p>First step, we try to verify if all genes are covered by TADB, if not, TADB may be wrong (since they should cover the whole genome)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># check if we have leave out any genes</span>

<span class="n">TADB_final</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TADB_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">rename</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#chr&quot;</span><span class="p">)</span>

<span class="c1"># find the corresponding TADB of one gene</span>

<span class="n">gene_TADB_tb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">left_join</span><span class="p">(</span><span class="n">all_gene</span><span class="p">,</span><span class="w"> </span><span class="n">TADB_final</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">relationship</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;many-to-many&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">filter</span><span class="p">(</span><span class="o">!</span><span class="p">(</span><span class="n">gene_start</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">end</span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w">  </span><span class="o">!</span><span class="p">(</span><span class="n">gene_end</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">start</span><span class="p">))</span>

<span class="n">genes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_TADB_tb</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>

<span class="n">all_genes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">all_gene</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>

<span class="n">dif</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">setdiff</span><span class="p">(</span><span class="n">all_genes</span><span class="p">,</span><span class="w"> </span><span class="n">genes</span><span class="p">)</span>

<span class="n">all_gene</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">gene_id</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">dif</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">chr</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'chrY'</li><li>'chrMT'</li><li>'ERCC_00002'</li><li>'ERCC_00003'</li><li>'ERCC_00004'</li><li>'ERCC_00009'</li><li>'ERCC_00012'</li><li>'ERCC_00013'</li><li>'ERCC_00014'</li><li>'ERCC_00016'</li><li>'ERCC_00017'</li><li>'ERCC_00019'</li><li>'ERCC_00022'</li><li>'ERCC_00024'</li><li>'ERCC_00025'</li><li>'ERCC_00028'</li><li>'ERCC_00031'</li><li>'ERCC_00033'</li><li>'ERCC_00034'</li><li>'ERCC_00035'</li><li>'ERCC_00039'</li><li>'ERCC_00040'</li><li>'ERCC_00041'</li><li>'ERCC_00042'</li><li>'ERCC_00043'</li><li>'ERCC_00044'</li><li>'ERCC_00046'</li><li>'ERCC_00048'</li><li>'ERCC_00051'</li><li>'ERCC_00053'</li><li>'ERCC_00054'</li><li>'ERCC_00057'</li><li>'ERCC_00058'</li><li>'ERCC_00059'</li><li>'ERCC_00060'</li><li>'ERCC_00061'</li><li>'ERCC_00062'</li><li>'ERCC_00067'</li><li>'ERCC_00069'</li><li>'ERCC_00071'</li><li>'ERCC_00073'</li><li>'ERCC_00074'</li><li>'ERCC_00075'</li><li>'ERCC_00076'</li><li>'ERCC_00077'</li><li>'ERCC_00078'</li><li>'ERCC_00079'</li><li>'ERCC_00081'</li><li>'ERCC_00083'</li><li>'ERCC_00084'</li><li>'ERCC_00085'</li><li>'ERCC_00086'</li><li>'ERCC_00092'</li><li>'ERCC_00095'</li><li>'ERCC_00096'</li><li>'ERCC_00097'</li><li>'ERCC_00098'</li><li>'ERCC_00099'</li><li>'ERCC_00104'</li><li>'ERCC_00108'</li><li>'ERCC_00109'</li><li>'ERCC_00111'</li><li>'ERCC_00112'</li><li>'ERCC_00113'</li><li>'ERCC_00116'</li><li>'ERCC_00117'</li><li>'ERCC_00120'</li><li>'ERCC_00123'</li><li>'ERCC_00126'</li><li>'ERCC_00130'</li><li>'ERCC_00131'</li><li>'ERCC_00134'</li><li>'ERCC_00136'</li><li>'ERCC_00137'</li><li>'ERCC_00138'</li><li>'ERCC_00142'</li><li>'ERCC_00143'</li><li>'ERCC_00144'</li><li>'ERCC_00145'</li><li>'ERCC_00147'</li><li>'ERCC_00148'</li><li>'ERCC_00150'</li><li>'ERCC_00154'</li><li>'ERCC_00156'</li><li>'ERCC_00157'</li><li>'ERCC_00158'</li><li>'ERCC_00160'</li><li>'ERCC_00162'</li><li>'ERCC_00163'</li><li>'ERCC_00164'</li><li>'ERCC_00165'</li><li>'ERCC_00168'</li><li>'ERCC_00170'</li><li>'ERCC_00171'</li></ol>
</div></div>
</div>
<p>So diff is the gene that is in the list of all genes, but not covered by TADB, I checked it again and find that they are all on chromosome Y, MT or ERCC genes. So that’s normal because we don’t care these genes, and our TAD data do not include it, so we will never cover these genes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># add the end of chromosome info so that we won&#39;t have cis window outside of chromosome</span>

<span class="n">cis_TADB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">left_join</span><span class="p">(</span><span class="n">gene_TADB_tb</span><span class="p">,</span><span class="w"> </span><span class="n">chrDF</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;chr&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">rename</span><span class="p">(</span><span class="n">chr_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">left</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">cis_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmax</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">gene_start</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">1000000</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">cis_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmin</span><span class="p">(</span><span class="n">chr_end</span><span class="p">,</span><span class="w"> </span><span class="n">gene_end</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1000000</span><span class="p">))</span>
<span class="w">    </span>
<span class="nf">head</span><span class="p">(</span><span class="n">cis_TADB</span><span class="p">)</span>

<span class="c1"># label overlap status 0/1</span>

<span class="n">cis_TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">cis_covered_status</span><span class="w"> </span><span class="o">=</span><span class="w"> </span>
<span class="w">                    </span><span class="nf">if_else</span><span class="p">((</span><span class="n">cis_start</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">&amp;</span>
<span class="w">                    </span><span class="n">cis_end</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end</span><span class="p">),</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">group_by</span><span class="p">(</span><span class="n">cis_covered_status</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">            </span><span class="nf">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 6 × 10</caption>
<thead>
	<tr><th scope=col>chr</th><th scope=col>gene_start</th><th scope=col>gene_end</th><th scope=col>gene_id</th><th scope=col>start</th><th scope=col>end</th><th scope=col>index</th><th scope=col>chr_end</th><th scope=col>cis_start</th><th scope=col>cis_end</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1</td><td>11868</td><td>14409</td><td>ENSG00000223972</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1014409</td></tr>
	<tr><td>chr1</td><td>14404</td><td>29570</td><td>ENSG00000227232</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1029570</td></tr>
	<tr><td>chr1</td><td>17369</td><td>17436</td><td>ENSG00000278267</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1017436</td></tr>
	<tr><td>chr1</td><td>29553</td><td>31109</td><td>ENSG00000243485</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1031109</td></tr>
	<tr><td>chr1</td><td>30365</td><td>30503</td><td>ENSG00000284332</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1030503</td></tr>
	<tr><td>chr1</td><td>34554</td><td>36081</td><td>ENSG00000237613</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1036081</td></tr>
</tbody>
</table>
</div><div class="output text_html"><table class="dataframe">
<caption>A grouped_df: 2 × 2</caption>
<thead>
	<tr><th scope=col>cis_covered_status</th><th scope=col>n</th></tr>
	<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
	<tr><td>0</td><td>51336</td></tr>
	<tr><td>1</td><td>20754</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>0 stands for cis window not fully covered by the TADB the gene lies in. So after we have TADB, we still
have many genes whose 1M cis window (total length: 2M) cannot be covered by its TADB region.</p>
</section>
<section id="construct-tadb-enhanced-cis-window">
<h3>Construct TADB-enhanced cis window<a class="headerlink" href="#construct-tadb-enhanced-cis-window" title="Link to this heading">#</a></h3>
<p>For each gene we get its 1M window (upstream and downstream 1M respectively, total length 2M), and see which TADB it’s in.  Then we extend the region to the smaller start and larger end.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cis_TADB</span><span class="w">  </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 6 × 10</caption>
<thead>
	<tr><th scope=col>chr</th><th scope=col>gene_start</th><th scope=col>gene_end</th><th scope=col>gene_id</th><th scope=col>start</th><th scope=col>end</th><th scope=col>index</th><th scope=col>chr_end</th><th scope=col>cis_start</th><th scope=col>cis_end</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1</td><td>11868</td><td>14409</td><td>ENSG00000223972</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1014409</td></tr>
	<tr><td>chr1</td><td>14404</td><td>29570</td><td>ENSG00000227232</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1029570</td></tr>
	<tr><td>chr1</td><td>17369</td><td>17436</td><td>ENSG00000278267</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1017436</td></tr>
	<tr><td>chr1</td><td>29553</td><td>31109</td><td>ENSG00000243485</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1031109</td></tr>
	<tr><td>chr1</td><td>30365</td><td>30503</td><td>ENSG00000284332</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1030503</td></tr>
	<tr><td>chr1</td><td>34554</td><td>36081</td><td>ENSG00000237613</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1036081</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">extended</span><span class="w"> </span><span class="o">=</span><span class="w">  </span><span class="n">cis_TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">true_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmin</span><span class="p">(</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">cis_start</span><span class="p">),</span>
<span class="w">          </span><span class="n">true_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmax</span><span class="p">(</span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="n">cis_end</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">arrange</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">true_start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">ungroup</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">select</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">gene_id</span><span class="p">,</span><span class="w"> </span><span class="n">true_start</span><span class="p">,</span><span class="w"> </span><span class="n">true_end</span><span class="p">)</span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">rename</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">true_start</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">true_end</span><span class="p">)</span>

<span class="n">extended_final</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">extended</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">group_by</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">gene_id</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">summarize</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">start</span><span class="p">),</span>
<span class="w">                                            </span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">end</span><span class="p">))</span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">arrange</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">)</span>

<span class="nf">nrow</span><span class="p">(</span><span class="n">extended_final</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`summarise()` has grouped output by &#39;chr&#39;. You can override using the `.groups`
argument.
</pre></div>
</div>
<div class="output text_html">60017</div></div>
</div>
<p>Now we have the same number of extended cis windows as the gene numbers.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## Distribution of extended regions</span>

<span class="n">extended_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">length</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2000000</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">)</span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">))</span><span class="w">  </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../../_images/d942ba901884e5184cf1906eb760ede687cbd5ed6eb7dbc2d3fe4878c4ae1a14.png"><img alt="../../_images/d942ba901884e5184cf1906eb760ede687cbd5ed6eb7dbc2d3fe4878c4ae1a14.png" src="../../_images/d942ba901884e5184cf1906eb760ede687cbd5ed6eb7dbc2d3fe4878c4ae1a14.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">extended_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">length</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">max</span><span class="p">()</span><span class="w"> </span>
<span class="n">extended_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">length</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mean</span><span class="p">()</span><span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">35080000</div><div class="output text_html">4653395.44833964</div></div>
</div>
<p>After extending, most of the extended regions have a length longer than 2M. There are still some regions with length shorter than 2M. For example if one gene is at the start of one chromosome, then extend to the left, the original cis window will shorter than 2M, so when merged with TADB this region can still be shorter than 2M.</p>
<p>We should no longer find again what genes are in each extended-TADB. Otherwise for these genes their 1M window may again not covered, then we fall into the loop until we indlude the whole genome.</p>
<p>We will only finemap those genes originally in not extended TADBs, using extended TADB as fine mapping region.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ordered_chr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">)),</span><span class="w"> </span><span class="s">&quot;chrX&quot;</span><span class="p">)</span>

<span class="n">extend_cis_ordered</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">extended_final</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ordered_chr</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">arrange</span><span class="p">(</span><span class="n">chr</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">rename</span><span class="p">(</span><span class="n">`#chr`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chr</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">select</span><span class="p">(</span><span class="n">`#chr`</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="n">gene_id</span><span class="p">)</span>

<span class="n">extend_cis_ordered</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">write_tsv</span><span class="p">(</span><span class="s">&quot;../../reference_data/TAD/TADB_enhanced_cis.bed&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="extended-tadb">
<h3>Extended TADB<a class="headerlink" href="#extended-tadb" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># add the end of chromosome info so that we won&#39;t have cis window outside of chromosome</span>

<span class="n">cis_TADB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">left_join</span><span class="p">(</span><span class="n">gene_TADB_tb</span><span class="p">,</span><span class="w"> </span><span class="n">chrDF</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;chr&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">rename</span><span class="p">(</span><span class="n">chr_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">left</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">cis_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmax</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">gene_start</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">1000000</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">cis_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmin</span><span class="p">(</span><span class="n">chr_end</span><span class="p">,</span><span class="w"> </span><span class="n">gene_end</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1000000</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">idx</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_TADB_tb</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">index</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">unique</span><span class="p">()</span>
<span class="n">idx2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TADB_final</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">index</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cis_TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 6 × 10</caption>
<thead>
	<tr><th scope=col>chr</th><th scope=col>gene_start</th><th scope=col>gene_end</th><th scope=col>gene_id</th><th scope=col>start</th><th scope=col>end</th><th scope=col>index</th><th scope=col>chr_end</th><th scope=col>cis_start</th><th scope=col>cis_end</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1</td><td>11868</td><td>14409</td><td>ENSG00000223972</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1014409</td></tr>
	<tr><td>chr1</td><td>14404</td><td>29570</td><td>ENSG00000227232</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1029570</td></tr>
	<tr><td>chr1</td><td>17369</td><td>17436</td><td>ENSG00000278267</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1017436</td></tr>
	<tr><td>chr1</td><td>29553</td><td>31109</td><td>ENSG00000243485</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1031109</td></tr>
	<tr><td>chr1</td><td>30365</td><td>30503</td><td>ENSG00000284332</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1030503</td></tr>
	<tr><td>chr1</td><td>34554</td><td>36081</td><td>ENSG00000237613</td><td>0</td><td>6480000</td><td>TADB1</td><td>248956422</td><td>0</td><td>1036081</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cis_summary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cis_TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">group_by</span><span class="p">(</span><span class="n">index</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">summarize</span><span class="p">(</span><span class="n">min_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">cis_start</span><span class="p">),</span><span class="w"> </span><span class="n">max_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">cis_end</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">extend_TADB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">left_join</span><span class="p">(</span><span class="n">cis_summary</span><span class="p">,</span><span class="w"> </span><span class="n">TADB_final</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;index&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">new_start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmin</span><span class="p">(</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">min_start</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">new_end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmax</span><span class="p">(</span><span class="n">end</span><span class="p">,</span><span class="w"> </span><span class="n">max_end</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">new_end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">new_start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">old_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length_diff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">length</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">old_length</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">select</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">new_start</span><span class="p">,</span><span class="w"> </span><span class="n">new_end</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">rename</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">new_start</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">new_end</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;TADB&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">row_number</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
<p>For the extended TADB, one gene can be in multiple extended TADBs, even more than 2, as expected.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">left_join</span><span class="p">(</span><span class="n">extend_TADB</span><span class="p">,</span><span class="w"> </span><span class="n">all_gene</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;chr&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">gene_start</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">gene_end</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">end</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">group_by</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">count</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">group_by</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message in left_join(extend_TADB, all_gene, by = &quot;chr&quot;):
“Detected an unexpected many-to-many relationship between `x` and `y`.
<span class=" -Color -Color-Cyan">ℹ</span> Row 1 of `x` matches multiple rows in `y`.
<span class=" -Color -Color-Cyan">ℹ</span> Row 1 of `y` matches multiple rows in `x`.
<span class=" -Color -Color-Cyan">ℹ</span> If a many-to-many relationship is expected, set `relationship =
  &quot;many-to-many&quot;` to silence this warning.”
Storing counts in `nn`, as `n` already present in input
<span class=" -Color -Color-Cyan">ℹ</span> Use `name = &quot;new_name&quot;` to pick a new name.
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A grouped_df: 5 × 2</caption>
<thead>
	<tr><th scope=col>n</th><th scope=col>nn</th></tr>
	<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>
</thead>
<tbody>
	<tr><td>1</td><td>12330</td></tr>
	<tr><td>2</td><td>32049</td></tr>
	<tr><td>3</td><td>13938</td></tr>
	<tr><td>4</td><td> 1611</td></tr>
	<tr><td>5</td><td>   89</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">nrow</span><span class="p">(</span><span class="n">extend_TADB</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">1381</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># extend_TADB length distribution</span>
<span class="n">extend_TADB</span><span class="w">  </span><span class="o">%&gt;%</span><span class="w"> </span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">length</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2000000</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">)</span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">))</span><span class="w">  </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../../_images/0f880c90692dc012e7a39c4bb65d5e44e0bf4a31f9fdd7fa0b05ff3a8ea6d32c.png"><img alt="../../_images/0f880c90692dc012e7a39c4bb65d5e44e0bf4a31f9fdd7fa0b05ff3a8ea6d32c.png" src="../../_images/0f880c90692dc012e7a39c4bb65d5e44e0bf4a31f9fdd7fa0b05ff3a8ea6d32c.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ordered_chr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">)),</span><span class="w"> </span><span class="s">&quot;chrX&quot;</span><span class="p">)</span>

<span class="n">extend_TADB</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">extend_TADB</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ordered_chr</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">arrange</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">rename</span><span class="p">(</span><span class="n">`#chr`</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">chr</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">index</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;TADB_&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">row_number</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">extend_TADB</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">write_tsv</span><span class="p">(</span><span class="s">&quot;../../reference_data/TAD/extended_TADB.bed&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Compared with TADB-enhanced cis window, extended TADB is longer in length.</p>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./code/reference_data"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="reference_data_preparation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Reference Data Standardization</p>
      </div>
    </a>
    <a class="right-next"
       href="ld_prune_reference.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Independent list of variants using LD clumping</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-manage-tad-redundancy">i. Manage TAD Redundancy</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-tads-with-neighboring-tabs">Merge TADs With Neighboring TABs</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#preliminary-merges">Preliminary Merges</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#determining-overlap">Determining Overlap</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#overlap-check">Overlap Check</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#recursively-merge-tads">Recursively Merge TADs</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-generate-tadb-enhanced-cis-windows-and-extended-tadbs">ii. Generate TADB-enhanced cis windows and extended TADBs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-find-boundaries-and-build-tadb">How to find boundaries and build TADB?</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tadb-coding">TADB - coding</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#find-left-boundaries">Find left boundaries</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#find-right-boundaries">Find right boundaries</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#map-each-tad-to-the-closest-left-right-boundary">Map each TAD to the closest left_right boundary</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#length-distribution-of-generalized-tads">Length distribution of generalized TADs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#length-distribution-of-general-tadbs">Length distribution of general TADBs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compare-tadb-and-cis-windows">Compare TADB and cis windows</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#construct-tadb-enhanced-cis-window">Construct TADB-enhanced cis window</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extended-tadb">Extended TADB</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>