
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Independent list of variants using LD clumping &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="RNA-seq expression" href="../molecular_phenotypes/bulk_expression.html" />
    <link rel="prev" title="Generation of Topologically Associated Domains and their Boundaries" href="generalized_TADB.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="reference_data.html">
   Reference Data
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="reference_data_preparation.html">
     Reference Data Standardization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generalized_TADB.html">
     Generation of Topologically Associated Domains and their Boundaries
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Independent list of variants using LD clumping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene Coordinate Annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype Data Imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden Factor Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate Data Formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pecotmr_integration/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/twas_mr.html">
     GWAS integration: TWAS and MR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Extract genome-wide data for multivariate analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/reference_data/ld_prune_reference.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/reference_data/ld_prune_reference.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/reference_data/ld_prune_reference.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#description">
   Description
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example-steps">
   Minimal Working Example Steps
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#i-extract-genotypes-per-ld-block">
     i. Extract genotypes per LD block
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ii-ld-pruning-within-each-ld-block">
     ii. LD pruning within each LD block
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="independent-list-of-variants-using-ld-clumping">
<h1>Independent list of variants using LD clumping<a class="headerlink" href="#independent-list-of-variants-using-ld-clumping" title="Permalink to this headline">¶</a></h1>
<div class="section" id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h2>
<p>This notebook creates a list of randomly selected variants that are in weak LD with each other, for given reference data-set. This is useful for the downstreams <code class="docutils literal notranslate"><span class="pre">mashr</span></code> analysis where a list of independent variants are used.</p>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Reference genotype, in this case, our ROSMAP European ancestry genotype of ~1,000 individuals, <code class="docutils literal notranslate"><span class="pre">ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.bed</span></code></p></li>
<li><p>A list of approximately independent LD blocks, which is optional but when it is included, we can parallel the LD clumping to all LD blocks which will speed up the process. This list is available from:  <a class="reference external" href="https://github.com/gaow/LDblocks_GRCh38/blob/master/data/deCODE_EUR_LD_blocks.bed">https://github.com/gaow/LDblocks_GRCh38/blob/master/data/deCODE_EUR_LD_blocks.bed</a> (a forked repo created from work described at <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2022.03.04.483057v2">https://www.biorxiv.org/content/10.1101/2022.03.04.483057v2</a>)</p></li>
</ul>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>A list of independent variants in <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> format</p></li>
</ul>
</div>
<div class="section" id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Permalink to this headline">¶</a></h2>
<div class="section" id="i-extract-genotypes-per-ld-block">
<h3>i. Extract genotypes per LD block<a class="headerlink" href="#i-extract-genotypes-per-ld-block" title="Permalink to this headline">¶</a></h3>
<p>Using the <code class="docutils literal notranslate"><span class="pre">genotype_formatting.ipynb</span></code> pipeline,</p>
<p><strong>FIXME: this is using the original bed file for blocks, which is a lot cleaner. Please test this</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run pipeline/genotype_formatting.ipynb genotype_by_region <span class="se">\</span>
    --region-list data/references/deCODE_EUR_LD_blocks.bed <span class="se">\</span>
    --genoFile data/genotype/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.bed <span class="se">\</span>
    --cwd output/geno_by_LDblock <span class="se">\</span>
    --container /mnt/vast/hpc/csg/containers_xqtl/bioinfo.sif <span class="se">\</span>
    --mem 200G -s build
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">genotype_by_region_1</span>: 
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="ii-ld-pruning-within-each-ld-block">
<h3>ii. LD pruning within each LD block<a class="headerlink" href="#ii-ld-pruning-within-each-ld-block" title="Permalink to this headline">¶</a></h3>
<p>We create a small pipeline for the purpose,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>global<span class="o">]</span>
parameter: <span class="nv">cwd</span> <span class="o">=</span> path<span class="o">(</span><span class="s2">&quot;output&quot;</span><span class="o">)</span>
parameter: <span class="nv">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
import re
parameter: <span class="nv">entrypoint</span><span class="o">=</span> <span class="o">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> + <span class="s1">&#39; &#39;</span> + re.sub<span class="o">(</span>r<span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span>, <span class="s1">&#39;&#39;</span>, container.split<span class="o">(</span><span class="s1">&#39;/&#39;</span><span class="o">)[</span>-1<span class="o">]))</span> <span class="k">if</span> container <span class="k">else</span> <span class="s2">&quot;&quot;</span>
parameter: <span class="nv">job_size</span> <span class="o">=</span> <span class="m">1</span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s2">&quot;1h&quot;</span>
parameter: <span class="nv">mem</span> <span class="o">=</span> <span class="s2">&quot;80G&quot;</span>
parameter: <span class="nv">numThreads</span> <span class="o">=</span> <span class="m">10</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>LD_pruning_1<span class="o">]</span>
parameter: <span class="nv">genotype_list</span> <span class="o">=</span> path
parameter: <span class="nv">window</span> <span class="o">=</span> <span class="m">50</span>
parameter: <span class="nb">shift</span> <span class="o">=</span> <span class="m">10</span>
parameter: <span class="nv">r2</span> <span class="o">=</span> 1e-3

import pandas as pd
<span class="nv">geno_path</span> <span class="o">=</span> pd.read_csv<span class="o">(</span>genotype, <span class="nv">sep</span> <span class="o">=</span> <span class="s2">&quot;\t&quot;</span><span class="o">)</span>
<span class="nv">input_df</span> <span class="o">=</span> geno_path.values.tolist<span class="o">()</span>
<span class="nv">input_blocks</span> <span class="o">=</span> <span class="o">[</span>x<span class="o">[</span><span class="m">0</span><span class="o">]</span> <span class="k">for</span> x <span class="k">in</span> input_df<span class="o">]</span>
<span class="nv">input_files</span> <span class="o">=</span> <span class="o">[</span>x<span class="o">[</span><span class="m">1</span>:<span class="o">]</span> <span class="k">for</span> x <span class="k">in</span> input_df<span class="o">]</span> 

input: input_files, <span class="nv">group_by</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">group_with</span> <span class="o">=</span> <span class="s2">&quot;input_blocks&quot;</span>
output: <span class="nv">prune</span> <span class="o">=</span> f<span class="s1">&#39;{cwd}/{step_name[:-2]}/{_input_blocks}.prune.in&#39;</span>
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span> <span class="o">=</span> job_size, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads, <span class="nv">tags</span> <span class="o">=</span> f<span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
bash: <span class="nv">container</span><span class="o">=</span>container, <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stdout&#39;</span>, <span class="nv">entrypoint</span><span class="o">=</span>entrypoint
    plink2 <span class="se">\</span>
        --bfile  <span class="si">${</span><span class="nv">_input</span><span class="p">:</span><span class="nv">anr</span><span class="si">}</span> <span class="se">\</span>
        --indep-pairwise  <span class="si">${</span><span class="nv">window</span><span class="si">}</span> <span class="si">${</span><span class="nv">shift</span><span class="si">}</span> <span class="si">${</span><span class="nv">r2</span><span class="si">}</span> <span class="se">\</span>
        --rm-dup force-first <span class="se">\</span>
        --out <span class="si">${</span><span class="nv">_output</span><span class="p">[</span><span class="s2">&quot;prune&quot;</span><span class="p">]:</span><span class="nv">annr</span><span class="si">}</span> <span class="se">\</span>
        --threads <span class="si">${</span><span class="nv">numThreads</span><span class="si">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>LD_pruning_2<span class="o">]</span>
input: <span class="nv">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
output: f<span class="s1">&#39;{cwd}/{step_name[:-2]}/LD_pruned_variants.txt&#39;</span>
R: <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stdout&#39;</span>, <span class="nv">container</span> <span class="o">=</span> container, <span class="nv">entrypoint</span> <span class="o">=</span> entrypoint
    library<span class="o">(</span>data.table<span class="o">)</span>
    library<span class="o">(</span>dplyr<span class="o">)</span>
    library<span class="o">(</span>stringr<span class="o">)</span>
    <span class="nv">per_block_variants</span> <span class="o">=</span> c<span class="o">(</span><span class="si">${</span><span class="nv">_input</span><span class="p">:</span><span class="nv">ar</span><span class="p">,</span><span class="si">}</span><span class="o">)</span>
    <span class="nv">uncorrelated_variants</span> <span class="o">=</span> as.data.frame<span class="o">(</span>rbindlist<span class="o">(</span>lapply<span class="o">(</span>per_block_variants, fread, <span class="nv">header</span> <span class="o">=</span> FALSE<span class="o">)))</span>
    <span class="nv">random_variants</span>  <span class="o">=</span> str_split<span class="o">(</span>gsub<span class="o">(</span><span class="s2">&quot;_&quot;</span>,<span class="s2">&quot;:&quot;</span>,uncorrelated_variants<span class="nv">$V1</span><span class="o">)</span>,<span class="s2">&quot;:&quot;</span>,simplify<span class="o">=</span>TRUE<span class="o">)</span>%&gt;%data.frame<span class="o">()</span>%&gt;%
                    rename<span class="o">(</span><span class="s2">&quot;chrom&quot;</span> <span class="o">=</span> <span class="s2">&quot;X1&quot;</span>,<span class="s2">&quot;pos&quot;</span> <span class="o">=</span> <span class="s2">&quot;X2&quot;</span>,<span class="s2">&quot;alt&quot;</span> <span class="o">=</span> <span class="s2">&quot;X3&quot;</span>,<span class="s2">&quot;ref&quot;</span><span class="o">=</span><span class="s2">&quot;X4&quot;</span><span class="o">)</span>%&gt;%
                    mutate<span class="o">(</span><span class="nv">variant_id</span> <span class="o">=</span> paste<span class="o">(</span>chrom, pos, alt, ref, <span class="nv">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="o">))</span>
    fwrite<span class="o">(</span>random_variants, <span class="si">${</span><span class="nv">_output</span><span class="p">:</span><span class="nv">r</span><span class="si">}</span>, row.names<span class="o">=</span>F,sep<span class="o">=</span><span class="s2">&quot;\t&quot;</span>,quote<span class="o">=</span>F<span class="o">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run ld_prune_reference.ipynb LD_pruning <span class="se">\</span>
   --cwd output/ROSMAP_LD_pruned <span class="se">\</span>
</pre></div>
</div>
</div>
</div>
<p><strong>FIXME: please note that I uused LD pruning step 2 to merge them. Please test it. Also you did not show how to actually run this. Please add it above</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">random_variants</span> <span class="o">=</span> data.table::fread<span class="o">(</span><span class="s2">&quot;LD_pruned_variants.txt&quot;</span><span class="o">)</span>
head<span class="o">(</span>random_variants<span class="o">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 5</caption>
<thead>
	<tr><th></th><th scope=col>chrom</th><th scope=col>pos</th><th scope=col>alt</th><th scope=col>ref</th><th scope=col>variant_id</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>chr1</td><td>16206</td><td>T</td><td>A</td><td>chr1:16206:T:A</td></tr>
	<tr><th scope=row>2</th><td>chr1</td><td>16433</td><td>C</td><td>G</td><td>chr1:16433:C:G</td></tr>
	<tr><th scope=row>3</th><td>chr1</td><td>16471</td><td>C</td><td>T</td><td>chr1:16471:C:T</td></tr>
	<tr><th scope=row>4</th><td>chr1</td><td>16542</td><td>C</td><td>A</td><td>chr1:16542:C:A</td></tr>
	<tr><th scope=row>5</th><td>chr1</td><td>16570</td><td>C</td><td>T</td><td>chr1:16570:C:T</td></tr>
	<tr><th scope=row>6</th><td>chr1</td><td>16635</td><td>G</td><td>A</td><td>chr1:16635:G:A</td></tr>
</tbody>
</table>
</div></div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "bash"
        },
        kernelOptions: {
            kernelName: "bash",
            path: "./code/reference_data"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'bash'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="generalized_TADB.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Generation of Topologically Associated Domains and their Boundaries</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../molecular_phenotypes/bulk_expression.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">RNA-seq expression</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>