
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Reference Data Standardization &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/reference_data/reference_data_preparation';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Generation of Topologically Associated Domains and their Boundaries" href="generalized_TADB.html" />
    <link rel="prev" title="Reference Data" href="reference_data.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="reference_data.html">Reference Data</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/reference_data/reference_data_preparation.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/reference_data/reference_data_preparation.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/reference_data/reference_data_preparation.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Reference Data Standardization</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-and-output">Input and Output</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#download-reference-data">Download Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gff3-to-gtf">GFF3 To GTF</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#format-reference-data">Format Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#format-gene-feature-data">Format Gene Feature Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-star-index">Generate STAR Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-rsem-index">Generate RSEM Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-refflat-annotation-for-picard">Generate RefFlat Annotation for Picard</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-suppa-annotation-for-psichomics">Generate SUPPA Annotation for Psichomics</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-rsids-for-known-variants">Extract rsIDs for known variants</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tad-annotation">TAD Annotation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-download-reference-data">i. Download Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-format-reference-data">ii. Format Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-format-gene-feature-data">iii. Format Gene Feature Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iv-generate-star-index">iv. Generate STAR Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#v-generate-rsem-index">v. Generate RSEM Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-generate-refflat-annotation-for-picard">vi. Generate RefFlat Annotation for Picard</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vii-generate-suppa-annotation-for-psichomics">vii. Generate SUPPA Annotation for Psichomics</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#viii-extract-rsids-for-known-variants">viii. Extract rsIDs for known variants</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-download">Data download</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gff3-to-gtf-formatting">GFF3 to GTF formatting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#hg-reference-file-preprocessing">HG reference file preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#transcript-and-gene-model-reference-processing">Transcript and gene model reference processing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-star">Generating index file for <code class="docutils literal notranslate"><span class="pre">STAR</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-output">Step Output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-rsem">Generating index file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-refflat-file">Generation of RefFlat file</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-suppa-annotation-for-psichomics">Generation of SUPPA annotation for psichomics.</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-psichomics-hg38-splicing-annotation">Generate psichomics Hg38 splicing annotation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-tad-associated-index">Generate TAD associated index</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="reference-data-standardization">
<h1>Reference Data Standardization<a class="headerlink" href="#reference-data-standardization" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>We follow the <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md">TOPMed workflow from Broad</a> for the aquisition and preparation of reference files for use throughout our pipeline. Our processed reference data may be found <a class="reference external" href="https://www.synapse.org/#!Synapse:syn36416587">in this folder on Synapse</a>. We have included the PDF document compiled by Data Standardization Working Group in the <a class="reference external" href="https://www.synapse.org/#!Synapse:syn36416587">on Synapse</a> as well as on <a class="reference external" href="https://www.niagads.org/adsp/content/adspgcadgenomeresources-v2pdf">ADSP Dashboard</a>. It contains the reference data to use for the project.</p>
</section>
<section id="input-and-output">
<h2>Input and Output<a class="headerlink" href="#input-and-output" title="Link to this heading">#</a></h2>
<section id="download-reference-data">
<h3>Download Reference Data<a class="headerlink" href="#download-reference-data" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: None<br />
<code class="docutils literal notranslate"><span class="pre">Output</span></code>:</p>
<ul class="simple">
<li><p>00-All.vcf.gz</p></li>
<li><p>00-All.vcf.gz.tbi</p></li>
<li><p>ERCC92.fa</p></li>
<li><p>ERCC92.gtf</p></li>
<li><p>ERCC92.zip</p></li>
<li><p>GRCh38_full_analysis_set_plus_decoy_hla.fa</p></li>
<li><p>Homo_sapiens.GRCh38.103.chr.gtf</p></li>
<li><p>Homo_sapiens.GRCh38.103.chr.gtf.gz</p></li>
</ul>
</section>
<section id="gff3-to-gtf">
<h3>GFF3 To GTF<a class="headerlink" href="#gff3-to-gtf" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: an uncompressed gff3 file.(i.e. can be view via cat)</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: a gtf file</p>
</section>
<section id="format-reference-data">
<h3>Format Reference Data<a class="headerlink" href="#format-reference-data" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>:  a reference fasta file.</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: a reference fasta file without HLA, ALT, Decoy but with ERCC:</p>
<ul class="simple">
<li><p>ERCC92.patched.fa</p></li>
<li><p>GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy.fasta</p></li>
<li><p>GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.dict</p></li>
<li><p>GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta</p></li>
<li><p>GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta.fai</p></li>
<li><p>Homo_sapiens.GRCh38.103.chr.reformatted.gtf</p></li>
</ul>
</section>
<section id="format-gene-feature-data">
<h3>Format Gene Feature Data<a class="headerlink" href="#format-gene-feature-data" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: a gtf file.</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>:  a gtf file with collapsed gene model.</p>
</section>
<section id="generate-star-index">
<h3>Generate STAR Index<a class="headerlink" href="#generate-star-index" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: a gtf file and an acompanying fasta file.</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: A folder containing STAR index files.</p>
</section>
<section id="generate-rsem-index">
<h3>Generate RSEM Index<a class="headerlink" href="#generate-rsem-index" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: a gtf file and an acompanying fasta file.</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: A folder containing RSEM index files.</p>
</section>
<section id="generate-refflat-annotation-for-picard">
<h3>Generate RefFlat Annotation for Picard<a class="headerlink" href="#generate-refflat-annotation-for-picard" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: a gtf file</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: a ref.flat file</p>
</section>
<section id="generate-suppa-annotation-for-psichomics">
<h3>Generate SUPPA Annotation for Psichomics<a class="headerlink" href="#generate-suppa-annotation-for-psichomics" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: a gtf file</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: a .SUPPA_annotation.rds file</p>
</section>
<section id="extract-rsids-for-known-variants">
<h3>Extract rsIDs for known variants<a class="headerlink" href="#extract-rsids-for-known-variants" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: a .vcf.gz file</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: a .variants.gz file</p>
</section>
<section id="tad-annotation">
<h3>TAD Annotation<a class="headerlink" href="#tad-annotation" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Input</span></code>: a file containing a list of TADs and a region list</p>
<p><code class="docutils literal notranslate"><span class="pre">Output</span></code>: a .annotation file</p>
</section>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<section id="i-download-reference-data">
<h3>i. Download Reference Data<a class="headerlink" href="#i-download-reference-data" title="Link to this heading">#</a></h3>
<p>Timing: ~2 hours</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb download_hg_reference --cwd ../reference_data
!sos run reference_data_preparation.ipynb download_gene_annotation --cwd ../reference_data
!sos run reference_data_preparation.ipynb download_ercc_reference --cwd ../reference_data
!sos run reference_data_preparation.ipynb download_dbsnp --cwd ../reference_data
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">download_hg_reference</span>: 
GRCh38_ful...lus_decoy_hla.fa:<span class=" -Color -Color-Green"> downloaded                                       </span>: 
INFO: <span class=" -Color -Color-Green">download_hg_reference</span> is <span class=" -Color -Color-Green">completed</span>.
INFO: <span class=" -Color -Color-Green">download_hg_reference</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.fa</span>
INFO: Workflow download_hg_reference (ID=w5027ac774a9eadc1) is executed successfully with 1 completed step.
INFO: Running <span class=" -Color -Color-Green">download_gene_annotation</span>: 
Homo_sapie...8.103.chr.gtf.gz:   0%|               | 0/49087092 [00:00&lt;?, ?it/s]
INFO: <span class=" -Color -Color-Green">download_gene_annotation</span> is <span class=" -Color -Color-Green">completed</span>.                   [00:07&lt;0
INFO: <span class=" -Color -Color-Green">download_gene_annotation</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.gtf</span>
INFO: Workflow download_gene_annotation (ID=w0bded47053a09483) is executed successfully with 1 completed step.
INFO: Running <span class=" -Color -Color-Green">download_ercc_reference</span>: 
ERCC92.zip:   0%|                                     | 0/28717 [00:00&lt;?, ?it/s]
INFO: <span class=" -Color -Color-Green">download_ercc_reference</span> is <span class=" -Color -Color-Green">completed</span>.                   5.95it/s]
INFO: <span class=" -Color -Color-Green">download_ercc_reference</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/ERCC92.gtf /restricted/projectnb/xqtl/xqtl_protocol/reference_data/ERCC92.fa</span>
INFO: Workflow download_ercc_reference (ID=w1ed1866b727d8c9e) is executed successfully with 1 completed step.
INFO: Running <span class=" -Color -Color-Green">download_dbsnp</span>: 
00-All.vcf.gz
00-All.vcf.gz.tbi
00-All.vcf.gz.tbi
00-All.vcf.gz.tbi
00-All.vcf.gz.tbi:<span class=" -Color -Color-Green"> downloaded                                                   </span>: 
00-All.vcf.gz:<span class=" -Color -Color-Green"> downloaded                                                       </span>: 
INFO: <span class=" -Color -Color-Green">download_dbsnp</span> is <span class=" -Color -Color-Green">completed</span>.
INFO: <span class=" -Color -Color-Green">download_dbsnp</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/00-All.vcf.gz /restricted/projectnb/xqtl/xqtl_protocol/reference_data/00-All.vcf.gz.tbi</span>
INFO: Workflow download_dbsnp (ID=wd4d1efccb2946bf1) is executed successfully with 1 completed step.
</pre></div>
</div>
</div>
</div>
</section>
<section id="ii-format-reference-data">
<h3>ii. Format Reference Data<a class="headerlink" href="#ii-format-reference-data" title="Link to this heading">#</a></h3>
<p>Timing: ~40 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb hg_reference \
    --cwd ../reference_data \
    --ercc-reference ../reference_data/ERCC92.fa \
    --hg-reference ../reference_data/GRCh38_full_analysis_set_plus_decoy_hla.fa 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">HLA ALT Decoy removal</span>: 
INFO: t52741959485e9056 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2339992 (&quot;job_t52741959485e9056&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">HLA ALT Decoy removal</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy.fasta</span>
INFO: Running <span class=" -Color -Color-Green">merge with ERCC reference</span>: 
INFO: t84b67f567fd35728 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2339996 (&quot;job_t84b67f567fd35728&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">merge with ERCC reference</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta</span>
INFO: Running <span class=" -Color -Color-Green">index the fasta file</span>: 
INFO: tf5b9c47f5c546b07 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2340092 (&quot;job_tf5b9c47f5c546b07&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">index the fasta file</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.dict</span>
INFO: Workflow hg_reference (ID=wc092cd2056576e99) is executed successfully with 3 completed steps and 3 completed tasks.
</pre></div>
</div>
</div>
</div>
<p>Change the –stranded option to –no-stranded if using un-stranded RNA-seq protocol.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb hg_gtf \
    --cwd ../reference_data \
    --hg-gtf ../reference_data/Homo_sapiens.GRCh38.103.chr.gtf \
    --hg-reference ../reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy.fasta \
    --stranded 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">add chr prefix to gtf file</span>: 
INFO: tf004f605413b7240 <span class=" -Color -Color-Green">re-execute completed</span>
INFO: tf004f605413b7240 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2473403 (&quot;job_tf004f605413b7240&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">add chr prefix to gtf file</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.gtf</span>
INFO: Running <span class=" -Color -Color-Green">collapsed gene model</span>: 
INFO: t764808e29ab699b0 <span class=" -Color -Color-Green">re-execute completed</span>
INFO: t764808e29ab699b0 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2473428 (&quot;job_t764808e29ab699b0&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">collapsed gene model</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.gtf</span>
INFO: Workflow hg_gtf (ID=wafa522e33646f5f8) is executed successfully with 2 completed steps and 2 completed tasks.
</pre></div>
</div>
</div>
</div>
</section>
<section id="iii-format-gene-feature-data">
<h3>iii. Format Gene Feature Data<a class="headerlink" href="#iii-format-gene-feature-data" title="Link to this heading">#</a></h3>
<p>Timing: ~1min</p>
<p>Change the –stranded option to –no-stranded if using un-stranded RNA-seq protocol.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb gene_annotation \
    --cwd ../reference_data \
    --ercc-gtf ../reference_data/ERCC92.gtf \
    --hg-gtf ../reference_data/Homo_sapiens.GRCh38.103.chr.gtf \
    --hg-reference ../reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy.fasta \
    --stranded 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">add chr prefix to gtf file</span>: 
INFO: Running <span class=" -Color -Color-Green">Preprocess ERCC gtf file</span>: 
INFO: Step <span class=" -Color -Color-Green">ercc_gtf</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> with signature constructed
INFO: Step <span class=" -Color -Color-Green">ercc_gtf</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> with signature constructed
INFO: <span class=" -Color -Color-Green">Preprocess ERCC gtf file</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/ERCC92.genes.patched.gtf</span>
INFO: Step <span class=" -Color -Color-Green">hg_gtf_1</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> with signature constructed
INFO: Step <span class=" -Color -Color-Green">hg_gtf_1</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> with signature constructed
INFO: <span class=" -Color -Color-Green">add chr prefix to gtf file</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.gtf</span>
INFO: Running <span class=" -Color -Color-Green">collapsed gene model</span>: 
INFO: Step <span class=" -Color -Color-Green">hg_gtf_2</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> with signature constructed
INFO: Step <span class=" -Color -Color-Green">hg_gtf_2</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> with signature constructed
INFO: <span class=" -Color -Color-Green">collapsed gene model</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.gtf</span>
INFO: Running <span class=" -Color -Color-Green">gene_annotation</span>: 
WARNING: Invalid tag &quot;gene_annotation_Homo_sapiens.GRCh38.103.chr.reformatted.ERCC Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC&quot; is added as &quot;gene_annotation_Homo_sapiens.GRCh38.103.chr.reformatted.ERCCHomo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC&quot;
INFO: tf66344318fec0849 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2473505 (&quot;job_tf66344318fec0849&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">gene_annotation</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.ERCC.gtf /restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf</span>
INFO: Workflow gene_annotation (ID=wd680c7eefcd76b7e) is executed successfully with 1 completed step, 3 ignored steps and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
<section id="iv-generate-star-index">
<h3>iv. Generate STAR Index<a class="headerlink" href="#iv-generate-star-index" title="Link to this heading">#</a></h3>
<p>Timing: ~30 min</p>
<p>Generate the STAR index without the GTF annotation file to allow for customized read length later in STAR alignment.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb STAR_index \
    --cwd ../reference_data \
    --hg-reference ../reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta \
    --numThreads 10 \
    --mem 40G
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">STAR_index</span>: 
INFO: t41bda200d129e222 <span class=" -Color -Color-Green">re-execute completed</span>
INFO: t41bda200d129e222 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2380397 (&quot;job_t41bda200d129e222&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">STAR_index</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/STAR_Index/chrName.txt /restricted/projectnb/xqtl/xqtl_protocol/reference_data/STAR_Index/SAindex... (8 items)</span>
INFO: Workflow STAR_index (ID=wc6567c60a9fc9f11) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
<section id="v-generate-rsem-index">
<h3>v. Generate RSEM Index<a class="headerlink" href="#v-generate-rsem-index" title="Link to this heading">#</a></h3>
<p>Timing: ~1min</p>
<p>Generate RSEM index with the uncollapsed gtf file (without the gene tag in its file name.)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb RSEM_index \
    --cwd ../reference_data \
    --hg-reference ../reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta \
    --hg-gtf ../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.ERCC.gtf
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">RSEM_index</span>: 
INFO: t3cd7c2d1d2347634 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2473576 (&quot;job_t3cd7c2d1d2347634&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">RSEM_index</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/RSEM_Index/rsem_reference.n2g.idx.fa /restricted/projectnb/xqtl/xqtl_protocol/reference_data/RSEM_Index/rsem_reference.grp... (7 items)</span>
INFO: Workflow RSEM_index (ID=wee55a025a3891eb8) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
<section id="vi-generate-refflat-annotation-for-picard">
<h3>vi. Generate RefFlat Annotation for Picard<a class="headerlink" href="#vi-generate-refflat-annotation-for-picard" title="Link to this heading">#</a></h3>
<p>Timing: ~1min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb RefFlat_generation \
    --cwd ../reference_data \
    --hg-gtf ../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.ERCC.gtf
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">RefFlat_generation</span>: 
INFO: t9e2919744de892d8 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2473638 (&quot;job_t9e2919744de892d8&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">RefFlat_generation</span> output:   <span class=" -Color -Color-Green">../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.ERCC.ref.flat</span>
INFO: Workflow RefFlat_generation (ID=w3f901e2f82fc33e0) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
<section id="vii-generate-suppa-annotation-for-psichomics">
<h3>vii. Generate SUPPA Annotation for Psichomics<a class="headerlink" href="#vii-generate-suppa-annotation-for-psichomics" title="Link to this heading">#</a></h3>
<p>Timing: ~1min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run reference_data_preparation.ipynb SUPPA_annotation \
    --cwd ../reference_data \
    --hg_gtf ../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.ERCC.gtf 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">SUPPA_annotation_1</span>: 
INFO: t043bdd93ad2e832e <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2473686 (&quot;job_t043bdd93ad2e832e&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">SUPPA_annotation_1</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/hg38.Homo_sapiens.GRCh38.103.chr.reformatted.ERCC_SE_strict.ioe</span>
INFO: Running <span class=" -Color -Color-Green">SUPPA_annotation_2</span>: 
INFO: tde3dad57fa2b8b7e <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2473726 (&quot;job_tde3dad57fa2b8b7e&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">SUPPA_annotation_2</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.ERCC.SUPPA_annotation.rds</span>
INFO: Workflow SUPPA_annotation (ID=wc2895daedb91047b) is executed successfully with 2 completed steps and 2 completed tasks.
</pre></div>
</div>
</div>
</div>
</section>
<section id="viii-extract-rsids-for-known-variants">
<h3>viii. Extract rsIDs for known variants<a class="headerlink" href="#viii-extract-rsids-for-known-variants" title="Link to this heading">#</a></h3>
<p>Timing: &lt;30min</p>
<p>Extract the rsID for the known varriants so that we can distinguish them from the novel variants we had in our study. The procedure/rationale is explained in this post</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run VCF_QC.ipynb dbsnp_annotate \
    --genoFile ../reference_data/00-All.vcf.gz \
    --cwd ../reference_data 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">dbsnp_annotate</span>: 
INFO: tb731222e5323dc2d <span class=" -Color -Color-Green">restart</span> from status <span class=" -Color -Color-Green">submitted</span>
INFO: tb731222e5323dc2d <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 2546318 (&quot;job_tb731222e5323dc2d&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">dbsnp_annotate</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/reference_data/00-All.variants.gz</span>
INFO: Workflow dbsnp_annotate (ID=w6bb14ba9b9686b80) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run reference_data_preparation.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run reference_data.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  download_hg_reference
  download_gene_annotation
  download_ercc_reference
  download_dbsnp
  gff3_to_gtf
  hg_reference
  hg_gtf
  ercc_gtf
  gene_annotation
  STAR_index
  RSEM_index
  RefFlat_generation
  SUPPA_annotation
  psi_hg38_annotation_modification

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files.
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option

Sections
  download_hg_reference:
  download_gene_annotation:
  download_ercc_reference:
  download_dbsnp:
  gff3_to_gtf:
    Workflow Options:
      --gff3-file VAL (as path, required)
  hg_reference_1:
    Workflow Options:
      --hg-reference VAL (as path, required)
                        Path to HG reference file
  hg_reference_2:
    Workflow Options:
      --ercc-reference VAL (as path, required)
  hg_reference_3:
  hg_gtf_1:
    Workflow Options:
      --hg-reference VAL (as path, required)
      --hg-gtf VAL (as path, required)
  hg_gtf_2:
    Workflow Options:
      --[no-]stranded (required)
  ercc_gtf:
    Workflow Options:
      --ercc-gtf VAL (as path, required)
  gene_annotation:
  STAR_index:
    Workflow Options:
      --hg-reference VAL (as path, required)
  RSEM_index:
    Workflow Options:
      --hg-gtf VAL (as path, required)
      --hg-reference VAL (as path, required)
  RefFlat_generation:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  SUPPA_annotation_1:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  SUPPA_annotation_2:
    Workflow Options:
      --hg-gtf VAL (as path, required)
  psi_hg38_annotation_modification_1:
    Workflow Options:
      --hg-gtf VAL (as path, required)
      --hgrc-db VAL (as path, required)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# The output directory for generated files.
parameter: cwd = path(&quot;output&quot;)
# For cluster jobs, number commands to run per job
parameter: job_size = 1
# Wall clock time expected
parameter: walltime = &quot;5h&quot;
# Memory expected
parameter: mem = &quot;16G&quot;
# Number of threads
parameter: numThreads = 8
# Software container option
parameter: container = &quot;&quot;
import re
parameter: entrypoint= (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;
cwd = path(f&#39;{cwd:a}&#39;)
from sos.utils import expand_size
</pre></div>
</div>
</div>
</div>
</section>
<section id="data-download">
<h2>Data download<a class="headerlink" href="#data-download" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[download_hg_reference]
output: f&quot;{cwd:a}/GRCh38_full_analysis_set_plus_decoy_hla.fa&quot;
download: dest_dir = cwd
    ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[download_gene_annotation]
output: f&quot;{cwd:a}/Homo_sapiens.GRCh38.103.chr.gtf&quot;
download: dest_dir = cwd, decompress=True
    http://ftp.ensembl.org/pub/release-103/gtf/homo_sapiens/Homo_sapiens.GRCh38.103.chr.gtf.gz
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[download_ercc_reference]
output: f&quot;{cwd:a}/ERCC92.gtf&quot;, f&quot;{cwd:a}/ERCC92.fa&quot;
download: dest_dir = cwd, decompress=True
    https://tools.thermofisher.com/content/sfs/manuals/ERCC92.zip
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[download_dbsnp]
output: f&quot;{cwd:a}/00-All.vcf.gz&quot;, f&quot;{cwd:a}/00-All.vcf.gz.tbi&quot;
download: dest_dir = cwd
    ftp://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606/VCF/00-All.vcf.gz
    ftp://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606/VCF/00-All.vcf.gz.tbi
</pre></div>
</div>
</div>
</div>
</section>
<section id="gff3-to-gtf-formatting">
<h2>GFF3 to GTF formatting<a class="headerlink" href="#gff3-to-gtf-formatting" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[gff3_to_gtf]
parameter: gff3_file = path
input: gff3_file
output: f&#39;{cwd}/{_input:bn}.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint = entrypoint
        gffread ${_input} -T -o ${_output}
</pre></div>
</div>
</div>
</div>
</section>
<section id="hg-reference-file-preprocessing">
<h2>HG reference file preprocessing<a class="headerlink" href="#hg-reference-file-preprocessing" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Remove the HLA/ALT/Decoy record from the fasta – because none of the downstreams RNA-seq calling pipeline component can handle them properly.</p></li>
<li><p>Adding in ERCC information to the fasta file – even if ERCC is not included in the RNA-seq library it does not harm to add them.</p></li>
<li><p>Generating index for the fasta file</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_reference_1 (HLA ALT Decoy removal)]
# Path to HG reference file
parameter: hg_reference = path
input: hg_reference
output:  f&#39;{cwd}/{_input:bn}.noALT_noHLA_noDecoy.fasta&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
python: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint = entrypoint
    with open(&#39;${_input}&#39;, &#39;r&#39;) as fasta:
        contigs = fasta.read()
        contigs = contigs.split(&#39;&gt;&#39;)
        contig_ids = [i.split(&#39; &#39;, 1)[0] for i in contigs]

        # exclude ALT, HLA and decoy contigs
        filtered_fasta = &#39;&gt;&#39;.join([c for i,c in zip(contig_ids, contigs)
        if not (i[-4:]==&#39;_alt&#39; or i[:3]==&#39;HLA&#39; or i[-6:]==&#39;_decoy&#39;)])
    
    with open(&#39;${_output}&#39;, &#39;w&#39;) as fasta:
        fasta.write(filtered_fasta)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_reference_2 (merge with ERCC reference)]
parameter: ercc_reference = path
output: f&#39;{cwd}/{_input:bn}_ERCC.fasta&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint = entrypoint
    sed &#39;s/ERCC-/ERCC_/g&#39; ${ercc_reference} &gt; ${ercc_reference:n}.patched.fa
    cat ${_input} ${ercc_reference:n}.patched.fa &gt; ${_output}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_reference_3 (index the fasta file)]
output: f&#39;{cwd}/{_input:bn}.dict&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint = entrypoint
    samtools faidx ${_input}
    picard CreateSequenceDictionary \
        R=${_input} \
        O=${_output}
</pre></div>
</div>
</div>
</div>
</section>
<section id="transcript-and-gene-model-reference-processing">
<h2>Transcript and gene model reference processing<a class="headerlink" href="#transcript-and-gene-model-reference-processing" title="Link to this heading">#</a></h2>
<p>This step modify the <code class="docutils literal notranslate"><span class="pre">gtf</span></code> file for following reasons:</p>
<ol class="arabic simple">
<li><p>RSEM require GTF input to have the same chromosome name format (with <code class="docutils literal notranslate"><span class="pre">chr</span></code> prefix) as the fasta file. <strong>Although for STAR, this problem can be solved by the <code class="docutils literal notranslate"><span class="pre">--sjdbGTFchrPrefix</span> <span class="pre">&quot;chr&quot;</span></code> option, we have to add <code class="docutils literal notranslate"><span class="pre">chr</span></code> to it for use with RSEM anyways.</strong> That is why we would like to do it here.</p></li>
<li><p>Gene model collapsing script <code class="docutils literal notranslate"><span class="pre">collapse_annotation.py</span></code> from GTEx require the gtf have <code class="docutils literal notranslate"><span class="pre">transcript_type</span></code> instead <code class="docutils literal notranslate"><span class="pre">transcript_biotype</span></code> in its annotation. We rename it here, although <strong>this problem can also be solved by modifying the collapse_annotation.py while building the docker, since we are already customizing the reference file for 1 above, we add this in as well, as another customization made.</strong></p></li>
<li><p>Adding in ERCC information to the <code class="docutils literal notranslate"><span class="pre">gtf</span></code> reference.</p></li>
</ol>
<p>We may reimplement 1 and 2 with the alternative approaches discussed above if the problem with RSEM is solved, or when RSEM is no longer needed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_gtf_1 (add chr prefix to gtf file)]
parameter: hg_reference = path
parameter: hg_gtf = path
input: hg_reference, hg_gtf
output: f&#39;{cwd}/{_input[1]:bn}.reformatted.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint = entrypoint
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;dplyr&quot;)
    options(scipen = 999)
    con &lt;- file(&quot;${_input[0]}&quot;,&quot;r&quot;)
    fasta &lt;- readLines(con,n=1)
    close(con)
    gtf = read_delim(&quot;${_input[1]}&quot;, &quot;\t&quot;,  col_names  = F, comment = &quot;#&quot;, col_types=&quot;ccccccccc&quot;)
    if(!str_detect(fasta,&quot;&gt;chr&quot;)) {
        gtf_mod = gtf%&gt;%mutate(X1 = str_remove_all(X1,&quot;chr&quot;))
    } else if (!any(str_detect(gtf$X1[1],&quot;chr&quot;))) {
        gtf_mod = gtf%&gt;%mutate(X1 = paste0(&quot;chr&quot;,X1))
    } else (gtf_mod = gtf)
    if(any(str_detect(gtf_mod$X9, &quot;transcript_biotype&quot;))) {
      gtf_mod = gtf_mod%&gt;%mutate(X9 = str_replace_all(X9,&quot;transcript_biotype&quot;,&quot;transcript_type&quot;))
    }
    gtf_mod%&gt;%write.table(&quot;${_output}&quot;,sep = &quot;\t&quot;,quote = FALSE,col.names = F,row.names = F)
</pre></div>
</div>
</div>
</div>
<p><strong>Text below is taken from <a class="github reference external" href="https://github.com/broadinstitute/gtex-pipeline/tree/master/gene_model">broadinstitute/gtex-pipeline</a></strong></p>
<p>Gene-level expression and eQTLs from the GTEx project are calculated based on a collapsed gene model (i.e., combining all isoforms of a gene into a single transcript), according to the following rules:</p>
<ol class="arabic simple">
<li><p>Transcripts annotated as “retained_intron” or “read_through” are excluded. Additionally, transcripts that overlap with annotated read-through transcripts may be blacklisted (blacklists for GENCODE v19, 24 &amp; 25 are provided in this repository; no transcripts were blacklisted for v26).</p></li>
<li><p>The union of all exon intervals of each gene is calculated.</p></li>
<li><p>Overlapping intervals between genes are excluded from all genes.</p></li>
</ol>
<p>The purpose of step 3 is primarily to exclude overlapping regions from genes annotated on both strands, which can’t be unambiguously quantified from unstranded RNA-seq (GTEx samples were sequenced using an unstranded protocol). For stranded protocols, this step can be skipped by adding the <code class="docutils literal notranslate"><span class="pre">--collapse_only</span></code> flag.</p>
<p>Further documentation is available on the <a class="reference external" href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">GTEx Portal</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[hg_gtf_2 (collapsed gene model)]
parameter: stranded = bool
output: f&#39;{_input:n}{&quot;.collapse_only&quot; if stranded else &quot;&quot;}.gene.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint = entrypoint
    collapse_annotation.py ${&quot;--collapse_only&quot; if stranded else &quot;&quot;} ${_input} ${_output}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[ercc_gtf (Preprocess ERCC gtf file)]
parameter: ercc_gtf = path
input: ercc_gtf
output: f&#39;{cwd}/{_input:bn}.genes.patched.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
python: expand = &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint = entrypoint
    with open(&#39;${_input}&#39;) as exon_gtf, open(&#39;${_output}&#39;, &#39;w&#39;) as gene_gtf:
        for line in exon_gtf:
            f = line.strip().split(&#39;\t&#39;)
            f[0] = f[0].replace(&#39;-&#39;,&#39;_&#39;)  # required for RNA-SeQC/GATK (no &#39;-&#39; in contig name)
        
            attr = f[8]
            if attr[-1]==&#39;;&#39;:
                attr = attr[:-1]
            attr = dict([i.split(&#39; &#39;) for i in attr.replace(&#39;&quot;&#39;,&#39;&#39;).split(&#39;; &#39;)])
            # add gene_name, gene_type
            attr[&#39;gene_name&#39;] = attr[&#39;gene_id&#39;]
            attr[&#39;gene_type&#39;] = &#39;ercc_control&#39;
            attr[&#39;gene_status&#39;] = &#39;KNOWN&#39;
            attr[&#39;level&#39;] = 2
            for k in [&#39;id&#39;, &#39;type&#39;, &#39;name&#39;, &#39;status&#39;]:
                attr[&#39;transcript_&#39;+k] = attr[&#39;gene_&#39;+k]
        
            attr_str = []
            for k in [&#39;gene_id&#39;, &#39;transcript_id&#39;, &#39;gene_type&#39;, &#39;gene_status&#39;, &#39;gene_name&#39;,
                &#39;transcript_type&#39;, &#39;transcript_status&#39;, &#39;transcript_name&#39;]:
                attr_str.append(&#39;{0:s} &quot;{1:s}&quot;;&#39;.format(k, attr[k]))
            attr_str.append(&#39;{0:s} {1:d};&#39;.format(&#39;level&#39;, attr[&#39;level&#39;]))
            f[8] = &#39; &#39;.join(attr_str)
        
            # write gene, transcript, exon
            gene_gtf.write(&#39;\t&#39;.join(f[:2]+[&#39;gene&#39;]+f[3:])+&#39;\n&#39;)
            gene_gtf.write(&#39;\t&#39;.join(f[:2]+[&#39;transcript&#39;]+f[3:])+&#39;\n&#39;)
            f[8] = &#39; &#39;.join(attr_str[:2])
            gene_gtf.write(&#39;\t&#39;.join(f[:2]+[&#39;exon&#39;]+f[3:])+&#39;\n&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[gene_annotation]
input: output_from(&quot;hg_gtf_1&quot;), output_from(&quot;hg_gtf_2&quot;), output_from(&quot;ercc_gtf&quot;)
output: f&#39;{cwd}/{_input[0]:bn}.ERCC.gtf&#39;, f&#39;{cwd}/{_input[1]:bn}.ERCC.gtf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container, entrypoint = entrypoint
    cat ${_input[0]} ${_input[2]} &gt; ${_output[0]}
    cat ${_input[1]} ${_input[2]} &gt; ${_output[1]}
</pre></div>
</div>
</div>
</div>
</section>
<section id="generating-index-file-for-star">
<h2>Generating index file for <code class="docutils literal notranslate"><span class="pre">STAR</span></code><a class="headerlink" href="#generating-index-file-for-star" title="Link to this heading">#</a></h2>
<p>This step generate the index file for STAR alignment. This file just need to generate once and can be re-used.</p>
<p><strong>At least 40GB of memory is needed</strong>.</p>
<section id="step-inputs">
<h3>Step Inputs<a class="headerlink" href="#step-inputs" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code> and <code class="docutils literal notranslate"><span class="pre">fasta</span></code>: path to reference sequence. Both of them needs to be unzipped. <code class="docutils literal notranslate"><span class="pre">gtf</span></code> should be the one prior to collapse by gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sjdbOverhang</span></code>: specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads. We use 100 here as recommended by the TOPMed pipeline. See here for <a class="reference external" href="https://groups.google.com/g/rna-star/c/h9oh10UlvhI/m/BfSPGivUHmsJ">some additional discussions</a>.</p></li>
</ul>
</section>
<section id="step-output">
<h3>Step Output<a class="headerlink" href="#step-output" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Indexing file stored in <code class="docutils literal notranslate"><span class="pre">{cwd}/STAR_Index</span></code>, which will be used by <code class="docutils literal notranslate"><span class="pre">STAR</span></code></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[STAR_index]
parameter: hg_reference = path
# Specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads.
# Default choice follows from TOPMed pipeline recommendation.
if expand_size(mem) &lt; expand_size(&#39;40G&#39;):
    print(&quot;Insufficent memory for STAR, changing to 40G&quot;)
    star_mem = &#39;40G&#39;
else:
    star_mem = mem
input: hg_reference
output: f&quot;{cwd}/STAR_Index/chrName.txt&quot;, 
        f&quot;{cwd}/STAR_Index/SAindex&quot;, f&quot;{cwd}/STAR_Index/SA&quot;, f&quot;{cwd}/STAR_Index/genomeParameters.txt&quot;, 
        f&quot;{cwd}/STAR_Index/chrStart.txt&quot;,
        f&quot;{cwd}/STAR_Index/chrLength.txt&quot;, 
        f&quot;{cwd}/STAR_Index/Genome&quot;, f&quot;{cwd}/STAR_Index/chrNameLength.txt&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output[0]:bd}&#39;, cores = numThreads
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[1]:n}.stderr&#39;, stdout = f&#39;{_output[1]:n}.stdout&#39;, entrypoint = entrypoint
    STAR --runMode genomeGenerate \
         --genomeDir ${_output[0]:d} \
         --genomeFastaFiles ${_input[0]} \
         --runThreadN ${numThreads}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="generating-index-file-for-rsem">
<h2>Generating index file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code><a class="headerlink" href="#generating-index-file-for-rsem" title="Link to this heading">#</a></h2>
<p>This step generate the indexing file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code>. This file just need to generate once.</p>
<section id="id1">
<h3>Step Inputs<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code> and <code class="docutils literal notranslate"><span class="pre">fasta</span></code>: path to reference sequence. <code class="docutils literal notranslate"><span class="pre">gtf</span></code> should be the one prior to collapse by gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sjdbOverhang</span></code>: specifies the length of the genomic sequence around the annotated junction to be used in constructing the splice junctions database. Ideally, this length should be equal to the ReadLength-1, where ReadLength is the length of the reads.</p></li>
</ul>
</section>
<section id="step-outputs">
<h3>Step Outputs<a class="headerlink" href="#step-outputs" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Indexing file stored in <code class="docutils literal notranslate"><span class="pre">RSEM_Index</span></code>, which will be used by <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[RSEM_index]
parameter: hg_gtf = path
parameter: hg_reference = path
input: hg_reference, hg_gtf
output: f&quot;{cwd}/RSEM_Index/rsem_reference.n2g.idx.fa&quot;, f&quot;{cwd}/RSEM_Index/rsem_reference.grp&quot;, 
        f&quot;{cwd}/RSEM_Index/rsem_reference.idx.fa&quot;, f&quot;{cwd}/RSEM_Index/rsem_reference.ti&quot;, 
        f&quot;{cwd}/RSEM_Index/rsem_reference.chrlist&quot;, f&quot;{cwd}/RSEM_Index/rsem_reference.seq&quot;, 
        f&quot;{cwd}/RSEM_Index/rsem_reference.transcripts.fa&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output[0]:bd}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[1]:n}.stderr&#39;, stdout = f&#39;{_output[1]:n}.stdout&#39;, entrypoint = entrypoint
    rsem-prepare-reference \
            ${_input[0]} \
            ${_output[1]:n} \
            --gtf ${_input[1]} \
            --num-threads ${numThreads}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="generation-of-refflat-file">
<h2>Generation of RefFlat file<a class="headerlink" href="#generation-of-refflat-file" title="Link to this heading">#</a></h2>
<p>This file is needed for picard CollectRnaSeqMetrics module, which in turn</p>
<blockquote>
<div><p>produces metrics describing the distribution of the bases within the transcripts. It calculates the total numbers and the fractions of nucleotides within specific genomic regions including untranslated regions (UTRs), introns, intergenic sequences (between discrete genes), and peptide-coding sequences (exons). This tool also determines the numbers of bases that pass quality filters that are specific to Illumina data (PF_BASES).</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[RefFlat_generation]
parameter: hg_gtf = path
input: hg_gtf
output: f&#39;{_input:n}.ref.flat&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output[0]:bd}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, entrypoint = entrypoint
    gtfToGenePred ${_input}  ${_output}.tmp -genePredExt -geneNameAsName2
    awk -F&#39;\t&#39; -v OFS=&quot;\t&quot; &#39;{$1=$12 OFS $1;}7&#39; ${_output}.tmp | cut -f 1-11 &gt; ${_output}
    rm ${_output}.tmp
</pre></div>
</div>
</div>
</div>
</section>
<section id="generation-of-suppa-annotation-for-psichomics">
<h2>Generation of SUPPA annotation for psichomics.<a class="headerlink" href="#generation-of-suppa-annotation-for-psichomics" title="Link to this heading">#</a></h2>
<p>The generation of custom alternative splicing annotation is based on <a class="reference external" href="https://rpubs.com/nuno-agostinho/preparing-AS-annotation">this tutorial</a>. The way to generate the local alternative splicing suppa output is documented on <a class="reference external" href="https://github.com/comprna/SUPPA">SUPPA github page</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[SUPPA_annotation_1]
parameter: hg_gtf = path
input: hg_gtf
output: f&#39;{cwd}/hg38.{_input:bn}_SE_strict.ioe&#39; # The stderr file must not shared the same start with the output file
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{cwd}/{_input:bn}.stderr&#39;, stdout = f&#39;{cwd}/{_input:bn}.stdout&#39;, entrypoint = entrypoint
    suppa.py generateEvents -i ${_input} -o ${cwd}/hg38.${_input:bn} -f ioe -e SE SS MX RI FL
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[SUPPA_annotation_2]
parameter: hg_gtf = path
output: f&#39;{cwd}/{hg_gtf:bn}.SUPPA_annotation.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, entrypoint = entrypoint
    library(&quot;psichomics&quot;)
    suppa &lt;- parseSuppaAnnotation(&quot;${_input:d}&quot;, genome=&quot;hg38&quot;) 
    annot &lt;- prepareAnnotationFromEvents(suppa)
    saveRDS(annot, file=${_output:r})
</pre></div>
</div>
</div>
</div>
</section>
<section id="generate-psichomics-hg38-splicing-annotation">
<h2>Generate psichomics Hg38 splicing annotation<a class="headerlink" href="#generate-psichomics-hg38-splicing-annotation" title="Link to this heading">#</a></h2>
<p>Since the original annotation provided by psichomics package is using gene symbols, we modified it to use Ensembl IDs. The modified annotation will be used for <a class="reference external" href="https://statfungen.github.io/xqtl-protocol/code/molecular_phenotypes/calling/splicing_calling.html">detecting RNA alternative splicing using psichomics</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/reference_data_preparation.ipynb psi_hg38_annotation \
    --hg-gtf reference_data/Homo_sapiens.GRCh38.103.chr.reformated.gtf \
    --hgrc-db reference_data/hgnc_database.txt
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[psi_hg38_annotation]
parameter: hg_gtf = path
# FIXME: Please document what this file is and where do we get it @xuanhe.
parameter: hgrc_db = path
input: hg_gtf, hgrc_db
output: f&#39;{cwd}/psichomics_hg38_annotation.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, entrypoint = entrypoint
    library(&quot;psichomics&quot;)
    library(&quot;purrr&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;data.table&quot;)
  
    # load psicomics default annotation, option of hg38 from listSplicingAnnotations()
    annotation &lt;- loadAnnotation(&quot;AH63657&quot;)

    
    # reduce the demension of annotation file
    annotation &lt;- 
      map(annotation, ~.x%&gt;%
                       tidyr::unnest(cols = `Gene`))
  
    # Create empty colomns for each event for easier mapping
    annotation[[&quot;Tandem UTR&quot;]][[&quot;SUPPA.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Tandem UTR&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Alternative first exon&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Alternative last exon&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
    annotation[[&quot;Mutually exclusive exon&quot;]][[&quot;VAST-TOOLS.Event.ID&quot;]] &lt;- NA
  
    # extract Ensembl ID substring from original SUPPA.ID and VASTTOOL.ID
    annotation &lt;- 
      map(annotation, ~.x%&gt;%
                       mutate(ENSG.SUPPA = substr(`SUPPA.Event.ID`, 1, 15))%&gt;%
                       mutate(ENSG.VAST = substr(`VAST-TOOLS.Event.ID`, 1, 15)))
  
    # Load gtf file
    gtf_sample &lt;- read.table(&#39;${_input[0]}&#39;,header = FALSE, sep = &#39;\t&#39;)
  
    # from the gtf file, seperate gene names and corresponding Ensembl ID
    gtf_sample &lt;- separate(gtf_sample, V9, sep = &quot;;&quot;,into = c(&quot;gene_id&quot;, &quot;transcript_id&quot;, &quot;exon_number&quot;, &quot;gene_name&quot;))
    gtf_sample &lt;- separate(gtf_sample, gene_id, sep = &quot; &quot;,into = c(&quot;gene_id&quot;, &quot;gene_id_val&quot;))
    gtf_sample &lt;- separate(gtf_sample, gene_name, sep = &quot;e &quot;,into = c(&quot;gene_name&quot;, &quot;gene_name_val&quot;))
  
    gtf_name_id_match &lt;- gtf_sample[,c(&quot;gene_id_val&quot;,&quot;gene_name_val&quot;)]
    gtf_name_id_match &lt;- gtf_name_id_match[!duplicated(gtf_name_id_match), ]
  
    # For any matched approved id in the psi hg38 annotation and gtf file, record the corresponding Ensembl ID
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                       mutate(`ENSG.GTF` = gtf_name_id_match$gene_id_val[match(`Gene`, gtf_name_id_match$gene_name_val)]))
  
    # load hgnc database
    hgnc_db &lt;- fread(&#39;${_input[1]}&#39;, fill = TRUE, header = TRUE, sep = &#39;\t&#39;, quote=&quot;&quot;)
  
    # Combine the `Ensembl.ID.supplied.by.Ensembl.` and `Ensembl.gene.ID` column, if there are any conflict use the former
    # For conflict ones (15 total) both the former and latter records are poiting to the same gene name in Ensembl website so the order should not matter
    hgnc_db &lt;- hgnc_db %&gt;%
    mutate(ENSG.ID = ifelse(`Ensembl ID(supplied by Ensembl)` == &quot;&quot;, `Ensembl gene ID`, `Ensembl ID(supplied by Ensembl)`))
  
    # Create a one to one reference list for approved names, previous names and aliases
    # There is no duplicate symbol and Ensembl id info for approved symbol so no need for chromosome verification
    hgnc_name_id_match &lt;- hgnc_db[,c(&quot;Approved symbol&quot;,&quot;ENSG.ID&quot;)]
    hgnc_name_prev_check &lt;- hgnc_db[,c(&quot;Previous symbols&quot;,&quot;Chromosome&quot;,&quot;ENSG.ID&quot;)]
    hgnc_name_alias_check &lt;- hgnc_db[,c(&quot;Alias symbols&quot;,&quot;Chromosome&quot;,&quot;ENSG.ID&quot;)]
  
    # Remove NAs
    hgnc_name_prev_check &lt;- hgnc_name_prev_check[hgnc_name_prev_check$ENSG.ID != &quot;&quot;,]
    hgnc_name_alias_check &lt;- hgnc_name_alias_check[hgnc_name_alias_check$ENSG.ID != &quot;&quot;,]

    hgnc_name_prev_check &lt;- hgnc_name_prev_check[hgnc_name_prev_check$&quot;Previous symbols&quot; != &quot;&quot;,] 
    hgnc_name_alias_check &lt;- hgnc_name_alias_check[hgnc_name_alias_check$&quot;Alias symbols&quot; != &quot;&quot;,]

    # Seperate symbol column values from list of sybols to individual rows with one each
    hgnc_name_prev_check &lt;- separate_rows(hgnc_name_prev_check, &quot;Previous symbols&quot;, convert = FALSE)
    hgnc_name_alias_check &lt;- separate_rows(hgnc_name_alias_check, &quot;Alias symbols&quot;, convert = FALSE)
  
    # Convert chomosome info in hgnc database to number for matching with other database
    hgnc_name_prev_check &lt;- separate(hgnc_name_prev_check, &quot;Chromosome&quot;, sep = &#39;p&#39;, into = &quot;Chrp&quot;, remove = FALSE)
    hgnc_name_prev_check &lt;- separate(hgnc_name_prev_check, &quot;Chromosome&quot;, sep = &#39;q&#39;, into = &quot;Chrq&quot;, remove = FALSE)
    hgnc_name_prev_check &lt;- hgnc_name_prev_check%&gt;%
                                mutate(Chr = ifelse(nchar(hgnc_name_prev_check$Chrp) &lt;= 2, Chrp, Chrq))
    
    hgnc_name_alias_check &lt;- separate(hgnc_name_alias_check, &quot;Chromosome&quot;, sep = &#39;p&#39;, into = &quot;Chrp&quot;, remove = FALSE)
    hgnc_name_alias_check &lt;- separate(hgnc_name_alias_check, &quot;Chromosome&quot;, sep = &#39;q&#39;, into = &quot;Chrq&quot;, remove = FALSE)
    hgnc_name_alias_check &lt;- hgnc_name_alias_check%&gt;%
                                mutate(Chr = ifelse(nchar(hgnc_name_alias_check$Chrp) &lt;= 2, Chrp, Chrq))
  
    # For any matched approved id in the psi hg38 annotation and hgnc database, record the corresponding Ensembl ID
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                       mutate(`ENSG.HGNC` = hgnc_name_id_match$`ENSG.ID`[match(`Gene`, hgnc_name_id_match$&quot;Approved symbol&quot;)]))
  
    # Drop hypothetical genes
    annotation&lt;-
      map(annotation, ~.x%&gt;%
            subset(`Gene` != &#39;Hypothetical&#39;))
  
    # IN remaining NAs, for any matched alias/previous names and chromosome in the psi hg38 annotation and hgnc database, record the corresponding Ensembl ID
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                      mutate(ENSG.HGNC = ifelse(is.na(`ENSG.HGNC`) | `ENSG.HGNC` == &quot;&quot;,
                                                hgnc_name_alias_check$`ENSG.ID`[match(`Gene`, hgnc_name_alias_check$&quot;Alias symbols&quot;) &amp; match(`Chromosome`, hgnc_name_alias_check$Chr)],
                                                `ENSG.HGNC`))%&gt;%
                      mutate(ENSG.HGNC = ifelse(is.na(`ENSG.HGNC`) | `ENSG.HGNC` == &quot;&quot;,
                                                hgnc_name_prev_check$`ENSG.ID`[match(`Gene`, hgnc_name_prev_check$&quot;Previous symbols&quot;) &amp; match(`Chromosome`, hgnc_name_prev_check$Chr)],
                                                `ENSG.HGNC`)))
  
    # Build the final Ensembl id column base on the gtf file first, then for remaining NAs check the HGNC database record, SUPPA and VASTTOOL record,
    # Drop special cases that Ensembl ID is not recorded in VASTTOOLs and SUPPA
    # finnally in the remaining NA Ensmbl ID if the gene name in original annotation is NCBI IDs just use it
    annotation &lt;-
      map(annotation, ~.x%&gt;%
                       mutate(`ENSG.ID` = `ENSG.GTF`)%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`),
                                               `ENSG.HGNC`,
                                                `ENSG.ID`))%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`),
                                               `ENSG.VAST`,
                                                `ENSG.ID`))%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`),
                                               `ENSG.SUPPA`,
                                                `ENSG.ID`))%&gt;%
                       mutate(`ENSG.ID` = ifelse(substr(`ENSG.ID`, 1, 4) == &#39;ENSG&#39;,
                                                 `ENSG.ID`,
                                                 NA))%&gt;%
                       mutate(`ENSG.ID` = ifelse(is.na(`ENSG.ID`) &amp; substr(`Gene`, 1, 3) == &#39;LOC&#39;,
                                               `Gene`,
                                                `ENSG.ID`)))

    # Use the Ensembl IDs to replace gene names, drop remaining NAs
    annotation &lt;-
    map(annotation, ~.x%&gt;%
                       mutate(`Gene` = `ENSG.ID`)%&gt;%
            drop_na(`Gene`)
          )

    # save modified annotation
    saveRDS(annotation, file = &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="generate-tad-associated-index">
<h2>Generate TAD associated index<a class="headerlink" href="#generate-tad-associated-index" title="Link to this heading">#</a></h2>
<p>We use <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7895846/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7895846/</a> to define the genotype cis region for fine mapping with different phenotypes. The actual TAD dataframe for each cell types and tissues are download from the hg38 TAD link from  <a class="reference external" href="http://3dgenome.fsm.northwestern.edu/publications.html">http://3dgenome.fsm.northwestern.edu/publications.html</a>. The generation of this TAD file follows three steps:</p>
<ol class="arabic simple">
<li><p>Union of TADs across tissues</p></li>
<li><p>Extend with TAD boundaries</p></li>
<li><p>Extend with cis windows</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[Tad_annotateion]
## The tad file downloads from 
parameter: TAD_list = path
parameter: region_list = path
parameter: TAD_genotype = path
input: TAD_list, region_list
output:f&#39;{cwd:a}/{region_list:bn}.annotation&#39; 
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
python: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, entrypoint = entrypoint
    import pandas as pd
    import numpy as np
    #tad = pd.read_csv(&quot;${_input[0]}&quot;,&quot;\t&quot;,header = None)
    #tad.columns = [&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;]
    #tad[&quot;tad_index&quot;] = [f&#39;tad{x}&#39; for x in np.arange(1,len(tad)+1)]
    tad = pd.read_csv(&quot;${_input[0]}&quot;,&quot;\t&quot;)
    region_list =  pd.read_csv(&quot;${_input[1]}&quot;,&quot;\t&quot;)
    tad = tad.merge(region_list, on = &quot;#chr&quot;).query(&quot;start_y &gt; start_x &amp;  end_y &lt; end_x&quot;).drop([&quot;start_y&quot;,&quot;end_y&quot;] , axis = 1).rename({&quot;start_x&quot;:&quot;start&quot;,&quot;end_x&quot;:&quot;end&quot;})
    tad.to_csv(&quot;${_output}&quot;,&quot;\t&quot;) 
    tad_geno_list = pd.read_csv(&quot;${TAD_genotype}&quot;,&quot;\t&quot;)
    tad_gene_id = tad.merge(tad_geno_list,left_on = &quot;tad_index&quot;,right_on = &quot;#id&quot;).iloc[:,[4,6]]
    tad_gene_id.columns = [&quot;ID&quot;,&quot;dir&quot;]
    tad_gene_id.to_csv(&quot;${_output}.genotype_files_list.tsv&quot;,&quot;\t&quot;) 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/hs3163/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument &#39;filepath_or_buffer&#39; will be keyword-only.
  exec(code_obj, self.user_global_ns, self.user_ns)
/home/hs3163/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument &#39;filepath_or_buffer&#39; will be keyword-only.
  exec(code_obj, self.user_global_ns, self.user_ns)
/home/hs3163/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument &#39;filepath_or_buffer&#39; will be keyword-only.
  exec(code_obj, self.user_global_ns, self.user_ns)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/reference_data"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="reference_data.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Reference Data</p>
      </div>
    </a>
    <a class="right-next"
       href="generalized_TADB.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Generation of Topologically Associated Domains and their Boundaries</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-and-output">Input and Output</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#download-reference-data">Download Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gff3-to-gtf">GFF3 To GTF</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#format-reference-data">Format Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#format-gene-feature-data">Format Gene Feature Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-star-index">Generate STAR Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-rsem-index">Generate RSEM Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-refflat-annotation-for-picard">Generate RefFlat Annotation for Picard</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-suppa-annotation-for-psichomics">Generate SUPPA Annotation for Psichomics</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-rsids-for-known-variants">Extract rsIDs for known variants</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tad-annotation">TAD Annotation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-download-reference-data">i. Download Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-format-reference-data">ii. Format Reference Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-format-gene-feature-data">iii. Format Gene Feature Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iv-generate-star-index">iv. Generate STAR Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#v-generate-rsem-index">v. Generate RSEM Index</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-generate-refflat-annotation-for-picard">vi. Generate RefFlat Annotation for Picard</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#vii-generate-suppa-annotation-for-psichomics">vii. Generate SUPPA Annotation for Psichomics</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#viii-extract-rsids-for-known-variants">viii. Extract rsIDs for known variants</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-download">Data download</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gff3-to-gtf-formatting">GFF3 to GTF formatting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#hg-reference-file-preprocessing">HG reference file preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#transcript-and-gene-model-reference-processing">Transcript and gene model reference processing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-star">Generating index file for <code class="docutils literal notranslate"><span class="pre">STAR</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-output">Step Output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-index-file-for-rsem">Generating index file for <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-refflat-file">Generation of RefFlat file</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-suppa-annotation-for-psichomics">Generation of SUPPA annotation for psichomics.</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-psichomics-hg38-splicing-annotation">Generate psichomics Hg38 splicing annotation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-tad-associated-index">Generate TAD associated index</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>