
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Hierarchical Multiple Testing &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/association_scan/qtl_association_postprocessing';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Mixture Multivariate Distribution Estimate" href="../multivariate_genome/multivariate_mixture_vignette.html" />
    <link rel="prev" title="Quantile regression for QTL association testing" href="quantile_models/qr_and_twas.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/association_scan/qtl_association_postprocessing.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/association_scan/qtl_association_postprocessing.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/association_scan/qtl_association_postprocessing.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Hierarchical Multiple Testing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description-fixme">Description [FIXME]</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-genetic-effect">i. Genetic Effect</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-interaction-effect">ii. Interaction Effect</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-quantile-qtl">iii. Quantile QTL</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="hierarchical-multiple-testing">
<h1>Hierarchical Multiple Testing<a class="headerlink" href="#hierarchical-multiple-testing" title="Link to this heading">#</a></h1>
<section id="description-fixme">
<h2>Description [FIXME]<a class="headerlink" href="#description-fixme" title="Link to this heading">#</a></h2>
<p>This protocol implements a three-step procedure:</p>
<ol class="arabic simple">
<li><p>Local adjustment: p-values of all cis-variants adjusted within each gene</p></li>
<li><p>Global adjustment: minimum adjusted p-values from Step 1 further adjusted across all genes</p></li>
<li><p>Global informed identification of significant xQTL: xQTL with locally adjusted p-value below the threshold</p></li>
</ol>
<p>It also reorganizes intermediate files produced in tensorQTL analysis to specified archive folder for book-keeping or deletion, to save space.</p>
<p><a class="reference external" href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6294523/">cf. Huang et al. 2018</a></p>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Output from TensorQTL in the form of <code class="docutils literal notranslate"><span class="pre">*.cis_qtl.regional.tsv.gz</span></code> files</p></li>
<li><p>A gene coordinates tsv file with <code class="docutils literal notranslate"><span class="pre">chr</span></code>, <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">end</span></code> and <code class="docutils literal notranslate"><span class="pre">gene_id</span></code> columns</p></li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*significant_qtl.filtered_bonferroni_BH_adjusted.tsv.gz</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*significant_qtl.original_bonferroni_BH_adjusted.tsv.gz</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*significant_qtl.permutation_adjusted.tsv.gz</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*significant_qtl.q_beta_adjusted_events_qvalue.tsv.gz</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*significant_events.filtered_bonferroni_BH_adjusted.tsv.gz</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*significant_events.original_bonferroni_BH_adjusted.tsv.gz</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*significant_events.permutation_adjusted.tsv.gz</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*summary.txt</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*multiple_testing_consolidated.rds</span></code></p></li>
</ol>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<section id="i-genetic-effect">
<h3>i. Genetic Effect<a class="headerlink" href="#i-genetic-effect" title="Link to this heading">#</a></h3>
<p>With</p>
<ol class="arabic simple">
<li><p>permutation testing</p></li>
<li><p>recounted number of variants limited to MAF and cis-windows for Bonferroni correction</p></li>
</ol>
<p>We recommend reassessing significance of Bonferroni correction if previous analysis use too large of window or contain many low frequency variants. Below we reassess with MAF &gt; 0.01 and cis-window size 1000000 up and downstreams of TSS/TES.</p>
<p>When <code class="docutils literal notranslate"><span class="pre">maf_cutoff</span></code> and <code class="docutils literal notranslate"><span class="pre">cis_window</span></code> are not zero, the program will first compute the number of variants after filtering by these metric and write files with <code class="docutils literal notranslate"><span class="pre">n_variants_suffix</span></code> in the same folder as QTL data, then use those numbers to create filtered list of variants for Bonferroni adjusted p-value.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/qtl_association_postprocessing.ipynb<span class="w"> </span>default<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>output/tensorqtl_cis<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--gene-coordinates<span class="w"> </span>data/look_up_gene_id.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--sub-dir<span class="w"> </span>.<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--tss_dist_col<span class="w"> </span>cis_window_start_distance<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--tes_dist_col<span class="w"> </span>cis_window_end_distance<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pecotmr-path<span class="w"> </span>pecotmr<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--maf-cutoff<span class="w"> </span><span class="m">0</span>.01<span class="w"> </span>--cis-window<span class="w"> </span><span class="m">1000000</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--regional-pattern<span class="w"> </span><span class="s2">&quot;*.cis_qtl.regional.tsv.gz</span>$<span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output-dir<span class="w"> </span>output/heirarchical_multi_test/output<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--archive-dir<span class="w"> </span>output/heirarchical_multi_test/archive<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-s<span class="w"> </span>force
</pre></div>
</div>
</section>
<section id="ii-interaction-effect">
<h3>ii. Interaction Effect<a class="headerlink" href="#ii-interaction-effect" title="Link to this heading">#</a></h3>
<p>With EigenMT method to get us the effective number of tests, it is recommended setting <code class="docutils literal notranslate"><span class="pre">maf_cutoff</span></code> and <code class="docutils literal notranslate"><span class="pre">cis_window</span></code> <strong>both to zero</strong> to tell the program to <strong>not</strong> recompute counts for Bonferroni correction,</p>
<p>The above approach seems recommended by <code class="docutils literal notranslate"><span class="pre">tensorqtl</span></code> package because it is their default option for interaction QTL analysis. We will also adopt it as a recommended practice.</p>
<p>To keep using variant count based Bonferroni correct, simply remove <code class="docutils literal notranslate"><span class="pre">--regional-pattern</span></code>,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>~/GIT/xqtl-protocol/code/association_scan/qtl_association_postprocessing.ipynb<span class="w"> </span>default<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--gene-coordinates<span class="w"> </span>look_up_gene_id.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>~/Downloads/snuc_DeJager_Ast_tensorQTL_MWE<span class="w"> </span>--sub-dir<span class="w"> </span><span class="s2">&quot;interaction/msex&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--maf-cutoff<span class="w"> </span><span class="m">0</span><span class="w"> </span>--cis-window<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pecotmr-path<span class="w"> </span>/path/to/pecotmr<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--regional-pattern<span class="w"> </span><span class="s2">&quot;*.cis_qtl_top_assoc.txt.gz</span>$<span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pvalue-pattern<span class="w"> </span><span class="s2">&quot;pvalue_.*_interaction&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--qvalue-pattern<span class="w"> </span><span class="s2">&quot;qvalue_interaction&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output-dir<span class="w"> </span>~/output<span class="w"> </span>--archive-dir<span class="w"> </span>~/archive<span class="w"> </span>-s<span class="w"> </span>force<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--fdr-threshold<span class="w"> </span><span class="m">0</span>.3<span class="w"> </span><span class="c1"># set it large for illustration purpose</span>
</pre></div>
</div>
</section>
<section id="iii-quantile-qtl">
<h3>iii. Quantile QTL<a class="headerlink" href="#iii-quantile-qtl" title="Link to this heading">#</a></h3>
<p>Here we summarize the significant QTL using both original variants as well as additionally considering MAF&gt;0.05 and cis-window 1Mb around a gene, for Bonferroni adjusted p-value,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>~/GIT/xqtl-protocol/code/association_scan/qtl_association_postprocessing.ipynb<span class="w"> </span>default<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--gene-coordinates<span class="w"> </span>look_up_gene_id.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>~/Downloads/QR_ROSMAP_Ast_mega_quantile_tensorQTL_MWE<span class="w"> </span>--sub-dir<span class="w"> </span><span class="s2">&quot;.&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--maf-cutoff<span class="w"> </span><span class="m">0</span>.05<span class="w"> </span>--cis-window<span class="w"> </span><span class="m">1000000</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pecotmr-path<span class="w"> </span>/path/to/pecotmr<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--af-col<span class="w"> </span>maf<span class="w"> </span>--tss-dist-col<span class="w"> </span>feature_tss<span class="w"> </span>--tes-dist-col<span class="w"> </span>feature_tes<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--qvalue-pattern<span class="w"> </span><span class="s2">&quot;^qvalue</span>$<span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output-dir<span class="w"> </span>~/output<span class="w"> </span>--archive-dir<span class="w"> </span>~/archive<span class="w"> </span>-s<span class="w"> </span>force
</pre></div>
</div>
</section>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run qtl_association_postprocessing.ipynb -h
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">association_scan</span><span class="o">/</span><span class="n">qtl_association_postprocessing</span><span class="o">.</span><span class="n">ipynb</span>
               <span class="p">[</span><span class="n">workflow_name</span> <span class="o">|</span> <span class="o">-</span><span class="n">t</span> <span class="n">targets</span><span class="p">]</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="p">[</span><span class="n">workflow_options</span><span class="p">]</span>
  <span class="n">workflow_name</span><span class="p">:</span>        <span class="n">Single</span> <span class="ow">or</span> <span class="n">combined</span> <span class="n">workflows</span> <span class="n">defined</span> <span class="ow">in</span> <span class="n">this</span> <span class="n">script</span>
  <span class="n">targets</span><span class="p">:</span>              <span class="n">One</span> <span class="ow">or</span> <span class="n">more</span> <span class="n">targets</span> <span class="n">to</span> <span class="n">generate</span>
  <span class="n">options</span><span class="p">:</span>              <span class="n">Single</span><span class="o">-</span><span class="n">hyphen</span> <span class="n">sos</span> <span class="n">parameters</span> <span class="p">(</span><span class="n">see</span> <span class="s2">&quot;sos run -h&quot;</span> <span class="k">for</span> <span class="n">details</span><span class="p">)</span>
  <span class="n">workflow_options</span><span class="p">:</span>     <span class="n">Double</span><span class="o">-</span><span class="n">hyphen</span> <span class="n">workflow</span><span class="o">-</span><span class="n">specific</span> <span class="n">parameters</span>

<span class="n">Workflows</span><span class="p">:</span>
  <span class="n">default</span>

<span class="n">Global</span> <span class="n">Workflow</span> <span class="n">Options</span><span class="p">:</span>
  <span class="o">--</span><span class="n">cwd</span> <span class="o">.</span> <span class="p">(</span><span class="k">as</span> <span class="n">path</span><span class="p">)</span>
  <span class="o">--</span><span class="n">pecotmr</span><span class="o">-</span><span class="n">path</span> <span class="n">VAL</span> <span class="p">(</span><span class="k">as</span> <span class="n">path</span><span class="p">,</span> <span class="n">required</span><span class="p">)</span>
                        <span class="n">path</span> <span class="n">to</span> <span class="n">pecotmr</span> <span class="n">repo</span>
  <span class="o">--</span><span class="n">gene</span><span class="o">-</span><span class="n">coordinates</span> <span class="n">VAL</span> <span class="p">(</span><span class="k">as</span> <span class="n">path</span><span class="p">,</span> <span class="n">required</span><span class="p">)</span>
  <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="n">VAL</span> <span class="p">(</span><span class="k">as</span> <span class="n">path</span><span class="p">,</span> <span class="n">required</span><span class="p">)</span>
  <span class="o">--</span><span class="n">archive</span><span class="o">-</span><span class="nb">dir</span> <span class="n">VAL</span> <span class="p">(</span><span class="k">as</span> <span class="n">path</span><span class="p">,</span> <span class="n">required</span><span class="p">)</span>
  <span class="o">--</span><span class="n">sub</span><span class="o">-</span><span class="nb">dir</span> <span class="n">VAL</span> <span class="p">(</span><span class="k">as</span> <span class="n">path</span><span class="p">,</span> <span class="n">required</span><span class="p">)</span>
  <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="n">cutoff</span> <span class="mf">0.01</span> <span class="p">(</span><span class="k">as</span> <span class="nb">float</span><span class="p">)</span>
  <span class="o">--</span><span class="n">cis</span><span class="o">-</span><span class="n">window</span> <span class="mi">1000000</span> <span class="p">(</span><span class="k">as</span> <span class="nb">int</span><span class="p">)</span>
  <span class="o">--</span><span class="n">tss</span><span class="o">-</span><span class="n">dist</span><span class="o">-</span><span class="n">col</span> <span class="s1">&#39;start_distance&#39;</span>
  <span class="o">--</span><span class="n">tes</span><span class="o">-</span><span class="n">dist</span><span class="o">-</span><span class="n">col</span> <span class="s1">&#39;end_distance&#39;</span>
  <span class="o">--</span><span class="n">af</span><span class="o">-</span><span class="n">col</span> <span class="n">af</span>
  <span class="o">--</span><span class="n">molecular</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="n">col</span> <span class="s1">&#39;molecular_trait_object_id&#39;</span>
  <span class="o">--</span><span class="n">pvalue</span><span class="o">-</span><span class="n">cutoff</span> <span class="mf">0.05</span> <span class="p">(</span><span class="k">as</span> <span class="nb">float</span><span class="p">)</span>
                        <span class="n">This</span> <span class="ow">is</span> <span class="k">for</span> <span class="n">selecting</span> <span class="n">the</span> <span class="n">subset</span> <span class="n">of</span> <span class="n">data</span> <span class="n">to</span> <span class="n">process</span> <span class="n">on</span> <span class="n">protential</span> <span class="n">signals</span> <span class="n">assuming</span> <span class="n">we</span> <span class="n">drop</span> <span class="n">those</span> <span class="n">above</span> <span class="n">this</span> <span class="n">threshold</span> <span class="n">This</span> <span class="n">might</span> <span class="n">lead</span> <span class="n">to</span> <span class="n">underestimates</span> <span class="ow">in</span>
                        <span class="n">qvalue</span> <span class="n">method</span> <span class="n">since</span> <span class="n">qvalue</span> <span class="o">&lt;</span> <span class="mf">0.05</span> <span class="n">may</span> <span class="n">contain</span> <span class="n">pvalue</span> <span class="o">&gt;</span> <span class="mf">0.05</span>
  <span class="o">--</span><span class="n">fdr</span><span class="o">-</span><span class="n">threshold</span> <span class="mf">0.05</span> <span class="p">(</span><span class="k">as</span> <span class="nb">float</span><span class="p">)</span>
                        <span class="n">This</span> <span class="ow">is</span> <span class="n">used</span> <span class="k">for</span> <span class="n">both</span> <span class="n">event</span> <span class="ow">and</span> <span class="n">variant</span> <span class="n">level</span> <span class="n">significance</span> <span class="nb">filter</span>
  <span class="o">--</span><span class="n">pvalue</span><span class="o">-</span><span class="n">pattern</span> <span class="s1">&#39;^pvalue$&#39;</span>
                        <span class="n">eg</span> <span class="s2">&quot;pvalue&quot;</span> <span class="k">for</span> <span class="n">pvalue</span><span class="p">,</span> <span class="s2">&quot;pvalue_.*_interaction&quot;</span> <span class="k">for</span> <span class="n">interaction</span>
  <span class="o">--</span><span class="n">qvalue</span><span class="o">-</span><span class="n">pattern</span> <span class="s1">&#39;^qvalue$&#39;</span>
                        <span class="n">eg</span> <span class="s2">&quot;qvalue&quot;</span> <span class="k">for</span> <span class="n">pvalue</span><span class="p">,</span> <span class="s2">&quot;qvalue_interaction&quot;</span> <span class="k">for</span> <span class="n">interaction</span>
  <span class="o">--</span><span class="n">regional</span><span class="o">-</span><span class="n">pattern</span> <span class="n">NULL</span>
                        <span class="n">eg</span> <span class="s2">&quot;*.cis_qtl.regional.tsv.gz$&quot;</span> <span class="k">for</span> <span class="n">genetic</span> <span class="n">effect</span> <span class="n">via</span> <span class="n">our</span> <span class="n">pipeline</span> <span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="s2">&quot;*.cis_qtl_top_assoc.txt.gz$&quot;</span> <span class="k">for</span> <span class="n">interaction</span> <span class="n">genetic</span> <span class="n">effect</span>
  <span class="o">--</span><span class="n">qtl</span><span class="o">-</span><span class="n">pattern</span> <span class="s1">&#39;*.cis_qtl.pairs.tsv.gz$&#39;</span>
  <span class="o">--</span><span class="n">n</span><span class="o">-</span><span class="n">variants</span><span class="o">-</span><span class="n">suffix</span> <span class="s1">&#39;cis_n_variants_stats.tsv.gz&#39;</span>

<span class="n">Sections</span>
  <span class="n">default</span><span class="p">:</span>

</pre></div>
</div>
</section>
<section id="workflow-implementation">
<h2>Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
parameter: cwd = path(&quot;.&quot;)
#path to pecotmr repo
parameter: pecotmr_path = path
parameter: gene_coordinates = path
parameter: output_dir = path
parameter: archive_dir = path
parameter: sub_dir = path
parameter: maf_cutoff = 0.01
parameter: cis_window = 1000000
parameter: tss_dist_col = &quot;start_distance&quot;
parameter: tes_dist_col = &quot;end_distance&quot;
parameter: af_col = &quot;af&quot;
parameter: molecular_id_col = &quot;molecular_trait_object_id&quot;
# This is for selecting the subset of data to process on protential signals 
# assuming we drop those above this threshold
# This might lead to underestimates in qvalue method since qvalue &lt; 0.05 may contain pvalue &gt; 0.05
parameter: pvalue_cutoff = 0.05
# This is used for both event and variant level significance filter
parameter: fdr_threshold = 0.05
# eg &quot;pvalue&quot; for pvalue, &quot;pvalue_.*_interaction&quot; for interaction
parameter: pvalue_pattern = &quot;^pvalue$&quot;
# eg &quot;qvalue&quot; for pvalue, &quot;qvalue_interaction&quot; for interaction
parameter: qvalue_pattern = &quot;^qvalue$&quot;
# eg &quot;*.cis_qtl.regional.tsv.gz$&quot; for genetic effect via our pipeline TensorQTL.ipynb
# &quot;*.cis_qtl_top_assoc.txt.gz$&quot; for interaction genetic effect
parameter: regional_pattern = &quot;NULL&quot;
parameter: qtl_pattern = &quot;*.cis_qtl.pairs.tsv.gz$&quot;
parameter: n_variants_suffix = &quot;cis_n_variants_stats.tsv.gz&quot;

work_dir = f&quot;{cwd:a}/{sub_dir}&quot;
if sub_dir == path(&quot;.&quot;):
    output_dir = f&quot;{output_dir:a}/{cwd:b}&quot;
    archive_dir = f&quot;{archive_dir:a}/{cwd:b}&quot;
    work_dir = f&quot;{cwd:a}&quot;
else:
    output_dir = f&quot;{output_dir:a}/{cwd:b}/{sub_dir}&quot;
    archive_dir = f&quot;{archive_dir:a}/{cwd:b}/{sub_dir}&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[default]
output: f&quot;{output_dir}/{cwd:b}_multiple_testing_consolidated.rds&quot;
task: trunk_workers = 1, tags = f&#39;tensorqtl_postprocessing_{_output:n}&#39;
R: expand = &quot;${ }&quot;

    params &lt;- list()
    params$workdir           &lt;- &quot;${work_dir}&quot;
    params$maf_cutoff        &lt;- ${maf_cutoff}
    params$cis_window        &lt;- ${cis_window}
    params$pvalue_cutoff     &lt;- ${pvalue_cutoff}
    params$fdr_threshold     &lt;- ${fdr_threshold}
    params$gene_coordinates  &lt;- &quot;${gene_coordinates:a}&quot;
    params$output_dir        &lt;- &quot;${output_dir}&quot;
    params$archive_dir       &lt;- &quot;${archive_dir}&quot;
    params$regional_pattern  &lt;- &quot;${regional_pattern}&quot;
    params$n_variants_suffix &lt;- &quot;${n_variants_suffix}&quot;
    params$qtl_pattern       &lt;- &quot;${qtl_pattern}&quot;
    params$pvalue_pattern    &lt;- &quot;${pvalue_pattern}&quot;
    params$qvalue_pattern    &lt;- &quot;${qvalue_pattern}&quot;
    params$start_distance_col &lt;- &quot;${tss_dist_col}&quot;
    params$end_distance_col   &lt;- &quot;${tes_dist_col}&quot;
    params$af_col &lt;- &quot;${af_col}&quot;
    params$molecular_id_col &lt;- &quot;${molecular_id_col}&quot;
    convert_null_strings &lt;- function(params) {
      if (is.list(params)) {
        # Apply the function to each element in the list
        result &lt;- lapply(params, convert_null_strings)
        return(result)
      } else {
        # For non-list elements, check if it&#39;s the string &quot;NULL&quot;
        if (is.character(params) &amp;&amp; params == &quot;NULL&quot;) {
          return(NULL)
        } else {
          return(params)
        }
      }
    }

    params &lt;- convert_null_strings(params)
    source(&quot;${pecotmr_path}/inst/code/tensorqtl_postprocessor.R&quot;)
    results &lt;- hierarchical_multiple_testing_correction(params)
    write_results(results, params$output_dir, params$workdir, to_cwd = &quot;regional&quot;)
    archive_files(params)
    saveRDS(results, ${_output:r})
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/association_scan"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="quantile_models/qr_and_twas.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Quantile regression for QTL association testing</p>
      </div>
    </a>
    <a class="right-next"
       href="../multivariate_genome/multivariate_mixture_vignette.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Mixture Multivariate Distribution Estimate</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description-fixme">Description [FIXME]</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-genetic-effect">i. Genetic Effect</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-interaction-effect">ii. Interaction Effect</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-quantile-qtl">iii. Quantile QTL</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>