
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>GREGOR enrichment analysis &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Stratified LD Score Regression" href="ldsc.html" />
    <link rel="prev" title="Multivariate TWAS prediction model" href="../cis_analysis/mrmash.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../reference_data/reference_data.html">
   Reference Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/reference_data_preparation.html">
     Reference Data Standardization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/generalized_TADB.html">
     Generation of Topologically Associated Domains and their Boundaries
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/ld_prune_reference.html">
     Independent list of variants using LD clumping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene Coordinate Annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype Data Imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden Factor Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate Data Formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pecotmr_integration/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/twas_mr.html">
     xQTL and GWAS integration: TWAS and MR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Extract genome-wide data for multivariate analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/enrichment/gregor.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/enrichment/gregor.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/enrichment/gregor.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#fixme">
   FIXME
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods-overview">
   Methods overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#reference-data-preparation">
   Reference data preparation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#annotation-files">
     Annotation files
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gregor-reference-files">
     GREGOR reference files
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#example-command">
   Example command
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#explaination-for-each-step">
   Explaination for each step
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#example">
   Example
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="gregor-enrichment-analysis">
<h1>GREGOR enrichment analysis<a class="headerlink" href="#gregor-enrichment-analysis" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://genome.sph.umich.edu/wiki/GREGOR">GREGOR (Genomic Regulatory Elements and Gwas Overlap algoRithm)</a> is a tool built to evaluate global enrichment of trait-associated variants in experimentally annotated epigenomic regulatory features.</p>
<p>Since GREGOR is released under UMichgan license, we build a Singularity container <code class="docutils literal notranslate"><span class="pre">gregor.sif</span></code> to run it locally without pushing the image into any container repositories.</p>
<div class="section" id="fixme">
<h2>FIXME<a class="headerlink" href="#fixme" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Ru: Include the dockerfile script under Methods Overview – DONE</p></li>
<li><p>Ru: Add <code class="docutils literal notranslate"><span class="pre">gregor_4</span></code> to include a default graphical summary – <code class="docutils literal notranslate"><span class="pre">gregor_fisher_plot</span></code>, there are 2 traits on the plot, and I think it would be better to seperate from gregor_1/2/3.</p></li>
<li><p>Hao: you may want to add base R environment to the container, so we can use container for all steps including <code class="docutils literal notranslate"><span class="pre">gregor_3</span></code> <code class="docutils literal notranslate"><span class="pre">gregor_4</span></code></p></li>
<li><p>Hao: we need to try to get hg38 working – possibly by changing GREGOR reference data. The reference data preparation step is not yet well-documented by Ru. Please revisit and improve the documentation.</p></li>
</ol>
</div>
<div class="section" id="methods-overview">
<h2>Methods overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">¶</a></h2>
<p>GREGOR performs enrichment analysis for given positive set of variants against matched negative set, matched by:</p>
<ul class="simple">
<li><p>MAF</p></li>
<li><p>LD proxy</p></li>
<li><p>distance to TSS</p></li>
<li><p>Gene density</p></li>
</ul>
<p>Methods to obtain the column “PValue” from GREGOR software can be found in the GREGOR paper: <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/25886982/">https://pubmed.ncbi.nlm.nih.gov/25886982/</a>. We have noticed some PValue are greater than 1. Also, p-value itself is not as informative compared to odds ratios of enrichment. We therefore parse intermediate outputs from GREGOR to get a 2 by 2 table where the rows are within annotation peak vs outside; the columns are positive set vs matched negative set. We perform Fisher’s exact test to get both p-value and odds ratios.</p>
</div>
<div class="section" id="reference-data-preparation">
<h2>Reference data preparation<a class="headerlink" href="#reference-data-preparation" title="Permalink to this headline">¶</a></h2>
<p>hg19.fa</p>
<ul class="simple">
<li><p>Genome Reference: hg19 (GRCh37) or hg38 (GRCh38).</p></li>
<li><p>Download <a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz">link</a></p></li>
</ul>
<div class="section" id="annotation-files">
<h3>Annotation files<a class="headerlink" href="#annotation-files" title="Permalink to this headline">¶</a></h3>
<p>Alkes Price’s lab has some annotation files in <code class="docutils literal notranslate"><span class="pre">bed</span></code> format available for download:</p>
<p><a class="reference external" href="https://data.broadinstitute.org/alkesgroup/LDSCORE/">https://data.broadinstitute.org/alkesgroup/LDSCORE/</a></p>
<p>eg <code class="docutils literal notranslate"><span class="pre">baselineLD_v2.2_bedfiles.tgz</span></code>.</p>
<p>Annotation files are in <code class="docutils literal notranslate"><span class="pre">bed</span></code> format (for example: Promoter_UCSC.bed).</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    chr1	9873	16361
    chr1	32610	36610
    chr1	67090	71090
    ...
</pre></div>
</div>
</div>
<div class="section" id="gregor-reference-files">
<h3>GREGOR reference files<a class="headerlink" href="#gregor-reference-files" title="Permalink to this headline">¶</a></h3>
<p>GREGOR reference files for hg19 can be prepared from downloaded reference data,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> \
    <span class="n">GREGOR</span><span class="o">.</span><span class="n">AFR</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">greater</span><span class="o">.</span><span class="n">than</span><span class="o">.</span><span class="mf">0.7</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">part</span><span class="o">.</span><span class="mi">00</span> \
    <span class="n">GREGOR</span><span class="o">.</span><span class="n">AFR</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">greater</span><span class="o">.</span><span class="n">than</span><span class="o">.</span><span class="mf">0.7</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">part</span><span class="o">.</span><span class="mi">01</span> \
    <span class="o">&gt;</span> <span class="n">GREGOR</span><span class="o">.</span><span class="n">AFR</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">greater</span><span class="o">.</span><span class="n">than</span><span class="o">.</span><span class="mf">0.7</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
<span class="n">tar</span> <span class="n">zxvf</span> <span class="n">GREGOR</span><span class="o">.</span><span class="n">AFR</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">r2</span><span class="o">.</span><span class="n">greater</span><span class="o">.</span><span class="n">than</span><span class="o">.</span><span class="mf">0.7</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>MD5SUM check:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AFR</span><span class="o">.</span><span class="n">part</span><span class="o">.</span><span class="mi">0</span>	<span class="p">(</span> <span class="n">MD5</span><span class="p">:</span> <span class="mi">9926904128</span><span class="n">dd58d6bf1ad4f1e90638af</span> <span class="p">)</span>
<span class="n">AFR</span><span class="o">.</span><span class="n">part</span><span class="o">.</span><span class="mi">1</span>	<span class="p">(</span> <span class="n">MD5</span><span class="p">:</span> <span class="n">c1d30aff89a584bfa8c1fa1bdc197f21</span> <span class="p">)</span>
</pre></div>
</div>
<p>Same for EUR data-set.</p>
</div>
</div>
<div class="section" id="example-command">
<h2>Example command<a class="headerlink" href="#example-command" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">gregor</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">gregor</span> \
    <span class="o">--</span><span class="n">gregor</span><span class="o">-</span><span class="n">db</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">GREGOR</span><span class="o">/</span><span class="n">REF</span> \
    <span class="o">--</span><span class="n">index</span><span class="o">-</span><span class="n">snp</span><span class="o">-</span><span class="n">file</span> <span class="n">GREGOR</span><span class="o">/</span><span class="n">example</span><span class="o">/</span><span class="n">example</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">snps</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="o">-</span><span class="n">index</span> <span class="n">GREGOR</span><span class="o">/</span><span class="n">example</span><span class="o">/</span><span class="n">example</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">file</span><span class="o">.</span><span class="n">index</span>  \
    <span class="o">--</span><span class="n">container</span> <span class="n">gregor</span><span class="o">.</span><span class="n">sif</span> 
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">gregor</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run gregor.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  gregor_conf
  gregor

Global Workflow Options:
  --cwd output (as path)
                        working directory
  --container &#39;&#39;
                        Software container option

Sections
  gregor_conf, gregor_1: make configuration file for GREGOR
    Workflow Options:
      --gregor-db VAL (as path, required)
      --pop EUR
      --index-snp-file VAL (as path, required)
      --bed-file-index VAL (as path, required)
      --ld-window-size 10000 (as int)
      --min-neighbor 10 (as int)
      --job-number 10 (as int)
  gregor_2:             run GREGOR
  gregor_3:             Fisher test of enrichment
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="explaination-for-each-step">
<h2>Explaination for each step<a class="headerlink" href="#explaination-for-each-step" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gregor_conf,</span> <span class="pre">gregor_1</span></code>: make the configuration file for GREGOR</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gregor_2</span></code>: run GREGOR, for the details, please see <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4612390/">Ellen et al.</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gregor_3</span></code>: The process for generating a 2x2 table for each annotation to conduct a Fisher test in the context of GREGOR analysis. The structure and computation of the elements in the 2x2 table are as follows:</p></li>
</ol>
<ul class="simple">
<li><p>Annotated Positive Set (N1): Derived from the <code class="docutils literal notranslate"><span class="pre">inBedIndexSNPNum</span></code> in the PValue.txt file located in each annotation folder. This value is integral to identifying the positive set that has been annotated.</p></li>
<li><p>Un-annotated Positive Set (Np): Calculated as the total number of lines in the <code class="docutils literal notranslate"><span class="pre">index_SNP/annotated.index.snp.txt</span></code> file. It represents the count of input signals that were successfully converted to chromosome positions, forming the overall positive set.</p></li>
<li><p>Annotated Negative Set (N2): This is computed by taking the total lines in the <code class="docutils literal notranslate"><span class="pre">neighbor*.txt</span></code> files (represented as N1 + N2) of each annotation folder and subtracting N1 from it, excluding any header lines. This gives the count of the negative set that has been annotated.</p></li>
<li><p>Un-annotated Negative Set (Nn): This is computed by taking the total line in the <code class="docutils literal notranslate"><span class="pre">neighbor_SNP/neighbor*.txt</span></code> files and subtracting Np from it, with header lines excluded. This represents the un-annotated portion of the negative set.</p></li>
</ul>
<ol class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gregor_fisher_test</span></code>: The visualization plot is designed to display the results of Fisher’s exact test, comparing two different traits. The input data for this plot is sourced from the ‘gregor_3’ outputs, with each dataset representing a distinct trait.</p></li>
</ol>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">gregor</span><span class="o">.</span><span class="n">ipynb</span>  <span class="n">gregor_fisher_plot</span> \
    <span class="o">--</span><span class="n">fisher1</span>  <span class="n">mvSuSiE_Basophill_perc_positive_cs</span><span class="o">.</span><span class="n">index_gregor_output</span><span class="o">/*</span><span class="n">_enrichment_results</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">fisher2</span> <span class="n">SuSiE_Basophill_perc_positive_cs</span><span class="o">.</span><span class="n">index_gregor_output</span><span class="o">/*</span><span class="n">_enrichment_results</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">test</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># working directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># make configuration file for GREGOR</span>
<span class="p">[</span><span class="n">gregor_conf</span><span class="p">,</span> <span class="n">gregor_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">gregor_db</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">pop</span> <span class="o">=</span> <span class="s1">&#39;EUR&#39;</span>
<span class="kn">parameter:</span> <span class="n">index_snp_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">bed_file_index</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span>  <span class="n">r2_threshold</span> <span class="o">=</span> <span class="mf">0.7</span>
<span class="kn">parameter:</span> <span class="n">ld_window_size</span> <span class="o">=</span> <span class="mi">10000</span>
<span class="kn">parameter:</span> <span class="n">min_neighbor</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">job_number</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">input:</span> <span class="n">index_snp_file</span><span class="p">,</span> <span class="n">bed_file_index</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.gregor.conf&#39;</span>
<span class="kn">report:</span> <span class="n">output</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="c1">##############################################################################</span>
    <span class="c1"># CHIPSEQ ENRICHMENT CONFIGURATION FILE</span>
    <span class="c1"># This configuration file contains run-time configuration of</span>
    <span class="c1"># CHIP_SEQ ENRICHMENT</span>
    <span class="c1">###############################################################################</span>
    <span class="c1">## KEY ELEMENTS TO CONFIGURE : NEED TO MODIFY</span>
    <span class="c1">###############################################################################</span>
    <span class="n">INDEX_SNP_FILE</span> <span class="o">=</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">BED_FILE_INDEX</span> <span class="o">=</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> 
    <span class="n">REF_DIR</span> <span class="o">=</span> <span class="p">{</span><span class="n">gregor_db</span><span class="p">}</span>
    <span class="n">R2THRESHOLD</span> <span class="o">=</span> <span class="p">{</span><span class="n">r2_threshold</span><span class="p">}</span> <span class="c1">## must be greater than 0.7</span>
    <span class="n">LDWINDOWSIZE</span> <span class="o">=</span> <span class="p">{</span><span class="n">ld_window_size</span><span class="p">}</span> <span class="c1">## must be less than this window; these two values define LD buddies</span>
    <span class="n">OUT_DIR</span> <span class="o">=</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="n">_gregor_output</span>
    <span class="n">MIN_NEIGHBOR_NUM</span> <span class="o">=</span> <span class="p">{</span><span class="n">min_neighbor</span><span class="p">}</span> <span class="c1">## define the minimum size of neighborhood</span>
    <span class="n">BEDFILE_IS_SORTED</span> <span class="o">=</span> <span class="n">true</span>  <span class="c1">## false, if the bed files are not sorted</span>
    <span class="n">POPULATION</span> <span class="o">=</span> <span class="p">{</span><span class="n">pop</span><span class="p">}</span>  <span class="c1">## define the population, you can specify EUR, AFR, AMR or ASN</span>
    <span class="n">TOPNBEDFILES</span> <span class="o">=</span> <span class="mi">2</span> 
    <span class="n">JOBNUMBER</span> <span class="o">=</span> <span class="p">{</span><span class="n">job_number</span><span class="p">}</span>
    <span class="c1">###############################################################################</span>
    <span class="c1">#BATCHTYPE = mosix ##  submit jobs on MOSIX</span>
    <span class="c1">#BATCHOPTS = -E/tmp -i -m2000 -j10,11,12,13,14,15,16,17,18,19,120,122,123,124,125 sh -c</span>
    <span class="c1">###############################################################################</span>
    <span class="c1">#BATCHTYPE = slurm   ##  submit jobs on SLURM</span>
    <span class="c1">#BATCHOPTS = --partition=broadwl --account=pi-mstephens --time=0:30:0</span>
    <span class="c1">###############################################################################</span>
    <span class="n">BATCHTYPE</span> <span class="o">=</span> <span class="n">local</span> <span class="c1">##  run jobs on local machine</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">sed</span> <span class="o">-</span><span class="n">i</span> <span class="s1">&#39;/^$/d&#39;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>GREGOR is written in <code class="docutils literal notranslate"><span class="pre">perl</span></code>. If you don’t use containers, some libraries are required before one can run GREGOR:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sudo</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">libdbi</span><span class="o">-</span><span class="n">perl</span> <span class="n">libswitch</span><span class="o">-</span><span class="n">perl</span> <span class="n">libdbd</span><span class="o">-</span><span class="n">sqlite3</span><span class="o">-</span><span class="n">perl</span>
</pre></div>
</div>
<p>With our docker image:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">GREGOR_folder</span>
<span class="n">docker</span> <span class="n">run</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;$PWD:/usr/src/myapp&quot;</span> <span class="o">-</span><span class="n">it</span> <span class="n">custom</span><span class="o">-</span><span class="n">perl</span>
<span class="n">perl</span> <span class="n">script</span><span class="o">/</span><span class="n">GREGOR</span><span class="o">.</span><span class="n">pl</span> <span class="o">--</span><span class="n">conf</span> <span class="n">example</span><span class="o">/</span><span class="n">mvsusie_annotation</span><span class="o">.</span><span class="n">conf</span>
<span class="n">perl</span> <span class="n">script</span><span class="o">/</span><span class="n">GREGOR</span><span class="o">.</span><span class="n">pl</span> <span class="o">--</span><span class="n">conf</span> <span class="n">example</span><span class="o">/</span><span class="n">susie_annotation</span><span class="o">.</span><span class="n">conf</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># run GREGOR</span>
<span class="p">[</span><span class="n">gregor_2</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">_gregor_output/StatisticSummaryFile.txt&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">perl</span> <span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">GREGOR</span><span class="o">/</span><span class="n">GREGOR</span><span class="o">/</span><span class="n">script</span><span class="o">/</span><span class="n">GREGOR</span><span class="o">.</span><span class="n">pl</span> <span class="o">--</span><span class="n">conf</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&amp;&amp;</span> <span class="n">touch</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Fisher test of enrichment</span>
<span class="p">[</span><span class="n">gregor_3</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">ad</span><span class="si">}</span><span class="s1">_variant_counts.txt&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">ad</span><span class="si">}</span><span class="s1">_enrichment_results.txt&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;$[ ]&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>

    <span class="n">cd</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">ad</span><span class="p">]</span><span class="o">/</span>  <span class="c1"># The issue was that,{_input:ad}_variant_counts.txt is not in the same dir as the input  </span>
    <span class="c1"># Loop through each subdirectory</span>
    <span class="k">for</span> <span class="nb">dir</span> <span class="ow">in</span> <span class="o">*/</span><span class="p">;</span> <span class="n">do</span>
        <span class="c1"># Ensure that the directory is not empty</span>
        <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot;$dir&quot;</span> <span class="p">]];</span> <span class="n">then</span>
            <span class="c1"># Calculate the total number of lines in neighbor.*.txt files</span>
            <span class="n">total_lines</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">find</span> <span class="s2">&quot;$dir&quot;</span> <span class="o">-</span><span class="kp">name</span> <span class="s1">&#39;neighbor.*.txt&#39;</span> <span class="o">-</span><span class="n">exec</span> <span class="n">cat</span> <span class="p">{}</span> <span class="o">+</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span><span class="p">)</span>

            <span class="c1"># Count the number of neighbor.*.txt files</span>
            <span class="n">file_count</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">find</span> <span class="s2">&quot;$dir&quot;</span> <span class="o">-</span><span class="kp">name</span> <span class="s1">&#39;neighbor.*.txt&#39;</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span><span class="p">)</span>

            <span class="c1"># Calculate the adjusted row number</span>
            <span class="n">row_num</span><span class="o">=</span><span class="err">$</span><span class="p">((</span><span class="n">total_lines</span> <span class="o">-</span> <span class="n">file_count</span><span class="p">))</span>

            <span class="c1"># Check if PValue.txt exists</span>
            <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">f</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">/PValue.txt&quot;</span> <span class="p">]];</span> <span class="n">then</span>
                <span class="c1"># Append the row count to PValue.txt and save as PValue_new.txt</span>
                <span class="p">(</span><span class="n">cat</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">/PValue.txt&quot;</span><span class="p">;</span> <span class="n">echo</span> <span class="s2">&quot;N1_N2 = $row_num&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">/PValue_N1_N2.txt&quot;</span>
            <span class="n">fi</span>
        <span class="n">fi</span>
    <span class="n">done</span>

    <span class="c1"># Header for the new summary file</span>
    <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;Bed_File</span><span class="se">\t</span><span class="s2">InBed_Index_SNP</span><span class="se">\t</span><span class="s2">ExpectNum_of_InBed_SNP</span><span class="se">\t</span><span class="s2">PValue</span><span class="se">\t</span><span class="s2">N1_N2</span><span class="se">\t</span><span class="s2">Np</span><span class="se">\t</span><span class="s2">Np_Nn&quot;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

    <span class="c1"># Calculate Np and Np_Nn for the neighbor_SNP directory</span>
    <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot;neighbor_SNP&quot;</span> <span class="p">]];</span> <span class="n">then</span>
        <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">f</span> <span class="s2">&quot;neighbor_SNP/index.snp.neighbors.txt&quot;</span> <span class="p">]];</span> <span class="n">then</span>
            <span class="n">Np</span><span class="o">=</span><span class="err">$</span><span class="p">((</span><span class="err">$</span><span class="p">(</span><span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">&lt;</span> <span class="s2">&quot;index_SNP/annotated.index.snp.txt&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
        <span class="k">else</span>
            <span class="n">Np</span><span class="o">=</span><span class="mi">0</span>
        <span class="n">fi</span>
        <span class="n">Np_Nn_files</span><span class="o">=</span><span class="p">(</span><span class="err">$</span><span class="p">(</span><span class="n">find</span> <span class="s2">&quot;neighbor_SNP&quot;</span> <span class="o">-</span><span class="kp">name</span> <span class="s1">&#39;neighbor.chr*txt&#39;</span><span class="p">))</span>
        <span class="n">Np_Nn</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="s2">&quot;$</span><span class="si">{Np_Nn_files[@]}</span><span class="s2">&quot;</span><span class="p">;</span> <span class="n">do</span>
            <span class="n">Np_Nn</span><span class="o">=</span><span class="err">$</span><span class="p">((</span><span class="err">$</span><span class="n">Np_Nn</span> <span class="o">+</span> <span class="err">$</span><span class="p">(</span><span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">&lt;</span> <span class="s2">&quot;$file&quot;</span><span class="p">)))</span>
        <span class="n">done</span>
        <span class="n">Np_Nn</span><span class="o">=</span><span class="err">$</span><span class="p">((</span><span class="err">$</span><span class="n">Np_Nn</span> <span class="o">-</span> <span class="err">$</span><span class="p">{</span><span class="c1">#Np_Nn_files[@]}))</span>
    <span class="n">fi</span>

    <span class="c1"># Loop through each subdirectory</span>
    <span class="k">for</span> <span class="nb">dir</span> <span class="ow">in</span> <span class="o">*/</span><span class="p">;</span> <span class="n">do</span>
        <span class="c1"># Ensure that the directory is not empty</span>
        <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot;$dir&quot;</span> <span class="p">]];</span> <span class="n">then</span>
            <span class="c1"># Check if PValue_new.txt exists</span>
            <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">f</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">PValue_N1_N2.txt&quot;</span> <span class="p">]];</span> <span class="n">then</span>
                <span class="c1"># Extract values from PValue_new.txt</span>
                <span class="n">inBedIndexSNPNum</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">grep</span> <span class="s2">&quot;inBedIndexSNPNum&quot;</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">PValue_N1_N2.txt&quot;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;=&#39;</span> <span class="o">-</span><span class="n">f2</span> <span class="o">|</span> <span class="n">tr</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;[:space:]&#39;</span><span class="p">)</span>
                <span class="n">expectedS</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">grep</span> <span class="s2">&quot;expectedS&quot;</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">PValue_N1_N2.txt&quot;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;=&#39;</span> <span class="o">-</span><span class="n">f2</span> <span class="o">|</span> <span class="n">tr</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;[:space:]&#39;</span><span class="p">)</span>
                <span class="n">p3</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">grep</span> <span class="s2">&quot;p3&quot;</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">PValue_N1_N2.txt&quot;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;=&#39;</span> <span class="o">-</span><span class="n">f2</span> <span class="o">|</span> <span class="n">tr</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;[:space:]&#39;</span><span class="p">)</span>
                <span class="n">N1_N2</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">grep</span> <span class="s2">&quot;N1_N2&quot;</span> <span class="s2">&quot;$</span><span class="si">{dir}</span><span class="s2">PValue_N1_N2.txt&quot;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;=&#39;</span> <span class="o">-</span><span class="n">f2</span> <span class="o">|</span> <span class="n">tr</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;[:space:]&#39;</span><span class="p">)</span>
                <span class="c1"># Get the directory name as the Bed_File</span>
                <span class="n">bedFile</span><span class="o">=</span><span class="err">$</span><span class="p">(</span><span class="n">basename</span> <span class="s2">&quot;$dir&quot;</span><span class="p">)</span>
                <span class="c1"># Append the data to the summary file</span>
                <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;$bedFile</span><span class="se">\t</span><span class="s2">$inBedIndexSNPNum</span><span class="se">\t</span><span class="s2">$expectedS</span><span class="se">\t</span><span class="s2">$p3</span><span class="se">\t</span><span class="s2">$N1_N2</span><span class="se">\t</span><span class="s2">$Np</span><span class="se">\t</span><span class="s2">$Np_Nn&quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="n">fi</span>
        <span class="n">fi</span>
    <span class="n">done</span>

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">res</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span> <span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span> 
    <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">res</span><span class="p">)){</span>
        <span class="n">n1</span> <span class="o">=</span> <span class="n">res</span><span class="err">$</span><span class="n">InBed_Index_SNP</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">n2</span> <span class="o">=</span> <span class="n">res</span><span class="err">$</span><span class="n">N1_N2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">n1</span>
        <span class="n">np</span> <span class="o">=</span> <span class="n">res</span><span class="err">$</span><span class="n">Np</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">nn</span> <span class="o">=</span> <span class="n">res</span><span class="err">$</span><span class="n">Np_Nn</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">np</span>

        <span class="c1"># Construct the contingency matrix</span>
        <span class="n">dat</span> <span class="o">=</span> <span class="n">matrix</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">,</span> <span class="n">np</span> <span class="o">-</span> <span class="n">n1</span><span class="p">,</span> <span class="n">nn</span> <span class="o">-</span> <span class="n">n2</span><span class="p">),</span> <span class="n">nrow</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Perform Fisher&#39;s exact test</span>
        <span class="n">test_res</span> <span class="o">=</span> <span class="n">fisher</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">alternative</span> <span class="o">=</span> <span class="s1">&#39;two.sided&#39;</span><span class="p">)</span>

        <span class="c1"># Store the results in &#39;res&#39;</span>
        <span class="n">res</span><span class="err">$</span><span class="n">odds</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">test_res</span><span class="err">$</span><span class="n">estimate</span>
        <span class="n">res</span><span class="err">$</span><span class="n">low</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">test_res</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">res</span><span class="err">$</span><span class="n">high</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">test_res</span><span class="err">$</span><span class="n">conf</span><span class="o">.</span><span class="n">int</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">res</span><span class="err">$</span><span class="n">p_fisher</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">test_res</span><span class="err">$</span><span class="n">p</span><span class="o">.</span><span class="n">value</span>
    <span class="p">}</span>

    <span class="n">res</span><span class="err">$</span><span class="n">odds</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">res</span><span class="err">$</span><span class="n">odds</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># enrichment plot with fisher test results</span>
<span class="p">[</span><span class="n">gregor_fisher_plot</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">fisher1</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">fisher2</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">fisher1</span><span class="p">,</span> <span class="n">fisher2</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">dbn</span><span class="si">}</span><span class="s1">_vs_</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">dbn</span><span class="si">}</span><span class="s1">_enrichment.pdf&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>    
    <span class="n">group1</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="n">group2</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="n">group1</span><span class="err">$</span><span class="n">group</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">dbnr</span><span class="p">}</span>
    <span class="n">group2</span><span class="err">$</span><span class="n">group</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">dbnr</span><span class="p">}</span>
    <span class="n">res_all</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="n">group1</span><span class="p">,</span> <span class="n">group2</span><span class="p">)</span>
    <span class="n">res_all</span><span class="err">$</span><span class="n">feature</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="n">res_all</span><span class="err">$</span><span class="n">Bed_File</span><span class="p">)</span>
    <span class="n">res_all</span> <span class="o">&lt;-</span> <span class="n">res_all</span> <span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">str_detect</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span><span class="s2">&quot;hg38&quot;</span><span class="p">)))</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">p_fisher</span> <span class="o">==</span> <span class="mi">1</span><span class="p">))</span><span class="o">%&gt;%</span> <span class="n">na</span><span class="o">.</span><span class="n">omit</span> 
    <span class="n">p</span> <span class="o">&lt;-</span> <span class="n">res_all</span><span class="o">%&gt;%</span>
        <span class="n">arrange</span><span class="p">(</span><span class="n">odds</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">ggplot</span><span class="p">()</span><span class="o">+</span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">odds</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">reorder</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span><span class="o">-</span><span class="n">odds</span><span class="p">),</span> <span class="n">color</span> <span class="o">=</span> <span class="n">group</span><span class="p">))</span><span class="o">+</span>
            <span class="n">geom_vline</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span> <span class="n">xintercept</span> <span class="o">=</span> <span class="mi">0</span> <span class="p">))</span><span class="o">+</span><span class="n">theme_bw</span><span class="p">()</span><span class="o">+</span><span class="n">theme</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="n">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="mi">20</span><span class="p">))</span><span class="o">+</span><span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;Odds Ratio&quot;</span><span class="p">)</span><span class="o">+</span>
                <span class="n">geom_linerange</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">xmin</span> <span class="o">=</span> <span class="n">low</span><span class="p">,</span> <span class="n">xmax</span><span class="o">=</span> <span class="n">high</span> <span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">feature</span> <span class="p">),</span> <span class="n">color</span><span class="o">=</span> <span class="s1">&#39;grey40&#39;</span><span class="p">)</span><span class="o">+</span><span class="n">ylab</span><span class="p">(</span><span class="s2">&quot;Functional Annotations&quot;</span><span class="p">)</span>
    <span class="n">ggsave</span><span class="p">(</span><span class="n">plot</span> <span class="o">=</span> <span class="n">p</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">height</span> <span class="o">=</span> <span class="mi">24</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/enrichment"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../cis_analysis/mrmash.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Multivariate TWAS prediction model</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="ldsc.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Stratified LD Score Regression</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>