

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Stratified LD Score Regression &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/enrichment/ldsc';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="The PECOTMR intergration framework" href="../pecotmr_integration/pecotmr.html" />
    <link rel="prev" title="GREGOR enrichment analysis" href="gregor.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype Data Imputation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">Overview of advanced cis-QTL analysis</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../cis_analysis/cis_workhorse.html">Advanced cis-QTL analysis with individual level data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/enrichment/ldsc.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/enrichment/ldsc.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/enrichment/ldsc.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Stratified LD Score Regression</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-brief-review-on-stratified-ld-score-regression">A brief review on Stratified LD score regression</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ldsc-model">LDSC model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#stratified-ldsc">Stratified LDSC</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#make-annotation-file">Make Annotation File</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#munge-summary-statistics-option-1-no-signed-summary-statistic">Munge Summary Statistics (Option 1: No Signed Summary Statistic)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#munge-summary-statistics-option-2-no-signed-summary-statistic">Munge Summary Statistics (Option 2: No Signed Summary Statistic)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-ld-scores">Calculate LD Scores</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-functional-enrichment-using-annotations">Calculate Functional Enrichment using Annotations</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="stratified-ld-score-regression">
<h1>Stratified LD Score Regression<a class="headerlink" href="#stratified-ld-score-regression" title="Permalink to this heading">#</a></h1>
<p>This notebook implements the pipepline of <a class="reference external" href="https://github.com/bulik/ldsc/wiki">S-LDSC</a> for LD score and functional enrichment analysis. It is written by Anmol Singh (<a class="reference external" href="mailto:singh&#46;anmol&#37;&#52;&#48;columbia&#46;edu">singh<span>&#46;</span>anmol<span>&#64;</span>columbia<span>&#46;</span>edu</a>), with input from Dr. Gao Wang.</p>
<p><strong>FIXME: the initial draft is complete but pending Gao’s review and documentation with minimal working example</strong></p>
<p>The pipeline is developed to integrate GWAS summary statistics data, annotation data, and LD reference panel data to compute functional enrichment for each of the epigenomic annotations that the user provides using the S-LDSC model. We will first start off with an introduction, instructions to set up, and the minimal working examples. Then the workflow code that can be run using SoS on any data will be at the end.</p>
<section id="a-brief-review-on-stratified-ld-score-regression">
<h2>A brief review on Stratified LD score regression<a class="headerlink" href="#a-brief-review-on-stratified-ld-score-regression" title="Permalink to this heading">#</a></h2>
<p>Here I briefly review LD Score Regression and what it is used for. For more in depth information on LD Score Regression please read the following three papers:</p>
<ol class="arabic simple">
<li><p>“LD Score regression distinguishes confounding from polygenicity in genome-wide association studies” by Sullivan et al (2015)</p></li>
<li><p>“Partitioning heritability by functional annotation using genome-wide association summary statistics” by Finucane et al (2015)</p></li>
<li><p>“Linkage disequilibrium–dependent architecture of human complex traits shows action of negative selection” by Gazal et al (2017)</p></li>
</ol>
<p>As stated in Sullivan et al 2015, confounding factors and polygenic effects can cause inflated test statistics and other methods cannot distinguish between inflation from confounding bias and a true signal. LD Score Regression (LDSC) is a technique that aims to identify the impact of confounding factors and polygenic effects using information from GWAS summary statistics.</p>
<p>This approach involves using regression to mesaure the relationship between Linkage Disequilibrium (LD) scores and test statistics of SNPs from the GWAS summary statistics. Variants in LD with a “causal” variant show an elevation in test statistics in association analysis proportional to their LD (measured by <span class="math notranslate nohighlight">\(r^2\)</span>) with the causal variant within a certain window size (could be 1 cM, 1kB, etc.). In contrast, inflation from confounders such as population stratification that occur purely from genetic drift will not correlate with LD. For a polygenic trait, SNPs with a high LD score will have more significant χ2 statistics on average than SNPs with a low LD score. Thus, if we regress the <span class="math notranslate nohighlight">\(\chi^2\)</span> statistics from GWAS against LD Score, the intercept minus one is an estimator of the mean contribution of confounding bias to the inflation in the test statistics. The regression model is known as LD Score regression.</p>
<section id="ldsc-model">
<h3>LDSC model<a class="headerlink" href="#ldsc-model" title="Permalink to this heading">#</a></h3>
<p>Under a polygenic assumption, in which effect sizes for variants are drawn independently from distributions with variance proportional to  <span class="math notranslate nohighlight">\(1/(p(1-p))\)</span> where p is the minor allele frequency (MAF), the expected <span class="math notranslate nohighlight">\(\chi^2\)</span> statistic of variant j is:</p>
<div class="math notranslate nohighlight">
\[E[\chi^2|l_j] = Nh^2l_j/M + Na + 1 \quad (1)\]</div>
<p>where <span class="math notranslate nohighlight">\(N\)</span> is the sample size; <span class="math notranslate nohighlight">\(M\)</span> is the number of SNPs, such that <span class="math notranslate nohighlight">\(h^2/M\)</span> is the average heritability explained per SNP; <span class="math notranslate nohighlight">\(a\)</span> measures the contribution of confounding biases, such as cryptic relatedness and population stratification; and <span class="math notranslate nohighlight">\(l_j = \sum_k r^2_{jk}\)</span> is the LD Score of variant <span class="math notranslate nohighlight">\(j\)</span>, which measures the amount of genetic variation tagged by <span class="math notranslate nohighlight">\(j\)</span>. A full derivation of this equation is provided in the Supplementary Note of Sullivan et al (2015). An alternative derivation is provided in Supplementary Note of Zhu and Stephens (2017) AoAS.</p>
<p>From this we can see that LD Score regression can be used to compute SNP-based heritability for a phenotype or trait, from GWAS summary statistics and does not require genotype information like other methods such as REML do.</p>
</section>
<section id="stratified-ldsc">
<h3>Stratified LDSC<a class="headerlink" href="#stratified-ldsc" title="Permalink to this heading">#</a></h3>
<p>Heritability is the proportion of phenotypic variation (VP) that is due to variation in genetic values (VG) and thus can tell us how much of the difference in observed phenotypes in a sample is due to difference in genetics in the sample. It can also be extended to analyze partitioned heritability for a phenotype/trait split over categories.</p>
<p>For Partitioned Heritability or Stratified LD Score Regression (S-LDSC) more power is added to our analysis by leveraging LD Score information as well as using SNPs that haven’t reached Genome Wide Significance to partition heritability for a trait over categories which many other methods do not do.</p>
<p>S-LDSC relies on the fact that the <span class="math notranslate nohighlight">\(\chi^2\)</span> association statistic for a given SNP includes the effects of all SNPs tagged by this SNP meaning that in a region of high LD in the genome the given SNP from the GWAS represents the effects of a group of SNPs in that region.</p>
<p>S-LDSC determines that a category of SNPs is enriched for heritability if SNPs with high LD to that category have more significant <span class="math notranslate nohighlight">\(\chi^2\)</span> statistics than SNPs with low LD to that category.</p>
<p>Here, enrichment of a category is defined as the proportion of SNP heritability in the category divided by the proportion of SNPs in that category.</p>
<p>More precisely, under a polygenic model, the expected <span class="math notranslate nohighlight">\(\chi^2\)</span> statistic of SNP <span class="math notranslate nohighlight">\(j\)</span> is</p>
<div class="math notranslate nohighlight">
\[E[\chi^2_j] = N\sum_CT_Cl(j,C) + Na + 1 \quad (2)\]</div>
<p>where <span class="math notranslate nohighlight">\(N\)</span> is sample size, C indexes categories, <span class="math notranslate nohighlight">\(ℓ(j, C)\)</span> is the LD score of SNP j with respect to category <span class="math notranslate nohighlight">\(l(j,C) = \sum_{k\epsilon C} r^2_{jk}\)</span>, <span class="math notranslate nohighlight">\(a\)</span> is a term that measures the contribution of confounding biases, and if the categories are disjoint, <span class="math notranslate nohighlight">\(\tau_C\)</span> is the per-SNP heritability in category <span class="math notranslate nohighlight">\(C\)</span>; if the categories overlap, then the per-SNP heritability of SNP j is <span class="math notranslate nohighlight">\(\sum_{C:j\epsilon C} \tau_C\)</span>.  Equation 2 allows us to estimate <span class="math notranslate nohighlight">\(\tau_C\)</span> via a (computationally simple) multiple regression of <span class="math notranslate nohighlight">\(\chi^2\)</span> against <span class="math notranslate nohighlight">\(ℓ(j, C)\)</span>, for either a quantitative or case-control study.</p>
<p>To see how these methods have been applied to real world data as well as a further discussion on methods and comparisons to other methods please read the three papers listed at the top of the document.</p>
</section>
</section>
<section id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run LDSC_Code.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run LDSC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  make_annot
  munge_sumstats_no_sign
  munge_sumstats_sign
  calc_ld_score
  calc_enrichment

Sections
  make_annot:
    Workflow Options:
      --bed VAL (as str, required)
                        path to bed file
      --bim VAL (as str, required)
                        path to bim file
      --annot VAL (as str, required)
                        name of output annotation file
  munge_sumstats_no_sign: This option is for when the summary statistic file
                        does not contain a signed summary statistic (Z or Beta).
                        In this case,the program will calculate Z for you based
                        on A1 being the risk allele
    Workflow Options:
      --sumst VAL (as str, required)
                        path to summary statistic file
      --alleles &#39;w_hm3.snplist&#39;
                        path to Hapmap3 SNPs file, keep all columns (SNP, A1,
                        and A2) for the munge_sumstats program
      --output VAL (as str, required)
                        path to output file
  munge_sumstats_sign:  This option is for when the summary statistic file does
                        contain a signed summary statistic (Z or Beta)
    Workflow Options:
      --sumst VAL (as str, required)
                        path to summary statistic file
      --alleles &#39;w_hm3.snplist&#39;
                        path to Hapmap3 SNPs file, keep all columns (SNP, A1,
                        and A2) for the munge_sumstats program
      --output VAL (as str, required)
                        path to output file
  calc_ld_score:        Calculate LD Scores **Make sure to delete SNP,CHR, and
                        BP columns from annotation files if they are present
                        otherwise this code will not work. Before deleting, if
                        these columns are present, make sure that the annotation
                        file is sorted.**
    Workflow Options:
      --bim VAL (as str, required)
                        Path to bim file
      --annot-file VAL (as str, required)
                        Path to annotation File. Make sure to remove the SNP,
                        CHR, and BP columns from the annotation file if present
                        before running.
      --output VAL (as str, required)
                        name of output file
      --snplist &#39;w_hm3.snplist&#39;
                        path to Hapmap3 SNPs file, remove the A1 and A2 columns
                        for the Calculate LD Scores program
  calc_enrichment:
    Workflow Options:
      --sumstats VAL (as str, required)
                        Path to Summary statistics File
      --ref-ld VAL (as str, required)
                        Path to Reference LD Scores Files (Base Annotation +
                        Annotation you want to analyze, format like minimal
                        working example)
      --w-ld VAL (as str, required)
                        Path to LD Weight Files (Format like minimal working
                        example)
      --frq-file VAL (as str, required)
                        path to frequency files (Format like minimal working
                        example)
      --output VAL (as str, required)
                        Output name
</pre></div>
</div>
</div>
</div>
</section>
<section id="make-annotation-file">
<h2>Make Annotation File<a class="headerlink" href="#make-annotation-file" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">make_annot</span><span class="p">]</span>

<span class="c1"># Make Annotated Bed File</span>

<span class="c1"># path to bed file</span>
<span class="kn">parameter:</span> <span class="n">bed</span> <span class="o">=</span> <span class="nb">str</span> 
<span class="c1">#path to bim file</span>
<span class="kn">parameter:</span> <span class="n">bim</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#name of output annotation file</span>
<span class="kn">parameter:</span> <span class="n">annot</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">make_annot</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span> <span class="p">{</span><span class="n">bed</span><span class="p">}</span> <span class="o">--</span><span class="n">bimfile</span> <span class="p">{</span><span class="n">bim</span><span class="p">}</span> <span class="o">--</span><span class="n">annot</span><span class="o">-</span><span class="n">file</span> <span class="p">{</span><span class="n">annot</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="munge-summary-statistics-option-1-no-signed-summary-statistic">
<h2>Munge Summary Statistics (Option 1: No Signed Summary Statistic)<a class="headerlink" href="#munge-summary-statistics-option-1-no-signed-summary-statistic" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#This option is for when the summary statistic file does not contain a signed summary statistic (Z or Beta). </span>
<span class="c1">#In this case,the program will calculate Z for you based on A1 being the risk allele</span>
<span class="p">[</span><span class="n">munge_sumstats_no_sign</span><span class="p">]</span>



<span class="c1">#path to summary statistic file</span>
<span class="kn">parameter:</span> <span class="n">sumst</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to Hapmap3 SNPs file, keep all columns (SNP, A1, and A2) for the munge_sumstats program</span>
<span class="kn">parameter:</span> <span class="n">alleles</span> <span class="o">=</span> <span class="s2">&quot;w_hm3.snplist&quot;</span>
<span class="c1">#path to output file</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">munge_sumstats</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">sumstats</span> <span class="p">{</span><span class="n">sumst</span><span class="p">}</span> <span class="o">--</span><span class="n">merge</span><span class="o">-</span><span class="n">alleles</span> <span class="p">{</span><span class="n">alleles</span><span class="p">}</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span> <span class="o">--</span><span class="n">a1</span><span class="o">-</span><span class="n">inc</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="munge-summary-statistics-option-2-no-signed-summary-statistic">
<h2>Munge Summary Statistics (Option 2: No Signed Summary Statistic)<a class="headerlink" href="#munge-summary-statistics-option-2-no-signed-summary-statistic" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># This option is for when the summary statistic file does contain a signed summary statistic (Z or Beta)</span>
<span class="p">[</span><span class="n">munge_sumstats_sign</span><span class="p">]</span>



<span class="c1">#path to summary statistic file</span>
<span class="kn">parameter:</span> <span class="n">sumst</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to Hapmap3 SNPs file, keep all columns (SNP, A1, and A2) for the munge_sumstats program</span>
<span class="kn">parameter:</span> <span class="n">alleles</span> <span class="o">=</span> <span class="s2">&quot;w_hm3.snplist&quot;</span>
<span class="c1">#path to output file</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">munge_sumstats</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">sumstats</span> <span class="p">{</span><span class="n">sumst</span><span class="p">}</span> <span class="o">--</span><span class="n">merge</span><span class="o">-</span><span class="n">alleles</span> <span class="p">{</span><span class="n">alleles</span><span class="p">}</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-ld-scores">
<h2>Calculate LD Scores<a class="headerlink" href="#calculate-ld-scores" title="Permalink to this heading">#</a></h2>
<p><strong>Make sure to delete SNP,CHR, and BP columns from annotation files if they are present otherwise this code will not work. Before deleting, if these columns are present, make sure that the annotation file is sorted.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculate LD Scores</span>
<span class="c1">#**Make sure to delete SNP,CHR, and BP columns from annotation files if they are present otherwise this code will not work. Before deleting, if these columns are present, make sure that the annotation file is sorted.**</span>
<span class="p">[</span><span class="n">calc_ld_score</span><span class="p">]</span>

<span class="c1">#Path to bim file</span>
<span class="kn">parameter:</span> <span class="n">bim</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Path to annotation File. Make sure to remove the SNP, CHR, and BP columns from the annotation file if present before running.</span>
<span class="kn">parameter:</span> <span class="n">annot_file</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#name of output file</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to Hapmap3 SNPs file, remove the A1 and A2 columns for the Calculate LD Scores program </span>
<span class="kn">parameter:</span> <span class="n">snplist</span> <span class="o">=</span> <span class="s2">&quot;w_hm3.snplist&quot;</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">ldsc</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">bfile</span> <span class="p">{</span><span class="n">bim</span><span class="p">}</span> <span class="o">--</span><span class="n">l2</span> <span class="o">--</span><span class="n">ld</span><span class="o">-</span><span class="n">wind</span><span class="o">-</span><span class="n">cm</span> <span class="mi">1</span> <span class="o">--</span><span class="n">annot</span> <span class="p">{</span><span class="n">annot_file</span><span class="p">}</span> <span class="o">--</span><span class="n">thin</span><span class="o">-</span><span class="n">annot</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span> <span class="o">--</span><span class="nb">print</span><span class="o">-</span><span class="n">snps</span> <span class="p">{</span><span class="n">snplist</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-functional-enrichment-using-annotations">
<h2>Calculate Functional Enrichment using Annotations<a class="headerlink" href="#calculate-functional-enrichment-using-annotations" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculate Enrichment Scores for Functional Annotations</span>
<span class="p">[</span><span class="n">calc_enrichment</span><span class="p">]</span>

<span class="c1">#Path to Summary statistics File</span>
<span class="kn">parameter:</span> <span class="n">sumstats</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Path to Reference LD Scores Files (Base Annotation + Annotation you want to analyze, format like minimal working example)</span>
<span class="kn">parameter:</span> <span class="n">ref_ld</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Path to LD Weight Files (Format like minimal working example)</span>
<span class="kn">parameter:</span> <span class="n">w_ld</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to frequency files (Format like minimal working example)</span>
<span class="kn">parameter:</span> <span class="n">frq_file</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Output name</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">ldsc</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">h2</span> <span class="p">{</span><span class="n">sumstats</span><span class="p">}</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="p">{</span><span class="n">ref_ld</span><span class="p">}</span> <span class="o">--</span><span class="n">w</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="p">{</span><span class="n">w_ld</span><span class="p">}</span> <span class="o">--</span><span class="n">overlap</span><span class="o">-</span><span class="n">annot</span> <span class="o">--</span><span class="n">frqfile</span><span class="o">-</span><span class="nb">chr</span> <span class="p">{</span><span class="n">frq_file</span><span class="p">}</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/enrichment"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="gregor.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">GREGOR enrichment analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="../pecotmr_integration/pecotmr.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">The PECOTMR intergration framework</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-brief-review-on-stratified-ld-score-regression">A brief review on Stratified LD score regression</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ldsc-model">LDSC model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#stratified-ldsc">Stratified LDSC</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#make-annotation-file">Make Annotation File</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#munge-summary-statistics-option-1-no-signed-summary-statistic">Munge Summary Statistics (Option 1: No Signed Summary Statistic)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#munge-summary-statistics-option-2-no-signed-summary-statistic">Munge Summary Statistics (Option 2: No Signed Summary Statistic)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-ld-scores">Calculate LD Scores</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-functional-enrichment-using-annotations">Calculate Functional Enrichment using Annotations</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>