
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Sample level RNA-seq quality control &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Bulk RNA-seq counts normalization" href="bulk_expression_normalization.html" />
    <link rel="prev" title="Quantifying expression from RNA-seq data" href="../calling/RNA_calling.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../bulk_expression.html">
   RNA-seq expression
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="../calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     Data overview
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/molecular_phenotypes/QC/bulk_expression_QC.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/molecular_phenotypes/QC/bulk_expression_QC.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/molecular_phenotypes/QC/bulk_expression_QC.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods-overview">
   Methods overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gtex-v8-qc-procedures">
     GTEx V8 QC procedures
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#our-implementation">
     Our implementation
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#three-statistics-to-identify-outliers">
   Three statistics to identify outliers
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#rle">
     RLE
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#hierarchical-clustering">
     Hierarchical clustering
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#d-statistic">
     D-statistic
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#workflow-implementation">
   Workflow implementation
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="sample-level-rna-seq-quality-control">
<h1>Sample level RNA-seq quality control<a class="headerlink" href="#sample-level-rna-seq-quality-control" title="Permalink to this headline">¶</a></h1>
<p>This pipeline implements RNA-seq QC that identifies and removes genes and samples from the TPM matrix.</p>
<div class="section" id="methods-overview">
<h2>Methods overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">¶</a></h2>
<div class="section" id="gtex-v8-qc-procedures">
<h3>GTEx V8 QC procedures<a class="headerlink" href="#gtex-v8-qc-procedures" title="Permalink to this headline">¶</a></h3>
<p>From <a class="reference external" href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods</a> :</p>
<ol class="simple">
<li><p>RNA-seq expression outliers were identified and excluded using a multidimensional extension of the statistic described in (Wright et al., Nat. Genet. 2014 ). Briefly, for each tissue, read counts from each sample were normalized using size factors calculated with DESeq2 and log-transformed with an offset of 1; genes with a log-transformed value &gt;1 in &gt;10% of samples were selected, and the resulting read counts were centered and unit-normalized. The resulting matrix was then hierarchically clustered (based on average and cosine distance), and a chi2 p-value was calculated based on Mahalanobis distance. Clusters with ≥60% samples with Bonferroni-corrected p-values &lt;0.05 were marked as outliers, and their samples were excluded.</p></li>
<li><p>Samples with &lt;10 million mapped reads were removed.</p></li>
<li><p>Genes were selected based on expression thresholds of &gt;0.1 TPM in at least 20% of samples and ≥6 reads in at least 20% of samples.</p></li>
</ol>
</div>
<div class="section" id="our-implementation">
<h3>Our implementation<a class="headerlink" href="#our-implementation" title="Permalink to this headline">¶</a></h3>
<p>For #1 and #2 above because the GTEx source code is not available, we adopted <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04307-0">some R scripts developed in this paper</a> to perform gene and sample QC using TPM data, which we implements GTEx recommendations. #3 above is implemented in the gene expression normalization module.</p>
</div>
</div>
<div class="section" id="three-statistics-to-identify-outliers">
<h2>Three statistics to identify outliers<a class="headerlink" href="#three-statistics-to-identify-outliers" title="Permalink to this headline">¶</a></h2>
<div class="section" id="rle">
<h3>RLE<a class="headerlink" href="#rle" title="Permalink to this headline">¶</a></h3>
<p>RLE(Relative Log Expression)is assume that most expressions in a sample should be near a mean value, only a few genes have differential expression, which means higher or lower than other genes. To calculate it, for each gene <em>i</em> , calculate it median in <em>N</em> sample as <em>Medi</em> , for each sample <em>j</em> , and expression value <em>eij</em>, count the difference between <em>eij</em> and <em>Medi</em> : <em>deij = eij-Medi</em> , than boxplot each sample base on <em>deij</em> and sort by IQR(interquartile range), the sample with lager IQR is more likely to be an outlier.</p>
</div>
<div class="section" id="hierarchical-clustering">
<h3>Hierarchical clustering<a class="headerlink" href="#hierarchical-clustering" title="Permalink to this headline">¶</a></h3>
<p>We use(1 - spearman’s correlation) as distances between samples and assume samples should be homogeneous, so all samples should have short distances between others and clustered homogeneously. So samples far away others may be outliers. So we first use distances which is 1-spearmen correlation to hierarchically clustered, than use top 100 gene sort by variance to calculate Mahalanobis Distance, then a chi2 p-value will be calculated based on mahalanobis distance. Then clusters with <span class="math notranslate nohighlight">\( \geq 60\%\)</span> samples with Bonferroni-corrected p-values &lt; 0.05 were marked as outliers.</p>
</div>
<div class="section" id="d-statistic">
<h3>D-statistic<a class="headerlink" href="#d-statistic" title="Permalink to this headline">¶</a></h3>
<p>For each sample, D-statistic represent the average correlation between its expression and other samples. A sample with lower D-statistic is more likely to be an outlier.</p>
</div>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>The MWE data can be found on <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1Rv2bWHBbX_tastTh49ToYVDMV6rFP5Wk">Google Drive</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">bulk_expression_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">tpm</span><span class="o">-</span><span class="n">gct</span> <span class="n">data</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">TPM</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">counts</span><span class="o">-</span><span class="n">gct</span> <span class="n">data</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">Counts</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">log</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;Gene expression profiles loaded successfully!&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;19 genes and 1118 samples are loaded from data/mwe.TPM.gct&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;0 genes are filtered, because &gt; 20 </span><span class="si">% s</span><span class="s2">amples have expression values &lt; 0.1&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;19 genes left, saving output.&quot;</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">log</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">ls</span> <span class="n">output</span><span class="o">/*.</span><span class="n">pdf</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">cluster</span><span class="o">.</span><span class="n">pdf</span>  <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">D_stat_hist</span><span class="o">.</span><span class="n">pdf</span>  <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">preQC_cluster</span><span class="o">.</span><span class="n">pdf</span>  <span class="n">output</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tpm</span><span class="o">.</span><span class="n">gct</span><span class="o">.</span><span class="n">RLEplot</span><span class="o">.</span><span class="n">pdf</span>
</pre></div>
</div>
</div>
<div class="section" id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">bulk_expression_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run bulk_expression_QC.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  qc

Global Workflow Options:
  --tpm-gct VAL (as path, required)
                        Required input is TPM file
  --counts-gct . (as path)
                        Raw counts file is optional and if available, it will be
                        filtered to match with the TPM file sample and genes
  --cwd output (as path)
  --container &#39;&#39;

Sections
  qc_1:
    Workflow Options:
      --low-expr-TPM 0.1 (as float)
      --low-expr-TPM-percent 0.2 (as float)
  qc_2:
    Workflow Options:
      --RLEFilterPercent 0.05 (as float)
      --DSFilterPercent 0.05 (as float)
      --topk-genes 100 (as int)
      --cluster-percent 0.6 (as float)
      --pvalue-cutoff 0.05 (as float)
      --cluster-level 5 (as int)
  qc_3:
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="workflow-implementation">
<h2>Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Required input is TPM file</span>
<span class="kn">parameter:</span> <span class="n">tpm_gct</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Raw counts file is optional and if available, it will be filtered to match with the TPM file sample and genes</span>
<span class="kn">parameter:</span> <span class="n">counts_gct</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span><span class="p">{(</span><span class="s1">&#39;micromamba run -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">])</span> <span class="k">if</span> <span class="n">container</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.sif&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;&#39;</span><span class="p">}</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>

<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">qc_1</span> <span class="p">(</span><span class="n">basic</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">low</span> <span class="n">expression</span> <span class="n">filtering</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">low_expr_TPM</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">low_expr_TPM_percent</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="kn">input:</span> <span class="n">tpm_gct</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.low_expression_filtered.tpm.gct.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.log&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="c1"># Load packages</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">purrr</span><span class="p">)</span>
    <span class="c1">## Setting inputs</span>
    <span class="n">low_expr_TPM</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">[</span><span class="n">low_expr_TPM</span><span class="p">]</span>
    <span class="n">low_expr_TPM_percent</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">[</span><span class="n">low_expr_TPM_percent</span><span class="p">]</span>
    <span class="c1">## QC</span>
    <span class="c1">### Data integrity check. The input file should has sample ID as column name; the first column is gene ID</span>
    <span class="n">TPM_data</span> <span class="o">&lt;-</span> <span class="n">read_tsv</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">ar</span><span class="p">],</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
    <span class="n">names</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;feature&quot;</span>
    <span class="k">if</span><span class="p">(</span> <span class="nb">sum</span><span class="p">(</span><span class="n">duplicated</span><span class="p">(</span><span class="n">TPM_data</span><span class="err">$</span><span class="n">feature</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">){</span>
      <span class="n">message</span><span class="p">(</span><span class="s2">&quot;feature (e.g. gene names) should be in the first column. Please remove duplicated feature IDs, Exit!&quot;</span><span class="p">)</span>
      <span class="n">quit</span><span class="p">(</span><span class="n">save</span> <span class="o">=</span> <span class="s2">&quot;no&quot;</span><span class="p">,</span> <span class="n">status</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">runLast</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="p">}</span><span class="k">else</span><span class="p">{</span>
      <span class="n">rownames</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">TPM_data</span><span class="err">$</span><span class="n">feature</span>
      <span class="n">TPM_data</span> <span class="o">=</span> <span class="n">TPM_data</span><span class="p">[,</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
      <span class="n">loaded_sample_count</span> <span class="o">&lt;-</span> <span class="n">ncol</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="c1">#### NA check</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">){</span>      
      <span class="n">message</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;NA is not allowed in the data, there are &quot;</span><span class="p">,</span><span class="nb">sum</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)),</span><span class="s2">&quot; NAs, Exit!&quot;</span><span class="p">))</span>
      <span class="n">quit</span><span class="p">(</span><span class="n">save</span> <span class="o">=</span> <span class="s2">&quot;no&quot;</span><span class="p">,</span> <span class="n">status</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">runLast</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)}</span>
    <span class="c1">#### make sure every expr column is in numeric type</span>
    <span class="n">matrix_check</span> <span class="o">&lt;-</span> <span class="nb">map</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">,</span> <span class="ow">is</span><span class="o">.</span><span class="n">numeric</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">unlist</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="err">!</span><span class="n">matrix_check</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">){</span>
      <span class="n">message</span><span class="p">(</span><span class="s2">&quot;The following column(s) in expression matrix is/are NOT in numeric type. Plese check, Proceed by excluding those samples&quot;</span><span class="p">)</span>
      <span class="n">message</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">names</span><span class="p">(</span><span class="n">matrix_check</span><span class="p">)[</span><span class="err">!</span><span class="n">matrix_check</span><span class="p">],</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;; &quot;</span><span class="p">))</span>
      <span class="n">TPM_data</span> <span class="o">=</span> <span class="n">TPM_data</span><span class="p">[,</span><span class="n">matrix_check</span><span class="p">]</span>
    <span class="p">}</span>
    <span class="n">message</span><span class="p">(</span><span class="s2">&quot;Gene expression profiles loaded successfully!&quot;</span><span class="p">)</span>
    <span class="n">message</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">),</span> <span class="s2">&quot;genes and&quot;</span><span class="p">,</span> <span class="n">ncol</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">),</span> <span class="s2">&quot;samples are loaded from&quot;</span><span class="p">,</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">],</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
    
    <span class="c1">#### Filter out low exp genes</span>
    <span class="n">keep_genes_idx</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">rowMeans</span><span class="p">(</span><span class="n">TPM_data</span><span class="o">&gt;</span><span class="n">low_expr_TPM</span><span class="p">)</span><span class="o">&gt;</span><span class="n">low_expr_TPM_percent</span><span class="p">)</span> 
    <span class="n">TPM_data</span> <span class="o">=</span> <span class="n">TPM_data</span><span class="p">[</span><span class="n">keep_genes_idx</span><span class="p">,]</span>
    <span class="n">message</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">keep_genes_idx</span><span class="p">),</span> <span class="s2">&quot;genes are filtered by filter: &gt;&quot;</span><span class="p">,</span> <span class="n">low_expr_TPM_percent</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">% s</span><span class="s2">amples have expression values &lt;&quot;</span><span class="p">,</span> <span class="n">low_expr_TPM</span><span class="p">))</span>
    <span class="n">message</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">keep_genes_idx</span><span class="p">),</span> <span class="s2">&quot;genes left, saving output to $[_output].&quot;</span><span class="p">))</span>
    <span class="n">TPM_data</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">(</span><span class="n">rownames</span> <span class="o">=</span> <span class="s2">&quot;gene_ID&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$[_output]&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>CAUTION: I (GW) am not sure what to do with the offset for log-transformation on TPM. GTEx suggests using offset = 1. Here I keep the recommendation from <a class="reference external" href="https://github.com/stormlovetao/eQTLQC/blob/master/Sample/src/report.Rmd">these authors</a> where both 0.0001 and 1 were used in different steps.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">qc_2</span> <span class="p">(</span><span class="n">remove</span> <span class="n">outliers</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">RLEFilterPercent</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="kn">parameter:</span> <span class="n">DSFilterPercent</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="kn">parameter:</span> <span class="n">topk_genes</span> <span class="o">=</span> <span class="mi">100</span>
<span class="kn">parameter:</span> <span class="n">cluster_percent</span> <span class="o">=</span> <span class="mf">0.6</span>
<span class="kn">parameter:</span> <span class="n">pvalue_cutoff</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="kn">parameter:</span> <span class="n">cluster_level</span> <span class="o">=</span> <span class="mi">5</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.outlier_removed.tpm.gct.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.log&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="c1">## Setting parameters</span>
    <span class="n">RLEFilterPercent</span> <span class="o">&lt;-</span><span class="err">$</span><span class="p">[</span><span class="n">RLEFilterPercent</span><span class="p">]</span>
    <span class="n">DSFilterPercent</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">[</span><span class="n">DSFilterPercent</span><span class="p">]</span>
    <span class="c1">### Hcluster parameter</span>
    <span class="n">pvalues</span><span class="o">.</span><span class="n">cut</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">[</span><span class="n">pvalue_cutoff</span><span class="p">]</span>
    <span class="n">topk_genes</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">[</span><span class="n">topk_genes</span><span class="p">]</span>
    <span class="n">cluster_percent</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">[</span><span class="n">cluster_percent</span><span class="p">]</span>
    <span class="n">treesNum</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">[</span><span class="n">cluster_level</span><span class="p">]</span>
    <span class="c1">## Define functions to accomodate rank deficient covariance matrices in https://github.com/cumc/xqtl-pipeline/issues/307</span>
    <span class="n">mahalanobis</span> <span class="o">=</span> <span class="n">function</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">center</span><span class="p">,</span> <span class="n">cov</span><span class="p">,</span> <span class="n">inverted</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span> 
        <span class="p">{</span>
            <span class="n">x</span> <span class="o">&lt;-</span> <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">vector</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> 
                <span class="n">matrix</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ncol</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
            <span class="k">else</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="n">isFALSE</span><span class="p">(</span><span class="n">center</span><span class="p">))</span> 
                <span class="n">x</span> <span class="o">&lt;-</span> <span class="n">sweep</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">2</span><span class="n">L</span><span class="p">,</span> <span class="n">center</span><span class="p">)</span>
            <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="n">inverted</span><span class="p">)</span> 
                <span class="n">cov</span> <span class="o">&lt;-</span> <span class="n">MASS</span><span class="p">::</span><span class="n">ginv</span><span class="p">(</span><span class="n">cov</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
            <span class="n">setNames</span><span class="p">(</span><span class="n">rowSums</span><span class="p">(</span><span class="n">x</span> <span class="o">%*%</span> <span class="n">cov</span> <span class="o">*</span> <span class="n">x</span><span class="p">),</span> <span class="n">rownames</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="p">}</span>
          
    <span class="c1">## laod Data</span>
    <span class="n">TPM_data</span> <span class="o">&lt;-</span> <span class="n">read_tsv</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">ar</span><span class="p">],</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">TPM_data</span><span class="err">$</span><span class="n">gene_ID</span>
    <span class="n">TPM_data</span> <span class="o">=</span> <span class="n">TPM_data</span><span class="p">[,</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">RLEFilterLength</span> <span class="o">&lt;-</span> <span class="n">RLEFilterPercent</span><span class="o">*</span><span class="n">ncol</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)</span>
    <span class="n">DSFilter</span> <span class="o">&lt;-</span> <span class="n">DSFilterPercent</span><span class="o">*</span><span class="n">ncol</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)</span>
  
    <span class="c1">## Outlier detection</span>
  
    <span class="c1">### RLE</span>
    <span class="c1"># https://github.com/stormlovetao/eQTLQC/blob/86dcc388c8da7f1bd5b223f4b9b26f09c907eb15/Sample/src/report.Rmd#L71</span>
    <span class="n">log_offset</span> <span class="o">&lt;-</span> <span class="mf">0.0001</span>
    <span class="n">logtpm</span> <span class="o">=</span> <span class="n">log10</span><span class="p">(</span><span class="n">TPM_data</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span> <span class="o">+</span> <span class="n">log_offset</span><span class="p">)</span>
    <span class="n">rle</span><span class="o">=</span><span class="n">logtpm</span><span class="o">-</span><span class="n">apply</span><span class="p">(</span><span class="n">logtpm</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">median</span><span class="p">)</span> <span class="c1"># change &quot;/&quot; to &quot;-&quot; so that we got log(fold-change) which centered on 0 on the RLE plot.</span>
    <span class="n">iqr</span> <span class="o">=</span> <span class="n">apply</span><span class="p">(</span><span class="n">rle</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="n">IQR</span><span class="p">)</span>
    <span class="n">rle</span><span class="o">=</span><span class="n">melt</span><span class="p">(</span> <span class="n">rle</span> <span class="p">,</span> <span class="n">variable</span><span class="o">.</span><span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;Sample&quot;</span><span class="p">,</span><span class="n">value</span><span class="o">.</span><span class="kp">name</span> <span class="o">=</span><span class="s2">&quot;TPM&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;feature&quot;</span><span class="p">)</span>
    <span class="n">names</span><span class="p">(</span><span class="n">rle</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;feature&quot;</span><span class="p">,</span><span class="s2">&quot;Sample&quot;</span><span class="p">,</span><span class="s2">&quot;TPM&quot;</span><span class="p">)</span>
    <span class="n">rle_IQR</span> <span class="o">&lt;-</span> <span class="n">rle</span> <span class="o">%&gt;%</span> <span class="n">group_by</span><span class="p">(</span><span class="n">Sample</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">summarise</span><span class="p">(</span><span class="n">IQR</span> <span class="o">=</span> <span class="n">IQR</span><span class="p">(</span><span class="n">TPM</span><span class="p">))</span>
    <span class="n">rle_IQR_range</span> <span class="o">&lt;-</span> <span class="n">rle_IQR</span><span class="err">$</span><span class="n">IQR</span> <span class="o">%&gt;%</span> <span class="nb">range</span> <span class="o">%&gt;%</span> <span class="nb">abs</span><span class="p">()</span> <span class="o">%&gt;%</span> <span class="nb">max</span><span class="p">()</span>
    <span class="n">rle_IQR_range</span> <span class="o">&lt;-</span> <span class="mi">2</span><span class="o">*</span><span class="n">rle_IQR_range</span> <span class="o">%&gt;%</span> <span class="n">ceiling</span><span class="p">()</span>
    <span class="n">bymedian</span> <span class="o">&lt;-</span> <span class="k">with</span><span class="p">(</span><span class="n">rle</span><span class="p">,</span> <span class="n">reorder</span><span class="p">(</span><span class="n">Sample</span><span class="p">,</span> <span class="n">TPM</span><span class="p">,</span> <span class="n">IQR</span><span class="p">))</span>  <span class="c1"># sort by IQR</span>
    <span class="n">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span>
    <span class="n">pdf</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output:n].RLEplot.pdf&quot;</span><span class="p">)</span>
    <span class="n">boxplot</span><span class="p">(</span><span class="n">TPM</span> <span class="o">~</span> <span class="n">bymedian</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">rle</span><span class="p">,</span> <span class="n">outline</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">ylim</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="o">-</span><span class="n">rle_IQR_range</span><span class="p">,</span> <span class="n">rle_IQR_range</span><span class="p">),</span> <span class="n">las</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">boxwex</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">cex</span><span class="o">.</span><span class="n">axis</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s2">&quot;RLE plot before QC&quot;</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">ylab</span><span class="o">=</span><span class="s2">&quot;Residual expression levels&quot;</span><span class="p">,</span> <span class="n">frame</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    <span class="n">ExpPerSample</span> <span class="o">&lt;-</span> <span class="n">nrow</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">)</span>
    <span class="n">RLEFilterList</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">bymedian</span><span class="p">)[((</span><span class="n">length</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">bymedian</span><span class="p">))</span><span class="o">-</span><span class="p">(</span><span class="n">RLEFilterLength</span><span class="p">))</span><span class="o">+</span><span class="mi">1</span><span class="p">):(</span><span class="n">length</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">bymedian</span><span class="p">))</span><span class="o">+</span><span class="mi">1</span><span class="p">)])</span>
    <span class="n">RLEFilterList</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">RLEFilterList</span><span class="p">)</span>
    <span class="n">message</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;The right most &quot;</span><span class="p">,</span> <span class="n">RLEFilterPercent</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">% s</span><span class="s2">amples (N = &quot;</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">RLEFilterList</span><span class="p">),</span> <span class="s2">&quot;) are marked as candidate outliers in this step:&quot;</span><span class="p">)</span> <span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">RLEFilterList</span><span class="p">)){</span>
      <span class="n">message</span><span class="p">(</span><span class="n">RLEFilterList</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
    <span class="p">}</span>

    <span class="c1">### hcluster  </span>
    <span class="n">sampleDists</span> <span class="o">&lt;-</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">cor</span><span class="p">(</span><span class="n">logtpm</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;spearman&#39;</span><span class="p">)</span>
    <span class="n">hc</span> <span class="o">&lt;-</span> <span class="n">hclust</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">dist</span><span class="p">(</span><span class="n">sampleDists</span><span class="p">),</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;complete&quot;</span><span class="p">)</span>
    <span class="n">hcphy</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">phylo</span><span class="p">(</span><span class="n">hc</span><span class="p">)</span>
  
    <span class="n">pdf</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output:n].preQC_cluster.pdf&quot;</span><span class="p">)</span>
    <span class="n">plot</span><span class="p">(</span><span class="n">hcphy</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s2">&quot;unrooted&quot;</span><span class="p">,</span> <span class="n">cex</span><span class="o">=.</span><span class="mi">2</span><span class="p">,</span> <span class="n">lab4ut</span><span class="o">=</span><span class="s1">&#39;axial&#39;</span><span class="p">,</span><span class="n">underscore</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s2">&quot;Sample clustering before QC (Spearman - Cor.)&quot;</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    <span class="c1"># https://github.com/stormlovetao/eQTLQC/blob/86dcc388c8da7f1bd5b223f4b9b26f09c907eb15/Sample/src/report.Rmd#L102</span>
    <span class="n">log_offset</span> <span class="o">&lt;-</span> <span class="mi">1</span>
    <span class="n">logtpm</span> <span class="o">=</span> <span class="n">log10</span><span class="p">(</span><span class="n">TPM_data</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span> <span class="o">+</span> <span class="n">log_offset</span><span class="p">)</span>
    <span class="n">ntop</span> <span class="o">&lt;-</span> <span class="n">topk_genes</span>
    <span class="n">Pvars</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">logtpm</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">var</span><span class="p">)</span>
    <span class="n">select</span> <span class="o">&lt;-</span> <span class="n">order</span><span class="p">(</span><span class="n">Pvars</span><span class="p">,</span> <span class="n">decreasing</span> <span class="o">=</span><span class="n">TRUE</span><span class="p">)[</span><span class="n">seq_len</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">ntop</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">Pvars</span><span class="p">)))]</span>
    <span class="n">MD_matrix</span> <span class="o">&lt;-</span> <span class="n">logtpm</span><span class="p">[</span><span class="n">select</span><span class="p">,</span> <span class="p">]</span>
    <span class="n">MahalanobisDistance</span> <span class="o">=</span> <span class="n">mahalanobis</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">MD_matrix</span><span class="p">),</span> <span class="n">colMeans</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">MD_matrix</span><span class="p">)),</span> <span class="n">cov</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">MD_matrix</span><span class="p">)))</span> 
    <span class="c1"># Note: t(MD_matrix)=sample_row*gene_column, Manhalanobis() returns one vector with length=row number</span>
    <span class="n">pvalues</span> <span class="o">=</span> <span class="n">pchisq</span><span class="p">(</span><span class="n">MahalanobisDistance</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">nrow</span><span class="p">(</span><span class="n">MD_matrix</span><span class="p">),</span> <span class="n">lower</span><span class="o">.</span><span class="n">tail</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
    <span class="n">pvalues</span><span class="o">.</span><span class="n">adjust</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="n">pvalues</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span><span class="s2">&quot;bonferroni&quot;</span><span class="p">)</span> <span class="c1"># adjusted pvalues for each sample</span>
    <span class="n">pvalues</span><span class="o">.</span><span class="n">low</span> <span class="o">&lt;-</span> <span class="n">pvalues</span><span class="o">.</span><span class="n">adjust</span><span class="p">[</span><span class="n">pvalues</span><span class="o">.</span><span class="n">adjust</span><span class="o">&lt;</span><span class="n">pvalues</span><span class="o">.</span><span class="n">cut</span><span class="p">]</span>
    
    <span class="n">HCoutliers</span> <span class="o">&lt;-</span> <span class="n">character</span><span class="p">()</span>
    <span class="k">for</span><span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">treesNum</span><span class="p">)){</span>
      <span class="n">trees</span> <span class="o">&lt;-</span> <span class="n">cutree</span><span class="p">(</span><span class="n">hc</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="n">x</span><span class="p">)</span>
      <span class="n">idx</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">x</span><span class="p">)</span><span class="c1">#which tree is checking</span>
      <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="n">idx</span><span class="p">)</span>
      <span class="p">{</span>
        <span class="n">group</span> <span class="o">&lt;-</span> <span class="n">hc</span><span class="err">$</span><span class="n">labels</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">trees</span> <span class="o">==</span> <span class="n">i</span><span class="p">)]</span>
        <span class="k">if</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">group</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">pvalues</span><span class="o">.</span><span class="n">low</span><span class="p">))</span><span class="o">/</span><span class="n">length</span><span class="p">(</span><span class="n">group</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">cluster_percent</span><span class="p">)</span>
        <span class="p">{</span>
          <span class="n">HCoutliers</span> <span class="o">&lt;-</span> <span class="n">union</span><span class="p">(</span><span class="n">HCoutliers</span><span class="p">,</span><span class="n">group</span><span class="p">)</span>
        <span class="p">}</span>
      <span class="p">}</span>
    <span class="p">}</span>
    
    <span class="n">message</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">HCoutliers</span><span class="p">),</span> <span class="s2">&quot;samples are marked as candidate outlier(s) in this step.&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">HCoutliers</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">){</span>
      <span class="n">message</span><span class="p">(</span><span class="s2">&quot;Sample outliers are marked in red as follows:&quot;</span><span class="p">)</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">HCoutliers</span><span class="p">)){</span>
          <span class="n">message</span><span class="p">(</span><span class="n">HCoutliers</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
      <span class="p">}</span>
      <span class="n">co1</span> <span class="o">=</span> <span class="n">hc</span><span class="err">$</span><span class="n">labels</span><span class="o">%</span><span class="ow">in</span><span class="o">%</span><span class="n">HCoutliers</span>
      <span class="n">co1</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">co1</span> <span class="o">==</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">)]</span><span class="o">&lt;-</span><span class="s2">&quot;gray0&quot;</span>
      <span class="n">co1</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">co1</span> <span class="o">==</span> <span class="s2">&quot;TRUE&quot;</span><span class="p">)]</span><span class="o">&lt;-</span><span class="s2">&quot;red&quot;</span>
      <span class="n">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span>
  
    <span class="n">pdf</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output:n].cluster.pdf&quot;</span><span class="p">)</span>
    <span class="n">plot</span><span class="p">(</span><span class="n">hcphy</span><span class="p">,</span> <span class="n">tip</span><span class="o">.</span><span class="n">col</span> <span class="o">=</span> <span class="n">co1</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s2">&quot;unrooted&quot;</span><span class="p">,</span> <span class="n">cex</span><span class="o">=.</span><span class="mi">2</span><span class="p">,</span> <span class="n">lab4ut</span><span class="o">=</span><span class="s1">&#39;axial&#39;</span><span class="p">,</span><span class="n">underscore</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s2">&quot;Label Outliers in Red&quot;</span><span class="p">)</span>
    <span class="n">Xcol</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;gray0&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">)</span>
    <span class="n">Xtext</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;Normal Sample&quot;</span><span class="p">,</span> <span class="s2">&quot;Outliers&quot;</span><span class="p">)</span>
    <span class="n">legend</span><span class="p">(</span><span class="s1">&#39;bottomleft&#39;</span><span class="p">,</span><span class="n">pch</span><span class="o">=</span><span class="mi">21</span><span class="p">,</span><span class="n">Xtext</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span><span class="n">pt</span><span class="o">.</span><span class="n">bg</span><span class="o">=</span><span class="n">Xcol</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    <span class="p">}</span><span class="k">else</span><span class="p">{</span>
      <span class="n">message</span><span class="p">(</span><span class="s2">&quot;No outlier detected.&quot;</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="c1">### D-s</span>
    <span class="n">D</span> <span class="o">=</span> <span class="n">apply</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">sampleDists</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">median</span><span class="p">)</span>
    <span class="n">pdf</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output:n].D_stat_hist.pdf&quot;</span><span class="p">)</span>
    <span class="n">hist</span><span class="p">(</span><span class="n">D</span><span class="p">,</span> <span class="n">breaks</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">ylab</span><span class="o">=</span><span class="s2">&quot;Number of samples&quot;</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;D-statistic&quot;</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s2">&quot;Histogram of Sample D-statistics before data QC&quot;</span><span class="p">)</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>
    
    <span class="n">DSFilter</span> <span class="o">&lt;-</span> <span class="n">sort</span><span class="p">(</span><span class="n">D</span><span class="p">)[</span><span class="n">DSFilter</span><span class="p">]</span>
    <span class="n">D</span><span class="o">&lt;-</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
    <span class="n">D</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Sample</span> <span class="o">=</span> <span class="n">rownames</span><span class="p">(</span><span class="n">D</span><span class="p">),</span><span class="n">D</span> <span class="o">=</span> <span class="n">D</span><span class="err">$</span><span class="n">D</span><span class="p">)</span>
    <span class="n">D_filterList</span> <span class="o">=</span> <span class="n">D</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="n">D</span> <span class="o">&lt;=</span> <span class="n">DSFilter</span><span class="p">)</span>
    <span class="n">D_filterList</span> <span class="o">&lt;-</span> <span class="n">D_filterList</span><span class="err">$</span><span class="n">Sample</span>
    <span class="n">D_filterList</span><span class="o">&lt;-</span><span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">D_filterList</span><span class="p">)</span>
    <span class="n">message</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;The right most &quot;</span><span class="p">,</span> <span class="n">DSFilterPercent</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">% s</span><span class="s2">amples (N=&quot;</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">D_filterList</span><span class="p">),</span> <span class="s2">&quot;) are marked as candidate outliers in this step:&quot;</span><span class="p">)</span> <span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">D_filterList</span><span class="p">)){</span>
      <span class="n">message</span><span class="p">(</span><span class="n">D_filterList</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
    <span class="p">}</span>
    
    <span class="c1">## Outliers are the intersect of three candidates list</span>
    <span class="n">outliersList</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">RLEFilterList</span><span class="p">,</span><span class="n">intersect</span><span class="p">(</span><span class="n">HCoutliers</span><span class="p">,</span><span class="n">D_filterList</span><span class="p">))</span>
    <span class="n">message</span><span class="p">(</span><span class="s2">&quot;Outliers:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">outliersList</span><span class="p">)){</span>
      <span class="n">message</span><span class="p">(</span><span class="n">outliersList</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
    <span class="p">}</span>
    <span class="n">outliersIndex</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">logtpm</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">outliersList</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="err">!</span><span class="n">length</span><span class="p">(</span><span class="n">outliersIndex</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">){</span>
        <span class="n">TPM_data</span> <span class="o">&lt;-</span> <span class="n">TPM_data</span><span class="p">[,</span><span class="o">-</span><span class="n">outliersIndex</span><span class="p">]</span>
    <span class="p">}</span>
    <span class="c1">## Add 2 header lines, https://github.com/getzlab/rnaseqc/blob/286f99dfd4164d33014241dd4f3149da0cddf5bf/src/RNASeQC.cpp#L426</span>
    <span class="n">cat</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;#1.2</span><span class="se">\n</span><span class="s2">#&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">),</span> <span class="n">ncol</span><span class="p">(</span><span class="n">TPM_data</span><span class="p">),</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">file</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">],</span> <span class="n">append</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">TPM_data</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">(</span><span class="n">rownames</span> <span class="o">=</span> <span class="s2">&quot;gene_ID&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">],</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">col_names</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
  
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.log&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">gzip</span> <span class="o">-</span><span class="n">f</span> <span class="o">--</span><span class="n">best</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Additional formatting:</p>
<ol class="simple">
<li><p>Filter out the geneCount table based on TPM table.</p></li>
<li><p>Adds two comment lines above the header of TPM and geneCount table to mimick the original output from RNASeQC.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">qc_3</span> <span class="p">(</span><span class="n">remove</span> <span class="n">gene</span> <span class="ow">and</span> <span class="n">samples</span> <span class="kn">from</span> <span class="nn">raw</span> <span class="n">counts</span><span class="p">)]</span>
<span class="n">stop_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">counts_gct</span><span class="o">.</span><span class="n">is_file</span><span class="p">())</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.geneCount.gct.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="c1"># Reason to use read.table: 1.accomodate both &quot; &quot; and &quot;\t&quot;</span>
    <span class="n">tpm</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">ar</span><span class="p">],</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span>
    <span class="n">geneCount</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">counts_gct</span><span class="p">:</span><span class="n">ar</span><span class="p">],</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">,</span><span class="n">col_names</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span>
    <span class="c1">## Make geneCount consistant with tpm, remove gene by filter() and remove samples by select()</span>
    <span class="n">geneCount</span> <span class="o">=</span> <span class="n">geneCount</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="n">gene_ID</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">tpm</span><span class="err">$</span><span class="n">gene_ID</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">tpm</span><span class="p">))</span>
    <span class="c1">## Add 2 header lines, https://github.com/getzlab/rnaseqc/blob/286f99dfd4164d33014241dd4f3149da0cddf5bf/src/RNASeQC.cpp#L426</span>
    <span class="n">cat</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;#1.2</span><span class="se">\n</span><span class="s2">#&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="p">(</span><span class="n">geneCount</span><span class="p">),</span> <span class="n">ncol</span><span class="p">(</span><span class="n">geneCount</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">file</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">],</span> <span class="n">append</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="c1">## Save each file with 3 header line</span>
    <span class="n">geneCount</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">],</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">col_names</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.log&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">gzip</span> <span class="o">-</span><span class="n">f</span> <span class="o">--</span><span class="n">best</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/molecular_phenotypes/QC"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../calling/RNA_calling.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Quantifying expression from RNA-seq data</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="bulk_expression_normalization.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Bulk RNA-seq counts normalization</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>