
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Normalization and phenotype table generation for splicingQTL analysis &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/molecular_phenotypes/QC/splicing_normalization';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Genotype data preprocessing" href="../../data_preprocessing/genotype_preprocessing.html" />
    <link rel="prev" title="Quantifying alternative splicing from RNA-seq data" href="../calling/splicing_calling.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../splicing.html">Alternative splicing from RNA-seq data</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/edit/gh-pages/code/molecular_phenotypes/QC/splicing_normalization.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/molecular_phenotypes/QC/splicing_normalization.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/molecular_phenotypes/QC/splicing_normalization.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Normalization and phenotype table generation for splicingQTL analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#leafcutter"><code class="docutils literal notranslate"><span class="pre">leafcutter</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#psichomics"><code class="docutils literal notranslate"><span class="pre">psichomics</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1"><code class="docutils literal notranslate"><span class="pre">leafcutter</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2"><code class="docutils literal notranslate"><span class="pre">psichomics</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-splicing-qc-and-normalization">ii. Splicing QC and Normalization</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-leafcutter">a. Leafcutter</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-psichomics">a. Psichomics</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-sample-list">0. Generate sample list</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#leafcutter-norm"><code class="docutils literal notranslate"><span class="pre">leafcutter_norm</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#parameter-annotations">Parameter Annotations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#things-to-keep-in-mind">Things to keep in mind</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#psichomics-norm"><code class="docutils literal notranslate"><span class="pre">psichomics_norm</span></code></a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="normalization-and-phenotype-table-generation-for-splicingqtl-analysis">
<h1>Normalization and phenotype table generation for splicingQTL analysis<a class="headerlink" href="#normalization-and-phenotype-table-generation-for-splicingqtl-analysis" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>Quality control and normalization are performed on output from the leafcutter and psichomics tools. The raw output data is first converted to bed format. Quality control involves the removal of features with high missingness across samples (default 40%) and any remaining missing values are retained. Then introns with less than a minimal variation (default of 0.005) are removed from the data. Quantile-Quantile normalization is performed on the quality controlled data. Imputation of missing values is done separately afterwards.</p>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<section id="leafcutter">
<h3><code class="docutils literal notranslate"><span class="pre">leafcutter</span></code><a class="headerlink" href="#leafcutter" title="Link to this heading">#</a></h3>
<p>The sample_list_intron_usage_perind.counts.gz file generated by previous splicing_calling.ipynb.</p>
<p>Minimal working example input may be found at <a class="reference external" href="https://drive.google.com/drive/folders/1lpcx3eKG2UpauntLUuJ6bMBjHyIhWW_R">google drive</a>, specifically the  <code class="docutils literal notranslate"><span class="pre">sample_fastq_bam_list_intron_usage_perind.counts.gz</span></code> file.</p>
</section>
<section id="psichomics">
<h3><code class="docutils literal notranslate"><span class="pre">psichomics</span></code><a class="headerlink" href="#psichomics" title="Link to this heading">#</a></h3>
<p>The psichomics_raw_data.tsv file generated by previous splicing_calling.ipynb.</p>
<p>Minimal working example input may be found at <a class="reference external" href="https://drive.google.com/drive/folders/1lpcx3eKG2UpauntLUuJ6bMBjHyIhWW_R">google drive</a>, specifically the  <code class="docutils literal notranslate"><span class="pre">psi_raw_data.tsv</span></code> file.</p>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<section id="id1">
<h3><code class="docutils literal notranslate"><span class="pre">leafcutter</span></code><a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">{sample_list}</span></code> below refers to the name of the meta-data file input for previous step.</p>
<p>Main output include:</p>
<p><code class="docutils literal notranslate"><span class="pre">{sample_list}_intron_usage_perind.counts.gz_raw_data.qqnorm.txt</span></code> a merged table with normalized intron usage ratio for each sample, ready to be phenotype input for tensorQTL in following format:
` a merged table with normalized intron usage ratio for each sample, ready to be phenotype input for tensorQTL in following format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#Chr       start        end        ID                                                      samp1 samp2 samp3 ...</span>
<span class="n">chromosome</span> <span class="n">intron_start</span> <span class="n">intron_end</span> <span class="p">{</span><span class="nb">chr</span><span class="p">}:{</span><span class="n">intron_start</span><span class="p">}:{</span><span class="n">intron_end</span><span class="p">}:{</span><span class="n">cluster_id</span><span class="p">}</span><span class="n">_</span><span class="p">{</span><span class="n">strand</span><span class="p">}</span> <span class="n">data</span>  <span class="n">data</span>  <span class="n">data</span>  <span class="o">...</span>  
</pre></div>
</div>
<p>(the strand info in “ID” column is calculated strandness of junctions via retools in previous workflow)</p>
</section>
<section id="id2">
<h3><code class="docutils literal notranslate"><span class="pre">psichomics</span></code><a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<p>Main output include:</p>
<p><code class="docutils literal notranslate"><span class="pre">psichomics_raw_data_bedded.qqnorm.txt</span></code> a merged table with normalized intron usage ratio for each sample, ready to be phenotype input for tensorQTL in following format:
` a merged table with normalized percent spliced in (psi) value for each sample, ready to be phenotype input for tensorQTL in following format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#Chr       start        end        ID                                               samp1 samp2 samp3 ...</span>
<span class="n">chromosome</span> <span class="n">intron_start</span> <span class="n">intron_end</span> <span class="p">{</span><span class="n">event_type</span><span class="p">}</span><span class="n">_</span><span class="p">{</span><span class="nb">chr</span><span class="p">}</span><span class="n">_</span><span class="p">{</span><span class="n">strand</span><span class="p">}</span><span class="n">_</span><span class="p">{</span><span class="n">coordinates</span><span class="p">}</span><span class="n">_</span><span class="p">{</span><span class="n">gene</span><span class="p">}</span> <span class="n">data</span>  <span class="n">data</span>  <span class="n">data</span>  <span class="o">...</span>  
</pre></div>
</div>
</section>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<section id="ii-splicing-qc-and-normalization">
<h3>ii. Splicing QC and Normalization<a class="headerlink" href="#ii-splicing-qc-and-normalization" title="Link to this heading">#</a></h3>
<section id="a-leafcutter">
<h4>a. Leafcutter<a class="headerlink" href="#a-leafcutter" title="Link to this heading">#</a></h4>
<p>Timing:  ~30min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run splicing_normalization.ipynb leafcutter_norm \
    --cwd ../../output_test/leafcutter/normalize \
    --ratios ../../output_test/leafcutter/PCC_sample_list_subset_leafcutter_intron_usage_perind.counts.gz \
    --container oras://ghcr.io/cumc/leafcutter_apptainer:latest \
    -s force -c ../csg.yml -q neurology
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">leafcutter_norm_1</span>: 
INFO: t7dd380a023cbb83f <span class=" -Color -Color-Green">re-execute completed</span>
INFO: t7dd380a023cbb83f <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 3983798 (&quot;job_t7dd380a023cbb83f&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">leafcutter_norm_1</span> output:   <span class=" -Color -Color-Green">../../output_test/leafcutter/PCC_sample_list_subset_leafcutter_intron_usage_perind.counts.gz_phenotype_file_list.txt</span>
INFO: Running <span class=" -Color -Color-Green">leafcutter_norm_2</span>: 
INFO: td8bff19d02211b90 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 3983807 (&quot;job_td8bff19d02211b90&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">leafcutter_norm_2</span> output:   <span class=" -Color -Color-Green">../../output_test/leafcutter/PCC_sample_list_subset_leafcutter_intron_usage_perind.counts.gz_raw_data.txt</span>
INFO: Running <span class=" -Color -Color-Green">leafcutter_norm_3</span>: 
INFO: t2b8b4754c0be4cbc <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 3983815 (&quot;job_t2b8b4754c0be4cbc&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">leafcutter_norm_3</span> output:   <span class=" -Color -Color-Green">../../output_test/leafcutter/PCC_sample_list_subset_leafcutter_intron_usage_perind.counts.gz_raw_data.qqnorm.txt</span>
INFO: Workflow leafcutter_norm (ID=w9a20332cfb71c800) is executed successfully with 3 completed steps and 3 completed tasks.
</pre></div>
</div>
</div>
</div>
<p>To perform leafcutter analysis with QC (setting cluster counts of 0 as NA), imputation (flashR), and normalization, use the following commands (RECOMMENDED):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Step 1: QC    </span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">splicing_normalization</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">leafcutter_norm</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">../../</span><span class="n">output_test</span><span class="o">/</span><span class="n">leafcutter</span><span class="o">/</span><span class="n">normalize</span> \
    <span class="o">--</span><span class="n">ratios</span> <span class="o">../../</span><span class="n">output_test</span><span class="o">/</span><span class="n">leafcutter</span><span class="o">/</span><span class="n">PCC_sample_list_subset_leafcutter_intron_usage_perind</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">leafcutter_apptainer</span><span class="p">:</span><span class="n">latest</span> \
    <span class="o">--</span><span class="n">no_norm</span> <span class="c1"># add no norm to skip last step (qqnorm) in leafcutter_norm</span>

<span class="c1"># Step 2: Imputation        </span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">phenotype_imputation</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">EBMF</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="o">&lt;</span><span class="n">output_from_step1</span><span class="o">&gt;</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">../../</span><span class="n">output_test</span><span class="o">/</span><span class="n">leafcutter</span><span class="o">/</span><span class="n">imputation</span> \
    <span class="o">--</span><span class="n">prior</span> <span class="n">ebnm_point_laplace</span> <span class="o">--</span><span class="n">varType</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">factor_analysis_apptainer</span><span class="p">:</span><span class="n">latest</span> \
    <span class="o">--</span><span class="n">mem</span> <span class="mi">40</span><span class="n">G</span> <span class="o">--</span><span class="n">numThreads</span> <span class="mi">20</span> <span class="o">--</span><span class="n">walltime</span> <span class="mi">100</span><span class="n">h</span> 

<span class="c1"># Step 3: Normalization     </span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">splicing_normalization</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">leafcutter_qqnorm</span> \
    <span class="o">--</span><span class="n">qced</span><span class="o">-</span><span class="n">data</span> <span class="o">&lt;</span><span class="n">output_from_step2</span><span class="o">&gt;</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">leafcutter_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
<p>Or if you are going to use the default mean imputation method, you can run it with one-step command (remove  –no_norm in the first step):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">splicing_normalization</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">leafcutter_norm</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">../../</span><span class="n">output_test</span><span class="o">/</span><span class="n">leafcutter</span><span class="o">/</span><span class="n">normalize</span> \
    <span class="o">--</span><span class="n">ratios</span> <span class="o">../../</span><span class="n">output_test</span><span class="o">/</span><span class="n">leafcutter</span><span class="o">/</span><span class="n">PCC_sample_list_subset_leafcutter_intron_usage_perind</span><span class="o">.</span><span class="n">counts</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">leafcutter_apptainer</span><span class="p">:</span><span class="n">latest</span> \
    <span class="o">--</span><span class="n">mean_impute</span>
</pre></div>
</div>
</section>
<section id="a-psichomics">
<h4>a. Psichomics<a class="headerlink" href="#a-psichomics" title="Link to this heading">#</a></h4>
<p>Timing: ~20min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run splicing_normalization.ipynb psichomics_norm \
    --cwd ../../output_test/psichomics/normalize \
    --ratios ../../output_test/psichomics/psi_raw_data.tsv \
    --container oras://ghcr.io/cumc/psichomics_apptainer:latest \
    -c ../csg.yml -q neurology
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">psichomics_norm_1</span>: 
INFO: <span class=" -Color -Color-Green">psichomics_norm_1</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> due to saved signature
INFO: <span class=" -Color -Color-Green">psichomics_norm_1</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> due to saved signature
INFO: <span class=" -Color -Color-Green">psichomics_norm_1</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/output_test/psichomics/normalize/psichomics_raw_data_bedded.txt</span>
INFO: Running <span class=" -Color -Color-Green">psichomics_norm_2</span>: 
INFO: t261c5bcc674fd156 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 4284374 (&quot;job_t261c5bcc674fd156&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">psichomics_norm_2</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/output_test/psichomics/normalize/psichomics_raw_data_bedded.qqnorm.txt</span>
INFO: Workflow psichomics_norm (ID=w414de35dd6b2a7af) is executed successfully with 1 completed step, 1 ignored step and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run splicing_normalization.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run splicing_normalization.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  leafcutter_norm

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files.
  --ratios VAL (as path, required)
                        intron usage ratio file wiht samples after QC
  --job-size 1 (as int)
                        Raw data directory, default to the same directory as
                        sample list parameter: data_dir = path(f&quot;{ratios:d}&quot;)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option

Sections
  leafcutter_norm_1:
    Workflow Options:
      --chr-blacklist  f&#39;{ratios:dd}/black_list.txt&#39;

  leafcutter_norm_2:
</pre></div>
</div>
</div>
</div>
</section>
<section id="setup-and-global-parameters">
<h2>Setup and global parameters<a class="headerlink" href="#setup-and-global-parameters" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# The output directory for generated files. 
parameter: cwd = path(&quot;output&quot;)
# intron usage ratio file wiht samples after QC
# optional parameter black list if user want to blacklist some chromosomes and not to analyze
parameter: chr_blacklist = path(&quot;.&quot;)
# For cluster jobs, number commands to run per job
parameter: job_size = 1
# Wall clock time expected
parameter: walltime = &quot;5h&quot;
# Memory expected
parameter: mem = &quot;16G&quot;
# Number of threads
parameter: numThreads = 8
# Software container option
parameter: container = &quot;&quot;
import re
parameter: entrypoint= (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;
from sos.utils import expand_size
cwd = path(f&#39;{cwd:a}&#39;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="generate-sample-list">
<h2>0. Generate sample list<a class="headerlink" href="#generate-sample-list" title="Link to this heading">#</a></h2>
<p>The rna seq sample IDs are different from the wgs samples. And Hao provided a table for that, but I need to cut that list in a same length with my samples in sQTL analysis.</p>
<section id="step-inputs">
<h3>Step Inputs:<a class="headerlink" href="#step-inputs" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ROSMAP_JointCall_sample_participant_lookup_fixed</span></code>: A sample comparison table for ROSMAP datasets.</p></li>
</ul>
</section>
<section id="step-outputs">
<h3>Step Outputs:<a class="headerlink" href="#step-outputs" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ROSMAP_JointCall_sample_participant_lookup_fixed.rnaseq</span></code>: A sample comparison table for ROSMAP datasets with same length as length with my samples.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[Junc_list]
parameter: junc_path = path
parameter: file_suffix = &quot;junc&quot;
parameter: leafcutter_version = &quot;leafcutter2&quot;
parameter: dataset = &quot;ROSMAP_DLPFC&quot;
output: f&#39;{cwd}/{leafcutter_version}/{dataset}_intron_usage_perind.junc&#39;
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stderr&#39; 
    cd ${cwd}
    realpath  ${junc_path}/*${file_suffix} &gt; ${_output}.tmp
    grep -v &quot;all&quot; ${_output}.tmp &gt; ${_output}
    rm ${_output}.tmp
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[Jointcall_samples]
parameter: sample_table = path
parameter: junc_list = path
parameter: leafcutter_version = &quot;leafcutter2&quot;
input: sample_table, junc_list
output: f&#39;{cwd}/{leafcutter_version}/{sample_table:b}.rnaseq&#39;
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stderr&#39; 
    library(data.table)
    library(tidyverse)
    sample_lookup_hao = read_delim(${_input[0]:r}, &quot;\t&quot; ,col_names = T)
    sample_lookup_hao$sample_id&lt;-sample_lookup_hao$sample_id%&gt;%gsub(&quot;.final&quot;,&quot;&quot;,.)
    sample_ids&lt;-list.files(paste0(&quot;${cwd}&quot;,&quot;/&quot;,&quot;${leafcutter_version}&quot;,&quot;/&quot;),&quot;junc$&quot;)%&gt;%stringr::str_split(.,&quot;[.]&quot;,simplify=T)%&gt;%.[,1]
    junc_list&lt;-read.table(${_input[1]:r})
    sample_ids&lt;-junc_list$V1%&gt;%basename%&gt;%stringr::str_split(.,&quot;[.]&quot;,simplify=T)%&gt;%.[,1]
    sample_lookup_hao&lt;-sample_lookup_hao[sample_lookup_hao$sample_id%in%sample_ids,]
    write.table(sample_lookup_hao,${_output:r},quote = F,row.names = F,sep=&#39;\t&#39;)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="leafcutter-norm">
<h2><code class="docutils literal notranslate"><span class="pre">leafcutter_norm</span></code><a class="headerlink" href="#leafcutter-norm" title="Link to this heading">#</a></h2>
<p>Documentation: <a class="reference external" href="https://davidaknowles.github.io/leafcutter/index.html"><code class="docutils literal notranslate"><span class="pre">leafcutter</span></code></a>. The choices of regtool parameters are <a class="reference external" href="https://github.com/davidaknowles/leafcutter/issues/127">discussed here</a>.</p>
<section id="parameter-annotations">
<h3>Parameter Annotations<a class="headerlink" href="#parameter-annotations" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>chr_blacklist: file of blacklisted chromosomes to exclude from analysis, one per line. If none is provided, will default blacklist nothing.</p></li>
</ul>
</section>
<section id="things-to-keep-in-mind">
<h3>Things to keep in mind<a class="headerlink" href="#things-to-keep-in-mind" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Seems leafcutter_norm_1 requires ~ 10G memory (or larger if having large input) or there will be segmentation fault.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[leafcutter_norm_1]
parameter: ratios = path
parameter: pseudo_ratio = False 
import os
if os.path.isfile(f&#39;{ratios:dd}/black_list.txt&#39;):
    chr_blacklist = f&#39;{ratios:dd}/black_list.txt&#39;
input: ratios, group_by = &#39;all&#39;
output: f&#39;{ratios}_phenotype_file_list.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container=container, entrypoint = entrypoint
    # code in [leafcutter_norm_1] and [leafcutter_norm_3] is modified from 
    # https://github.com/davidaknowles/leafcutter/blob/master/scripts/prepare_phenotype_table.py
    import sys
    import gzip
    import os
    import numpy as np
    import pandas as pd
    import scipy as sc
    import pickle

    from sklearn import linear_model

    def stream_table(f, ss = &#39;&#39;):
        fc = &#39;#&#39;
        while fc[0] == &quot;#&quot;:
            fc = f.readline().strip()
            head = fc.split(ss)

        for ln in f:
            ln = ln.strip().split(ss)
            attr = {}

            for i in range(len(head)):
                try: attr[head[i]] = ln[i]
                except: break
            yield attr

    def get_chromosomes(ratio_file):
        &quot;&quot;&quot;Get chromosomes from table. Returns set of chromosome names&quot;&quot;&quot;
        try: open(ratio_file)
        except:
            sys.stderr.write(&quot;Can&#39;t find %s..exiting\n&quot;%(ratio_file))
            return
        sys.stderr.write(&quot;Parsing chromosome names...\n&quot;)
        chromosomes = set()
        with gzip.open(ratio_file, &#39;rt&#39;) as f:
                f.readline()
                for line in f:
                    chromosomes.add(line.split(&quot;:&quot;)[0])
        return(chromosomes)

    def get_blacklist_chromosomes(chromosome_blacklist_file):
        &quot;&quot;&quot;
        Get list of chromosomes to ignore from a file with one blacklisted
        chromosome per line. Returns list. eg. [&#39;X&#39;, &#39;Y&#39;, &#39;MT&#39;]
        &quot;&quot;&quot;

        if os.path.isfile(chromosome_blacklist_file):
            with open(chromosome_blacklist_file, &#39;r&#39;) as f:
                return(f.read().splitlines())
        else:
            return([])

    def create_phenotype_table(ratio_file, chroms, blacklist_chroms):
        dic_pop, fout = {}, {}
        try: open(ratio_file)
        except:
            sys.stderr.write(&quot;Can&#39;t find %s..exiting\n&quot;%(ratio_file))
            return

        sys.stderr.write(&quot;Starting...\n&quot;)
        for i in chroms:
            fout[i] = open(ratio_file+&quot;.phen_&quot;+i, &#39;w&#39;)
            fout_ave = open(ratio_file+&quot;.ave&quot;, &#39;w&#39;)
        valRows, valRowsnn, geneRows = [], [], []
        finished = False
        header = gzip.open(ratio_file, &#39;rt&#39;).readline().split()[1:]

        for i in fout:
            fout[i].write(&quot;\t&quot;.join([&quot;#chr&quot;,&quot;start&quot;, &quot;end&quot;, &quot;ID&quot;]+header)+&#39;\n&#39;)

        for dic in stream_table(gzip.open(ratio_file, &#39;rt&#39;),&#39; &#39;):

            chrom = dic[&#39;chrom&#39;]
            chr_ = chrom.split(&quot;:&quot;)[0]
            if chr_ in blacklist_chroms: continue
            NA_indices, aveReads = [], []
            tmpvalRow = []

            i = 0
            for sample in header:

                try: count = dic[sample]
                except: print([chrom, len(dic)])
                num, denom = count.split(&#39;/&#39;)
                if float(denom) &lt; 1:
                    count = &quot;NA&quot;
                    tmpvalRow.append(&quot;NA&quot;)
                    NA_indices.append(i)
                else:
                    # add a 0.5 pseudocount
                    if ${pseudo_ratio}:
                        pseudocount = 0.01 * float(denom)
                    else:
                        pseudocount = 0.5
                    count = (float(num)+pseudocount)/((float(denom))+pseudocount)
                    tmpvalRow.append(count)
                    aveReads.append(count)

            parts = chrom.split(&quot;:&quot;)

            # Check the number of parts to determine how to unpack
            if len(parts) == 5:
                chr_, s, e, clu, category = parts #leafcutter2 would have 5 columns, the last column is function category
            elif len(parts) == 4:
                chr_, s, e, clu = parts #leafcutter would have 4 columns
                category = None  # or some default value, if necessary
            else:
                raise ValueError(&quot;Unexpected format of &#39;chrom&#39; string&quot;)

            if len(tmpvalRow) &gt; 0:
                fout[chr_].write(&quot;\t&quot;.join([chr_,s,e,chrom]+[str(x) for x in tmpvalRow])+&#39;\n&#39;)
                fout_ave.write(&quot; &quot;.join([&quot;%s&quot;%chrom]+[str(min(aveReads)), str(max(aveReads)), str(np.mean(aveReads))])+&#39;\n&#39;)

                valRows.append(tmpvalRow)
                geneRows.append(&quot;\t&quot;.join([chr_,s,e,chrom]))
                if len(geneRows) % 1000 == 0:
                    sys.stderr.write(&quot;Parsed %s introns...\n&quot;%len(geneRows))

        for i in fout:
            fout[i].close()

        matrix = np.array(valRows)

        # write the corrected tables

        sample_names = []

        for name in header:
            sample_names.append(name.replace(&#39;.Aligned.sortedByCoord.out.md&#39;, &#39;&#39;))

        fout = {}
        for i in chroms:
            fn=&quot;%s.qqnorm_%s&quot;%(ratio_file,i)
            print(&quot;Outputting: &quot; + fn)
            fout[i] = open(fn, &#39;w&#39;)
            fout[i].write(&quot;\t&quot;.join([&#39;#Chr&#39;,&#39;start&#39;,&#39;end&#39;,&#39;ID&#39;] + sample_names)+&#39;\n&#39;)
        lst = []
        for i in range(len(matrix)):
            chrom, s = geneRows[i].split()[:2]

            lst.append((chrom, int(s), &quot;\t&quot;.join([geneRows[i]] + [str(x) for x in  matrix[i]])+&#39;\n&#39;))

        lst.sort()
        for ln in lst:
            fout[ln[0]].write(ln[2])

        fout_run = open(&quot;%s_phenotype_file_list.txt&quot;%ratio_file, &#39;w&#39;)

        fout_run.write(&quot;#chr\t#dir\n&quot;)

        for i in fout:
            fout[i].close()
            fout_run.write(&quot;%s\t&quot;%(i))
            fout_run.write(&quot;%s.qqnorm_%s\n&quot;%(ratio_file, i))
        fout_run.close()

    ratio_file = f&#39;${_input}&#39;
    chroms = get_chromosomes(f&#39;${_input}&#39;)
    blacklist_chroms = get_blacklist_chromosomes(f&#39;${chr_blacklist}&#39;)

    create_phenotype_table(ratio_file, chroms, blacklist_chroms)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[leafcutter_norm_2]
parameter: autosomes = True
import pandas as pd
molecular_pheno_chr_inv = pd.read_csv(f&#39;{_input[0]}&#39;,sep = &quot;\t&quot;)
# filter the autosomes or not
if autosomes:
    print(&quot;Analyzing autosomes 1 to 22...&quot;)
    mask = molecular_pheno_chr_inv[&#39;#chr&#39;].str.match(r&#39;chr(\d+)$&#39;)
    chr_numbers = molecular_pheno_chr_inv[&#39;#chr&#39;].str.extract(r&#39;chr(\d+)&#39;, expand=False).dropna().astype(int)
    molecular_pheno_chr_inv = molecular_pheno_chr_inv[mask &amp; chr_numbers.between(1, 22)]
    
molecular_pheno_chr_inv = molecular_pheno_chr_inv.values.tolist()
file_inv = [x[1] for x in molecular_pheno_chr_inv]
input: file_inv # This design is necessary to avoid using for_each, as sos can not take chr number as an input.
output: f&#39;{_input[0]:n}_raw_data.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container=container, entrypoint = entrypoint
    head -1 ${_input[0]:r}  &gt; ${_output}
    cat ${_input:r} | grep -v &quot;#Chr&quot; &gt;&gt; ${_output}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[leafcutter_norm_3,leafcutter_qqnorm]
parameter: no_norm = False
stop_if(no_norm)
parameter: mean_impute = False
# minimal NA rate with in sample values for a possible alternative splicing event to be kept (default 0.4, chosen according to leafcutter default na rate)
parameter: na_rate = 0.4
# minimal variance across samples for a possible alternative splicing event to be kept (default 0.001, chosen according to psichomics suggested minimal variance)
parameter: min_variance = 0.001
# qced_data refers to data that has undergone leafcutter_norm1/2 normalization or has already been imputed. 
# When this data is provided, mean imputation is automatically enabled (mean imputation set to true). Thus, the default imputation method is mean imputation. 
# However, if the data provided is already imputed (i.e., contains no NA values), no further imputation will be applied.
parameter: qced_data = path(&quot;.&quot;)
parameter: bgzip=False
import os
if qced_data.is_dir():
    print(&#39;Target Qced file with suffix &quot;_raw_data.txt&quot;&#39;)
    input_file = paths([os.path.join(qced_data, x) for x in os.listdir(qced_data) if x.endswith(&#39;_raw_data.txt&#39;)])
    bgzip=True
else:
    print(&#39;Loading QCed data as inputed file&#39;)
    input_file = paths(qced_data)
    print(&#39;Setting mean_imput to True. If the input has already been imputed (i.e., no NA values are present), nothing will change. Otherwise, mean imputation will be applied to NA values.&#39;)
    mean_impute = True
input: input_file
output: f&#39;{_input:n}.qqnorm.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container=container, entrypoint = entrypoint
    import numpy as np
    import pandas as pd
    from sklearn import preprocessing

    from scipy.stats import norm
    from scipy.stats import rankdata

    def qqnorm(x):
        n=len(x)
        a=3.0/8.0 if n&lt;=10 else 0.5
        return(norm.ppf( (rankdata(x,nan_policy=&quot;omit&quot;)-a)/(n+1.0-2.0*a) ))
        #return(norm.ppf( (rankdata(x)-a)/(n+1.0-2.0*a) ))

    raw_df = pd.read_csv(f&#39;${_input}&#39;,sep = &quot;\t&quot;)
    valRows = raw_df.iloc[:,4:]
    headers = list(raw_df)

    drop_list = []
    na_limit = len(valRows.columns)*${na_rate}

    for index, row in valRows.iterrows():

        # If ratio is missing for over 40% of the samples, drop
        if (row.isna().sum()) &gt; na_limit:
            drop_list.append(index)
        # Set missing values as the mean of existed values in a row
        if ${mean_impute}:
            valRows.iloc[index, ] = row.fillna(row.mean())
        # else: 
        #     row.fillna(row.mean())
        # drop introns with variance smaller than some minimal value
        if np.nanstd(row) &lt; ${min_variance}:
            drop_list.append(index)

    # save the intron information and sample values for remaining introns/rows
    newtable = raw_df.drop(drop_list).iloc[:,0:4]
    valRows = valRows.drop(drop_list)

    # scale normalize on each row
    valRows_matrix = []
    for c in (valRows.values.tolist()):
        c = preprocessing.scale(c)
        valRows_matrix.append(c)
    
    # qqnorms on the columns
    matrix = np.array(valRows_matrix)
    for i in range(len(matrix[0,:])):
        matrix[:,i] = qqnorm(matrix[:,i])
    normalized_table = pd.DataFrame(matrix)

    # reset row index for the saved intron infomation so the index will match sample values
    newtable = newtable.reset_index(drop=True)
    # merge the two parts of table
    output = pd.concat([newtable, normalized_table], axis=1)
    output.columns = headers

    # write normalized table
    output.to_csv(f&#39;${_input:n}.qqnorm.txt&#39;, sep=&quot;\t&quot;, index=None)

bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container=container, entrypoint = entrypoint
    if ${&#39;true&#39; if bgzip else &#39;false&#39;}; then
        # Compress the sorted BED file
        bgzip -c &quot;${_output}&quot; &gt; &quot;${_output:n}.bed.gz&quot;
        # Index the compressed BED file
        tabix -p bed &quot;${_output:n}.bed.gz&quot;
    fi
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="psichomics-norm">
<h2><code class="docutils literal notranslate"><span class="pre">psichomics_norm</span></code><a class="headerlink" href="#psichomics-norm" title="Link to this heading">#</a></h2>
<p>Documentation: <a class="reference external" href="http://bioconductor.org/packages/release/bioc/html/psichomics.html"><code class="docutils literal notranslate"><span class="pre">psichomics</span></code></a>
Consider retaining more information, the only QC on PSI values here are NA removal and a minimal variance filter, however, psichomics team suggested some further QC which can be checked <a class="reference external" href="https://github.com/nuno-agostinho/psichomics/issues/450">here</a>.
For reference, default minimal variance in leafcutter QC is 0.005.</p>
<p>Due to the difference of alternative splicing event type in psichomics outputs, we are not performing normalization on each sample’s data (columns) but only on each event (rows).</p>
<p><em><strong>TO DO:</strong></em> rename the .qqnorm. and test downstreams</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[psichomics_norm_1]
parameter: ratios = path
input: ratios, group_by = &#39;all&#39;
output: f&#39;{cwd}/psichomics_raw_data_bedded.txt&#39; 
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
R: expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container=container, entrypoint = entrypoint
    library(psichomics)
    library(data.table)

    psi_data &lt;- as.matrix(fread(&quot;${_input}&quot;),rownames=1)
    psi_data = as.data.frame(psi_data)
  
    # Process PSI df into bed file for tensorQTL. (This part of code is modified from Ryan Yordanoff&#39;s work)
    parsed_events &lt;- parseSplicingEvent(row.names(psi_data))
    
    # Create bedfile df and fill values with parsed values
    bed_file &lt;- data.frame(&quot;chr&quot;=parsed_events$chrom,&quot;start&quot;=parsed_events$start,&quot;end&quot;=parsed_events$end,&quot;ID&quot;=row.names(parsed_events),psi_data,check.names = FALSE)
    names(bed_file)[1] &lt;- &quot;#Chr&quot;
    bed_file$&#39;#Chr&#39; &lt;- sub(&quot;^&quot;, &quot;chr&quot;, bed_file$&#39;#Chr&#39;)
    row.names(bed_file) &lt;- NULL   

    bed_file &lt;- bed_file[order(bed_file$&#39;#Chr&#39;,bed_file$start,bed_file$end),]
  
    # Create BED file output
    write.table(x=bed_file, file = &quot;${cwd}/psichomics_raw_data_bedded.txt&quot;, quote = FALSE, row.names = FALSE, sep = &quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[psichomics_norm_2]
# minimal NA rate with in sample values for a possible alternative splicing event to be kept (default 0.4, chosen according to leafcutter default na rate)
parameter: na_rate = 0.4
# minimal variance across samples for a possible alternative splicing event to be kept (default 0.001, chosen according to psichomics suggested minimal variance)
parameter: min_variance = 0.001
output: f&#39;{_input:n}.qqnorm.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container=container, entrypoint = entrypoint
    import numpy as np
    import pandas as pd
    from sklearn import preprocessing

    from scipy.stats import norm
    from scipy.stats import rankdata

    def qqnorm(x):
        n=len(x)
        a=3.0/8.0 if n&lt;=10 else 0.5
        return(norm.ppf( (rankdata(x,nan_policy=&quot;omit&quot;)-a)/(n+1.0-2.0*a) ))
        #return(norm.ppf( (rankdata(x)-a)/(n+1.0-2.0*a) ))

    raw_df = pd.read_csv(f&#39;${_input}&#39;,sep = &quot;\t&quot;)
    valRows = raw_df.iloc[:,4:]
    headers = list(raw_df)

    drop_list = []
    na_limit = len(valRows.columns)*${na_rate}

    for index, row in valRows.iterrows():

        # If ratio is missing for over 40% of the samples, drop
        if (row.isna().sum()) &gt; na_limit:
            drop_list.append(index)
            # drop introns with variance smaller than some minimal value
        elif np.std(row) &lt; ${min_variance}:
            drop_list.append(index)

    # save the intron information and sample values for remaining introns/rows
    newtable = raw_df.drop(drop_list).iloc[:,0:4]
    valRows = valRows.drop(drop_list)

    # scale normalize on each row
    valRows_matrix = []
    for c in (valRows.values.tolist()):
        c = preprocessing.scale(c)
        valRows_matrix.append(c)

    # qqnorms on the columns
    matrix = np.array(valRows_matrix)
    for i in range(len(matrix[0,:])):
        matrix[:,i] = qqnorm(matrix[:,i])
    normalized_table = pd.DataFrame(matrix)

    # reset row index for the saved intron infomation so the index will match sample values
    newtable = newtable.reset_index(drop=True)
    # merge the two parts of table
    output = pd.concat([newtable, normalized_table], axis=1)
    output.columns = headers

    # write normalized table, avoid scientific notation
    output.to_csv(f&#39;${_input:n}.qqnorm.txt&#39;, sep=&quot;\t&quot;, index=None, float_format=&#39;%.16f&#39;)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/molecular_phenotypes/QC"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../calling/splicing_calling.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Quantifying alternative splicing from RNA-seq data</p>
      </div>
    </a>
    <a class="right-next"
       href="../../data_preprocessing/genotype_preprocessing.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genotype data preprocessing</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#leafcutter"><code class="docutils literal notranslate"><span class="pre">leafcutter</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#psichomics"><code class="docutils literal notranslate"><span class="pre">psichomics</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1"><code class="docutils literal notranslate"><span class="pre">leafcutter</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2"><code class="docutils literal notranslate"><span class="pre">psichomics</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-splicing-qc-and-normalization">ii. Splicing QC and Normalization</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-leafcutter">a. Leafcutter</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-psichomics">a. Psichomics</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-sample-list">0. Generate sample list</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#leafcutter-norm"><code class="docutils literal notranslate"><span class="pre">leafcutter_norm</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#parameter-annotations">Parameter Annotations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#things-to-keep-in-mind">Things to keep in mind</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#psichomics-norm"><code class="docutils literal notranslate"><span class="pre">psichomics_norm</span></code></a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>