
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Extract genome-wide data for multivariate analysis &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="A multivariate EBNM approach for mixture multivariate distribution estimate" href="mixture_prior.html" />
    <link rel="prev" title="Multivariate association with MASH" href="../MASH_pipeline.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../reference_data/reference_data.html">
   Reference Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../reference_data/reference_data_preparation.html">
     Reference Data Standardization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../reference_data/generalized_TADB.html">
     Generation of Topologically Associated Domains and their Boundaries
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../reference_data/ld_prune_reference.html">
     Independent list of variants using LD clumping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene Coordinate Annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype Data Imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden Factor Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate Data Formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../pecotmr_integration/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_integration/twas_mr.html">
     GWAS integration: TWAS and MR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_integration/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Extract genome-wide data for multivariate analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/multivariate_genome/MASH/mash_preprocessing.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate_genome/MASH/mash_preprocessing.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/multivariate_genome/MASH/mash_preprocessing.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#description">
   Description
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example">
     Example
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-the-random-and-null-effects-per-analysis-unit">
   Get the random and null effects per analysis unit
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="extract-genome-wide-data-for-multivariate-analysis">
<h1>Extract genome-wide data for multivariate analysis<a class="headerlink" href="#extract-genome-wide-data-for-multivariate-analysis" title="Permalink to this headline">¶</a></h1>
<div class="section" id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h2>
<p>This notebook prepares input data for Utimate Decomposition to generate mixture prior (for mvSuSiE) or to use for MASH analysis. It outputs 3 sets of data: <span class="math notranslate nohighlight">\(Z_s\)</span>, <span class="math notranslate nohighlight">\(Z_n\)</span> and <span class="math notranslate nohighlight">\(Z_r\)</span> (strong, null and random)</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(Z_s\)</span>: <strong>this is now extracted from genome-wide cis analysis fine-mapping results</strong>. We extract the top loci data frame of each condition, where the CS threshold is set to be 0.7. Then we merge the z-scores of them into one data frame.</p></li>
<li><p><span class="math notranslate nohighlight">\(Z_n\)</span>: (null <span class="math notranslate nohighlight">\(Z\)</span>-scores): we first extract up to <span class="math notranslate nohighlight">\(M\)</span> candidate SNPs from each region which satisify <span class="math notranslate nohighlight">\(|z| \le 2\)</span>, then overlap it with the list of independent SNPs to keep only independent variants, then finally take the union of the extracted.</p></li>
<li><p><span class="math notranslate nohighlight">\(Z_r\)</span>: we randomly extract variants based on input independent list of variants.</p></li>
</ul>
<p><strong>FIXME: We need to apply the independent list of variants Anqi developed and use it here to filter and get <span class="math notranslate nohighlight">\(Z_n\)</span> and <span class="math notranslate nohighlight">\(Z_r\)</span>. This logic shoud be added to <code class="docutils literal notranslate"><span class="pre">processing_1</span></code>. Also, it might be a good idea we take some of these utility functions into pecotmr package for better maintenance. For example <code class="docutils literal notranslate"><span class="pre">processing_1</span></code> the function to load regional summary stats from tensorQTL into a matrix should be packed into pecotmr; plus this one function <code class="docutils literal notranslate"><span class="pre">handle_nan_etc</span></code>. processing_2 can stay as is; the <code class="docutils literal notranslate"><span class="pre">susie_signal</span></code> step can also go into <code class="docutils literal notranslate"><span class="pre">pecotmr</span></code> as a way for users to summarize signals from SuSiE for other purposes</strong></p>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p><strong>Marginal summary statistics files</strong>: Bgzipped summary statistics for chromosomes 1-22, generated by tensorQTL cis-analysis and indexed by <code class="docutils literal notranslate"><span class="pre">tabix</span></code>.</p></li>
<li><p><strong>Fine-mapping results file index</strong>: Path to lists of fine-mapped RDS files from finemapping output.</p></li>
<li><p><strong>Genome region partition</strong> (optional): Defines genomic regions for each gene as enhanced cis regions where we should extract <span class="math notranslate nohighlight">\(Z_n\)</span> and <span class="math notranslate nohighlight">\(Z_r\)</span> from. This list is used for fine-mapping, so if the complete list of fine-mapping RDS (rather than a handful of it) is already avaiable (#2 above) then there is no need to provide this file. Otherwise, it’s going to be limited to only certaion regions, which is also good for testing purpose.</p></li>
</ol>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>A list of 10 elements:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">List</span> <span class="n">of</span> <span class="mi">10</span>
 <span class="err">$</span> <span class="n">random</span><span class="o">.</span><span class="n">z</span><span class="p">:</span> <span class="n">num</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">...</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;dimnames&quot;</span><span class="p">)</span><span class="o">=</span><span class="n">List</span> <span class="n">of</span> <span class="mi">2</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">]</span> <span class="s2">&quot;1:97960:A:G&quot;</span> <span class="s2">&quot;1:138565:G:A&quot;</span> <span class="s2">&quot;1:15112:C:T&quot;</span> <span class="s2">&quot;1:189947:G:A&quot;</span> <span class="o">...</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
 <span class="err">$</span> <span class="n">null</span><span class="o">.</span><span class="n">z</span>  <span class="p">:</span> <span class="n">num</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">-</span><span class="mf">0.785</span> <span class="o">...</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;dimnames&quot;</span><span class="p">)</span><span class="o">=</span><span class="n">List</span> <span class="n">of</span> <span class="mi">2</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">]</span> <span class="s2">&quot;1:93692:C:T&quot;</span> <span class="s2">&quot;1:273645:A:G&quot;</span> <span class="s2">&quot;1:10442:CCTA:.&quot;</span> <span class="s2">&quot;1:198942:A:C&quot;</span> <span class="o">...</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
 <span class="err">$</span> <span class="n">random</span><span class="o">.</span><span class="n">b</span><span class="p">:</span> <span class="n">num</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">...</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;dimnames&quot;</span><span class="p">)</span><span class="o">=</span><span class="n">List</span> <span class="n">of</span> <span class="mi">2</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">]</span> <span class="s2">&quot;1:97960:A:G&quot;</span> <span class="s2">&quot;1:138565:G:A&quot;</span> <span class="s2">&quot;1:15112:C:T&quot;</span> <span class="s2">&quot;1:189947:G:A&quot;</span> <span class="o">...</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
 <span class="err">$</span> <span class="n">null</span><span class="o">.</span><span class="n">b</span>  <span class="p">:</span> <span class="n">num</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">-</span><span class="mf">0.123</span> <span class="o">...</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;dimnames&quot;</span><span class="p">)</span><span class="o">=</span><span class="n">List</span> <span class="n">of</span> <span class="mi">2</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">]</span> <span class="s2">&quot;1:93692:C:T&quot;</span> <span class="s2">&quot;1:273645:A:G&quot;</span> <span class="s2">&quot;1:10442:CCTA:.&quot;</span> <span class="s2">&quot;1:198942:A:C&quot;</span> <span class="o">...</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
 <span class="err">$</span> <span class="n">null</span><span class="o">.</span><span class="n">s</span>  <span class="p">:</span> <span class="n">num</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="o">...</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;dimnames&quot;</span><span class="p">)</span><span class="o">=</span><span class="n">List</span> <span class="n">of</span> <span class="mi">2</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">]</span> <span class="s2">&quot;1:93692:C:T&quot;</span> <span class="s2">&quot;1:273645:A:G&quot;</span> <span class="s2">&quot;1:10442:CCTA:.&quot;</span> <span class="s2">&quot;1:198942:A:C&quot;</span> <span class="o">...</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
 <span class="err">$</span> <span class="n">random</span><span class="o">.</span><span class="n">s</span><span class="p">:</span> <span class="n">num</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="mf">0.157</span> <span class="o">...</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;dimnames&quot;</span><span class="p">)</span><span class="o">=</span><span class="n">List</span> <span class="n">of</span> <span class="mi">2</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">36</span><span class="p">]</span> <span class="s2">&quot;1:97960:A:G&quot;</span> <span class="s2">&quot;1:138565:G:A&quot;</span> <span class="s2">&quot;1:15112:C:T&quot;</span> <span class="s2">&quot;1:189947:G:A&quot;</span> <span class="o">...</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
 <span class="err">$</span> <span class="n">strong</span><span class="o">.</span><span class="n">b</span><span class="p">:</span><span class="n">Classes</span> <span class="err">‘</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="err">’</span> <span class="ow">and</span> <span class="s1">&#39;data.frame&#39;</span><span class="p">:</span>	<span class="mi">1</span> <span class="n">obs</span><span class="o">.</span> <span class="n">of</span>  <span class="mi">2</span> <span class="n">variables</span><span class="p">:</span>
  <span class="o">..</span><span class="err">$</span> <span class="n">A</span><span class="p">:</span> <span class="n">num</span> <span class="o">-</span><span class="mf">0.217</span>
  <span class="o">..</span><span class="err">$</span> <span class="n">B</span><span class="p">:</span> <span class="n">num</span> <span class="o">-</span><span class="mf">0.217</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;.internal.selfref&quot;</span><span class="p">)</span><span class="o">=&lt;</span><span class="n">externalptr</span><span class="o">&gt;</span> 
 <span class="err">$</span> <span class="n">strong</span><span class="o">.</span><span class="n">s</span><span class="p">:</span><span class="n">Classes</span> <span class="err">‘</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="err">’</span> <span class="ow">and</span> <span class="s1">&#39;data.frame&#39;</span><span class="p">:</span>	<span class="mi">1</span> <span class="n">obs</span><span class="o">.</span> <span class="n">of</span>  <span class="mi">2</span> <span class="n">variables</span><span class="p">:</span>
  <span class="o">..</span><span class="err">$</span> <span class="n">A</span><span class="p">:</span> <span class="n">num</span> <span class="mf">0.0481</span>
  <span class="o">..</span><span class="err">$</span> <span class="n">B</span><span class="p">:</span> <span class="n">num</span> <span class="mf">0.0481</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;.internal.selfref&quot;</span><span class="p">)</span><span class="o">=&lt;</span><span class="n">externalptr</span><span class="o">&gt;</span> 
 <span class="err">$</span> <span class="n">strong</span><span class="o">.</span><span class="n">z</span><span class="p">:</span><span class="n">Classes</span> <span class="err">‘</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="err">’</span> <span class="ow">and</span> <span class="s1">&#39;data.frame&#39;</span><span class="p">:</span>	<span class="mi">1</span> <span class="n">obs</span><span class="o">.</span> <span class="n">of</span>  <span class="mi">2</span> <span class="n">variables</span><span class="p">:</span>
  <span class="o">..</span><span class="err">$</span> <span class="n">A</span><span class="p">:</span> <span class="n">num</span> <span class="o">-</span><span class="mf">4.5</span>
  <span class="o">..</span><span class="err">$</span> <span class="n">B</span><span class="p">:</span> <span class="n">num</span> <span class="o">-</span><span class="mf">4.5</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;.internal.selfref&quot;</span><span class="p">)</span><span class="o">=&lt;</span><span class="n">externalptr</span><span class="o">&gt;</span> 
 <span class="err">$</span> <span class="n">XtX</span>     <span class="p">:</span> <span class="n">num</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="mf">20.3</span> <span class="mf">20.3</span> <span class="mf">20.3</span> <span class="mf">20.3</span>
  <span class="o">..-</span> <span class="n">attr</span><span class="p">(</span><span class="o">*</span><span class="p">,</span> <span class="s2">&quot;dimnames&quot;</span><span class="p">)</span><span class="o">=</span><span class="n">List</span> <span class="n">of</span> <span class="mi">2</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
  <span class="o">..</span> <span class="o">..</span><span class="err">$</span> <span class="p">:</span> <span class="nb">chr</span> <span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="s2">&quot;A&quot;</span> <span class="s2">&quot;B&quot;</span>
</pre></div>
</div>
<div class="section" id="example">
<h3>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># generate random and null only</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_preprocessing</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">processing</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_protein</span> \
    <span class="o">--</span><span class="n">sum_files</span> <span class="n">test_pQTL_asso_list</span> \
               <span class="n">test_pQTL_asso_list</span> \
    <span class="o">--</span><span class="n">region_file</span> <span class="n">test</span><span class="o">.</span><span class="n">region</span> \
    <span class="o">--</span><span class="n">traits</span> <span class="n">A</span> <span class="n">B</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># generate strong only</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_preprocessing</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_signal</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_protein</span> \
    <span class="o">--</span><span class="n">susie_list</span> <span class="n">protocol_example_protein</span><span class="o">.</span><span class="n">susie_output</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">traits</span> <span class="n">A</span> <span class="n">B</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># generate mashr input directly</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_preprocessing</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash_input</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_protein</span> \
    <span class="o">--</span><span class="n">sum_files</span> <span class="n">test_pQTL_asso_list</span> \
               <span class="n">test_pQTL_asso_list</span> \
    <span class="o">--</span><span class="n">region_file</span> <span class="n">test</span><span class="o">.</span><span class="n">region</span> \
    <span class="o">--</span><span class="n">susie_list</span> <span class="n">protocol_example_protein</span><span class="o">.</span><span class="n">susie_output</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">traits</span> <span class="n">A</span> <span class="n">B</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1"># Path to work directory where output locates</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./output&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">seed</span> <span class="o">=</span> <span class="mi">999</span>
<span class="kn">parameter:</span> <span class="n">n_random</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">n_null</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">expected_ncondition</span> <span class="o">=</span> <span class="mi">0</span>
<span class="kn">parameter:</span> <span class="n">exclude_condition</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Containers that contains the necessary packages</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;10min&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;8G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># This is in principle required; but in practice it can be optional if we are not exactly stringent about getting independent SNPs</span>
<span class="kn">parameter:</span> <span class="n">independent_variant_list</span> <span class="o">=</span> <span class="n">path</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># extract data for MASH from summary stats</span>
<span class="p">[</span><span class="n">susie_to_mash_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span><span class="mi">100</span>
<span class="kn">parameter:</span> <span class="n">susie_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># first 3 col are chr start end, 4th column is region ID, 5th col are file names, 6 col is all the condition names comma split</span>
<span class="n">susie_data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;c(&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">y</span><span class="si">}</span><span class="s2">&#39;&quot;</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">())</span> <span class="o">+</span> <span class="s2">&quot;)&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">susie_list</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()]</span> 
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="n">susie_data</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">per_chunk</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s2">_cache/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s2">_batch</span><span class="si">{</span><span class="n">_index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">.rds&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">merge_data</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">one_data</span><span class="p">)</span> <span class="p">{</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
          <span class="k">return</span><span class="p">(</span><span class="n">one_data</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">res</span><span class="err">$</span><span class="n">random</span><span class="o">.</span><span class="n">b</span><span class="p">)</span><span class="o">|</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">res</span><span class="err">$</span><span class="n">null</span><span class="o">.</span><span class="n">b</span><span class="p">)))</span> <span class="p">{</span>
          <span class="k">return</span><span class="p">(</span><span class="n">one_data</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">one_data</span><span class="p">))</span> <span class="p">{</span>
          <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="k">for</span> <span class="p">(</span><span class="n">d</span> <span class="ow">in</span> <span class="n">names</span><span class="p">(</span><span class="n">one_data</span><span class="p">))</span> <span class="p">{</span>
            <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">one_data</span><span class="p">[[</span><span class="n">d</span><span class="p">]]))</span> <span class="p">{</span>
              <span class="nb">next</span>
            <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">res</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">res</span><span class="p">[[</span><span class="n">d</span><span class="p">]],</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">one_data</span><span class="p">[[</span><span class="n">d</span><span class="p">]])))</span>
            <span class="p">}</span>
          <span class="p">}</span>
          <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
      <span class="p">}</span>
    <span class="p">}</span>
    <span class="n">exclude_condition</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">repr</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">exclude_condition</span><span class="p">])})</span>
    <span class="n">meta_df</span> <span class="o">=</span> <span class="n">rbind</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">_susie_data</span><span class="p">)}}</span>
    <span class="n">res</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">nrows</span><span class="p">(</span><span class="n">meta_df</span><span class="p">))</span> <span class="p">{</span>
      <span class="n">line</span> <span class="o">=</span> <span class="n">meta_df</span><span class="p">[</span><span class="n">i</span><span class="p">,]</span>
      <span class="n">region_files</span> <span class="o">=</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span> <span class="n">split</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
      <span class="n">traits</span> <span class="o">=</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span> <span class="n">split</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
      <span class="n">strong</span> <span class="o">&lt;-</span> <span class="n">pecotmr</span><span class="p">::</span><span class="n">load_multitrait_R_sumstat</span><span class="p">(</span><span class="n">region_files</span><span class="p">,</span> <span class="n">top_loci</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="n">ran_null</span> <span class="o">&lt;-</span> <span class="n">pecotmr</span><span class="p">::</span><span class="n">load_multitrait_R_sumstat</span><span class="p">(</span><span class="n">region_files</span><span class="p">,</span> <span class="n">filter_file</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">independent_variant_list</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">exclude_condition</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
          <span class="n">strong</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">strong</span><span class="o">.</span><span class="n">z</span><span class="o">=</span><span class="n">strong</span><span class="err">$</span><span class="n">bhat</span><span class="p">[,</span><span class="o">-</span><span class="n">exclude_condition</span><span class="p">]</span><span class="o">/</span><span class="n">strong</span><span class="err">$</span><span class="n">sbhat</span><span class="p">[,</span><span class="o">-</span><span class="n">exclude_condition</span><span class="p">],</span> <span class="n">strong</span><span class="o">.</span><span class="n">b</span><span class="o">=</span><span class="n">strong</span><span class="err">$</span><span class="n">bhat</span><span class="p">[,</span><span class="o">-</span><span class="n">exclude_condition</span><span class="p">],</span> <span class="n">strong</span><span class="o">=</span><span class="n">strong</span><span class="err">$</span><span class="n">sbhat</span><span class="p">[,</span><span class="o">-</span><span class="n">exclude_condition</span><span class="p">])</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="n">strong</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">strong</span><span class="o">.</span><span class="n">z</span><span class="o">=</span><span class="n">strong</span><span class="err">$</span><span class="n">bhat</span><span class="o">/</span><span class="n">strong</span><span class="err">$</span><span class="n">sbhat</span><span class="p">,</span> <span class="n">strong</span><span class="o">.</span><span class="n">b</span><span class="o">=</span><span class="n">strong</span><span class="err">$</span><span class="n">bhat</span><span class="p">,</span> <span class="n">strong</span><span class="o">=</span><span class="n">strong</span><span class="err">$</span><span class="n">sbhat</span><span class="p">)</span>
      <span class="p">}</span>
      <span class="n">ran_null</span> <span class="o">&lt;-</span> <span class="n">pecotmr</span><span class="p">::</span><span class="n">mash_ran_null_sample</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">n_random</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">n_null</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">expected_ncondition</span><span class="p">},</span> <span class="n">exclude_condition</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">seed</span><span class="p">})</span>
      <span class="n">res</span> <span class="o">&lt;-</span> <span class="n">merge_data</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="n">strong</span><span class="p">,</span> <span class="n">ran_null</span><span class="p">))</span>
    <span class="p">}</span>
    
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s2">&quot;xz&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">susie_to_mash_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s2">.mash_input.rds&quot;</span> 
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;1h&#39;</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;16G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>

  <span class="c1"># Function to merge data</span>
  <span class="n">merge_data</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">one_data</span><span class="p">)</span> <span class="p">{</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
      <span class="k">return</span><span class="p">(</span><span class="n">one_data</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">one_data</span><span class="p">))</span> <span class="p">{</span>
      <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">d</span> <span class="ow">in</span> <span class="n">names</span><span class="p">(</span><span class="n">one_data</span><span class="p">))</span> <span class="p">{</span>
        <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">one_data</span><span class="p">[[</span><span class="n">d</span><span class="p">]]))</span> <span class="p">{</span>
          <span class="nb">next</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="n">res</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">res</span><span class="p">[[</span><span class="n">d</span><span class="p">]],</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">one_data</span><span class="p">[[</span><span class="n">d</span><span class="p">]])))</span>
        <span class="p">}</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
    <span class="p">}</span>
  <span class="p">}</span>

  <span class="n">dat</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
  <span class="k">for</span> <span class="p">(</span><span class="n">f</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}))</span> <span class="p">{</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">merge_data</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">f</span><span class="p">))</span>
  <span class="p">}</span>
  <span class="n">dat</span><span class="err">$</span><span class="n">ZtZ</span> <span class="o">=</span> <span class="n">t</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">strong</span><span class="o">.</span><span class="n">z</span><span class="p">))</span> <span class="o">%*%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">strong</span><span class="o">.</span><span class="n">z</span><span class="p">)</span> <span class="o">/</span> <span class="n">nrow</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">strong</span><span class="o">.</span><span class="n">z</span><span class="p">)</span>
  <span class="n">saveRDS</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s2">&quot;xz&quot;</span><span class="p">)</span>
 
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="kp">name</span><span class="p">}</span><span class="n">_cache</span><span class="o">/</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="get-the-random-and-null-effects-per-analysis-unit">
<h2>Get the random and null effects per analysis unit<a class="headerlink" href="#get-the-random-and-null-effects-per-analysis-unit" title="Permalink to this headline">¶</a></h2>
<p><strong>FIXME: notice that we no longer rely on tensorQTL results for MASH analysis. Our new protocol uses SuSiE output. We keep this piece of code to extract from tensorQTL which may still be relevant for trans-analysis. But currently it is limited to random and null variants extraction</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">random_null_tensorqtl_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sum_files</span> <span class="o">=</span> <span class="n">paths</span>
<span class="kn">parameter:</span> <span class="n">region_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">traits</span> <span class="o">=</span> <span class="n">paths</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="k">def</span> <span class="nf">find_matching_files_for_region</span><span class="p">(</span><span class="n">chr_id</span><span class="p">):</span>
    <span class="n">chr_number</span> <span class="o">=</span> <span class="n">chr_id</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>  <span class="c1"># subset 1 from chr1</span>
    <span class="n">pattern_str</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;\.</span><span class="si">{chr_number}</span><span class="s2">\.&quot;</span>
    <span class="n">pattern</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">pattern_str</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chr_number</span><span class="o">=</span><span class="n">chr_number</span><span class="p">))</span>
    <span class="n">paths</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">sum_file</span> <span class="ow">in</span> <span class="n">sum_files</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">sum_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">af</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">aline</span> <span class="ow">in</span> <span class="n">af</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">pattern</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">aline</span><span class="p">):</span>
                    <span class="n">paths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aline</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
    <span class="k">return</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">paths</span><span class="p">)</span>

<span class="n">updated_regions</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">region_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">regions</span><span class="p">:</span>
    <span class="n">header</span> <span class="o">=</span> <span class="n">regions</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
    <span class="n">updated_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">header</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">path</span><span class="se">\t</span><span class="s2">region&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">:</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">chr_id</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gene_id</span> <span class="o">=</span> <span class="n">parts</span>
        <span class="n">paths</span> <span class="o">=</span> <span class="n">find_matching_files_for_region</span><span class="p">(</span><span class="n">chr_id</span><span class="p">)</span>
        <span class="n">updated_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chr_id</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">paths</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">chr_id</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">meta_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">updated_regions</span><span class="p">[</span><span class="mi">1</span><span class="p">:]],</span> <span class="n">columns</span><span class="o">=</span><span class="n">updated_regions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="n">meta</span> <span class="o">=</span> <span class="n">meta_df</span><span class="p">[[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;path&#39;</span><span class="p">,</span> <span class="s1">&#39;region&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">orient</span><span class="o">=</span><span class="s1">&#39;records&#39;</span><span class="p">)</span>

<span class="kn">input:</span> <span class="n">for_each</span><span class="o">=</span><span class="s1">&#39;meta&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">_cache/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_meta</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">region</span> <span class="o">&lt;-</span> <span class="s2">&quot;$</span><span class="si">{_meta[&#39;region&#39;]}</span><span class="s2">&quot;</span>
    <span class="c1"># FIXME I am sure there is a more elegant way to put together the path, via SoS</span>
    <span class="n">phenotype_path</span> <span class="o">&lt;-</span> <span class="n">unlist</span><span class="p">(</span><span class="n">strsplit</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_meta[&#39;path&#39;]}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">))</span>
    <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">tryCatch</span><span class="p">(</span>
      <span class="p">{</span>
        <span class="c1"># Try to run the function</span>
         <span class="n">pecotmr</span><span class="p">::</span><span class="n">load_multitrait_tensorqtl_sumstat</span><span class="p">(</span><span class="n">phenotype_path</span> <span class="o">=</span> <span class="n">phenotype_path</span><span class="p">,</span> <span class="n">region</span> <span class="o">=</span> <span class="n">region</span><span class="p">,</span> 
          <span class="n">trait_names</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">traits</span><span class="p">:</span><span class="n">r</span><span class="p">,}),</span> <span class="n">filter_file</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">remove_any_missing</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">max_rows_selected</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span> <span class="n">na_remove</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;T&quot;</span> <span class="k">if</span> <span class="n">na_remove</span> <span class="k">else</span> <span class="s2">&quot;F&quot;</span><span class="p">})</span>
      <span class="p">},</span>
      <span class="n">error</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">e</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Attempt remove chr in region ID to load the data.&quot;</span><span class="p">)</span>
        <span class="c1"># If an error occurs, modify the region and try again</span>
        <span class="n">pecotmr</span><span class="p">::</span><span class="n">load_multitrait_tensorqtl_sumstat</span><span class="p">(</span><span class="n">phenotype_path</span> <span class="o">=</span> <span class="n">phenotype_path</span><span class="p">,</span> <span class="n">region</span> <span class="o">=</span>  <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">region</span><span class="p">),</span> 
          <span class="n">trait_names</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">traits</span><span class="p">:</span><span class="n">r</span><span class="p">,}),</span> <span class="n">filter_file</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">remove_any_missing</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">max_rows_selected</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span> <span class="n">na_remove</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;T&quot;</span> <span class="k">if</span> <span class="n">na_remove</span> <span class="k">else</span> <span class="s2">&quot;F&quot;</span><span class="p">})</span>
      <span class="p">}</span>
    <span class="p">)</span>
    <span class="n">exclude_condition</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">repr</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">exclude_condition</span><span class="p">])})</span>
    <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">pecotmr</span><span class="p">::</span><span class="n">mash_ran_null_sample</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">n_random</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">n_null</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">expected_ncondition</span><span class="p">},</span> <span class="n">exclude_condition</span><span class="p">,</span> <span class="n">z_only</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">z_only</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">},</span> <span class="n">seed</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">seed</span><span class="p">})</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s2">&quot;xz&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/multivariate_genome/MASH"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../MASH_pipeline.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Multivariate association with MASH</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="mixture_prior.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">A multivariate EBNM approach for mixture multivariate distribution estimate</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>