
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>MASH analysis pipeline with posterior computation &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="watershed analysis for rare xQTL" href="../../rare_xqtl/watershed.html" />
    <link rel="prev" title="MASH analysis pipeline with data-driven prior matrices" href="mash_fit.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../pecotmr_integration/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_integration/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_integration/ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_integration/fine_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_integration/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/multivariate_genome/MASH/mash_posterior.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate_genome/MASH/mash_posterior.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/multivariate_genome/MASH/mash_posterior.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#example">
   example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#posterior">
     posterior
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#contrast-analysis-works">
     contrast analysis: works
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#posterior-results">
   Posterior results
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id1">
   Posterior
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#posterior-contrast">
   Posterior contrast
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#feature-score-from-contrast-results">
   Feature score from contrast results
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#feature-score-meta">
     feature_score_meta
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#feature-score-finemap">
     feature_score_finemap
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#feature-score-nsig">
     feature_score_nsig
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#feature-pval-pair">
     feature_pval_pair
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="mash-analysis-pipeline-with-posterior-computation">
<h1>MASH analysis pipeline with posterior computation<a class="headerlink" href="#mash-analysis-pipeline-with-posterior-computation" title="Permalink to this headline">¶</a></h1>
<p>In the GTEx V6 paper we assumed one eQTL per gene and applied the model learned above to those SNPs. Under that assumption, the input data for posterior calculation will be the <code class="docutils literal notranslate"><span class="pre">dat$strong.*</span></code> matrices.
It is a fairly straightforward procedure as shown in <a class="reference external" href="https://stephenslab.github.io/mashr/articles/eQTL_outline.html">this vignette</a>.</p>
<p>But it is often more interesting to apply MASH to given list of eQTLs, eg, from those from fine-mapping results. In GTEx V8 analysis we obtain such gene-SNP pairs from DAP-G fine-mapping analysis. See <a class="reference external" href="https://stephenslab.github.io/gtex-eqtls/analysis/Independent_eQTL_Results.html">this notebook</a> for how the input data is prepared. The workflow below takes a number of input chunks (each chunk is a list of matrices <code class="docutils literal notranslate"><span class="pre">dat$Bhat</span></code> and <code class="docutils literal notranslate"><span class="pre">dat$Shat</span></code>)
and computes posterior for each chunk. It is therefore suited for running in parallel posterior computation for all gene-SNP pairs, if input data chunks are provided.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">JOB_OPT</span><span class="o">=</span><span class="s2">&quot;-c midway2.yml -q midway2&quot;</span>
<span class="n">DATA_DIR</span><span class="o">=/</span><span class="n">project</span><span class="o">/</span><span class="n">compbio</span><span class="o">/</span><span class="n">GTEx_eQTL</span><span class="o">/</span><span class="n">independent_eQTL</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">workflows</span><span class="o">/</span><span class="n">mashr_flashr_workflow</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">posterior</span> \
    <span class="err">$</span><span class="n">JOB_OPT</span> \
    <span class="o">--</span><span class="n">posterior</span><span class="o">-</span><span class="nb">input</span> <span class="err">$</span><span class="n">DATA_DIR</span><span class="o">/</span><span class="n">DAPG_pip_gt_0</span><span class="o">.</span><span class="mi">01</span><span class="o">-</span><span class="n">AllTissues</span><span class="o">/</span><span class="n">DAPG_pip_gt_0</span><span class="o">.</span><span class="mi">01</span><span class="o">-</span><span class="n">AllTissues</span><span class="o">.*.</span><span class="n">rds</span> \
                      <span class="err">$</span><span class="n">DATA_DIR</span><span class="o">/</span><span class="n">ConditionalAnalysis_AllTissues</span><span class="o">/</span><span class="n">ConditionalAnalysis_AllTissues</span><span class="o">.*.</span><span class="n">rds</span>
</pre></div>
</div>
<div class="section" id="example">
<h2>example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="section" id="posterior">
<h3>posterior<a class="headerlink" href="#posterior" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">MASH</span><span class="o">/</span><span class="n">MWE</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#slice</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">posterior</span>  \
    <span class="o">--</span><span class="n">analysis</span><span class="o">-</span><span class="n">units</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">cat</span>  <span class="n">MWE</span><span class="o">.</span><span class="n">list</span><span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">2</span><span class="p">)</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">MWE_udr</span>  \
    <span class="o">--</span><span class="n">output_prefix</span>  <span class="n">MWE_udr</span> \
    <span class="o">--</span><span class="n">vhat</span> <span class="n">mle</span> \
    <span class="o">--</span><span class="n">mash_model</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">MWE_udr</span><span class="o">.</span><span class="n">EZ</span><span class="o">.</span><span class="n">V_mle</span><span class="o">.</span><span class="n">mash_model</span><span class="o">.</span><span class="n">rds</span> \
    <span class="o">--</span><span class="n">slice_method</span> <span class="kc">True</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">stephenslab</span><span class="o">.</span><span class="n">sif</span> 
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="contrast-analysis-works">
<h3>contrast analysis: works<a class="headerlink" href="#contrast-analysis-works" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># no group</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash_posterior_contrast</span>   \
    <span class="o">--</span><span class="n">posterior_file</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">mash_output_list_all</span> \
    <span class="o">--</span><span class="n">sum_file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">list</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">test</span> \
    <span class="o">--</span><span class="n">cells</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># 2 group</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash_posterior_contrast</span>   \
    <span class="o">--</span><span class="n">posterior_file</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">mash_output_list_all</span> \
    <span class="o">--</span><span class="n">sum_file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">list</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">test_2group</span> \
    <span class="o">--</span><span class="n">cells</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> \
    <span class="o">--</span><span class="n">group1</span> <span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> \
    <span class="o">--</span><span class="n">group2</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># recipe with 2 group</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash_posterior_contrast</span>   \
    <span class="o">--</span><span class="n">posterior_file</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">mash_output_list_all</span> \
    <span class="o">--</span><span class="n">sum_file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">list</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">test_2group_recipe</span> \
    <span class="o">--</span><span class="n">cells</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> \
    <span class="o">--</span><span class="n">grouping_recipe</span> <span class="n">recipe</span>
</pre></div>
</div>
</div>
</div>
<p>group recipe looks like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Ast</span><span class="p">,</span><span class="n">Exc</span>
<span class="n">Mic</span><span class="p">,</span><span class="n">MiGA_GFM</span><span class="p">,</span><span class="n">MiGA_GTS</span>
<span class="n">Oli</span><span class="p">,</span><span class="n">DLPFC_pQTL</span>
<span class="c1"># ... add more groups as needed, separated by commas</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="c1"># Work directory &amp; output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./&#39;</span><span class="p">)</span>
<span class="c1"># The filename prefix for output data</span>
<span class="kn">parameter:</span> <span class="kp">name</span><span class="o">=</span><span class="s2">&quot;test&quot;</span>
<span class="kn">parameter:</span> <span class="n">cells</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">]</span><span class="c1">#order is important</span>
<span class="kn">parameter:</span> <span class="n">group1</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">group2</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">group3</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="c1"># handle N = per_chunk data-set in one job</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1">###add for test</span>
<span class="kn">parameter:</span> <span class="n">output_prefix</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">output_suffix</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="c1"># Exchangable effect (EE) or exchangable z-scores (EZ)</span>
<span class="kn">parameter:</span> <span class="n">effect_model</span> <span class="o">=</span> <span class="s1">&#39;EZ&#39;</span>
<span class="c1"># Identifier of $\hat{V}$ estimate file</span>
<span class="c1"># Options are &quot;identity&quot;, &quot;simple&quot;, &quot;mle&quot;, &quot;vhat_corshrink_xcondition&quot;, &quot;vhat_simple_specific&quot;</span>
<span class="kn">parameter:</span> <span class="n">vhat</span> <span class="o">=</span> <span class="s1">&#39;simple&#39;</span>
<span class="kn">parameter:</span> <span class="n">data</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;fastqtl_to_mash_output/FastQTLSumStats.mash.rds&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">p_cut</span><span class="o">=</span><span class="mf">0.00001</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">cwd</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">output_prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="n">vhat_data</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.V_</span><span class="si">{</span><span class="n">vhat</span><span class="si">}</span><span class="s2">.rds&quot;</span><span class="p">)</span>
<span class="n">mash_model</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.V_</span><span class="si">{</span><span class="n">vhat</span><span class="si">}</span><span class="s2">.mash_model.rds&quot;</span><span class="p">)</span>
<span class="n">posterior_list</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">output_suffix</span><span class="si">}</span><span class="s2">.posterior_list&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">sort_uniq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="n">seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq</span> <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="n">seen</span> <span class="ow">or</span> <span class="n">seen</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="posterior-results">
<h2>Posterior results<a class="headerlink" href="#posterior-results" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>The outcome of the <code class="docutils literal notranslate"><span class="pre">[posterior]</span></code> step should produce a number of serialized R objects <code class="docutils literal notranslate"><span class="pre">*.batch_*.posterior.rds</span></code> (can be loaded to R via <code class="docutils literal notranslate"><span class="pre">readRDS()</span></code>) – I chopped data to batches to take advantage of computing in multiple cluster nodes. It should be self-explanary but please let me know otherwise.</p></li>
<li><p>Other posterior related files are:</p>
<ol class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.batch_*.yaml</span></code>: gene-SNP pairs of interest, identified elsewhere (eg. fine-mapping analysis).</p></li>
<li><p>The corresponding univariate analysis summary statistics for gene-SNPs from <code class="docutils literal notranslate"><span class="pre">*.batch_*.yaml</span></code> are extracted and saved to <code class="docutils literal notranslate"><span class="pre">*.batch_*.rds</span></code>, creating input to the <code class="docutils literal notranslate"><span class="pre">[posterior]</span></code> step.</p></li>
<li><p>Note the <code class="docutils literal notranslate"><span class="pre">*.batch_*.stdout</span></code> file documents some SNPs found in fine-mapping results but not found in the original <code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> output.</p></li>
</ol>
</li>
</ol>
</div>
<div class="section" id="id1">
<h2>Posterior<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>take all the 13K genes,
if <code class="docutils literal notranslate"><span class="pre">slice_method</span> <span class="pre">=</span> <span class="pre">True</span></code> for those with missing conditions we just drop those corresponding rows and cols in the prior model</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Apply posterior calculations with slice NA and set NaN/Inf 0/1E3, output_posterior_cov = T </span>
<span class="p">[</span><span class="n">posterior_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">analysis_units</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">regions</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">analysis_units</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">mash_model</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="n">mash_model</span> <span class="o">=</span> <span class="n">mash_model</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">posterior_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">posterior_vhat_files</span> <span class="o">=</span> <span class="n">paths</span><span class="p">()</span>
<span class="c1"># eg, if data is saved in R list as data$strong, then</span>
<span class="c1"># when you specify `--data-table-name strong` it will read the data as</span>
<span class="c1"># readRDS(&#39;{_input:r}&#39;)$strong</span>
<span class="kn">parameter:</span> <span class="n">data_table_name</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">bhat_table_name</span> <span class="o">=</span> <span class="s1">&#39;bhat&#39;</span>
<span class="kn">parameter:</span> <span class="n">shat_table_name</span> <span class="o">=</span> <span class="s1">&#39;sbhat&#39;</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="s1">&#39;100&#39;</span>
<span class="c1">##  conditions can be excluded if needs arise. If nothing to exclude keep the default 0</span>
<span class="kn">parameter:</span> <span class="n">exclude_condition</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span><span class="s2">&quot;3&quot;</span><span class="p">]</span>

<span class="kn">parameter:</span> <span class="n">slice_method</span> <span class="o">=</span> <span class="kc">False</span> 
<span class="n">skip_if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">posterior_input</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;No posterior input data to compute on. Please specify it using --posterior-input.&quot;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">posterior_vhat_files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">posterior_vhat_files</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">posterior_input</span><span class="p">),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;length of --posterior-input and --posterior-vhat-files do not agree.&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">posterior_input</span><span class="p">:</span>
    <span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">p</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Cannot find posterior input file ``</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">``&#39;</span><span class="p">)</span>

<span class="kn">input:</span> <span class="n">posterior_input</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">per_chunk</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/cache/mash_output_list_</span><span class="si">{</span><span class="n">_index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="n">mashr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="c1">#library(ttt)</span>
    <span class="n">handle_nan_etc</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">x</span><span class="err">$</span><span class="n">bhat</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">x</span><span class="err">$</span><span class="n">bhat</span><span class="p">))]</span> <span class="o">=</span> <span class="mi">0</span>
      <span class="n">x</span><span class="err">$</span><span class="n">sbhat</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">x</span><span class="err">$</span><span class="n">sbhat</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">infinite</span><span class="p">(</span><span class="n">x</span><span class="err">$</span><span class="n">sbhat</span><span class="p">))]</span> <span class="o">=</span> <span class="mf">1E3</span>
      <span class="k">return</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="c1"># Slice matrices</span>
    <span class="n">slice_and_update_data</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">vhat</span><span class="p">,</span> <span class="n">snps</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">data</span><span class="err">$</span><span class="n">bhat</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="err">$</span><span class="n">bhat</span><span class="p">[</span><span class="n">snps</span><span class="p">,</span> <span class="n">samples</span><span class="p">]</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span>
        <span class="n">data</span><span class="err">$</span><span class="n">sbhat</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="err">$</span><span class="n">sbhat</span><span class="p">[</span><span class="n">snps</span><span class="p">,</span> <span class="n">samples</span><span class="p">]</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span>
        <span class="n">data</span><span class="err">$</span><span class="n">Z</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="err">$</span><span class="n">Z</span><span class="p">[</span><span class="n">snps</span><span class="p">,</span> <span class="n">samples</span><span class="p">]</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span>
        <span class="n">vhat</span> <span class="o">&lt;-</span> <span class="n">vhat</span><span class="p">[</span><span class="n">samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">]</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span>

        <span class="c1"># Filter SNPs and update column names</span>
        <span class="n">data</span><span class="err">$</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="err">$</span><span class="n">snp</span><span class="p">[</span><span class="n">data</span><span class="err">$</span><span class="n">snp</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">snps</span><span class="p">]</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="err">$</span><span class="n">bhat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="err">$</span><span class="n">sbhat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="err">$</span><span class="n">Z</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">vhat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">samples</span>

        <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">vhat</span> <span class="o">=</span> <span class="n">vhat</span><span class="p">))</span>
    <span class="p">}</span>
  
    <span class="c1"># Remove covariance matrices that are not needed</span>
    <span class="n">remove_unnecessary_cov_matrices</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">cov</span><span class="p">,</span> <span class="n">all_samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">unwanted_samples</span> <span class="o">&lt;-</span> <span class="n">setdiff</span><span class="p">(</span><span class="nb">all</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">d</span> <span class="ow">in</span> <span class="n">names</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span> <span class="p">{</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">d</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">unwanted_samples</span> <span class="o">||</span> <span class="n">d</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;ED_&quot;</span><span class="p">,</span> <span class="n">unwanted_samples</span><span class="p">))</span> <span class="p">{</span>
          <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">cov</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Update or adjust the covariance matrices</span>
    <span class="n">adjust_cov_matrices</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">cov</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span> <span class="p">{</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">d</span> <span class="ow">in</span> <span class="n">names</span><span class="p">(</span><span class="n">cov</span><span class="p">))</span> <span class="p">{</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">d</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">samples</span><span class="p">)</span> <span class="p">{</span>
          <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span> <span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">))</span>
          <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]][</span><span class="n">which</span><span class="p">(</span><span class="n">samples</span> <span class="o">==</span> <span class="n">d</span><span class="p">),</span> <span class="n">which</span><span class="p">(</span><span class="n">samples</span> <span class="o">==</span> <span class="n">d</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="mi">1</span>
        <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="n">d</span> <span class="o">==</span> <span class="s2">&quot;identity&quot;</span><span class="p">)</span> <span class="p">{</span>
          <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span> <span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">))</span>
          <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]][</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="mi">1</span>  
        <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]])))</span> <span class="p">{</span>
          <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]][</span><span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">)]</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]][</span><span class="n">samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">]</span>
        <span class="p">}</span>
        <span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">cov</span><span class="p">[[</span><span class="n">d</span><span class="p">]])</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">cov</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Main function to update the covariance in the MASH model</span>
    <span class="n">update_mash_model_cov</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">mash_model</span><span class="p">,</span> <span class="n">all_samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">cov</span> <span class="o">&lt;-</span> <span class="n">mash_model</span><span class="err">$</span><span class="n">fitted_g</span><span class="err">$</span><span class="n">Ulist</span>
      <span class="c1"># Remove matrices that are not required</span>
      <span class="n">cov</span> <span class="o">&lt;-</span> <span class="n">remove_unnecessary_cov_matrices</span><span class="p">(</span><span class="n">cov</span><span class="p">,</span> <span class="n">all_samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
  
      <span class="c1"># Update or reshape the covariance matrices</span>
      <span class="n">cov</span> <span class="o">&lt;-</span> <span class="n">adjust_cov_matrices</span><span class="p">(</span><span class="n">cov</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
  
      <span class="c1"># Update the covariance matrices in the model</span>
      <span class="n">mash_model</span><span class="err">$</span><span class="n">fitted_g</span><span class="err">$</span><span class="n">Ulist</span> <span class="o">&lt;-</span> <span class="n">cov</span>
  
      <span class="c1"># Update the &#39;pi&#39; attribute of the model</span>
      <span class="n">unwanted_samples</span> <span class="o">&lt;-</span> <span class="n">setdiff</span><span class="p">(</span><span class="nb">all</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">s</span> <span class="ow">in</span> <span class="n">unwanted_samples</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">mash_model</span><span class="err">$</span><span class="n">fitted_g</span><span class="err">$</span><span class="n">pi</span> <span class="o">&lt;-</span> <span class="n">mash_model</span><span class="err">$</span><span class="n">fitted_g</span><span class="err">$</span><span class="n">pi</span><span class="p">[</span><span class="o">-</span><span class="n">grep</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">names</span><span class="p">(</span><span class="n">mash_model</span><span class="err">$</span><span class="n">fitted_g</span><span class="err">$</span><span class="n">pi</span><span class="p">))]</span>
      <span class="p">}</span>

      <span class="k">return</span><span class="p">(</span><span class="n">mash_model</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="n">outlist</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">f</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}))</span> <span class="k">try</span><span class="p">({</span>

      <span class="n">data</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="err">$</span><span class="p">{(</span><span class="s1">&#39;$&#39;</span> <span class="o">+</span> <span class="n">data_table_name</span><span class="p">)</span> <span class="k">if</span> <span class="n">data_table_name</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span>
      <span class="n">data</span> <span class="o">&lt;-</span> <span class="n">handle_nan_etc</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

      <span class="k">if</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exclude_condition</span><span class="p">)})[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">){</span>
        <span class="n">message</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Excluding condition $</span><span class="si">{exclude_condition}</span><span class="s2"> from the analysis&quot;</span><span class="p">))</span>
        <span class="n">data</span><span class="err">$</span><span class="n">bhat</span> <span class="o">=</span> <span class="n">data</span><span class="err">$</span><span class="n">bhat</span><span class="p">[,</span><span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exclude_condition</span><span class="p">)})]</span>
        <span class="n">data</span><span class="err">$</span><span class="n">sbhat</span> <span class="o">=</span> <span class="n">data</span><span class="err">$</span><span class="n">sbhat</span><span class="p">[,</span><span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exclude_condition</span><span class="p">)})]</span>
        <span class="n">data</span><span class="err">$</span><span class="n">Z</span> <span class="o">=</span> <span class="n">data</span><span class="err">$</span><span class="n">Z</span><span class="p">[,</span><span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exclude_condition</span><span class="p">)})]</span>
      <span class="p">}</span>

      <span class="n">vhat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="s2">&quot;${vhat_data if len(posterior_vhat_files) == 0 else posterior_vhat_files[_index]}&quot;</span><span class="p">)</span>
      <span class="n">mash_model</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{mash_model}</span><span class="s2">&quot;</span><span class="p">)</span>
  
      <span class="n">slice_method</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;TRUE&#39;</span> <span class="k">if</span> <span class="n">slice_method</span> <span class="k">else</span> <span class="s1">&#39;FALSE&#39;</span><span class="p">}</span>
      <span class="k">if</span><span class="p">(</span><span class="n">slice_method</span><span class="p">){</span>
        <span class="c1"># All additional operations from the second script go here</span>

        <span class="nb">all</span><span class="o">.</span><span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="err">$</span><span class="n">bhat</span><span class="p">)</span>
        <span class="nb">all</span><span class="o">.</span><span class="n">snps</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">data</span><span class="err">$</span><span class="n">bhat</span><span class="p">)</span>    

        <span class="c1">#remove the rows and cols containing NA</span>
        <span class="n">na</span><span class="o">.</span><span class="n">test</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="err">$</span><span class="n">bhat</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span> <span class="o">%&gt;%</span> <span class="n">select_if</span><span class="p">(</span><span class="o">~</span><span class="nb">any</span><span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="o">.</span><span class="p">)))</span> <span class="o">%&gt;%</span> <span class="n">na</span><span class="o">.</span><span class="n">omit</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span>

        <span class="c1">#recording meaningful rows and cols</span>
        <span class="n">samples</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">test</span><span class="p">)</span>
        <span class="n">snps</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">test</span><span class="p">)</span>

        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="nb">all</span><span class="o">.</span><span class="n">snps</span><span class="p">)</span><span class="o">!=</span><span class="n">length</span><span class="p">(</span><span class="n">snps</span><span class="p">)</span> <span class="o">|</span> <span class="n">length</span><span class="p">(</span><span class="nb">all</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span><span class="o">!=</span><span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">)){</span>
            <span class="c1"># slice data matrix</span>
            <span class="n">data</span> <span class="o">&lt;-</span> <span class="n">slice_and_update_data</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">vhat</span><span class="p">,</span> <span class="n">snps</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>

            <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="nb">all</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span><span class="o">!=</span><span class="n">length</span><span class="p">(</span><span class="n">samples</span><span class="p">)){</span>
                <span class="c1">##slice the prior</span>
                <span class="n">mash_model</span> <span class="o">&lt;-</span> <span class="n">update_mash_model_cov</span><span class="p">(</span><span class="n">mash_model</span><span class="p">,</span> <span class="n">all_samples</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
            <span class="p">}</span>
        <span class="p">}</span>
      <span class="p">}</span>

      <span class="n">mash_data</span> <span class="o">=</span> <span class="n">mash_set_data</span><span class="p">(</span><span class="n">data</span><span class="err">$$</span><span class="p">{</span><span class="n">bhat_table_name</span><span class="p">},</span> <span class="n">Shat</span><span class="o">=</span><span class="n">data</span><span class="err">$$</span><span class="p">{</span><span class="n">shat_table_name</span><span class="p">},</span> <span class="n">alpha</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="mi">1</span> <span class="k">if</span> <span class="n">effect_model</span> <span class="o">==</span> <span class="s1">&#39;EZ&#39;</span> <span class="k">else</span> <span class="mi">0</span><span class="p">},</span> <span class="n">V</span><span class="o">=</span><span class="n">vhat</span><span class="p">,</span> <span class="n">zero_Bhat_Shat_reset</span> <span class="o">=</span> <span class="mf">1E3</span><span class="p">)</span>
      <span class="n">mash_output</span> <span class="o">=</span> <span class="n">mash_compute_posterior_matrices</span><span class="p">(</span><span class="n">mash_model</span><span class="p">,</span> <span class="n">mash_data</span><span class="p">,</span> <span class="n">output_posterior_cov</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
      <span class="n">mash_output</span><span class="err">$</span><span class="n">snps</span> <span class="o">=</span> <span class="n">data</span><span class="err">$</span><span class="n">snps</span>
      <span class="n">samplename</span> <span class="o">&lt;-</span> <span class="n">str_split</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="o">.</span><span class="p">[</span><span class="n">length</span><span class="p">(</span><span class="o">.</span><span class="p">)]</span> <span class="o">%&gt;%</span> <span class="n">gsub</span><span class="p">(</span><span class="s1">&#39;.rds&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
      <span class="n">saveRDS</span><span class="p">(</span><span class="n">mash_output</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:d}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="n">samplename</span><span class="p">,</span> <span class="s2">&quot;.posterior.rds&quot;</span><span class="p">))</span>
      <span class="n">outlist</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="n">outlist</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:d}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="n">samplename</span><span class="p">,</span> <span class="s2">&quot;.posterior.rds&quot;</span><span class="p">))</span>

    <span class="p">})</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">outlist</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">posterior_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/mash_output_list_</span><span class="si">{</span><span class="n">output_suffix</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span>
     <span class="n">cd</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span>
     <span class="n">cat</span> <span class="n">mash_output_list_</span><span class="o">*</span><span class="p">[</span><span class="mi">0</span><span class="o">-</span><span class="mi">9</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="n">posterior_file_list</span>
     <span class="n">awk</span> <span class="o">-</span><span class="n">F</span> <span class="s1">&#39;cis_long_table.&#39;</span> <span class="s1">&#39;{print $2}&#39;</span> <span class="n">posterior_file_list</span><span class="o">|</span> <span class="n">awk</span> <span class="o">-</span><span class="n">F</span> <span class="s1">&#39;.posterior.rds&#39;</span> <span class="s1">&#39;{print $1}&#39;</span><span class="o">|</span><span class="n">paste</span> <span class="o">-</span> <span class="n">posterior_file_list</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>
     <span class="n">rm</span> <span class="n">posterior_file_list</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="posterior-contrast">
<h2>Posterior contrast<a class="headerlink" href="#posterior-contrast" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>Add group in this module</p>
<p>e.g. add MiGA-Microglia-scRNA data to this analysis to supplement the Mic cell count shortage, but the MiGA data comes from four sites with four datasets, and here I handled it in such a way that the different Mic’s were always in the same state during the deviation contrast analysis. If Mic is the one be compared, each Mic sample would be set as (n_populations - 1)/n_Mic instead of n_populations</p>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># perform mash posterior contrast for sliced data</span>
<span class="p">[</span><span class="n">mash_posterior_contrast_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">grouping_recipe</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">posterior_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">sum_file</span> <span class="o">=</span> <span class="n">path</span>

<span class="c1"># Extract data from posterior_file</span>
<span class="n">paths_posterior</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">posterior_file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)]</span>
<span class="c1"># Create a dictionary from sum_file for quick lookup</span>
<span class="n">dict_sum</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">sum_file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)])</span>
<span class="c1"># Use genes from posterior_file to fetch corresponding paths from sum_file</span>
<span class="n">paths_sum</span> <span class="o">=</span> <span class="p">[</span><span class="n">dict_sum</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">posterior_file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)]</span>

<span class="kn">input:</span> <span class="n">paths_posterior</span><span class="p">,</span> <span class="n">paired_with</span><span class="o">=</span><span class="s1">&#39;paths_sum&#39;</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s2">bnn</span><span class="si">}</span><span class="s2">_posterior_contrast.rds&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="c1"># Load necessary libraries</span>
    <span class="n">library</span><span class="p">(</span><span class="n">mashr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">RhpcBLASctl</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">magrittr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
    <span class="c1">#library(ttt)</span>

    <span class="c1"># Set number of threads for BLAS operations</span>
    <span class="n">blas_set_num_threads</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Create a function for pairwise contrast columns</span>
    <span class="n">MakePairwiseContrastCols</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">contrast_left</span><span class="p">,</span> <span class="n">orig_vector</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">orig_vector</span><span class="p">[</span><span class="n">contrast_left</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="mi">1</span>
        <span class="n">orig_vector</span><span class="p">[</span><span class="n">contrast_left</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="o">-</span><span class="mi">1</span>
        <span class="n">orig_vector</span>
    <span class="p">}</span>

    <span class="c1"># Function to fit contrast data</span>
    <span class="n">FitContrast</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">orig_mean</span><span class="p">,</span> <span class="n">posterior_mean</span><span class="p">,</span> <span class="n">posterior_vcov</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">population_names</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">posterior_mean</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">str_remove_all</span><span class="p">(</span><span class="s2">&quot;BETA_&quot;</span><span class="p">)</span>

        <span class="n">orig_mean_vector</span> <span class="o">&lt;-</span> <span class="n">orig_mean</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
        <span class="n">names</span><span class="p">(</span><span class="n">orig_mean_vector</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">population_names</span>
        <span class="n">orig_mean_nonzero</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">vector</span><span class="p">(</span><span class="n">orig_mean_vector</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">orig_mean_tested</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">orig_mean_vector</span><span class="p">[</span><span class="n">orig_mean_nonzero</span><span class="p">])</span>
        
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">orig_mean_tested</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">){</span>
            <span class="n">n_populations</span> <span class="o">&lt;-</span> <span class="n">length</span><span class="p">(</span><span class="n">orig_mean_tested</span><span class="p">)</span>

            <span class="n">pairwise_vector</span> <span class="o">&lt;-</span> <span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_populations</span><span class="p">)</span>
            <span class="n">names</span><span class="p">(</span><span class="n">pairwise_vector</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">orig_mean_tested</span>

            <span class="n">grouping</span> <span class="o">&lt;-</span> <span class="n">grouping_all</span><span class="p">[</span><span class="n">orig_mean_tested</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">n_populations</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">n_populations</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">)</span> <span class="p">{</span>
                    <span class="c1">#####1. deviation contrast</span>
                    <span class="n">deviation_contrasts</span> <span class="o">&lt;-</span> <span class="n">rep</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_populations</span><span class="o">^</span><span class="mi">2</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">matrix</span><span class="p">(</span><span class="n">nrow</span> <span class="o">=</span> <span class="n">n_populations</span><span class="p">,</span> <span class="n">ncol</span> <span class="o">=</span> <span class="n">n_populations</span><span class="p">)</span>
                    <span class="n">diag</span><span class="p">(</span><span class="n">deviation_contrasts</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">n_populations</span> <span class="o">-</span> <span class="mi">1</span>
                    <span class="n">rownames</span><span class="p">(</span><span class="n">deviation_contrasts</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">orig_mean_tested</span>
                    <span class="n">colnames</span><span class="p">(</span><span class="n">deviation_contrasts</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">orig_mean_tested</span>
                    <span class="n">deviation_contrasts_tested</span> <span class="o">&lt;-</span> <span class="n">deviation_contrasts</span><span class="p">[,</span> <span class="n">orig_mean_tested</span><span class="p">]</span>

                    <span class="n">unique_groups</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">grouping</span><span class="p">)</span>
                    <span class="k">for</span> <span class="p">(</span><span class="n">grp</span> <span class="ow">in</span> <span class="n">unique_groups</span><span class="p">[</span><span class="n">unique_groups</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">])</span> <span class="p">{</span>
                       <span class="c1">#same celltype (e.g. MIC) with different populations would get 1/n for their weight,</span>
                        <span class="n">diag</span><span class="p">(</span><span class="n">deviation_contrasts_tested</span><span class="p">)[</span><span class="n">grouping</span> <span class="o">==</span> <span class="n">grp</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">n_populations</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">length</span><span class="p">(</span><span class="n">grouping</span><span class="p">[</span><span class="n">grouping</span> <span class="o">==</span> <span class="n">grp</span><span class="p">])</span>
                        <span class="n">deviation_contrasts_tested</span><span class="p">[</span><span class="n">grouping</span> <span class="o">==</span> <span class="n">grp</span><span class="p">,</span> <span class="n">grouping</span> <span class="o">==</span> <span class="n">grp</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">n_populations</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">length</span><span class="p">(</span><span class="n">grouping</span><span class="p">[</span><span class="n">grouping</span> <span class="o">==</span> <span class="n">grp</span><span class="p">])</span>
                    <span class="p">}</span>

                    <span class="n">colnames</span><span class="p">(</span><span class="n">deviation_contrasts_tested</span><span class="p">)</span> <span class="o">%&lt;&gt;%</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;_deviation&quot;</span><span class="p">)</span>

                    <span class="c1">####2. pairwise contrast</span>
                    <span class="n">two_combn</span> <span class="o">&lt;-</span> <span class="n">combn</span><span class="p">(</span><span class="n">orig_mean_tested</span><span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="n">pairwise_names</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">two_combn</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">str_c</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">)</span>
                    <span class="n">pairwise_contrast</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">two_combn</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">MakePairwiseContrastCols</span><span class="p">,</span> <span class="n">pairwise_vector</span><span class="p">)</span>

                    <span class="n">colnames</span><span class="p">(</span><span class="n">pairwise_contrast</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">pairwise_names</span>

                    <span class="c1"># Create a new matrix to store the adjusted values</span>
                    <span class="n">pairwise_contrast_new</span> <span class="o">&lt;-</span> <span class="n">pairwise_contrast</span>

                    <span class="c1"># Loop through each column to archieve such goal: e.g.</span>
                    <span class="c1"># microglia populations would get 1/n_Mic for their weight,</span>
                    <span class="c1"># and Mic vs Mic would still be 1 vs -1 to estimate the internal difference among microglia datasets</span>
                    <span class="k">for</span> <span class="p">(</span><span class="n">col</span> <span class="ow">in</span> <span class="n">colnames</span><span class="p">(</span><span class="n">pairwise_contrast</span><span class="p">))</span> <span class="p">{</span>
                      <span class="c1"># Split column names to get group names</span>
                      <span class="n">groups</span> <span class="o">&lt;-</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">)[[</span><span class="mi">1</span><span class="p">]]</span>

                      <span class="c1"># Get the grouping values for the two groups</span>
                      <span class="n">group_values</span> <span class="o">&lt;-</span> <span class="n">grouping</span><span class="p">[</span><span class="n">names</span><span class="p">(</span><span class="n">grouping</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">groups</span><span class="p">]</span>

                      <span class="c1"># Identify groups with non-zero grouping values</span>
                      <span class="n">relevant_groups</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">group_values</span><span class="p">[</span><span class="n">group_values</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">])</span>

                      <span class="c1"># Check if there are multiple distinct groups</span>
                      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">group_values</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="o">&amp;&amp;</span> <span class="n">length</span><span class="p">(</span><span class="n">relevant_groups</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">distinct_groups</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">group_values</span><span class="p">[</span><span class="n">group_values</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">])</span>

                        <span class="k">for</span> <span class="p">(</span><span class="n">distinct_grp</span> <span class="ow">in</span> <span class="n">distinct_groups</span><span class="p">)</span> <span class="p">{</span>
                          <span class="c1"># Identify rows belonging to the current group</span>
                          <span class="n">rows_in_group</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">grouping</span><span class="p">[</span><span class="n">grouping</span> <span class="o">==</span> <span class="n">distinct_grp</span><span class="p">])</span>

                          <span class="c1"># Adjust the pairwise_contrast values for each row in the group</span>
                          <span class="n">pairwise_contrast_new</span><span class="p">[</span><span class="n">rows_in_group</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">pairwise_contrast</span><span class="p">[</span><span class="n">rows_in_group</span><span class="p">[</span><span class="n">rows_in_group</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">groups</span><span class="p">],</span> <span class="n">col</span><span class="p">]</span> <span class="o">/</span> <span class="n">length</span><span class="p">(</span><span class="n">rows_in_group</span><span class="p">)</span>
                        <span class="p">}</span>
                      <span class="p">}</span>
                    <span class="p">}</span>

                    <span class="c1"># Replace the original matrix with the new one</span>
                    <span class="n">pairwise_contrast</span> <span class="o">&lt;-</span> <span class="n">pairwise_contrast_new</span>

                    <span class="c1">#### 3. combine them</span>
                    <span class="n">contrast_design</span> <span class="o">&lt;-</span> <span class="n">cbind</span><span class="p">(</span><span class="n">deviation_contrasts_tested</span> <span class="o">/</span> <span class="p">(</span><span class="n">n_populations</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">pairwise_contrast</span><span class="p">)</span>

                <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">pairwise_vector</span><span class="p">[</span><span class="n">orig_mean_tested</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="mi">1</span>
                    <span class="n">pairwise_vector</span><span class="p">[</span><span class="n">orig_mean_tested</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="o">-</span><span class="mi">1</span>
                    <span class="n">contrast_design</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">pairwise_vector</span><span class="p">)</span>
                    <span class="n">colnames</span><span class="p">(</span><span class="n">contrast_design</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">str_c</span><span class="p">(</span><span class="n">orig_mean_tested</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">orig_mean_tested</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="p">}</span>

                <span class="n">posterior_mean_subset</span> <span class="o">&lt;-</span> <span class="n">posterior_mean</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
                <span class="n">posterior_mean_subset2</span> <span class="o">&lt;-</span> <span class="n">posterior_mean_subset</span><span class="p">[</span><span class="n">orig_mean_tested</span><span class="p">]</span>
                <span class="n">posterior_vcov_subset</span> <span class="o">&lt;-</span> <span class="n">posterior_vcov</span><span class="p">[,,</span><span class="n">index</span><span class="p">]</span>
                <span class="n">posterior_vcov_subset2</span> <span class="o">&lt;-</span> <span class="n">posterior_vcov_subset</span><span class="p">[</span><span class="n">orig_mean_tested</span><span class="p">,</span><span class="n">orig_mean_tested</span><span class="p">]</span>

                <span class="n">contrast_diff</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">contrast_design</span><span class="p">)</span> <span class="o">%*%</span> <span class="n">posterior_mean_subset2</span>
                <span class="n">contrast_vcov</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">contrast_design</span><span class="p">)</span> <span class="o">%*%</span> <span class="n">posterior_vcov_subset2</span> <span class="o">%*%</span> <span class="n">contrast_design</span>
                <span class="n">contrast_se</span> <span class="o">&lt;-</span> <span class="n">diag</span><span class="p">(</span><span class="n">contrast_vcov</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">sqrt</span>

                <span class="n">contrast_p</span> <span class="o">&lt;-</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">pnorm</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">contrast_diff</span><span class="p">)</span> <span class="o">/</span> <span class="n">contrast_se</span><span class="p">))</span>

                <span class="n">contrast_diff_df</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">contrast_diff</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">as_tibble</span>
                <span class="n">colnames</span><span class="p">(</span><span class="n">contrast_diff_df</span><span class="p">)</span> <span class="o">%&lt;&gt;%</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;mean_contrast_&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
                <span class="n">contrast_se_df</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">contrast_se</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">as_tibble</span>
                <span class="n">colnames</span><span class="p">(</span><span class="n">contrast_se_df</span><span class="p">)</span> <span class="o">%&lt;&gt;%</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;se_contrast_&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
                <span class="n">contrast_p_df</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">contrast_p</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">as_tibble</span>
                <span class="n">colnames</span><span class="p">(</span><span class="n">contrast_p_df</span><span class="p">)</span> <span class="o">%&lt;&gt;%</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;p_contrast_&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>

                <span class="n">contrast_df</span> <span class="o">&lt;-</span> <span class="n">bind_cols</span><span class="p">(</span><span class="n">contrast_diff_df</span><span class="p">,</span> <span class="n">contrast_se_df</span><span class="p">,</span> <span class="n">contrast_p_df</span><span class="p">)</span>
            <span class="p">}</span> <span class="k">else</span> <span class="k">if</span><span class="p">(</span><span class="n">grouping</span><span class="p">[</span><span class="n">orig_mean_tested</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">!=</span><span class="n">grouping</span><span class="p">[</span><span class="n">orig_mean_tested</span><span class="p">][</span><span class="mi">2</span><span class="p">]){</span>
                <span class="n">contrast_vector</span> <span class="o">&lt;-</span> <span class="n">rep</span><span class="p">(</span><span class="n">NA</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">population_names</span><span class="p">))</span>
                <span class="n">names</span><span class="p">(</span><span class="n">contrast_vector</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;mean_contrast_&quot;</span><span class="p">,</span> <span class="n">population_names</span><span class="p">,</span> <span class="s2">&quot;_deviation&quot;</span><span class="p">)</span>
                <span class="n">contrast_df</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">contrast_vector</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">as_tibble</span>
            <span class="p">}</span>
         
        <span class="n">contrast_df</span> <span class="o">&lt;-</span> <span class="n">contrast_df</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span>
        <span class="n">rownames</span><span class="p">(</span><span class="n">contrast_df</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">posterior_mean</span><span class="p">)[</span><span class="n">index</span><span class="p">]</span>
        <span class="k">return</span><span class="p">(</span><span class="n">contrast_df</span><span class="p">)</span>
        <span class="p">}</span>
    
    <span class="p">}</span>

    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{cells}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">){</span>
        <span class="c1"># All the cells</span>
        <span class="n">cells</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;${&quot;</span><span class="p">,</span> <span class="s2">&quot;.join(cells)}&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span> 

        <span class="c1"># Automatically set grouping categories based on the recipe， set0 for the celltypes without multiple populations</span>
        <span class="n">grouping_all</span> <span class="o">&lt;-</span> <span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">cells</span><span class="p">))</span>
        <span class="n">names</span><span class="p">(</span><span class="n">grouping_all</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">cells</span>
        
  
        <span class="c1"># Read groupings from the recipe</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{group1}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">){</span>
            <span class="n">cell_groups</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">(</span>
              <span class="err">$</span><span class="p">{</span><span class="s2">&quot;group1 = c(&quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;&#39;&quot;</span> <span class="o">+</span> <span class="n">item</span> <span class="o">+</span> <span class="s2">&quot;&#39;&quot;</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">group1</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;)&quot;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">group1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> 
              <span class="err">$</span><span class="p">{</span><span class="s2">&quot;, group2 = c(&quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;&#39;&quot;</span> <span class="o">+</span> <span class="n">item</span> <span class="o">+</span> <span class="s2">&quot;&#39;&quot;</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">group2</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;)&quot;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">group2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> 
              <span class="err">$</span><span class="p">{</span><span class="s2">&quot;, group3 = c(&quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;&#39;&quot;</span> <span class="o">+</span> <span class="n">item</span> <span class="o">+</span> <span class="s2">&quot;&#39;&quot;</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">group3</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;)&quot;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">group3</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span>
            <span class="p">)</span>
            <span class="k">if</span><span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">cell_groups</span><span class="p">))</span> <span class="p">{</span>
              <span class="n">cell_groups</span> <span class="o">&lt;-</span> <span class="nb">map</span><span class="p">(</span><span class="n">cell_groups</span><span class="p">,</span> <span class="o">~</span><span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">())</span>
            <span class="p">}</span>
        <span class="p">}</span>
  
        <span class="k">if</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{grouping_recipe}</span><span class="s2">&quot;</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">){</span>
            <span class="n">cell_groups</span> <span class="o">&lt;-</span> <span class="n">readLines</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{grouping_recipe}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">cell_groups</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">cell_groups</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">)[[</span><span class="mi">1</span><span class="p">]])</span>
        <span class="p">}</span>

        <span class="k">if</span><span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">cell_groups</span><span class="p">)){</span>
            <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="n">seq_along</span><span class="p">(</span><span class="n">cell_groups</span><span class="p">))</span> <span class="p">{</span>
              <span class="n">grouping_all</span><span class="p">[</span><span class="n">cell_groups</span><span class="p">[[</span><span class="n">i</span><span class="p">]]]</span> <span class="o">&lt;-</span> <span class="n">i</span>
            <span class="p">}</span>
        <span class="p">}</span>
    <span class="p">}</span>

    
    <span class="c1"># Read the data files</span>
    <span class="n">orig_data</span> <span class="o">&lt;-</span> <span class="n">read_rds</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_paths_sum[0]}</span><span class="s2">&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">bhat</span>
    <span class="n">posterior_data</span> <span class="o">&lt;-</span> <span class="n">read_rds</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">posterior_mean</span> <span class="o">&lt;-</span> <span class="n">posterior_data</span><span class="err">$</span><span class="n">PosteriorMean</span>
    <span class="n">posterior_cov</span> <span class="o">&lt;-</span> <span class="n">posterior_data</span><span class="err">$</span><span class="n">PosteriorCov</span>

    <span class="c1"># Align data and clean-up NaN values</span>
    <span class="n">orig_data</span> <span class="o">&lt;-</span> <span class="n">orig_data</span><span class="p">[,</span> <span class="n">colnames</span><span class="p">(</span><span class="n">posterior_mean</span><span class="p">),</span> <span class="n">drop</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">]</span>
    <span class="n">orig_data</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">orig_data</span><span class="p">))]</span> <span class="o">&lt;-</span> <span class="mi">0</span> <span class="c1"># Placeholder for NaNs</span>

    <span class="c1"># Apply the FitContrast function and consolidate results</span>
    <span class="n">contrast_result</span> <span class="o">&lt;-</span> <span class="nb">map</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">posterior_mean</span><span class="p">),</span> <span class="n">FitContrast</span><span class="p">,</span> <span class="n">orig_data</span><span class="p">,</span> <span class="n">posterior_mean</span><span class="p">,</span> <span class="n">posterior_cov</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">bind_rows</span> <span class="o">%&gt;%</span>
        <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;mean_contrast.*deviation&quot;</span><span class="p">),</span> <span class="n">matches</span><span class="p">(</span><span class="s2">&quot;mean_contrast.*_vs_&quot;</span><span class="p">),</span> 
               <span class="n">matches</span><span class="p">(</span><span class="s2">&quot;se_contrast.*deviation&quot;</span><span class="p">),</span> <span class="n">matches</span><span class="p">(</span><span class="s2">&quot;se_contrast.*_vs_&quot;</span><span class="p">),</span> 
               <span class="n">matches</span><span class="p">(</span><span class="s2">&quot;p_contrast.*deviation&quot;</span><span class="p">),</span> <span class="n">matches</span><span class="p">(</span><span class="s2">&quot;p_contrast.*_vs_&quot;</span><span class="p">))</span>
    <span class="c1">#rownames(contrast_result) &lt;- rownames(posterior_mean)</span>

    <span class="n">write_rds</span><span class="p">(</span><span class="n">contrast_result</span><span class="p">,</span>  <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># merge the contrast data with slice data</span>
<span class="p">[</span><span class="n">mash_posterior_contrast_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/posterior_sum.csv&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> 
        <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
        <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
        <span class="n">library</span><span class="p">(</span><span class="n">ggnewscale</span><span class="p">)</span>
        
        <span class="nb">all</span><span class="o">.</span><span class="n">list</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
       
        <span class="n">p_cut</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">p_cut</span><span class="p">}</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span>

        <span class="n">cells</span><span class="o">&lt;-</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;${&quot;</span><span class="p">,</span><span class="s2">&quot;.join(cells)}&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="p">,</span><span class="s2">&quot;,&quot;</span><span class="p">,</span><span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">character</span> <span class="c1">#ggnewscale cannot use a specified order, I can not find a good way to order them by category for now</span>
        <span class="n">conditions</span> <span class="o">&lt;-</span><span class="n">combn</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">%&gt;%</span><span class="n">apply</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">str_c</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">)</span> 
        
        <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">ncol</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">conditions</span><span class="p">),</span> <span class="n">nrow</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">conditions</span>
        <span class="n">rownames</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span> <span class="s2">&quot;n_sig_snp&quot;</span><span class="p">,</span> <span class="s2">&quot;n_snp&quot;</span><span class="p">,</span><span class="s2">&quot;n_sig_feature&quot;</span><span class="p">,</span><span class="s2">&quot;n_all_feature&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">con</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">feature</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">feature</span> <span class="o">&lt;-</span> <span class="mi">0</span>

            <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="nb">all</span><span class="o">.</span><span class="n">list</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="nb">all</span><span class="o">.</span><span class="n">list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                <span class="n">p</span><span class="o">.</span><span class="n">mtx</span> <span class="o">&lt;-</span> <span class="n">tmp</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;p_contrast.*_vs_&quot;</span><span class="p">))</span>
                <span class="n">p</span><span class="o">.</span><span class="n">mtx</span><span class="o">.</span><span class="n">con</span> <span class="o">&lt;-</span> <span class="n">p</span><span class="o">.</span><span class="n">mtx</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="n">con</span><span class="p">))</span>
                <span class="c1"># print(n.sig.snp)</span>
                <span class="k">if</span><span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">mtx</span><span class="o">.</span><span class="n">con</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">){</span>
                    <span class="n">p</span><span class="o">.</span><span class="n">mtx</span><span class="o">.</span><span class="n">con</span><span class="o">&lt;-</span><span class="n">na</span><span class="o">.</span><span class="n">omit</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">mtx</span><span class="o">.</span><span class="n">con</span><span class="p">)</span>
                    <span class="n">n</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="nb">sum</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">mtx</span><span class="o">.</span><span class="n">con</span> <span class="o">&lt;</span> <span class="n">p_cut</span><span class="p">)</span>
                    <span class="n">n</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">nrow</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">mtx</span><span class="o">.</span><span class="n">con</span><span class="p">)</span>
                <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">n</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span><span class="mi">0</span>
                <span class="p">}</span>
                <span class="c1"># print(n.snp)</span>
                <span class="n">n</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">feature</span> <span class="o">&lt;-</span> <span class="n">ifelse</span><span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="n">n</span><span class="o">.</span><span class="n">feature</span><span class="o">&lt;-</span> <span class="n">ifelse</span> <span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">snp</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

                <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span> <span class="o">+</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span>
                <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">snp</span> <span class="o">+</span> <span class="n">n</span><span class="o">.</span><span class="n">snp</span>
                <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">feature</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">feature</span> <span class="o">+</span> <span class="n">n</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">feature</span>
                <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">feature</span> <span class="o">&lt;-</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">feature</span> <span class="o">+</span> <span class="n">n</span><span class="o">.</span><span class="n">feature</span>
            <span class="p">}</span>
            <span class="n">df</span><span class="p">[,</span> <span class="n">con</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">snp</span><span class="p">,</span><span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">snp</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">feature</span><span class="p">,</span><span class="n">n</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">feature</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="n">write</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot contrast result</span>
<span class="p">[</span><span class="n">mash_posterior_contrast_3</span><span class="p">,</span> <span class="n">posterior_cntrast_plot</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/posterior_sum.png&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> 
        <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
        <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
        <span class="n">library</span><span class="p">(</span><span class="n">ggnewscale</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="s2">&quot;${_input:a}&quot;</span><span class="p">,</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">&lt;-</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;DLPFC_&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="n">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">df</span><span class="p">))</span> <span class="p">{</span>
            <span class="n">con1</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">)[</span><span class="n">i</span><span class="p">],</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span> <span class="mi">1</span><span class="p">]</span>
            <span class="n">con2</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">)[</span><span class="n">i</span><span class="p">],</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span> <span class="mi">2</span><span class="p">]</span>

            <span class="k">if</span> <span class="p">(</span><span class="n">con1</span> <span class="o">&gt;</span> <span class="n">con2</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">new</span><span class="o">.</span><span class="kp">name</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="n">con2</span><span class="p">,</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">con1</span><span class="p">)</span>
                <span class="n">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">new</span><span class="o">.</span><span class="kp">name</span>
            <span class="p">}</span>
        <span class="p">}</span>

        <span class="c1">## summarizxse with approach 1: snp-feature pair</span>
        <span class="n">snp</span><span class="o">.</span><span class="n">ratio</span> <span class="o">&lt;-</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;n_sig_snp&quot;</span><span class="p">,</span> <span class="p">]</span> <span class="o">/</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;n_snp&quot;</span><span class="p">,</span> <span class="p">]</span>
        <span class="n">snp</span><span class="o">.</span><span class="n">ratio</span> <span class="o">&lt;-</span> <span class="n">snp</span><span class="o">.</span><span class="n">ratio</span> <span class="o">%&gt;%</span>
            <span class="n">t</span><span class="p">()</span> <span class="o">%&gt;%</span>
            <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">snp</span><span class="o">.</span><span class="n">ratio</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">&quot;ratio&quot;</span>
        <span class="n">snp</span><span class="o">.</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">&lt;-</span> <span class="s2">&quot;snp&quot;</span>

        <span class="c1">## summarize with approach 2: feature</span>
        <span class="n">fet</span><span class="o">.</span><span class="n">ratio</span> <span class="o">&lt;-</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;n_sig_feature&quot;</span><span class="p">,</span> <span class="p">]</span> <span class="o">/</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;n_all_feature&quot;</span><span class="p">,</span> <span class="p">]</span>
        <span class="n">fet</span><span class="o">.</span><span class="n">ratio</span> <span class="o">&lt;-</span> <span class="n">fet</span><span class="o">.</span><span class="n">ratio</span> <span class="o">%&gt;%</span>
            <span class="n">t</span><span class="p">()</span> <span class="o">%&gt;%</span>
            <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
        <span class="n">rownames</span><span class="p">(</span><span class="n">fet</span><span class="o">.</span><span class="n">ratio</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">fet</span><span class="o">.</span><span class="n">ratio</span><span class="p">),</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span> <span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">fet</span><span class="o">.</span><span class="n">ratio</span><span class="p">),</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">fet</span><span class="o">.</span><span class="n">ratio</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">&quot;ratio&quot;</span>
        <span class="n">fet</span><span class="o">.</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">&lt;-</span> <span class="s2">&quot;feature&quot;</span>
        <span class="n">ratio</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="n">snp</span><span class="o">.</span><span class="n">ratio</span><span class="p">,</span> <span class="n">fet</span><span class="o">.</span><span class="n">ratio</span><span class="p">)</span>

        <span class="c1">## I need to add the below to make it Simmetrie</span>

        <span class="n">cons</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">ratio</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="o">.</span><span class="p">[,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">%&gt;%</span>
            <span class="n">unique</span><span class="p">()</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">cons</span><span class="p">))</span> <span class="p">{</span>
            <span class="n">new</span><span class="o">.</span><span class="kp">name</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="n">cons</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">cons</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">ratio</span><span class="p">[</span><span class="n">new</span><span class="o">.</span><span class="kp">name</span><span class="p">,</span> <span class="p">]</span> <span class="o">&lt;-</span> <span class="mi">0</span>
        <span class="p">}</span>

        <span class="n">ratio</span><span class="err">$</span><span class="n">con1</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">ratio</span><span class="p">),</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">ratio</span><span class="err">$</span><span class="n">con2</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">ratio</span><span class="p">),</span> <span class="s2">&quot;_vs_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span> <span class="mi">2</span><span class="p">]</span>

        <span class="c1">## prepare for the plot, score1 is for snp-feature pair, score2 is for feature only</span>
        <span class="n">ratio</span><span class="err">$</span><span class="n">score1</span> <span class="o">&lt;-</span> <span class="n">ratio</span><span class="err">$</span><span class="n">score2</span> <span class="o">&lt;-</span> <span class="mi">0</span>
        <span class="n">ratio</span><span class="err">$</span><span class="n">score1</span><span class="p">[</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">ratio</span><span class="err">$</span><span class="n">ratio</span><span class="p">[</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">]</span>
        <span class="n">ratio</span><span class="err">$</span><span class="n">score2</span><span class="p">[</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">==</span> <span class="s2">&quot;feature&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">ratio</span><span class="err">$</span><span class="n">ratio</span><span class="p">[</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">==</span> <span class="s2">&quot;feature&quot;</span><span class="p">]</span>
        <span class="n">ratio</span><span class="err">$</span><span class="n">label</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">ratio</span><span class="err">$</span><span class="n">ratio</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">)</span>
        <span class="n">ratio</span><span class="err">$</span><span class="n">label</span><span class="p">[</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">NA</span>

        <span class="c1"># plot</span>
        <span class="n">num_cols</span> <span class="o">&lt;-</span> <span class="n">length</span><span class="p">(</span><span class="n">cons</span><span class="p">)</span>
        <span class="n">height</span> <span class="o">&lt;-</span> <span class="n">width</span> <span class="o">&lt;-</span> <span class="mi">4</span> <span class="o">+</span> <span class="n">num_cols</span> <span class="o">*</span> <span class="mf">0.5</span> 
        <span class="n">ggplot</span><span class="p">(</span><span class="n">ratio</span><span class="p">[</span><span class="n">ratio</span><span class="err">$</span><span class="n">group</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">,</span> <span class="p">],</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">con1</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">con2</span><span class="p">))</span> <span class="o">+</span>
            <span class="n">geom_tile</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span> <span class="o">=</span> <span class="n">score1</span><span class="p">))</span> <span class="o">+</span>
            <span class="n">scale_fill_gradient2</span><span class="p">(</span><span class="s2">&quot;SNP_Feature pair&quot;</span><span class="p">,</span>
                <span class="n">low</span> <span class="o">=</span> <span class="s2">&quot;#762A83&quot;</span><span class="p">,</span> <span class="n">mid</span> <span class="o">=</span> <span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="n">high</span> <span class="o">=</span> <span class="s2">&quot;#1B7837&quot;</span>
            <span class="p">)</span> <span class="o">+</span>
            <span class="n">new_scale</span><span class="p">(</span><span class="s2">&quot;fill&quot;</span><span class="p">)</span> <span class="o">+</span>
            <span class="n">geom_tile</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span> <span class="o">=</span> <span class="n">score2</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">subset</span><span class="p">(</span><span class="n">ratio</span><span class="p">,</span> <span class="n">group</span> <span class="o">!=</span> <span class="s2">&quot;snp&quot;</span><span class="p">))</span> <span class="o">+</span>
            <span class="n">scale_fill_gradient2</span><span class="p">(</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span>
                <span class="n">low</span> <span class="o">=</span> <span class="s2">&quot;#1B7837&quot;</span><span class="p">,</span> <span class="n">mid</span> <span class="o">=</span> <span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="n">high</span> <span class="o">=</span> <span class="s2">&quot;#762A83&quot;</span>
            <span class="p">)</span> <span class="o">+</span>
            <span class="n">geom_text</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">ratio</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">label</span> <span class="o">=</span> <span class="n">label</span><span class="p">))</span> <span class="o">+</span>
            <span class="n">theme_bw</span><span class="p">()</span>
        <span class="c1">#geom_text(data=ratio, aes(label = label, color = factor(group))) +theme_bw()</span>
        <span class="c1">#ggsave(gsub(&quot;.csv&quot;,&quot;.png&quot;,filename))</span>

        <span class="n">ggsave</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">width</span> <span class="o">=</span> <span class="n">width</span><span class="p">,</span> <span class="n">height</span> <span class="o">=</span> <span class="n">height</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="feature-score-from-contrast-results">
<h2>Feature score from contrast results<a class="headerlink" href="#feature-score-from-contrast-results" title="Permalink to this headline">¶</a></h2>
<p><strong>FIXME: Haven’t cleaned up yet, need to talk to Gao to decide which one should be kept.</strong></p>
<div class="section" id="feature-score-meta">
<h3>feature_score_meta<a class="headerlink" href="#feature-score-meta" title="Permalink to this headline">¶</a></h3>
<p>Meta approach: Meta analysis with each cell deviation contrast result of each feature (a loop for each column in deviation results). And then perform meta analysis with pairwise contrasts to get a pvalue to find to understand what the specific differences are. But there is a problem that meta analysis with so many snps would cost too much time to compute, I am trying to downsample for input, Dan suggest to LD prune with a more permissive threshold.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute feature score from contrast results</span>
<span class="p">[</span><span class="n">feature_score_meta_1</span><span class="p">]</span>
<span class="c1">#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]</span>
<span class="kn">parameter:</span> <span class="n">plink_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/mfs/cluster/bin/plink1.9.10/plink&quot;</span>
<span class="kn">parameter:</span> <span class="n">bfile_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/WashU_genotype/genotype_qc/MAP_Brain-xQTL_Gwas_geno_0.1_maf_0.0005.filtered.&quot;</span>
<span class="kn">parameter:</span> <span class="n">extract</span> <span class="o">=</span> <span class="s2">&quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/snp.list&quot;</span>
<span class="kn">parameter:</span> <span class="n">window_size</span> <span class="o">=</span> <span class="s1">&#39;100&#39;</span>
<span class="kn">parameter:</span> <span class="n">step_size</span> <span class="o">=</span> <span class="s1">&#39;10&#39;</span>
<span class="kn">parameter:</span> <span class="n">r2_threshold</span> <span class="o">=</span> <span class="s1">&#39;0.2&#39;</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="s1">&#39;100&#39;</span>
<span class="kn">parameter:</span> <span class="n">LD_prune</span> <span class="o">=</span> <span class="s2">&quot;TRUE&quot;</span>
<span class="kn">parameter:</span> <span class="n">downsample_ratio</span> <span class="o">=</span> <span class="s1">&#39;1&#39;</span>
<span class="kn">parameter:</span> <span class="n">meta_method</span> <span class="o">=</span> <span class="s1">&#39;REML&#39;</span>
<span class="kn">parameter:</span> <span class="n">LDcache</span> <span class="o">=</span> <span class="s2">&quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/&quot;</span>
<span class="kn">parameter:</span> <span class="n">contrast_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">contrast_input</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">per_chunk</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/feature_score_metaLD/cache/mash_posterior_contrast_featurescore</span><span class="si">{</span><span class="n">_index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">.rds&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="c1"># Library Loading</span>
    <span class="n">suppressMessages</span><span class="p">({</span>
      <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>
      <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
      <span class="n">library</span><span class="p">(</span><span class="n">metafor</span><span class="p">)</span>
    <span class="p">})</span>

    <span class="c1"># Environment Configuration</span>
    <span class="n">plink_path</span> <span class="o">&lt;-</span> <span class="s2">&quot;$</span><span class="si">{plink_path}</span><span class="s2">&quot;</span>
    <span class="n">bfile_path</span> <span class="o">&lt;-</span> <span class="s2">&quot;$</span><span class="si">{bfile_path}</span><span class="s2">&quot;</span>
    <span class="n">indep_pairwise</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{window_size}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{step_size}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{r2_threshold}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
    <span class="nb">set</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">999</span><span class="p">)</span>

    <span class="c1"># Define helper functions</span>

    <span class="n">plink_ld_pruning</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="n">bfile_path</span><span class="p">,</span> <span class="n">LDcache</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">plink_path</span><span class="p">,</span> <span class="n">indep_pairwise</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">extract</span> <span class="o">&lt;-</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{LDcache}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;_snp.list&quot;</span><span class="p">)</span>
      <span class="n">output</span> <span class="o">&lt;-</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{LDcache}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;_output&quot;</span><span class="p">)</span>
      <span class="n">tmp_contrast_results</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>
      <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">),</span> <span class="n">extract</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
      <span class="c1"># system(paste0(&quot;rm &quot;, output, &quot;*&quot;), intern = TRUE)</span>
      <span class="n">bfile</span> <span class="o">&lt;-</span> <span class="n">str_c</span><span class="p">(</span><span class="n">bfile_path</span><span class="p">,</span> <span class="nb">chr</span><span class="p">)</span>
      <span class="n">command</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="n">plink_path</span><span class="p">,</span> <span class="s2">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">bfile</span><span class="p">,</span> <span class="s2">&quot;--extract&quot;</span><span class="p">,</span> <span class="n">extract</span><span class="p">,</span> <span class="s2">&quot;--indep-pairwise&quot;</span><span class="p">,</span> <span class="n">indep_pairwise</span><span class="p">,</span> <span class="s2">&quot;--out&quot;</span><span class="p">,</span> <span class="n">output</span><span class="p">)</span>
      <span class="n">system</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">intern</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
      <span class="n">ld_output</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{LDcache}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;_output.prune.in&quot;</span><span class="p">))</span>
      <span class="k">return</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">[</span><span class="n">ld_output</span><span class="err">$</span><span class="n">V1</span><span class="p">,</span> <span class="p">]</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">())</span>
    <span class="p">}</span>

    <span class="n">calculate_feature_scores</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">,</span> <span class="n">meta_method</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">effect_sizes</span> <span class="o">&lt;-</span> <span class="n">tmp_contrast_results</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;mean_contrast.*deviation&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>
      <span class="n">se_values</span> <span class="o">&lt;-</span> <span class="n">tmp_contrast_results</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;se_contrast.*deviation&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>
      <span class="n">feature_scores</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">effect_sizes</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">effect_sizes_condition</span> <span class="o">&lt;-</span> <span class="n">effect_sizes</span><span class="p">[,</span> <span class="n">i</span><span class="p">]</span>
        <span class="n">se_values_condition</span> <span class="o">&lt;-</span> <span class="n">se_values</span><span class="p">[,</span> <span class="n">i</span><span class="p">]</span>
        <span class="n">absolute_effect_sizes</span> <span class="o">&lt;-</span> <span class="nb">abs</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">effect_sizes_condition</span><span class="p">))</span>
        <span class="n">pairwise_standard_errors</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">se_values_condition</span><span class="p">)</span>
        <span class="n">meta_result</span> <span class="o">&lt;-</span> <span class="n">rma</span><span class="p">(</span><span class="n">yi</span> <span class="o">=</span> <span class="n">absolute_effect_sizes</span><span class="p">,</span> <span class="n">sei</span> <span class="o">=</span> <span class="n">pairwise_standard_errors</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="n">meta_method</span><span class="p">)</span>
        <span class="n">z_scores</span> <span class="o">&lt;-</span> <span class="n">meta_result</span><span class="err">$</span><span class="n">b</span> <span class="o">/</span> <span class="n">meta_result</span><span class="err">$</span><span class="n">se</span>
        <span class="n">feature_scores_condition</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">ZScore</span> <span class="o">=</span> <span class="n">z_scores</span><span class="p">)</span>
        <span class="n">feature_scores</span> <span class="o">&lt;-</span> <span class="n">rbindlist</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">,</span> <span class="n">feature_scores_condition</span><span class="p">))</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Main Loop</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}))</span> <span class="k">try</span><span class="p">({</span>
      <span class="nb">chr</span> <span class="o">&lt;-</span> <span class="n">res</span> <span class="o">%&gt;%</span> <span class="n">basename</span><span class="p">()</span> <span class="o">%&gt;%</span> <span class="n">str_extract</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">.(.*?)</span><span class="se">\\</span><span class="s2">.&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="n">str_replace_all</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">str_replace_all</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
      <span class="n">gene</span> <span class="o">&lt;-</span> <span class="n">basename</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;norminal.cis_long_table.&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="o">.</span><span class="p">[,</span> <span class="mi">2</span><span class="p">]</span> <span class="o">%&gt;%</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;_posterior_contrast.rds&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>

      <span class="n">tmp_contrast_results</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>

      <span class="n">LD_prune</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;TRUE&#39;</span> <span class="k">if</span> <span class="n">LD_prune</span> <span class="k">else</span> <span class="s1">&#39;FALSE&#39;</span><span class="p">}</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">LD_prune</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">tmp_contrast_results</span> <span class="o">&lt;-</span> <span class="n">plink_ld_pruning</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="n">bfile_path</span><span class="p">,</span> <span class="n">LDcache</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">plink_path</span><span class="p">,</span> <span class="n">indep_pairwise</span><span class="p">)</span>
      <span class="p">}</span>

      <span class="k">if</span> <span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">tmp_contrast_results</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;mean_contrast.*deviation&quot;</span><span class="p">)))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">feature_scores</span> <span class="o">&lt;-</span> <span class="n">calculate_feature_scores</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{meta_method}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">gene</span>
        <span class="n">condition_name</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">effect_sizes</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;mean_contrast_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;_deviation&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
        <span class="n">feature_scores</span> <span class="o">&lt;-</span> <span class="n">feature_scores</span><span class="p">[,</span> <span class="n">condition</span> <span class="o">:=</span> <span class="n">condition_name</span><span class="p">]</span>
        <span class="n">saveRDS</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">,</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:d}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.rds&quot;</span><span class="p">,</span> <span class="s2">&quot;_featurescore.rds&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">res</span><span class="p">))))</span>
      <span class="p">}</span>

      <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">out</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">feature_scores</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">merge</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="n">feature_scores</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="nb">all</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>
    <span class="p">})</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="feature-score-finemap">
<h3>feature_score_finemap<a class="headerlink" href="#feature-score-finemap" title="Permalink to this headline">¶</a></h3>
<p>feature score with fine mapped QTL (the top 1 in each CS)</p>
<p>Fine-mapped approach(more recommend by Dan). pick the top SNP in each CS from each cell type fine mapped results. Then pick the most significant one from contrast result. Get the z-score from that snp, which should be the score of that cell type.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute feature score from contrast results with eQTL and pQTL finemapped signals</span>
<span class="p">[</span><span class="n">feature_score_finemap_1</span><span class="p">]</span>
<span class="c1">#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="s1">&#39;100&#39;</span>
<span class="kn">parameter:</span> <span class="n">pfine_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/QTL_fine_mapping/pqtl.all_variants.tsv&quot;</span>
<span class="kn">parameter:</span> <span class="n">efine_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/vast/hpc/csg/hs3393/susie_eqtl/Result_new/tsv/&quot;</span>
<span class="kn">parameter:</span> <span class="n">efine_suffix</span> <span class="o">=</span> <span class="s2">&quot;.unisusie.fit.variant.tsv&quot;</span>
<span class="kn">parameter:</span> <span class="n">gene_ref_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list&quot;</span>
<span class="kn">parameter:</span> <span class="n">contrast_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">contrast_input</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">per_chunk</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/feature_score_finemap/cache/mash_posterior_contrast_featurescore</span><span class="si">{</span><span class="n">_index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">.rds&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
   <span class="c1"># Load necessary libraries</span>
    <span class="n">suppressMessages</span><span class="p">({</span>
        <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>
        <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
    <span class="p">})</span>

    <span class="c1"># Function to calculate scores from fine-mapping results</span>
    <span class="n">score_from_cs</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">fine</span><span class="o">.</span><span class="n">file</span><span class="p">,</span> <span class="n">contrast_results</span><span class="p">,</span> <span class="n">con</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">CSs</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">fine</span><span class="o">.</span><span class="n">file</span><span class="err">$</span><span class="n">cs_order</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="o">.</span><span class="p">[</span><span class="o">.</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>
        <span class="nb">max</span><span class="o">.</span><span class="n">cs</span><span class="o">.</span><span class="n">df</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">cs</span> <span class="ow">in</span> <span class="n">CSs</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">fine</span><span class="o">.</span><span class="n">file</span><span class="p">[</span><span class="n">fine</span><span class="o">.</span><span class="n">file</span><span class="err">$</span><span class="n">cs_order</span> <span class="o">==</span> <span class="n">cs</span><span class="p">,</span> <span class="p">]</span>
            <span class="nb">max</span><span class="o">.</span><span class="n">cs</span><span class="o">.</span><span class="n">df</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="nb">max</span><span class="o">.</span><span class="n">cs</span><span class="o">.</span><span class="n">df</span><span class="p">,</span> <span class="n">tmp</span><span class="p">[</span><span class="n">tmp</span><span class="err">$</span><span class="n">pip</span> <span class="o">==</span> <span class="nb">max</span><span class="p">(</span><span class="n">tmp</span><span class="err">$</span><span class="n">pip</span><span class="p">),</span> <span class="p">])</span>
        <span class="p">}</span>
        <span class="n">snps</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">contrast_results</span><span class="p">),</span> <span class="nb">max</span><span class="o">.</span><span class="n">cs</span><span class="o">.</span><span class="n">df</span><span class="err">$</span><span class="n">variants</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">snps</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="k">return</span><span class="p">(</span><span class="n">NA</span><span class="p">)</span>

        <span class="c1"># Extract contrast p-values</span>
        <span class="n">contrast_p</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">select</span><span class="p">(</span><span class="n">contrast_results</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(),</span> <span class="n">matches</span><span class="p">(</span><span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;p_contrast_&quot;</span><span class="p">,</span> <span class="n">con</span><span class="p">,</span> <span class="s2">&quot;_deviation&quot;</span><span class="p">))))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">contrast_p</span> <span class="o">&lt;-</span> <span class="n">contrast_results</span><span class="p">[</span><span class="n">snps</span><span class="p">,</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;p_contrast_&quot;</span><span class="p">,</span> <span class="n">con</span><span class="p">,</span> <span class="s2">&quot;_deviation&quot;</span><span class="p">),</span> <span class="n">drop</span> <span class="o">=</span> <span class="n">F</span><span class="p">]</span>
        <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">select</span><span class="p">(</span><span class="n">contrast_results</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(),</span> <span class="n">matches</span><span class="p">(</span><span class="s2">&quot;p_contrast.*_vs_*&quot;</span><span class="p">)))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">contrast_p</span> <span class="o">&lt;-</span> <span class="n">contrast_results</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;p_contrast.*_vs_*&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="o">.</span><span class="p">[</span><span class="n">snps</span><span class="p">,</span> <span class="p">,</span> <span class="n">drop</span> <span class="o">=</span> <span class="n">F</span><span class="p">]</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">contrast_p</span><span class="p">))</span> <span class="k">return</span><span class="p">(</span><span class="n">NA</span><span class="p">)</span>

        <span class="nb">max</span><span class="o">.</span><span class="n">snp</span> <span class="o">&lt;-</span> <span class="n">contrast_p</span><span class="p">[</span><span class="n">contrast_p</span><span class="p">[,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="nb">max</span><span class="p">(</span><span class="n">contrast_p</span><span class="p">[,</span> <span class="mi">1</span><span class="p">]),</span> <span class="p">,</span> <span class="n">drop</span> <span class="o">=</span> <span class="n">F</span><span class="p">]</span> <span class="o">%&gt;%</span> <span class="n">rownames</span><span class="p">()</span>
        <span class="n">score</span> <span class="o">&lt;-</span> <span class="nb">max</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">contrast_results</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;mean_contrast.*_vs_*&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="o">.</span><span class="p">[</span><span class="nb">max</span><span class="o">.</span><span class="n">snp</span><span class="p">,</span> <span class="p">])</span> <span class="o">/</span>
                     <span class="p">(</span><span class="n">contrast_results</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;se_contrast.*_vs_*&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="o">.</span><span class="p">[</span><span class="nb">max</span><span class="o">.</span><span class="n">snp</span><span class="p">,</span> <span class="p">]))</span>
        <span class="k">return</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Read pQTL file</span>
    <span class="n">pfine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{pfine_path}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">pfine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="err">$</span><span class="n">Gene</span> <span class="o">&lt;-</span> <span class="n">str_split</span><span class="p">(</span><span class="n">pfine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="err">$</span><span class="n">molecular_trait_id</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)[,</span> <span class="mi">2</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">ref</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{gene_ref_path}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">()</span>

    <span class="c1"># Loop through each input and compute scores</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}))</span> <span class="p">{</span>
        <span class="n">tmp_contrast_results</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>

        <span class="n">gene</span> <span class="o">&lt;-</span> <span class="n">basename</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;norminal.cis_long_table.&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="o">.</span><span class="p">[,</span> <span class="mi">2</span><span class="p">]</span> <span class="o">%&gt;%</span>
            <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;_posterior_contrast.rds&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
        <span class="n">ensemble</span> <span class="o">&lt;-</span> <span class="n">gene</span><span class="o">.</span><span class="n">ref</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">ref</span><span class="err">$</span><span class="n">V5</span> <span class="o">==</span> <span class="n">gene</span><span class="p">,</span> <span class="p">]</span><span class="err">$</span><span class="n">V4</span>
        <span class="n">efine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{efine_path}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">ensemble</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{efine_suffix}</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">full</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
        <span class="n">pfine</span><span class="o">.</span><span class="n">file</span> <span class="o">&lt;-</span> <span class="n">pfine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="n">pfine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="err">$</span><span class="n">Gene</span> <span class="o">==</span> <span class="n">gene</span> <span class="o">&amp;</span> <span class="n">pfine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="err">$</span><span class="n">cs_order</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="p">]</span>

        <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">pfine</span><span class="o">.</span><span class="n">file</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">message</span><span class="p">(</span><span class="s2">&quot;Extracting signal from pQTL in &quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">score_from_cs</span><span class="p">(</span><span class="n">pfine</span><span class="o">.</span><span class="n">file</span><span class="p">,</span> <span class="n">tmp_contrast_results</span><span class="p">,</span> <span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">efine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">fine</span><span class="o">.</span><span class="n">conditions</span> <span class="o">&lt;-</span> <span class="n">basename</span><span class="p">(</span><span class="n">efine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="n">sub</span><span class="p">(</span><span class="s2">&quot;demo.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="n">sub</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">ensemble</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{efine_suffix}</span><span class="s2">&quot;</span><span class="p">),</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="o">.</span><span class="p">[</span><span class="o">.</span> <span class="o">!=</span> <span class="s2">&quot;ALL&quot;</span><span class="p">]</span> <span class="o">%&gt;%</span>
                <span class="o">.</span><span class="p">[</span><span class="o">.</span> <span class="o">!=</span> <span class="s2">&quot;End&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">con</span> <span class="ow">in</span> <span class="n">fine</span><span class="o">.</span><span class="n">conditions</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">con</span><span class="o">.</span><span class="n">file</span> <span class="o">&lt;-</span> <span class="n">efine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="n">con</span><span class="p">,</span> <span class="n">efine</span><span class="o">.</span><span class="n">mapped</span><span class="o">.</span><span class="n">result</span><span class="p">)]</span> <span class="o">%&gt;%</span> <span class="n">fread</span><span class="p">()</span>
                <span class="n">con</span><span class="o">.</span><span class="n">fine</span><span class="o">.</span><span class="n">file</span> <span class="o">&lt;-</span> <span class="n">con</span><span class="o">.</span><span class="n">file</span><span class="p">[</span><span class="n">con</span><span class="o">.</span><span class="n">file</span><span class="err">$</span><span class="n">cs_order</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="p">]</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">con</span><span class="o">.</span><span class="n">fine</span><span class="o">.</span><span class="n">file</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
                    <span class="n">df</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="n">con</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">score_from_cs</span><span class="p">(</span><span class="n">con</span><span class="o">.</span><span class="n">fine</span><span class="o">.</span><span class="n">file</span><span class="p">,</span> <span class="n">tmp_contrast_results</span><span class="p">,</span> <span class="n">con</span><span class="p">)</span>
                <span class="p">}</span>
            <span class="p">}</span>
        <span class="p">}</span>
        <span class="n">saveRDS</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{cwd}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;/feature_score_finemap/cache/&quot;</span><span class="p">,</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.rds&quot;</span><span class="p">,</span><span class="s2">&quot;_featurescore.rds&quot;</span><span class="p">,</span><span class="n">basename</span><span class="p">(</span><span class="n">res</span><span class="p">))))</span>
        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">rbindlist</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">keep</span><span class="o">.</span><span class="n">rownames</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)),</span> <span class="n">use</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># merge the feature score from contrast results with eQTL and pQTL finemapped signals</span>
<span class="p">[</span><span class="n">feature_score_finemap_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/feature_score_finemap/posterior_feature_score_sum.csv&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> 
    <span class="n">suppressMessages</span><span class="p">(</span><span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">))</span>
    <span class="n">suppressMessages</span><span class="p">(</span><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">))</span>
    
    <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">()</span>

    <span class="nb">all</span><span class="o">.</span><span class="n">list</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="nb">all</span><span class="o">.</span><span class="n">list</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">feature_scores</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="c1"># Print the feature scores for each condition</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">))</span> <span class="p">{</span>
           <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">out</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">out</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">,</span> <span class="n">keep</span><span class="o">.</span><span class="n">rownames</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
              <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">out</span> <span class="o">&lt;-</span>  <span class="n">rbindlist</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">,</span> <span class="n">keep</span><span class="o">.</span><span class="n">rownames</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)),</span> <span class="n">use</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
              <span class="p">}</span>
        <span class="p">}</span>
    <span class="p">}</span>
    <span class="n">write</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="feature-score-nsig">
<h3>feature_score_nsig<a class="headerlink" href="#feature-score-nsig" title="Permalink to this headline">¶</a></h3>
<p>Calculate the number of meaningful SNPs in each feature of each cell type (actually, ratio)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute feature score from contrast results</span>
<span class="p">[</span><span class="n">feature_score_nsig_1</span><span class="p">]</span>
<span class="c1">#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="s1">&#39;100&#39;</span>
<span class="kn">parameter:</span> <span class="n">contrast_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">contrast_input</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">per_chunk</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/feature_score_nsig/cache/mash_posterior_contrast_featurescore_nsig</span><span class="si">{</span><span class="n">_index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">.rds&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
   <span class="c1"># Load required libraries</span>
    <span class="n">suppressMessages</span><span class="p">({</span>
        <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>
        <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
        <span class="n">library</span><span class="p">(</span><span class="n">metafor</span><span class="p">)</span>
    <span class="p">})</span>

    <span class="c1"># Convert p_cut to numeric</span>
    <span class="n">p_cut</span> <span class="o">&lt;-</span> <span class="err">$</span><span class="p">{</span><span class="n">p_cut</span><span class="p">}</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span>

    <span class="c1"># Initialize the output data table</span>
    <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
    <span class="nb">set</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">999</span><span class="p">)</span>

    <span class="c1"># Function to extract the gene name from the file path</span>
    <span class="n">get_gene_name</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">basename</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;norminal.cis_long_table.&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="o">.</span><span class="p">[,</span> <span class="mi">2</span><span class="p">]</span> <span class="o">%&gt;%</span>
            <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;_posterior_contrast.rds&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Loop through each input RDS file</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}))</span> <span class="p">{</span>
        <span class="k">try</span><span class="p">({</span>
            <span class="c1"># Extract gene name</span>
            <span class="n">gene</span> <span class="o">&lt;-</span> <span class="n">get_gene_name</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

            <span class="c1"># Load the contrast results and filter columns of interest</span>
            <span class="n">tmp_contrast_results</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;p_contrast_.*deviation&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
                <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>

            <span class="c1"># Initialize the feature output data table</span>
            <span class="n">feature</span><span class="o">.</span><span class="n">out</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">()</span>

            <span class="c1"># Calculate the ratio for each condition</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">condition_name</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span> <span class="o">%&gt;%</span>
                    <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;p_contrast_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
                    <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;_deviation&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>

                <span class="n">n_sig_snp</span> <span class="o">&lt;-</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">[,</span> <span class="n">i</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">p_cut</span><span class="p">,</span> <span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
                <span class="n">n_snp</span> <span class="o">&lt;-</span> <span class="nb">sum</span><span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">tmp_contrast_results</span><span class="p">[,</span> <span class="n">i</span><span class="p">]))</span>

                <span class="n">ratio</span> <span class="o">&lt;-</span> <span class="n">n_sig_snp</span> <span class="o">/</span> <span class="n">n_snp</span>
                <span class="n">feature</span><span class="o">.</span><span class="n">out</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span> <span class="n">condition</span> <span class="o">=</span> <span class="n">condition_name</span><span class="p">,</span> <span class="n">ratio</span> <span class="o">=</span> <span class="n">ratio</span><span class="p">))</span>
            <span class="p">}</span>

            <span class="c1"># Save individual gene results</span>
            <span class="n">saveRDS</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:d}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.rds&quot;</span><span class="p">,</span> <span class="s2">&quot;_n_sig_ratio.rds&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">res</span><span class="p">))))</span>

            <span class="c1"># Merge with the overall output</span>
            <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">out</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">feature</span><span class="o">.</span><span class="n">out</span>
            <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">merge</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="nb">all</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
            <span class="p">}</span>
        <span class="p">},</span> <span class="n">silent</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span> <span class="c1"># Error handling to proceed even if one iteration fails</span>
    <span class="p">}</span>

    <span class="c1"># Save the combined output</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="feature-pval-pair">
<h3>feature_pval_pair<a class="headerlink" href="#feature-pval-pair" title="Permalink to this headline">¶</a></h3>
<p>find the specific different cell type pair</p>
<p>perform meta analysis with pairwise contrasts to get a pvalue to find to understand what the specific differences are.</p>
<p>metaanalysis return a warning as “Warning message: “Ratio of largest to smallest sampling variance extremely large. May not be able to obtain stable results.””
Which is due to the big difference between max(pairwise_standard_errors^2) and min(pairwise_standard_errors^2) So I have delete the snps with small pairwise_standard_errors as Xuewei suggested</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute feature score from contrast results</span>
<span class="p">[</span><span class="n">feature_pval_pair_1</span><span class="p">]</span>
<span class="c1">#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]</span>
<span class="kn">parameter:</span> <span class="n">plink_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/mfs/cluster/bin/plink1.9.10/plink&quot;</span>
<span class="kn">parameter:</span> <span class="n">bfile_path</span> <span class="o">=</span> <span class="s2">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/WashU_genotype/genotype_qc/MAP_Brain-xQTL_Gwas_geno_0.1_maf_0.0005.filtered.&quot;</span>
<span class="kn">parameter:</span> <span class="n">extract</span> <span class="o">=</span> <span class="s2">&quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/snp.list&quot;</span>
<span class="kn">parameter:</span> <span class="n">window_size</span> <span class="o">=</span> <span class="s1">&#39;100&#39;</span>
<span class="kn">parameter:</span> <span class="n">step_size</span> <span class="o">=</span> <span class="s1">&#39;10&#39;</span>
<span class="kn">parameter:</span> <span class="n">r2_threshold</span> <span class="o">=</span> <span class="s1">&#39;0.2&#39;</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="s1">&#39;100&#39;</span>
<span class="kn">parameter:</span> <span class="n">LD_prune</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">parameter:</span> <span class="n">downsample_ratio</span> <span class="o">=</span> <span class="s1">&#39;1&#39;</span>
<span class="kn">parameter:</span> <span class="n">meta_method</span> <span class="o">=</span> <span class="s1">&#39;REML&#39;</span>
<span class="kn">parameter:</span> <span class="n">se_cutoff</span> <span class="o">=</span> <span class="s1">&#39;1E-03&#39;</span>
<span class="kn">parameter:</span> <span class="n">LDcache</span> <span class="o">=</span> <span class="s2">&quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/&quot;</span>
<span class="kn">parameter:</span> <span class="n">contrast_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">contrast_input</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="n">per_chunk</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/feature_pval_pair/cache/mash_posterior_contrast_featurescore</span><span class="si">{</span><span class="n">_index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">.rds&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="c1"># Library Loading</span>
    <span class="n">suppressMessages</span><span class="p">({</span>
      <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>
      <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
      <span class="n">library</span><span class="p">(</span><span class="n">metafor</span><span class="p">)</span>
    <span class="p">})</span>

    <span class="c1"># Initializations</span>
    <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
    <span class="nb">set</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">999</span><span class="p">)</span>

    <span class="c1"># Helper function: extract gene name from result</span>
    <span class="n">extract_gene</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">basename</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;norminal.cis_long_table.&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="o">.</span><span class="p">[,</span> <span class="mi">2</span><span class="p">]</span> <span class="o">%&gt;%</span>
        <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;_posterior_contrast.rds&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Helper function: perform meta analysis per cell</span>
    <span class="n">meta_analysis_per_cell</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">effect_sizes</span><span class="p">,</span> <span class="n">se_values</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">se_cutoff</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">conditions</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">effect_sizes</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="n">sub</span><span class="p">(</span><span class="s2">&quot;mean_contrast_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="n">unique</span><span class="p">()</span>
      <span class="n">cells</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;_vs_.*&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">conditions</span><span class="p">),</span> <span class="n">sub</span><span class="p">(</span><span class="s2">&quot;.*_vs_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">conditions</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="n">unique</span><span class="p">()</span>

      <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">()</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">cell</span> <span class="ow">in</span> <span class="n">cells</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">cell</span><span class="o">.</span><span class="n">b</span> <span class="o">&lt;-</span> <span class="n">effect_sizes</span><span class="p">[,</span> <span class="n">grep</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">colnames</span><span class="p">(</span><span class="n">effect_sizes</span><span class="p">)),</span> <span class="n">drop</span> <span class="o">=</span> <span class="n">F</span><span class="p">]</span>
        <span class="n">cell</span><span class="o">.</span><span class="n">se</span> <span class="o">&lt;-</span> <span class="n">se_values</span><span class="p">[,</span> <span class="n">grep</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">colnames</span><span class="p">(</span><span class="n">se_values</span><span class="p">)),</span> <span class="n">drop</span> <span class="o">=</span> <span class="n">F</span><span class="p">]</span>
        <span class="n">feature_pvals</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">()</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">b</span><span class="p">))</span> <span class="p">{</span>
          <span class="n">effect_sizes_condition</span> <span class="o">&lt;-</span> <span class="n">cell</span><span class="o">.</span><span class="n">b</span><span class="p">[,</span> <span class="n">i</span><span class="p">]</span>
          <span class="n">se_values_condition</span> <span class="o">&lt;-</span> <span class="n">cell</span><span class="o">.</span><span class="n">b</span><span class="p">[,</span> <span class="n">i</span><span class="p">]</span>
          <span class="c1"># Filter out based on se cutoff</span>
          <span class="n">effect_sizes_condition</span> <span class="o">&lt;-</span> <span class="n">effect_sizes_condition</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">se_values_condition</span> <span class="o">&gt;</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">se_cutoff</span><span class="p">))]</span>
          <span class="n">se_values_condition</span> <span class="o">&lt;-</span> <span class="n">se_values_condition</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">se_values_condition</span> <span class="o">&gt;</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">se_cutoff</span><span class="p">))]</span>
          <span class="c1"># Meta-analysis</span>
          <span class="n">absolute_effect_sizes</span> <span class="o">&lt;-</span> <span class="nb">abs</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">effect_sizes_condition</span><span class="p">))</span>
          <span class="n">pairwise_standard_errors</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">se_values_condition</span><span class="p">)</span>
          <span class="n">meta_result</span> <span class="o">&lt;-</span> <span class="n">rma</span><span class="p">(</span><span class="n">yi</span> <span class="o">=</span> <span class="n">absolute_effect_sizes</span><span class="p">,</span> <span class="n">sei</span> <span class="o">=</span> <span class="n">pairwise_standard_errors</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;REML&quot;</span><span class="p">)</span>
          <span class="n">feature_pval_condition</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">pavlue</span> <span class="o">=</span> <span class="n">meta_result</span><span class="err">$</span><span class="n">pval</span><span class="p">)</span>
          <span class="n">feature_pvals</span> <span class="o">&lt;-</span> <span class="n">rbindlist</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">feature_pvals</span><span class="p">,</span> <span class="n">feature_pval_condition</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">feature_pvals</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">gene</span>
        <span class="n">condition_name</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">cell</span><span class="o">.</span><span class="n">b</span><span class="p">)</span> <span class="o">%&gt;%</span>
          <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;mean_contrast_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">.</span><span class="p">)</span>
        <span class="n">feature_pvals</span> <span class="o">&lt;-</span> <span class="n">feature_pvals</span><span class="p">[,</span> <span class="n">condition</span> <span class="o">:=</span> <span class="n">condition_name</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">rbindlist</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">feature_pvals</span><span class="p">))</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Main Loop</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">res</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}))</span> <span class="p">{</span>
      <span class="n">gene</span> <span class="o">&lt;-</span> <span class="n">extract_gene</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
      <span class="n">tmp_contrast_results</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
      <span class="n">ld_output</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{LDcache}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;_output.prune.in&quot;</span><span class="p">))</span>
      <span class="n">tmp_contrast_results</span> <span class="o">&lt;-</span> <span class="n">tmp_contrast_results</span><span class="p">[</span><span class="n">ld_output</span><span class="err">$</span><span class="n">V1</span><span class="p">,</span> <span class="p">]</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">keep</span><span class="o">.</span><span class="n">rownames</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>

      <span class="n">effect_sizes</span> <span class="o">&lt;-</span> <span class="n">tmp_contrast_results</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;mean_contrast.*_vs_*&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>
      <span class="n">se_values</span> <span class="o">&lt;-</span> <span class="n">tmp_contrast_results</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="n">matches</span><span class="p">(</span><span class="s2">&quot;se_contrast.*_vs_*&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>

      <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">meta_analysis_per_cell</span><span class="p">(</span><span class="n">effect_sizes</span><span class="p">,</span> <span class="n">se_values</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{se_cutoff}</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="n">saveRDS</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">str_c</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:d}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;.rds&quot;</span><span class="p">,</span> <span class="s2">&quot;_featurescore_pw.rds&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">(</span><span class="n">res</span><span class="p">))))</span>

      <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">out</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">df</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">merge</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="nb">all</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">allow</span><span class="o">.</span><span class="n">cartesian</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
      <span class="p">}</span>
    <span class="p">}</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># merge the feature score from contrast results</span>
<span class="p">[</span><span class="n">feature_pval_pair_2</span><span class="p">,</span> <span class="n">feature_score_meta_2</span><span class="p">,</span> <span class="n">feature_score_nsig_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/feature_pval_pair/posterior_feature_score_sum.csv&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;10G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> 
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
    <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">NULL</span>

    <span class="nb">all</span><span class="o">.</span><span class="n">list</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="nb">all</span><span class="o">.</span><span class="n">list</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">feature_scores</span> <span class="o">&lt;-</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="c1"># Print the feature scores for each condition</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">feature_scores</span><span class="p">))</span> <span class="p">{</span>
            <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">out</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">feature_scores</span>
            <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">merge</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="n">feature_scores</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="nb">all</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
            <span class="p">}</span>
        <span class="p">}</span>
    <span class="p">}</span>
      <span class="n">out</span><span class="o">&lt;-</span><span class="n">t</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
      <span class="c1">#out&lt;-out[-1,]</span>
    <span class="n">write</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="n">out</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/multivariate_genome/MASH"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="mash_fit.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">MASH analysis pipeline with data-driven prior matrices</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../../rare_xqtl/watershed.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">watershed analysis for rare xQTL</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>