
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>TensorQTL QTL association testing &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="APEX QTL association testing" href="../APEX/APEX.html" />
    <link rel="prev" title="QTL association analysis" href="../qtl_association_testing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../qtl_association_testing.html">
   QTL association analysis
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with individual level data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     Data overview
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/association_scan/TensorQTL/TensorQTL.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/association_scan/TensorQTL/TensorQTL.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/association_scan/TensorQTL/TensorQTL.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   TensorQTL QTL association testing
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     Input
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#output">
     Output
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#minimal-working-example">
     Minimal working example
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#global-parameter-settings">
     Global parameter settings
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cis-xqtl-association-testing">
     cis-xQTL association testing
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#transqtl-association-testing">
     TransQTL association testing
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#association-results-processing">
     Association results processing
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="tensorqtl-qtl-association-testing">
<h1>TensorQTL QTL association testing<a class="headerlink" href="#tensorqtl-qtl-association-testing" title="Permalink to this headline">¶</a></h1>
<p>This notebook implements a workflow using <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1836-7">tensorQTL</a> to perform QTL association testing.</p>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>List of molecular phenotype files: a list of <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> files containing the table for the molecular phenotype. It should have a companion index file in <code class="docutils literal notranslate"><span class="pre">tbi</span></code> format.</p>
<p>The header of the bed.gz is per the <a class="reference external" href="https://github.com/broadinstitute/tensorqtl">TensorQTL</a> convention:</p>
<blockquote>
<div><p>Phenotypes must be provided in BED format, with a single header line starting with # and the first four columns corresponding to: chr, start, end, phenotype_id, with the remaining columns corresponding to samples (the identifiers must match those in the genotype input). The BED file should specify the center of the cis-window (usually the TSS), with start == end-1.</p>
</div></blockquote>
</li>
<li><p>List of genotypes in PLINK binary format (<code class="docutils literal notranslate"><span class="pre">bed</span></code>/<code class="docutils literal notranslate"><span class="pre">bim</span></code>/<code class="docutils literal notranslate"><span class="pre">fam</span></code>) for each chromosome, previously processed through our genotype QC pipelines.</p></li>
<li><p>Covariate file, a file with #id + samples name as colnames and each row a covariate: fixed and known covariates as well as hidden covariates recovered from factor analysis.</p></li>
<li><p>Optionally, a list of traits (genes, regions of molecular features etc) to analyze.</p></li>
</ul>
<p>For cis-analysis:</p>
<ul class="simple">
<li><p>Optionally, a list of genomic regions associate with each molecular features to analyze. The default cis-analysis will use a window around TSS. This can be customized to take given start and end genomic coordinates.</p></li>
</ul>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>For each chromosome, several of summary statistics files are generated, including both nominal test statistics for each test, as well as region (gene) level association evidence.</p>
<p>The columns of nominal association result are as follows:</p>
<ul class="simple">
<li><p>phenotype_id: Molecular trait identifier.(gene)</p></li>
<li><p>variant_id: ID of the variant (rsid or chr:position:ref:alt)</p></li>
<li><p>tss_distance: Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>af: The allele frequency of this SNPs</p></li>
<li><p>ma_samples: Number of samples carrying the minor allele</p></li>
<li><p>ma_count: Total number of minor alleles across individuals</p></li>
<li><p>pval: Nominal P-value from linear regression</p></li>
<li><p>beta: Slope of the linear regression</p></li>
<li><p>se: Standard error of beta</p></li>
<li><p>chr : Variant chromosome.</p></li>
<li><p>pos : Variant chromosomal position (basepairs).</p></li>
<li><p>ref : Variant reference allele (A, C, T, or G).</p></li>
<li><p>alt : Variant alternate allele.</p></li>
</ul>
<p>The column specification of region (gene) level association evidence are as follows:</p>
<ul class="simple">
<li><p>phenotype_id - Molecular trait identifier. (gene)</p></li>
<li><p>num_var - Total number of variants tested in cis</p></li>
<li><p>beta_shape1 - First parameter value of the fitted beta distribution</p></li>
<li><p>beta_shape2 - Second parameter value of the fitted beta distribution</p></li>
<li><p>true_df - Effective degrees of freedom the beta distribution approximation</p></li>
<li><p>pval_true_df - Empirical P-value for the beta distribution approximation</p></li>
<li><p>variant_id - ID of the top variant (rsid or chr:position:ref:alt)</p></li>
<li><p>tss_distance - Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>ma_samples - Number of samples carrying the minor allele</p></li>
<li><p>ma_count - Total number of minor alleles across individuals</p></li>
<li><p>maf - Minor allele frequency in MiGA cohort</p></li>
<li><p>ref_factor - Flag indicating if the alternative allele is the minor allele in the cohort (1 if AF &lt;= 0.5, -1 if not)</p></li>
<li><p>pval_nominal - Nominal P-value from linear regression</p></li>
<li><p>slope - Slope of the linear regression</p></li>
<li><p>slope_se - Standard error of the slope</p></li>
<li><p>pval_perm - First permutation P-value directly obtained from the permutations with the direct method</p></li>
<li><p>pval_beta - Second permutation P-value obtained via beta approximation. This is the one to use for downstream analysis</p></li>
</ul>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="command-interface">
<h1>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run TensorQTL.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  cis
  trans

Global Workflow Options:
  --covariate-file VAL (as path, required)
                        Covariate file
  --genotype-file VAL (as path, required)
                        For cis, genotype file in PLINK binary format
                        (bed/bam/fam) format, per chrom, for trans, 1 whole
                        genome genotype file in plink binary format
  --region-list . (as path)
                        An optional subset of regions of molecular features to
                        analyze
  --customized-cis-windows . (as path)
                        An optional list documenting the custom cis window for
                        each region to analyze, with four column, chr, start,
                        end, region ID (eg gene ID). If this list is not
                        provided, the default `window` parameter (see below)
                        will be used.
  --cwd output (as path)
                        Path to the work directory of the analysis.
  --phenotype-file VAL (as path, required)
                        Phenotype file, if cis a list of phenotype per chrom, if
                        trans, 1 whole genome phenotype file.
  --name  f&quot;{phenotype_file:bn}_{covariate_file:bn}&quot;

                        Prefix for the analysis output
  --MAC 0 (as int)
                        Minor allele count cutoff
  --region-name &#39;gene_id&#39;
                        The name of phenotype corresponding to gene_id or
                        gene_name in the region
  --phenotype-group . (as path)
                        The phenotype group file to group molecule_trait into
                        molecule_trait_object This is applicable to when there
                        are multiple molecular events in the same region, such
                        as sQTL analysis.
  --window 1000000 (as int)
                        Specify the cis window for the up and downstream radius
                        to analyze around the region of interest, in units of bp
                        This parameter will be set to zero if
                        `customized_cis_windows` is provided.
  --numThreads 8 (as int)
                        Number of threads
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 12h
  --mem 16G
  --container &#39;&#39;
                        Container option for software to run the analysis:
                        docker or singularity
  --entrypoint {(&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + container.split(&#39;/&#39;)[-1][:-4]) if container.endswith(&#39;.sif&#39;) else f&#39;&#39;}

  --maf-threshold  MAC/(2.0*N)

                        Minor allele frequency cutoff. It will overwrite minor
                        allele cutoff.

Sections
  cis_1:
    Workflow Options:
      --skip-nominal-if-exist  false

                        skip nominal association results if the files exists
                        already This is false by default which means to
                        recompute everything This is only relevant when the
                        `parquet` files for nominal results exist but not the
                        other files and you want to avoid computing the nominal
                        results again
  trans_1:
    Workflow Options:
      --batch-size 50000 (as int)
      --pval-threshold 1.0 (as float)
  cis_2:
  trans_2:
</pre></div>
</div>
</div>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>An MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1yjTwoO0DYGi-J9ouMsh9fHKfDmsXJ_4I?usp=sharing">google drive</a>.
The singularity image (sif) for running this MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1mLOS3AVQM8yTaWtCbO8Q3xla98Nr5bZQ">google drive</a></p>
<p>FIXME: need to update these links.</p>
<p>FIXME: Also need to update the example commands below using our new example dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">cis</span> \
    <span class="o">--</span><span class="n">genotype</span><span class="o">-</span><span class="n">file</span> <span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">recipe</span> \
    <span class="o">--</span><span class="n">covariate</span><span class="o">-</span><span class="n">file</span> <span class="n">ALL</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">BiCV</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">MAC</span> <span class="mi">5</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">trans</span> \
    <span class="o">--</span><span class="n">genotype</span><span class="o">-</span><span class="n">file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">mol_phe</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">covariate</span><span class="o">-</span><span class="n">file</span> <span class="n">ALL</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">BiCV</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">MAC</span> <span class="mi">5</span> <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="kp">name</span>  <span class="n">gene_name</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">trans</span> \
    <span class="o">--</span><span class="n">genotype</span><span class="o">-</span><span class="n">file</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">genotype_qced</span><span class="o">/</span><span class="n">GRCh38_liftedover_sorted_all</span><span class="o">.</span><span class="n">add_chr</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">renamed</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">renamed</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">file</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">eight_tissue_analysis</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">ALL</span><span class="o">/</span><span class="n">phenotype_data</span><span class="o">/</span><span class="n">ALL</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">covariate</span><span class="o">-</span><span class="n">file</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">eight_tissue_analysis</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">ALL</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ALL</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">ALL</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span><span class="n">trans_tensorQTL</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">eight_tissue_analysis</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">AD_genes</span><span class="o">.</span><span class="n">region_list</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">MAC</span> <span class="mi">5</span> <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="kp">name</span> <span class="n">ID</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="global-parameter-settings">
<h2>Global parameter settings<a class="headerlink" href="#global-parameter-settings" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Covariate file</span>
<span class="kn">parameter:</span> <span class="n">covariate_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># For cis, genotype file in PLINK binary format (bed/bam/fam) format, per chrom, for trans, 1 whole genome genotype file in plink binary format</span>
<span class="kn">parameter:</span> <span class="n">genotype_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># An optional subset of regions of molecular features to analyze</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># An optional list documenting the custom cis window for each region to analyze, with four column, chr, start, end, region ID (eg gene ID).</span>
<span class="c1"># If this list is not provided, the default `window` parameter (see below) will be used.</span>
<span class="kn">parameter:</span> <span class="n">customized_cis_windows</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Path to the work directory of the analysis.</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;output&#39;</span><span class="p">)</span>
<span class="c1"># Phenotype file, if cis a list of phenotype per chrom, if trans, 1 whole genome phenotype file.</span>
<span class="kn">parameter:</span> <span class="n">phenotype_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Prefix for the analysis output</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">phenotype_file</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">covariate_file</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="c1"># Minor allele count cutoff</span>
<span class="kn">parameter:</span> <span class="n">MAC</span> <span class="o">=</span> <span class="mi">0</span>
<span class="c1"># The name of phenotype corresponding to gene_id or gene_name in the region</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span>
<span class="c1"># The phenotype group file to group molecule_trait into molecule_trait_object</span>
<span class="c1"># This is applicable to when there are multiple molecular events in the same region, such as sQTL analysis.</span>
<span class="kn">parameter:</span> <span class="n">phenotype_group</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span> 

<span class="c1"># Specify the cis window for the up and downstream radius to analyze around the region of interest, in units of bp</span>
<span class="c1"># This parameter will be set to zero if `customized_cis_windows` is provided.</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">1000000</span>

<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;16G&#39;</span>
<span class="c1"># Container option for software to run the analysis: docker or singularity</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span><span class="p">{(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">])</span> <span class="k">if</span> <span class="n">container</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.sif&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;&#39;</span><span class="p">}</span>

<span class="c1"># Use the header of covariate file to decide the sample size</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covariate_file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">nrows</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>

<span class="c1"># Minor allele frequency cutoff. It will overwrite minor allele cutoff.</span>
<span class="kn">parameter:</span> <span class="n">maf_threshold</span> <span class="o">=</span> <span class="n">MAC</span><span class="o">/</span><span class="p">(</span><span class="mf">2.0</span><span class="o">*</span><span class="n">N</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="cis-xqtl-association-testing">
<h2>cis-xQTL association testing<a class="headerlink" href="#cis-xqtl-association-testing" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cis_1</span><span class="p">]</span>
<span class="c1"># parse input file lists</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">molecular_pheno_files</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">phenotype_file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">genotype_files</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genotype_file</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">genotype_files</span><span class="p">[</span><span class="s2">&quot;#id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genotype_files</span><span class="p">[</span><span class="s2">&quot;#id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">molecular_pheno_files</span><span class="p">[</span><span class="s2">&quot;#id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">molecular_pheno_files</span><span class="p">[</span><span class="s2">&quot;#id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

<span class="n">input_files</span> <span class="o">=</span> <span class="n">molecular_pheno_files</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">genotype_files</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s2">&quot;#id&quot;</span><span class="p">)</span>
<span class="n">input_files</span> <span class="o">=</span> <span class="n">input_files</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">input_chroms</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">input_files</span><span class="p">]</span>
<span class="n">input_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">input_files</span><span class="p">]</span>

<span class="c1"># skip nominal association results if the files exists already</span>
<span class="c1"># This is false by default which means to recompute everything</span>
<span class="c1"># This is only relevant when the `parquet` files for nominal results exist but not the other files</span>
<span class="c1"># and you want to avoid computing the nominal results again</span>
<span class="kn">parameter:</span> <span class="n">skip_nominal_if_exist</span> <span class="o">=</span> <span class="kc">False</span>
<span class="kn">input:</span> <span class="n">input_files</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">input_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;input_chroms&quot;</span> 
<span class="kn">output:</span> <span class="n">parquet</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.cis_qtl_pairs.</span><span class="si">{</span><span class="n">_input_chroms</span><span class="si">}</span><span class="s1">.parquet&#39;</span><span class="p">,</span> <span class="c1"># This convention is necessary to match the pattern of map_norminal output</span>
        <span class="n">regional</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_input_chroms</span><span class="si">}</span><span class="s1">.cis_qtl_regional.gz&#39;</span><span class="p">,</span>
        <span class="n">nominal</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_input_chroms</span><span class="si">}</span><span class="s1">.cis_qtl_nominal.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">tensorqtl</span>
    <span class="kn">from</span> <span class="nn">tensorqtl</span> <span class="kn">import</span> <span class="n">genotypeio</span><span class="p">,</span> <span class="n">cis</span><span class="p">,</span> <span class="n">trans</span>
    <span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chi2</span>
        
    <span class="c1">## Define paths</span>
    <span class="n">plink_prefix_path</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">nar</span><span class="p">]</span>
    <span class="n">expression_bed</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">ar</span><span class="p">]</span>
    <span class="n">covariates_file</span> <span class="o">=</span> <span class="s2">&quot;$[covariate_file:a]&quot;</span>
    <span class="n">window</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">window</span><span class="p">]</span>
    <span class="c1">## Load Data</span>
    <span class="n">phenotype_df</span><span class="p">,</span> <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">tensorqtl</span><span class="o">.</span><span class="n">read_phenotype_bed</span><span class="p">(</span><span class="n">expression_bed</span><span class="p">)</span>
    <span class="n">phenotype_id</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="kp">name</span>
    <span class="c1">## Analyze only the regions listed</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">[</span><span class="n">region_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">()]:</span>
        <span class="n">region</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;$[region_list:a]&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">keep_gene</span> <span class="o">=</span> <span class="n">region</span><span class="p">[</span><span class="n">phenotype_id</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
        <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_gene</span><span class="p">)]</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="p">[</span><span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_gene</span><span class="p">)]</span>

    <span class="c1">## use custom cis windows</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">[</span><span class="n">customized_cis_windows</span><span class="o">.</span><span class="n">is_file</span><span class="p">()]:</span>
        <span class="n">cis_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;$[customized_cis_windows:a]&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span> <span class="c1">#move the phenotype id index to a new column of the dataframe</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cis_list</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="n">phenotype_id</span><span class="p">],</span><span class="n">right_on</span> <span class="o">=</span> <span class="p">[</span><span class="n">cis_list</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">cis_list</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">3</span><span class="p">]])</span><span class="c1">#in some cases (gene expression for eQTLs) the phenotype_id may be in the cis_list file</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">phenotype_id</span><span class="p">)[[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">]]</span> <span class="c1"># The final phenotype_pos_df will have three columns(chr, start, end) and index is the phenotype ID</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;cannot uniquely match all the phentoype data in the input to customized cis windows list&quot;</span><span class="p">)</span>
        <span class="n">window</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># In the updated tensorQTL, by default if there is a customized cis window, the actual cis window will be start - window &amp; end + window, so it is necessary to change the window parameter to 0</span>
    
    <span class="n">covariates_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">pr</span> <span class="o">=</span> <span class="n">genotypeio</span><span class="o">.</span><span class="n">PlinkReader</span><span class="p">(</span><span class="n">plink_prefix_path</span><span class="p">)</span>
    <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">load_genotypes</span><span class="p">()</span>
    <span class="n">variant_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;snp&#39;</span><span class="p">)[[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">]]</span>
    <span class="c1">## Retaining only common samples</span>
    <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span>
    <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">genotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)]</span>
    <span class="n">covariates_df</span> <span class="o">=</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">transpose</span><span class="p">()[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    
    <span class="c1">## To simplify things, there should really not be &quot;chr&quot; prefix</span>
    <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">chr</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">chr</span><span class="p">]</span>
    <span class="n">variant_df</span><span class="o">.</span><span class="n">chrom</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">variant_df</span><span class="o">.</span><span class="n">chrom</span><span class="p">]</span>
    <span class="c1">## Read phenotype group if availble</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">[</span><span class="n">phenotype_group</span><span class="o">.</span><span class="n">is_file</span><span class="p">()]:</span>
        <span class="n">group_s</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">phenotype_group</span><span class="p">:</span><span class="n">r</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">group_s</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1">## cis-QTL mapping: nominal associations for all variant-phenotype pairs</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">skip_nominal_if_exist</span><span class="p">]</span> <span class="ow">and</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">is_file</span><span class="p">()]):</span>
        <span class="n">cis</span><span class="o">.</span><span class="n">map_nominal</span><span class="p">(</span><span class="n">genotype_df</span><span class="p">,</span> <span class="n">variant_df</span><span class="p">,</span>
                <span class="n">phenotype_df</span><span class="p">,</span>
                <span class="n">phenotype_pos_df</span><span class="p">,</span>
                <span class="s2">&quot;$[_output[0]:nnn]&quot;</span><span class="p">,</span> 
                <span class="n">covariates_df</span><span class="o">=</span><span class="n">covariates_df</span><span class="p">,</span> 
                <span class="n">window</span><span class="o">=</span><span class="n">window</span><span class="p">,</span> 
                <span class="n">maf_threshold</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">maf_threshold</span><span class="p">],</span>
                <span class="n">group_s</span><span class="o">=</span><span class="n">group_s</span><span class="p">)</span>

    <span class="c1">## Load the parquet and save it as txt</span>
    <span class="n">pairs_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">)</span>
    <span class="c1">## Adds the group columns to pairs_df, if there is group_s use group_s, else use phenotype_id</span>
    <span class="k">if</span> <span class="n">group_s</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pairs_df</span> <span class="o">=</span> <span class="n">pairs_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span> <span class="p">{</span><span class="s2">&quot;molecular_trait_object_id&quot;</span><span class="p">:</span> <span class="n">group_s</span><span class="p">}),</span><span class="n">left_on</span> <span class="o">=</span> <span class="s2">&quot;phenotype_id&quot;</span><span class="p">,</span> <span class="n">right_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pairs_df</span><span class="p">[</span><span class="s2">&quot;molecular_trait_object_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pairs_df</span><span class="o">.</span><span class="n">phenotype_id</span>

    <span class="c1"># genomic inflation factor lambda</span>
    <span class="n">lambda_col</span> <span class="o">=</span> <span class="n">pairs_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;molecular_trait_object_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>  <span class="n">chi2</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="mf">1.</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">pval_nominal</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="n">chi2</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;molecular_trait_id&quot;</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;pvalue&quot;</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;beta&quot;</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;se&quot;</span>
    <span class="n">pairs_df</span> <span class="o">=</span> <span class="n">pairs_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">maf</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;af&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">af</span><span class="p">:</span><span class="n">af</span> <span class="k">if</span> <span class="n">af</span> <span class="o">&lt;</span> <span class="mf">0.5</span> <span class="k">else</span> <span class="mi">1</span><span class="o">-</span><span class="n">af</span><span class="p">)</span> <span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;af&quot;</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span>  <span class="mi">1</span><span class="p">)</span>
    <span class="n">pairs_df</span><span class="p">[</span><span class="s2">&quot;n&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">pairs_df</span> <span class="o">=</span> <span class="n">pairs_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">alt</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">2</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[2]]&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
    <span class="n">cis_df</span> <span class="o">=</span> <span class="n">cis</span><span class="o">.</span><span class="n">map_cis</span><span class="p">(</span><span class="n">genotype_df</span><span class="p">,</span> 
                        <span class="n">variant_df</span><span class="p">,</span> 
                        <span class="n">phenotype_df</span><span class="p">,</span>
                        <span class="n">phenotype_pos_df</span><span class="p">,</span>
                        <span class="n">covariates_df</span><span class="o">=</span><span class="n">covariates_df</span><span class="p">,</span> 
                        <span class="n">seed</span><span class="o">=</span><span class="mi">999</span><span class="p">,</span> 
                        <span class="n">window</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">window</span><span class="p">],</span> 
                        <span class="n">maf_threshold</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">maf_threshold</span><span class="p">],</span>
                        <span class="n">group_s</span><span class="o">=</span><span class="n">group_s</span><span class="p">)</span>
    
    <span class="n">cis_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;molecular_trait_id&quot;</span>
    <span class="c1">## Add groups columns for eQTL analysis</span>
    <span class="k">if</span> <span class="s2">&quot;group_id&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cis_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">cis_df</span><span class="p">[</span><span class="s2">&quot;group_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cis_df</span><span class="o">.</span><span class="n">index</span>
        <span class="n">cis_df</span><span class="p">[</span><span class="s2">&quot;group_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">cis_df</span> <span class="o">=</span> <span class="n">cis_df</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s2">&quot;group_id&quot;</span><span class="p">:</span><span class="s2">&quot;molecular_trait_object_id&quot;</span><span class="p">,</span><span class="s2">&quot;group_size&quot;</span><span class="p">:</span><span class="s2">&quot;n_traits&quot;</span><span class="p">,</span><span class="s2">&quot;num_var&quot;</span> <span class="p">:</span> <span class="s2">&quot;n_variants&quot;</span><span class="p">,</span><span class="s2">&quot;variant_id&quot;</span><span class="p">:</span><span class="s2">&quot;variant&quot;</span><span class="p">,</span><span class="s2">&quot;pval_perm&quot;</span><span class="p">:</span><span class="s2">&quot;p_perm&quot;</span><span class="p">,</span> <span class="s2">&quot;pval_beta&quot;</span><span class="p">:</span><span class="s2">&quot;p_beta&quot;</span> <span class="p">},</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">cis_df</span> <span class="o">=</span> <span class="n">cis_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">genomic_inflation</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span> <span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s2">&quot;molecular_trait_object_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">molecular_trait_object_id</span><span class="p">:</span><span class="n">lambda_col</span><span class="p">[</span><span class="n">molecular_trait_object_id</span><span class="p">]))</span>
    <span class="n">cis_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>

<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
        <span class="n">stdout</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info: $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;This is the file containing the immediate output of TensorQTL&#39;s map_nominal function &quot;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">done</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info: $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_headerow:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">V</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">zcat</span> <span class="err">$</span><span class="n">i</span>  <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info: $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_headerow:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">V</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">zcat</span> <span class="err">$</span><span class="n">i</span>  <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="transqtl-association-testing">
<h2>TransQTL association testing<a class="headerlink" href="#transqtl-association-testing" title="Permalink to this headline">¶</a></h2>
<p>With a genotype file of size 3.2G, it takes at least 55 GB of memory, also 8 threads is not enough</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">trans_1</span><span class="p">]</span>

<span class="c1"># fail_if(not region_list.is_file(), msg = &quot;For trans analysis, ``region_list`` must be provided.&quot;)</span>

<span class="kn">input:</span> <span class="n">phenotype_file</span><span class="p">,</span> <span class="n">genotype_file</span>
<span class="kn">output:</span> <span class="n">nominal</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.trans_qtl_nominal.gz&#39;</span>
<span class="kn">parameter:</span> <span class="n">batch_size</span> <span class="o">=</span> <span class="mi">50000</span>
<span class="kn">parameter:</span> <span class="n">pval_threshold</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">tensorqtl</span>
    <span class="kn">from</span> <span class="nn">tensorqtl</span> <span class="kn">import</span> <span class="n">genotypeio</span><span class="p">,</span> <span class="n">cis</span><span class="p">,</span> <span class="n">trans</span>
    <span class="c1">## Define paths</span>
    <span class="n">plink_prefix_path</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">nar</span><span class="p">]</span>
    <span class="n">expression_bed</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">ar</span><span class="p">]</span>
    <span class="n">covariates_file</span> <span class="o">=</span> <span class="s2">&quot;$[covariate_file:a]&quot;</span>
    <span class="n">window</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">window</span><span class="p">]</span>
    <span class="c1">## Loading Data</span>
    <span class="n">phenotype_df</span><span class="p">,</span> <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">tensorqtl</span><span class="o">.</span><span class="n">read_phenotype_bed</span><span class="p">(</span><span class="n">expression_bed</span><span class="p">)</span>
    <span class="n">phenotype_id</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="kp">name</span>
    <span class="c1">## Analyze only the regions listed</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">[</span><span class="n">region_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">()]:</span>
        <span class="n">region</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;$[region_list:a]&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">keep_gene</span> <span class="o">=</span> <span class="n">region</span><span class="p">[</span><span class="n">phenotype_id</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
        <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_gene</span><span class="p">)]</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="p">[</span><span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_gene</span><span class="p">)]</span>
    <span class="c1">## use custom cis windows</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">[</span><span class="n">customized_cis_windows</span><span class="o">.</span><span class="n">is_file</span><span class="p">()]:</span>
        <span class="n">cis_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;$[customized_cis_windows:a]&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span> <span class="c1">#move the phenotype id index to a new column of the dataframe</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cis_list</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="n">phenotype_id</span><span class="p">],</span><span class="n">right_on</span> <span class="o">=</span> <span class="p">[</span><span class="n">cis_list</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">cis_list</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">3</span><span class="p">]])</span><span class="c1">#in some cases (gene expression for eQTLs) the phenotype_id may be in the cis_list file</span>
        <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">phenotype_id</span><span class="p">)[[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">]]</span> <span class="c1"># The final phenotype_pos_df will have three columns(chr, start, end) and index is the phenotype ID</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;cannot uniquely match all the phentoype data in the input to customized cis windows list&quot;</span><span class="p">)</span>
        <span class="n">window</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># In the updated tensorQTL, by default if there is a customized cis window, the actual cis window will be start - window &amp; end + window, so it is necessary to change the window parameter to 0</span>
    
    <span class="n">covariates_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">pr</span> <span class="o">=</span> <span class="n">genotypeio</span><span class="o">.</span><span class="n">PlinkReader</span><span class="p">(</span><span class="n">plink_prefix_path</span><span class="p">)</span>
    <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">load_genotypes</span><span class="p">()</span>
    <span class="n">variant_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;snp&#39;</span><span class="p">)[[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">]]</span>
    <span class="c1">## Retaining only common samples</span>
    <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span>
    <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">genotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)]</span>
    <span class="n">covariates_df</span> <span class="o">=</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">transpose</span><span class="p">()[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="c1">## Trans analysis</span>
    <span class="n">trans_df</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">map_trans</span><span class="p">(</span><span class="n">genotype_df</span><span class="p">,</span> 
                            <span class="n">phenotype_df</span><span class="p">,</span>
                            <span class="n">covariates_df</span><span class="p">,</span> 
                            <span class="n">batch_size</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">batch_size</span><span class="p">],</span>
                            <span class="n">return_sparse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
                            <span class="n">return_r2</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> 
                            <span class="n">pval_threshold</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">pval_threshold</span><span class="p">],</span> 
                            <span class="n">maf_threshold</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">maf_threshold</span><span class="p">])</span>
    <span class="c1">## Filter out cis signal, again if customized cis windows are used, the windows is [start-win,end + win] where win = 0, else it is [start - win, start + win]</span>
    <span class="n">trans_df</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">filter_cis</span><span class="p">(</span><span class="n">trans_df</span><span class="p">,</span> <span class="n">phenotype_pos_df</span><span class="p">,</span> <span class="n">variant_df</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="n">window</span><span class="p">)</span>   
    <span class="c1">## Permutation</span>
    <span class="n">perm_df</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">map_permutations</span><span class="p">(</span><span class="n">genotype_df</span><span class="p">,</span> <span class="n">covariates_df</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">batch_size</span><span class="p">],</span>
                             <span class="n">maf_threshold</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">maf_threshold</span><span class="p">])</span>
    <span class="n">perm_output</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">apply_permutations</span><span class="p">(</span><span class="n">perm_df</span><span class="p">,</span><span class="n">trans_df</span><span class="p">)</span>
    <span class="c1">## Output</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;molecular_trait_id&quot;</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;pvalue&quot;</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;beta&quot;</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;se&quot;</span>
    <span class="n">trans_df</span> <span class="o">=</span> <span class="n">trans_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">maf</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;af&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">af</span><span class="p">:</span><span class="n">af</span> <span class="k">if</span> <span class="n">af</span> <span class="o">&lt;</span> <span class="mf">0.5</span> <span class="k">else</span> <span class="mi">1</span><span class="o">-</span><span class="n">af</span><span class="p">)</span> <span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;af&quot;</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span>  <span class="mi">1</span><span class="p">)</span>
    <span class="n">trans_df</span> <span class="o">=</span> <span class="n">trans_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">alt</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output]&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>

<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
        <span class="n">stdout</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info: $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_headerow:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">V</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">cat</span> <span class="err">$</span><span class="n">i</span>  <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="association-results-processing">
<h2>Association results processing<a class="headerlink" href="#association-results-processing" title="Permalink to this headline">¶</a></h2>
<p>For both cis and trans: Generate the recipe for yml processing
For cis: Also process the consolidates emprical data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">cis_2</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;nominal&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.cis_qtl.meta_info&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;nominal&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.cis_qtl.column_info&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;regional&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.fdr.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;regional&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nnn</span><span class="si">}</span><span class="s1">.summary.txt&#39;</span>   
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

    <span class="n">data_temp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s2">&quot;sumstat_dir&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;regional&quot;</span><span class="p">]:</span><span class="n">r</span><span class="p">,]],</span>
    <span class="s2">&quot;column_info&quot;</span> <span class="p">:</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">]</span>
    <span class="p">})</span>
    <span class="n">column_info_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="s2">&quot;molecular_trait_id,chromosome,position,ref,alt&quot;</span><span class="p">,</span>
      <span class="s2">&quot;chromosome&quot;</span><span class="p">:</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span>
      <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span>
      <span class="s2">&quot;variant&quot;</span><span class="p">:</span> <span class="s2">&quot;variant_id&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ref&quot;</span><span class="p">:</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span>
      <span class="s2">&quot;alt&quot;</span><span class="p">:</span> <span class="s2">&quot;alt&quot;</span><span class="p">,</span>
      <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="s2">&quot;beta&quot;</span><span class="p">,</span>
      <span class="s2">&quot;se&quot;</span><span class="p">:</span> <span class="s2">&quot;se&quot;</span><span class="p">,</span>
      <span class="s2">&quot;pvalue&quot;</span><span class="p">:</span> <span class="s2">&quot;pvalue&quot;</span><span class="p">,</span>
      <span class="s2">&quot;TSS_D&quot;</span><span class="p">:</span> <span class="s2">&quot;tss_distance&quot;</span><span class="p">,</span>
      <span class="s2">&quot;maf&quot;</span><span class="p">:</span> <span class="s2">&quot;maf&quot;</span><span class="p">,</span>
      <span class="s2">&quot;n&quot;</span> <span class="p">:</span> <span class="s2">&quot;n&quot;</span> <span class="p">,</span>
      <span class="s2">&quot;ma_samples&quot;</span><span class="p">:</span> <span class="s2">&quot;ma_samples&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ac&quot;</span><span class="p">:</span> <span class="s2">&quot;ma_count&quot;</span><span class="p">,</span>
      <span class="s2">&quot;molecular_trait_id&quot;</span><span class="p">:</span> <span class="s2">&quot;molecular_trait_id&quot;</span><span class="p">,</span> 
      <span class="s2">&quot;molecular_trait_object_id&quot;</span><span class="p">:</span> <span class="s2">&quot;molecular_trait_object_id&quot;</span><span class="p">}),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;TensorQTL&quot;</span><span class="p">]</span> <span class="p">)</span>

    <span class="n">data_temp</span><span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="p">[</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;nominal&quot;</span><span class="p">]:</span><span class="n">r</span><span class="p">,]]]</span>
    <span class="n">data_temp</span> <span class="o">=</span> <span class="n">data_temp</span><span class="p">[[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;sumstat_dir&#39;</span><span class="p">,</span> <span class="s1">&#39;column_info&#39;</span><span class="p">]]</span>
    <span class="n">data_temp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">)</span>
    <span class="n">column_info_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">)</span>

<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span> 
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tidyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;qvalue&quot;</span><span class="p">)</span>
    <span class="n">emprical_pd</span> <span class="o">=</span> <span class="n">tibble</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;regional&quot;</span><span class="p">]:</span><span class="n">r</span><span class="p">,]),</span> <span class="o">~</span><span class="n">read_delim</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)))</span><span class="o">%&gt;%</span><span class="n">unnest</span><span class="p">()</span>
    <span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;q_beta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tryCatch</span><span class="p">(</span><span class="n">qvalue</span><span class="p">(</span><span class="n">emprical_pd</span><span class="err">$</span><span class="n">p_beta</span><span class="p">)</span><span class="err">$</span><span class="n">qvalue</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">e</span><span class="p">){</span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Too few pvalue to calculate qvalue, fall back to BH&quot;</span><span class="p">)</span> 
                                                                                              <span class="n">qvalue</span><span class="p">(</span><span class="n">emprical_pd</span><span class="err">$</span><span class="n">p_beta</span><span class="p">,</span><span class="n">pi0</span> <span class="o">=</span> <span class="mi">1</span> <span class="p">)</span><span class="err">$</span><span class="n">qvalue</span><span class="p">})</span>  

    <span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;q_perm&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tryCatch</span><span class="p">(</span><span class="n">qvalue</span><span class="p">(</span><span class="n">emprical_pd</span><span class="err">$</span><span class="n">p_perm</span><span class="p">)</span><span class="err">$</span><span class="n">qvalue</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">e</span><span class="p">){</span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Too few pvalue to calculate qvalue, fall back to BH&quot;</span><span class="p">)</span> 
                                                                                              <span class="n">qvalue</span><span class="p">(</span><span class="n">emprical_pd</span><span class="err">$</span><span class="n">p_perm</span><span class="p">,</span><span class="n">pi0</span> <span class="o">=</span> <span class="mi">1</span> <span class="p">)</span><span class="err">$</span><span class="n">qvalue</span><span class="p">})</span>
    <span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;fdr_beta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="n">emprical_pd</span><span class="err">$</span><span class="n">p_beta</span><span class="p">,</span><span class="s2">&quot;fdr&quot;</span><span class="p">)</span>    
    <span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;fdr_perm&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="n">emprical_pd</span><span class="err">$</span><span class="n">p_perm</span><span class="p">,</span><span class="s2">&quot;fdr&quot;</span><span class="p">)</span>    
    <span class="n">summary</span> <span class="o">=</span> <span class="n">tibble</span><span class="p">(</span><span class="s2">&quot;fdr_perm_0.05&quot;</span> <span class="o">=</span>  <span class="nb">sum</span><span class="p">(</span><span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;fdr_perm&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)</span> <span class="p">,</span> 
                      <span class="s2">&quot;fdr_beta_0.05&quot;</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;fdr_beta&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">),</span>
                      <span class="s2">&quot;q_perm_0.05&quot;</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;q_perm&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)</span> <span class="p">,</span>
                      <span class="s2">&quot;q_beta_0.05&quot;</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;q_beta&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)</span> <span class="p">,</span>
                       <span class="s2">&quot;q_perm_0.01&quot;</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;q_perm&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.01</span><span class="p">)</span> <span class="p">,</span>
                      <span class="s2">&quot;q_beta_0.01&quot;</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">emprical_pd</span><span class="p">[</span><span class="s2">&quot;q_beta&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.01</span><span class="p">)</span>  <span class="p">)</span>
    <span class="n">emprical_pd</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$[_output[2]]&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">summary</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$[_output[3]]&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
        <span class="n">stdout</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info: $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_headerow:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">V</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">cat</span> <span class="err">$</span><span class="n">i</span>  <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">trans_2</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;nominal&quot;</span><span class="p">]</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.trans_qtl.meta_info&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;nominal&quot;</span><span class="p">]</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.trans_qtl.column_info&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> 
    <span class="n">data_temp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s2">&quot;sumstat_dir&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="s2">&quot;nominal&quot;</span><span class="p">]:</span><span class="n">r</span><span class="p">,]],</span>
    <span class="s2">&quot;column_info&quot;</span> <span class="p">:</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">]</span>
    <span class="p">})</span>
    <span class="n">column_info_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="s2">&quot;molecular_trait_id,chromosome,position,ref,alt&quot;</span><span class="p">,</span>
      <span class="s2">&quot;chromosome&quot;</span><span class="p">:</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span>
      <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span>
      <span class="s2">&quot;ref&quot;</span><span class="p">:</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span>
      <span class="s2">&quot;alt&quot;</span><span class="p">:</span> <span class="s2">&quot;alt&quot;</span><span class="p">,</span>
      <span class="s2">&quot;variant&quot;</span><span class="p">:</span> <span class="s2">&quot;variant_id&quot;</span><span class="p">,</span>
      <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="s2">&quot;beta&quot;</span><span class="p">,</span>
      <span class="s2">&quot;se&quot;</span><span class="p">:</span> <span class="s2">&quot;se&quot;</span><span class="p">,</span>
      <span class="s2">&quot;pvalue&quot;</span><span class="p">:</span> <span class="s2">&quot;pval&quot;</span><span class="p">,</span>
      <span class="s2">&quot;maf&quot;</span><span class="p">:</span> <span class="s2">&quot;maf&quot;</span><span class="p">,</span>
      <span class="s2">&quot;molecular_trait_id&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_ID&quot;</span><span class="p">}),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;TensorQTL&quot;</span><span class="p">]</span> <span class="p">)</span>
    <span class="n">data_temp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">)</span>
    <span class="n">column_info_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">)</span>

<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
        <span class="n">stdout</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info: $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_headerow:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">cat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">V</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">cat</span> <span class="err">$</span><span class="n">i</span>  <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;##&quot;</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/association_scan/TensorQTL"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../qtl_association_testing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">QTL association analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../APEX/APEX.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">APEX QTL association testing</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>