
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Quantile regression for QTL association testing &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/association_scan/quantile_models/qr_and_twas';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Hierarchical Multiple Testing" href="../qtl_association_postprocessing.html" />
    <link rel="prev" title="QTL Association Testing" href="../TensorQTL/TensorQTL.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../qtl_association_testing.html">QTL Association Analysis</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL Modifier Score</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_modifier_score/ems_training.html">xQTL Model Training Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_modifier_score/ems_prediction.html">EMS Prediction for xQTL Modifier Scores</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/association_scan/quantile_models/qr_and_twas.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/association_scan/quantile_models/qr_and_twas.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/association_scan/quantile_models/qr_and_twas.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Quantile regression for QTL association testing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-association-window-file">Example association-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantile-qtl-and-quantile-twas-weight">Quantile QTL and quantile TWAS weight</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="quantile-regression-for-qtl-association-testing">
<h1>Quantile regression for QTL association testing<a class="headerlink" href="#quantile-regression-for-qtl-association-testing" title="Link to this heading">#</a></h1>
<p>This notebook implements a workflow for using <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/37333162/">quantile regression of QTL</a>  and quantile TWAS to perform quantile QTL association testing and get TWAS weight.</p>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>A list of regions to be analyzed (optional); the last column of this file should be region name.</p></li>
<li><p>Either a list of per chromosome genotype files, or one file for genotype data of the entire genome. Genotype data has to be in PLINK <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p></li>
<li><p>Vector of lists of phenotype files per region to be analyzed, in UCSC <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> with index in <code class="docutils literal notranslate"><span class="pre">bed.gz.tbi</span></code> formats.</p></li>
<li><p>Vector of covariate files corresponding to the lists above.</p></li>
<li><p>Customized association windows file for variants (cis or trans). If it is not provided, a fixed sized window will be used around the region (a cis-window)</p></li>
<li><p>Optionally a vector of names of the phenotypic conditions in the form of <code class="docutils literal notranslate"><span class="pre">cond1</span> <span class="pre">cond2</span> <span class="pre">cond3</span></code> separated with whitespace.</p></li>
</ol>
<p>Input 2 and 3 should be outputs from <code class="docutils literal notranslate"><span class="pre">genotype_per_region</span></code> and <code class="docutils literal notranslate"><span class="pre">annotate_coord</span></code> modules in previous preprocessing steps. 4 should be output of <code class="docutils literal notranslate"><span class="pre">covariate_preprocessing</span></code> pipeline that contains genotype PC, phenotypic hidden confounders and fixed covariates.</p>
<section id="example-genotype-data">
<h3>Example genotype data<a class="headerlink" href="#example-genotype-data" title="Link to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr        path</span>
<span class="n">chr21</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21</span><span class="o">.</span><span class="n">bed</span>
<span class="n">chr22</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr22</span><span class="o">.</span><span class="n">bed</span>
</pre></div>
</div>
<p>Alternatively, simply use <code class="docutils literal notranslate"><span class="pre">protocol_example.genotype.chr21_22.bed</span></code> if all chromosomes are in the same file.</p>
</section>
<section id="example-phenotype-list">
<h3>Example phenotype list<a class="headerlink" href="#example-phenotype-list" title="Link to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end ID  path</span>
<span class="n">chr12</span>   <span class="mi">752578</span>  <span class="mi">752579</span>  <span class="n">ENSG00000060237</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">990508</span>  <span class="mi">990509</span>  <span class="n">ENSG00000082805</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">2794969</span> <span class="mi">2794970</span> <span class="n">ENSG00000004478</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">4649113</span> <span class="mi">4649114</span> <span class="n">ENSG00000139180</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6124769</span> <span class="mi">6124770</span> <span class="n">ENSG00000110799</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6534516</span> <span class="mi">6534517</span> <span class="n">ENSG00000111640</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="example-association-window-file">
<h3>Example association-window file<a class="headerlink" href="#example-association-window-file" title="Link to this heading">#</a></h3>
<p>It should have strictly 4 columns, with the header a commented out line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start    end    gene_id</span>
<span class="n">chr10</span>   <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008128</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008130</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000067606</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7101193</span>    <span class="n">ENSG00000069424</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7960000</span>    <span class="n">ENSG00000069812</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078369</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078808</span>
</pre></div>
</div>
<p>The key is that the 4th column ID should match with the 4th column ID in the phenotype list. Otherwise the association-window to analyze will not be found.</p>
</section>
<section id="about-indels">
<h3>About indels<a class="headerlink" href="#about-indels" title="Link to this heading">#</a></h3>
<p>Option <code class="docutils literal notranslate"><span class="pre">--no-indel</span></code> will remove indel from analysis. FIXME: Gao need to provide more guidelines how to deal with indels in practice.</p>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<p>For each analysis region, the output is quantile qtl model and quantile twas fitted and saved in RDS format.</p>
<p>For each gene, several of summary statistics files are generated, including all quantile qtl nominal test statistics for each test.</p>
<p>The columns of significant quantile qtl nominal association result are as follows:</p>
<ul class="simple">
<li><p>chr : Variant chromosome.</p></li>
<li><p>pos : Variant chromosomal position (basepairs).</p></li>
<li><p>ref : Variant reference allele (A, C, T, or G).</p></li>
<li><p>alt : Variant alternate allele.</p></li>
<li><p>phenotype_id: Molecular trait identifier.(gene)</p></li>
<li><p>variant_id: ID of the variant (rsid or chr:position:ref:alt)</p></li>
<li><p>p_qr(composite-p value using cauchy combination method): the integrated QR p-value across multiple quantile levels.</p></li>
<li><p>p_qr_0.05 to p_qr_0.95: quantile-specific QR p-values for the quantile levels 0.05, 0.1, …, 0.95.</p></li>
<li><p>qvalue_qr: the q-value of p_qr.</p></li>
<li><p>qvalue_qr_0.05 to qvalue_qr_0.95: quantile-specific QR q-values for the quantile levels 0.05, 0.1, …, 0.95.</p></li>
<li><p>zscore_qr_0.05 to zscore_qr_0.95: quantile-specific QR standard errors for the quantile levels 0.05, 0.1, …, 0.95.</p></li>
<li><p>coef_qr_0.05 to coef_qr_0.95 only for snps after LD clumping: quantile-specific QR coefficients for the quantile levels 0.05, 0.1, …, 0.95.</p></li>
</ul>
<p>Additionally, the matrix of quantile TWAS weights and pseudo R-squares is calculated and saved using Koenker and Machado’s Pseudo R² method, as described in their <a class="reference external" href="https://www.maths.usyd.edu.au/u/jchan/GLM/Koenker&amp;amp;Machado1999InferenceQuantileReg.pdf">Koenker and Machado, 1999: Inference in Quantile Regression</a></p>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<p>Timing [FIXME]</p>
<p>Below we duplicate the examples for phenotype and covariates to demonstrate that when there are multiple phenotypes for the same genotype it is possible to use this pipeline to analyze all of them (more than two is accepted as well).</p>
<p>Here using <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> we focus the analysis on 3 genes. In practice if this parameter is dropped, the union of all regions in all phenotype region lists will be analyzed. It is possible for some of the regions there are no genotype data, in which case the pipeline will output RDS files with a warning message to indicate the lack of genotype data to analyze.</p>
<p><strong>Note:</strong> Suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/quantile_twas.ipynb quantile_qtl_twas_weight  \
    --name protocol_example_protein  \
    --genoFile input/xqtl_association/protocol_example.genotype.chr21_22.bed   \
    --phenoFile output/phenotype/protocol_example.protein.region_list.txt \
                output/phenotype/protocol_example.protein.region_list.txt \
    --covFile output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz \
              output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz  \
    --customized-association-windows input/xqtl_association/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --region-name ENSG00000241973_P42356 ENSG00000160209_O00764 ENSG00000100412_Q99798 \
    --phenotype-names trait_A trait_B \
    --container oras://ghcr.io/cumc/pecotmr_apptainer:latest
</pre></div>
</div>
</div>
</div>
<p>It is also possible to analyze a selected list of regions using option <code class="docutils literal notranslate"><span class="pre">--region-list</span></code>. The last column of this file will be used for the list to analyze. Here for example use the same list of regions as we used for customized association-window:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run xqtl-protocol/pipeline/quantile_twas.ipynb quantile_qtl_twas_weight  \
    --name protocol_example_protein  \
    --genoFile xqtl_association/protocol_example.genotype.chr21_22.bed   \
    --phenoFile output/phenotype/protocol_example.protein.region_list.txt \
                output/phenotype/protocol_example.protein.region_list.txt \
    --covFile output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz \
              output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz  \
    --customized-association-windows xqtl_association/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --region-list xqtl_association/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --phenotype-names trait_A trait_B \
    --container oras://ghcr.io/cumc/pecotmr_apptainer:latest
</pre></div>
</div>
</div>
</div>
<p><strong>Note:</strong> When both <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> and <code class="docutils literal notranslate"><span class="pre">--region-list</span></code> are used, the union of regions from these parameters will be analyzed.</p>
<p>It is also possible to specify a subset of samples to analyze, using <code class="docutils literal notranslate"><span class="pre">--keep-samples</span></code> parameter. For example we create a file to keep the ID of 50 samples,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>zcat output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz | head -1 | awk &#39;{for (i=2; i&lt;=51; i++) printf $i &quot; &quot;; print &quot;&quot;}&#39;&gt; output/keep_samples.txt
</pre></div>
</div>
</div>
</div>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run quantile_twas.ipynb -h
</pre></div>
</div>
</div>
</div>
</section>
<section id="workflow-implementation">
<h2>Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# It is required to input the name of the analysis
parameter: name = str
parameter: cwd = path(&quot;output&quot;)
# A list of file paths for genotype data, or the genotype data itself. 
parameter: genoFile = path
# One or multiple lists of file paths for phenotype data.
parameter: phenoFile = paths
# One or multiple lists of file paths for phenotype ID mapping file. The first column should be the original ID, the 2nd column should be the ID to be mapped to.
parameter: phenoIDFile = paths()
# Covariate file path
parameter: covFile = paths
# Optional: if a region list is provide the analysis will be focused on provided region. 
# The LAST column of this list will contain the ID of regions to focus on
# Otherwise, all regions with both genotype and phenotype files will be analyzed
parameter: region_list = path()
# Optional: if a region name is provided 
# the analysis would be focused on the union of provides region list and region names
parameter: region_name = []
# Only focus on a subset of samples
parameter: keep_samples = path()
# An optional list documenting the custom association window for each region to analyze, with four column, chr, start, end, region ID (eg gene ID).
# If this list is not provided, the default `window` parameter (see below) will be used.
parameter: customized_association_windows = path()
# Specify the cis window for the up and downstream radius to analyze around the region of interest in units of bp
# When this is set to negative, we will rely on using customized_association_windows
parameter: cis_window = -1
# save data object or not
parameter: save_data = False
# Name of phenotypes
parameter: phenotype_names = [f&#39;{x:bn}&#39; for x in phenoFile]
parameter: seed = 999
# remove a variant if it has more than imiss missing individual level data
parameter: imiss = 1.0
# MAF cutoff
parameter: maf = 0.0025
# MAC cutoff, on top of MAF cutoff
parameter: mac = 5
# Remove indels if indel = False
parameter: indel = True
parameter: min_twas_maf = 0.01
parameter: screen_threshold = 0.01
parameter: ld_reference_meta_file = path()
# Analysis environment settings
parameter: container = &quot;&quot;
import re
parameter: entrypoint= (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;
# For cluster jobs, number commands to run per job
parameter: job_size = 200
# Wall clock time expected
parameter: walltime = &quot;1h&quot;
# Memory expected
parameter: mem = &quot;20G&quot;
# Number of threads
parameter: numThreads = 1

if len(phenoFile) != len(covFile):
    raise ValueError(&quot;Number of input phenotypes files must match that of covariates files&quot;)
if len(phenoFile) != len(phenotype_names):
    raise ValueError(&quot;Number of input phenotypes files must match the number of phenotype names&quot;)
if len(phenoIDFile) &gt; 0 and len(phenoFile) != len(phenoIDFile):
    raise ValueError(&quot;Number of input phenotypes files must match the number of phenotype ID mapping files&quot;)

def group_by_region(lst, partition):
    # from itertools import accumulate
    # partition = [len(x) for x in partition]
    # Compute the cumulative sums once
    # cumsum_vector = list(accumulate(partition))
    # Use slicing based on the cumulative sums
    # return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]
    return partition

import os
import pandas as pd

def adapt_file_path(file_path, reference_file):
    &quot;&quot;&quot;
    Adapt a single file path based on its existence and a reference file&#39;s path.

    Args:
    - file_path (str): The file path to adapt.
    - reference_file (str): File path to use as a reference for adaptation.

    Returns:
    - str: Adapted file path.

    Raises:
    - FileNotFoundError: If no valid file path is found.
    &quot;&quot;&quot;
    reference_path = os.path.dirname(reference_file)

    # Check if the file exists
    if os.path.isfile(file_path):
        return file_path

    # Check file name without path
    file_name = os.path.basename(file_path)
    if os.path.isfile(file_name):
        return file_name

    # Check file name in reference file&#39;s directory
    file_in_ref_dir = os.path.join(reference_path, file_name)
    if os.path.isfile(file_in_ref_dir):
        return file_in_ref_dir

    # Check original file path prefixed with reference file&#39;s directory
    file_prefixed = os.path.join(reference_path, file_path)
    if os.path.isfile(file_prefixed):
        return file_prefixed

    # If all checks fail, raise an error
    raise FileNotFoundError(f&quot;No valid path found for file: {file_path}&quot;)

def adapt_file_path_all(df, column_name, reference_file):
    return df[column_name].apply(lambda x: adapt_file_path(x, reference_file))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[get_analysis_regions: shared = &quot;regional_data&quot;]
# input is genoFile, phenoFile, covFile and optionally region_list. If region_list presents then we only analyze what&#39;s contained in the list.
# regional_data should be a dictionary like:
#{&#39;data&#39;: [(&quot;genotype_1.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;covariate_1.gz&quot;), (&quot;genotype_2.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;phenotype_2.bed.gz&quot;, &quot;covariate_1.gz&quot;, &quot;covariate_2.gz&quot;) ... ],
# &#39;meta_info&#39;: [(&quot;chr12:752578-752579&quot;,&quot;chr12:752577-752580&quot;, &quot;gene_1&quot;, &quot;trait_1&quot;), (&quot;chr13:852580-852581&quot;,&quot;chr13:852579-852580&quot;, &quot;gene_2&quot;, &quot;trait_1&quot;, &quot;trait_2&quot;) ... ]}
import numpy as np

def preload_id_map(id_map_files):
    id_maps = {}
    for id_map_file in id_map_files:
        if id_map_file is not None and os.path.isfile(id_map_file):
            df = pd.read_csv(id_map_file, sep=&#39;\s+&#39;, header=None, comment=&#39;#&#39;, names=[&#39;old_ID&#39;, &#39;new_ID&#39;])
            id_maps[id_map_file] = df.set_index(&#39;old_ID&#39;)[&#39;new_ID&#39;].to_dict()
    return id_maps

def load_and_apply_id_map(pheno_path, id_map_path, preloaded_id_maps):
    pheno_df = pd.read_csv(pheno_path, sep=&quot;\s+&quot;, header=0)
    pheno_df[&#39;Original_ID&#39;] = pheno_df[&#39;ID&#39;]
    if id_map_path in preloaded_id_maps:
        id_map = preloaded_id_maps[id_map_path]
        pheno_df[&#39;ID&#39;] = pheno_df[&#39;ID&#39;].map(id_map).fillna(pheno_df[&#39;ID&#39;])
    return pheno_df

def filter_by_region_ids(data, region_ids):
    if region_ids is not None and len(region_ids) &gt; 0:
        data = data[data[&#39;ID&#39;].isin(region_ids)]
    return data

def custom_join(series):
    # Initialize an empty list to hold the processed items
    result = []
    for item in series:
        if &#39;,&#39; in item:
            # If the item contains commas, split by comma and convert to tuple
            result.append(tuple(item.split(&#39;,&#39;)))
        else:
            # If the item does not contain commas, add it directly
            result.append(item)
    # Convert the list of items to a tuple and return
    return tuple(result)

def aggregate_phenotype_data(accumulated_pheno_df):
    if not accumulated_pheno_df.empty:
        accumulated_pheno_df = accumulated_pheno_df.groupby([&#39;#chr&#39;,&#39;ID&#39;,&#39;cond&#39;,&#39;path&#39;,&#39;cov_path&#39;], as_index=False).agg({
            &#39;#chr&#39;: lambda x: np.unique(x).astype(str)[0],
            &#39;ID&#39;: lambda x: np.unique(x).astype(str)[0],
            &#39;Original_ID&#39;: &#39;,&#39;.join,
            &#39;start&#39;: &#39;min&#39;,
            &#39;end&#39;: &#39;max&#39;
        }).groupby([&#39;#chr&#39;,&#39;ID&#39;], as_index=False).agg({
            &#39;cond&#39;: &#39;,&#39;.join,
            &#39;path&#39;: &#39;,&#39;.join,
            &#39;Original_ID&#39;: custom_join,
            &#39;cov_path&#39;: &#39;,&#39;.join,
            &#39;start&#39;: &#39;min&#39;,
            &#39;end&#39;: &#39;max&#39;
        })
    return accumulated_pheno_df

def process_cis_files(pheno_files, cov_files, phenotype_names, pheno_id_files, region_ids, preloaded_id_maps):
    &#39;&#39;&#39;
    Example output:
    #chr    start      end    ID  Original_ID   path     cov_path             cond
    chr12   752578   752579  ENSG00000060237  Q9H4A3,P62873  protocol_example.protein_1.bed.gz,protocol_example.protein_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B
    &#39;&#39;&#39;
    accumulated_pheno_df = pd.DataFrame()
    pheno_id_files = [None] * len(pheno_files) if len(pheno_id_files) == 0 else pheno_id_files
    for pheno_path, cov_path, phenotype_name, id_map_path in zip(pheno_files, cov_files, phenotype_names, pheno_id_files):
        if not os.path.isfile(cov_path):
            raise FileNotFoundError(f&quot;No valid path found for file: {cov_path}&quot;)
        pheno_df = load_and_apply_id_map(pheno_path, id_map_path, preloaded_id_maps)
        pheno_df = filter_by_region_ids(pheno_df, region_ids)
        if not pheno_df.empty:
            pheno_df.iloc[:, 4] = adapt_file_path_all(pheno_df, pheno_df.columns[4], f&quot;{pheno_path:a}&quot;)
            pheno_df = pheno_df.assign(cov_path=str(cov_path), cond=phenotype_name)           
            accumulated_pheno_df = pd.concat([accumulated_pheno_df, pheno_df], ignore_index=True)

    accumulated_pheno_df = aggregate_phenotype_data(accumulated_pheno_df)
    return accumulated_pheno_df

def process_trans_files(pheno_files, cov_files, phenotype_names, pheno_id_files, region_ids, customized_association_windows):
    &#39;&#39;&#39;
    Example output:
    #chr    start      end    ID  Original_ID   path     cov_path             cond
    chr21   0   0  chr21_18133254_19330300  carnitine,benzoate,hippurate  metabolon_1.bed.gz,metabolon_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B
    &#39;&#39;&#39;
    
    if not os.path.isfile(customized_association_windows):
        raise ValueError(&quot;Customized association analysis window must be specified for trans analysis.&quot;)
    accumulated_pheno_df = pd.DataFrame()
    pheno_id_files = [None] * len(pheno_files) if len(pheno_id_files) == 0 else pheno_id_files
    genotype_windows = pd.read_csv(customized_association_windows, comment=&quot;#&quot;, header=None, names=[&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;ID&quot;], sep=&quot;\t&quot;)
    genotype_windows = filter_by_region_ids(genotype_windows, region_ids)
    if genotype_windows.empty:
        return accumulated_pheno_df
    
    for pheno_path, cov_path, phenotype_name, id_map_path in zip(pheno_files, cov_files, phenotype_names, pheno_id_files):
        if not os.path.isfile(cov_path):
            raise FileNotFoundError(f&quot;No valid path found for file: {cov_path}&quot;)
        pheno_df = pd.read_csv(pheno_path, sep=&quot;\s+&quot;, header=0, names=[&#39;Original_ID&#39;, &#39;path&#39;])
        if not pheno_df.empty:
            pheno_df.iloc[:, -1] = adapt_file_path_all(pheno_df, pheno_df.columns[-1], f&quot;{pheno_path:a}&quot;)
            pheno_df = pheno_df.assign(cov_path=str(cov_path), cond=phenotype_name)
            # Here we combine genotype_windows which contains &quot;#chr&quot; and &quot;ID&quot; to pheno_df by creating a cartesian product
            pheno_df = pd.merge(genotype_windows.assign(key=1), pheno_df.assign(key=1), on=&#39;key&#39;).drop(&#39;key&#39;, axis=1)
            # then set start and end columns to zero
            pheno_df[&#39;start&#39;] = 0
            pheno_df[&#39;end&#39;] = 0
            if id_map_path is not None:
                # Filter pheno_df by specific association-window and phenotype pairs
                association_analysis_pair = pd.read_csv(id_map_path, sep=&#39;\s+&#39;, header=None, comment=&#39;#&#39;, names=[&#39;ID&#39;, &#39;Original_ID&#39;])
                pheno_df = pd.merge(pheno_df, association_analysis_pair, on=[&#39;ID&#39;, &#39;Original_ID&#39;])
            accumulated_pheno_df = pd.concat([accumulated_pheno_df, pheno_df], ignore_index=True)

    accumulated_pheno_df = aggregate_phenotype_data(accumulated_pheno_df)
    return accumulated_pheno_df

# Load genotype meta data
if f&quot;{genoFile:x}&quot; == &quot;.bed&quot;:
    geno_meta_data = pd.DataFrame([(&quot;chr&quot;+str(x), f&quot;{genoFile:a}&quot;) for x in range(1,23)] + [(&quot;chrX&quot;, f&quot;{genoFile:a}&quot;)], columns=[&#39;#chr&#39;, &#39;geno_path&#39;])
else:
    geno_meta_data = pd.read_csv(f&quot;{genoFile:a}&quot;, sep = &quot;\s+&quot;, header=0)
    geno_meta_data.iloc[:, 1] = adapt_file_path_all(geno_meta_data, geno_meta_data.columns[1], f&quot;{genoFile:a}&quot;)
    geno_meta_data.columns = [&#39;#chr&#39;, &#39;geno_path&#39;]
    geno_meta_data[&#39;#chr&#39;] = geno_meta_data[&#39;#chr&#39;].apply(lambda x: str(x) if str(x).startswith(&#39;chr&#39;) else f&#39;chr{x}&#39;)

# Checking the DataFrame
valid_chr_values = [f&#39;chr{x}&#39; for x in range(1, 23)] + [&#39;chrX&#39;]
if not all(value in valid_chr_values for value in geno_meta_data[&#39;#chr&#39;]):
    raise ValueError(&quot;Invalid chromosome values found. Allowed values are chr1 to chr22 and chrX.&quot;)

region_ids = []
# If region_list is provided, read the file and extract IDs
if region_list.is_file():
    region_list_df = pd.read_csv(region_list, delim_whitespace=True, header=None, comment = &quot;#&quot;)
    region_ids = region_list_df.iloc[:, -1].unique()  # Extracting the last column for IDs
# If region_name is provided, include those IDs as well
# --region-name A B C will result in a list of [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;] here
if len(region_name) &gt; 0:
    region_ids = list(set(region_ids).union(set(region_name)))

trans_analysis = False
if trans_analysis:
    meta_data = process_trans_files(phenoFile, covFile, phenotype_names, phenoIDFile, region_ids, customized_association_windows)
else:
    meta_data = process_cis_files(phenoFile, covFile, phenotype_names, phenoIDFile, region_ids, preload_id_map(phenoIDFile))

if not meta_data.empty:
    meta_data = meta_data.merge(geno_meta_data, on=&#39;#chr&#39;, how=&#39;inner&#39;)
    # Adjust association-window
    if os.path.isfile(customized_association_windows):
        print(f&quot;Loading customized association analysis window from {customized_association_windows}&quot;)
        association_windows_list = pd.read_csv(customized_association_windows, comment=&quot;#&quot;, header=None, names=[&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;ID&quot;], sep=&quot;\t&quot;)
        meta_data = pd.merge(meta_data, association_windows_list, on=[&#39;#chr&#39;, &#39;ID&#39;], how=&#39;left&#39;, suffixes=(&#39;&#39;, &#39;_association&#39;))
        mismatches = meta_data[meta_data[&#39;start_association&#39;].isna()]
        if not mismatches.empty:
            raise ValueError(f&quot;{len(mismatches)} regions to analyze cannot be found in ``{customized_association_windows}``. Please check your ``{customized_association_windows}`` database to make sure it contains all association-window definitions. &quot;)
    else:
        if cis_window &lt; 0 :
            raise ValueError(&quot;Please either input valid path to association-window file via ``--customized-association-windows``, or set ``--cis-window`` to a non-negative integer.&quot;)
        if cis_window == 0:
            print(&quot;Warning: only variants within the range of start and end of molecular phenotype will be considered since cis_window is set to zero and no customized association window file was found. Please make sure this is by design.&quot;)
        meta_data[&#39;start_association&#39;] = meta_data[&#39;start&#39;].apply(lambda x: max(x - cis_window, 0))
        meta_data[&#39;end_association&#39;] = meta_data[&#39;end&#39;] + cis_window

    # Example meta_data:
    # #chr    start      end    start_association       end_association           ID  Original_ID   path     cov_path             cond             coordinate     geno_path
    # 0  chr12   752578   752579  652578   852579  ENSG00000060237  Q9H4A3,P62873  protocol_example.protein_1.bed.gz,protocol_example.protein_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B    chr12:752578-752579  protocol_example.genotype.chr21_22.bed       
    # Create the final dictionary
    regional_data = {
        &#39;data&#39;: [(row[&#39;geno_path&#39;], *row[&#39;path&#39;].split(&#39;,&#39;), *row[&#39;cov_path&#39;].split(&#39;,&#39;)) for _, row in meta_data.iterrows()],
        &#39;meta_info&#39;: [(f&quot;{row[&#39;#chr&#39;]}:{row[&#39;start&#39;]}-{row[&#39;end&#39;]}&quot;, # this is the phenotypic region to extract data from
                       f&quot;{row[&#39;#chr&#39;]}:{row[&#39;start_association&#39;]}-{row[&#39;end_association&#39;]}&quot;, # this is the association window region
                       row[&#39;ID&#39;], row[&#39;Original_ID&#39;], *row[&#39;cond&#39;].split(&#39;,&#39;)) for _, row in meta_data.iterrows()]
    }
else:
    regional_data = {&#39;data&#39;:[], &#39;meta_info&#39;:[]}
</pre></div>
</div>
</div>
</div>
</section>
<section id="quantile-qtl-and-quantile-twas-weight">
<h2>Quantile QTL and quantile TWAS weight<a class="headerlink" href="#quantile-qtl-and-quantile-twas-weight" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[quantile_qtl_twas_weight]
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name}/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.univariate_qr_twas_weights.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    options(warn=1)
    library(pecotmr)
    start_time_total &lt;- proc.time()
    
    phenotype_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])})
    covariate_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])})
    conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4:]])})
    region = ${(&quot;&#39;%s&#39;&quot; % _meta_info[0]) if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &#39;NULL&#39;} # if the end position is zero return NULL
    association_window = &quot;${_meta_info[1]}&quot;
    extract_region_name = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])})
    phenotype_header = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;}
    region_name_col = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;} 


    # extract subset of samples
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }

    # Load regional association data
    tryCatch({
    fdat = load_regional_association_data(genotype = ${_input[0]:anr},
                                          phenotype = phenotype_files,
                                          covariate = covariate_files,
                                          region = ${(&quot;&#39;%s&#39;&quot; % _meta_info[0]) if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &#39;NULL&#39;}, # if the end position is zero return NULL
                                          association_window = &quot;${_meta_info[1]}&quot;,
                                          conditions = conditions,
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                          keep_samples = keep_samples,
                                          extract_region_name = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}),
                                          phenotype_header = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                          region_name_col = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                          scale_residuals = FALSE)
    }, NoSNPsError = function(e) {
        message(&quot;Error: &quot;, paste(e$message, &quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot;))
        saveRDS(list(${_meta_info[2]} = e$message), ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    })
  
    if (${&quot;TRUE&quot; if save_data else &quot;FALSE&quot;}) {
        # save data object for debug purpose
        saveRDS(list(${_meta_info[2]} = fdat), &quot;${_output:ann}.univariate.rds&quot;, compress=&#39;xz&#39;)
    }
    
    # setup univariate analysis pipeline options
    if (&quot;${_meta_info[2]}&quot; != &quot;${_meta_info[3]}&quot;) {
        region_name = c(&quot;${_meta_info[2]}&quot;, c(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}))
    } else {
        region_name = &quot;${_meta_info[2]}&quot;
    }
  
    region_info = list(region_coord=parse_region(&quot;${_meta_info[0]}&quot;), grange=parse_region(&quot;${_meta_info[1]}&quot;), region_name=region_name)
    
    fitted = list()
    condition_names = vector()
    r = 1
    while (r&lt;=length(fdat$Y)) {
        X &lt;- fdat$X_data[[r]]  
        Y &lt;- fdat$Y[[r]]
        if (is.null(dim(Y))) {
            Y &lt;- matrix(Y, nrow = length(Y), ncol = 1)
        }
        colnames(Y) &lt;- extract_region_name[[r]] 
        Z &lt;- fdat$covar[[r]]     
        # Update condition names first
        new_names = names(fdat$residual_Y)[r]
        new_col_names = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])})[[r]]
        if (is.null(new_col_names)) {
            new_col_names = 1:ncol(fdat$residual_Y[[r]])
        }
        if(!(identical(new_names, new_col_names))) {
            new_names = paste(new_names, new_col_names, sep=&quot;_&quot;)
        }
                
        column_results &lt;- lapply(1:ncol(Y), function(i) {
            Y_col &lt;- matrix(Y[,i], ncol=1)
            colnames(Y_col) &lt;- colnames(Y)[i]
            
            qr_results = quantile_twas_weight_pipeline(
                X = X, 
                Y = Y_col,
                Z = Z, 
                ld_reference_meta_file=${(&#39;&quot;%s&quot;&#39; % ld_reference_meta_file) if not ld_reference_meta_file.is_dir() else &quot;NULL&quot;},
                maf = fdat$maf[[r]],
                twas_maf_cutoff = ${min_twas_maf},                                            
                region_id = paste0(colnames(Y_col), &quot;_&quot;, names(fdat$residual_Y)[r]),
                quantile_qtl_tau_list = seq(0.05, 0.95, by = 0.05),
                quantile_twas_tau_list = seq(0.01, 0.99, by = 0.01),
                screen_threshold = ${screen_threshold}
            )
            
            if (!is.null(qr_results$message)) {
                message(qr_results$message)
            }
            
            qr_results$region_info = region_info
            qr_results$maf = fdat$maf[[r]]
            
            return(qr_results)
        })
        
        fitted &lt;- c(fitted, column_results)
        condition_names &lt;- c(condition_names, new_names)
        
        if (length(new_names) &gt; 0) {
            message(&quot;Analysis completed for: &quot;, paste(new_names, collapse=&quot;,&quot;))
        }
        
        # Release memory
        fdat$residual_X[[r]] &lt;- NA
        fdat$residual_Y[[r]] &lt;- NA      
        r = r + 1
    }

    # Set names for the final results
    if (length(fitted) &gt; 0) {
        names(fitted) &lt;- condition_names
    }

    saveRDS(list(&quot;${_meta_info[2]}&quot; = fitted), ${_output:ar}, compress=&#39;xz&#39;)
    end_time_total &lt;- proc.time()
    total_time &lt;- end_time_total - start_time_total
    print(total_time)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/association_scan/quantile_models"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../TensorQTL/TensorQTL.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">QTL Association Testing</p>
      </div>
    </a>
    <a class="right-next"
       href="../qtl_association_postprocessing.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Hierarchical Multiple Testing</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-association-window-file">Example association-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantile-qtl-and-quantile-twas-weight">Quantile QTL and quantile TWAS weight</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>