
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>xQTL Model Training Tutorial &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/xqtl_modifier_score/ems_training';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="EMS Prediction for xQTL Modifier Scores" href="ems_prediction.html" />
    <link rel="prev" title="Stratified LD Score Regression" href="../enrichment/sldsc_enrichment.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL Modifier Score</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">xQTL Model Training Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="ems_prediction.html">EMS Prediction for xQTL Modifier Scores</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/xqtl_modifier_score/ems_training.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/xqtl_modifier_score/ems_training.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/xqtl_modifier_score/ems_training.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>xQTL Model Training Tutorial</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-weighted-catboost-algorithm">Feature-Weighted CatBoost Algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#training-data-construction">Training Data Construction</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-data-feature-categories-from-manuscript-table-c">Input Data: Feature Categories from Manuscript Table C</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-categories">Feature Categories</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distance-features">1. Distance Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-type-regulatory-features">2. Cell-Type Regulatory Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genetics-features">3. Population Genetics Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#conservation-features">4. Conservation Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#deep-learning-predictions">5. Deep Learning Predictions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output-data-generated"><strong>Output Data Generated</strong></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#what-we-re-trying-to-predict-y">What We‚Äôre Trying to Predict (Y)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description-of-toy-dataset">Description of Toy Dataset</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dataset-overview">Dataset Overview</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#data-split-strategy">Data Split Strategy</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#key-parameters">Key Parameters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-5-hyperparameters">Model 5 Hyperparameters</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-weighting-strategy">Feature Weighting Strategy:</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cross-validation-configuration">Cross-Validation Configuration</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#leave-one-chromosome-out-loco-procedure">Leave-One-Chromosome-Out (LOCO) Procedure</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mathematical-framework">Mathematical Framework</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#detailed-workflow">Detailed Workflow</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-running-the-pipeline">Step 1: Running the Pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-understanding-training-process-overview">Step 2: Understanding Training Process Overview</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-results-results-analysis">Step 3: Results &amp; Results Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#results-analysis">Results Analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#strong-performance-indicators-our-results">üéØ <strong>Strong Performance Indicators (Our Results):</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#what-this-means">‚úÖ <strong>What This Means:</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#key-biological-findings">üî¨ <strong>Key Biological Findings:</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-generated-output-files">Step 4: Generated Output Files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-application">Model Application</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-using-your-trained-models">Step 5: Using Your Trained Models</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="xqtl-model-training-tutorial">
<h1>xQTL Model Training Tutorial<a class="headerlink" href="#xqtl-model-training-tutorial" title="Link to this heading">#</a></h1>
<p>This notebook implements the sc-EMS (single-cell Expression Modifier Scores) methodology from <a class="reference internal" href="#FIXME"><span class="xref myst">our previous research (FIXME)</span></a>. The original research demonstrated that machine learning can predict which genetic variants affect gene expression in specific brain cell types, particularly for Alzheimer‚Äôs disease research.</p>
<section id="motivation">
<h2>Motivation<a class="headerlink" href="#motivation" title="Link to this heading">#</a></h2>
<p>Traditional genetic studies only explain 5% of Alzheimer‚Äôs disease heritability. The remaining 95% ‚Äúmissing heritability‚Äù occurs because bulk tissue approaches cannot capture regulatory effects that occur in specific cell types. Our machine learning approach identifies genetic variants that affect gene expression specifically in brain cell types like microglia, potentially explaining previously undetected disease mechanisms.</p>
<ul class="simple">
<li><p>Our Motivation is to create a reproducible pipeline that accurately predicts functional genetic variants using cell-type specific genomic annotations, transforming complex research findings into a standardized tool for the broader scientific community.</p></li>
</ul>
</section>
<section id="methods-overview">
<h2>Methods Overview<a class="headerlink" href="#methods-overview" title="Link to this heading">#</a></h2>
<section id="feature-weighted-catboost-algorithm">
<h3>Feature-Weighted CatBoost Algorithm<a class="headerlink" href="#feature-weighted-catboost-algorithm" title="Link to this heading">#</a></h3>
<p>We use a single <strong>CatBoost</strong> (Categorical Boosting) gradient boosting model optimized for genomic data. CatBoost builds an ensemble of decision trees sequentially, where each tree learns from previous errors, making it particularly effective for high-dimensional biological datasets with mixed data types.</p>
<p><strong>Our Training Strategy</strong>: Single feature-weighted model (Model 5) that emphasizes biology-informed features while maintaining comprehensive genomic context.</p>
</section>
<section id="training-data-construction">
<h3>Training Data Construction<a class="headerlink" href="#training-data-construction" title="Link to this heading">#</a></h3>
<p><strong>Data Source</strong>: Chromosome 2 variants with non-overlapping splits</p>
<ul class="simple">
<li><p><strong>Training set</strong>: 3,056 variants (80% of available data)</p></li>
<li><p><strong>Testing set</strong>: 761 variants (20% of available data)</p></li>
<li><p><strong>Critical validation</strong>: Zero variant overlap between training and testing sets</p></li>
</ul>
<p><strong>Label Definition (Y)</strong>:</p>
<ul class="simple">
<li><p>Y = 1: Functional eQTL (PIP &gt; 0.1) - variant significantly affects gene expression</p></li>
<li><p>Y = 0: Non-functional variant (PIP &lt; 0.01) - no detectable expression effect</p></li>
<li><p><strong>Class distribution</strong>: 9% positive rate reflects biological reality that most variants are non-functional</p></li>
</ul>
</section>
</section>
<section id="input-data-feature-categories-from-manuscript-table-c">
<h2>Input Data: Feature Categories from Manuscript Table C<a class="headerlink" href="#input-data-feature-categories-from-manuscript-table-c" title="Link to this heading">#</a></h2>
<p>Based on <a class="reference internal" href="#FIXME"><span class="xref myst">our previous research (FIXME)</span></a> (Figure 1, Table C), we use these broad feature categories. Each category represents a branch of features that can be applied to many different contexts and datasets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load gene constraint data</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="n">gene_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;data/41588_2024_1820_MOESM4_ESM.xlsx&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Sample output showing GeneBayes constraint scores:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ensg</span>            <span class="n">hgnc</span>        <span class="n">chrom</span>  <span class="n">obs_lof</span>  <span class="n">exp_lof</span>   <span class="n">post_mean</span>    <span class="n">post_lower_95</span>  <span class="n">post_upper_95</span>
<span class="n">ENSG00000198488</span> <span class="n">HGNC</span><span class="p">:</span><span class="mi">24141</span>  <span class="n">chr11</span>    <span class="mi">12</span>      <span class="mf">8.9777</span>   <span class="mf">6.46629E-05</span>  <span class="mf">7.16256E-06</span>   <span class="mf">0.00017805</span>
<span class="n">ENSG00000164363</span> <span class="n">HGNC</span><span class="p">:</span><span class="mi">26441</span>  <span class="n">chr5</span>     <span class="mi">31</span>      <span class="mf">28.55</span>    <span class="mf">0.00016062</span>   <span class="mf">2.59918E-05</span>   <span class="mf">0.00044175</span>
<span class="n">ENSG00000159337</span> <span class="n">HGNC</span><span class="p">:</span><span class="mi">30038</span>  <span class="n">chr15</span>    <span class="mi">28</span>      <span class="mf">41.84</span>    <span class="mf">0.00016978</span>   <span class="mf">0.000018674</span>   <span class="mf">0.00053317</span>
</pre></div>
</div>
<section id="feature-categories">
<h3>Feature Categories<a class="headerlink" href="#feature-categories" title="Link to this heading">#</a></h3>
<section id="distance-features">
<h4>1. Distance Features<a class="headerlink" href="#distance-features" title="Link to this heading">#</a></h4>
<p><strong>Biological Rationale</strong>: Physical proximity determines regulatory potential</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">abs_distance_TSS_log</span></code>: Log-transformed distance to transcription start site</p></li>
<li><p><strong>Example</strong>: Value 8.2 = ~160,000 base pairs from gene start</p></li>
<li><p><strong>Impact</strong>: Closest predictor of regulatory effects (importance: 23.58)</p></li>
</ul>
</section>
<section id="cell-type-regulatory-features">
<h4>2. Cell-Type Regulatory Features<a class="headerlink" href="#cell-type-regulatory-features" title="Link to this heading">#</a></h4>
<p><strong>Biological Rationale</strong>: Microglia-specific regulatory mechanisms</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ABC_score_microglia</span></code>: Activity-by-Contact regulatory activity score</p></li>
<li><p><strong>Example</strong>: Score 0.73 indicates high regulatory potential in microglia</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">diff_32_ENCFF140MBA</span></code>: Cell-type chromatin accessibility change</p></li>
<li><p><strong>Example</strong>: Value 1.83 indicates strong accessibility difference</p></li>
<li><p><strong>Impact</strong>: Multiple diff_32_* features contribute consistently (0.5-1.8 importance)</p></li>
</ul>
</section>
<section id="population-genetics-features">
<h4>3. Population Genetics Features<a class="headerlink" href="#population-genetics-features" title="Link to this heading">#</a></h4>
<p><strong>Biological Rationale</strong>: Evolutionary constraint and population frequency</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gnomad_MAF</span></code>: Minor allele frequency from population database</p></li>
<li><p><strong>Example</strong>: 0.15 = variant found in 15% of human population</p></li>
<li><p><strong>Impact</strong>: Secondary predictor (importance: 0.88)</p></li>
</ul>
</section>
<section id="conservation-features">
<h4>4. Conservation Features<a class="headerlink" href="#conservation-features" title="Link to this heading">#</a></h4>
<p><strong>Biological Rationale</strong>: Evolutionary preservation indicates functional importance</p>
<ul class="simple">
<li><p>Cross-species conservation scores and constraint metrics</p></li>
<li><p><strong>Impact</strong>: Moderate contribution to overall prediction</p></li>
</ul>
</section>
<section id="deep-learning-predictions">
<h4>5. Deep Learning Predictions<a class="headerlink" href="#deep-learning-predictions" title="Link to this heading">#</a></h4>
<p><strong>Biological Rationale</strong>: Sequence-to-function model predictions</p>
<ul class="simple">
<li><p>Enformer and BPNet chromatin accessibility predictions</p></li>
<li><p><strong>Impact</strong>: Complementary information beyond experimental features</p></li>
</ul>
</section>
</section>
</section>
<section id="output-data-generated">
<h2><strong>Output Data Generated</strong><a class="headerlink" href="#output-data-generated" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Output Type</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Research Use</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Trained models</strong></p></td>
<td><p>Single feature-weighted CatBoost classifier</p></td>
<td><p>Variant effect prediction</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Performance metrics</strong></p></td>
<td><p>AP/AUC scores</p></td>
<td><p>Model comparison</p></td>
</tr>
<tr class="row-even"><td><p><strong>Feature importance</strong></p></td>
<td><p>Genomic feature rankings</p></td>
<td><p>Biological interpretation</p></td>
</tr>
</tbody>
</table>
</div>
<section id="what-we-re-trying-to-predict-y">
<h3>What We‚Äôre Trying to Predict (Y)<a class="headerlink" href="#what-we-re-trying-to-predict-y" title="Link to this heading">#</a></h3>
<p><strong>Binary Classification:</strong></p>
<ul class="simple">
<li><p>Y = 1: Functional variant (affects gene expression)</p></li>
<li><p>Y = 0: Non-functional variant (no effect on genes)</p></li>
</ul>
<p><strong>Training Data Breakdown:</strong></p>
<ul class="simple">
<li><p>Total: 3,056 variants</p></li>
<li><p>Functional (Y=1): 278 variants (9%)</p></li>
<li><p>Non-functional (Y=0): 2,778 variants (91%)</p></li>
<li><p>This 9% rate reflects biological reality - most genetic variants don‚Äôt affect genes</p></li>
</ul>
</section>
</section>
<section id="description-of-toy-dataset">
<h2>Description of Toy Dataset<a class="headerlink" href="#description-of-toy-dataset" title="Link to this heading">#</a></h2>
<section id="dataset-overview">
<h3>Dataset Overview<a class="headerlink" href="#dataset-overview" title="Link to this heading">#</a></h3>
<p>This tutorial demonstrates the sc-EMS (single-cell Expression Modifier Scores) methodology from <a class="reference internal" href="#FIXME"><span class="xref myst">our previous research (FIXME)</span></a>, which predicts brain cell type eQTLs for Alzheimer‚Äôs disease research. Our toy dataset represents a subset of the full study data:</p>
<p><strong>Dataset specifications:</strong></p>
<ul class="simple">
<li><p><strong>Training variants</strong>: 3,056 genetic variants from chromosome 2</p></li>
<li><p><strong>Testing variants</strong>: 761 genetic variants from chromosome 2 (non-overlapping)</p></li>
<li><p><strong>Feature dimensions</strong>: 4,839 genomic annotations per variant</p></li>
<li><p><strong>Target prediction</strong>: Binary classification (functional vs. non-functional eQTLs)</p></li>
</ul>
<p><strong>What makes this a ‚Äútoy‚Äù dataset:</strong></p>
<ul class="simple">
<li><p>We use a smaller, simpler version of the full research study</p></li>
<li><p>The full study used all 22 chromosomes with thousands of people</p></li>
<li><p>Our tutorial uses only chromosome 2 data from a smaller sample</p></li>
</ul>
<p><strong>Why we can still prevent data leakage:</strong></p>
<ul class="simple">
<li><p>We split chromosome 2 data into two completely separate groups</p></li>
<li><p>Training data: 3,056 variants with looser selection criteria</p></li>
<li><p>Test data: 761 variants with stricter selection criteria</p></li>
<li><p>No variant appears in both groups</p></li>
</ul>
</section>
<section id="data-split-strategy">
<h3>Data Split Strategy<a class="headerlink" href="#data-split-strategy" title="Link to this heading">#</a></h3>
<p><strong>Critical Design</strong>: Our dataset is smaller and simpler than the real research, but it teaches the same machine learning method and maintains scientific rigor by keeping training and testing data completely separate</p>
</section>
</section>
<section id="key-parameters">
<h2>Key Parameters<a class="headerlink" href="#key-parameters" title="Link to this heading">#</a></h2>
<section id="model-5-hyperparameters">
<h3>Model 5 Hyperparameters<a class="headerlink" href="#model-5-hyperparameters" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><strong>Tree depth</strong>: 6 (balances complexity vs. overfitting)</p></li>
<li><p><strong>L2 regularization</strong>: 3.0 (controls genomic noise)</p></li>
<li><p><strong>Learning rate</strong>: 0.03 (conservative for stability)</p></li>
<li><p><strong>Iterations</strong>: 1000 (sufficient for convergence)</p></li>
</ul>
<section id="feature-weighting-strategy">
<h4>Feature Weighting Strategy:<a class="headerlink" href="#feature-weighting-strategy" title="Link to this heading">#</a></h4>
<ul class="simple">
<li><p><strong>Standard features</strong>: Weight = 1.0 (distance, conservation, population genetics)</p></li>
<li><p><strong>High-priority regulatory features</strong>: Weight = 10.0</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">chrombpnet_positive</span></code>: Chromatin accessibility predictions</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tf_positive</span></code>: Transcription factor binding sites</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">diff_*</span></code>: Cell-type specific differential signals</p></li>
</ul>
</li>
</ul>
<p><strong>Biological Rationale</strong>: Experimentally validated regulatory features receive higher weight than purely computational predictions, reflecting greater confidence in biology-based annotations.</p>
</section>
</section>
<section id="cross-validation-configuration">
<h3>Cross-Validation Configuration<a class="headerlink" href="#cross-validation-configuration" title="Link to this heading">#</a></h3>
<section id="leave-one-chromosome-out-loco-procedure">
<h4>Leave-One-Chromosome-Out (LOCO) Procedure<a class="headerlink" href="#leave-one-chromosome-out-loco-procedure" title="Link to this heading">#</a></h4>
<p>The sc-EMS methodology uses chromosome-level cross-validation to prevent data leakage:</p>
<ul class="simple">
<li><p><strong>Training approach</strong>: For chromosome <em>i</em>, train on all other chromosomes <em>j</em> where <em>j ‚â† i</em></p></li>
<li><p><strong>Testing approach</strong>: Evaluate performance on held-out chromosome <em>i</em></p></li>
<li><p><strong>Implementation</strong>: 22 separate models trained (one per chromosome)</p></li>
</ul>
</section>
</section>
<section id="mathematical-framework">
<h3>Mathematical Framework<a class="headerlink" href="#mathematical-framework" title="Link to this heading">#</a></h3>
<p>For chromosome validation:</p>
</section>
</section>
<section id="detailed-workflow">
<h2>Detailed Workflow<a class="headerlink" href="#detailed-workflow" title="Link to this heading">#</a></h2>
<p><em>This section explains how to use our xQTL training pipeline</em></p>
<section id="step-1-running-the-pipeline">
<h3>Step 1: Running the Pipeline<a class="headerlink" href="#step-1-running-the-pipeline" title="Link to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>~/xqtl-protocol/code/xqtl_modifier_score/
python<span class="w"> </span>model_training_model5_only.py<span class="w"> </span>Mic_mega_eQTL<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--data_config<span class="w"> </span>data_config.yaml<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--model_config<span class="w"> </span>model_config.yaml
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="err">```</span><span class="n">python</span>
<span class="c1"># file check</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="n">required_files</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s1">&#39;model_training_simplified.py&#39;</span><span class="p">,</span>
    <span class="s1">&#39;data_config.yaml&#39;</span><span class="p">,</span>
    <span class="s1">&#39;model_config.yaml&#39;</span>
<span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Configuration File Check:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">required_files</span><span class="p">:</span>
    <span class="n">status</span> <span class="o">=</span> <span class="s2">&quot;‚úÖ Found&quot;</span> <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;‚ùå Missing&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-2-understanding-training-process-overview">
<h3>Step 2: Understanding Training Process Overview<a class="headerlink" href="#step-2-understanding-training-process-overview" title="Link to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Load leak-free training data (3,056 variants)</p></li>
<li><p>Train single feature-weighted CatBoost model</p></li>
<li><p>Evaluate on held-out test set (761 variants)</p></li>
<li><p>Generate feature importance rankings and predictions</p></li>
</ol>
</section>
<section id="step-3-results-results-analysis">
<h3>Step 3: Results &amp; Results Analysis<a class="headerlink" href="#step-3-results-results-analysis" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Performance results </span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Model Performance Results:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=&quot;</span> <span class="o">*</span> <span class="mi">50</span><span class="p">)</span>

<span class="c1"># Single model results (corrected from previous 4-model approach)</span>
<span class="n">corrected_results</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Feature-Weighted Model (Model 5)&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;AP&#39;</span><span class="p">:</span> <span class="mf">0.5050</span><span class="p">,</span> <span class="s1">&#39;AUC&#39;</span><span class="p">:</span> <span class="mf">0.8978</span><span class="p">}</span>
<span class="p">}</span>

<span class="k">for</span> <span class="n">model</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">corrected_results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Average Precision: </span><span class="si">{</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;AP&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  AUC Score: </span><span class="si">{</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;AUC&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Feature Importance (Top 10):&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="mi">40</span><span class="p">)</span>
<span class="n">top_features</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">(</span><span class="s2">&quot;abs_distance_TSS_log&quot;</span><span class="p">,</span> <span class="mf">23.58</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;diff_32_ENCFF140MBA&quot;</span><span class="p">,</span> <span class="mf">1.83</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;diff_32_ENCFF457MJJ&quot;</span><span class="p">,</span> <span class="mf">1.12</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;gnomad_MAF&quot;</span><span class="p">,</span> <span class="mf">0.88</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;diff_32_ENCFF455KGB&quot;</span><span class="p">,</span> <span class="mf">0.84</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;diff_32_ENCFF579IKK&quot;</span><span class="p">,</span> <span class="mf">0.80</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;ABC_score_microglia&quot;</span><span class="p">,</span> <span class="mf">0.54</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;microglia_enhancer_promoter_union_atac&quot;</span><span class="p">,</span> <span class="mf">0.58</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;diff_32_ENCFF098QMC&quot;</span><span class="p">,</span> <span class="mf">0.66</span><span class="p">),</span>
    <span class="p">(</span><span class="s2">&quot;diff_32_ENCFF757EYT&quot;</span><span class="p">,</span> <span class="mf">0.66</span><span class="p">)</span>
<span class="p">]</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">feature</span><span class="p">,</span> <span class="n">importance</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">top_features</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">:</span><span class="s2">2d</span><span class="si">}</span><span class="s2">. </span><span class="si">{</span><span class="n">feature</span><span class="si">:</span><span class="s2">&lt;35</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">importance</span><span class="si">:</span><span class="s2">&gt;6.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="results-analysis">
<h3>Results Analysis<a class="headerlink" href="#results-analysis" title="Link to this heading">#</a></h3>
<section id="strong-performance-indicators-our-results">
<h4>üéØ <strong>Strong Performance Indicators (Our Results):</strong><a class="headerlink" href="#strong-performance-indicators-our-results" title="Link to this heading">#</a></h4>
<ul class="simple">
<li><p><strong>AUC: 0.8978</strong> - Excellent discrimination (89.78% &gt; target of 75%)</p></li>
<li><p><strong>Average Precision: 0.5050</strong> - Strong precision (50.5% &gt; target of 40%)</p></li>
<li><p><strong>Distance dominance</strong>: Clear biological pattern (23.58 importance)</p></li>
<li><p><strong>Feature diversity</strong>: Multiple regulatory signals contribute (0.5-1.8 importance)</p></li>
</ul>
</section>
<section id="what-this-means">
<h4>‚úÖ <strong>What This Means:</strong><a class="headerlink" href="#what-this-means" title="Link to this heading">#</a></h4>
<ul class="simple">
<li><p><strong>Realistic genomic performance</strong> - Not artificially perfect, indicates genuine learning</p></li>
<li><p><strong>Biology makes sense</strong> - Distance matters most, regulatory features add value</p></li>
<li><p><strong>Model works correctly</strong> - Can distinguish functional from non-functional variants</p></li>
</ul>
</section>
<section id="key-biological-findings">
<h4>üî¨ <strong>Key Biological Findings:</strong><a class="headerlink" href="#key-biological-findings" title="Link to this heading">#</a></h4>
<ul class="simple">
<li><p><strong>Distance rules everything</strong>: 20x more important than other features</p></li>
<li><p><strong>Cell-type features help</strong>: Microglia-specific signals provide additional information</p></li>
<li><p><strong>Feature weighting worked</strong>: 10x weighting successfully elevated regulatory signals</p></li>
<li><p><strong>Population genetics secondary</strong>: MAF contributes but less than regulatory data</p></li>
</ul>
</section>
</section>
<section id="step-4-generated-output-files">
<h3>Step 4: Generated Output Files<a class="headerlink" href="#step-4-generated-output-files" title="Link to this heading">#</a></h3>
<p><strong>Model File</strong>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">model_feature_weighted_chr_chr2_NPR_10.joblib</span></code> - Trained CatBoost classifier</p></li>
</ul>
<p><strong>Analysis Results</strong>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">summary_dict_catboost_1model_chr_chr2_NPR_10.pkl</span></code> - Performance metrics and validation statistics</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">features_importance_1model_chr_chr2_NPR_10.csv</span></code> - Complete feature importance rankings</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">predictions_1model_chr2.tsv</span></code> - Per-variant prediction probabilities</p></li>
</ul>
</section>
<section id="model-application">
<h3>Model Application<a class="headerlink" href="#model-application" title="Link to this heading">#</a></h3>
<p>The trained model predicts functional genetic variants with 89.78% discriminative accuracy, enabling:</p>
<ul class="simple">
<li><p>Identification of disease-relevant variants in cell-type specific contexts</p></li>
<li><p>Prioritization of variants for experimental validation</p></li>
<li><p>Understanding of regulatory mechanisms in neurodegenerative diseases</p></li>
</ul>
</section>
<section id="step-5-using-your-trained-models">
<h3>Step 5: Using Your Trained Models<a class="headerlink" href="#step-5-using-your-trained-models" title="Link to this heading">#</a></h3>
<p>Once training is complete, load the trained models for predictions. Please refer to <strong>EMS Predictions</strong> (<a class="reference external" href="https://statfungen.github.io/xqtl-protocol/code/xqtl_modifier_score/ems_prediction.html">https://statfungen.github.io/xqtl-protocol/code/xqtl_modifier_score/ems_prediction.html</a>) for detailed prediction workflows and variant scoring.</p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./code/xqtl_modifier_score"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../enrichment/sldsc_enrichment.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Stratified LD Score Regression</p>
      </div>
    </a>
    <a class="right-next"
       href="ems_prediction.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">EMS Prediction for xQTL Modifier Scores</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motivation">Motivation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-weighted-catboost-algorithm">Feature-Weighted CatBoost Algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#training-data-construction">Training Data Construction</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-data-feature-categories-from-manuscript-table-c">Input Data: Feature Categories from Manuscript Table C</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-categories">Feature Categories</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#distance-features">1. Distance Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-type-regulatory-features">2. Cell-Type Regulatory Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genetics-features">3. Population Genetics Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#conservation-features">4. Conservation Features</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#deep-learning-predictions">5. Deep Learning Predictions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output-data-generated"><strong>Output Data Generated</strong></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#what-we-re-trying-to-predict-y">What We‚Äôre Trying to Predict (Y)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description-of-toy-dataset">Description of Toy Dataset</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dataset-overview">Dataset Overview</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#data-split-strategy">Data Split Strategy</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#key-parameters">Key Parameters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-5-hyperparameters">Model 5 Hyperparameters</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-weighting-strategy">Feature Weighting Strategy:</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cross-validation-configuration">Cross-Validation Configuration</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#leave-one-chromosome-out-loco-procedure">Leave-One-Chromosome-Out (LOCO) Procedure</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mathematical-framework">Mathematical Framework</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#detailed-workflow">Detailed Workflow</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-running-the-pipeline">Step 1: Running the Pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-understanding-training-process-overview">Step 2: Understanding Training Process Overview</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-results-results-analysis">Step 3: Results &amp; Results Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#results-analysis">Results Analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#strong-performance-indicators-our-results">üéØ <strong>Strong Performance Indicators (Our Results):</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#what-this-means">‚úÖ <strong>What This Means:</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#key-biological-findings">üî¨ <strong>Key Biological Findings:</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-generated-output-files">Step 4: Generated Output Files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-application">Model Application</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-using-your-trained-models">Step 5: Using Your Trained Models</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      ¬© Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>