
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Fine-mapping with PolyFun &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Overview of design" href="../../misc/summary_stats_standardizer.html" />
    <link rel="prev" title="Stratified LD Score Regression" href="../enrichment/ldsc.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   ADSP-FG xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../functional_annotation.html">
   Functional annotation integration
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="../enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Overview of design
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/integrative_analysis/SuSiE_Ann/polyfun.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/integrative_analysis/SuSiE_Ann/polyfun.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/integrative_analysis/SuSiE_Ann/polyfun.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#aim">
   Aim
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods-overview">
   Methods Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#workflow">
   Workflow
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-1-obtain-functional-annotations">
     Step 1: Obtain functional annotations
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-2-compute-ld-scores-for-annotations">
     Step 2: Compute LD-scores for annotations
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-1-compute-with-reference-panel-of-sequenced-individuals">
       Method 1: Compute with reference panel of sequenced individuals
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-2-compute-with-pre-computed-uk-biobank-ld-matrices">
       Method 2: Compute with pre-computed UK Biobank LD matrices
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-3-compute-with-own-pre-computed-ld-matrices">
       Method 3: Compute with own pre-computed LD matrices
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-3-compute-prior-causal-probabilities">
     Step 3: Compute Prior Causal Probabilities
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-1-use-precomputed-prior-causal-probabilities">
       Method 1: Use precomputed prior causal probabilities
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred">
       Method 2: Compute via L2-regularized extension of S-LDSC (preferred)
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-3-compute-non-parametrically">
       Method 3: Compute Non-parametrically
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-4-functionally-informed-fine-mapping-with-finemapper">
     Step 4: Functionally informed fine mapping with finemapper
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal Working Example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example-1-functionally-informed-fine-mapping-using-summary-statistics-file-with-precomputed-prior-causal-probabilities">
     Example 1: Functionally-informed fine-mapping using summary statistics file with precomputed prior causal probabilities
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example-2-functionally-informed-fine-mapping-using-summary-statistics-file-generated-from-pre-obtained-annotation-and-ld-score-files">
     Example 2: Functionally-informed fine-mapping using summary statistics file generated from pre-obtained annotation and LD-score files
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#summary">
     Summary
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#example-1">
       Example 1:
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#example-2">
       Example 2:
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="fine-mapping-with-polyfun">
<h1>Fine-mapping with PolyFun<a class="headerlink" href="#fine-mapping-with-polyfun" title="Permalink to this headline">¶</a></h1>
<div class="section" id="aim">
<h2>Aim<a class="headerlink" href="#aim" title="Permalink to this headline">¶</a></h2>
<p>The purpose of this notebook ipmlements commands for <a class="reference external" href="https://github.com/omerwe/polyfun/wiki">a functionally-informed fine-mapping workflow using the PolyFun method</a>.</p>
</div>
<div class="section" id="methods-overview">
<h2>Methods Overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">PolyFun</span></code> offers the following features:</p>
<ol class="simple">
<li><p>Using and/or creating Functional Annotations</p></li>
<li><p>Estimating Functional Enrichment Using <code class="docutils literal notranslate"><span class="pre">S-LDSC</span></code></p></li>
<li><p>Using and/or computating Prior Causal Probabilities from 1</p></li>
<li><p>Functionally Informed Fine Mapping with Finemapper</p></li>
<li><p>Polygenic Localization with <code class="docutils literal notranslate"><span class="pre">PolyLoc</span></code></p></li>
</ol>
<p><strong>Notice: this workflow does not implements 5 the <code class="docutils literal notranslate"><span class="pre">PolyLoc</span></code>.</strong></p>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol>
<li><p>GWAS summary statistics including the following variables:</p>
<ul class="simple">
<li><p>variant_id - variant ID</p></li>
<li><p>P - p-value</p></li>
<li><p>CHR - chromosome number</p></li>
<li><p>BP - base pair position</p></li>
<li><p>A1 - The effect allele (i.e., the sign of the effect size is with respect to A1)</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>MAF - minor allele frequency</p></li>
<li><p>BETA - effect size</p></li>
<li><p>SE - effect size standard error</p></li>
</ul>
</li>
<li><p>Functional annotation files including the following columns:</p>
<ul class="simple">
<li><p>CHR - chromosome number</p></li>
<li><p>BP base pair position (in hg19 coordinates)</p></li>
<li><p>SNP - dbSNP reference number</p></li>
<li><p>A1 - The effect allele</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>Arbitrary additional columns representing annotations</p></li>
</ul>
</li>
<li><p>A <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> white-space delimited file containing a single line with the sums of the columns of each annotation</p></li>
<li><p>LD-score files</p>
<ul class="simple">
<li><p>Strongly recommended that LD-score files include A1,A2 columns</p></li>
</ul>
</li>
<li><p>LD information, taken from one of three possible data sources:</p>
<ul class="simple">
<li><p>plink files with genotypes from a reference panel</p></li>
<li><p>bgen file with genotypes from a reference panel</p></li>
<li><p>pre-computed LD matrix</p></li>
</ul>
<p>Optional if (4) is obtained and no plans to compute prior causal probabilities non-parametrically</p>
</li>
<li><p>Ld-score weights files.</p>
<ul class="simple">
<li><p>Strongly recommended that weight files include A1,A2 columns</p></li>
</ul>
</li>
</ol>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>A <code class="docutils literal notranslate"><span class="pre">.gz</span></code> file containing input summary statistics columns and additionally the following columns:</p>
<ul class="simple">
<li><p>PIP - posterior causal probability</p></li>
<li><p>BETA_MEAN - posterior mean of causal effect size (in standardized genotype scale)</p></li>
<li><p>BETA_SD - posterior standard deviation of causal effect size (in standardized genotype scale)</p></li>
<li><p>CREDIBLE_SET - the index of the first (typically smallest) credible set that the SNP belongs to (0 means none).</p></li>
</ul>
</div>
<div class="section" id="workflow">
<h2>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h2>
<p>Step 1 and 2 are optional if using pre-computed prior causal probabilities</p>
<div class="section" id="step-1-obtain-functional-annotations">
<h3>Step 1: Obtain functional annotations<a class="headerlink" href="#step-1-obtain-functional-annotations" title="Permalink to this headline">¶</a></h3>
<p>For each chromosome, the following files need to be obtained:</p>
<ol class="simple">
<li><p>A <code class="docutils literal notranslate"><span class="pre">.gz</span></code> or <code class="docutils literal notranslate"><span class="pre">.parquet</span></code> annotations file containing the following columns:</p></li>
</ol>
<ul class="simple">
<li><p>CHR - chromosome number</p></li>
<li><p>BP base pair position</p></li>
<li><p>SNP - dbSNP reference number</p></li>
<li><p>A1 - The effect allele</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>Arbitrary additional columns representing annotations</p></li>
</ul>
<ol class="simple">
<li><p>A <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> white-space delimited file containing a single line with the sums of the columns of each annotation</p></li>
<li><p>(Optional) A <code class="docutils literal notranslate"><span class="pre">l2.M_5_50</span></code> file that is the <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> file but only containing common SNPS (MAF between 5% and 50%)</p></li>
</ol>
<p>The above files can be obtained either by using existing function annotation files, or by creating your own through other software such as <code class="docutils literal notranslate"><span class="pre">TORUS</span></code>.</p>
<p>Existing function annotation files example: functional annotations for ~19 million UK Biobank imputed SNPs with MAF&gt;0.1%, based on the baseline-LF 2.2.UKB annotations.</p>
<p>Download (30G): <a class="reference external" href="https://data.broadinstitute.org/alkesgroup/LDSCORE/baselineLF_v2.2.UKB.polyfun.tar.gz">https://data.broadinstitute.org/alkesgroup/LDSCORE/baselineLF_v2.2.UKB.polyfun.tar.gz</a></p>
</div>
<div class="section" id="step-2-compute-ld-scores-for-annotations">
<h3>Step 2: Compute LD-scores for annotations<a class="headerlink" href="#step-2-compute-ld-scores-for-annotations" title="Permalink to this headline">¶</a></h3>
<p>Precomputed LD-score files can be used. LD-score files can also be generated through the methods below:</p>
<div class="section" id="method-1-compute-with-reference-panel-of-sequenced-individuals">
<h4>Method 1: Compute with reference panel of sequenced individuals<a class="headerlink" href="#method-1-compute-with-reference-panel-of-sequenced-individuals" title="Permalink to this headline">¶</a></h4>
<p>Reference panel should have at least 3000 sequenced individuals from target population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;/mnt/mfs/statgen/containers/xqtl_pipeline_sif/polyfun.sif&quot;</span>
<span class="kn">parameter:</span> <span class="n">wd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">exe_dir</span> <span class="o">=</span> <span class="s2">&quot;/usr/local/bin/&quot;</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;demo&quot;</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">annot_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">sumstats</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ld_score</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">annot_file</span><span class="p">,</span> <span class="n">genoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.ref.ldscore.parquet&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">compute_ldscores</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">annot</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="method-2-compute-with-pre-computed-uk-biobank-ld-matrices">
<h4>Method 2: Compute with pre-computed UK Biobank LD matrices<a class="headerlink" href="#method-2-compute-with-pre-computed-uk-biobank-ld-matrices" title="Permalink to this headline">¶</a></h4>
<p>Matrices download: <a class="reference external" href="https://data.broadinstitute.org/alkesgroup/UKBB_LD">https://data.broadinstitute.org/alkesgroup/UKBB_LD</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ld_score_ukb</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">annot_file</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.ukb.ldscore.parquet&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">compute_ldscores_from_ld</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">annot</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">ukb</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="method-3-compute-with-own-pre-computed-ld-matrices">
<h4>Method 3: Compute with own pre-computed LD matrices<a class="headerlink" href="#method-3-compute-with-own-pre-computed-ld-matrices" title="Permalink to this headline">¶</a></h4>
<p>Own pre-computed LD matrices should be in <code class="docutils literal notranslate"><span class="pre">.bcor</span></code> format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ld_score_own</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sample_size</span> <span class="o">=</span> <span class="nb">int</span>
<span class="kn">parameter:</span> <span class="n">bcor_files</span> <span class="o">=</span> <span class="n">paths</span>
<span class="kn">input:</span> <span class="n">annot_file</span><span class="p">,</span><span class="n">bcor_files</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.original.ldscore.parquet&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">compute_ldscores_from_ld</span><span class="o">.</span><span class="n">py</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">annot</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">n</span> <span class="err">$</span><span class="p">[</span><span class="n">sample_size</span><span class="p">]</span> \
        
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-3-compute-prior-causal-probabilities">
<h3>Step 3: Compute Prior Causal Probabilities<a class="headerlink" href="#step-3-compute-prior-causal-probabilities" title="Permalink to this headline">¶</a></h3>
<div class="section" id="method-1-use-precomputed-prior-causal-probabilities">
<h4>Method 1: Use precomputed prior causal probabilities<a class="headerlink" href="#method-1-use-precomputed-prior-causal-probabilities" title="Permalink to this headline">¶</a></h4>
<p>Use precomputed prior causal probabilities of 19 million imputed UK Biobank SNPs with MAF&gt;0.1%, based on a meta-analysis of 15 UK Biobank traits.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">prior_causal_prob</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">sumstats</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.pcp.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">extract_snpvar</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">sumstats</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">missing</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred">
<h4>Method 2: Compute via L2-regularized extension of S-LDSC (preferred)<a class="headerlink" href="#method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred" title="Permalink to this headline">¶</a></h4>
<p>Compute via an L2-regularized extension of stratified LD-score regression (S-LDSC). Use the annotation and LD-score files produced in Step1.</p>
<ol class="simple">
<li><p>Create a munged summary statistics file in a PolyFun-friendly parquet format.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">munged_sumstats</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sample_size</span> <span class="o">=</span> <span class="mi">472868</span> 
<span class="kn">parameter:</span> <span class="n">min_info</span> <span class="o">=</span> <span class="mf">0.6</span>
<span class="kn">parameter:</span> <span class="n">min_maf</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="kn">input:</span> <span class="n">sumstats</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.sumstats_munged.parquet&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;127G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">munge_polyfun_sumstats</span><span class="o">.</span><span class="n">py</span> \
      <span class="o">--</span><span class="n">sumstats</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">]</span> \
      <span class="o">--</span><span class="n">n</span> <span class="err">$</span><span class="p">[</span><span class="n">sample_size</span><span class="p">]</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> \
      <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">info</span> <span class="err">$</span><span class="p">[</span><span class="n">min_info</span><span class="p">]</span> \
      <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">maf</span> <span class="err">$</span><span class="p">[</span><span class="n">min_maf</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<ol class="simple">
<li><p>Run PolyFun with L2-regularized S-LDSC</p></li>
</ol>
<ul class="simple">
<li><p>Require at least 45 GB of mem</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">L2_SLDSC</span><span class="p">]</span>
<span class="c1"># a ld score file with surfix l2.ldscore.parquet</span>
<span class="kn">parameter:</span> <span class="n">ref_ld</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># another ld score file with surfix l2.ldscore.parquet, different from ref_ld</span>
<span class="kn">parameter:</span> <span class="n">ref_wgt</span> <span class="o">=</span> <span class="n">ref_ld</span>
<span class="kn">parameter:</span> <span class="n">partitions</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">input:</span> <span class="n">ref_ld</span><span class="p">,</span> <span class="n">ref_wgt</span><span class="p">,</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;munged_sumstats&quot;</span><span class="p">)</span>
<span class="c1"># parameter: sumstat = _input[2]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.ldsldsc.parquent&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;127G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">h2</span><span class="o">-</span><span class="n">L2</span> \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">sumstats</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">]</span><span class="o">.</span>\
        <span class="o">--</span><span class="n">w</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">nnnn</span><span class="p">]</span><span class="o">.</span> \
        <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">missing</span> <span class="err">$</span><span class="p">[</span><span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="n">partitions</span> <span class="k">else</span> <span class="s2">&quot;--no-partitions&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="method-3-compute-non-parametrically">
<h4>Method 3: Compute Non-parametrically<a class="headerlink" href="#method-3-compute-non-parametrically" title="Permalink to this headline">¶</a></h4>
<ol class="simple">
<li><p>Create a munged summary statistics file in a PolyFun-friendly parquet format.
Duplicated cells are commented out, the input of [ld_snpbin] is the output from [L2_regu_SLDSC]</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#[munged_sumstats2]</span>
<span class="c1">#parameter: sumstats = AD_sumstats_Jansenetal_2019sept.txt.gz</span>
<span class="c1">#parameter: sample_size = int</span>
<span class="c1">#parameter: container = none</span>
<span class="c1">#bash: container = container </span>
<span class="c1">#    mkdir -p SLDSC_output</span>
<span class="c1">#    python munge_polyfun_sumstats.py \</span>
<span class="c1">#      --sumstats sumstats \</span>
<span class="c1">#      --n sample_size \</span>
<span class="c1">#      --out /SLDSC_output/sumstats_munged.parquet \</span>
<span class="c1">#      --min-info 0 \</span>
<span class="c1">#      --min-maf 0</span>
</pre></div>
</div>
</div>
</div>
<ol class="simple">
<li><p>Run PolyFun with L2-regularized S-LDSC</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># [L2_regu_SLDSC2]</span>
<span class="c1"># </span>
<span class="c1"># parameter: container = none</span>
<span class="c1"># paramter: ref_ld = example_data/annotations.</span>
<span class="c1"># parameter: ref_wgt = example_data/weights.</span>
<span class="c1"># bash: container=container</span>
<span class="c1">#     python polyfun.py \</span>
<span class="c1">#     --compute-h2-L2 \</span>
<span class="c1">#     --output-prefix output/testrun \</span>
<span class="c1">#     --sumstats example_data/sumstats.parquet \</span>
<span class="c1">#     --ref-ld-chr ref_ld \</span>
<span class="c1">#     --w-ld-chr ref_wgt</span>
</pre></div>
</div>
</div>
</div>
<ol class="simple">
<li><p>Compute LD-scores for each SNP bin</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ld_snpbin</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">sos_step</span><span class="p">(</span><span class="s2">&quot;L2_regu_SLDSC&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">chrom</span> <span class="o">=</span> <span class="nb">int</span>
<span class="kn">input:</span> <span class="n">annot_file</span><span class="p">,</span> <span class="n">genoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.snpbin.ldscore.parquet&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
     <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">ldscores</span> \
        <span class="o">--</span><span class="n">bfile</span><span class="o">-</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> \
        <span class="o">--</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<ol class="simple">
<li><p>Re-estimate per-SNP heritabilities via S-LDSC</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#[re_SLDSC]</span>
<span class="c1">#bash:</span>
<span class="c1">#    python polyfun.py \</span>
<span class="c1">#    --compute-h2-bins \</span>
<span class="c1">#    --output-prefix output/testrun \</span>
<span class="c1">#    --sumstats example_data/sumstats.parquet \</span>
<span class="c1">#    --w-ld-chr example_data/weights.</span>

<span class="p">[</span><span class="n">L2_SLDSC_bins</span><span class="p">]</span>
<span class="kn">paramter:</span> <span class="n">ref_ld</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">ref_wgt</span> <span class="o">=</span> <span class="n">ref_ld</span>
<span class="kn">parameter:</span> <span class="n">partitions</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">input:</span> <span class="n">ref_ld</span><span class="p">,</span> <span class="n">ref_wgt</span><span class="p">,</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;munged_sumstats&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.txt.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">h2</span><span class="o">-</span><span class="n">bins</span> \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">sumstats</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>\
        <span class="o">--</span><span class="n">w</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> 
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-4-functionally-informed-fine-mapping-with-finemapper">
<h3>Step 4: Functionally informed fine mapping with finemapper<a class="headerlink" href="#step-4-functionally-informed-fine-mapping-with-finemapper" title="Permalink to this headline">¶</a></h3>
<p>Input summary statistics file must have <code class="docutils literal notranslate"><span class="pre">SNPVAR</span></code> column (per-SNP heritability) to perform functionally-informed fine-mapping. To fine-map without annotations, use additional parameter <code class="docutils literal notranslate"><span class="pre">--non-funct</span></code>. The summary statistical file then will not require the <code class="docutils literal notranslate"><span class="pre">SNPVAR</span></code> column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">fine_mapping</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sample_size</span> <span class="o">=</span> <span class="mi">383290</span>
<span class="kn">parameter:</span> <span class="n">chrom</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="n">start</span> <span class="o">=</span> <span class="mi">46000001</span>
<span class="kn">parameter:</span> <span class="n">end</span> <span class="o">=</span> <span class="mi">49000001</span>
<span class="c1">#parameter: output_path = &quot;output/finemap.1.46000001.49000001.gz&quot;</span>
<span class="kn">parameter:</span> <span class="n">max_num_causal</span> <span class="o">=</span> <span class="mi">5</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span><span class="n">sumstats</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/output/finemap.</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s1">.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;30G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">python</span> <span class="err">$</span><span class="p">[</span><span class="n">exe_dir</span><span class="p">]</span><span class="o">/</span><span class="n">finemapper</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">geno</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">sumstats</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>  \
        <span class="o">--</span><span class="n">n</span> <span class="err">$</span><span class="p">[</span><span class="n">sample_size</span><span class="p">]</span> \
        <span class="o">--</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">start</span> <span class="err">$</span><span class="p">[</span><span class="n">start</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">end</span> <span class="err">$</span><span class="p">[</span><span class="n">end</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">method</span> <span class="n">susie</span> \
        <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">num</span><span class="o">-</span><span class="n">causal</span> <span class="err">$</span><span class="p">[</span><span class="n">max_num_causal</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">cache</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">d</span><span class="p">]</span><span class="o">/</span><span class="n">cache</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal Working Example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<div class="section" id="example-1-functionally-informed-fine-mapping-using-summary-statistics-file-with-precomputed-prior-causal-probabilities">
<h3>Example 1: Functionally-informed fine-mapping using summary statistics file with precomputed prior causal probabilities<a class="headerlink" href="#example-1-functionally-informed-fine-mapping-using-summary-statistics-file-with-precomputed-prior-causal-probabilities" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">prior_causal_prob</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">at3535</span><span class="o">/</span><span class="n">polyfun</span><span class="o">/</span><span class="n">AD_sumstats_Jansenetal_2019sept</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> \
    <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fine_mapping</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">at3535</span><span class="o">/</span><span class="n">polyfun</span><span class="o">/</span><span class="n">AD_sumstats_Jansenetal_2019sept</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">ROSMAP_xqtl</span><span class="o">/</span><span class="n">dataset</span><span class="o">/</span><span class="n">snvCombinedPlink</span><span class="o">/</span><span class="n">chr1</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> \
    <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="example-2-functionally-informed-fine-mapping-using-summary-statistics-file-generated-from-pre-obtained-annotation-and-ld-score-files">
<h3>Example 2: Functionally-informed fine-mapping using summary statistics file generated from pre-obtained annotation and LD-score files<a class="headerlink" href="#example-2-functionally-informed-fine-mapping-using-summary-statistics-file-generated-from-pre-obtained-annotation-and-ld-score-files" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">munged_sumstats</span>  \
<span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">at3535</span><span class="o">/</span><span class="n">polyfun</span><span class="o">/</span><span class="n">GCST90012877_buildGRCh37_colrenamed</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> \
<span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">L2_SLDSC</span>  \
<span class="o">--</span><span class="n">sumstats</span> <span class="n">demo</span><span class="o">.</span><span class="n">sumstats_munged</span><span class="o">.</span><span class="n">parquet</span> \
<span class="o">--</span><span class="n">ref_ld</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">baselineLF2</span><span class="o">.</span><span class="mf">2.</span><span class="n">UKB</span><span class="o">/</span><span class="n">baselineLF2</span><span class="o">.</span><span class="mf">2.</span><span class="n">UKB</span><span class="o">.</span><span class="mf">1.</span><span class="n">l2</span><span class="o">.</span><span class="n">ldscore</span><span class="o">.</span><span class="n">parquet</span> \
<span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fine_mapping</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="n">demo</span><span class="o">.</span><span class="n">ldsldsc</span><span class="o">.</span><span class="n">parquent</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">ROSMAP_xqtl</span><span class="o">/</span><span class="n">dataset</span><span class="o">/</span><span class="n">snvCombinedPlink</span><span class="o">/</span><span class="n">chr1</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="summary">
<h3>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h3>
<div class="section" id="example-1">
<h4>Example 1:<a class="headerlink" href="#example-1" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">bash:</span>
    <span class="n">gzcat</span> <span class="n">output</span><span class="o">/</span><span class="n">finemap</span><span class="o">.</span><span class="mf">1.46000001</span><span class="o">.</span><span class="mf">49000001.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CHR	SNP	BP	A1	A2	SNPVAR	Z	N	P	PIP	BETA_MEAN	BETA_SD	DISTANCE_FROM_CENTER	CREDIBLE_SET
1	rs2088102	46032974	T	C	1.70060e-06	1.25500e+01	383290	3.97510e-36	1.00000e+00	-2.03917e-02	1.61901e-03	1456799	1
1	rs7528714	47966058	G	A	1.18040e-06	5.14320e+00	383290	2.70098e-07	9.97870e-01	7.42146e-03	1.62305e-03	476285	2
1	rs7528075	47870271	G	A	1.18040e-06	4.40160e+00	383290	1.07456e-05	9.76545e-01	-5.98945e-03	1.81667e-03	380498	3
1	rs212968	48734666	G	A	1.70060e-06	-3.01130e+00	383290	2.60132e-03	3.75823e-01	-1.56305e-03	2.23942e-03	1244893	0
1	rs2622911	47837404	C	A	1.70060e-06	3.12520e+00	383290	1.77684e-03	3.71804e-01	1.54312e-03	2.22812e-03	347631	0
1	rs4511165	48293181	G	A	1.70060e-06	-1.18940e+00	383290	2.34282e-01	5.75970e-02	1.60630e-04	7.52226e-04	803408	0
1	rs3766196	47284526	C	A	6.93040e-06	-5.92360e-02	383290	9.52764e-01	4.89776e-02	-5.06776e-07	3.48039e-04	205247	0
1	rs12567716	48197570	T	C	1.18040e-06	2.14810e+00	383290	3.17058e-02	4.45128e-02	-1.28281e-04	6.81457e-04	707797	0
1	rs4927234	48384796	G	A	1.70060e-06	-7.59600e-01	383290	4.47494e-01	3.41123e-02	7.74978e-05	5.04755e-04	895023	0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;output/finemap.1.46000001.49000001.gz&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    
<span class="n">num_var_cs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">])</span>
<span class="n">total_cs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">CREDIBLE_SET</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span><span class="o">-</span> <span class="mi">1</span>
<span class="n">avg_var_cs</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">/</span> <span class="n">total_cs</span>
<span class="n">pip50</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">pip95</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.95</span><span class="p">)</span>

<span class="n">result</span> <span class="o">=</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.5: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip50</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.95: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip95</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of variants that have credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of unique credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">total_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Average number of variants per credible set: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">avg_var_cs</span><span class="p">)</span> 


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">the_file</span><span class="p">:</span>
    <span class="n">the_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">contents</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">contents</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;Number of variants with PIP &gt; 0.5: 3\n&#39;, &#39;Number of variants with PIP &gt; 0.95: 3\n&#39;, &#39;Number of variants that have credible sets: 3\n&#39;, &#39;Number of unique credible sets: 3\n&#39;, &#39;Average number of variants per credible set: 1.0&#39;]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="example-2">
<h4>Example 2:<a class="headerlink" href="#example-2" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">bash:</span>
    <span class="n">gzcat</span> <span class="n">output</span><span class="o">/</span><span class="n">finemap</span><span class="o">.</span><span class="mf">1.460000010</span><span class="o">.</span><span class="mf">49000000.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;output/finemap.1.46000000.49000000.gz&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    
<span class="n">num_var_cs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">])</span>
<span class="n">total_cs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">CREDIBLE_SET</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span><span class="o">-</span> <span class="mi">1</span>
<span class="n">avg_var_cs</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">/</span> <span class="n">total_cs</span>
<span class="n">pip50</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">pip95</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.95</span><span class="p">)</span>

<span class="n">result</span> <span class="o">=</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.5: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip50</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.95: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip95</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of variants that have credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of unique credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">total_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Average number of variants per credible set: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">avg_var_cs</span><span class="p">)</span> 


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">the_file</span><span class="p">:</span>
    <span class="n">the_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">contents</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">contents</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;Number of variants with PIP &gt; 0.5: 3\n&#39;, &#39;Number of variants with PIP &gt; 0.95: 3\n&#39;, &#39;Number of variants that have credible sets: 3\n&#39;, &#39;Number of unique credible sets: 3\n&#39;, &#39;Average number of variants per credible set: 1.0&#39;]
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/integrative_analysis/SuSiE_Ann"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../enrichment/ldsc.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Stratified LD Score Regression</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../../misc/summary_stats_standardizer.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Overview of design</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>