{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5839a8ae-e13c-44ba-b06c-8346a0e65655",
   "metadata": {},
   "source": [
    "# Multivariate Fine-Mapping with mvSuSiE and mr.mash\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b860370-77bb-4095-b9b1-1613ed681bf4",
   "metadata": {},
   "source": [
    "Multivariate fine-mapping using mvSuSiE and mr.mash is also available in our pipeline.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a10f540-933d-497c-908f-f868c965007f",
   "metadata": {},
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f0bcf9f-55c4-43fa-b846-170c935403b9",
   "metadata": {},
   "source": [
    "\n",
    "    \n",
    "`--genoFile`: path to a text file contatining information on genotype files. For example:\n",
    "```\n",
    "#id     #path\n",
    "21      $PATH/protocol_example.genotype.chr21_22.21.bed\n",
    "22      $PATH/protocol_example.genotype.chr21_22.22.bed\n",
    "```\n",
    "`--phenoFile`: a tab delimited file containing chr, start, end, ID and path for the regions. For example:\n",
    "```\n",
    "#chr    start   end     ID      path\n",
    "chr21   0       14120807        TADB_1297       $PATH/protocol_example.ha.bed.gz\n",
    "chr21   10840000        16880069        TADB_1298       $PATH/protocol_example.ha.bed.gz\n",
    "```\n",
    "\n",
    "`--covFile`: path to a gzipped file containing covariates in the rows, and sample ids in the columns.  \n",
    "`--customized-association-windows`: a tab delimited file containing chr, start, end, and ID regions. For example:\n",
    "```\n",
    "#chr    start   end     ID\n",
    "chr21   0       14120807        TADB_1297\n",
    "chr21   10840000        16880069        TADB_1298\n",
    "```\n",
    "`--region-name`: if you only wish to analyze one region, then include the ID of a region found in the `customized-association-windows` file\n",
    "\n",
    "`--mixture_prior`: rds file from mr.mash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58b21adc-571d-483c-b033-0fb2662be3f5",
   "metadata": {},
   "source": [
    "## Output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25e3b8e3-4434-487c-b8ad-eb34905540dc",
   "metadata": {},
   "source": [
    "* `*.multicontext_data.rds`\n",
    "```\n",
    "> str(readRDS(\"/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_protocol/output/mnm_old/data/test_mnm.chr11_ENSG00000073921.multicontext_data.rds\"))\n",
    "List of 8\n",
    " $ residual_Y       : num [1:150, 1:6] 0.1241 0.2981 -0.6054 0.0435 0.0388 ...\n",
    "  ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. ..$ : chr [1:150] \"sample0\" \"sample1\" \"sample10\" \"sample100\" ...\n",
    "  .. ..$ : chr [1:6] \"Ast\" \"Exc\" \"Inh\" \"Mic\" ...\n",
    " $ residual_Y_scalar: num [1:6] 1 1 1 1 1 1\n",
    " $ dropped_sample   :List of 3\n",
    "  ..$ X    :List of 6\n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  ..$ Y    :List of 6\n",
    "  .. ..$ : chr \"strand\"\n",
    "  .. ..$ : chr \"strand\"\n",
    "  .. ..$ : chr \"strand\"\n",
    "  .. ..$ : chr \"strand\"\n",
    "  .. ..$ : chr \"strand\"\n",
    "  .. ..$ : chr \"strand\"\n",
    "  ..$ covar:List of 6\n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    "  .. ..$ : chr(0) \n",
    " $ X                : num [1:150, 1:9970] 0 0 1 1 1 0 0 1 1 1 ...\n",
    "  ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. ..$ : chr [1:150] \"sample0\" \"sample1\" \"sample10\" \"sample100\" ...\n",
    "  .. ..$ : chr [1:9970] \"chr11:84957209:G:C\" \"chr11:84957879:A:G\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" ...\n",
    " $ maf              : Named num [1:9970] 0.33 0.327 0.103 0.343 0.343 ...\n",
    "  ..- attr(*, \"names\")= chr [1:9970] \"chr11:84957209:G:C\" \"chr11:84957879:A:G\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" ...\n",
    " $ chrom            : chr \"chr11\"\n",
    " $ grange           : chr [1:2] \"85957174\" \"86069881\"\n",
    " $ X_variance       : Named num [1:9970] 0.441 0.443 0.192 0.458 0.458 ...\n",
    "  ..- attr(*, \"names\")= chr [1:9970] \"chr11:84957209:G:C\" \"chr11:84957879:A:G\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" ...\n",
    "```\n",
    "\n",
    "* `*.multicontext_bvsr.rds`\n",
    "```\n",
    "> str(readRDS(\"/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_protocol/output/mnm_old/multivariate_fine_mapping/test_mnm.chr11_ENSG00000073921.multicontext_bvsr.rds\"),max.level=2)\n",
    "List of 1\n",
    " $ ENSG00000073921:List of 10\n",
    "  ..$ mrmash_fitted              :List of 14\n",
    "  .. ..- attr(*, \"class\")= chr [1:2] \"mr.mash\" \"list\"\n",
    "  ..$ reweighted_mixture_prior   :Classes 'MashInitializer', 'R6' <MashInitializer>\n",
    "  Public:\n",
    "    clone: function (deep = FALSE) \n",
    "    compute_prior_inv: function () \n",
    "    initialize: function (Ulist, grid, prior_weights = NULL, null_weight = 0, \n",
    "    n_component: active binding\n",
    "    n_condition: active binding\n",
    "    precompute_cov_matrices: function (d, algorithm = c(\"R\", \"cpp\")) \n",
    "    precomputed: active binding\n",
    "    prior_variance: active binding\n",
    "    remove_precomputed: function () \n",
    "    scale_prior_variance: function (sigma) \n",
    "  Private:\n",
    "    inv_mats: NULL\n",
    "    U: NULL\n",
    "    xU: list \n",
    "  ..$ reweighted_mixture_prior_cv:List of 5\n",
    "  ..$ mvsusie_fitted             :List of 28\n",
    "  .. ..- attr(*, \"class\")= chr \"mvsusie\"\n",
    "  ..$ variant_names              : chr [1:6454] \"chr11:84957209:G:C\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" \"chr11:84959602:T:C\" ...\n",
    "  ..$ analysis_script            : chr \"\\nlibrary(pecotmr)\\n\\nphenotype_files = c(\\\"/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_protocol/output/p\"| __truncated__\n",
    "  ..$ other_quantities           :List of 1\n",
    "  ..$ context_names              : chr [1:6] \"Ast\" \"Exc\" \"Inh\" \"Mic\" ...\n",
    "  ..$ total_time_elapsed         : 'proc_time' Named num [1:5] 104.38 0.653 105.828 0 0\n",
    "  .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  ..$ region_info                :List of 3\n",
    "```\n",
    "\n",
    "* `*.multicontext_twas_weights.rds`\n",
    "```\n",
    "> str(readRDS(\"/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_protocol/output/mnm_old/multivariate_twas_weights/test_mnm.chr11_ENSG00000073921.multicontext_twas_weights.rds\"), max.level=3)\n",
    "List of 1\n",
    " $ ENSG00000073921:List of 6\n",
    "  ..$ Ast_ENSG00000073921:List of 6\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. ..$ variant_names     : chr [1:6454] \"chr11:84957209:G:C\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" \"chr11:84959602:T:C\" ...\n",
    "  .. ..$ twas_cv_result    :List of 4\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 1645.11 3.4 1652.73 0.01 0.02\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  ..$ Exc_ENSG00000073921:List of 6\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. ..$ variant_names     : chr [1:6454] \"chr11:84957209:G:C\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" \"chr11:84959602:T:C\" ...\n",
    "  .. ..$ twas_cv_result    :List of 4\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 1645.11 3.4 1652.73 0.01 0.02\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  ..$ Inh_ENSG00000073921:List of 6\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. ..$ variant_names     : chr [1:6454] \"chr11:84957209:G:C\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" \"chr11:84959602:T:C\" ...\n",
    "  .. ..$ twas_cv_result    :List of 4\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 1645.11 3.4 1652.73 0.01 0.02\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  ..$ Mic_ENSG00000073921:List of 6\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. ..$ variant_names     : chr [1:6454] \"chr11:84957209:G:C\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" \"chr11:84959602:T:C\" ...\n",
    "  .. ..$ twas_cv_result    :List of 4\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 1645.11 3.4 1652.73 0.01 0.02\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  ..$ OPC_ENSG00000073921:List of 6\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. ..$ variant_names     : chr [1:6454] \"chr11:84957209:G:C\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" \"chr11:84959602:T:C\" ...\n",
    "  .. ..$ twas_cv_result    :List of 4\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 1645.11 3.4 1652.73 0.01 0.02\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  ..$ Oli_ENSG00000073921:List of 6\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. ..$ variant_names     : chr [1:6454] \"chr11:84957209:G:C\" \"chr11:84958236:C:A\" \"chr11:84958954:T:C\" \"chr11:84959602:T:C\" ...\n",
    "  .. ..$ twas_cv_result    :List of 4\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 1645.11 3.4 1652.73 0.01 0.02\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "397a4dc3-1e4c-4845-ad8b-beed40f1f408",
   "metadata": {},
   "source": [
    "## Minimal Working Example Steps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bca0962-ae57-422c-a79f-892ef3b7f7ae",
   "metadata": {},
   "source": [
    "### iv. [Run the Fine-Mapping with mvSuSiE](https://statfungen.github.io/xqtl-protocol/code/mnm_analysis/mnm_methods/mnm_regression.html#iii-mvsusie-with-twas-weights-from-mr-mash)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dff2c979-52a1-4ec8-a0be-51b289327d37",
   "metadata": {},
   "outputs": [],
   "source": [
    "sos run pipeline/mnm_regression.ipynb mnm \\\n",
    "    --name test_mnm --cwd output/mnm \\\n",
    "    --genoFile output/genotype_by_chrom/wgs.merged.plink_qc.genotype_by_chrom_files.txt \\\n",
    "    --phenoFile output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    --covFile output/covariate/bulk_rnaseq_tpm_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    output/covariate/bulk_rnaseq_tpm_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    output/covariate/bulk_rnaseq_tpm_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    output/covariate/bulk_rnaseq_tpm_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    output/covariate/bulk_rnaseq_tpm_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    output/covariate/bulk_rnaseq_tpm_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    --customized-association-windows reference_data/TAD/TADB_enhanced_cis.bed \\\n",
    "    --region-name ENSG00000073921 --save-data --no-skip-twas-weights \\\n",
    "    --phenotype-names Ast Exc Inh Mic OPC Oli \\\n",
    "    --mixture_prior output/multivariate_mixture/MWE_ed_bovy.EE.prior.rds \\\n",
    "    --max_cv_variants 5000 \\\n",
    "\t--ld_reference_meta_file data/ld_meta_file_with_bim.tsv "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc7691e6-954c-4dc1-bac3-25d56026ab96",
   "metadata": {},
   "source": [
    "## Anticipated Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5418a39f-f2a6-48ac-8440-38d692cc3c38",
   "metadata": {},
   "source": [
    "For each gene, multivariate finemapping will produce a file containing results for the top hits and a file containing twas weights produced by susie."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "199a2637-cc0c-4aff-b4ea-a2546936f073",
   "metadata": {},
   "source": [
    "`test_mnm.chr11_ENSG00000073921.multicontext_bvsr.rds`:\n",
    "* For each gene of interest, this file contains:\n",
    "    1. mrmash_fitted\n",
    "    2. reweighted_mixture_prior\n",
    "    3. reweighted_mixture_prior_cv\n",
    "    4. mvsusie_fitted\n",
    "    5. variant_names\n",
    "    6. analysis_script\n",
    "    7. other_quantities\n",
    "    8. context_names\n",
    "    9. top_loci\n",
    "    10. susie_result_trimmed\n",
    "    11. total_time_elapsed\n",
    "    12. region_info\n",
    "\n",
    "`test_mnm.chr11_ENSG00000073921.multicontext_data.rds`: (from the --save-data argument)\n",
    "* see [pecotmr code](https://github.com/statfungen/pecotmr/blob/68d87ca1d0a059022bf4e55339621cbddc8993cc/R/file_utils.R#L461) for description \n",
    "\n",
    "`test_mnm.chr11_ENSG00000073921.multicontext_twas_weights.rds`:\n",
    "* For each gene of interest and phenotype, this file contains:\n",
    "    1. twas_weights - weights mrmash and mvsusie methods\n",
    "    2. twas_predictions - twas predictions for mrmash and mvsusie methods\n",
    "    3. variant_names\n",
    "    5. twas_cv_result\n",
    "    6. total_time_elapsed\n",
    "    8. region_info\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "159e580c-e290-4403-a701-d9dfa1043341",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
