{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5839a8ae-e13c-44ba-b06c-8346a0e65655",
   "metadata": {},
   "source": [
    "# Multivariate Fine-Mapping for multiple genes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5161522b-01c0-42f3-b527-5e4e376648c0",
   "metadata": {},
   "source": [
    "## Description"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b860370-77bb-4095-b9b1-1613ed681bf4",
   "metadata": {},
   "source": [
    "Multi gene fine-mapping and TWAS may also be conducted with our pipeline. This considers multiple genes jointly within specific TAD windows.\n",
    "\n",
    "This step is similar to the multivariate fine-mapping with two main differences. 1) TAD windows with multiple genes need to be defined. The `--pheno_id_map_file` parameter is used for this. 2) To speed things up, the genes are filtered out if they don't have a univariate fine mapped region. Genes may also be filtered out if they do have a univariate fine-mapped signal, but the signal is nowhere close to that of other genes.  The `--skip-analysis-pip-cutoff` parameter is used for this.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8201581-f9fd-4c8c-8540-8daf0a2dcf76",
   "metadata": {},
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f0bcf9f-55c4-43fa-b846-170c935403b9",
   "metadata": {},
   "source": [
    "\n",
    "    \n",
    "`--genoFile`: path to a plink bed file containin genotypes. Include the `.bed`\n",
    "\n",
    "`--phenoFile`: a tab delimited file containing chr, start, end, ID and path for the regions. For example:\n",
    "```\n",
    "#chr    start   end     ID      path\n",
    "chr12   389319  389320  ENSG00000073614 $PATH/snuc_pseudo_bulk.Mic.mega.normalized.log2cpm.bed.gz\n",
    "chr12   752578  752579  ENSG00000060237 $PATH/snuc_pseudo_bulk.Mic.mega.normalized.log2cpm.bed.gz\n",
    "```\n",
    "\n",
    "`--covFile`: path to a gzipped file containing covariates in the rows, and sample ids in the columns.  \n",
    "\n",
    "`--customized-association-windows`: a tab delimited file containing chr, start, end, and ID regions. For example:\n",
    "```\n",
    "#chr    start   end     TAD_id\n",
    "chr1    0       10985501        chr1_0_10985501\n",
    "chr1    5101787 11630758        chr1_5101787_11630758\n",
    "```\n",
    "\n",
    "`--phenotype-names`: the names of the phenotypes if multiple are included. There should be one for each phenotype file you include.\n",
    "\n",
    "`--max-cv-variants`: maximum number of variants for cross-validation.\n",
    "\n",
    "`--ld_reference_meta_file`: path to file containing chrom, start, end and path for linkage disequilibrium region information. For example:\n",
    "```\n",
    "#chrom  start   end     path\n",
    "chr1    101384274       104443097       chr1/chr1_101384274_104443097.cor.xz,chr1/chr1_101384274_104443097.cor.xz.bim\n",
    "chr1    104443097       106225286       chr1/chr1_104443097_106225286.cor.xz,chr1/chr1_104443097_106225286.cor.xz.bim\n",
    "```\n",
    "\n",
    "`--independent_variant_list`: a gzipped file containing variant information. These should be independent from one another in terms of linkage disequilibrium.\n",
    "For example:\n",
    "```\n",
    "chrom   pos     alt     ref     variant_id\n",
    "chr1    16206   T       A       chr1:16206:T:A\n",
    "chr1    16433   C       G       chr1:16433:C:G\n",
    "```\n",
    "\n",
    "`--fine_mapping_meta`: A file containg a list of gene and region information and other conditions.\n",
    "For example:\n",
    "```\n",
    "#chr    start   end     region_id       TSS     original_data   combined_data   combined_data_sumstats  conditions      conditions_top_loci\n",
    "chr1    0       6480000 ENSG00000008128 1724356 KNIGHT_pQTL.ENSG00000008128.univariate_susie_twas_weights.rds,MiGA_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,MSBB_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_Bennett_Klein_pQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_DeJager_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_Kellis_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_mega_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,STARNET_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds  $PATH/Fungen_xQTL.ENSG00000008128.cis_results_db.export.rds        $PATH/Fungen_xQTL.ENSG00000008128.cis_results_db.export_sumstats.rds       Knight_eQTL_brain,MiGA_GFM_eQTL,MiGA_GTS_eQTL,MiGA_SVZ_eQTL,MiGA_THA_eQTL,BM_10_MSBB_eQTL,BM_22_MSBB_eQTL,BM_36_MSBB_eQTL,BM_44_MSBB_eQTL,monocyte_ROSMAP_eQTL,Mic_DeJager_eQTL,Ast_DeJager_eQTL,Oli_DeJager_eQTL,Exc_DeJager_eQTL,Inh_DeJager_eQTL,DLPFC_DeJager_eQTL,PCC_DeJager_eQTL,AC_DeJager_eQTL,Mic_Kellis_eQTL,Ast_Kellis_eQTL,Oli_Kellis_eQTL,OPC_Kellis_eQTL,Exc_Kellis_eQTL,Inh_Kellis_eQTL,Ast_mega_eQTL,Exc_mega_eQTL,Inh_mega_eQTL,Oli_mega_eQTL,STARNET_eQTL_Mac       Knight_eQTL_brain,MiGA_GFM_eQTL,MiGA_GTS_eQTL,MiGA_SVZ_eQTL,MiGA_THA_eQTL,BM_10_MSBB_eQTL,BM_22_MSBB_eQTL,BM_36_MSBB_eQTL,BM_44_MSBB_eQTL,monocyte_ROSMAP_eQTL,Mic_DeJager_eQTL,Ast_DeJager_eQTL,Oli_DeJager_eQTL,Exc_DeJager_eQTL,Inh_DeJager_eQTL,DLPFC_DeJager_eQTL,PCC_DeJager_eQTL,AC_DeJager_eQTL,Mic_Kellis_eQTL,Ast_Kellis_eQTL,Oli_Kellis_eQTL,OPC_Kellis_eQTL,Exc_Kellis_eQTL,Inh_Kellis_eQTL,Ast_mega_eQTL,Exc_mega_eQTL,Inh_mega_eQTL,Oli_mega_eQTL,STARNET_eQTL_Mac\n",
    "```\n",
    "\n",
    "`--phenoIDFile`: A bed file containing a list of genes and their LD region.\n",
    "For example:\n",
    "```\n",
    "TAD_id  ID\n",
    "chr19_0_13957223        ENSG00000172270\n",
    "chr19_0_13957223        ENSG00000099864\n",
    "chr19_0_13957223        ENSG00000011304\n",
    "```\n",
    "\n",
    "`--skip-analysis-pip-cutoff`: A number of the pip cutoff. \n",
    "\n",
    "`--coverage`\n",
    "\n",
    "`--maf`\n",
    "\n",
    "`--pheno_id_map_file`: A file containing IDs and genes.\n",
    "For example:\n",
    "```\n",
    "ID      gene\n",
    "chr20:50940933:50941105:clu_44490_-:ENSG00000000419     ENSG00000000419\n",
    "chr20:50940933:50941129:clu_44490_-:ENSG00000000419     ENSG00000000419\n",
    "chr20:50936262:50942031:clu_44490_-:ENSG00000000419     ENSG00000000419\n",
    "```\n",
    "\n",
    "`--prior-canonical-matrices`\n",
    "\n",
    "`--save-data`: whether to save intermediate data or not\n",
    "\n",
    "`--twas-cv-folds`: Perform K folds valiation CV for TWAS. Set this to zero to skip\n",
    "\n",
    "`--trans-analysis`: Include this if doing trans-analysis (not using phenotypic coordinate information)\n",
    "\n",
    "`--region-name`: if you only wish to analyze one region, then include the ID of a region found in the `customized-association-windows` file\n",
    "\n",
    "`--cwd`: output file path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "929ceb29-6395-4fd5-8ef9-a198626dffa4",
   "metadata": {},
   "source": [
    "## Output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2dbef18-cc4f-40af-9951-6673d8b4e991",
   "metadata": {},
   "source": [
    "* `*.multigene_bvsr.rds`\n",
    "\n",
    "```\n",
    "> str(readRDS(\"ROSMAP_Ast_mega_eQTL.chr11_chr11_77324757_82556425.multigene_bvsr.rds\"))\n",
    "List of 1\n",
    " $ chr11_77324757_82556425:List of 12\n",
    "  ..$ mrmash_fitted              :List of 14\n",
    "  .. ..$ mu1          : num [1:14830, 1:5] 3.02e-06 3.81e-06 -1.59e-05 -1.59e-05 -1.36e-05 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ S1           : num [1:5, 1:5, 1:14830] 6.07e-07 1.55e-09 1.78e-09 1.62e-09 1.44e-09 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 3\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ w1           : num [1:14830, 1:50] 0.999 0.998 0.997 0.997 0.999 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. .. ..$ : chr [1:50] \"null\" \"singleton1_grid1\" \"singleton1_grid2\" \"singleton1_grid3\" ...\n",
    "  .. ..$ V            : num [1:5, 1:5] 0.40797 0.00349 0.016 -0.01287 0.00417 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ w0           : Named num [1:50] 9.98e-01 7.08e-05 1.12e-04 2.20e-04 4.38e-04 ...\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:50] \"null\" \"singleton1_grid1\" \"singleton1_grid2\" \"singleton1_grid3\" ...\n",
    "  .. ..$ S0           : num [1:5, 1:5, 1:50] 1e-08 0e+00 0e+00 0e+00 0e+00 0e+00 1e-08 0e+00 0e+00 0e+00 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 3\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:50] \"null\" \"singleton1_grid1\" \"singleton1_grid2\" \"singleton1_grid3\" ...\n",
    "  .. ..$ intercept    : Named num [1:5] 0.0913 -0.0112 -0.3494 -0.1176 -0.1019\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ fitted       : num [1:737, 1:5] 0.0207 0.0475 0.1043 -0.1477 0.0628 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ G            : num [1:5, 1:5] 0.00883 0.00382 0.01266 0.00541 0.00374 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ pve          : Named num [1:5] 0.0212 0.0123 0.1862 0.0254 0.0548\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ progress     :'data.frame':    55 obs. of  5 variables:\n",
    "  .. .. ..$ iter        : num [1:55] 1 2 3 4 5 6 7 8 9 10 ...\n",
    "  .. .. ..$ timing      : num [1:55] 13.1 13.6 14 14 14 ...\n",
    "  .. .. ..$ mu1_max.diff: num [1:55] 0.11127 0.06216 0.04167 0.00867 0.00708 ...\n",
    "  .. .. ..$ ELBO_diff   : num [1:55] Inf 49.14 8.237 1.527 0.788 ...\n",
    "  .. .. ..$ ELBO        : num [1:55] -3296 -3247 -3239 -3237 -3236 ...\n",
    "  .. ..$ converged    : logi TRUE\n",
    "  .. ..$ ELBO         : num -3233\n",
    "  .. ..$ analysis_time: Named num 546\n",
    "  .. .. ..- attr(*, \"names\")= chr \"elapsed\"\n",
    "  .. ..- attr(*, \"class\")= chr [1:2] \"mr.mash\" \"list\"\n",
    "  ..$ reweighted_mixture_prior   :Classes 'MashInitializer', 'R6' <environment: 0x300d828> \n",
    "  ..$ reweighted_mixture_prior_cv:List of 2\n",
    "  .. ..$ :Classes 'MashInitializer', 'R6' <environment: 0x300d828> \n",
    "  .. ..$ :Classes 'MashInitializer', 'R6' <environment: 0x300d828> \n",
    "  ..$ mvsusie_fitted             :List of 28\n",
    "  .. ..$ alpha             : num [1:20, 1:14830] 3.97e-46 4.03e-12 1.38e-05 5.70e-05 5.88e-05 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:20] \"l1\" \"l2\" \"l3\" \"l4\" ...\n",
    "  .. .. .. ..$ : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ b1                : num [1:20, 1:14830, 1:5] -7.80e-49 1.28e-14 -5.73e-08 0.00 0.00 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 3\n",
    "  .. .. .. ..$ single_effect: chr [1:20] \"l1\" \"l2\" \"l3\" \"l4\" ...\n",
    "  .. .. .. ..$ variable     : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. .. ..$ condition    : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ b2                : num [1:20, 1:14830, 1:5] 4.66e-50 7.22e-16 2.27e-09 0.00 0.00 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 3\n",
    "  .. .. .. ..$ single_effect: chr [1:20] \"l1\" \"l2\" \"l3\" \"l4\" ...\n",
    "  .. .. .. ..$ variable     : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. .. ..$ condition    : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ KL                : num [1:20] 17.4903 11.5316 7.7359 0.109 0.0886 ...\n",
    "  .. ..$ lbf               : num [1:20] 92.5381 14.961 0.6533 -0.109 -0.0886 ...\n",
    "  .. ..$ lbf_variable      : num [1:20, 1:14830] -2.396 -1.671 -0.936 0 0 ...\n",
    "  .. ..$ V                 : num [1:20] 0.03724 0.01042 0.00352 0 0 ...\n",
    "  .. ..$ sigma2            : num [1:5, 1:5] 0.40797 0.00349 0.016 -0.01287 0.00417 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ elbo              : num [1:31] -3285 -3227 -3223 -3221 -3219 ...\n",
    "  .. ..$ niter             : int 31\n",
    "  .. ..$ convergence       :List of 2\n",
    "  .. .. ..$ delta    : num 0.000904\n",
    "  .. .. ..$ converged: logi TRUE\n",
    "  .. ..$ coef              : num [1:14831, 1:5] 7.65e-02 -8.66e-08 -1.61e-08 -1.66e-06 -1.66e-06 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:14831] \"(Intercept)\" \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ b1_rescaled       : num [1:20, 1:14831, 1:5] -4.34e-03 6.76e-02 1.32e-02 5.85e-18 5.85e-18 ...\n",
    "  .. ..$ mixture_weights   : num [1:20, 1:14830, 1:11] 0 0 0 0 0 0 0 0 0 0 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 3\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. ..$ conditional_lfsr  : num [1:20, 1:14830, 1:5] 0.812 0.814 0.702 0.614 0.604 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 3\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. ..$ lfsr              : num [1:14830, 1:5] 1 1 1 1 1 ...\n",
    "  .. ..$ single_effect_lfsr: num [1:20, 1:5] 4.37e-01 3.90e-07 4.55e-01 5.82e-01 5.82e-01 ...\n",
    "  .. ..$ alpha_history     :List of 3\n",
    "  .. .. ..$ : NULL\n",
    "  .. .. ..$ : NULL\n",
    "  .. .. ..$ : NULL\n",
    "  .. ..$ lbf_history       :List of 3\n",
    "  .. .. ..$ : NULL\n",
    "  .. .. ..$ : NULL\n",
    "  .. .. ..$ : NULL\n",
    "  .. ..$ prior_history     :List of 3\n",
    "  .. .. ..$ : NULL\n",
    "  .. .. ..$ : NULL\n",
    "  .. .. ..$ : NULL\n",
    "  .. ..$ fitted            : num [1:737, 1:5] 0.0687 0.069 0.0714 -0.2329 0.0696 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ intercept         : Named num [1:5] 0.0765 0.0335 -0.3115 -0.1172 -0.1102\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ pip               : Named num [1:14830] 1.38e-05 1.07e-05 1.05e-04 1.05e-04 1.61e-05 ...\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ walltime          : 'proc_time' Named num [1:5] 364.51 4.63 370.42 0 0\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ sets              :List of 5\n",
    "  .. .. ..$ cs                :List of 2\n",
    "  .. .. .. ..$ L1: int [1:11] 1814 1827 1839 1856 1863 1892 1919 1937 1940 1943 ...\n",
    "  .. .. .. ..$ L2: int [1:242] 458 467 470 474 510 513 516 517 520 521 ...\n",
    "  .. .. ..$ purity            :'data.frame':    2 obs. of  3 variables:\n",
    "  .. .. .. ..$ min.abs.corr   : num [1:2] 0.979 0.898\n",
    "  .. .. .. ..$ mean.abs.corr  : num [1:2] 0.989 0.975\n",
    "  .. .. .. ..$ median.abs.corr: num [1:2] 0.988 0.973\n",
    "  .. .. ..$ cs_index          : int [1:2] 1 2\n",
    "  .. .. ..$ coverage          : num [1:2] 0.954 0.951\n",
    "  .. .. ..$ requested_coverage: num 0.95\n",
    "  .. ..$ residual_variance : num [1:5, 1:5] 0.40797 0.00349 0.016 -0.01287 0.00417 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ condition_names   : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ variable_names    : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..- attr(*, \"class\")= chr \"mvsusie\"\n",
    "  ..$ variant_names              : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  ..$ analysis_script            : chr \"\\nlibrary(data.table)\\nlibrary(dplyr)\\nlibrary(pecotmr)\\ncombine_result_list <- function(univariate_finemapping\"| __truncated__\n",
    "  ..$ other_quantities           :List of 1\n",
    "  .. ..$ dropped_samples: NULL\n",
    "  ..$ context_names              : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  ..$ top_loci                   :'data.frame': 263 obs. of  4 variables:\n",
    "  .. ..$ variant_id      : chr [1:263] \"chr11:77534902:C:T\" \"chr11:77539867:G:A\" \"chr11:77543221:C:T\" \"chr11:77546759:C:T\" ...\n",
    "  .. ..$ maf             : num [1:263] 0.107 0.107 0.107 0.107 0.107 ...\n",
    "  .. ..$ pip             : num [1:263] 0.000857 0.000857 0.000932 0.000857 0.000857 ...\n",
    "  .. ..$ cs_coverage_0.95: num [1:263] 2 2 2 2 2 2 2 2 2 2 ...\n",
    "  ..$ susie_result_trimmed       :List of 12\n",
    "  .. ..$ pip           : num [1:14830] 1.38e-05 1.07e-05 1.05e-04 1.05e-04 1.61e-05 ...\n",
    "  .. ..$ sets          :List of 5\n",
    "  .. .. ..$ cs                :List of 2\n",
    "  .. .. .. ..$ L1: int [1:11] 1814 1827 1839 1856 1863 1892 1919 1937 1940 1943 ...\n",
    "  .. .. .. ..$ L2: int [1:242] 458 467 470 474 510 513 516 517 520 521 ...\n",
    "  .. .. ..$ purity            :'data.frame':    2 obs. of  3 variables:\n",
    "  .. .. .. ..$ min.abs.corr   : num [1:2] 0.979 0.877\n",
    "  .. .. .. ..$ mean.abs.corr  : num [1:2] 0.989 0.972\n",
    "  .. .. .. ..$ median.abs.corr: num [1:2] 0.988 0.973\n",
    "  .. .. ..$ cs_index          : int [1:2] 1 2\n",
    "  .. .. ..$ coverage          : num [1:2] 0.954 0.951\n",
    "  .. .. ..$ requested_coverage: num 0.95\n",
    "  .. ..$ cs_corr       : num [1:2, 1:2] 1 -0.184 -0.184 1\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:2] \"L1\" \"L2\"\n",
    "  .. .. .. ..$ : chr [1:2] \"L1\" \"L2\"\n",
    "  .. ..$ sets_secondary: NULL\n",
    "  .. ..$ alpha         : num [1:3, 1:14830] 3.97e-46 4.03e-12 1.38e-05 3.77e-46 3.82e-12 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:3] \"l1\" \"l2\" \"l3\"\n",
    "  .. .. .. ..$ : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ lbf_variable  : num [1:3, 1:14830] -2.396 -1.671 -0.936 -2.449 -1.726 ...\n",
    "  .. ..$ V             : num [1:3] 0.03724 0.01042 0.00352\n",
    "  .. ..$ niter         : int 31\n",
    "  .. ..$ max_L         : int 20\n",
    "  .. ..$ b1_rescaled   : num [1:3, 1:14831, 1:5] -4.34e-03 6.76e-02 1.32e-02 -1.18e-48 1.94e-14 ...\n",
    "  .. ..$ coef          : num [1:14831, 1:5] 7.65e-02 -8.66e-08 -1.61e-08 -1.66e-06 -1.66e-06 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
    "  .. .. .. ..$ : chr [1:14831] \"(Intercept)\" \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" ...\n",
    "  .. .. .. ..$ : chr [1:5] \"ENSG00000074201\" \"ENSG00000048649\" \"ENSG00000159063\" \"ENSG00000188997\" ...\n",
    "  .. ..$ clfsr         : num [1:3, 1:14830, 1:5] 0.812 0.814 0.702 0.872 0.862 ...\n",
    "  .. .. ..- attr(*, \"dimnames\")=List of 3\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. .. .. ..$ : NULL\n",
    "  .. ..- attr(*, \"class\")= chr \"susie\"\n",
    "  ..$ total_time_elapsed         : 'proc_time' Named num [1:5] 982 13 999 0 0\n",
    "  .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  ..$ region_info                :List of 3\n",
    "  .. ..$ region_coord:'data.frame':     1 obs. of  3 variables:\n",
    "  .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. ..$ start: int 77324757\n",
    "  .. .. ..$ end  : int 82556425\n",
    "  .. ..$ grange      :'data.frame':     1 obs. of  3 variables:\n",
    "  .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. ..$ start: int 77324757\n",
    "  .. .. ..$ end  : int 82556425\n",
    "  .. ..$ region_name : chr [1:12] \"chr11_77324757_82556425\" \"ENSG00000149269\" \"ENSG00000074201\" \"ENSG00000048649\" ...\n",
    "```\n",
    "\n",
    "* `*.multigene_twas_weights.rds`\n",
    "\n",
    "```\n",
    "> str(readRDS(\"ROSMAP_Ast_mega_eQTL.chr11_chr11_77324757_82556425.multigene_twas_weights.rds\"))\n",
    "List of 1\n",
    " $ chr11_77324757_82556425:List of 5\n",
    "  ..$ ENSG00000074201:List of 5\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. .. ..$ mrmash_weights : Named num [1:14830] 3.02e-06 3.81e-06 -1.59e-05 -1.59e-05 -1.36e-05 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. ..$ mvsusie_weights: Named num [1:14830] -8.66e-08 -1.61e-08 -1.66e-06 -1.66e-06 -8.16e-08 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. .. ..$ mrmash_predicted : Named num [1:737] -0.0706 -0.0438 0.013 -0.239 -0.0285 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. .. ..$ mvsusie_predicted: Named num [1:737] -0.00776 -0.00753 -0.00509 -0.30936 -0.00694 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. ..$ variant_names     : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 0.725 0.179 0.981 0.011 0.018\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  .. .. ..$ region_coord:'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ grange      :'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ region_name : chr [1:12] \"chr11_77324757_82556425\" \"ENSG00000149269\" \"ENSG00000074201\" \"ENSG00000048649\" ...\n",
    "  ..$ ENSG00000048649:List of 5\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. .. ..$ mrmash_weights : Named num [1:14830] -5.97e-07 -1.07e-06 -1.76e-05 -1.76e-05 -2.98e-06 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. ..$ mvsusie_weights: Named num [1:14830] -1.01e-07 -4.81e-08 -1.17e-06 -1.17e-06 -4.58e-08 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. .. ..$ mrmash_predicted : Named num [1:737] 0.07298 0.03071 0.00939 -0.07178 0.04762 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. .. ..$ mvsusie_predicted: Named num [1:737] 2.05e-02 1.89e-02 4.65e-06 -1.70e-01 1.98e-02 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. ..$ variant_names     : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 0.725 0.179 0.981 0.011 0.018\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  .. .. ..$ region_coord:'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ grange      :'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ region_name : chr [1:12] \"chr11_77324757_82556425\" \"ENSG00000149269\" \"ENSG00000074201\" \"ENSG00000048649\" ...\n",
    "  ..$ ENSG00000159063:List of 5\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. .. ..$ mrmash_weights : Named num [1:14830] -2.51e-06 -9.33e-06 -3.67e-04 -3.67e-04 4.71e-06 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. ..$ mvsusie_weights: Named num [1:14830] -2.12e-07 -2.77e-07 -3.14e-05 -3.14e-05 2.22e-07 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. .. ..$ mrmash_predicted : Named num [1:737] 0.4017 0.3971 -0.0145 -0.0315 0.4016 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. .. ..$ mvsusie_predicted: Named num [1:737] 0.3727 0.3725 -0.0192 -0.1086 0.3728 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. ..$ variant_names     : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 0.725 0.179 0.981 0.011 0.018\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  .. .. ..$ region_coord:'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ grange      :'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ region_name : chr [1:12] \"chr11_77324757_82556425\" \"ENSG00000149269\" \"ENSG00000074201\" \"ENSG00000048649\" ...\n",
    "  ..$ ENSG00000188997:List of 5\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. .. ..$ mrmash_weights : Named num [1:14830] -1.31e-07 9.65e-08 -1.04e-05 -1.04e-05 -7.16e-07 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. ..$ mvsusie_weights: Named num [1:14830] -9.39e-08 4.32e-08 2.20e-06 2.20e-06 -5.78e-08 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. .. ..$ mrmash_predicted : Named num [1:737] 0.14819 0.14679 -0.00484 -0.02042 0.15488 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. .. ..$ mvsusie_predicted: Named num [1:737] 0.1523 0.1519 -0.02 -0.0413 0.1534 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. ..$ variant_names     : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 0.725 0.179 0.981 0.011 0.018\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  .. .. ..$ region_coord:'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ grange      :'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ region_name : chr [1:12] \"chr11_77324757_82556425\" \"ENSG00000149269\" \"ENSG00000074201\" \"ENSG00000048649\" ...\n",
    "  ..$ ENSG00000033327:List of 5\n",
    "  .. ..$ twas_weights      :List of 2\n",
    "  .. .. ..$ mrmash_weights : Named num [1:14830] -7.89e-07 -2.21e-06 -3.37e-05 -3.37e-05 1.50e-06 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. .. ..$ mvsusie_weights: Named num [1:14830] -1.02e-07 -1.15e-07 -3.05e-06 -3.05e-06 1.75e-07 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ twas_predictions  :List of 2\n",
    "  .. .. ..$ mrmash_predicted : Named num [1:737] 0.11803 0.12448 -0.00522 -0.00542 0.12055 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. .. ..$ mvsusie_predicted: Named num [1:737] 0.126205 0.127183 -0.005335 0.000314 0.12657 ...\n",
    "  .. .. .. ..- attr(*, \"names\")= chr [1:737] \"MAP15387421\" \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" ...\n",
    "  .. ..$ variant_names     : chr [1:14830] \"chr11:77325990:C:T\" \"chr11:77326116:G:A\" \"chr11:77326354:A:G\" \"chr11:77326475:G:C\" ...\n",
    "  .. ..$ total_time_elapsed: 'proc_time' Named num [1:5] 0.725 0.179 0.981 0.011 0.018\n",
    "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
    "  .. ..$ region_info       :List of 3\n",
    "  .. .. ..$ region_coord:'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ grange      :'data.frame':  1 obs. of  3 variables:\n",
    "  .. .. .. ..$ chrom: chr \"11\"\n",
    "  .. .. .. ..$ start: int 77324757\n",
    "  .. .. .. ..$ end  : int 82556425\n",
    "  .. .. ..$ region_name : chr [1:12] \"chr11_77324757_82556425\" \"ENSG00000149269\" \"ENSG00000074201\" \"ENSG00000048649\" ...\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "397a4dc3-1e4c-4845-ad8b-beed40f1f408",
   "metadata": {},
   "source": [
    "## Minimal Working Example Steps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bca0962-ae57-422c-a79f-892ef3b7f7ae",
   "metadata": {},
   "source": [
    "### ii. [Run the Fine-Mapping with mvSuSiE](https://statfungen.github.io/xqtl-protocol/code/mnm_analysis/mnm_methods/mnm_regression.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b6b5d17-0846-4ed3-ad9a-b8f2c0665dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sos run pipeline/mnm_regression.ipynb mnm_genes \\\n",
    "    --name ROSMAP_Ast_mega_eQTL \\\n",
    "    --genoFile data/mnm_genes/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.11.bed \\\n",
    "    --phenoFile data/mnm_genes/snuc_pseudo_bulk.Ast.mega.normalized.log2cpm.region_list.txt \\\n",
    "    --covFile data/mnm_genes/snuc_pseudo_bulk.Ast.mega.normalized.log2cpm.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.snuc_pseudo_bulk_mega.related.plink_qc.extracted.pca.projected.Marchenko_PC.gz \\\n",
    "    --customized-association-windows data/mnm_genes/extended_TADB.bed \\\n",
    "    --phenotype-names Ast_mega_eQTL \\\n",
    "    --max-cv-variants 5000 \\\n",
    "    --ld_reference_meta_file data/ld_meta_file_with_bim.tsv \\\n",
    "    --independent_variant_list data/mnm_genes/ld_pruned_variants.txt.gz \\\n",
    "    --fine_mapping_meta data/mnm_genes/combined_data_updated.tsv \\\n",
    "    --phenoIDFile data/mnm_genes/phenoIDFile_extended_TADB.bed \\\n",
    "    --region-name chr11_77324757_82556425 \\\n",
    "    --skip-analysis-pip-cutoff 0 \\\n",
    "    --maf 0.01 \\\n",
    "    --coverage 0.95 \\\n",
    "    --pheno_id_map_file data/mnm_genes/pheno_id_map_file.txt \\\n",
    "    --prior-canonical-matrices \\\n",
    "    --twas-cv-folds 0 \\\n",
    "    --trans-analysis \\\n",
    "    --cwd output/mnm_regression/mnm_genes -s build"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc7691e6-954c-4dc1-bac3-25d56026ab96",
   "metadata": {},
   "source": [
    "## Anticipated Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08582666-774d-4620-a3bf-31f35689e7da",
   "metadata": {},
   "source": [
    "For each gene and region, multivariate multigene finemapping will produce a file containing results for the top hits and a file containing twas weights produced by susie."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "199a2637-cc0c-4aff-b4ea-a2546936f073",
   "metadata": {},
   "source": [
    "`ROSMAP_Ast_DeJager_eQTL.chr11_77324757_86627922.multigene_bvrs.rds`:\n",
    "* for each region name, includes:\n",
    "1. mrmash_fitted\n",
    "2. reweighted_mixture_prior\n",
    "3. reweighted_mixture_prior_cv\n",
    "4. mvsusie_fitted\n",
    "5. variant_names\n",
    "6. analysis_script\n",
    "7. other_quantitites\n",
    "8. analysis_script\n",
    "9. top_loci\n",
    "10. susie_result_trimmed\n",
    "11. total_time_elapsed\n",
    "12. region_info\n",
    "\n",
    "`ROSMAP_Ast_DeJager_eQTL.chr11_77324757_86627922.multigene_data.rds`:(from the --save-data argument)\n",
    "* see [pecotmr code](https://github.com/statfungen/pecotmr/blob/68d87ca1d0a059022bf4e55339621cbddc8993cc/R/file_utils.R#L461) for description \n",
    "\n",
    "`ROSMAP_Ast_DeJager_eQTL.chr11_77324757_86627922.multigene_twas_weights.rds`:\n",
    "* for each region name and for each gene within that region, includes:\n",
    "1. twas_weights - from mrmash and mvsusie\n",
    "2. twas_predictions - from mrmash and mvsusie\n",
    "3. variant_names\n",
    "4. total_time_elapsed\n",
    "5. region_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "159e580c-e290-4403-a701-d9dfa1043341",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
