

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>High-dimensional regression with summary statistics &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/mnm_analysis/rss_analysis';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Fine-mapping with PolyFun" href="../prototype_drafts/polyfun.html" />
    <link rel="prev" title="Multivariate TWAS prediction model" href="mrmash.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l1"><a class="reference internal" href="SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced Genome-wide Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">High-dimensional regression with summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/polyfun.html">Fine-mapping with PolyFun</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/MRAID_QTL.html">Mendelian Randomization using MRAID</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/edit/gh-pages/code/mnm_analysis/rss_analysis.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/mnm_analysis/rss_analysis.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/mnm_analysis/rss_analysis.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>High-dimensional regression with summary statistics</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="high-dimensional-regression-with-summary-statistics">
<h1>High-dimensional regression with summary statistics<a class="headerlink" href="#high-dimensional-regression-with-summary-statistics" title="Permalink to this heading">#</a></h1>
<p>This notebook take a list of LD reference files and a list of sumstat files from various association studies, and perform:</p>
<ol class="arabic simple">
<li><p>Fine-mapping with SuSiE RSS model</p></li>
<li><p>TWAS / PRS weights …</p></li>
</ol>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<p>I. <strong>GWAS Summary Statistics Files</strong></p>
<ul class="simple">
<li><p><strong>Input</strong>: Vector of files for one or more GWAS studies.</p></li>
<li><p><strong>Format</strong>:</p>
<ul>
<li><p>Tab-delimited files.</p></li>
<li><p>First 4 columns: <code class="docutils literal notranslate"><span class="pre">chrom</span></code>, <code class="docutils literal notranslate"><span class="pre">pos</span></code>, <code class="docutils literal notranslate"><span class="pre">A1</span></code>, <code class="docutils literal notranslate"><span class="pre">A2</span></code></p></li>
<li><p>Additional columns can be loaded using column mapping file see below</p></li>
</ul>
</li>
<li><p><strong>Column Mapping files (optional)</strong>:</p>
<ul>
<li><p>Optional YAML file for custom column mapping.</p></li>
<li><p>Required columns: <code class="docutils literal notranslate"><span class="pre">chrom</span></code>, <code class="docutils literal notranslate"><span class="pre">pos</span></code>, <code class="docutils literal notranslate"><span class="pre">A1</span></code>, <code class="docutils literal notranslate"><span class="pre">A2</span></code>, either <code class="docutils literal notranslate"><span class="pre">z</span></code> or (<code class="docutils literal notranslate"><span class="pre">beta</span></code> and <code class="docutils literal notranslate"><span class="pre">se</span></code>), <code class="docutils literal notranslate"><span class="pre">n_sample</span></code>, <code class="docutils literal notranslate"><span class="pre">n_case</span></code>, <code class="docutils literal notranslate"><span class="pre">n_control</span></code>. Note: You can only provide <code class="docutils literal notranslate"><span class="pre">n_sample</span></code> or (<code class="docutils literal notranslate"><span class="pre">n_case</span></code> &amp; <code class="docutils literal notranslate"><span class="pre">n_control</span></code>, together for case control study), fill with 0 if do not provide them. If none of these are known, fill <code class="docutils literal notranslate"><span class="pre">n_sample</span></code>, <code class="docutils literal notranslate"><span class="pre">n_case</span></code>, <code class="docutils literal notranslate"><span class="pre">n_control</span></code> with 0.</p></li>
<li><p>Optional columns: <code class="docutils literal notranslate"><span class="pre">var_y</span></code>.</p></li>
</ul>
</li>
</ul>
<p>II. <strong>GWAS Summary Statistics Meta-File</strong>: this is optional and helpful when there are lots of GWAS summary statistics data to process via the same command. All studies will run at the same time.</p>
<p>The meta file should contain 7 columns.</p>
<ul class="simple">
<li><p><strong>Columns</strong>:</p></li>
</ul>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">study_id</span></code>: can be any string / number.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">chrom</span></code>: the chromosome that summary statistics cover. Chromosome number <code class="docutils literal notranslate"><span class="pre">0</span></code> indicates a genome-wide summary statistics file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">file_path</span></code>: the path to the summary statistics file. The format of summary statistics file can be tsv, csv.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">column_mapping_file</span></code>: served to unify the column names in file_path. Map those column names to a standard column name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_sample</span></code>: (effective) sample size of the study. If not known, fill with 0.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_case</span></code>: (effective) case number in the study if it’s case control study. If not known, fill with 0.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_control</span></code>: control number in the study if it’s case control study. If not known, fill with 0.</p></li>
</ol>
<p><strong>Note:</strong> we only need n_sample or (n_case &amp; n_control). If is a case control study but n_case n_control not known, using n_sample also works.</p>
<p>eg: <code class="docutils literal notranslate"><span class="pre">gwas_meta.tsv</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">study_id</span>	<span class="n">chrom</span>	<span class="n">file_path</span>	<span class="n">column_mapping_file</span>	<span class="n">n_sample</span>	<span class="n">n_case</span>	<span class="n">n_control</span>
<span class="n">AD_1</span>	<span class="mi">0</span>	<span class="n">data1</span><span class="o">.</span><span class="n">txt</span>	<span class="n">mapping1</span><span class="o">.</span><span class="n">yml</span>	<span class="mi">0</span>	<span class="mi">111326</span>	<span class="mi">677663</span>
<span class="n">AD_2</span>	<span class="mi">0</span>	<span class="n">data2</span><span class="o">.</span><span class="n">txt</span>	<span class="n">mapping2</span><span class="o">.</span><span class="n">yml</span>	<span class="mi">123455</span>	<span class="mi">0</span>	<span class="mi">0</span>
<span class="n">AD_3</span>	<span class="mi">2</span>	<span class="n">data3</span><span class="o">.</span><span class="n">txt</span>	<span class="n">mapping3</span><span class="o">.</span><span class="n">yml</span>	<span class="mi">0</span>	<span class="mi">21982</span>	<span class="mi">41944</span>
</pre></div>
</div>
<p>If both summary stats file (I) and meta data file (II) are specified we will take the union of the two.</p>
<p>eg. <code class="docutils literal notranslate"><span class="pre">column_mapping.yml</span></code> left: standard name (please hold on to it). Right: original column name in your summary statistics file. (do not add space before and after “:”)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chrom</span><span class="p">:</span><span class="n">chromosome</span>
<span class="n">pos</span><span class="p">:</span><span class="n">base_pair_location</span>
<span class="n">A1</span><span class="p">:</span><span class="n">effect_allele</span>
<span class="n">A2</span><span class="p">:</span><span class="n">other_allele</span>
<span class="n">beta</span><span class="p">:</span><span class="n">beta</span>
<span class="n">se</span><span class="p">:</span><span class="n">standard_error</span>
<span class="n">pvalue</span><span class="p">:</span><span class="n">p_value</span>
<span class="n">maf</span><span class="p">:</span><span class="n">maf</span>
<span class="n">z</span><span class="p">:</span><span class="n">z</span>
<span class="n">n_case</span><span class="p">:</span><span class="n">n_cases</span>
<span class="n">n_control</span><span class="p">:</span><span class="n">n_controls</span>
<span class="n">n_sample</span><span class="p">:</span><span class="n">n</span>
</pre></div>
</div>
<p>III. <strong>LD Reference Metadata File</strong></p>
<ul class="simple">
<li><p><strong>Format</strong>: Single TSV file.</p></li>
<li><p><strong>Contents</strong>:</p>
<ul>
<li><p>Columns: <code class="docutils literal notranslate"><span class="pre">#chrom</span></code>, <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">end</span></code>, path to the LD matrix, genomic build.</p></li>
<li><p>LD matrix path format: comma-separated, first entry is the LD matrix, second is the bim file.</p></li>
</ul>
</li>
<li><p><strong>Documentation</strong>: Refer to our LD reference preparation document for detailed information (Tosin pending update).</p></li>
</ul>
<p>IV. For analyzing specific genomic regions, you can specify them using the <code class="docutils literal notranslate"><span class="pre">--region-names</span></code> option in the ‘chr:start-end’ format, where multiple regions are accepted. Alternatively, you may provide a file containing a list of regions through the <code class="docutils literal notranslate"><span class="pre">--region-list</span></code> option, also adhering to the ‘chr:start-end’ format. When both <code class="docutils literal notranslate"><span class="pre">--region-names</span></code> and <code class="docutils literal notranslate"><span class="pre">--region-list</span></code> are provided, union of these options will be used to analyze. In cases where neither option is specified, the analysis defaults to encompass all regions specified in the LD reference metadata.</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<p>Each rds file is the finemapping result of one LD block. Including 10 elements in each rds file.</p>
<ol class="arabic simple">
<li><p>single_effect_regression: Assume only one causal variant, simple regression.</p></li>
<li><p>noqc: Susie finemapping, the sumstat is only allele_qced.</p></li>
<li><p>qc_impute: Susie finemapping, the summary statistics are QCed (suspecious outliers of z scores removed) and imputed (all outliers and missing variants from reference panel).</p></li>
<li><p>qc_only: Susie finemapping, the summary statistics are only QCed .</p></li>
<li><p>conditional_regression_noqc: Bayesian conditional regression, original sumstat</p></li>
<li><p>conditional_regression_qc_only: Bayesian conditional regression, sumstat QCed</p></li>
<li><p>conditional_regression_qc_impute: Bayesian conditional regression, sumstat QCed and imputed.</p></li>
<li><p>sumstats_allele_qc_only: a table of sumstat after allele qc (check each variant to see if ref/alt are flipped, and correct them)</p></li>
<li><p>sumstats_qc_impute: a table of sumstat after qc and imputation.</p></li>
<li><p>sumstats_qc_impute: a table of sumstat after qc and imputation, bad quality imputation variants removed by empirical thresholds. (the actual input of all qc_impute results)</p></li>
</ol>
<p>Each rds file is accompanied by 2 tsv files reporting elements 8 and 9 in tsv format for the sake of convenience.</p>
</section>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">rss_analysis</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">univariate_rss</span> \
    <span class="o">--</span><span class="n">ld</span><span class="o">-</span><span class="n">meta</span><span class="o">-</span><span class="n">data</span> <span class="n">ADSP_R4_EUR</span><span class="o">.</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span> \
    <span class="o">--</span><span class="n">gwas</span><span class="o">-</span><span class="n">meta</span><span class="o">-</span><span class="n">data</span> <span class="n">AD_sumstat_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">impute</span> <span class="o">--</span><span class="n">qc</span><span class="o">-</span><span class="n">method</span> <span class="n">dentist</span> <span class="o">--</span><span class="n">finemapping</span><span class="o">-</span><span class="n">method</span> <span class="n">susie_rss</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">pecotmr_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output/&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">gwas_meta_data</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">ld_meta_data</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">gwas_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">gwas_data</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">column_mapping</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">skip_regions</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">5</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;10h&quot;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">partition</span><span class="p">):</span>
    <span class="c1"># from itertools import accumulate</span>
    <span class="c1"># partition = [len(x) for x in partition]</span>
    <span class="c1"># Compute the cumulative sums once</span>
    <span class="c1"># cumsum_vector = list(accumulate(partition))</span>
    <span class="c1"># Use slicing based on the cumulative sums</span>
    <span class="c1"># return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]</span>
    <span class="k">return</span> <span class="n">partition</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">region_list</span><span class="p">))</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_name</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">region_list</span> <span class="o">=</span> <span class="n">ld_meta_data</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">get_analysis_regions</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="s2">&quot;regional_data&quot;</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>

<span class="k">def</span> <span class="nf">file_exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">relative_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if a file exists at the given path or relative to a specified path.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">elif</span> <span class="n">relative_path</span><span class="p">:</span>
        <span class="n">relative_file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">relative_path</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">relative_file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">relative_file_path</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">False</span>

<span class="k">def</span> <span class="nf">check_required_columns</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">required_columns</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if the required columns are present in the dataframe.&quot;&quot;&quot;</span>
    <span class="n">missing_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">required_columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">missing_columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing required columns: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">missing_columns</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">parse_region</span><span class="p">(</span><span class="n">region</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse a region string in &#39;chr:start-end&#39; format into a list [chr, start, end].&quot;&quot;&quot;</span>
    <span class="n">chrom</span><span class="p">,</span> <span class="n">rest</span> <span class="o">=</span> <span class="n">region</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">rest</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)]</span>

<span class="k">def</span> <span class="nf">extract_regional_data</span><span class="p">(</span><span class="n">gwas_meta_data</span><span class="p">,</span> <span class="n">gwas_name</span><span class="p">,</span> <span class="n">gwas_data</span><span class="p">,</span> <span class="n">column_mapping</span><span class="p">,</span> <span class="n">region_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">region_list</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts data from GWAS metadata files and additional GWAS data provided. </span>
<span class="sd">    Optionally filters data based on specified regions.</span>

<span class="sd">    Args:</span>
<span class="sd">    - gwas_meta_data (str): File path to the GWAS metadata file.</span>
<span class="sd">    - gwas_name (list): Vector of GWAS study names.</span>
<span class="sd">    - gwas_data (list): Vector of GWAS data.</span>
<span class="sd">    - column_mapping (list, optional): Vector of column mapping files.</span>
<span class="sd">    - region_name (list, optional): List of region names in &#39;chr:start-end&#39; format.</span>
<span class="sd">    - region_list (str, optional): File path to a file containing regions.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - GWAS Dictionary: Maps study IDs to a list containing chromosome number, </span>
<span class="sd">      GWAS file path, and optional column mapping file path.</span>
<span class="sd">    - Region Dictionary: Maps region names to lists [chr, start, end].</span>

<span class="sd">    Raises:</span>
<span class="sd">    - FileNotFoundError: If any specified file path does not exist.</span>
<span class="sd">    - ValueError: If required columns are missing in the input files or vector lengths mismatch.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check vector lengths</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gwas_name</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gwas_data</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gwas_name and gwas_data must be of equal length&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column_mapping</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">column_mapping</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gwas_name</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If column_mapping is provided, it must be of the same length as gwas_name and gwas_data&quot;</span><span class="p">)</span>

    <span class="c1"># Required columns for GWAS file type</span>
    <span class="n">required_gwas_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">,</span> <span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;file_path&#39;</span><span class="p">]</span>

    <span class="c1"># Base directory of the metadata files</span>
    <span class="n">gwas_base_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">gwas_meta_data</span><span class="p">)</span>
    
    <span class="c1"># Reading the GWAS metadata file</span>
    <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gwas_meta_data</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">check_required_columns</span><span class="p">(</span><span class="n">gwas_df</span><span class="p">,</span> <span class="n">required_gwas_columns</span><span class="p">)</span>
    <span class="n">gwas_dict</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

    <span class="c1"># Process additional GWAS data from vectors</span>
    <span class="k">for</span> <span class="kp">name</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">mapping</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">gwas_name</span><span class="p">,</span> <span class="n">gwas_data</span><span class="p">,</span> <span class="n">column_mapping</span> <span class="ow">or</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">gwas_name</span><span class="p">)):</span>
        <span class="n">gwas_dict</span><span class="p">[</span><span class="kp">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="p">[</span><span class="n">data</span><span class="p">,</span> <span class="n">mapping</span><span class="p">]}</span>

    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gwas_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">file_path</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;file_path&#39;</span><span class="p">]</span>
        <span class="n">mapping_file</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;column_mapping_file&#39;</span><span class="p">)</span>
        <span class="n">n_sample</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;n_sample&#39;</span><span class="p">)</span>
        <span class="n">n_case</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;n_case&#39;</span><span class="p">)</span>
        <span class="n">n_control</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;n_control&#39;</span><span class="p">)</span>

        <span class="c1"># Check if the file and optional mapping file exist</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">gwas_base_dir</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">mapping_file</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">mapping_file</span><span class="p">,</span> <span class="n">gwas_base_dir</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> not found for </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1"># Adjust paths if necessary</span>
        <span class="n">file_path</span> <span class="o">=</span> <span class="n">file_path</span> <span class="k">if</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gwas_base_dir</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mapping_file</span><span class="p">:</span>
            <span class="n">mapping_file</span> <span class="o">=</span> <span class="n">mapping_file</span> <span class="k">if</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">mapping_file</span><span class="p">)</span> <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gwas_base_dir</span><span class="p">,</span> <span class="n">mapping_file</span><span class="p">)</span>
        
        <span class="c1"># Create or update the entry for the study_id</span>
        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gwas_dict</span><span class="p">:</span>
            <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Expand chrom 0 to chrom 1-22 or use the specified chrom</span>
        <span class="n">chrom_range</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]]</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chrom_range</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]]:</span>
                <span class="n">existing_entry</span> <span class="o">=</span> <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]][</span><span class="n">chrom</span><span class="p">]</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Duplicate chromosome specification for study_id </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">, chrom </span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">. &quot;</span>
                                 <span class="sa">f</span><span class="s2">&quot;Conflicting entries: </span><span class="si">{</span><span class="n">existing_entry</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="p">[</span><span class="n">file_path</span><span class="p">,</span><span class="w"> </span><span class="n">mapping_file</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]][</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">file_path</span><span class="p">,</span> <span class="n">mapping_file</span><span class="p">,</span> <span class="n">n_sample</span><span class="p">,</span> <span class="n">n_case</span><span class="p">,</span> <span class="n">n_control</span><span class="p">]</span>

         <span class="c1"># Process region_list and region_name</span>
            <span class="n">region_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">region_list</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">region_list</span><span class="p">):</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">region_list</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
                        <span class="c1"># Skip empty lines</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
                            <span class="k">continue</span>
                        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                            <span class="k">continue</span>
                        <span class="n">parts</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                            <span class="n">region</span> <span class="o">=</span> <span class="n">parse_region</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
                            <span class="n">region</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)),</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">2</span><span class="p">])]</span>
                        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">4</span> <span class="ow">and</span>  <span class="n">region_list</span> <span class="o">!=</span> <span class="n">ld_meta_data</span> <span class="p">:</span> <span class="c1"># for eQTL where chr:start:end:gene_id:gene_name, and path if LD_meta are used.</span>
                            <span class="n">region</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)),</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">2</span><span class="p">]),</span><span class="n">parts</span><span class="p">[</span><span class="mi">3</span><span class="p">]]</span>

        
                        <span class="k">else</span><span class="p">:</span>
                            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid region format in region_list&quot;</span><span class="p">)</span>
                
                        <span class="n">region_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">region</span>
                
    <span class="k">if</span> <span class="n">region_name</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">region_name</span><span class="p">:</span>
            <span class="n">parsed_region</span> <span class="o">=</span> <span class="n">parse_region</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
            <span class="n">region_key</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">parsed_region</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">parsed_region</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">parsed_region</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">region_key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">region_dict</span><span class="p">:</span>
                <span class="n">region_dict</span><span class="p">[</span><span class="n">region_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">parsed_region</span>

    <span class="k">return</span> <span class="n">gwas_dict</span><span class="p">,</span> <span class="n">region_dict</span>

<span class="n">gwas_dict</span><span class="p">,</span> <span class="n">region_dict</span> <span class="o">=</span> <span class="n">extract_regional_data</span><span class="p">(</span><span class="n">gwas_meta_data</span><span class="p">,</span> <span class="n">gwas_name</span><span class="p">,</span> <span class="n">gwas_data</span><span class="p">,</span> <span class="n">column_mapping</span><span class="p">,</span> <span class="n">region_name</span><span class="p">,</span> <span class="n">region_list</span><span class="p">)</span>
<span class="n">regional_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="s2">&quot;GWAS&quot;</span><span class="p">,</span> <span class="n">gwas_dict</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;regions&quot;</span><span class="p">,</span> <span class="n">region_dict</span><span class="p">)])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[univariate_rss]
parameter: L = 5
parameter: max_L = 10
# If available the column that indicates sample size within the sumstats
# filtering threshold for raiss imputation
parameter: rcond = 0.01
parameter: lamb = 0.01
parameter: R2_threshold = 0.6
parameter: l_step = 5
parameter: pip_cutoff = 0.025
parameter: skip_analysis_pip_cutoff = 0.025
parameter: minimum_ld = 5
parameter: coverage = [0.95, 0.7, 0.5]
# Whether to impute the sumstat for all the snp in LD but not in sumstat.
parameter: impute = False 
# summary stats QC methods: rss_qc, dentist, slalom
parameter: qc_method = &quot;&quot;
# analysis method: single_effect, susie_rss, bayesian_conditional_regression
parameter: finemapping_method = &quot;single_effect&quot;
depends: sos_variable(&quot;regional_data&quot;)
regions = list(regional_data[&#39;regions&#39;].keys())
input: for_each = &quot;regions&quot;
output: f&#39;{cwd:a}/{step_name}/{_regions.replace(&quot;:&quot;, &quot;_&quot;)}.{&quot;%s&quot; % qc_method.upper() if qc_method else &quot;noQC&quot;}{&quot;_RAISS_imputed&quot; if impute else &quot;&quot;}.univariate{&quot;_%s&quot; % finemapping_method if finemapping_method else &quot;&quot;}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    library(pecotmr)
    library(data.table)
    skip_region = c(${&#39;, &#39;.join([&#39;&quot;{}&quot;&#39;.format(item) for item in skip_regions])})
    studies = c(${&#39;, &#39;.join([&#39;&quot;{}&quot;&#39;.format(item) for item in regional_data[&quot;GWAS&quot;].keys()])})
    sumstat_paths = c(${paths([regional_data[&#39;GWAS&#39;][x][regional_data[&#39;regions&#39;][_regions][0]][0] for x in regional_data[&quot;GWAS&quot;].keys()]):r,})
    column_file_paths = c(${paths([regional_data[&#39;GWAS&#39;][x][regional_data[&#39;regions&#39;][_regions][0]][1] for x in regional_data[&quot;GWAS&quot;].keys()]):r,})
    n_samples = c(${paths([regional_data[&#39;GWAS&#39;][x][regional_data[&#39;regions&#39;][_regions][0]][2] for x in regional_data[&quot;GWAS&quot;].keys()]):r,})
    n_cases = c(${paths([regional_data[&#39;GWAS&#39;][x][regional_data[&#39;regions&#39;][_regions][0]][3] for x in regional_data[&quot;GWAS&quot;].keys()]):r,})
    n_controls = c(${paths([regional_data[&#39;GWAS&#39;][x][regional_data[&#39;regions&#39;][_regions][0]][4] for x in regional_data[&quot;GWAS&quot;].keys()]):r,})
    LD_data = load_LD_matrix(&quot;${ld_meta_data}&quot;, &#39;${&quot;chr%s:%s-%s&quot; % (regional_data[&#39;regions&#39;][_regions][0], regional_data[&#39;regions&#39;][_regions][1], regional_data[&#39;regions&#39;][_regions][2])}&#39;) 
    res = setNames(replicate(length(studies), list(), simplify = FALSE), studies)
    for (r in 1:length(res)) {
        tryCatch({
        res[[r]] = rss_analysis_pipeline(sumstat_path = sumstat_paths[r], column_file_path = column_file_paths[r], 
                                        LD_data = LD_data, n_sample = as.numeric(n_samples[r]), n_case = as.numeric(n_cases[r]), 
                                        n_control = as.numeric(n_controls[r]), skip_region = skip_region,
                                        qc_method = ${&quot;NULL&quot; if len(qc_method) == 0 else &quot;&#39;%s&#39;&quot; % qc_method},
                                        impute = ${&quot;TRUE&quot; if impute else &quot;FALSE&quot;},
                                        impute_opts = list(rcond = ${rcond}, R2_threshold = ${R2_threshold}, minimum_ld = ${minimum_ld}, lamb=${lamb}),
                                        finemapping_method = ${&quot;NULL&quot; if len(finemapping_method) == 0 else &quot;&#39;%s&#39;&quot; % finemapping_method}, 
                                        finemapping_opts = list(init_L = ${L}, max_L = ${max_L}, l_step = ${l_step}, coverage = c(${&quot;,&quot;.join([str(x) for x in coverage])}), signal_cutoff = ${pip_cutoff}),
                                        pip_cutoff_to_skip = ${skip_analysis_pip_cutoff} 
                                        )
        fwrite(res[[r]]$rss_data_analyzed, file = paste0(&quot;${_output:nn}.&quot;, studies[r], &quot;.sumstats.tsv.gz&quot;), sep = &quot;\t&quot;, col.names = TRUE, row.names = FALSE, quote = FALSE, compress = &quot;gzip&quot;)
        if(length(res[[r]]) == 1){
            # the analysis is skipped because of pip_cutoff_to_skip  
            res[[r]] = NULL
        }
      }, error = function(e) {
        res[[r]] = NULL
        cat(paste(&quot;Error processing file &quot;, studies[r], &quot;: &quot;, conditionMessage(e), &quot;\n&quot;))
      }
      )}
    saveRDS(res, file = &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[univariate_plot]
output: pip_plot = f&quot;{cwd}/{_input:bn}.png&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = &#39;12h&#39;, mem = &#39;20G&#39;, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint = entrypoint
    res = readRDS(${_input:r})
    png(${_output[0]:r}, width = 14, height=6, unit=&#39;in&#39;, res=300)
    par(mfrow=c(1,2))
    susieR::susie_plot(res, y= &quot;PIP&quot;, pos=list(attr=&#39;pos&#39;,start=res$pos[1],end=res$pos[length(res$pos)]), add_legend=T, xlab=&quot;position&quot;)
    susieR::susie_plot(res, y= &quot;z&quot;, pos=list(attr=&#39;pos&#39;,start=res$pos[1],end=res$pos[length(res$pos)]), add_legend=T, xlab=&quot;position&quot;, ylab=&quot;-log10(p)&quot;)
    dev.off()
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/mnm_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="mrmash.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Multivariate TWAS prediction model</p>
      </div>
    </a>
    <a class="right-next"
       href="../prototype_drafts/polyfun.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Fine-mapping with PolyFun</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>