
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Univariate Fine-Mapping and TWAS with SuSiE &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/mnm_analysis/univariate_fine_mapping_twas_vignette';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Multivariate Fine-Mapping for multiple genes" href="multivariate_multigene_fine_mapping_vignette.html" />
    <link rel="prev" title="Integrative Analysis with High-Dimensional Regression" href="mnm_miniprotocol.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL Modifier Score</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_modifier_score/ems_training.html">xQTL Model Training Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../xqtl_modifier_score/ems_prediction.html">EMS Prediction for xQTL Modifier Scores</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/mnm_analysis/univariate_fine_mapping_twas_vignette.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/mnm_analysis/univariate_fine_mapping_twas_vignette.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/mnm_analysis/univariate_fine_mapping_twas_vignette.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Univariate Fine-Mapping and TWAS with SuSiE</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-run-the-fine-mapping-and-twas-with-susie">i. Run the Fine-Mapping and TWAS with SuSiE</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#anticipated-results">Anticipated Results</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="univariate-fine-mapping-and-twas-with-susie">
<h1>Univariate Fine-Mapping and TWAS with SuSiE<a class="headerlink" href="#univariate-fine-mapping-and-twas-with-susie" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>Our pipeline is capable of performing univariate fine-mapping with SuSiE with TWAS weights. The TWAS portion makes use of TWAS weights, linkage disequilibrium data and GWAS summary statistics. Preset variants used are taken from the linkage disequilibrium data and used only for TWAS. TWAS cross validation tell us which of the four methods (enet, lasso, mrash, SuSiE) are best to use. By default, we limit to under 5000 variants for cross validation. In cross validation, the data is split into five parts. Training is done on four parts, and prediction is done on the fifth. Linear regression is used to assess the results and get r squared and pvalues.</p>
<p>Fine mapping with SuSiE follows the formulat y=xb+e where x has many highly correlated variables due to linkage disequilibrium. Therefore, true effects (b), are very sparse. The SuSiE wrapper looks for five independent signals in each region to increase convergence speed. However, if five signals are found, then the the upper limit is increased. SuSiE does not allow for the inclusion of covariates. Therefore, covariates are regressed in.</p>
</section>
<section id="input">
<h2>Input:<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">--genoFile</span></code>: path to a plink bed file containing genotypes.<br />
<code class="docutils literal notranslate"><span class="pre">--phenoFile</span></code>: at least one tab delimited file containing chr, start, end, ID and path for the regions. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#chr    start   end     ID      path
chr12   389272  389273  ENSG00000120647 $PATH/Ast.log2cpm.bed.gz
chr12   389319  389320  ENSG00000073614 $PATH/Ast.log2cpm.bed.gz
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--covFile</span></code>: at least one tab delimited file containing covariates in the rows and samples in the columns.<br />
<code class="docutils literal notranslate"><span class="pre">--customized-association-windows</span></code>: a tab delimited file containing chr, start, end, and gene_id. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end     gene_id</span>
<span class="n">chr1</span>    <span class="mi">0</span>       <span class="mi">6480000</span> <span class="n">ENSG00000008128</span>
<span class="n">chr1</span>    <span class="mi">0</span>       <span class="mi">6480000</span> <span class="n">ENSG00000008130</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">----phenotype-names</span></code>: the names of the phenotypes if multiple are included. There should be one for each phenotype file you include.<br />
<code class="docutils literal notranslate"><span class="pre">--max-cv-variants</span></code>: maximum number of variants for cross-validation.<br />
<code class="docutils literal notranslate"><span class="pre">--ld_reference_meta_file</span></code>: path to file containing chrom, start, end and path for linkage disequilibrium region information. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#chrom  start   end     path
chr1    101384274       104443097       $PATH/chr1_101384274_104443097.cor.xz.bim
chr1    104443097       106225286       $PATH/chr1_104443097_106225286.cor.xz.bim
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--region-name</span></code>: if you only wish to analyze one region, then include the gene_id of a region found in the <code class="docutils literal notranslate"><span class="pre">customized-association-windows</span></code> file</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.univariate_bvrs.rds</span></code></p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; str(readRDS(&quot;/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_protocol/output/mnm_regression/susie_twas/fine_mapping/test_susie_twas.chr11_ENSG00000073921.univariate_bvsr.rds&quot;))
List of 1
 $ ENSG00000073921:List of 1
  ..$ test_pheno_ENSG00000073921:List of 10
  .. ..$ variant_names         : chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. ..$ analysis_script       : chr &quot;options(warn=1)\nlibrary(pecotmr)\nphenotype_files = c(\&quot;/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_prot&quot;| __truncated__
  .. ..$ other_quantities      :List of 1
  .. .. ..$ dropped_samples:List of 3
  .. .. .. ..$ X    : NULL
  .. .. .. ..$ y    : NULL
  .. .. .. ..$ covar: NULL
  .. ..$ sumstats              :List of 5
  .. .. ..$ betahat  : Named num [1:9970] 0.00479 0.00573 0.09947 0.02035 0.02035 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ sebetahat: Named num [1:9970] 0.0563 0.0568 0.0791 0.0551 0.0551 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ z_scores : Named num [1:9970] 0.085 0.101 1.258 0.369 0.369 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ p_values : Named num [1:9970] 0.932 0.92 0.208 0.712 0.712 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ q_values : Named num [1:9970] 0.922 0.92 0.633 0.859 0.859 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. ..$ sample_names          : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. ..$ top_loci              :&#39;data.frame&#39;:	3 obs. of  9 variables:
  .. .. ..$ variant_id      : chr [1:3] &quot;chr11:85179671:G:A&quot; &quot;chr11:85266101:G:C&quot; &quot;chr11:85314573:G:A&quot;
  .. .. ..$ betahat         : num [1:3] -0.743 -0.747 -0.747
  .. .. ..$ sebetahat       : num [1:3] 0.147 0.147 0.147
  .. .. ..$ z               : num [1:3] -5.06 -5.1 -5.1
  .. .. ..$ maf             : num [1:3] 0.0235 0.0233 0.0233
  .. .. ..$ pip             : num [1:3] 0.292 0.349 0.349
  .. .. ..$ cs_coverage_0.95: num [1:3] 1 1 1
  .. .. ..$ cs_coverage_0.7 : num [1:3] 1 1 1
  .. .. ..$ cs_coverage_0.5 : num [1:3] 0 1 1
  .. ..$ susie_result_trimmed  :List of 12
  .. .. ..$ pip                   : num [1:9970] 0.000431 0.00043 0.000673 0.000431 0.000431 ...
  .. .. ..$ sets                  :List of 5
  .. .. .. ..$ cs                :List of 1
  .. .. .. .. ..$ L1: int [1:3] 653 910 1069
  .. .. .. ..$ purity            :&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. .. ..$ min.abs.corr   : num 1
  .. .. .. .. ..$ mean.abs.corr  : num 1
  .. .. .. .. ..$ median.abs.corr: num 1
  .. .. .. ..$ cs_index          : int 1
  .. .. .. ..$ coverage          : num 0.989
  .. .. .. ..$ requested_coverage: num 0.95
  .. .. ..$ cs_corr               : logi NA
  .. .. ..$ sets_secondary        :List of 2
  .. .. .. ..$ coverage_0.7:List of 2
  .. .. .. .. ..$ sets   :List of 5
  .. .. .. .. .. ..$ cs                :List of 1
  .. .. .. .. .. .. ..$ L1: int [1:3] 653 910 1069
  .. .. .. .. .. ..$ purity            :&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. .. .. .. ..$ min.abs.corr   : num 1
  .. .. .. .. .. .. ..$ mean.abs.corr  : num 1
  .. .. .. .. .. .. ..$ median.abs.corr: num 1
  .. .. .. .. .. ..$ cs_index          : int 1
  .. .. .. .. .. ..$ coverage          : num 0.989
  .. .. .. .. .. ..$ requested_coverage: num 0.7
  .. .. .. .. ..$ cs_corr: logi NA
  .. .. .. ..$ coverage_0.5:List of 2
  .. .. .. .. ..$ sets   :List of 5
  .. .. .. .. .. ..$ cs                :List of 1
  .. .. .. .. .. .. ..$ L1: int [1:2] 910 1069
  .. .. .. .. .. ..$ purity            :&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. .. .. .. ..$ min.abs.corr   : num 1
  .. .. .. .. .. .. ..$ mean.abs.corr  : num 1
  .. .. .. .. .. .. ..$ median.abs.corr: num 1
  .. .. .. .. .. ..$ cs_index          : int 1
  .. .. .. .. .. ..$ coverage          : num 0.697
  .. .. .. .. .. ..$ requested_coverage: num 0.5
  .. .. .. .. ..$ cs_corr: logi NA
  .. .. ..$ alpha                 : num [1:8, 1:9970] 1.52e-07 6.35e-05 6.05e-05 5.93e-05 5.99e-05 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : NULL
  .. .. .. .. ..$ : chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ lbf_variable          : num [1:8, 1:9970] -1.697 -0.402 -0.437 -0.451 -0.444 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : NULL
  .. .. .. .. ..$ : chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ V                     : num [1:8] 0.019 0.00097 0.0011 0.00115 0.00112 ...
  .. .. ..$ niter                 : int 10
  .. .. ..$ max_L                 : int 8
  .. .. ..$ X_column_scale_factors: Named num [1:9970] 0.517 0.512 0.366 0.528 0.528 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ mu                    : num [1:8, 1:9970] 0.00227 0.00643 0.00672 0.00683 0.00677 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : NULL
  .. .. .. .. ..$ : chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..$ mu2                   : num [1:8, 1:9970] 0.000638 0.000432 0.000455 0.000464 0.00046 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : NULL
  .. .. .. .. ..$ : chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  .. .. ..- attr(*, &quot;class&quot;)= chr &quot;susie&quot;
  .. ..$ total_time_elapsed    : &#39;proc_time&#39; Named num [1:5] 32.189 0.339 43.517 0 0
  .. .. ..- attr(*, &quot;names&quot;)= chr [1:5] &quot;user.self&quot; &quot;sys.self&quot; &quot;elapsed&quot; &quot;user.child&quot; ...
  .. ..$ region_info           :List of 3
  .. .. ..$ region_coord:&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. ..$ chrom: chr &quot;11&quot;
  .. .. .. ..$ start: int 85957174
  .. .. .. ..$ end  : int 86069881
  .. .. ..$ grange      :&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. ..$ chrom: chr &quot;11&quot;
  .. .. .. ..$ start: int 84957175
  .. .. .. ..$ end  : int 87360000
  .. .. ..$ region_name : chr [1:2] &quot;ENSG00000073921&quot; &quot;ENSG00000073921&quot;
  .. ..$ preset_variants_result:List of 9
  .. .. ..$ susie_fitted        :List of 22
  .. .. .. ..$ alpha                 : num [1:8, 1:6454] 0.00015 0.000148 0.000148 0.000148 0.000149 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ mu                    : num [1:8, 1:6454] 0.000134 0.000169 0.000177 0.000169 0.000152 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ mu2                   : num [1:8, 1:6454] 5.39e-05 6.83e-05 7.19e-05 6.85e-05 6.13e-05 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ Xr                    : num [1:150] 0.000696 -0.000919 -0.002423 -0.00093 0.000305 ...
  .. .. .. ..$ KL                    : num [1:8] 0.0345 0.0445 0.047 0.0446 0.0395 ...
  .. .. .. ..$ lbf                   : num [1:8] 0.000194 0.000317 0.000353 0.000319 0.000253 ...
  .. .. .. ..$ lbf_variable          : num [1:8, 1:6454] -0.0334 -0.0427 -0.045 -0.0428 -0.0381 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ sigma2                : num 0.124
  .. .. .. ..$ V                     : num [1:8] 5.76e-05 7.44e-05 7.87e-05 7.46e-05 6.61e-05 ...
  .. .. .. ..$ pi                    : num [1:6454] 0.000155 0.000155 0.000155 0.000155 0.000155 ...
  .. .. .. ..$ null_index            : num 0
  .. .. .. ..$ correct_zR_discrepancy:List of 5
  .. .. .. .. ..$ to_correct          : logi FALSE
  .. .. .. .. ..$ outlier_index       : logi(0) 
  .. .. .. .. ..$ is_init             : logi TRUE
  .. .. .. .. ..$ outlier_stabilize   : num 5
  .. .. .. .. ..$ outlier_stable_count: num 0
  .. .. .. ..$ force_iterate         : logi FALSE
  .. .. .. ..$ converged             : logi TRUE
  .. .. .. ..$ elbo                  : num [1:3] -56.5 -56.5 -56.5
  .. .. .. ..$ niter                 : int 3
  .. .. .. ..$ intercept             : num -0.00273
  .. .. .. ..$ fitted                : Named num [1:150] 0.000696 -0.000919 -0.002423 -0.00093 0.000305 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. ..$ sets                  :List of 3
  .. .. .. .. ..$ cs                : NULL
  .. .. .. .. ..$ coverage          : NULL
  .. .. .. .. ..$ requested_coverage: num 0.95
  .. .. .. ..$ pip                   : Named num [1:6454] 0.0012 0.00124 0.0012 0.0012 0.0012 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ X_column_scale_factors: Named num [1:6454] 0.664 0.438 0.677 0.702 0.677 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ time_elapsed          : &#39;proc_time&#39; Named num [1:5] 0.851 0.008 0.861 0 0
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:5] &quot;user.self&quot; &quot;sys.self&quot; &quot;elapsed&quot; &quot;user.child&quot; ...
  .. .. .. ..- attr(*, &quot;class&quot;)= chr &quot;susie&quot;
  .. .. ..$ variant_names       : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. ..$ analysis_script     : chr &quot;options(warn=1)\nlibrary(pecotmr)\nphenotype_files = c(\&quot;/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_prot&quot;| __truncated__
  .. .. ..$ other_quantities    :List of 1
  .. .. .. ..$ dropped_samples:List of 3
  .. .. .. .. ..$ X    : NULL
  .. .. .. .. ..$ y    : NULL
  .. .. .. .. ..$ covar: NULL
  .. .. ..$ sumstats            :List of 5
  .. .. .. ..$ betahat  : Named num [1:6454] 0.0029 0.06935 0.01237 -0.00911 0.01237 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ sebetahat: Named num [1:6454] 0.0438 0.0661 0.0429 0.0414 0.0429 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ z_scores : Named num [1:6454] 0.0662 1.0485 0.288 -0.22 0.288 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ p_values : Named num [1:6454] 0.947 0.294 0.773 0.826 0.773 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ q_values : Named num [1:6454] 0.998 0.892 0.991 0.998 0.991 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. ..$ sample_names        : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. ..$ susie_result_trimmed:List of 12
  .. .. .. ..$ pip                   : num [1:6454] 0.0012 0.00124 0.0012 0.0012 0.0012 ...
  .. .. .. ..$ sets                  :List of 3
  .. .. .. .. ..$ cs                : NULL
  .. .. .. .. ..$ coverage          : NULL
  .. .. .. .. ..$ requested_coverage: num 0.95
  .. .. .. ..$ cs_corr               : logi NA
  .. .. .. ..$ sets_secondary        :List of 2
  .. .. .. .. ..$ coverage_0.7:List of 2
  .. .. .. .. .. ..$ sets   :List of 3
  .. .. .. .. .. .. ..$ cs                : NULL
  .. .. .. .. .. .. ..$ coverage          : NULL
  .. .. .. .. .. .. ..$ requested_coverage: num 0.7
  .. .. .. .. .. ..$ cs_corr: logi NA
  .. .. .. .. ..$ coverage_0.5:List of 2
  .. .. .. .. .. ..$ sets   :List of 3
  .. .. .. .. .. .. ..$ cs                : NULL
  .. .. .. .. .. .. ..$ coverage          : NULL
  .. .. .. .. .. .. ..$ requested_coverage: num 0.5
  .. .. .. .. .. ..$ cs_corr: logi NA
  .. .. .. ..$ alpha                 : num [1:8, 1:6454] 0.00015 0.000148 0.000148 0.000148 0.000149 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ lbf_variable          : num [1:8, 1:6454] -0.0334 -0.0427 -0.045 -0.0428 -0.0381 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ V                     : num [1:8] 5.76e-05 7.44e-05 7.87e-05 7.46e-05 6.61e-05 ...
  .. .. .. ..$ niter                 : int 3
  .. .. .. ..$ max_L                 : int 8
  .. .. .. ..$ X_column_scale_factors: Named num [1:6454] 0.664 0.438 0.677 0.702 0.677 ...
  .. .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ mu                    : num [1:8, 1:6454] 0.000134 0.000169 0.000177 0.000169 0.000152 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..$ mu2                   : num [1:8, 1:6454] 5.39e-05 6.83e-05 7.19e-05 6.85e-05 6.13e-05 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. ..- attr(*, &quot;class&quot;)= chr &quot;susie&quot;
  .. .. ..$ total_time_elapsed  : &#39;proc_time&#39; Named num [1:5] 10.991 0.012 11.046 0 0
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:5] &quot;user.self&quot; &quot;sys.self&quot; &quot;elapsed&quot; &quot;user.child&quot; ...
  .. .. ..$ region_info         :List of 3
  .. .. .. ..$ region_coord:&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. .. ..$ chrom: chr &quot;11&quot;
  .. .. .. .. ..$ start: int 85957174
  .. .. .. .. ..$ end  : int 86069881
  .. .. .. ..$ grange      :&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. .. ..$ chrom: chr &quot;11&quot;
  .. .. .. .. ..$ start: int 84957175
  .. .. .. .. ..$ end  : int 87360000
  .. .. .. ..$ region_name : chr [1:2] &quot;ENSG00000073921&quot; &quot;ENSG00000073921&quot;
</pre></div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.univariate_data.rds</span></code></p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; str(readRDS(&quot;output/mnm_regression/susie_twas/data/test_susie_twas.chr11_ENSG00000073921.univariate_data.rds&quot;))
List of 1
 $ ENSG00000073921:List of 10
  ..$ residual_Y       :List of 1
  .. ..$ test_pheno: num [1:150, 1] 0.1241 0.2981 -0.6054 0.0435 0.0388 ...
  .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  ..$ residual_X       :List of 1
  .. ..$ : num [1:150, 1:9970] -0.2386 -0.8293 0.4173 0.7408 -0.0231 ...
  .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. ..$ : chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  ..$ residual_Y_scalar: num 1
  ..$ residual_X_scalar: num 1
  ..$ dropped_sample   :List of 3
  .. ..$ X    :List of 1
  .. .. ..$ : chr(0) 
  .. ..$ Y    :List of 1
  .. .. ..$ : chr &quot;strand&quot;
  .. ..$ covar:List of 1
  .. .. ..$ : chr(0) 
  ..$ maf              :List of 1
  .. ..$ : Named num [1:9970] 0.33 0.327 0.103 0.343 0.343 ...
  .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  ..$ X                : num [1:150, 1:9970] 0 0 1 1 1 0 0 1 1 1 ...
  .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. ..$ : chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
  ..$ chrom            : chr &quot;chr11&quot;
  ..$ grange           : chr [1:2] &quot;85957174&quot; &quot;86069881&quot;
  ..$ X_variance       :List of 1
  .. ..$ : Named num [1:9970] 0.267 0.262 0.134 0.279 0.279 ...
  .. .. ..- attr(*, &quot;names&quot;)= chr [1:9970] &quot;chr11:84957209:G:C&quot; &quot;chr11:84957879:A:G&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; ...
</pre></div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.univariate_twas_weights.rds</span></code></p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; str(readRDS(&quot;/restricted/projectnb/xqtl/xqtl_protocol/toy_xqtl_protocol/output/mnm_regression/susie_twas/twas_weights/test_susie_twas.chr11_ENSG00000073921.univariate_twas_weights.rds&quot;))
List of 1
 $ ENSG00000073921:List of 1
  ..$ test_pheno_ENSG00000073921:List of 7
  .. ..$ susie_weights_intermediate:List of 4
  .. .. ..$ mu                    : num [1:8, 1:6454] 0.000134 0.000169 0.000177 0.000169 0.000152 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : NULL
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. ..$ lbf_variable          : num [1:8, 1:6454] -0.0334 -0.0427 -0.045 -0.0428 -0.0381 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : NULL
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. ..$ X_column_scale_factors: Named num [1:6454] 0.664 0.438 0.677 0.702 0.677 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. ..$ pip                   : Named num [1:6454] 0.0012 0.00124 0.0012 0.0012 0.0012 ...
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. ..$ twas_weights              :List of 6
  .. .. ..$ enet_weights   : num [1:6454, 1] 0 0 0 0 0 0 0 0 0 0 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ lasso_weights  : num [1:6454, 1] 0 0 0 0 0 0 0 0 0 0 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ bayes_r_weights: num [1:6454, 1] 4.31e-05 4.73e-04 2.44e-04 -8.48e-05 2.30e-04 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ bayes_l_weights: num [1:6454, 1] 2.04e-05 3.66e-04 1.03e-04 -1.54e-04 3.47e-04 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ mrash_weights  : num [1:6454, 1] 6.01e-05 2.05e-04 7.02e-05 -4.71e-05 7.02e-05 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ susie_weights  : num [1:6454, 1] 2.55e-07 5.89e-06 1.00e-06 -7.53e-07 1.00e-06 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. ..$ twas_predictions          :List of 6
  .. .. ..$ enet_predicted   : num [1:150, 1] 0 0 0 0 0 0 0 0 0 0 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ lasso_predicted  : num [1:150, 1] 0 0 0 0 0 0 0 0 0 0 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ bayes_r_predicted: num [1:150, 1] 0.1971 0.1333 -0.0176 0.1234 0.2194 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ bayes_l_predicted: num [1:150, 1] 0.20151 0.13565 -0.00533 0.13823 0.23318 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ mrash_predicted  : num [1:150, 1] 0.08183 0.05465 -0.00379 0.05346 0.09283 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ susie_predicted  : num [1:150, 1] 0.003423 0.001808 0.000304 0.001797 0.003032 ...
  .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. ..$ twas_cv_result            :List of 4
  .. .. ..$ sample_partition:&#39;data.frame&#39;:	150 obs. of  2 variables:
  .. .. .. ..$ Sample: chr [1:150] &quot;sample108&quot; &quot;sample52&quot; &quot;sample103&quot; &quot;sample29&quot; ...
  .. .. .. ..$ Fold  : int [1:150] 1 1 1 1 1 1 1 1 1 1 ...
  .. .. ..$ prediction      :List of 4
  .. .. .. ..$ bayes_r_predicted: num [1:150, 1] 0.3606 -0.0263 0.0687 0.0989 0.1158 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. .. ..$ bayes_l_predicted: num [1:150, 1] 0.2274 -0.0389 0.0379 0.0513 0.0706 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. .. ..$ mrash_predicted  : num [1:150, 1] 0.0692 -0.0021 0.0183 0.0231 0.0353 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. .. ..$ susie_predicted  : num [1:150, 1] 0.05794 -0.00105 -0.01131 0 0 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr [1:150] &quot;sample0&quot; &quot;sample1&quot; &quot;sample10&quot; &quot;sample100&quot; ...
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. ..$ performance     :List of 4
  .. .. .. ..$ bayes_r_performance: num [1, 1:6] 0.022483 0.000505 -0.006248 0.784784 0.40514 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. .. .. .. ..$ : chr [1:6] &quot;corr&quot; &quot;rsq&quot; &quot;adj_rsq&quot; &quot;pval&quot; ...
  .. .. .. ..$ bayes_l_performance: num [1, 1:6] -8.04e-03 6.47e-05 -6.69e-03 9.22e-01 3.91e-01 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. .. .. .. ..$ : chr [1:6] &quot;corr&quot; &quot;rsq&quot; &quot;adj_rsq&quot; &quot;pval&quot; ...
  .. .. .. ..$ mrash_performance  : num [1, 1:6] 0.05206 0.00271 -0.00403 0.52697 0.35394 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. .. .. .. ..$ : chr [1:6] &quot;corr&quot; &quot;rsq&quot; &quot;adj_rsq&quot; &quot;pval&quot; ...
  .. .. .. ..$ susie_performance  : num [1, 1:6] -0.03878 0.0015 -0.00524 0.6375 0.35463 ...
  .. .. .. .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. .. .. .. ..$ : chr &quot;ENSG00000073921&quot;
  .. .. .. .. .. ..$ : chr [1:6] &quot;corr&quot; &quot;rsq&quot; &quot;adj_rsq&quot; &quot;pval&quot; ...
  .. .. ..$ time_elapsed    : &#39;proc_time&#39; Named num [1:5] 188.023 0.581 189.175 0 0
  .. .. .. ..- attr(*, &quot;names&quot;)= chr [1:5] &quot;user.self&quot; &quot;sys.self&quot; &quot;elapsed&quot; &quot;user.child&quot; ...
  .. ..$ total_time_elapsed        : &#39;proc_time&#39; Named num [1:5] 242.761 0.972 249.471 0.013 0.024
  .. .. ..- attr(*, &quot;names&quot;)= chr [1:5] &quot;user.self&quot; &quot;sys.self&quot; &quot;elapsed&quot; &quot;user.child&quot; ...
  .. ..$ variant_names             : chr [1:6454] &quot;chr11:84957209:G:C&quot; &quot;chr11:84958236:C:A&quot; &quot;chr11:84958954:T:C&quot; &quot;chr11:84959602:T:C&quot; ...
  .. ..$ region_info               :List of 3
  .. .. ..$ region_coord:&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. ..$ chrom: chr &quot;11&quot;
  .. .. .. ..$ start: int 85957174
  .. .. .. ..$ end  : int 86069881
  .. .. ..$ grange      :&#39;data.frame&#39;:	1 obs. of  3 variables:
  .. .. .. ..$ chrom: chr &quot;11&quot;
  .. .. .. ..$ start: int 84957175
  .. .. .. ..$ end  : int 87360000
  .. .. ..$ region_name : chr [1:2] &quot;ENSG00000073921&quot; &quot;ENSG00000073921&quot;
</pre></div>
</div>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<section id="i-run-the-fine-mapping-and-twas-with-susie">
<h3>i. <a class="reference external" href="https://statfungen.github.io/xqtl-protocol/code/mnm_analysis/mnm_methods/mnm_regression.html#i-susie-with-twas-weights-from-multiple-methods">Run the Fine-Mapping and TWAS with SuSiE</a><a class="headerlink" href="#i-run-the-fine-mapping-and-twas-with-susie" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">test_susie_twas</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">genotype_by_chrom</span><span class="o">/</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="mf">.1</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">phenotype_by_chrom_for_cis</span><span class="o">/</span><span class="n">bulk_rnaseq</span><span class="o">.</span><span class="n">phenotype_by_chrom_files</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">bulk_rnaseq_tmp_matrix</span><span class="o">.</span><span class="n">low_expression_filtered</span><span class="o">.</span><span class="n">outlier_removed</span><span class="o">.</span><span class="n">tmm</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">covariates</span><span class="o">.</span><span class="n">wgs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">association</span><span class="o">-</span><span class="n">windows</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">TAD</span><span class="o">/</span><span class="n">TADB_enhanced_cis</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">test_pheno</span> \
    <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">cv</span><span class="o">-</span><span class="n">variants</span> <span class="mi">5000</span> <span class="o">--</span><span class="n">ld_reference_meta_file</span> <span class="n">data</span><span class="o">/</span><span class="n">ld_meta_file_with_bim</span><span class="o">.</span><span class="n">tsv</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="n">name</span> <span class="n">ENSG00000049246</span> <span class="n">ENSG00000054116</span> <span class="n">ENSG00000116678</span> \
    <span class="o">--</span><span class="n">save</span><span class="o">-</span><span class="n">data</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">/</span><span class="n">susie_twas</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">get_analysis_regions</span><span class="p">:</span> 
<span class="n">Loading</span> <span class="n">customized</span> <span class="n">association</span> <span class="n">analysis</span> <span class="n">window</span> <span class="kn">from</span><span class="w"> </span><span class="nn">reference_data</span><span class="o">/</span><span class="n">TAD</span><span class="o">/</span><span class="n">TADB_enhanced_cis</span><span class="o">.</span><span class="n">bed</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">get_analysis_regions</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">get_analysis_regions</span> <span class="n">output</span><span class="p">:</span>   <span class="n">regional_data</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Running</span> <span class="n">susie_twas</span><span class="p">:</span> 
<span class="n">INFO</span><span class="p">:</span> <span class="n">susie_twas</span> <span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">susie_twas</span> <span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">susie_twas</span> <span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="ow">is</span> <span class="n">completed</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">susie_twas</span> <span class="n">output</span><span class="p">:</span>   <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">/</span><span class="n">susie_twas</span><span class="o">/</span><span class="n">fine_mapping</span><span class="o">/</span><span class="n">test_susie_twas</span><span class="o">.</span><span class="n">chr1_ENSG00000049246</span><span class="o">.</span><span class="n">univariate_bvsr</span><span class="o">.</span><span class="n">rds</span> <span class="o">/</span><span class="n">restricted</span><span class="o">/</span><span class="n">projectnb</span><span class="o">/</span><span class="n">xqtl</span><span class="o">/</span><span class="n">xqtl_protocol</span><span class="o">/</span><span class="n">toy_xqtl_protocol</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">/</span><span class="n">susie_twas</span><span class="o">/</span><span class="n">twas_weights</span><span class="o">/</span><span class="n">test_susie_twas</span><span class="o">.</span><span class="n">chr1_ENSG00000049246</span><span class="o">.</span><span class="n">univariate_twas_weights</span><span class="o">.</span><span class="n">rds</span><span class="o">...</span> <span class="p">(</span><span class="mi">6</span> <span class="n">items</span> <span class="ow">in</span> <span class="mi">3</span> <span class="n">groups</span><span class="p">)</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Workflow</span> <span class="n">susie_twas</span> <span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="n">w487c342397d09ccb</span><span class="p">)</span> <span class="ow">is</span> <span class="n">executed</span> <span class="n">successfully</span> <span class="k">with</span> <span class="mi">2</span> <span class="n">completed</span> <span class="n">steps</span> <span class="ow">and</span> <span class="mi">4</span> <span class="n">completed</span> <span class="n">substeps</span><span class="o">.</span>
</pre></div>
</div>
</section>
</section>
<section id="anticipated-results">
<h2>Anticipated Results<a class="headerlink" href="#anticipated-results" title="Link to this heading">#</a></h2>
<p>Univariate finemapping will produce a file containing results for the top hits and a file containing twas weights produced by susie.</p>
<p><code class="docutils literal notranslate"><span class="pre">ROSMAP_mega_eQTL.chr11_ENSG00000073921.univariate_bvrs.rds</span></code>:</p>
<ul class="simple">
<li><p>For each gene of interest and phenotype, this file contains:</p>
<ol class="arabic simple">
<li><p>susie_fitted</p></li>
<li><p>variant_names</p></li>
<li><p>analysis_script</p></li>
<li><p>other quantities - information on dropped samples, if any</p></li>
<li><p>summary statistics (beta and se) for each variant</p></li>
<li><p>sample names</p></li>
<li><p>summary statistics for top loci</p></li>
<li><p>susie results trimmed - includes pip, sets, cs_corr, sets_secondary, alpha, lbf_variable, V, niter, X_column_scale_factors, mu, mu2</p></li>
<li><p>total time elapsed</p></li>
<li><p>region info</p></li>
<li><p>preset_variants_results</p></li>
</ol>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">ROSMAP_mega_eQTL.chr11_ENSG00000073921.univariate_data.rds</span></code> (from the –save-data argument):</p>
<ul class="simple">
<li><p>For each gene of interest, contains residuals for each sample and phenotype</p></li>
<li><p>see <a class="reference external" href="https://github.com/statfungen/pecotmr/blob/68d87ca1d0a059022bf4e55339621cbddc8993cc/R/file_utils.R#L461">pecotmr code</a> for description at <code class="docutils literal notranslate"><span class="pre">load_regional_association_data</span></code> function</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">ROSMAP_mega_eQTL.chr11_ENSG00000073921.univariate_twas_weights.rds</span></code>:</p>
<ul class="simple">
<li><p>For each gene of interest and phenotype, this file contains:</p>
<ol class="arabic simple">
<li><p>twas_weights - weights for enet, lasso, bayes r, mrash and susie methods</p></li>
<li><p>twas_predictions - twas predictions for enet, lasso, bayes r, mrash and susie methods</p></li>
<li><p>twas_cv_result</p></li>
<li><p>total_time_elapsed</p></li>
<li><p>variant_names</p></li>
<li><p>region_info</p></li>
</ol>
</li>
</ul>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./code/mnm_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="mnm_miniprotocol.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Integrative Analysis with High-Dimensional Regression</p>
      </div>
    </a>
    <a class="right-next"
       href="multivariate_multigene_fine_mapping_vignette.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Multivariate Fine-Mapping for multiple genes</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-run-the-fine-mapping-and-twas-with-susie">i. Run the Fine-Mapping and TWAS with SuSiE</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#anticipated-results">Anticipated Results</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>