{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# TWAS, cTWAS and MR\n",
    "\n",
    "## Introduction\n",
    "\n",
    "This module provides software implementations for transcriptome-wide association analysis (TWAS), Quantile TWAS and performs variant selection for providing sparse signals for cTWAS (causal TWAS) analysis as described in Qian et al (2024+) the multi-group cTWAS method. It will additionally perform Mendelian Randomization using fine-mapping instrumental variables (IV) as described in Zhang et al 2020 for \"causal\" effects estimation and model validation, with the unit of analysis being a single gene-trait pair.\n",
    "\n",
    "This procedure is a continuation of the SuSiE-TWAS workflow --- it assumes that xQTL fine-mapping has been performed and moleuclar traits prediction weights pre-computed (to be used for TWAS). Cross validation for TWAS weights is optional but highly recommended.\n",
    "\n",
    "Quantile TWAS extends traditional TWAS by testing genetic effects at different quantiles of the trait distribution, which provides insights into genetic associations that vary across the distribution rather than just at the mean.\n",
    "\n",
    "GWAS data required are GWAS summary statistics and LD matrix for the region of interest.\n",
    "\n",
    "### Step 1: TWAS \n",
    "\n",
    "1. Extract GWAS z-score for region of interest and corresponding LD matrix.\n",
    "2. (Optional) perform allele matching QC for the LD matrix with summary stats.\n",
    "3. Process weights: for a number of methods such as LASSO, Elastic Net and mr.ash we have to take the weights as is for QTL variants overlapping with GWAS variants. For SuSiE weights it can be adjusted to exactly match GWAS variants.\n",
    "4. Perofrm TWAS test for multiple sets of weights. \n",
    "5. For each gene, filter TWAS results by keeping the best model selected by CV. Drop the genes that don't show good evidence of TWAS prediction weights.\n",
    "\n",
    "### Step 2: Variant Selection for Imputable Genes via the Best Prediction Methods\n",
    "1. Determine if the gene is imputable at each context based on the twas_cv performance by adjusted $r^2$ (>=0.01) and p-values (<0.05).\n",
    "2. The imputable gene-context pair will go through variant selection step. Maximum 10 variants with top pip selected from either `top_loci` table or SuSiE CS set. \n",
    "3. Harmonize weights against LD reference and udpate SuSiE weight. \n",
    "4. Extract weights by best model for the context then by the variant names were selected from the previous step\n",
    "\n",
    "### Step 3: cTWAS analysis\n",
    "\n",
    "**FIXME: add more documentation here**\n",
    "\n",
    "### Step 4: MR for candidate genes\n",
    "\n",
    "1. Limit MR only to those showing some evidence of cTWAS significance AND have strong instrumental variable (fine-mapping PIP or CS). \n",
    "2. Use fine-mapped xQTL with GWAS data to perform MR. \n",
    "3. For multiple IV, aggregate individual IV estimates using a fixed-effect meta-analysis procedure.\n",
    "4. Identify and exclude results with severe violations of the exclusion restriction (ER) assumption.\n",
    "\n",
    "### Step 5: Quantile TWAS analysis\n",
    "- Use pre-computed TWAS weights (beta for now) for quantile-specific testing.\n",
    "- For each quantile level, cluster and integrate by fixed and dynamic region groups, and extract relevant GWAS z-scores and LD matrix for the region of interest.\n",
    "- Perform quantile region-specific association tests, identifying genetic variants with effects that vary across different quantile regions of the phenotype distribution.\n",
    "\n",
    "## Input\n",
    "\n",
    "### GWAS Data Input Interface (Similar to `susie_rss`)\n",
    "\n",
    "I. **GWAS Summary Statistics Files**\n",
    "- **Input**: Vector of files for one or more GWAS studies.\n",
    "- **Format**: \n",
    "  - Tab-delimited files that is tabix-indexed by the first `chrom`(or `#chrom`) and second `pos` column. \n",
    "  - First 4 columns: `chrom` or `#chrom`, `pos`, `A1`, `A2`\n",
    "  - Additional columns can be loaded using column mapping file see below  \n",
    "- **Column Mapping files (optional)**:\n",
    "  - Optional YAML file for custom column mapping.\n",
    "  - Required columns: `chrom`, `pos`, `A1`, `A2`, `z` or (`betahat` and `sebetahat`).\n",
    "  - Optional columns: `n`, `var_y` (relevant to fine-mapping).\n",
    "\n",
    "II. **GWAS Summary Statistics Meta-File**: this is optional and helpful when there are lots of GWAS data to process via the same command\n",
    "- **Columns**: `study_id`, chromosome number, path to summary statistics file, optional path to column mapping file.\n",
    "- **Note**: Chromosome number `0` indicates a genome-wide file.\n",
    "\n",
    "eg: `gwas_meta.tsv`\n",
    "\n",
    "```\n",
    "study_id    chrom    file_path                 column_mapping_file\n",
    "study1      1        gwas1.tsv.gz         column_mapping.yml\n",
    "study1      2        gwas2.tsv.gz         column_mapping.yml\n",
    "study2      0        gwas3.tsv.gz         column_mapping.yml\n",
    "```\n",
    "\n",
    "If both summary stats file (I) and meta data file (II) are specified we will take the union of the two.\n",
    "\n",
    "\n",
    "III. **LD Reference Metadata File**\n",
    "- **Format**: Single TSV file.\n",
    "- **Contents**:\n",
    "  - Columns: `#chrom`, `start`, `end`, path to the LD matrix, genomic build.\n",
    "  - LD matrix path format: comma-separated, first entry is the LD matrix, second is the bim file.\n",
    "- **Documentation**: Refer to [our LD reference preparation document](https://cumc.github.io/xqtl-protocol/code/reference_data/ld_reference_generation.html) for detailed information.\n",
    "\n",
    "### Output of Fine-Mapping & TWAS Pipeline\n",
    "\n",
    "**xQTL Weight Database Metadata File**: \n",
    "- **Essential columns**: `#chr`, `start`, `end`, `TSS`, `region_id`, `original_data`\n",
    "- **Structure of the weight database**: \n",
    "  - RDS format.\n",
    "  - Organized hierarchically: region → context → weight matrix.\n",
    "  - Each column represents a different method.\n",
    "\n",
    "eg: `xqtl_meta.tsv`\n",
    "\n",
    "```\n",
    "#chr start end region_id TSS original_data combined_data combined_data_sumstats contexts contexts_top_loci\n",
    "chr1 0 6480000 ENSG00000008128 1724356 \"KNIGHT_pQTL.ENSG00000008128.univariate_susie_twas_weights.rds, MiGA_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds, MSBB_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds, ROSMAP_Bennett_Klein_pQTL.ENSG00000008128.univariate_susie_twas_weights.rds, ROSMAP_DeJager_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds, ROSMAP_Kellis_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds, ROSMAP_mega_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds, STARNET_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds\" Fungen_xQTL.ENSG00000008128.cis_results_db.export.rds Fungen_xQTL.ENSG00000008128.cis_results_db.export_sumstats.rds Knight_eQTL_brain,MiGA_GFM_eQTL,MiGA_GTS_eQTL,MiGA_SVZ_eQTL,MiGA_THA_eQTL,BM_10_MSBB_eQTL,BM_22_MSBB_eQTL,BM_36_MSBB_eQTL,BM_44_MSBB_eQTL,monocyte_ROSMAP_eQTL,Mic_DeJager_eQTL,Ast_DeJager_eQTL,Oli_DeJager_eQTL,Exc_DeJager_eQTL,Inh_DeJager_eQTL,DLPFC_DeJager_eQTL,PCC_DeJager_eQTL,AC_DeJager_eQTL,Mic_Kellis_eQTL,Ast_Kellis_eQTL,Oli_Kellis_eQTL,OPC_Kellis_eQTL,Exc_Kellis_eQTL,Inh_Kellis_eQTL,Ast_mega_eQTL,Exc_mega_eQTL,Inh_mega_eQTL,Oli_mega_eQTL,STARNET_eQTL_Mac Knight_eQTL_brain,MiGA_GFM_eQTL,MiGA_GTS_eQTL,MiGA_SVZ_eQTL,MiGA_THA_eQTL,BM_10_MSBB_eQTL,BM_22_MSBB_eQTL,BM_36_MSBB_eQTL,BM_44_MSBB_eQTL,monocyte_ROSMAP_eQTL,Mic_DeJager_eQTL,Ast_DeJager_eQTL,Oli_DeJager_eQTL,Exc_DeJager_eQTL,Inh_DeJager_eQTL,DLPFC_DeJager_eQTL,PCC_DeJager_eQTL,AC_DeJager_eQTL,Mic_Kellis_eQTL,Ast_Kellis_eQTL,Oli_Kellis_eQTL,OPC_Kellis_eQTL,Exc_Kellis_eQTL,Inh_Kellis_eQTL,Ast_mega_eQTL,Exc_mega_eQTL,Inh_mega_eQTL,Oli_mega_eQTL,STARNET_eQTL_Mac\n",
    "```\n",
    "\n",
    "This file is automatically generated as part of the FunGen-xQTL protocol, although only the essential columns are relevant to our application here.\n",
    "\n",
    "\n",
    "### TWAS region information\n",
    "\n",
    "This is required for cTWAS analysis, where multiple TWAS and SNP data within each region are combined for joint inference to select the variables, either genes or SNPs, to figure out which variables are likely to be directly associated with the phenotype of interest, rather than being associated through correlations with true causal variables.\n",
    "\n",
    "```\n",
    "chrom    start    end    block_id  \n",
    "1        1000     5000   block1    \n",
    "2        2000     6000   block2\n",
    "3        3000     7000   block3\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## Output\n",
    "\n",
    "I. A table with the following contents\n",
    "\n",
    "```\n",
    "gwas_study, chrom, start, end, block, gene, TSS, context, is_imputable, method, is_selected_method, rsq_adj_cv, pval_cv, twas_z, twas_pval\n",
    "```\n",
    "\n",
    "where\n",
    "\n",
    "- if `twas_z` is `NA` it means the context is not imputable for the method of choice\n",
    "\n",
    "II. a list of  `refined_twas_weights_data` per block, in RDS format, of this structure:\n",
    "\n",
    "```\n",
    "$ region_id\n",
    "   $ gene \n",
    "      $ context\n",
    "        $ selected_model\n",
    "        $ is_imputable \n",
    "        $ selected_top_variants\n",
    "        $ selected_model_weights\n",
    "\n",
    "```\n",
    "\n",
    "This will only contain imputatable genes. It should come with a meta-data file like this:\n",
    "\n",
    "```\n",
    "chrom    start    end    block_id  refined_twas_db\n",
    "1        1000     5000   block1    block1.rds\n",
    "2        2000     6000   block2    block2.rds\n",
    "3        3000     7000   block3    block3.rds\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Example\n",
    "```\n",
    "sos run xqtl-protocol/code/pecotmr_integration/twas.ipynb twas \\\n",
    "   --cwd /output/ --name test \\\n",
    "   --gwas_meta_data /mnt/vast/hpc/csg/cl4215/mrmash/workflow/GWAS/gwas_meta.tsv \\\n",
    "   --ld_meta_data /mnt/vast/hpc/csg/data_public/20240409_ADSP_LD_matrix/ld_meta_file.tsv \\\n",
    "   --regions /mnt/vast/hpc/csg/cl4215/mrmash/workflow/pipeline_data/picalm_EUR_LD_blocks.bed \\\n",
    "   --xqtl_meta_data /mnt/vast/hpc/csg/cl4215/mrmash/workflow/pipeline_data/twas_157_genes_table_2024.tsv \\\n",
    "   --xqtl_type_table /mnt/vast/hpc/csg/cl4215/mrmash/workflow/pipeline_data/data_type_table.txt \\\n",
    "   --rsq_pval_cutoff 0.05 --rsq_cutoff 0.01\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "```\n",
    "sos run xqtl-protocol/code/pecotmr_integration/twas.ipynb ctwas \\\n",
    "   --cwd /output/ --name test \\\n",
    "   --gwas_meta_data /mnt/vast/hpc/csg/cl4215/mrmash/workflow/GWAS/gwas_meta.tsv \\\n",
    "   --ld_meta_data /mnt/vast/hpc/csg/data_public/20240409_ADSP_LD_matrix/ld_meta_file.tsv \\\n",
    "   --regions /mnt/vast/hpc/csg/cl4215/mrmash/workflow/pipeline_data/picalm_EUR_LD_blocks.bed \\\n",
    "   --xqtl_meta_data /mnt/vast/hpc/csg/cl4215/mrmash/workflow/pipeline_data/twas_157_genes_table_2024.tsv \\\n",
    "   --twas_weight_cutoff 0 --focal_region_ids \"10_80126158_82231647\" \"11_84267999_86714492\"    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "Here using `--region-name` we focus the analysis on 2 blocks: format as `chr_start_stop`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "```\n",
    "sos run xqtl-protocol/code/pecotmr_integration/twas.ipynb quantile_twas \\\n",
    "   --cwd /output/ --name test \\\n",
    "   --gwas_meta_data /mnt/vast/hpc/csg/cl4215/mrmash/workflow/GWAS/gwas_meta.tsv \\\n",
    "   --ld_meta_data /mnt/vast/hpc/csg/data_public/20240409_ADSP_LD_matrix/ld_meta_file.tsv \\\n",
    "   --region_name chr11_84267999_86714492 chr7_54681006_57314931 \\\n",
    "   --xqtl_meta_data /home/al4225/project/quantile_twas/quantile_twas_analysis/test_data/small_region_gene_meta_data_test.tsv\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "It is also possible to analyze a selected list of regions using option `--regions`. The 3 columns(chr, start, stop) of this file will be used for the block list to analyze. Here for example use the same list of regions as we used for LD block:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "```\n",
    "sos run xqtl-protocol/code/pecotmr_integration/twas.ipynb quantile_twas \\\n",
    "   --cwd /output/ --name test \\\n",
    "   --gwas_meta_data /mnt/vast/hpc/csg/cl4215/mrmash/workflow/GWAS/gwas_meta.tsv \\\n",
    "   --ld_meta_data /mnt/vast/hpc/csg/data_public/20240409_ADSP_LD_matrix/ld_meta_file.tsv \\\n",
    "   --regions /mnt/vast/hpc/csg/cl4215/mrmash/workflow/pipeline_data/EUR_LD_blocks.bed \\\n",
    "   --xqtl_meta_data /home/al4225/project/quantile_twas/quantile_twas_analysis/test_data/small_region_gene_meta_data_test.tsv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "parameter: cwd = path(\"output/\")\n",
    "parameter: gwas_meta_data = path()\n",
    "parameter: xqtl_meta_data = path()\n",
    "parameter: ld_meta_data = path()\n",
    "parameter: xqtl_type_table = ''\n",
    "parameter: gwas_name = []\n",
    "parameter: gwas_data = []\n",
    "parameter: column_mapping = []\n",
    "parameter: regions = path()\n",
    "# Optional: if a region name is provided \n",
    "# the analysis would be focused on the union of provides region list and region names\n",
    "parameter: region_name = []\n",
    "parameter: name = f\"{xqtl_meta_data:bn}.{gwas_meta_data:bn}\"\n",
    "parameter: container = ''\n",
    "import re\n",
    "parameter: entrypoint= ('micromamba run -a \"\" -n' + ' ' + re.sub(r'(_apptainer:latest|_docker:latest|\\.sif)$', '', container.split('/')[-1])) if container else \"\"\n",
    "parameter: job_size = 100\n",
    "parameter: walltime = \"5m\"\n",
    "parameter: mem = \"8G\"\n",
    "parameter: numThreads = 1\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "def adapt_file_path(file_path, reference_file):\n",
    "    \"\"\"\n",
    "    Adapt a single file path based on its existence and a reference file's path.\n",
    "\n",
    "    Args:\n",
    "    - file_path (str): The file path to adapt.\n",
    "    - reference_file (str): File path to use as a reference for adaptation.\n",
    "\n",
    "    Returns:\n",
    "    - str: Adapted file path.\n",
    "\n",
    "    Raises:\n",
    "    - FileNotFoundError: If no valid file path is found.\n",
    "    \"\"\"\n",
    "    reference_path = os.path.dirname(reference_file)\n",
    "\n",
    "    # Check if the file exists\n",
    "    if os.path.isfile(file_path):\n",
    "        return file_path\n",
    "\n",
    "    # Check file name without path\n",
    "    file_name = os.path.basename(file_path)\n",
    "    if os.path.isfile(file_name):\n",
    "        return file_name\n",
    "\n",
    "    # Check file name in reference file's directory\n",
    "    file_in_ref_dir = os.path.join(reference_path, file_name)\n",
    "    if os.path.isfile(file_in_ref_dir):\n",
    "        return file_in_ref_dir\n",
    "\n",
    "    # Check original file path prefixed with reference file's directory\n",
    "    file_prefixed = os.path.join(reference_path, file_path)\n",
    "    if os.path.isfile(file_prefixed):\n",
    "        return file_prefixed\n",
    "\n",
    "    # If all checks fail, raise an error\n",
    "    raise FileNotFoundError(f\"No valid path found for file: {file_path}\")\n",
    "\n",
    "def group_by_region(lst, partition):\n",
    "    # from itertools import accumulate\n",
    "    # partition = [len(x) for x in partition]\n",
    "    # Compute the cumulative sums once\n",
    "    # cumsum_vector = list(accumulate(partition))\n",
    "    # Use slicing based on the cumulative sums\n",
    "    # return [lst[(cumsum_vector[i-1] if i > 0 else 0):cumsum_vector[i]] for i in range(len(partition))]\n",
    "    return partition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "[get_analysis_regions: shared = [\"filtered_region_info\", \"filtered_regional_xqtl_files\", \"regional_data\"]]\n",
    "from collections import OrderedDict\n",
    "\n",
    "def check_required_columns(df, required_columns):\n",
    "    \"\"\"Check if the required columns are present in the dataframe.\"\"\"\n",
    "    missing_columns = [col for col in required_columns if col not in list(df.columns)]\n",
    "    if missing_columns:\n",
    "        raise ValueError(f\"Missing required columns: {', '.join(missing_columns)}\")\n",
    "\n",
    "def extract_regional_data(gwas_meta_data, xqtl_meta_data, regions, region_name, gwas_name, gwas_data, column_mapping):\n",
    "    \"\"\"\n",
    "    Extracts data from GWAS and xQTL metadata files and additional GWAS data provided. \n",
    "\n",
    "    Args:\n",
    "    - gwas_meta_data (str): File path to the GWAS metadata file.\n",
    "    - xqtl_meta_data (str): File path to the xQTL weight metadata file.\n",
    "    - gwas_name (list): vector of GWAS study names.\n",
    "    - gwas_data (list): vector of GWAS data.\n",
    "    - column_mapping (list, optional): vector of column mapping files.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple of two dictionaries:\n",
    "        - GWAS Dictionary: Maps study IDs to a list containing chromosome number, \n",
    "          GWAS file path, and optional column mapping file path.\n",
    "        - xQTL Dictionary: Nested dictionary with region IDs as keys.\n",
    "\n",
    "    Raises:\n",
    "    - FileNotFoundError: If any specified file path does not exist.\n",
    "    - ValueError: If required columns are missing in the input files or vector lengths mismatch.\n",
    "    \"\"\"\n",
    "    # Check vector lengths\n",
    "    if len(gwas_name) != len(gwas_data):\n",
    "        raise ValueError(\"gwas_name and gwas_data must be of equal length\")\n",
    "    \n",
    "    if len(column_mapping)>0 and len(column_mapping) != len(gwas_name):\n",
    "        raise ValueError(\"If column_mapping is provided, it must be of the same length as gwas_name and gwas_data\")\n",
    "\n",
    "    # Required columns for each file type\n",
    "    required_gwas_columns = ['study_id', 'chrom', 'file_path']\n",
    "    required_xqtl_columns = ['region_id', '#chr', 'start', 'end', \"TSS\", 'original_data'] #region_id here is gene name\n",
    "    required_ld_columns = ['chr', 'start', 'stop']\n",
    "    \n",
    "    # Reading the GWAS metadata file\n",
    "    gwas_df = pd.read_csv(gwas_meta_data, sep=\"\\t\")\n",
    "    check_required_columns(gwas_df, required_gwas_columns)\n",
    "    gwas_dict = OrderedDict()\n",
    "    \n",
    "    # Reading LD regions info\n",
    "    # Initialize empty DataFrame for regions\n",
    "    regions_df = pd.DataFrame(columns=['chr', 'start', 'stop'])\n",
    "\n",
    "    # Check if regions file exists and read it\n",
    "    if os.path.isfile(regions):\n",
    "        file_regions_df = pd.read_csv(regions, sep=\"\\t\", skipinitialspace=True)\n",
    "        file_regions_df.columns = [col.strip() for col in file_regions_df.columns]  # Strip spaces from column names\n",
    "        file_regions_df['chr'] = file_regions_df['chr'].str.strip()\n",
    "        check_required_columns(file_regions_df, required_ld_columns)\n",
    "        regions_df = pd.concat([regions_df, file_regions_df])\n",
    "    # Process region_name if provided: \n",
    "    # fomat: region_name = [\"chr1_16103_2888443\", \"chr1_4320284_5853833\"]\n",
    "    if len(region_name) > 0:\n",
    "        # Split region_name entries into chr, start, and stop columns\n",
    "        extra_regions = [name.split(\"_\") for name in region_name]\n",
    "        extra_regions_df = pd.DataFrame(extra_regions, columns=['chr', 'start', 'stop'])\n",
    "        extra_regions_df['start'] = extra_regions_df['start'].astype(int)\n",
    "        extra_regions_df['stop'] = extra_regions_df['stop'].astype(int)\n",
    "        # Add extra regions to regions_df\n",
    "        regions_df = pd.concat([regions_df, extra_regions_df])\n",
    "    # Remove duplicates and reset index\n",
    "    regions_df = regions_df.drop_duplicates().reset_index(drop=True)\n",
    "    regions_dict = OrderedDict()\n",
    "\n",
    "    # Reading the xQTL weight metadata file\n",
    "    xqtl_df = pd.read_csv(xqtl_meta_data, sep=\" \")\n",
    "    check_required_columns(xqtl_df, required_xqtl_columns)\n",
    "    xqtl_dict = OrderedDict()\n",
    "\n",
    "    # Process additional GWAS data from R vectors\n",
    "    for name, data, mapping in zip(gwas_name, gwas_data, column_mapping or [None]*len(gwas_name)):\n",
    "        gwas_dict[name] = {0: [data, mapping]}\n",
    "\n",
    "    for _, row in gwas_df.iterrows():\n",
    "        file_path = row['file_path']\n",
    "        mapping_file = row.get('column_mapping_file')\n",
    "        \n",
    "        # Adjust paths if necessary\n",
    "        file_path = adapt_file_path(file_path, gwas_meta_data)\n",
    "        if mapping_file:\n",
    "            mapping_file = adapt_file_path(mapping_file,  gwas_meta_data)\n",
    "\n",
    "       # Create or update the entry for the study_id\n",
    "        if row['study_id'] not in gwas_dict:\n",
    "            gwas_dict[row['study_id']] = {}\n",
    "\n",
    "        # Expand chrom 0 to chrom 1-22 or use the specified chrom\n",
    "        chrom_range = range(1, 23) if row['chrom'] == 0 else [row['chrom']]\n",
    "        for chrom in chrom_range:\n",
    "            if chrom in gwas_dict[row['study_id']]:\n",
    "                existing_entry = gwas_dict[row['study_id']][f'chr{chrom}']\n",
    "                raise ValueError(f\"Duplicate chromosome specification for study_id {row['study_id']}, chrom {chrom}. \"\n",
    "                                 f\"Conflicting entries: {existing_entry} and {[file_path, mapping_file]}\")\n",
    "            gwas_dict[row['study_id']][f'chr{chrom}'] = [file_path, mapping_file]\n",
    "            \n",
    "    for _, row in regions_df.iterrows():\n",
    "        LD_region_id = f\"{row['chr']}_{row['start']}_{row['stop']}\"\n",
    "        overlapping_xqtls = xqtl_df[(xqtl_df['#chr'] == row['chr']) & \n",
    "                                     (xqtl_df['TSS'] <= row['stop']) & \n",
    "                                     (xqtl_df['TSS'] >= (row['start']))]\n",
    "        file_paths = []\n",
    "        mapped_genes = []\n",
    "        # Collect file paths for xQTLs overlapping this region\n",
    "        for _, xqtl_row in overlapping_xqtls.iterrows():\n",
    "            original_data = xqtl_row['original_data']\n",
    "            file_list = original_data.split(',') if ',' in original_data else [original_data]\n",
    "            file_paths.extend([adapt_file_path(fp.strip(), xqtl_meta_data) for fp in file_list])\n",
    "            mapped_genes.extend([xqtl_row['region_id']] * len(file_list))\n",
    "\n",
    "        # Store metadata and files in the dictionary\n",
    "        regions_dict[LD_region_id] = {\n",
    "            \"meta_info\": [row['chr'], row['start'], row['stop'], LD_region_id, mapped_genes],\n",
    "            \"files\": file_paths\n",
    "        }\n",
    "        \n",
    "    for _, row in xqtl_df.iterrows():\n",
    "        file_paths = [adapt_file_path(fp.strip(), xqtl_meta_data) for fp in row['original_data'].split(',')]  # Splitting and stripping file paths\n",
    "        xqtl_dict[row['region_id']] = {\"meta_info\": [row['#chr'], row['start'], row['end'], row['region_id'], row['contexts']],\n",
    "                                       \"files\": file_paths}\n",
    "    return gwas_dict, xqtl_dict, regions_dict\n",
    "\n",
    "\n",
    "gwas_dict, xqtl_dict, regions_dict = extract_regional_data(gwas_meta_data, xqtl_meta_data,regions,region_name,gwas_name, gwas_data, column_mapping)\n",
    "regional_data = dict([(\"GWAS\", gwas_dict), (\"xQTL\", xqtl_dict), (\"Regions\", regions_dict)])\n",
    "\n",
    "\n",
    "# get regions data \n",
    "region_info = [x[\"meta_info\"] for x in regional_data['Regions'].values()]\n",
    "regional_xqtl_files = [x[\"files\"] for x in regional_data['Regions'].values()]\n",
    "\n",
    "# Filter out empty xQTL file paths\n",
    "filtered_region_info = []\n",
    "filtered_regional_xqtl_files = []\n",
    "skipped_regions =[]\n",
    "\n",
    "for region, files in zip(region_info, regional_xqtl_files):\n",
    "    if files:\n",
    "        filtered_region_info.append(region)\n",
    "        filtered_regional_xqtl_files.append(files)\n",
    "    else:\n",
    "        skipped_regions.append(region)\n",
    "print(f\"Skipping {len(skipped_regions)} out of {len(regional_xqtl_files)} regions, no overlapping xQTL weights found. \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "outputs": [],
   "source": [
    "[twas]\n",
    "depends: sos_variable(\"filtered_regional_xqtl_files\")\n",
    "parameter: coverage = \"cs_coverage_0.95\"\n",
    "# Threashold for rsq and pvalue for imputability determination for a model \n",
    "parameter: rsq_cutoff = 0.01\n",
    "parameter: rsq_pval_cutoff = 0.05\n",
    "parameter: mr_pval_cutoff = 0.05\n",
    "parameter: save_ctwas_data = True\n",
    "parameter: save_mr_result = True\n",
    "parameter: rsq_option=\"adj_rsq\"\n",
    "parameter: rsq_pval_option=[\"adj_rsq_pval\", \"pval\"]\n",
    "\n",
    "input: filtered_regional_xqtl_files, group_by = lambda x: group_by_region(x, filtered_regional_xqtl_files), group_with = \"filtered_region_info\"\n",
    "output_files = [f'{cwd:a}/{step_name}/{name}.{_filtered_region_info[3]}.twas.tsv.gz']\n",
    "if save_ctwas_data:\n",
    "    output_files.append(f'{cwd:a}/{step_name}/{name}.{_filtered_region_info[3]}.twas_data.rds')\n",
    "if save_mr_result:\n",
    "    output_files.append(f'{cwd:a}/{step_name}/{name}.{_filtered_region_info[3]}.mr_result.tsv.gz')\n",
    "output: output_files\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:n}.stdout\", stderr = f\"{_output[0]:n}.stderr\", container = container, entrypoint = entrypoint\n",
    "\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    library(pecotmr)\n",
    "    library(readr)\n",
    "\n",
    "    # get xQTL weight information \n",
    "    xqtl_meta_df <- fread(\"${xqtl_meta_data}\", data.table=FALSE) # Get related gene information from the xqtl_meta data table\n",
    "    xqtl_type_table <- if (isTRUE(file.exists(\"${xqtl_type_table}\"))) fread(\"${xqtl_type_table}\") else NULL\n",
    "    gwas_studies = c(${paths(regional_data[\"GWAS\"].keys()):r,})\n",
    "    gwas_files = c(${paths([v[_filtered_region_info[0]] for k, v in regional_data[\"GWAS\"].items()]):r,})\n",
    "    gene_list <- c(${', '.join([f\"'{gene}'\" for gene in _filtered_region_info[4]])})\n",
    "\n",
    "    # Process weights \n",
    "    twas_weights_results <- list()\n",
    "    weight_db_list <- c(${_input:r,})\n",
    "    names(weight_db_list) <- gene_list\n",
    "    weight_db_list <- split(weight_db_list, names(weight_db_list), c)\n",
    "    # remove weight db files that only contain messages: \"No SNPs found in the specified region ...\"\n",
    "    weight_db_list_update <- Filter(Negate(is.null), lapply(weight_db_list, function(file_list){do.call(c,lapply(file_list, function(file) if (file.size(file) > 200) file))}))\n",
    "    if(length(weight_db_list_update) <= 0) stop(paste0(\"No weight information available from \", paste0(weight_db_list, collapse=\",\"), \". \"))\n",
    "    \n",
    "    # Step 1: Load TWAS weights and generate twas weight db, output export_twas_weights_db \n",
    "    for (gene_db in names(weight_db_list_update)) {\n",
    "        weight_dbs <- weight_db_list_update[[gene_db]]\n",
    "        twas_weights_results[[gene_db]] = load_twas_weights(weight_dbs, variable_name_obj=\"variant_names\", \n",
    "                                                           susie_obj=\"susie_weights_intermediate\",\n",
    "                                                           twas_weights_table = \"twas_weights\")\n",
    "        if (length(twas_weights_results[[gene_db]]) > 1) {\n",
    "          twas_weights_results[[gene_db]]$data_type <- setNames(lapply(names(twas_weights_results[[gene_db]]$weights), function(context) xqtl_type_table$type[sapply(xqtl_type_table$context,\n",
    "                                                                function(x) grepl(x, context))]), names(twas_weights_results[[gene_db]]$weights))      \n",
    "        } else {\n",
    "          twas_weights_results[[gene_db]] <- NULL\n",
    "        }                   \n",
    "    }\n",
    "\n",
    "    # Step 2: twas analysis for imputable genes across contexts\n",
    "    twas_results_db <- twas_pipeline(twas_weights_results, \n",
    "                                     \"${ld_meta_data}\", \n",
    "                                     \"${gwas_meta_data}\",  \n",
    "                                     region_block=\"${_filtered_region_info[3]}\",\n",
    "                                     rsq_cutoff = ${rsq_cutoff}, \n",
    "                                     rsq_option = \"${rsq_option}\", \n",
    "                                     rsq_pval_cutoff = ${rsq_pval_cutoff}, \n",
    "                                     rsq_pval_option=c(${\", \".join([f'\"{x}\"' for x in rsq_pval_option])}), \n",
    "                                     mr_pval_cutoff = ${mr_pval_cutoff},\n",
    "                                     mr_coverage_column = \"${coverage}\", \n",
    "                                     output_twas_data = ${\"TRUE\" if save_ctwas_data else \"FALSE\"})\n",
    "    if(!is.null(twas_results_db)){\n",
    "        twas_results_db$twas_result <- merge( twas_results_db$twas_result, xqtl_meta_df[, c(\"region_id\",\"TSS\",\"start\",\"end\")], by.x=\"molecular_id\", by.y=\"region_id\")\n",
    "        fwrite(twas_results_db$twas_result[, c(2,1,14:16,3:13)], file = ${_output[0]:r}, sep = \"\\t\", compress = \"gzip\")\n",
    "    } else {\n",
    "        fwrite(data.frame(), file = ${_output[0]:r}, sep = \"\\t\", compress = \"gzip\")\n",
    "    }\n",
    "    # Step 3: reformat for follow up cTWAS analysis\n",
    "    if (${\"TRUE\" if save_ctwas_data else \"FALSE\"}) {\n",
    "        saveRDS(twas_results_db$twas_data, \"${_output[0]:nnn}.twas_data.rds\", compress='xz')\n",
    "    }\n",
    "    if (${\"TRUE\" if save_mr_result else \"FALSE\"}) {\n",
    "        if(!is.null(twas_results_db)) {\n",
    "            fwrite(twas_results_db$mr_result, file = \"${_output[0]:nnn}.mr_result.tsv.gz\", sep = \"\\t\", compress = \"gzip\")\n",
    "        } else {\n",
    "            fwrite(data.frame(), file = \"${_output[0]:nnn}.mr_result.tsv.gz\", sep = \"\\t\", compress = \"gzip\")\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS",
    "tags": [],
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "[ctwas]\n",
    "depends: sos_variable(\"filtered_region_info\")\n",
    "# twas weight cutoff for ctwas variant selection\n",
    "parameter: twas_weight_cutoff = 0\n",
    "# cs_min_cor cutoff in susie finemapping result for variant selection\n",
    "parameter: cs_min_cor = 0\n",
    "# minimum pip cutoff in susie finemapping result for variant selection\n",
    "parameter: min_pip_cutoff =0 \n",
    "# A scalar in (0,1]. The proportion of SNPs, reduce runtime at the cost of accuracy \n",
    "parameter: thin = 0.1\n",
    "# A list of regions to be subset for screening and fine-mapping, for example: \"10_80126158_82231647\"\n",
    "parameter: focal_region_ids = []\n",
    "\n",
    "region_blocks = ', '.join([f\"'{region_info[3]}'\" for region_info in filtered_region_info])\n",
    "twas_output_files = [f'{cwd:a}/twas/{name}.{region_info[3]}.twas_data.rds' for region_info in filtered_region_info]\n",
    "if focal_region_ids:\n",
    "    focal_region_ids = 'c(' + ', '.join(f'\"{focal_region_id}\"' for focal_region_id in focal_region_ids) + ')'    \n",
    "else:\n",
    "    focal_region_ids = \"NULL\"  # Default to NULL if no focal_region_ids provided\n",
    "    \n",
    "input: twas_output_files, group_by = \"all\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output:n}.stdout\", stderr = f\"{_output:n}.stderr\", container = container, entrypoint = entrypoint\n",
    "\n",
    "    library(data.table)\n",
    "    library(ctwas) # multigroup_ctwas\n",
    "    library(pecotmr)\n",
    "    \n",
    "    outputdir = \"${cwd:a}/ctwas/\"\n",
    "    if (!dir.exists(file.path(outputdir))) dir.create(file.path(outputdir))\n",
    "\n",
    "    focal_region_ids <- if (is.null(${focal_region_ids})) names(region_data) else ${focal_region_ids}\n",
    "    maxSNP = 20000\n",
    "\n",
    "    # load genome-wide data across all regions\n",
    "    gwas_studies <- unique(fread(\"${gwas_meta_data}\",data.table=FALSE, select = \"study_id\"))[,1]\n",
    "    data <- lapply(c(${_input:r,}), readRDS)\n",
    "    names(data) <- c(${region_blocks})\n",
    "\n",
    "    weights <- do.call(c, lapply(names(data), function(region_block){\n",
    "        trim_ctwas_variants(data[[region_block]], twas_weight_cutoff=${twas_weight_cutoff}, cs_min_cor=${cs_min_cor},\n",
    "                            min_pip_cutoff=${min_pip_cutoff}, max_num_variants=Inf)\n",
    "    }))\n",
    "    \n",
    "    # compile z_snp, z_gene, and snp_info across multiple regions\n",
    "    snp_info <- do.call(c, lapply(unname(data), function(x) x$snp_info))\n",
    "    snp_map <- snp_info[!duplicated(names(snp_info))]\n",
    "    z_gene_dat <- do.call(c, lapply(unname(data), function(x)x$z_gene))\n",
    "    z_snp_dat <-  do.call(c, lapply(unname(data), function(x)x$z_snp))\n",
    "    z_gene <- setNames(lapply(gwas_studies, function(study)  do.call(rbind, z_gene_dat[names(z_gene_dat) == study])), gwas_studies)\n",
    "    z_gene <- lapply(z_gene, function(x) x[!is.na(x$z) & x$id %in% names(weights),])\n",
    "    z_snp <- setNames(lapply(gwas_studies, function(study) {\n",
    "        df <- do.call(rbind, z_snp_dat[names(z_snp_dat) == study])\n",
    "        df[!duplicated(df$id), ] \n",
    "    }), gwas_studies)\n",
    "    \n",
    "    regions_data <- get_ctwas_meta_data(\"${ld_meta_data}\", names(snp_map))\n",
    "    region_info <- regions_data$region_info\n",
    "    LD_map <- regions_data$LD_info\n",
    "    rm(data)\n",
    "  \n",
    "    for (study in gwas_studies){\n",
    "      outname = paste0(\"${name}\", \"_\", study)\n",
    "  \n",
    "      # assemble regions\n",
    "      region_data_file <- file.path(outputdir, paste0(outname, \".ctwas_region_data.thin\", ${thin}, \".rds\"))\n",
    "      boundary_gene_file <- file.path(outputdir, paste0(outname, \".ctwas_boundary_genes.rds\"))\n",
    "      if (file.exists(region_data_file)) {\n",
    "        message(\"Region assemble data \",region_data_file, \" already exists, using pre-computed region data for analysis. \" )\n",
    "        region_data <- readRDS(region_data_file)\n",
    "        boundary_genes <- readRDS(boundary_gene_file)\n",
    "      } else{\n",
    "        res <- assemble_region_data(region_info,\n",
    "                                    z_snp[[study]],\n",
    "                                    z_gene[[study]],\n",
    "                                    weights,\n",
    "                                    snp_map,\n",
    "                                    thin = ${thin},\n",
    "                                    maxSNP = maxSNP,\n",
    "                                    min_group_size = 1,\n",
    "                                    ncore = 5)\n",
    "        region_data <- res$region_data\n",
    "        boundary_genes <- res$boundary_genes\n",
    "        saveRDS(region_data, region_data_file, compress='xz')\n",
    "        saveRDS(boundary_genes, boundary_gene_file, compress='xz')\n",
    "      }\n",
    "\n",
    "      # assess global parameters\n",
    "      group_prior_file <- file.path(outputdir, paste0(outname, \".ctwas_param.rds\"))\n",
    "      if (file.exists(group_prior_file)) {\n",
    "        message(\"Global parameter \",group_prior_file, \" already exists, using pre-computed global parameter for analysis. \" )\n",
    "        group_prior_data <- readRDS(group_prior_file)\n",
    "        group_prior <- group_prior_data$group_prior\n",
    "        group_prior_var <- group_prior_data$group_prior_var\n",
    "      } else {\n",
    "        param <- est_param(region_data,\n",
    "                         init_group_prior = NULL,\n",
    "                         init_group_prior_var = NULL,\n",
    "                         group_prior_var_structure = \"shared_nonSNP\", \n",
    "                         niter_prefit = 3,\n",
    "                         niter = 30,\n",
    "                         min_var = 2,\n",
    "                         min_gene = 1,\n",
    "                         min_group_size = 1,\n",
    "                         min_p_single_effect = 0.8,\n",
    "                         ncore = ${numThreads},\n",
    "                         verbose = TRUE)\n",
    "       group_prior <- param$group_prior\n",
    "       group_prior_var <- param$group_prior_var \n",
    "       saveRDS(param, group_prior_file, compress='xz')\n",
    "      }\n",
    "\n",
    "      # screen regions\n",
    "      screen_regions_file <- file.path(outputdir, paste0(outname, \".ctwas_screen_res.rds\"))\n",
    "      if (file.exists(screen_regions_file)) {\n",
    "        message(\"Region screening result \",screen_regions_file, \" already exists, using pre-computed screen region result for analysis. \" )\n",
    "        screen_res <- readRDS(screen_regions_file)\n",
    "      } else {\n",
    "        screen_res <- screen_regions(region_data[focal_region_ids],\n",
    "                                     LD_map = LD_map,\n",
    "                                     weights = weights,\n",
    "                                     filter_L = TRUE,\n",
    "                                     filter_nonSNP_PIP = FALSE,\n",
    "                                     min_L = 0,\n",
    "                                     ncore = ncore_LD,\n",
    "                                     LD_format = \"custom\",\n",
    "                                     LD_loader_fun = ctwas_ld_loader,\n",
    "                                     snpinfo_loader_fun = ctwas_bimfile_loader,\n",
    "                                     logfile = file.path(outputdir, paste0(outname, \".screen_regions.log\")))\n",
    "        screened_region_data <- screen_res$screened_region_data\n",
    "        screen_summary <- screen_res$screen_summary\n",
    "        screened_region_L <- screen_res$screened_region_L\n",
    "        # Expand screened region_data with all SNPs in the regions\n",
    "        screened_region_data <- expand_region_data(screened_region_data,\n",
    "                                                   snp_map,\n",
    "                                                   z_snp[[study]],\n",
    "                                                   maxSNP = maxSNP,\n",
    "                                                   ncore = ${numThreads})\n",
    "        screen_res$screened_region_data <- screened_region_data\n",
    "        saveRDS(screen_res, screen_regions_file, compress='xz')\n",
    "      }\n",
    "\n",
    "\n",
    "      # finemapping\n",
    "      finemap_res_file <- file.path(outputdir, paste0(outname, \".ctwas_finemap_res.rds\"))\n",
    "      susie_alpha_file <- file.path(outputdir, paste0(outname, \".ctwas_susie_alpha_res.rds\"))\n",
    "      if (file.exists(finemap_res_file)) {\n",
    "        finemap_res <- readRDS(finemap_res_file)\n",
    "        susie_alpha_res <- readRDS(susie_alpha_file)\n",
    "      } else{\n",
    "        screened_region_data <- screen_res$screened_region_data\n",
    "        screened_region_L <- screen_res$screened_region_L\n",
    "        res <- finemap_regions(screened_region_data,\n",
    "                              LD_map = LD_map,\n",
    "                              weights = weights,\n",
    "                              group_prior = param$group_prior,\n",
    "                              group_prior_var = param$group_prior_var,\n",
    "                              L = screened_region_L,\n",
    "                              ncore = ncore_LD,\n",
    "                              save_cor = TRUE,\n",
    "                              cor_dir = file.path(outputdir, \"cor_matrix\"),\n",
    "                              LD_format = \"custom\",\n",
    "                              LD_loader_fun = ctwas_ld_loader,\n",
    "                              snpinfo_loader_fun = ctwas_bimfile_loader,\n",
    "                              verbose = TRUE,\n",
    "                              logfile = file.path(outputdir, paste0(outname, \".finemap_regions.log\")))\n",
    "        finemap_res <- res$finemap_res\n",
    "        finemap_res$pval <- z2p(finemap_res$z)\n",
    "        susie_alpha_res <- res$susie_alpha_res\n",
    "        saveRDS(finemap_res, finemap_res_file, compress='xz')\n",
    "        saveRDS(susie_alpha_res, susie_alpha_file, compress='xz')\n",
    "        \n",
    "        # LD mismatch diagnosis \n",
    "        diag_LD_res_file <- file.path(outputdir, paste0(outname, \".ctwas_diag_LD_res.rds\"))\n",
    "        diag_res <- diagnose_LD_mismatch_susie(region_ids = focal_region_ids, \n",
    "                                        z_snp = z_snp[[study]], \n",
    "                                        LD_map = LD_map, \n",
    "                                        gwas_n = nrow(z_snp[[study]]),\n",
    "                                        p_diff_thresh = 5e-8,\n",
    "                                        ncore = ncore_LD,\n",
    "                                        LD_format = \"custom\",\n",
    "                                        LD_loader_fun = ctwas_ld_loader,\n",
    "                                        snpinfo_loader_fun = ctwas_bimfile_loader)\n",
    "        problematic_snps <- diag_res$problematic_snps\n",
    "        flipped_snps <- diag_res$flipped_snps\n",
    "        condz_stats <- diag_res$condz_stats\n",
    "        saveRDS(diag_res, diag_LD_res_file, compress='xz')\n",
    "      }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[quantile_twas]\n",
    "depends: sos_variable(\"filtered_regional_xqtl_files\")\n",
    "parameter: save_ctwas_data = True\n",
    "parameter: save_mr_result = False\n",
    "input: filtered_regional_xqtl_files, group_by = lambda x: group_by_region(x, filtered_regional_xqtl_files), group_with = \"filtered_region_info\"\n",
    "output_files = [f'{cwd:a}/{step_name}/{name}.{_filtered_region_info[3]}.quantile_twas.tsv.gz']\n",
    "if save_ctwas_data:\n",
    "    output_files.append(f'{cwd:a}/{step_name}/{name}.{_filtered_region_info[3]}.quantile_twas_data.rds')\n",
    "if save_mr_result:\n",
    "    output_files.append(f'{cwd:a}/{step_name}/{name}.{_filtered_region_info[3]}.quantile_mr_result.tsv.gz')\n",
    "output: output_files\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:n}.stdout\", stderr = f\"{_output[0]:n}.stderr\", container = container, entrypoint = entrypoint\n",
    "\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    library(pecotmr)\n",
    "    library(readr)\n",
    "\n",
    "    # get xQTL weight information\n",
    "    xqtl_meta_df <- fread(\"${xqtl_meta_data}\") # Get related gene information from the xqtl_meta data table\n",
    "    xqtl_type_table <- if (isTRUE(file.exists(\"${xqtl_type_table}\"))) fread(\"${xqtl_type_table}\") else NULL\n",
    "    gwas_studies = c(${paths(regional_data[\"GWAS\"].keys()):r,})\n",
    "    gwas_files = c(${paths([v[_filtered_region_info[0]] for k, v in regional_data[\"GWAS\"].items()]):r,})\n",
    "    gene_list <- c(${', '.join([f\"'{gene}'\" for gene in _filtered_region_info[4]])})\n",
    "\n",
    "    # Initialize export_twas_weights_db\n",
    "    export_twas_weights_db <- list()\n",
    "    export_twas_weights_db[[\"${_filtered_region_info[3]}\"]] <- list()\n",
    "\n",
    "    # Initialize twas_weights_results\n",
    "    twas_weights_results <- list()\n",
    "\n",
    "    # Create initial weight_db_list\n",
    "    weight_db_list <- c(${_input:r,})\n",
    "    names(weight_db_list) <- gene_list\n",
    "\n",
    "    # Split weight_db_list\n",
    "    weight_db_list <- split(weight_db_list, names(weight_db_list))\n",
    "\n",
    "    # Update weight_db_list\n",
    "    weight_db_list_update <- lapply(weight_db_list, function(file_list){\n",
    "        do.call(c, lapply(file_list, function(file) if (file.size(file) > 200) file))\n",
    "    })\n",
    "\n",
    "    # Check if weight_db_list_update is empty\n",
    "    if(length(weight_db_list_update) <= 0) {\n",
    "        stop(paste0(\"No weight information available from \", paste0(weight_db_list, collapse=\",\"), \". \"))\n",
    "    }\n",
    "\n",
    "    # Define tau_values\n",
    "    tau_values <- seq(0.01, 0.99, 0.01)\n",
    "\n",
    "    # Main processing loop\n",
    "    for (gene_db in names(weight_db_list_update)) {\n",
    "        weight_dbs <- weight_db_list_update[[gene_db]]\n",
    "        \n",
    "        twas_weights_results[[gene_db]] <- load_quantile_twas_weights(\n",
    "            weight_db_files = weight_dbs,\n",
    "            tau_values = tau_values,\n",
    "            between_cluster = 0.8,\n",
    "            num_intervals = 3\n",
    "        )\n",
    "        \n",
    "        if (!is.null(twas_weights_results[[gene_db]]) && !is.null(twas_weights_results[[gene_db]]$weights)) {\n",
    "            twas_weights_results[[gene_db]]$data_type <- setNames(\n",
    "                lapply(names(twas_weights_results[[gene_db]]$weights), function(context) {\n",
    "                    xqtl_type_table$type[sapply(xqtl_type_table$context, function(x) grepl(x, context))]\n",
    "                }),\n",
    "                names(twas_weights_results[[gene_db]]$weights)\n",
    "            )\n",
    "        } else {\n",
    "            print(paste(\"Warning: No valid weights found for gene:\", gene_db))\n",
    "        }\n",
    "    }\n",
    "\n",
    "    #Step 2: twas analysis for imputable genes across contexts\n",
    "    twas_results_db <- twas_pipeline(twas_weights_data = twas_weights_results, \n",
    "                                    ld_meta_file_path = \"${ld_meta_data}\", \n",
    "                                    gwas_meta_file = \"${gwas_meta_data}\", \n",
    "                                    region_block = \"${_filtered_region_info[3]}\", \n",
    "                                    quantile_twas = TRUE,  \n",
    "                                    output_twas_data = ${\"TRUE\" if save_ctwas_data else \"FALSE\"} )\n",
    "\n",
    "    # Merging with xQTL meta-data      \n",
    "    message(\"Merging twas_result with xqtl_meta_df...\")    \n",
    "    twas_results_db$twas_result <- merge(twas_results_db$twas_result, xqtl_meta_df[, c(\"region_id\",\"TSS\",\"start\",\"end\")], by.x=\"molecular_id\", by.y=\"region_id\")\n",
    "    twas_results_db$twas_result <- unique(twas_results_db$twas_result)\n",
    "    fwrite(twas_results_db$twas_result[, c(2, 1, (ncol(twas_results_db$twas_result)-2):ncol(twas_results_db$twas_result), 3:(ncol(twas_results_db$twas_result)-3))], file = ${_output[0]:r}, sep = \"\\t\", compress = \"gzip\")\n",
    "\n",
    "    # Step 3: reformat for follow up cTWAS analysis\n",
    "    if (${\"TRUE\" if save_ctwas_data else \"FALSE\"}) {\n",
    "        saveRDS(twas_results_db$twas_data, \"${_output[0]:nnn}.quantile_twas_data.rds\", compress='xz')\n",
    "    }\n",
    "    if (${\"TRUE\" if save_mr_result else \"FALSE\"}) {\n",
    "        fwrite(twas_results_db$mr_result, file = \"${_output[0]:nnn}.quantile_mr_result.tsv.gz\", sep = \"\\t\", compress = \"gzip\")\n",
    "    }\n",
    "    message(\"quantile twas analysis is completed in this block.\")  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     "shell"
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.24.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
