{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# GWAS integration: TWAS and MR\n",
    "\n",
    "## Introduction\n",
    "\n",
    "This module provides software implementations for transcriptome-wide association analysis (TWAS), aimed at facilitating rigorous causal inference connecting genes to complex traits.\n",
    "\n",
    "## Mendelian Randomization using Fine-Mapping IV\n",
    "\n",
    "This component focuses on Mendelian Randomization using fine-mapping instrumental variables (IV). The procedures implemented in the R script encompass PTWAS estimation and model validation, with the unit of analysis being a single gene-trait pair.\n",
    "\n",
    "## Overview\n",
    "\n",
    "- **Purpose**: Organize eQTL fine-mapping results, eQTL summary statistics, and GWAS summary statistics into a format for ptwas_est processing.\n",
    "- **Execution**: Run ptwas_est using R scripts for causal effect estimation and validation.\n",
    "\n",
    "### PTWAS Analysis from SuSiE Objects\n",
    "\n",
    "#### Step 1: PTWAS (twas_z_format)\n",
    "1. Load GWAS z-score.\n",
    "2. Load LD matrices within the TAD region of each gene.\n",
    "3. Use `allele_qc()` function to QC the LD matrix with summary stats.\n",
    "4. Extract weights: susie, lasso, enet, and mr_ash.\n",
    "5. Use `twas_z()` to compute TWAS results from multiple weights.\n",
    "\n",
    "#### Step 2: MR (twas_candidate_genes)\n",
    "1. Apply a p-value cutoff to identify TWAS regions of interest.\n",
    "2. Save QC-ed GWAS data in a format compatible with [mr.R](https://github.com/cumc/pecotmr/blob/main/R/mr.R).\n",
    "\n",
    "### Input Requirements for TWAS\n",
    "\n",
    "- `susie_path`: List of file paths for susie results.\n",
    "- `GWAS_path`: GWAS summary statistics path (for loading the GWAS sumstats dataframe).\n",
    "- `LD_path`: LD block matrix path (list of file paths of LD block).\n",
    "- `TAD_path`: TAD region path (dataframe of TAD region for each gene, used to subset the LD matrix).\n",
    "\n",
    "### Output Specifications for TWAS\n",
    "\n",
    "#### Step 1: TWAS\n",
    "1. TWAS results from multiple weights.\n",
    "2. GWAS summary statistics after QC.\n",
    "3. Input format for the `twas_z()` function.\n",
    "   - `twas_z_format`: Dataframe for input format of the `twas_z()` function.\n",
    "   - `gene_weights_pq`: Dataframe of output from `twas_z()` function. TWAS_z is applied to weights of four methods (susie, lasso, enet, and mr_ash) to calculate p-values and corresponding q-values.\n",
    "   - `AD_allele_flip`: AD GWAS sumstats after QC.\n",
    "\n",
    "#### Step 2: MR\n",
    "1. Candidate genes from TWAS.\n",
    "2. Input format for [mr.R](https://github.com/cumc/pecotmr/blob/main/R/mr.R).\n",
    "3. MR results.\n",
    "\n",
    "#### Output File Format for TWAS\n",
    "- Each row corresponds to a single SNP inferred as a member of a signal cluster, with columns including:\n",
    "   - `snp`: SNP name.\n",
    "   - `beta_eQTL`: eQTL effect.\n",
    "   - `se_eQTL`: Standard error of estimated eQTL effect.\n",
    "   - `beta_GWAS`: GWAS effect.\n",
    "   - `se_GWAS`: Standard error of GWAS effect.\n",
    "   - `cluster`: Signal cluster ID (credible sets index).\n",
    "   - `pip`: SNP posterior inclusion probability (PIP).\n",
    "   - `gene_id`: Gene name.\n",
    "\n",
    "### Causal Effect Estimation and Validation\n",
    "\n",
    "#### Estimation Process\n",
    "- Select strong eQTLs by thresholding on signal-level PIPs (cpip = 0.5).\n",
    "- Aggregate individual eQTL estimates using a fixed-effect meta-analysis procedure.\n",
    "\n",
    "#### Validation Process\n",
    "- Identify and exclude cases with severe violations of the exclusion restriction (ER) assumption.\n",
    "\n",
    "### File Formats\n",
    "\n",
    "#### Input File Format\n",
    "- Output from the first step of the analysis.\n",
    "\n",
    "#### Output File Format\n",
    "- **Contents**: The file format includes the following columns for each gene:\n",
    "    - `gene_id`: Gene name.\n",
    "    - `num_cluster`: Number of credible sets of the gene.\n",
    "    - `num_instruments`: Number of instruments included in the gene.\n",
    "    - `spip`: Sum of PIP for credible sets of each gene.\n",
    "    - `grp_beta`: Signal-level estimates, combining SNP-level estimates from member SNPs weighted by their PIPs.\n",
    "    - `grp_se`: Standard error of signal-level estimates.\n",
    "    - `meta`: Gene-level estimate of the causal effect, aggregating signal-level estimates using a fixed-effect meta-analysis model.\n",
    "    - `se_meta`: Standard error of the gene-level estimate of the causal effect.\n",
    "    - `Q`: Cochranâ€™s Q statistic.\n",
    "    - `I2`: $I^2$ statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Workdir\n",
    "parameter: susie_path = paths\n",
    "parameter: cwd = path(\"output\")\n",
    "parameter: container = ''\n",
    "import re\n",
    "parameter: entrypoint= ('micromamba run -a \"\" -n' + ' ' + re.sub(r'(_apptainer:latest|_docker:latest|\\.sif)$', '', container.split('/')[-1])) if container else \"\"\n",
    "parameter: job_size = 100\n",
    "parameter: walltime = \"1h\"\n",
    "parameter: mem = \"16G\"\n",
    "parameter: numThreads = 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "[twas_z_format_1]\n",
    "\n",
    "import pandas as pd\n",
    "s_path = pd.read_csv(susie_path, header=None)\n",
    "# split_path = s_path[0].str.split(\".\", expand=True)\n",
    "# ID = pd.DataFrame({'ID': split_path[1]})\n",
    "# path = pd.DataFrame({'path': s_path[0]})\n",
    "# input_df = pd.concat([ID, path], axis=1)\n",
    "input_df = s_path.values.tolist()\n",
    "\n",
    "parameter: GWAS_path = path\n",
    "parameter: LD_path = path\n",
    "parameter: TAD_path = path\n",
    "#input: susie_path\n",
    "input: [x for x in input_df], group_by = 1\n",
    "output: f'{cwd}/{step_name[:-2]}/{_input:bnn}.twas_z.rds'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "R: expand= \"${ }\", stderr = f'{_output:nn}.stderr', stdout = f'{_output:nn}.stdout'\n",
    "\n",
    "    options(bedtools.path = \"/home/aw3600/bedtools2/bin/\")\n",
    "    library(pecotmr)\n",
    "    library(dplyr)\n",
    "    library(stringr)\n",
    "    library(data.table)\n",
    "\n",
    "    susie_path = str_split(${_input:ar},\"\\\\.\",simplify=T)[,2]%>%cbind(.,${_input:ar})%>%data.frame()%>%setNames(c(\"ID\",\"path\"))\n",
    "    TAD_region = fread(${TAD_path:ar})%>%rename(\"ID\"=\"gene_id\")\n",
    "    gene_name = susie_path$ID\n",
    "    region = TAD_region%>%filter(ID==gene_name)%>%rename(\"chr\"=\"#chr\")\n",
    "    chr = str_sub(region$chr,4)\n",
    "    AD_dataset = fread(paste0(${GWAS_path:ar},\"/ADGWAS_Bellenguez_2022.\",chr,\"/ADGWAS2022.chr\",chr,\".sumstat.tsv\",sep=\"\"))\n",
    "    GWAS_data = AD_dataset%>%mutate(chr = paste0(\"chr\",chromosome))%>%mutate(pos = as.character(position))%>%select(-chromosome)%>%rename(\"A1\"=\"ref\",\"A2\"=\"alt\")%>%mutate(z=beta/se)%>%select(-position)\n",
    "    LD_list = read.table(${LD_path:ar},header=F,sep=\"\\t\")\n",
    "    LD_block_path = str_split(LD_list$V1,\"_\",simplify=T)%>%cbind(.,LD_list)%>%rename(\"chr\"=\"1\",\"start\"=\"2\",\"end\"=\"3\",\"variants\"=\"V1\",\"path\" = \"V2\")\n",
    "    twas_z = twas_scan(susie_path, region, GWAS_data, LD_block_path)\n",
    "    saveRDS(twas_z, ${_output:ar}, compress='xz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[twas_z_format_2]\n",
    "# Path to the input molecular phenotype data.\n",
    "input: group_by = \"all\"\n",
    "output: f'{cwd}/{step_name[:-2]}/twas_z_files.txt'\n",
    "python: expand= \"$[ ]\", stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout', container = container, entrypoint = entrypoint\n",
    "import pandas as pd\n",
    "pd.DataFrame({\"output\" : [$[_input:ar,]]}).to_csv(\"$[_output]\",index = False ,header = False, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[twas_candidate_gene]\n",
    "parameter: weights_path = path\n",
    "parameter: pval_threshold = 0.05\n",
    "parameter: cpip_cutoff = 0.5\n",
    "input: weights_path\n",
    "output: cand_genes = f'{cwd:a}/{step_name}.rds',\n",
    "        mr_output = f'{cwd:a}/mr_output.rds'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "R: expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout'\n",
    "    library(data.table)\n",
    "    library(dplyr)\n",
    "    library(stringr)\n",
    "    library(qvalue)\n",
    "    library(doMC)\n",
    "    library(pecotmr)\n",
    "\n",
    "    ####\n",
    "    susie_list = fread(${susie_path:ar},header=F)\n",
    "    susie_path = str_split(susie_list$V1,\"\\\\.\",simplify=T)[,2]%>%cbind(.,susie_list)%>%setNames(c(\"ID\",\"path\"))\n",
    "\n",
    "    #detectCores()\n",
    "    registerDoMC(100)\n",
    "    weights_file = fread(${_input:ar},header=F)\n",
    "    weights_file_path = str_split(weights_file$V1,\"\\\\.\",simplify=T)[,2]%>%cbind(.,weights_file)%>%setNames(c(\"ID\",\"path\"))\n",
    "    #ptm = proc.time()\n",
    "    gene_pq = foreach(k=1:dim(weights_file)[1], .combine=rbind) %dopar% {\n",
    "       weights_pq = readRDS(weights_file_path$path[k])$gene_weights_pq\n",
    "    }\n",
    "    #proc.time() - ptm\n",
    "    padj = apply(gene_pq[,c(\"susie_pval\",\"lasso_pval\",\"enet_pval\",\"mr_ash_pval\")],2,function(x) p.adjust(x, method = \"bonferroni\",))%>%data.frame()\n",
    "    gene_pq_adj = gene_pq%>%cbind(.,padj)\n",
    "    names(gene_pq_adj)[15:18] = c(\"susie_pval_adj\",\"lasso_pval_adj\",\"enet_pval_adj\",\"mr_ash_pval_adj\")\n",
    "    gene.ref = read.table(\"/mnt/vast/hpc/csg/molecular_phenotype_calling/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list\",header=F)\n",
    "    gene_pq_adj$gene_id = gene.ref[match(gene_pq_adj$gene_name,gene.ref$V4),\"V5\"]\n",
    "\n",
    "    cand_genes = NULL\n",
    "    cand_genes$gene_pq_adj = gene_pq_adj\n",
    "    cand_genes$susie= cand_genes$gene_pq_adj%>%filter(susie_pval<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "    cand_genes$lasso= cand_genes$gene_pq_adj%>%filter(lasso_pval<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "    cand_genes$enet= cand_genes$gene_pq_adj%>%filter(enet_pval<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "    cand_genes$mr_ash= cand_genes$gene_pq_adj%>%filter(mr_ash_pval<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "    cand_genes$susie_adj= cand_genes$gene_pq_adj%>%filter(susie_pval_adj<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "    cand_genes$lasso_adj= cand_genes$gene_pq_adj%>%filter(lasso_pval_adj<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "    cand_genes$enet_adj= cand_genes$gene_pq_adj%>%filter(enet_pval_adj<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "    cand_genes$mr_ash_adj= cand_genes$gene_pq_adj%>%filter(mr_ash_pval_adj<${pval_threshold})%>%select(gene_name,gene_id)\n",
    "\n",
    "    # cand_genes$susie_format_input = twas_mr_format_input(cand_genes$susie$gene_name,${susie_path:ar},weights_file_path)\n",
    "    # cand_genes$lasso_format_input = twas_mr_format_input(cand_genes$lasso$gene_name,${susie_path:ar},weights_file_path)\n",
    "    # cand_genes$enet_format_input = twas_mr_format_input(cand_genes$enet$gene_name,${susie_path:ar},weights_file_path)\n",
    "    # cand_genes$mr_ash_format_input = twas_mr_format_input(cand_genes$mr_ash$gene_name,${susie_path:ar},weights_file_path)\n",
    "    cand_genes$susie_adj_format_input = twas_mr_format_input(cand_genes$susie_adj,susie_path,weights_file_path)\n",
    "    cand_genes$lasso_adj_format_input = twas_mr_format_input(cand_genes$lasso_adj,susie_path,weights_file_path)\n",
    "    cand_genes$enet_adj_format_input = twas_mr_format_input(cand_genes$enet_adj,susie_path,weights_file_path)\n",
    "    cand_genes$mr_ash_adj_format_input = twas_mr_format_input(cand_genes$mr_ash_adj,susie_path,weights_file_path)\n",
    "\n",
    "    saveRDS(cand_genes, ${_output[0]:ar}, compress='xz')\n",
    "\n",
    "    mr_output = NULL\n",
    "    # mr_output$susie = fine_mr(cand_genes$susie_format_input,${cpip_cutoff})\n",
    "    # mr_output$lasso = fine_mr(cand_genes$lasso_format_input,${cpip_cutoff})\n",
    "    # mr_output$enet = fine_mr(cand_genes$enet_format_input,${cpip_cutoff})\n",
    "    # mr_output$mr_ash = fine_mr(cand_genes$mr_ash_format_input,${cpip_cutoff})\n",
    "    if(!is.null(cand_genes$susie_adj_format_input))\n",
    "    {mr_output$susie_adj = fine_mr(cand_genes$susie_adj_format_input,${cpip_cutoff})}\n",
    "    if(!is.null(cand_genes$lasso_adj_format_input))\n",
    "    {mr_output$lasso_adj = fine_mr(cand_genes$lasso_adj_format_input,${cpip_cutoff})}\n",
    "    if(!is.null(cand_genes$enet_adj_format_input))\n",
    "    {mr_output$enet_adj = fine_mr(cand_genes$enet_adj_format_input,${cpip_cutoff})}\n",
    "    if(!is.null(cand_genes$mr_ash_adj_format_input))\n",
    "    {mr_output$mr_ash_adj = fine_mr(cand_genes$mr_ash_adj_format_input,${cpip_cutoff})}\n",
    "\n",
    "    saveRDS(mr_output, ${_output[1]:ar}, compress='xz')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     "shell"
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.20.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
