
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Overview of design &#8212; ADSP-FG Project</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Summary statistics in VCF format" href="rds_to_vcf.html" />
    <link rel="prev" title="Fine-mapping with PolyFun" href="../integrative_analysis/SuSiE_Ann/polyfun.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   ADSP-FG xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Overview of design
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/misc/summary_stats_standardizer.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/misc/summary_stats_standardizer.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/misc/summary_stats_standardizer.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Overview of design
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#individual-xqtl-studies">
     Individual xQTL studies
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multiple-xqtl-integration">
     Multiple xQTL integration
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#column-name-standardization">
   Column name standardization
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#software-input-headers">
     Software (input) headers
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#our-effect-level-summary">
     Our effect level summary
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#trait-level-qtl-summary-multiple-testing-corrected">
     Trait level QTL summary (multiple-testing corrected)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#some-technical-notes">
   Some technical notes
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pre-requisites">
   Pre-requisites
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-minimal-format-of-the-input-yaml-file">
     The minimal format of the input yaml file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#memory-usage">
   Memory usage
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mwe-example-command">
   MWE Example command
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#target-generation">
     Target generation
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sumstat-standardization">
     sumstat_standardization
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#workflow-codes">
   Workflow codes
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <p>This pipeline module contains codes to process summary statistics from conventional QTL association scan to standard formats for public distribution. It will also export multiple QTL studies to formats easily accessible for data integration methods to query and analyze the summary statistics.</p>
<div class="section" id="overview-of-design">
<h1>Overview of design<a class="headerlink" href="#overview-of-design" title="Permalink to this headline">¶</a></h1>
<div class="section" id="individual-xqtl-studies">
<h2>Individual xQTL studies<a class="headerlink" href="#individual-xqtl-studies" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Reorganize the QTL marginal statistics to standard formats (see section <code class="docutils literal notranslate"><span class="pre">Column</span> <span class="pre">name</span> <span class="pre">standardization</span></code> for details).</p></li>
<li><p>Report separately cis-QTL and trans-QTL. Additionally report filtered results, e.g. QTLs that survived multiple-testing correction.</p></li>
<li><p>Our column conventions are based on studies such as GTEx and eQTL category. Our design is “modular” in the sense that we do not provide information that could be trivially annotated after, such as <strong>rsID, gene symbol, gene biotype, gene start and end positions</strong>; or for information that can be inferred from other columns such as type of allele (SNP or INDEL).</p></li>
<li><p>We use GRCh38 reference allele and alternative allele, and the effect allele is adjusted, as necessary, to the alternative allele.</p></li>
<li><p>Summary statistics will be in conventional TSV format which can be converted to <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02248-0">GWAS-VCF format</a> as needed.</p></li>
</ol>
</div>
<div class="section" id="multiple-xqtl-integration">
<h2>Multiple xQTL integration<a class="headerlink" href="#multiple-xqtl-integration" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>We will <strong>NOT</strong> standardize integrative results. Instead we will distribute the original outputs from the data integration methods we chose.</p></li>
<li><p>For meta-analysis we’ll rely on output from <a class="reference external" href="http://genetics.cs.ucla.edu/meta/">METASOFT</a> using <code class="docutils literal notranslate"><span class="pre">effect_id</span></code> as SNP ID, created from <code class="docutils literal notranslate"><span class="pre">variant</span></code>, <code class="docutils literal notranslate"><span class="pre">molecular_trait_id</span></code> and <code class="docutils literal notranslate"><span class="pre">molecular_trait_object_id</span></code> as needed.</p></li>
</ul>
<p>To build internal database for multiple xQTL data-sets, we:</p>
<ol class="simple">
<li><p>Include variants that presents in at least one xQTL.</p></li>
<li><p>Unify allele strand and frequency flips to GRCh38 reference.</p></li>
</ol>
</div>
</div>
<div class="section" id="column-name-standardization">
<h1>Column name standardization<a class="headerlink" href="#column-name-standardization" title="Permalink to this headline">¶</a></h1>
<p>The header of actual output sumstat depends on how we configure it (see section <code class="docutils literal notranslate"><span class="pre">Input</span></code> for details). However, all of them will have the column of <code class="docutils literal notranslate"><span class="pre">chromosome,</span> <span class="pre">position,</span> <span class="pre">ref,</span> <span class="pre">alt,</span> <span class="pre">variant_id,</span> <span class="pre">beta,</span> <span class="pre">se,</span> <span class="pre">pvalue</span></code>.</p>
<div class="section" id="software-input-headers">
<h2>Software (input) headers<a class="headerlink" href="#software-input-headers" title="Permalink to this headline">¶</a></h2>
<p>For example, when the input sumstat is from TensorQTL, the column specification is:</p>
<ul class="simple">
<li><p>GENE: Molecular trait identifier.(gene)</p></li>
<li><p>CHR: Variant chromosome.</p></li>
<li><p>POS: Variant chromosomal position (basepairs).</p></li>
<li><p>A0: Variant reference allele (A, C, T, or G).</p></li>
<li><p>A1: Variant alternate allele.</p></li>
<li><p>TSS_D: Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>AF: The allele frequency of this SNPs</p></li>
<li><p>MA_SAMPLES: Number of samples carrying the minor allele</p></li>
<li><p>MA_COUNT: Total number of minor alleles across individuals</p></li>
<li><p>P: Nominal P-value from linear regression</p></li>
<li><p>STAT: Slope of the linear regression</p></li>
<li><p>SE: Standard error of beta</p></li>
</ul>
<p>when the input sumstat is from APEX, the column specification is:</p>
<ul class="simple">
<li><p>GENE: Molecular trait identifier.(gene)</p></li>
<li><p>CHR: Variant chromosome.</p></li>
<li><p>POS: Variant chromosomal position (basepairs).</p></li>
<li><p>A0: Variant reference allele (A, C, T, or G).</p></li>
<li><p>A1: Variant alternate allele.</p></li>
<li><p>P: Nominal P-value from linear regression</p></li>
<li><p>STAT: Slope of the linear regression</p></li>
<li><p>SE: Standard error of beta</p></li>
</ul>
</div>
<div class="section" id="our-effect-level-summary">
<h2>Our effect level summary<a class="headerlink" href="#our-effect-level-summary" title="Permalink to this headline">¶</a></h2>
<p>Our proposed xQTL summary statistics fields should include (cf. <a class="reference external" href="https://www.niagads.org/adsp/content/xqtl-fileformats-110921v3sharedxlsx">our xQTL format draft V3, Jan 2021</a>, <a class="reference external" href="https://github.com/eQTL-Catalogue/eQTL-Catalogue-resources/blob/master/tabix/Columns.md">eQTL catalog</a>, <a class="reference external" href="https://www.gtexportal.org/home/datasets">GTEx</a>, <a class="reference external" href="https://www.metabrain.nl/">metaBrain</a>):</p>
<ul class="simple">
<li><p><strong>variant</strong> - The variant ID (chromosome_position_ref_alt) e.g. chr19_226776_C_T. Based on GRCh38 coordinates and reference genome, with ‘chr’ prefix added to the chromosome number.</p></li>
<li><p><strong>chromosome</strong> - GRCh38 chromosome name of the variant (e.g. 1,2,3 …,X).</p></li>
<li><p><strong>position</strong> - GRCh38 position of the variant.</p></li>
<li><p><strong>ref</strong> - GRCh38 reference allele.</p></li>
<li><p><strong>alt</strong> - GRCh38 alternative allele (also the effect allele).</p></li>
<li><p><strong>imputation_quality</strong> - Optional imputation quality score from the imputation software, can be replaced with NA if not available.</p></li>
<li><p><strong>molecular_trait_object_id</strong> - For phenotypes with multiple correlated alternatives (multiple alternative transcripts or exons within a gene, multple alternative promoters in txrevise, multiple alternative intons in Leafcutter), this defines the level at which the phenotypes were aggregated. Permutation p-values are calculated across this set of alternatives.</p></li>
<li><p><strong>molecular_trait_id</strong> - ID of the molecular trait used for QTL mapping. Depending on the quantification method used, this can be either a gene id, exon id, transcript id or a txrevise promoter, splicing or 3’end event id. Examples: ENST00000356937, ENSG00000008128.</p></li>
<li><p><strong>maf</strong> - Minor allele frequency within a QTL mapping context (e.g. cell type or tissues within a study).</p></li>
<li><p><strong>beta</strong> - Regression coefficient from the linear model.</p></li>
<li><p><strong>se</strong> - Standard error of the beta.</p></li>
<li><p><strong>pvalue</strong> - Nominal p-value of association between the variant and the molecular trait.</p></li>
<li><p><strong>n</strong> - Total number of samples without missing data.</p></li>
<li><p><strong>ac</strong> - Count of the alternative allele.</p></li>
<li><p><strong>ma_samples</strong> - Number of samples carrying at least one copy of the minor allele.</p></li>
</ul>
</div>
<div class="section" id="trait-level-qtl-summary-multiple-testing-corrected">
<h2>Trait level QTL summary (multiple-testing corrected)<a class="headerlink" href="#trait-level-qtl-summary-multiple-testing-corrected" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><strong>molecular_trait_object_id</strong></p></li>
<li><p><strong>molecular_trait_id</strong></p></li>
<li><p><strong>n_traits</strong> - The number of molecular traits over which permutation p-values were calculated (e.g. the number of transcripts per gene). Note that the permutations are performed accross all molecular traits within the same molecular trait object (e.g. all transcripts of a gene) and the results are reported for the most significant variant and molecular trait pair.</p></li>
<li><p><strong>n_variants</strong> - number of genetic variants tested within the cis region of the molecular trait.</p></li>
<li><p><strong>variant</strong></p></li>
<li><p><strong>chromosome</strong> - GRCh38 chromosome name of the variant (e.g. 1,2,3 …,X).</p></li>
<li><p><strong>position</strong> - GRCh38 position of the variant.</p></li>
<li><p><strong>ref</strong> - GRCh38 reference allele.</p></li>
<li><p><strong>alt</strong> - GRCh38 alternative allele (also the effect allele).</p></li>
<li><p><strong>p_perm</strong> - Empirical p-value calculated from 1000 permutations.</p></li>
<li><p><strong>p_beta</strong> - Estimated empirical p-value based on the beta distribution. This is the column that you want to use for filtering the results. See the FastQTL <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btv722">paper</a> for more details.</p></li>
<li><p><strong>qvalue</strong> - FDR based on Storey’s q-value.</p></li>
</ul>
<p>Other summary:</p>
<ul class="simple">
<li><p>Quantiles of molecular phenotypes</p></li>
</ul>
</div>
</div>
<div class="section" id="some-technical-notes">
<h1>Some technical notes<a class="headerlink" href="#some-technical-notes" title="Permalink to this headline">¶</a></h1>
<ol class="simple">
<li><p>If there are duplicated INDELs in the summary statistics, they will be removed. For example, two SNPs at 10000 on chr1. one’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> is <code class="docutils literal notranslate"><span class="pre">T</span></code>, and <code class="docutils literal notranslate"><span class="pre">A1</span></code> is <code class="docutils literal notranslate"><span class="pre">TC</span></code>. Whereas the other one’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> is <code class="docutils literal notranslate"><span class="pre">TC</span></code>, and <code class="docutils literal notranslate"><span class="pre">A1</span></code> is <code class="docutils literal notranslate"><span class="pre">T</span></code>. Both of them will be removed. More about INDEL issues(<a class="reference external" href="https://github.com/statgenetics/UKBB_GWAS_dev/issues/81#issuecomment-1015556800">https://github.com/statgenetics/UKBB_GWAS_dev/issues/81#issuecomment-1015556800</a>). For SNPs, <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> can be easily standardized to ref/alt in GRCh38 reference genome.</p></li>
<li><p>If duplicated <code class="docutils literal notranslate"><span class="pre">chr:pos</span></code> (GWAS) or <code class="docutils literal notranslate"><span class="pre">gene:chr:pos</span></code> (TWAS) exist, run a recursive match for each pair of them between two summary statistic files (<code class="docutils literal notranslate"><span class="pre">query</span></code>(each of inputs) and <code class="docutils literal notranslate"><span class="pre">subject</span></code> (target file)).</p></li>
<li><p>under the same <code class="docutils literal notranslate"><span class="pre">chr:pos</span></code> or <code class="docutils literal notranslate"><span class="pre">gene:chr:pos</span></code>, The variants’ <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> are matched by exact, flip, reverse, or flip+reverse models. Only one of them is <code class="docutils literal notranslate"><span class="pre">True</span></code>, the variant in two files are matched. If they are matched by flip or flip+reverse, the sign of <code class="docutils literal notranslate"><span class="pre">query</span></code>’s <code class="docutils literal notranslate"><span class="pre">STAT</span></code> will be inversed. And the <code class="docutils literal notranslate"><span class="pre">query</span></code>’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> will be the same as the <code class="docutils literal notranslate"><span class="pre">subject</span></code>’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code>.  <strong>FIXME: should we standardize it to GRCh38 first?</strong></p></li>
</ol>
</div>
<div class="section" id="pre-requisites">
<h1>Pre-requisites<a class="headerlink" href="#pre-requisites" title="Permalink to this headline">¶</a></h1>
<p>Make sure you install the pre-requisited before running this notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">cugg</span>
</pre></div>
</div>
</div>
<div class="section" id="input">
<h1>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--cwd</span></code>, the path of working directory</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--yml_list</span></code>, the path to a list of yaml file</p></li>
<li></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--keep-ambiguous</span></code>, boolean. default False. if add –keep-ambiguous parameter, keep ambiguous alleles which can not be decided from flip or reverse, such as A/T or C/G. Otherwise, remove them.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--intersect</span></code>, boolean. default False. if add –intersect parameter, output intersect SNPs in all input files.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--TARGET_list</span></code>a path to a list of reference file, with the column name CHR, POS, REF, ALT that represent the correct reference allele. If these reference is not availble, it can be generated by the TARGET_generation step of this workflow.</p></li>
<li><p>TARGET</p>
<ul>
<li><p>The target file is a reference summary statistic file or a file with at least variant ID relevant columns. When provided with standard <code class="docutils literal notranslate"><span class="pre">chr,</span> <span class="pre">pos,</span> <span class="pre">ref,</span> <span class="pre">alt</span></code> based on GRCh38, it can serve the purpose to standardize the REF/ALT alleles.</p></li>
</ul>
</li>
</ul>
<div class="section" id="the-minimal-format-of-the-input-yaml-file">
<h2>The minimal format of the input yaml file<a class="headerlink" href="#the-minimal-format-of-the-input-yaml-file" title="Permalink to this headline">¶</a></h2>
<p>For GWAS summary statistics,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">INPUT:</span>
  <span class="o">-</span> <span class="o">./</span><span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/*.</span><span class="n">gz</span><span class="p">:</span>
        <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
        <span class="n">variant</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span>
        <span class="n">chromosome</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">position</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">ref</span><span class="p">:</span> <span class="n">A0</span>
        <span class="n">alt</span><span class="p">:</span> <span class="n">A1</span>
        <span class="n">beta</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">se</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">pvalue</span><span class="p">:</span> <span class="n">P</span>
  <span class="o">-</span> <span class="o">./</span><span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/</span><span class="n">flip</span><span class="o">/</span><span class="n">snps500_flip</span><span class="o">.</span><span class="n">regenie</span><span class="o">.</span><span class="n">snp_stats</span><span class="o">.</span><span class="n">gz</span><span class="p">:</span>
        <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
        <span class="n">variant</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span>
        <span class="n">chromosome</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">position</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">ref</span><span class="p">:</span> <span class="n">A0</span>
        <span class="n">alt</span><span class="p">:</span> <span class="n">A1</span>
        <span class="n">beta</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">se</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">pvalue</span><span class="p">:</span> <span class="n">P</span>
        
<span class="kn">OUTPUT:</span> <span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/</span><span class="n">output</span><span class="o">/</span>
</pre></div>
</div>
<p>For xQTL summary statistics, <code class="docutils literal notranslate"><span class="pre">molecular_trait_object_id</span></code> is required because a variant can be made association with multiple molecular traits.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">INPUT:</span>
  <span class="o">-</span> <span class="n">data</span><span class="o">/</span><span class="n">twas</span><span class="o">/*.</span><span class="n">txt</span><span class="p">:</span>
        <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
        <span class="n">variant</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span>
        <span class="n">chromosome</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">position</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">ref</span><span class="p">:</span> <span class="n">A0</span>
        <span class="n">alt</span><span class="p">:</span> <span class="n">A1</span>
        <span class="n">molecular_trait_id</span><span class="p">:</span> <span class="n">GENE</span>
        <span class="n">beta</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">se</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">pvalue</span><span class="p">:</span> <span class="n">P</span> 

<span class="kn">OUTPUT:</span> <span class="o">../</span><span class="n">data</span><span class="o">/</span><span class="n">twas</span><span class="o">/</span><span class="n">output</span><span class="o">/</span>
</pre></div>
</div>
<p>There are three parts in the input yaml file.</p>
<ul class="simple">
<li><p>INPUT</p>
<ul>
<li><p>A list of yml file, as the output from yml_generator, each yml file documents a set of input</p>
<ul>
<li><p>the input summary statistic files with the column names in below.</p></li>
<li><p>the input files can be from multiple directory and from different format. The input paths must follow the rules related to Unix shell. the format is to pair the column names with required keys. If not provided, the column names of the input file will be considered as the default keys.</p></li>
<li><p>The input summary statistic file cannot have duplicated chr:pos</p></li>
<li><p>The input summary statstic file cannot have # in its header
-<code class="docutils literal notranslate"><span class="pre">variant</span></code> in yml is the rule to generate a unique identifier for each SNP, the content of variant ID shall be a combination of other columns such as chrom, position, ref, alt, build, but not taken from existing id columns in the original file.</p></li>
</ul>
</li>
</ul>
</li>
<li><p>OUTPUT</p>
<ul>
<li><p>the path of an output directory for new summary statistic files</p></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="output">
<h1>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h1>
<p>New summary statistic files with common SNPs in all input files. the sign of statistics has been corrected to make it consistent in different data.</p>
<ul class="simple">
<li><p>for each input sumstat file, a standardized version of it will be generated.</p></li>
<li><p>The generated sumstat files will have header standardized header names. The minimal set of headers will be “chromosome, position, ref, alt, variant_id, beta, se, pvalue”</p></li>
<li><p>The generated sumstat files will be in gz format.</p></li>
</ul>
</div>
<div class="section" id="memory-usage">
<h1>Memory usage<a class="headerlink" href="#memory-usage" title="Permalink to this headline">¶</a></h1>
<p>For merging two sumstat with ~85000 rows and of size of ~5MB, 1 GB of memory is needed</p>
<p>For merging two sumstat with ~2000000 rows and of size of ~1 GB, at least 50 GB of memory is needed.</p>
</div>
<div class="section" id="mwe-example-command">
<h1>MWE Example command<a class="headerlink" href="#mwe-example-command" title="Permalink to this headline">¶</a></h1>
<div class="section" id="target-generation">
<h2>Target generation<a class="headerlink" href="#target-generation" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">TARGET_generation</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">yml</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">yml_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
      <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">2</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="n">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
<div class="section" id="sumstat-standardization">
<h2>sumstat_standardization<a class="headerlink" href="#sumstat-standardization" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">sumstat_standardization</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">yml</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">yml_list</span><span class="o">.</span><span class="n">txt</span>    \
      <span class="o">--</span><span class="n">TARGET_list</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">TARGET</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">list</span> \
      <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">2</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="n">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
<p>output/data_intergration/TensorQTL/MWE.3.yml.TARGET.ref.list</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span>  <span class="n">pipeline</span><span class="o">/</span><span class="n">summary_stats_standardizer</span><span class="o">.</span><span class="n">ipynb</span>   <span class="n">sumstat_to_vcf</span>  \
      <span class="o">--</span><span class="n">sumstat</span><span class="o">-</span><span class="nb">list</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ROSMAP_methy_QTL</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">qced_sumstat_list</span><span class="o">.</span><span class="n">txt</span>   \
      <span class="o">--</span><span class="n">cwd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ROSMAP_methy_QTL</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">23</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg2</span> <span class="o">--</span><span class="kp">mem</span> <span class="mi">50</span><span class="n">G</span> <span class="o">--</span><span class="kp">walltime</span> <span class="mi">48</span><span class="n">h</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">TARGET_generation</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> 
<span class="c1"># Work directory where output will be saved to</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>

<span class="c1">#if add --keep-ambiguous parameter, keep ambiguous alleles which can not be decided from flip or reverse, such as A/T or C/G. Otherwise, remove them.</span>
<span class="kn">parameter:</span> <span class="n">keep_ambiguous</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># if add --intersect parameter, output intersect SNPs in all input files.</span>
<span class="kn">parameter:</span> <span class="n">intersect</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Containers that contains the necessary packages</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Walltime </span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;5h&#39;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;3G&#39;</span>
<span class="c1"># The directory of the output sumstat</span>
<span class="kn">parameter:</span> <span class="n">sumstat_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">sumstat_path</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sumstat_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;#chr&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="kp">name</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sumstat_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;#chr&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="c1">## Whether to rename the Chr name.</span>
<span class="kn">parameter:</span> <span class="n">remame</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;A&quot;</span> <span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span><span class="s2">&quot;Y&quot;</span><span class="p">,</span><span class="s2">&quot;MT&quot;</span><span class="p">],</span><span class="s2">&quot;X&quot;</span> <span class="p">:</span> <span class="p">[</span> <span class="sa">f</span><span class="s1">&#39;chr</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span><span class="s2">&quot;Y&quot;</span><span class="p">,</span><span class="s2">&quot;MT&quot;</span><span class="p">]]})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/chr_name&#39;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">None</span> <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="workflow-codes">
<h1>Workflow codes<a class="headerlink" href="#workflow-codes" title="Permalink to this headline">¶</a></h1>
<p>The first session is to generate a stand alone target file that can be changed into vcf and then standardized based on GTF. It include three step: Take the union of all snps without allele fliping, create a pseudo-vcf file, use bcftools to standardized the result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">TARGET_generation_1</span><span class="p">]</span>
<span class="c1">## path to a list of yml file , with columns #chr and dir</span>
<span class="kn">parameter:</span> <span class="n">yml_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">yml_path</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">yml_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">chr_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">yml_path</span><span class="p">]</span>
<span class="n">file_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">yml_path</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">file_inv</span> <span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;chr_inv&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_chr_inv</span><span class="si">}</span><span class="s1">.all_snp.vcf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="kn">import</span> <span class="nn">os</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">from</span> <span class="nn">cugg.sumstat</span> <span class="kn">import</span> <span class="n">read_sumstat</span>
    <span class="kn">from</span> <span class="nn">cugg.sumstat</span> <span class="kn">import</span> <span class="n">ss_2_vcf</span>
    <span class="kn">from</span> <span class="nn">cugg.utils</span> <span class="kn">import</span> <span class="o">*</span>
    <span class="n">yml</span> <span class="o">=</span> <span class="n">load_yaml</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">input_dict</span> <span class="o">=</span> <span class="n">parse_input</span><span class="p">(</span><span class="n">yml</span><span class="p">[</span><span class="s1">&#39;INPUT&#39;</span><span class="p">])</span>
    <span class="c1">## Remap the YML field name</span>
    <span class="n">new_key</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;molecular_trait_id&quot;</span><span class="p">,</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span> <span class="p">,</span><span class="s2">&quot;alt&quot;</span><span class="p">,</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span>
    <span class="n">old_key</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GENE&quot;</span> <span class="p">,</span>  <span class="s2">&quot;CHR&quot;</span> <span class="p">,</span>  <span class="s2">&quot;POS&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A0&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">old_key</span><span class="p">,</span><span class="n">new_key</span><span class="p">):</span>
            <span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">ss_2_vcf</span><span class="p">(</span><span class="n">ss_df</span><span class="p">,</span><span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;name&quot;</span><span class="p">):</span>
        <span class="c1">## Geno field</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">if</span> <span class="s2">&quot;SNP&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">ss_df</span><span class="p">[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="n">ss_df</span><span class="o">.</span><span class="n">CHR</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">POS</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">A0</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">A1</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;#CHROM&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;REF&#39;</span><span class="p">,</span> <span class="s1">&#39;ALT&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">ss_df</span><span class="p">[[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="s1">&#39;A0&#39;</span><span class="p">,</span> <span class="s1">&#39;A1&#39;</span><span class="p">]]</span>
        <span class="c1">## Info field(Empty)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;QUAL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;FILTER&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;PASS&quot;</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;INFO&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span>
        <span class="n">fix_header</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">]</span>
        <span class="n">header_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="s2">&quot;GENE&quot;</span> <span class="ow">in</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ss_df</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="p">[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;INFO&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;GENE=&quot;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span>
            <span class="n">fix_header</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">]</span>
            <span class="n">header_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##INFO=&lt;ID=GENE,Number=1,Type=String,Description=&quot;The name of genes&quot;&gt;&#39;</span><span class="p">]</span>
        <span class="c1">### Fix headers</span>
        <span class="kn">import</span> <span class="nn">time</span>
        <span class="n">header</span> <span class="o">=</span> <span class="s1">&#39;##fileformat=VCFv4.2</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="s1">&#39;##FILTER=&lt;ID=PASS,Description=&quot;All filters passed&quot;&gt;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="sa">f</span><span class="s1">&#39;##fileDate=</span><span class="si">{</span><span class="n">time</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y%m</span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">time</span><span class="o">.</span><span class="n">localtime</span><span class="p">())</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">+</span> \
        <span class="s1">&#39;##FORMAT=&lt;ID=STAT,Number=1,Type=Float,Description=&quot;Effect size estimate relative to the alternative allele&quot;&gt;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="s1">&#39;##FORMAT=&lt;ID=SE,Number=1,Type=Float,Description=&quot;Standard error of effect size estimate&quot;&gt;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="s1">&#39;##FORMAT=&lt;ID=P,Number=1,Type=Float,Description=&quot;The Pvalue corresponding to ES&quot;&gt;&#39;</span> 
        <span class="c1">### Customized Field headers</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">fix_header</span><span class="p">:</span>
                <span class="n">Prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;##FORMAT=&lt;ID=</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">,Number=1,Type=&#39;</span>
                <span class="n">Type</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">ss_df</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="mi">0</span><span class="p">]))</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&lt;class </span><span class="se">\&#39;</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&gt;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;numpy.&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;64&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Int&quot;</span><span class="p">,</span><span class="s2">&quot;Integer&quot;</span><span class="p">)</span>
                <span class="n">Surfix</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;,Description=&quot;Customized Field </span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&quot;&gt;&#39;</span>
                <span class="n">header_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Prefix</span><span class="o">+</span><span class="n">Type</span><span class="o">+</span><span class="n">Surfix</span><span class="p">)</span>
        <span class="c1">## format and sample field</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;FORMAT&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">]</span>  <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">fix_header</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span> <span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">],</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="c1">## Rearrangment</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;#CHROM&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;REF&#39;</span><span class="p">,</span> <span class="s1">&#39;ALT&#39;</span><span class="p">,</span> <span class="s1">&#39;QUAL&#39;</span><span class="p">,</span> <span class="s1">&#39;FILTER&#39;</span><span class="p">,</span> <span class="s1">&#39;INFO&#39;</span><span class="p">,</span><span class="s1">&#39;FORMAT&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;#CHROM&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">])</span>
        <span class="c1"># Add headers</span>
        <span class="n">header</span> <span class="o">=</span> <span class="n">header</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">header_list</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">df</span><span class="p">,</span><span class="n">header</span>

    <span class="c1">## Verify data uniqueness</span>
    <span class="n">lst_sumstats_file</span> <span class="o">=</span> <span class="p">[</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span><span class="o">&lt;</span><span class="nb">len</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">):</span> 
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;There are duplicated names in </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span>
    <span class="c1">#read all sumstats</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">input_dict</span><span class="p">)</span>
    <span class="n">lst_sumstats</span> <span class="o">=</span> <span class="p">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">i</span><span class="p">):</span><span class="n">read_sumstat</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">,)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="c1">## Retaining only chrom/pos/ref/alt, and dropping the duplicates (drop twice to reduce mem usage)</span>
    <span class="n">union_snp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">x</span><span class="p">[[</span><span class="s2">&quot;CHR&quot;</span> <span class="p">,</span>  <span class="s2">&quot;POS&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A0&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lst_sumstats</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="p">])</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span> 
    <span class="c1">## Create fake header</span>
    <span class="n">union_snp</span><span class="p">[[</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">sumstats</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="n">ss_2_vcf</span><span class="p">(</span><span class="n">union_snp</span><span class="p">,</span><span class="s2">&quot;PseudoVCF&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
    <span class="n">sumstats</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">TARGET_generation_2</span><span class="p">]</span>
<span class="c1">## The reference fasta</span>
<span class="kn">parameter:</span> <span class="n">fasta</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.TARGET.ref&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;CHR</span><span class="se">\t</span><span class="s2">POS</span><span class="se">\t</span><span class="s2">A0</span><span class="se">\t</span><span class="s2">A1&quot;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
    <span class="n">bgzip</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span>
    <span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">vcf</span> <span class="o">-</span><span class="n">f</span>  <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span><span class="o">.</span><span class="n">gz</span>
    <span class="c1">## our fasta required chr* as chromosome name format</span>
    <span class="n">bcftools</span> <span class="n">annotate</span> <span class="o">--</span><span class="n">rename</span><span class="o">-</span><span class="n">chrs</span> <span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="n">chr_name</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">Oz</span> <span class="o">|</span> \
    <span class="n">bcftools</span> <span class="n">norm</span>  <span class="o">-</span><span class="n">N</span> <span class="o">--</span><span class="n">check</span><span class="o">-</span><span class="n">ref</span> <span class="n">ws</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">fasta</span><span class="p">}</span> <span class="o">|</span> <span class="n">bcftools</span> <span class="n">query</span> <span class="o">-</span><span class="n">f</span> <span class="s1">&#39;%CHROM</span><span class="se">\t</span><span class="s1">%POS</span><span class="se">\t</span><span class="s1">%REF</span><span class="se">\t</span><span class="s1">%ALT</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">TARGET_generation_3</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/TARGET.ref.list&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="n">target_path</span> <span class="o">=</span> <span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}]</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">target_path</span> <span class="p">]</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;#chr&quot;</span><span class="p">:</span> <span class="n">chrom</span> <span class="p">,</span> <span class="s2">&quot;TARGET&quot;</span> <span class="p">:</span> <span class="n">target_path</span> <span class="p">})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">sumstat_standardization</span><span class="p">]</span>
<span class="c1">## path to a list of yml file , with columns #chr and dir</span>
<span class="kn">parameter:</span> <span class="n">yml_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">yml_path</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">yml_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="kn">depends:</span> <span class="n">Py_Module</span><span class="p">(</span><span class="s1">&#39;cugg&#39;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">TARGET_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">TARGET_path</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">TARGET_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">yml_path</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">yml_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">TARGET_path</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s2">&quot;#chr&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">file_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">yml_path</span><span class="p">]</span>
<span class="n">TARGET_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">yml_path</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">file_inv</span> <span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;TARGET_inv&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="kn">import</span> <span class="nn">os</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">from</span> <span class="nn">cugg.sumstat</span> <span class="kn">import</span> <span class="n">read_sumstat</span>
    <span class="kn">from</span> <span class="nn">cugg.utils</span> <span class="kn">import</span> <span class="o">*</span>
    
    <span class="n">yml</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span>
    <span class="n">keep_ambiguous</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">keep_ambiguous</span><span class="p">}</span>
    <span class="n">intersect</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">intersect</span><span class="p">}</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">yml</span><span class="p">,</span> <span class="n">keep_ambiguous</span><span class="p">,</span><span class="n">intersect</span><span class="p">)</span>
    <span class="c1">#parse yaml</span>
    <span class="n">yml</span> <span class="o">=</span> <span class="n">load_yaml</span><span class="p">(</span><span class="n">yml</span><span class="p">)</span>
    <span class="n">input_dict</span> <span class="o">=</span> <span class="n">parse_input</span><span class="p">(</span><span class="n">yml</span><span class="p">[</span><span class="s1">&#39;INPUT&#39;</span><span class="p">])</span>
    <span class="c1">## Remap the YML field name</span>
    <span class="n">new_key</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;molecular_trait_id&quot;</span><span class="p">,</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;alt&quot;</span><span class="p">,</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span><span class="s2">&quot;se&quot;</span><span class="p">,</span><span class="s2">&quot;pvalue&quot;</span><span class="p">,</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span>
    <span class="n">old_key</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GENE&quot;</span> <span class="p">,</span>  <span class="s2">&quot;CHR&quot;</span> <span class="p">,</span>  <span class="s2">&quot;POS&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A0&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A1&quot;</span> <span class="p">,</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">]</span>
    <span class="n">name_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">old_key</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span> <span class="n">new_key</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_key</span><span class="p">))}</span>
    <span class="n">name_map_rev</span> <span class="o">=</span> <span class="p">{</span><span class="n">new_key</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span> <span class="n">old_key</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_key</span><span class="p">))}</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">x</span><span class="p">[</span><span class="n">name_map_rev</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">k</span><span class="p">)]</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">old_key</span><span class="p">,</span><span class="n">new_key</span><span class="p">):</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">)</span>
    <span class="n">target_dict</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{_TARGET_inv}</span><span class="s2">&quot;</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">yml</span><span class="p">[</span><span class="s1">&#39;OUTPUT&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">lst_sumstats_file</span> <span class="o">=</span> <span class="p">[</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Total number of sumstats: &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span><span class="o">&lt;</span><span class="nb">len</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;There are duplicated names in </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span>
    <span class="c1">#read all sumstats</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">input_dict</span><span class="p">)</span>
    <span class="n">lst_sumstats</span> <span class="o">=</span> <span class="p">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">i</span><span class="p">):</span><span class="n">read_sumstat</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">,)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="n">nqs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#Readin the reference target file</span>
    <span class="n">subject</span> <span class="o">=</span> <span class="n">check_indels</span><span class="p">(</span><span class="n">read_sumstat</span><span class="p">(</span><span class="n">target_dict</span><span class="p">,</span><span class="kc">None</span><span class="p">,</span><span class="kc">True</span><span class="p">)[[</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">]])</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">query</span> <span class="ow">in</span> <span class="n">lst_sumstats</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="c1">#check duplicated indels and remove them.</span>
        <span class="n">query</span> <span class="o">=</span> <span class="n">check_indels</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
        <span class="c1"># Set the snp column to be the second column to satisify the requirement of compare_snps() function</span>
        <span class="n">column</span> <span class="o">=</span>  <span class="n">query</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;SNP&quot;</span><span class="p">)</span>
        <span class="n">query</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="n">column</span> <span class="p">)</span>
        <span class="c1">#under the same chr:pos or gene:chr:pos. match A0 and A1 by exact, flip, reverse, or flip+reverse.</span>
        <span class="c1">#if duplicated chr_pos or gene_chr_pos exist, run a recursive match for each pair of them between query and subject.</span>
        <span class="c1"># If GENE info is in query but not subject, added it.</span>
        <span class="k">if</span> <span class="s2">&quot;GENE&quot;</span> <span class="ow">in</span> <span class="n">query</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s2">&quot;GENE&quot;</span> <span class="ow">not</span>  <span class="ow">in</span> <span class="n">subject</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">subject</span> <span class="o">=</span> <span class="n">subject</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">query</span><span class="p">[[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">]])</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
            <span class="c1">## It is crucial that the index was built via this function, where the order of A0/A1 was removed. Otherwise will cause error is issue #306</span>
            <span class="n">subject</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">namebyordA0_A1</span><span class="p">(</span><span class="n">subject</span><span class="p">[[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">]],</span><span class="n">cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">])</span>
            <span class="n">subject</span> <span class="o">=</span> <span class="n">subject</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
        <span class="n">nq</span><span class="p">,</span><span class="n">_</span> <span class="o">=</span> <span class="n">snps_match</span><span class="p">(</span><span class="n">query</span><span class="p">,</span><span class="n">subject</span><span class="p">,</span><span class="n">keep_ambiguous</span><span class="p">)</span>
        <span class="n">nq</span> <span class="o">=</span> <span class="n">nq</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="o">~</span><span class="n">nq</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()]</span> <span class="c1"># Remove duplicated columns due to order of columns difference in subject and query</span>
        <span class="n">nqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nq</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">intersect</span><span class="p">:</span>
            <span class="c1">#get common snps</span>
            <span class="n">common_snps</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">nq</span><span class="o">.</span><span class="n">SNP</span><span class="p">)</span> <span class="k">for</span> <span class="n">nq</span> <span class="ow">in</span> <span class="n">nqs</span><span class="p">])</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Total number of common SNPs: &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">common_snps</span><span class="p">))</span>
            <span class="c1">#write out new sumstats</span>
            <span class="k">for</span> <span class="n">output_sumstats</span><span class="p">,</span><span class="n">nq</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">,</span><span class="n">nqs</span><span class="p">):</span>
                <span class="n">sumstats</span> <span class="o">=</span> <span class="n">nq</span><span class="p">[</span><span class="n">nq</span><span class="o">.</span><span class="n">SNP</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_snps</span><span class="p">)]</span>
                <span class="n">sumstats</span><span class="p">[</span><span class="s2">&quot;variants&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">CHR</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">POS</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">A0</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">A1</span>
                <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">name_map</span> <span class="p">)</span>
                <span class="n">sumstats</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">output_sumstats</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">output_sumstats</span><span class="p">,</span><span class="n">nq</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">,</span><span class="n">nqs</span><span class="p">):</span>
                <span class="n">nq</span><span class="p">[</span><span class="s2">&quot;variants&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">nq</span><span class="o">.</span><span class="n">CHR</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">nq</span><span class="o">.</span><span class="n">POS</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">nq</span><span class="o">.</span><span class="n">A0</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">nq</span><span class="o">.</span><span class="n">A1</span>
                <span class="n">nq</span> <span class="o">=</span> <span class="n">nq</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">name_map</span> <span class="p">)</span>
                <span class="c1">#output match SNPs with target SNPs.</span>
                <span class="n">nq</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">output_sumstats</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;All are done&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">sumstat_to_vcf_1</span> <span class="p">]</span>
<span class="kn">input:</span>  <span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;sumstat_path&quot;</span>
<span class="kn">output:</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="si">:</span><span class="s1">an</span><span class="si">}</span><span class="s1">.vcf&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_sumstat_path</span><span class="p">]</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">path</span><span class="p">(</span><span class="n">_sumstat_path</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/output.stdout&#39;</span>
    <span class="kn">from</span> <span class="nn">cugg.sumstat</span> <span class="kn">import</span> <span class="n">ss_2_vcf</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">from</span> <span class="nn">sos.targets</span> <span class="kn">import</span> <span class="n">path</span>
    <span class="k">def</span> <span class="nf">ss_2_vcf</span><span class="p">(</span><span class="n">ss_df</span><span class="p">,</span><span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;name&quot;</span><span class="p">):</span>
        <span class="c1">## Geno field</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">if</span> <span class="s2">&quot;SNP&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">ss_df</span><span class="p">[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="n">ss_df</span><span class="o">.</span><span class="n">CHR</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">POS</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">A0</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">A1</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;#CHROM&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;REF&#39;</span><span class="p">,</span> <span class="s1">&#39;ALT&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">ss_df</span><span class="p">[[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="s1">&#39;A0&#39;</span><span class="p">,</span> <span class="s1">&#39;A1&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">])</span>
        <span class="c1">## Info field(Empty)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;QUAL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;FILTER&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;PASS&quot;</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;INFO&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span>
        <span class="n">fix_header</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">]</span>
        <span class="n">header_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="s2">&quot;GENE&quot;</span> <span class="ow">in</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ss_df</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="p">[</span><span class="s1">&#39;SNP&#39;</span><span class="p">]</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;INFO&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;GENE=&quot;</span> <span class="o">+</span> <span class="n">ss_df</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span>
            <span class="n">fix_header</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">,</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">]</span>
            <span class="n">header_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;##INFO=&lt;ID=GENE,Number=1,Type=String,Description=&quot;The name of genes&quot;&gt;&#39;</span><span class="p">]</span>
        <span class="c1">### Fix headers</span>
        <span class="kn">import</span> <span class="nn">time</span>
        <span class="n">header</span> <span class="o">=</span> <span class="s1">&#39;##fileformat=VCFv4.2</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="s1">&#39;##FILTER=&lt;ID=PASS,Description=&quot;All filters passed&quot;&gt;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="sa">f</span><span class="s1">&#39;##fileDate=</span><span class="si">{</span><span class="n">time</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y%m</span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">time</span><span class="o">.</span><span class="n">localtime</span><span class="p">())</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">+</span> \
        <span class="s1">&#39;##FORMAT=&lt;ID=STAT,Number=1,Type=Float,Description=&quot;Effect size estimate relative to the alternative allele&quot;&gt;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="s1">&#39;##FORMAT=&lt;ID=SE,Number=1,Type=Float,Description=&quot;Standard error of effect size estimate&quot;&gt;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> \
        <span class="s1">&#39;##FORMAT=&lt;ID=P,Number=1,Type=Float,Description=&quot;The Pvalue corresponding to ES&quot;&gt;&#39;</span>
        <span class="c1">### Customized Field headers</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">fix_header</span><span class="p">:</span>
                <span class="n">Prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;##FORMAT=&lt;ID=</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">,Number=1,Type=&#39;</span>
                <span class="n">Type</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">ss_df</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="mi">0</span><span class="p">]))</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&lt;class </span><span class="se">\&#39;</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&gt;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;numpy.&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;64&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Int&quot;</span><span class="p">,</span><span class="s2">&quot;Integer&quot;</span><span class="p">)</span>
                <span class="n">Surfix</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;,Description=&quot;Customized Field </span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&quot;&gt;&#39;</span>
                <span class="n">header_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Prefix</span><span class="o">+</span><span class="n">Type</span><span class="o">+</span><span class="n">Surfix</span><span class="p">)</span>
        <span class="c1">## format and sample field</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;FORMAT&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">]</span>  <span class="o">+</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">fix_header</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ss_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span> <span class="p">[</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span><span class="s2">&quot;A1&quot;</span><span class="p">,</span><span class="s2">&quot;A0&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;CHR&quot;</span><span class="p">],</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="c1">## Rearrangment</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;#CHROM&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;REF&#39;</span><span class="p">,</span> <span class="s1">&#39;ALT&#39;</span><span class="p">,</span> <span class="s1">&#39;QUAL&#39;</span><span class="p">,</span> <span class="s1">&#39;FILTER&#39;</span><span class="p">,</span> <span class="s1">&#39;INFO&#39;</span><span class="p">,</span><span class="s1">&#39;FORMAT&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]]</span>
        <span class="c1"># Add headers</span>
        <span class="n">header</span> <span class="o">=</span> <span class="n">header</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">header_list</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">df</span><span class="p">,</span><span class="n">header</span>
    <span class="n">sumstat_path_list</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_sumstat_path</span><span class="p">}</span>
    <span class="kp">name</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="kp">name</span><span class="p">}</span>
    <span class="c1">## Remap the YML field name</span>
    <span class="n">new_key</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;molecular_trait_id&quot;</span><span class="p">,</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;alt&quot;</span><span class="p">,</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span><span class="s2">&quot;se&quot;</span><span class="p">,</span><span class="s2">&quot;pvalue&quot;</span><span class="p">,</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span>
    <span class="n">old_key</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GENE&quot;</span> <span class="p">,</span>  <span class="s2">&quot;CHR&quot;</span> <span class="p">,</span>  <span class="s2">&quot;POS&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A0&quot;</span> <span class="p">,</span>   <span class="s2">&quot;A1&quot;</span> <span class="p">,</span><span class="s2">&quot;STAT&quot;</span><span class="p">,</span><span class="s2">&quot;SE&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">,</span><span class="s2">&quot;SNP&quot;</span><span class="p">]</span>
    <span class="n">name_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">old_key</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span> <span class="n">new_key</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_key</span><span class="p">))}</span>
    <span class="n">name_map_rev</span> <span class="o">=</span> <span class="p">{</span><span class="n">new_key</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span> <span class="n">old_key</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_key</span><span class="p">))}</span>
    <span class="k">for</span> <span class="n">x</span><span class="p">,</span><span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sumstat_path_list</span><span class="p">,</span><span class="kp">name</span><span class="p">):</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">name_map_rev</span> <span class="p">)</span>
        <span class="n">sumstats</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="n">ss_2_vcf</span><span class="p">(</span><span class="n">sumstats</span><span class="p">,</span><span class="n">y</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="si">:</span><span class="s1">an</span><span class="si">}</span><span class="s1">.vcf&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
        <span class="n">sumstats</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">path</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="si">:</span><span class="s1">an</span><span class="si">}</span><span class="s1">.vcf&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">sumstat_to_vcf_2</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.merged.vcf.gz&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="kp">name</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="kp">name</span><span class="p">))</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">};</span> <span class="n">do</span>
    <span class="n">bgzip</span> <span class="o">-</span><span class="n">k</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="n">i</span> 
    <span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">vcf</span> <span class="o">-</span><span class="n">f</span>  <span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">gz</span><span class="p">;</span> <span class="n">done</span>
    <span class="n">bcftools</span> <span class="n">merge</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="si">}</span><span class="s1">.gz&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">])}</span> <span class="o">--</span><span class="n">force</span><span class="o">-</span><span class="n">samples</span> <span class="o">-</span><span class="n">m</span> <span class="nb">id</span>  <span class="o">-</span><span class="n">Oz</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">a</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/misc"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../integrative_analysis/SuSiE_Ann/polyfun.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Fine-mapping with PolyFun</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="rds_to_vcf.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Summary statistics in VCF format</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>