
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Mendelian Randomization using MRAID &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="GREGOR enrichment analysis" href="../enrichment/gregor.html" />
    <link rel="prev" title="Fine-mapping with PolyFun" href="polyfun.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pecotmr_analysis/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/fine_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/trans_analysis/MRAID_QTL.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/trans_analysis/MRAID_QTL.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/trans_analysis/MRAID_QTL.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#global-parameters">
   Global parameters
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="mendelian-randomization-using-mraid">
<h1>Mendelian Randomization using MRAID<a class="headerlink" href="#mendelian-randomization-using-mraid" title="Permalink to this headline">¶</a></h1>
<p>This notebook is to perform MRAID analysis for different targets with QTL data</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">~/</span><span class="n">Work</span><span class="o">/</span><span class="n">MR</span><span class="o">/</span><span class="mf">2023.4</span><span class="n">_MR</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">metabolics</span><span class="o">/</span><span class="n">Metabolon_Bile_Biocrate_targets_df</span><span class="o">.</span><span class="n">csv</span><span class="o">|</span><span class="n">cut</span> <span class="o">-</span><span class="n">f1</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>841
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span>    \
    <span class="o">--</span><span class="n">n</span> <span class="mi">840</span> \
    <span class="o">--</span><span class="n">targets_df</span> <span class="s2">&quot;~/Work/MR/2023.4_MR/output/metabolics/Metabolon_Bile_Biocrate_targets_df.csv&quot;</span>  \
    <span class="o">--</span><span class="n">con</span> <span class="s2">&quot;metabolics_pval_beta_0.001&quot;</span>    \
    <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;metaQTL&quot;</span>    \
    <span class="o">--</span><span class="n">p_cut</span> <span class="mf">0.001</span> \
    <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">1</span> \
    <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span>   <span class="o">&amp;&gt;</span> <span class="n">mraid_meta_1</span><span class="o">.</span><span class="mf">2.</span><span class="n">log</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">MR</span><span class="o">/</span><span class="mf">2023.4</span><span class="n">_MR</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">ADlist_lit</span><span class="o">/</span><span class="n">Causal_AD_genes_from_literature_targets_df</span><span class="o">.</span><span class="n">csv</span><span class="o">|</span><span class="n">cut</span> <span class="o">-</span><span class="n">f1</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39; &#39;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>19
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="mf">0.01</span> <span class="mf">0.001</span> <span class="mf">0.0001</span> <span class="mf">0.00001</span><span class="p">;</span>
<span class="n">do</span>
    <span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span> \
        <span class="o">--</span><span class="n">n</span> <span class="mi">19</span> \
        <span class="o">--</span><span class="n">targets_df</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature_targets_df.csv&quot;</span> \
        <span class="o">--</span><span class="n">con</span> <span class="n">ADlist_lit_eQTL_GWAS_</span><span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;eQTL&quot;</span> \
        <span class="o">--</span><span class="n">p_cut</span> <span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">0</span> \
        <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;&gt;</span> <span class="n">mraid_eQTL_</span><span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span><span class="o">.</span><span class="n">log</span> <span class="o">&amp;</span>
<span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="mf">0.01</span> <span class="mf">0.001</span> <span class="mf">0.0001</span> <span class="mf">0.00001</span><span class="p">;</span>
<span class="n">do</span>
    <span class="n">n</span><span class="o">=</span><span class="err">$</span><span class="p">((</span><span class="err">$</span><span class="p">(</span><span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">&lt;</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">MR</span><span class="o">/</span><span class="mf">2023.4</span><span class="n">_MR</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">ADlist_lit</span><span class="o">/</span><span class="n">Causal_AD_genes_from_literature_targets_df</span><span class="o">.</span><span class="n">csv</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span> \
        <span class="o">--</span><span class="n">n</span> <span class="err">$</span><span class="n">n</span> \
        <span class="o">--</span><span class="n">targets_df</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature_targets_df.csv&quot;</span> \
        <span class="o">--</span><span class="n">con</span> <span class="n">ADlist_lit_eQTL_GWAS_</span><span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;eQTL&quot;</span> \
        <span class="o">--</span><span class="n">p_cut</span> <span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">0</span> \
        <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> 
<span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#library(tidyverse)</span>
<span class="c1">#dir.create(&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit_pQTL_GWAS&quot;)</span>
<span class="c1">#AD_df&lt;-data.table::fread(&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature.csv&quot;)</span>
<span class="c1">#targets_df&lt;-data.frame(targets=AD_df$ID,chromosome=gsub(&quot;chr&quot;,&quot;&quot;,AD_df$chr))%&gt;%.[.$targets!=&quot;&quot;,]</span>
<span class="c1">#data.table::fwrite(targets_df,&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit_pQTL_GWAS/Causal_AD_genes_from_literature_pqtl_targets.csv&quot;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message in dir.create(&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit_pQTL_GWAS&quot;):
“&#39;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit_pQTL_GWAS&#39; already exists”
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#for i in 0.01 0.001 0.0001 0.00001;</span>
<span class="c1">#do</span>
<span class="c1">#    n=$(($(wc -l &lt; /mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature_targets_df.csv) - 1))</span>
<span class="c1">#    sos run ~/codes/xqtl-pipeline/pipeline/MRAID_QTL.ipynb/ mraid_qtl \</span>
<span class="c1">#        --n $n \</span>
<span class="c1">#        --AD_df &quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature.csv&quot; \</span>
<span class="c1">#        --qtl &quot;pQTL&quot; \</span>
<span class="c1">#        --p_cut ${i} \</span>
<span class="c1">#        --pval_beta 0 \</span>
<span class="c1">#        -s build -J 200 -q csg -c ~/test/csg.yml </span>
<span class="c1">#done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="mf">0.01</span> <span class="mf">0.001</span> <span class="mf">0.0001</span> <span class="mf">0.0001</span> <span class="mf">0.00001</span><span class="p">;</span>
<span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span> \
        <span class="o">--</span><span class="n">AD_df</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature.csv&quot;</span> \
        <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;pQTL&quot;</span> \
        <span class="o">--</span><span class="n">p_cut</span> <span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">0</span> \
        <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> 
<span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span> \
        <span class="o">--</span><span class="n">AD_df</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature.csv&quot;</span>  \
        <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;pQTL&quot;</span> \
        <span class="o">--</span><span class="n">p_cut</span> <span class="mf">0.01</span> \
        <span class="o">--</span><span class="n">out</span> <span class="s2">&quot;eQTL&quot;</span> \
        <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">0</span> \
        <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">out</span> <span class="ow">in</span> <span class="s2">&quot;GWAS&quot;</span> <span class="s2">&quot;pQTL&quot;</span><span class="p">;</span>
<span class="n">do</span> 

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="mf">0.01</span> <span class="mf">0.001</span> <span class="mf">0.0001</span> <span class="mf">0.0001</span> <span class="mf">0.00001</span><span class="p">;</span>
<span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span> \
        <span class="o">--</span><span class="n">AD_df</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature.csv&quot;</span> \
        <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;eQTL&quot;</span> \
        <span class="o">--</span><span class="n">p_cut</span> <span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">out</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">0</span> \
        <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> 
<span class="n">done</span>

<span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">ad</span> <span class="ow">in</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature.csv&quot;</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist/Causal_AD_genes.csv&quot;</span><span class="p">;</span>
<span class="n">do</span>

<span class="k">for</span> <span class="n">out</span> <span class="ow">in</span> <span class="s2">&quot;GWAS&quot;</span> <span class="s2">&quot;pQTL&quot;</span><span class="p">;</span>
<span class="n">do</span> 

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="mf">0.01</span> <span class="mf">0.001</span> <span class="mf">0.0001</span> <span class="mf">0.0001</span> <span class="mf">0.00001</span><span class="p">;</span>
<span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span> \
        <span class="o">--</span><span class="n">AD_df</span> <span class="err">$</span><span class="p">{</span><span class="n">ad</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;eQTL&quot;</span> \
        <span class="o">--</span><span class="n">p_cut</span> <span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">out</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">0</span> \
        <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> 
<span class="n">done</span>

<span class="n">done</span>

<span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">ad</span> <span class="ow">in</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist_lit/Causal_AD_genes_from_literature.csv&quot;</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/MR/2023.4_MR/output/ADlist/Causal_AD_genes.csv&quot;</span><span class="p">;</span>
<span class="n">do</span>

<span class="k">for</span> <span class="n">out</span> <span class="ow">in</span> <span class="s2">&quot;GWAS&quot;</span> <span class="s2">&quot;eQTL&quot;</span><span class="p">;</span>
<span class="n">do</span> 

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="mf">0.01</span> <span class="mf">0.001</span> <span class="mf">0.0001</span> <span class="mf">0.0001</span> <span class="mf">0.00001</span><span class="p">;</span>
<span class="n">do</span>
    <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span> \
        <span class="o">--</span><span class="n">AD_df</span> <span class="err">$</span><span class="p">{</span><span class="n">ad</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;pQTL&quot;</span> \
        <span class="o">--</span><span class="n">p_cut</span> <span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">out</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">0</span> \
        <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> 
<span class="n">done</span>

<span class="n">done</span>

<span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">ad</span> <span class="ow">in</span> <span class="s2">&quot;~/Work/MR/2023.4_MR/output/metabolics/Metabolon_Bile_Biocrate.csv&quot;</span><span class="p">;</span>
<span class="n">do</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="mf">0.01</span> <span class="mf">0.001</span> <span class="mf">0.0001</span> <span class="mf">0.0001</span> <span class="mf">0.00001</span><span class="p">;</span>
<span class="n">do</span>
   <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">MRAID_QTL</span><span class="o">.</span><span class="n">ipynb</span><span class="o">/</span> <span class="n">mraid_qtl</span>    \
       <span class="o">--</span><span class="n">AD_df</span> <span class="err">$</span><span class="p">{</span><span class="n">ad</span><span class="p">}</span> \
       <span class="o">--</span><span class="n">qtl</span> <span class="s2">&quot;metaQTL&quot;</span>    \
       <span class="o">--</span><span class="n">p_cut</span> <span class="err">$</span><span class="p">{</span><span class="n">i</span><span class="p">}</span> \
       <span class="o">--</span><span class="n">pval_beta</span> <span class="mi">1</span> \
       <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span>   
<span class="n">done</span>
<span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="global-parameters">
<h2>Global parameters<a class="headerlink" href="#global-parameters" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="c1"># Work directory &amp; output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./&#39;</span><span class="p">)</span>
<span class="c1"># The filename prefix for output data</span>
<span class="kn">parameter:</span> <span class="kp">name</span><span class="o">=</span><span class="s2">&quot;test&quot;</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">table_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1">#parameter: con = str</span>
<span class="kn">parameter:</span> <span class="n">qtl</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">out</span> <span class="o">=</span> <span class="s2">&quot;GWAS&quot;</span>
<span class="kn">parameter:</span> <span class="n">p_cut</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="kn">parameter:</span> <span class="n">AD_df</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">samplen1</span> <span class="o">=</span> <span class="mi">900</span>
<span class="kn">parameter:</span> <span class="n">samplen2</span> <span class="o">=</span> <span class="mi">788989</span>
<span class="c1">##  conditions can be excluded if needs arise. If nothing to exclude keep the default 0</span>
<span class="kn">parameter:</span> <span class="n">datadir</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1">#parameter: analysis_units = path</span>
<span class="kn">parameter:</span> <span class="n">pval_beta</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># set as 1 if you use metabolic data, which would use pavlue_beta in sumstats result</span>
<span class="c1"># handle N = per_chunk data-set in one job</span>
<span class="c1">#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_units).readlines() if x.strip() and not x.strip().startswith(&#39;target&#39;)]</span>
<span class="kn">parameter:</span> <span class="n">con</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">AD_df</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">qtl</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="n">targets_df</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">_target_df.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">mraid_qtl_1</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">AD_df</span>
<span class="kn">output:</span> <span class="n">targets_df</span>

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> 
    <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
       <span class="n">AD_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{AD_df}</span><span class="s2">&quot;</span><span class="p">)</span>
       <span class="n">qtl</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{qtl}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;eQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ensembl</span><span class="p">,</span><span class="n">chromosome</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="n">AD_df</span><span class="err">$</span><span class="nb">chr</span><span class="p">))</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;pQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ID</span><span class="p">,</span><span class="n">chromosome</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="n">AD_df</span><span class="err">$</span><span class="nb">chr</span><span class="p">))</span><span class="o">%&gt;%.</span><span class="p">[</span><span class="o">.</span><span class="err">$</span><span class="n">targets</span><span class="o">!=</span><span class="s2">&quot;&quot;</span><span class="p">,]</span>

            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;mQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">NA</span> <span class="c1">#need to add sth..</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;haQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">NA</span> <span class="c1">#need to add sth..</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;metaQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">molecular_trait_id</span><span class="p">,</span><span class="n">chromosome</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">group</span><span class="p">)</span>
            <span class="p">}</span>  
      <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fwrite</span><span class="p">(</span><span class="n">targets_df</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># perform mraid analysis with QTL data</span>
<span class="p">[</span><span class="n">mraid_qtl_2</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">_target_df.csv&quot;</span><span class="p">)</span>
<span class="n">n</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">n</span>  <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s1">&#39;n&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">qtl</span><span class="si">}</span><span class="s2">/sig_</span><span class="si">{</span><span class="n">p_cut</span><span class="si">}</span><span class="s2">/res_mraid_</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">_p_cut_</span><span class="si">{</span><span class="n">p_cut</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">_n</span><span class="si">}</span><span class="s2">.csv&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;80G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> 
    <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">UpSetR</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;data.table&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">RcppCNPy</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">arrow</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">MRAID</span><span class="p">)</span>    
    <span class="n">find_SNP_from_sum</span><span class="o">&lt;-</span><span class="n">function</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span><span class="n">refpath</span><span class="p">,</span><span class="n">SNP</span><span class="p">,</span><span class="n">ID</span><span class="o">=</span><span class="s2">&quot;variant&quot;</span><span class="p">,</span><span class="n">full_chr</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span><span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span><span class="n">ref_with_fullpath</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span><span class="n">IDsep</span><span class="o">=</span><span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="n">suffix</span><span class="o">=</span><span class="s1">&#39;.norminal.cis_long_table.txt&#39;</span><span class="p">,</span><span class="n">IDsep_in_sum</span><span class="o">=</span><span class="s2">&quot;_&quot;</span><span class="p">){</span>
        <span class="nb">chr</span><span class="o">&lt;-</span><span class="n">unique</span><span class="p">(</span><span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">SNP</span><span class="p">,</span><span class="n">IDsep</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="n">T</span><span class="p">)[,</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span><span class="p">(</span><span class="n">isTRUE</span><span class="p">(</span><span class="n">full_chr</span><span class="p">))</span><span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="o">&lt;-</span><span class="nb">chr</span> <span class="k">else</span> <span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="o">&lt;-</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="nb">chr</span><span class="p">)</span>
        <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="p">))</span><span class="k">try</span><span class="p">({</span>
            <span class="n">c</span><span class="o">&lt;-</span><span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">use</span><span class="o">.</span><span class="n">ref</span><span class="o">&lt;-</span><span class="n">ref</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;[.]&quot;</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">suffix</span><span class="p">),</span><span class="n">ref</span><span class="p">)]</span>
            <span class="k">if</span><span class="p">(</span><span class="n">isTRUE</span><span class="p">(</span><span class="n">ref_with_fullpath</span><span class="p">)){</span>
                <span class="n">tmp</span><span class="o">.</span><span class="n">f</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="n">use</span><span class="o">.</span><span class="n">ref</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">tmp</span><span class="o">.</span><span class="n">f</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">refpath</span><span class="p">,</span><span class="s2">&quot;/&quot;</span><span class="p">,</span><span class="n">use</span><span class="o">.</span><span class="n">ref</span><span class="p">),</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span><span class="p">(</span><span class="n">stringr</span><span class="p">::</span><span class="n">str_split_fixed</span><span class="p">(</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]],</span><span class="n">IDsep</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">)[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span><span class="o">!=</span><span class="nb">chr</span><span class="p">[</span><span class="n">i</span><span class="p">]){</span>
                <span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]]</span><span class="o">&lt;-</span><span class="n">sub</span><span class="p">(</span><span class="n">IDsep_in_sum</span><span class="p">,</span><span class="n">IDsep</span><span class="p">,</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]])</span>
            <span class="p">}</span>
            <span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="o">&lt;-</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]]</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">SNP</span><span class="p">,]</span>
            <span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="o">&lt;-</span><span class="n">rbind</span><span class="p">(</span><span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="p">,</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">)</span>    
        <span class="p">})</span>
        <span class="k">return</span><span class="p">(</span><span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="p">)</span>
    <span class="p">}</span>
        <span class="c1"># mraid function</span>
        <span class="n">mraid</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">target</span><span class="p">,</span><span class="n">path</span><span class="p">,</span> <span class="n">qtl</span><span class="p">,</span> <span class="n">res</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">p_cut</span><span class="p">,</span> <span class="n">targets_df</span><span class="p">,</span> <span class="n">con</span> <span class="p">,</span><span class="n">outcome</span><span class="p">,</span><span class="n">pval_beta</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span><span class="n">samplen1</span><span class="p">,</span><span class="n">samplen2</span><span class="p">)</span> <span class="p">{</span>
        <span class="nb">chr</span><span class="o">=</span><span class="n">targets_df</span><span class="p">[</span><span class="n">targets_df</span><span class="err">$</span><span class="n">targets</span> <span class="o">==</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">]</span><span class="o">%&gt;%</span><span class="n">unique</span><span class="p">()</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;eQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/eQTL/dlpfc_batch_all.rnaseqc.low_expression_filtered.outlier_removed.tmm.expression.bed.processed_phenotype.per_chrom_dlpfc_batch_all.rnaseqc.ROSMAP_covariates.ROSMAP_NIA_WGS.pca.PEER.txt.&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;pQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/pQTL/rf_standard/pheno_recipe_pheno.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.filtered.filtered.prune.pca.resid.PEER.&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
                <span class="c1"># there are so many duplicated rows in exp matrix</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="p">[</span><span class="o">-</span><span class="n">which</span><span class="p">(</span><span class="n">duplicated</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="p">)),</span> <span class="p">]</span>
                <span class="c1"># in pQTL files, the snps are chrx_tacg_... but in most files should be chrx:tacg_...</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">variant</span> <span class="o">&lt;-</span> <span class="n">sub</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">variant</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;mQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/mQTL/ROSMAP_arrayMethylation_covariates.sesame.methyl.beta.sample_matched.bed_BMIQ.bed.filter_na.bed.softImputed.bed.processed_phenotype.per_chrom_methy.peer.pca.chr&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">%&gt;%</span> <span class="n">rename</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span> <span class="o">=</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;haQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/haQTL/h3k9ac_bed_recipe_h3k9ac_whole.k9_cov.xqtl_protocol_data.filtered.related.filtered.extracted.pca.projected.resid.PEER.merged.&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">%&gt;%</span> <span class="n">rename</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span> <span class="o">=</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;metaQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/metaQTL/&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span><span class="s2">&quot;/rf/&quot;</span><span class="p">),</span><span class="s2">&quot;_rf.norminal.trans_long_table.txt$&quot;</span><span class="p">,</span><span class="n">full</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">T</span><span class="p">))</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">%&gt;%</span> <span class="n">rename</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span> <span class="o">=</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">position</span> <span class="o">&lt;-</span> <span class="n">str_split</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">position</span><span class="p">,</span><span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span><span class="mi">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">()</span>
            <span class="p">}</span>
                <span class="c1">## 2. prapare the input for MR</span>
                <span class="c1">### 2.1 exposure GWAS sumstats</span>
                <span class="c1"># get qtl Data</span>
                <span class="n">exp</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="p">[</span><span class="n">molecular_trait_id</span> <span class="o">==</span> <span class="n">target</span><span class="p">]</span>
                <span class="n">snps_ana_gene</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="err">$</span><span class="n">variant</span>
                <span class="k">if</span><span class="p">(</span><span class="n">pval_beta</span><span class="o">==</span><span class="mi">1</span><span class="p">){</span><span class="n">exp</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="p">[</span><span class="n">pval_beta</span> <span class="o">&lt;</span> <span class="n">p_cut</span><span class="p">,</span> <span class="p">]}</span> <span class="k">else</span> <span class="n">exp</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="p">[</span><span class="n">pvalue</span> <span class="o">&lt;</span> <span class="n">p_cut</span><span class="p">,</span> <span class="p">]</span>
                <span class="n">snps_asso_gene</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="err">$</span><span class="n">variant</span>
                <span class="n">res</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span> <span class="n">n_snps_ana</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">snps_ana_gene</span><span class="p">),</span> <span class="n">p_cut</span> <span class="o">=</span> <span class="n">p_cut</span><span class="p">,</span> <span class="n">n_snps_asso</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">))</span>
                <span class="n">message</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">),</span> <span class="s2">&quot; SNPs associated with &quot;</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
                    <span class="c1">### 2.2 outcome GWAS sumstats</span>
                    <span class="c1"># get snps&#39; GWAS data</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">outcome</span> <span class="o">==</span> <span class="s2">&quot;GWAS&quot;</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">path</span> <span class="o">&lt;-</span> <span class="s2">&quot;/home/rf2872/Work/MR/2023.4_MR/data/sumtats/GWAS/&quot;</span>
                        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">find_SNP_from_sum</span><span class="p">(</span><span class="n">ref</span> <span class="o">=</span> <span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">path</span><span class="p">),</span> <span class="n">refpath</span> <span class="o">=</span> <span class="n">path</span><span class="p">,</span> <span class="n">SNP</span> <span class="o">=</span> <span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">suffix</span> <span class="o">=</span> <span class="s2">&quot;.sumstat.tsv&quot;</span><span class="p">,</span> <span class="n">full_chr</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">IDsep_in_sum</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">path</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;/home/rf2872/Work/MR/2023.4_MR/data/sumtats/&quot;</span><span class="p">,</span><span class="n">outcome</span><span class="p">)</span>
                        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">find_SNP_from_sum</span><span class="p">(</span><span class="n">ref</span> <span class="o">=</span> <span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">path</span><span class="p">),</span> <span class="n">refpath</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">SNP</span> <span class="o">=</span> <span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">full_chr</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">IDsep_in_sum</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
                        <span class="c1">###find out data</span>
                        <span class="k">if</span> <span class="p">(</span><span class="n">outcome</span> <span class="o">==</span> <span class="s2">&quot;eQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="err">$</span><span class="n">molecular_trait_id</span><span class="o">==</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ensembl</span><span class="p">[</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ID</span><span class="o">==</span><span class="n">target</span><span class="p">],]}</span>
                        <span class="k">if</span> <span class="p">(</span><span class="n">outcome</span> <span class="o">==</span> <span class="s2">&quot;pQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="err">$</span><span class="n">molecular_trait_id</span><span class="o">==</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ID</span><span class="p">[</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ensembl</span><span class="o">==</span><span class="n">target</span><span class="p">],]}</span>
                             
                        <span class="n">int_snps</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">out</span><span class="err">$</span><span class="n">variant</span><span class="p">)</span>
                    <span class="p">}</span>
                    <span class="n">int_snps</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">out</span><span class="err">$</span><span class="n">variant</span><span class="p">)</span>
                    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">int_snps</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">){</span>
                    <span class="c1">### 2.3/2.4 exposure LD matrix</span>
                    <span class="c1"># since the LD files are saved separately accrodding to their postions, I&#39;d like to use bedtools to map the SNPs to LD files</span>
                    <span class="n">LD</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s2">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/LD/output/1300_hg38_EUR_LD_blocks_LD/ROSMAP_NIA_WGS.leftnorm.filtered.filtered.ld.list&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">bed</span> <span class="o">&lt;-</span> <span class="n">str_split</span><span class="p">(</span><span class="n">LD</span><span class="err">$</span><span class="n">V1</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">cbind</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="n">LD</span><span class="p">)</span>
                    
                    <span class="c1"># if(qtl == &quot;eQTL&quot; | qtl == &quot;pQTL&quot;){</span>
                    <span class="n">exp_LD</span><span class="o">.</span><span class="n">bed</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">chrome</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="err">$</span><span class="n">chromosome</span><span class="p">),</span> <span class="n">start</span> <span class="o">=</span> <span class="n">exp</span><span class="err">$</span><span class="n">position</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="p">(</span><span class="n">exp</span><span class="err">$</span><span class="n">position</span> <span class="o">+</span> <span class="n">nchar</span><span class="p">(</span><span class="n">exp</span><span class="err">$</span><span class="n">ref</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
                    <span class="c1"># }</span>
                    <span class="c1"># if(qtl == &quot;mQTL&quot;){</span>
                    <span class="c1"># exp_LD.bed &lt;- data.frame(chrome = exp$chrom, start = exp$pos, end = (exp$pos + nchar(exp$ref) - 1))</span>
                    <span class="c1"># }</span>
                    <span class="n">options</span><span class="p">(</span><span class="n">bedtools</span><span class="o">.</span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;/home/rf2872/software/bedtools2/bin/&quot;</span><span class="p">)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">files</span> <span class="o">&lt;-</span> <span class="n">bedtoolsr</span><span class="p">::</span><span class="n">bt</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">a</span> <span class="o">=</span> <span class="n">LD</span><span class="o">.</span><span class="n">bed</span><span class="p">,</span> <span class="n">b</span> <span class="o">=</span> <span class="n">exp_LD</span><span class="o">.</span><span class="n">bed</span><span class="p">)</span>
                      
                    <span class="c1"># load python tools to read npz files, but there would be an error for npz1$f[[&quot;arr_1&quot;]], so I need to read the bim file to locate the snps. (Hao says they are in sam order)</span>
                    <span class="n">np</span> <span class="o">&lt;-</span> <span class="n">import</span><span class="p">(</span><span class="s2">&quot;numpy&quot;</span><span class="p">)</span>
                    <span class="n">LD_files</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">files</span><span class="err">$</span><span class="n">V5</span><span class="p">)</span>
                    
                    <span class="c1"># some snp sets could locate in more than 1 ld files. so I have to read those files separately and combine them (I don&#39;t know which function can do this easily, so I just write a new one)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">list</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
                    <span class="k">for</span> <span class="p">(</span><span class="n">f</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">LD_files</span><span class="p">))</span> <span class="p">{</span>
                        <span class="n">npz1</span> <span class="o">&lt;-</span> <span class="n">np</span><span class="err">$</span><span class="n">load</span><span class="p">(</span><span class="n">LD_files</span><span class="p">[</span><span class="n">f</span><span class="p">])</span>
                        <span class="n">ld</span><span class="o">.</span><span class="n">mtx</span> <span class="o">&lt;-</span> <span class="n">npz1</span><span class="err">$</span><span class="n">f</span><span class="p">[[</span><span class="s2">&quot;arr_0&quot;</span><span class="p">]]</span>
                        <span class="n">ld</span><span class="o">.</span><span class="n">snps</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">LD_files</span><span class="p">[</span><span class="n">f</span><span class="p">],</span> <span class="s2">&quot;[.]&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="o">.</span><span class="p">[,</span> <span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="o">.</span><span class="p">),</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="o">.</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))]</span> <span class="o">%&gt;%</span>
                            <span class="n">paste</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="n">paste0</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;.bim&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="o">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="o">.</span><span class="p">[,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">colnames</span><span class="p">(</span><span class="n">ld</span><span class="o">.</span><span class="n">mtx</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">ld</span><span class="o">.</span><span class="n">mtx</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">ld</span><span class="o">.</span><span class="n">snps</span>
                        <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">ld</span><span class="o">.</span><span class="n">mtx</span><span class="p">)</span>
                        <span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]),</span> <span class="n">int_snps</span><span class="p">)</span>
                        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]][</span><span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">,</span> <span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
                    <span class="n">colnames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]])</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]])</span> <span class="o">&lt;-</span> <span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">}</span> <span class="k">else</span> <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]</span><span class="o">&lt;-</span><span class="n">NA</span> 
                        
                    <span class="p">}</span>      
                    <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[</span><span class="n">sapply</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">,</span><span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nb">all</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">x</span><span class="p">)))]</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
                    
                    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">){</span>
                    <span class="n">u</span> <span class="o">&lt;-</span> <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="mi">1</span><span class="p">]]</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
                        <span class="k">for</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="mi">2</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">))</span> <span class="p">{</span>
                            <span class="n">t</span><span class="o">.</span><span class="n">col</span> <span class="o">&lt;-</span> <span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">+</span> <span class="n">ncol</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span>
                            <span class="n">u</span><span class="o">.</span><span class="n">n</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">ncol</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">col</span><span class="p">,</span> <span class="n">nrow</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">col</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                            <span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">),</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
                            <span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">[(</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">),</span> <span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span>
                            <span class="n">colnames</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">u</span><span class="p">),</span> <span class="n">colnames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">x</span><span class="p">]]))</span>
                            <span class="n">u</span> <span class="o">&lt;-</span> <span class="n">u</span><span class="o">.</span><span class="n">n</span>
                        <span class="p">}</span>
                    <span class="p">}</span>
                      
                    <span class="c1"># the snps in LD matrix and sumstats files are not exactly same, so I need to get the overlapped ones</span>
                    <span class="n">realsnps</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">u</span><span class="p">),</span> <span class="n">int_snps</span><span class="p">)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">m</span> <span class="o">&lt;-</span> <span class="n">LD</span><span class="o">.</span><span class="n">exp</span><span class="o">.</span><span class="n">m</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">u</span><span class="p">[</span><span class="n">realsnps</span><span class="p">,</span> <span class="n">realsnps</span><span class="p">])</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="p">)]</span>
                    <span class="n">message</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="p">),</span> <span class="s2">&quot; SNPs in &quot;</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot; analyzed in MRAID&quot;</span><span class="p">)</span>
                      
                    <span class="c1">### 2.5 exposure sample size</span>
                    <span class="c1"># from [DLPFC ROSMAP](https://github.com/cumc/fungen-xqtl-analysis/blob/main/data/descriptor/qtl/ROSMAP_DLPFC_expression_qtl.md) (The number on this page is wrong on 4/11/2023, it should be ~950, let&#39;s set 900 for now)</span>
                    <span class="n">samplen1</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">samplen1</span><span class="p">)</span>
                    <span class="c1">### 2.6 outcome sample size</span>
                    <span class="c1"># from [Bellenguez GWAS paper](https://pubmed.ncbi.nlm.nih.gov/35379992/)(111,326 AD cases and 677,663 controls)</span>
                    <span class="n">samplen2</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">samplen2</span><span class="p">)</span>
                    
                      
                    <span class="c1">## 3. RUN MRAID</span>
                    <span class="n">exp</span><span class="o">.</span><span class="n">real</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="p">[</span><span class="n">exp</span><span class="err">$</span><span class="n">variant</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">realsnps</span><span class="p">,</span> <span class="p">]</span>
                    <span class="n">exp</span><span class="o">.</span><span class="n">z</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">beta</span><span class="p">)</span> <span class="o">/</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">se</span><span class="p">)</span>
                      
                    <span class="n">out</span><span class="o">.</span><span class="n">real</span> <span class="o">&lt;-</span> <span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="err">$</span><span class="n">variant</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">realsnps</span><span class="p">,</span> <span class="p">]</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">z</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">beta</span><span class="p">)</span> <span class="o">/</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">se</span><span class="p">)</span>
                      
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="p">)]</span>
                        
                    <span class="c1"># you can check the parameters info for MRAID  in [here](https://github.com/yuanzhongshang/MRAID/blob/main/man/MRAID.Rd)</span>
                    <span class="n">message</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="s2">&quot; MRAID analysis...&quot;</span><span class="p">)</span>
                      
                    <span class="c1"># running MRAID with a big LD matrix would cost so much time, I would like to document the time used</span>
                    <span class="n">start</span><span class="o">.</span><span class="n">time</span> <span class="o">&lt;-</span> <span class="n">Sys</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">tryCatch</span><span class="p">(</span><span class="n">MRAID</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">z</span><span class="p">,</span> <span class="n">out</span><span class="o">.</span><span class="n">z</span><span class="p">,</span> <span class="n">LD</span><span class="o">.</span><span class="n">exp</span><span class="o">.</span><span class="n">m</span><span class="p">,</span> <span class="n">LD</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">m</span><span class="p">,</span> <span class="n">samplen1</span><span class="p">,</span> <span class="n">samplen2</span><span class="p">,</span>
                        <span class="n">Gibbsnumber</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">burninproportion</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">pi_beta_shape</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
                        <span class="n">pi_beta_scale</span> <span class="o">=</span> <span class="mf">4.5</span><span class="p">,</span> <span class="n">pi_c_shape</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">pi_c_scale</span> <span class="o">=</span> <span class="mf">9.5</span><span class="p">,</span> <span class="n">pi_1_shape</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">pi_1_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span> <span class="n">pi_0_shape</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">pi_0_scale</span> <span class="o">=</span> <span class="mf">9.95</span>
                    <span class="p">),</span> <span class="n">error</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">message</span><span class="p">(</span><span class="s2">&quot;Can&#39;t get a valid LD matrix for &quot;</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
                        <span class="k">return</span><span class="p">(</span><span class="n">NA</span><span class="p">)</span>
                    <span class="p">})</span>
                    <span class="n">end</span><span class="o">.</span><span class="n">time</span> <span class="o">&lt;-</span> <span class="n">Sys</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">NA</span>
                <span class="p">}</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">NA</span>
                <span class="p">}</span>
                <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">NA</span>
                <span class="p">}</span>
                
                <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">result</span><span class="p">[[</span><span class="mi">1</span><span class="p">]]))</span> <span class="p">{</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">nrow</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ncol</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span> <span class="n">NA</span><span class="p">)</span>
                    <span class="n">colnames</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;causal_effect&quot;</span><span class="p">,</span> <span class="s2">&quot;causal_pvalue&quot;</span><span class="p">,</span> <span class="s2">&quot;correlated_pleiotropy_effect&quot;</span><span class="p">,</span> <span class="s2">&quot;sigmabeta&quot;</span><span class="p">,</span> <span class="s2">&quot;sigmaeta&quot;</span><span class="p">,</span> <span class="s2">&quot;sigma_error_1&quot;</span><span class="p">,</span> <span class="s2">&quot;sigma_error_2&quot;</span><span class="p">)</span>
                <span class="p">}</span>
                
                <span class="n">result</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">result</span><span class="p">)[,</span> <span class="n">target</span> <span class="o">:=</span> <span class="n">target</span><span class="p">]</span>
    
                <span class="n">res</span> <span class="o">&lt;-</span> <span class="n">tryCatch</span><span class="p">(</span><span class="n">merge</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;target&quot;</span><span class="p">),</span> <span class="n">error</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                    <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
                <span class="p">})</span>
                <span class="c1"># I am afraid of receiveing error in final result, so I&#39;d like to save the result per target....</span>
                <span class="c1">#fwrite(res, paste0(outpath, &quot;res_mraid_&quot;, qtl, &quot;_&quot;, con, &quot;_p_cut_&quot;, p_cut, &quot;_&quot;, target, &quot;.csv.gz&quot;))</span>
                <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
          <span class="p">}</span>
      <span class="c1">#out&lt;-data.frame()</span>
      <span class="c1">#for (f in c(${_input:r,})) {</span>
       <span class="n">AD_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{AD_df}</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="c1"># qtl=&quot;${qtl}&quot;</span>
      <span class="c1">#      if (qtl == &quot;eQTL&quot;) {</span>
      <span class="c1">#          targets_df&lt;-data.frame(targets=AD_df$ensembl,chromosome=gsub(&quot;chr&quot;,&quot;&quot;,AD_df$chr))</span>
      <span class="c1">#      }</span>
      <span class="c1">#      if (qtl == &quot;pQTL&quot;) {</span>
      <span class="c1">#         targets_df&lt;-data.frame(targets=AD_df$ID,chromosome=gsub(&quot;chr&quot;,&quot;&quot;,AD_df$chr))%&gt;%.[.$targets!=&quot;&quot;,]</span>
      <span class="c1">#      }</span>
      <span class="c1">#      if (qtl == &quot;mQTL&quot;) {</span>
      <span class="c1">#          targets_df&lt;-NA #need to add sth..</span>
      <span class="c1">#      }</span>
      <span class="c1">#      if (qtl == &quot;haQTL&quot;) {</span>
      <span class="c1">#          targets_df&lt;-NA #need to add sth..</span>
      <span class="c1">#      }</span>
      <span class="c1">#      if (qtl == &quot;metaQTL&quot;) {</span>
      <span class="c1">#          targets_df&lt;-data.frame(targets=AD_df$molecular_trait_id,chromosome=AD_df$group)</span>
      <span class="c1">#      }  </span>
       <span class="c1">#target=targets_df$targets[${_n}+1]</span>
  
       <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{targets_df}</span><span class="s2">&quot;</span><span class="p">)</span>
       <span class="n">target</span><span class="o">=</span><span class="n">targets_df</span><span class="err">$</span><span class="n">targets</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">_n</span><span class="p">}]</span>
       <span class="n">eqtl</span><span class="o">.</span><span class="n">mr</span> <span class="o">&lt;-</span> <span class="n">mraid</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">target</span><span class="p">,</span><span class="n">path</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;${cwd:a}&quot;</span><span class="p">,</span><span class="s2">&quot;/&quot;</span><span class="p">),</span><span class="n">qtl</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{qtl}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">p_cut</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">p_cut</span><span class="p">},</span><span class="n">con</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{con}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">targets_df</span><span class="o">=</span><span class="n">targets_df</span><span class="p">,</span><span class="n">pval_beta</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">pval_beta</span><span class="p">},</span><span class="n">outcome</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{out}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">samplen1</span><span class="o">=</span><span class="s1">&#39;$</span><span class="si">{samplen1}</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">samplen2</span><span class="o">=</span><span class="s1">&#39;$</span><span class="si">{samplen2}</span><span class="s1">&#39;</span><span class="p">)</span>
       <span class="c1">#fwrite(eqtl.mr, paste0(&quot;${_output:d}&quot;,&quot;res_mraid_&quot;,&quot;${con}&quot;,&quot;_p_cut_&quot;,&quot;{p_cut}&quot;,f,&quot;.csv&quot;))</span>
       <span class="c1">#out&lt;-rbind(out,eqtl.mr)</span>
      <span class="c1">#}</span>
    <span class="n">fwrite</span><span class="p">(</span><span class="n">eqtl</span><span class="o">.</span><span class="n">mr</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># perform mraid analysis with QTL data</span>
<span class="c1">#[mraid_qtl_1]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1">#parameter: analysis_units = path</span>
<span class="c1">#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_units).readlines() if x.strip() and not x.strip().startswith(&#39;target&#39;)]</span>
<span class="kn">parameter:</span> <span class="n">n</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">n</span>  <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s1">&#39;n&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">qtl</span><span class="si">}</span><span class="s2">/sig_</span><span class="si">{</span><span class="n">p_cut</span><span class="si">}</span><span class="s2">/res_mraid_</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">_p_cut_</span><span class="si">{</span><span class="n">p_cut</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">_n</span><span class="si">}</span><span class="s2">.csv&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;80G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span><span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> 
    <span class="n">library</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">UpSetR</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;data.table&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">RcppCNPy</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">arrow</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">MRAID</span><span class="p">)</span>    
    <span class="n">find_SNP_from_sum</span><span class="o">&lt;-</span><span class="n">function</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span><span class="n">refpath</span><span class="p">,</span><span class="n">SNP</span><span class="p">,</span><span class="n">ID</span><span class="o">=</span><span class="s2">&quot;variant&quot;</span><span class="p">,</span><span class="n">full_chr</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span><span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span><span class="n">ref_with_fullpath</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span><span class="n">IDsep</span><span class="o">=</span><span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="n">suffix</span><span class="o">=</span><span class="s1">&#39;.norminal.cis_long_table.txt&#39;</span><span class="p">,</span><span class="n">IDsep_in_sum</span><span class="o">=</span><span class="s2">&quot;_&quot;</span><span class="p">){</span>
        <span class="nb">chr</span><span class="o">&lt;-</span><span class="n">unique</span><span class="p">(</span><span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">SNP</span><span class="p">,</span><span class="n">IDsep</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="n">T</span><span class="p">)[,</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span><span class="p">(</span><span class="n">isTRUE</span><span class="p">(</span><span class="n">full_chr</span><span class="p">))</span><span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="o">&lt;-</span><span class="nb">chr</span> <span class="k">else</span> <span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="o">&lt;-</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="nb">chr</span><span class="p">)</span>
        <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="p">))</span><span class="k">try</span><span class="p">({</span>
            <span class="n">c</span><span class="o">&lt;-</span><span class="nb">chr</span><span class="o">.</span><span class="n">use</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">use</span><span class="o">.</span><span class="n">ref</span><span class="o">&lt;-</span><span class="n">ref</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;[.]&quot;</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">suffix</span><span class="p">),</span><span class="n">ref</span><span class="p">)]</span>
            <span class="k">if</span><span class="p">(</span><span class="n">isTRUE</span><span class="p">(</span><span class="n">ref_with_fullpath</span><span class="p">)){</span>
                <span class="n">tmp</span><span class="o">.</span><span class="n">f</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="n">use</span><span class="o">.</span><span class="n">ref</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">tmp</span><span class="o">.</span><span class="n">f</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">refpath</span><span class="p">,</span><span class="s2">&quot;/&quot;</span><span class="p">,</span><span class="n">use</span><span class="o">.</span><span class="n">ref</span><span class="p">),</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span><span class="p">(</span><span class="n">stringr</span><span class="p">::</span><span class="n">str_split_fixed</span><span class="p">(</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]],</span><span class="n">IDsep</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">)[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span><span class="o">!=</span><span class="nb">chr</span><span class="p">[</span><span class="n">i</span><span class="p">]){</span>
                <span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]]</span><span class="o">&lt;-</span><span class="n">sub</span><span class="p">(</span><span class="n">IDsep_in_sum</span><span class="p">,</span><span class="n">IDsep</span><span class="p">,</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]])</span>
            <span class="p">}</span>
            <span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="o">&lt;-</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">[[</span><span class="n">ID</span><span class="p">]]</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">SNP</span><span class="p">,]</span>
            <span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="o">&lt;-</span><span class="n">rbind</span><span class="p">(</span><span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="p">,</span><span class="n">tmp</span><span class="o">.</span><span class="n">f</span><span class="p">)</span>    
        <span class="p">})</span>
        <span class="k">return</span><span class="p">(</span><span class="n">snp</span><span class="o">.</span><span class="n">df</span><span class="p">)</span>
    <span class="p">}</span>
        <span class="c1"># mraid function</span>
        <span class="n">mraid</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">target</span><span class="p">,</span><span class="n">path</span><span class="p">,</span> <span class="n">qtl</span><span class="p">,</span> <span class="n">res</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">p_cut</span><span class="p">,</span> <span class="n">targets_df</span><span class="p">,</span> <span class="n">con</span> <span class="p">,</span><span class="n">outcome</span><span class="p">,</span><span class="n">pval_beta</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span><span class="n">samplen1</span><span class="p">,</span><span class="n">samplen2</span><span class="p">)</span> <span class="p">{</span>
        <span class="nb">chr</span><span class="o">=</span><span class="n">targets_df</span><span class="p">[</span><span class="n">targets_df</span><span class="err">$</span><span class="n">targets</span> <span class="o">==</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">]</span><span class="o">%&gt;%</span><span class="n">unique</span><span class="p">()</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;eQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/eQTL/dlpfc_batch_all.rnaseqc.low_expression_filtered.outlier_removed.tmm.expression.bed.processed_phenotype.per_chrom_dlpfc_batch_all.rnaseqc.ROSMAP_covariates.ROSMAP_NIA_WGS.pca.PEER.txt.&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;pQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/pQTL/rf_standard/pheno_recipe_pheno.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.filtered.filtered.prune.pca.resid.PEER.&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
                <span class="c1"># there are so many duplicated rows in exp matrix</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="p">[</span><span class="o">-</span><span class="n">which</span><span class="p">(</span><span class="n">duplicated</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="p">)),</span> <span class="p">]</span>
                <span class="c1"># in pQTL files, the snps are chrx_tacg_... but in most files should be chrx:tacg_...</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">variant</span> <span class="o">&lt;-</span> <span class="n">sub</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">variant</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;mQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/mQTL/ROSMAP_arrayMethylation_covariates.sesame.methyl.beta.sample_matched.bed_BMIQ.bed.filter_na.bed.softImputed.bed.processed_phenotype.per_chrom_methy.peer.pca.chr&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">%&gt;%</span> <span class="n">rename</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span> <span class="o">=</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;haQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/haQTL/h3k9ac_bed_recipe_h3k9ac_whole.k9_cov.xqtl_protocol_data.filtered.related.filtered.extracted.pca.projected.resid.PEER.merged.&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="s2">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">))</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">%&gt;%</span> <span class="n">rename</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span> <span class="o">=</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;metaQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">fread</span><span class="p">(</span><span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;~/Work/MR/2023.4_MR/data/sumtats/metaQTL/&quot;</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span><span class="s2">&quot;/rf/&quot;</span><span class="p">),</span><span class="s2">&quot;_rf.norminal.trans_long_table.txt$&quot;</span><span class="p">,</span><span class="n">full</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">T</span><span class="p">))</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span> <span class="o">%&gt;%</span> <span class="n">rename</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span> <span class="o">=</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span> <span class="o">=</span> <span class="s2">&quot;pos&quot;</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span> <span class="o">&lt;-</span> <span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">chromosome</span><span class="p">)</span>
                <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">position</span> <span class="o">&lt;-</span> <span class="n">str_split</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="err">$</span><span class="n">position</span><span class="p">,</span><span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)[,</span><span class="mi">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">()</span>
            <span class="p">}</span>
                <span class="c1">## 2. prapare the input for MR</span>
                <span class="c1">### 2.1 exposure GWAS sumstats</span>
                <span class="c1"># get qtl Data</span>
                <span class="n">exp</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="o">.</span><span class="n">chr</span><span class="p">[</span><span class="n">molecular_trait_id</span> <span class="o">==</span> <span class="n">target</span><span class="p">]</span>
                <span class="n">snps_ana_gene</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="err">$</span><span class="n">variant</span>
                <span class="k">if</span><span class="p">(</span><span class="n">pval_beta</span><span class="o">==</span><span class="mi">1</span><span class="p">){</span><span class="n">exp</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="p">[</span><span class="n">pval_beta</span> <span class="o">&lt;</span> <span class="n">p_cut</span><span class="p">,</span> <span class="p">]}</span> <span class="k">else</span> <span class="n">exp</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="p">[</span><span class="n">pvalue</span> <span class="o">&lt;</span> <span class="n">p_cut</span><span class="p">,</span> <span class="p">]</span>
                <span class="n">snps_asso_gene</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="err">$</span><span class="n">variant</span>
                <span class="n">res</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span> <span class="n">n_snps_ana</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">snps_ana_gene</span><span class="p">),</span> <span class="n">p_cut</span> <span class="o">=</span> <span class="n">p_cut</span><span class="p">,</span> <span class="n">n_snps_asso</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">))</span>
                <span class="n">message</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">),</span> <span class="s2">&quot; SNPs associated with &quot;</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
                    <span class="c1">### 2.2 outcome GWAS sumstats</span>
                    <span class="c1"># get snps&#39; GWAS data</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">outcome</span> <span class="o">==</span> <span class="s2">&quot;GWAS&quot;</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">path</span> <span class="o">&lt;-</span> <span class="s2">&quot;/home/rf2872/Work/MR/2023.4_MR/data/sumtats/GWAS/&quot;</span>
                        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">find_SNP_from_sum</span><span class="p">(</span><span class="n">ref</span> <span class="o">=</span> <span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">path</span><span class="p">),</span> <span class="n">refpath</span> <span class="o">=</span> <span class="n">path</span><span class="p">,</span> <span class="n">SNP</span> <span class="o">=</span> <span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">suffix</span> <span class="o">=</span> <span class="s2">&quot;.sumstat.tsv&quot;</span><span class="p">,</span> <span class="n">full_chr</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">IDsep_in_sum</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                        <span class="n">out</span><span class="o">.</span><span class="n">path</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;/home/rf2872/Work/MR/2023.4_MR/data/sumtats/&quot;</span><span class="p">,</span><span class="n">outcome</span><span class="p">)</span>
                        <span class="n">out</span> <span class="o">&lt;-</span> <span class="n">find_SNP_from_sum</span><span class="p">(</span><span class="n">ref</span> <span class="o">=</span> <span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">path</span><span class="p">),</span> <span class="n">refpath</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">SNP</span> <span class="o">=</span> <span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">full_chr</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">IDsep_in_sum</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
                        <span class="n">int_snps</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">out</span><span class="err">$</span><span class="n">variant</span><span class="p">)</span>
                    <span class="p">}</span>
                    <span class="n">int_snps</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">snps_asso_gene</span><span class="p">,</span> <span class="n">out</span><span class="err">$</span><span class="n">variant</span><span class="p">)</span>
                    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">int_snps</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">){</span>
                    <span class="c1">### 2.3/2.4 exposure LD matrix</span>
                    <span class="c1"># since the LD files are saved separately accrodding to their postions, I&#39;d like to use bedtools to map the SNPs to LD files</span>
                    <span class="n">LD</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s2">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/LD/output/1300_hg38_EUR_LD_blocks_LD/ROSMAP_NIA_WGS.leftnorm.filtered.filtered.ld.list&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">bed</span> <span class="o">&lt;-</span> <span class="n">str_split</span><span class="p">(</span><span class="n">LD</span><span class="err">$</span><span class="n">V1</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">cbind</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="n">LD</span><span class="p">)</span>
                    
                    <span class="c1"># if(qtl == &quot;eQTL&quot; | qtl == &quot;pQTL&quot;){</span>
                    <span class="n">exp_LD</span><span class="o">.</span><span class="n">bed</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">chrome</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="n">exp</span><span class="err">$</span><span class="n">chromosome</span><span class="p">),</span> <span class="n">start</span> <span class="o">=</span> <span class="n">exp</span><span class="err">$</span><span class="n">position</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="p">(</span><span class="n">exp</span><span class="err">$</span><span class="n">position</span> <span class="o">+</span> <span class="n">nchar</span><span class="p">(</span><span class="n">exp</span><span class="err">$</span><span class="n">ref</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
                    <span class="c1"># }</span>
                    <span class="c1"># if(qtl == &quot;mQTL&quot;){</span>
                    <span class="c1"># exp_LD.bed &lt;- data.frame(chrome = exp$chrom, start = exp$pos, end = (exp$pos + nchar(exp$ref) - 1))</span>
                    <span class="c1"># }</span>
                    <span class="n">options</span><span class="p">(</span><span class="n">bedtools</span><span class="o">.</span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;/home/rf2872/software/bedtools2/bin/&quot;</span><span class="p">)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">files</span> <span class="o">&lt;-</span> <span class="n">bedtoolsr</span><span class="p">::</span><span class="n">bt</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">a</span> <span class="o">=</span> <span class="n">LD</span><span class="o">.</span><span class="n">bed</span><span class="p">,</span> <span class="n">b</span> <span class="o">=</span> <span class="n">exp_LD</span><span class="o">.</span><span class="n">bed</span><span class="p">)</span>
                      
                    <span class="c1"># load python tools to read npz files, but there would be an error for npz1$f[[&quot;arr_1&quot;]], so I need to read the bim file to locate the snps. (Hao says they are in sam order)</span>
                    <span class="n">np</span> <span class="o">&lt;-</span> <span class="n">import</span><span class="p">(</span><span class="s2">&quot;numpy&quot;</span><span class="p">)</span>
                    <span class="n">LD_files</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">files</span><span class="err">$</span><span class="n">V5</span><span class="p">)</span>
                    
                    <span class="c1"># some snp sets could locate in more than 1 ld files. so I have to read those files separately and combine them (I don&#39;t know which function can do this easily, so I just write a new one)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">list</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">()</span>
                    <span class="k">for</span> <span class="p">(</span><span class="n">f</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">LD_files</span><span class="p">))</span> <span class="p">{</span>
                        <span class="n">npz1</span> <span class="o">&lt;-</span> <span class="n">np</span><span class="err">$</span><span class="n">load</span><span class="p">(</span><span class="n">LD_files</span><span class="p">[</span><span class="n">f</span><span class="p">])</span>
                        <span class="n">ld</span><span class="o">.</span><span class="n">mtx</span> <span class="o">&lt;-</span> <span class="n">npz1</span><span class="err">$</span><span class="n">f</span><span class="p">[[</span><span class="s2">&quot;arr_0&quot;</span><span class="p">]]</span>
                        <span class="n">ld</span><span class="o">.</span><span class="n">snps</span> <span class="o">&lt;-</span> <span class="n">stringr</span><span class="p">::</span><span class="n">str_split</span><span class="p">(</span><span class="n">LD_files</span><span class="p">[</span><span class="n">f</span><span class="p">],</span> <span class="s2">&quot;[.]&quot;</span><span class="p">,</span> <span class="n">simplify</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="o">.</span><span class="p">[,</span> <span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="o">.</span><span class="p">),</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="o">.</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))]</span> <span class="o">%&gt;%</span>
                            <span class="n">paste</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="n">paste0</span><span class="p">(</span><span class="o">.</span><span class="p">,</span> <span class="s2">&quot;.bim&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="o">.</span><span class="p">)</span> <span class="o">%&gt;%</span>
                            <span class="o">.</span><span class="p">[,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">colnames</span><span class="p">(</span><span class="n">ld</span><span class="o">.</span><span class="n">mtx</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">ld</span><span class="o">.</span><span class="n">mtx</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">ld</span><span class="o">.</span><span class="n">snps</span>
                        <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">ld</span><span class="o">.</span><span class="n">mtx</span><span class="p">)</span>
                        <span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]),</span> <span class="n">int_snps</span><span class="p">)</span>
                        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]][</span><span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">,</span> <span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span>
                    <span class="n">colnames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]])</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]])</span> <span class="o">&lt;-</span> <span class="n">realsnps</span><span class="o">.</span><span class="n">tmp</span><span class="p">}</span> <span class="k">else</span> <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">f</span><span class="p">]]</span><span class="o">&lt;-</span><span class="n">NA</span> 
                        
                    <span class="p">}</span>      
                    <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[</span><span class="n">sapply</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">,</span><span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nb">all</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">x</span><span class="p">)))]</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
                    
                    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">){</span>
                    <span class="n">u</span> <span class="o">&lt;-</span> <span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="mi">1</span><span class="p">]]</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
                        <span class="k">for</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="mi">2</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">))</span> <span class="p">{</span>
                            <span class="n">t</span><span class="o">.</span><span class="n">col</span> <span class="o">&lt;-</span> <span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">+</span> <span class="n">ncol</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span>
                            <span class="n">u</span><span class="o">.</span><span class="n">n</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">ncol</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">col</span><span class="p">,</span> <span class="n">nrow</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">col</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                            <span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">),</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
                            <span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">[(</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">),</span> <span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span><span class="n">ncol</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span>
                            <span class="n">colnames</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">u</span><span class="o">.</span><span class="n">n</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">u</span><span class="p">),</span> <span class="n">colnames</span><span class="p">(</span><span class="n">LD</span><span class="o">.</span><span class="n">list</span><span class="p">[[</span><span class="n">x</span><span class="p">]]))</span>
                            <span class="n">u</span> <span class="o">&lt;-</span> <span class="n">u</span><span class="o">.</span><span class="n">n</span>
                        <span class="p">}</span>
                    <span class="p">}</span>
                      
                    <span class="c1"># the snps in LD matrix and sumstats files are not exactly same, so I need to get the overlapped ones</span>
                    <span class="n">realsnps</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">u</span><span class="p">),</span> <span class="n">int_snps</span><span class="p">)</span>
                    <span class="n">LD</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">m</span> <span class="o">&lt;-</span> <span class="n">LD</span><span class="o">.</span><span class="n">exp</span><span class="o">.</span><span class="n">m</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">u</span><span class="p">[</span><span class="n">realsnps</span><span class="p">,</span> <span class="n">realsnps</span><span class="p">])</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="p">)]</span>
                    <span class="n">message</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="p">),</span> <span class="s2">&quot; SNPs in &quot;</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot; analyzed in MRAID&quot;</span><span class="p">)</span>
                      
                    <span class="c1">### 2.5 exposure sample size</span>
                    <span class="c1"># from [DLPFC ROSMAP](https://github.com/cumc/fungen-xqtl-analysis/blob/main/data/descriptor/qtl/ROSMAP_DLPFC_expression_qtl.md) (The number on this page is wrong on 4/11/2023, it should be ~950, let&#39;s set 900 for now)</span>
                    <span class="n">samplen1</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">samplen1</span><span class="p">)</span>
                    <span class="c1">### 2.6 outcome sample size</span>
                    <span class="c1"># from [Bellenguez GWAS paper](https://pubmed.ncbi.nlm.nih.gov/35379992/)(111,326 AD cases and 677,663 controls)</span>
                    <span class="n">samplen2</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">samplen2</span><span class="p">)</span>
                    
                      
                    <span class="c1">## 3. RUN MRAID</span>
                    <span class="n">exp</span><span class="o">.</span><span class="n">real</span> <span class="o">&lt;-</span> <span class="n">exp</span><span class="p">[</span><span class="n">exp</span><span class="err">$</span><span class="n">variant</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">realsnps</span><span class="p">,</span> <span class="p">]</span>
                    <span class="n">exp</span><span class="o">.</span><span class="n">z</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">beta</span><span class="p">)</span> <span class="o">/</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">se</span><span class="p">)</span>
                      
                    <span class="n">out</span><span class="o">.</span><span class="n">real</span> <span class="o">&lt;-</span> <span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="err">$</span><span class="n">variant</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">realsnps</span><span class="p">,</span> <span class="p">]</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">z</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">beta</span><span class="p">)</span> <span class="o">/</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">real</span><span class="err">$</span><span class="n">se</span><span class="p">)</span>
                      
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="n">length</span><span class="p">(</span><span class="n">realsnps</span><span class="p">)]</span>
                        
                    <span class="c1"># you can check the parameters info for MRAID  in [here](https://github.com/yuanzhongshang/MRAID/blob/main/man/MRAID.Rd)</span>
                    <span class="n">message</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="s2">&quot; MRAID analysis...&quot;</span><span class="p">)</span>
                      
                    <span class="c1"># running MRAID with a big LD matrix would cost so much time, I would like to document the time used</span>
                    <span class="n">start</span><span class="o">.</span><span class="n">time</span> <span class="o">&lt;-</span> <span class="n">Sys</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">tryCatch</span><span class="p">(</span><span class="n">MRAID</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">z</span><span class="p">,</span> <span class="n">out</span><span class="o">.</span><span class="n">z</span><span class="p">,</span> <span class="n">LD</span><span class="o">.</span><span class="n">exp</span><span class="o">.</span><span class="n">m</span><span class="p">,</span> <span class="n">LD</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">m</span><span class="p">,</span> <span class="n">samplen1</span><span class="p">,</span> <span class="n">samplen2</span><span class="p">,</span>
                        <span class="n">Gibbsnumber</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">burninproportion</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">pi_beta_shape</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
                        <span class="n">pi_beta_scale</span> <span class="o">=</span> <span class="mf">4.5</span><span class="p">,</span> <span class="n">pi_c_shape</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">pi_c_scale</span> <span class="o">=</span> <span class="mf">9.5</span><span class="p">,</span> <span class="n">pi_1_shape</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">pi_1_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span> <span class="n">pi_0_shape</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">pi_0_scale</span> <span class="o">=</span> <span class="mf">9.95</span>
                    <span class="p">),</span> <span class="n">error</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">message</span><span class="p">(</span><span class="s2">&quot;Can&#39;t get a valid LD matrix for &quot;</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
                        <span class="k">return</span><span class="p">(</span><span class="n">NA</span><span class="p">)</span>
                    <span class="p">})</span>
                    <span class="n">end</span><span class="o">.</span><span class="n">time</span> <span class="o">&lt;-</span> <span class="n">Sys</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">NA</span>
                <span class="p">}</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">NA</span>
                <span class="p">}</span>
                <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">res</span><span class="p">[,</span> <span class="n">n_snps_real</span> <span class="o">:=</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">NA</span>
                <span class="p">}</span>
                
                <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">result</span><span class="p">[[</span><span class="mi">1</span><span class="p">]]))</span> <span class="p">{</span>
                    <span class="n">result</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">nrow</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ncol</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span> <span class="n">NA</span><span class="p">)</span>
                    <span class="n">colnames</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;causal_effect&quot;</span><span class="p">,</span> <span class="s2">&quot;causal_pvalue&quot;</span><span class="p">,</span> <span class="s2">&quot;correlated_pleiotropy_effect&quot;</span><span class="p">,</span> <span class="s2">&quot;sigmabeta&quot;</span><span class="p">,</span> <span class="s2">&quot;sigmaeta&quot;</span><span class="p">,</span> <span class="s2">&quot;sigma_error_1&quot;</span><span class="p">,</span> <span class="s2">&quot;sigma_error_2&quot;</span><span class="p">)</span>
                <span class="p">}</span>
                
                <span class="n">result</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">result</span><span class="p">)[,</span> <span class="n">target</span> <span class="o">:=</span> <span class="n">target</span><span class="p">]</span>
    
                <span class="n">res</span> <span class="o">&lt;-</span> <span class="n">tryCatch</span><span class="p">(</span><span class="n">merge</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;target&quot;</span><span class="p">),</span> <span class="n">error</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                    <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
                <span class="p">})</span>
                <span class="c1"># I am afraid of receiveing error in final result, so I&#39;d like to save the result per target....</span>
                <span class="c1">#fwrite(res, paste0(outpath, &quot;res_mraid_&quot;, qtl, &quot;_&quot;, con, &quot;_p_cut_&quot;, p_cut, &quot;_&quot;, target, &quot;.csv.gz&quot;))</span>
                <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
          <span class="p">}</span>
      <span class="c1">#out&lt;-data.frame()</span>
      <span class="c1">#for (f in c(${_input:r,})) {</span>
       <span class="n">AD_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{AD_df}</span><span class="s2">&quot;</span><span class="p">)</span>
       <span class="n">qtl</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{qtl}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;eQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ensembl</span><span class="p">,</span><span class="n">chromosome</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="n">AD_df</span><span class="err">$</span><span class="nb">chr</span><span class="p">))</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;pQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">ID</span><span class="p">,</span><span class="n">chromosome</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="n">AD_df</span><span class="err">$</span><span class="nb">chr</span><span class="p">))</span><span class="o">%&gt;%.</span><span class="p">[</span><span class="o">.</span><span class="err">$</span><span class="n">targets</span><span class="o">!=</span><span class="s2">&quot;&quot;</span><span class="p">,]</span>

            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;mQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">NA</span> <span class="c1">#need to add sth..</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;haQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">NA</span> <span class="c1">#need to add sth..</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">qtl</span> <span class="o">==</span> <span class="s2">&quot;metaQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">targets_df</span><span class="o">&lt;-</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">molecular_trait_id</span><span class="p">,</span><span class="n">chromosome</span><span class="o">=</span><span class="n">AD_df</span><span class="err">$</span><span class="n">group</span><span class="p">)</span>
            <span class="p">}</span>  
       <span class="c1">#target=targets_df$targets[${_n}+1]</span>
       <span class="n">target</span><span class="o">=</span><span class="n">targets_df</span><span class="err">$</span><span class="n">targets</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">_n</span><span class="p">}]</span>
       <span class="n">eqtl</span><span class="o">.</span><span class="n">mr</span> <span class="o">&lt;-</span> <span class="n">mraid</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">target</span><span class="p">,</span><span class="n">path</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;${cwd:a}&quot;</span><span class="p">,</span><span class="s2">&quot;/&quot;</span><span class="p">),</span><span class="n">qtl</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{qtl}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">p_cut</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">p_cut</span><span class="p">},</span><span class="n">con</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{con}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">targets_df</span><span class="o">=</span><span class="n">targets_df</span><span class="p">,</span><span class="n">pval_beta</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">pval_beta</span><span class="p">},</span><span class="n">outcome</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{out}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">samplen1</span><span class="o">=</span><span class="s1">&#39;$</span><span class="si">{samplen1}</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">samplen2</span><span class="o">=</span><span class="s1">&#39;$</span><span class="si">{samplen2}</span><span class="s1">&#39;</span><span class="p">)</span>
       <span class="c1">#fwrite(eqtl.mr, paste0(&quot;${_output:d}&quot;,&quot;res_mraid_&quot;,&quot;${con}&quot;,&quot;_p_cut_&quot;,&quot;{p_cut}&quot;,f,&quot;.csv&quot;))</span>
       <span class="c1">#out&lt;-rbind(out,eqtl.mr)</span>
      <span class="c1">#}</span>
    <span class="n">fwrite</span><span class="p">(</span><span class="n">eqtl</span><span class="o">.</span><span class="n">mr</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">mraid_qtl_3</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">qtl</span><span class="si">}</span><span class="s2">/sig_</span><span class="si">{</span><span class="n">p_cut</span><span class="si">}</span><span class="s2">/all_res_mraid_</span><span class="si">{</span><span class="n">con</span><span class="si">}</span><span class="s2">_p_cut_</span><span class="si">{</span><span class="n">p_cut</span><span class="si">}</span><span class="s2">.csv&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;2h&#39;</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;16G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>

  
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span><span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span>
     <span class="n">cd</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">d</span><span class="p">}</span>
     <span class="n">echo</span> <span class="s1">&#39;target,n_snps_ana,p_cut,n_snps_asso,n_snps_real,causal_effect,causal_pvalue,correlated_pleiotropy_effect,sigmabeta,sigmaeta,sigma_error_1,sigma_error_2&#39;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">b</span><span class="p">}</span>
     <span class="n">cat</span> <span class="o">*.</span><span class="n">csv</span><span class="o">|</span><span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">n_snps_ana</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">b</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/trans_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="polyfun.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Fine-mapping with PolyFun</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../enrichment/gregor.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">GREGOR enrichment analysis</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>