{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8595abca-c08a-4e57-9718-29414df835e9",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## Generate posterior output for mash\n",
    "This notebook is a extraction of the original mashr_flashr_workflow.ipynb, the reason for extracting this part is otherwise the mashr_flashr_workflow will run the MASH steps again when ask to only compute the covariate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b09a4103-8c6e-4b7f-90c7-b08bbf7bfa71",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Input and output\n",
    "Input: \n",
    "    A path to:\n",
    "        1. The Vhat RDS object generated by mashr_flashr_workflow\n",
    "        2. The model RDS object generated by mashr_flashr_workflow\n",
    "    A space connected, collections of path to the data object of each of the gene\n",
    "\n",
    "Output: \n",
    "    A list of path to a RDS object which contains\n",
    "        1. the output of mash_compute_posterior_matrices function based on the input.\n",
    "        2. The snps names in the form of chr:pos_alt_ref for each of the variant in the input."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "040ad665-81dd-4f45-b717-03099bd42bb1",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "parameter: cwd = path('./mashr_flashr_workflow_output')\n",
    "# Input summary statistics data\n",
    "parameter: data = path(\"fastqtl_to_mash_output/FastQTLSumStats.mash.rds\")\n",
    "# Prefix of output files. If not specified, it will derive it from data.\n",
    "# If it is specified, for example, `--output-prefix AnalysisResults`\n",
    "# It will save output files as `{cwd}/AnalysisResults*`.\n",
    "parameter: output_prefix = ''\n",
    "# Exchangable effect (EE) or exchangable z-scores (EZ)\n",
    "parameter: effect_model = 'EZ'\n",
    "# Identifier of $\\hat{V}$ estimate file\n",
    "# Options are \"identity\", \"simple\", \"mle\", \"vhat_corshrink_xcondition\", \"vhat_simple_specific\"\n",
    "parameter: vhat = 'simple'\n",
    "parameter: mixture_components = ['flash', 'flash_nonneg', 'pca',\"canonical\"]\n",
    "parameter: container = str\n",
    "data = data.absolute()\n",
    "cwd = cwd.absolute()\n",
    "if len(output_prefix) == 0:\n",
    "    output_prefix = f\"{data:bn}\"\n",
    "prior_data = file_target(f\"{cwd:a}/{output_prefix}.{effect_model}.prior.rds\")\n",
    "vhat_data = file_target(f\"{cwd:a}/{output_prefix}.{effect_model}.V_{vhat}.rds\")\n",
    "mash_model = file_target(f\"{cwd:a}/{output_prefix}.{effect_model}.V_{vhat}.mash_model.rds\")\n",
    "\n",
    "def sort_uniq(seq):\n",
    "    seen = set()\n",
    "    return [x for x in seq if not (x in seen or seen.add(x))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ef5356b2-109f-4873-a5bb-c623c2464a9a",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Apply posterior calculations\n",
    "[posterior_1]\n",
    "parameter: mash_model = path\n",
    "parameter: posterior_input = paths()\n",
    "parameter: posterior_vhat_files = paths()\n",
    "# eg, if data is saved in R list as data$strong, then\n",
    "# when you specify `--data-table-name strong` it will read the data as\n",
    "# readRDS('{_input:r}')$strong\n",
    "parameter: data_table_name = ''\n",
    "parameter: bhat_table_name = 'bhat'\n",
    "parameter: shat_table_name = 'sbhat'\n",
    "\n",
    "skip_if(len(posterior_input) == 0, msg = \"No posterior input data to compute on. Please specify it using --posterior-input.\")\n",
    "fail_if(len(posterior_vhat_files) > 1 and len(posterior_vhat_files) != len(posterior_input), msg = \"length of --posterior-input and --posterior-vhat-files do not agree.\")\n",
    "for p in posterior_input:\n",
    "    fail_if(not p.is_file(), msg = f'Cannot find posterior input file ``{p}``')\n",
    "\n",
    "depends: R_library(\"mashr\")\n",
    "input: posterior_input, group_by = 1\n",
    "output: f\"{cwd}/mash_rds/{_input:bn}.posterior.rds\"\n",
    "task: trunk_workers = 1, walltime = '20h', trunk_size = 1, mem = '20G', cores = 1, tags = f'{_output:bn}'\n",
    "R: expand = \"${ }\", workdir = cwd, stderr = f\"{_output:n}.stderr\", stdout = f\"{_output:n}.stdout\"\n",
    "    library(mashr)\n",
    "    data = readRDS(${_input:r})${('$' + data_table_name) if data_table_name else ''}\n",
    "    vhat = readRDS(\"${vhat_data if len(posterior_vhat_files) == 0 else posterior_vhat_files[_index]}\")\n",
    "    mash_data = mash_set_data(data$${bhat_table_name}, Shat=data$${shat_table_name}, alpha=${1 if effect_model == 'EZ' else 0}, V=vhat, zero_Bhat_Shat_reset = 1E3)\n",
    "    mash_output = mash_compute_posterior_matrices(readRDS(${mash_model:r}), mash_data)\n",
    "    mash_output$snps = data$snps\n",
    "    saveRDS(mash_output, ${_output:r})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3123954c-0b37-4c3b-8bf5-4dea20a9efc6",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[posterior_2]\n",
    "input: group_by = \"all\"\n",
    "output:f\"{cwd}/mash_output_list\"\n",
    "python: expand = \"$[ ]\", workdir = cwd, stderr = f\"{_output:n}.stderr\", stdout = f\"{_output:n}.stdout\"\n",
    "    import pandas as pd\n",
    "    pd.DataFrame({\"#mash_result\" :  [$[_input:ar,]] }).to_csv(\"$[_output]\",index = False ,header = False, sep = \"t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.22.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
