
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Multivariate SuSiE and ENLOC model &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Functional annotation integration" href="../functional_annotation.html" />
    <link rel="prev" title="Colocalization with FastENLOC" href="fastenloc.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   ADSP-FG xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Overview of design
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/integrative_analysis/COLOC/mvenloc.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/integrative_analysis/COLOC/mvenloc.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/integrative_analysis/COLOC/mvenloc.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#aim">
   Aim
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods-overview">
   Methods overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#summary">
     Summary
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#implementation">
   Implementation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-0-data-formatting">
     Step 0: data formatting
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#extract-common-snps-between-the-gwas-summary-statistics-and-eqtl-data">
       Extract common SNPS between the GWAS summary statistics and eQTL data
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#extract-common-snps-between-the-summary-statistics-and-ld">
       Extract common SNPS between the summary statistics and LD
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-1-fine-mapping">
     Step 1: fine-mapping
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-2-fine-mapping-results-processing">
     Step 2: fine-mapping results processing
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#construct-eqtl-annotation-file-using-eqtl-snp-pips-and-credible-sets">
       Construct eQTL annotation file using eQTL SNP PIPs and credible sets
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#export-gwas-pip">
       Export GWAS PIP
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-3-colocalization-with-fastenloc">
     Step 3: Colocalization with FastEnloc
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="multivariate-susie-and-enloc-model">
<h1>Multivariate SuSiE and ENLOC model<a class="headerlink" href="#multivariate-susie-and-enloc-model" title="Permalink to this headline">¶</a></h1>
<div class="section" id="aim">
<h2>Aim<a class="headerlink" href="#aim" title="Permalink to this headline">¶</a></h2>
<p>This notebook aims to demonstrate a workflow of generating posterior inclusion probabilities (PIPs) from GWAS summary statistics using SuSiE regression and construsting SNP signal clusters from global eQTL analysis data obtained from multivariate SuSiE models.</p>
</div>
<div class="section" id="methods-overview">
<h2>Methods overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">¶</a></h2>
<p>This procedure assumes that molecular phenotype summary statistics and GWAS summary statistics are aligned and harmonized to have consistent allele coding (see <span class="xref myst">this module</span> for implementation details). Both molecular phenotype QTL and GWAS should be fine-mapped beforehand using mvSusiE or SuSiE. We further assume (and require) that molecular phenotype and GWAS data come from the same population ancestry. Violations from this assumption may not cause an error in the analysis computational workflow but the results obtained may not be valid.</p>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>GWAS Summary Statistics with the following columns:</p>
<ul class="simple">
<li><p>chr: chromosome number</p></li>
<li><p>bp: base pair position</p></li>
<li><p>a1: effect allele</p></li>
<li><p>a2: other allele</p></li>
<li><p>beta: effect size</p></li>
<li><p>se: standard error of beta</p></li>
<li><p>z: z score</p></li>
</ul>
</li>
<li><p>eQTL data from multivariate SuSiE model with the following columns:</p>
<ul class="simple">
<li><p>chr: chromosome number</p></li>
<li><p>bp: base pair position</p></li>
<li><p>a1: effect allele</p></li>
<li><p>a2: other allele</p></li>
<li><p>pip: posterior inclusion probability</p></li>
</ul>
</li>
<li><p>LD correlation matrix</p></li>
</ol>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>Intermediate files:</p>
<ol class="simple">
<li><p>GWAS PIP file with the following columns</p>
<ul class="simple">
<li><p>var_id</p></li>
<li><p>ld_block</p></li>
<li><p>snp_pip</p></li>
<li><p>block_pip</p></li>
</ul>
</li>
<li><p>eQTL annotation file with the following columns</p>
<ul class="simple">
<li><p>chr</p></li>
<li><p>bp</p></li>
<li><p>var_id</p></li>
<li><p>a1</p></li>
<li><p>a2</p></li>
<li><p>annotations, in the format: <code class="docutils literal notranslate"><span class="pre">gene:cs_num&#64;tissue=snp_pip[cs_pip:cs_total_snps]</span></code></p></li>
</ul>
</li>
</ol>
<p>Final Outputs:</p>
<ol class="simple">
<li><p>Enrichment analysis result prefix.enloc.enrich.rst: estimated enrichment parameters and standard errors.</p></li>
<li><p>Signal-level colocalization result prefix.enloc.sig.out: the main output from the colocalization analysis with the following format</p>
<ul class="simple">
<li><p>column 1: signal cluster name (from eQTL analysis)</p></li>
<li><p>column 2: number of member SNPs</p></li>
<li><p>column 3: cluster PIP of eQTLs</p></li>
<li><p>column 4: cluster PIP of GWAS hits (without eQTL prior)</p></li>
<li><p>column 5: cluster PIP of GWAS hits (with eQTL prior)</p></li>
<li><p>column 6: regional colocalization probability (RCP)</p></li>
</ul>
</li>
<li><p>SNP-level colocalization result prefix.enloc.snp.out: SNP-level colocalization output with the following form at</p>
<ul class="simple">
<li><p>column 1: signal cluster name</p></li>
<li><p>column 2: SNP name</p></li>
<li><p>column 3: SNP-level PIP of eQTLs</p></li>
<li><p>column 4: SNP-level PIP of GWAS (without eQTL prior)</p></li>
<li><p>column 5: SNP-level PIP of GWAS (with eQTL prior)</p></li>
<li><p>column 6: SNP-level colocalization probability</p></li>
</ul>
</li>
<li><p>Sorted list of colocalization signals</p></li>
</ol>
<p>Takes into consideration 3 situations:</p>
<ol class="simple">
<li><p>“Major” and “minor” alleles flipped</p></li>
<li><p>Different strand but same variant</p></li>
<li><p>Remove variants with A/T and C/G alleles due to ambiguity</p></li>
</ol>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">mvenloc</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">merge</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">eqtl</span><span class="o">-</span><span class="n">sumstats</span> <span class="o">..</span> \
    <span class="o">--</span><span class="n">gwas</span><span class="o">-</span><span class="n">sumstats</span> <span class="o">..</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">mvenloc</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">eqtl</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">sumstats</span><span class="o">-</span><span class="n">file</span> <span class="o">..</span> \
    <span class="o">--</span><span class="n">ld</span><span class="o">-</span><span class="n">region</span> <span class="o">..</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">mvenloc</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">gwas</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">sumstats</span><span class="o">-</span><span class="n">file</span> <span class="o">..</span> \
    <span class="o">--</span><span class="n">ld</span><span class="o">-</span><span class="n">region</span> <span class="o">..</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">mvenloc</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">enloc</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">eqtl</span><span class="o">-</span><span class="n">pip</span> <span class="o">..</span> \
    <span class="o">--</span><span class="n">gwas</span><span class="o">-</span><span class="n">pip</span> <span class="o">..</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="summary">
<h3>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">enloc</span><span class="o">.</span><span class="n">enrich</span><span class="o">.</span><span class="n">out</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">enloc</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">out</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">enloc</span><span class="o">.</span><span class="n">snp</span><span class="o">.</span><span class="n">out</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">mvenloc</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="implementation">
<h2>Implementation<a class="headerlink" href="#implementation" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="step-0-data-formatting">
<h3>Step 0: data formatting<a class="headerlink" href="#step-0-data-formatting" title="Permalink to this headline">¶</a></h3>
<div class="section" id="extract-common-snps-between-the-gwas-summary-statistics-and-eqtl-data">
<h4>Extract common SNPS between the GWAS summary statistics and eQTL data<a class="headerlink" href="#extract-common-snps-between-the-gwas-summary-statistics-and-eqtl-data" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">merger</span><span class="p">]</span>
<span class="c1"># eQTL summary statistics as a list of RData</span>
<span class="kn">parameter:</span> <span class="n">eqtl_sumstats</span> <span class="o">=</span> <span class="n">path</span> 
<span class="c1"># GWAS summary stats in gz format</span>
<span class="kn">parameter:</span> <span class="n">gwas_sumstats</span> <span class="o">=</span> <span class="n">path</span> 
<span class="kn">input:</span> <span class="n">eqtl_sumstats</span><span class="p">,</span> <span class="n">gwas_sumstats</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">eqtl_sumstats</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">.standardized.gz&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gwas_sumstats</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">.standardized.gz&quot;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span>

    <span class="c1">###</span>
    <span class="c1"># functions</span>
    <span class="c1">###</span>

    <span class="n">allele</span><span class="o">.</span><span class="n">qc</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">ref1</span><span class="p">,</span><span class="n">ref2</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># a1 and a2 are the first data-set</span>
        <span class="c1"># ref1 and ref2 are the 2nd data-set</span>
        <span class="c1"># Make all the alleles into upper-case, as A,T,C,G:</span>
            <span class="n">a1</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">a1</span><span class="p">)</span>
            <span class="n">a2</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">a2</span><span class="p">)</span>
            <span class="n">ref1</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">ref1</span><span class="p">)</span>
            <span class="n">ref2</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">ref2</span><span class="p">)</span>
        <span class="c1"># Strand flip, to change the allele representation in the 2nd data-set</span>
        <span class="n">strand_flip</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">flip</span> <span class="o">=</span> <span class="n">ref</span>
            <span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;A&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;T&quot;</span>
            <span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
            <span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;G&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span>
            <span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;G&quot;</span>
            <span class="n">flip</span>
        <span class="p">}</span>
        <span class="n">flip1</span> <span class="o">=</span> <span class="n">strand_flip</span><span class="p">(</span><span class="n">ref1</span><span class="p">)</span>
        <span class="n">flip2</span> <span class="o">=</span> <span class="n">strand_flip</span><span class="p">(</span><span class="n">ref2</span><span class="p">)</span>
        <span class="n">snp</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="c1"># Remove strand ambiguous SNPs (scenario 3)</span>
        <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="err">!</span><span class="p">((</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;A&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;T&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;T&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;A&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;C&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;G&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;G&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;C&quot;</span><span class="p">))</span>
        <span class="c1"># Remove non-ATCG coding</span>
        <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]][</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;A&quot;</span> <span class="o">&amp;</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;T&quot;</span> <span class="o">&amp;</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;G&quot;</span> <span class="o">&amp;</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;C&quot;</span> <span class="p">]</span> <span class="o">=</span> <span class="n">F</span>
        <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]][</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;A&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;T&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;G&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;C&quot;</span> <span class="p">]</span> <span class="o">=</span> <span class="n">F</span>
        <span class="c1"># as long as scenario 1 is involved, sign_flip will return TRUE</span>
        <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;sign_flip&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">ref2</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">ref1</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">flip2</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">flip1</span><span class="p">)</span>
        <span class="c1"># as long as scenario 2 is involved, strand_flip will return TRUE</span>
        <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;strand_flip&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">flip1</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">flip2</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">flip2</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">flip1</span><span class="p">)</span>
        <span class="c1"># remove other cases, eg, tri-allelic, one dataset is A C, the other is A G, for example.</span>
        <span class="n">exact_match</span> <span class="o">=</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">ref1</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">ref2</span><span class="p">)</span> 
        <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]][</span><span class="err">!</span><span class="p">(</span><span class="n">exact_match</span> <span class="o">|</span> <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;sign_flip&quot;</span><span class="p">]]</span> <span class="o">|</span> <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;strand_flip&quot;</span><span class="p">]])]</span> <span class="o">=</span> <span class="n">F</span>
        <span class="k">return</span><span class="p">(</span><span class="n">snp</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Extract information from RData</span>
    <span class="n">eqtl</span><span class="o">.</span><span class="n">split</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">eqtl</span><span class="p">){</span>
      <span class="n">rows</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">eqtl</span><span class="p">)</span>
      <span class="nb">chr</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
      <span class="n">pos</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
      <span class="n">a1</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
      <span class="n">a2</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">rows</span><span class="p">){</span>
        <span class="n">split1</span> <span class="o">=</span> <span class="n">str_split</span><span class="p">(</span><span class="n">eqtl</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s2">&quot;:&quot;</span><span class="p">)</span>
        <span class="n">split2</span> <span class="o">=</span> <span class="n">str_split</span><span class="p">(</span><span class="n">split1</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="nb">chr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">=</span> <span class="n">split1</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">split2</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">split2</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">split2</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">3</span><span class="p">]</span>

      <span class="p">}</span>
      <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">eqtl</span><span class="p">,</span><span class="nb">chr</span><span class="p">,</span><span class="n">pos</span><span class="p">,</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">remove</span><span class="o">.</span><span class="n">dup</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">df</span><span class="p">){</span>
      <span class="n">df</span> <span class="o">=</span> <span class="n">df</span> <span class="o">%&gt;%</span> <span class="n">arrange</span><span class="p">(</span><span class="n">PosGRCh37</span><span class="p">,</span> <span class="o">-</span><span class="n">N</span><span class="p">)</span>
      <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">df</span><span class="err">$</span><span class="n">PosGRCh37</span><span class="p">),]</span>
      <span class="k">return</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="c1">###</span>
    <span class="c1"># Code</span>
    <span class="c1">###</span>
    
    <span class="c1"># gene regions: </span>
    <span class="c1"># 1 = ENSG00000203710</span>
    <span class="c1"># 2 = ENSG00000064687</span>
    <span class="c1"># 3 = ENSG00000203710</span>
    
    <span class="c1"># eqtl</span>
    <span class="n">gene</span><span class="o">.</span><span class="kp">name</span> <span class="o">=</span> <span class="n">scan</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">what</span><span class="o">=</span><span class="s1">&#39;character&#39;</span><span class="p">)</span>

    <span class="c1"># initial filter of gwas variants that are in eqtl</span>
    <span class="n">gwas</span> <span class="o">=</span> <span class="n">gwas_sumstats</span>
    <span class="n">gwas_filter</span> <span class="o">=</span> <span class="n">gwas</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">gwas</span><span class="err">$</span><span class="nb">id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">var</span><span class="p">),]</span>

    <span class="c1"># create eqtl df</span>
    <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">eqtl</span><span class="err">$</span><span class="n">var</span><span class="p">)</span>

    <span class="c1"># allele flip</span>
    <span class="n">f_gwas</span> <span class="o">=</span> <span class="n">gwas</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="nb">chr</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="err">$</span><span class="nb">chr</span> <span class="o">&amp;</span> <span class="n">PosGRCh37</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="err">$</span><span class="n">pos</span><span class="p">)</span>
    <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">pos</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">PosGRCh37</span><span class="p">)</span>

    <span class="c1"># check if there are duplicate pos</span>
    <span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">f_gwas</span><span class="err">$</span><span class="n">PosGRCh37</span><span class="p">))</span>

    <span class="c1"># multiple snps with same pos</span>
    <span class="n">dup</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">f_gwas</span> <span class="o">%&gt;%</span> <span class="n">group_by</span><span class="p">(</span><span class="n">PosGRCh37</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">n</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> 

    <span class="n">f_gwas</span> <span class="o">=</span> <span class="n">remove</span><span class="o">.</span><span class="n">dup</span><span class="p">(</span><span class="n">f_gwas</span><span class="p">)</span>

    <span class="n">qc</span> <span class="o">=</span> <span class="n">allele</span><span class="o">.</span><span class="n">qc</span><span class="p">(</span><span class="n">f_gwas</span><span class="err">$</span><span class="n">testedAllele</span><span class="p">,</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">otherAllele</span><span class="p">,</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a1</span><span class="p">,</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a2</span><span class="p">)</span>
    <span class="n">keep</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">)</span>
    <span class="n">sign</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">)</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">)</span>

    <span class="c1"># sign flip</span>
    <span class="n">f_gwas</span><span class="err">$</span><span class="n">z</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">z</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span>
    <span class="n">f_gwas</span><span class="err">$</span><span class="n">testedAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a1</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span>
    <span class="n">f_gwas</span><span class="err">$</span><span class="n">otherAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a2</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span>

    <span class="n">f_gwas</span><span class="err">$</span><span class="n">testedAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a1</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span>
    <span class="n">f_gwas</span><span class="err">$</span><span class="n">otherAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a2</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span>

    <span class="c1"># remove ambigiuous </span>
    <span class="k">if</span> <span class="p">(</span> <span class="nb">sum</span><span class="p">(</span><span class="err">!</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">)</span> <span class="p">{</span>
      <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">,]</span>
      <span class="n">f_gwas</span> <span class="o">=</span> <span class="n">f_gwas</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">,]</span>
    <span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="extract-common-snps-between-the-summary-statistics-and-ld">
<h4>Extract common SNPS between the summary statistics and LD<a class="headerlink" href="#extract-common-snps-between-the-summary-statistics-and-ld" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">eqtl_1</span><span class="p">,</span> <span class="n">gwas_1</span> <span class="p">(</span><span class="nb">filter</span> <span class="n">LD</span> <span class="n">file</span> <span class="ow">and</span> <span class="n">sumstat</span> <span class="n">file</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">sumstat_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># LD and region information: chr, start, end, LD file</span>
<span class="kn">paramter:</span> <span class="n">ld_region</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">sumstat_file</span><span class="p">,</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s1">&#39;ld_region&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sumstat_file</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">.z.rds&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sumstat_file</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">.ld.rds&quot;</span>
<span class="kn">R:</span>
    <span class="c1"># FIXME: need to filter both ways for sumstats and for LD</span>
    <span class="c1"># lds filtered</span>
  
    <span class="n">eqtl_id</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="n">var</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">eqtl</span><span class="p">)</span>
    <span class="n">ld_f</span> <span class="o">=</span> <span class="n">ld</span><span class="p">[</span><span class="n">eqtl_id</span><span class="p">,</span> <span class="n">eqtl_id</span><span class="p">]</span>

    <span class="c1"># ld missing</span>
    <span class="n">miss</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">ld_f</span><span class="p">),</span> <span class="n">arr</span><span class="o">.</span><span class="n">ind</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
    <span class="n">miss_r</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">miss</span><span class="p">)</span><span class="err">$</span><span class="n">row</span><span class="p">)</span>
    <span class="n">miss_c</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">miss</span><span class="p">)</span><span class="err">$</span><span class="n">col</span><span class="p">)</span>

    <span class="n">total_miss</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="n">union</span><span class="p">(</span><span class="n">miss_r</span><span class="p">,</span><span class="n">miss_c</span><span class="p">))</span>
    <span class="c1"># FIXME: LD should not have missing data if properly processed by our pipeline</span>
    <span class="c1"># In the future we should throw an error when it happens</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">total_miss</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">){</span>
    <span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f</span><span class="p">[</span><span class="o">-</span><span class="n">total_miss</span><span class="p">,]</span>
    <span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f2</span><span class="p">[,</span><span class="o">-</span><span class="n">total_miss</span><span class="p">]</span>
    <span class="n">dim</span><span class="p">(</span><span class="n">ld_f2</span><span class="p">)</span>
    <span class="p">}</span><span class="k">else</span><span class="p">{</span><span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f</span><span class="p">}</span>
    
    <span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">f_gwas</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="nb">id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">eqtl</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-1-fine-mapping">
<h3>Step 1: fine-mapping<a class="headerlink" href="#step-1-fine-mapping" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">eqtl_2</span><span class="p">,</span> <span class="n">gwas_2</span> <span class="p">(</span><span class="n">finemapping</span><span class="p">)]</span>
<span class="c1"># FIXME: RDS file should have included region information</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">nn</span><span class="si">}</span><span class="s2">.susieR.rds&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">nn</span><span class="si">}</span><span class="s2">.susieR_plot.rds&quot;</span>
<span class="kn">R:</span>
    <span class="n">susie_results</span> <span class="o">=</span> <span class="n">susieR</span><span class="p">::</span><span class="n">susie_rss</span><span class="p">(</span><span class="n">z</span> <span class="o">=</span> <span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">z</span><span class="p">,</span><span class="n">R</span> <span class="o">=</span> <span class="n">ld_f2</span><span class="p">,</span> <span class="n">check_prior</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
    <span class="n">susieR</span><span class="p">::</span><span class="n">susie_plot</span><span class="p">(</span><span class="n">susie_results</span><span class="p">,</span><span class="s2">&quot;PIP&quot;</span><span class="p">)</span>
    <span class="n">susie_results</span><span class="err">$</span><span class="n">z</span> <span class="o">=</span> <span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">z</span>
    <span class="n">susieR</span><span class="p">::</span><span class="n">susie_plot</span><span class="p">(</span><span class="n">susie_results</span><span class="p">,</span><span class="s2">&quot;z_original&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="step-2-fine-mapping-results-processing">
<h3>Step 2: fine-mapping results processing<a class="headerlink" href="#step-2-fine-mapping-results-processing" title="Permalink to this headline">¶</a></h3>
<div class="section" id="construct-eqtl-annotation-file-using-eqtl-snp-pips-and-credible-sets">
<h4>Construct eQTL annotation file using eQTL SNP PIPs and credible sets<a class="headerlink" href="#construct-eqtl-annotation-file-using-eqtl-snp-pips-and-credible-sets" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">eqtl_3</span> <span class="p">(</span><span class="n">create</span> <span class="n">signal</span> <span class="n">cluster</span> <span class="n">using</span> <span class="n">CS</span><span class="p">)]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">nn</span><span class="si">}</span><span class="s2">.enloc_annot.gz&quot;</span>
<span class="kn">R:</span>
    <span class="n">cs</span> <span class="o">=</span> <span class="n">eqtl</span><span class="p">[[</span><span class="s2">&quot;sets&quot;</span><span class="p">]][[</span><span class="s2">&quot;cs&quot;</span><span class="p">]][[</span><span class="s2">&quot;L1&quot;</span><span class="p">]]</span>
    <span class="n">o_id</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="n">var</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">eqtl</span><span class="p">)</span>
    <span class="n">pip</span> <span class="o">=</span> <span class="n">eqtl</span><span class="err">$</span><span class="n">pip</span><span class="p">[</span><span class="n">o_id</span><span class="p">]</span>
    <span class="n">eqtl_annot</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span><span class="p">,</span> <span class="n">pip</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">mutate</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="kp">name</span><span class="p">,</span><span class="n">cluster</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">cluster_pip</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">total_snps</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">for</span><span class="p">(</span><span class="n">snp</span> <span class="ow">in</span> <span class="n">cs</span><span class="p">){</span>
      <span class="n">eqtl_annot</span><span class="err">$</span><span class="n">cluster</span><span class="p">[</span><span class="n">snp</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
      <span class="n">eqtl_annot</span><span class="err">$</span><span class="n">cluster_pip</span><span class="p">[</span><span class="n">snp</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="p">[[</span><span class="s2">&quot;sets&quot;</span><span class="p">]][[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]]</span>
       <span class="n">eqtl_annot</span><span class="err">$</span><span class="n">total_snps</span><span class="p">[</span><span class="n">snp</span><span class="p">]</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">eqtl_annot1</span> <span class="o">=</span> <span class="n">eqtl_annot</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">cluster</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">%&gt;%</span> 
      <span class="n">mutate</span><span class="p">(</span><span class="n">annot</span> <span class="o">=</span> <span class="n">sprintf</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">:</span><span class="si">%d</span><span class="s2">@=</span><span class="si">%e</span><span class="s2">[</span><span class="si">%e</span><span class="s2">:</span><span class="si">%d</span><span class="s2">]&quot;</span><span class="p">,</span><span class="n">gene</span><span class="p">,</span><span class="n">cluster</span><span class="p">,</span><span class="n">pip</span><span class="p">,</span><span class="n">cluster_pip</span><span class="p">,</span><span class="n">total_snps</span><span class="p">))</span> <span class="o">%&gt;%</span>
      <span class="n">select</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="nb">chr</span><span class="p">,</span><span class="n">pos</span><span class="p">,</span><span class="n">eqtl</span><span class="p">,</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">annot</span><span class="p">))</span>
      
    
    <span class="c1"># FIXME: repeats whole process (extracting+fine-mapping+cs creation) 3 times before this next step </span>
    <span class="n">eqtl_annot_comb</span> <span class="o">=</span> <span class="n">rbind</span><span class="p">(</span><span class="n">eqtl_annot3</span><span class="p">,</span> <span class="n">eqtl_annot1</span><span class="p">,</span> <span class="n">eqtl_annot2</span><span class="p">)</span>

    <span class="c1"># FIXME: write to a zip file</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">eqtl_annot_comb</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;eqtl.annot.txt&quot;</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="export-gwas-pip">
<h4>Export GWAS PIP<a class="headerlink" href="#export-gwas-pip" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">gwas_3</span> <span class="p">(</span><span class="nb">format</span> <span class="n">PIP</span> <span class="n">into</span> <span class="n">enloc</span> <span class="n">GWAS</span> <span class="nb">input</span><span class="p">)]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">nn</span><span class="si">}</span><span class="s2">.enloc_gwas.gz&quot;</span>
<span class="kn">R:</span>
    <span class="n">gwas_annot1</span> <span class="o">=</span> <span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span> <span class="o">%&gt;%</span> <span class="n">mutate</span><span class="p">(</span><span class="n">pip</span> <span class="o">=</span> <span class="n">susie_results</span><span class="err">$</span><span class="n">pip</span><span class="p">)</span>
    
    <span class="c1"># FIXME: repeat whole process (extracting common snps + fine-mapping) 3 times before the next steps</span>
    
    <span class="n">gwas_annot_comb</span> <span class="o">=</span> <span class="n">rbind</span><span class="p">(</span><span class="n">gwas_annot3</span><span class="p">,</span> <span class="n">gwas_annot1</span><span class="p">,</span> <span class="n">gwas_annot2</span><span class="p">)</span>
    <span class="n">gwas_loc_annot</span> <span class="o">=</span> <span class="n">gwas_annot_comb</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="nb">id</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="n">PosGRCh37</span><span class="p">,</span><span class="n">z</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">gwas_loc_annot</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;loc.gwas.txt&quot;</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>

<span class="kn">bash:</span>    
    <span class="n">perl</span> <span class="n">format2torus</span><span class="o">.</span><span class="n">pl</span> <span class="n">loc</span><span class="o">.</span><span class="n">gwas</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="n">loc2</span><span class="o">.</span><span class="n">gwas</span><span class="o">.</span><span class="n">txt</span>
    
<span class="kn">R:</span>
    <span class="n">loc</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="s2">&quot;loc2.gwas.txt&quot;</span><span class="p">)</span>
    <span class="n">loc</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[[</span><span class="s2">&quot;V2&quot;</span><span class="p">]]</span>
    
    <span class="n">gwas_annot_comb2</span> <span class="o">=</span> <span class="n">gwas_annot_comb</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="nb">id</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="n">PosGRCh37</span><span class="p">,</span><span class="n">pip</span><span class="p">)</span>
    <span class="n">gwas_annot_comb2</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">gwas_annot_comb2</span><span class="p">,</span> <span class="n">loc</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="nb">id</span><span class="p">,</span> <span class="n">loc</span><span class="p">,</span> <span class="n">pip</span><span class="p">)</span>
    
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">gwas_annot_comb2</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;gwas.pip.txt&quot;</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>

<span class="kn">bash:</span>
    <span class="n">perl</span> <span class="n">format2torus</span><span class="o">.</span><span class="n">pl</span> <span class="n">gwas</span><span class="o">.</span><span class="n">pip</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">gzip</span> <span class="o">--</span><span class="n">best</span> <span class="o">&gt;</span> <span class="n">gwas</span><span class="o">.</span><span class="n">pip</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="step-3-colocalization-with-fastenloc">
<h3>Step 3: Colocalization with FastEnloc<a class="headerlink" href="#step-3-colocalization-with-fastenloc" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">enloc</span><span class="p">]</span>
<span class="c1"># eQTL summary statistics as a list of RData</span>
<span class="c1"># FIXME: to replace later</span>
<span class="kn">parameter:</span> <span class="n">eqtl_pip</span> <span class="o">=</span> <span class="n">path</span> 
<span class="c1"># GWAS summary stats in gz format</span>
<span class="kn">parameter:</span> <span class="n">gwas_pip</span> <span class="o">=</span> <span class="n">path</span> 
<span class="kn">input:</span> <span class="n">eqtl_pip</span><span class="p">,</span> <span class="n">gwas_pip</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">eqtl_pip</span><span class="si">:</span><span class="s2">bnn</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">gwas_pip</span><span class="si">:</span><span class="s2">bnn</span><span class="si">}</span><span class="s2">.xx.gz&quot;</span>
<span class="kn">bash:</span>
    <span class="n">fastenloc</span> <span class="o">-</span><span class="n">eqtl</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">annot</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">gwas</span> <span class="n">gwas</span><span class="o">.</span><span class="n">pip</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>
    <span class="n">sort</span> <span class="o">-</span><span class="n">grk6</span> <span class="n">prefix</span><span class="o">.</span><span class="n">enloc</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">out</span> <span class="o">|</span> <span class="n">gzip</span> <span class="o">--</span><span class="n">best</span> <span class="o">&gt;</span> <span class="n">prefix</span><span class="o">.</span><span class="n">enloc</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">sorted</span><span class="o">.</span><span class="n">gz</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="n">prefix</span><span class="o">.</span><span class="n">enloc</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">out</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/integrative_analysis/COLOC"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="fastenloc.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Colocalization with FastENLOC</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../functional_annotation.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Functional annotation integration</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>