
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>xQTL and GWAS region alignment &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="TWAS for xQTL" href="ptwas.html" />
    <link rel="prev" title="The PECOTMR intergration framework" href="pecotmr.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/MRAID_QTL.html">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="fine_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/pecotmr_analysis/pecotmr_preprocessing.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_analysis/pecotmr_preprocessing.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/pecotmr_analysis/pecotmr_preprocessing.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="xqtl-and-gwas-region-alignment">
<h1>xQTL and GWAS region alignment<a class="headerlink" href="#xqtl-and-gwas-region-alignment" title="Permalink to this headline">¶</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">intact_pipeline</span><span class="o">/</span><span class="n">Padding</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">padding</span> \
    <span class="o">--</span><span class="n">qtl_file</span> <span class="err">`</span><span class="n">ls</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">susie_per_gene_tad</span><span class="o">/</span><span class="n">cache</span><span class="o">/*</span><span class="n">rds</span><span class="err">`</span>  \
    <span class="o">--</span><span class="n">gwas_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ftp_lisanwanglab_sync</span><span class="o">/</span><span class="n">ftp_fgc_xqtl</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">GWAS_Finemapping_Results</span><span class="o">/</span><span class="n">Bellenguez</span> <span class="o">-</span><span class="n">J</span> <span class="mi">50</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">job_size</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">intact_pipeline</span><span class="o">/</span><span class="n">Padding</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">padding</span> \
    <span class="o">--</span><span class="n">qtl_file</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">eqtl</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">susie_per_gene_tad</span><span class="o">/</span><span class="n">cache</span><span class="o">/</span><span class="n">demo</span><span class="o">.</span><span class="n">ENSG00000288681</span><span class="o">.</span><span class="n">unisusie</span><span class="o">.</span><span class="n">fit</span><span class="o">.</span><span class="n">rds</span> \
    <span class="o">--</span><span class="n">gwas_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ftp_lisanwanglab_sync</span><span class="o">/</span><span class="n">ftp_fgc_xqtl</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">GWAS_Finemapping_Results</span><span class="o">/</span><span class="n">Bellenguez</span> <span class="o">-</span><span class="n">J</span> <span class="mi">50</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">test</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">job_size</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Workdir</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># dataset </span>
<span class="kn">parameter:</span> <span class="n">tissue</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="c1"># QTL data type</span>
<span class="kn">parameter:</span> <span class="n">QTL</span> <span class="o">=</span> <span class="s1">&#39;eQTL&#39;</span>
<span class="kn">parameter:</span> <span class="n">GWAS</span> <span class="o">=</span> <span class="s1">&#39;GWAS&#39;</span>
<span class="kn">parameter:</span> <span class="n">qtl_file</span> <span class="o">=</span> <span class="n">paths</span>
<span class="kn">parameter:</span> <span class="n">gwas_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;8G&quot;</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">padding_1</span><span class="p">]</span>

<span class="kn">input:</span> <span class="n">qtl_file</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">QTL</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">QTL</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.newpadding.rds&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">GWAS</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">GWAS</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.newpadding.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">nn</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">nn</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">suppressPackageStartupMessages</span><span class="p">(</span><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">))</span>
    <span class="c1"># read all GWAS finemapping result file name to extract the chr, start, end and corresponding file paths  </span>
    <span class="n">gwas_finemap_result</span> <span class="o">=</span> <span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{gwas_path}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">full</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">T</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">.rds$&quot;</span><span class="p">)</span>
    <span class="n">gwas_file_tb</span> <span class="o">=</span> <span class="n">tibble</span><span class="p">(</span><span class="n">file_path</span> <span class="o">=</span> <span class="n">gwas_finemap_result</span><span class="p">)</span>

    <span class="c1"># use regex to extract the start, end and chromosome of LD blocks from the file name</span>
    <span class="n">match_pattern</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">filename</span><span class="p">){</span>
        <span class="n">pattern</span> <span class="o">&lt;-</span> <span class="s2">&quot;chr[0-9XY]+_(</span><span class="se">\\</span><span class="s2">d+_</span><span class="se">\\</span><span class="s2">d+)&quot;</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">regmatches</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">regexpr</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span> <span class="n">filename</span><span class="p">))</span>
        <span class="k">return</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
    <span class="p">}</span>
      
    <span class="n">LD_block_position</span> <span class="o">=</span> <span class="n">gwas_file_tb</span> <span class="o">%&gt;%</span> <span class="n">mutate</span><span class="p">(</span><span class="n">chr_pos</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">match_pattern</span><span class="p">))</span> <span class="o">%&gt;%</span> 
      <span class="n">separate</span><span class="p">(</span><span class="n">chr_pos</span><span class="p">,</span> <span class="n">into</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="n">relocate</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="o">.</span><span class="n">after</span> <span class="o">=</span> <span class="n">end</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="n">mutate</span><span class="p">(</span><span class="n">start</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="n">end</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">end</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="n">arrange</span><span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">)</span> 
      
    <span class="c1"># give the absolute path to read RDS of qtl finemapping result rds file  </span>
    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;$</span><span class="si">{QTL}</span><span class="s2">&quot;</span> <span class="o">==</span> <span class="s2">&quot;eQTL&quot;</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">qtl_file</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">dlpfc_eqtl</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
      <span class="n">qtl_file</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># after reading the variants in the rds file, we can use the first variant and last one to extract the chr, start and end of eqtl variants</span>
    <span class="n">qtl_chr</span> <span class="o">=</span> <span class="n">regmatches</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">regexpr</span><span class="p">(</span><span class="s2">&quot;(chr[0-9]+)&quot;</span><span class="p">,</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">qtl_start</span>  <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;.*:(</span><span class="se">\\</span><span class="s2">d+)_.*&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">1&quot;</span><span class="p">,</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">qtl_end</span>  <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;.*:(</span><span class="se">\\</span><span class="s2">d+)_.*&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">1&quot;</span><span class="p">,</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span><span class="p">[</span><span class="n">length</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span><span class="p">)]))</span>

    <span class="c1"># use the regions in qtl to get those LD blocks that overlap with qtl regions</span>
    <span class="n">related_LD</span> <span class="o">=</span> <span class="n">LD_block_position</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="nb">chr</span> <span class="o">==</span> <span class="n">qtl_chr</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">((</span><span class="n">start</span> <span class="o">&lt;=</span> <span class="n">qtl_start</span> <span class="o">&amp;</span> <span class="n">end</span> <span class="o">&gt;=</span> <span class="n">qtl_start</span><span class="p">)</span> <span class="o">|</span>
                                            <span class="p">(</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="n">qtl_start</span> <span class="o">&amp;</span> <span class="n">end</span> <span class="o">&lt;=</span> <span class="n">qtl_end</span><span class="p">)</span> <span class="o">|</span> 
                                            <span class="p">(</span><span class="n">start</span> <span class="o">&lt;=</span> <span class="n">qtl_end</span> <span class="o">&amp;</span> <span class="n">end</span> <span class="o">&gt;=</span> <span class="n">qtl_end</span><span class="p">))</span>

    <span class="c1"># extract those related GWAS finemapping result lbf matrix to form the larger one</span>
    <span class="n">cnt</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">c</span><span class="p">()</span>
    <span class="n">lbf_mtx</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">file</span> <span class="ow">in</span> <span class="n">related_LD</span><span class="err">$</span><span class="n">file_path</span><span class="p">){</span>
        <span class="n">rds</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="n">rds</span><span class="err">$</span><span class="n">variants</span><span class="p">)</span>
        <span class="n">lbf_mtx</span><span class="p">[[</span><span class="n">cnt</span><span class="p">]]</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">rds</span><span class="err">$</span><span class="n">lbf_variable</span><span class="p">)</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">lbf_mtx</span><span class="p">[[</span><span class="n">cnt</span><span class="p">]])</span> <span class="o">=</span> <span class="n">rds</span><span class="err">$</span><span class="n">variants</span>
        <span class="n">cnt</span> <span class="o">=</span> <span class="n">cnt</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="p">}</span>

    <span class="c1"># after combining the matrices, fill those NA with 0 to form the whole matrix</span>
    <span class="n">lbf_whole_mtx</span> <span class="o">=</span> <span class="n">bind_rows</span><span class="p">(</span><span class="n">lbf_mtx</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">replace</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="o">.</span><span class="p">),</span> <span class="mi">0</span><span class="p">)</span>

    <span class="c1"># get the shared variants between gwas and qtl</span>

    <span class="c1"># here is one problem: maybe now the output of finemapping does not need $dlpfc... and the variant name have different format</span>
    <span class="c1"># chr:9999 or chr_999, so here we change : to _; this brings a lot of trouble</span>
    <span class="c1"># if now variable name are uniformed, then things are good, remove the str_replcace line</span>
      
    <span class="n">shared_variant</span> <span class="o">=</span> <span class="n">unlist</span><span class="p">(</span><span class="n">intersect</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span> <span class="o">%&gt;%</span> 
      <span class="nb">map</span><span class="p">(</span><span class="o">~</span> <span class="n">str_replace_all</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)),</span> <span class="n">colnames</span><span class="p">(</span><span class="n">lbf_whole_mtx</span><span class="p">)))</span>

      
    <span class="c1"># remove those columns that does not share SNP, only keep those shared by two datasetl</span>
    <span class="n">GWAS_lbf_matrix</span> <span class="o">=</span> <span class="n">lbf_whole_mtx</span><span class="p">[,</span> <span class="n">shared_variant</span><span class="p">]</span>
      

    <span class="c1"># again, change the variant names</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">lbf_variable</span><span class="p">)</span> <span class="o">=</span> <span class="n">unlist</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span> <span class="o">%&gt;%</span> 
      <span class="nb">map</span><span class="p">(</span><span class="o">~</span> <span class="n">str_replace_all</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)))</span>

    <span class="c1"># because in the susie output the cs are recorded by index, so we get the index that are removed</span>
    <span class="n">rm_index</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">lbf_variable</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">shared_variant</span><span class="p">))</span>
      
    <span class="c1"># also, for qtl data, only keep variants that are shared</span>
    <span class="n">qtl_lbf_mtx</span> <span class="o">=</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">lbf_variable</span><span class="p">[,</span><span class="n">shared_variant</span><span class="p">]</span>
      

    <span class="c1"># convert lbf to alpha</span>
    <span class="n">lbf_to_alpha_vector</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">lbf</span><span class="p">,</span> <span class="n">prior_weights</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">)</span> <span class="p">{</span>
      <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">prior_weights</span><span class="p">))</span> <span class="n">prior_weights</span> <span class="o">=</span> <span class="n">rep</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">length</span><span class="p">(</span><span class="n">lbf</span><span class="p">),</span> <span class="n">length</span><span class="p">(</span><span class="n">lbf</span><span class="p">))</span>
      
      <span class="n">maxlbf</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">lbf</span><span class="p">)</span>
      
      <span class="c1"># If maxlbf is 0, return a vector of zeros</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">maxlbf</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
        <span class="k">return</span><span class="p">(</span><span class="n">setNames</span><span class="p">(</span><span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">lbf</span><span class="p">)),</span> <span class="n">names</span><span class="p">(</span><span class="n">lbf</span><span class="p">)))</span>
      <span class="p">}</span>
      
      <span class="c1"># w is proportional to BF, subtract max for numerical stability.</span>
      <span class="n">w</span> <span class="o">=</span> <span class="n">exp</span><span class="p">(</span><span class="n">lbf</span> <span class="o">-</span> <span class="n">maxlbf</span><span class="p">)</span>
      
      <span class="c1"># Posterior prob for each SNP.</span>
      <span class="n">w_weighted</span> <span class="o">=</span> <span class="n">w</span> <span class="o">*</span> <span class="n">prior_weights</span>
      <span class="n">weighted_sum_w</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">w_weighted</span><span class="p">)</span>
      <span class="n">alpha</span> <span class="o">=</span> <span class="n">w_weighted</span> <span class="o">/</span> <span class="n">weighted_sum_w</span>
      
      <span class="k">return</span><span class="p">(</span><span class="n">alpha</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">lbf_to_alpha</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">lbf</span><span class="p">)</span> <span class="n">t</span><span class="p">(</span><span class="n">apply</span><span class="p">(</span><span class="n">lbf</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">lbf_to_alpha_vector</span><span class="p">))</span>

    <span class="c1"># convert lbf to pip</span>
    <span class="n">lbf_to_pip</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">lbf</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">alpha</span> <span class="o">=</span> <span class="n">lbf_to_alpha</span><span class="p">(</span><span class="n">lbf</span><span class="p">)</span>
        <span class="k">return</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">vector</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">apply</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">alpha</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="n">prod</span><span class="p">)))</span>
    <span class="p">}</span>


    <span class="c1"># use new lbf matrix to compute alpha and pip</span>
    <span class="n">shared_variable_name</span><span class="o">&lt;-</span><span class="n">str_replace</span><span class="p">(</span><span class="n">shared_variant</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;:&quot;</span><span class="p">)</span>

    <span class="n">GWAS_alpha</span> <span class="o">=</span> <span class="n">lbf_to_alpha</span><span class="p">(</span><span class="n">GWAS_lbf_matrix</span><span class="p">)</span>
    <span class="n">GWAS_pip</span> <span class="o">=</span> <span class="n">lbf_to_pip</span><span class="p">(</span><span class="n">GWAS_lbf_matrix</span><span class="p">)</span>

    <span class="n">qtl_alpha</span> <span class="o">=</span> <span class="n">lbf_to_alpha</span><span class="p">(</span><span class="n">qtl_lbf_mtx</span><span class="p">)</span>
    <span class="n">qtl_pip</span> <span class="o">=</span> <span class="n">lbf_to_pip</span><span class="p">(</span><span class="n">qtl_lbf_mtx</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">qtl_alpha</span><span class="p">)</span> <span class="o">=</span><span class="n">colnames</span><span class="p">(</span><span class="n">GWAS_alpha</span><span class="p">)</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">qtl_pip</span><span class="p">)</span>  <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">GWAS_pip</span><span class="p">)</span> <span class="o">=</span> <span class="n">shared_variable_name</span>
    <span class="c1"># for now, we remove the variants in cs, if that variant is not shared by two traits</span>
    <span class="c1">## note: this part may be changed!! now is just a rough strategy</span>
    <span class="n">cs_number</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">)</span>
    <span class="n">new_cs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">if</span><span class="p">(</span><span class="n">cs_number</span> <span class="o">==</span> <span class="mi">0</span><span class="p">){</span>
        <span class="n">new_cs</span> <span class="o">=</span> <span class="n">NA</span>
    <span class="p">}</span><span class="k">else</span><span class="p">{</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">cs_number</span><span class="p">)){</span>
        <span class="n">new_cs</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">setdiff</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">[[</span><span class="n">i</span><span class="p">]],</span> <span class="n">rm_index</span><span class="p">)</span>
        <span class="p">}</span>
    <span class="p">}</span>

    <span class="c1"># the output can be: GWAS+ qtl lbf matrix; alpha; pip</span>
    <span class="n">qtl_mu</span> <span class="o">&lt;-</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">mu</span><span class="p">[,</span><span class="n">shared_variable_name</span><span class="p">]</span>
    <span class="n">names</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">X_column_scale_factors</span><span class="p">)</span> <span class="o">=</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">variable_name</span>
  
    <span class="n">qtl_X_column_scale_factors</span> <span class="o">&lt;-</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">X_column_scale_factors</span><span class="p">[</span><span class="n">shared_variable_name</span><span class="p">]</span>
    <span class="n">qtl_coef</span><span class="o">&lt;-</span><span class="n">colSums</span><span class="p">(</span><span class="n">qtl_alpha</span><span class="o">*</span><span class="n">qtl_mu</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">qtl_X_column_scale_factors</span><span class="p">)</span>
    <span class="n">qtl_intercept</span> <span class="o">=</span> <span class="n">qtl_file</span><span class="err">$</span><span class="n">intercept</span>
  
    <span class="n">new_qtl_obj</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">(</span> <span class="n">pip</span> <span class="o">=</span><span class="n">qtl_pip</span><span class="p">,</span>
                        <span class="n">coef</span> <span class="o">=</span> <span class="n">qtl_coef</span><span class="p">,</span>
                        <span class="n">mu</span> <span class="o">=</span> <span class="n">qtl_mu</span><span class="p">,</span>
                        <span class="n">alpha</span> <span class="o">=</span> <span class="n">qtl_alpha</span><span class="p">,</span>
                        <span class="n">sets</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">cs</span> <span class="o">=</span> <span class="n">new_cs</span><span class="p">),</span>
                        <span class="n">variable_name</span> <span class="o">=</span> <span class="n">shared_variable_name</span><span class="p">,</span>
                        <span class="n">intercept</span> <span class="o">=</span> <span class="n">qtl_intercept</span><span class="p">,</span>
                        <span class="n">X_column_scale_factors</span> <span class="o">=</span> <span class="n">qtl_X_column_scale_factors</span>
    <span class="p">)</span>
    <span class="n">names</span><span class="p">(</span><span class="n">new_qtl_obj</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">)</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">qtl_file</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">)</span>

    <span class="n">new_gwas_obj</span> <span class="o">&lt;-</span> <span class="nb">list</span><span class="p">(</span> <span class="n">pip</span> <span class="o">=</span><span class="n">GWAS_pip</span><span class="p">,</span>
                        <span class="n">alpha</span> <span class="o">=</span> <span class="n">GWAS_alpha</span><span class="p">,</span>
                        <span class="n">variable_name</span> <span class="o">=</span> <span class="n">shared_variable_name</span>
    <span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">new_qtl_obj</span><span class="p">,</span><span class="s2">&quot;$</span><span class="si">{_output[0]}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">new_gwas_obj</span><span class="p">,</span><span class="s2">&quot;$</span><span class="si">{_output[1]}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">padding_2</span><span class="p">,</span> <span class="n">padding_file_table</span><span class="p">]</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/susie_table.stdout&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/susie_table.stderr&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="kn">import</span> <span class="nn">os</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="k">def</span> <span class="nf">generate_susie_table</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">output_file</span><span class="p">):</span>
        <span class="n">filenames</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">filepaths</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.rds&quot;</span><span class="p">):</span>
                <span class="n">middle_name</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span> 
                <span class="n">filenames</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">middle_name</span><span class="p">)</span>
                <span class="n">full_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">file</span><span class="p">)</span>
                <span class="n">filepaths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">full_path</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;molecular_trait_id&#39;</span><span class="p">:</span> <span class="n">filenames</span><span class="p">,</span>
            <span class="s1">&#39;susie_file&#39;</span><span class="p">:</span> <span class="n">filepaths</span>
        <span class="p">})</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">generate_susie_table</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">QTL</span><span class="si">}</span><span class="s1">.susie_table.txt&#39;</span><span class="p">)</span>
    <span class="n">generate_susie_table</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">GWAS</span><span class="si">}</span><span class="s1">.susie_table.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/pecotmr_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="pecotmr.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">The PECOTMR intergration framework</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="ptwas.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">TWAS for xQTL</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>