
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Mendelian Randomization using fine-mapping IV &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="TWAS and COLOC integration" href="intact.html" />
    <link rel="prev" title="TWAS for xQTL" href="ptwas.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/MRAID_QTL.html">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   Enrichment analysis: TORUS and GREGOR
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/pecotmr_analysis/ptwas_mr.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_analysis/ptwas_mr.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/pecotmr_analysis/ptwas_mr.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#output">
     Output
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#example">
   Example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generate-the-units-of-analysis">
     Generate the units of analysis
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#id1">
       Input
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#output-file-format">
       Output file format
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#causal-effect-estimation-and-validation">
     Causal effect estimation and validation
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#estimation">
       Estimation
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#validation">
       Validation
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#inuput-file-format">
       Inuput file format
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#id2">
       Output file format
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#id3">
       Input
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#ptwas-est">
       ptwas_est
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#id4">
       Output
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="mendelian-randomization-using-fine-mapping-iv">
<h1>Mendelian Randomization using fine-mapping IV<a class="headerlink" href="#mendelian-randomization-using-fine-mapping-iv" title="Permalink to this headline">¶</a></h1>
<p>The procedures for PTWAS estimation and model validation that are implemented in the R script. The unit of analysis is a single gene-trait pair, whose information is summarized in a text file. First, we generate the units of analysis. Then, we conduct the causal effect estimation and validation procedure.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Organize the eQTL fine-mapping results, eQTL summary statistics, and GWAS summary statistics data into a format that can be processed by ptwas_est.</p></li>
<li><p>run ptwas_est with R codes: causal effect estimation and validation.</p></li>
</ul>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>eQTL finemapping result:</p>
<ul>
<li><p>This result contain the information of credible sets and pip. <br/>
<br/></p></li>
</ul>
</li>
<li><p>eqtl sumstats: se and beta</p></li>
<li><p>gwas sumstats: se and beta</p></li>
</ul>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>data format for ptwas_est</p></li>
<li><p>ptwas_est resuls</p></li>
</ul>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="section" id="generate-the-units-of-analysis">
<h3>Generate the units of analysis<a class="headerlink" href="#generate-the-units-of-analysis" title="Permalink to this headline">¶</a></h3>
<div class="section" id="id1">
<h4>Input<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><strong>eQTL finemapping</strong>: /mnt/vast/hpc/csg/rf2872/Work/INTACT/fastenloc_test/eqtl_susie_table.txt</p></li>
<li><p><strong>eQTL sumstats</strong>: /mnt/vast/hpc/csg/ftp_lisanwanglab_sync/ftp_fgc_xqtl/projects/rna-seq/BU/ROSMAP_DLPFC/eQTL/association_scan/norminal_qced_files</p></li>
<li><p><strong>GWAS sumstats</strong>: /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022</p></li>
</ul>
</div>
<div class="section" id="output-file-format">
<h4>Output file format<a class="headerlink" href="#output-file-format" title="Permalink to this headline">¶</a></h4>
<p>Each row of the file corresponds to a single SNP that is inferred as a memeber of a signal cluster, the columns are:</p>
<ul class="simple">
<li><p>snp: SNP name</p></li>
<li><p>beta_eQTL: eQTL effect</p></li>
<li><p>se_eQTL: standard error of estimated eQTL effect</p></li>
<li><p>beta_GWAS: GWAS effect</p></li>
<li><p>se_GWAS: standard error of GWAS effect</p></li>
<li><p>cluster: Signal cluster ID (credible sets index)</p></li>
<li><p>pip: SNP posterior inclusiion probability (PIP)</p></li>
<li><p>gene_id: gene name</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#######organize the data of eqtl finemapping, eqtl summary statistics and GWAS summary statistics</span>
<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>

<span class="c1">#####load the the eqtl finemapping folders</span>
<span class="n">eqtl_table_list</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="s">&quot;/mnt/vast/hpc/csg/rf2872/Work/INTACT/fastenloc_test/eqtl_susie_table.txt&quot;</span><span class="p">)</span>
<span class="n">finemapping_folder</span> <span class="o">=</span> <span class="s">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/susie_per_gene_tad/cache&quot;</span>

<span class="c1">#####load the eqtl sumstats</span>
<span class="n">eqtl_folder</span> <span class="o">=</span> <span class="s">&quot;/mnt/vast/hpc/csg/ftp_lisanwanglab_sync/ftp_fgc_xqtl/projects/rna-seq/BU/ROSMAP_DLPFC/eQTL/association_scan/norminal_qced_files&quot;</span>
<span class="c1">#####load the GWAS sumstats</span>
<span class="n">GWAS_folder</span> <span class="o">=</span> <span class="s">&quot;/mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022&quot;</span>

<span class="n">gene_name_list</span> <span class="o">=</span> <span class="n">eqtl_table_list</span><span class="o">$</span><span class="n">molecular_trait_i</span>
<span class="n">eqtl</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">k</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">){</span>
<span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene_name_list</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
<span class="n">eqtl_finemapp</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">finemapping_folder</span><span class="p">,</span><span class="s">&quot;/demo.&quot;</span><span class="p">,</span><span class="n">gene_name</span><span class="p">,</span><span class="s">&quot;.unisusie.fit.rds&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
<span class="n">cluster_num</span> <span class="o">=</span> <span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">sets</span><span class="o">$</span><span class="n">cs_index</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cluster_num</span><span class="p">)){</span>
    <span class="n">chr</span> <span class="o">=</span> <span class="nf">str_sub</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="nf">str_split</span><span class="p">(</span><span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">variable_name</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">1</span><span class="p">]),</span><span class="m">4</span><span class="p">)</span>
    <span class="n">eqtl_data</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">eqtl_folder</span><span class="p">,</span><span class="s">&quot;/dlpfc_batch_all.rnaseqc.low_expression_filtered.outlier_removed.tmm.expression.bed.processed_phenotype.per_chrom_dlpfc_batch_all.rnaseqc.ROSMAP_covariates.ROSMAP_NIA_WGS.pca.PEER.txt.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
    <span class="n">GWAS_data</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">GWAS_folder</span><span class="p">,</span><span class="s">&quot;/ADGWAS_Bellenguez_2022.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;/ADGWAS2022.chr&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;.sumstat.tsv&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
    <span class="nf">for </span><span class="p">(</span><span class="n">i</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">cluster_num</span><span class="p">)){</span>
       <span class="n">cluster_id</span> <span class="o">=</span> <span class="n">cluster_num</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
       <span class="n">cluster_id_name</span>  <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;L&quot;</span><span class="p">,</span><span class="n">cluster_id</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
       <span class="n">crediblesets</span> <span class="o">=</span> <span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">sets</span><span class="o">$</span><span class="n">cs</span><span class="p">[[</span><span class="n">cluster_id_name</span><span class="p">]]</span>
       <span class="n">variants_name</span> <span class="o">=</span> <span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">variable_name</span><span class="p">[</span><span class="n">crediblesets</span><span class="p">]</span>
       <span class="c1">#####get the chromosome fo the variants_name</span>
       <span class="n">chr</span> <span class="o">=</span> <span class="nf">str_sub</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="nf">str_split</span><span class="p">(</span><span class="n">variants_name</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">1</span><span class="p">]),</span><span class="m">4</span><span class="p">)</span>
       <span class="n">pip</span> <span class="o">=</span> <span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">pip</span><span class="p">[</span><span class="n">crediblesets</span><span class="p">])</span>
       <span class="n">eqtl_cluster</span> <span class="o">=</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">snp</span> <span class="o">=</span> <span class="n">variants_name</span><span class="p">,</span> <span class="n">cluster</span>  <span class="o">=</span> <span class="nf">rep</span><span class="p">(</span><span class="n">cluster_id</span><span class="p">,</span> <span class="nf">length</span><span class="p">(</span><span class="n">variants_name</span><span class="p">)),</span><span class="n">pip</span> <span class="o">=</span> <span class="n">pip</span><span class="p">,</span><span class="n">gene_id</span> <span class="o">=</span> <span class="nf">rep</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="nf">length</span><span class="p">(</span><span class="n">variants_name</span><span class="p">)))</span>
       <span class="n">eqtl_beta_se</span> <span class="o">=</span> <span class="n">eqtl_data</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">molecular_trait_id</span><span class="o">%in%</span><span class="n">gene_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">variant</span><span class="o">%in%</span><span class="n">variants_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span><span class="n">beta</span><span class="p">,</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;snp&quot;</span> <span class="o">=</span> <span class="s">&quot;variant&quot;</span><span class="p">,</span><span class="s">&quot;beta_eQTL&quot;</span><span class="o">=</span><span class="s">&quot;beta&quot;</span><span class="p">,</span><span class="s">&quot;se_eQTL&quot;</span><span class="o">=</span><span class="s">&quot;se&quot;</span><span class="p">)</span>
       <span class="n">GWAS_beta_se</span> <span class="o">=</span> <span class="n">GWAS_data</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">snp</span> <span class="o">=</span> <span class="nf">sub</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">variant</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">snp</span><span class="o">%in%</span><span class="n">variants_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">snp</span><span class="p">,</span><span class="n">beta</span><span class="p">,</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;beta_GWAS&quot;</span><span class="o">=</span><span class="s">&quot;beta&quot;</span><span class="p">,</span><span class="s">&quot;se_GWAS&quot;</span><span class="o">=</span><span class="s">&quot;se&quot;</span><span class="p">)</span>
       <span class="n">df</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="n">eqtl_beta_se</span><span class="p">,</span> <span class="n">GWAS_beta_se</span><span class="p">,</span> <span class="n">eqtl_cluster</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">reduce</span><span class="p">(</span><span class="n">full_join</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s">&quot;snp&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">drop_na</span><span class="p">()</span>
       <span class="n">eqtl</span> <span class="o">=</span> <span class="nf">rbind</span><span class="p">(</span><span class="n">eqtl</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="nf">rm</span><span class="p">(</span><span class="n">eqtl_data</span><span class="p">)</span>
    <span class="nf">rm</span><span class="p">(</span><span class="n">GWAS_data</span><span class="p">)</span>
    <span class="nf">fwrite</span><span class="p">(</span><span class="n">eqtl</span><span class="p">,</span><span class="n">file</span>  <span class="o">=</span> <span class="s">&quot;/home/aw3600/PTWAS/PTWAS_test_data.txt&quot;</span><span class="p">)</span>
<span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="causal-effect-estimation-and-validation">
<h3>Causal effect estimation and validation<a class="headerlink" href="#causal-effect-estimation-and-validation" title="Permalink to this headline">¶</a></h3>
<div class="section" id="estimation">
<h4>Estimation<a class="headerlink" href="#estimation" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>First, we select strong eQTLs by thresholding on the corresponding signal-level PIPs (cpip = 0.5).</p></li>
<li><p>Then, we aggregate the estimates from the individual eQTLs using a fixed-effect meta-analysis procedure.</p></li>
</ul>
</div>
<div class="section" id="validation">
<h4>Validation<a class="headerlink" href="#validation" title="Permalink to this headline">¶</a></h4>
<p>Identify severe violations of the exclusion restriction (ER) assumption and exclude them from the downstream estimation procedure.</p>
</div>
<div class="section" id="inuput-file-format">
<h4>Inuput file format<a class="headerlink" href="#inuput-file-format" title="Permalink to this headline">¶</a></h4>
<p>The output from the first step</p>
</div>
<div class="section" id="id2">
<h4>Output file format<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h4>
<p>Each row of the file corresponds to a gene and the columns are:</p>
<ul class="simple">
<li><p>gene_id: gene name</p></li>
<li><p>num_cluster: the number of credible sets of gene</p></li>
<li><p>num_instruments: the number of instruments included in the gene</p></li>
<li><p>spip: the sum of pip for credible sets of each gene</p></li>
<li><p>grp_beta: signal-level estimates (combine the SNP-level estimates from the member SNPs weighted by their corresponding PIPs)</p></li>
<li><p>grp_se: standard error of the signal-level estimates</p></li>
<li><p>meta: gene-level estimate of the causal effect (aggregate the signal-level estimates by applying a fixed-effect meta-analysis model)</p></li>
<li><p>se_meta: standard error of the gene-level estimate of the causal effect</p></li>
<li><p>Q: Cochran’s Q statistic</p></li>
<li><p>I2: <span class="math notranslate nohighlight">\(I^2\)</span> statistics (quantify the heterogeneity, range from 0 to 1, <span class="math notranslate nohighlight">\(I^2 \rightarrow 0\)</span> indicates reasonable consistency among estimates by multiple eQTLs, whereas <span class="math notranslate nohighlight">\(I^2 \rightarrow 1\)</span> 1 implies severe departures from exclusion restriction (ER).)</p></li>
</ul>
</div>
<div class="section" id="id3">
<h4>Input<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">formatted_input</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="s">&quot;/home/aw3600/PTWAS/PTWAS_test_data.txt&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">formatted_input</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.table: 6 × 8</caption>
<thead>
	<tr><th scope=col>snp</th><th scope=col>beta_eQTL</th><th scope=col>se_eQTL</th><th scope=col>beta_GWAS</th><th scope=col>se_GWAS</th><th scope=col>cluster</th><th scope=col>pip</th><th scope=col>gene_id</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1:169880000_A_G</td><td>-0.2883383</td><td>0.03356421</td><td>0.0031</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169880823_G_A</td><td>-0.2883383</td><td>0.03356421</td><td>0.0034</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169880877_T_C</td><td>-0.2883383</td><td>0.03356421</td><td>0.0033</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169882224_A_G</td><td>-0.2859217</td><td>0.03388142</td><td>0.0024</td><td>0.0138</td><td>1</td><td>0.003048243</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169882228_G_A</td><td>-0.2859217</td><td>0.03388142</td><td>0.0023</td><td>0.0138</td><td>1</td><td>0.003048243</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169882581_G_T</td><td>-0.2883383</td><td>0.03356421</td><td>0.0025</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
</tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="ptwas-est">
<h4>ptwas_est<a class="headerlink" href="#ptwas-est" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#####define the threshold of the cumulative cluster pip</span>
<span class="n">cpip</span> <span class="o">=</span> <span class="m">0.5</span>

<span class="c1">######heterogeneity:  calculate I2 statistics based on the Cochran&#39;s Q statistic</span>
<span class="n">calc_I2</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">Q</span><span class="p">,</span> <span class="n">Est</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">Q</span> <span class="o">=</span> <span class="n">Q</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span>
    <span class="n">Est</span> <span class="o">=</span> <span class="nf">length</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">Est</span><span class="p">))</span>
    <span class="n">I2</span> <span class="o">=</span> <span class="nf">if </span><span class="p">(</span><span class="n">Q</span> <span class="o">&gt;</span> <span class="m">1e-3</span><span class="p">)</span> <span class="p">(</span><span class="n">Q</span> <span class="o">-</span> <span class="n">Est</span> <span class="o">+</span> <span class="m">1</span><span class="p">)</span><span class="o">/</span><span class="n">Q</span> <span class="n">else</span> <span class="m">0</span>
    <span class="nf">return</span><span class="p">(</span><span class="nf">if </span><span class="p">(</span><span class="n">I2</span> <span class="o">&lt;</span> <span class="m">0</span><span class="p">)</span> <span class="m">0</span> <span class="n">else</span> <span class="n">I2</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1">#####ptwas_est</span>
<span class="n">output</span> <span class="o">=</span> <span class="n">formatted_input</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span>
        <span class="n">beta_eQTL</span> <span class="o">=</span> <span class="n">beta_eQTL</span><span class="o">/</span><span class="n">se_eQTL</span><span class="p">,</span>
        <span class="n">se_eQTL</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">group_by</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">cluster</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">spip</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">(</span><span class="n">pip</span><span class="p">))</span> <span class="o">%&gt;%</span>
    <span class="nf">filter</span><span class="p">(</span><span class="n">spip</span> <span class="o">&gt;=</span> <span class="n">cpip</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="c1"># Cumulative cluster pip greater than a user defined cumulative pip threshold</span>
    <span class="nf">group_by</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">cluster</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span>
        <span class="n">beta_yx</span> <span class="o">=</span> <span class="n">beta_GWAS</span><span class="o">/</span><span class="n">beta_eQTL</span><span class="p">,</span>
        <span class="n">se_yx</span> <span class="o">=</span> <span class="nf">sqrt</span><span class="p">(</span>
            <span class="p">(</span><span class="n">se_GWAS</span><span class="o">^</span><span class="m">2</span><span class="o">/</span><span class="n">beta_eQTL</span><span class="o">^</span><span class="m">2</span><span class="p">)</span> <span class="o">+</span> <span class="p">((</span><span class="n">beta_GWAS</span><span class="o">^</span><span class="m">2</span><span class="o">*</span><span class="n">se_eQTL</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="o">/</span><span class="n">beta_eQTL</span><span class="o">^</span><span class="m">4</span><span class="p">)),</span>
        <span class="n">grp_beta</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">((</span><span class="n">beta_yx</span><span class="o">*</span><span class="n">pip</span><span class="p">)</span><span class="o">/</span><span class="n">spip</span><span class="p">),</span>
        <span class="n">grp_se</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">((</span><span class="n">beta_yx</span><span class="o">^</span><span class="m">2</span> <span class="o">+</span> <span class="n">se_yx</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="o">*</span><span class="n">pip</span><span class="o">/</span><span class="n">spip</span><span class="p">))</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span>
        <span class="n">grp_se</span> <span class="o">=</span> <span class="nf">sqrt</span><span class="p">(</span><span class="n">grp_se</span> <span class="o">-</span> <span class="n">grp_beta</span><span class="o">^</span><span class="m">2</span><span class="p">),</span>
        <span class="n">wv</span> <span class="o">=</span> <span class="n">grp_se</span><span class="o">^</span><span class="m">-2</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">ungroup</span><span class="p">()</span> <span class="o">%&gt;%</span>
    <span class="nf">group_by</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span>
        <span class="n">meta</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">wv</span><span class="p">)</span> <span class="o">*</span> <span class="nf">unique</span><span class="p">(</span><span class="n">grp_beta</span><span class="p">)),</span>
        <span class="n">sum_w</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">wv</span><span class="p">)),</span>
        <span class="n">se_meta</span> <span class="o">=</span> <span class="nf">sqrt</span><span class="p">(</span><span class="n">sum_w</span><span class="o">^</span><span class="m">-1</span><span class="p">),</span>
        <span class="n">num_cluster</span> <span class="o">=</span> <span class="nf">length</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">cluster</span><span class="p">)))</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span>
        <span class="n">num_instruments</span> <span class="o">=</span> <span class="nf">length</span><span class="p">(</span><span class="n">snp</span><span class="p">),</span>
        <span class="n">meta</span> <span class="o">=</span> <span class="n">meta</span><span class="o">/</span><span class="n">sum_w</span><span class="p">,</span>
        <span class="c1">######sum(unique(wv)*(unique(grp_beta) - unique(meta))^2)</span>
        <span class="n">Q</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">wv</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">grp_beta</span><span class="p">)</span> <span class="o">-</span> <span class="nf">unique</span><span class="p">(</span><span class="n">meta</span><span class="p">))</span><span class="o">^</span><span class="m">2</span><span class="p">),</span>
        <span class="n">I2</span> <span class="o">=</span> <span class="nf">calc_I2</span><span class="p">(</span><span class="n">Q</span><span class="p">,</span> <span class="n">grp_beta</span><span class="p">))</span> <span class="o">%&gt;%</span>
        <span class="nf">ungroup</span><span class="p">()</span> <span class="o">%&gt;%</span>
    <span class="nf">distinct</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">.keep_all</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span>
        <span class="n">spip</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">spip</span><span class="p">,</span> <span class="m">3</span><span class="p">),</span>
        <span class="n">grp_beta</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">grp_beta</span><span class="p">,</span> <span class="m">3</span><span class="p">),</span>
        <span class="n">meta</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span> <span class="m">3</span><span class="p">),</span>
        <span class="n">se_meta</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">se_meta</span><span class="p">,</span> <span class="m">3</span><span class="p">),</span>
        <span class="n">Q</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">Q</span><span class="p">,</span> <span class="m">3</span><span class="p">),</span>
        <span class="n">I2</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">I2</span><span class="p">,</span> <span class="m">3</span><span class="p">))</span> <span class="o">%&gt;%</span>
    <span class="nf">select</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">num_cluster</span><span class="p">,</span> <span class="n">num_instruments</span><span class="p">,</span> <span class="n">spip</span><span class="p">,</span> <span class="n">grp_beta</span><span class="p">,</span> <span class="n">grp_se</span><span class="p">,</span> <span class="n">meta</span><span class="p">,</span> <span class="n">se_meta</span><span class="p">,</span> <span class="n">Q</span><span class="p">,</span> <span class="n">I2</span><span class="p">)</span>
    <span class="nf">fwrite</span><span class="p">(</span><span class="n">output</span><span class="p">,</span><span class="n">file</span>  <span class="o">=</span> <span class="s">&quot;/home/aw3600/PTWAS/PTWAS_est_data.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="id4">
<h4>Output<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">output</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 51 × 10</caption>
<thead>
	<tr><th scope=col>gene_id</th><th scope=col>num_cluster</th><th scope=col>num_instruments</th><th scope=col>spip</th><th scope=col>grp_beta</th><th scope=col>grp_se</th><th scope=col>meta</th><th scope=col>se_meta</th><th scope=col>Q</th><th scope=col>I2</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>ENSG00000000457</td><td>2</td><td> 74</td><td>0.696</td><td> 0.000</td><td>0.0016118763</td><td>-0.001</td><td>0.001</td><td>0.157</td><td>0.000</td></tr>
	<tr><td>ENSG00000000971</td><td>1</td><td> 22</td><td>0.739</td><td>-0.001</td><td>0.0010414938</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001084</td><td>1</td><td> 10</td><td>0.881</td><td> 0.000</td><td>0.0010095378</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001167</td><td>1</td><td> 16</td><td>0.737</td><td> 0.009</td><td>0.0029981468</td><td> 0.009</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001460</td><td>1</td><td> 16</td><td>0.745</td><td>-0.001</td><td>0.0006707913</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001461</td><td>1</td><td> 47</td><td>0.679</td><td>-0.005</td><td>0.0028313411</td><td>-0.005</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001561</td><td>1</td><td>  9</td><td>0.586</td><td> 0.001</td><td>0.0006746664</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001626</td><td>1</td><td> 28</td><td>0.801</td><td> 0.001</td><td>0.0011200443</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001629</td><td>2</td><td> 48</td><td>0.678</td><td> 0.000</td><td>0.0019744363</td><td> 0.000</td><td>0.001</td><td>0.046</td><td>0.000</td></tr>
	<tr><td>ENSG00000001630</td><td>1</td><td> 67</td><td>0.807</td><td> 0.001</td><td>0.0011870369</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000002822</td><td>2</td><td> 19</td><td>0.926</td><td>-0.001</td><td>0.0030673516</td><td> 0.001</td><td>0.002</td><td>0.897</td><td>0.000</td></tr>
	<tr><td>ENSG00000002834</td><td>1</td><td>  1</td><td>0.999</td><td>-0.001</td><td>0.0012142297</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000002933</td><td>2</td><td> 40</td><td>0.985</td><td>-0.001</td><td>0.0013062166</td><td> 0.001</td><td>0.001</td><td>1.257</td><td>0.205</td></tr>
	<tr><td>ENSG00000003137</td><td>1</td><td> 21</td><td>0.595</td><td> 0.001</td><td>0.0014258335</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003147</td><td>1</td><td>  8</td><td>0.901</td><td>-0.003</td><td>0.0015701290</td><td>-0.003</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003249</td><td>2</td><td> 17</td><td>0.688</td><td> 0.002</td><td>0.0020332468</td><td> 0.002</td><td>0.001</td><td>0.096</td><td>0.000</td></tr>
	<tr><td>ENSG00000003393</td><td>1</td><td>  5</td><td>0.901</td><td> 0.002</td><td>0.0011071698</td><td> 0.002</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003402</td><td>1</td><td>  1</td><td>0.998</td><td> 0.002</td><td>0.0020686803</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003509</td><td>1</td><td> 14</td><td>0.582</td><td>-0.001</td><td>0.0012611594</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003987</td><td>1</td><td>  5</td><td>0.828</td><td> 0.001</td><td>0.0044738629</td><td> 0.001</td><td>0.004</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003989</td><td>1</td><td>  3</td><td>0.985</td><td> 0.000</td><td>0.0007597256</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004142</td><td>1</td><td>  3</td><td>0.985</td><td> 0.000</td><td>0.0005755454</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004399</td><td>3</td><td>123</td><td>0.624</td><td> 0.003</td><td>0.0014414321</td><td> 0.003</td><td>0.001</td><td>0.489</td><td>0.000</td></tr>
	<tr><td>ENSG00000004455</td><td>1</td><td> 72</td><td>0.755</td><td> 0.002</td><td>0.0015330947</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004487</td><td>1</td><td> 12</td><td>0.550</td><td>-0.001</td><td>0.0013412799</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004534</td><td>1</td><td> 28</td><td>0.871</td><td> 0.000</td><td>0.0003097528</td><td> 0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004700</td><td>1</td><td> 16</td><td>0.737</td><td> 0.000</td><td>0.0025675503</td><td> 0.000</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004766</td><td>1</td><td>  1</td><td>0.957</td><td> 0.001</td><td>0.0035742160</td><td> 0.001</td><td>0.004</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005020</td><td>1</td><td>107</td><td>0.628</td><td> 0.000</td><td>0.0008228963</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005059</td><td>2</td><td> 48</td><td>0.684</td><td>-0.001</td><td>0.0018060197</td><td> 0.000</td><td>0.002</td><td>2.463</td><td>0.594</td></tr>
	<tr><td>ENSG00000005075</td><td>1</td><td> 11</td><td>0.550</td><td> 0.002</td><td>0.0019552991</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005108</td><td>2</td><td> 36</td><td>0.604</td><td> 0.002</td><td>0.0013832122</td><td> 0.001</td><td>0.001</td><td>0.217</td><td>0.000</td></tr>
	<tr><td>ENSG00000005156</td><td>2</td><td> 30</td><td>0.890</td><td> 0.000</td><td>0.0016157681</td><td> 0.000</td><td>0.001</td><td>0.002</td><td>0.000</td></tr>
	<tr><td>ENSG00000005189</td><td>1</td><td> 88</td><td>0.783</td><td> 0.001</td><td>0.0026267194</td><td> 0.001</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005206</td><td>1</td><td>  7</td><td>0.949</td><td> 0.000</td><td>0.0015403997</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005249</td><td>1</td><td> 68</td><td>0.898</td><td> 0.000</td><td>0.0032628529</td><td> 0.000</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005436</td><td>1</td><td>  2</td><td>0.986</td><td> 0.001</td><td>0.0016112891</td><td> 0.001</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005448</td><td>1</td><td>  2</td><td>0.770</td><td> 0.007</td><td>0.0033168502</td><td> 0.007</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005471</td><td>1</td><td>  2</td><td>1.000</td><td> 0.000</td><td>0.0015087191</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005700</td><td>1</td><td> 19</td><td>0.537</td><td>-0.001</td><td>0.0029638984</td><td>-0.001</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005981</td><td>2</td><td> 21</td><td>0.848</td><td> 0.001</td><td>0.0014226597</td><td> 0.001</td><td>0.001</td><td>0.598</td><td>0.000</td></tr>
	<tr><td>ENSG00000006016</td><td>1</td><td>  2</td><td>0.832</td><td> 0.001</td><td>0.0008338203</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006042</td><td>2</td><td> 15</td><td>0.730</td><td> 0.000</td><td>0.0009331976</td><td> 0.000</td><td>0.001</td><td>0.080</td><td>0.000</td></tr>
	<tr><td>ENSG00000006116</td><td>1</td><td>  5</td><td>0.919</td><td>-0.004</td><td>0.0018333447</td><td>-0.004</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006125</td><td>1</td><td> 79</td><td>0.788</td><td>-0.001</td><td>0.0014991831</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006194</td><td>1</td><td> 23</td><td>0.605</td><td> 0.001</td><td>0.0013802571</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006282</td><td>2</td><td>  6</td><td>1.000</td><td>-0.001</td><td>0.0005764921</td><td> 0.000</td><td>0.000</td><td>2.708</td><td>0.631</td></tr>
	<tr><td>ENSG00000006283</td><td>1</td><td>  7</td><td>0.853</td><td> 0.002</td><td>0.0016798225</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006453</td><td>1</td><td> 76</td><td>0.585</td><td>-0.001</td><td>0.0010935407</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006459</td><td>1</td><td> 50</td><td>0.704</td><td> 0.000</td><td>0.0018713348</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006555</td><td>1</td><td> 26</td><td>0.776</td><td> 0.000</td><td>0.0024193024</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
</tbody>
</table>
</div></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./code/pecotmr_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="ptwas.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">TWAS for xQTL</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="intact.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">TWAS and COLOC integration</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>