
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>TWAS for xQTL &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Mendelian Randomization using fine-mapping IV" href="ptwas_mr.html" />
    <link rel="prev" title="xQTL and GWAS region alignment" href="pecotmr_preprocessing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/MRAID_QTL.html">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   Enrichment analysis: TORUS and GREGOR
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="ptwas_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/pecotmr_analysis/ptwas.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_analysis/ptwas.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/pecotmr_analysis/ptwas.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ouput">
     Ouput
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ptwas">
   PTWAS
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#output">
     Output
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id2">
   PTWAS
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id4">
     Output
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example">
     Example
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#bulk-data-set">
       bulk data set
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="twas-for-xqtl">
<h1>TWAS for xQTL<a class="headerlink" href="#twas-for-xqtl" title="Permalink to this headline">¶</a></h1>
<p>This module contains the software implementations to perform transcriptome-wide association analysis (TWAS). These methods are designed to perform rigorous causal inference connecting genes to complex traits.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The goal of this module is to perform PTWAS analysis from SuSiE objects, including:</p>
<ul class="simple">
<li><p><strong>Step 1: PTWAS (twas_z_format)</strong></p></li>
</ul>
<ol class="simple">
<li><p>load GWAS z-score</p></li>
<li><p>load the corresponding LD matrices within the TAD region of each gene</p></li>
<li><p>use allele_qc() function to QC the LD matrix with summary stats</p></li>
<li><p>extract susie, lasso, enet and mr_ash weights</p></li>
<li><p>use twas_z() to compute TWAS results from multipile weights</p></li>
</ol>
<ul class="simple">
<li><p><strong>Step 2: MR (twas_candidate_genes)</strong></p></li>
</ul>
<ol class="simple">
<li><p>use some p-value cutoff to loosely pick TWAS regions of interest then if the region passes the cutoff, save the QC-ed GWAS data in the format compatabile to <a class="reference external" href="https://github.com/cumc/pecotmr/blob/main/R/mr.R">mr.R</a></p></li>
</ol>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>Step 1: PTWAS</strong></p></li>
</ul>
<ol class="simple">
<li><p>QTL susie table：</p></li>
<li><p>GWAS sumstats results (tsv format)</p></li>
<li><p>LD reference</p></li>
<li><p>TAD region</p></li>
</ol>
<ul class="simple">
<li><p><strong>Step 2: MR</strong></p></li>
</ul>
<ol class="simple">
<li><p>output of Step 1</p></li>
</ol>
</div>
<div class="section" id="ouput">
<h3>Ouput<a class="headerlink" href="#ouput" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>Step 1: PTWAS</strong></p></li>
</ul>
<ol class="simple">
<li><p>TWAS results from multiple weights</p></li>
<li><p>GWAS summary statistics after QC</p></li>
<li><p>the input format of twas_z() function</p></li>
</ol>
<ul class="simple">
<li><p><strong>Step 2: MR</strong></p></li>
</ul>
<ol class="simple">
<li><p>the candidate genes of TWAS</p></li>
<li><p>the input format of <a class="reference external" href="https://github.com/cumc/pecotmr/blob/main/R/mr.R">mr.R</a></p></li>
<li><p>the MR results</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">global</span><span class="p">]</span>
<span class="c1"># Workdir</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">susie_path</span> <span class="o">=</span> <span class="n">paths</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="nf">path</span><span class="p">(</span><span class="s">&quot;output&quot;</span><span class="p">)</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">container</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
<span class="n">import</span> <span class="n">re</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s">&#39; &#39;</span> <span class="o">+</span> <span class="nf">re.sub</span><span class="p">(</span><span class="n">r</span><span class="s">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">,</span> <span class="nf">container.split</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)[</span><span class="m">-1</span><span class="p">]))</span> <span class="n">if</span> <span class="n">container</span> <span class="n">else</span> <span class="s">&quot;&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="m">100</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">walltime</span> <span class="o">=</span> <span class="s">&quot;1h&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">mem</span> <span class="o">=</span> <span class="s">&quot;16G&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="m">20</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">allele_qc_R</span> <span class="o">=</span> <span class="nf">path</span><span class="p">(</span><span class="s">&quot;~/pecotmr/R/allele_qc.R&quot;</span><span class="p">)</span>
<span class="n">parameter</span><span class="o">:</span> 
<span class="n">allele_qc_R</span> <span class="o">=</span> <span class="n">f</span><span class="s">&quot;{allele_qc_R:a}&quot;</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="ptwas">
<h2>PTWAS<a class="headerlink" href="#ptwas" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Input<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>susie_path: A list of file paths for susie results</p></li>
<li><p>GWAS_path: gwas summary statistics path (to load the gwas sumstats dataframe)</p></li>
<li><p>LD_path: LD block matrix path (a list of file paths of LD block)</p></li>
<li><p>TAD_path: TAD region path (a dataframe  of TAD region of each gene, this one is applied to subset the LD matrix for each gene)</p></li>
</ul>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>twas_z_format: a dataframe of input format for the twas_z() function</p></li>
<li><p>gene_weights_pq: a dataframe of output of twas_z() function, we apply twas_z function to the  weights of four methods (susie, lasso, enet and mr_ash) to calculate the pvalue, after obtaining the pvalue, we also calculate the corresponding qvalue</p></li>
<li><p>AD_allele_flip: AD gwas sumstats after QC</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_z_format_1</span><span class="p">]</span>

<span class="n">import</span> <span class="n">pandas</span> <span class="n">as</span> <span class="n">pd</span>
<span class="n">s_path</span> <span class="o">=</span> <span class="nf">pd.read_csv</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">None</span><span class="p">)</span>
<span class="c1"># split_path = s_path[0].str.split(&quot;.&quot;, expand=True)</span>
<span class="c1"># ID = pd.DataFrame({&#39;ID&#39;: split_path[1]})</span>
<span class="c1"># path = pd.DataFrame({&#39;path&#39;: s_path[0]})</span>
<span class="c1"># input_df = pd.concat([ID, path], axis=1)</span>
<span class="n">input_df</span> <span class="o">=</span> <span class="nf">s_path.values.tolist</span><span class="p">()</span>

<span class="n">parameter</span><span class="o">:</span> <span class="n">GWAS_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">LD_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">TAD_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">allele_qc_R</span> <span class="o">=</span> <span class="nf">path</span><span class="p">(</span><span class="s">&quot;~/pecotmr/R/allele_qc.R&quot;</span><span class="p">)</span>
<span class="n">allele_qc_R</span> <span class="o">=</span> <span class="n">f</span><span class="s">&quot;{allele_qc_R:a}&quot;</span>
<span class="c1">#input: susie_path</span>
<span class="n">input</span><span class="o">:</span> <span class="p">[</span><span class="n">x</span> <span class="n">for</span> <span class="n">x</span> <span class="n">in</span> <span class="n">input_df</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="m">1</span>
<span class="n">output</span><span class="o">:</span> <span class="n">f</span><span class="s">&#39;{cwd}/{step_name[:-2]}/{_input:bnn}.twas_z.rds&#39;</span>
<span class="n">task</span><span class="o">:</span> <span class="n">trunk_workers</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="n">walltime</span> <span class="o">=</span> <span class="n">walltime</span><span class="p">,</span> <span class="n">mem</span> <span class="o">=</span> <span class="n">mem</span><span class="p">,</span> <span class="n">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="n">R</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:nn}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:nn}.stdout&#39;</span><span class="p">,</span><span class="n">input</span><span class="o">=</span><span class="n">allele_qc_R</span>
<span class="c1">#twas_z_format = ptwas(susie_path = &quot;${_input:ar}&quot;,TAD_path = &quot;${TAD_path:ar}&quot;, GWAS_path = &quot;${GWAS_path:ar}&quot;, LD_path = &quot;${LD_path:ar}&quot;)</span>
<span class="c1">#fwrite(twas_z_format,${_output:ar},row.names=F,sep=&quot;\t&quot;,quote=F)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">plink2R</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">pillar</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">matrixcalc</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">genio</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">bigsnpr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Rlab</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">qvalue</span><span class="p">)</span>
<span class="n">np</span> <span class="o">=</span> <span class="nf">import</span><span class="p">(</span><span class="s">&quot;numpy&quot;</span><span class="p">)</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bedtools.path</span> <span class="o">=</span> <span class="s">&quot;/home/aw3600/bedtools2/bin/&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">bedtoolsr</span><span class="p">)</span>



<span class="n">twas_z</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">weights</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">R</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span> <span class="p">{</span>
    <span class="nf">if </span><span class="p">(</span><span class="nf">is.null</span><span class="p">(</span><span class="n">R</span><span class="p">))</span> <span class="p">{</span>
        <span class="c1"># mean impute X</span>
        <span class="n">genetype_data_imputed</span> <span class="o">&lt;-</span> <span class="nf">apply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span>
            <span class="n">pos</span> <span class="o">&lt;-</span> <span class="nf">which</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
            <span class="nf">if </span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">!=</span> <span class="m">0</span><span class="p">){</span>
                <span class="n">x</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="nf">mean</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">na.rm</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
                <span class="p">}</span>
                <span class="nf">return</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="p">})</span>
        <span class="c1"># need to add `large = T` in Rfast::cora, or could get wrong results </span>
        <span class="c1"># R &lt;- cora(genetype_data_imputed, large = T)</span>
        <span class="c1"># FIXME: test and enable using cora if Rfast is installed</span>
        <span class="c1"># See how susieR did it</span>
        <span class="n">R</span> <span class="o">&lt;-</span> <span class="nf">cor</span><span class="p">(</span><span class="n">genetype_data_imputed</span><span class="p">)</span> 
        <span class="nf">colnames</span><span class="p">(</span><span class="n">R</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">R</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">genetype_data_imputed</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">stat</span> <span class="o">&lt;-</span> <span class="nf">t</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span> <span class="o">%*%</span> <span class="n">z</span>
    <span class="n">denom</span> <span class="o">&lt;-</span> <span class="nf">t</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span> <span class="o">%*%</span> <span class="n">R</span> <span class="o">%*%</span> <span class="n">weights</span>
    <span class="n">zscore</span> <span class="o">&lt;-</span> <span class="n">stat</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="n">denom</span><span class="p">)</span>
    <span class="n">pval</span> <span class="o">&lt;-</span> <span class="nf">pchisq</span><span class="p">(</span> <span class="n">zscore</span> <span class="o">*</span> <span class="n">zscore</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="n">lower.tail</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
    <span class="nf">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">z</span><span class="o">=</span><span class="n">zscore</span><span class="p">,</span> <span class="n">pval</span><span class="o">=</span><span class="n">pval</span><span class="p">))</span>
<span class="p">}</span>



<span class="n">ptwas</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span><span class="n">TAD_path</span><span class="p">,</span><span class="n">GWAS_path</span><span class="p">,</span><span class="n">LD_path</span><span class="p">){</span>
<span class="c1">#####load susie_res_path</span>
<span class="n">susie</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">susie_path</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">data.frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>
<span class="n">gene_name</span> <span class="o">=</span> <span class="n">susie</span><span class="o">$</span><span class="n">ID</span>
<span class="n">susie_res</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">susie</span><span class="o">$</span><span class="n">path</span><span class="p">)</span>
<span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">sets</span><span class="p">))){</span>
<span class="c1">#####load LD path list</span>
<span class="n">LD_list</span> <span class="o">=</span> <span class="nf">read.table</span><span class="p">(</span><span class="n">LD_path</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="n">LD_list_pos.bed</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">LD_list</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;_&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">LD_list</span><span class="p">)</span>
<span class="c1">#####load TAD_region</span>
<span class="n">TAD_region</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="n">TAD_path</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="o">=</span><span class="s">&quot;gene_id&quot;</span><span class="p">)</span>
<span class="n">qtl_select</span> <span class="o">=</span> <span class="n">TAD_region</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">ID</span><span class="o">==</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">chr</span> <span class="o">=</span> <span class="nf">str_sub</span><span class="p">(</span><span class="n">qtl_select</span><span class="o">$</span><span class="n">`#chr`</span><span class="p">,</span><span class="m">4</span><span class="p">)</span>

<span class="n">qtl_reference</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">variant_names</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">simplify</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">data.frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="s">&quot;pos&quot;</span><span class="p">,</span><span class="s">&quot;A1&quot;</span><span class="p">,</span><span class="s">&quot;A2&quot;</span><span class="p">))</span>
<span class="c1">######load GWAS summary statistics</span>
<span class="n">AD_dataset</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">GWAS_path</span><span class="p">,</span><span class="s">&quot;/ADGWAS_Bellenguez_2022.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;/ADGWAS2022.chr&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;.sumstat.tsv&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
<span class="c1">##transform the AD_dataset to the form of allele_qc format</span>
<span class="n">AD_data</span> <span class="o">=</span> <span class="n">AD_dataset</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="n">chromosome</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">pos</span> <span class="o">=</span> <span class="nf">as.character</span><span class="p">(</span><span class="n">position</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">chromosome</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;A1&quot;</span><span class="o">=</span><span class="s">&quot;ref&quot;</span><span class="p">,</span><span class="s">&quot;A2&quot;</span><span class="o">=</span><span class="s">&quot;alt&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">z</span><span class="o">=</span><span class="n">beta</span><span class="o">/</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">position</span><span class="p">)</span>
<span class="n">AD_allele_flip</span> <span class="o">=</span> <span class="nf">allele_qc</span><span class="p">(</span><span class="n">AD_data</span><span class="p">,</span><span class="n">qtl_reference</span><span class="p">,</span><span class="n">match.min.prop</span><span class="o">=</span><span class="m">0.2</span><span class="p">,</span><span class="n">remove_dups</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">flip</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">remove</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="o">%&gt;%</span>
              <span class="nf">mutate</span><span class="p">(</span><span class="n">variant_allele_flip</span> <span class="o">=</span> <span class="nf">paste</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span><span class="n">pos</span><span class="p">,</span><span class="n">A1.sumstats</span><span class="p">,</span><span class="n">A2.sumstats</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;:&quot;</span><span class="p">))</span>
<span class="c1">####load LD matrix</span>
<span class="n">LD.files</span> <span class="o">&lt;-</span> <span class="n">bedtoolsr</span><span class="o">::</span><span class="nf">bt.intersect</span><span class="p">(</span><span class="n">a</span> <span class="o">=</span> <span class="n">LD_list_pos.bed</span><span class="p">,</span> <span class="n">b</span> <span class="o">=</span> <span class="n">qtl_select</span><span class="p">)</span>
<span class="n">LD.files.name</span> <span class="o">=</span> <span class="nf">unique</span><span class="p">(</span><span class="n">LD.files</span><span class="o">$</span><span class="n">V5</span><span class="p">)</span>
<span class="n">LD.list</span> <span class="o">=</span> <span class="nf">list</span><span class="p">()</span>
<span class="n">LD.matrix.names</span><span class="o">=</span><span class="kc">NULL</span>
<span class="nf">for </span><span class="p">(</span><span class="n">k</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">LD.files.name</span><span class="p">)){</span>
 <span class="n">npz</span> <span class="o">=</span> <span class="n">np</span><span class="o">$</span><span class="nf">load</span><span class="p">(</span><span class="n">LD.files.name</span><span class="p">[</span><span class="n">k</span><span class="p">])</span>
 <span class="n">LD.matrix</span> <span class="o">=</span> <span class="n">npz</span><span class="o">$</span><span class="n">f</span><span class="p">[[</span><span class="s">&quot;arr_0&quot;</span><span class="p">]]</span>
 <span class="n">LD.snps</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">LD.files.name</span><span class="p">[</span><span class="n">k</span><span class="p">],</span><span class="s">&quot;[.]&quot;</span><span class="p">,</span><span class="n">simplify</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">.</span><span class="p">[,</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">.</span><span class="p">),(</span><span class="nf">length</span><span class="p">(</span><span class="n">.</span><span class="p">)</span><span class="m">-1</span><span class="p">))]</span><span class="o">%&gt;%</span><span class="nf">paste</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">collapse</span><span class="o">=</span><span class="s">&quot;.&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">paste0</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="s">&quot;.bim&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">read.table</span><span class="p">(</span><span class="n">.</span><span class="p">)</span>
 <span class="c1">#head(LD.snps)</span>
 <span class="n">LD_names</span> <span class="o">=</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">LD.matrix</span><span class="p">)</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">LD.matrix</span><span class="p">)</span> <span class="o">=</span> <span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">LD.snps</span><span class="o">$</span><span class="n">V2</span><span class="p">)</span>
 <span class="n">snp_merge</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">LD_names</span><span class="p">,</span><span class="n">AD_allele_flip</span><span class="o">$</span><span class="n">variant_allele_flip</span><span class="p">)</span>
 <span class="n">LD.select</span> <span class="o">=</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="n">LD.matrix</span><span class="p">[</span><span class="n">snp_merge</span><span class="p">,</span><span class="n">snp_merge</span><span class="p">])</span>
 <span class="n">LD.list</span><span class="p">[[</span><span class="n">k</span><span class="p">]]</span> <span class="o">=</span> <span class="n">LD.select</span>
 <span class="n">LD.matrix.names</span> <span class="o">=</span> <span class="nf">append</span><span class="p">(</span><span class="n">LD.matrix.names</span><span class="p">,</span><span class="n">snp_merge</span><span class="p">)</span>
<span class="p">}</span>
   <span class="n">LD.block</span> <span class="o">=</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="nf">bdiag</span><span class="p">(</span><span class="n">LD.list</span><span class="p">))</span>
   <span class="nf">upperTriangle</span><span class="p">(</span><span class="n">LD.block</span><span class="p">,</span><span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span> <span class="o">=</span> <span class="nf">lowerTriangle</span><span class="p">(</span><span class="n">LD.block</span><span class="p">)</span>
   <span class="nf">colnames</span><span class="p">(</span><span class="n">LD.block</span><span class="p">)</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">LD.block</span><span class="p">)</span> <span class="o">=</span> <span class="n">LD.matrix.names</span>
<span class="c1">####generate the twas_z format input</span>
<span class="n">twas_z_format</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">LD.matrix.names</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene_name</span><span class="p">)</span><span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span> <span class="o">=</span> <span class="n">chr</span><span class="p">)</span><span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">AD_allele_flip</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">LD.matrix.names</span><span class="p">,</span><span class="n">AD_allele_flip</span><span class="o">$</span><span class="n">variant_allele_flip</span><span class="p">),]</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="n">se</span><span class="p">,</span><span class="n">z</span><span class="p">))</span><span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">susie_weights</span> <span class="o">=</span> <span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">susie_weights</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">LD.matrix.names</span><span class="p">,</span><span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">variant_names</span><span class="p">)])</span><span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">enet_weights</span> <span class="o">=</span>  <span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">enet_weights</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">LD.matrix.names</span><span class="p">,</span><span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">variant_names</span><span class="p">)])</span><span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">lasso_weights</span> <span class="o">=</span> <span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">lasso_weights</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">LD.matrix.names</span><span class="p">,</span><span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">variant_names</span><span class="p">)])</span><span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">mr_ash_weights</span> <span class="o">=</span> <span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">mr_ash_weights</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">LD.matrix.names</span><span class="p">,</span><span class="n">susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">variant_names</span><span class="p">)])</span><span class="o">%&gt;%</span>
    <span class="nf">rename</span><span class="p">(</span><span class="s">&quot;variants_name&quot;</span><span class="o">=</span><span class="s">&quot;LD.matrix.names&quot;</span><span class="p">)</span>
<span class="n">weights</span> <span class="o">=</span> <span class="nf">apply</span><span class="p">(</span><span class="n">twas_z_format</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_weights&quot;</span><span class="p">,</span><span class="s">&quot;enet_weights&quot;</span><span class="p">,</span><span class="s">&quot;lasso_weights&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_weights&quot;</span><span class="p">)],</span><span class="m">2</span><span class="p">,</span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">twas_z</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">twas_z_format</span><span class="o">$</span><span class="n">z</span><span class="p">,</span><span class="n">R</span> <span class="o">=</span> <span class="n">LD.block</span><span class="p">))</span>
<span class="n">twas_weights</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">gene_name</span><span class="o">=</span><span class="n">gene_name</span><span class="p">,</span><span class="n">chr</span> <span class="o">=</span> <span class="n">chr</span><span class="p">,</span><span class="n">weights</span><span class="o">$</span><span class="n">susie_weights</span><span class="o">$</span><span class="n">pval</span><span class="p">,</span><span class="n">weights</span><span class="o">$</span><span class="n">susie_weights</span><span class="o">$</span><span class="n">z</span><span class="p">,</span>
                          <span class="n">weights</span><span class="o">$</span><span class="n">lasso_weights</span><span class="o">$</span><span class="n">pval</span><span class="p">,</span><span class="n">weights</span><span class="o">$</span><span class="n">lasso_weights</span><span class="o">$</span><span class="n">z</span><span class="p">,</span><span class="n">weights</span><span class="o">$</span><span class="n">enet_weights</span><span class="o">$</span><span class="n">pval</span><span class="p">,</span><span class="n">weights</span><span class="o">$</span><span class="n">enet_weights</span><span class="o">$</span><span class="n">z</span><span class="p">,</span>
                          <span class="n">weights</span><span class="o">$</span><span class="n">mr_ash_weights</span><span class="o">$</span><span class="n">pval</span><span class="p">,</span><span class="n">weights</span><span class="o">$</span><span class="n">mr_ash_weights</span><span class="o">$</span><span class="n">z</span><span class="p">)</span>
  <span class="nf">names</span><span class="p">(</span><span class="n">twas_weights</span><span class="p">)</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;gene_name&quot;</span><span class="p">,</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="s">&quot;susie_pval&quot;</span><span class="p">,</span><span class="s">&quot;susie_z&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval&quot;</span><span class="p">,</span><span class="s">&quot;lasso_z&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval&quot;</span><span class="p">,</span><span class="s">&quot;enet_z&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_z&quot;</span><span class="p">)</span>
  <span class="n">p_values</span> <span class="o">=</span> <span class="n">twas_weights</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval&quot;</span><span class="p">)]</span>
  <span class="n">p_values</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">p_values</span><span class="p">)]</span> <span class="o">=</span> <span class="m">1</span>
  <span class="n">q_values</span> <span class="o">=</span> <span class="nf">qvalue</span><span class="p">(</span><span class="n">p_values</span><span class="p">,</span><span class="n">lambda</span><span class="o">=</span><span class="m">0</span><span class="p">)</span><span class="o">$</span><span class="n">qvalues</span>
  <span class="n">gene_weights_pq</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">twas_weights</span><span class="p">,</span><span class="n">qvalue</span> <span class="o">=</span> <span class="n">q_values</span><span class="p">)</span>
  <span class="nf">names</span><span class="p">(</span><span class="n">gene_weights_pq</span><span class="p">)[</span><span class="m">11</span><span class="o">:</span><span class="m">14</span><span class="p">]</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_qval&quot;</span><span class="p">,</span><span class="s">&quot;lasso_qval&quot;</span><span class="p">,</span><span class="s">&quot;enet_qval&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_qval&quot;</span><span class="p">)</span>
<span class="c1">####calculate the pvalue and zscore using twas_z function    </span>
<span class="nf">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">twas_z_format</span> <span class="o">=</span> <span class="n">twas_z_format</span><span class="p">,</span>
            <span class="n">gene_weights_pq</span> <span class="o">=</span> <span class="n">gene_weights_pq</span><span class="p">,</span>
            <span class="n">AD_allele_flip</span> <span class="o">=</span> <span class="n">AD_allele_flip</span><span class="p">))</span>
 <span class="p">}</span>
<span class="n">else</span> <span class="p">{</span>
    <span class="nf">cat</span><span class="p">(</span><span class="s">&quot;The &#39;susie_result&#39; is NULL, so no output is generated.\n&quot;</span><span class="p">)</span>
<span class="p">}</span>
<span class="p">}</span>

<span class="n">twas_z</span> <span class="o">=</span> <span class="nf">ptwas</span><span class="p">(</span><span class="n">susie_path</span> <span class="o">=</span> <span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">TAD_path</span> <span class="o">=</span> <span class="o">$</span><span class="p">{</span><span class="n">TAD_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">GWAS_path</span> <span class="o">=</span> <span class="o">$</span><span class="p">{</span><span class="n">GWAS_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">LD_path</span> <span class="o">=</span> <span class="o">$</span><span class="p">{</span><span class="n">LD_path</span><span class="o">:</span><span class="n">ar</span><span class="p">})</span>
<span class="nf">saveRDS</span><span class="p">(</span><span class="n">twas_z</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_z_format_2</span><span class="p">]</span>
<span class="c1"># Path to the input molecular phenotype data.</span>
<span class="n">input</span><span class="o">:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s">&quot;all&quot;</span>
<span class="n">output</span><span class="o">:</span> <span class="n">f</span><span class="s">&#39;{cwd}/{step_name[:-2]}/twas_z_files.txt&#39;</span>
<span class="n">python</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:n}.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
<span class="n">import</span> <span class="n">pandas</span> <span class="n">as</span> <span class="n">pd</span>
<span class="nf">pd.DataFrame</span><span class="p">({</span><span class="s">&quot;output&quot;</span> <span class="o">:</span> <span class="p">[</span><span class="o">$</span><span class="p">[</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">,]]})</span><span class="nf">.to_csv</span><span class="p">(</span><span class="s">&quot;$[_output]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="n">False</span> <span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="n">False</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="id2">
<h2>PTWAS<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id3">
<h3>Input<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>weights_path: A list of file paths of the PTWAS output</p></li>
<li><p>pval_threshold: pvalue threshold for selecting the candidate genes, it is default to be 0.05</p></li>
<li><p>cpip_cutoff: cpip cutoff threshold for MR method</p></li>
</ul>
</div>
<div class="section" id="id4">
<h3>Output<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>cand_genes.rds: the candidate significant genes obtained using the multiple weights</p></li>
<li><p>mr_output.rds: the MR results using the candidate genes (cand_genes.rds)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_candidate_gene</span><span class="p">]</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">weights_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">pval_threshold</span> <span class="o">=</span> <span class="m">0.05</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">cpip_cutoff</span> <span class="o">=</span> <span class="m">0.5</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">mr_R</span> <span class="o">=</span> <span class="nf">path</span><span class="p">(</span><span class="s">&quot;~/pecotmr/R/mr.R&quot;</span><span class="p">)</span>
<span class="n">mr_R</span> <span class="o">=</span> <span class="n">f</span><span class="s">&quot;{mr_R:a}&quot;</span>

<span class="n">input</span><span class="o">:</span> <span class="n">weights_path</span>
<span class="n">output</span><span class="o">:</span> <span class="n">cand_genes</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{cwd:a}/{step_name}.rds&#39;</span><span class="p">,</span>
        <span class="n">mr_output</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{cwd:a}/mr_output.rds&#39;</span>
<span class="n">task</span><span class="o">:</span> <span class="n">trunk_workers</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="n">walltime</span> <span class="o">=</span> <span class="n">walltime</span><span class="p">,</span> <span class="n">mem</span> <span class="o">=</span> <span class="n">mem</span><span class="p">,</span> <span class="n">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="n">R</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output[0]:n}.stdout&#39;</span><span class="p">,</span><span class="n">input</span><span class="o">=</span><span class="n">mr_R</span>


<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">qvalue</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">doMC</span><span class="p">)</span>

<span class="n">twas_mr_format_input</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">cand_genes</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">){</span>
<span class="n">mr_format_input</span> <span class="o">=</span> <span class="kc">NULL</span>
<span class="n">susie_path</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">susie</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">susie_path</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">susie_path</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>
<span class="nf">for </span><span class="p">(</span><span class="n">k</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">cand_genes</span><span class="p">)){</span>
<span class="n">gene_name</span> <span class="o">=</span> <span class="n">cand_genes</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
<span class="n">qtl_susie_res</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">susie</span><span class="o">$</span><span class="n">path</span><span class="p">[</span><span class="n">susie</span><span class="o">$</span><span class="n">ID</span><span class="o">==</span><span class="n">gene_name</span><span class="p">])</span>
<span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">qtl_susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">sets</span><span class="o">$</span><span class="n">cs</span><span class="p">))){</span>
<span class="n">qtl_finemap</span> <span class="o">=</span> <span class="n">qtl_susie_res</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">top_loci</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">cs_index_primary</span><span class="o">&gt;=</span><span class="m">1</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">cs_index_secondary</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">X_ID</span><span class="o">=</span><span class="n">gene_name</span><span class="p">)</span><span class="o">%&gt;%</span>
           <span class="nf">rename</span><span class="p">(</span><span class="s">&quot;snp&quot;</span><span class="o">=</span><span class="s">&quot;variant_id&quot;</span><span class="p">,</span><span class="s">&quot;bhat_x&quot;</span> <span class="o">=</span> <span class="s">&quot;bhat&quot;</span><span class="p">,</span><span class="s">&quot;sbhat_x&quot;</span><span class="o">=</span><span class="s">&quot;sbhat&quot;</span><span class="p">,</span><span class="s">&quot;cs&quot;</span> <span class="o">=</span> <span class="s">&quot;cs_index_primary&quot;</span><span class="p">)</span>
<span class="n">AD_allele_flip</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">weights_file_path</span><span class="o">$</span><span class="n">path</span><span class="p">[</span><span class="n">weights_file_path</span><span class="o">$</span><span class="n">ID</span><span class="o">==</span><span class="n">gene_name</span><span class="p">])</span><span class="o">$</span><span class="n">AD_allele_flip</span>
<span class="n">merge_snp</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">qtl_finemap</span><span class="o">$</span><span class="n">snp</span><span class="p">,</span><span class="n">AD_allele_flip</span><span class="o">$</span><span class="n">variant_allele_flip</span><span class="p">)</span>
<span class="n">format_input</span> <span class="o">=</span> <span class="n">qtl_finemap</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">merge_snp</span><span class="p">,</span><span class="n">qtl_finemap</span><span class="o">$</span><span class="n">snp</span><span class="p">),]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">AD_allele_flip</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">merge_snp</span><span class="p">,</span><span class="n">AD_allele_flip</span><span class="o">$</span><span class="n">variant_allele_flip</span><span class="p">),]</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;bhat_y&quot;</span><span class="o">=</span><span class="s">&quot;beta&quot;</span><span class="p">,</span><span class="s">&quot;sbhat_y&quot;</span><span class="o">=</span><span class="s">&quot;se&quot;</span><span class="p">))</span>
<span class="n">mr_format_input</span> <span class="o">=</span> <span class="nf">rbind</span><span class="p">(</span><span class="n">mr_format_input</span><span class="p">,</span> <span class="n">format_input</span><span class="p">)</span>
    <span class="p">}</span>
<span class="p">}</span>
<span class="nf">return</span><span class="p">(</span><span class="n">mr_format_input</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1">#detectCores()</span>
<span class="nf">registerDoMC</span><span class="p">(</span><span class="m">100</span><span class="p">)</span>
<span class="n">weights_file</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">weights_file_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">weights_file</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">weights_file</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>
<span class="c1">#ptm = proc.time()</span>
<span class="n">gene_pq</span> <span class="o">=</span> <span class="nf">foreach</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="nf">dim</span><span class="p">(</span><span class="n">weights_file</span><span class="p">)[</span><span class="m">1</span><span class="p">],</span> <span class="n">.combine</span><span class="o">=</span><span class="n">rbind</span><span class="p">)</span> <span class="o">%dopar%</span> <span class="p">{</span>
   <span class="n">weights_pq</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">weights_file_path</span><span class="o">$</span><span class="n">path</span><span class="p">[</span><span class="n">k</span><span class="p">])</span><span class="o">$</span><span class="n">gene_weights_pq</span>
<span class="p">}</span>
<span class="c1">#proc.time() - ptm</span>
<span class="n">padj</span> <span class="o">=</span> <span class="nf">apply</span><span class="p">(</span><span class="n">gene_pq</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval&quot;</span><span class="p">)],</span><span class="m">2</span><span class="p">,</span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">p.adjust</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;bonferroni&quot;</span><span class="p">,))</span><span class="o">%&gt;%</span><span class="nf">data.frame</span><span class="p">()</span>
<span class="n">gene_pq_adj</span> <span class="o">=</span> <span class="n">gene_pq</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">padj</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">gene_pq_adj</span><span class="p">)[</span><span class="m">15</span><span class="o">:</span><span class="m">18</span><span class="p">]</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval_adj&quot;</span><span class="p">)</span>

<span class="n">cand_genes</span> <span class="o">=</span> <span class="kc">NULL</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span> <span class="o">=</span> <span class="n">gene_pq_adj</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">susie</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">susie_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">lasso_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">enet</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">enet_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">mr_ash_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">susie_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">lasso_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">enet_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">mr_ash_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>

<span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">weights_file_path</span><span class="p">)</span>

<span class="nf">saveRDS</span><span class="p">(</span><span class="n">cand_genes</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="p">[</span><span class="m">0</span><span class="p">]</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>

<span class="n">mr_output</span> <span class="o">=</span> <span class="kc">NULL</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">susie</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">lasso</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">enet</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">mr_ash</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">susie_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">lasso_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">enet_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>
<span class="n">mr_output</span><span class="o">$</span><span class="n">mr_ash_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})</span>

<span class="nf">saveRDS</span><span class="p">(</span><span class="n">mr_output</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="example">
<h3>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h3>
<div class="section" id="bulk-data-set">
<h4>bulk data set<a class="headerlink" href="#bulk-data-set" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>fine mapping only on the 612 genes</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE.ipynb</span> <span class="n">susie</span> \
     <span class="o">--</span><span class="n">name</span> <span class="n">AD_DLPFC_bulk</span> \
     <span class="o">--</span><span class="n">genoFile</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">FunGen_xQTL</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">Genotype</span><span class="o">/</span><span class="n">geno_by_region</span><span class="o">/</span><span class="n">TADB_enhanced_cis_genotype_by_region</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.genotype_by_region_files.txt</span> \
     <span class="o">--</span><span class="n">phenoFile</span>   <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">pheno</span><span class="o">/</span><span class="n">process</span><span class="o">/</span><span class="n">DLPFC_bulk_geneid_eqtl_data.region_list.txt</span> \
     <span class="o">--</span><span class="n">covFile</span>  <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">DLPFC_covs_genoPCs_marchPCs.txt</span> \
     <span class="o">--</span><span class="n">phenotype_names</span> <span class="n">AD</span> \
     <span class="o">--</span><span class="n">utils</span><span class="o">-</span><span class="n">R</span>  <span class="o">~/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">xqtl_utils.R</span> \
     <span class="o">--</span><span class="n">maf</span> <span class="m">0.01</span> \
     <span class="o">--</span><span class="n">pip_cutoff</span> <span class="m">0.1</span> \
     <span class="o">--</span><span class="n">coverage</span> <span class="m">0.95</span> \
     <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="n">list</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">AD_phenotype_select.txt</span> \
     <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">indel</span> \
     <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">DLPFC_bulk_AD_SuSiE_results</span> \
     <span class="o">--</span><span class="n">mem</span> <span class="m">200</span><span class="n">G</span> <span class="o">-</span><span class="n">J</span> <span class="m">50</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">csg.yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">ptwas.ipynb</span> <span class="n">twas_z_format</span> \
        <span class="o">--</span><span class="n">susie_path</span>  <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">DLPFC_bulk_AD_SuSiE_results</span><span class="o">/</span><span class="n">AD_DLPFC_bulk.susie_output.txt</span> \
        <span class="o">--</span><span class="n">TAD_path</span> <span class="o">~/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">resource</span><span class="o">/</span><span class="n">TADB_enhanced_cis.bed</span> \
        <span class="o">--</span><span class="n">GWAS_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">ADGWAS</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">ADGWAS2022</span> \
        <span class="o">--</span><span class="n">LD_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">LD</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="m">1300</span>_<span class="n">hg38_EUR_LD_blocks_LD</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS.leftnorm.filtered.filtered.ld.list</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span> \
        <span class="o">--</span><span class="n">mem</span> <span class="m">80</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">ptwas.ipynb</span> <span class="n">twas_candidate_gene</span> \
        <span class="o">--</span><span class="n">susie_path</span>  <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">DLPFC_bulk_AD_SuSiE_results</span><span class="o">/</span><span class="n">AD_DLPFC_bulk.susie_output.txt</span> \
        <span class="o">--</span><span class="n">weights_path</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">twas_z_format</span><span class="o">/</span><span class="n">AD_DLPFC_bulk.twas_z_files.txt</span> \
        <span class="o">--</span><span class="n">pval_threshold</span> <span class="m">0.05</span> \
        <span class="o">--</span><span class="n">cpip_cutoff</span> <span class="m">0.5</span> \
        <span class="o">--</span><span class="n">mr_R</span> <span class="o">~/</span><span class="n">pecotmr</span><span class="o">/</span><span class="n">R</span><span class="o">/</span><span class="n">mr.R</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">DLPFC_bulk_data</span> \
        <span class="o">--</span><span class="n">mem</span> <span class="m">80</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="n">DLPFC_genes</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;~/MR_KMT_analysis/PTWAS/DLPFC_bulk_data/twas_candidate_gene.rds&quot;</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>482</li><li>18</li></ol>
</div><div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 18</caption>
<thead>
	<tr><th></th><th scope=col>gene_name</th><th scope=col>chr</th><th scope=col>susie_pval</th><th scope=col>susie_z</th><th scope=col>lasso_pval</th><th scope=col>lasso_z</th><th scope=col>enet_pval</th><th scope=col>enet_z</th><th scope=col>mr_ash_pval</th><th scope=col>mr_ash_z</th><th scope=col>susie_qval</th><th scope=col>lasso_qval</th><th scope=col>enet_qval</th><th scope=col>mr_ash_qval</th><th scope=col>susie_pval_adj</th><th scope=col>lasso_pval_adj</th><th scope=col>enet_pval_adj</th><th scope=col>mr_ash_pval_adj</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>ENSG00000188157</td><td>1</td><td>0.94223145</td><td> 0.07246551</td><td>       NaN</td><td>        NaN</td><td>      NaN</td><td>        NaN</td><td>0.896210155</td><td>-0.1304503</td><td>1.00000000</td><td>1.00000000</td><td>1.00000000</td><td>1.000000000</td><td>1</td><td>NaN</td><td>NaN</td><td>1.00000</td></tr>
	<tr><th scope=row>2</th><td>ENSG00000074800</td><td>1</td><td>0.45220217</td><td>-0.75174874</td><td>0.89523247</td><td> 0.13168624</td><td>0.6625478</td><td>-0.43639828</td><td>0.499605350</td><td>-0.6751108</td><td>0.88339705</td><td>0.89523247</td><td>0.88339705</td><td>0.883397048</td><td>1</td><td>  1</td><td>  1</td><td>1.00000</td></tr>
	<tr><th scope=row>3</th><td>ENSG00000177000</td><td>1</td><td>0.12225353</td><td>-1.54538351</td><td>0.03928815</td><td>-2.06115615</td><td>0.0426312</td><td>-2.02730527</td><td>0.085125875</td><td>-1.7216890</td><td>0.12225353</td><td>0.08526241</td><td>0.08526241</td><td>0.113501166</td><td>1</td><td>  1</td><td>  1</td><td>1.00000</td></tr>
	<tr><th scope=row>4</th><td>ENSG00000117118</td><td>1</td><td>0.73754554</td><td>-0.33510546</td><td>0.95718135</td><td>-0.05369101</td><td>0.9518644</td><td> 0.06036566</td><td>0.761034824</td><td>-0.3041222</td><td>0.95718135</td><td>0.95718135</td><td>0.95718135</td><td>0.957181348</td><td>1</td><td>  1</td><td>  1</td><td>1.00000</td></tr>
	<tr><th scope=row>5</th><td>ENSG00000117115</td><td>1</td><td>0.26937576</td><td>-1.10450124</td><td>0.07594689</td><td>-1.77470331</td><td>0.1178218</td><td>-1.56398218</td><td>0.522042520</td><td> 0.6402001</td><td>0.35916769</td><td>0.23564351</td><td>0.23564351</td><td>0.522042520</td><td>1</td><td>  1</td><td>  1</td><td>1.00000</td></tr>
	<tr><th scope=row>6</th><td>ENSG00000158748</td><td>1</td><td>0.08305581</td><td> 1.73322428</td><td>0.06390926</td><td> 1.85281234</td><td>0.0634120</td><td> 1.85629165</td><td>0.001551992</td><td> 3.1647820</td><td>0.08305581</td><td>0.08305581</td><td>0.08305581</td><td>0.006207967</td><td>1</td><td>  1</td><td>  1</td><td>0.74806</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>After fine mapping, there are only 482 genes have significant results. Thus, we only list the pvalue and qvalue of 482 genes.</p>
<p>We set pvalue threshold to be 0.05 and compare the bonferroni adjusted pvalue with the 0.05,</p>
<ul class="simple">
<li><p>For the susie method, we obtained 28 candidate significant genes;</p></li>
<li><p>For the lasso method, we obtained 12 candidate significant genes;</p></li>
<li><p>For the enet method, we obtained 12 candidate significant genes;</p></li>
<li><p>For the mr_ash method, we obtained 23 candidate significant genes</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">length</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">28</div><div class="output text_html">12</div><div class="output text_html">12</div><div class="output text_html">23</div></div>
</div>
<ul class="simple">
<li><p>After merge the candidate genes of four methods, we finally obtained genes</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">candidate_genes_merge</span> <span class="o">=</span> <span class="nf">unique</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">))</span>
<span class="nf">length</span><span class="p">(</span><span class="n">candidate_genes_merge</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">40</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">candidate_genes_merge</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'ENSG00000117322'</li><li>'ENSG00000203710'</li><li>'ENSG00000196549'</li><li>'ENSG00000196126'</li><li>'ENSG00000234745'</li><li>'ENSG00000240065'</li><li>'ENSG00000095970'</li><li>'ENSG00000086288'</li><li>'ENSG00000106351'</li><li>'ENSG00000095585'</li><li>'ENSG00000109919'</li><li>'ENSG00000110079'</li><li>'ENSG00000100599'</li><li>'ENSG00000103550'</li><li>'ENSG00000099364'</li><li>'ENSG00000099381'</li><li>'ENSG00000103495'</li><li>'ENSG00000103510'</li><li>'ENSG00000149922'</li><li>'ENSG00000156885'</li><li>'ENSG00000108556'</li><li>'ENSG00000030582'</li><li>'ENSG00000005379'</li><li>'ENSG00000159640'</li><li>'ENSG00000007255'</li><li>'ENSG00000073008'</li><li>'ENSG00000130204'</li><li>'ENSG00000161640'</li><li>'ENSG00000204520'</li><li>'ENSG00000120885'</li><li>'ENSG00000064300'</li><li>'ENSG00000182087'</li><li>'ENSG00000104853'</li><li>'ENSG00000150672'</li><li>'ENSG00000104856'</li><li>'ENSG00000104884'</li><li>'ENSG00000130208'</li><li>'ENSG00000138442'</li><li>'ENSG00000127184'</li><li>'ENSG00000197943'</li></ol>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mr_output</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;~/MR_KMT_analysis/PTWAS/DLPFC_bulk_data/mr_output.rds&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">DLPFC_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 9</caption>
<thead>
	<tr><th></th><th scope=col>snp</th><th scope=col>maf</th><th scope=col>bhat_x</th><th scope=col>sbhat_x</th><th scope=col>pip</th><th scope=col>cs</th><th scope=col>X_ID</th><th scope=col>bhat_y</th><th scope=col>sbhat_y</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>chr1:207505962:A:G</td><td>0.2008929</td><td>-0.1833420</td><td>0.03515377</td><td>0.03532058</td><td>1</td><td>ENSG00000117322</td><td>-0.1093</td><td>0.0102</td></tr>
	<tr><th scope=row>2</th><td>chr1:207510847:T:G</td><td>0.1932398</td><td>-0.1878704</td><td>0.03512318</td><td>0.05683321</td><td>1</td><td>ENSG00000117322</td><td>-0.1234</td><td>0.0104</td></tr>
	<tr><th scope=row>3</th><td>chr1:207512441:T:C</td><td>0.2028061</td><td>-0.1862302</td><td>0.03513435</td><td>0.05114225</td><td>1</td><td>ENSG00000117322</td><td>-0.1115</td><td>0.0102</td></tr>
	<tr><th scope=row>4</th><td>chr1:207512620:A:C</td><td>0.1934866</td><td>-0.1923521</td><td>0.03509215</td><td>0.11236450</td><td>1</td><td>ENSG00000117322</td><td>-0.1240</td><td>0.0104</td></tr>
	<tr><th scope=row>5</th><td>chr1:207518704:A:G</td><td>0.1926020</td><td>-0.1950117</td><td>0.03507337</td><td>0.16032051</td><td>1</td><td>ENSG00000117322</td><td>-0.1253</td><td>0.0104</td></tr>
	<tr><th scope=row>6</th><td>chr1:207524699:T:C</td><td>0.2021684</td><td>-0.1877230</td><td>0.03512419</td><td>0.06340718</td><td>1</td><td>ENSG00000117322</td><td>-0.1126</td><td>0.0102</td></tr>
</tbody>
</table>
</div></div>
</div>
<ul class="simple">
<li><p>Apply mr method to the susie_adj_format_input</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mr_output</span><span class="o">$</span><span class="n">susie_adj</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 6 × 10</caption>
<thead>
	<tr><th scope=col>X_ID</th><th scope=col>num_CS</th><th scope=col>num_IV</th><th scope=col>meta_eff</th><th scope=col>se_meta_eff</th><th scope=col>meta_pval</th><th scope=col>meta_qval</th><th scope=col>Q</th><th scope=col>Q_pval</th><th scope=col>I2</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>ENSG00000117322</td><td>1</td><td> 19</td><td> 0.022</td><td>0.005</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>ENSG00000203710</td><td>1</td><td> 12</td><td> 0.010</td><td>0.001</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><td>ENSG00000110079</td><td>1</td><td>124</td><td> 0.013</td><td>0.004</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><td>ENSG00000103550</td><td>1</td><td> 36</td><td>-0.005</td><td>0.001</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><td>ENSG00000108556</td><td>1</td><td> 11</td><td> 0.005</td><td>0.001</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
	<tr><td>ENSG00000161640</td><td>1</td><td>  5</td><td> 0.005</td><td>0.001</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>We can obtain six causal genes using the mr method. In fact, meta_pvals of ENSG00000108556 and ENSG00000161640 are equal to 0.0002 and 0.0003. Because we use round(x,3) to keep meta_pval to three decimal, all the meta_pval are zero in the table. And the Q value of first gene is equal to <span class="math notranslate nohighlight">\(5.538\times10^{-31}\)</span>, so the Q_pval is equal to 0.</p>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./code/pecotmr_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="pecotmr_preprocessing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">xQTL and GWAS region alignment</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="ptwas_mr.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Mendelian Randomization using fine-mapping IV</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>