
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>TWAS for xQTL &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Mendelian Randomization using fine-mapping IV" href="ptwas_mr.html" />
    <link rel="prev" title="xQTL and GWAS region alignment" href="pecotmr_preprocessing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/MRAID_QTL.html">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR: enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="ptwas_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/pecotmr_analysis/ptwas.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_analysis/ptwas.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/pecotmr_analysis/ptwas.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ouput">
     Ouput
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#twas">
   TWAS
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#output">
     Output
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mr">
   MR
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Output
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example">
     Example
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mic-pseudo-bulk-dataset">
       Mic pseudo bulk dataset
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mic-twas-results">
       Mic TWAS results
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mic-mr-results">
       Mic MR results
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="twas-for-xqtl">
<h1>TWAS for xQTL<a class="headerlink" href="#twas-for-xqtl" title="Permalink to this headline">¶</a></h1>
<p>This module contains the software implementations to perform transcriptome-wide association analysis (TWAS). These methods are designed to perform rigorous causal inference connecting genes to complex traits.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The goal of this module is to perform PTWAS analysis from SuSiE objects, including:</p>
<ul class="simple">
<li><p><strong>Step 1: PTWAS (twas_z_format)</strong></p></li>
</ul>
<ol class="simple">
<li><p>load GWAS z-score</p></li>
<li><p>load the corresponding LD matrices within the TAD region of each gene</p></li>
<li><p>use allele_qc() function to QC the LD matrix with summary stats</p></li>
<li><p>extract susie, lasso, enet and mr_ash weights</p></li>
<li><p>use twas_z() to compute TWAS results from multipile weights</p></li>
</ol>
<ul class="simple">
<li><p><strong>Step 2: MR (twas_candidate_genes)</strong></p></li>
</ul>
<ol class="simple">
<li><p>use some p-value cutoff to loosely pick TWAS regions of interest then if the region passes the cutoff, save the QC-ed GWAS data in the format compatabile to <a class="reference external" href="https://github.com/cumc/pecotmr/blob/main/R/mr.R">mr.R</a></p></li>
</ol>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>Step 1: TWAS</strong></p></li>
</ul>
<ol class="simple">
<li><p>QTL susie table：</p></li>
<li><p>GWAS sumstats results (tsv format)</p></li>
<li><p>LD reference</p></li>
<li><p>TAD region</p></li>
</ol>
<ul class="simple">
<li><p><strong>Step 2: MR</strong></p></li>
</ul>
<ol class="simple">
<li><p>output of Step 1</p></li>
</ol>
</div>
<div class="section" id="ouput">
<h3>Ouput<a class="headerlink" href="#ouput" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>Step 1: TWAS</strong></p></li>
</ul>
<ol class="simple">
<li><p>TWAS results from multiple weights</p></li>
<li><p>GWAS summary statistics after QC</p></li>
<li><p>the input format of twas_z() function</p></li>
</ol>
<ul class="simple">
<li><p><strong>Step 2: MR</strong></p></li>
</ul>
<ol class="simple">
<li><p>the candidate genes of TWAS</p></li>
<li><p>the input format of <a class="reference external" href="https://github.com/cumc/pecotmr/blob/main/R/mr.R">mr.R</a></p></li>
<li><p>the MR results</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">global</span><span class="p">]</span>
<span class="c1"># Workdir</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">susie_path</span> <span class="o">=</span> <span class="n">paths</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="nf">path</span><span class="p">(</span><span class="s">&quot;output&quot;</span><span class="p">)</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">container</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
<span class="n">import</span> <span class="n">re</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s">&#39; &#39;</span> <span class="o">+</span> <span class="nf">re.sub</span><span class="p">(</span><span class="n">r</span><span class="s">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">,</span> <span class="nf">container.split</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)[</span><span class="m">-1</span><span class="p">]))</span> <span class="n">if</span> <span class="n">container</span> <span class="n">else</span> <span class="s">&quot;&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="m">100</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">walltime</span> <span class="o">=</span> <span class="s">&quot;1h&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">mem</span> <span class="o">=</span> <span class="s">&quot;16G&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="m">20</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="twas">
<h2>TWAS<a class="headerlink" href="#twas" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Input<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>susie_path: A list of file paths for susie results</p></li>
<li><p>GWAS_path: gwas summary statistics path (to load the gwas sumstats dataframe)</p></li>
<li><p>LD_path: LD block matrix path (a list of file paths of LD block)</p></li>
<li><p>TAD_path: TAD region path (a dataframe  of TAD region of each gene, this one is applied to subset the LD matrix for each gene)</p></li>
</ul>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>twas_z_format: a dataframe of input format for the twas_z() function</p></li>
<li><p>gene_weights_pq: a dataframe of output of twas_z() function, we apply twas_z function to the  weights of four methods (susie, lasso, enet and mr_ash) to calculate the pvalue, after obtaining the pvalue, we also calculate the corresponding qvalue</p></li>
<li><p>AD_allele_flip: AD gwas sumstats after QC</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_z_format_1</span><span class="p">]</span>

<span class="n">import</span> <span class="n">pandas</span> <span class="n">as</span> <span class="n">pd</span>
<span class="n">s_path</span> <span class="o">=</span> <span class="nf">pd.read_csv</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">None</span><span class="p">)</span>
<span class="c1"># split_path = s_path[0].str.split(&quot;.&quot;, expand=True)</span>
<span class="c1"># ID = pd.DataFrame({&#39;ID&#39;: split_path[1]})</span>
<span class="c1"># path = pd.DataFrame({&#39;path&#39;: s_path[0]})</span>
<span class="c1"># input_df = pd.concat([ID, path], axis=1)</span>
<span class="n">input_df</span> <span class="o">=</span> <span class="nf">s_path.values.tolist</span><span class="p">()</span>

<span class="n">parameter</span><span class="o">:</span> <span class="n">GWAS_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">LD_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">TAD_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1">#input: susie_path</span>
<span class="n">input</span><span class="o">:</span> <span class="p">[</span><span class="n">x</span> <span class="n">for</span> <span class="n">x</span> <span class="n">in</span> <span class="n">input_df</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="m">1</span>
<span class="n">output</span><span class="o">:</span> <span class="n">f</span><span class="s">&#39;{cwd}/{step_name[:-2]}/{_input:bnn}.twas_z.rds&#39;</span>
<span class="n">task</span><span class="o">:</span> <span class="n">trunk_workers</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="n">walltime</span> <span class="o">=</span> <span class="n">walltime</span><span class="p">,</span> <span class="n">mem</span> <span class="o">=</span> <span class="n">mem</span><span class="p">,</span> <span class="n">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="n">R</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:nn}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:nn}.stdout&#39;</span>

<span class="nf">options</span><span class="p">(</span><span class="n">bedtools.path</span> <span class="o">=</span> <span class="s">&quot;/home/aw3600/bedtools2/bin/&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">pecotmr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>

<span class="n">susie_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">data.frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>
<span class="n">TAD_region</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">TAD_path</span><span class="o">:</span><span class="n">ar</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="o">=</span><span class="s">&quot;gene_id&quot;</span><span class="p">)</span>
<span class="n">gene_name</span> <span class="o">=</span> <span class="n">susie_path</span><span class="o">$</span><span class="n">ID</span>
<span class="n">region</span> <span class="o">=</span> <span class="n">TAD_region</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">ID</span><span class="o">==</span><span class="n">gene_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="o">=</span><span class="s">&quot;#chr&quot;</span><span class="p">)</span>
<span class="n">chr</span> <span class="o">=</span> <span class="nf">str_sub</span><span class="p">(</span><span class="n">region</span><span class="o">$</span><span class="n">chr</span><span class="p">,</span><span class="m">4</span><span class="p">)</span>
<span class="n">AD_dataset</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">GWAS_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="s">&quot;/ADGWAS_Bellenguez_2022.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;/ADGWAS2022.chr&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;.sumstat.tsv&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
<span class="n">GWAS_data</span> <span class="o">=</span> <span class="n">AD_dataset</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="n">chromosome</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">pos</span> <span class="o">=</span> <span class="nf">as.character</span><span class="p">(</span><span class="n">position</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">chromosome</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;A1&quot;</span><span class="o">=</span><span class="s">&quot;ref&quot;</span><span class="p">,</span><span class="s">&quot;A2&quot;</span><span class="o">=</span><span class="s">&quot;alt&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">z</span><span class="o">=</span><span class="n">beta</span><span class="o">/</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">position</span><span class="p">)</span>
<span class="n">LD_list</span> <span class="o">=</span> <span class="nf">read.table</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">LD_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="n">LD_block_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">LD_list</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;_&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">LD_list</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="o">=</span><span class="s">&quot;1&quot;</span><span class="p">,</span><span class="s">&quot;start&quot;</span><span class="o">=</span><span class="s">&quot;2&quot;</span><span class="p">,</span><span class="s">&quot;end&quot;</span><span class="o">=</span><span class="s">&quot;3&quot;</span><span class="p">,</span><span class="s">&quot;variants&quot;</span><span class="o">=</span><span class="s">&quot;V1&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span> <span class="o">=</span> <span class="s">&quot;V2&quot;</span><span class="p">)</span>
<span class="n">twas_z</span> <span class="o">=</span> <span class="nf">twas_scan</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span> <span class="n">GWAS_data</span><span class="p">,</span> <span class="n">LD_block_path</span><span class="p">)</span>
<span class="nf">saveRDS</span><span class="p">(</span><span class="n">twas_z</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_z_format_2</span><span class="p">]</span>
<span class="c1"># Path to the input molecular phenotype data.</span>
<span class="n">input</span><span class="o">:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s">&quot;all&quot;</span>
<span class="n">output</span><span class="o">:</span> <span class="n">f</span><span class="s">&#39;{cwd}/{step_name[:-2]}/twas_z_files.txt&#39;</span>
<span class="n">python</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:n}.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
<span class="n">import</span> <span class="n">pandas</span> <span class="n">as</span> <span class="n">pd</span>
<span class="nf">pd.DataFrame</span><span class="p">({</span><span class="s">&quot;output&quot;</span> <span class="o">:</span> <span class="p">[</span><span class="o">$</span><span class="p">[</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">,]]})</span><span class="nf">.to_csv</span><span class="p">(</span><span class="s">&quot;$[_output]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="n">False</span> <span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="n">False</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="mr">
<h2>MR<a class="headerlink" href="#mr" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id2">
<h3>Input<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>weights_path: A list of file paths of the PTWAS output</p></li>
<li><p>pval_threshold: pvalue threshold for selecting the candidate genes, it is default to be 0.05</p></li>
<li><p>cpip_cutoff: cpip cutoff threshold for MR method</p></li>
</ul>
</div>
<div class="section" id="id3">
<h3>Output<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>twas_cand_genes.rds: the candidate significant genes obtained using the multiple weights</p></li>
<li><p>mr_output.rds: the MR results using the candidate genes (cand_genes.rds)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_candidate_gene</span><span class="p">]</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">weights_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">pval_threshold</span> <span class="o">=</span> <span class="m">0.05</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">cpip_cutoff</span> <span class="o">=</span> <span class="m">0.5</span>


<span class="n">input</span><span class="o">:</span> <span class="n">weights_path</span>
<span class="n">output</span><span class="o">:</span> <span class="n">cand_genes</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{cwd:a}/{step_name}.rds&#39;</span><span class="p">,</span>
        <span class="n">mr_output</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{cwd:a}/mr_output.rds&#39;</span>
<span class="n">task</span><span class="o">:</span> <span class="n">trunk_workers</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="n">walltime</span> <span class="o">=</span> <span class="n">walltime</span><span class="p">,</span> <span class="n">mem</span> <span class="o">=</span> <span class="n">mem</span><span class="p">,</span> <span class="n">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="n">R</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output[0]:n}.stdout&#39;</span>


<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">qvalue</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">doMC</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">pecotmr</span><span class="p">)</span>

<span class="c1">####</span>
<span class="n">susie_list</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">susie_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">susie_list</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">susie_list</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>

<span class="c1">#detectCores()</span>
<span class="nf">registerDoMC</span><span class="p">(</span><span class="m">100</span><span class="p">)</span>
<span class="n">weights_file</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">weights_file_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">weights_file</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">weights_file</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>
<span class="c1">#ptm = proc.time()</span>
<span class="n">gene_pq</span> <span class="o">=</span> <span class="nf">foreach</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="nf">dim</span><span class="p">(</span><span class="n">weights_file</span><span class="p">)[</span><span class="m">1</span><span class="p">],</span> <span class="n">.combine</span><span class="o">=</span><span class="n">rbind</span><span class="p">)</span> <span class="o">%dopar%</span> <span class="p">{</span>
   <span class="n">weights_pq</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">weights_file_path</span><span class="o">$</span><span class="n">path</span><span class="p">[</span><span class="n">k</span><span class="p">])</span><span class="o">$</span><span class="n">gene_weights_pq</span>
<span class="p">}</span>
<span class="c1">#proc.time() - ptm</span>
<span class="n">padj</span> <span class="o">=</span> <span class="nf">apply</span><span class="p">(</span><span class="n">gene_pq</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval&quot;</span><span class="p">)],</span><span class="m">2</span><span class="p">,</span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">p.adjust</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;bonferroni&quot;</span><span class="p">,))</span><span class="o">%&gt;%</span><span class="nf">data.frame</span><span class="p">()</span>
<span class="n">gene_pq_adj</span> <span class="o">=</span> <span class="n">gene_pq</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">padj</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">gene_pq_adj</span><span class="p">)[</span><span class="m">15</span><span class="o">:</span><span class="m">18</span><span class="p">]</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval_adj&quot;</span><span class="p">)</span>
<span class="n">gene.ref</span> <span class="o">=</span> <span class="nf">read.table</span><span class="p">(</span><span class="s">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_id</span> <span class="o">=</span> <span class="n">gene.ref</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene.ref</span><span class="o">$</span><span class="n">V4</span><span class="p">),</span><span class="s">&quot;V5&quot;</span><span class="p">]</span>

<span class="n">cand_genes</span> <span class="o">=</span> <span class="kc">NULL</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span> <span class="o">=</span> <span class="n">gene_pq_adj</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">susie</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">susie_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">lasso_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">enet</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">enet_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">mr_ash_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">susie_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">lasso_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">enet_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">mr_ash_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>

<span class="c1"># cand_genes$susie_format_input = twas_mr_format_input(cand_genes$susie$gene_name,${susie_path:ar},weights_file_path)</span>
<span class="c1"># cand_genes$lasso_format_input = twas_mr_format_input(cand_genes$lasso$gene_name,${susie_path:ar},weights_file_path)</span>
<span class="c1"># cand_genes$enet_format_input = twas_mr_format_input(cand_genes$enet$gene_name,${susie_path:ar},weights_file_path)</span>
<span class="c1"># cand_genes$mr_ash_format_input = twas_mr_format_input(cand_genes$mr_ash$gene_name,${susie_path:ar},weights_file_path)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>
<span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>

<span class="nf">saveRDS</span><span class="p">(</span><span class="n">cand_genes</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="p">[</span><span class="m">0</span><span class="p">]</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>

<span class="n">mr_output</span> <span class="o">=</span> <span class="kc">NULL</span>
<span class="c1"># mr_output$susie = fine_mr(cand_genes$susie_format_input,${cpip_cutoff})</span>
<span class="c1"># mr_output$lasso = fine_mr(cand_genes$lasso_format_input,${cpip_cutoff})</span>
<span class="c1"># mr_output$enet = fine_mr(cand_genes$enet_format_input,${cpip_cutoff})</span>
<span class="c1"># mr_output$mr_ash = fine_mr(cand_genes$mr_ash_format_input,${cpip_cutoff})</span>
<span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span><span class="p">))</span>
<span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">susie_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>
<span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span><span class="p">))</span>
<span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">lasso_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>
<span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span><span class="p">))</span>
<span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">enet_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>
<span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span><span class="p">))</span>
<span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">mr_ash_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>

<span class="nf">saveRDS</span><span class="p">(</span><span class="n">mr_output</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="example">
<h3>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h3>
<div class="section" id="mic-pseudo-bulk-dataset">
<h4>Mic pseudo bulk dataset<a class="headerlink" href="#mic-pseudo-bulk-dataset" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>fine mapping only on the 612 genes</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">cis_analysis</span><span class="o">/</span><span class="n">cis_workhorse.ipynb</span> <span class="n">susie</span> \
     <span class="o">--</span><span class="n">name</span> <span class="n">Mic_pseudo_bulk</span> \
     <span class="o">--</span><span class="n">genoFile</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">FunGen_xQTL</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">Genotype</span><span class="o">/</span><span class="n">geno_by_region</span><span class="o">/</span><span class="n">TADB_enhanced_cis_genotype_by_region</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.genotype_by_region_files.txt</span>  \
     <span class="o">--</span><span class="n">phenoFile</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Mic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">phenotype_data</span><span class="o">/</span><span class="n">Mic.log2cpm.region_list.txt</span> \
     <span class="o">--</span><span class="n">covFile</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Mic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">covariate_data</span><span class="o">/</span><span class="n">Mic.log2cpm.Mic.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.snuc_pseudo_bulk.related.plink_qc.extracted.pca.projected.Marchenko_PC.gz</span> \
     <span class="o">--</span><span class="n">phenotype_names</span> <span class="n">AD</span> \
     <span class="o">--</span><span class="n">maf</span> <span class="m">0.01</span> \
     <span class="o">--</span><span class="n">pip_cutoff</span> <span class="m">0.1</span> \
     <span class="o">--</span><span class="n">coverage</span> <span class="m">0.95</span> \
     <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="n">list</span>   <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">Mic_AD_phenotype_select.txt</span> \
     <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">indel</span> \
     <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">Mic_pseudo_bulk_SuSiE_results</span> \
     <span class="o">--</span><span class="n">mem</span> <span class="m">200</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">pecotmr_analysis</span><span class="o">/</span><span class="n">ptwas.ipynb</span> <span class="n">twas_z_format</span> \
        <span class="o">--</span><span class="n">susie_path</span>  <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">Mic_pseudo_bulk_SuSiE_results</span><span class="o">/</span><span class="n">Mic_pseudo_bulk.cis_workhorse_output.txt</span> \
        <span class="o">--</span><span class="n">TAD_path</span> <span class="o">~/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">resource</span><span class="o">/</span><span class="n">TADB_enhanced_cis.bed</span> \
        <span class="o">--</span><span class="n">GWAS_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">ADGWAS</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">ADGWAS2022</span> \
        <span class="o">--</span><span class="n">LD_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">LD</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="m">1300</span>_<span class="n">hg38_EUR_LD_blocks_LD</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS.leftnorm.filtered.filtered.ld.list</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">Mic</span> \
        <span class="o">--</span><span class="n">mem</span> <span class="m">40</span><span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">pecotmr_analysis</span><span class="o">/</span><span class="n">ptwas.ipynb</span> <span class="n">twas_candidate_gene</span> \
        <span class="o">--</span><span class="n">susie_path</span>  <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">Mic_pseudo_bulk_SuSiE_results</span><span class="o">/</span><span class="n">Mic_pseudo_bulk.cis_workhorse_output.txt</span> \
        <span class="o">--</span><span class="n">weights_path</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">Mic</span><span class="o">/</span><span class="n">twas_z_format</span><span class="o">/</span><span class="n">twas_z_files.txt</span> \
        <span class="o">--</span><span class="n">cpip_cutoff</span> <span class="m">0.5</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">Mic</span> \
        <span class="o">--</span><span class="n">mem</span> <span class="m">80</span><span class="n">G</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="n">Mic_genes</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;~/MR_KMT_analysis/PTWAS/Mic/twas_candidate_gene.rds&quot;</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">arrange</span><span class="p">(</span><span class="n">susie_pval_adj</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Attaching package: ‘dplyr’


The following objects are masked from ‘package:data.table’:

    between, first, last


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union
</pre></div>
</div>
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>337</li><li>19</li></ol>
</div><div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 19</caption>
<thead>
	<tr><th></th><th scope=col>gene_name</th><th scope=col>chr</th><th scope=col>susie_pval</th><th scope=col>susie_z</th><th scope=col>lasso_pval</th><th scope=col>lasso_z</th><th scope=col>enet_pval</th><th scope=col>enet_z</th><th scope=col>mr_ash_pval</th><th scope=col>mr_ash_z</th><th scope=col>susie_qval</th><th scope=col>lasso_qval</th><th scope=col>enet_qval</th><th scope=col>mr_ash_qval</th><th scope=col>susie_pval_adj</th><th scope=col>lasso_pval_adj</th><th scope=col>enet_pval_adj</th><th scope=col>mr_ash_pval_adj</th><th scope=col>gene_id</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>ENSG00000136717</td><td>2 </td><td>5.538175e-80</td><td> 18.938078</td><td>1.201514e-19</td><td>  9.068966</td><td>2.431026e-19</td><td>  8.991850</td><td>8.077433e-88</td><td> 19.865604</td><td>1.107635e-79</td><td>1.602019e-19</td><td>2.431026e-19</td><td>3.230973e-87</td><td>1.866365e-77</td><td>3.688649e-17</td><td>7.536179e-17</td><td>2.722095e-85</td><td>BIN1  </td></tr>
	<tr><th scope=row>2</th><td>ENSG00000073921</td><td>11</td><td>2.887120e-37</td><td>-12.755982</td><td>2.217436e-38</td><td>-12.954442</td><td>5.241860e-38</td><td>-12.888252</td><td>1.746735e-36</td><td>-12.614948</td><td>3.849494e-37</td><td>8.869746e-38</td><td>1.048372e-37</td><td>1.746735e-36</td><td>9.729596e-35</td><td>6.807530e-36</td><td>1.624977e-35</td><td>5.886497e-34</td><td>PICALM</td></tr>
	<tr><th scope=row>3</th><td>ENSG00000130203</td><td>19</td><td>1.020363e-18</td><td>  8.832856</td><td>1.308259e-15</td><td>  7.993815</td><td>3.405317e-15</td><td>  7.875078</td><td>1.871431e-15</td><td>  7.949581</td><td>4.081452e-18</td><td>2.495242e-15</td><td>3.405317e-15</td><td>2.495242e-15</td><td>3.438623e-16</td><td>4.016356e-13</td><td>1.055648e-12</td><td>6.306724e-13</td><td>APOE  </td></tr>
	<tr><th scope=row>4</th><td>ENSG00000087589</td><td>20</td><td>1.173876e-15</td><td> -8.007160</td><td>1.000017e-04</td><td> -3.890588</td><td>3.961193e-03</td><td> -2.881236</td><td>2.222454e-15</td><td> -7.928256</td><td>4.444907e-15</td><td>1.333356e-04</td><td>3.961193e-03</td><td>4.444907e-15</td><td>3.955963e-13</td><td>3.070051e-02</td><td>1.000000e+00</td><td>7.489669e-13</td><td>CASS4 </td></tr>
	<tr><th scope=row>5</th><td>ENSG00000120899</td><td>8 </td><td>2.414907e-14</td><td>  7.626352</td><td>2.427304e-12</td><td>  7.007433</td><td>5.437822e-12</td><td>  6.893651</td><td>2.785009e-14</td><td>  7.607939</td><td>5.570019e-14</td><td>3.236406e-12</td><td>5.437822e-12</td><td>5.570019e-14</td><td>8.138237e-12</td><td>7.451825e-10</td><td>1.685725e-09</td><td>9.385482e-12</td><td>PTK2B </td></tr>
	<tr><th scope=row>6</th><td>ENSG00000130208</td><td>19</td><td>1.862955e-13</td><td>  7.358280</td><td>4.141127e-03</td><td>  2.867206</td><td>1.910014e-03</td><td>  3.103879</td><td>2.157805e-12</td><td>  7.023886</td><td>7.451820e-13</td><td>4.141127e-03</td><td>2.546685e-03</td><td>4.315610e-12</td><td>6.278158e-11</td><td>1.000000e+00</td><td>5.921042e-01</td><td>7.271803e-10</td><td>APOC1 </td></tr>
</tbody>
</table>
</div></div>
</div>
<p>After fine mapping, there are only 337 genes have significant results. Thus, we only list the pvalue and qvalue of 337 genes.</p>
</div>
<div class="section" id="mic-twas-results">
<h4>Mic TWAS results<a class="headerlink" href="#mic-twas-results" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>There are eleven candidate significant genes obtained by four methods: <strong>BIN1</strong>, <strong>RASGEF1C</strong>, <strong>JAZF1</strong>, <strong>PTK2B</strong>, <strong>TP53INP1</strong>, <strong>USP6NL</strong>, <strong>BLNK</strong>, <strong>MS4A4A</strong>, <strong>PICALM</strong>, <strong>RIN3</strong>, <strong>APOE</strong></p></li>
<li><p>union of genes: 43 genes</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">length</span><span class="p">(</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">)</span>
<span class="n">susie</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">$</span><span class="n">gene_id</span>
<span class="n">lasso</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">$</span><span class="n">gene_id</span>
<span class="n">enet</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">$</span><span class="n">gene_id</span>
<span class="n">mr_ash</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">$</span><span class="n">gene_id</span>
<span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="n">susie</span><span class="p">,</span><span class="n">lasso</span><span class="p">),</span><span class="n">enet</span><span class="p">),</span><span class="n">mr_ash</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">susie</span><span class="p">,</span><span class="n">lasso</span><span class="p">,</span><span class="n">enet</span><span class="p">,</span><span class="n">mr_ash</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">40</div><div class="output text_html">14</div><div class="output text_html">13</div><div class="output text_html">34</div><div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'BIN1'</li><li>'RASGEF1C'</li><li>'JAZF1'</li><li>'PTK2B'</li><li>'TP53INP1'</li><li>'USP6NL'</li><li>'BLNK'</li><li>'MS4A4A'</li><li>'PICALM'</li><li>'RIN3'</li><li>'APOE'</li></ol>
</div><div class="output text_html">43</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">VennDiagram</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggvenn</span><span class="p">)</span>
<span class="n">D</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="s">&#39;SuSiE&#39;</span> <span class="o">=</span> <span class="n">susie</span><span class="p">,</span><span class="s">&#39;Lasso&#39;</span> <span class="o">=</span> <span class="n">lasso</span><span class="p">,</span> <span class="s">&#39;Enet&#39;</span> <span class="o">=</span>  <span class="n">enet</span><span class="p">,</span> <span class="s">&#39;MR.ASH&#39;</span> <span class="o">=</span> <span class="n">mr_ash</span><span class="p">)</span>
<span class="nf">ggvenn</span><span class="p">(</span><span class="n">D</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;SuSiE&quot;</span><span class="p">,</span><span class="s">&quot;Lasso&quot;</span><span class="p">,</span><span class="s">&quot;Enet&quot;</span><span class="p">,</span><span class="s">&quot;MR.ASH&quot;</span><span class="p">),</span><span class="n">show_percentage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">fill_color</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="s">&quot;orange&quot;</span><span class="p">,</span><span class="s">&quot;blue&quot;</span><span class="p">,</span><span class="s">&quot;green&quot;</span><span class="p">))</span>
<span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span> <span class="o">=</span> <span class="m">16</span><span class="p">,</span> <span class="n">repr.plot.height</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/ptwas_19_0.png" src="../../_images/ptwas_19_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggnewscale</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span>
<span class="n">min_pval</span> <span class="o">=</span> <span class="nf">apply</span><span class="p">(</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval&quot;</span><span class="p">)],</span><span class="m">1</span><span class="p">,</span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">min</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">na.rm</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
<span class="n">data_all_gene</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="n">gene_id</span><span class="p">,</span><span class="n">susie_pval</span><span class="p">,</span><span class="n">lasso_pval</span><span class="p">,</span><span class="n">enet_pval</span><span class="p">,</span><span class="n">mr_ash_pval</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">min_pval</span> <span class="o">=</span> <span class="n">min_pval</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span><span class="o">=</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">chr</span><span class="p">))</span>
<span class="n">gene_pos</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="s">&quot;~/MR_KMT_analysis/PTWAS/Mic_AD_phenotype_select.txt&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span> <span class="o">=</span> <span class="nf">as.numeric</span><span class="p">(</span><span class="nf">str_sub</span><span class="p">(</span><span class="n">`#chr`</span><span class="p">,</span><span class="m">4</span><span class="p">)))</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">`#chr`</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;start_bp&quot;</span><span class="p">,</span><span class="s">&quot;end_bp&quot;</span><span class="p">,</span><span class="s">&quot;gene_name&quot;</span><span class="p">,</span><span class="s">&quot;chr&quot;</span><span class="p">))</span>
<span class="n">gene_pos_pval</span> <span class="o">=</span> <span class="nf">merge</span><span class="p">(</span><span class="n">data_all_gene</span><span class="p">,</span><span class="n">gene_pos</span><span class="p">,</span><span class="n">by</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;gene_name&quot;</span><span class="p">,</span><span class="s">&quot;chr&quot;</span><span class="p">))</span>
<span class="n">susie_select</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">susie_pval</span><span class="o">&lt;</span><span class="p">(</span><span class="m">2.5</span><span class="o">*</span><span class="m">10</span><span class="o">^</span><span class="p">(</span><span class="m">-6</span><span class="p">)</span><span class="o">/</span><span class="m">4</span><span class="p">)]</span>
<span class="n">lasso_select</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">lasso_pval</span><span class="o">&lt;</span><span class="p">(</span><span class="m">2.5</span><span class="o">*</span><span class="m">10</span><span class="o">^</span><span class="p">(</span><span class="m">-6</span><span class="p">)</span><span class="o">/</span><span class="m">4</span><span class="p">)]</span>
<span class="n">enet_select</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">enet_pval</span><span class="o">&lt;</span><span class="p">(</span><span class="m">2.5</span><span class="o">*</span><span class="m">10</span><span class="o">^</span><span class="p">(</span><span class="m">-6</span><span class="p">)</span><span class="o">/</span><span class="m">4</span><span class="p">)]</span>
<span class="n">ash_select</span> <span class="o">=</span> <span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">mr_ash_pval</span><span class="o">&lt;</span><span class="p">(</span><span class="m">2.5</span><span class="o">*</span><span class="m">10</span><span class="o">^</span><span class="p">(</span><span class="m">-6</span><span class="p">)</span><span class="o">/</span><span class="m">4</span><span class="p">)]</span>
<span class="n">gene_annotate</span> <span class="o">=</span> <span class="nf">unique</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">lasso_select</span><span class="p">,</span><span class="n">enet_select</span><span class="p">,</span><span class="n">ash_select</span><span class="p">))</span>
<span class="c1">#### more than 2 methods have intersection</span>
<span class="n">d1</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">lasso_select</span><span class="p">)</span>
<span class="n">d2</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">enet_select</span><span class="p">)</span>
<span class="n">d3</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">ash_select</span><span class="p">)</span>
<span class="n">d4</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">lasso_select</span><span class="p">,</span><span class="n">enet_select</span><span class="p">)</span>
<span class="n">d5</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">lasso_select</span><span class="p">,</span><span class="n">ash_select</span><span class="p">)</span>
<span class="n">d6</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="n">enet_select</span><span class="p">,</span><span class="n">ash_select</span><span class="p">)</span>
<span class="n">d7</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">lasso_select</span><span class="p">),</span><span class="n">ash_select</span><span class="p">)</span>
<span class="n">d8</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">enet_select</span><span class="p">),</span><span class="n">ash_select</span><span class="p">)</span>
<span class="n">d9</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">lasso_select</span><span class="p">),</span><span class="n">enet_select</span><span class="p">)</span>
<span class="n">d10</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="n">ash_select</span><span class="p">,</span><span class="n">lasso_select</span><span class="p">),</span><span class="n">enet_select</span><span class="p">)</span>
<span class="n">d11</span> <span class="o">=</span> <span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="n">susie_select</span><span class="p">,</span><span class="n">lasso_select</span><span class="p">),</span><span class="n">enet_select</span><span class="p">),</span><span class="n">ash_select</span><span class="p">)</span>
<span class="n">two_more</span> <span class="o">=</span> <span class="nf">unique</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">d1</span><span class="p">,</span><span class="n">d2</span><span class="p">,</span><span class="n">d3</span><span class="p">,</span><span class="n">d4</span><span class="p">,</span><span class="n">d5</span><span class="p">,</span><span class="n">d6</span><span class="p">,</span><span class="n">d7</span><span class="p">,</span><span class="n">d8</span><span class="p">,</span><span class="n">d9</span><span class="p">,</span><span class="n">d10</span><span class="p">,</span><span class="n">d11</span><span class="p">))</span>
<span class="n">don</span> <span class="o">&lt;-</span> <span class="n">gene_pos_pval</span> <span class="o">%&gt;%</span> 
  <span class="c1">#Compute chromosome size</span>
  <span class="nf">group_by</span><span class="p">(</span><span class="n">chr</span><span class="p">)</span> <span class="o">%&gt;%</span> 
  <span class="nf">summarise</span><span class="p">(</span><span class="n">chr_len</span><span class="o">=</span><span class="nf">max</span><span class="p">((</span><span class="n">start_bp</span><span class="o">+</span><span class="n">end_bp</span><span class="p">)</span><span class="o">/</span><span class="m">2</span><span class="p">))</span> <span class="o">%&gt;%</span>  
  
  <span class="c1"># Calculate cumulative position of each chromosome</span>
   <span class="nf">mutate</span><span class="p">(</span><span class="n">tot</span><span class="o">=</span><span class="nf">cumsum</span><span class="p">(</span><span class="n">chr_len</span><span class="p">)</span><span class="o">-</span><span class="n">chr_len</span><span class="p">)</span> <span class="o">%&gt;%</span>
   <span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">chr_len</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="c1"># </span>
  <span class="c1"># # Add this info to the initial dataset</span>
   <span class="nf">left_join</span><span class="p">(</span><span class="n">gene_pos_pval</span><span class="p">,</span> <span class="n">.</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="o">=</span><span class="s">&quot;chr&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
  <span class="c1"># </span>
  <span class="c1"># # Add a cumulative position of each SNP</span>
   <span class="nf">arrange</span><span class="p">(</span><span class="n">chr</span><span class="p">,</span> <span class="p">(</span><span class="n">start_bp</span><span class="o">+</span><span class="n">end_bp</span><span class="p">)</span><span class="o">/</span><span class="m">2</span><span class="p">)</span> <span class="o">%&gt;%</span>
   <span class="nf">mutate</span><span class="p">(</span> <span class="n">BPcum</span><span class="o">=</span><span class="p">(</span><span class="n">start_bp</span><span class="o">+</span><span class="n">end_bp</span><span class="p">)</span><span class="o">/</span><span class="m">2</span><span class="o">+</span><span class="n">tot</span><span class="p">)</span> <span class="o">%&gt;%</span>
   <span class="nf">mutate</span><span class="p">(</span> <span class="n">susie_highlight</span><span class="o">=</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">%in%</span> <span class="n">susie_select</span><span class="p">,</span> <span class="s">&quot;yes&quot;</span><span class="p">,</span> <span class="s">&quot;no&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
   <span class="nf">mutate</span><span class="p">(</span> <span class="n">lasso_highlight</span><span class="o">=</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">%in%</span> <span class="n">lasso_select</span><span class="p">,</span> <span class="s">&quot;yes&quot;</span><span class="p">,</span> <span class="s">&quot;no&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
   <span class="nf">mutate</span><span class="p">(</span> <span class="n">enet_highlight</span><span class="o">=</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">%in%</span> <span class="n">enet_select</span><span class="p">,</span> <span class="s">&quot;yes&quot;</span><span class="p">,</span> <span class="s">&quot;no&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
   <span class="nf">mutate</span><span class="p">(</span> <span class="n">ash_highlight</span><span class="o">=</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">%in%</span> <span class="n">ash_select</span><span class="p">,</span> <span class="s">&quot;yes&quot;</span><span class="p">,</span> <span class="s">&quot;no&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
   <span class="nf">mutate</span><span class="p">(</span> <span class="n">two_highlight</span> <span class="o">=</span> <span class="nf">ifelse</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">%in%</span> <span class="n">two_more</span><span class="p">,</span> <span class="s">&quot;yes&quot;</span><span class="p">,</span> <span class="s">&quot;no&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
   <span class="nf">mutate</span><span class="p">(</span> <span class="n">gene_annotate</span> <span class="o">=</span> <span class="nf">ifelse</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">%in%</span> <span class="n">gene_annotate</span><span class="p">,</span> <span class="s">&quot;yes&quot;</span><span class="p">,</span> <span class="s">&quot;no&quot;</span><span class="p">))</span>  
  
<span class="n">axisdf</span> <span class="o">=</span> <span class="n">don</span> <span class="o">%&gt;%</span> <span class="nf">group_by</span><span class="p">(</span><span class="n">chr</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">summarize</span><span class="p">(</span><span class="n">center</span><span class="o">=</span><span class="p">(</span> <span class="nf">max</span><span class="p">(</span><span class="n">BPcum</span><span class="p">)</span> <span class="o">+</span> <span class="nf">min</span><span class="p">(</span><span class="n">BPcum</span><span class="p">)</span> <span class="p">)</span> <span class="o">/</span> <span class="m">2</span> <span class="p">)</span>
<span class="c1"># ######TWAS-GKF</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">don</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">BPcum</span><span class="p">,</span> <span class="n">y</span><span class="o">=-</span><span class="nf">log10</span><span class="p">(</span><span class="n">min_pval</span><span class="p">)))</span> <span class="o">+</span>
  
  <span class="c1"># Show all points</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="nf">as.factor</span><span class="p">(</span><span class="n">chr</span><span class="p">),</span><span class="n">show.legend</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="m">0.8</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span> <span class="o">=</span> <span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;#b6b6b6&quot;</span><span class="p">,</span><span class="s">&quot;#9f9f9f&quot;</span><span class="p">),</span> <span class="m">22</span><span class="p">),</span><span class="n">guide</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="o">+</span>
  <span class="c1"># custom X axis:</span>
  <span class="nf">scale_x_continuous</span><span class="p">(</span> <span class="n">label</span> <span class="o">=</span> <span class="n">axisdf</span><span class="o">$</span><span class="n">chr</span><span class="p">,</span> <span class="n">breaks</span><span class="o">=</span> <span class="n">axisdf</span><span class="o">$</span><span class="n">center</span> <span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">expand</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">))</span> <span class="o">+</span>  <span class="c1"># remove space between plot area and x axis</span>
  <span class="nf">ylim</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">90</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">new_scale_color</span><span class="p">()</span><span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="nf">subset</span><span class="p">(</span><span class="n">don</span><span class="p">,</span><span class="n">susie_highlight</span><span class="o">==</span><span class="s">&quot;yes&quot;</span><span class="p">),</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;red&quot;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="nf">subset</span><span class="p">(</span><span class="n">don</span><span class="p">,</span><span class="n">lasso_highlight</span><span class="o">==</span><span class="s">&quot;yes&quot;</span><span class="p">),</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;orange&quot;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="nf">subset</span><span class="p">(</span><span class="n">don</span><span class="p">,</span><span class="n">enet_highlight</span><span class="o">==</span><span class="s">&quot;yes&quot;</span><span class="p">),</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;green&quot;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="nf">subset</span><span class="p">(</span><span class="n">don</span><span class="p">,</span><span class="n">ash_highlight</span><span class="o">==</span><span class="s">&quot;yes&quot;</span><span class="p">),</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;blue&quot;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="nf">subset</span><span class="p">(</span><span class="n">don</span><span class="p">,</span><span class="n">two_highlight</span><span class="o">==</span><span class="s">&quot;yes&quot;</span><span class="p">),</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;black&quot;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">geom_label_repel</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="nf">subset</span><span class="p">(</span><span class="n">don</span><span class="p">,</span><span class="n">gene_annotate</span><span class="o">==</span><span class="s">&quot;yes&quot;</span><span class="p">),</span><span class="nf">aes</span><span class="p">(</span><span class="n">label</span> <span class="o">=</span> <span class="n">gene_id</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="n">max.overlaps</span> <span class="o">=</span> <span class="kc">Inf</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="m">0.75</span><span class="p">)</span><span class="o">+</span>
  <span class="c1">#geom_label_repel(data = subset(don,S_annotate==&quot;yes&quot;),aes(label = gene_name),size=12,max.overlaps = Inf,box.padding = 2.5,color = &quot;red&quot;,alpha=0.75)+</span>
  <span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="s">&quot;orange&quot;</span><span class="p">,</span><span class="s">&quot;green&quot;</span><span class="p">,</span><span class="s">&quot;blue&quot;</span><span class="p">,</span><span class="s">&quot;black&quot;</span><span class="p">),</span><span class="n">breaks</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="s">&quot;orange&quot;</span><span class="p">,</span><span class="s">&quot;green&quot;</span><span class="p">,</span><span class="s">&quot;blue&quot;</span><span class="p">,</span><span class="s">&quot;black&quot;</span><span class="p">),</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;Methods&quot;</span><span class="p">,</span><span class="n">labels</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;SuSiE&quot;</span><span class="p">,</span><span class="s">&quot;Lasso&quot;</span><span class="p">,</span><span class="s">&quot;Enet&quot;</span><span class="p">,</span><span class="s">&quot;MR.ASH&quot;</span><span class="p">,</span><span class="s">&quot;&gt;=2 methods&quot;</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Chromosome&quot;</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;-log10(pval)&quot;</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="m">2.5</span><span class="o">*</span><span class="m">10</span><span class="o">^</span><span class="p">(</span><span class="m">-6</span><span class="p">)</span><span class="o">/</span><span class="m">4</span><span class="p">),</span> <span class="n">linetype</span> <span class="o">=</span> <span class="s">&quot;dashed&quot;</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">)</span><span class="o">+</span>
  <span class="nf">theme_bw</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span> 
    <span class="c1">#legend.position=&quot;none&quot;,</span>
    <span class="n">plot.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">45</span><span class="p">),</span>
    <span class="n">legend.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">face</span> <span class="o">=</span> <span class="s">&quot;bold&quot;</span><span class="p">,</span><span class="n">size</span> <span class="o">=</span> <span class="m">45</span><span class="p">),</span>
    <span class="n">legend.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">30</span><span class="p">),</span>
    <span class="n">axis.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">20</span><span class="p">),</span>
    <span class="n">axis.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">45</span><span class="p">),</span>
    <span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;top&quot;</span><span class="p">,</span>
    <span class="n">panel.border</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
    <span class="n">panel.grid.major.x</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
    <span class="n">panel.grid.minor.x</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">()</span>
  <span class="p">)</span>
<span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span> <span class="o">=</span> <span class="m">25</span><span class="p">,</span> <span class="n">repr.plot.height</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
“There was 1 warning in `mutate()`.
<span class=" -Color -Color-Cyan">ℹ</span> In argument: `chr = as.numeric(str_sub(`#chr`, 4))`.
Caused by warning:
<span class=" -Color -Color-Yellow">!</span> NAs introduced by coercion”
Warning message in geom_point(aes(color = as.factor(chr), show.legend = FALSE), :
“Ignoring unknown aesthetics: <span class=" -Color -Color-Green">show.legend</span>”
Scale for <span class=" -Color -Color-Green">y</span> is already present.
Adding another scale for <span class=" -Color -Color-Green">y</span>, which will replace the existing scale.
Warning message:
“The `guide` argument in `scale_*()` cannot be `FALSE`. This was deprecated in
ggplot2 3.3.4.
<span class=" -Color -Color-Cyan">ℹ</span> Please use &quot;none&quot; instead.”
</pre></div>
</div>
<img alt="../../_images/ptwas_20_1.png" src="../../_images/ptwas_20_1.png" />
</div>
</div>
</div>
<div class="section" id="mic-mr-results">
<h4>Mic MR results<a class="headerlink" href="#mic-mr-results" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>intersection: there are eleven causal genes obtained by four methods, <strong>BIN1</strong>, <strong>RASGEF1C</strong>, <strong>JAZF1</strong>, <strong>PTK2B</strong>, <strong>TP53INP1</strong>, <strong>USP6NL</strong>, <strong>BLNK</strong>, <strong>MS4A4A</strong>, <strong>PICALM</strong>, <strong>RIN3</strong>, <strong>APOE</strong></p></li>
<li><p>union of four methods (25 genes): <strong>BIN1</strong>, <strong>INPP5D</strong>, <strong>RHOH</strong>, <strong>RASGEF1C</strong>, <strong>CD2AP</strong>, <strong>JAZF1</strong>, <strong>EPDR1</strong>, <strong>ZYX</strong>, <strong>PTK2B</strong>, <strong>TP53INP1</strong>, <strong>USP6NL</strong>, <strong>CCDC6</strong>, <strong>BLNK</strong>, <strong>PLEKHA1</strong>, <strong>MS4A6A</strong>, <strong>MS4A4A</strong>, <strong>PICALM</strong>, <strong>RIN3</strong>, <strong>LACTB</strong>, <strong>RABEP1</strong>, <strong>APOE</strong>, <strong>APOC1</strong>, <strong>CASS4</strong>, <strong>NCK2</strong>, <strong>HLA-C</strong></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">Mic_mr</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;~/MR_KMT_analysis/PTWAS/Mic/mr_output.rds&quot;</span><span class="p">)</span>
<span class="n">susie_mr</span> <span class="o">=</span> <span class="n">Mic_mr</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_mr</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">$</span><span class="n">meta_qval</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">]</span>
<span class="n">lasso_mr</span> <span class="o">=</span> <span class="n">Mic_mr</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_mr</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">$</span><span class="n">meta_qval</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">]</span>
<span class="n">enet_mr</span> <span class="o">=</span> <span class="n">Mic_mr</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_mr</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">$</span><span class="n">meta_qval</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">]</span>
<span class="n">ash_mr</span> <span class="o">=</span> <span class="n">Mic_mr</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">$</span><span class="n">gene_id</span><span class="p">[</span><span class="n">Mic_mr</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">$</span><span class="n">meta_qval</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">]</span>
<span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="nf">intersect</span><span class="p">(</span><span class="n">susie_mr</span><span class="p">,</span><span class="n">lasso_mr</span><span class="p">),</span><span class="n">enet_mr</span><span class="p">),</span><span class="n">ash_mr</span><span class="p">)</span>
<span class="n">Mic_mr_merge</span> <span class="o">=</span> <span class="nf">unique</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">susie_mr</span><span class="p">,</span><span class="n">lasso_mr</span><span class="p">,</span><span class="n">enet_mr</span><span class="p">,</span><span class="n">ash_mr</span><span class="p">))</span>
<span class="n">Mic_mr_merge</span>
<span class="n">Mic_mr</span><span class="o">$</span><span class="n">susie_adj</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'BIN1'</li><li>'RASGEF1C'</li><li>'JAZF1'</li><li>'PTK2B'</li><li>'TP53INP1'</li><li>'USP6NL'</li><li>'BLNK'</li><li>'MS4A4A'</li><li>'PICALM'</li><li>'RIN3'</li><li>'APOE'</li></ol>
</div><div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'BIN1'</li><li>'INPP5D'</li><li>'RHOH'</li><li>'RASGEF1C'</li><li>'CD2AP'</li><li>'JAZF1'</li><li>'EPDR1'</li><li>'ZYX'</li><li>'PTK2B'</li><li>'TP53INP1'</li><li>'USP6NL'</li><li>'CCDC6'</li><li>'BLNK'</li><li>'PLEKHA1'</li><li>'MS4A6A'</li><li>'MS4A4A'</li><li>'PICALM'</li><li>'RIN3'</li><li>'LACTB'</li><li>'RABEP1'</li><li>'APOE'</li><li>'APOC1'</li><li>'CASS4'</li><li>'NCK2'</li><li>'HLA-C'</li></ol>
</div><div class="output text_html"><table class="dataframe">
<caption>A tibble: 25 × 12</caption>
<thead>
	<tr><th scope=col>X_ID</th><th scope=col>num_CS</th><th scope=col>num_IV</th><th scope=col>cpip</th><th scope=col>gene_id</th><th scope=col>meta_eff</th><th scope=col>se_meta_eff</th><th scope=col>meta_pval</th><th scope=col>meta_qval</th><th scope=col>Q</th><th scope=col>Q_pval</th><th scope=col>I2</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>ENSG00000136717</td><td>1</td><td> 1</td><td>0.996</td><td>BIN1    </td><td> 0.018</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000168918</td><td>1</td><td> 2</td><td>0.596</td><td>INPP5D  </td><td> 0.009</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000168421</td><td>1</td><td> 1</td><td>0.994</td><td>RHOH    </td><td>-0.008</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000146090</td><td>4</td><td>28</td><td>0.899</td><td>RASGEF1C</td><td>-0.004</td><td>0.001</td><td>0.000</td><td>0.000</td><td>16.936</td><td>0.001</td><td>0.823</td></tr>
	<tr><td>ENSG00000198087</td><td>1</td><td>53</td><td>0.878</td><td>CD2AP   </td><td> 0.009</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000153814</td><td>2</td><td> 7</td><td>0.801</td><td>JAZF1   </td><td>-0.002</td><td>0.000</td><td>0.000</td><td>0.000</td><td> 0.009</td><td>0.924</td><td>0.000</td></tr>
	<tr><td>ENSG00000086289</td><td>1</td><td>27</td><td>0.879</td><td>EPDR1   </td><td>-0.007</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000159840</td><td>2</td><td> 6</td><td>0.956</td><td>ZYX     </td><td>-0.008</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 8.833</td><td>0.003</td><td>0.887</td></tr>
	<tr><td>ENSG00000120899</td><td>1</td><td> 5</td><td>0.999</td><td>PTK2B   </td><td> 0.005</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000164938</td><td>1</td><td>20</td><td>0.806</td><td>TP53INP1</td><td>-0.005</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000148429</td><td>1</td><td> 1</td><td>0.980</td><td>USP6NL  </td><td> 0.004</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000108091</td><td>1</td><td>15</td><td>0.883</td><td>CCDC6   </td><td>-0.005</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000095585</td><td>1</td><td>22</td><td>0.716</td><td>BLNK    </td><td> 0.013</td><td>0.003</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000107679</td><td>1</td><td>23</td><td>0.876</td><td>PLEKHA1 </td><td> 0.006</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000110077</td><td>1</td><td>14</td><td>0.768</td><td>MS4A6A  </td><td> 0.007</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000110079</td><td>1</td><td>94</td><td>0.817</td><td>MS4A4A  </td><td> 0.006</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000073921</td><td>2</td><td> 9</td><td>0.900</td><td>PICALM  </td><td>-0.008</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.137</td><td>0.711</td><td>0.000</td></tr>
	<tr><td>ENSG00000100599</td><td>4</td><td> 9</td><td>0.962</td><td>RIN3    </td><td>-0.002</td><td>0.001</td><td>0.000</td><td>0.000</td><td>34.460</td><td>0.000</td><td>0.913</td></tr>
	<tr><td>ENSG00000103642</td><td>1</td><td>10</td><td>0.605</td><td>LACTB   </td><td> 0.005</td><td>0.001</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000029725</td><td>1</td><td> 9</td><td>0.793</td><td>RABEP1  </td><td> 0.009</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000130203</td><td>1</td><td>37</td><td>0.792</td><td>APOE    </td><td> 0.010</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000130208</td><td>1</td><td> 8</td><td>0.679</td><td>APOC1   </td><td> 0.008</td><td>0.002</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000087589</td><td>1</td><td> 3</td><td>0.585</td><td>CASS4   </td><td>-0.015</td><td>0.003</td><td>0.000</td><td>0.000</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000071051</td><td>1</td><td> 1</td><td>0.991</td><td>NCK2    </td><td>-0.034</td><td>0.010</td><td>0.001</td><td>0.001</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000204525</td><td>1</td><td> 7</td><td>0.766</td><td>HLA-C   </td><td> 0.019</td><td>0.006</td><td>0.001</td><td>0.001</td><td> 0.000</td><td>1.000</td><td>0.000</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">VennDiagram</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggvenn</span><span class="p">)</span>
<span class="n">D</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="s">&#39;SuSiE_MR&#39;</span> <span class="o">=</span> <span class="n">susie_mr</span><span class="p">,</span><span class="s">&#39;Lasso_MR&#39;</span> <span class="o">=</span> <span class="n">lasso_mr</span><span class="p">,</span> <span class="s">&#39;Enet_MR&#39;</span> <span class="o">=</span>  <span class="n">enet_mr</span><span class="p">,</span> <span class="s">&#39;MR.ASH_MR&#39;</span> <span class="o">=</span> <span class="n">ash_mr</span><span class="p">)</span>
<span class="nf">ggvenn</span><span class="p">(</span><span class="n">D</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;SuSiE_MR&quot;</span><span class="p">,</span><span class="s">&quot;Lasso_MR&quot;</span><span class="p">,</span><span class="s">&quot;Enet_MR&quot;</span><span class="p">,</span><span class="s">&quot;MR.ASH_MR&quot;</span><span class="p">),</span><span class="n">show_percentage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">fill_color</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="s">&quot;orange&quot;</span><span class="p">,</span><span class="s">&quot;blue&quot;</span><span class="p">,</span><span class="s">&quot;green&quot;</span><span class="p">))</span>
<span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span> <span class="o">=</span> <span class="m">16</span><span class="p">,</span> <span class="n">repr.plot.height</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/ptwas_23_0.png" src="../../_images/ptwas_23_0.png" />
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./code/pecotmr_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="pecotmr_preprocessing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">xQTL and GWAS region alignment</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="ptwas_mr.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Mendelian Randomization using fine-mapping IV</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>