
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Association result processing &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="SuSIE results post process" href="fine_mapping_post_processing.html" />
    <link rel="prev" title="Summary statistics in VCF format" href="rds_to_vcf.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../reference_data/reference_data.html">
   Reference Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/reference_data_preparation.html">
     Reference Data Standardization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/generalized_TADB.html">
     Generation of Topologically Associated Domains and their Boundaries
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/ld_prune_reference.html">
     Independent list of variants using LD clumping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene Coordinate Annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype Data Imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden Factor Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate Data Formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pecotmr_integration/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/fine_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_integration/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/post_processing/association_scan_post_processing.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/post_processing/association_scan_post_processing.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/post_processing/association_scan_post_processing.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="association-result-processing">
<h1>Association result processing<a class="headerlink" href="#association-result-processing" title="Permalink to this headline">¶</a></h1>
<p>This notebook analysis the result of association scan based on a merged vcf file.
see <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/issues/215">https://github.com/cumc/xqtl-pipeline/issues/215</a> for detail
Input:</p>
<ul class="simple">
<li><p>list of vcf</p></li>
</ul>
<p>Output:</p>
<ul class="simple">
<li><p>list of vcf with adjusted pvalue</p></li>
<li><p>a table summerized the number of significant finding, after multiple testing correction</p></li>
<li><p>a table of adjusted, best norminal pvalue for each gene.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">assoc_result_processing</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">genome</span> \
    <span class="o">--</span><span class="n">vcf</span>  <span class="err">`</span><span class="n">ls</span> <span class="n">output</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/*</span><span class="n">merged</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="err">`</span> \
    <span class="o">--</span><span class="n">padjust</span><span class="o">-</span><span class="n">method</span> <span class="s2">&quot;bonferroni&quot;</span>  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">assoc_result_processing</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">genome</span> \
    <span class="o">--</span><span class="n">vcf</span>  <span class="err">`</span><span class="n">ls</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">ROSMAP_methy_QTL</span><span class="o">/</span><span class="n">data_intergration</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">/</span><span class="n">methyl_QTL</span><span class="o">.*/*.</span><span class="n">vcf</span><span class="err">`</span> \
    <span class="o">--</span><span class="n">padjust</span><span class="o">-</span><span class="n">method</span> <span class="s2">&quot;bonferroni&quot;</span>  <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># list of vcf to be analysised, best to be for all chromosome by `ls *.merged.vcf.gz` in the output folder of sumstat merger</span>
<span class="kn">parameter:</span> <span class="n">vcf</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">20</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">per_chrom</span><span class="p">]</span>
<span class="c1"># Way to do gene level multiple correction for each snps, bonferroni or fdr</span>
<span class="kn">parameter:</span> <span class="n">padjust_method</span> <span class="o">=</span> <span class="s2">&quot;fdr&quot;</span>
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;vcf&quot;</span>
<span class="kn">output:</span> <span class="n">rds</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_vcf</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.rds&#39;</span><span class="p">,</span> <span class="n">processed_vcf</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_vcf</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.p_adjusted.vcf.bgz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
   <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
   <span class="c1"># Read VCF</span>
   <span class="n">data</span> <span class="o">=</span> <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">readVcf</span><span class="p">(</span><span class="s2">&quot;$[_vcf]&quot;</span><span class="p">)</span>
   <span class="c1"># Extract p-val</span>
   <span class="n">pval</span> <span class="o">=</span> <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">(</span><span class="n">rownames</span> <span class="o">=</span> <span class="s2">&quot;ID&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">GENE</span> <span class="o">=</span> <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">info</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">GENE</span><span class="p">,</span> <span class="n">ID</span> <span class="o">=</span> <span class="n">sub</span><span class="p">(</span><span class="s2">&quot;[^:]*:&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">ID</span><span class="p">)</span> <span class="p">)</span>
   <span class="c1"># Adjust pvalue per gene based on fdr or beferroni</span>
   <span class="n">pval_adj</span> <span class="o">=</span> <span class="n">pval</span><span class="o">%&gt;%</span><span class="n">group_by</span><span class="p">(</span><span class="n">GENE</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">across</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P</span><span class="p">),</span><span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span><span class="n">p</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="s2">&quot;$[padjust_method]&quot;</span><span class="p">)))</span>
   <span class="c1">## Get the number of sig association</span>
   <span class="n">n_asso</span> <span class="o">=</span> <span class="p">(</span><span class="n">pval_adj</span><span class="o">%&gt;%</span><span class="n">ungroup</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">ID</span><span class="p">,</span><span class="o">-</span><span class="n">GENE</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">colSums</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">t</span><span class="o">%&gt;%</span><span class="n">as_tibble</span>
   <span class="c1">## Get the number of sig snps</span>
   <span class="n">n_snp</span> <span class="o">=</span> <span class="p">(</span><span class="n">pval_adj</span><span class="o">%&gt;%</span><span class="n">ungroup</span><span class="o">%&gt;%</span><span class="n">group_by</span><span class="p">(</span><span class="n">ID</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">summarize</span><span class="p">(</span><span class="n">across</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P</span><span class="p">),</span><span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span><span class="n">coalesce</span><span class="p">(</span><span class="mi">9</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nb">min</span><span class="p">))</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">ID</span><span class="p">)</span><span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">colSums</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">t</span><span class="o">%&gt;%</span><span class="n">as_tibble</span>
   <span class="c1">## Get the top  snps for each genes</span>
   <span class="n">pval_adj_gene</span> <span class="o">=</span> <span class="n">pval_adj</span><span class="o">%&gt;%</span><span class="n">group_by</span><span class="p">(</span><span class="n">GENE</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">summarize</span><span class="p">(</span><span class="n">across</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P</span><span class="p">),</span><span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span><span class="nb">min</span><span class="p">))</span>
   <span class="n">pval_adj_gene_snps</span> <span class="o">=</span> <span class="n">pval_adj</span><span class="o">%&gt;%</span><span class="n">group_by</span><span class="p">(</span><span class="n">GENE</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">summarize</span><span class="p">(</span><span class="n">across</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P</span><span class="p">),</span><span class="o">~</span><span class="nb">list</span><span class="p">(</span><span class="n">pval_adj</span><span class="p">[(</span><span class="n">which</span><span class="p">(</span><span class="o">.</span><span class="n">x</span> <span class="o">==</span> <span class="nb">min</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">))),</span><span class="mi">1</span><span class="p">])))</span>
   <span class="n">output</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s2">&quot;n_asso&quot;</span> <span class="o">=</span> <span class="n">n_asso</span><span class="p">,</span><span class="s2">&quot;n_snp&quot;</span> <span class="o">=</span> <span class="n">n_snp</span><span class="p">,</span><span class="s2">&quot;pval_adj_gene&quot;</span> <span class="o">=</span> <span class="n">pval_adj_gene</span><span class="p">,</span><span class="s2">&quot;pval_adj_gene_snps&quot;</span> <span class="o">=</span> <span class="n">pval_adj_gene_snps</span> <span class="p">)</span>
   <span class="n">saveRDS</span><span class="p">(</span><span class="n">output</span><span class="p">,</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">)</span>
   <span class="c1"># Edit the VCF and output it</span>
   <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">header</span><span class="p">(</span><span class="n">data</span><span class="p">))[</span><span class="s2">&quot;P_ADJ&quot;</span><span class="p">,]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span><span class="s2">&quot;Float&quot;</span><span class="p">,</span><span class="s2">&quot;P value adjusted by $[padjust_method] among each genes&quot;</span><span class="p">)</span>
   <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P_ADJ</span> <span class="o">=</span> <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P</span>
   <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">geno</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="err">$</span><span class="n">P_ADJ</span> <span class="o">=</span> <span class="n">pval_adj</span><span class="o">%&gt;%</span><span class="n">ungroup</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">ID</span><span class="p">,</span><span class="o">-</span><span class="n">GENE</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span>
   <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">meta</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">header</span><span class="p">(</span><span class="n">data</span><span class="p">))</span><span class="err">$</span><span class="n">n_sig_asso</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">n_asso</span><span class="p">,</span><span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">)</span>
   <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">meta</span><span class="p">(</span><span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">header</span><span class="p">(</span><span class="n">data</span><span class="p">))</span><span class="err">$</span><span class="n">n_sig_snp</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">n_snp</span><span class="p">,</span><span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">)</span>
   <span class="n">VariantAnnotation</span><span class="p">::</span><span class="n">writeVcf</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="s2">&quot;$[_output[1]:n]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">genome</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;per_chrom&quot;</span><span class="p">)[</span><span class="s2">&quot;rds&quot;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.n_sig.txt&#39;</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.gene_p_adj.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
   <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
   <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
   <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tidyr&quot;</span><span class="p">)</span>
   <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
   <span class="c1"># Read data</span>
   <span class="n">input_ls</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,])</span>
   <span class="n">data</span> <span class="o">=</span> <span class="n">tibble</span><span class="p">(</span><span class="n">input_ls</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">rds</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">input_ls</span><span class="p">,</span><span class="o">~</span><span class="n">read_rds</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)),</span>
                                    <span class="n">n_assoc</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">rds</span><span class="p">,</span><span class="o">~.</span><span class="n">x</span><span class="err">$</span><span class="n">n_asso</span><span class="p">),</span>
                                    <span class="n">n_snp</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">rds</span><span class="p">,</span><span class="o">~.</span><span class="n">x</span><span class="err">$</span><span class="n">n_snp</span><span class="p">),</span>
                                    <span class="n">pval_adj_gene</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">rds</span><span class="p">,</span><span class="o">~.</span><span class="n">x</span><span class="err">$</span><span class="n">pval_adj_gene</span><span class="p">)</span>
                                      <span class="p">)</span>
   <span class="c1"># Total number of significant association, snps, and genes</span>
   <span class="n">n_assoc</span> <span class="o">=</span> <span class="n">data</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">n_assoc</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">unnest</span><span class="p">(</span><span class="s2">&quot;n_assoc&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">colSums</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
   <span class="n">n_snp</span> <span class="o">=</span> <span class="n">data</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">n_snp</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">unnest</span><span class="p">(</span><span class="s2">&quot;n_snp&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">colSums</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
   <span class="n">pval_adj_gene_adj</span> <span class="o">=</span> <span class="n">data</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">pval_adj_gene</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">unnest</span><span class="p">(</span><span class="s2">&quot;pval_adj_gene&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">across</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">n_assoc</span><span class="p">),</span><span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span><span class="n">p</span><span class="o">.</span><span class="n">adjust</span><span class="p">(</span><span class="s2">&quot;fdr&quot;</span><span class="p">)</span> <span class="p">))</span>
   <span class="n">n_gene</span> <span class="o">=</span> <span class="p">(</span><span class="n">pval_adj_gene_adj</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">GENE</span><span class="p">)</span><span class="o">&lt;</span> <span class="mf">0.05</span> <span class="p">)</span><span class="o">%&gt;%</span><span class="n">colSums</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
   <span class="n">n_sum</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">n_assoc</span><span class="p">,</span><span class="n">n_snp</span><span class="p">,</span><span class="n">n_gene</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">(</span><span class="n">rownames</span> <span class="o">=</span> <span class="s2">&quot;tissue&quot;</span><span class="p">)</span>
   <span class="n">n_sum</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
   <span class="n">pval_adj_gene_adj</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/post_processing"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="rds_to_vcf.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Summary statistics in VCF format</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="fine_mapping_post_processing.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">SuSIE results post process</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>