
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Gene Coordinate Annotation &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/phenotype/gene_annotation';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Phenotype data imputation" href="phenotype_imputation.html" />
    <link rel="prev" title="Phenotype data preprocessing" href="../phenotype_preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../phenotype_preprocessing.html">Phenotype data preprocessing</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../mnm_analysis/mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/data_preprocessing/phenotype/gene_annotation.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/phenotype/gene_annotation.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/data_preprocessing/phenotype/gene_annotation.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Gene Coordinate Annotation</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alternative-implementation">Alternative implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-cooridnate-annotation">i. Cooridnate Annotation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-gene-expression">a. Gene Expression</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#b-proteins">b. Proteins</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotate-coord">Annotate coord</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Input:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Output:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-pyqtl">Implementation using <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-biomart">Implementation using biomaRt</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-of-leafcutter-isoform">Annotation of leafcutter isoform</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-of-psichomics-output">Processing of psichomics output</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="gene-coordinate-annotation">
<h1>Gene Coordinate Annotation<a class="headerlink" href="#gene-coordinate-annotation" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>We use a gene coordinate annotation pipeline based on <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/src/eqtl_prepare_expression.py"><code class="docutils literal notranslate"><span class="pre">pyqtl</span></code>, as demonstrated here</a>. This adds genomic coordinate annotations to gene-level molecular phenotype files generated in <code class="docutils literal notranslate"><span class="pre">gct</span></code> format and converts them to <code class="docutils literal notranslate"><span class="pre">bed</span></code> format for downstreams analysis.</p>
<section id="alternative-implementation">
<h3>Alternative implementation<a class="headerlink" href="#alternative-implementation" title="Link to this heading">#</a></h3>
<p>Previously we use <code class="docutils literal notranslate"><span class="pre">biomaRt</span></code> package in R instead of code from <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code>. The core function calls are:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="w">    </span><span class="n">ensembl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">useEnsembl</span><span class="p">(</span><span class="n">biomart</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;ensembl&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dataset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">version</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;$[ensembl_version]&quot;</span><span class="p">)</span>
<span class="w">    </span><span class="n">ensembl_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;start_position&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;end_position&quot;</span><span class="p">),</span><span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
</pre></div>
</div>
<p>We require ENSEMBL version to be specified explicitly in this pipeline. As of 2021 for the Brain xQTL project, we use ENSEMBL version 103.</p>
</section>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Molecular phenotype data in <code class="docutils literal notranslate"><span class="pre">gct</span></code> format, with the first column being ENSEMBL ID and other columns being sample names.</p></li>
<li><p>GTF for collapsed gene model</p>
<ul class="simple">
<li><p>the gene names must be consistent with the molecular phenotype data matrices (eg ENSG00000000003 vs. ENSG00000000003.1 will not work)</p></li>
</ul>
</li>
<li><p>(Optional) Meta-data to match between sample names in expression data and genotype files</p>
<ul class="simple">
<li><p>Tab delimited with header</p></li>
<li><p>Only 2 columns: first column is sample name in expression data, 2nd column is sample name in genotype data</p></li>
<li><p><strong>must contains all the sample name in expression matrices even if they don’t existing in genotype data</strong></p></li>
</ul>
</li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<p>Molecular phenotype data in <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<p>The MWE is uploaded to <a class="reference external" href="https://www.synapse.org/#!Synapse:syn36416559/files/">Synapse</a></p>
<section id="i-cooridnate-annotation">
<h3>i. Cooridnate Annotation<a class="headerlink" href="#i-cooridnate-annotation" title="Link to this heading">#</a></h3>
<section id="a-gene-expression">
<h4>a. Gene Expression<a class="headerlink" href="#a-gene-expression" title="Link to this heading">#</a></h4>
<p>Timing: &lt;1 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run gene_annotation.ipynb annotate_coord \
    --cwd ../../../output_test/phenotype_by_chrom \
    --phenoFile ../../../mwe_data/xQTL_discovery/MWE.log2cpm.mol_phe.bed.gz \
    --coordinate-annotation ../../../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf \
    --phenotype-id-column gene_id
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">annotate_coord_gene</span>: 
INFO: tc987661f8bb5f0b1 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 3690363 (&quot;job_tc987661f8bb5f0b1&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">annotate_coord_gene</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/output_test/phenotype_by_chrom/MWE.log2cpm.mol_phe.bed.bed.gz /restricted/projectnb/xqtl/xqtl_protocol/output_test/phenotype_by_chrom/MWE.log2cpm.mol_phe.bed.region_list.txt</span>
INFO: Workflow annotate_coord_gene (ID=w90bd1b9b7fb38fca) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
<section id="b-proteins">
<h4>b. Proteins<a class="headerlink" href="#b-proteins" title="Link to this heading">#</a></h4>
<p>Timing: &lt;1 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run gene_annotation.ipynb annotate_coord \
    --cwd ../../../output_test/phenotype_by_chrom \
    --phenoFile ../../../mwe_data/protocol_data/input/xqtl_association/protocol_example.protein.csv \
    --coordinate-annotation ../../../reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.ERCC.gtf \
    --sample-participant-lookup ../../../mwe_data/protocol_data/output/protocol_example.protein.sample_overlap.txt \
    --phenotype-id-column gene_name \
    --molecular-trait-type protein \
    --sep &quot;,&quot;
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">annotate_coord_protein</span>: 
INFO: tbecf36ea3e901ea1 <span class=" -Color -Color-Green">re-execute completed</span>
INFO: tbecf36ea3e901ea1 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 3961996 (&quot;job_tbecf36ea3e901ea1&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">annotate_coord_protein</span> output:   <span class=" -Color -Color-Green">/restricted/projectnb/xqtl/xqtl_protocol/output_test/phenotype_by_chrom/protocol_example.protein.bed.gz /restricted/projectnb/xqtl/xqtl_protocol/output_test/phenotype_by_chrom/protocol_example.protein.region_list.txt</span>
INFO: Workflow annotate_coord_protein (ID=w75d21004c6718efd) is executed successfully with 1 completed step and 1 completed task.
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run gene_annotation.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run gene_annotation.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  region_list_generation
  annotate_coord
  annotate_coord_biomart
  map_leafcutter_cluster_to_gene
  annotate_leafcutter_isoforms
  annotate_psichomics_isoforms

Global Workflow Options:
  --cwd output (as path)
                        Work directory &amp; output directory
  --phenoFile VAL (as path, required)
                        Molecular phenotype matrix
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 1 (as int)
                        Number of threads
  --container &#39;&#39;
  --entrypoint &#39;&#39;

Sections
  region_list_generation:
  annotate_coord:
    Workflow Options:
      --sample-participant-lookup . (as path)
                        A file to map sample ID from expression to genotype,
                        must contain two columns, sample_id and participant_id,
                        mapping IDs in the expression files to IDs in the
                        genotype (these can be the same).
      --molecular-trait-type gene
                        Options: gene, protein, atac
      --coordinate-annotation VAL (as path, required)
                        gtf annotation for RNA-seq data; other types of
                        annotation index for protein and atac-seq
      --phenotype-id-column &#39;gene_id&#39;
                        gene_id or gene_name for RNA-seq data, SOMAseqID for
                        proteins for example
      --auxiliary-id-mapping . (as path)
                        Optional mapping file (protein_name_index for protein,
                        etc.)
      --[no-]strip-id (default to False)
                        Optional processing flag
  annotate_coord_biomart:
    Workflow Options:
      --ensembl-version VAL (as int, required)
  map_leafcutter_cluster_to_gene:
    Workflow Options:
      --intron-count VAL (as path, required)
                        Extract the code in case psichromatic needs to be
                        processed the same way PheoFile in this step is the
                        intron_count file
      --map-stra site
                        Defines the mapping strategy options: &#39;site&#39; or
                        &#39;region&#39;, with &#39;site&#39; as the default. The &#39;site&#39;
                        strategy maps introns to the start and end of each exon.
                        The &#39;region&#39; strategy, to be recommended in leafcutter2,
                        maps each intron based on it overlapping more than
                        overlap_ratio of a gene&#39;s region.
      --overlap-ratio 0.8 (as float)
                        Define the overlap ratio as the proportion of the
                        cluster length that intersects with a gene, used to
                        determine mapping to the gene.
  annotate_leafcutter_isoforms:
    Workflow Options:
      --sample-participant-lookup . (as path)
  annotate_psichomics_isoforms:
    Workflow Options:
      --sample-participant-lookup . (as path)
</pre></div>
</div>
</div>
</div>
</section>
<section id="setup-and-global-parameters">
<h2>Setup and global parameters<a class="headerlink" href="#setup-and-global-parameters" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# Work directory &amp; output directory
parameter: cwd = path(&quot;output&quot;)
# Molecular phenotype matrix
parameter: phenoFile = path
# For cluster jobs, number commands to run per job
parameter: job_size = 1
# Wall clock time expected
parameter: walltime = &quot;5h&quot;
# Memory expected
parameter: mem = &quot;16G&quot;
# Number of threads
parameter: numThreads = 1
parameter: container = &quot;&quot;
parameter: entrypoint= &quot;&quot;
</pre></div>
</div>
</div>
</div>
</section>
<section id="annotate-coord">
<h2>Annotate coord<a class="headerlink" href="#annotate-coord" title="Link to this heading">#</a></h2>
<section id="id1">
<h3>Input:<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>A gtf file used to generated the bed</p></li>
<li><p>A phenotype bed file, must have a gene_id column indicating the name of genes.</p></li>
</ol>
</section>
<section id="id2">
<h3>Output:<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>annotated phenotype bed file</p></li>
<li><p>Optional: a tsv file only has the first several columns indicating phenotype_id and names. For RNA-seq, it contains five columns: chr,start,end,gene_id,gene_name</p></li>
</ol>
</section>
</section>
<section id="implementation-using-pyqtl">
<h2>Implementation using <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code><a class="headerlink" href="#implementation-using-pyqtl" title="Link to this heading">#</a></h2>
<p>Implementation based on <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/src/eqtl_prepare_expression.py">GTEx pipeline</a>.</p>
<p>Following step serves to annotate coord for gene expression file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_coord]
# A file to map sample ID from expression to genotype, must contain two columns, sample_id and participant_id, mapping IDs in the expression files to IDs in the genotype (these can be the same).
parameter: sample_participant_lookup = path()
# Options: gene, protein, atac
parameter: molecular_trait_type = &quot;gene&quot; 
# gtf annotation for RNA-seq data; other types of annotation index for protein and atac-seq
parameter: coordinate_annotation = path
# gene_id or gene_name for RNA-seq data, SOMAseqID for proteins for example
parameter: phenotype_id_column = &quot;gene_id&quot;
# Optional mapping file (protein_name_index for protein, etc.)
parameter: auxiliary_id_mapping = path()
# Optional processing flag
parameter: strip_id = False
parameter: sep = &quot;\t&quot;

input: phenoFile, coordinate_annotation
output: f&#39;{cwd:a}/{_input[0]:bn}.bed.gz&#39;, f&#39;{cwd:a}/{_input[0]:bn}.region_list.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;

    import pandas as pd
    import numpy as np
    import qtl.io
    from pathlib import Path
    from collections import defaultdict
    import warnings

    # Function to convert gtf to bed
    def gtf_to_bed(annotation_gtf, feature=&#39;gene&#39;, exclude_chrs=[], phenotype_id=&#39;gene_id&#39;):
        &quot;&quot;&quot;
        Parse genes from GTF and return DataFrame with min start and max end positions.
        For each gene, uses the smallest position as start and largest as end position.
        &quot;&quot;&quot;
        
        # Data collections
        gene_data = defaultdict(lambda: {&#39;chr&#39;: &#39;&#39;, &#39;start&#39;: float(&#39;inf&#39;), &#39;end&#39;: 0, &#39;strand&#39;: &#39;&#39;})
        gene_names = {}  # To store gene_name for each gene_id
        
        if annotation_gtf.endswith(&#39;.gz&#39;):
            opener = gzip.open(annotation_gtf, &#39;rt&#39;)
        else:
            opener = open(annotation_gtf, &#39;r&#39;)
            
        with opener as gtf:
            for row in gtf:
                row = row.strip().split(&#39;\t&#39;)
                if row[0][0] == &#39;#&#39; or row[2] != feature: 
                    continue  # Skip header or non-matching features
                
                # Parse attributes
                attributes = defaultdict()
                for a in row[8].replace(&#39;&quot;&#39;, &#39;&#39;).split(&#39;;&#39;)[:-1]:
                    kv = a.strip().split(&#39; &#39;)
                    if kv[0] != &#39;tag&#39;:
                        attributes[kv[0]] = kv[1]
                    else:
                        attributes.setdefault(&#39;tags&#39;, []).append(kv[1])
                
                # Get gene identifiers
                curr_gene_id = attributes[&#39;gene_id&#39;]
                curr_gene_name = attributes[&#39;gene_name&#39;]
                gene_names[curr_gene_id] = curr_gene_name
                
                # Update gene data
                data = gene_data[curr_gene_id]
                if data[&#39;chr&#39;] == &#39;&#39;:  # First entry for this gene
                    data[&#39;chr&#39;] = row[0]
                    data[&#39;strand&#39;] = row[6]
                
                # Update min start and max end
                start_pos = int(row[3]) - 1  # Convert to 0-based for BED
                end_pos = int(row[4])
                data[&#39;start&#39;] = min(data[&#39;start&#39;], start_pos)
                data[&#39;end&#39;] = max(data[&#39;end&#39;], end_pos)
        
        # Convert to dataframe format
        chrom, start, end, ids, names = [], [], [], [], []
        for gene_id, data in gene_data.items():
            chrom.append(data[&#39;chr&#39;])
            start.append(data[&#39;start&#39;])
            end.append(data[&#39;end&#39;])
            ids.append(gene_id)
            names.append(gene_names[gene_id])
        
        # Create DataFrame based on phenotype_id
        if phenotype_id == &#39;gene_id&#39;:
            bed_df = pd.DataFrame({
                &#39;chr&#39;: chrom,
                &#39;start&#39;: start,
                &#39;end&#39;: end,
                &#39;gene_id&#39;: ids
            }, columns=[&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;, &#39;gene_id&#39;], index=ids)
        elif phenotype_id == &#39;gene_name&#39;:
            bed_df = pd.DataFrame({
                &#39;chr&#39;: chrom,
                &#39;start&#39;: start,
                &#39;end&#39;: end,
                &#39;gene_id&#39;: names
            }, columns=[&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;, &#39;gene_id&#39;], index=names)
        
        # Filter out excluded chromosomes
        mask = np.ones(len(chrom), dtype=bool)
        for k in exclude_chrs:
            mask = mask &amp; (bed_df[&#39;chr&#39;] != k)
        bed_df = bed_df[mask]
        
        # Sort by chromosome and start position
        bed_df = bed_df.groupby(&#39;chr&#39;, sort=False, group_keys=False).apply(lambda x: x.sort_values([&#39;start&#39;, &#39;end&#39;]))
        
        return bed_df

    def prepare_bed(df, bed_template_df, chr_subset=None):
        bed_df = pd.merge(bed_template_df, df, left_index=True, right_index=True)
        bed_df = bed_df.groupby(&#39;#chr&#39;, sort=False, group_keys=False).apply(lambda x: x.sort_values(&#39;start&#39;))
        if chr_subset is not None:
            bed_df = bed_df[bed_df.chr.isin(chr_subset)]
        return bed_df

    def load_and_preprocess_data(input_path, drop_columns):
        df = pd.read_csv(input_path, sep=&quot;${sep}&quot;, skiprows=0)
        dc = [col for col in df.columns if col in drop_columns] # Take intersection between df.columns and drop_columns
        df = df.drop(dc, axis=1) # drop the intersect
        if len(df.columns) &lt; 2:
            raise ValueError(&quot;There are too few columns in the loaded dataframe, please check the delimiter of the input file. The default delimiter is tab&quot;)
        return df

    def rename_samples_using_lookup(df, lookup_path):
        lookup_file = Path(lookup_path)
        
        if not lookup_file.is_file():
            return df
            
        if lookup_file.suffix == &#39;.csv&#39;:
            sep = &#39;,&#39;
        elif lookup_file.suffix == &#39;.tsv&#39;:
            sep = &#39;\t&#39;
        else:
            sep = &#39;\t&#39;  # Default to tab if unknown extension
            
        try:
            sample_participant_lookup = pd.read_csv(lookup_file, sep=sep, dtype={0:str, 1:str})
            if &quot;genotype_id&quot; in sample_participant_lookup.columns:
                sample_participant_lookup = sample_participant_lookup.set_index(sample_participant_lookup.columns[1])
                df.rename(columns=sample_participant_lookup.to_dict()[&quot;genotype_id&quot;], inplace=True)
            else:
                # Fall back to the no header style
                sample_participant_lookup = pd.read_csv(lookup_file, sep=sep, header=None, dtype={0:str, 1:str})
                rename_dict = dict(zip(sample_participant_lookup[0], sample_participant_lookup[1]))
                df.rename(columns=rename_dict, inplace=True)
        except Exception as e:
            print(f&quot;Warning: Error processing sample lookup file: {e}&quot;)
        return df

    def load_bed_template_from_gtf(input_path, phenotype_id_column):
        if sum(gtf_to_bed(input_path, feature=&#39;gene&#39;, phenotype_id=&quot;gene_id&quot;).index.duplicated()) &gt; 0:
            raise ValueError(f&quot;gtf file {input_path} needs to be collapsed into gene model by reference data processing module&quot;)

        bed_template_df_id = gtf_to_bed(input_path, feature=&#39;transcript&#39;, phenotype_id=&quot;gene_id&quot;)
        bed_template_df_name = gtf_to_bed(input_path, feature=&#39;transcript&#39;, phenotype_id=&quot;gene_name&quot;)
        bed_template_df = bed_template_df_id.merge(bed_template_df_name, on=[&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;])
        bed_template_df.columns = [&quot;#chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;gene_id&quot;, &quot;gene_name&quot;]
        bed_template_df = bed_template_df.set_index(phenotype_id_column, drop=False)
        return bed_template_df

    drop_cols = [&quot;#chr&quot;, &quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;stop&quot;, &quot;annot.seqnames&quot;, &quot;annot.start&quot;, &quot;annot.end&quot;]
    df = load_and_preprocess_data(${_input[0]:ar}, drop_cols)
    molecular_trait_type = &quot;${molecular_trait_type}&quot;
    df.set_index(df.columns[0], inplace=True)
    if ${True if sample_participant_lookup.is_file() else False}:
        df = rename_samples_using_lookup(df, &quot;${sample_participant_lookup:a}&quot;)    
    # First column is always the phenotype ID column
    phenotype_id = df.columns.values[0]
    
    # Type-specific processing
    if molecular_trait_type == &quot;gene&quot;:
        if ${strip_id}:
            # Update the index by stripping the pattern &quot;A | B&quot; to &quot;B&quot;
            df.index = df.index.map(lambda x: x.split(&#39;|&#39;)[-1].strip() if &#39;|&#39; in x else x)
            
        bed_template_df = load_bed_template_from_gtf(${_input[1]:ar}, &quot;${phenotype_id_column}&quot;)
        bed_df = prepare_bed(df, bed_template_df).drop(&quot;gene_name&quot;, axis=1)
        bed_df = bed_df.drop_duplicates(&quot;gene_id&quot;, keep=False)
        bed_df = bed_df.rename(columns={&#39;gene_id&#39;: &#39;ID&#39;})
        
    elif molecular_trait_type == &quot;protein&quot;:
        auxiliary_mapping = Path(&quot;${auxiliary_id_mapping:a}&quot;)
        
        if auxiliary_mapping.is_file():
            # Use protein name mapping from auxiliary file
            df_info = pd.read_csv(auxiliary_mapping).rename(
                columns={phenotype_id: phenotype_id, &#39;EntrezGeneSymbol&#39;: &#39;gene_name&#39;}
            )[[&#39;gene_name&#39;, phenotype_id, &#39;UniProt&#39;]]
            df = df_info.merge(df, on=phenotype_id).drop(phenotype_id, axis=1)
        else:
            # Extract UniProt ID from phenotype ID if no mapping file
            df[[phenotype_id, &#39;UniProt&#39;]] = df[phenotype_id].astype(str).str.split(&#39;|&#39;, 1, expand=True)
            
        bed_template_df = load_bed_template_from_gtf(${_input[1]:ar}, &quot;${phenotype_id_column}&quot;)
        bed_df = prepare_bed(df, bed_template_df)
        bed_df[&quot;ID&quot;] = bed_df[&quot;gene_id&quot;] + &quot;_&quot; + bed_df[&quot;UniProt&quot;]
        bed_df = bed_df.drop_duplicates(&quot;ID&quot;, keep=False)[[&quot;#chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;ID&quot;] + df.drop([&quot;UniProt&quot;], axis=1).columns.values.tolist()]
        
    elif molecular_trait_type == &quot;atac&quot;:
        # For ATAC, the coordinate_annotation is directly a BED file
        bed_template_df = pd.read_csv(&quot;${_input[1]}&quot;, sep=&quot;\t&quot;).set_index(&quot;ID&quot;)
        bed_template_df = bed_template_df.assign(ID=bed_template_df.index)
        bed_df = prepare_bed(df, bed_template_df)
        bed_df = bed_df.drop_duplicates(&quot;ID&quot;, keep=False)
    else:
        raise ValueError(f&quot;Unsupported molecular_trait_type: {molecular_trait_type}&quot;)
        
    qtl.io.write_bed(bed_df, ${_output[0]:r})
    bed_df[[&quot;#chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;ID&quot;]].assign(path=${_output[0]:r}).to_csv(${_output[1]:r}, &quot;\t&quot;, index=False)

    # If annotating gene or protein, save a gene_list.tsv for gene partitioning
    if molecular_trait_type in [&quot;gene&quot;, &quot;protein&quot;]:
        # Save the region list
        region_list = bed_template_df[bed_template_df.${phenotype_id_column}.isin(df.index)]
        region_list.to_csv(&#39;${cwd:a}/${_input[0]:bn}.gene_list.tsv&#39;, sep=&quot;\t&quot;, index=False)
    else:
        # For ATAC, we cannot partition it by gene.
        warnings.warn(f&quot;Gene partitioning is not applicable for molecular trait type: {molecular_trait_type}. No gene_list.tsv will be generated.&quot;)
        


bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
        stdout=$[_output[0]:n].stdout
        for i in $[_output[0]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
        for i in $[_output[1]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `cat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="implementation-using-biomart">
<h2>Implementation using biomaRt<a class="headerlink" href="#implementation-using-biomart" title="Link to this heading">#</a></h2>
<p>This workflow adds the annotations of chr pos(TSS where start = end -1) and gene_ID to the <code class="docutils literal notranslate"><span class="pre">bed</span></code> file. <strong>This workflow is obsolete</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_coord_biomart]
parameter: ensembl_version=int
input: phenoFile
output: f&#39;{cwd:a}/{_input:bn}.bed.gz&#39;,
        f&#39;{cwd:a}/{_input:bn}.region_list.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output[0]:bn}&#39;  
R:  expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39; ,container = container, entrypoint=entrypoint

    library(biomaRt)
    library(dplyr)
    library(readr)

    # Clear biomart cache
    biomartCacheClear()

    # Read gene expression data
    gene_exp &lt;- read_delim(&quot;$[_input[0]]&quot;, delim = &quot;\t&quot;)

    # If column &quot;#chr&quot; exists, remove the first 3 columns (chr, start, and end)
    if(&quot;#chr&quot; %in% colnames(gene_exp)){
      gene_exp &lt;- gene_exp[, 4:ncol(gene_exp)]
    }

    # Connect to Ensembl biomart
    ensembl &lt;- useEnsembl(biomart = &quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;, version = &quot;$[ensembl_version]&quot;)

    # Get gene annotations from Ensembl
    ensembl_df &lt;- getBM(attributes = c(&quot;ensembl_gene_id&quot;, &quot;chromosome_name&quot;, &quot;start_position&quot;, &quot;end_position&quot;), mart = ensembl)

    # Filter and rename columns
    my_genes_ann &lt;- ensembl_df %&gt;%
      filter(ensembl_gene_id %in% gene_exp$gene_ID, chromosome_name %in% 1:23) %&gt;%
      rename(&quot;#chr&quot; = chromosome_name, 
            &quot;start&quot; = start_position, 
            &quot;end&quot; = end_position, 
            &quot;gene_ID&quot; = ensembl_gene_id) %&gt;%
      filter(gene_ID != &quot;NA&quot;)

    # Save the annotations
    my_genes_ann %&gt;%
      select(`#chr`, start, end, gene_ID) %&gt;%
      write_delim(path = &quot;$[_output[1]]&quot;, &quot;\t&quot;)

    # Merge the annotations with the gene expression data, after modifying &#39;end&#39; to be &#39;start + 1&#39;
    my_gene_bed &lt;- my_genes_ann %&gt;%
      mutate(end = start + 1) %&gt;%
      inner_join(gene_exp, by = &quot;gene_ID&quot;) %&gt;%
      arrange(`#chr`, start)

    # Save the final merged data
    write_tsv(my_gene_bed, path = &quot;$[_output[0]:n]&quot;, na = &quot;NA&quot;, append = FALSE, col_names = TRUE, quote_escape = &quot;double&quot;)


bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container = container, entrypoint=entrypoint
    bgzip -f $[_output[0]:n]
    tabix -p bed $[_output[0]] -f
</pre></div>
</div>
</div>
</div>
</section>
<section id="annotation-of-leafcutter-isoform">
<h2>Annotation of leafcutter isoform<a class="headerlink" href="#annotation-of-leafcutter-isoform" title="Link to this heading">#</a></h2>
<p>The following steps processed the output files of leafcutter so that they are TensorQTL ready. Shown below are three intemediate files</p>
<p>Exon list</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>chr</p></th>
<th class="head"><p>start</p></th>
<th class="head"><p>end</p></th>
<th class="head"><p>strand</p></th>
<th class="head"><p>gene_id</p></th>
<th class="head"><p>gene_name</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>29554</p></td>
<td><p>30039</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>30564</p></td>
<td><p>30667</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>30976</p></td>
<td><p>31097</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>35721</p></td>
<td><p>36081</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>35277</p></td>
<td><p>35481</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>34554</p></td>
<td><p>35174</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>65419</p></td>
<td><p>65433</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>65520</p></td>
<td><p>65573</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>69037</p></td>
<td><p>71585</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
</tbody>
</table>
</div>
<p>clusters_to_genes</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>clu</p></th>
<th class="head"><p>genes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1:clu_1_+</p></td>
<td><p>ENSG00000116288</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_10_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_11_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_12_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_14_-</p></td>
<td><p>ENSG00000126709</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_15_-</p></td>
<td><p>ENSG00000121753</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_16_-</p></td>
<td><p>ENSG00000121753</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_17_-</p></td>
<td><p>ENSG00000116560</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_18_-</p></td>
<td><p>ENSG00000143549</p></td>
</tr>
</tbody>
</table>
</div>
<p>phenotype_group</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>X1</p></th>
<th class="head"><p>X2</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>7:102476270:102478811:clu_309_-:ENSG00000005075</p></td>
<td><p>ENSG00000005075</p></td>
</tr>
<tr class="row-odd"><td><p>7:102476270:102478808:clu_309_-:ENSG00000005075</p></td>
<td><p>ENSG00000005075</p></td>
</tr>
<tr class="row-even"><td><p>X:47572961:47574002:clu_349_-:ENSG00000008056</p></td>
<td><p>ENSG00000008056</p></td>
</tr>
<tr class="row-odd"><td><p>X:47572999:47574002:clu_349_-:ENSG00000008056</p></td>
<td><p>ENSG00000008056</p></td>
</tr>
<tr class="row-even"><td><p>8:27236905:27239971:clu_322_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>8:27239279:27239971:clu_322_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-even"><td><p>8:27241262:27241677:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>8:27241262:27242397:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-even"><td><p>8:27241757:27242397:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>1:35558223:35559107:clu_4_+:ENSG00000020129</p></td>
<td><p>ENSG00000020129</p></td>
</tr>
</tbody>
</table>
</div>
<p>The gtf used here should be the collapsed gtf, i.e. the final output of reference_data gtf processing and the one used to called rnaseq.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[map_leafcutter_cluster_to_gene]
## Extract the code in case psichromatic needs to be processed the same way
## PheoFile in this step is the intron_count file
parameter: intron_count = path
# Defines the mapping strategy options: &#39;site&#39; or &#39;region&#39;, with &#39;site&#39; as the default. 
# The &#39;site&#39; strategy maps introns to the start and end of each exon. 
# The &#39;region&#39; strategy, to be recommended in leafcutter2, maps each intron based on it overlapping more than overlap_ratio of a gene&#39;s region.
parameter: map_stra = &quot;site&quot; 
# Define the overlap ratio as the proportion of the cluster length that intersects with a gene, used to determine mapping to the gene.
parameter: overlap_ratio = 0.8
input: intron_count, annotation_gtf
output: f&#39;{cwd}/{_input[0]:b}.exon_list&#39;, f&#39;{cwd}/{_input[0]:b}.leafcutter.clusters_to_genes.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output[0]:bn}&#39;  
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    import pandas as pd
    import qtl.annotation

    gtf = ${_input[1]:r}
    #there is no &quot;gene_type&quot; when using the uncollapsed gtf. Replace &quot;gene_biotype&quot; with &quot;gene_type&quot; in the gtf and temporarily save for use in the annotation
    has_gene_type = True
    with open(gtf, &#39;r&#39;) as file:
        for line in file:
            if line[0] == &#39;#&#39;:
                continue

            row = line.strip().split(&#39;\t&#39;)
            chrom = row[0]
            # source = row[1]
            annot_type = row[2]

            if annot_type == &quot;gene&quot;:
                if &quot;gene_type&quot; not in line:
                    has_gene_type = False
                break

    if has_gene_type == False:
        
        with open(gtf, &#39;r&#39;) as file:
            file_contents = file.read()
        updated_contents = file_contents.replace(&quot;gene_biotype&quot;, &quot;gene_type&quot;)
        gtf = ${_input[1]:rn}.tmp${_input[1]:rx}
        print(gtf)
        with open(gtf, &#39;w&#39;) as file:
            file.write(updated_contents)


    # Load data
    #annot = qtl.annotation.Annotation(${_input[1]:r})
    annot = qtl.annotation.Annotation(gtf)
    exon_df = pd.DataFrame([[g.chr, e.start_pos, e.end_pos, g.strand, g.id, g.name]
                        for g in annot.genes for e in g.transcripts[0].exons],
                       columns=[&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;, &#39;strand&#39;, &#39;gene_id&#39;, &#39;gene_name&#39;])
    exon_df.to_csv(${_output[0]:r}, sep=&#39;\t&#39;, index=False)
    
    
R:expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    suppressMessages(library(dplyr, quietly=TRUE))
    suppressMessages(library(stringr, quietly=TRUE))
    suppressMessages(library(foreach, quietly=TRUE))
    # leafcutter functions:
    
    #&#39; Make a data.frame of meta data about the introns
    #&#39; @param introns Names of the introns
    #&#39; @return Data.frame with chr, start, end, cluster id
    #&#39; @export
    get_intron_meta &lt;- function(introns) {
      intron_meta &lt;- do.call(rbind, strsplit(introns, &quot;:&quot;))  
      if (ncol(intron_meta) == 5) {
        colnames(intron_meta) &lt;- c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;clu&quot;, &quot;category&quot;)
      } else {
        colnames(intron_meta) &lt;- c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;clu&quot;)
      }

      intron_meta &lt;- as.data.frame(intron_meta, stringsAsFactors = FALSE)
      intron_meta$start &lt;- as.numeric(intron_meta$start)
      intron_meta$end &lt;- as.numeric(intron_meta$end)

      return(intron_meta)
    }
    #&#39; Work out which gene each cluster belongs to. Note the chromosome names used in the two inputs must match.
    #&#39; @param intron_meta Data frame describing the introns, usually from get_intron_meta
    #&#39; @param exons_table Table of exons, see e.g. /data/gencode19_exons.txt.gz
    #&#39; @return Data.frame with cluster ids and genes separated by commas
    #&#39; @import dplyr
    #&#39; @export
    map_clusters_to_genes_site &lt;- function(intron_meta, exons_table) {
      suppressMessages(library(foreach, quietly=TRUE))
      gene_df &lt;- foreach (chr=sort(unique(intron_meta$chr)), .combine=rbind) %dopar% {
    
        intron_chr &lt;- intron_meta[ intron_meta$chr==chr, ]
        exons_chr &lt;- exons_table[exons_table$chr==chr, ]
    
        exons_chr$temp &lt;- exons_chr$start
        intron_chr$temp &lt;- intron_chr$end
        three_prime_matches &lt;- inner_join( intron_chr, exons_chr, by=&quot;temp&quot;)
    
        exons_chr$temp &lt;- exons_chr$end
        intron_chr$temp &lt;- intron_chr$start
        five_prime_matches &lt;- inner_join( intron_chr, exons_chr, by=&quot;temp&quot;)
    
        all_matches &lt;- rbind(three_prime_matches, five_prime_matches)[ , c(&quot;clu&quot;, &quot;gene_name&quot;)]
    
        all_matches &lt;- all_matches[!duplicated(all_matches),]
    
        if (nrow(all_matches)==0) return(NULL)
        all_matches$clu &lt;- paste(chr,all_matches$clu,sep=&#39;:&#39;)
        all_matches
      }
    
      clu_df &lt;- gene_df %&gt;% group_by(clu) %&gt;% summarize(genes=paste(gene_name, collapse = &quot;,&quot;))
      class(clu_df) &lt;- &quot;data.frame&quot;
      clu_df
    }
    
    map_clusters_to_genes_region &lt;- function(intron_meta, exons_table, f) {
        #combine the exon position to get gene region
        merged_df &lt;- exons_table %&gt;%
            group_by(chr, gene_id) %&gt;%
            summarize(start = min(start), end = max(end), .groups = &#39;drop&#39;) %&gt;% as.data.frame()%&gt;%
            .[,c(&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;gene_id&quot;)]
        # use bedtools to map the intron region to gene region 
         options(bedtools.path = &quot;/home/rf2872/software/bedtools2/bin/&quot;) #Fixme: build bedtools in leafcutter sif
                map_res &lt;- bedtoolsr::bt.intersect(a = intron_meta, b = merged_df,f = f, wa = TRUE, wb=TRUE)
        #organize output of mapped file
        exon_map &lt;- map_res %&gt;% 
            mutate(clu = paste(V1, V4, sep=&quot;:&quot;), 
                   genes = .[[paste0(&quot;V&quot;, 4 + ncol(intron_meta))]]) %&gt;%
                select(clu,genes)
          return(exon_map)
    }

    cat(&quot;LeafCutter: mapping clusters to genes\n&quot;)
    intron_counts &lt;- read.table(${_input[0]:r}, header=TRUE, check.names=FALSE, row.names=4)
    intron_meta &lt;- get_intron_meta(rownames(intron_counts))
    exon_table &lt;- read.table(${_output[0]:r}, header=TRUE, stringsAsFactors=FALSE)
    if(!str_detect(intron_meta$chr[1],&quot;chr&quot;)) {
        exon_table = exon_table %&gt;% mutate(chr = str_remove_all(chr,&quot;chr&quot;))
    } else if (!any(str_detect(exon_table$chr[1],&quot;chr&quot;))) {
        exon_table = exon_table %&gt;% mutate(chr = paste0(&quot;chr&quot;,chr))
    } else {exon_table = exon_table}
    stopifnot(is.element(&#39;gene_id&#39;, colnames(exon_table)))
    exon_table[, &#39;gene_name&#39;] &lt;- exon_table[, &#39;gene_id&#39;]
    if(&quot;${map_stra}&quot; == &quot;site&quot;){
      cat(&quot;mapping clusters to genes by splicing site\n&quot;)
      m &lt;- map_clusters_to_genes_site(intron_meta, exon_table)
      } else if(&quot;${map_stra}&quot; == &quot;region&quot;){
      cat(&quot;mapping clusters to genes by overlapping gene region\n&quot;)
      m &lt;- map_clusters_to_genes_region(intron_meta, exon_table, f = ${overlap_ratio})
    } else {
      stop(&quot;Map strategy should only be &#39;site&#39; or &#39;region&#39;.&quot;)
    }
    write.table(m, ${_output[1]:r}, sep = &quot;\t&quot;, quote=FALSE, row.names=FALSE)
  
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
        stdout=$[_output[0]:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_leafcutter_isoforms]
parameter: sample_participant_lookup = path()
input: phenoFile, annotation_gtf,output_from(&quot;map_leafcutter_cluster_to_gene&quot;)
output: f&#39;{cwd:a}/{_input[0]:bn}.formated.bed.gz&#39;, f&#39;{cwd:a}/{_input[0]:bn}.phenotype_group.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output[0]:bn}&#39;  
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    import pandas as pd
    import numpy as np
    import qtl.io
    import os
    from pathlib import Path

    # Load data
    if os.path.exists(${_input[1]:rn}.tmp${_input[1]:rx}):
        gtf = ${_input[1]:rn}.tmp${_input[1]:rx}
    else:
        gtf = ${_input[1]:r}
    #tss_df = qtl.io.gtf_to_tss_bed(${_input[1]:r})
    tss_df = qtl.io.gtf_to_tss_bed(gtf)
    bed_df = pd.read_csv(${_input[0]:ar}, sep=&#39;\t&#39;, skiprows=0)
    bed_df.columns.values[0] = &quot;#chr&quot; # Temporary
    sample_participant_lookup = Path(&quot;${sample_participant_lookup:a}&quot;)
    cluster2gene_dict = pd.read_csv(${_input[3]:r}, sep=&#39;\t&#39;, index_col=0).to_dict()
    cluster2gene_dict = cluster2gene_dict[&#39;genes&#39;]
    print(&#39;    ** assigning introns to gene mapping(s)&#39;)
    n = 0
    gene_bed_df = []
    group_s = {}
    for _,r in bed_df.iterrows():
        s = r[&#39;ID&#39;].split(&#39;:&#39;)
        cluster_id = s[0]+&#39;:&#39;+s[3]
        if cluster_id in cluster2gene_dict:
            gene_ids = cluster2gene_dict[cluster_id].split(&#39;,&#39;)
            for g in gene_ids:
                gi = r[&#39;ID&#39;]+&#39;:&#39;+g
                gene_bed_df.append(tss_df.loc[g, [&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;]].tolist() + [gi] + r.iloc[4:].tolist())
                group_s[gi] = g
        else:
            n += 1
    if n &gt; 0:
        print(f&#39;    ** discarded {n} introns without a gene mapping&#39;)

    print(&#39;  * writing BED files for QTL mapping&#39;)
    gene_bed_df = pd.DataFrame(gene_bed_df, columns=bed_df.columns)
    # sort by TSS
    gene_bed_df = gene_bed_df.groupby(&#39;#chr&#39;, sort=False, group_keys=False).apply(lambda x: x.sort_values(&#39;start&#39;))
    #rename the samples if they named by file name (simply pick the first element with [.])
    gene_bed_df.columns = list(gene_bed_df.columns[:4]) + [name.split(&#39;.&#39;)[0] if &#39;junc&#39; in name else name for name in gene_bed_df.columns[4:]]
    # change sample IDs to participant IDs
    if sample_participant_lookup.is_file():
        sample_participant_lookup_s = pd.read_csv(sample_participant_lookup, sep=&quot;\t&quot;, index_col=0, dtype={0:str,1:str})
        #gene_bed_df.rename(columns=sample_participant_lookup_s.to_dict(), inplace=True)
        # Create a dictionary mapping from sample_id to participant_id
        column_mapping = dict(zip(sample_participant_lookup_s.index, sample_participant_lookup_s[&#39;participant_id&#39;]))
        # Get the column names to be replaced
        column_names = gene_bed_df.columns[4:]
        # Replace the column names using the mapping dictionary
        #new_column_names = [column_mapping.get(col, &#39;missing_data&#39;) for col in column_names]
        #it should overlap with genotype in downstream anyways
        new_column_names = [column_mapping.get(col, col) for col in column_names]
        gene_bed_df.rename(columns=dict(zip(column_names, new_column_names)), inplace=True)

    gene_bed_df = gene_bed_df.drop_duplicates()
    qtl.io.write_bed(gene_bed_df, ${_output[0]:r})
    gene_bed_df[[&#39;start&#39;, &#39;end&#39;]] = gene_bed_df[[&#39;start&#39;, &#39;end&#39;]].astype(np.int32)
    gene_bed_df[gene_bed_df.columns[4:]] = gene_bed_df[gene_bed_df.columns[4:]].astype(np.float32)
    group_s_df =  pd.Series(group_s).sort_values().reset_index()
    group_s_df.columns = [&#39;ID&#39;, &#39;gene&#39;]
    group_s_df.to_csv(${_output[1]:r}, sep=&#39;\t&#39;, index=False, header=True)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
        stdout=$[_output[0]:n].stdout

        rm $[_input[1]:rn].tmp$[_input[1]:rx]

        for i in $[_output[0]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
        for i in $[_output[1]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `cat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="processing-of-psichomics-output">
<h2>Processing of psichomics output<a class="headerlink" href="#processing-of-psichomics-output" title="Link to this heading">#</a></h2>
<p>It occurs that the psichomatic by default grouped the isoforms by gene name, so only thing needs to be done is to extract this information and potentially renamed the gene symbol into ENSG ID</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[annotate_psichomics_isoforms]
parameter: sample_participant_lookup = path()
input: phenoFile, annotation_gtf
output: f&#39;{cwd:a}/{_input[0]:bn}.formated.bed.gz&#39;, f&#39;{cwd:a}/{_input[0]:bn}.phenotype_group.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output[0]:bn}&#39;  
python: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    import pandas as pd
    import numpy as np
    import qtl.io
    from pathlib import Path
    tss_df = qtl.io.gtf_to_tss_bed(${_input[1]:r}, feature=&#39;gene&#39;,phenotype_id = &quot;gene_id&quot; )
    bed_df = pd.read_csv(${_input[0]:ar}, sep=&#39;\t&#39;, skiprows=0)
    bed_df[&quot;gene_id&quot;]  = [x[-1] for x in bed_df.ID.str.split(&quot;_&quot;)]
    sample_participant_lookup = Path(&quot;${sample_participant_lookup:a}&quot;)
    if &quot;start&quot; in  bed_df.columns:
        bed_df = bed_df.drop([&quot;#Chr&quot;,&quot;start&quot;,&quot;end&quot;],axis = 1)
    output = tss_df.merge(bed_df, left_on = [&quot;gene_id&quot;], right_on = [&quot;gene_id&quot;],how = &quot;right&quot;).sort_values([&quot;chr&quot;,&quot;start&quot;])
    # change sample IDs to participant IDs
    if sample_participant_lookup.is_file():
        sample_participant_lookup_s = pd.read_csv(sample_participant_lookup, sep=&quot;\t&quot;, index_col=0, dtype={0:str,1:str})

        column_mapping = dict(zip(sample_participant_lookup_s.index, sample_participant_lookup_s[&#39;participant_id&#39;]))

        column_names = output.columns[4:]
        print(column_names)
        new_column_names = [column_mapping.get(col, col) for col in column_names]
        output.rename(columns=dict(zip(column_names, new_column_names)), inplace=True)

        #output.rename(columns=sample_participant_lookup_s.to_dict(), inplace=True)

    # Old code grouping by each gene
    #bed_output = output.drop(&quot;gene_id&quot; , axis = 1)
    #phenotype_group = output[[&quot;ID&quot;,&quot;gene_id&quot;]]

    bed_output = output.drop(&quot;gene_id&quot; , axis = 1)

    # R version 
    #output = output%&gt;%mutate(gene_type_id = map_chr(str_split( ID ,&quot;_&quot;),~paste0(.x[length(.x)],&quot;_&quot;,.x[1] )) )
    
    # corrected in python
    output = output.assign(gene_type_id = output[&#39;ID&#39;].str.split(&#39;_&#39;).map(lambda x: x[-1] + &#39;_&#39; + x[0]))
    phenotype_group = output[[&quot;ID&quot;,&quot;gene_type_id&quot;]]

    bed_output.to_csv(${_output[0]:nr},&quot;\t&quot;,index = False)
    phenotype_group.to_csv(${_output[1]:r},&quot;\t&quot;,index = False,header=False)

bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
    bgzip -f ${_output[0]:n}
    tabix ${_output[0]}

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
        stdout=$[_output[0]:n].stdout
        for i in $[_output[0]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
        for i in $[_output[1]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_headerow:&quot; `cat $i | grep &quot;##&quot; | wc -l `   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | grep -V &quot;##&quot; | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing/phenotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../phenotype_preprocessing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Phenotype data preprocessing</p>
      </div>
    </a>
    <a class="right-next"
       href="phenotype_imputation.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Phenotype data imputation</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alternative-implementation">Alternative implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-cooridnate-annotation">i. Cooridnate Annotation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-gene-expression">a. Gene Expression</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#b-proteins">b. Proteins</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotate-coord">Annotate coord</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Input:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Output:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-pyqtl">Implementation using <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-using-biomart">Implementation using biomaRt</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-of-leafcutter-isoform">Annotation of leafcutter isoform</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-of-psichomics-output">Processing of psichomics output</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>