
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Gene coordinate annotation &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Phenotype data formatting" href="phenotype_formatting.html" />
    <link rel="prev" title="Phenotype data preprocessing" href="../phenotype_preprocessing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   ADSP-FG xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Overview of design
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/data_preprocessing/phenotype/gene_annotation.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/phenotype/gene_annotation.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/phenotype/gene_annotation.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#alternative-implementation">
     Alternative implementation
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#region-list-generation">
   Region List generation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#implementation-using-pyqtl">
   Implementation using
   <code class="docutils literal notranslate">
    <span class="pre">
     pyqtl
    </span>
   </code>
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#implementation-using-biomart">
   Implementation using biomaRt
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#annotation-of-leafcutter-isoform">
   Annotation of leafcutter isoform
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#processing-of-psichomics-output">
   Processing of psichomics output
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="gene-coordinate-annotation">
<h1>Gene coordinate annotation<a class="headerlink" href="#gene-coordinate-annotation" title="Permalink to this headline">¶</a></h1>
<p>This workflow adds genomic coordinate annotation to gene-level molecular phenotype files generated in <code class="docutils literal notranslate"><span class="pre">gct</span></code> format and convert them to <code class="docutils literal notranslate"><span class="pre">bed</span></code> format for downstreams analysis.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This pipeline is based on <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/src/eqtl_prepare_expression.py"><code class="docutils literal notranslate"><span class="pre">pyqtl</span></code>, as demonstrated here</a>.</p>
<p><strong>FIXME: please explain here what we do with gene symbol vs gene ID</strong></p>
<div class="section" id="alternative-implementation">
<h3>Alternative implementation<a class="headerlink" href="#alternative-implementation" title="Permalink to this headline">¶</a></h3>
<p>Previously we use <code class="docutils literal notranslate"><span class="pre">biomaRt</span></code> package in R instead of code from <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code>. The core function calls are:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span>    <span class="n">ensembl</span> <span class="o">=</span> <span class="nf">useEnsembl</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s">&quot;ensembl&quot;</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;$[ensembl_version]&quot;</span><span class="p">)</span>
    <span class="n">ensembl_df</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="s">&quot;start_position&quot;</span><span class="p">,</span> <span class="s">&quot;end_position&quot;</span><span class="p">),</span><span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
</pre></div>
</div>
<p>We require ENSEMBL version to be specified explicitly in this pipeline. As of 2021 for the Brain xQTL project, we use ENSEMBL version 103.</p>
</div>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Molecular phenotype data in <code class="docutils literal notranslate"><span class="pre">gct</span></code> format, with the first column being ENSEMBL ID and other columns being sample names.</p></li>
<li><p>GTF for collapsed gene model</p>
<ul class="simple">
<li><p>the gene names must be consistent with the molecular phenotype data matrices (eg ENSG00000000003 vs. ENSG00000000003.1 will not work)</p></li>
</ul>
</li>
<li><p>(Optional) Meta-data to match between sample names in expression data and genotype files</p>
<ul class="simple">
<li><p>Required input</p></li>
<li><p>Tab delimited with header</p></li>
<li><p>Only 2 columns: first column is sample name in expression data, 2nd column is sample name in genotype data</p></li>
<li><p><strong>must contains all the sample name in expression matrices even if they don’t existing in genotype data</strong></p></li>
</ul>
</li>
</ol>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>Molecular phenotype data in <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>The MWE is uploaded to the <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1Rv2bWHBbX_tastTh49ToYVDMV6rFP5Wk">Google Drive</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">annotate_coord</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">pheno_log2cpm</span><span class="o">.</span><span class="n">tsv</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">annotation</span><span class="o">-</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">103.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">sample</span><span class="o">-</span><span class="n">participant</span><span class="o">-</span><span class="n">lookup</span> <span class="n">data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="nb">id</span><span class="o">-</span><span class="nb">type</span> <span class="n">gene_name</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run gene_annotation.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  annotate_coord
  annotate_coord_biomart

Global Workflow Options:
  --cwd output (as path)
                        Work directory &amp; output directory
  --annotation-gtf VAL (as path, required)
                        gene gtf annotation table
  --phenoFile VAL (as path, required)
                        Molecular phenotype matrix
  --phenotype-id-type &#39;gene_id&#39;
                        Whether the input data is named by gene_id or gene_name.
                        By default it is gene_id, if not, please change it to
                        gene_name
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 1 (as int)
                        Number of threads
  --container &#39;&#39;

Sections
  annotate_coord:
    Workflow Options:
      --sample-participant-lookup . (as path)
                        A file to map sample ID from expression to genotype,
                        must contain two columns, sample_id and participant_id,
                        mapping IDs in the expression files to IDs in the
                        genotype (these can be the same).
  annotate_coord_biomart:
    Workflow Options:
      --ensembl-version VAL (as int, required)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Work directory &amp; output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1">#  gene gtf annotation table</span>
<span class="kn">parameter:</span> <span class="n">annotation_gtf</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Molecular phenotype matrix</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Whether the input data is named by gene_id or gene_name. By default it is gene_id, if not, please change it to gene_name</span>
<span class="kn">parameter:</span> <span class="n">phenotype_id_type</span> <span class="o">=</span> <span class="s1">&#39;gene_id&#39;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="region-list-generation">
<h2>Region List generation<a class="headerlink" href="#region-list-generation" title="Permalink to this headline">¶</a></h2>
<p>To partitioning the data by genes require a region list file which:</p>
<ol class="simple">
<li><p>have 5 columns: chr,start,end,gene_id,gene_name</p></li>
<li><p>have the same gene as or less gene than that of the bed file</p></li>
</ol>
<p>Input:</p>
<ol class="simple">
<li><p>A gtf file used to generated the bed</p></li>
<li><p>A phenotype bed file, must have a gene_id column indicating the name of genes.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">region_list_generation</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">phenoFile</span><span class="p">,</span> <span class="n">annotation_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.region_list&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">qtl.io</span>
    <span class="c1"># get the five column data</span>
    <span class="n">bed_template_df_id</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">gtf_to_tss_bed</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;transcript&#39;</span><span class="p">,</span><span class="n">phenotype_id</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span> <span class="p">)</span>
    <span class="n">bed_template_df_name</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">gtf_to_tss_bed</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;transcript&#39;</span><span class="p">,</span><span class="n">phenotype_id</span> <span class="o">=</span> <span class="s2">&quot;gene_name&quot;</span> <span class="p">)</span>
    <span class="n">bed_template_df</span> <span class="o">=</span> <span class="n">bed_template_df_id</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">bed_template_df_name</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">])</span>
    <span class="n">bed_template_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># Retaining only the genes in the data</span>
    <span class="n">region_list</span> <span class="o">=</span> <span class="n">bed_template_df</span><span class="p">[</span><span class="n">bed_template_df</span><span class="o">.</span><span class="err">$</span><span class="p">{</span><span class="n">phenotype_id_type</span><span class="p">}</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">pheno</span><span class="o">.</span><span class="n">gene_id</span><span class="p">)]</span>
    <span class="n">region_list</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="implementation-using-pyqtl">
<h2>Implementation using <code class="docutils literal notranslate"><span class="pre">pyqtl</span></code><a class="headerlink" href="#implementation-using-pyqtl" title="Permalink to this headline">¶</a></h2>
<p>Implementation based on <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/src/eqtl_prepare_expression.py">GTEx pipeline</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">annotate_coord</span><span class="p">]</span>
<span class="c1"># A file to map sample ID from expression to genotype, must contain two columns, sample_id and participant_id, mapping IDs in the expression files to IDs in the genotype (these can be the same).</span>
<span class="kn">parameter:</span> <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">input:</span> <span class="n">phenoFile</span><span class="p">,</span> <span class="n">annotation_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.bed.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>

    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">qtl.io</span>
    <span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
    <span class="k">def</span> <span class="nf">prepare_bed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">bed_template_df</span><span class="p">,</span> <span class="n">chr_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">bed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">bed_template_df</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># sort by start position</span>
        <span class="n">bed_df</span> <span class="o">=</span> <span class="n">bed_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">group_keys</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;start&#39;</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">chr_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># subset chrs from VCF</span>
            <span class="n">bed_df</span> <span class="o">=</span> <span class="n">bed_df</span><span class="p">[</span><span class="n">bed_df</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">chr_subset</span><span class="p">)]</span>
        <span class="k">return</span> <span class="n">bed_df</span>
    <span class="c1"># Load data</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;${sample_participant_lookup:a}&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;chr&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s2">&quot;start&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span>  <span class="s2">&quot;end&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span> <span class="p">])</span>
    <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># change sample IDs to participant IDs</span>
    <span class="k">if</span> <span class="n">sample_participant_lookup</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
        <span class="n">sample_participant_lookup_s</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample_participant_lookup</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="nb">str</span><span class="p">},</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">sample_participant_lookup_s</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">gtf_to_tss_bed</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span><span class="n">phenotype_id</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span> <span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">())</span> <span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GTF file $</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">ar</span><span class="si">}</span><span class="s2"> needs to be collapsed into gene model by reference data processing module&quot;</span><span class="p">)</span>
         
    <span class="n">bed_template_df</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">gtf_to_tss_bed</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;transcript&#39;</span><span class="p">,</span><span class="n">phenotype_id</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{phenotype_id_type}</span><span class="s2">&quot;</span> <span class="p">)</span>

    <span class="c1">### Detect duplicated gene_id</span>
    <span class="n">dup_count</span> <span class="o">=</span> <span class="n">bed_template_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">bed_template_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">cumcount</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">dup_count</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="sa">f</span><span class="s1">&#39;.</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.0&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">dup_count</span><span class="p">])</span>
    <span class="c1">### Add surfix to the duplicated gene_id</span>
    <span class="n">bed_template_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">bed_template_df</span><span class="o">.</span><span class="n">index</span> <span class="o">+</span>  <span class="n">dup_count</span>
    <span class="n">bed_template_df</span><span class="o">.</span><span class="n">gene_id</span> <span class="o">=</span> <span class="n">bed_template_df</span><span class="o">.</span><span class="n">index</span>
    <span class="n">bed_df</span> <span class="o">=</span> <span class="n">prepare_bed</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">bed_template_df</span><span class="p">)</span>
    <span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">write_bed</span><span class="p">(</span><span class="n">bed_df</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="implementation-using-biomart">
<h2>Implementation using biomaRt<a class="headerlink" href="#implementation-using-biomart" title="Permalink to this headline">¶</a></h2>
<p>This workflow adds the annotations of chr pos(TSS where start = end -1) and gene_ID to the <code class="docutils literal notranslate"><span class="pre">bed</span></code> file. <strong>This workflow is obsolete</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">annotate_coord_biomart</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">ensembl_version</span><span class="o">=</span><span class="nb">int</span>

<span class="kn">input:</span> <span class="n">phenoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.bed.gz&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.region_list&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;biomaRt&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="n">biomartCacheClear</span><span class="p">()</span>
    <span class="n">gene_exp</span> <span class="o">=</span> <span class="n">readr</span><span class="p">::</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_input[0]]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="s2">&quot;#chr&quot;</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">colnames</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">)</span> <span class="p">){</span>
      <span class="c1"># need to re-annotate</span>
      <span class="n">gene_exp</span> <span class="o">=</span> <span class="n">gene_exp</span><span class="p">[,</span><span class="mi">4</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">)]</span>
    <span class="p">}</span>
    <span class="n">ensembl</span> <span class="o">=</span> <span class="n">useEnsembl</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s2">&quot;ensembl&quot;</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="s2">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="s2">&quot;$[ensembl_version]&quot;</span><span class="p">)</span>
    <span class="n">ensembl_df</span> <span class="o">&lt;-</span> <span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s2">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="s2">&quot;start_position&quot;</span><span class="p">,</span> <span class="s2">&quot;end_position&quot;</span><span class="p">),</span><span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
    <span class="n">my_genes</span> <span class="o">=</span> <span class="n">gene_exp</span><span class="err">$</span><span class="n">gene_ID</span>
    <span class="n">keep_genes</span> <span class="o">=</span>  <span class="n">my_genes</span>
    <span class="n">my_genes_ann</span> <span class="o">=</span> <span class="n">ensembl_df</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="n">my_genes</span><span class="p">,</span> <span class="n">ensembl_df</span><span class="err">$</span><span class="n">ensembl_gene_id</span><span class="p">),]</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="n">chromosome_name</span><span class="o">%</span><span class="ow">in</span><span class="o">%</span><span class="mi">1</span><span class="p">:</span><span class="mi">23</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">dplyr</span><span class="p">::</span><span class="n">rename</span><span class="p">(</span> <span class="s2">&quot;#chr&quot;</span> <span class="o">=</span> <span class="n">chromosome_name</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span> <span class="o">=</span> <span class="n">start_position</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span> <span class="o">=</span> <span class="n">end_position</span><span class="p">,</span><span class="s2">&quot;gene_ID&quot;</span> <span class="o">=</span> <span class="n">ensembl_gene_id</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="n">gene_ID</span><span class="o">!=</span><span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">gene_ID</span><span class="o">%</span><span class="ow">in</span><span class="o">%</span><span class="n">keep_genes</span><span class="p">)</span>
    <span class="n">my_genes_ann</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="err">`</span><span class="c1">#chr`,start,end,gene_ID)%&gt;%write_delim(path = &quot;$[_output[1]]&quot;,&quot;\t&quot;)</span>
    <span class="n">my_gene_bed</span> <span class="o">=</span> <span class="n">inner_join</span><span class="p">(</span><span class="n">my_genes_ann</span> <span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="err">`</span><span class="c1">#chr`,start,end,gene_ID),gene_exp,by = &quot;gene_ID&quot; )%&gt;%arrange(`#chr`,start) </span>
    <span class="n">my_gene_bed</span><span class="o">%&gt;%</span><span class="n">readr</span><span class="p">::</span><span class="n">write_tsv</span><span class="p">(</span> <span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;$[_output[0]:n]&quot;</span><span class="p">,</span> <span class="n">na</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">quote_escape</span> <span class="o">=</span> <span class="s2">&quot;double&quot;</span><span class="p">)</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
        <span class="n">bgzip</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span>
        <span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">bed</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">-</span><span class="n">f</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="annotation-of-leafcutter-isoform">
<h2>Annotation of leafcutter isoform<a class="headerlink" href="#annotation-of-leafcutter-isoform" title="Permalink to this headline">¶</a></h2>
<p>The following steps processed the output files of leafcutter so that they are TensorQTL ready. Shown below are three intemediate files</p>
<p>Exon list</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>chr</p></th>
<th class="head"><p>start</p></th>
<th class="head"><p>end</p></th>
<th class="head"><p>strand</p></th>
<th class="head"><p>gene_id</p></th>
<th class="head"><p>gene_name</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>29554</p></td>
<td><p>30039</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>30564</p></td>
<td><p>30667</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>30976</p></td>
<td><p>31097</p></td>
<td><p>+</p></td>
<td><p>ENSG00000243485</p></td>
<td><p>MIR1302-2HG</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>35721</p></td>
<td><p>36081</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>35277</p></td>
<td><p>35481</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>34554</p></td>
<td><p>35174</p></td>
<td><p>-</p></td>
<td><p>ENSG00000237613</p></td>
<td><p>FAM138A</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>65419</p></td>
<td><p>65433</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
<tr class="row-odd"><td><p>chr1</p></td>
<td><p>65520</p></td>
<td><p>65573</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
<tr class="row-even"><td><p>chr1</p></td>
<td><p>69037</p></td>
<td><p>71585</p></td>
<td><p>+</p></td>
<td><p>ENSG00000186092</p></td>
<td><p>OR4F5</p></td>
</tr>
</tbody>
</table>
<p>clusters_to_genes</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>clu</p></th>
<th class="head"><p>genes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1:clu_1_+</p></td>
<td><p>ENSG00000116288</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_10_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_11_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_12_+</p></td>
<td><p>ENSG00000143774</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_14_-</p></td>
<td><p>ENSG00000126709</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_15_-</p></td>
<td><p>ENSG00000121753</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_16_-</p></td>
<td><p>ENSG00000121753</p></td>
</tr>
<tr class="row-odd"><td><p>1:clu_17_-</p></td>
<td><p>ENSG00000116560</p></td>
</tr>
<tr class="row-even"><td><p>1:clu_18_-</p></td>
<td><p>ENSG00000143549</p></td>
</tr>
</tbody>
</table>
<p>phenotype_group</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>X1</p></th>
<th class="head"><p>X2</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>7:102476270:102478811:clu_309_-:ENSG00000005075</p></td>
<td><p>ENSG00000005075</p></td>
</tr>
<tr class="row-odd"><td><p>7:102476270:102478808:clu_309_-:ENSG00000005075</p></td>
<td><p>ENSG00000005075</p></td>
</tr>
<tr class="row-even"><td><p>X:47572961:47574002:clu_349_-:ENSG00000008056</p></td>
<td><p>ENSG00000008056</p></td>
</tr>
<tr class="row-odd"><td><p>X:47572999:47574002:clu_349_-:ENSG00000008056</p></td>
<td><p>ENSG00000008056</p></td>
</tr>
<tr class="row-even"><td><p>8:27236905:27239971:clu_322_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>8:27239279:27239971:clu_322_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-even"><td><p>8:27241262:27241677:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>8:27241262:27242397:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-even"><td><p>8:27241757:27242397:clu_323_-:ENSG00000015592</p></td>
<td><p>ENSG00000015592</p></td>
</tr>
<tr class="row-odd"><td><p>1:35558223:35559107:clu_4_+:ENSG00000020129</p></td>
<td><p>ENSG00000020129</p></td>
</tr>
</tbody>
</table>
<p>The gtf used here should be the collapsed gtf, i.e. the final output of reference_data gtf processing and the one used to called rnaseq.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">map_leafcutter_cluster_to_gene</span><span class="p">]</span>
<span class="c1">## Extract the code in case psichromatic needs to be processed the same way</span>
<span class="c1">## PheoFile in this step is the intron_count file</span>
<span class="kn">parameter:</span> <span class="n">intron_count</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">input:</span> <span class="n">intron_count</span><span class="p">,</span> <span class="n">annotation_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">.exon_list&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">b</span><span class="si">}</span><span class="s1">.leafcutter.clusters_to_genes.txt&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">qtl.annotation</span>
    <span class="c1"># Load data</span>
    <span class="n">annot</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">Annotation</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">exon_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([[</span><span class="n">g</span><span class="o">.</span><span class="n">chr</span><span class="p">,</span> <span class="n">e</span><span class="o">.</span><span class="n">start_pos</span><span class="p">,</span> <span class="n">e</span><span class="o">.</span><span class="n">end_pos</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="kp">name</span><span class="p">]</span>
                        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">annot</span><span class="o">.</span><span class="n">genes</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">exons</span><span class="p">],</span>
                       <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">])</span>
    <span class="n">exon_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


<span class="kn">R:</span><span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">suppressMessages</span><span class="p">(</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">,</span> <span class="n">quietly</span><span class="o">=</span><span class="n">TRUE</span><span class="p">))</span>
    <span class="n">suppressMessages</span><span class="p">(</span><span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">,</span> <span class="n">quietly</span><span class="o">=</span><span class="n">TRUE</span><span class="p">))</span>
    <span class="n">suppressMessages</span><span class="p">(</span><span class="n">library</span><span class="p">(</span><span class="n">foreach</span><span class="p">,</span> <span class="n">quietly</span><span class="o">=</span><span class="n">TRUE</span><span class="p">))</span>
    <span class="c1"># leafcutter functions:</span>
    
    <span class="c1">#&#39; Make a data.frame of meta data about the introns</span>
    <span class="c1">#&#39; @param introns Names of the introns</span>
    <span class="c1">#&#39; @return Data.frame with chr, start, end, cluster id</span>
    <span class="c1">#&#39; @export</span>
    <span class="n">get_intron_meta</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">introns</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">intron_meta</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">strsplit</span><span class="p">(</span><span class="n">introns</span><span class="p">,</span><span class="s2">&quot;:&quot;</span><span class="p">))</span>
      <span class="n">colnames</span><span class="p">(</span><span class="n">intron_meta</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;clu&quot;</span><span class="p">)</span>
      <span class="n">intron_meta</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">intron_meta</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
      <span class="n">intron_meta</span><span class="err">$</span><span class="n">start</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">intron_meta</span><span class="err">$</span><span class="n">start</span><span class="p">)</span>
      <span class="n">intron_meta</span><span class="err">$</span><span class="n">end</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">intron_meta</span><span class="err">$</span><span class="n">end</span><span class="p">)</span>
      <span class="n">intron_meta</span>
    <span class="p">}</span>
    
    <span class="c1">#&#39; Work out which gene each cluster belongs to. Note the chromosome names used in the two inputs must match.</span>
    <span class="c1">#&#39; @param intron_meta Data frame describing the introns, usually from get_intron_meta</span>
    <span class="c1">#&#39; @param exons_table Table of exons, see e.g. /data/gencode19_exons.txt.gz</span>
    <span class="c1">#&#39; @return Data.frame with cluster ids and genes separated by commas</span>
    <span class="c1">#&#39; @import dplyr</span>
    <span class="c1">#&#39; @export</span>
    <span class="n">map_clusters_to_genes</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">intron_meta</span><span class="p">,</span> <span class="n">exons_table</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">gene_df</span> <span class="o">&lt;-</span> <span class="n">foreach</span> <span class="p">(</span><span class="nb">chr</span><span class="o">=</span><span class="n">sort</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">intron_meta</span><span class="err">$</span><span class="nb">chr</span><span class="p">)),</span> <span class="o">.</span><span class="n">combine</span><span class="o">=</span><span class="n">rbind</span><span class="p">)</span> <span class="o">%</span><span class="n">dopar</span><span class="o">%</span> <span class="p">{</span>
    
        <span class="n">intron_chr</span> <span class="o">&lt;-</span> <span class="n">intron_meta</span><span class="p">[</span> <span class="n">intron_meta</span><span class="err">$</span><span class="nb">chr</span><span class="o">==</span><span class="nb">chr</span><span class="p">,</span> <span class="p">]</span>
        <span class="n">exons_chr</span> <span class="o">&lt;-</span> <span class="n">exons_table</span><span class="p">[</span><span class="n">exons_table</span><span class="err">$</span><span class="nb">chr</span><span class="o">==</span><span class="nb">chr</span><span class="p">,</span> <span class="p">]</span>
    
        <span class="n">exons_chr</span><span class="err">$</span><span class="n">temp</span> <span class="o">&lt;-</span> <span class="n">exons_chr</span><span class="err">$</span><span class="n">start</span>
        <span class="n">intron_chr</span><span class="err">$</span><span class="n">temp</span> <span class="o">&lt;-</span> <span class="n">intron_chr</span><span class="err">$</span><span class="n">end</span>
        <span class="n">three_prime_matches</span> <span class="o">&lt;-</span> <span class="n">inner_join</span><span class="p">(</span> <span class="n">intron_chr</span><span class="p">,</span> <span class="n">exons_chr</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;temp&quot;</span><span class="p">)</span>
    
        <span class="n">exons_chr</span><span class="err">$</span><span class="n">temp</span> <span class="o">&lt;-</span> <span class="n">exons_chr</span><span class="err">$</span><span class="n">end</span>
        <span class="n">intron_chr</span><span class="err">$</span><span class="n">temp</span> <span class="o">&lt;-</span> <span class="n">intron_chr</span><span class="err">$</span><span class="n">start</span>
        <span class="n">five_prime_matches</span> <span class="o">&lt;-</span> <span class="n">inner_join</span><span class="p">(</span> <span class="n">intron_chr</span><span class="p">,</span> <span class="n">exons_chr</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;temp&quot;</span><span class="p">)</span>
    
        <span class="n">all_matches</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="n">three_prime_matches</span><span class="p">,</span> <span class="n">five_prime_matches</span><span class="p">)[</span> <span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;clu&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">)]</span>
    
        <span class="n">all_matches</span> <span class="o">&lt;-</span> <span class="n">all_matches</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">all_matches</span><span class="p">),]</span>
    
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">all_matches</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">)</span> <span class="k">return</span><span class="p">(</span><span class="n">NULL</span><span class="p">)</span>
        <span class="n">all_matches</span><span class="err">$</span><span class="n">clu</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="nb">chr</span><span class="p">,</span><span class="n">all_matches</span><span class="err">$</span><span class="n">clu</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>
        <span class="n">all_matches</span>
      <span class="p">}</span>
    
      <span class="n">clu_df</span> <span class="o">&lt;-</span> <span class="n">gene_df</span> <span class="o">%&gt;%</span> <span class="n">group_by</span><span class="p">(</span><span class="n">clu</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">summarize</span><span class="p">(</span><span class="n">genes</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">))</span>
      <span class="n">class</span><span class="p">(</span><span class="n">clu_df</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">&quot;data.frame&quot;</span>
      <span class="n">clu_df</span>
    <span class="p">}</span>

    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;LeafCutter: mapping clusters to genes</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">intron_counts</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">check</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">intron_meta</span> <span class="o">&lt;-</span> <span class="n">get_intron_meta</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">intron_counts</span><span class="p">))</span>
    <span class="n">exon_table</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="err">!</span><span class="n">str_detect</span><span class="p">(</span><span class="n">intron_meta</span><span class="p">,</span><span class="s2">&quot;chr&quot;</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">exon_table</span> <span class="o">=</span> <span class="n">exon_table</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="nb">chr</span> <span class="o">=</span> <span class="n">str_remove_all</span><span class="p">(</span><span class="nb">chr</span><span class="p">,</span><span class="s2">&quot;chr&quot;</span><span class="p">))</span>
    <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="nb">any</span><span class="p">(</span><span class="n">str_detect</span><span class="p">(</span><span class="n">exon_table</span><span class="err">$</span><span class="nb">chr</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;chr&quot;</span><span class="p">)))</span> <span class="p">{</span>
        <span class="n">exon_table</span> <span class="o">=</span> <span class="n">exon_table</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="nb">chr</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="nb">chr</span><span class="p">))</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">(</span><span class="n">exon_table</span> <span class="o">=</span> <span class="n">exon_table</span><span class="p">)</span>
    <span class="n">stopifnot</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">element</span><span class="p">(</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="n">colnames</span><span class="p">(</span><span class="n">exon_table</span><span class="p">)))</span>
    <span class="n">exon_table</span><span class="p">[,</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">exon_table</span><span class="p">[,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">]</span>
    <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">map_clusters_to_genes</span><span class="p">(</span><span class="n">intron_meta</span><span class="p">,</span> <span class="n">exon_table</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">annotate_leafcutter_isoforms</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">input:</span> <span class="n">phenoFile</span><span class="p">,</span> <span class="n">annotation_gtf</span><span class="p">,</span><span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;map_leafcutter_cluster_to_gene&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.formated.bed.gz&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.phenotype_group.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">qtl.io</span>
    <span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
    <span class="c1"># Load data</span>
    <span class="n">tss_df</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">gtf_to_tss_bed</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">bed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">bed_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#chr&quot;</span> <span class="c1"># Temporary</span>
    <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;${sample_participant_lookup:a}&quot;</span><span class="p">)</span>
    <span class="n">cluster2gene_dict</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;    ** assigning introns to gene mapping(s)&#39;</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">gene_bed_df</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">group_s</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span><span class="n">r</span> <span class="ow">in</span> <span class="n">bed_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>
        <span class="n">cluster_id</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:&#39;</span><span class="o">+</span><span class="n">s</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">cluster_id</span> <span class="ow">in</span> <span class="n">cluster2gene_dict</span><span class="p">:</span>
            <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">cluster2gene_dict</span><span class="p">[</span><span class="n">cluster_id</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">:</span>
                <span class="n">gi</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:&#39;</span><span class="o">+</span><span class="n">g</span>
                <span class="n">gene_bed_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tss_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">g</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">+</span> <span class="p">[</span><span class="n">gi</span><span class="p">]</span> <span class="o">+</span> <span class="n">r</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
                <span class="n">group_s</span><span class="p">[</span><span class="n">gi</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;    ** discarded </span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s1"> introns without a gene mapping&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  * writing BED files for QTL mapping&#39;</span><span class="p">)</span>
    <span class="n">gene_bed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_bed_df</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">bed_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="c1"># sort by TSS</span>
    <span class="n">gene_bed_df</span> <span class="o">=</span> <span class="n">gene_bed_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">group_keys</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;start&#39;</span><span class="p">))</span>
    <span class="c1"># change sample IDs to participant IDs</span>
    <span class="k">if</span> <span class="n">sample_participant_lookup</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
        <span class="n">sample_participant_lookup_s</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample_participant_lookup</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="nb">str</span><span class="p">},</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">gene_bed_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">sample_participant_lookup_s</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">gene_bed_df</span> <span class="o">=</span> <span class="n">gene_bed_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">write_bed</span><span class="p">(</span><span class="n">gene_bed_df</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">gene_bed_df</span><span class="p">[[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">gene_bed_df</span><span class="p">[[</span><span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="n">gene_bed_df</span><span class="p">[</span><span class="n">gene_bed_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]]</span> <span class="o">=</span> <span class="n">gene_bed_df</span><span class="p">[</span><span class="n">gene_bed_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">group_s</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="processing-of-psichomics-output">
<h2>Processing of psichomics output<a class="headerlink" href="#processing-of-psichomics-output" title="Permalink to this headline">¶</a></h2>
<p>It occurs that the psichomatic by default grouped the isoforms by gene name, so only thing needs to be done is to extract this information and potentially renamed the gene symbol into ENSG ID</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">annotate_psichomics_isoforms</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">input:</span> <span class="n">phenoFile</span><span class="p">,</span> <span class="n">annotation_gtf</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.formated.bed.gz&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.phenotype_group.txt&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">qtl.io</span>
    <span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
    <span class="n">tss_df</span> <span class="o">=</span> <span class="n">qtl</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">gtf_to_tss_bed</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">feature</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span><span class="n">phenotype_id</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span> <span class="p">)</span>
    <span class="n">bed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">ar</span><span class="p">},</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">bed_df</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span>  <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">bed_df</span><span class="o">.</span><span class="n">ID</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)]</span>
    <span class="n">sample_participant_lookup</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;${sample_participant_lookup:a}&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;start&quot;</span> <span class="ow">in</span>  <span class="n">bed_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">bed_df</span> <span class="o">=</span> <span class="n">bed_df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;#Chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">],</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">tss_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">bed_df</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">],</span> <span class="n">right_on</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">],</span><span class="n">how</span> <span class="o">=</span> <span class="s2">&quot;right&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">])</span>
    <span class="c1"># change sample IDs to participant IDs</span>
    <span class="k">if</span> <span class="n">sample_participant_lookup</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
        <span class="n">sample_participant_lookup_s</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample_participant_lookup</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="nb">str</span><span class="p">},</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">output</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">sample_participant_lookup_s</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">bed_output</span> <span class="o">=</span> <span class="n">output</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;gene_id&quot;</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">phenotype_group</span> <span class="o">=</span> <span class="n">output</span><span class="p">[[</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]]</span>
    <span class="n">bed_output</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nr</span><span class="p">},</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">phenotype_group</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">bgzip</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span>
    <span class="n">tabix</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/data_preprocessing/phenotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../phenotype_preprocessing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Phenotype data preprocessing</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="phenotype_formatting.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Phenotype data formatting</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021+, ADSP-FG xQTL methods and data integration working group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>