

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Mendelian Randomization using fine-mapping IV &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=c5ced968eda925caa686" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=c5ced968eda925caa686"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/pecotmr_integration/fine_mr';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="TWAS and COLOC integration" href="intact.html" />
    <link rel="prev" title="TWAS for xQTL" href="ptwas.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden factor analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">Overview of advanced cis-QTL analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../cis_analysis/cis_workhorse.html">Advanced cis-QTL analysis with individual level data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="pecotmr.html">The PECOTMR intergration framework</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="pecotmr_preprocessing.html">xQTL and GWAS region alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="ptwas.html">TWAS for xQTL</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Mendelian Randomization using fine-mapping IV</a></li>
<li class="toctree-l2"><a class="reference internal" href="intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/pecotmr_integration/fine_mr.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_integration/fine_mr.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/pecotmr_integration/fine_mr.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Mendelian Randomization using fine-mapping IV</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-the-units-of-analysis">Generate the units of analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Input</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#output-file-format">Output file format</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#causal-effect-estimation-and-validation">Causal effect estimation and validation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#estimation">Estimation</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#validation">Validation</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inuput-file-format">Inuput file format</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Output file format</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Input</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptwas-est">ptwas_est</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Output</a></li>
</ul>
</li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="mendelian-randomization-using-fine-mapping-iv">
<h1>Mendelian Randomization using fine-mapping IV<a class="headerlink" href="#mendelian-randomization-using-fine-mapping-iv" title="Permalink to this heading">#</a></h1>
<p>The procedures for PTWAS estimation and model validation that are implemented in the R script. The unit of analysis is a single gene-trait pair, whose information is summarized in a text file. First, we generate the units of analysis. Then, we conduct the causal effect estimation and validation procedure.</p>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Organize the eQTL fine-mapping results, eQTL summary statistics, and GWAS summary statistics data into a format that can be processed by ptwas_est.</p></li>
<li><p>run ptwas_est with R codes: causal effect estimation and validation.</p></li>
</ul>
<section id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>eQTL finemapping result:</p>
<ul>
<li><p>This result contain the information of credible sets and pip. <br/>
<br/></p></li>
</ul>
</li>
<li><p>eqtl sumstats: se and beta</p></li>
<li><p>gwas sumstats: se and beta</p></li>
</ul>
</section>
<section id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>data format for ptwas_est</p></li>
<li><p>ptwas_est resuls</p></li>
</ul>
</section>
</section>
<section id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this heading">#</a></h2>
<section id="generate-the-units-of-analysis">
<h3>Generate the units of analysis<a class="headerlink" href="#generate-the-units-of-analysis" title="Permalink to this heading">#</a></h3>
<section id="id1">
<h4>Input<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h4>
<ul class="simple">
<li><p><strong>eQTL finemapping</strong>: /mnt/vast/hpc/csg/rf2872/Work/INTACT/fastenloc_test/eqtl_susie_table.txt</p></li>
<li><p><strong>eQTL sumstats</strong>: /mnt/vast/hpc/csg/ftp_lisanwanglab_sync/ftp_fgc_xqtl/projects/rna-seq/BU/ROSMAP_DLPFC/eQTL/association_scan/norminal_qced_files</p></li>
<li><p><strong>GWAS sumstats</strong>: /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022</p></li>
</ul>
</section>
<section id="output-file-format">
<h4>Output file format<a class="headerlink" href="#output-file-format" title="Permalink to this heading">#</a></h4>
<p>Each row of the file corresponds to a single SNP that is inferred as a memeber of a signal cluster, the columns are:</p>
<ul class="simple">
<li><p>snp: SNP name</p></li>
<li><p>beta_eQTL: eQTL effect</p></li>
<li><p>se_eQTL: standard error of estimated eQTL effect</p></li>
<li><p>beta_GWAS: GWAS effect</p></li>
<li><p>se_GWAS: standard error of GWAS effect</p></li>
<li><p>cluster: Signal cluster ID (credible sets index)</p></li>
<li><p>pip: SNP posterior inclusiion probability (PIP)</p></li>
<li><p>gene_id: gene name</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#######organize the data of eqtl finemapping, eqtl summary statistics and GWAS summary statistics</span>
<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>

<span class="c1">#####load the the eqtl finemapping folders</span>
<span class="n">eqtl_table_list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">fread</span><span class="p">(</span><span class="s">&quot;/mnt/vast/hpc/csg/rf2872/Work/INTACT/fastenloc_test/eqtl_susie_table.txt&quot;</span><span class="p">)</span>
<span class="n">finemapping_folder</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/susie_per_gene_tad/cache&quot;</span>

<span class="c1">#####load the eqtl sumstats</span>
<span class="n">eqtl_folder</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;/mnt/vast/hpc/csg/ftp_lisanwanglab_sync/ftp_fgc_xqtl/projects/rna-seq/BU/ROSMAP_DLPFC/eQTL/association_scan/norminal_qced_files&quot;</span>
<span class="c1">#####load the GWAS sumstats</span>
<span class="n">GWAS_folder</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;/mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022&quot;</span>

<span class="n">gene_name_list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eqtl_table_list</span><span class="o">$</span><span class="n">molecular_trait_i</span>
<span class="n">eqtl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">k</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">){</span>
<span class="n">gene_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_name_list</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
<span class="n">eqtl_finemapp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">finemapping_folder</span><span class="p">,</span><span class="s">&quot;/demo.&quot;</span><span class="p">,</span><span class="n">gene_name</span><span class="p">,</span><span class="s">&quot;.unisusie.fit.rds&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
<span class="n">cluster_num</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">sets</span><span class="o">$</span><span class="n">cs_index</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cluster_num</span><span class="p">)){</span>
<span class="w">    </span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">str_sub</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="nf">str_split</span><span class="p">(</span><span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">variable_name</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">1</span><span class="p">]),</span><span class="m">4</span><span class="p">)</span>
<span class="w">    </span><span class="n">eqtl_data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">fread</span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">eqtl_folder</span><span class="p">,</span><span class="s">&quot;/dlpfc_batch_all.rnaseqc.low_expression_filtered.outlier_removed.tmm.expression.bed.processed_phenotype.per_chrom_dlpfc_batch_all.rnaseqc.ROSMAP_covariates.ROSMAP_NIA_WGS.pca.PEER.txt.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;.norminal.cis_long_table.txt&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
<span class="w">    </span><span class="n">GWAS_data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">fread</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">GWAS_folder</span><span class="p">,</span><span class="s">&quot;/ADGWAS_Bellenguez_2022.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;/ADGWAS2022.chr&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;.sumstat.tsv&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
<span class="w">    </span><span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">cluster_num</span><span class="p">)){</span>
<span class="w">       </span><span class="n">cluster_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cluster_num</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="w">       </span><span class="n">cluster_id_name</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;L&quot;</span><span class="p">,</span><span class="n">cluster_id</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="w">       </span><span class="n">crediblesets</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">sets</span><span class="o">$</span><span class="n">cs</span><span class="p">[[</span><span class="n">cluster_id_name</span><span class="p">]]</span>
<span class="w">       </span><span class="n">variants_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">variable_name</span><span class="p">[</span><span class="n">crediblesets</span><span class="p">]</span>
<span class="w">       </span><span class="c1">#####get the chromosome fo the variants_name</span>
<span class="w">       </span><span class="n">chr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">str_sub</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="nf">str_split</span><span class="p">(</span><span class="n">variants_name</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">1</span><span class="p">]),</span><span class="m">4</span><span class="p">)</span>
<span class="w">       </span><span class="n">pip</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">eqtl_finemapp</span><span class="o">$</span><span class="n">dlpfc_eqtl</span><span class="o">$</span><span class="n">pip</span><span class="p">[</span><span class="n">crediblesets</span><span class="p">])</span>
<span class="w">       </span><span class="n">eqtl_cluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">tibble</span><span class="p">(</span><span class="n">snp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">variants_name</span><span class="p">,</span><span class="w"> </span><span class="n">cluster</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="n">cluster_id</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">variants_name</span><span class="p">)),</span><span class="n">pip</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pip</span><span class="p">,</span><span class="n">gene_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="nf">length</span><span class="p">(</span><span class="n">variants_name</span><span class="p">)))</span>
<span class="w">       </span><span class="n">eqtl_beta_se</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eqtl_data</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">molecular_trait_id</span><span class="o">%in%</span><span class="n">gene_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">variant</span><span class="o">%in%</span><span class="n">variants_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span><span class="n">beta</span><span class="p">,</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;snp&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;variant&quot;</span><span class="p">,</span><span class="s">&quot;beta_eQTL&quot;</span><span class="o">=</span><span class="s">&quot;beta&quot;</span><span class="p">,</span><span class="s">&quot;se_eQTL&quot;</span><span class="o">=</span><span class="s">&quot;se&quot;</span><span class="p">)</span>
<span class="w">       </span><span class="n">GWAS_beta_se</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GWAS_data</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">snp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sub</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">,</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="n">variant</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">snp</span><span class="o">%in%</span><span class="n">variants_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">snp</span><span class="p">,</span><span class="n">beta</span><span class="p">,</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;beta_GWAS&quot;</span><span class="o">=</span><span class="s">&quot;beta&quot;</span><span class="p">,</span><span class="s">&quot;se_GWAS&quot;</span><span class="o">=</span><span class="s">&quot;se&quot;</span><span class="p">)</span>
<span class="w">       </span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">eqtl_beta_se</span><span class="p">,</span><span class="w"> </span><span class="n">GWAS_beta_se</span><span class="p">,</span><span class="w"> </span><span class="n">eqtl_cluster</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">reduce</span><span class="p">(</span><span class="n">full_join</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;snp&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">drop_na</span><span class="p">()</span>
<span class="w">       </span><span class="n">eqtl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rbind</span><span class="p">(</span><span class="n">eqtl</span><span class="p">,</span><span class="w"> </span><span class="n">df</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span><span class="nf">rm</span><span class="p">(</span><span class="n">eqtl_data</span><span class="p">)</span>
<span class="w">    </span><span class="nf">rm</span><span class="p">(</span><span class="n">GWAS_data</span><span class="p">)</span>
<span class="w">    </span><span class="nf">fwrite</span><span class="p">(</span><span class="n">eqtl</span><span class="p">,</span><span class="n">file</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;/home/aw3600/PTWAS/PTWAS_test_data.txt&quot;</span><span class="p">)</span>
<span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="causal-effect-estimation-and-validation">
<h3>Causal effect estimation and validation<a class="headerlink" href="#causal-effect-estimation-and-validation" title="Permalink to this heading">#</a></h3>
<section id="estimation">
<h4>Estimation<a class="headerlink" href="#estimation" title="Permalink to this heading">#</a></h4>
<ul class="simple">
<li><p>First, we select strong eQTLs by thresholding on the corresponding signal-level PIPs (cpip = 0.5).</p></li>
<li><p>Then, we aggregate the estimates from the individual eQTLs using a fixed-effect meta-analysis procedure.</p></li>
</ul>
</section>
<section id="validation">
<h4>Validation<a class="headerlink" href="#validation" title="Permalink to this heading">#</a></h4>
<p>Identify severe violations of the exclusion restriction (ER) assumption and exclude them from the downstream estimation procedure.</p>
</section>
<section id="inuput-file-format">
<h4>Inuput file format<a class="headerlink" href="#inuput-file-format" title="Permalink to this heading">#</a></h4>
<p>The output from the first step</p>
</section>
<section id="id2">
<h4>Output file format<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h4>
<p>Each row of the file corresponds to a gene and the columns are:</p>
<ul class="simple">
<li><p>gene_id: gene name</p></li>
<li><p>num_cluster: the number of credible sets of gene</p></li>
<li><p>num_instruments: the number of instruments included in the gene</p></li>
<li><p>spip: the sum of pip for credible sets of each gene</p></li>
<li><p>grp_beta: signal-level estimates (combine the SNP-level estimates from the member SNPs weighted by their corresponding PIPs)</p></li>
<li><p>grp_se: standard error of the signal-level estimates</p></li>
<li><p>meta: gene-level estimate of the causal effect (aggregate the signal-level estimates by applying a fixed-effect meta-analysis model)</p></li>
<li><p>se_meta: standard error of the gene-level estimate of the causal effect</p></li>
<li><p>Q: Cochran’s Q statistic</p></li>
<li><p>I2: <span class="math notranslate nohighlight">\(I^2\)</span> statistics (quantify the heterogeneity, range from 0 to 1, <span class="math notranslate nohighlight">\(I^2 \rightarrow 0\)</span> indicates reasonable consistency among estimates by multiple eQTLs, whereas <span class="math notranslate nohighlight">\(I^2 \rightarrow 1\)</span> 1 implies severe departures from exclusion restriction (ER).)</p></li>
</ul>
</section>
<section id="id3">
<h4>Input<a class="headerlink" href="#id3" title="Permalink to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">formatted_input</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">fread</span><span class="p">(</span><span class="s">&quot;/home/aw3600/PTWAS/PTWAS_test_data.txt&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">formatted_input</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.table: 6 × 8</caption>
<thead>
	<tr><th scope=col>snp</th><th scope=col>beta_eQTL</th><th scope=col>se_eQTL</th><th scope=col>beta_GWAS</th><th scope=col>se_GWAS</th><th scope=col>cluster</th><th scope=col>pip</th><th scope=col>gene_id</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
	<tr><td>chr1:169880000_A_G</td><td>-0.2883383</td><td>0.03356421</td><td>0.0031</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169880823_G_A</td><td>-0.2883383</td><td>0.03356421</td><td>0.0034</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169880877_T_C</td><td>-0.2883383</td><td>0.03356421</td><td>0.0033</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169882224_A_G</td><td>-0.2859217</td><td>0.03388142</td><td>0.0024</td><td>0.0138</td><td>1</td><td>0.003048243</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169882228_G_A</td><td>-0.2859217</td><td>0.03388142</td><td>0.0023</td><td>0.0138</td><td>1</td><td>0.003048243</td><td>ENSG00000000457</td></tr>
	<tr><td>chr1:169882581_G_T</td><td>-0.2883383</td><td>0.03356421</td><td>0.0025</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
<section id="ptwas-est">
<h4>ptwas_est<a class="headerlink" href="#ptwas-est" title="Permalink to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#####define the threshold of the cumulative cluster pip</span>
<span class="n">cpip</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span>

<span class="c1">######heterogeneity:  calculate I2 statistics based on the Cochran&#39;s Q statistic</span>
<span class="n">calc_I2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">Q</span><span class="p">,</span><span class="w"> </span><span class="n">Est</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">Q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Q</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span>
<span class="w">    </span><span class="n">Est</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">Est</span><span class="p">))</span>
<span class="w">    </span><span class="n">I2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">if </span><span class="p">(</span><span class="n">Q</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1e-3</span><span class="p">)</span><span class="w"> </span><span class="p">(</span><span class="n">Q</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">Est</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="o">/</span><span class="n">Q</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="m">0</span>
<span class="w">    </span><span class="nf">return</span><span class="p">(</span><span class="nf">if </span><span class="p">(</span><span class="n">I2</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="n">I2</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1">#####ptwas_est</span>
<span class="n">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">formatted_input</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">        </span><span class="n">beta_eQTL</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">beta_eQTL</span><span class="o">/</span><span class="n">se_eQTL</span><span class="p">,</span>
<span class="w">        </span><span class="n">se_eQTL</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">group_by</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span><span class="w"> </span><span class="n">cluster</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span><span class="n">spip</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">pip</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">filter</span><span class="p">(</span><span class="n">spip</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">cpip</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># Cumulative cluster pip greater than a user defined cumulative pip threshold</span>
<span class="w">    </span><span class="nf">group_by</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span><span class="w"> </span><span class="n">cluster</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">        </span><span class="n">beta_yx</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">beta_GWAS</span><span class="o">/</span><span class="n">beta_eQTL</span><span class="p">,</span>
<span class="w">        </span><span class="n">se_yx</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span>
<span class="w">            </span><span class="p">(</span><span class="n">se_GWAS</span><span class="o">^</span><span class="m">2</span><span class="o">/</span><span class="n">beta_eQTL</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="p">((</span><span class="n">beta_GWAS</span><span class="o">^</span><span class="m">2</span><span class="o">*</span><span class="n">se_eQTL</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="o">/</span><span class="n">beta_eQTL</span><span class="o">^</span><span class="m">4</span><span class="p">)),</span>
<span class="w">        </span><span class="n">grp_beta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">((</span><span class="n">beta_yx</span><span class="o">*</span><span class="n">pip</span><span class="p">)</span><span class="o">/</span><span class="n">spip</span><span class="p">),</span>
<span class="w">        </span><span class="n">grp_se</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">((</span><span class="n">beta_yx</span><span class="o">^</span><span class="m">2</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">se_yx</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="o">*</span><span class="n">pip</span><span class="o">/</span><span class="n">spip</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">        </span><span class="n">grp_se</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="n">grp_se</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">grp_beta</span><span class="o">^</span><span class="m">2</span><span class="p">),</span>
<span class="w">        </span><span class="n">wv</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">grp_se</span><span class="o">^</span><span class="m">-2</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">ungroup</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">group_by</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">        </span><span class="n">meta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">wv</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">grp_beta</span><span class="p">)),</span>
<span class="w">        </span><span class="n">sum_w</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">wv</span><span class="p">)),</span>
<span class="w">        </span><span class="n">se_meta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="n">sum_w</span><span class="o">^</span><span class="m">-1</span><span class="p">),</span>
<span class="w">        </span><span class="n">num_cluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">cluster</span><span class="p">)))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">        </span><span class="n">num_instruments</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">snp</span><span class="p">),</span>
<span class="w">        </span><span class="n">meta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">meta</span><span class="o">/</span><span class="n">sum_w</span><span class="p">,</span>
<span class="w">        </span><span class="c1">######sum(unique(wv)*(unique(grp_beta) - unique(meta))^2)</span>
<span class="w">        </span><span class="n">Q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">wv</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">grp_beta</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">meta</span><span class="p">))</span><span class="o">^</span><span class="m">2</span><span class="p">),</span>
<span class="w">        </span><span class="n">I2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">calc_I2</span><span class="p">(</span><span class="n">Q</span><span class="p">,</span><span class="w"> </span><span class="n">grp_beta</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">        </span><span class="nf">ungroup</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">distinct</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span><span class="w"> </span><span class="n">.keep_all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">        </span><span class="n">spip</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">spip</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span>
<span class="w">        </span><span class="n">grp_beta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">grp_beta</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span>
<span class="w">        </span><span class="n">meta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span>
<span class="w">        </span><span class="n">se_meta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">se_meta</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span>
<span class="w">        </span><span class="n">Q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">Q</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span>
<span class="w">        </span><span class="n">I2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">I2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">    </span><span class="nf">select</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span><span class="w"> </span><span class="n">num_cluster</span><span class="p">,</span><span class="w"> </span><span class="n">num_instruments</span><span class="p">,</span><span class="w"> </span><span class="n">spip</span><span class="p">,</span><span class="w"> </span><span class="n">grp_beta</span><span class="p">,</span><span class="w"> </span><span class="n">grp_se</span><span class="p">,</span><span class="w"> </span><span class="n">meta</span><span class="p">,</span><span class="w"> </span><span class="n">se_meta</span><span class="p">,</span><span class="w"> </span><span class="n">Q</span><span class="p">,</span><span class="w"> </span><span class="n">I2</span><span class="p">)</span>
<span class="w">    </span><span class="nf">fwrite</span><span class="p">(</span><span class="n">output</span><span class="p">,</span><span class="n">file</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;/home/aw3600/PTWAS/PTWAS_est_data.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="id4">
<h4>Output<a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">output</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 51 × 10</caption>
<thead>
	<tr><th scope=col>gene_id</th><th scope=col>num_cluster</th><th scope=col>num_instruments</th><th scope=col>spip</th><th scope=col>grp_beta</th><th scope=col>grp_se</th><th scope=col>meta</th><th scope=col>se_meta</th><th scope=col>Q</th><th scope=col>I2</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>ENSG00000000457</td><td>2</td><td> 74</td><td>0.696</td><td> 0.000</td><td>0.0016118763</td><td>-0.001</td><td>0.001</td><td>0.157</td><td>0.000</td></tr>
	<tr><td>ENSG00000000971</td><td>1</td><td> 22</td><td>0.739</td><td>-0.001</td><td>0.0010414938</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001084</td><td>1</td><td> 10</td><td>0.881</td><td> 0.000</td><td>0.0010095378</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001167</td><td>1</td><td> 16</td><td>0.737</td><td> 0.009</td><td>0.0029981468</td><td> 0.009</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001460</td><td>1</td><td> 16</td><td>0.745</td><td>-0.001</td><td>0.0006707913</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001461</td><td>1</td><td> 47</td><td>0.679</td><td>-0.005</td><td>0.0028313411</td><td>-0.005</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001561</td><td>1</td><td>  9</td><td>0.586</td><td> 0.001</td><td>0.0006746664</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001626</td><td>1</td><td> 28</td><td>0.801</td><td> 0.001</td><td>0.0011200443</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000001629</td><td>2</td><td> 48</td><td>0.678</td><td> 0.000</td><td>0.0019744363</td><td> 0.000</td><td>0.001</td><td>0.046</td><td>0.000</td></tr>
	<tr><td>ENSG00000001630</td><td>1</td><td> 67</td><td>0.807</td><td> 0.001</td><td>0.0011870369</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000002822</td><td>2</td><td> 19</td><td>0.926</td><td>-0.001</td><td>0.0030673516</td><td> 0.001</td><td>0.002</td><td>0.897</td><td>0.000</td></tr>
	<tr><td>ENSG00000002834</td><td>1</td><td>  1</td><td>0.999</td><td>-0.001</td><td>0.0012142297</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000002933</td><td>2</td><td> 40</td><td>0.985</td><td>-0.001</td><td>0.0013062166</td><td> 0.001</td><td>0.001</td><td>1.257</td><td>0.205</td></tr>
	<tr><td>ENSG00000003137</td><td>1</td><td> 21</td><td>0.595</td><td> 0.001</td><td>0.0014258335</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003147</td><td>1</td><td>  8</td><td>0.901</td><td>-0.003</td><td>0.0015701290</td><td>-0.003</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003249</td><td>2</td><td> 17</td><td>0.688</td><td> 0.002</td><td>0.0020332468</td><td> 0.002</td><td>0.001</td><td>0.096</td><td>0.000</td></tr>
	<tr><td>ENSG00000003393</td><td>1</td><td>  5</td><td>0.901</td><td> 0.002</td><td>0.0011071698</td><td> 0.002</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003402</td><td>1</td><td>  1</td><td>0.998</td><td> 0.002</td><td>0.0020686803</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003509</td><td>1</td><td> 14</td><td>0.582</td><td>-0.001</td><td>0.0012611594</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003987</td><td>1</td><td>  5</td><td>0.828</td><td> 0.001</td><td>0.0044738629</td><td> 0.001</td><td>0.004</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000003989</td><td>1</td><td>  3</td><td>0.985</td><td> 0.000</td><td>0.0007597256</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004142</td><td>1</td><td>  3</td><td>0.985</td><td> 0.000</td><td>0.0005755454</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004399</td><td>3</td><td>123</td><td>0.624</td><td> 0.003</td><td>0.0014414321</td><td> 0.003</td><td>0.001</td><td>0.489</td><td>0.000</td></tr>
	<tr><td>ENSG00000004455</td><td>1</td><td> 72</td><td>0.755</td><td> 0.002</td><td>0.0015330947</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004487</td><td>1</td><td> 12</td><td>0.550</td><td>-0.001</td><td>0.0013412799</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004534</td><td>1</td><td> 28</td><td>0.871</td><td> 0.000</td><td>0.0003097528</td><td> 0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004700</td><td>1</td><td> 16</td><td>0.737</td><td> 0.000</td><td>0.0025675503</td><td> 0.000</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000004766</td><td>1</td><td>  1</td><td>0.957</td><td> 0.001</td><td>0.0035742160</td><td> 0.001</td><td>0.004</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005020</td><td>1</td><td>107</td><td>0.628</td><td> 0.000</td><td>0.0008228963</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005059</td><td>2</td><td> 48</td><td>0.684</td><td>-0.001</td><td>0.0018060197</td><td> 0.000</td><td>0.002</td><td>2.463</td><td>0.594</td></tr>
	<tr><td>ENSG00000005075</td><td>1</td><td> 11</td><td>0.550</td><td> 0.002</td><td>0.0019552991</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005108</td><td>2</td><td> 36</td><td>0.604</td><td> 0.002</td><td>0.0013832122</td><td> 0.001</td><td>0.001</td><td>0.217</td><td>0.000</td></tr>
	<tr><td>ENSG00000005156</td><td>2</td><td> 30</td><td>0.890</td><td> 0.000</td><td>0.0016157681</td><td> 0.000</td><td>0.001</td><td>0.002</td><td>0.000</td></tr>
	<tr><td>ENSG00000005189</td><td>1</td><td> 88</td><td>0.783</td><td> 0.001</td><td>0.0026267194</td><td> 0.001</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005206</td><td>1</td><td>  7</td><td>0.949</td><td> 0.000</td><td>0.0015403997</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005249</td><td>1</td><td> 68</td><td>0.898</td><td> 0.000</td><td>0.0032628529</td><td> 0.000</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005436</td><td>1</td><td>  2</td><td>0.986</td><td> 0.001</td><td>0.0016112891</td><td> 0.001</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005448</td><td>1</td><td>  2</td><td>0.770</td><td> 0.007</td><td>0.0033168502</td><td> 0.007</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005471</td><td>1</td><td>  2</td><td>1.000</td><td> 0.000</td><td>0.0015087191</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005700</td><td>1</td><td> 19</td><td>0.537</td><td>-0.001</td><td>0.0029638984</td><td>-0.001</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000005981</td><td>2</td><td> 21</td><td>0.848</td><td> 0.001</td><td>0.0014226597</td><td> 0.001</td><td>0.001</td><td>0.598</td><td>0.000</td></tr>
	<tr><td>ENSG00000006016</td><td>1</td><td>  2</td><td>0.832</td><td> 0.001</td><td>0.0008338203</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006042</td><td>2</td><td> 15</td><td>0.730</td><td> 0.000</td><td>0.0009331976</td><td> 0.000</td><td>0.001</td><td>0.080</td><td>0.000</td></tr>
	<tr><td>ENSG00000006116</td><td>1</td><td>  5</td><td>0.919</td><td>-0.004</td><td>0.0018333447</td><td>-0.004</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006125</td><td>1</td><td> 79</td><td>0.788</td><td>-0.001</td><td>0.0014991831</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006194</td><td>1</td><td> 23</td><td>0.605</td><td> 0.001</td><td>0.0013802571</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006282</td><td>2</td><td>  6</td><td>1.000</td><td>-0.001</td><td>0.0005764921</td><td> 0.000</td><td>0.000</td><td>2.708</td><td>0.631</td></tr>
	<tr><td>ENSG00000006283</td><td>1</td><td>  7</td><td>0.853</td><td> 0.002</td><td>0.0016798225</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006453</td><td>1</td><td> 76</td><td>0.585</td><td>-0.001</td><td>0.0010935407</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006459</td><td>1</td><td> 50</td><td>0.704</td><td> 0.000</td><td>0.0018713348</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
	<tr><td>ENSG00000006555</td><td>1</td><td> 26</td><td>0.776</td><td> 0.000</td><td>0.0024193024</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./code/pecotmr_integration"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="ptwas.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">TWAS for xQTL</p>
      </div>
    </a>
    <a class="right-next"
       href="intact.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">TWAS and COLOC integration</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-the-units-of-analysis">Generate the units of analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Input</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#output-file-format">Output file format</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#causal-effect-estimation-and-validation">Causal effect estimation and validation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#estimation">Estimation</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#validation">Validation</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inuput-file-format">Inuput file format</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Output file format</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Input</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptwas-est">ptwas_est</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Output</a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>