

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>GWAS integration: enrichment and colocalization &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/pecotmr_integration/SuSiE_enloc';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="GWAS integration: TWAS and MR" href="twas_mr.html" />
    <link rel="prev" title="The PECOTMR intergration framework" href="pecotmr.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced Genome-wide Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/polyfun.html">Fine-mapping with PolyFun</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/MRAID_QTL.html">Mendelian Randomization using MRAID</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="pecotmr.html">The PECOTMR intergration framework</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/edit/gh-pages/code/pecotmr_integration/SuSiE_enloc.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_integration/SuSiE_enloc.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/pecotmr_integration/SuSiE_enloc.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>GWAS integration: enrichment and colocalization</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analytical-logic">Analytical Logic</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="gwas-integration-enrichment-and-colocalization">
<h1>GWAS integration: enrichment and colocalization<a class="headerlink" href="#gwas-integration-enrichment-and-colocalization" title="Permalink to this heading">#</a></h1>
<p>This workflow processes fine-mapping results for xQTL, generated by <code class="docutils literal notranslate"><span class="pre">susie_twas</span></code> in the <code class="docutils literal notranslate"><span class="pre">mnm_regression.ipynb</span></code> notebook for cis xQTL, and GWAS fine-mapping results produced by <code class="docutils literal notranslate"><span class="pre">susie_rss</span></code> in the <code class="docutils literal notranslate"><span class="pre">rss_analysis.ipynb</span></code> notebook. It is designed to perform enrichment and colocalization analysis, particularly when fine-mapping results originate from different regions in the case of cis-xQTL and GWAS. The pipeline is capable to integrate and analyze data across these distinct regions. Originally tailored for cis-xQTL and GWAS integration, this pipeline can be applied to other pairwise integrations. An example of such application is in trans analysis, where the fine-mapped regions might be identical between trans-xQTL and GWAS, representing a special case of this broader implementation.</p>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<p>Lists of SuSiE fine-mapping output objects, in RDS format, of <code class="docutils literal notranslate"><span class="pre">class(susie)</span></code> in R.</p>
<ul class="simple">
<li><p>For xQTL the list is meta-data of format: <code class="docutils literal notranslate"><span class="pre">chr</span></code>, <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">end</span></code>, <code class="docutils literal notranslate"><span class="pre">original_data</span></code>, <code class="docutils literal notranslate"><span class="pre">conditions_top_loci</span></code>, <code class="docutils literal notranslate"><span class="pre">block_top_loci</span></code> where <code class="docutils literal notranslate"><span class="pre">original_data</span></code> is an RDS file, <code class="docutils literal notranslate"><span class="pre">conditions_top_loci</span></code> is showing which contexts have top loci table (potential signals) <code class="docutils literal notranslate"><span class="pre">block_top_loci</span></code> is the blocks have overlapped top loci variant with xQTL data. That file could be output from <code class="docutils literal notranslate"><span class="pre">fine_mapping_post_processing/overlap_qtl_gwas</span></code>.</p></li>
<li><p>For GWAS the list is meta-data of format: <code class="docutils literal notranslate"><span class="pre">chr</span></code>, <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">end</span></code>, <code class="docutils literal notranslate"><span class="pre">original_data</span></code>, <code class="docutils literal notranslate"><span class="pre">block_top_loci...</span></code> where <code class="docutils literal notranslate"><span class="pre">original_data</span></code> is an RDS file. That file could be output from <code class="docutils literal notranslate"><span class="pre">fine_mapping_post_processing/gwas_results_export</span></code>.</p></li>
<li><p>Context meta is a metafile that shows the analysis_names and the contained contexts. It can be used to identify the corresponding raw data for each context.</p></li>
</ul>
</section>
<section id="analytical-logic">
<h2>Analytical Logic<a class="headerlink" href="#analytical-logic" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Enrichment</p></li>
</ul>
<ol class="arabic simple">
<li><p><strong>Identify GWAS Blocks:</strong> Select GWAS blocks with top loci from Bellenguez data and using a single variant regression method. Locate the original filenames (<code class="docutils literal notranslate"><span class="pre">original_data</span></code>) for these blocks and map the analysis regions to identify <strong>overlapping gene analysis regions</strong> and corresponding QTL files with top loci table (<code class="docutils literal notranslate"><span class="pre">condition_top_loci</span></code>).</p></li>
<li><p><strong>Contexts in xQTL Meta-data:</strong> Find contexts within the xQTL meta-data that include top loci results for each gene. Retrieve the corresponding original xQTL files (<code class="docutils literal notranslate"><span class="pre">original_data</span></code>) by referencing the context meta-data.</p></li>
<li><p><strong>Execute xQTL GWAS Enrichment:</strong> Perform <code class="docutils literal notranslate"><span class="pre">xqtl_gwas_enrichment</span></code> analysis to generate enrichment parameters (a0, a1) and calculate priors (p1, p2, p12).</p></li>
</ol>
<ul class="simple">
<li><p>Coloc</p></li>
</ul>
<ol class="arabic simple">
<li><p><strong>xQTL Meta-data Contexts:</strong> Within the xQTL meta-data, identify contexts that include top loci results for each gene, indicated by <code class="docutils literal notranslate"><span class="pre">condition_top_loci</span></code>. Retrieve the corresponding original xQTL files (<code class="docutils literal notranslate"><span class="pre">original_data</span></code>) by referring back to the context meta-data. (Optional) Subset the QTL table with specifies gene list.</p></li>
<li><p><strong>Original GWAS Files:</strong> Identify GWAS blocks that contain <strong>overlapping top loci variants</strong> for each gene (<code class="docutils literal notranslate"><span class="pre">block_top_loci</span></code>) in above file. Utilize the GWAS list to locate the original filenames (<code class="docutils literal notranslate"><span class="pre">original_data</span></code>) for the identified GWAS block candidates.</p></li>
<li><p><strong>Enrichment and Coloc Analysis:</strong> Automatically execute <code class="docutils literal notranslate"><span class="pre">xqtl_gwas_enrichment</span></code> to enable enrichment analysis, generating priors for subsequent coloc analysis through logic defined in <code class="docutils literal notranslate"><span class="pre">enrichment</span></code>. Then, apply <code class="docutils literal notranslate"><span class="pre">susie_coloc</span></code> for coloc analysis on the selected xQTL and GWAS original files, analyzing each gene under each identified condition.</p></li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Enrichment analysis results — this is a global enrichment estimate that combines all input data for each context.</p></li>
<li><p>Colocalization results for regions of interest — if <code class="docutils literal notranslate"><span class="pre">ld_meta_file_path</span></code> is provided, would also report coloc cs by coloc postprocessing.</p></li>
</ol>
</section>
<section id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this heading">#</a></h2>
<p>enrichment</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE_enloc</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">xqtl_gwas_enrichment</span>   \
    <span class="o">--</span><span class="n">gwas_meta_data</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">pecotmr</span><span class="o">/</span><span class="n">encoloc_test</span><span class="o">/</span><span class="n">gwas</span><span class="o">.</span><span class="n">block_results_db</span><span class="o">.</span><span class="n">tsv</span>     \
    <span class="o">--</span><span class="n">xqtl_meta_data</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">gwas</span><span class="o">/</span><span class="n">overlap_test</span><span class="o">/</span><span class="n">ROSMAP_eQTL</span><span class="o">.</span><span class="n">overlapped</span><span class="o">.</span><span class="n">gwas</span><span class="o">.</span><span class="n">tsv</span>         \
    <span class="o">--</span><span class="n">xqtl_finemapping_obj</span> <span class="n">preset_variants_result</span> <span class="n">susie_result_trimmed</span>              \
    <span class="o">--</span><span class="n">xqtl_varname_obj</span> <span class="n">preset_variants_result</span> <span class="n">variant_names</span>             \
    <span class="o">--</span><span class="n">gwas_finemapping_obj</span> <span class="n">AD_Bellenguez_2022</span> <span class="n">single_effect_regression</span> <span class="n">susie_result_trimmed</span>          \
    <span class="o">--</span><span class="n">gwas_varname_obj</span>  <span class="n">AD_Bellenguez_2022</span> <span class="n">single_effect_regression</span> <span class="n">variant_names</span>         \
    <span class="o">--</span><span class="n">xqtl_region_obj</span>  <span class="n">region_info</span>   <span class="n">grange</span>   \
    <span class="o">--</span><span class="n">qtl</span><span class="o">-</span><span class="n">path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">susie_2024_new</span><span class="o">/</span><span class="n">rds_files</span>         \
    <span class="o">--</span><span class="n">gwas_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">hs3393</span><span class="o">/</span><span class="n">RSS_QC</span><span class="o">/</span><span class="n">GWAS_finemapping_Feb22_7dataset</span><span class="o">/</span><span class="n">SuSiE_RSS</span> \
    <span class="o">--</span><span class="n">context_meta</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">homes</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">codes</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">resource</span><span class="o">/</span><span class="n">context_meta</span><span class="o">.</span><span class="n">tsv</span> 
</pre></div>
</div>
<p>coloc (would trigger enrichment automatically)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE_enloc</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_coloc</span>    \
    <span class="o">--</span><span class="n">gwas_meta_data</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">pecotmr</span><span class="o">/</span><span class="n">encoloc_test</span><span class="o">/</span><span class="n">gwas</span><span class="o">.</span><span class="n">block_results_db</span><span class="o">.</span><span class="n">tsv</span>         \
    <span class="o">--</span><span class="n">xqtl_meta_data</span>  <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">gwas</span><span class="o">/</span><span class="n">overlap_test</span><span class="o">/</span><span class="n">ROSMAP_eQTL</span><span class="o">.</span><span class="n">overlapped</span><span class="o">.</span><span class="n">gwas</span><span class="o">.</span><span class="n">tsv</span>      \
    <span class="o">--</span><span class="n">xqtl_finemapping_obj</span> <span class="n">preset_variants_result</span> <span class="n">susie_result_trimmed</span>              \
    <span class="o">--</span><span class="n">xqtl_varname_obj</span> <span class="n">preset_variants_result</span> <span class="n">variant_names</span>             \
    <span class="o">--</span><span class="n">gwas_finemapping_obj</span> <span class="n">AD_Bellenguez_2022</span> <span class="n">single_effect_regression</span> <span class="n">susie_result_trimmed</span>              \
    <span class="o">--</span><span class="n">gwas_varname_obj</span>  <span class="n">AD_Bellenguez_2022</span> <span class="n">single_effect_regression</span> <span class="n">variant_names</span>             \
    <span class="o">--</span><span class="n">xqtl_region_obj</span>  <span class="n">region_info</span>   <span class="n">grange</span>       \
    <span class="o">--</span><span class="n">qtl</span><span class="o">-</span><span class="n">path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">susie_2024_new</span><span class="o">/</span><span class="n">rds_files</span>             \
    <span class="o">--</span><span class="n">gwas_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">hs3393</span><span class="o">/</span><span class="n">RSS_QC</span><span class="o">/</span><span class="n">GWAS_finemapping_Feb22_7dataset</span><span class="o">/</span><span class="n">SuSiE_RSS</span>     \
    <span class="o">--</span><span class="n">context_meta</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">homes</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">codes</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">resource</span><span class="o">/</span><span class="n">context_meta</span><span class="o">.</span><span class="n">tsv</span>      \
    <span class="o">--</span><span class="n">ld_meta_file_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">data_public</span><span class="o">/</span><span class="mi">20240120</span><span class="n">_ADSP_LD_matrix</span><span class="o">/</span><span class="n">ld_meta_file</span><span class="o">.</span><span class="n">tsv</span> 
</pre></div>
</div>
<ul class="simple">
<li><p>eg: <code class="docutils literal notranslate"><span class="pre">gwas_meta_data</span></code><br />
output from <code class="docutils literal notranslate"><span class="pre">fine_mapping_post_processing/gwas_results_export</span></code></p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr	start	end	region_id	TSS	original_data	combined_data	combined_data_sumstats	conditions	conditions_top_loci</span>
<span class="n">chr1</span>	<span class="mi">101384274</span>	<span class="mi">104443097</span>	<span class="mi">1_101384274</span><span class="o">-</span><span class="mi">104443097</span>	<span class="n">NA</span>	<span class="mi">1_101384274</span><span class="o">-</span><span class="mf">104443097.</span><span class="n">susie_rss</span><span class="o">.</span><span class="n">rds</span>	<span class="n">gwas</span><span class="mf">.1_101384274</span><span class="o">-</span><span class="mf">104443097.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">rds</span>	<span class="n">gwas</span><span class="mf">.1_101384274</span><span class="o">-</span><span class="mf">104443097.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export_sumstats</span><span class="o">.</span><span class="n">rds</span>	<span class="n">NA</span>	<span class="n">NA</span>
<span class="n">chr19</span>	<span class="mi">44935906</span>	<span class="mi">46842901</span>	<span class="mi">19_44935906</span><span class="o">-</span><span class="mi">46842901</span>	<span class="n">NA</span>	<span class="mi">19_44935906</span><span class="o">-</span><span class="mf">46842901.</span><span class="n">susie_rss</span><span class="o">.</span><span class="n">rds</span>	<span class="n">gwas</span><span class="mf">.19_44935906</span><span class="o">-</span><span class="mf">46842901.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">rds</span>	<span class="n">gwas</span><span class="mf">.19_44935906</span><span class="o">-</span><span class="mf">46842901.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export_sumstats</span><span class="o">.</span><span class="n">rds</span>	<span class="n">AD_Bellenguez_2022</span><span class="p">,</span><span class="n">AD_Jansen_2021</span><span class="p">,</span><span class="n">AD_Kunkle_Stage1_2019</span><span class="p">,</span><span class="n">AD_Wightman_Full_2021</span><span class="p">,</span><span class="n">AD_Wightman_Excluding23andMe_2021</span><span class="p">,</span><span class="n">AD_Wightman_ExcludingUKBand23andME_2021</span>	<span class="n">AD_Wightman_ExcludingUKBand23andME_2021</span><span class="o">.</span><span class="n">qc_impute</span><span class="p">,</span><span class="n">AD_Wightman_ExcludingUKBand23andME_2021</span><span class="o">.</span><span class="n">qc_only</span>	<span class="n">AD_Wightman_ExcludingUKBand23andME_2021</span><span class="o">.</span><span class="n">qc_impute</span>
</pre></div>
</div>
<ul class="simple">
<li><p>eg: <code class="docutils literal notranslate"><span class="pre">xqtl_meta_data</span></code></p></li>
</ul>
<p>for enrichment: output from <code class="docutils literal notranslate"><span class="pre">fine_mapping_post_processing/cis_results_export</span></code>; for coloc: output from <code class="docutils literal notranslate"><span class="pre">fine_mapping_post_processing/overlap_qtl_gwas</span></code>, the difference between them is without or with last 2 columns: <code class="docutils literal notranslate"><span class="pre">block_top_loci</span></code>, <code class="docutils literal notranslate"><span class="pre">final_combined_data</span></code> to document overlapped block info at variant level</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr	start	end	region_id	TSS	original_data	combined_data	combined_data_sumstats	conditions	conditions_top_loci	block_top_loci	final_combined_data</span>
<span class="n">chr1</span>	<span class="mi">167600000</span>	<span class="mi">171480000</span>	<span class="n">ENSG00000000457</span>	<span class="mi">169894266</span>	<span class="n">MSBB_eQTL</span><span class="o">.</span><span class="n">ENSG00000000457</span><span class="o">.</span><span class="n">univariate_susie_twas_weights</span><span class="o">.</span><span class="n">rds</span><span class="p">,</span> <span class="n">ROSMAP_Kellis_eQTL</span><span class="o">.</span><span class="n">ENSG00000000457</span><span class="o">.</span><span class="n">univariate_susie_twas_weights</span><span class="o">.</span><span class="n">rds</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000000457</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">rds</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000000457</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export_sumstats</span><span class="o">.</span><span class="n">rds</span>	<span class="n">BM_10_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_22_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_36_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_44_MSBB_eQTL</span><span class="p">,</span><span class="n">Ast_Kellis_eQTL</span>	<span class="n">BM_22_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_44_MSBB_eQTL</span>		
<span class="n">chr19</span>	<span class="mi">41840000</span>	<span class="mi">47960000</span>	<span class="n">ENSG00000130203</span>	<span class="mi">44905790</span>	<span class="n">ROSMAP_Kellis_eQTL</span><span class="o">.</span><span class="n">ENSG00000130203</span><span class="o">.</span><span class="n">univariate_susie_twas_weights</span><span class="o">.</span><span class="n">rds</span><span class="p">,</span> <span class="n">ROSMAP_Bennett_Klein_pQTL</span><span class="o">.</span><span class="n">ENSG00000130203</span><span class="o">.</span><span class="n">univariate_susie_twas_weights</span><span class="o">.</span><span class="n">rds</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000130203</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">rds</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000130203</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export_sumstats</span><span class="o">.</span><span class="n">rds</span>	<span class="n">Mic_Kellis_eQTL</span><span class="p">,</span><span class="n">DLPFC_Bennett_pQTL</span>	<span class="n">Mic_Kellis_eQTL</span>	<span class="mi">19_44935906</span><span class="o">-</span><span class="mi">46842901</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000130203</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">overlapped</span><span class="o">.</span><span class="n">gwas</span><span class="o">.</span><span class="n">rds</span>
<span class="n">chr20</span>	<span class="mi">49934867</span>	<span class="mi">53560000</span>	<span class="n">ENSG00000000419</span>	<span class="mi">50958554</span>	<span class="n">MSBB_eQTL</span><span class="o">.</span><span class="n">ENSG00000000419</span><span class="o">.</span><span class="n">univariate_susie_twas_weights</span><span class="o">.</span><span class="n">rds</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000000419</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">rds</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000000419</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export_sumstats</span><span class="o">.</span><span class="n">rds</span>	<span class="n">BM_10_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_22_MSBB_eQTL</span>	<span class="n">BM_22_MSBB_eQTL</span>	<span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">ENSG00000000419</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">overlapped</span><span class="o">.</span><span class="n">gwas</span><span class="o">.</span><span class="n">rds</span>
</pre></div>
</div>
<ul class="simple">
<li><p>eg: <code class="docutils literal notranslate"><span class="pre">context_meta</span></code></p></li>
</ul>
<p>should be updated as more analysis is done, we used analysis_name and context in this analysis</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">analysis_name</span>	<span class="n">cohort</span>	<span class="n">context</span>
<span class="n">KNIGHT_pQTL</span>	<span class="n">KNIGHT</span>	<span class="n">Knight_pQTL_brain</span><span class="p">,</span><span class="n">Knight_eQTL_brain</span>
<span class="n">MSBB_eQTL</span>	<span class="n">MSBB</span>	<span class="n">BM_10_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_22_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_36_MSBB_eQTL</span><span class="p">,</span><span class="n">BM_44_MSBB_eQTL</span>
<span class="n">MIGA_eQTL</span>	<span class="n">MIGA</span>	<span class="n">MiGA_GTS_eQTL</span><span class="p">,</span><span class="n">MiGA_SVZ_eQTL</span><span class="p">,</span><span class="n">MiGA_THA_eQTL</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Workdir</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># A list of file paths for fine-mapped GWAS results. </span>
<span class="kn">parameter:</span> <span class="n">gwas_meta_data</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># A list of file paths for fine-mapped xQTL results. </span>
<span class="kn">parameter:</span> <span class="n">xqtl_meta_data</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Optional: if a region list is provide the enrichment analysis will be focused on provided region. </span>
<span class="c1"># The LAST column of this list will contain the ID of regions to focus on</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Optional: if a region name is provided </span>
<span class="c1"># the analysis would be focused on the union of provides region list and region names</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># It is required to input the name of the analysis</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">xqtl_meta_data</span><span class="si">:</span><span class="s2">bnn</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">gwas_meta_data</span><span class="si">:</span><span class="s2">bnn</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">200</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5m&quot;</span>
<span class="c1"># Memory expected: quite large for enrichment analysis but small for xQTL colocalization</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1">#Optional table name in xQTL RDS files to get fine mapping results (eg &#39;susie_result_trimmed &#39;).</span>
<span class="kn">parameter:</span> <span class="n">xqtl_finemapping_obj</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1">#Optional table name in xQTL RDS files to get variable names (eg &#39; variant_names &#39;).</span>
<span class="kn">parameter:</span> <span class="n">xqtl_varname_obj</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1">#Optional table name in GWAS RDS files to get fine mapping results (eg &#39;AD_Bellenguez_2022 single_effect_regression susie_result_trimmed &#39;).</span>
<span class="kn">parameter:</span> <span class="n">gwas_finemapping_obj</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1">#Optional table name in GWAS RDS files to get variable names(eg &#39;AD_Bellenguez_2022 single_effect_regression variant_names &#39;).</span>
<span class="kn">parameter:</span> <span class="n">gwas_varname_obj</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1">#Optional table name in xQTL RDS files to get region info (eg &#39;susie_result_trimmed region_info grange &#39;).</span>
<span class="kn">parameter:</span> <span class="n">xqtl_region_obj</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1">#Optional table name in GWAS RDS files to get region info (eg &#39;AD_Bellenguez_2022 single_effect_regression region_info grange &#39;).</span>
<span class="kn">parameter:</span> <span class="n">gwas_region_obj</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1">#Directory path for GWAS orignal finemapping results </span>
<span class="kn">parameter:</span> <span class="n">gwas_path</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="c1">#Directory path for xQTL orignal finemapping results </span>
<span class="kn">parameter:</span> <span class="n">qtl_path</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="c1"># a meta file showing the context and corresponding analysis_name</span>
<span class="kn">parameter:</span> <span class="n">context_meta</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Optional: if a region list is provide the analysis will be focused on provided region. </span>
<span class="c1"># The LAST column of this list will contain the ID of regions to focus on</span>
<span class="c1"># Otherwise, all regions with both genotype and phenotype files will be analyzed</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Optional: if a region name is provided </span>
<span class="c1"># the analysis would be focused on the union of provides region list and region names</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">partition</span><span class="p">):</span>
    <span class="c1"># from itertools import accumulate</span>
    <span class="c1"># partition = [len(x) for x in partition]</span>
    <span class="c1"># Compute the cumulative sums once</span>
    <span class="c1"># cumsum_vector = list(accumulate(partition))</span>
    <span class="c1"># Use slicing based on the cumulative sums</span>
    <span class="c1"># return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]</span>
    <span class="k">return</span> <span class="n">partition</span>

<span class="k">def</span> <span class="nf">make_unique_data</span><span class="p">(</span><span class="n">row</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Make the GWAS data unique by removing duplicates.&quot;&quot;&quot;</span>
    <span class="n">unique_data</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">unique_data</span>

<span class="k">def</span> <span class="nf">generate_meta_dataframe</span><span class="p">(</span><span class="n">meta_data_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a new long metadata by converting to context:analysis_name (1:1).&quot;&quot;&quot;</span>
    <span class="n">meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">meta_data_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">new_meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">contexts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s1">&#39;context&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">stack</span><span class="p">()</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>

    <span class="k">for</span> <span class="n">context</span> <span class="ow">in</span> <span class="n">contexts</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[</span><span class="s1">&#39;context&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">context</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;context&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">context</span><span class="p">],</span>
            <span class="s1">&#39;analysis_name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span> <span class="s1">&#39;analysis_name&#39;</span><span class="p">])]</span>
        <span class="p">})</span>
        <span class="n">new_meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">new_meta</span><span class="p">,</span> <span class="n">tmp</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_meta</span>

<span class="k">def</span> <span class="nf">generate_condition_based_dataframe</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a new long table by converting to condition:orginal_data (1:1).&quot;&quot;&quot;</span>
    <span class="c1"># only consider the QTL contexts have top loci data</span>
    <span class="n">conditions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;conditions_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">stack</span><span class="p">()</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
    <span class="n">new_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="n">escaped_condition</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">(</span><span class="n">condition</span><span class="p">)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;conditions_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">escaped_condition</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;condition&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">condition</span><span class="p">],</span>
            <span class="s1">&#39;QTL_original_data&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">xqtl_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span> <span class="s1">&#39;original_data&#39;</span><span class="p">])],</span>
            <span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">xqtl_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span> <span class="s1">&#39;block_data&#39;</span><span class="p">])]</span>
        <span class="p">})</span>
        <span class="n">new_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">new_df</span><span class="p">,</span> <span class="n">tmp</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_df</span>

<span class="k">def</span> <span class="nf">merge_and_filter_dfs</span><span class="p">(</span><span class="n">new_df</span><span class="p">,</span> <span class="n">new_meta</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Merge and filter the dataframes based on condition/context and analysis_name to pick the corresponding original files.&quot;&quot;&quot;</span>
    <span class="n">new_df</span> <span class="o">=</span> <span class="n">new_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;condition&#39;</span><span class="p">,</span> <span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">])[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">stack</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="kp">name</span><span class="o">=</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;level_2&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;analysis_name_prefix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">new_df</span><span class="p">,</span> <span class="n">new_meta</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s1">&#39;condition&#39;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s1">&#39;context&#39;</span><span class="p">)</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;analysis_name_prefix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;analysis_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span>

    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;analysis_name_prefix&#39;</span><span class="p">,</span> <span class="s1">&#39;context&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">make_unique_data</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">filtered_df</span>

<span class="k">def</span> <span class="nf">prepare_final_paths</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">,</span> <span class="n">qtl_path</span><span class="p">,</span> <span class="n">gwas_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepare final paths for QTL and GWAS original data.&quot;&quot;&quot;</span>
    <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">qtl_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">file_name</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gwas_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">file_name</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">filtered_df</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># get overlapped regions by gene and block region for enrichment analysis</span>
<span class="p">[</span><span class="n">get_analysis_regions</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="s2">&quot;regional_data&quot;</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="k">def</span> <span class="nf">process_dataframes_with_toploci</span><span class="p">(</span><span class="n">xqtl_meta_data</span><span class="p">,</span> <span class="n">gwas_meta_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process the XQTL and GWAS dataframes.&quot;&quot;&quot;</span>
    <span class="n">xqtl_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">xqtl_meta_data</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># filter xQTL data with the ones have overlapped top loci variants with GWAS data</span>
    <span class="n">xqtl_df</span> <span class="o">=</span> <span class="n">xqtl_df</span><span class="p">[</span><span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;conditions_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()]</span>

    <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gwas_meta_data</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">gwas_df</span><span class="p">[</span><span class="n">gwas_df</span><span class="p">[</span><span class="s1">&#39;conditions_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()]</span>
    <span class="c1"># e.g. gwas_finemapping_obj = [&#39;AD_Bellenguez_2022&#39;, &#39;single_effect_regression&#39;, &#39;susie_result_trimmed&#39;]</span>
    <span class="c1"># filter gwas data with find variants in specified cohort &#39;AD_Bellenguez_2022&#39; and method &#39;single_effect_regression&#39;</span>
    <span class="n">gwas_finemapping_obj_filtered</span> <span class="o">=</span> <span class="p">[</span><span class="n">obj</span> <span class="k">for</span> <span class="n">obj</span> <span class="ow">in</span> <span class="n">gwas_finemapping_obj</span> <span class="k">if</span> <span class="n">obj</span> <span class="o">!=</span> <span class="s1">&#39;susie_result_trimmed&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gwas_finemapping_obj_filtered</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">pattern</span> <span class="o">=</span> <span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gwas_finemapping_obj_filtered</span><span class="p">)</span>
        <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">gwas_df</span><span class="p">[</span><span class="n">gwas_df</span><span class="p">[</span><span class="s1">&#39;conditions_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">pattern</span><span class="p">)]</span>

    <span class="n">xqtl_df</span> <span class="o">=</span> <span class="n">overlapped_analysis_region</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">,</span> <span class="n">gwas_df</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">xqtl_df</span><span class="p">,</span> <span class="n">gwas_df</span>



<span class="k">def</span> <span class="nf">overlapped_analysis_region</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">,</span> <span class="n">gwas_df</span><span class="p">):</span>
    <span class="c1"># Create an empty dataframe to store overlapping xQTL records</span>
    <span class="n">overlapped_xqtl_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    
    <span class="c1"># Iterate over each row in the GWAS dataframe</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">gwas_row</span> <span class="ow">in</span> <span class="n">gwas_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">gwas_chr</span> <span class="o">=</span> <span class="n">gwas_row</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span>
        <span class="n">gwas_start</span> <span class="o">=</span> <span class="n">gwas_row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>
        <span class="n">gwas_end</span> <span class="o">=</span> <span class="n">gwas_row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>
        <span class="n">gwas_block_id</span> <span class="o">=</span> <span class="n">gwas_row</span><span class="p">[</span><span class="s1">&#39;original_data&#39;</span><span class="p">]</span>  <span class="c1"># Assuming each GWAS row has a unique block ID</span>
        
        <span class="c1"># Filter xQTL dataframe for the same chromosome</span>
        <span class="n">same_chr_xqtl_df</span> <span class="o">=</span> <span class="n">xqtl_df</span><span class="p">[</span><span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gwas_chr</span><span class="p">]</span>
        
        <span class="c1"># Find overlapping xQTL regions with the current GWAS region</span>
        <span class="n">overlapping_xqtl</span> <span class="o">=</span> <span class="n">same_chr_xqtl_df</span><span class="p">[((</span><span class="n">same_chr_xqtl_df</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">gwas_end</span><span class="p">)</span> <span class="o">&amp;</span> 
                                             <span class="p">(</span><span class="n">same_chr_xqtl_df</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">gwas_start</span><span class="p">))]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># Add .copy() to avoid SettingWithCopyWarning</span>
        
        <span class="c1"># Check if there are any overlapping xQTLs</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">overlapping_xqtl</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="c1"># Use .loc to safely assign &#39;block_data&#39; to avoid the SettingWithCopyWarning</span>
            <span class="n">overlapping_xqtl</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;block_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gwas_block_id</span>
            
            <span class="c1"># Append the overlapping xQTLs to the accumulated dataframe</span>
            <span class="n">overlapped_xqtl_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">overlapped_xqtl_df</span><span class="p">,</span> <span class="n">overlapping_xqtl</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
    <span class="c1"># Group by xQTL region and concatenate &#39;block_data&#39;</span>
    <span class="n">overlapped_xqtl_df</span><span class="p">[</span><span class="s1">&#39;block_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">overlapped_xqtl_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">])[</span><span class="s1">&#39;block_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
    <span class="n">overlapped_xqtl_df</span> <span class="o">=</span> <span class="n">overlapped_xqtl_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">overlapped_xqtl_df</span>


<span class="n">xqtl_df</span><span class="p">,</span> <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">process_dataframes_with_toploci</span><span class="p">(</span><span class="n">xqtl_meta_data</span><span class="p">,</span> <span class="n">gwas_meta_data</span><span class="p">)</span>

<span class="n">region_ids</span><span class="o">=</span><span class="p">[]</span>
<span class="c1"># If region_list is provided, read the file and extract IDs</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">region_list</span><span class="o">.</span><span class="n">is_dir</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">region_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
        <span class="n">region_list_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">region_list</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span>
        <span class="n">region_ids</span> <span class="o">=</span> <span class="n">region_list_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>  <span class="c1"># Extracting the last column for IDs</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The region_list path provided is not a file.&quot;</span><span class="p">)</span>
        
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_name</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">region_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_name</span><span class="p">)))</span>
    
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">xqtl_df</span> <span class="o">=</span> <span class="n">xqtl_df</span><span class="p">[</span><span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;region_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)]</span>
    
<span class="n">xqtl_df</span><span class="p">,</span> <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">process_dataframes_with_toploci</span><span class="p">(</span><span class="n">xqtl_meta_data</span><span class="p">,</span> <span class="n">gwas_meta_data</span><span class="p">)</span>

<span class="n">new_df</span> <span class="o">=</span> <span class="n">generate_condition_based_dataframe</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">)</span>

<span class="k">if</span> <span class="p">(</span><span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">context_meta</span><span class="o">.</span><span class="n">is_file</span><span class="p">()):</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">new_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">make_unique_data</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">merge_and_filter_dfs</span><span class="p">(</span><span class="n">new_df</span><span class="p">,</span> <span class="n">generate_meta_dataframe</span><span class="p">(</span><span class="n">context_meta</span><span class="p">))</span>

<span class="n">filtered_df</span> <span class="o">=</span> <span class="n">prepare_final_paths</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">,</span> <span class="n">qtl_path</span><span class="p">,</span> <span class="n">gwas_path</span><span class="p">)</span>
<span class="n">regional_data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;data&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()],</span>
    <span class="s1">&#39;conditions&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;condition&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># get overlapped regions from flatten table, to get the regions with overlapped variants and select those regions for coloc analysis</span>
<span class="p">[</span><span class="n">get_overlapped_analysis_regions</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="s2">&quot;overlapped_regional_data&quot;</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="k">def</span> <span class="nf">map_blocks_to_data</span><span class="p">(</span><span class="n">blocks</span><span class="p">,</span> <span class="n">mapping_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map blocks to data using a provided mapping dictionary.&quot;&quot;&quot;</span>
    <span class="n">mapped_data</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mapping_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span> <span class="s1">&#39;NA&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">block</span> <span class="ow">in</span> <span class="n">blocks</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">mapped_data</span>

<span class="k">def</span> <span class="nf">process_dataframes</span><span class="p">(</span><span class="n">xqtl_meta_data</span><span class="p">,</span> <span class="n">gwas_meta_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process the XQTL and GWAS dataframes.&quot;&quot;&quot;</span>
    <span class="n">xqtl_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">xqtl_meta_data</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># filter xQTL data with the ones have overlapped top loci variants with GWAS data</span>
    <span class="n">xqtl_df</span> <span class="o">=</span> <span class="n">xqtl_df</span><span class="p">[</span><span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;block_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()]</span>

    <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gwas_meta_data</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>    
    <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">gwas_df</span><span class="p">[</span><span class="n">gwas_df</span><span class="p">[</span><span class="s1">&#39;conditions_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()]</span>
    <span class="c1"># e.g. gwas_finemapping_obj = [&#39;AD_Bellenguez_2022&#39;, &#39;single_effect_regression&#39;, &#39;susie_result_trimmed&#39;]</span>
    <span class="c1"># filter gwas data with find variants in specified cohort &#39;AD_Bellenguez_2022&#39; and method &#39;single_effect_regression&#39;</span>
    <span class="n">gwas_finemapping_obj_filtered</span> <span class="o">=</span> <span class="p">[</span><span class="n">obj</span> <span class="k">for</span> <span class="n">obj</span> <span class="ow">in</span> <span class="n">gwas_finemapping_obj</span> <span class="k">if</span> <span class="n">obj</span> <span class="o">!=</span> <span class="s1">&#39;susie_result_trimmed&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gwas_finemapping_obj_filtered</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">pattern</span> <span class="o">=</span> <span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gwas_finemapping_obj_filtered</span><span class="p">)</span>
        <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">gwas_df</span><span class="p">[</span><span class="n">gwas_df</span><span class="p">[</span><span class="s1">&#39;conditions_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">pattern</span><span class="p">)]</span>

    <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">gwas_df</span><span class="p">[</span><span class="n">gwas_df</span><span class="p">[</span><span class="s1">&#39;region_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;block_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">explode</span><span class="p">()</span><span class="o">.</span><span class="n">unique</span><span class="p">())]</span>

    <span class="n">region_to_combined_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">gwas_df</span><span class="p">[</span><span class="s1">&#39;region_id&#39;</span><span class="p">],</span> <span class="n">gwas_df</span><span class="p">[</span><span class="s1">&#39;original_data&#39;</span><span class="p">]))</span>
    <span class="c1"># and only consider the overlapped GWAS blocks</span>
    <span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;block_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">xqtl_df</span><span class="p">[</span><span class="s1">&#39;block_top_loci&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">map_blocks_to_data</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">region_to_combined_data</span><span class="p">,))</span>

    <span class="k">return</span> <span class="n">xqtl_df</span><span class="p">,</span> <span class="n">gwas_df</span>

<span class="n">xqtl_df</span><span class="p">,</span> <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">process_dataframes_with_toploci</span><span class="p">(</span><span class="n">xqtl_meta_data</span><span class="p">,</span> <span class="n">gwas_meta_data</span><span class="p">)</span>

<span class="n">new_df</span> <span class="o">=</span> <span class="n">generate_condition_based_dataframe</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">)</span>

<span class="k">if</span> <span class="p">(</span><span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">context_meta</span><span class="o">.</span><span class="n">is_file</span><span class="p">()):</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">new_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">make_unique_data</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">merge_and_filter_dfs</span><span class="p">(</span><span class="n">new_df</span><span class="p">,</span> <span class="n">generate_meta_dataframe</span><span class="p">(</span><span class="n">context_meta</span><span class="p">))</span>

<span class="n">filtered_df</span> <span class="o">=</span> <span class="n">prepare_final_paths</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">,</span> <span class="n">qtl_path</span><span class="p">,</span> <span class="n">gwas_path</span><span class="p">)</span>
<span class="n">regional_data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;data&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;QTL_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()],</span>
    <span class="s1">&#39;conditions&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;condition&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;GWAS_original_data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[xqtl_gwas_enrichment]
depends: sos_variable(&quot;regional_data&quot;)
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;No files left for analysis&#39;)

meta = regional_data[&#39;conditions&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta&quot;
output: f&#39;{cwd:a}/{name}.{_meta[0]}.enrichment.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    library(tidyverse)
    # RDS files for GWAS data
    gwas_finemapped_data = c(${paths([x for x in _meta[1:len(_meta)]]):r,})
    # RDS files for xQTL data
    xqtl_finemapped_data = c(${paths([x for x in _input]):r,})
    enrich_result = xqtl_enrichment_wrapper(gwas_files = gwas_finemapped_data, xqtl_files = xqtl_finemapped_data, 
                                                xqtl_finemapping_obj =  c(&quot;${_meta[0]}&quot;,${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x  for x in xqtl_finemapping_obj]) if len(xqtl_finemapping_obj) != 0 else &quot;NULL&quot;}), 
                                                xqtl_varname_obj =   c(&quot;${_meta[0]}&quot;,${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x  for x in xqtl_varname_obj]) if len(xqtl_varname_obj) != 0 else &quot;NULL&quot;}), 
                                                gwas_finemapping_obj =  c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in gwas_finemapping_obj]) if len(gwas_finemapping_obj) != 0 else &quot;NULL&quot;}), 
                                                gwas_varname_obj =  c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in gwas_varname_obj]) if len(gwas_varname_obj) != 0 else &quot;NULL&quot;}))
    saveRDS(enrich_result, ${_output:ar})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_coloc]
depends: sos_variable(&quot;overlapped_regional_data&quot;), sos_step(&quot;xqtl_gwas_enrichment&quot;)
stop_if(len(overlapped_regional_data[&#39;data&#39;]) == 0, f&#39;No files left for analysis&#39;)

parameter: ld_meta_file_path = path()
meta = overlapped_regional_data[&#39;conditions&#39;]
input: overlapped_regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, overlapped_regional_data[&quot;data&quot;]), group_with = &quot;meta&quot;
# output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta[0]}.coloc.rds&#39;
output: f&#39;{cwd:a}/{step_name}/{name}.{_meta[0]}.coloc.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    library(tidyverse)
    library(pecotmr)
    library(coloc) 
    # RDS files for xQTL data
    xqtl_finemapped_datas = c(${paths([x for x in _input]):r,})
    coloc_res &lt;- list()
    # are we doing something like this? that means results are by each context, and each one has 
    for(xqtl_finemapped_data in xqtl_finemapped_datas){
      qtl_dat &lt;- readRDS(xqtl_finemapped_data)
      gene = names(qtl_dat)
      gene_region = pecotmr:::get_nested_element(qtl_dat[[1]],  c(&quot;${_meta[0]}&quot;,&quot;region_info&quot;, &quot;grange&quot;))
  
      # Step 1: find relevant GWAS regions that overlap each the xQTL region of interest
      gwas_finemapped_datas &lt;- c(${ &quot;,&quot;.join(set(&#39;&quot;%s&quot;&#39; % path for sublist in [&#39;,&#39;.join(x[1:]) for x in overlapped_regional_data[&#39;conditions&#39;]] for path in sublist.split(&#39;,&#39;)))}) 
 
      gwas_regions &lt;- gwas_finemapped_datas %&gt;% basename %&gt;% gsub(&#39;.susie_rss.rds&#39;,&#39;&#39;,.)
      overlap_index &lt;- NULL
      for (i in 1:length(gwas_regions)) {
        region &lt;- gwas_regions[i]
        split_region &lt;- unlist(strsplit(region, &quot;_&quot;))
        block_chrom &lt;- as.numeric(split_region[1])
        block_start &lt;- as.numeric(split_region[2] %&gt;% strsplit(., &quot;-&quot;) %&gt;% unlist %&gt;% .[1])
        block_end &lt;- as.numeric(split_region[2] %&gt;% strsplit(., &quot;-&quot;) %&gt;% unlist %&gt;% .[2])
        if (gene_region$chrom == block_chrom &amp;&amp; (gene_region$start &lt;= block_end |  gene_region$end &gt;= block_start)) {
          overlap_index &lt;- c(overlap_index, i)
        }
      }
  
     if (!is.null(overlap_index)) {
       gwas_finemapped_data &lt;- gwas_finemapped_datas[overlap_index]
  
       # Step 2: load enrichment analysis results
       # Extract values for p1, p2, and p12
       enrich_file = paste0(&#39;${cwd:a}&#39;,&#39;/&#39;, &#39;${name}&#39;, &#39;.&#39; ,&#39;${_meta[0]}&#39;,&#39;.enrichment.rds&#39;)
       p1 &lt;-  readRDS(enrich_file)[[1]]$`Alternative (coloc) p1`
       p2 &lt;-  readRDS(enrich_file)[[1]]$`Alternative (coloc) p2`
       p12 &lt;-  readRDS(enrich_file)[[1]]$`Alternative (coloc) p12`
  
       message(&quot;Priors are P1:&quot;, p1, &quot;; p2: &quot;, p2, &quot;; p12: &quot;, p12)
       # Step 3: Apply colocalization analysis between each condition and GWAS
       coloc_res[[gene]] &lt;- coloc_wrapper(xqtl_file = xqtl_finemapped_data, gwas_files = gwas_finemapped_data, 
                                          xqtl_finemapping_obj =  c(&quot;${_meta[0]}&quot;,${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x  for x in xqtl_finemapping_obj]) if len(xqtl_finemapping_obj) != 0 else &quot;NULL&quot;}), 
                                          xqtl_varname_obj =   c(&quot;${_meta[0]}&quot;,${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x  for x in xqtl_varname_obj]) if len(xqtl_varname_obj) != 0 else &quot;NULL&quot;}), 
                                          gwas_finemapping_obj =  c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in gwas_finemapping_obj]) if len(gwas_finemapping_obj) != 0 else &quot;NULL&quot;}), 
                                          gwas_varname_obj =  c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in gwas_varname_obj]) if len(gwas_varname_obj) != 0 else &quot;NULL&quot;}),
                                          xqtl_region_obj =   c(&quot;${_meta[0]}&quot;,${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x  for x in xqtl_region_obj]) if len(xqtl_region_obj) != 0 else &quot;NULL&quot;}), 
                                          gwas_region_obj =  c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in gwas_region_obj]) if len(gwas_region_obj) != 0 else &quot;NULL&quot;}),
                                          p1 = p1, p2 = p2, p12 = p12)

        if (${&quot;TRUE&quot; if ld_meta_file_path.is_file() else &quot;FALSE&quot;}) {
          if(length(coloc_res[[gene]]) &gt; 2) {
              coloc_res[[gene]] &lt;- coloc_post_processor(coloc_res[[gene]], LD_meta_file_path = ${ld_meta_file_path:r}, analysis_region = coloc_res[[gene]]$analysis_region)
          } else { message(&quot;No coloc results were generated.&quot;) }
        }
      
      } else {
        message(&quot;No overlap was found between GWAS blocks and QTL analysis region.&quot;)
        coloc_res &lt;-  &quot;No overlap was found between GWAS blocks and QTL analysis region.&quot;
      }
    }
    saveRDS(coloc_res, ${_output:r})
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/pecotmr_integration"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="pecotmr.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">The PECOTMR intergration framework</p>
      </div>
    </a>
    <a class="right-next"
       href="twas_mr.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">GWAS integration: TWAS and MR</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analytical-logic">Analytical Logic</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>