
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>GWAS integration: TWAS and MR &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="TWAS and COLOC integration" href="intact.html" />
    <link rel="prev" title="xQTL and GWAS region alignment" href="SuSiE_enloc.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../reference_data/reference_data.html">
   Reference Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/reference_data_preparation.html">
     Reference Data Standardization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/generalized_TADB.html">
     Generation of Topologically Associated Domains and their Boundaries
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/ld_prune_reference.html">
     Independent list of variants using LD clumping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene Coordinate Annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype Data Imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden Factor Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate Data Formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="SuSiE_enloc.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     GWAS integration: TWAS and MR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Extract genome-wide data for multivariate analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/pecotmr_integration/twas_mr.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_integration/twas_mr.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/pecotmr_integration/twas_mr.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-1-twas">
     Step 1: TWAS
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-2-mr-for-candidate-genes">
     Step 2: MR for candidate genes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     Input
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#gwas-data-input-interface-similar-to-susie-rss">
       GWAS Data Input Interface (Similar to
       <code class="docutils literal notranslate">
        <span class="pre">
         susie_rss
        </span>
       </code>
       )
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#output-of-fine-mapping-twas-pipeline">
       Output of Fine-Mapping &amp; TWAS Pipeline
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="gwas-integration-twas-and-mr">
<h1>GWAS integration: TWAS and MR<a class="headerlink" href="#gwas-integration-twas-and-mr" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This module provides software implementations for transcriptome-wide association analysis (TWAS), and Mendelian Randomization using fine-mapping instrumental variables (IV). The procedures implements the MR procedure described in Zhang et al 2020 for “causal” effects estimation and model validation, with the unit of analysis being a single gene-trait pair.</p>
<p>This procedure is based on the SuSiE-TWAS workflow — it assumes that xQTL fine-mapping has been performed (to be used for both TWAS and MR) and moleuclar traits prediction weights pre-computed (to be used for TWAS). Cross validation for TWAS weights is optional but highly recommended.</p>
<p>GWAS data required are GWAS summary statistics and LD matrix for the region of interest.</p>
<div class="section" id="step-1-twas">
<h3>Step 1: TWAS<a class="headerlink" href="#step-1-twas" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>Extract GWAS z-score for region of interest and corresponding LD matrix.</p></li>
<li><p>(Optional) perform allele matching QC for the LD matrix with summary stats.</p></li>
<li><p>Process weights: for LASSO, Elastic Net and mr.ash we have to take the weights as is for QTL variants overlapping with GWAS variants. For SuSiE weights it can be adjusted to exactly match GWAS variants.</p></li>
<li><p>Perofrm TWAS test for multiple sets of weights.</p></li>
<li><p>For each gene, filter TWAS results by keeping the best model selected by CV. Drop the genes that don’t show good evidence of TWAS prediction weights.</p></li>
</ol>
</div>
<div class="section" id="step-2-mr-for-candidate-genes">
<h3>Step 2: MR for candidate genes<a class="headerlink" href="#step-2-mr-for-candidate-genes" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>Limit MR only to those showing some evidence of TWAS significance AND have strong instrumental variable (fine-mapping PIP or CS).</p></li>
<li><p>Use fine-mapped xQTL with GWAS data to perform MR.</p></li>
<li><p>For multiple IV, aggregate individual IV estimates using a fixed-effect meta-analysis procedure.</p></li>
<li><p>Identify and exclude results with severe violations of the exclusion restriction (ER) assumption.</p></li>
</ol>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="section" id="gwas-data-input-interface-similar-to-susie-rss">
<h4>GWAS Data Input Interface (Similar to <code class="docutils literal notranslate"><span class="pre">susie_rss</span></code>)<a class="headerlink" href="#gwas-data-input-interface-similar-to-susie-rss" title="Permalink to this headline">¶</a></h4>
<p><strong>GWAS Summary Statistics Files</strong></p>
<ul class="simple">
<li><p><strong>Input</strong>: Vector of files for one or more GWAS studies.</p></li>
<li><p><strong>Format</strong>:</p>
<ul>
<li><p>Tab-delimited files.</p></li>
<li><p>First 4 columns: <code class="docutils literal notranslate"><span class="pre">chr</span></code>, <code class="docutils literal notranslate"><span class="pre">pos</span></code>, <code class="docutils literal notranslate"><span class="pre">a0</span></code>, <code class="docutils literal notranslate"><span class="pre">a1</span></code></p></li>
<li><p>Additional columns can be loaded using column mapping file see below</p></li>
</ul>
</li>
<li><p><strong>Column Mapping files (optional)</strong>:</p>
<ul>
<li><p>Optional YAML file for custom column mapping.</p></li>
<li><p>Required columns: <code class="docutils literal notranslate"><span class="pre">chr</span></code>, <code class="docutils literal notranslate"><span class="pre">pos</span></code>, <code class="docutils literal notranslate"><span class="pre">a0</span></code>, <code class="docutils literal notranslate"><span class="pre">a1</span></code>, either <code class="docutils literal notranslate"><span class="pre">z</span></code> or (<code class="docutils literal notranslate"><span class="pre">betahat</span></code> and <code class="docutils literal notranslate"><span class="pre">sebetahat</span></code>).</p></li>
<li><p>Optional columns: <code class="docutils literal notranslate"><span class="pre">n</span></code>, <code class="docutils literal notranslate"><span class="pre">var_y</span></code>.</p></li>
</ul>
</li>
</ul>
<p><strong>GWAS Summary Statistics Meta-File</strong>: this is optional and helpful when there are lots of GWAS data to process via the same command</p>
<ul class="simple">
<li><p><strong>Columns</strong>: <code class="docutils literal notranslate"><span class="pre">study_id</span></code>, chromosome number, path to summary statistics file, optional path to column mapping file.</p></li>
<li><p><strong>Note</strong>: Chromosome number <code class="docutils literal notranslate"><span class="pre">0</span></code> indicates a genome-wide file.</p></li>
</ul>
<p>eg: <code class="docutils literal notranslate"><span class="pre">gwas_meta.tsv</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">study_id</span>    <span class="n">chrom</span>    <span class="n">file_path</span>                 <span class="n">column_mapping_file</span>
<span class="n">study1</span>      <span class="mi">1</span>        <span class="n">gwas1</span><span class="o">.</span><span class="n">tsv</span><span class="o">.</span><span class="n">gz</span>         <span class="n">column_mapping</span><span class="o">.</span><span class="n">yml</span>
<span class="n">study1</span>      <span class="mi">2</span>        <span class="n">gwas2</span><span class="o">.</span><span class="n">tsv</span><span class="o">.</span><span class="n">gz</span>         <span class="n">column_mapping</span><span class="o">.</span><span class="n">yml</span>
<span class="n">study2</span>      <span class="mi">0</span>        <span class="n">gwas3</span><span class="o">.</span><span class="n">tsv</span><span class="o">.</span><span class="n">gz</span>         <span class="n">column_mapping</span><span class="o">.</span><span class="n">yml</span>
</pre></div>
</div>
<p><strong>LD Reference Metadata File</strong></p>
<ul class="simple">
<li><p><strong>Format</strong>: Single TSV file.</p></li>
<li><p><strong>Contents</strong>:</p>
<ul>
<li><p>Columns: <code class="docutils literal notranslate"><span class="pre">chr</span></code>, <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">end</span></code>, path to the LD matrix, genomic build.</p></li>
<li><p>LD matrix path format: comma-separated, first entry is the LD matrix, second is the bim file.</p></li>
</ul>
</li>
<li><p><strong>Documentation</strong>: Refer to our LD reference preparation document for detailed information (Tosin pending update).</p></li>
</ul>
</div>
<div class="section" id="output-of-fine-mapping-twas-pipeline">
<h4>Output of Fine-Mapping &amp; TWAS Pipeline<a class="headerlink" href="#output-of-fine-mapping-twas-pipeline" title="Permalink to this headline">¶</a></h4>
<p><strong>xQTL Weight Database files</strong></p>
<ul class="simple">
<li><p>path to various weight DB files, comma delimited.</p></li>
</ul>
<p><strong>xQTL Weight Database Metadata File</strong>: this is optional and helpful when TWAS is done genome-wide for many regions via the same command</p>
<ul class="simple">
<li><p><strong>Types</strong>: Gene-based or TAD-based.</p></li>
<li><p><strong>Structure</strong>:</p>
<ul>
<li><p>RDS format.</p></li>
<li><p>Organized hierarchically: region → condition → weight matrix.</p></li>
<li><p>Each column represents a different method.</p></li>
</ul>
</li>
<li><p><strong>Format</strong>: <code class="docutils literal notranslate"><span class="pre">chrom</span></code>, <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">end</span></code>, <code class="docutils literal notranslate"><span class="pre">region_id</span></code>, <code class="docutils literal notranslate"><span class="pre">condition</span></code> (e.g., tissue type, QTL), path to various weight DB files, comma delimited.</p></li>
</ul>
<p>eg: <code class="docutils literal notranslate"><span class="pre">xqtl_meta.tsv</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chrom</span>    <span class="n">start</span>    <span class="n">end</span>    <span class="n">region_id</span>    <span class="n">condition</span>    <span class="n">file_path</span>
<span class="mi">1</span>        <span class="mi">1000</span>     <span class="mi">5000</span>   <span class="n">region1</span>      <span class="n">cohor1</span><span class="p">:</span><span class="n">tissue1</span><span class="p">:</span><span class="n">eQTL</span>      <span class="n">weight1</span><span class="o">.</span><span class="n">rds</span><span class="p">,</span> <span class="n">weight2</span><span class="o">.</span><span class="n">rds</span>
<span class="mi">2</span>        <span class="mi">2000</span>     <span class="mi">6000</span>   <span class="n">region2</span>      <span class="n">cohor1</span><span class="p">:</span><span class="n">tissue1</span><span class="p">:</span><span class="n">eQTL</span>      <span class="n">weight3</span><span class="o">.</span><span class="n">rds</span>
<span class="mi">3</span>        <span class="mi">3000</span>     <span class="mi">7000</span>   <span class="n">region3</span>      <span class="n">cohor1</span><span class="p">:</span><span class="n">tissue1</span><span class="p">:</span><span class="n">eQTL</span>      <span class="n">weight4</span><span class="o">.</span><span class="n">rds</span><span class="p">,</span> <span class="n">weight5</span><span class="o">.</span><span class="n">rds</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>TWAS FIXME this is incorrect for now.</p>
<ul class="simple">
<li><p>Each row corresponds to a single SNP inferred as a member of a signal cluster, with columns including:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">snp</span></code>: SNP name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">beta_eQTL</span></code>: eQTL effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">se_eQTL</span></code>: Standard error of estimated eQTL effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">beta_GWAS</span></code>: GWAS effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">se_GWAS</span></code>: Standard error of GWAS effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cluster</span></code>: Signal cluster ID (credible sets index).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pip</span></code>: SNP posterior inclusion probability (PIP).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id</span></code>: Gene name.</p></li>
</ul>
</li>
</ul>
<p>MR</p>
<ul class="simple">
<li><p>The output includes the following columns for each gene:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id</span></code>: Gene name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">num_cluster</span></code>: Number of credible sets of the gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">num_instruments</span></code>: Number of instruments included in the gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spip</span></code>: Sum of PIP for credible sets of each gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">grp_beta</span></code>: Signal-level estimates, combining SNP-level estimates from member SNPs weighted by their PIPs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">grp_se</span></code>: Standard error of signal-level estimates.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">meta</span></code>: Gene-level estimate of the causal effect, aggregating signal-level estimates using a fixed-effect meta-analysis model.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">se_meta</span></code>: Standard error of the gene-level estimate of the causal effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Q</span></code>: Cochran’s Q statistic.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">I2</span></code>: <span class="math notranslate nohighlight">\(I^2\)</span> statistics</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Workdir</span>
<span class="kn">parameter:</span> <span class="n">gwas_meta_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">xqtl_meta_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">100</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;1h&quot;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">2</span>

<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">partition</span><span class="p">):</span>
    <span class="c1"># from itertools import accumulate</span>
    <span class="c1"># partition = [len(x) for x in partition]</span>
    <span class="c1"># Compute the cumulative sums once</span>
    <span class="c1"># cumsum_vector = list(accumulate(partition))</span>
    <span class="c1"># Use slicing based on the cumulative sums</span>
    <span class="c1"># return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]</span>
    <span class="k">return</span> <span class="n">partition</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">get_analysis_regions</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="s2">&quot;regional_data&quot;</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="k">def</span> <span class="nf">file_exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">relative_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check if a file exists at the given path or relative to a specified path.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">elif</span> <span class="n">relative_path</span><span class="p">:</span>
        <span class="c1"># Constructing the relative path</span>
        <span class="n">relative_file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">relative_path</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">relative_file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">relative_file_path</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">False</span>

<span class="k">def</span> <span class="nf">check_required_columns</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">required_columns</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check if the required columns are present in the dataframe.&quot;&quot;&quot;</span>
    <span class="n">missing_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">required_columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)]</span>
    <span class="k">if</span> <span class="n">missing_columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing required columns: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">missing_columns</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">extract_regional_data</span><span class="p">(</span><span class="n">gwas_meta_file</span><span class="p">,</span> <span class="n">xqtl_meta_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts data from GWAS and xQTL metadata files. Checks if files exist at specified paths,</span>
<span class="sd">    or relative to the metadata file&#39;s directory if not found initially. Also checks for required columns in the input files.</span>

<span class="sd">    Args:</span>
<span class="sd">    - gwas_meta_file (str): File path to the GWAS metadata file.</span>
<span class="sd">    - xqtl_meta_file (str): File path to the xQTL weight metadata file.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Tuple of two dictionaries:</span>
<span class="sd">        - GWAS Dictionary: Maps study IDs to a list containing chromosome number, </span>
<span class="sd">          GWAS file path, and optional column mapping file path.</span>
<span class="sd">        - xQTL Dictionary: Nested dictionary with region IDs as keys. Each key maps to a </span>
<span class="sd">          dictionary containing &#39;meta_info&#39; (list of chromosome, start, end, condition)</span>
<span class="sd">          and &#39;files&#39; (list of file paths).</span>

<span class="sd">    Raises:</span>
<span class="sd">    - FileNotFoundError: If any specified file path does not exist, even when considered</span>
<span class="sd">      relative to the metadata file&#39;s directory.</span>
<span class="sd">    - ValueError: If required columns are missing in the input files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
    <span class="c1"># Required columns for each file type</span>
    <span class="n">required_gwas_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">,</span> <span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;file_path&#39;</span><span class="p">]</span>
    <span class="n">required_xqtl_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;region_id&#39;</span><span class="p">,</span> <span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;condition&#39;</span><span class="p">,</span> <span class="s1">&#39;file_path&#39;</span><span class="p">]</span>

    <span class="c1"># Base directory of the metadata files</span>
    <span class="n">gwas_base_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">gwas_meta_file</span><span class="p">)</span>
    <span class="n">xqtl_base_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">xqtl_meta_file</span><span class="p">)</span>
    
    <span class="c1"># Reading the GWAS metadata file</span>
    <span class="n">gwas_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gwas_meta_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">check_required_columns</span><span class="p">(</span><span class="n">gwas_df</span><span class="p">,</span> <span class="n">required_gwas_columns</span><span class="p">)</span>
    <span class="n">gwas_dict</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gwas_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">file_path</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;file_path&#39;</span><span class="p">]</span>
        <span class="n">mapping_file</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;column_mapping_file&#39;</span><span class="p">)</span>

        <span class="c1"># Check if the file and optional mapping file exist</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">gwas_base_dir</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">mapping_file</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">mapping_file</span><span class="p">,</span> <span class="n">gwas_base_dir</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> not found for </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1"># Adjust paths if necessary</span>
        <span class="n">file_path</span> <span class="o">=</span> <span class="n">file_path</span> <span class="k">if</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gwas_base_dir</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mapping_file</span><span class="p">:</span>
            <span class="n">mapping_file</span> <span class="o">=</span> <span class="n">mapping_file</span> <span class="k">if</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">mapping_file</span><span class="p">)</span> <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gwas_base_dir</span><span class="p">,</span> <span class="n">mapping_file</span><span class="p">)</span>
        
       <span class="c1"># Create or update the entry for the study_id</span>
        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gwas_dict</span><span class="p">:</span>
            <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Expand chrom 0 to chrom 1-22 or use the specified chrom</span>
        <span class="n">chrom_range</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]]</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chrom_range</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]]:</span>
                <span class="n">existing_entry</span> <span class="o">=</span> <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]][</span><span class="n">chrom</span><span class="p">]</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Duplicate chromosome specification for study_id </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">, chrom </span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">. &quot;</span>
                                 <span class="sa">f</span><span class="s2">&quot;Conflicting entries: </span><span class="si">{</span><span class="n">existing_entry</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="p">[</span><span class="n">file_path</span><span class="p">,</span> <span class="n">mapping_file</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">gwas_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;study_id&#39;</span><span class="p">]][</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">file_path</span><span class="p">,</span> <span class="n">mapping_file</span><span class="p">]</span>

    <span class="c1"># Reading the xQTL weight metadata file</span>
    <span class="n">xqtl_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">xqtl_meta_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">check_required_columns</span><span class="p">(</span><span class="n">xqtl_df</span><span class="p">,</span> <span class="n">required_xqtl_columns</span><span class="p">)</span>
    <span class="n">xqtl_dict</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">xqtl_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">file_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">fp</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">fp</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;file_path&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)]</span>  <span class="c1"># Splitting and stripping file paths</span>

        <span class="c1"># Check if each file path exists</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">fp</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">file_paths</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">fp</span><span class="p">,</span> <span class="n">xqtl_base_dir</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">fp</span><span class="si">}</span><span class="s2"> not found in </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;region_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_paths</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">fp</span> <span class="k">if</span> <span class="n">file_exists</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span> <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">xqtl_base_dir</span><span class="p">,</span> <span class="n">fp</span><span class="p">)</span>

        <span class="n">xqtl_dict</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;region_id&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;meta_info&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;region_id&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;condition&#39;</span><span class="p">]],</span>
                                       <span class="s2">&quot;files&quot;</span><span class="p">:</span> <span class="n">file_paths</span><span class="p">}</span>

    <span class="k">return</span> <span class="n">gwas_dict</span><span class="p">,</span> <span class="n">xqtl_dict</span>

<span class="c1"># Example usage</span>
<span class="n">gwas_dict</span><span class="p">,</span> <span class="n">xqtl_dict</span> <span class="o">=</span> <span class="n">extract_regional_data</span><span class="p">(</span><span class="n">gwas_meta_file</span><span class="p">,</span> <span class="n">xqtl_meta_file</span><span class="p">)</span>
<span class="n">regional_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="s2">&quot;GWAS&quot;</span><span class="p">,</span> <span class="n">gwas_dict</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;xQTL&quot;</span><span class="p">,</span> <span class="n">xqtl_dict</span><span class="p">)])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">regional_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_scan_1</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s2">&quot;regional_data&quot;</span><span class="p">)</span>
<span class="n">regions</span> <span class="o">=</span> <span class="n">regional_data</span><span class="p">[</span><span class="s1">&#39;xQTL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="n">meta_info</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;meta_info&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regional_data</span><span class="p">[</span><span class="s1">&#39;xQTL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
<span class="n">xqtl_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;files&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regional_data</span><span class="p">[</span><span class="s1">&#39;xQTL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
<span class="kn">input:</span> <span class="n">xqtl_files</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">group_by_region</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">xqtl_files</span><span class="p">),</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;meta_info&quot;</span>
<span class="c1">#output: f&#39;{cwd:a}/{step_name[:-2]}/{condition.replace(&quot;:&quot;, &quot;_&quot;)}.{_meta_info[3]}.twas.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="c1"># we have potentially multiple weight db RDS files for each region of interest</span>
    <span class="n">weight_db</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,})</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">start</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> 
    <span class="n">end</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">2</span><span class="p">]}</span>
    <span class="n">region</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{_meta_info[3]}</span><span class="s2">&quot;</span>
    <span class="n">xqtl_conditions</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">(</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">4</span><span class="p">:]):</span><span class="n">r</span><span class="p">,})</span>
    <span class="n">gwas_studies</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">(</span><span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;GWAS&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span><span class="n">r</span><span class="p">,})</span>
    <span class="c1"># load gwas data file for this particular chrom</span>
    <span class="n">gwas_files</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">paths</span><span class="p">([</span><span class="n">v</span><span class="p">[</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;GWAS&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()]):</span><span class="n">r</span><span class="p">,})</span>
    <span class="n">library</span><span class="p">(</span><span class="n">pecotmr</span><span class="p">)</span>
    <span class="c1"># Step 0: Load GWAS data for the region of interest, for each study</span>
    <span class="n">gwas_data</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">s</span> <span class="ow">in</span> <span class="n">length</span><span class="p">(</span><span class="n">gwas_studies</span><span class="p">))</span> <span class="p">{</span>
      <span class="n">gwas_data</span><span class="p">[[</span><span class="n">gwas_studies</span><span class="p">[</span><span class="n">s</span><span class="p">]]]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
      <span class="c1"># FIXME: Anqi please fix this</span>
      <span class="n">gwas_data</span><span class="p">[[</span><span class="n">gwas_studies</span><span class="p">[</span><span class="n">s</span><span class="p">]]][</span><span class="s2">&quot;sumstats&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tabix_region</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gwas_files</span><span class="p">[</span><span class="n">s</span><span class="p">])</span>
      <span class="n">gwas_data</span><span class="p">[[</span><span class="n">gwas_studies</span><span class="p">[</span><span class="n">s</span><span class="p">]]][</span><span class="s2">&quot;LD&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">load_LD_matrix</span><span class="p">(</span> <span class="o">...</span> <span class="p">)</span>
    <span class="p">}</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">condition</span> <span class="ow">in</span> <span class="n">xqtl_conditions</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># For each region of interest for a particular xQTL (condtion), we can perform TWAS for all the input studies, with all of the weights available</span>
        <span class="c1"># Step 1: load the weight matrix: a matrix of weights for the specified condition and region; each column is weight for a method. The row names should be variant names.</span>
        <span class="n">weights</span> <span class="o">=</span> <span class="n">load_twas_weights</span><span class="p">(</span><span class="n">weight_db_files</span><span class="p">,</span> <span class="n">condition</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span> <span class="n">variable_name_obj</span> <span class="o">=</span> <span class="s2">&quot;variant_name&quot;</span><span class="p">,</span> <span class="n">twas_weights_table</span> <span class="o">=</span> <span class="s2">&quot;twas_weights&quot;</span><span class="p">)</span>
        <span class="c1"># Step 2: for each study we perform TWAS: </span>
        <span class="c1"># take overlap between weights and the gwas_data summstats for each study. If there are some variants in sumstats but not in weights then simply make weight zero</span>
        <span class="n">weights</span> <span class="o">=</span> <span class="o">...</span>
        <span class="c1"># Step 2-1: NOT YET IMPLEMEMENTED (FIXME) adjust susie_weights if possible --- if susie_fit exist in weight_db. This should replace the origional SuSiE_weight computed above</span>
        <span class="c1"># Step 3: perform TWAS on the weights and sumstats</span>
    <span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/pecotmr_integration"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="SuSiE_enloc.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">xQTL and GWAS region alignment</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="intact.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">TWAS and COLOC integration</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>