
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>GWAS integration: TWAS and MR &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="TWAS and COLOC integration" href="intact.html" />
    <link rel="prev" title="xQTL and GWAS region alignment" href="SuSiE_enloc.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../reference_data/reference_data.html">
   Reference Data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/reference_data_preparation.html">
     Reference Data Standardization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/generalized_TADB.html">
     Generation of Topologically Associated Domains and their Boundaries
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../reference_data/ld_prune_reference.html">
     Independent list of variants using LD clumping
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene Coordinate Annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype Data Imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden Factor Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate Data Formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cis_analysis/mrmash.html">
   Multivariate TWAS prediction model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   GREGOR enrichment analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="SuSiE_enloc.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     GWAS integration: TWAS and MR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">
     Extract genome-wide data for multivariate analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/pecotmr_integration/twas_mr.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/pecotmr_integration/twas_mr.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/pecotmr_integration/twas_mr.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mendelian-randomization-using-fine-mapping-iv">
   Mendelian Randomization using Fine-Mapping IV
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ptwas-analysis-from-susie-objects">
     PTWAS Analysis from SuSiE Objects
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#step-1-ptwas-twas-z-format">
       Step 1: PTWAS (twas_z_format)
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#step-2-mr-twas-candidate-genes">
       Step 2: MR (twas_candidate_genes)
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input-requirements-for-twas">
     Input Requirements for TWAS
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#output-specifications-for-twas">
     Output Specifications for TWAS
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#step-1-twas">
       Step 1: TWAS
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#step-2-mr">
       Step 2: MR
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#output-file-format-for-twas">
       Output File Format for TWAS
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#causal-effect-estimation-and-validation">
     Causal Effect Estimation and Validation
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#estimation-process">
       Estimation Process
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#validation-process">
       Validation Process
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#file-formats">
     File Formats
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#input-file-format">
       Input File Format
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#output-file-format">
       Output File Format
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="gwas-integration-twas-and-mr">
<h1>GWAS integration: TWAS and MR<a class="headerlink" href="#gwas-integration-twas-and-mr" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This module provides software implementations for transcriptome-wide association analysis (TWAS), aimed at facilitating rigorous causal inference connecting genes to complex traits.</p>
</div>
<div class="section" id="mendelian-randomization-using-fine-mapping-iv">
<h2>Mendelian Randomization using Fine-Mapping IV<a class="headerlink" href="#mendelian-randomization-using-fine-mapping-iv" title="Permalink to this headline">¶</a></h2>
<p>This component focuses on Mendelian Randomization using fine-mapping instrumental variables (IV). The procedures implemented in the R script encompass PTWAS estimation and model validation, with the unit of analysis being a single gene-trait pair.</p>
</div>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><strong>Purpose</strong>: Organize eQTL fine-mapping results, eQTL summary statistics, and GWAS summary statistics into a format for ptwas_est processing.</p></li>
<li><p><strong>Execution</strong>: Run ptwas_est using R scripts for causal effect estimation and validation.</p></li>
</ul>
<div class="section" id="ptwas-analysis-from-susie-objects">
<h3>PTWAS Analysis from SuSiE Objects<a class="headerlink" href="#ptwas-analysis-from-susie-objects" title="Permalink to this headline">¶</a></h3>
<div class="section" id="step-1-ptwas-twas-z-format">
<h4>Step 1: PTWAS (twas_z_format)<a class="headerlink" href="#step-1-ptwas-twas-z-format" title="Permalink to this headline">¶</a></h4>
<ol class="simple">
<li><p>Load GWAS z-score.</p></li>
<li><p>Load LD matrices within the TAD region of each gene.</p></li>
<li><p>Use <code class="docutils literal notranslate"><span class="pre">allele_qc()</span></code> function to QC the LD matrix with summary stats.</p></li>
<li><p>Extract weights: susie, lasso, enet, and mr_ash.</p></li>
<li><p>Use <code class="docutils literal notranslate"><span class="pre">twas_z()</span></code> to compute TWAS results from multiple weights.</p></li>
</ol>
</div>
<div class="section" id="step-2-mr-twas-candidate-genes">
<h4>Step 2: MR (twas_candidate_genes)<a class="headerlink" href="#step-2-mr-twas-candidate-genes" title="Permalink to this headline">¶</a></h4>
<ol class="simple">
<li><p>Apply a p-value cutoff to identify TWAS regions of interest.</p></li>
<li><p>Save QC-ed GWAS data in a format compatible with <a class="reference external" href="https://github.com/cumc/pecotmr/blob/main/R/mr.R">mr.R</a>.</p></li>
</ol>
</div>
</div>
<div class="section" id="input-requirements-for-twas">
<h3>Input Requirements for TWAS<a class="headerlink" href="#input-requirements-for-twas" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">susie_path</span></code>: List of file paths for susie results.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GWAS_path</span></code>: GWAS summary statistics path (for loading the GWAS sumstats dataframe).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">LD_path</span></code>: LD block matrix path (list of file paths of LD block).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">TAD_path</span></code>: TAD region path (dataframe of TAD region for each gene, used to subset the LD matrix).</p></li>
</ul>
</div>
<div class="section" id="output-specifications-for-twas">
<h3>Output Specifications for TWAS<a class="headerlink" href="#output-specifications-for-twas" title="Permalink to this headline">¶</a></h3>
<div class="section" id="step-1-twas">
<h4>Step 1: TWAS<a class="headerlink" href="#step-1-twas" title="Permalink to this headline">¶</a></h4>
<ol class="simple">
<li><p>TWAS results from multiple weights.</p></li>
<li><p>GWAS summary statistics after QC.</p></li>
<li><p>Input format for the <code class="docutils literal notranslate"><span class="pre">twas_z()</span></code> function.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">twas_z_format</span></code>: Dataframe for input format of the <code class="docutils literal notranslate"><span class="pre">twas_z()</span></code> function.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_weights_pq</span></code>: Dataframe of output from <code class="docutils literal notranslate"><span class="pre">twas_z()</span></code> function. TWAS_z is applied to weights of four methods (susie, lasso, enet, and mr_ash) to calculate p-values and corresponding q-values.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">AD_allele_flip</span></code>: AD GWAS sumstats after QC.</p></li>
</ul>
</li>
</ol>
</div>
<div class="section" id="step-2-mr">
<h4>Step 2: MR<a class="headerlink" href="#step-2-mr" title="Permalink to this headline">¶</a></h4>
<ol class="simple">
<li><p>Candidate genes from TWAS.</p></li>
<li><p>Input format for <a class="reference external" href="https://github.com/cumc/pecotmr/blob/main/R/mr.R">mr.R</a>.</p></li>
<li><p>MR results.</p></li>
</ol>
</div>
<div class="section" id="output-file-format-for-twas">
<h4>Output File Format for TWAS<a class="headerlink" href="#output-file-format-for-twas" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>Each row corresponds to a single SNP inferred as a member of a signal cluster, with columns including:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">snp</span></code>: SNP name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">beta_eQTL</span></code>: eQTL effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">se_eQTL</span></code>: Standard error of estimated eQTL effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">beta_GWAS</span></code>: GWAS effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">se_GWAS</span></code>: Standard error of GWAS effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cluster</span></code>: Signal cluster ID (credible sets index).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pip</span></code>: SNP posterior inclusion probability (PIP).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id</span></code>: Gene name.</p></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="causal-effect-estimation-and-validation">
<h3>Causal Effect Estimation and Validation<a class="headerlink" href="#causal-effect-estimation-and-validation" title="Permalink to this headline">¶</a></h3>
<div class="section" id="estimation-process">
<h4>Estimation Process<a class="headerlink" href="#estimation-process" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>Select strong eQTLs by thresholding on signal-level PIPs (cpip = 0.5).</p></li>
<li><p>Aggregate individual eQTL estimates using a fixed-effect meta-analysis procedure.</p></li>
</ul>
</div>
<div class="section" id="validation-process">
<h4>Validation Process<a class="headerlink" href="#validation-process" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>Identify and exclude cases with severe violations of the exclusion restriction (ER) assumption.</p></li>
</ul>
</div>
</div>
<div class="section" id="file-formats">
<h3>File Formats<a class="headerlink" href="#file-formats" title="Permalink to this headline">¶</a></h3>
<div class="section" id="input-file-format">
<h4>Input File Format<a class="headerlink" href="#input-file-format" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>Output from the first step of the analysis.</p></li>
</ul>
</div>
<div class="section" id="output-file-format">
<h4>Output File Format<a class="headerlink" href="#output-file-format" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><strong>Contents</strong>: The file format includes the following columns for each gene:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id</span></code>: Gene name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">num_cluster</span></code>: Number of credible sets of the gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">num_instruments</span></code>: Number of instruments included in the gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spip</span></code>: Sum of PIP for credible sets of each gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">grp_beta</span></code>: Signal-level estimates, combining SNP-level estimates from member SNPs weighted by their PIPs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">grp_se</span></code>: Standard error of signal-level estimates.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">meta</span></code>: Gene-level estimate of the causal effect, aggregating signal-level estimates using a fixed-effect meta-analysis model.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">se_meta</span></code>: Standard error of the gene-level estimate of the causal effect.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Q</span></code>: Cochran’s Q statistic.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">I2</span></code>: <span class="math notranslate nohighlight">\(I^2\)</span> statistics</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">global</span><span class="p">]</span>
<span class="c1"># Workdir</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">susie_path</span> <span class="o">=</span> <span class="n">paths</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="nf">path</span><span class="p">(</span><span class="s">&quot;output&quot;</span><span class="p">)</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">container</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
<span class="n">import</span> <span class="n">re</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s">&#39; &#39;</span> <span class="o">+</span> <span class="nf">re.sub</span><span class="p">(</span><span class="n">r</span><span class="s">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">,</span> <span class="nf">container.split</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)[</span><span class="m">-1</span><span class="p">]))</span> <span class="n">if</span> <span class="n">container</span> <span class="n">else</span> <span class="s">&quot;&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="m">100</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">walltime</span> <span class="o">=</span> <span class="s">&quot;1h&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">mem</span> <span class="o">=</span> <span class="s">&quot;16G&quot;</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="m">20</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_z_format_1</span><span class="p">]</span>

<span class="n">import</span> <span class="n">pandas</span> <span class="n">as</span> <span class="n">pd</span>
<span class="n">s_path</span> <span class="o">=</span> <span class="nf">pd.read_csv</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">None</span><span class="p">)</span>
<span class="c1"># split_path = s_path[0].str.split(&quot;.&quot;, expand=True)</span>
<span class="c1"># ID = pd.DataFrame({&#39;ID&#39;: split_path[1]})</span>
<span class="c1"># path = pd.DataFrame({&#39;path&#39;: s_path[0]})</span>
<span class="c1"># input_df = pd.concat([ID, path], axis=1)</span>
<span class="n">input_df</span> <span class="o">=</span> <span class="nf">s_path.values.tolist</span><span class="p">()</span>

<span class="n">parameter</span><span class="o">:</span> <span class="n">GWAS_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">LD_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">TAD_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1">#input: susie_path</span>
<span class="n">input</span><span class="o">:</span> <span class="p">[</span><span class="n">x</span> <span class="n">for</span> <span class="n">x</span> <span class="n">in</span> <span class="n">input_df</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="m">1</span>
<span class="n">output</span><span class="o">:</span> <span class="n">f</span><span class="s">&#39;{cwd}/{step_name[:-2]}/{_input:bnn}.twas_z.rds&#39;</span>
<span class="n">task</span><span class="o">:</span> <span class="n">trunk_workers</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="n">walltime</span> <span class="o">=</span> <span class="n">walltime</span><span class="p">,</span> <span class="n">mem</span> <span class="o">=</span> <span class="n">mem</span><span class="p">,</span> <span class="n">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="n">R</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:nn}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:nn}.stdout&#39;</span>

    <span class="nf">options</span><span class="p">(</span><span class="n">bedtools.path</span> <span class="o">=</span> <span class="s">&quot;/home/aw3600/bedtools2/bin/&quot;</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">pecotmr</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>

    <span class="n">susie_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">data.frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>
    <span class="n">TAD_region</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">TAD_path</span><span class="o">:</span><span class="n">ar</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="o">=</span><span class="s">&quot;gene_id&quot;</span><span class="p">)</span>
    <span class="n">gene_name</span> <span class="o">=</span> <span class="n">susie_path</span><span class="o">$</span><span class="n">ID</span>
    <span class="n">region</span> <span class="o">=</span> <span class="n">TAD_region</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">ID</span><span class="o">==</span><span class="n">gene_name</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="o">=</span><span class="s">&quot;#chr&quot;</span><span class="p">)</span>
    <span class="n">chr</span> <span class="o">=</span> <span class="nf">str_sub</span><span class="p">(</span><span class="n">region</span><span class="o">$</span><span class="n">chr</span><span class="p">,</span><span class="m">4</span><span class="p">)</span>
    <span class="n">AD_dataset</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">GWAS_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="s">&quot;/ADGWAS_Bellenguez_2022.&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;/ADGWAS2022.chr&quot;</span><span class="p">,</span><span class="n">chr</span><span class="p">,</span><span class="s">&quot;.sumstat.tsv&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
    <span class="n">GWAS_data</span> <span class="o">=</span> <span class="n">AD_dataset</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">chr</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="n">chromosome</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">pos</span> <span class="o">=</span> <span class="nf">as.character</span><span class="p">(</span><span class="n">position</span><span class="p">))</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">chromosome</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;A1&quot;</span><span class="o">=</span><span class="s">&quot;ref&quot;</span><span class="p">,</span><span class="s">&quot;A2&quot;</span><span class="o">=</span><span class="s">&quot;alt&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">mutate</span><span class="p">(</span><span class="n">z</span><span class="o">=</span><span class="n">beta</span><span class="o">/</span><span class="n">se</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">position</span><span class="p">)</span>
    <span class="n">LD_list</span> <span class="o">=</span> <span class="nf">read.table</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">LD_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
    <span class="n">LD_block_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">LD_list</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;_&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">LD_list</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">rename</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="o">=</span><span class="s">&quot;1&quot;</span><span class="p">,</span><span class="s">&quot;start&quot;</span><span class="o">=</span><span class="s">&quot;2&quot;</span><span class="p">,</span><span class="s">&quot;end&quot;</span><span class="o">=</span><span class="s">&quot;3&quot;</span><span class="p">,</span><span class="s">&quot;variants&quot;</span><span class="o">=</span><span class="s">&quot;V1&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span> <span class="o">=</span> <span class="s">&quot;V2&quot;</span><span class="p">)</span>
    <span class="n">twas_z</span> <span class="o">=</span> <span class="nf">twas_scan</span><span class="p">(</span><span class="n">susie_path</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span> <span class="n">GWAS_data</span><span class="p">,</span> <span class="n">LD_block_path</span><span class="p">)</span>
    <span class="nf">saveRDS</span><span class="p">(</span><span class="n">twas_z</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_z_format_2</span><span class="p">]</span>
<span class="c1"># Path to the input molecular phenotype data.</span>
<span class="n">input</span><span class="o">:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s">&quot;all&quot;</span>
<span class="n">output</span><span class="o">:</span> <span class="n">f</span><span class="s">&#39;{cwd}/{step_name[:-2]}/twas_z_files.txt&#39;</span>
<span class="n">python</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output:n}.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
<span class="n">import</span> <span class="n">pandas</span> <span class="n">as</span> <span class="n">pd</span>
<span class="nf">pd.DataFrame</span><span class="p">({</span><span class="s">&quot;output&quot;</span> <span class="o">:</span> <span class="p">[</span><span class="o">$</span><span class="p">[</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">,]]})</span><span class="nf">.to_csv</span><span class="p">(</span><span class="s">&quot;$[_output]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="n">False</span> <span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="n">False</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">twas_candidate_gene</span><span class="p">]</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">weights_path</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">pval_threshold</span> <span class="o">=</span> <span class="m">0.05</span>
<span class="n">parameter</span><span class="o">:</span> <span class="n">cpip_cutoff</span> <span class="o">=</span> <span class="m">0.5</span>
<span class="n">input</span><span class="o">:</span> <span class="n">weights_path</span>
<span class="n">output</span><span class="o">:</span> <span class="n">cand_genes</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{cwd:a}/{step_name}.rds&#39;</span><span class="p">,</span>
        <span class="n">mr_output</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{cwd:a}/mr_output.rds&#39;</span>
<span class="n">task</span><span class="o">:</span> <span class="n">trunk_workers</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="n">walltime</span> <span class="o">=</span> <span class="n">walltime</span><span class="p">,</span> <span class="n">mem</span> <span class="o">=</span> <span class="n">mem</span><span class="p">,</span> <span class="n">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{step_name}_{_output:bn}&#39;</span>
<span class="n">R</span><span class="o">:</span> <span class="n">expand</span><span class="o">=</span> <span class="s">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">qvalue</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">doMC</span><span class="p">)</span>
    <span class="nf">library</span><span class="p">(</span><span class="n">pecotmr</span><span class="p">)</span>

    <span class="c1">####</span>
    <span class="n">susie_list</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="o">$</span><span class="p">{</span><span class="n">susie_path</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
    <span class="n">susie_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">susie_list</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">susie_list</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>

    <span class="c1">#detectCores()</span>
    <span class="nf">registerDoMC</span><span class="p">(</span><span class="m">100</span><span class="p">)</span>
    <span class="n">weights_file</span> <span class="o">=</span> <span class="nf">fread</span><span class="p">(</span><span class="o">$</span><span class="p">{</span>_<span class="n">input</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
    <span class="n">weights_file_path</span> <span class="o">=</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">weights_file</span><span class="o">$</span><span class="n">V1</span><span class="p">,</span><span class="s">&quot;\\.&quot;</span><span class="p">,</span><span class="n">simplify</span><span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="m">2</span><span class="p">]</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">weights_file</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nf">setNames</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;path&quot;</span><span class="p">))</span>
    <span class="c1">#ptm = proc.time()</span>
    <span class="n">gene_pq</span> <span class="o">=</span> <span class="nf">foreach</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="nf">dim</span><span class="p">(</span><span class="n">weights_file</span><span class="p">)[</span><span class="m">1</span><span class="p">],</span> <span class="n">.combine</span><span class="o">=</span><span class="n">rbind</span><span class="p">)</span> <span class="o">%dopar%</span> <span class="p">{</span>
       <span class="n">weights_pq</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">weights_file_path</span><span class="o">$</span><span class="n">path</span><span class="p">[</span><span class="n">k</span><span class="p">])</span><span class="o">$</span><span class="n">gene_weights_pq</span>
    <span class="p">}</span>
    <span class="c1">#proc.time() - ptm</span>
    <span class="n">padj</span> <span class="o">=</span> <span class="nf">apply</span><span class="p">(</span><span class="n">gene_pq</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval&quot;</span><span class="p">)],</span><span class="m">2</span><span class="p">,</span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">p.adjust</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;bonferroni&quot;</span><span class="p">,))</span><span class="o">%&gt;%</span><span class="nf">data.frame</span><span class="p">()</span>
    <span class="n">gene_pq_adj</span> <span class="o">=</span> <span class="n">gene_pq</span><span class="o">%&gt;%</span><span class="nf">cbind</span><span class="p">(</span><span class="n">.</span><span class="p">,</span><span class="n">padj</span><span class="p">)</span>
    <span class="nf">names</span><span class="p">(</span><span class="n">gene_pq_adj</span><span class="p">)[</span><span class="m">15</span><span class="o">:</span><span class="m">18</span><span class="p">]</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;susie_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;lasso_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;enet_pval_adj&quot;</span><span class="p">,</span><span class="s">&quot;mr_ash_pval_adj&quot;</span><span class="p">)</span>
    <span class="n">gene.ref</span> <span class="o">=</span> <span class="nf">read.table</span><span class="p">(</span><span class="s">&quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list&quot;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
    <span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_id</span> <span class="o">=</span> <span class="n">gene.ref</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">gene_pq_adj</span><span class="o">$</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene.ref</span><span class="o">$</span><span class="n">V4</span><span class="p">),</span><span class="s">&quot;V5&quot;</span><span class="p">]</span>

    <span class="n">cand_genes</span> <span class="o">=</span> <span class="kc">NULL</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span> <span class="o">=</span> <span class="n">gene_pq_adj</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">susie</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">susie_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">lasso_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">enet</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">enet_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">mr_ash_pval</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">susie_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">lasso_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">enet_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="o">=</span> <span class="n">cand_genes</span><span class="o">$</span><span class="n">gene_pq_adj</span><span class="o">%&gt;%</span><span class="nf">filter</span><span class="p">(</span><span class="n">mr_ash_pval_adj</span><span class="o">&lt;$</span><span class="p">{</span><span class="n">pval_threshold</span><span class="p">})</span><span class="o">%&gt;%</span><span class="nf">select</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span><span class="n">gene_id</span><span class="p">)</span>

    <span class="c1"># cand_genes$susie_format_input = twas_mr_format_input(cand_genes$susie$gene_name,${susie_path:ar},weights_file_path)</span>
    <span class="c1"># cand_genes$lasso_format_input = twas_mr_format_input(cand_genes$lasso$gene_name,${susie_path:ar},weights_file_path)</span>
    <span class="c1"># cand_genes$enet_format_input = twas_mr_format_input(cand_genes$enet$gene_name,${susie_path:ar},weights_file_path)</span>
    <span class="c1"># cand_genes$mr_ash_format_input = twas_mr_format_input(cand_genes$mr_ash$gene_name,${susie_path:ar},weights_file_path)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>
    <span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span> <span class="o">=</span> <span class="nf">twas_mr_format_input</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj</span><span class="p">,</span><span class="n">susie_path</span><span class="p">,</span><span class="n">weights_file_path</span><span class="p">)</span>

    <span class="nf">saveRDS</span><span class="p">(</span><span class="n">cand_genes</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="p">[</span><span class="m">0</span><span class="p">]</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>

    <span class="n">mr_output</span> <span class="o">=</span> <span class="kc">NULL</span>
    <span class="c1"># mr_output$susie = fine_mr(cand_genes$susie_format_input,${cpip_cutoff})</span>
    <span class="c1"># mr_output$lasso = fine_mr(cand_genes$lasso_format_input,${cpip_cutoff})</span>
    <span class="c1"># mr_output$enet = fine_mr(cand_genes$enet_format_input,${cpip_cutoff})</span>
    <span class="c1"># mr_output$mr_ash = fine_mr(cand_genes$mr_ash_format_input,${cpip_cutoff})</span>
    <span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span><span class="p">))</span>
    <span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">susie_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">susie_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>
    <span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span><span class="p">))</span>
    <span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">lasso_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">lasso_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>
    <span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span><span class="p">))</span>
    <span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">enet_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">enet_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>
    <span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span><span class="p">))</span>
    <span class="p">{</span><span class="n">mr_output</span><span class="o">$</span><span class="n">mr_ash_adj</span> <span class="o">=</span> <span class="nf">fine_mr</span><span class="p">(</span><span class="n">cand_genes</span><span class="o">$</span><span class="n">mr_ash_adj_format_input</span><span class="p">,</span><span class="o">$</span><span class="p">{</span><span class="n">cpip_cutoff</span><span class="p">})}</span>

    <span class="nf">saveRDS</span><span class="p">(</span><span class="n">mr_output</span><span class="p">,</span> <span class="o">$</span><span class="p">{</span>_<span class="n">output</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="o">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s">&#39;xz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./code/pecotmr_integration"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="SuSiE_enloc.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">xQTL and GWAS region alignment</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="intact.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">TWAS and COLOC integration</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>