
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Hidden factor analysis &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Covariate data formatting" href="covariate_formatting.html" />
    <link rel="prev" title="Covariate data preprocessing" href="../covariate_preprocessing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/qtl_association_testing.html">
   QTL association analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with individual level data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     Data overview
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/data_preprocessing/covariate/covariate_hidden_factor.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/covariate/covariate_hidden_factor.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/covariate/covariate_hidden_factor.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#main-outputs">
     Main outputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#other-outputs">
     Other outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#number-of-factors-to-use">
   Number of factors to use
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#peer">
     PEER
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#apex">
     APEX
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#pca">
     PCA
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     PEER
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     PCA
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     APEX
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#implementation">
   Implementation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reference">
     Reference
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#bicv-factor-analysis-with-apex">
     BiCV factor analysis with APEX
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="hidden-factor-analysis">
<h1>Hidden factor analysis<a class="headerlink" href="#hidden-factor-analysis" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This workflow implements 3 procedures for hidden factor analysis from omcis data:</p>
<ol class="simple">
<li><p>The <a class="reference external" href="https://github.com/PMBio/peer/wiki/Tutorial">Probabilistic Estimation of Expression Residuals (PEER) method</a>, a method also used for GTEx eQTL data analysis.</p></li>
<li><p>Factor analysis using Bi-Cross validation, Owen, Art &amp; Wang, Jingshu. (2015). Bi-Cross-Validation for Factor Analysis. Statistical Science. 31. 10.1214/15-STS539. with software package <code class="docutils literal notranslate"><span class="pre">APEX</span></code> (Corbin Quick, Li Guan, Zilin Li, Xihao Li, Rounak Dey, Yaowu Liu, Laura Scott, Xihong Lin, bioRxiv 2020.12.18.423490; doi: <a class="reference external" href="https://doi.org/10.1101/2020.12.18.423490">https://doi.org/10.1101/2020.12.18.423490</a>)</p></li>
<li><p>PCA with automatic determination of the number of factors to use. This is mainly inspired by a <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02761-4">recent benchmark from Jessica Li’s group</a>.</p></li>
</ol>
<p>Overall, we will pick PCA based approach for the xQTL project, although additional considerations should be taken for single-cell eQTL analysis as investigated in <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-023-02873-5">this paper</a>.</p>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Molecular phenotype in <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> file format with indexing:</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">readr:</span><span class="p">:</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;data/example_data.bed.gz&quot;</span><span class="p">,</span><span class="n">show_col_types</span><span class="o">=</span> <span class="n">F</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 3 × 8</caption>
<thead>
	<tr><th scope=col>#chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>gene_name</th><th scope=col>Sample1</th><th scope=col>Sample2</th><th scope=col>Sample3</th><th scope=col>Sample4</th></tr>
	<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>1</td><td>  1</td><td>100</td><td>Gene1</td><td> -1.627455</td><td> 1.684725</td><td> -0.0807924</td><td> 1.510522</td></tr>
	<tr><td>1</td><td>101</td><td>200</td><td>Gene2</td><td>-12.626715</td><td>17.746163</td><td>-14.2519447</td><td> 5.254536</td></tr>
	<tr><td>1</td><td>201</td><td>300</td><td>Gene3</td><td>  3.337390</td><td>-7.259074</td><td>  6.2643526</td><td>-1.375911</td></tr>
</tbody>
</table>
</div></div>
</div>
<ol class="simple">
<li><p>A covariate file in text format with the first column <code class="docutils literal notranslate"><span class="pre">#id</span></code> that lists the name of all covariates, followed by other columns each being a sample with sample names matching that in the molecular phenotype file:</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">readr:</span><span class="p">:</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;data/example_cov.txt&quot;</span><span class="p">,</span><span class="n">show_col_types</span><span class="o">=</span> <span class="n">F</span><span class="p">)[,</span><span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 2 × 8</caption>
<thead>
	<tr><th scope=col>#id</th><th scope=col>Sample1</th><th scope=col>Sample2</th><th scope=col>Sample3</th><th scope=col>Sample4</th><th scope=col>Sample5</th><th scope=col>Sample6</th><th scope=col>Sample7</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>Cov1</td><td> 1.6243454</td><td>-2.301539</td><td> 1.462108</td><td>-1.0998913</td><td>-1.100619</td><td>-0.6837279</td><td>-0.6916608</td></tr>
	<tr><td>Cov2</td><td>-0.6117564</td><td> 1.744812</td><td>-2.060141</td><td>-0.1724282</td><td> 1.144724</td><td>-0.1228902</td><td>-0.3967535</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Covariates file is optional to the PEER method but it should be supplied in order to construct a proper PEER model in the context of xQTL mapping.</p>
<ol class="simple">
<li><p>For <code class="docutils literal notranslate"><span class="pre">APEX</span></code> we need an indexed VCF file to get the command to work, although this file is not used.</p></li>
</ol>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<div class="section" id="main-outputs">
<h3>Main outputs<a class="headerlink" href="#main-outputs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.cov.gz</span></code>: hidden factors inferred.</p></li>
</ul>
</div>
<div class="section" id="other-outputs">
<h3>Other outputs<a class="headerlink" href="#other-outputs" title="Permalink to this headline">¶</a></h3>
<p>PEER:</p>
<ul class="simple">
<li><p>Diagnosis polts for check of convergence</p></li>
<li><p>peer precision /residuals /weights</p></li>
</ul>
</div>
</div>
<div class="section" id="number-of-factors-to-use">
<h2>Number of factors to use<a class="headerlink" href="#number-of-factors-to-use" title="Permalink to this headline">¶</a></h2>
<div class="section" id="peer">
<h3>PEER<a class="headerlink" href="#peer" title="Permalink to this headline">¶</a></h3>
<p>By default the pipeline applies recommendations from <a class="reference external" href="https://gtexportal.org/home/documentationPage">GTEx</a>:</p>
<blockquote>
<div><p>A set of covariates identified using the Probabilistic Estimation of Expression Residuals (PEER) method (Stegle et al., PLoS Comp. Biol., 2010 ), calculated for the normalized expression matrices (described below). For eQTL analyses, the number of PEER factors was determined as function of sample size (N): 15 factors for N&lt;150, 30 factors for 150≤ N&lt;250, 45 factors for 250≤ N&lt;350, and 60 factors for N≥350, as a result of optimizing for the number of eGenes discovered. For sQTL analyses, 15 PEER factors were computed for each tissue.</p>
</div></blockquote>
<p>Although one can specify fixed number of factors using <code class="docutils literal notranslate"><span class="pre">--N</span></code> option in this pipeline.</p>
</div>
<div class="section" id="apex">
<h3>APEX<a class="headerlink" href="#apex" title="Permalink to this headline">¶</a></h3>
<p>We will do the same as PEER</p>
</div>
<div class="section" id="pca">
<h3>PCA<a class="headerlink" href="#pca" title="Permalink to this headline">¶</a></h3>
<p>We will use Marchenko’s approach to determine the number of PC to use. FIXME: citation.</p>
</div>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>An MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1yjTwoO0DYGi-J9ouMsh9fHKfDmsXJ_4I?usp=sharing">google drive</a>.
The singularity image (sif) for running this MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1mLOS3AVQM8yTaWtCbO8Q3xla98Nr5bZQ">google drive</a></p>
<p>FIXME: need updated links</p>
<div class="section" id="id1">
<h3>PEER<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>Because the MWE only contains 10 genes but 400+ samples. The computed N will be far greater than the number of genes. Therefore in the MWE the N is fixed to be 3.</p>
<p>FIXME: need updated examples</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">PEER_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">PEER</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
   <span class="o">--</span><span class="n">phenoFile</span> <span class="n">ALL</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">chr12</span><span class="o">.</span><span class="n">mol_phe</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>  \
   <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">PEER</span><span class="o">.</span><span class="n">sif</span>  \
   <span class="o">--</span><span class="n">N</span> <span class="mi">3</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">./</span><span class="n">output</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">./output</span>
├── MWE.Cov_PEER.PEER.cov
├── MWE.Cov_PEER.PEER.cov.stderr
├── MWE.Cov_PEER.PEER.cov.stdout
├── MWE.Cov_PEER.PEER_diagnosis.pdf
├── MWE.Cov_PEER.PEER.peer_precisions.txt
├── MWE.Cov_PEER.PEER.peer_residuals.txt
└── MWE.Cov_PEER.PEER.peer_weights.txt

0 directories, 7 files
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">Cov_PEER</span><span class="o">.</span><span class="n">PEER_diagnosis</span><span class="o">.</span><span class="n">pdf</span> <span class="o">-</span><span class="n">s</span> <span class="n">png</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ./peer_output/example_data.bed_diagnosis_plot.jpeg (22.2 KiB):</div></div><img alt="../../../_images/covariate_hidden_factor_14_1.jpg" src="../../../_images/covariate_hidden_factor_14_1.jpg" />
<div class="output text_html"><div class="sos_hint">> ./peer_output/example_data.bed_diagnosis_plot.jpeg:</div></div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>bash: ./peer_output/example_data.bed_diagnosis_plot.jpeg: Permission denied
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span><span class="n">MWE</span><span class="o">.</span><span class="n">Cov_PEER</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">stdout</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>PEER: loading input data... PEER: loading covariate file ...done!
PEER: Input summary: 200 samples with 100 genes and 2 covariates 
Setting initialization parameters ...Done.
PEER: estimating hidden confounders (30)
	iteration 0/1000
	iteration 1/1000
	iteration 2/1000
	iteration 3/1000
	iteration 4/1000
	iteration 5/1000
Converged (var(residuals)) after 5 iterations

Exporting results ... Done.
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="id2">
<h3>PCA<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>In our pipeline, it is designed to include the pca as part of our cov input. This can be achieved by the merge_covariate step</p>
<p><strong>FIXME: complete MWE</strong></p>
</div>
<div class="section" id="id3">
<h3>APEX<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">BiCV_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">BiCV</span> \
   <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
   <span class="o">--</span><span class="n">phenoFile</span> <span class="n">ALL</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">chr12</span><span class="o">.</span><span class="n">mol_phe</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>  \
   <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">apex</span><span class="o">.</span><span class="n">sif</span>  \
   <span class="o">--</span><span class="n">N</span> <span class="mi">3</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">covariate_hidden_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run covariate_hidden_factor.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  Marchenko_PC
  PEER
  BiCV

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files. MUST BE FULL
                        PATH
  --covFile VAL (as path, required)
                        Merged Covariates File
  --phenoFile VAL (as path, required)
                        Path to the input molecular phenotype data.
  --name  f&#39;{phenoFile:bnn}.{covFile:bn}&#39;

  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option
  --entrypoint {(&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + container.split(&#39;/&#39;)[-1][:-4]) if container.endswith(&#39;.sif&#39;) else f&#39;&#39;}


Sections
  *_1:
  Marchenko_PC_2:
  PEER_2:
    Workflow Options:
      --N 0 (as int)
                        N PEER factors, If do not specify or specified as 0,
                        default values suggested by GTEx (based on different
                        sample size) will be used
      --iteration 1000 (as int)
                        Default values from PEER software: The number of max
                        iteration
      --tol 0.001 (as float)
                        Prior parameters parameter: Alpha_a = 0.001 parameter:
                        Alpha_b = 0.1 parameter: Eps_a = 0.1 parameter: Eps_b =
                        10.0 Tolarance parameters
      --[no-]r2-tol (default to False)
                        parameter: var_tol = 0.00001 minimum variance explained
                        criteria to drop factors while training
      --convergence-mode fast
                        Convergence mode: Convergence mode for MOFAr &quot;slow&quot;,
                        &quot;medium&quot; or &quot;fast&quot;, corresponding to 1e-5%, 1e-4% or
                        1e-3% deltaELBO change.
  PEER_3:
  BiCV_2:
  BiCV_3:
    Workflow Options:
      --N 0 (as int)
                        N factors, if not specify, calculated based on sample
                        size according to GTeX
      --iteration 10 (as int)
                        The number of iteration
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="implementation">
<h2>Implementation<a class="headerlink" href="#implementation" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># The output directory for generated files. MUST BE FULL PATH</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># Merged Covariates File</span>
<span class="kn">parameter:</span> <span class="n">covFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Path to the input molecular phenotype data.</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">phenoFile</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">covFile</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span><span class="p">{(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">])</span> <span class="k">if</span> <span class="n">container</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.sif&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;&#39;</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="o">*</span><span class="n">_1</span><span class="p">(</span><span class="n">computing</span> <span class="n">residual</span> <span class="n">on</span> <span class="n">merged</span> <span class="n">covariates</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">mean_impute_missing</span> <span class="o">=</span> <span class="kc">False</span>
<span class="kn">input:</span> <span class="n">phenoFile</span><span class="p">,</span> <span class="n">covFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.residual.bed.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.stdout&#39;</span> <span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>

    <span class="n">mean_impute</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">d</span><span class="p">){</span>
      <span class="n">f</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">mean</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">f</span><span class="p">))</span> <span class="n">d</span><span class="p">[,</span><span class="n">i</span><span class="p">][</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">d</span><span class="p">[,</span><span class="n">i</span><span class="p">]))]</span> <span class="o">&lt;-</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
      <span class="k">return</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">covariate</span><span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> 

    <span class="c1"># Extract samples in both files (Outliers will be removed in here as they are no longer presented in the header of factor)</span>
    <span class="n">extraction_sample_list</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">),</span> <span class="n">colnames</span><span class="p">(</span><span class="n">covariate</span><span class="p">))</span> 
    
    
    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">extraction_sample_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">){</span>
      <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;No samples are overlapped in two files!&quot;</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="c1">## Report sample counts:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">paste</span><span class="p">((</span><span class="n">ncol</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span> <span class="o">-</span> <span class="mi">4</span><span class="p">),</span> <span class="s2">&quot;samples are in the phenotype file&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">paste</span><span class="p">((</span><span class="n">ncol</span><span class="p">(</span><span class="n">covariate</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;samples are in the covariate file&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>

    <span class="c1">## Report identical samples:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">extraction_sample_list</span><span class="p">),</span> <span class="s2">&quot;samples overlap between phenotype &amp; covariate files and are included in the analysis:&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">extraction_sample_list</span><span class="p">)</span>

    <span class="c1">## Report non-overlapping samples :</span>
    <span class="n">covariate_missing</span> <span class="o">=</span> <span class="n">covariate</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">all_of</span><span class="p">(</span><span class="s1">&#39;#id&#39;</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">all_of</span><span class="p">(</span><span class="n">extraction_sample_list</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">covariate_missing</span><span class="p">),</span> <span class="s2">&quot;samples in the covariate file are missing from the phenotype file:&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">covariate_missing</span><span class="p">))</span>

    <span class="n">pheno_missing</span> <span class="o">=</span> <span class="n">pheno</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">all_of</span><span class="p">(</span><span class="n">extraction_sample_list</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">pheno_missing</span><span class="p">),</span> <span class="s2">&quot;samples in the phenotype file are missing from the covariate file:&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">pheno_missing</span><span class="p">))</span>

    <span class="c1"># Subset the data:</span>
    <span class="n">covariate</span> <span class="o">=</span> <span class="n">covariate</span><span class="p">[,</span><span class="n">extraction_sample_list</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">t</span><span class="p">()</span>
    <span class="n">pheno_id</span> <span class="o">=</span> <span class="n">pheno</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">)</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">pheno</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">all_of</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">covariate</span><span class="p">)))</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">t</span><span class="p">()</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;T&quot;</span> <span class="k">if</span> <span class="n">mean_impute_missing</span> <span class="k">else</span> <span class="s2">&quot;F&quot;</span><span class="p">})</span> <span class="p">{</span>
      <span class="n">pheno</span> <span class="o">=</span> <span class="n">mean_impute</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
      <span class="k">if</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">pheno</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">){</span> <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;NA in phenotype input is not allowed!&quot;</span><span class="p">)</span> <span class="p">}</span>
    <span class="p">}</span>
    <span class="c1"># Get residual </span>
    <span class="n">pheno_resid</span> <span class="o">=</span> <span class="o">.</span><span class="n">lm</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">covariate</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="n">pheno</span><span class="p">)</span><span class="err">$</span><span class="n">residuals</span>
    <span class="n">pheno_output</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">pheno_id</span><span class="p">,</span> <span class="n">pheno_resid</span><span class="o">%&gt;%</span><span class="n">t</span><span class="p">())</span>
    <span class="n">pheno_output</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:n}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
  
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">bgzip</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span>
    <span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">bed</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>

    <span class="n">stdout</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">stdout</span> 
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span> <span class="p">;</span> <span class="n">do</span> 
    <span class="n">echo</span> <span class="s2">&quot;output_info: $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
    <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
    <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
    <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
    <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
    <span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">Marchenko_PC_2</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.Marchenko_PC.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;PCAtools&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s1">&#39;BiocSingular&#39;</span><span class="p">)</span>
    <span class="n">residExpPath</span>  <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span>
    <span class="n">covPath</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{covFile}</span><span class="s2">&quot;</span>
    <span class="n">residExpDF</span> <span class="o">&lt;-</span> <span class="n">read_delim</span><span class="p">(</span><span class="n">residExpPath</span><span class="p">,</span> <span class="n">show_col_types</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">covDF</span> <span class="o">&lt;-</span> <span class="n">read_delim</span><span class="p">(</span><span class="n">covPath</span><span class="p">,</span> <span class="n">show_col_types</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="n">commonMPSamples</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">covDF</span><span class="p">),</span> <span class="n">colnames</span><span class="p">(</span><span class="n">residExpDF</span><span class="p">))</span>
    <span class="n">covDFcommon</span> <span class="o">&lt;-</span> <span class="n">cbind</span><span class="p">(</span><span class="n">covDF</span><span class="p">[,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">covDF</span><span class="p">[,</span> <span class="n">commonMPSamples</span><span class="p">])</span>
  
    <span class="n">residExpPC</span> <span class="o">&lt;-</span> <span class="n">pca</span><span class="p">(</span>
        <span class="n">residExpDF</span><span class="p">[,</span><span class="n">commonMPSamples</span><span class="p">],</span> <span class="c1"># The first four columns are: chr, start, end, and gene_id; so we skip those.</span>
        <span class="n">scale</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
        <span class="n">center</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
        <span class="n">BSPARAM</span> <span class="o">=</span> <span class="n">ExactParam</span><span class="p">())</span>
    <span class="n">M</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">residExpDF</span><span class="p">[,</span> <span class="n">commonMPSamples</span><span class="p">],</span> <span class="mi">1</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">X</span><span class="p">){</span> <span class="p">(</span><span class="n">X</span> <span class="o">-</span> <span class="n">mean</span><span class="p">(</span><span class="n">X</span><span class="p">))</span><span class="o">/</span><span class="n">sqrt</span><span class="p">(</span><span class="n">var</span><span class="p">(</span><span class="n">X</span><span class="p">))});</span>
    <span class="n">residSigma2</span> <span class="o">&lt;-</span> <span class="n">var</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">vector</span><span class="p">(</span><span class="n">M</span><span class="p">));</span>
    <span class="n">paste</span><span class="p">(</span><span class="s1">&#39;sigma2:&#39;</span><span class="p">,</span> <span class="n">residSigma2</span><span class="p">)</span>
    
    <span class="n">MPPCNum</span> <span class="o">&lt;-</span> <span class="n">chooseMarchenkoPastur</span><span class="p">(</span>
        <span class="o">.</span><span class="n">dim</span> <span class="o">=</span> <span class="n">dim</span><span class="p">(</span><span class="n">residExpDF</span><span class="p">[,</span> <span class="n">commonMPSamples</span><span class="p">]),</span> <span class="n">var</span><span class="o">.</span><span class="n">explained</span><span class="o">=</span><span class="n">residExpPC</span><span class="err">$</span><span class="n">sdev</span><span class="o">^</span><span class="mi">2</span><span class="p">,</span> <span class="n">noise</span><span class="o">=</span><span class="n">residSigma2</span><span class="p">)</span>
    
    <span class="n">MPPCsDF</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">residExpPC</span><span class="err">$</span><span class="n">rotated</span><span class="p">[,</span> <span class="mi">1</span><span class="p">:</span><span class="n">MPPCNum</span><span class="p">])</span>

    <span class="n">MPColMatrix</span> <span class="o">&lt;-</span>  <span class="n">matrix</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">rep</span><span class="p">(</span><span class="s1">&#39;Hidden_Factor_PC&#39;</span><span class="p">,</span> <span class="n">times</span><span class="o">=</span><span class="n">MPPCNum</span><span class="p">),</span> <span class="n">seq</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">MPPCNum</span><span class="p">)),</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">nrow</span><span class="o">=</span><span class="n">MPPCNum</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">MPPCsDF</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">apply</span><span class="p">(</span><span class="n">MPColMatrix</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">X</span><span class="p">){</span><span class="k">return</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">X</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">X</span><span class="p">[</span><span class="mi">2</span><span class="p">]))})</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">MPPCsDF</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">residExpDF</span><span class="p">[,</span> <span class="mi">5</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">residExpDF</span><span class="p">)])</span>
    <span class="c1"># Add #id Column</span>
    <span class="n">MPPCsDF</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">MPPCsDF</span><span class="p">))</span>
    <span class="n">MPPCsDF</span><span class="err">$</span><span class="nb">id</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">MPPCsDF</span><span class="p">)</span>
    <span class="n">MPPCsDF</span> <span class="o">&lt;-</span> <span class="n">MPPCsDF</span> <span class="o">%&gt;%</span> <span class="n">select</span><span class="p">(</span><span class="nb">id</span><span class="p">,</span> <span class="n">everything</span><span class="p">())</span> <span class="o">%&gt;%</span> <span class="n">rename</span><span class="p">(</span><span class="s2">&quot;#id&quot;</span> <span class="o">=</span> <span class="s2">&quot;id&quot;</span><span class="p">)</span>
    <span class="n">write_delim</span><span class="p">((</span><span class="n">rbind</span><span class="p">(</span><span class="n">covDFcommon</span><span class="p">,</span> <span class="n">MPPCsDF</span><span class="p">)),</span> <span class="s2">&quot;$</span><span class="si">{_output}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
        <span class="n">stdout</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">PEER_2</span><span class="p">]</span>
<span class="c1"># N PEER factors, If do not specify or specified as 0, default values suggested by </span>
<span class="c1"># GTEx (based on different sample size) will be used</span>
<span class="kn">parameter:</span> <span class="n">N</span> <span class="o">=</span> <span class="mi">0</span>
<span class="c1"># Default values from PEER software:</span>
<span class="c1">## The number of max iteration</span>
<span class="kn">parameter:</span> <span class="n">iteration</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="c1">### Prior parameters</span>
<span class="c1">#parameter: Alpha_a = 0.001</span>
<span class="c1">#parameter: Alpha_b = 0.1</span>
<span class="c1">#parameter: Eps_a = 0.1</span>
<span class="c1">#parameter: Eps_b = 10.0</span>
<span class="c1"># Tolarance parameters</span>
<span class="kn">parameter:</span> <span class="n">tol</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="c1">#parameter: var_tol = 0.00001</span>
<span class="c1"># minimum variance explained criteria to drop factors while training</span>
<span class="kn">parameter:</span> <span class="n">r2_tol</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Convergence mode: Convergence mode for MOFAr &quot;slow&quot;, &quot;medium&quot; or &quot;fast&quot;, corresponding to 1e-5%, 1e-4% or 1e-3% deltaELBO change.</span>
<span class="kn">parameter:</span> <span class="n">convergence_mode</span> <span class="o">=</span> <span class="s2">&quot;fast&quot;</span>
<span class="c1"># input is the residual file from the first step</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.PEER_MODEL.hd5&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">python:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="kn">from</span> <span class="nn">mofapy2.run.entry_point</span> <span class="kn">import</span> <span class="n">entry_point</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">os</span>
    <span class="c1">## Force the number of threads used in numpy.</span>
    <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;OMP_NUM_THREADS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{numThreads}</span><span class="s2">&quot;</span> <span class="c1"># export OMP_NUM_THREADS=4</span>
    <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;OPENBLAS_NUM_THREADS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{numThreads}</span><span class="s2">&quot;</span> <span class="c1"># export OPENBLAS_NUM_THREADS=4 </span>
    <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;MKL_NUM_THREADS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{numThreads}</span><span class="s2">&quot;</span> <span class="c1"># export MKL_NUM_THREADS=6</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="c1">## Data</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">},</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index_col</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">],</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="c1"># initialise the entry point</span>
    <span class="n">ent</span> <span class="o">=</span> <span class="n">entry_point</span><span class="p">()</span>
    <span class="c1"># Guess number of factor</span>
    <span class="c1"># Suggest the number of factors to use if no input value</span>
    <span class="n">num_factor</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">N</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">num_factor</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">150</span><span class="p">:</span>
            <span class="n">num_factor</span> <span class="o">=</span> <span class="mi">15</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">250</span><span class="p">:</span>
            <span class="n">num_factor</span> <span class="o">=</span> <span class="mi">30</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">350</span><span class="p">:</span>
            <span class="n">num_factor</span> <span class="o">=</span> <span class="mi">45</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">num_factor</span> <span class="o">=</span> <span class="mi">60</span>
    <span class="c1"># Set data</span>
    <span class="c1"># MOFA is a multi-view and multi-group inference framework. </span>
    <span class="c1"># If usig only a single view and a single group (as in PEER), the data needs to be embedded into a nested list</span>
    <span class="n">ent</span><span class="o">.</span><span class="n">set_data_matrix</span><span class="p">([[</span><span class="n">data</span><span class="o">.</span><span class="n">transpose</span><span class="p">()]],</span><span class="n">samples_names</span><span class="o">=</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()],</span> <span class="n">features_names</span><span class="o">=</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()])</span>
    <span class="n">ent</span><span class="o">.</span><span class="n">set_model_options</span><span class="p">(</span><span class="n">factors</span><span class="o">=</span> <span class="n">num_factor</span> <span class="p">,</span> <span class="n">spikeslab_weights</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ard_weights</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ent</span><span class="o">.</span><span class="n">set_train_options</span><span class="p">(</span><span class="nb">iter</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">iteration</span><span class="p">},</span> <span class="n">convergence_mode</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{convergence_mode}</span><span class="s2">&quot;</span> <span class="err">$</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;,dropR2 = </span><span class="si">{</span><span class="n">r2_tol</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">r2_tol</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="p">},</span> <span class="n">startELBO</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">freqELBO</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">tolerance</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">tol</span><span class="p">},</span> <span class="n">gpu_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
    <span class="n">ent</span><span class="o">.</span><span class="n">build</span><span class="p">()</span>
    <span class="n">ent</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
    <span class="n">ent</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">ar</span><span class="p">})</span>
    <span class="c1">## To fix issue https://github.com/cumc/xqtl-pipeline/issues/420</span>
    <span class="kn">import</span> <span class="nn">h5py</span>
    <span class="n">right_name</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s2">&quot;UTF-8&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>  <span class="n">ent</span><span class="o">.</span><span class="n">data_opts</span><span class="p">[</span><span class="s2">&quot;features_names&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> 
    <span class="n">new_hd5</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">ar</span><span class="p">},</span> <span class="s2">&quot;r+&quot;</span><span class="p">)</span>
    <span class="k">del</span> <span class="n">new_hd5</span><span class="p">[</span><span class="s2">&quot;features/view0&quot;</span><span class="p">]</span>
    <span class="n">new_hd5</span><span class="p">[</span><span class="s2">&quot;features&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span><span class="s2">&quot;view0&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">right_name</span><span class="p">))</span>
    <span class="n">new_hd5</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">PEER_3</span><span class="p">]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.PEER.gz&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.PEER.diag.pdf&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span>  <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="c1">####################</span>
    <span class="c1">## Load libraries ##</span>
    <span class="c1">####################</span>
    
    <span class="n">library</span><span class="p">(</span><span class="n">MOFA2</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">readr</span><span class="p">)</span>
    <span class="c1">################</span>
    <span class="c1">## Load model ##</span>
    <span class="c1">################</span>
    
    <span class="n">file</span> <span class="o">&lt;-</span> <span class="s2">&quot;$</span><span class="si">{_input}</span><span class="s2">&quot;</span>
    <span class="n">model</span> <span class="o">&lt;-</span> <span class="n">load_model</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    
    <span class="c1">####################################</span>
    <span class="c1">## (Optional) add sample metadata ##</span>
    <span class="c1">####################################</span>
    
    <span class="c1"># Important: </span>
    <span class="c1"># (1) the row names in the sample metadata data.frame have to match the sample names in the MOFA model</span>
    <span class="c1"># (2) the sample name has to contain at least two columns: </span>
    <span class="c1"># 		- sample: sample names</span>
    <span class="c1">#		- group: group name. Only required for multi-group inference. In your case just set the entire column to the same value</span>
    
    <span class="c1"># Load metadata</span>
    
    <span class="c1"># stopifnot(all(samples(model)[[1]]==metadata$sample))</span>
    <span class="c1"># samples_metadata(model) &lt;- metadata</span>
    
    <span class="c1">###############################</span>
    <span class="c1">## (Optional) Subset factors ##</span>
    <span class="c1">###############################</span>
    
    <span class="c1"># We can remove factors that explain little variance (in this case, we require at least 0.01%)</span>
    <span class="c1"># r2 &lt;- model@cache$variance_explained$r2_per_factor</span>
    <span class="c1"># factors &lt;- sapply(r2, function(x) x[,1]&gt;0.0001)</span>
    <span class="c1"># model &lt;- subset_factors(model, which(apply(factors,1,sum) &gt;= 1))</span>
    
    <span class="c1"># fetch factors in matrix format, with dimensions (nfactors,nsamples)</span>
    <span class="n">factors</span> <span class="o">&lt;-</span> <span class="n">get_factors</span><span class="p">(</span><span class="n">model</span><span class="p">)[[</span><span class="mi">1</span><span class="p">]]</span>
    <span class="n">dim</span><span class="p">(</span><span class="n">factors</span><span class="p">)</span>
    <span class="n">factors</span> <span class="o">=</span> <span class="n">as_tibble</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">factors</span><span class="p">),</span><span class="n">rownames</span> <span class="o">=</span> <span class="s2">&quot;#id&quot;</span><span class="p">)</span>
    <span class="n">covPath</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{covFile}</span><span class="s2">&quot;</span>
    <span class="n">options</span><span class="p">(</span><span class="n">readr</span><span class="o">.</span><span class="n">show_col_types</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="n">covDF</span> <span class="o">&lt;-</span> <span class="n">read_delim</span><span class="p">(</span><span class="n">covPath</span><span class="p">,</span> <span class="n">delim</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">commonMPSamples</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">covDF</span><span class="p">),</span> <span class="n">colnames</span><span class="p">(</span><span class="n">factors</span><span class="p">))</span>      
    <span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">covDF</span><span class="p">[,</span> <span class="n">commonMPSamples</span><span class="p">],</span> <span class="n">factors</span><span class="p">[,</span> <span class="n">commonMPSamples</span><span class="p">])</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output[0]}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>
    <span class="c1"># Plot save to vcf</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output[1]}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1">#############################</span>
    <span class="c1">## Plot variance explained ##</span>
    <span class="c1">#############################</span>
    
    <span class="c1"># Plot variance explained using individual factors</span>
    <span class="n">plot_variance_explained</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">factors</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">)</span>
    <span class="n">plot_variance_explained</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">factors</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span>
    
    <span class="c1"># Plot total variance explained using all factors</span>
    <span class="n">plot_variance_explained</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">plot_total</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)[[</span><span class="mi">2</span><span class="p">]]</span>
    
    <span class="c1">########################</span>
    <span class="c1">## Plot factor values ##</span>
    <span class="c1">########################</span>
    
    <span class="n">plot_factor</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">factor</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">plot_factors</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">factor</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
    
    <span class="c1">###########################</span>
    <span class="c1">## Plot feature loadings ##</span>
    <span class="c1">###########################</span>
    
    <span class="c1"># The weights or loadings provide a score for each gene on each factor. </span>
    <span class="c1"># Genes with no association with the factor are expected to have values close to zero</span>
    <span class="c1"># Genes with strong association with the factor are expected to have large absolute values. </span>
    <span class="c1"># The sign of the loading indicates the direction of the effect: a positive loading indicates that the feature is more active in the cells with positive factor values, and viceversa.</span>
    
    <span class="c1"># Plot the distribution of loadings for Factor 1.</span>
    <span class="n">plot_weights</span><span class="p">(</span><span class="n">model</span><span class="p">,</span>
      <span class="n">view</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
      <span class="n">factor</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
      <span class="n">nfeatures</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>     <span class="c1"># Top number of features to highlight</span>
      <span class="n">scale</span> <span class="o">=</span> <span class="n">T</span>           <span class="c1"># Scale loadings from -1 to 1</span>
    <span class="p">)</span>
    
    <span class="c1"># If you are not interested in the full distribution, but just on the top loadings:</span>
    <span class="n">plot_top_weights</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> 
      <span class="n">view</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> 
      <span class="n">factor</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> 
      <span class="n">nfeatures</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
      <span class="n">scale</span> <span class="o">=</span> <span class="n">T</span>
    <span class="p">)</span>
    
    <span class="c1">######################################</span>
    <span class="c1">## Plot correlation between factors ##</span>
    <span class="c1">######################################</span>
    
    <span class="n">plot_factor_cor</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
    
    <span class="c1">###################</span>
    <span class="c1">## Fetch factors ##</span>
    <span class="c1">###################</span>
    <span class="n">invisible</span><span class="p">(</span><span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">())</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;output_info $i &quot;</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_size:&quot;</span> <span class="err">`</span><span class="n">ls</span> <span class="o">-</span><span class="n">lh</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">5</span>  <span class="o">-</span><span class="n">d</span>  <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_rows:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span> <span class="o">-</span><span class="n">d</span> <span class="s2">&quot; &quot;</span><span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_column:&quot;</span> <span class="err">`</span><span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">w</span> <span class="err">`</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">echo</span> <span class="s2">&quot;output_preview:&quot;</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">stdout</span><span class="p">;</span>
        <span class="n">zcat</span> <span class="err">$</span><span class="n">i</span> <span class="o">|</span> <span class="n">head</span>  <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span>   <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">stdout</span> <span class="p">;</span> <span class="n">done</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="reference">
<h3>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>PEER codes are adapted from <a class="reference external" href="https://github.com/RTIInternational/biocloud_docker_tools/blob/master/peer/v1.3/run_peer.R">here</a></p></li>
<li><p>GTEx recommandation of PEER factors is <a class="reference external" href="https://gtexportal.org/home/documentationPage">here</a></p></li>
<li><p>Examples by PEER is at <a class="reference external" href="https://github.com/PMBio/peer/wiki/Tutorial">github</a></p></li>
</ul>
</div>
<div class="section" id="bicv-factor-analysis-with-apex">
<h3>BiCV factor analysis with APEX<a class="headerlink" href="#bicv-factor-analysis-with-apex" title="Permalink to this headline">¶</a></h3>
<p>Notice that the command options are different from those on the APEX website documentation. The commands on the documentation page does not work (last updated September 2021). The commands below were constructed and tested by our team based on our understanding of the program, without input from APEX authors.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">BiCV_2</span><span class="p">]</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.fake.vcf.gz&#39;</span>
<span class="c1">#task: trunk_workers = 1,trunk_size = job_size , walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="c1">## Add fake header</span>
    <span class="n">cat</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;##fileformat=VCFv4.2</span><span class="se">\n</span><span class="s2">##fileDate=$[time.strftime(&quot;</span><span class="o">%</span><span class="n">Y</span><span class="o">%</span><span class="n">m</span><span class="o">%</span><span class="n">d</span><span class="s2">&quot;,time.localtime())]</span><span class="se">\n</span><span class="s2">##source=FAKE</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">file</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">],</span> <span class="n">append</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    <span class="c1">## Add colnames based on bed</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_input]&quot;</span><span class="p">,</span> <span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">n_max</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;#CHROM&quot;</span><span class="p">,</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span><span class="s2">&quot;ID&quot;</span><span class="p">)</span> 
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">pheno</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">REF</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">ALT</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">QUAL</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span><span class="n">FILTER</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">INFO</span> <span class="o">=</span> <span class="s2">&quot;PR&quot;</span><span class="p">,</span> <span class="n">FORMAT</span> <span class="o">=</span> <span class="s2">&quot;GT&quot;</span><span class="p">),</span><span class="n">pheno</span><span class="p">[,</span><span class="mi">5</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">pheno</span><span class="p">)])</span>
    <span class="n">pheno</span><span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">],</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">col_names</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">bgzip</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">]</span>
    <span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">vcf</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">BiCV_3</span><span class="p">]</span>
<span class="c1"># N factors, if not specify, calculated based on sample size according to GTeX</span>
<span class="kn">parameter:</span> <span class="n">N</span> <span class="o">=</span> <span class="mi">0</span>
<span class="c1"># The number of iteration</span>
<span class="kn">parameter:</span> <span class="n">iteration</span> <span class="o">=</span> <span class="mi">10</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index_col</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">],</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="k">if</span> <span class="n">N</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">150</span><span class="p">:</span>
        <span class="n">N</span> <span class="o">=</span> <span class="mi">15</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">250</span><span class="p">:</span>
        <span class="n">N</span> <span class="o">=</span> <span class="mi">30</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">350</span><span class="p">:</span>
        <span class="n">N</span> <span class="o">=</span> <span class="mi">45</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">N</span> <span class="o">=</span> <span class="mi">60</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">output_from</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.BiCV.gz&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">apex</span> <span class="n">factor</span> \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nn</span><span class="p">]</span> \
        <span class="o">--</span><span class="nb">iter</span> <span class="err">$</span><span class="p">[</span><span class="n">iteration</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">factors</span> <span class="err">$</span><span class="p">[</span><span class="n">N</span><span class="p">]</span> \
        <span class="o">--</span><span class="n">bed</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">vcf</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
        <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">[</span><span class="n">numThreads</span><span class="p">]</span>  <span class="err">$</span><span class="p">[</span> <span class="sa">f</span><span class="s1">&#39;--cov </span><span class="si">{</span><span class="n">covFile</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">covFile</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/data_preprocessing/covariate"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../covariate_preprocessing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Covariate data preprocessing</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="covariate_formatting.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Covariate data formatting</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>