
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>eQTL summary statistics formatting &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="A multivariate EBNM approach for mixture multivariate distribution estimate" href="mixture_prior.html" />
    <link rel="prev" title="Multivariate association with MASH" href="../MASH_pipeline.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/qtl_association_testing.html">
   QTL association analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with individual level data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mashr.html">
     Data overview
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/multivariate/MASH/fastqtl_to_mash.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate/MASH/fastqtl_to_mash.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/multivariate/MASH/fastqtl_to_mash.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input-data">
   Input data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-list-of-summary-statistics">
     A list of summary statistics
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-list-of-gene-names-optional">
     A list of gene names (optional)
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#run-analysis">
   Run analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#note-on-job-option">
     Note on
     <code class="docutils literal notranslate">
      <span class="pre">
       JOB_OPTION
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#more-options">
   More options
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output-data">
   Output data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pipeline-details">
   Pipeline details
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#hdf5-utilities">
     HDF5 utilities
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#convert-to-hdf5-format">
     Convert to HDF5 format
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#per-study-tissue-conversion">
       Per study (tissue) conversion
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#merge-per-study-tissue-hdf5-to-one-hdf5">
       Merge per study (tissue) HDF5 to one HDF5
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#extract-data-for-mash-model">
     Extract data for MASH model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#compute-z-score-and-save-to-rds">
     Compute Z score and save to RDS
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#export-workflow-documentation-to-html-format">
   Export workflow documentation to HTML format
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="eqtl-summary-statistics-formatting">
<h1>eQTL summary statistics formatting<a class="headerlink" href="#eqtl-summary-statistics-formatting" title="Permalink to this headline">¶</a></h1>
<p>This workflow converts <code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> eQTL analysis summary statistics text output to formats more friendly to R analysis. In particular:</p>
<ol class="simple">
<li><p>It converts single study results to HDF5 format grouped by genes.</p></li>
<li><p>It combines multiple studies into one single HDF5 file. In the context of GTEx each study is result from one tissue.</p></li>
<li><p>For MASH analysis in particular, it extracts from the complete data a subset of results to compute data driven MASH prior covariance, and to fit the MASH mixture model.</p></li>
</ol>
<div class="section" id="input-data">
<h2>Input data<a class="headerlink" href="#input-data" title="Permalink to this headline">¶</a></h2>
<div class="section" id="a-list-of-summary-statistics">
<h3>A list of summary statistics<a class="headerlink" href="#a-list-of-summary-statistics" title="Permalink to this headline">¶</a></h3>
<p>Summary statistics from <code class="docutils literal notranslate"><span class="pre">fasteqtl</span></code>-like output are in text format, one row per gene-snp pair. Columns in our example data-set are:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gene_id</span> 
<span class="n">variant_id</span>      
<span class="n">tss_distance</span>    
<span class="n">ma_samples</span>      
<span class="n">ma_count</span>        
<span class="n">maf</span>     
<span class="n">pval_nominal</span>    
<span class="n">slope</span>   
<span class="n">slope_se</span>
</pre></div>
</div>
<p>This is format of GTEx version 8 summary statistics data. Each analysis (“tissue” for GTEx) has a separate text file. Additionally there are support files of gene transcription start site coordinates, and SNP coordinates.</p>
<p>The workflow takes a list of summary statistics file names, eg, <code class="docutils literal notranslate"><span class="pre">data/fastqtl/FastQTLSumStats.list</span></code> (can be configured) that has the contents:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tissue_1</span><span class="o">.</span><span class="n">fastqtl</span><span class="o">.</span><span class="n">gz</span>
<span class="n">Tissue_2</span><span class="o">.</span><span class="n">fastqtl</span><span class="o">.</span><span class="n">gz</span>
<span class="o">...</span>
</pre></div>
</div>
<p>The first two columns of these files have to be <code class="docutils literal notranslate"><span class="pre">gene_id</span></code> and <code class="docutils literal notranslate"><span class="pre">variant_id</span></code> (column name does not matter). For other contents we only need columns <span class="math notranslate nohighlight">\(\hat{\beta}\)</span>, <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> and p-value for the summary statistics. In the <code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> output format above it is columns <code class="docutils literal notranslate"><span class="pre">8,</span> <span class="pre">9,</span> <span class="pre">7</span></code>. If your summary statistics file has a different format you can use <code class="docutils literal notranslate"><span class="pre">--cols</span></code> parameter to pass the proper column numbers. There are two ways to do it:</p>
<ol class="simple">
<li><p>Specify 3 numbers for  <span class="math notranslate nohighlight">\(\hat{\beta}\)</span>, <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> and p-value</p></li>
<li><p>Specify 2 numbers for  <span class="math notranslate nohighlight">\(\hat{\beta}\)</span> and p-value. <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> will then be computed on the fly assuming that the test is two-sided test and that <span class="math notranslate nohighlight">\(\hat{\beta}/\text{SE}(\hat{\beta})\)</span> is standard normal.</p></li>
</ol>
<p>Current default is <code class="docutils literal notranslate"><span class="pre">--cols</span> <span class="pre">8</span> <span class="pre">9</span> <span class="pre">7</span></code>. <strong>Currently this pipeline only supports and requires summary statistics <span class="math notranslate nohighlight">\(\hat{\beta}\)</span>, <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> and p-value, not other quantities (eg t statistic)</strong>.</p>
</div>
<div class="section" id="a-list-of-gene-names-optional">
<h3>A list of gene names (optional)<a class="headerlink" href="#a-list-of-gene-names-optional" title="Permalink to this headline">¶</a></h3>
<p>To speed up merging multiple HDF5 files it helps to provide a list of gene names. Otherwise it takes too much time to figure them out from individual HDF5 files before merger can happen. The gene list has the contents like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ENSG00000186092</span><span class="o">.</span><span class="mi">4</span>
<span class="n">ENSG00000227232</span><span class="o">.</span><span class="mi">5</span>
<span class="n">ENSG00000228463</span><span class="o">.</span><span class="mi">9</span>
<span class="n">ENSG00000241860</span><span class="o">.</span><span class="mi">6</span>
<span class="n">ENSG00000268903</span><span class="o">.</span><span class="mi">1</span>
<span class="n">ENSG00000269981</span><span class="o">.</span><span class="mi">1</span>
<span class="n">ENSG00000279457</span><span class="o">.</span><span class="mi">4</span>
<span class="n">ENSG00000279928</span><span class="o">.</span><span class="mi">2</span>
<span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="run-analysis">
<h2>Run analysis<a class="headerlink" href="#run-analysis" title="Permalink to this headline">¶</a></h2>
<p>Under the same folder as this list file, you keep all these listed data files. Then you run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">JOB_OPTION</span><span class="o">=</span><span class="s2">&quot;-j 8&quot;</span>
<span class="c1">#JOB_OPTION=&quot;-q midway2 -c midway2.yml&quot;</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">workflows</span><span class="o">/</span><span class="n">fastqtl_to_mash</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">convert</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">list</span> <span class="n">data</span><span class="o">/</span><span class="n">fastqtl</span><span class="o">/</span><span class="n">FastQTLSumStats</span><span class="o">.</span><span class="n">list</span> \
    <span class="o">--</span><span class="n">gene</span><span class="o">-</span><span class="nb">list</span> <span class="n">data</span><span class="o">/</span><span class="n">fastqtl</span><span class="o">/</span><span class="n">GTEx_genes</span><span class="o">.</span><span class="n">txt</span> \
    <span class="err">$</span><span class="n">JOB_OPTION</span>
</pre></div>
</div>
<p>to convert to HDF5 only, and</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">workflows</span><span class="o">/</span><span class="n">fastqtl_to_mash</span><span class="o">.</span><span class="n">ipynb</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">list</span> <span class="n">data</span><span class="o">/</span><span class="n">fastqtl</span><span class="o">/</span><span class="n">FastQTLSumStats</span><span class="o">.</span><span class="n">list</span> \
    <span class="o">--</span><span class="n">gene</span><span class="o">-</span><span class="nb">list</span> <span class="n">data</span><span class="o">/</span><span class="n">fastqtl</span><span class="o">/</span><span class="n">GTEx_genes</span><span class="o">.</span><span class="n">txt</span> \
    <span class="err">$</span><span class="n">JOB_OPTION</span>
</pre></div>
</div>
<p>to convert to HDF5 AND extract MASH input.</p>
<div class="section" id="note-on-job-option">
<h3>Note on <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code><a class="headerlink" href="#note-on-job-option" title="Permalink to this headline">¶</a></h3>
<p>With bash variable <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code> set to <code class="docutils literal notranslate"><span class="pre">-q</span> <span class="pre">midway2</span> <span class="pre">-c</span> <span class="pre">midway2.yml</span></code>, each workflow is configured to run on UChicago RCC’s midway2 cluster, via <code class="docutils literal notranslate"><span class="pre">task</span></code> option.
When executed as is, it takes 33hrs to convert the summary statistics for GTEx data (Release V8).
You might need to configure <code class="docutils literal notranslate"><span class="pre">task</span></code> differently for your computing environment if you are familiar with <code class="docutils literal notranslate"><span class="pre">SoS</span></code>.
Otherwise, you can use the <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code> <code class="docutils literal notranslate"><span class="pre">-j</span> <span class="pre">8</span></code> to use 8 CPU threads.</p>
</div>
</div>
<div class="section" id="more-options">
<h2>More options<a class="headerlink" href="#more-options" title="Permalink to this headline">¶</a></h2>
<div class="cell tag_report_cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">fastqtl_to_mash</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run fastqtl_to_mash.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  convert
  default
  document_it

Global Workflow Options:
  --cwd fastqtl_to_mash_output (as path)
                        Work directory / output directory
  --data-list data/eQTL_summary_files.txt (as path)
                        A text file containing data-set names
  --gene-list . (as path)
                        Optionally, a list of gene names.
  --msg &#39;eQTL mapping summary statistics&#39;
                        Meta-info tag for HDF5 database
  --maxsize 1000 (as int)
                        maximum number of groups per HDF5 file
  --cols 8 9 7 (as list)
                        1-based indexing of 3 numbers indicating columns for
                        betahat, se(betahat) and p-value, or 2 numbesr for
                        betahat and p-value
  --keep-ensg-version 0 (as int)
  --common-suffix &#39;&#39;
                        for GTEx data, parameter: common_suffix =
                        &quot;.allpairs.txt&quot;

Sections
  convert_0, default_0: Generate utility functions
  convert_1, default_1: Convert summary stats gzip format to HDF5
  convert_2, default_2: Merge single study data to multivariate data
  default_3:            Extract data to fit MASH model
    Workflow Options:
      --random-per-gene 9 (as int)
                        Number of random SNPs to draw per gene for fitting MASH
                        mixture Use -1 to export all
      --best-per-gene 1 (as int)
                        Use 1 to extract strongest gene-snp pair Use 0 to avoid
                        extracting the info
  default_4:            Subset and split data, generate Z-score and save to RDS
    Workflow Options:
      --random-snp-size 0.5 (as float)
                        Size of mash random SNPs. This specifies the size of
                        `random` SNP set and the rest of random SNPs go into
                        `random_test` SNP set. if input is float it means half
                        random SNPs the input can also be a integer to use
                        specified number.
      --effects-list NULL (as path)
                        A list of effect names (SNP names) to include in mash
                        analysis.
      --conditions-list NULL (as path)
                        A list of condition names (tissue names) to include in
                        mash analysis
  document_it:          Export workflow to HTML document
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="output-data">
<h2>Output data<a class="headerlink" href="#output-data" title="Permalink to this headline">¶</a></h2>
<p>If you run the entire workflow, you should find under <code class="docutils literal notranslate"><span class="pre">./fastqtl_to_mash_output</span></code> (can be configured):</p>
<ul class="simple">
<li><p>Study (tissue) specific HDF5 files of summary statistics</p></li>
<li><p>Merged HDF5 from multiple studies</p></li>
<li><p>“*.portable.h5” data extracted for MASH computations</p></li>
<li><p>“*.mash.rds” data in RDS format splitted to <code class="docutils literal notranslate"><span class="pre">strong</span></code>/<code class="docutils literal notranslate"><span class="pre">random</span></code> and optionally <code class="docutils literal notranslate"><span class="pre">random_test</span></code>, with Z-scores computed from p-values.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Work directory / output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./fastqtl_to_mash_output&#39;</span><span class="p">)</span>
<span class="c1"># A text file containing data-set names</span>
<span class="kn">parameter:</span> <span class="n">data_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;data/eQTL_summary_files.txt&quot;</span><span class="p">)</span>
<span class="c1"># Optionally, a list of gene names.</span>
<span class="kn">parameter:</span> <span class="n">gene_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Meta-info tag for HDF5 database</span>
<span class="kn">parameter:</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;eQTL mapping summary statistics&quot;</span>
<span class="c1"># maximum number of groups per HDF5 file</span>
<span class="kn">parameter:</span> <span class="n">maxsize</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="c1"># 1-based indexing of 3 numbers indicating columns for betahat, se(betahat) and p-value, </span>
<span class="c1"># or 2 numbesr for betahat and p-value</span>
<span class="kn">parameter:</span> <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">keep_ensg_version</span> <span class="o">=</span> <span class="mi">0</span>
<span class="c1"># for GTEx data, parameter: common_suffix = &quot;.allpairs.txt&quot;</span>
<span class="kn">parameter:</span> <span class="n">common_suffix</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
    <span class="n">cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="pipeline-details">
<h2>Pipeline details<a class="headerlink" href="#pipeline-details" title="Permalink to this headline">¶</a></h2>
<div class="section" id="hdf5-utilities">
<h3>HDF5 utilities<a class="headerlink" href="#hdf5-utilities" title="Permalink to this headline">¶</a></h3>
<p>The HDF5 Python interface <code class="docutils literal notranslate"><span class="pre">pytables</span></code> is used to program the data format conversion workhorse. You can use command, eg, <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">tables</span></code> to install <code class="docutils literal notranslate"><span class="pre">pytables</span></code> package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate utility functions</span>
<span class="p">[</span><span class="n">utils</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="s1">&#39;.sos/utils.py&#39;</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">Py_Module</span><span class="p">(</span><span class="s1">&#39;tables&#39;</span><span class="p">),</span> <span class="n">Py_Module</span><span class="p">(</span><span class="s1">&#39;pandas&gt;0.23.0&#39;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="s1">&#39;.sos/utils.py&#39;</span>
<span class="kn">report:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">output</span> <span class="o">=</span> <span class="n">_output</span>
    <span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">re</span><span class="o">,</span> <span class="nn">copy</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span><span class="o">,</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span><span class="o">,</span> <span class="nn">tables</span> <span class="k">as</span> <span class="nn">tb</span>
    <span class="n">tb</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">MAX_GROUP_WIDTH</span> <span class="o">=</span> <span class="mi">51200</span>
    <span class="c1"># tb.parameters.NODE_CACHE_SLOTS = -51200</span>
    <span class="c1"># tb.parameters.METADATA_CACHE_SIZE = 1048576 * 100000</span>
    <span class="c1"># tb.parameters.CHUNK_CACHE_SIZE = 2097152 * 100000</span>
    <span class="c1"># tb.parameters.CHUNK_CACHE_NELMTS = 521</span>
    <span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="k">as</span> <span class="nn">st</span>

    <span class="k">class</span> <span class="nc">Environment</span><span class="p">:</span>
        <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">float</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">float32</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_tag</span> <span class="o">=</span> <span class="s1">&#39;_duplicated_&#39;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">common_suffix</span> <span class="o">=</span> <span class="s1">&#39;$</span><span class="si">{common_suffix}</span><span class="s1">.h5&#39;</span>

    <span class="kp">env</span> <span class="o">=</span> <span class="n">Environment</span><span class="p">()</span>

    <span class="k">class</span> <span class="nc">TBData</span><span class="p">(</span><span class="nb">dict</span><span class="p">):</span>
        <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">root</span> <span class="o">=</span> <span class="s1">&#39;/&#39;</span><span class="p">,</span> <span class="n">complib</span> <span class="o">=</span> <span class="s1">&#39;bzip2&#39;</span><span class="p">):</span>
            <span class="sd">&#39;&#39;&#39;bzip2 may not be compatible with other hdf5 applications; but zlib is fine&#39;&#39;&#39;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__root</span> <span class="o">=</span> <span class="n">root</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__group</span> <span class="o">=</span> <span class="kp">name</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__msg</span> <span class="o">=</span> <span class="n">msg</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">dict</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
                <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">str</span><span class="p">:</span>
                    <span class="c1"># is file name</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">__load</span><span class="p">(</span><span class="n">tb</span><span class="o">.</span><span class="n">open_file</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># is file stream</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">__load</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">tb</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">NoSuchNodeError</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Cannot find dataset </span><span class="si">{}</span><span class="s1">!&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="kp">name</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tb_filters</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">Filters</span><span class="p">(</span><span class="n">complevel</span> <span class="o">=</span> <span class="mi">9</span><span class="p">,</span> <span class="n">complib</span><span class="o">=</span><span class="n">complib</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">sink</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
            <span class="k">with</span> <span class="n">tb</span><span class="o">.</span><span class="n">open_file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">__root</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">__root</span><span class="p">)</span>
                    <span class="k">except</span><span class="p">:</span>
                        <span class="k">pass</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="c1"># there is existing data -- have to merge with current data</span>
                    <span class="c1"># have to do this because the input file lines are not grouped by gene names!!</span>
                    <span class="c1"># use try ... except to hopefully faster than if ... else</span>
                    <span class="c1"># e.g., if not f.__contains__(&#39;/{}&#39;.format(self.__group)) ... else ...</span>
                    <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">list_nodes</span><span class="p">(</span><span class="s1">&#39;/</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__root</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">)):</span>
                        <span class="k">if</span> <span class="n">element</span><span class="o">.</span><span class="kp">name</span> <span class="o">!=</span> <span class="s1">&#39;colnames&#39;</span><span class="p">:</span>
                            <span class="bp">self</span><span class="p">[</span><span class="n">element</span><span class="o">.</span><span class="kp">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">element</span><span class="p">[:],</span> <span class="bp">self</span><span class="p">[</span><span class="n">element</span><span class="o">.</span><span class="kp">name</span><span class="p">]))</span>
                <span class="k">except</span> <span class="n">tb</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">NoSuchNodeError</span><span class="p">:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">__root</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">,</span>
                                   <span class="bp">self</span><span class="o">.</span><span class="n">__msg</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">__msg</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">__store_array</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
                <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

        <span class="k">def</span> <span class="nf">dump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table</span><span class="p">,</span> <span class="n">output</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">output</span><span class="p">:</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="bp">self</span><span class="p">[</span><span class="s1">&#39;colnames&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="p">:</span> <span class="bp">self</span><span class="p">[</span><span class="n">table</span><span class="p">][:,</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="s1">&#39;colnames&#39;</span><span class="p">]))},</span> <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;rownames&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span> <span class="n">na_rep</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span><span class="p">)</span>
                <span class="k">return</span> <span class="kc">None</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="bp">self</span><span class="p">[</span><span class="s1">&#39;colnames&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="p">:</span> <span class="bp">self</span><span class="p">[</span><span class="n">table</span><span class="p">][:,</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="s1">&#39;colnames&#39;</span><span class="p">]))},</span> <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;rownames&#39;</span><span class="p">])</span>

        <span class="k">def</span> <span class="nf">__load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fstream</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">fstream</span><span class="o">.</span><span class="n">list_nodes</span><span class="p">(</span><span class="s1">&#39;/</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__root</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">)):</span>
                    <span class="bp">self</span><span class="p">[</span><span class="n">element</span><span class="o">.</span><span class="kp">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">element</span><span class="p">[:]</span>
                <span class="n">fstream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">fstream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                <span class="k">raise</span>

        <span class="k">def</span> <span class="nf">__roll_back</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group</span><span class="p">,</span> <span class="kp">name</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">n</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="kp">name</span><span class="p">)</span>
                <span class="n">n</span><span class="o">.</span><span class="n">_f_remove</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                <span class="k">pass</span>

        <span class="k">def</span> <span class="nf">__store_array</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="n">fstream</span><span class="p">):</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">__root</span><span class="p">:</span>
                <span class="n">element</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">fstream</span><span class="o">.</span><span class="n">root</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">__root</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">element</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">fstream</span><span class="o">.</span><span class="n">root</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">)</span>
            <span class="n">arr</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="kp">name</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">arr</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">list</span><span class="p">:</span>
                <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">arr</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__roll_back</span><span class="p">(</span><span class="n">element</span><span class="p">,</span> <span class="kp">name</span><span class="p">)</span>
            <span class="c1">#</span>
            <span class="k">if</span> <span class="n">arr</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,):</span>
                <span class="n">ds</span> <span class="o">=</span> <span class="n">fstream</span><span class="o">.</span><span class="n">create_carray</span><span class="p">(</span><span class="n">element</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="n">tb</span><span class="o">.</span><span class="n">Atom</span><span class="o">.</span><span class="n">from_dtype</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">dtype</span><span class="p">),</span> <span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span>
                                           <span class="n">filters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tb_filters</span><span class="p">)</span>
                <span class="n">ds</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">arr</span>

    <span class="k">def</span> <span class="nf">get_tb_grps</span><span class="p">(</span><span class="n">filenames</span><span class="p">,</span> <span class="n">group_name</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filenames</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">filenames</span><span class="p">]</span>
        <span class="n">names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">tb</span><span class="o">.</span><span class="n">open_file</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">names</span><span class="o">.</span><span class="n">update</span><span class="p">([</span><span class="n">node</span><span class="o">.</span><span class="n">_v_name</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">root</span> <span class="k">if</span> <span class="n">group_name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">root</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">group_name</span><span class="p">)))])</span>
        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">names</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_se</span><span class="p">(</span><span class="n">bhat</span><span class="p">,</span> <span class="n">pval</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">bhat</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">z</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="n">pval</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">z</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">pval</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">z</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">se</span> <span class="o">=</span> <span class="n">bhat</span><span class="o">/</span><span class="n">z</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">se</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="k">return</span> <span class="n">se</span>

    <span class="k">class</span> <span class="nc">SSData</span><span class="p">:</span>
        <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;buffer&#39;</span><span class="p">:{</span><span class="s1">&#39;data&#39;</span><span class="p">:[],</span> <span class="s1">&#39;rownames&#39;</span><span class="p">:[]},</span> <span class="s1">&#39;output&#39;</span><span class="p">:{}}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">header</span> <span class="o">=</span> <span class="n">header</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">previous_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_name</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">min_line_len</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]},</span> <span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">]},</span> <span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">]}])</span>

        <span class="k">def</span> <span class="nf">parse</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line</span><span class="p">,</span> <span class="n">ensg_version</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">include</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()):</span>
            <span class="c1"># input line is snp, gene, beta, t, pval and optionally se(bhat)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">__reset</span><span class="p">()</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">current_name</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">return</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">):</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">header</span><span class="p">:</span>
                <span class="c1"># header line has to be accounted for</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">header</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">return</span> <span class="mi">0</span>
            <span class="c1">#</span>
            <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">include</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">include</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>
            <span class="c1"># the line is not long enough ... there is format issues</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_line_len</span><span class="p">:</span>
                <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>
            <span class="c1">#</span>
            <span class="k">if</span> <span class="n">ensg_version</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">previous_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">previous_name</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">current_name</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_name</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">previous_name</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">__reset</span><span class="p">()</span>
            <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">]}</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># the two numbers are bhat and p-value, we need to compute se(bhat)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">line</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">}],</span> <span class="nb">str</span><span class="p">(</span><span class="n">get_se</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">}]),</span> <span class="nb">float</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">}]))),</span> <span class="n">line</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">}]])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;data&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">line</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">}],</span> <span class="n">line</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">}],</span> <span class="n">line</span><span class="p">[</span><span class="err">$</span><span class="p">{</span><span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">}]])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__format_variant_id</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
            <span class="k">return</span> <span class="mi">0</span>
        
        <span class="nd">@staticmethod</span>
        <span class="k">def</span> <span class="nf">__format_variant_id</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">value</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
                <span class="c1"># keep it chr, pos, ref, alt</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">value</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span>
            <span class="k">return</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">__reset</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;data&#39;</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="kp">env</span><span class="o">.</span><span class="n">float</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">][</span><span class="s1">&#39;colnames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;beta&#39;</span><span class="p">,</span><span class="s1">&#39;se&#39;</span><span class="p">,</span><span class="s1">&#39;pval&#39;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;output&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;buffer&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;data&#39;</span><span class="p">:[],</span> <span class="s1">&#39;rownames&#39;</span><span class="p">:[]}</span>

        <span class="k">def</span> <span class="nf">dump</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;output&#39;</span><span class="p">]</span>

    <span class="k">class</span> <span class="nc">DataMerger</span><span class="p">(</span><span class="n">TBData</span><span class="p">):</span>
        <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">files</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
            <span class="n">TBData</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="p">{},</span> <span class="kp">name</span><span class="p">,</span> <span class="n">msg</span><span class="p">,</span> <span class="n">complib</span> <span class="o">=</span> <span class="s2">&quot;zlib&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__group</span> <span class="o">=</span> <span class="kp">name</span>

        <span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">one_snp</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">failure_ct</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># Collect data</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">:</span>
                <span class="n">tissue</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">$&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="kp">env</span><span class="o">.</span><span class="n">common_suffix</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">item</span><span class="p">))</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">]</span> <span class="o">=</span> <span class="n">TBData</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">one_snp</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="n">one_snp</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;data&#39;</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">]]),</span> <span class="s1">&#39;rownames&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
                    <span class="n">failure_ct</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1"># Fix row name</span>
                <span class="c1"># Because in GTEx data file there are duplicated gene-snp pairs having different sumstats!!</span>
                <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">__dedup</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">],</span> <span class="n">item</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">failure_ct</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">):</span>
                <span class="k">return</span> <span class="mi">1</span>
            <span class="c1"># Merge data</span>
            <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">([</span><span class="s1">&#39;beta&#39;</span><span class="p">,</span><span class="s1">&#39;se&#39;</span><span class="p">,</span><span class="s1">&#39;pval&#39;</span><span class="p">]):</span>
                <span class="bp">self</span><span class="p">[</span><span class="n">item</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                    <span class="p">{</span><span class="n">tissue</span> <span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;data&#39;</span><span class="p">][:,</span><span class="n">idx</span><span class="p">]},</span>
                    <span class="n">index</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="p">[</span><span class="n">one_snp</span><span class="p">]</span>
                    <span class="p">)</span> <span class="k">for</span> <span class="n">tissue</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">())],</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
                <span class="k">if</span> <span class="s1">&#39;rownames&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
                    <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">str</span><span class="p">)</span>
                <span class="k">if</span> <span class="s1">&#39;colnames&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
                    <span class="bp">self</span><span class="p">[</span><span class="s1">&#39;colnames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">dtype</span> <span class="o">=</span> <span class="nb">str</span><span class="p">)</span>
                <span class="bp">self</span><span class="p">[</span><span class="n">item</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="kp">env</span><span class="o">.</span><span class="n">float</span><span class="p">)</span>
            <span class="c1"># np.savetxt(sys.stdout, self[&#39;pval&#39;], fmt=&#39;%10.5f&#39;)</span>
            <span class="c1"># print(self[&#39;rownames&#39;])</span>
            <span class="c1"># print(self[&#39;colnames&#39;])</span>
            <span class="k">return</span> <span class="mi">0</span>

        <span class="k">def</span> <span class="nf">__dedup</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
            <span class="n">seen</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">dups</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">def</span> <span class="nf">__is_seen</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">seen</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seen</span><span class="p">:</span>
                    <span class="n">seen</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="k">return</span> <span class="mi">0</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">seen</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">dups</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
                    <span class="k">return</span> <span class="mi">1</span>
            <span class="c1"># Tag them</span>
            <span class="n">obs</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">__is_seen</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">seen</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;</span><span class="si">%s%s%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="kp">env</span><span class="o">.</span><span class="n">duplicate_tag</span><span class="p">,</span> <span class="n">seen</span><span class="p">[</span><span class="n">x</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">]</span>
            <span class="c1"># Log them</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dups</span><span class="p">):</span>
                <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span> <span class="o">+</span> <span class="s1">&#39;.error&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">dups</span><span class="p">:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">:</span><span class="si">{}</span><span class="s1"> appeared </span><span class="si">{}</span><span class="s1"> times in </span><span class="si">{}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span>\
                                <span class="nb">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__group</span><span class="p">,</span> <span class="n">item</span><span class="p">,</span> <span class="n">seen</span><span class="p">[</span><span class="n">item</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">filename</span><span class="p">))</span>
            <span class="k">return</span> <span class="n">obs</span>

    <span class="k">def</span> <span class="nf">get_gs_pairs</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="n">num</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;equal_space&#39;</span><span class="p">):</span>
        <span class="sd">&#39;&#39;&#39;choose gene-snp pairs from data, controlled by num = (a, b)</span>
<span class="sd">        for the best gene-snp pair (a = 0 or 1), and b other random </span>
<span class="sd">        gene-snp pairs&#39;&#39;&#39;</span>

        <span class="k">def</span> <span class="nf">random_sample</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">k</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">k</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">k</span><span class="p">))))</span>

        <span class="k">def</span> <span class="nf">equal_sample</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">k</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">k</span> <span class="ow">or</span> <span class="n">k</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="n">f</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">m</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">*</span><span class="n">n</span><span class="o">//</span><span class="n">m</span> <span class="o">+</span> <span class="n">n</span><span class="o">//</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">)]</span>
            <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))]))</span>

        <span class="n">output</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;colnames&#39;</span> <span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;colnames&#39;</span><span class="p">]}</span>
        <span class="n">lp</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="s1">&#39;pval&#39;</span><span class="p">)</span>
        <span class="n">shat</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="s1">&#39;se&#39;</span><span class="p">)</span>
        <span class="c1"># pval cannot be nan or zero (in fastqtl output, Shat nan has pval zero)</span>
        <span class="n">lp</span> <span class="o">=</span> <span class="n">lp</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">lp</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">shat</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>
        <span class="c1">#</span>
        <span class="k">if</span> <span class="n">lp</span><span class="o">.</span><span class="n">empty</span> <span class="ow">and</span> <span class="n">num</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="c1"># Find strong SNP-gene pair</span>
        <span class="n">lp</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">lp</span><span class="p">)</span>
        <span class="n">rowidx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">lp</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">idxmax</span><span class="p">())[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">num</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">output</span><span class="p">[</span><span class="s1">&#39;strong_rownames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="kp">name</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;rownames&#39;</span><span class="p">][</span><span class="n">rowidx</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">())]</span>
            <span class="n">output</span><span class="p">[</span><span class="s1">&#39;strong&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;beta&#39;</span><span class="p">,</span> <span class="s1">&#39;pval&#39;</span><span class="p">,</span> <span class="s1">&#39;se&#39;</span><span class="p">]:</span>
                <span class="n">output</span><span class="p">[</span><span class="s1">&#39;strong&#39;</span><span class="p">][</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">rowidx</span><span class="p">,</span> <span class="p">:]</span>
        <span class="k">if</span> <span class="n">num</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">all_randomidxes</span> <span class="o">=</span> <span class="p">[</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;rownames&#39;</span><span class="p">])</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lp</span><span class="o">.</span><span class="n">index</span> <span class="ow">and</span> <span class="n">y</span> <span class="o">!=</span> <span class="n">rowidx</span><span class="p">]</span>
            <span class="n">sample_randomidxes</span> <span class="o">=</span> <span class="n">random_sample</span><span class="p">(</span><span class="n">all_randomidxes</span><span class="p">,</span> <span class="n">num</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;random&#39;</span> <span class="k">else</span> <span class="n">equal_sample</span><span class="p">(</span><span class="n">all_randomidxes</span><span class="p">,</span> <span class="n">num</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">output</span><span class="p">[</span><span class="s1">&#39;random_rownames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="kp">name</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;rownames&#39;</span><span class="p">][</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sample_randomidxes</span><span class="p">]</span>
            <span class="n">output</span><span class="p">[</span><span class="s1">&#39;random&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;beta&#39;</span><span class="p">,</span> <span class="s1">&#39;pval&#39;</span><span class="p">,</span> <span class="s1">&#39;se&#39;</span><span class="p">]:</span>
                <span class="n">output</span><span class="p">[</span><span class="s1">&#39;random&#39;</span><span class="p">][</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">sample_randomidxes</span><span class="p">,</span> <span class="p">:]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="s1">&#39;strong&#39;</span> <span class="ow">in</span> <span class="n">output</span> <span class="ow">and</span> <span class="ow">not</span> <span class="s1">&#39;random&#39;</span> <span class="ow">in</span> <span class="n">output</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">output</span>

    <span class="k">def</span> <span class="nf">merge_tmp_h5</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
        <span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>
        <span class="n">tmpfiles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="n">output</span> <span class="o">+</span> <span class="s2">&quot;_*.tmp&quot;</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">output</span><span class="o">+</span><span class="s1">&#39;.h5&#39;</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">output</span><span class="o">+</span><span class="s1">&#39;.h5&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">tmpfiles</span><span class="p">):</span>
            <span class="k">for</span> <span class="kp">name</span> <span class="ow">in</span> <span class="n">get_tb_grps</span><span class="p">(</span><span class="n">item</span><span class="p">):</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="s1">&#39;h5copy -i </span><span class="si">{0}</span><span class="s1"> -o </span><span class="si">{2}</span><span class="s1"> -s &quot;/</span><span class="si">{1}</span><span class="s1">&quot; -d &quot;/</span><span class="si">{1}</span><span class="s1">&quot;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="n">output</span><span class="o">+</span><span class="s1">&#39;.h5&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
                <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="convert-to-hdf5-format">
<h3>Convert to HDF5 format<a class="headerlink" href="#convert-to-hdf5-format" title="Permalink to this headline">¶</a></h3>
<div class="section" id="per-study-tissue-conversion">
<h4>Per study (tissue) conversion<a class="headerlink" href="#per-study-tissue-conversion" title="Permalink to this headline">¶</a></h4>
<p>For per study conversion I use <code class="docutils literal notranslate"><span class="pre">bzip2</span></code> compression method and <code class="docutils literal notranslate"><span class="pre">float32</span></code> to achieve higher compression rate. This workflow step will generate one HDF5 file per summary statistics file. <code class="docutils literal notranslate"><span class="pre">h5copy</span></code> program can be installed via <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">conda-forge</span> <span class="pre">hdf5</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Convert summary stats gzip format to HDF5</span>
<span class="p">[</span><span class="n">convert_1</span><span class="p">,</span> <span class="n">default_1</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="s1">&#39;.sos/utils.py&#39;</span><span class="p">,</span> <span class="n">executable</span><span class="p">(</span><span class="s2">&quot;h5copy&quot;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">data_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">(),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;Need data list file!&#39;</span><span class="p">)</span>
<span class="n">data_files</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_output</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;awk &#39;</span><span class="se">{{</span><span class="s2">print $1</span><span class="se">}}</span><span class="s2">&#39; </span><span class="si">{</span><span class="n">data_list</span><span class="si">:</span><span class="s2">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">))</span>
<span class="n">fail_if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;Need input data files!&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">input:</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">data_list</span><span class="si">:</span><span class="s1">d</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">data_list</span><span class="si">:</span><span class="s1">d</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_files</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.h5&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;36h&#39;</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;4G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="n">_</span><span class="o">*.</span><span class="n">tmp</span>
    
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="nb">input</span> <span class="o">=</span> <span class="n">_depends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="kn">import</span> <span class="nn">warnings</span>
    <span class="kn">import</span> <span class="nn">gzip</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">gene_list</span><span class="p">:</span><span class="n">r</span><span class="p">})</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span> <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">gene_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">()}</span> <span class="k">else</span> <span class="p">[])</span>
    <span class="n">ssp</span> <span class="o">=</span> <span class="n">SSData</span><span class="p">(</span><span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">group_counts</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
            <span class="n">quit</span> <span class="o">=</span> <span class="n">ssp</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">keep_ensg_version</span><span class="p">},</span> <span class="n">include</span> <span class="o">=</span> <span class="n">gene_names</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">quit</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">ssp</span><span class="o">.</span><span class="n">current_name</span> <span class="o">!=</span> <span class="n">ssp</span><span class="o">.</span><span class="n">previous_name</span><span class="p">:</span>
                <span class="n">group_counts</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
                    <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">FlavorWarning</span><span class="p">)</span>
                    <span class="n">data</span> <span class="o">=</span> <span class="n">TBData</span><span class="p">(</span><span class="n">ssp</span><span class="o">.</span><span class="n">dump</span><span class="p">(),</span> <span class="n">ssp</span><span class="o">.</span><span class="n">previous_name</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{msg}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">data</span><span class="o">.</span><span class="n">sink</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:n}</span><span class="s2">_</span><span class="si">%i</span><span class="s2">.tmp&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">group_counts</span> <span class="o">/</span> <span class="err">$</span><span class="p">{</span><span class="n">maxsize</span><span class="p">}))</span> 
                                <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">maxsize</span><span class="p">}</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
                <span class="n">ssp</span><span class="o">.</span><span class="n">previous_name</span> <span class="o">=</span> <span class="n">ssp</span><span class="o">.</span><span class="n">current_name</span>
            <span class="k">if</span> <span class="n">quit</span><span class="p">:</span>
                <span class="k">break</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">maxsize</span><span class="p">}</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">merge_tmp_h5</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">})</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="n">_</span><span class="o">*.</span><span class="n">tmp</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="merge-per-study-tissue-hdf5-to-one-hdf5">
<h4>Merge per study (tissue) HDF5 to one HDF5<a class="headerlink" href="#merge-per-study-tissue-hdf5-to-one-hdf5" title="Permalink to this headline">¶</a></h4>
<p>This step creates tables for each summary statistic per gene from multiple studies. Rows are effects (SNPs), columns are conditions (tissue names). This time I use <code class="docutils literal notranslate"><span class="pre">zlib</span></code> compression for better compatibility with other HDF5 routines (eg <code class="docutils literal notranslate"><span class="pre">rhdf5</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge single study data to multivariate data</span>
<span class="p">[</span><span class="n">convert_2</span><span class="p">,</span> <span class="n">default_2</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="s1">&#39;.sos/utils.py&#39;</span><span class="p">,</span> <span class="n">executable</span><span class="p">(</span><span class="s2">&quot;h5copy&quot;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">data_list</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.h5&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;36h&#39;</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;4G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="n">_</span><span class="o">*.</span><span class="n">tmp</span>

<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="nb">input</span> <span class="o">=</span> <span class="n">_depends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="kn">import</span> <span class="nn">warnings</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">gene_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">()}:</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">gene_list</span><span class="p">:</span><span class="n">r</span><span class="p">})</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="n">get_tb_grps</span><span class="p">([</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}])</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">keep_ensg_version</span><span class="p">}</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">gene_names</span><span class="p">]</span>
    <span class="n">failure_ct</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_names</span><span class="p">):</span>
        <span class="n">ssm</span> <span class="o">=</span> <span class="n">DataMerger</span><span class="p">([</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}],</span> <span class="n">item</span><span class="p">,</span> <span class="s2">&quot;$</span><span class="si">{msg}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">FlavorWarning</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ssm</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">ssm</span><span class="o">.</span><span class="n">sink</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:n}</span><span class="s2">_</span><span class="si">%i</span><span class="s2">.tmp&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">((</span><span class="n">idx</span> <span class="o">+</span> <span class="mf">1.0</span><span class="p">)</span> <span class="o">/</span> <span class="err">$</span><span class="p">{</span><span class="n">maxsize</span><span class="p">}))</span> 
                        <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">maxsize</span><span class="p">}</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">failure_ct</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:n}</span><span class="s2">.log&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> out of </span><span class="si">%s</span><span class="s2"> groups merged!</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_names</span><span class="p">)</span> <span class="o">-</span> <span class="n">failure_ct</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_names</span><span class="p">)))</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">maxsize</span><span class="p">}</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">merge_tmp_h5</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nr</span><span class="p">})</span>
            
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="n">_</span><span class="o">*.</span><span class="n">tmp</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="extract-data-for-mash-model">
<h3>Extract data for MASH model<a class="headerlink" href="#extract-data-for-mash-model" title="Permalink to this headline">¶</a></h3>
<p>We need to extract the “top” signals based on single tissue analysis to compute MASH priors, as well as some random SNP sets to fit MASH mixture model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract data to fit MASH model</span>
<span class="p">[</span><span class="n">default_3</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="s1">&#39;.sos/utils.py&#39;</span>
<span class="c1"># Number of random SNPs to draw per gene for fitting MASH mixture</span>
<span class="c1"># Use -1 to export all</span>
<span class="kn">parameter:</span> <span class="n">random_per_gene</span> <span class="o">=</span> <span class="mi">9</span>
<span class="c1"># Use 1 to extract strongest gene-snp pair</span>
<span class="c1"># Use 0 to avoid extracting the info</span>
<span class="kn">parameter:</span> <span class="n">best_per_gene</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.portable.h5&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;36h&#39;</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;4G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>

<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="nb">input</span> <span class="o">=</span> <span class="n">_depends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="kn">import</span> <span class="nn">warnings</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">gene_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">()}:</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">gene_list</span><span class="p">:</span><span class="n">r</span><span class="p">})</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="n">get_tb_grps</span><span class="p">([</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">,}])</span>
    <span class="k">if</span> <span class="err">$</span><span class="p">{</span><span class="n">keep_ensg_version</span><span class="p">}</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">gene_names</span><span class="p">]</span>
    <span class="n">output</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="n">output</span><span class="p">[</span><span class="s1">&#39;random&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;colnames&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;rownames&#39;</span><span class="p">:</span> <span class="p">[],</span> <span class="s1">&#39;beta&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;se&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;pval&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
    <span class="n">output</span><span class="p">[</span><span class="s1">&#39;strong&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;colnames&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;rownames&#39;</span><span class="p">:</span> <span class="p">[],</span> <span class="s1">&#39;beta&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;se&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;pval&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
    <span class="n">failure_ct</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="kp">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_names</span><span class="p">):</span>
        <span class="c1"># extract the best gene-snp pair or some random gene-snp pairs</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">TBData</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="kp">name</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">get_gs_pairs</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="kp">name</span><span class="p">,</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">best_per_gene</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">random_per_gene</span><span class="p">}))</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="n">res</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1">#</span>
        <span class="k">if</span> <span class="n">res</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">failure_ct</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">continue</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">output</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">for</span> <span class="n">kk</span> <span class="ow">in</span> <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">kk</span> <span class="o">==</span> <span class="s1">&#39;rownames&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_rownames&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">)]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;rownames&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_rownames&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">)])</span>
                <span class="k">elif</span> <span class="n">kk</span> <span class="o">==</span> <span class="s1">&#39;colnames&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;colnames&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;colnames&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s1">&#39;colnames&#39;</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">kk</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">kk</span><span class="p">],</span> <span class="n">res</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">kk</span><span class="p">]))</span> <span class="k">if</span> <span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">kk</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">res</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">kk</span><span class="p">]</span>
    <span class="c1">#</span>
    <span class="k">if</span> <span class="n">failure_ct</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_names</span><span class="p">):</span>
        <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">FlavorWarning</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">output</span><span class="p">:</span>
                <span class="n">TBData</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="n">k</span><span class="p">]),</span> <span class="n">k</span><span class="p">,</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2"> gene-snp pair&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="s2">&quot;$</span><span class="si">{msg}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">k</span><span class="p">),</span> <span class="n">complib</span> <span class="o">=</span> <span class="s1">&#39;zlib&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">sink</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output:n}</span><span class="s2">.log&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> out of </span><span class="si">%s</span><span class="s2"> groups extracted!</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_names</span><span class="p">)</span> <span class="o">-</span> <span class="n">failure_ct</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_names</span><span class="p">)))</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="compute-z-score-and-save-to-rds">
<h3>Compute Z score and save to RDS<a class="headerlink" href="#compute-z-score-and-save-to-rds" title="Permalink to this headline">¶</a></h3>
<p>Here I also implemented two command options to subset data:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--effects-list</span></code>: a list of effect names (row names) to include from output MASH data-set. This option was used in MASH paper to remove SNPs in LD with each other.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--conditions-list</span></code>: a list of conditions (column names) to include in output MASH data-set. Conditions not in this list are excluded from output. This option was used in MASH paper to focus analysis on brain / no-brain tissues.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Subset and split data, generate Z-score and save to RDS</span>
<span class="p">[</span><span class="n">default_4</span><span class="p">]</span>
<span class="c1"># Size of mash random SNPs. </span>
<span class="c1"># This specifies the size of `random` SNP set </span>
<span class="c1"># and the rest of random SNPs go into `random_test` SNP set.</span>
<span class="c1"># if input is float it means half random SNPs</span>
<span class="c1"># the input can also be a integer to use specified number.</span>
<span class="kn">parameter:</span> <span class="n">random_snp_size</span> <span class="o">=</span> <span class="mf">0.5</span>
<span class="c1"># A list of effect names (SNP names) to include in mash analysis.</span>
<span class="kn">parameter:</span> <span class="n">effects_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;NULL&#39;</span><span class="p">)</span>
<span class="c1"># A list of condition names (tissue names) to include in mash analysis</span>
<span class="kn">parameter:</span> <span class="n">conditions_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;NULL&#39;</span><span class="p">)</span>

<span class="kn">depends:</span> <span class="n">R_library</span><span class="p">(</span><span class="s1">&#39;rhdf5&#39;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.mash.rds&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">str</span><span class="p">(</span><span class="n">_input</span><span class="p">)</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.portable.h5&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s2">nn</span><span class="si">}</span><span class="s2">.mash.rds&quot;</span>

<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;36h&#39;</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;4G&#39;</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>

<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span>
    <span class="n">ConvertP2Z</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">pval</span><span class="p">,</span> <span class="n">beta</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">z</span> <span class="o">&lt;-</span> <span class="nb">abs</span><span class="p">(</span><span class="n">qnorm</span><span class="p">(</span><span class="n">pval</span> <span class="o">/</span> <span class="mi">2</span><span class="p">))</span>
      <span class="n">z</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">beta</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">z</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">beta</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)]</span>
      <span class="k">return</span><span class="p">(</span><span class="n">z</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">handle_nan_b</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
          <span class="n">x</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span> <span class="o">=</span> <span class="mi">0</span>
          <span class="k">return</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">handle_nan_s</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
          <span class="n">x</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">infinite</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">|</span> <span class="n">x</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)]</span> <span class="o">=</span> <span class="mf">1E3</span>
          <span class="k">return</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="n">handle_nan_z</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">z</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="p">{</span>
          <span class="n">z</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">infinite</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">|</span> <span class="n">s</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
          <span class="n">z</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="o">|</span> <span class="n">b</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
          <span class="n">z</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">nan</span><span class="p">(</span><span class="n">z</span><span class="p">)</span> <span class="o">|</span> <span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">z</span><span class="p">))]</span> <span class="o">=</span> <span class="mi">0</span>
          <span class="k">return</span><span class="p">(</span><span class="n">z</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="n">GetSS</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">db</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">rhdf5</span><span class="p">::</span><span class="n">h5read</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">table</span><span class="p">)</span>
      <span class="n">dat</span><span class="err">$</span><span class="s2">&quot;z&quot;</span> <span class="o">&lt;-</span> <span class="n">ConvertP2Z</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="s2">&quot;pval&quot;</span><span class="p">,</span> <span class="n">dat</span><span class="err">$</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span>
      <span class="k">for</span> <span class="p">(</span><span class="kp">name</span> <span class="ow">in</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="s2">&quot;se&quot;</span><span class="p">,</span> <span class="s2">&quot;pval&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">dat</span><span class="p">[[</span><span class="kp">name</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="kp">name</span><span class="p">]]))</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="kp">name</span><span class="p">]])</span> <span class="o">&lt;-</span> <span class="n">dat</span><span class="err">$</span><span class="n">colnames</span>
        <span class="n">rownames</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="kp">name</span><span class="p">]])</span> <span class="o">&lt;-</span> <span class="n">dat</span><span class="err">$</span><span class="n">rownames</span>
      <span class="p">}</span>
      <span class="n">dat</span><span class="err">$</span><span class="n">colnames</span> <span class="o">&lt;-</span> <span class="n">dat</span><span class="err">$</span><span class="n">rownames</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
      <span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;z&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">handle_nan_z</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;z&#39;</span><span class="p">]],</span> <span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;beta&#39;</span><span class="p">]],</span> <span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;se&#39;</span><span class="p">]])</span>
      <span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;se&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">handle_nan_s</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;se&#39;</span><span class="p">]])</span>
      <span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;beta&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">handle_nan_b</span><span class="p">(</span><span class="n">dat</span><span class="p">[[</span><span class="s1">&#39;beta&#39;</span><span class="p">]])</span>
      <span class="k">return</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="n">SplitTrainTest</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">table</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># load data</span>
        <span class="n">strong</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">strong</span><span class="p">[[</span><span class="n">table</span><span class="p">]]</span>
        <span class="n">random</span> <span class="o">=</span> <span class="n">dat</span><span class="err">$</span><span class="n">random</span><span class="p">[[</span><span class="n">table</span><span class="p">]]</span>
        <span class="c1"># select rows to keep</span>
        <span class="n">num_train</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">random_snp_size</span><span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">num_train</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span>
          <span class="n">num_train</span> <span class="o">=</span> <span class="n">nrow</span><span class="p">(</span><span class="n">random</span><span class="p">)</span> <span class="o">*</span> <span class="n">num_train</span>
        <span class="n">num_train</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">integer</span><span class="p">(</span><span class="n">num_train</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">num_train</span> <span class="o">&gt;</span> <span class="n">nrow</span><span class="p">(</span><span class="n">random</span><span class="p">))</span> <span class="p">{</span>
            <span class="n">num_train</span> <span class="o">=</span> <span class="n">nrow</span><span class="p">(</span><span class="n">random</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="n">train</span> <span class="o">=</span> <span class="n">random</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="n">num_train</span><span class="p">,]</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">num_train</span> <span class="o">==</span> <span class="n">nrow</span><span class="p">(</span><span class="n">random</span><span class="p">))</span> <span class="p">{</span>
            <span class="n">test</span> <span class="o">=</span> <span class="n">NULL</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
            <span class="n">test</span> <span class="o">=</span> <span class="n">random</span><span class="p">[(</span><span class="n">num_train</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span><span class="n">nrow</span><span class="p">(</span><span class="n">random</span><span class="p">),]</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">effects_list</span><span class="p">:</span><span class="n">r</span><span class="p">}</span> <span class="o">!=</span> <span class="s1">&#39;NULL&#39;</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">pout</span> <span class="o">=</span> <span class="n">scan</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">effects_list</span><span class="p">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">what</span><span class="o">=</span><span class="s2">&quot;character&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">NULL</span><span class="p">)</span>
            <span class="n">strong</span> <span class="o">=</span> <span class="n">strong</span><span class="p">[(</span><span class="n">rownames</span><span class="p">(</span><span class="n">strong</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pout</span><span class="p">),]</span>
            <span class="n">train</span> <span class="o">=</span> <span class="n">train</span><span class="p">[(</span><span class="n">rownames</span><span class="p">(</span><span class="n">train</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pout</span><span class="p">),]</span>
            <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">test</span><span class="p">))</span>
                <span class="n">test</span> <span class="o">=</span> <span class="n">test</span><span class="p">[(</span><span class="n">rownames</span><span class="p">(</span><span class="n">test</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pout</span><span class="p">),]</span>                                         
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">conditions_list</span><span class="p">:</span><span class="n">r</span><span class="p">}</span> <span class="o">!=</span> <span class="s1">&#39;NULL&#39;</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">rout</span> <span class="o">=</span> <span class="n">scan</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">conditions_list</span><span class="p">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">what</span><span class="o">=</span><span class="s2">&quot;character&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">NULL</span><span class="p">)</span>
            <span class="n">strong</span> <span class="o">=</span> <span class="n">strong</span><span class="p">[,(</span><span class="n">colnames</span><span class="p">(</span><span class="n">strong</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">rout</span><span class="p">),</span><span class="n">drop</span><span class="o">=</span><span class="n">F</span><span class="p">]</span>
            <span class="n">train</span> <span class="o">=</span> <span class="n">train</span><span class="p">[,(</span><span class="n">colnames</span><span class="p">(</span><span class="n">train</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">rout</span><span class="p">),</span><span class="n">drop</span><span class="o">=</span><span class="n">F</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">test</span><span class="p">))</span>
                <span class="n">test</span> <span class="o">=</span> <span class="n">test</span><span class="p">[,(</span><span class="n">colnames</span><span class="p">(</span><span class="n">test</span><span class="p">)</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">rout</span><span class="p">),</span><span class="n">drop</span><span class="o">=</span><span class="n">F</span><span class="p">]</span>
        <span class="p">}</span> 

        <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">random</span> <span class="o">=</span> <span class="n">train</span><span class="p">,</span>
               <span class="n">random</span><span class="o">.</span><span class="n">test</span> <span class="o">=</span> <span class="n">test</span><span class="p">,</span>
               <span class="n">strong</span> <span class="o">=</span> <span class="n">strong</span><span class="p">))</span>
    <span class="p">}</span>
      
    <span class="n">SS_data</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">strong</span> <span class="o">=</span> <span class="n">GetSS</span><span class="p">(</span><span class="s1">&#39;strong&#39;</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">}),</span> <span class="n">random</span> <span class="o">=</span> <span class="n">GetSS</span><span class="p">(</span><span class="s1">&#39;random&#39;</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">}))</span>
    <span class="n">ztable</span> <span class="o">=</span> <span class="n">SplitTrainTest</span><span class="p">(</span><span class="n">SS_data</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">)</span>
    <span class="n">btable</span> <span class="o">=</span> <span class="n">SplitTrainTest</span><span class="p">(</span><span class="n">SS_data</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">)</span>
    <span class="n">stable</span> <span class="o">=</span> <span class="n">SplitTrainTest</span><span class="p">(</span><span class="n">SS_data</span><span class="p">,</span> <span class="s2">&quot;se&quot;</span><span class="p">)</span>
    <span class="c1"># save output</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">z</span> <span class="o">=</span> <span class="n">ztable</span><span class="err">$</span><span class="n">random</span><span class="p">,</span>
                 <span class="n">random</span><span class="o">.</span><span class="n">test</span><span class="o">.</span><span class="n">z</span> <span class="o">=</span> <span class="n">ztable</span><span class="err">$</span><span class="n">random</span><span class="o">.</span><span class="n">test</span><span class="p">,</span>
                 <span class="n">strong</span><span class="o">.</span><span class="n">z</span> <span class="o">=</span> <span class="n">ztable</span><span class="err">$</span><span class="n">strong</span><span class="p">,</span>
                 <span class="n">random</span><span class="o">.</span><span class="n">b</span> <span class="o">=</span> <span class="n">btable</span><span class="err">$</span><span class="n">random</span><span class="p">,</span>
                 <span class="n">random</span><span class="o">.</span><span class="n">test</span><span class="o">.</span><span class="n">b</span> <span class="o">=</span> <span class="n">btable</span><span class="err">$</span><span class="n">random</span><span class="o">.</span><span class="n">test</span><span class="p">,</span>
                 <span class="n">strong</span><span class="o">.</span><span class="n">b</span> <span class="o">=</span> <span class="n">btable</span><span class="err">$</span><span class="n">strong</span><span class="p">,</span>
                 <span class="n">random</span><span class="o">.</span><span class="n">s</span> <span class="o">=</span> <span class="n">stable</span><span class="err">$</span><span class="n">random</span><span class="p">,</span>
                 <span class="n">random</span><span class="o">.</span><span class="n">test</span><span class="o">.</span><span class="n">s</span> <span class="o">=</span> <span class="n">stable</span><span class="err">$</span><span class="n">random</span><span class="o">.</span><span class="n">test</span><span class="p">,</span>
                 <span class="n">strong</span><span class="o">.</span><span class="n">s</span> <span class="o">=</span> <span class="n">stable</span><span class="err">$</span><span class="n">strong</span><span class="p">),</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<p>The output is an R list of data matrices:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="s1">&#39;FastQTLSumStats.mash.rds&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">names</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;random.z&quot;</span>      <span class="s2">&quot;random.test.z&quot;</span> <span class="s2">&quot;strong.z&quot;</span>      <span class="s2">&quot;random.b&quot;</span>     
<span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="s2">&quot;random.test.b&quot;</span> <span class="s2">&quot;strong.b&quot;</span>      <span class="s2">&quot;random.s&quot;</span>      <span class="s2">&quot;random.test.s&quot;</span>
<span class="p">[</span><span class="mi">9</span><span class="p">]</span> <span class="s2">&quot;strong.s&quot;</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">*.z</span></code>, <code class="docutils literal notranslate"><span class="pre">*.b</span></code> and <code class="docutils literal notranslate"><span class="pre">*.s</span></code> stands for z-scores, <span class="math notranslate nohighlight">\(\hat{\beta}\)</span> and <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> (standard error). <code class="docutils literal notranslate"><span class="pre">strong</span></code> is data of the “strongest” gene-SNP pair; <code class="docutils literal notranslate"><span class="pre">random</span></code> is a random subset of gene-SNP pairs. They are used to find pattern of sharing in strong signals and learning mixture weights. Additionally <code class="docutils literal notranslate"><span class="pre">random.test.*</span></code> is a seperate random data-set that can be used to assess fit / overfit of the MASH mixture model.</p>
</div>
</div>
<div class="section" id="export-workflow-documentation-to-html-format">
<h2>Export workflow documentation to HTML format<a class="headerlink" href="#export-workflow-documentation-to-html-format" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">workflow</span><span class="o">/</span><span class="n">fastqtl_to_mash</span> <span class="n">document_it</span>
</pre></div>
</div>
<p>will export the narrative in this workflow notebook to HTML format as a documentation file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Export workflow to HTML document</span>
<span class="p">[</span><span class="n">document_it</span><span class="p">]</span>
<span class="kn">input:</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="k">if</span> <span class="n">item</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.ipynb&#39;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">item</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.html&#39;</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="k">if</span> <span class="n">item</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.ipynb&#39;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">2</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="n">sos</span> <span class="n">convert</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span> <span class="o">--</span><span class="n">template</span> <span class="n">sos</span><span class="o">-</span><span class="n">report</span><span class="o">-</span><span class="n">toc</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/multivariate/MASH"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../MASH_pipeline.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Multivariate association with MASH</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="mixture_prior.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">A multivariate EBNM approach for mixture multivariate distribution estimate</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>