
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>MASH analysis pipeline with data-driven prior matrices &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="MASH analysis pipeline with posterior computation" href="mash_posterior.html" />
    <link rel="prev" title="A multivariate EBNM approach for mixture multivariate distribution estimate" href="mixture_prior.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../cis_analysis/cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../cis_analysis/cis_workhorse.html">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cis_analysis/mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../trans_analysis/rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../trans_analysis/polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../trans_analysis/MRAID_QTL.html">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/gregor.html">
   Enrichment analysis: TORUS and GREGOR
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../pecotmr_analysis/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/ptwas_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../pecotmr_analysis/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/multivariate/MASH/mash_fit.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate/MASH/mash_fit.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/multivariate/MASH/mash_fit.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods">
   Methods
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multivariate-adaptive-shrinkage-mash-analysis-of-eqtl-data">
     Multivariate adaptive shrinkage (MASH) analysis of eQTL data
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mwe-data">
   MWE Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multivariate-analysis-with-mash-model">
   Multivariate analysis with MASH model
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#global-parameter-settings">
   Global parameter settings
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mashr-mixture-model-fitting">
   <code class="docutils literal notranslate">
    <span class="pre">
     mashr
    </span>
   </code>
   mixture model fitting
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#optional-posterior-computations">
   Optional posterior computations
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="mash-analysis-pipeline-with-data-driven-prior-matrices">
<h1>MASH analysis pipeline with data-driven prior matrices<a class="headerlink" href="#mash-analysis-pipeline-with-data-driven-prior-matrices" title="Permalink to this headline">¶</a></h1>
<p>In this notebook, we utilize the MASH prior, referred to as the <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/blob/6c637645ce16aee2aa7dc86bbc334fb6bb66b9d9/code/multivariate/MASH/mixture_prior.ipynb#L4">mixture_prior</a>, from a previous step. Our objective is to conduct a multivariate analysis under the MASH model. After fitting the model, we subsequently compute the posteriors for our variables of interest.</p>
<div class="section" id="methods">
<h2>Methods<a class="headerlink" href="#methods" title="Permalink to this headline">¶</a></h2>
<div class="section" id="multivariate-adaptive-shrinkage-mash-analysis-of-eqtl-data">
<h3>Multivariate adaptive shrinkage (MASH) analysis of eQTL data<a class="headerlink" href="#multivariate-adaptive-shrinkage-mash-analysis-of-eqtl-data" title="Permalink to this headline">¶</a></h3>
<p>Since we published Urbut 2019, we have improved implementation of MASH algorithm and made a new R package, <a class="reference external" href="https://github.com/stephenslab/mashr"><code class="docutils literal notranslate"><span class="pre">mashr</span></code></a>. Major improvements compared to Urbut 2019 are:</p>
<ol class="simple">
<li><p>Faster computation of likelihood and posterior quantities via matrix algebra tricks and a C++ implementation.</p></li>
<li><p>Faster computation of MASH mixture via convex optimization.</p></li>
<li><p>New ways to estimate prior in place of the <code class="docutils literal notranslate"><span class="pre">SFA</span></code> approach, see <code class="docutils literal notranslate"><span class="pre">mixture_prior.ipynb</span></code> for details.</p></li>
<li><p>Improve estimate of residual variance <span class="math notranslate nohighlight">\(\hat{V}\)</span>.</p></li>
</ol>
<p>At this point, the input data have already been converted from the original association summary statistics to a format convenient for analysis in MASH.</p>
</div>
</div>
<div class="section" id="mwe-data">
<h2>MWE Data<a class="headerlink" href="#mwe-data" title="Permalink to this headline">¶</a></h2>
<p>Avaiable on <a class="reference external" href="https://www.synapse.org/#%21Synapse:syn52624471">synapse.org</a></p>
</div>
<div class="section" id="multivariate-analysis-with-mash-model">
<h2>Multivariate analysis with MASH model<a class="headerlink" href="#multivariate-analysis-with-mash-model" title="Permalink to this headline">¶</a></h2>
<p>Using MWE with <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/blob/main/code/multivariate/MASH/mixture_prior.ipynb">prior previously generated</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_fit</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash</span> \
    <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="n">protocol_example</span> \
    <span class="o">--</span><span class="n">data</span> <span class="n">output</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">mashr_input</span><span class="o">.</span><span class="n">rds</span> \
    <span class="o">--</span><span class="n">vhat</span><span class="o">-</span><span class="n">data</span> <span class="n">output</span><span class="o">/</span><span class="n">mashr</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">ed_mixture</span><span class="o">.</span><span class="n">EZ</span><span class="o">.</span><span class="n">V_simple</span><span class="o">.</span><span class="n">rds</span> \
    <span class="o">--</span><span class="n">prior</span><span class="o">-</span><span class="n">data</span> <span class="n">output</span><span class="o">/</span><span class="n">mashr</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">ed_mixture</span><span class="o">.</span><span class="n">EZ</span><span class="o">.</span><span class="n">prior</span><span class="o">.</span><span class="n">rds</span> \
    <span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">posterior</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">mashr</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">stephenslab_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">mash_1</span>: Fit MASH mixture model (time estimate: &lt;15min for 70K by 49 matrix)
INFO: <span class=" -Color -Color-Green">mash_1</span> (index=0) is <span class=" -Color -Color-Green">ignored</span> due to saved signature
INFO: <span class=" -Color -Color-Green">mash_1</span> output:   <span class=" -Color -Color-Green">/home/gw/Documents/xQTL/output/mashr/protocol_example.EZ.mash_model.rds</span>
INFO: Running <span class=" -Color -Color-Green">mash_2</span>: Compute posterior for the &quot;strong&quot; set of data as in Urbut et al 2017. This is optional because most of the time we want to apply the MASH model learned on much larger data-set.
INFO: <span class=" -Color -Color-Green">mash_2</span> is <span class=" -Color -Color-Green">completed</span>.
INFO: <span class=" -Color -Color-Green">mash_2</span> output:   <span class=" -Color -Color-Green">/home/gw/Documents/xQTL/output/mashr/protocol_example.EZ.posterior.rds</span>
INFO: Workflow mash (ID=wdb1a3918af0fd3c3) is executed successfully with 1 completed step and 1 ignored step.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">readRDS</span><span class="p">(</span><span class="s1">&#39;output/mashr/protocol_example.EZ.posterior.rds&#39;</span><span class="p">)</span><span class="err">$</span><span class="n">lfsr</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<caption>A matrix: 17 × 8 of type dbl</caption>
<thead>
	<tr><th scope=col>ALL</th><th scope=col>Ast</th><th scope=col>End</th><th scope=col>Exc</th><th scope=col>Inh</th><th scope=col>Mic</th><th scope=col>OPC</th><th scope=col>Oli</th></tr>
</thead>
<tbody>
	<tr><td>1.000725e-06</td><td>1.206902e-01</td><td>0.38024501</td><td>2.341416e-05</td><td>0.021238026</td><td>2.438671e-02</td><td>2.822754e-01</td><td>4.950394e-02</td></tr>
	<tr><td>6.195229e-01</td><td>6.725207e-01</td><td>0.65046008</td><td>6.835928e-01</td><td>0.582288619</td><td>6.287649e-01</td><td>6.903822e-01</td><td>4.929722e-01</td></tr>
	<tr><td>5.502451e-08</td><td>6.952073e-05</td><td>0.42849153</td><td>3.848081e-09</td><td>0.001349139</td><td>2.974207e-01</td><td>2.267577e-01</td><td>2.193946e-01</td></tr>
	<tr><td>1.141256e-02</td><td>3.673827e-01</td><td>0.29984283</td><td>8.920542e-03</td><td>0.062520281</td><td>1.557580e-01</td><td>1.438823e-01</td><td>3.652727e-01</td></tr>
	<tr><td>4.803276e-01</td><td>1.793390e-01</td><td>0.01916176</td><td>1.752551e-01</td><td>0.384661176</td><td>3.138391e-01</td><td>4.888775e-01</td><td>3.793039e-01</td></tr>
	<tr><td>3.240321e-01</td><td>1.006760e-04</td><td>0.28423516</td><td>2.692014e-10</td><td>0.464610846</td><td>8.359305e-04</td><td>8.654122e-21</td><td>1.358764e-09</td></tr>
	<tr><td>4.744668e-01</td><td>3.330669e-16</td><td>0.35728521</td><td>2.052281e-05</td><td>0.437127648</td><td>3.887724e-02</td><td>1.843843e-05</td><td>1.374751e-06</td></tr>
	<tr><td>3.166308e-03</td><td>0.000000e+00</td><td>0.34573821</td><td>2.111164e-04</td><td>0.424465824</td><td>1.686171e-02</td><td>9.992007e-16</td><td>2.119553e-07</td></tr>
	<tr><td>4.826046e-04</td><td>3.121070e-02</td><td>0.31763916</td><td>3.181940e-01</td><td>0.000107790</td><td>6.340894e-19</td><td>2.832298e-02</td><td>3.138958e-02</td></tr>
	<tr><td>2.426264e-01</td><td>3.950080e-01</td><td>0.16192905</td><td>1.023271e-01</td><td>0.284297524</td><td>3.167406e-04</td><td>1.449052e-02</td><td>4.348748e-01</td></tr>
	<tr><td>3.999852e-09</td><td>8.353939e-02</td><td>0.47399869</td><td>2.161030e-06</td><td>0.017915140</td><td>1.450814e-01</td><td>2.240465e-01</td><td>4.253355e-01</td></tr>
	<tr><td>8.830382e-06</td><td>5.623546e-03</td><td>0.23046666</td><td>7.952619e-08</td><td>0.465964274</td><td>9.821045e-02</td><td>6.871004e-02</td><td>2.039190e-01</td></tr>
	<tr><td>4.616982e-05</td><td>2.434419e-02</td><td>0.32092168</td><td>2.148510e-03</td><td>0.001318834</td><td>4.224729e-04</td><td>1.671158e-01</td><td>2.667831e-01</td></tr>
	<tr><td>1.929503e-01</td><td>2.808606e-01</td><td>0.00177875</td><td>2.362496e-02</td><td>0.327244770</td><td>3.820119e-01</td><td>3.156853e-01</td><td>1.851656e-01</td></tr>
	<tr><td>2.744579e-02</td><td>5.054301e-02</td><td>0.15913565</td><td>1.581546e-04</td><td>0.086397136</td><td>3.507312e-01</td><td>9.995466e-02</td><td>1.706981e-01</td></tr>
	<tr><td>2.349326e-01</td><td>2.796353e-01</td><td>0.24466478</td><td>2.804030e-01</td><td>0.251516032</td><td>3.344815e-01</td><td>1.402243e-02</td><td>3.982418e-01</td></tr>
	<tr><td>1.711193e-09</td><td>6.569349e-03</td><td>0.03457580</td><td>4.631637e-01</td><td>0.422135645</td><td>3.970892e-01</td><td>2.515601e-01</td><td>5.054854e-14</td></tr>
</tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="global-parameter-settings">
<h2>Global parameter settings<a class="headerlink" href="#global-parameter-settings" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./mashr_workflow_output&#39;</span><span class="p">)</span>
<span class="c1"># Input summary statistics data</span>
<span class="kn">parameter:</span> <span class="n">data</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">prior_data</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">vhat_data</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Prefix of output files. If not specified, it will derive it from data.</span>
<span class="c1"># If it is specified, for example, `--output-prefix AnalysisResults`</span>
<span class="c1"># It will save output files as `{cwd}/AnalysisResults*`.</span>
<span class="kn">parameter:</span> <span class="n">output_prefix</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">output_suffix</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="c1"># Exchangable effect (EE) or exchangable z-scores (EZ)</span>
<span class="kn">parameter:</span> <span class="n">effect_model</span> <span class="o">=</span> <span class="s1">&#39;EZ&#39;</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">cwd</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="n">prior_data</span> <span class="o">=</span> <span class="n">prior_data</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="n">vhat_data</span> <span class="o">=</span> <span class="n">vhat_data</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">output_prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">&quot;</span>

<span class="n">mash_model</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.mash_model.rds&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">sort_uniq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="n">seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq</span> <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="n">seen</span> <span class="ow">or</span> <span class="n">seen</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="mashr-mixture-model-fitting">
<h2><code class="docutils literal notranslate"><span class="pre">mashr</span></code> mixture model fitting<a class="headerlink" href="#mashr-mixture-model-fitting" title="Permalink to this headline">¶</a></h2>
<p>Main reference are our <code class="docutils literal notranslate"><span class="pre">mashr</span></code> vignettes <a class="reference external" href="https://stephenslab.github.io/mashr/articles/eQTL_outline.html">this for mashr eQTL outline</a> and <a class="reference external" href="https://github.com/stephenslab/mashr/blob/master/vignettes/flash_mash.Rmd">this for using FLASH prior</a>.</p>
<p>The outcome of this workflow should be found under <code class="docutils literal notranslate"><span class="pre">./mashr_workflow_output</span></code> folder (can be configured). File names have pattern <code class="docutils literal notranslate"><span class="pre">*.mash_model_*.rds</span></code>. They can be used to computer posterior for input list of gene-SNP pairs (see next section).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Fit MASH mixture model (time estimate: &lt;15min for 70K by 49 matrix)</span>
<span class="p">[</span><span class="n">mash_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">output_level</span> <span class="o">=</span> <span class="mi">4</span>
<span class="kn">input:</span> <span class="n">data</span><span class="p">,</span> <span class="n">vhat_data</span><span class="p">,</span> <span class="n">prior_data</span>
<span class="kn">output:</span> <span class="n">mash_model</span>

<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="n">mashr</span><span class="p">)</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">vhat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">U</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span><span class="err">$</span><span class="n">U</span>
    <span class="n">mash_data</span> <span class="o">=</span> <span class="n">mash_set_data</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">random</span><span class="o">.</span><span class="n">b</span><span class="p">,</span> <span class="n">Shat</span><span class="o">=</span><span class="n">dat</span><span class="err">$</span><span class="n">random</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="mi">1</span> <span class="k">if</span> <span class="n">effect_model</span> <span class="o">==</span> <span class="s1">&#39;EZ&#39;</span> <span class="k">else</span> <span class="mi">0</span><span class="p">},</span> <span class="n">V</span><span class="o">=</span><span class="n">vhat</span><span class="p">,</span> <span class="n">zero_Bhat_Shat_reset</span> <span class="o">=</span> <span class="mf">1E3</span><span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">mash_model</span> <span class="o">=</span> <span class="n">mash</span><span class="p">(</span><span class="n">mash_data</span><span class="p">,</span> <span class="n">Ulist</span> <span class="o">=</span> <span class="n">U</span><span class="p">,</span> <span class="n">outputlevel</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">output_level</span><span class="p">}),</span> <span class="n">vhat_file</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">},</span> <span class="n">prior_file</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">r</span><span class="p">}),</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="optional-posterior-computations">
<h2>Optional posterior computations<a class="headerlink" href="#optional-posterior-computations" title="Permalink to this headline">¶</a></h2>
<p>Additionally provide posterior for the “strong” set in MASH input data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute posterior for the &quot;strong&quot; set of data as in Urbut et al 2017.</span>
<span class="c1"># This is optional because most of the time we want to apply the </span>
<span class="c1"># MASH model learned on much larger data-set.</span>
<span class="p">[</span><span class="n">mash_2</span><span class="p">]</span>
<span class="c1"># default to True; use --no-compute-posterior to disable this</span>
<span class="kn">parameter:</span> <span class="n">compute_posterior</span> <span class="o">=</span> <span class="kc">True</span>
<span class="c1"># input Vhat file for the batch of posterior data</span>
<span class="n">skip_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">compute_posterior</span><span class="p">)</span>

<span class="kn">input:</span> <span class="n">data</span><span class="p">,</span> <span class="n">vhat_data</span><span class="p">,</span> <span class="n">mash_model</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.posterior.rds&quot;</span>

<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>  
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">cwd</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span><span class="o">=</span><span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="n">mashr</span><span class="p">)</span>
    <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">vhat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">r</span><span class="p">})</span>
    <span class="n">mash_data</span> <span class="o">=</span> <span class="n">mash_set_data</span><span class="p">(</span><span class="n">dat</span><span class="err">$</span><span class="n">strong</span><span class="o">.</span><span class="n">b</span><span class="p">,</span> <span class="n">Shat</span><span class="o">=</span><span class="n">dat</span><span class="err">$</span><span class="n">strong</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="mi">1</span> <span class="k">if</span> <span class="n">effect_model</span> <span class="o">==</span> <span class="s1">&#39;EZ&#39;</span> <span class="k">else</span> <span class="mi">0</span><span class="p">},</span> <span class="n">V</span><span class="o">=</span><span class="n">vhat</span><span class="p">,</span> <span class="n">zero_Bhat_Shat_reset</span> <span class="o">=</span> <span class="mf">1E3</span><span class="p">)</span>
    <span class="n">mash_model</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">ar</span><span class="p">})</span><span class="err">$</span><span class="n">mash_model</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">mash_compute_posterior_matrices</span><span class="p">(</span><span class="n">mash_model</span><span class="p">,</span> <span class="n">mash_data</span><span class="p">),</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/multivariate/MASH"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="mixture_prior.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">A multivariate EBNM approach for mixture multivariate distribution estimate</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="mash_posterior.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">MASH analysis pipeline with posterior computation</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>