
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Advanced regression models for association analysis with individual level data &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/mnm_analysis/mnm_methods/mnm_regression';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="High-dimensional regression with summary statistics" href="rss_analysis.html" />
    <link rel="prev" title="Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE" href="../summary_stats_finemapping_vignette.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">Hierarchical Multiple Testing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/mnm_analysis/mnm_methods/mnm_regression.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/mnm_analysis/mnm_methods/mnm_regression.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/mnm_analysis/mnm_methods/mnm_regression.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Advanced regression models for association analysis with individual level data</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-association-window-file">Example association-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-susie-with-twas-weights-from-multiple-methods">i. SuSiE with TWAS weights from multiple methods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-fsusie">ii. fSuSiE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-mvsusie-with-twas-weights-from-mr-mash">iii. mvSuSiE with TWAS weights from mr.mash</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-analysis-susie-and-twas">Univariate analysis: SuSiE and TWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-analysis-mvsusie-and-mr-mash">Multivariate analysis: mvSuSiE and mr.mash</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#xqtl-fine-mapping-across-multiple-genes">xQTL fine-mapping across multiple genes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-for-epigenomic-qtl-fine-mapping">Functional regression fSuSiE for epigenomic QTL fine-mapping</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-with-other-modality">Functional regression fSuSiE with other modality</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="advanced-regression-models-for-association-analysis-with-individual-level-data">
<h1>Advanced regression models for association analysis with individual level data<a class="headerlink" href="#advanced-regression-models-for-association-analysis-with-individual-level-data" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>This notebook performs various advanced statistical analysis on multiple xQTL in a given region. Current procedures implemented include:</p>
<ol class="arabic simple">
<li><p>Univariate analysis</p>
<ul class="simple">
<li><p>SuSiE</p></li>
<li><p>Univeriate TWAS weights: LASSO, Elastic net, mr.ash, SuSiE, Bayes alphabet soup</p></li>
<li><p>Cross validation of TWAS methods (optional but highly recommended if TWAS weights are computed)</p></li>
</ul>
</li>
<li><p>Functional data (epigenomic xQTL) analysis</p>
<ul class="simple">
<li><p>fSuSiE</p></li>
</ul>
</li>
<li><p>Multivariate analysis</p>
<ul class="simple">
<li><p>mvSuSiE</p></li>
<li><p>Multivariate TWAS weights: mvSuSiE and mr.mash</p></li>
</ul>
</li>
</ol>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>A list of regions to be analyzed (optional); the last column of this file should be region name.</p></li>
<li><p>Either a list of per chromosome genotype files, or one file for genotype data of the entire genome. Genotype data has to be in PLINK <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p></li>
<li><p>Vector of lists of phenotype files per region to be analyzed, in UCSC <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> with index in <code class="docutils literal notranslate"><span class="pre">bed.gz.tbi</span></code> formats.</p></li>
<li><p>Vector of covariate files corresponding to the lists above.</p></li>
<li><p>Customized association windows file for variants (cis or trans). If it is not provided, a fixed sized window will be used around the region (a cis-window)</p></li>
<li><p>Optionally a vector of names of the phenotypic conditions in the form of <code class="docutils literal notranslate"><span class="pre">cond1</span> <span class="pre">cond2</span> <span class="pre">cond3</span></code> separated with whitespace.</p></li>
</ol>
<p>Input 2 and 3 should be outputs from <code class="docutils literal notranslate"><span class="pre">genotype_per_region</span></code> and <code class="docutils literal notranslate"><span class="pre">annotate_coord</span></code> modules in previous preprocessing steps. 4 should be output of <code class="docutils literal notranslate"><span class="pre">covariate_preprocessing</span></code> pipeline that contains genotype PC, phenotypic hidden confounders and fixed covariates.</p>
<section id="example-genotype-data">
<h3>Example genotype data<a class="headerlink" href="#example-genotype-data" title="Link to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr        path</span>
<span class="n">chr21</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21</span><span class="o">.</span><span class="n">bed</span>
<span class="n">chr22</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr22</span><span class="o">.</span><span class="n">bed</span>
</pre></div>
</div>
<p>Alternatively, simply use <code class="docutils literal notranslate"><span class="pre">protocol_example.genotype.chr21_22.bed</span></code> if all chromosomes are in the same file.</p>
</section>
<section id="example-phenotype-list">
<h3>Example phenotype list<a class="headerlink" href="#example-phenotype-list" title="Link to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end ID  path</span>
<span class="n">chr12</span>   <span class="mi">752578</span>  <span class="mi">752579</span>  <span class="n">ENSG00000060237</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">990508</span>  <span class="mi">990509</span>  <span class="n">ENSG00000082805</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">2794969</span> <span class="mi">2794970</span> <span class="n">ENSG00000004478</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">4649113</span> <span class="mi">4649114</span> <span class="n">ENSG00000139180</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6124769</span> <span class="mi">6124770</span> <span class="n">ENSG00000110799</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6534516</span> <span class="mi">6534517</span> <span class="n">ENSG00000111640</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="example-association-window-file">
<h3>Example association-window file<a class="headerlink" href="#example-association-window-file" title="Link to this heading">#</a></h3>
<p>It should have strictly 4 columns, with the header a commented out line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start    end    gene_id</span>
<span class="n">chr10</span>   <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008128</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008130</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000067606</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7101193</span>    <span class="n">ENSG00000069424</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7960000</span>    <span class="n">ENSG00000069812</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078369</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078808</span>
</pre></div>
</div>
<p>The key is that the 4th column ID should match with the 4th column ID in the phenotype list. Otherwise the association-window to analyze will not be found.</p>
</section>
<section id="about-indels">
<h3>About indels<a class="headerlink" href="#about-indels" title="Link to this heading">#</a></h3>
<p>Option <code class="docutils literal notranslate"><span class="pre">--no-indel</span></code> will remove indel from analysis. FIXME: Gao need to provide more guidelines how to deal with indels in practice.</p>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<p>For each analysis region, the output is SuSiE model fitted and saved in RDS format.</p>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<section id="i-susie-with-twas-weights-from-multiple-methods">
<h3>i. SuSiE with TWAS weights from multiple methods<a class="headerlink" href="#i-susie-with-twas-weights-from-multiple-methods" title="Link to this heading">#</a></h3>
<p>Timing [FIXME]</p>
<p>Below we duplicate the examples for phenotype and covariates to demonstrate that when there are multiple phenotypes for the same genotype it is possible to use this pipeline to analyze all of them (more than two is accepted as well).</p>
<p>Here using <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> we focus the analysis on 3 genes. In practice if this parameter is dropped, the union of all regions in all phenotype region lists will be analyzed. It is possible for some of the regions there are no genotype data, in which case the pipeline will output RDS files with a warning message to indicate the lack of genotype data to analyze.</p>
<p><strong>Note:</strong> Suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/mnm_regression.ipynb susie_twas  \
    --name protocol_example_protein  \
    --genoFile input/xqtl_association/protocol_example.genotype.chr21_22.bed   \
    --phenoFile output/phenotype/protocol_example.protein.region_list.txt \
                output/phenotype/protocol_example.protein.region_list.txt \
    --covFile output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz \
              output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz  \
    --customized-association-windows input/xqtl_association/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --region-name ENSG00000241973_P42356 ENSG00000160209_O00764 ENSG00000100412_Q99798 \
    --phenotype-names trait_A trait_B \
    --container oras://ghcr.io/cumc/pecotmr_apptainer:latest
</pre></div>
</div>
</div>
</div>
<p>It is also possible to analyze a selected list of regions using option <code class="docutils literal notranslate"><span class="pre">--region-list</span></code>. The last column of this file will be used for the list to analyze. Here for example use the same list of regions as we used for customized association-window:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run xqtl-protocol/pipeline/mnm_regression.ipynb susie_twas  \
    --name protocol_example_protein  \
    --genoFile xqtl_association/protocol_example.genotype.chr21_22.bed   \
    --phenoFile output/phenotype/protocol_example.protein.region_list.txt \
                output/phenotype/protocol_example.protein.region_list.txt \
    --covFile output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz \
              output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz  \
    --customized-association-windows xqtl_association/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --region-list xqtl_association/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --phenotype-names trait_A trait_B \
    --container oras://ghcr.io/cumc/pecotmr_apptainer:latest
</pre></div>
</div>
</div>
</div>
<p><strong>Note:</strong> When both <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> and <code class="docutils literal notranslate"><span class="pre">--region-list</span></code> are used, the union of regions from these parameters will be analyzed.</p>
<p>FIXME: We should probably just explain these parameters, will work better for conversion script</p>
<p>To perform fine-mapping only without TWAS weights,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">twas</span><span class="o">-</span><span class="n">weights</span> <span class="o">...</span> <span class="c1"># rest of parameters the same. </span>
</pre></div>
</div>
<p>To perform fine-mapping and TWAS weights without cross validation,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">twas</span><span class="o">-</span><span class="n">cv</span><span class="o">-</span><span class="n">folds</span> <span class="mi">0</span> <span class="o">...</span> <span class="c1"># rest of parameters the same. </span>
</pre></div>
</div>
<p>It is also possible to specify a subset of samples to analyze, using <code class="docutils literal notranslate"><span class="pre">--keep-samples</span></code> parameter. For example we create a file to keep the ID of 50 samples,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>zcat output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz | head -1 | awk &#39;{for (i=2; i&lt;=51; i++) printf $i &quot; &quot;; print &quot;&quot;}&#39;&gt; output/keep_samples.txt
</pre></div>
</div>
</div>
</div>
<p>then use them in our analysis,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">txt</span> <span class="o">...</span> <span class="c1"># rest of parameters the same</span>
</pre></div>
</div>
</section>
<section id="ii-fsusie">
<h3>ii. fSuSiE<a class="headerlink" href="#ii-fsusie" title="Link to this heading">#</a></h3>
<p>Timing [FIXME]</p>
<p><strong>Note:</strong> Suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/mnm_regression.ipynb fsusie \
    --name protocol_example_methylation \
    --genoFile xqtl_association/protocol_example.genotype.chr21_22.plink_per_chrom.txt \
    --phenoFile output/phenotype_by_region/protocol_example.methylation.bed.phenotype_by_region_files.txt \
                output/phenotype_by_region/protocol_example.methylation.bed.phenotype_by_region_files.txt  \
    --covFile output/covariate/protocol_example.methylation.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz \
              output/covariate/protocol_example.methylation.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz \
    --container oras://ghcr.io/cumc/pecotmr_apptainer:latest
</pre></div>
</div>
</div>
</div>
</section>
<section id="iii-mvsusie-with-twas-weights-from-mr-mash">
<h3>iii. mvSuSiE with TWAS weights from mr.mash<a class="headerlink" href="#iii-mvsusie-with-twas-weights-from-mr-mash" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/mnm_regression.ipynb mnm \
    --name multi_trait \
    --genoFile /mnt/vast/hpc/csg/FunGen_xQTL/ROSMAP/Genotype/geno_by_chrom/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.11.bed \
    --phenoFile /mnt/vast/hpc/csg/wanggroup/fungen-xqtl-analysis/analysis/Wang_Columbia/ROSMAP/pseudo_bulk_eqtl/Mic/output/data_preprocessing/phenotype_data/Mic.log2cpm.region_list.txt \
      /mnt/vast/hpc/csg/wanggroup/fungen-xqtl-analysis/analysis/Wang_Columbia/ROSMAP/pseudo_bulk_eqtl/Ast/output/data_preprocessing/phenotype_data/Ast.log2cpm.region_list.txt \
      /mnt/vast/hpc/csg/wanggroup/fungen-xqtl-analysis/analysis/Wang_Columbia/ROSMAP/pseudo_bulk_eqtl/Oli/output/data_preprocessing/phenotype_data/Oli.log2cpm.region_list.txt \
    --covFile /mnt/vast/hpc/csg/wanggroup/fungen-xqtl-analysis/analysis/Wang_Columbia/ROSMAP/pseudo_bulk_eqtl/Mic/output/data_preprocessing/covariate_data/Mic.log2cpm.Mic.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.snuc_pseudo_bulk.related.plink_qc.extracted.pca.projected.Marchenko_PC.gz \
      /mnt/vast/hpc/csg/wanggroup/fungen-xqtl-analysis/analysis/Wang_Columbia/ROSMAP/pseudo_bulk_eqtl/Ast/output/data_preprocessing/covariate_data/Ast.log2cpm.Ast.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.snuc_pseudo_bulk.related.plink_qc.extracted.pca.projected.Marchenko_PC.gz \
      /mnt/vast/hpc/csg/wanggroup/fungen-xqtl-analysis/analysis/Wang_Columbia/ROSMAP/pseudo_bulk_eqtl/Oli/output/data_preprocessing/covariate_data/Oli.log2cpm.Oli.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.snuc_pseudo_bulk.related.plink_qc.extracted.pca.projected.Marchenko_PC.gz \
    --customized-association-windows /mnt/vast/hpc/csg/cl4215/mrmash/workflow/susie_twas/TADB_enhanced_cis.coding.bed \
    --cwd output/ \
    --region-name ENSG00000073921 \
    --phenotype-names Mic_DeJager_eQTL Ast_DeJager_eQTL Oli_DeJager_eQTL \
    --mixture_prior /home/aw3600/MR_KMT_analysis/PTWAS/twas_mr_test/mash_test/mixture_prior_example.EZ.prior.rds \
    --max_cv_variants 1000 \
    --ld_reference_meta_file /mnt/vast/hpc/csg/data_public/20240120_ADSP_LD_matrix/ld_meta_file.tsv
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run mnm_regression.ipynb -h
</pre></div>
</div>
</div>
</div>
</section>
<section id="workflow-implementation">
<h2>Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# It is required to input the name of the analysis
parameter: name = str
parameter: cwd = path(&quot;output&quot;)
# A list of file paths for genotype data, or the genotype data itself. 
parameter: genoFile = path
# One or multiple lists of file paths for phenotype data.
parameter: phenoFile = paths
# One or multiple lists of file paths for phenotype ID mapping file. The first column should be the original ID, the 2nd column should be the ID to be mapped to.
parameter: phenoIDFile = paths()
# Covariate file path
parameter: covFile = paths
# Optional: if a region list is provide the analysis will be focused on provided region. 
# The LAST column of this list will contain the ID of regions to focus on
# Otherwise, all regions with both genotype and phenotype files will be analyzed
parameter: region_list = path()
# Optional: if a region name is provided 
# the analysis would be focused on the union of provides region list and region names
parameter: region_name = []
# Only focus on a subset of samples
parameter: keep_samples = path()
# Only focus on a subset of variants
parameter: keep_variants = path()
# An optional list documenting the custom association window for each region to analyze, with four column, chr, start, end, region ID (eg gene ID).
# If this list is not provided, the default `window` parameter (see below) will be used.
parameter: customized_association_windows = path()
# Specify the cis window for the up and downstream radius to analyze around the region of interest in units of bp
# When this is set to negative, we will rely on using customized_association_windows
parameter: cis_window = -1
# save data object or not
parameter: save_data = False
# Name of phenotypes
parameter: phenotype_names = [f&#39;{x:bn}&#39; for x in phenoFile]
# And indicator whether it is trans-analysis, ie, not using phenotypic coordinate information
parameter: trans_analysis = False
parameter: seed = 999

# association analysis paramters
# initial number of single effects for SuSiE
parameter: init_L = 8
# maximum number of single effects to use for SuSiE
parameter: max_L = 30
# remove a variant if it has more than imiss missing individual level data
parameter: imiss = 1.0
# MAF and variance of X cutoff
parameter: maf = 0.0025
parameter: xvar_cutoff = 0.0
# MAC cutoff, on top of MAF cutoff
parameter: mac = 5
# Remove indels if indel = False
parameter: indel = True
parameter: pip_cutoff = 0.025
parameter: coverage = [0.95, 0.7, 0.5]
# If this value is not 0, then an initial single effect analysis will be performed 
# to determine if follow up analysis will be continued or to simply return NULL
# If this is negative we use a default way to determine this cutoff which is conservative but still useful
parameter: skip_analysis_pip_cutoff = []
# Skip fine-mapping
parameter: skip_fine_mapping = False
# Skip TWAS weights computation
parameter: skip_twas_weights = False
# Perform K folds valiation CV for TWAS
# Set it to zero if this is to be skipped
parameter: twas_cv_folds = 5
parameter: twas_cv_threads = twas_cv_folds
# maximum number of variants to consider for CV
# We will randomly pick a subset of it for CV purpose
# We can set it to eg 8000 to save computational burden althought may risk overfitting for methods comparison purpose
# When set to -1 we don&#39;t use this feature
parameter: max_cv_variants = -1
parameter: ld_reference_meta_file = path()
# Analysis environment settings
parameter: container = &quot;&quot;
import re
parameter: entrypoint= (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;
# For cluster jobs, number commands to run per job
parameter: job_size = 200
# Wall clock time expected
parameter: walltime = &quot;1h&quot;
# Memory expected
parameter: mem = &quot;20G&quot;
# Number of threads
parameter: numThreads = 1

if len(phenoFile) != len(covFile):
    raise ValueError(&quot;Number of input phenotypes files must match that of covariates files&quot;)
if len(phenoFile) != len(phenotype_names):
    raise ValueError(&quot;Number of input phenotypes files must match the number of phenotype names&quot;)
if len(phenoIDFile) &gt; 0 and len(phenoFile) != len(phenoIDFile):
    raise ValueError(&quot;Number of input phenotypes files must match the number of phenotype ID mapping files&quot;)

if len(skip_analysis_pip_cutoff) == 0:
    skip_analysis_pip_cutoff = [0.0] * len(phenoFile)
if len(skip_analysis_pip_cutoff) == 1:
    skip_analysis_pip_cutoff = skip_analysis_pip_cutoff * len(phenoFile)
if len(skip_analysis_pip_cutoff) != len(phenoFile):
    raise ValueError(f&quot;``skip_analysis_pip_cutoff`` should have either length 1 or length the same as phenotype files ({len(phenoFile)} in this case)&quot;)

# make it into an R List string
skip_analysis_pip_cutoff = [f&quot;&#39;{y}&#39;={x}&quot; for x,y in zip(skip_analysis_pip_cutoff, phenotype_names)]
    
def group_by_region(lst, partition):
    # from itertools import accumulate
    # partition = [len(x) for x in partition]
    # Compute the cumulative sums once
    # cumsum_vector = list(accumulate(partition))
    # Use slicing based on the cumulative sums
    # return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]
    return partition

import os
import pandas as pd

def adapt_file_path(file_path, reference_file):
    &quot;&quot;&quot;
    Adapt a single file path based on its existence and a reference file&#39;s path.

    Args:
    - file_path (str): The file path to adapt.
    - reference_file (str): File path to use as a reference for adaptation.

    Returns:
    - str: Adapted file path.

    Raises:
    - FileNotFoundError: If no valid file path is found.
    &quot;&quot;&quot;
    reference_path = os.path.dirname(reference_file)

    # Check if the file exists
    if os.path.isfile(file_path):
        return file_path

    # Check file name without path
    file_name = os.path.basename(file_path)
    if os.path.isfile(file_name):
        return file_name

    # Check file name in reference file&#39;s directory
    file_in_ref_dir = os.path.join(reference_path, file_name)
    if os.path.isfile(file_in_ref_dir):
        return file_in_ref_dir

    # Check original file path prefixed with reference file&#39;s directory
    file_prefixed = os.path.join(reference_path, file_path)
    if os.path.isfile(file_prefixed):
        return file_prefixed

    # If all checks fail, raise an error
    raise FileNotFoundError(f&quot;No valid path found for file: {file_path}&quot;)

def adapt_file_path_all(df, column_name, reference_file):
    return df[column_name].apply(lambda x: adapt_file_path(x, reference_file))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[get_analysis_regions: shared = &quot;regional_data&quot;]
# input is genoFile, phenoFile, covFile and optionally region_list. If region_list presents then we only analyze what&#39;s contained in the list.
# regional_data should be a dictionary like:
#{&#39;data&#39;: [(&quot;genotype_1.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;covariate_1.gz&quot;), (&quot;genotype_2.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;phenotype_2.bed.gz&quot;, &quot;covariate_1.gz&quot;, &quot;covariate_2.gz&quot;) ... ],
# &#39;meta_info&#39;: [(&quot;chr12:752578-752579&quot;,&quot;chr12:752577-752580&quot;, &quot;gene_1&quot;, &quot;trait_1&quot;), (&quot;chr13:852580-852581&quot;,&quot;chr13:852579-852580&quot;, &quot;gene_2&quot;, &quot;trait_1&quot;, &quot;trait_2&quot;) ... ]}
import numpy as np

def preload_id_map(id_map_files):
    id_maps = {}
    for id_map_file in id_map_files:
        if id_map_file is not None and os.path.isfile(id_map_file):
            df = pd.read_csv(id_map_file, sep=r&quot;\s+&quot;, header=None, comment=&#39;#&#39;, names=[&#39;old_ID&#39;, &#39;new_ID&#39;])
            id_maps[id_map_file] = df.set_index(&#39;old_ID&#39;)[&#39;new_ID&#39;].to_dict()
    return id_maps

def load_and_apply_id_map(pheno_path, id_map_path, preloaded_id_maps):
    pheno_df = pd.read_csv(pheno_path, sep=r&quot;\s+&quot;, header=0)
    pheno_df[&#39;Original_ID&#39;] = pheno_df[&#39;ID&#39;]
    if id_map_path in preloaded_id_maps:
        id_map = preloaded_id_maps[id_map_path]
        pheno_df[&#39;ID&#39;] = pheno_df[&#39;ID&#39;].map(id_map).fillna(pheno_df[&#39;ID&#39;])
    return pheno_df

def filter_by_region_ids(data, region_ids):
    if region_ids is not None and len(region_ids) &gt; 0:
        data = data[data[&#39;ID&#39;].isin(region_ids)]
    return data

def custom_join(series):
    # Initialize an empty list to hold the processed items
    result = []
    for item in series:
        if &#39;,&#39; in item:
            # If the item contains commas, split by comma and convert to tuple
            result.append(tuple(item.split(&#39;,&#39;)))
        else:
            # If the item does not contain commas, add it directly
            result.append(item)
    # Convert the list of items to a tuple and return
    return tuple(result)

def aggregate_phenotype_data(accumulated_pheno_df):
    if not accumulated_pheno_df.empty:
        accumulated_pheno_df = accumulated_pheno_df.groupby([&#39;#chr&#39;,&#39;ID&#39;,&#39;cond&#39;,&#39;path&#39;,&#39;cov_path&#39;], as_index=False).agg({
            &#39;#chr&#39;: lambda x: np.unique(x).astype(str)[0],
            &#39;ID&#39;: lambda x: np.unique(x).astype(str)[0],
            &#39;Original_ID&#39;: &#39;,&#39;.join,
            &#39;start&#39;: &#39;min&#39;,
            &#39;end&#39;: &#39;max&#39;
        }).groupby([&#39;#chr&#39;,&#39;ID&#39;], as_index=False).agg({
            &#39;cond&#39;: &#39;,&#39;.join,
            &#39;path&#39;: &#39;,&#39;.join,
            &#39;Original_ID&#39;: custom_join,
            &#39;cov_path&#39;: &#39;,&#39;.join,
            &#39;start&#39;: &#39;min&#39;,
            &#39;end&#39;: &#39;max&#39;
        })
    return accumulated_pheno_df

def process_cis_files(pheno_files, cov_files, phenotype_names, pheno_id_files, region_ids, preloaded_id_maps):
    &#39;&#39;&#39;
    Example output:
    #chr    start      end    ID  Original_ID   path     cov_path             cond
    chr12   752578   752579  ENSG00000060237  Q9H4A3,P62873  protocol_example.protein_1.bed.gz,protocol_example.protein_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B
    &#39;&#39;&#39;
    accumulated_pheno_df = pd.DataFrame()
    pheno_id_files = [None] * len(pheno_files) if len(pheno_id_files) == 0 else pheno_id_files

    for pheno_path, cov_path, phenotype_name, id_map_path in zip(pheno_files, cov_files, phenotype_names, pheno_id_files):
        if not os.path.isfile(cov_path):
            raise FileNotFoundError(f&quot;No valid path found for file: {cov_path}&quot;)
        pheno_df = load_and_apply_id_map(pheno_path, id_map_path, preloaded_id_maps)
        pheno_df = filter_by_region_ids(pheno_df, region_ids)
        
        if not pheno_df.empty:
            pheno_df.iloc[:, 4] = adapt_file_path_all(pheno_df, pheno_df.columns[4], f&quot;{pheno_path:a}&quot;)
            pheno_df = pheno_df.assign(cov_path=str(cov_path), cond=phenotype_name)           
            accumulated_pheno_df = pd.concat([accumulated_pheno_df, pheno_df], ignore_index=True)

    accumulated_pheno_df = aggregate_phenotype_data(accumulated_pheno_df)
    return accumulated_pheno_df

def process_trans_files(pheno_files, cov_files, phenotype_names, pheno_id_files, region_ids, customized_association_windows):
    &#39;&#39;&#39;
    Example output:
    #chr    start      end    ID  Original_ID   path     cov_path             cond
    chr21   0   0  chr21_18133254_19330300  carnitine,benzoate,hippurate  metabolon_1.bed.gz,metabolon_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B
    &#39;&#39;&#39;
    
    if not os.path.isfile(customized_association_windows):
        raise ValueError(&quot;Customized association analysis window must be specified for trans analysis.&quot;)
    accumulated_pheno_df = pd.DataFrame()
    pheno_id_files = [None] * len(pheno_files) if len(pheno_id_files) == 0 else pheno_id_files
    genotype_windows = pd.read_csv(customized_association_windows, comment=&quot;#&quot;, header=None, names=[&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;ID&quot;], sep=&quot;\t&quot;)
    genotype_windows = filter_by_region_ids(genotype_windows, region_ids)
    if genotype_windows.empty:
        return accumulated_pheno_df
    
    for pheno_path, cov_path, phenotype_name, id_map_path in zip(pheno_files, cov_files, phenotype_names, pheno_id_files):
        if not os.path.isfile(cov_path):
            raise FileNotFoundError(f&quot;No valid path found for file: {cov_path}&quot;)
        pheno_df = pd.read_csv(pheno_path, sep=r&quot;\s+&quot;, header=0, names=[&#39;Original_ID&#39;, &#39;path&#39;])
        if not pheno_df.empty:
            pheno_df.iloc[:, -1] = adapt_file_path_all(pheno_df, pheno_df.columns[-1], f&quot;{pheno_path:a}&quot;)
            pheno_df = pheno_df.assign(cov_path=str(cov_path), cond=phenotype_name)
            # Here we combine genotype_windows which contains &quot;#chr&quot; and &quot;ID&quot; to pheno_df by creating a cartesian product
            pheno_df = pd.merge(genotype_windows.assign(key=1), pheno_df.assign(key=1), on=&#39;key&#39;).drop(&#39;key&#39;, axis=1)
            # then set start and end columns to zero
            pheno_df[&#39;start&#39;] = 0
            pheno_df[&#39;end&#39;] = 0
            if id_map_path is not None:
                # Filter pheno_df by specific association-window and phenotype pairs
                association_analysis_pair = pd.read_csv(id_map_path, sep=r&quot;\s+&quot;, header=None, comment=&#39;#&#39;, names=[&#39;ID&#39;, &#39;Original_ID&#39;])
                pheno_df = pd.merge(pheno_df, association_analysis_pair, on=[&#39;ID&#39;, &#39;Original_ID&#39;])
            accumulated_pheno_df = pd.concat([accumulated_pheno_df, pheno_df], ignore_index=True)

    accumulated_pheno_df = aggregate_phenotype_data(accumulated_pheno_df)
    return accumulated_pheno_df

# Load genotype meta data
if f&quot;{genoFile:x}&quot; == &quot;.bed&quot;:
    geno_meta_data = pd.DataFrame([(&quot;chr&quot;+str(x), f&quot;{genoFile:a}&quot;) for x in range(1,23)] + [(&quot;chrX&quot;, f&quot;{genoFile:a}&quot;)], columns=[&#39;#chr&#39;, &#39;geno_path&#39;])
else:
    geno_meta_data = pd.read_csv(f&quot;{genoFile:a}&quot;, sep =r&quot;\s+&quot;, header=0)
    geno_meta_data.iloc[:, 1] = adapt_file_path_all(geno_meta_data, geno_meta_data.columns[1], f&quot;{genoFile:a}&quot;)
    geno_meta_data.columns = [&#39;#chr&#39;, &#39;geno_path&#39;]
    geno_meta_data[&#39;#chr&#39;] = geno_meta_data[&#39;#chr&#39;].apply(lambda x: str(x) if str(x).startswith(&#39;chr&#39;) else f&#39;chr{x}&#39;)

# Checking the DataFrame
valid_chr_values = [f&#39;chr{x}&#39; for x in range(1, 23)] + [&#39;chrX&#39;]
if not all(value in valid_chr_values for value in geno_meta_data[&#39;#chr&#39;]):
    raise ValueError(&quot;Invalid chromosome values found. Allowed values are chr1 to chr22 and chrX.&quot;)

region_ids = []
# If region_list is provided, read the file and extract IDs
if region_list.is_file():
    region_list_df = pd.read_csv(region_list, delim_whitespace=True, header=None, comment = &quot;#&quot;)
    region_ids = region_list_df.iloc[:, -1].unique()  # Extracting the last column for IDs

# If region_name is provided, include those IDs as well
# --region-name A B C will result in a list of [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;] here
if len(region_name) &gt; 0:
    region_ids = list(set(region_ids).union(set(region_name)))

if trans_analysis:
    meta_data = process_trans_files(phenoFile, covFile, phenotype_names, phenoIDFile, region_ids, customized_association_windows)
else:
    meta_data = process_cis_files(phenoFile, covFile, phenotype_names, phenoIDFile, region_ids, preload_id_map(phenoIDFile))
    
if not meta_data.empty:
    meta_data = meta_data.merge(geno_meta_data, on=&#39;#chr&#39;, how=&#39;inner&#39;)
    # Adjust association-window
    if os.path.isfile(customized_association_windows):
        print(f&quot;Loading customized association analysis window from {customized_association_windows}&quot;)
        association_windows_list = pd.read_csv(customized_association_windows, comment=&quot;#&quot;, header=None, names=[&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;,&quot;ID&quot;], sep=&quot;\t&quot;)
        meta_data = pd.merge(meta_data, association_windows_list, on=[&#39;#chr&#39;, &#39;ID&#39;], how=&#39;left&#39;, suffixes=(&#39;&#39;, &#39;_association&#39;))
        mismatches = meta_data[meta_data[&#39;start_association&#39;].isna()]
        if not mismatches.empty:
            print(&quot;First 5 mismatches:&quot;)
            print(mismatches[[&#39;ID&#39;]].head())
            raise ValueError(f&quot;{len(mismatches)} regions to analyze cannot be found in ``{customized_association_windows}``. Please check your ``{customized_association_windows}`` database to make sure it contains all association-window definitions. &quot;)
    else:
        if cis_window &lt; 0 :
            raise ValueError(&quot;Please either input valid path to association-window file via ``--customized-association-windows``, or set ``--cis-window`` to a non-negative integer.&quot;)
        if cis_window == 0:
            print(&quot;Warning: only variants within the range of start and end of molecular phenotype will be considered since cis_window is set to zero and no customized association window file was found. Please make sure this is by design.&quot;)
        meta_data[&#39;start_association&#39;] = meta_data[&#39;start&#39;].apply(lambda x: max(x - cis_window, 0))
        meta_data[&#39;end_association&#39;] = meta_data[&#39;end&#39;] + cis_window

    # Example meta_data:
    # #chr    start      end    start_association       end_association           ID  Original_ID   path     cov_path             cond             coordinate     geno_path
    # 0  chr12   752578   752579  652578   852579  ENSG00000060237  Q9H4A3,P62873  protocol_example.protein_1.bed.gz,protocol_example.protein_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B    chr12:752578-752579  protocol_example.genotype.chr21_22.bed       
    # Create the final dictionary
    regional_data = {
        &#39;data&#39;: [(row[&#39;geno_path&#39;], *row[&#39;path&#39;].split(&#39;,&#39;), *row[&#39;cov_path&#39;].split(&#39;,&#39;)) for _, row in meta_data.iterrows()],
        &#39;meta_info&#39;: [(f&quot;{row[&#39;#chr&#39;]}:{row[&#39;start&#39;]}-{row[&#39;end&#39;]}&quot;, # this is the phenotypic region to extract data from
                       f&quot;{row[&#39;#chr&#39;]}:{row[&#39;start_association&#39;]}-{row[&#39;end_association&#39;]}&quot;, # this is the association window region
                       row[&#39;ID&#39;], row[&#39;Original_ID&#39;], *row[&#39;cond&#39;].split(&#39;,&#39;)) for _, row in meta_data.iterrows()]
    }
else:
    regional_data = {&#39;data&#39;:[], &#39;meta_info&#39;:[]}
</pre></div>
</div>
</div>
</div>
</section>
<section id="univariate-analysis-susie-and-twas">
<h2>Univariate analysis: SuSiE and TWAS<a class="headerlink" href="#univariate-analysis-susie-and-twas" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_twas]
# Further limit TWAS to only using selected variants
parameter: min_twas_maf = 0.01
parameter: min_twas_xvar = 0.01
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)
meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;

if skip_fine_mapping and skip_twas_weights:
    save_data = True
    output_files = [f&#39;{cwd:a}/data/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.univariate_data.rds&#39;]
elif not skip_fine_mapping and skip_twas_weights:
    output_files = [f&#39;{cwd:a}/fine_mapping/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.univariate_bvsr.rds&#39;]
elif skip_fine_mapping and not skip_twas_weights:
    output_files = [f&#39;{cwd:a}/twas_weights/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.univariate_twas_weights.rds&#39;]
else:
    output_files = [f&#39;{cwd:a}/fine_mapping/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.univariate_bvsr.rds&#39;,
    f&#39;{cwd:a}/twas_weights/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.univariate_twas_weights.rds&#39;]
output: output_files
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bnn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.susie_twas.stdout&quot;, stderr = f&quot;{_output[0]:nn}.susie_twas.stderr&quot;, container = container, entrypoint = entrypoint
    options(warn=1)
    library(pecotmr)
    phenotype_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])})
    covariate_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])})
    conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4:]])})
    pip_cutoff_to_skip = list(${&quot;,&quot;.join(skip_analysis_pip_cutoff)})
    pip_cutoff_to_skip = unlist(c(pip_cutoff_to_skip[conditions]))
    # extract subset of samples
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }
    # Load regional association data
    tryCatch({
        fdat = load_regional_univariate_data(genotype = ${_input[0]:anr},
                                              phenotype = phenotype_files,
                                              covariate = covariate_files,
                                              region = ${(&quot;&#39;%s&#39;&quot; % _meta_info[0]) if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &#39;NULL&#39;}, # if the end position is zero return NULL
                                              association_window = &quot;${_meta_info[1]}&quot;,
                                              conditions = conditions,
                                              maf_cutoff = ${maf},
                                              mac_cutoff = ${mac},
                                              imiss_cutoff = ${imiss},
                                              keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                              keep_samples = keep_samples,
                                              keep_variants = ${&#39;&quot;%s&quot;&#39; % keep_variants if not keep_variants.is_dir() else &quot;NULL&quot;}, # use &quot;not keep_variants.is_dir()&quot; in case user input filename is wrong without realizing
                                              extract_region_name = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}),
                                              phenotype_header = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                              region_name_col = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                              scale_residuals = FALSE)
    }, NoSNPsError = function(e) {
        message(&quot;Error: &quot;, paste(e$message, &quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot;))
        data_dir &lt;- paste0(&quot;${cwd:a}&quot;, &quot;/data/&quot;)
        if (!dir.exists(data_dir)) {
            dir.create(data_dir, recursive = TRUE)
        }
        saveRDS(list(${_meta_info[2]} = e$message),
                paste0(data_dir, &quot;${_output[0]:bnn}.univariate_data.rds&quot;), 
                compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    })
    # save data-set
    if (${&quot;TRUE&quot; if save_data else &quot;FALSE&quot;}) {
        saveRDS(list(${_meta_info[2]} = fdat), 
                    paste0(&quot;${cwd:a}&quot;, &quot;/data/&quot;, &quot;${_output[0]:bnn}.univariate_data.rds&quot;), 
                    compress=&#39;xz&#39;)
    }
    if (${&quot;FALSE&quot; if skip_fine_mapping and skip_twas_weights else &quot;TRUE&quot;}) {
      if (&quot;${_meta_info[2]}&quot; != &quot;${_meta_info[3]}&quot;) {
          region_name = c(&quot;${_meta_info[2]}&quot;, c(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}))
      } else {
          region_name = &quot;${_meta_info[2]}&quot;
      }

      region_info = list(region_coord=parse_region(&quot;${_meta_info[0]}&quot;), grange=parse_region(&quot;${_meta_info[1]}&quot;), region_name=region_name)

      finemapping_result = list()
      preset_variants_result = list()
      condition_names = vector()
      empty_elements_cnt = 0
      r = 1

      while (r&lt;=length(fdat$residual_Y)) {
          # Update condition names
          dropped_samples = list(X=fdat$dropped_sample$dropped_samples_X[[r]], 
                               y=fdat$dropped_sample$dropped_samples_Y[[r]], 
                               covar=fdat$dropped_sample$dropped_samples_covar[[r]])
          new_names = names(fdat$residual_Y)[r]
          ##FIXME:  residule Y lost their colname when there was only 1 column
          # new_col_names = colnames(fdat$residual_Y[[r]])
          new_col_names = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])})[[r]]
          if (is.null(new_col_names)) {
            # column names does not exist, create generic names instead
            new_col_names = 1:ncol(fdat$residual_Y[[r]])
          }
          # if the name is different as region name ie, these original ID the same as gene ID, then give the new name
          if(!(identical(new_names, new_col_names)))
          new_names = paste(new_names, new_col_names, sep=&quot;_&quot;) # DLPFC_iso1 DLPFC_iso2

          out &lt;- list()
          # Run the first round focused on fine-mapping with many variants included
          # fine-mapping with complete data-set
          if (${&quot;FALSE&quot; if skip_fine_mapping else &quot;TRUE&quot;}) {
              out$finemapping &lt;- lapply(1:ncol(fdat$residual_Y[[r]]), function(i) {
                                  set.seed(${seed})
                                  rs &lt;- univariate_analysis_pipeline( X=fdat$residual_X[[r]], 
                                                                      Y=fdat$residual_Y[[r]][,i,drop=FALSE],
                                                                      maf=fdat$maf[[r]], 
                                                                      X_scalar=fdat$residual_X_scalar[[r]], 
                                                                      Y_scalar=if (fdat$residual_Y_scalar[[r]] == 1) 1 else fdat$residual_Y_scalar[[r]][,i,drop=FALSE],
                                                                      X_variance=fdat$X_variance[[r]],
                                                                      other_quantities = list(dropped_samples = dropped_samples),
                                                                      # filters
                                                                      imiss_cutoff = ${imiss},
                                                                      maf_cutoff = NULL,
                                                                      xvar_cutoff = 0,
                                                                      ld_reference_meta_file = NULL,
                                                                      pip_cutoff_to_skip = pip_cutoff_to_skip[r],
                                                                      # methods parameter configuration
                                                                      init_L = ${init_L},
                                                                      max_L = ${max_L},
                                                                      l_step = 5,
                                                                      # fine-mapping results summary
                                                                      signal_cutoff = ${pip_cutoff},
                                                                      coverage = c(${&quot;,&quot;.join([str(x) for x in coverage])}),
                                                                      # TWAS weights and CV for TWAS weights
                                                                      twas_weights = FALSE, 
                                                                      max_cv_variants=${max_cv_variants},
                                                                      cv_folds=${twas_cv_folds},
                                                                      cv_threads = ${twas_cv_threads}
                                                                  )
                                                                  return(rs)
                                                                  })
          }
          if (${&quot;FALSE&quot; if skip_twas_weights else &quot;TRUE&quot;}) {
              common_cols = intersect(colnames(fdat$X), colnames(fdat$residual_X[[r]]))
              X_r = fdat$X[rownames(fdat$residual_X[[r]]), common_cols, drop=F]
              maf = fdat$maf[[r]][common_cols]
              # Run the second round focused on TWAS using selected variants
              out$twas_models &lt;- lapply(1:ncol(fdat$residual_Y[[r]]), function(i) {
                                  set.seed(${seed})
                                  rs &lt;- univariate_analysis_pipeline(X=X_r, # use original X matrix for TWAS model: 1) to be comparable to multivariate analysis in CV; 2) to avoid overfitting in CV
                                                                  Y=fdat$residual_Y[[r]][,i,drop=FALSE],
                                                                  maf=maf, 
                                                                  X_scalar=fdat$residual_X_scalar[[r]], 
                                                                  Y_scalar=if (fdat$residual_Y_scalar[[r]] == 1) 1 else fdat$residual_Y_scalar[[r]][,i,drop=FALSE],
                                                                  X_variance=fdat$X_variance[[r]],
                                                                  other_quantities = list(dropped_samples = dropped_samples),
                                                                  # filters
                                                                  imiss_cutoff = ${imiss},
                                                                  maf_cutoff = ${min_twas_maf},
                                                                  xvar_cutoff = ${min_twas_xvar},
                                                                  ld_reference_meta_file=${(&#39;&quot;%s&quot;&#39; % ld_reference_meta_file) if not ld_reference_meta_file.is_dir() else &quot;NULL&quot;},
                                                                  pip_cutoff_to_skip = pip_cutoff_to_skip[r],
                                                                  # methods parameter configuration
                                                                  init_L = ${init_L},
                                                                  max_L = ${max_L},
                                                                  l_step = 5,
                                                                  # fine-mapping results summary
                                                                  signal_cutoff = ${pip_cutoff},
                                                                  coverage = c(${&quot;,&quot;.join([str(x) for x in coverage])}),
                                                                  # TWAS weights and CV for TWAS weights
                                                                  twas_weights = TRUE, 
                                                                  max_cv_variants=${max_cv_variants},
                                                                  cv_folds=${twas_cv_folds},
                                                                  cv_threads = ${twas_cv_threads}
                                                                  )
                                                                  return(rs)
                                                                  })
          }

          empty_elements_idx &lt;- unique(do.call(c, lapply(out, function(results) which(sapply(results, function(x) is.list(x) &amp;&amp; length(x) == 0)))))
          if (length(empty_elements_idx)&gt;0) {
            empty_elements_cnt &lt;- empty_elements_cnt + length(empty_elements_idx)
            if (!is.null(out$finemapping)) {
                out$finemapping &lt;- out$finemapping[-empty_elements_idx]
            }
            if (!is.null(out$twas_models)) {
                out$twas_models &lt;- out$twas_models[-empty_elements_idx]
            }
            new_names &lt;- new_names[-empty_elements_idx]
          }
          if (!is.null(out$finemapping)) {
              finemapping_result = c(finemapping_result, out$finemapping)
          }
          if (!is.null(out$twas_models)) {
              preset_variants_result = c(preset_variants_result, out$twas_models)
          }
          condition_names = c(condition_names, new_names)
          if (length(new_names)&gt;0) {
            message(&quot;Analysis completed for: &quot;, paste(new_names, collapse=&quot;,&quot;))
          }
          # original data no longer relevant, set to NA to release memory
          fdat$residual_X[[r]] &lt;- NA
          fdat$residual_Y[[r]] &lt;- NA
          r = r + 1
      }
      # Reorganize outputs
      # The BVSR model in this case contain different versions of SuSiE fits
      # preset_variants_result
      # Clean up TWAS weights output
      twas_output &lt;- list()
      if (length(preset_variants_result)&gt;0) {
          names(preset_variants_result) &lt;- condition_names
          for (r in condition_names) {
            twas_output[[r]] &lt;- preset_variants_result[[r]]$twas_weights_result
            preset_variants_result[[r]]$twas_weights_result &lt;- NULL
            twas_output[[r]]$variant_names &lt;- preset_variants_result[[r]]$variant_names
            twas_output[[r]]$region_info &lt;- region_info
            preset_variants_result[[r]]$region_info &lt;- region_info
          }
      }
      # Clean up fine-mapping output
      finemapping_output &lt;- list()
      if (length(finemapping_result)&gt;0) {
          names(finemapping_result) &lt;- condition_names
      }
      for (r in condition_names) {
          if (r %in% names(finemapping_result)) {
              finemapping_output[[r]] &lt;- finemapping_result[[r]]
              finemapping_output[[r]]$region_info &lt;- region_info
              finemapping_output[[r]]$susie_fitted &lt;- NULL
          }
          if (r %in% names(preset_variants_result)) {
              finemapping_output[[r]]$preset_variants_result &lt;- preset_variants_result[[r]]
          }
      } 
      if (empty_elements_cnt&gt;0) {
          msg = paste0(empty_elements_cnt, &quot; analysis are skipped for failing to pass initial screen for potential association signals&quot;, sep=&quot;&quot;)
          message(msg)
          if(length(finemapping_output)==0){
            finemapping_output = msg
          }
          if(length(twas_output)==0){
            twas_output = msg
          }
      }
    
      if (${&quot;FALSE&quot; if skip_fine_mapping else &quot;TRUE&quot;}){
        saveRDS(list(&quot;${_meta_info[2]}&quot; = finemapping_output), 
              paste0(&quot;${cwd:a}&quot;, &quot;/fine_mapping/&quot;, &quot;${_output[0]:bnn}.univariate_bvsr.rds&quot;),
              compress=&#39;xz&#39;)
      }
      if (${&quot;FALSE&quot; if skip_twas_weights else &quot;TRUE&quot;}) {
         saveRDS(list(&quot;${_meta_info[2]}&quot; = twas_output), 
                                           paste0(&quot;${cwd:a}&quot;, &quot;/twas_weights/&quot;, &quot;${_output[0]:bnn}.univariate_twas_weights.rds&quot;), 
                                           compress=&#39;xz&#39;)
      }
    }
</pre></div>
</div>
</div>
</div>
</section>
<section id="multivariate-analysis-mvsusie-and-mr-mash">
<h2>Multivariate analysis: mvSuSiE and mr.mash<a class="headerlink" href="#multivariate-analysis-mvsusie-and-mr-mash" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mnm]
# Prior model file generated from mashr. 
# Default will be used if it does not exist.
parameter: mixture_prior = path()
parameter: mixture_prior_cv = path()
parameter: prior_weights_min = 5e-4
parameter: prior_canonical_matrices = False
parameter: sample_partition = path() 
parameter: mvsusie_max_iter = 200
parameter: mrmash_max_iter = 5000
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)

meta_info = regional_data[&#39;meta_info&#39;]

input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
if skip_fine_mapping and skip_twas_weights:
    save_data = True
    output_files = [f&#39;{cwd:a}/data/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multicontext_data.rds&#39;]
elif not skip_fine_mapping and skip_twas_weights:
    output_files = [f&#39;{cwd:a}/multivariate_fine_mapping/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multicontext_bvsr.rds&#39;]
elif skip_fine_mapping and not skip_twas_weights:
    # Warning that fine-mapping is required for TWAS weights
    if not ld_reference_meta_file.is_dir():
        print(&quot;WARNING: Fine-mapping is required to generate TWAS weights. Setting skip_fine_mapping to False. &quot;
              &quot;Fine-mapping will be performed and saved on variants listed in %s.&quot; % ld_reference_meta_file)
    else:
        print(&quot;WARNING: Fine-mapping is required to generate TWAS weights. Setting skip_fine_mapping to False.&quot;)
    skip_fine_mapping = False
    output_files = [f&#39;{cwd:a}/multivariate_fine_mapping/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multicontext_bvsr.rds&#39;,
                    f&#39;{cwd:a}/multivariate_twas_weights/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multicontext_twas_weights.rds&#39;]
else:    
    output_files = [f&#39;{cwd:a}/multivariate_fine_mapping/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multicontext_bvsr.rds&#39;,
                    f&#39;{cwd:a}/multivariate_twas_weights/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multicontext_twas_weights.rds&#39;]   
output: output_files
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.mnm.stdout&quot;, stderr = f&quot;{_output[0]:nn}.mnm.stderr&quot;, container = container, entrypoint = entrypoint

    library(pecotmr)
    
    phenotype_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])})
    covariate_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])})
    conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4:]])})
    pip_cutoff_to_skip = list(${&quot;,&quot;.join(skip_analysis_pip_cutoff)})
    pip_cutoff_to_skip = unlist(pip_cutoff_to_skip[conditions])
    # extract subset of samples
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }
    
    # Load regional association data
    tryCatch({
    fdat = load_regional_multivariate_data(genotype = ${_input[0]:anr},
                                          phenotype = phenotype_files,
                                          covariate = covariate_files,
                                          region = &quot;${_meta_info[0]}&quot;,
                                          conditions = conditions,
                                          association_window = &quot;${_meta_info[1]}&quot;,
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                          keep_samples = keep_samples,
                                          keep_variants = ${&#39;&quot;%s&quot;&#39; % keep_variants if not keep_variants.is_dir() else &quot;NULL&quot;},
                                          extract_region_name = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}),
                                          phenotype_header = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                          region_name_col = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                          scale_residuals = FALSE)
        }, NoSNPsError = function(e) {
        message(&quot;Error: &quot;, paste(e$message, &quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot;))
        data_dir &lt;- paste0(&quot;${cwd:a}&quot;, &quot;/data/&quot;)
        if (!dir.exists(data_dir)) {
            dir.create(data_dir, recursive = TRUE)
        }
        saveRDS(list(${_meta_info[2]} = e$message),
                paste0(data_dir, &quot;${_output[0]:bnn}.multicontext_data.rds&quot;), 
                compress=&#39;xz&#39;)            
        quit(save=&quot;no&quot;)
    })
    if (${&quot;TRUE&quot; if save_data else &quot;FALSE&quot;}) {
        saveRDS(fdat, paste0(&quot;${cwd:a}&quot;, &quot;/data/&quot;, &quot;${_output[0]:bnn}.multicontext_data.rds&quot;), compress=&#39;xz&#39;)
    }
    if (${&quot;FALSE&quot; if skip_fine_mapping and skip_twas_weights else &quot;TRUE&quot;}) {
      if (&quot;${_meta_info[2]}&quot; != &quot;${_meta_info[3]}&quot;) {
          region_name = c(&quot;${_meta_info[2]}&quot;, c(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}))
      } else {
          region_name = &quot;${_meta_info[2]}&quot;
      }

      context_regions &lt;- list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])})
      names(context_regions) &lt;- colnames(fdat$residual_Y)
      region_info = list(region_coord=parse_region(&quot;${_meta_info[0]}&quot;), grange=parse_region(&quot;${_meta_info[1]}&quot;), region_name=region_name)

      dd_prior &lt;- if (${mixture_prior:r} == &#39;.&#39; || ${mixture_prior:r} == &#39;&#39;) NULL else readRDS(${mixture_prior:r})
      dd_prior_cv &lt;- if (${mixture_prior_cv:r} == &#39;.&#39; || ${mixture_prior_cv:r} == &#39;&#39;) NULL else readRDS(${mixture_prior_cv:r})
      use_canonical_prior = ${&quot;TRUE&quot; if prior_canonical_matrices else &quot;FALSE&quot;} &amp;&amp; ${&quot;FALSE&quot; if not mixture_prior.is_file() else &quot;TRUE&quot;}
      if (is.null(dd_prior)) use_canonical_prior = TRUE
      set.seed(${seed})
      result &lt;- multivariate_analysis_pipeline(
                      X = fdat$X,
                      Y = fdat$residual_Y,
                      maf = fdat$maf,
                      X_variance = fdat$X_variance,
                      other_quantities = list(dropped_samples = fdat$dropped_samples),
                      # filters
                      imiss_cutoff = ${imiss}, 
                      maf_cutoff = ${maf},
                      xvar_cutoff = ${xvar_cutoff},
                      ld_reference_meta_file = ${(&#39;&quot;%s&quot;&#39; % ld_reference_meta_file) if not ld_reference_meta_file.is_dir() else &quot;NULL&quot;},
                      pip_cutoff_to_skip = pip_cutoff_to_skip,
                      # methods parameter configuration
                      max_L = -1,
                      data_driven_prior_matrices = dd_prior, 
                      data_driven_prior_matrices_cv = dd_prior_cv, 
                      data_driven_prior_weights_cutoff = ${prior_weights_min}, 
                      canonical_prior_matrices = use_canonical_prior,
                      mvsusie_max_iter = ${mvsusie_max_iter}, 
                      mrmash_max_iter = ${mrmash_max_iter},
                      # fine-mapping results summary
                      signal_cutoff = ${pip_cutoff},
                      coverage = c(${&quot;,&quot;.join([str(x) for x in coverage])}),
                      # TWAS weights and CV for TWAS weights
                      twas_weights = ${&quot;TRUE&quot; if not skip_twas_weights else &quot;FALSE&quot;},
                      sample_partition = ${&quot;NULL&quot; if sample_partition==&#39;.&#39; or sample_partition==&#39;&#39; else sample_partition},
                      max_cv_variants = ${max_cv_variants}, 
                      cv_folds = ${twas_cv_folds},
                      cv_threads = ${twas_cv_threads}            
      )
      result$region_info &lt;- region_info
      if (!is.null(result$twas_weights_result)) {
          for (context in names(result$twas_weights_result)) {
              result$twas_weights_result[[context]]$region_info &lt;- region_info
              context_region_name = context_regions[[context]]
              if (is.null(context_region_name)) context_region_name = context 
              if(!(identical(context, context_region_name))) new_context_name = paste(context, context_region_name, sep=&quot;_&quot;) 
              names(result$twas_weights_result) [match(context, names(result$twas_weights_result))] &lt;- new_context_name
          }
          saveRDS(list(&quot;${_meta_info[2]}&quot; = result$twas_weights_result), paste0(&quot;${cwd:a}&quot;, &quot;/multivariate_twas_weights/&quot;, &quot;${_output[0]:bnn}.multicontext_twas_weights.rds&quot;), compress=&#39;xz&#39;)
          result$twas_weights_result &lt;- NULL
      }
      saveRDS(list(&quot;${_meta_info[2]}&quot; = result), paste0(&quot;${cwd:a}&quot;, &quot;/multivariate_fine_mapping/&quot;, &quot;${_output[0]:bnn}.multicontext_bvsr.rds&quot;), compress=&#39;xz&#39;)
    }
</pre></div>
</div>
</div>
</div>
</section>
<section id="xqtl-fine-mapping-across-multiple-genes">
<h2>xQTL fine-mapping across multiple genes<a class="headerlink" href="#xqtl-fine-mapping-across-multiple-genes" title="Link to this heading">#</a></h2>
<p>This pipeline jointly fine-map multiple genes in a given region. To determine if a gene is eligable for joint fine-mapping with other genes, it will first load the univeriate fine-mapping result of a gene and assess if its top_loci candidates overlap with any of the other genes. Genes don’t statisfy this requirement will be dropped. It is therefore required that the univariate fine-mapping results and its meta data are available, in the format of <code class="docutils literal notranslate"><span class="pre">susie_twas</span></code> output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mnm_genes]
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)
 
meta_info = regional_data[&#39;meta_info&#39;]
genes = meta_info[0][3]
if isinstance(genes, tuple):
    genes = genes[0] if isinstance(genes[0], tuple) else list(genes)
else:
    genes = [genes]

if len(skip_analysis_pip_cutoff) == 0:
    skip_analysis_pip_cutoff = [0.0] * len(genes)

if len(skip_analysis_pip_cutoff) == 1:
    skip_analysis_pip_cutoff = skip_analysis_pip_cutoff * len(genes)

skip_analysis_pip_cutoff = [&quot;&#39;{}&#39;={}&quot;.format(genes[i], skip_analysis_pip_cutoff[i].split(&#39;=&#39;)[1]) for i in range(len(genes))]

parameter: pheno_id_map_file = path
parameter: fine_mapping_meta = path
parameter: data_driven_prior = False
# Parameters below are relevant when we use data driven prior
# which is an experimental feature
parameter: n_random = 10
parameter: n_null = 10
parameter: independent_variant_list = path()
parameter: prior_weights_min = 5e-4
parameter: prior_canonical_matrices = False
# This is relevant to cross-validation
parameter: sample_partition = path() 
parameter: mvsusie_max_iter = 200
parameter: mrmash_max_iter = 5000

if not data_driven_prior:
     prior_canonical_matrices = True
input: regional_data[&quot;data&quot;],group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;

if skip_fine_mapping and skip_twas_weights:
    save_data = True
    output_files = [f&#39;{cwd:a}/data/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multigene_data.rds&#39;]
elif not skip_fine_mapping and skip_twas_weights:
    output_files = [f&#39;{cwd:a}/multivariate_fine_mapping/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multigene_bvsr.rds&#39;]
elif skip_fine_mapping and not skip_twas_weights:
    output_files = [f&#39;{cwd:a}/multivariate_twas_weights/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multigene_twas_weights.rds&#39;]
else:    
    output_files = [f&#39;{cwd:a}/multivariate_fine_mapping/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multigene_bvsr.rds&#39;,
    f&#39;{cwd:a}/multivariate_twas_weights/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.multigene_twas_weights.rds&#39;]    
output: output_files
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.mnm_genes.stdout&quot;, stderr = f&quot;{_output[0]:nn}.mnm_genes.stderr&quot;, container = container, entrypoint = entrypoint

    library(data.table)
    library(dplyr)
    library(pecotmr)
    combine_result_list &lt;- function(univariate_finemapping_metadata, condition_name, conditions = NULL, extracted_region){
        regional_window_combined_susie_result &lt;- list()
        regional_window_combined_sumstats_result &lt;- list()
        extracted_univariate_finemapping_metadata &lt;- univariate_finemapping_metadata%&gt;%filter(region_id%in%extracted_region)
        for(i in 1:dim(extracted_univariate_finemapping_metadata)[1]){
           gene_id &lt;- extracted_univariate_finemapping_metadata$region_id[i]
           susie_result &lt;- readRDS(extracted_univariate_finemapping_metadata$combined_data[i])
           sumstats_result &lt;- readRDS(extracted_univariate_finemapping_metadata$combined_data_sumstats[i]) 
           if (!is.null(conditions)) {
              if (any(names(susie_result[[gene_id]]) %in% conditions)) {
                  extracted_conditions = names(susie_result[[gene_id]])[names(susie_result[[gene_id]]) %in% conditions]
                 for (j in 1:length(extracted_conditions)){
                     regional_window_combined_susie_result[[condition_name]][[extracted_conditions[j]]] &lt;- susie_result[[gene_id]][[extracted_conditions[j]]]
                     regional_window_combined_sumstats_result[[condition_name]][[extracted_conditions[j]]] &lt;- sumstats_result[[gene_id]][[extracted_conditions[j]]]
                 }
              }
           } else {
                    regional_window_combined_susie_result[[condition_name]][[gene_id]] &lt;- susie_result[[gene_id]][[condition_name]]
                    regional_window_combined_sumstats_result[[condition_name]][[gene_id]] &lt;- sumstats_result[[gene_id]][[condition_name]]
                  }
          }       
           return(list(regional_window_combined_susie_result = regional_window_combined_susie_result, regional_window_combined_sumstats_result = regional_window_combined_sumstats_result))
     }

  
    extract_non_null_cs_list &lt;- function(combined_list, condition_name, extracted_region) {
        # List to store results for genes passing the first step
        extracted_regional_window_combined_susie_result &lt;- list()
        extracted_regional_window_combined_sumstats_result &lt;- list()

        # Iterate over each gene&#39;s susie and sumstats results
        filtered_region &lt;- extracted_region[extracted_region%in%names(combined_list[[&quot;regional_window_combined_susie_result&quot;]][[condition_name]])]
        for (gene_id in filtered_region) {
            susie_result &lt;- get_nested_element(combined_list,c(&quot;regional_window_combined_susie_result&quot;,condition_name,gene_id))

           # Check if there exits top_loci.
           if (!is.null(susie_result[[&quot;top_loci&quot;]])&amp;&amp;nrow(susie_result[[&quot;top_loci&quot;]])[1]!=0) {
              # If not all zeros, add this gene&#39;s susie and sumstats results to the lists
              extracted_regional_window_combined_susie_result[[condition_name]][[gene_id]] &lt;- get_nested_element(combined_list,c(&quot;regional_window_combined_susie_result&quot;,condition_name,gene_id))
              extracted_regional_window_combined_sumstats_result[[condition_name]][[gene_id]] &lt;- get_nested_element(combined_list,c(&quot;regional_window_combined_sumstats_result&quot;,condition_name,gene_id))
           }
         }
         return(list(extracted_regional_window_combined_susie_result = extracted_regional_window_combined_susie_result, extracted_regional_window_combined_sumstats_result = extracted_regional_window_combined_sumstats_result))
     }
  
    process_and_compare_variants &lt;- function(combined_list, condition_name) {
       extracted_regional_window_combined_susie_result &lt;- list()
       extracted_regional_window_combined_sumstats_result &lt;- list()
       compare_variants &lt;- function(variants1, variants2) {
            return(any(variants1 %in% variants2) | any(variants2 %in% variants1))
       }
       subset_top_loci_variants &lt;- function(df) {
          find_non_zero_rows &lt;- function(df) {
                non_zero_indices &lt;- which(rowSums(df!= 0) &gt; 0)
                return(non_zero_indices)
           }
           non_zero_indices &lt;- find_non_zero_rows(df)

           if (length(non_zero_indices) &gt; 0) {
              non_zero_index_variants &lt;- df[non_zero_indices, , drop = FALSE]%&gt;%select(variant_id)%&gt;%pull(variant_id)
              zero_index_filtered_variants &lt;- df[-non_zero_indices, ,drop = FALSE]%&gt;%filter(pip&gt;=0.01)%&gt;%select(variant_id)%&gt;%pull(variant_id)
              top_loci_variants &lt;- c(non_zero_index_variants, zero_index_filtered_variants)
           } else {
              zero_index_filtered_variants &lt;- df%&gt;%filter(pip&gt;=0.01)%&gt;%select(variant_id)%&gt;%pull(variant_id)
              top_loci_variants &lt;- zero_index_filtered_variants
           }
             return(top_loci_variants)
       }
  
       for (i in 1:length(combined_list[[&quot;extracted_regional_window_combined_susie_result&quot;]][[condition_name]])) {
           susie_result_1 &lt;- combined_list[[&quot;extracted_regional_window_combined_susie_result&quot;]][[condition_name]][[i]]
           gene_id &lt;- names(combined_list[[&quot;extracted_regional_window_combined_susie_result&quot;]][[condition_name]])[i]
    
           # Extract top_loci variants
           variants1 &lt;- subset_top_loci_variants(susie_result_1[[&quot;top_loci&quot;]])
           if (length(variants1) &gt; 0) {
              different_in_all_genes &lt;- TRUE
    
           for (j in 1:length(combined_list[[&quot;extracted_regional_window_combined_susie_result&quot;]][[condition_name]])) {
               if (i != j) {
                  susie_result_2 &lt;- combined_list[[&quot;extracted_regional_window_combined_susie_result&quot;]][[condition_name]][[j]]
                  variants2 &lt;- subset_top_loci_variants(susie_result_2[[&quot;top_loci&quot;]])
        
                  # Compare variants
                  if (length(variants2) &gt; 0) {
                    if (compare_variants(variants1, variants2)) {
                        # If variants are not totally different across all genes, set the flag to FALSE
                       different_in_all_genes &lt;- FALSE
                       break
                     }
                   }
                 }
               }
    
            # After the loop, check the flag
            if (!different_in_all_genes) {
               extracted_regional_window_combined_susie_result[[condition_name]][[gene_id]] &lt;- get_nested_element(combined_list, c(&quot;extracted_regional_window_combined_susie_result&quot;, condition_name, gene_id))
               extracted_regional_window_combined_sumstats_result[[condition_name]][[gene_id]] &lt;- get_nested_element(combined_list, c(&quot;extracted_regional_window_combined_sumstats_result&quot;, condition_name, gene_id))
            }
          }
        }
      return(list(extracted_regional_window_combined_susie_result = extracted_regional_window_combined_susie_result, extracted_regional_window_combined_sumstats_result = extracted_regional_window_combined_sumstats_result))
    }

    fine_mapping_metadata = fread(${fine_mapping_meta:r})
    phenotype_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])})
    covariate_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])})
    condition_name = ${(&quot;&#39;%s&#39;&quot; % _meta_info[-1])}
    region_names = ${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}
    if (any(region_names %in% fine_mapping_metadata$region_id)) {
       filtered_region &lt;- fine_mapping_metadata%&gt;%filter(region_id%in%region_names)%&gt;%select(region_id)%&gt;%pull(region_id)
       filtered_event_names &lt;- filtered_region
    } else {
       pheno_id_map &lt;- fread(${pheno_id_map_file:r})
       gene_ids &lt;- pheno_id_map%&gt;%filter(ID%in%region_names)%&gt;%select(gene)%&gt;%pull(gene)%&gt;%{.[!duplicated(.)]}
       filtered_region &lt;- fine_mapping_metadata%&gt;%filter(region_id%in%gene_ids)%&gt;%select(region_id)%&gt;%pull(region_id)
       filtered_event_names &lt;- pheno_id_map%&gt;%filter(ID%in%region_names&amp;gene%in%filtered_region)%&gt;%select(ID)%&gt;%pull(ID)
     }
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }
                                     
    fdat = load_regional_multivariate_data(genotype = ${_input[0]:anr},
                                          phenotype = phenotype_files,
                                          covariate = covariate_files,
                                          region = ${(&quot;&#39;%s&#39;&quot; % _meta_info[1]) if int(_meta_info[1].split(&#39;-&#39;)[-1])&gt;0 else &#39;NULL&#39;},
                                          maf_cutoff = ${maf},       
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          conditions = NULL,
                                          association_window = ${(&quot;&#39;%s&#39;&quot; % _meta_info[1]) if int(_meta_info[1].split(&#39;-&#39;)[-1])&gt;0 else &#39;NULL&#39;},
                                          extract_region_name = list(filtered_event_names),
                                          region_name_col = ${&quot;4&quot; if int(_meta_info[1].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                          keep_samples = keep_samples,
                                          keep_variants = ${&#39;&quot;%s&quot;&#39; % keep_variants if not keep_variants.is_dir() else &quot;NULL&quot;},
                                          phenotype_header = ${&quot;4&quot; if int(_meta_info[1].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;}, # skip first 4 rows of transposed phenotype for chr, start, end and ID
                                          scale_residuals = FALSE)
    if (${&quot;TRUE&quot; if save_data else &quot;FALSE&quot;}) {
        saveRDS(fdat, &quot;${_output[0]:ann}.multigene_data.rds&quot;, compress=&#39;xz&#39;)
    }
    if (${&quot;FALSE&quot; if skip_fine_mapping and skip_twas_weights else &quot;TRUE&quot;}) {
        # Extract and consolidate univariate results to determine which genes to focus on
        if (grepl(&quot;sQTL&quot;, condition_name)) {conditions &lt;- paste(condition_name, filtered_event_names, sep = &quot;_&quot;)} else {conditions &lt;- NULL}
        combined_regional_window_list &lt;- combine_result_list(fine_mapping_metadata, condition_name, conditions, filtered_region)
        if (grepl(&quot;sQTL&quot;,condition_name)) { filtered_region &lt;- paste(condition_name, filtered_event_names, sep = &quot;_&quot;) }
        extracted_non_null_combined_susie_list &lt;- extract_non_null_cs_list(combined_list = combined_regional_window_list, condition_name, filtered_region)
        if (length(extracted_non_null_combined_susie_list[[&quot;extracted_regional_window_combined_susie_result&quot;]])!=0) {
           extracted_combined_susie_list &lt;- process_and_compare_variants(combined_list = extracted_non_null_combined_susie_list, condition_name)
        } else {
           extracted_combined_susie_list &lt;- NULL
        }
        if (!is.null(extracted_combined_susie_list)&amp;length(extracted_combined_susie_list[[&quot;extracted_regional_window_combined_susie_result&quot;]])!=0) {
           if (grepl(&quot;sQTL&quot;, condition_name)) {
              filtered_event_names &lt;- sub(paste0(&quot;^&quot;, condition_name, &quot;_&quot;), &quot;&quot;, names(get_nested_element(extracted_combined_susie_list,c(&quot;extracted_regional_window_combined_susie_result&quot;, condition_name))))
           }  else if (grepl(&quot;pQTL&quot;, condition_name)) {
              filtered_event_names &lt;- pheno_id_map %&gt;%
                              filter(ID%in%region_names)%&gt;%
                              filter(gene %in% names(get_nested_element(extracted_combined_susie_list, c(&quot;extracted_regional_window_combined_susie_result&quot;, condition_name)))) %&gt;%
                              select(ID) %&gt;% pull(ID)
          } else {
            filtered_event_names &lt;- names(get_nested_element(extracted_combined_susie_list, c(&quot;extracted_regional_window_combined_susie_result&quot;, condition_name)))
          }
        } else {
          filtered_event_names &lt;- NULL
        }

        if (&quot;${_meta_info[2]}&quot; != &quot;${_meta_info[3]}&quot;) {
          region_name = c(&quot;${_meta_info[2]}&quot;, c(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}))
        } else {
          region_name = &quot;${_meta_info[2]}&quot;
        }

        region_info = list(region_coord=parse_region(&quot;${_meta_info[0]}&quot;), grange=parse_region(&quot;${_meta_info[1]}&quot;), region_name=region_name)
            
        dd_prior &lt;- NULL
        if (${&quot;TRUE&quot; if data_driven_prior else &quot;FALSE&quot;}) {
            # FIXME: please complete this function call
            dd_prior &lt;- multigene_udr(extracted_combined_susie_list, coverage = paste0(&quot;cs_coverage_&quot;,${coverage[0]},sep=&quot;&quot;), independent_variant_list = ${independent_variant_list:r}, n_random = ${n_random}, n_null = ${n_null}, seed = ${seed})
        }
        if (length(filtered_event_names)&gt;=2) {
           pip_cutoff_to_skip = list(${&quot;,&quot;.join(skip_analysis_pip_cutoff)})
           pip_cutoff_to_skip = unlist(pip_cutoff_to_skip[filtered_event_names])
           dd_prior_cv &lt;- dd_prior
           set.seed(${seed})
           result &lt;- multivariate_analysis_pipeline(
                            X=fdat$X,
                            Y=fdat$residual_Y[,filtered_event_names],
                            maf=fdat$maf,
                            X_variance = fdat$X_variance,
                            other_quantities = list(dropped_samples = fdat$dropped_samples),
                            # filters
                            imiss_cutoff = ${imiss}, 
                            maf_cutoff = ${maf},
                            xvar_cutoff = ${xvar_cutoff},
                            ld_reference_meta_file = ${(&#39;&quot;%s&quot;&#39; % ld_reference_meta_file) if not ld_reference_meta_file.is_dir() else &quot;NULL&quot;},
                            pip_cutoff_to_skip = pip_cutoff_to_skip,
                            # methods parameter configuration
                            max_L = -1,
                            data_driven_prior_matrices = dd_prior, 
                            data_driven_prior_matrices_cv = dd_prior_cv, 
                            data_driven_prior_weights_cutoff = ${prior_weights_min}, 
                            canonical_prior_matrices = ${&quot;TRUE&quot; if prior_canonical_matrices else &quot;FALSE&quot;},
                            mvsusie_max_iter = ${mvsusie_max_iter}, 
                            mrmash_max_iter = ${mrmash_max_iter},
                            # fine-mapping results summary
                            signal_cutoff = ${pip_cutoff},
                            coverage = c(${&quot;,&quot;.join([str(x) for x in coverage])}),
                            # TWAS weights and CV for TWAS weights
                            twas_weights = ${&quot;TRUE&quot; if not skip_twas_weights else &quot;FALSE&quot;},
                            sample_partition = ${&quot;NULL&quot; if sample_partition==&#39;.&#39; or sample_partition==&#39;&#39; else sample_partition},
                            max_cv_variants = ${max_cv_variants}, 
                            cv_folds = ${twas_cv_folds},
                            cv_threads = ${twas_cv_threads}            
                            )
          result$region_info &lt;- region_info
          if (!is.null(result$twas_weights_result)) {
              for (r in names(result$twas_weights_result)) {
                  result$twas_weights_result[[r]]$region_info &lt;- region_info
              }
              saveRDS(list(&quot;${_meta_info[2]}&quot; = result$twas_weights_result), &quot;${_output[0]:ann}.multigene_twas_weights.rds&quot;, compress=&#39;xz&#39;)
              result$twas_weights_result &lt;- NULL
          }
          saveRDS(list(&quot;${_meta_info[2]}&quot; = result), &quot;${_output[0]:ann}.multigene_bvsr.rds&quot;, compress=&#39;xz&#39;)        
        } else {
          for (f in c(${_output:ar,})) {
            saveRDS(NULL, f)
          }
        }
    }
</pre></div>
</div>
</div>
</div>
</section>
<section id="functional-regression-fsusie-for-epigenomic-qtl-fine-mapping">
<h2>Functional regression fSuSiE for epigenomic QTL fine-mapping<a class="headerlink" href="#functional-regression-fsusie-for-epigenomic-qtl-fine-mapping" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fsusie]
# prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: prior = &quot;mixture_normal&quot;
parameter: max_SNP_EM = 100
# Max scale is such that 2^max_scale being the number of phenotypes in the transformed space. Default to 2^10  = 1024. Don&#39;t change it unless you know what you are doing. Max_scale should be at least larger than 5.
parameter:  max_scale = 10
# Purity and coverage used to call cs
parameter:  min_purity = 0.5
# Epigenetics mark filter
parameter: epigenetics_mark_treshold = 16
# Run susie for top pc of the fsusie input
parameter: susie_top_pc = 10
# Run fsusie with this post-processing option. Available options are TI, HMM, and none. TI is recommended as default. HMM performs particularly well when analyzing data with, say, few sampling points (i.e., ncol(Y)&lt;30) or when the data are particularly noisy (low signal-to-noise ratio).
parameter: post_processing =  &quot;TI&quot;
# Run fsusie with small sample correction 
parameter: small_sample_correction = False

depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)
meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0].replace(&quot;:&quot;, &quot;_&quot;).replace(&quot;-&quot;, &quot;_&quot;)}.fsusie_{prior}_{post_processing}_{&quot;_top_pc_weights&quot; if not skip_twas_weights else &quot;&quot;}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    options(warn=1)
    # extract subset of samples
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }

    # Load regional functional data
    library(pecotmr)
    tryCatch({
    fdat = load_regional_functional_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])}),
                                          region = &quot;${_meta_info[0]}&quot;,
                                          association_window = &quot;${_meta_info[1]}&quot;,
                                          conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4:]])}),
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                          keep_samples = keep_samples,
                                          keep_variants = ${&#39;&quot;%s&quot;&#39; % keep_variants if not keep_variants.is_dir() else &quot;NULL&quot;},
                                          tabix_header = TRUE,
                                          phenotype_header = 4,
                                          region_name_col = 4,
                                          scale_residuals = FALSE)
    }, NoSNPsError = function(e) {
        message(&quot;Error: &quot;, paste(e$message, &quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot;))
        saveRDS(list(&quot;${_meta_info[0]}&quot; = e$message), ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    })
    # Filter out list fdat that with less than a treshold of epigenomic marker.
    library(tidyverse)
    filter_fdat_except_specific_names &lt;- function(fdat, n) {
        # Identify which elements in list1 meet the row count criteria
        indices_to_keep &lt;- sapply(fdat$Y_coordinates, function(x) nrow(x) &gt;= n)
        fdat_filtered &lt;- map(fdat[!names(fdat) %in% c(&quot;dropped_sample&quot;, &quot;X&quot;, &quot;chrom&quot;)],~.x[indices_to_keep]) 
        return(c(fdat_filtered,fdat[names(fdat) %in% c(&quot;dropped_sample&quot;, &quot;X&quot;, &quot;chrom&quot;)]))
    }

    fdat = filter_fdat_except_specific_names(fdat, n = ${epigenetics_mark_treshold})
    # Check if Y_coordinates is empty after filtering
    if (length(fdat$Y_coordinates) == 0) {
        e_msg = paste0(&quot;None of the study have more than or equal to &quot;,${epigenetics_mark_treshold}, &quot; epigenetics marks, region skipped&quot;)
        message(e_msg)
        saveRDS(list(&quot;${_meta_info[0]}&quot; = e_msg ),  ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    }

    if (${&quot;TRUE&quot; if save_data else &quot;FALSE&quot;}) {
      # save data object for debug purpose
      saveRDS(list(&quot;${_meta_info[0]}&quot; = fdat), &quot;${_output:ann}.${epigenetics_mark_treshold}_marks.dataset.rds&quot;, compress=&#39;xz&#39;)
    }
  
    fitted = setNames(replicate(length(fdat$residual_Y), list(), simplify = FALSE), names(fdat$residual_Y))
    for (r in 1:length(fitted)) {
        st = proc.time()
        fitted[[r]] = list()
        message(paste(&quot;Dimension of Y matrix is &quot;, nrow(fdat$residual_Y[[r]]), &quot;rows by&quot;, ncol(fdat$residual_Y[[r]]), &quot;columns.&quot;))
        
        # Run SuSiE on top K PC
        if(${&quot;TRUE&quot; if (susie_top_pc &gt;0 or not skip_twas_weights) else &quot;FALSE&quot;}) {
            top_pc_data &lt;- prcomp(fdat$residual_Y[[r]], center = TRUE, scale. = TRUE)$x
            if (${susie_top_pc} &lt; ncol(top_pc_data)) {
                top_pc_data &lt;- top_pc_data[,1:${susie_top_pc},drop=F]
            }
            fitted[[r]]$susie_on_top_pc &lt;- list()
            for (i in 1:ncol(top_pc_data)) {
                susie_on_top_pc &lt;- susie_wrapper(fdat$residual_X[[r]], top_pc_data[,i], init_L=${init_L}, max_L=${max_L}, refine=TRUE, coverage = ${coverage[0]})
                fitted[[r]]$susie_on_top_pc[[i]] &lt;- susie_post_processor(susie_on_top_pc, fdat$residual_X[[r]], top_pc_data[,i], fdat$residual_X_scalar[[r]], 1, fdat$maf[[r]],
                                           secondary_coverage = c(${&quot;,&quot;.join([str(x) for x in coverage[1:]])}), signal_cutoff = ${pip_cutoff},
                                           other_quantities = list(dropped_samples = list(X=fdat$dropped_sample$dropped_samples_X[[r]], 
                                                                   y=fdat$dropped_sample$dropped_samples_Y[[r]], 
                                                                   covar=fdat$dropped_sample$dropped_samples_covar[[r]])))
            }
            # Run TWAS weights on the first PC
            # Exactly the same codes copied from susie_twas
            if ( ${&quot;TRUE&quot; if not skip_twas_weights else &quot;FALSE&quot;} ) {
                twas_weights_output &lt;- twas_weights_pipeline(fdat$residual_X[[r]], top_pc_data[,1], 
                                         susie_fit=fitted[[r]]$susie_on_top_pc[[1]]$susie_result_trimmed,
                                         cv_folds = ${twas_cv_folds}, max_cv_variants=${max_cv_variants}, 
                                         cv_threads=${twas_cv_threads})
                # clean up the output database
                fitted[[r]] = c(fitted[[r]], twas_weights_output)
                fitted[[r]]$twas_weights = lapply(fitted[[r]]$twas_weights, function(x) { rownames(x) &lt;- NULL; return(x) })
            }
        }
        # Run fSuSiE -- this can take a while
        fitted[[r]]$fsusie_result &lt;- fsusie_wrapper(X = fdat$residual_X[[r]],
                                      Y = fdat$residual_Y[[r]],
                                      pos=fdat$Y_coordinates[[r]]$start,
                                      L=${max_L},
                                      prior=&quot;${prior}&quot;,
                                      max_SNP_EM=${max_SNP_EM}, 
                                      max_scale = ${max_scale},
                                      min.purity = ${min_purity},
                                      cov_lev = ${coverage[0]},
                                      post_processing = &#39;${post_processing}&#39;
                                      ${&quot;, cor_small = TRUE&quot;  if small_sample_correction else &quot;&quot; }
                                      )
        fitted[[r]]$Y_coordinates = fdat$Y_coordinates[[r]] # Add Y coord 
        names(fitted[[r]]$fsusie_result$pip) = colnames(fdat$residual_X[[r]]) 
        fitted[[r]]$fsusie_summary &lt;- susie_post_processor(fitted[[r]]$fsusie_result, fdat$residual_X[[r]], NULL, fdat$residual_X_scalar[[r]], 1, fdat$maf[[r]], 
                                                          secondary_coverage = c(${&quot;,&quot;.join([str(x) for x in coverage[1:]])}), signal_cutoff = ${pip_cutoff},
                                                          other_quantities = list(dropped_samples = list(X=fdat$dropped_sample$dropped_samples_X[[r]], y=fdat$dropped_sample$dropped_samples_Y[[r]], 
                                                                                  covar=fdat$dropped_sample$dropped_samples_covar[[r]])))
        fitted[[r]]$total_time_elapsed = proc.time() - st
        fitted[[r]]$region_info = list(region_coord=parse_region(&quot;${_meta_info[0]}&quot;), grange=parse_region(&quot;${_meta_info[1]}&quot;), region_name=&quot;${_meta_info[2]}&quot;)
        # original data no longer relevant, set to NA to release memory
        fdat$residual_X[[r]] &lt;- NA
        fdat$residual_Y[[r]] &lt;- NA
    }
    saveRDS(list(&quot;${_meta_info[0]}&quot; = fitted), ${_output:ar}, compress=&#39;xz&#39;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="functional-regression-fsusie-with-other-modality">
<h2>Functional regression fSuSiE with other modality<a class="headerlink" href="#functional-regression-fsusie-with-other-modality" title="Link to this heading">#</a></h2>
<p><strong>This is still WIP — mvfSuSiE package is still being developed</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mvfsusie]
# prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: prior  = &quot;mixture_normal_per_scale&quot;
parameter: max_SNP_EM = 1000
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0]}.mvfsusie_{prior}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    # Load regional association data
    fdat = load_regional_association_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1::2]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[2::2]])}),
                                          region = ${&#39;&quot;%s:%s-%s&quot;&#39; % (_meta_info[1], _meta_info[2], _meta_info[3])},
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss})
    # Fine-mapping with mvfSuSiE
    library(&quot;mvf.susie.alpha&quot;)
    Y = map(fdat$residual_Y, ~left_join(fdat$X[,1]%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;), .x%&gt;%t%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;) , by = &quot;rowname&quot;)%&gt;%select(-2)%&gt;%column_to_rownames(&quot;rowname&quot;)%&gt;%as.matrix )
    fitted &lt;- multfsusie(Y_f = list(Y[[1]],Y[[3]]), 
                         Y_u = Reduce(cbind, Y[[2]]),
                         pos = list(pos1 =fdat$phenotype_coordiates[[1]], pos2 = fdat$phenotype_coordiates[[3]]),
                         X=X,
                         L=${max_L},
                         data.format=&quot;list_df&quot;)
    saveRDS(fitted, ${_output:ar})
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/mnm_analysis/mnm_methods"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../summary_stats_finemapping_vignette.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</p>
      </div>
    </a>
    <a class="right-next"
       href="rss_analysis.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">High-dimensional regression with summary statistics</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-association-window-file">Example association-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-susie-with-twas-weights-from-multiple-methods">i. SuSiE with TWAS weights from multiple methods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-fsusie">ii. fSuSiE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-mvsusie-with-twas-weights-from-mr-mash">iii. mvSuSiE with TWAS weights from mr.mash</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-analysis-susie-and-twas">Univariate analysis: SuSiE and TWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-analysis-mvsusie-and-mr-mash">Multivariate analysis: mvSuSiE and mr.mash</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#xqtl-fine-mapping-across-multiple-genes">xQTL fine-mapping across multiple genes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-for-epigenomic-qtl-fine-mapping">Functional regression fSuSiE for epigenomic QTL fine-mapping</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-with-other-modality">Functional regression fSuSiE with other modality</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>