
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Advanced cis-QTL analysis with individual level data &#8212; FunGen-xQTL Consortium</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="SuSiE fine-mapping workflow" href="SuSiE.html" />
    <link rel="prev" title="Overview of advanced cis-QTL analysis" href="cis_advanced_overview.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">FunGen-xQTL Consortium</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   FunGen-xQTL Computational Protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../xqtl_protocol_demo.html">
   Illustration of xQTL protocol
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">
   Univariate xQTL Discovery
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">
     Quantifying alternative splicing from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">
     Normalization and phenotype table generation for splicingQTL analysis
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK File Quality Control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic Relationship Matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype Data Formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype Data Formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">
     Phenotype data imputation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate Data Preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">
     Hidden factor analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/qtl_association_testing.html">
   QTL Association Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL Association Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">
   QTL data postprocessing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced cis-QTL Analysis
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="cis_advanced_overview.html">
   Overview of advanced cis-QTL analysis
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Advanced cis-QTL analysis with individual level data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="SuSiE.html">
   SuSiE fine-mapping workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="fSuSiE.html">
   fSuSiE workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mvSuSiE.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Advanced Genome-wide Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/rss_analysis.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/polyfun.html">
   Fine-mapping with PolyFun
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trans_analysis/MRAID_QTL.html">
   Mendelian Randomization using MRAID
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Enrichment Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/gregor.html">
   Enrichment analysis: TORUS and GREGOR
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../enrichment/ldsc.html">
   Stratified LD Score Regression
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  xQTL-GWAS Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../pecotmr_analysis/pecotmr.html">
   The PECOTMR intergration framework
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/pecotmr_preprocessing.html">
     xQTL and GWAS region alignment
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/ptwas.html">
     TWAS for xQTL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/ptwas_mr.html">
     Mendelian Randomization using fine-mapping IV
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../pecotmr_analysis/intact.html">
     TWAS and COLOC integration
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_preprocessing.html">
     Summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_fit.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mash_posterior.html">
     MASH analysis pipeline with posterior computation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Rare xQTL
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../rare_xqtl/watershed.html">
   watershed analysis for rare xQTL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Post Processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/sumstats_standardizer.html">
   Summary statistics standardization and export
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/association_scan_post_processing.html">
   Association result processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">
   SuSIE results post process
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/code/cis_analysis/cis_workhorse.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/cis_analysis/cis_workhorse.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/cis_analysis/cis_workhorse.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example-genotype-list">
     Example genotype list
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example-phenotype-list">
     Example phenotype list
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#susie-with-twas-weights">
     SuSiE with TWAS weights
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fsusie">
     fSuSiE
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#univariate-analysis-susie-and-twas">
   Univariate analysis: SuSiE and TWAS
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multivariate-susie">
   Multivariate SuSiE
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#univariate-fsusie">
   Univariate fSuSiE
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multivariate-fsusie">
   Multivariate fSuSiE
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="advanced-cis-qtl-analysis-with-individual-level-data">
<h1>Advanced cis-QTL analysis with individual level data<a class="headerlink" href="#advanced-cis-qtl-analysis-with-individual-level-data" title="Permalink to this headline">¶</a></h1>
<p>This notebook performs various advanced statistical analysis on multiple xQTL in a given region. Current procedures implemented include:</p>
<ul class="simple">
<li><p>SuSiE</p></li>
<li><p>fSuSiE</p></li>
<li><p>mvSuSiE</p></li>
<li><p>mr.mash</p></li>
</ul>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>A list of regions to be analyzed (optional); the last column of this file should be region name.</p></li>
<li><p>A list of genotype files per region to be analyzed, in PLINK <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p></li>
<li><p>vector of lists of phenotype files per region to be analyzed, in UCSC <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> with index in <code class="docutils literal notranslate"><span class="pre">bed.gz.tbi</span></code> formats.</p></li>
<li><p>vector of covariate files corresponding to the lists above.</p></li>
<li><p>Optionally a vector of names of the phenotypic conditions in the form of <code class="docutils literal notranslate"><span class="pre">cond1</span> <span class="pre">cond2</span> <span class="pre">cond3</span></code> separated with whitespace.</p></li>
</ol>
<p>Input 2 and 3 should be outputs from <code class="docutils literal notranslate"><span class="pre">genotype_per_region</span></code> and <code class="docutils literal notranslate"><span class="pre">annotate_coord</span></code> modules in previous preprocessing steps. 4 should be output of <code class="docutils literal notranslate"><span class="pre">covariate_preprocessing</span></code> pipeline that contains genotype PC, phenotypic hidden confounders and fixed covariates.</p>
<div class="section" id="example-genotype-list">
<h3>Example genotype list<a class="headerlink" href="#example-genotype-list" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># region        path</span>
<span class="n">ENSG00000000457</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000000457</span><span class="o">.</span><span class="n">bed</span>
<span class="n">ENSG00000000460</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000000460</span><span class="o">.</span><span class="n">bed</span>
<span class="n">ENSG00000000938</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000000938</span><span class="o">.</span><span class="n">bed</span>
<span class="n">ENSG00000000971</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000000971</span><span class="o">.</span><span class="n">bed</span>
<span class="n">ENSG00000001036</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000001036</span><span class="o">.</span><span class="n">bed</span>
<span class="n">ENSG00000001084</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000001084</span><span class="o">.</span><span class="n">bed</span>
<span class="n">ENSG00000001167</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000001167</span><span class="o">.</span><span class="n">bed</span>
<span class="n">ENSG00000001460</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">genotype_per_region</span><span class="o">/</span><span class="n">ENSG00000001460</span><span class="o">.</span><span class="n">bed</span>
</pre></div>
</div>
</div>
<div class="section" id="example-phenotype-list">
<h3>Example phenotype list<a class="headerlink" href="#example-phenotype-list" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end ID  path</span>
<span class="n">chr12</span>   <span class="mi">752578</span>  <span class="mi">752579</span>  <span class="n">ENSG00000060237</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">990508</span>  <span class="mi">990509</span>  <span class="n">ENSG00000082805</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">2794969</span> <span class="mi">2794970</span> <span class="n">ENSG00000004478</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">4649113</span> <span class="mi">4649114</span> <span class="n">ENSG00000139180</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6124769</span> <span class="mi">6124770</span> <span class="n">ENSG00000110799</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6534516</span> <span class="mi">6534517</span> <span class="n">ENSG00000111640</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>For each analysis region, the output is SuSiE model fitted and saved in RDS format.</p>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<div class="section" id="susie-with-twas-weights">
<h3>SuSiE with TWAS weights<a class="headerlink" href="#susie-with-twas-weights" title="Permalink to this headline">¶</a></h3>
<p>Below we duplicate the examples for phenotype and covariates to demonstrate that when there are multiple phenotypes for the same genotype it is possible to use this pipeline to analyze all of them (more than two is accepted as well)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">protocol_example_protein</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22_genotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">genotype_by_region_files</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">A</span> <span class="n">B</span> \
    <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">indel</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">stephenslab_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
<p>It is also possible to only analyze a selected list of regions by name, using either option <code class="docutils literal notranslate"><span class="pre">--region-list</span></code> or option <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> or both. The command below will include 6 regions to analyze:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Define the data</span>
<span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;#chr&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">],</span>
    <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000000</span><span class="p">,</span> <span class="mi">1010000</span><span class="p">,</span> <span class="mi">1020000</span><span class="p">],</span>
    <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000100</span><span class="p">,</span> <span class="mi">1010100</span><span class="p">,</span> <span class="mi">1020100</span><span class="p">],</span>
    <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ENSG00000159082_O43426&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSG00000159131_P22102&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSG00000205726_Q15811&quot;</span><span class="p">]</span>
<span class="p">}</span>

<span class="c1"># Create a DataFrame</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="c1"># Save the DataFrame to a tab-separated file</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;output/selected_regions.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">protocol_example_protein</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22_genotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">genotype_by_region_files</span><span class="o">.</span><span class="n">txt</span>\
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">A</span> <span class="n">B</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">output</span><span class="o">/</span><span class="n">selected_regions</span><span class="o">.</span><span class="n">tsv</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="n">name</span> <span class="n">ENSG00000154654_O15394</span> <span class="n">ENSG00000142192_P05067</span> <span class="n">ENSG00000159082_O43426</span> \
    <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">indel</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">stephenslab_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
<div class="section" id="fsusie">
<h3>fSuSiE<a class="headerlink" href="#fsusie" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fsusie</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">protocol_example_methylation</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22_genotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">genotype_by_region_files</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">phenotype_by_region_files</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">phenotype_by_region_files</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">indel</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">stephenslab_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># A list of file paths for genotype data. </span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># One or multiple lists of file paths for phenotype data.</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># Covariate file path</span>
<span class="kn">parameter:</span> <span class="n">covFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># Optional: if a region list is provide the analysis will be focused on provided region. </span>
<span class="c1"># The LAST column of this list will contain the ID of regions to focus on</span>
<span class="c1"># Otherwise, all regions with both genotype and phenotype files will be analyzed</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Optional: if a region name is provided </span>
<span class="c1"># the analysis would be focused on the union of provides region list and region names</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># It is required to input the name of the analysis</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">200</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;100h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;80G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># Name of phenotypes</span>
<span class="kn">parameter:</span> <span class="n">phenotype_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">phenoFile</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">get_analysis_regions</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="s2">&quot;regional_data&quot;</span><span class="p">]</span>
<span class="c1"># input is genoFile, phenoFile, covFile and optionally region_list. If region_list presents then we only analyze what&#39;s contained in the list.</span>
<span class="c1"># regional_data should be a dictionary with:</span>
<span class="c1"># 1. a list of tuples: {data: [(gene_1.genotype, condition_1, cov_1), (gene_2.genotype, condition_1, cov_1, condition_2, cov_2), ...]} each element may not be of the same length</span>
<span class="c1"># 2. a list of region meta_info: {meta_info: ( &quot;chr:start-end&quot;,gene_1,&quot;cond_1&quot;), (&quot;chr:start-end&quot;,gene_2, &quot;cond_1&#39;,&#39;cond_2&quot;), ...]}</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="n">genoFile</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genoFile</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">covFile</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of input phenotypes files must match that of covariates files&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_names</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of input phenotypes files must match the number of phenotype names&quot;</span><span class="p">)</span>
<span class="c1">## pos and covar are condition specific, this way when there is no phenotype file, there is na in the corresponding column.</span>
<span class="n">phenoFile</span> <span class="o">=</span> <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">pos</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">y</span><span class="p">:</span><span class="n">y</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:&#39;</span><span class="o">+</span><span class="n">y</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">+</span>
                                              <span class="n">y</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">cov_path</span> <span class="o">=</span> <span class="n">z</span><span class="p">,</span> <span class="n">cond</span> <span class="o">=</span> <span class="n">a</span> <span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span><span class="s2">&quot;#id&quot;</span><span class="p">})</span>   
             <span class="k">for</span> <span class="n">x</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span><span class="n">covFile</span><span class="p">,</span><span class="n">phenotype_names</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)):</span>
    <span class="n">genoFile</span> <span class="o">=</span> <span class="n">genoFile</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;#id&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">suffixes</span> <span class="o">=</span> <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">_x&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">_y&#39;</span><span class="p">))</span>

<span class="c1"># remove id that has no phenotype.</span>

<span class="n">genoFile</span> <span class="o">=</span> <span class="n">genoFile</span><span class="p">[</span><span class="o">~</span><span class="n">genoFile</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#id&#39;</span><span class="p">,</span><span class="s2">&quot;#path&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>    

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genoFile</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No region overlap between genotype #id and any of the phenotypes ID&quot;</span><span class="p">)</span>

<span class="c1"># Get position for meta_data</span>
<span class="n">pos_col</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">genoFile</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;pos&#39;</span><span class="p">)]</span>
<span class="n">genoFile</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">genoFile</span><span class="p">[</span><span class="n">pos_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="c1"># Get the conditions strings for each ID</span>
<span class="n">cond_col</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">genoFile</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;cond&#39;</span><span class="p">)]</span>
<span class="n">genoFile</span><span class="p">[</span><span class="s2">&quot;phenotype_names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&#39;,&#39;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">((</span><span class="n">x</span><span class="p">))</span><span class="o">.</span><span class="n">dropna</span><span class="p">())</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">genoFile</span><span class="p">[</span><span class="n">cond_col</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;split&quot;</span><span class="p">)[</span><span class="s2">&quot;data&quot;</span><span class="p">]]</span>
<span class="c1"># Clean up</span>
<span class="n">genoFile</span> <span class="o">=</span> <span class="n">genoFile</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">cond_col</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">pos_col</span><span class="p">)</span>


<span class="n">region_ids</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># If region_list is provided, read the file and extract IDs</span>
<span class="k">if</span> <span class="n">region_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
    <span class="n">region_list_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">region_list</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span>
    <span class="n">region_ids</span> <span class="o">=</span> <span class="n">region_list_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>  <span class="c1"># Extracting the last column for IDs</span>

<span class="c1"># If region_name is provided, include those IDs as well</span>
<span class="c1"># --region-name A B C will result in a list of [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;] here</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_name</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">region_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_name</span><span class="p">)))</span>

<span class="c1"># If either region_list or region_name is provided, filter the genoFile</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">genoFile</span> <span class="o">=</span> <span class="n">genoFile</span><span class="p">[</span><span class="n">genoFile</span><span class="p">[</span><span class="s1">&#39;#id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)]</span>

<span class="n">file_inv</span> <span class="o">=</span> <span class="n">genoFile</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;#id&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype_names&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;split&quot;</span><span class="p">)</span>
<span class="n">file_inv</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[[</span><span class="n">value</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">sublist</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">value</span><span class="p">)]</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">file_inv</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">]]</span> 


<span class="c1">## There will alwayse be genotype file due to left join,</span>
<span class="c1">## There will alwayse be covar file as len(covFile) must == len(PhenoFile), and covar column is the same string accross all rows</span>
<span class="c1">## So only if there is no bed.gz there will be problem.</span>
<span class="n">regional_data</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span><span class="n">file_inv</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">],</span><span class="s2">&quot;meta_info&quot;</span><span class="p">:</span> <span class="n">genoFile</span><span class="p">[[</span><span class="s2">&quot;#id&quot;</span><span class="p">,</span><span class="s2">&quot;phenotype_names&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;split&quot;</span><span class="p">)[</span><span class="s1">&#39;data&#39;</span><span class="p">]</span> <span class="p">}</span>

<span class="c1"># Recreate file_inv based on the filtered genoFile</span>
<span class="n">file_inv</span> <span class="o">=</span> <span class="n">genoFile</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#id&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype_names&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;split&quot;</span><span class="p">)</span>
<span class="n">file_inv</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[[</span><span class="n">value</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">sublist</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">value</span><span class="p">)]</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">file_inv</span><span class="p">[</span><span class="s1">&#39;data&#39;</span><span class="p">]]</span> 

<span class="c1"># Recreate the regional_data based on the filtered data</span>
<span class="n">regional_data</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">file_inv</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">],</span>
                 <span class="s2">&quot;meta_info&quot;</span><span class="p">:</span> <span class="n">genoFile</span><span class="p">[[</span><span class="s2">&quot;#id&quot;</span><span class="p">,</span> <span class="s2">&quot;phenotype_names&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;split&quot;</span><span class="p">)[</span><span class="s1">&#39;data&#39;</span><span class="p">]}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="univariate-analysis-susie-and-twas">
<h2>Univariate analysis: SuSiE and TWAS<a class="headerlink" href="#univariate-analysis-susie-and-twas" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">susie_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">init_L</span> <span class="o">=</span> <span class="mi">12</span>
<span class="kn">parameter:</span> <span class="n">max_L</span> <span class="o">=</span> <span class="mi">30</span>
<span class="c1"># remove a variant if it has more than imiss missing individual level data</span>
<span class="kn">parameter:</span> <span class="n">imiss</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="c1"># MAF cutoff</span>
<span class="kn">parameter:</span> <span class="n">maf</span> <span class="o">=</span> <span class="mf">0.005</span>
<span class="c1"># MAC cutoff, on top of MAF cutoff</span>
<span class="kn">parameter:</span> <span class="n">mac</span> <span class="o">=</span> <span class="mi">5</span>
<span class="c1"># Remove indels if indel = False</span>
<span class="kn">parameter:</span> <span class="n">indel</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">parameter:</span> <span class="n">pip_cutoff</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">coverage</span> <span class="o">=</span> <span class="mf">0.95</span>
<span class="kn">parameter:</span> <span class="n">secondary_coverage</span> <span class="o">=</span> <span class="mf">0.7</span>
<span class="kn">parameter:</span> <span class="n">twas_weights</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s2">&quot;regional_data&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="n">vector</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">lst</span><span class="p">[</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="p">]):</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">])]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vector</span><span class="p">))]</span>

<span class="n">meta_info</span> <span class="o">=</span> <span class="n">regional_data</span><span class="p">[</span><span class="s1">&#39;meta_info&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">group_by_region</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">]),</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;meta_info&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">step_name</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">.susie_fitted.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="n">library</span><span class="p">(</span><span class="n">pecotmr</span><span class="p">)</span>
    <span class="c1"># Load regional association data</span>
    <span class="n">fdat</span> <span class="o">=</span> <span class="n">load_regional_finemapping_data</span><span class="p">(</span><span class="n">genotype</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">anr</span><span class="p">},</span>
                                          <span class="n">phenotype</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">covariate</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">region</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{_meta_info[0]}</span><span class="s2">&quot;</span><span class="p">,</span>
                                          <span class="n">conditions</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;$</span><span class="si">{_meta_info[2]}</span><span class="s1">&#39;</span><span class="p">),</span>
                                          <span class="n">maf_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">maf</span><span class="p">},</span>
                                          <span class="n">mac_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">mac</span><span class="p">},</span>
                                          <span class="n">imiss_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">imiss</span><span class="p">},</span>
                                          <span class="n">y_as_matrix</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
                                          <span class="n">keep_indel</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">indel</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span>
   
    <span class="c1"># Univeriate analysis suite</span>
    <span class="n">fitted</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">r</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">))</span> <span class="p">{</span>
      <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span> <span class="o">=</span> <span class="n">susie_wrapper</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_X_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="err">$</span><span class="p">{</span><span class="n">L_init</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">max_L</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">coverage</span><span class="p">})</span>
      <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span> <span class="o">=</span> <span class="n">susie_twas_wrapper</span><span class="p">(</span><span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">fdat</span><span class="err">$</span><span class="n">residual_X_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">fdat</span><span class="err">$</span><span class="n">residual_X_sd</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_sd</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> 
                                       <span class="n">fdat</span><span class="err">$</span><span class="n">maf</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">secondary_coverage</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">secondary_coverage</span><span class="p">},</span> <span class="n">signal_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">pip_cutoff</span><span class="p">},</span> 
                                       <span class="n">twas_weights</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">twas_weights</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">},</span>   
                                       <span class="n">other_quantities</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">dropped_samples</span> <span class="o">=</span> <span class="n">fdat</span><span class="err">$</span><span class="n">dropped_sample</span><span class="p">[[</span><span class="n">r</span><span class="p">]]))</span>
    <span class="p">}</span>
    <span class="n">names</span><span class="p">(</span><span class="n">fitted</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">fitted</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">ar</span><span class="p">},</span> <span class="n">compress</span><span class="o">=</span><span class="s1">&#39;xz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="multivariate-susie">
<h2>Multivariate SuSiE<a class="headerlink" href="#multivariate-susie" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">mvsusie_1</span><span class="p">]</span>
<span class="c1"># Prior model file generated from mashr. </span>
<span class="c1"># Default will be used if it does not exist.</span>
<span class="kn">parameter:</span> <span class="n">mixture_prior</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="kn">parameter:</span> <span class="n">max_L</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># remove a variant if it has more than imiss missing individual level data</span>
<span class="kn">parameter:</span> <span class="n">imiss</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="c1"># MAF cutoff</span>
<span class="kn">parameter:</span> <span class="n">maf</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># MAC cutoff, on top of MAF cutoff</span>
<span class="c1"># Here I set default to mac = 10 rather than using an MAF cutoff</span>
<span class="c1"># I don&#39;t set it to 5 because I&#39;m not so sure of performance of SuSiE on somewhat infrequent variants</span>
<span class="c1"># MAC = 10 would not be too infrequenty for xQTL data where sample size is about ~1,000 at most (as of 2022)</span>
<span class="kn">parameter:</span> <span class="n">mac</span> <span class="o">=</span> <span class="mi">10</span>
<span class="c1"># Remove indels if indel = False</span>
<span class="kn">parameter:</span> <span class="n">indel</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s2">&quot;regional_data&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="n">vector</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">lst</span><span class="p">[</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="p">]):</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">])]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vector</span><span class="p">))]</span>

<span class="n">meta_info</span> <span class="o">=</span> <span class="n">regional_data</span><span class="p">[</span><span class="s1">&#39;meta_info&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">group_by_region</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">]),</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;meta_info&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">step_name</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.susie_fitted.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
   
    <span class="n">get_prior_indices</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="n">U</span><span class="p">)</span> <span class="p">{</span>
      <span class="c1"># make sure the prior col/rows match the colnames of the Y matrix</span>
      <span class="n">y_names</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
      <span class="n">u_names</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">U</span><span class="p">)</span>
      <span class="k">if</span> <span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">y_names</span><span class="p">)</span> <span class="o">||</span> <span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">u_names</span><span class="p">))</span> <span class="p">{</span>
          <span class="k">return</span><span class="p">(</span><span class="n">NULL</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="k">if</span> <span class="p">(</span><span class="n">identical</span><span class="p">(</span><span class="n">y_names</span><span class="p">,</span> <span class="n">u_names</span><span class="p">))</span> <span class="p">{</span>
          <span class="k">return</span><span class="p">(</span><span class="n">NULL</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="k">return</span><span class="p">(</span><span class="n">match</span><span class="p">(</span><span class="n">y_names</span><span class="p">,</span> <span class="n">u_names</span><span class="p">))</span>
      <span class="p">}</span>
    <span class="p">}</span>

    <span class="c1"># Load regional association data</span>
    <span class="n">fdat</span> <span class="o">=</span> <span class="n">load_regional_finemapping_data</span><span class="p">(</span><span class="n">genotype</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">anr</span><span class="p">},</span>
                                          <span class="n">phenotype</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">covariate</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">region</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">:</span><span class="si">%s</span><span class="s1">-</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">_meta_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">_meta_info</span><span class="p">[</span><span class="mi">3</span><span class="p">])},</span>
                                          <span class="n">maf_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">maf</span><span class="p">},</span>
                                          <span class="n">mac_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">mac</span><span class="p">},</span>
                                          <span class="n">imiss_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">imiss</span><span class="p">},</span>
                                          <span class="n">y_as_matrix</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                                          <span class="n">keep_indel</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">indel</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span>

    <span class="c1"># univariate summary statistics</span>
    <span class="n">non_missing</span> <span class="o">=</span> <span class="n">lapply</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">),</span> <span class="n">function</span><span class="p">(</span><span class="n">r</span><span class="p">))</span> <span class="n">which</span><span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">[,</span><span class="n">r</span><span class="p">]))</span>
    <span class="n">univariate_res</span> <span class="o">=</span> <span class="n">lapply</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">),</span> <span class="n">function</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="n">susieR</span><span class="p">:::</span><span class="n">univariate_regression</span><span class="p">(</span><span class="n">X</span><span class="p">[</span><span class="n">non_missing</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="p">],</span> <span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">[</span><span class="n">non_missing</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">r</span><span class="p">]))</span>
    <span class="n">sumstat</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">bhat</span><span class="o">=</span><span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">cbind</span><span class="p">,</span> <span class="n">lapply</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">),</span> <span class="n">function</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="n">univariate_res</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">betahat</span><span class="p">)),</span>
                   <span class="n">sbhat</span><span class="o">=</span><span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">cbind</span><span class="p">,</span> <span class="n">lapply</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">),</span> <span class="n">function</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="n">univariate_res</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">sebetahat</span><span class="p">)))</span>
  
    <span class="c1"># Multivariate fine-mapping</span>
    <span class="c1"># FIXME: handle it when prior does not exist</span>
    <span class="n">prior</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">mixture_prior</span><span class="p">:</span><span class="n">r</span><span class="p">})</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Number of components in the mixture prior:&quot;</span><span class="p">,</span> <span class="n">length</span><span class="p">(</span><span class="n">prior</span><span class="err">$</span><span class="n">U</span><span class="p">)))</span>
    <span class="n">prior</span> <span class="o">=</span> <span class="n">mvsusieR</span><span class="p">::</span><span class="n">create_mash_prior</span><span class="p">(</span><span class="n">mixture_prior</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">weights</span><span class="o">=</span><span class="n">prior</span><span class="err">$</span><span class="n">w</span><span class="p">,</span> <span class="n">matrices</span><span class="o">=</span><span class="n">prior</span><span class="err">$</span><span class="n">U</span><span class="p">),</span> <span class="n">include_indices</span> <span class="o">=</span> <span class="n">get_prior_indices</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">,</span> <span class="n">prior</span><span class="err">$</span><span class="n">U</span><span class="p">[[</span><span class="mi">1</span><span class="p">]]),</span> <span class="n">max_mixture_len</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>   
    <span class="n">resid_Y</span> <span class="o">=</span> <span class="n">compute_cov_flash</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">)</span>
    <span class="n">st</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">fitted</span> <span class="o">=</span> <span class="n">mvsusieR</span><span class="p">::</span><span class="n">mvsusie</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">X</span><span class="p">,</span> 
                               <span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">,</span> 
                               <span class="n">L</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">max_L</span><span class="p">},</span> 
                               <span class="n">prior_variance</span><span class="o">=</span><span class="n">prior</span><span class="p">,</span> 
                               <span class="n">residual_variance</span><span class="o">=</span><span class="n">resid_Y</span><span class="p">,</span> 
                               <span class="n">precompute_covariances</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> 
                               <span class="n">compute_objective</span><span class="o">=</span><span class="n">T</span><span class="p">,</span> 
                               <span class="n">estimate_residual_variance</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> 
                               <span class="n">estimate_prior_variance</span><span class="o">=</span><span class="n">T</span><span class="p">,</span> 
                               <span class="n">estimate_prior_method</span><span class="o">=</span><span class="s1">&#39;EM&#39;</span><span class="p">,</span>
                               <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> 
                               <span class="n">n_thread</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">},</span> 
                               <span class="n">approximate</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">analysis_time</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">st</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">cs_corr</span> <span class="o">=</span> <span class="n">susieR</span><span class="p">::</span><span class="n">get_cs_correlation</span><span class="p">(</span><span class="n">fitted</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">fdat</span><span class="err">$</span><span class="n">X</span><span class="p">)</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">cs_snps</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">fitted</span><span class="err">$</span><span class="n">X_column_scale_factors</span><span class="p">[</span><span class="n">unlist</span><span class="p">(</span><span class="n">fitted</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">)])</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">variable_name</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">fitted</span><span class="err">$</span><span class="n">pip</span><span class="p">)</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">analysis_script</span> <span class="o">=</span> <span class="n">load_script</span><span class="p">()</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">dropped_samples</span> <span class="o">=</span> <span class="n">fdat</span><span class="err">$</span><span class="n">dropped_sample</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">sample_names</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">)</span>
    <span class="n">fitted</span><span class="err">$</span><span class="n">residual_y</span> <span class="o">=</span> <span class="n">resid_Y</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">fitted</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">ar</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="univariate-fsusie">
<h2>Univariate fSuSiE<a class="headerlink" href="#univariate-fsusie" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">fsusie_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">max_L</span> <span class="o">=</span> <span class="mi">30</span>
<span class="c1"># remove a variant if it has more than imiss missing individual level data</span>
<span class="kn">parameter:</span> <span class="n">imiss</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="c1"># MAF cutoff</span>
<span class="kn">parameter:</span> <span class="n">maf</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># MAC cutoff, on top of MAF cutoff</span>
<span class="c1"># Here I set default to mac = 10 rather than using an MAF cutoff</span>
<span class="c1"># I don&#39;t set it to 5 because I&#39;m not so sure of performance of SuSiE on somewhat infrequent variants</span>
<span class="c1"># MAC = 10 would not be too infrequenty for xQTL data where sample size is about ~1,000 at most (as of 2022)</span>
<span class="kn">parameter:</span> <span class="n">mac</span> <span class="o">=</span> <span class="mi">10</span>
<span class="c1"># prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]</span>
<span class="kn">parameter:</span> <span class="n">prior</span>  <span class="o">=</span> <span class="s2">&quot;mixture_normal_per_scale&quot;</span>
<span class="kn">parameter:</span> <span class="n">max_SNP_EM</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="c1"># Remove indels if indel = False</span>
<span class="kn">parameter:</span> <span class="n">indel</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s2">&quot;regional_data&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="n">vector</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">lst</span><span class="p">[</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="p">]):</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">])]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vector</span><span class="p">))]</span>

<span class="n">meta_info</span> <span class="o">=</span> <span class="n">regional_data</span><span class="p">[</span><span class="s1">&#39;meta_info&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">group_by_region</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">]),</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;meta_info&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">step_name</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.fsusie_</span><span class="si">{</span><span class="n">prior</span><span class="si">}</span><span class="s1">.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="c1"># Load regional association data</span>
    <span class="n">fdat</span> <span class="o">=</span> <span class="n">load_regional_association_data</span><span class="p">(</span><span class="n">genotype</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">anr</span><span class="p">},</span>
                                          <span class="n">phenotype</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">covariate</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">region</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">:</span><span class="si">%s</span><span class="s1">-</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">_meta_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">_meta_info</span><span class="p">[</span><span class="mi">3</span><span class="p">])},</span>
                                          <span class="n">maf_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">maf</span><span class="p">},</span>
                                          <span class="n">mac_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">mac</span><span class="p">},</span>
                                          <span class="n">imiss_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">imiss</span><span class="p">},</span>
                                          <span class="n">y_as_matrix</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
                                          <span class="n">keep_indel</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s2">&quot;TRUE&quot;</span> <span class="k">if</span> <span class="n">indel</span> <span class="k">else</span> <span class="s2">&quot;FALSE&quot;</span><span class="p">})</span>
    <span class="c1"># Fine-mapping with fSuSiE</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;susiF.alpha&quot;</span><span class="p">)</span>
    <span class="n">fitted</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">r</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">st</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>  
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">susiF</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_X_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span>
                             <span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span>
                             <span class="n">pos</span><span class="o">=</span><span class="n">fdat</span><span class="err">$</span><span class="n">phenotype_coordiates</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="c1">#FIXME: this needs to be edited and added dto load_regional_association_data</span>
                             <span class="n">L</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">max_L</span><span class="p">},</span>
                             <span class="n">prior</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{prior}</span><span class="s2">&quot;</span><span class="p">,</span>
                             <span class="n">max_SNP_EM</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">max_SNP_EM</span><span class="p">})</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">analysis_time</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">st</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">cs_corr</span> <span class="o">=</span> <span class="n">get_cs_correlation</span><span class="p">(</span><span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]],</span> <span class="n">X</span><span class="o">=</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_X_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]])</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">cs_snps</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">X_column_scale_factors</span><span class="p">[</span><span class="n">unlist</span><span class="p">(</span><span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">sets</span><span class="err">$</span><span class="n">cs</span><span class="p">)])</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">variable_name</span> <span class="o">=</span> <span class="n">names</span><span class="p">(</span><span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">pip</span><span class="p">)</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">coef</span> <span class="o">=</span> <span class="n">coef</span><span class="o">.</span><span class="n">susie</span><span class="p">(</span><span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]])</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">analysis_script</span> <span class="o">=</span> <span class="n">load_script</span><span class="p">()</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">analysis_name</span> <span class="o">=</span> <span class="n">fdat</span><span class="err">$</span><span class="n">traits</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">dropped_samples</span> <span class="o">=</span> <span class="n">fdat</span><span class="err">$</span><span class="n">dropped_sample</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span>
        <span class="n">fitted</span><span class="p">[[</span><span class="n">r</span><span class="p">]]</span><span class="err">$</span><span class="n">sample_names</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">[[</span><span class="n">r</span><span class="p">]])</span>
    <span class="p">}</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">fitted</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">ar</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="multivariate-fsusie">
<h2>Multivariate fSuSiE<a class="headerlink" href="#multivariate-fsusie" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">mvfsusie_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">max_L</span> <span class="o">=</span> <span class="mi">30</span>
<span class="c1"># remove a variant if it has more than imiss missing individual level data</span>
<span class="kn">parameter:</span> <span class="n">imiss</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="c1"># MAF cutoff</span>
<span class="kn">parameter:</span> <span class="n">maf</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># MAC cutoff, on top of MAF cutoff</span>
<span class="c1"># Here I set default to mac = 10 rather than using an MAF cutoff</span>
<span class="c1"># I don&#39;t set it to 5 because I&#39;m not so sure of performance of SuSiE on somewhat infrequent variants</span>
<span class="c1"># MAC = 10 would not be too infrequenty for xQTL data where sample size is about ~1,000 at most (as of 2022)</span>
<span class="kn">parameter:</span> <span class="n">mac</span> <span class="o">=</span> <span class="mi">10</span>
<span class="c1"># prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]</span>
<span class="kn">parameter:</span> <span class="n">prior</span>  <span class="o">=</span> <span class="s2">&quot;mixture_normal_per_scale&quot;</span>
<span class="kn">parameter:</span> <span class="n">max_SNP_EM</span> <span class="o">=</span> <span class="mi">1000</span>

<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s2">&quot;regional_data&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="n">vector</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">lst</span><span class="p">[</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="p">]):</span><span class="nb">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">[:</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">])]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vector</span><span class="p">))]</span>

<span class="n">meta_info</span> <span class="o">=</span> <span class="n">regional_data</span><span class="p">[</span><span class="s1">&#39;meta_info&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">group_by_region</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regional_data</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">]),</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;meta_info&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">step_name</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.mvfsusie_</span><span class="si">{</span><span class="n">prior</span><span class="si">}</span><span class="s1">.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stdout&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s2">n</span><span class="si">}</span><span class="s2">.stderr&quot;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="c1"># Load regional association data</span>
    <span class="n">fdat</span> <span class="o">=</span> <span class="n">load_regional_association_data</span><span class="p">(</span><span class="n">genotype</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">anr</span><span class="p">},</span>
                                          <span class="n">phenotype</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">covariate</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">2</span><span class="p">]])}),</span>
                                          <span class="n">region</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;&quot;</span><span class="si">%s</span><span class="s1">:</span><span class="si">%s</span><span class="s1">-</span><span class="si">%s</span><span class="s1">&quot;&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">_meta_info</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">_meta_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">_meta_info</span><span class="p">[</span><span class="mi">3</span><span class="p">])},</span>
                                          <span class="n">maf_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">maf</span><span class="p">},</span>
                                          <span class="n">mac_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">mac</span><span class="p">},</span>
                                          <span class="n">imiss_cutoff</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">imiss</span><span class="p">},</span>
                                          <span class="n">y_as_matrix</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    <span class="c1"># Fine-mapping with mvfSuSiE</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;mvf.susie.alpha&quot;</span><span class="p">)</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">residual_Y_scaled</span><span class="p">,</span> <span class="o">~</span><span class="n">left_join</span><span class="p">(</span><span class="n">fdat</span><span class="err">$</span><span class="n">X</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="o">%&gt;%</span><span class="n">rownames_to_column</span><span class="p">(</span><span class="s2">&quot;rowname&quot;</span><span class="p">),</span> <span class="o">.</span><span class="n">x</span><span class="o">%&gt;%</span><span class="n">t</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="o">%&gt;%</span><span class="n">rownames_to_column</span><span class="p">(</span><span class="s2">&quot;rowname&quot;</span><span class="p">)</span> <span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;rowname&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">column_to_rownames</span><span class="p">(</span><span class="s2">&quot;rowname&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span> <span class="p">)</span>
    <span class="n">fitted</span> <span class="o">&lt;-</span> <span class="n">multfsusie</span><span class="p">(</span><span class="n">Y_f</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">Y</span><span class="p">[[</span><span class="mi">1</span><span class="p">]],</span><span class="n">Y</span><span class="p">[[</span><span class="mi">3</span><span class="p">]]),</span> 
                         <span class="n">Y_u</span> <span class="o">=</span> <span class="n">Reduce</span><span class="p">(</span><span class="n">cbind</span><span class="p">,</span> <span class="n">Y</span><span class="p">[[</span><span class="mi">2</span><span class="p">]]),</span>
                         <span class="n">pos</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pos1</span> <span class="o">=</span><span class="n">fdat</span><span class="err">$</span><span class="n">phenotype_coordiates</span><span class="p">[[</span><span class="mi">1</span><span class="p">]],</span> <span class="n">pos2</span> <span class="o">=</span> <span class="n">fdat</span><span class="err">$</span><span class="n">phenotype_coordiates</span><span class="p">[[</span><span class="mi">3</span><span class="p">]]),</span>
                         <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">,</span>
                         <span class="n">L</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">max_L</span><span class="p">},</span>
                         <span class="n">data</span><span class="o">.</span><span class="n">format</span><span class="o">=</span><span class="s2">&quot;list_df&quot;</span><span class="p">)</span>
    <span class="n">saveRDS</span><span class="p">(</span><span class="n">fitted</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">ar</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="o">*</span><span class="n">_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.cis_workhorse_output.txt&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">entrypoint</span> <span class="o">=</span> <span class="n">entrypoint</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;output&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">ar</span><span class="p">,]]})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output]&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span> <span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/cis_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="cis_advanced_overview.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Overview of advanced cis-QTL analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="SuSiE.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">SuSiE fine-mapping workflow</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium<br/>
        
            &copy; Copyright 2021+, FunGen xQTL Analysis Working Group.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>